<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: MANUAL: apt-copynumber-cyto (1.18.2)</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>MANUAL: apt-copynumber-cyto (1.18.2) </h1>  </div>
</div>
<div class="contents">
<h2><a class="anchor" id="Contents"></a>
Contents</h2>
<ul>
<li>
<a href="#intro">Introduction.</a> </li>
<li>
<a href="#quickstart">Quick Start</a> </li>
<li>
<a href="#runtimeperformance">Runtime Performance</a> </li>
<li>
<a href="#faq">FAQ - Frequently Asked Questions.</a> </li>
</ul>
<p><a class="anchor" id="intro"></a> </p>
<h2><a class="anchor" id="intro"></a>
Introduction</h2>
<p>apt-copynumber-cyto is a program for finding de novo copy number changes and Loss of Heterozygosity (LOH) on a per sample basis with respect to a reference set of samples. The copy number algorithm it implements assumes that the reference set comprises a mix of normal human males (with XY chromosomes) and normal human females (with XX chromosomes). The algorithms assume that in this reference for each autosomal marker (SNP or Copy Number probe) the predominant Copy Number is 2, and for the sex chromosomes the copy number is determined by the gender.</p>
<p>apt-copynumber-cyto implements two distinct workflows, a reference workflow that uses as a a set of CEL files that are input to outputs one or more reference files including copynumber and SNP information; and conceptually, a single-sample workflow that compares each CEL file to a pre-computed reference for analysis. For efficiency of computation the "single-sample workflow" operates on a set of input CEL files at a time, but the output for any CEL file is unaffected by any of the other CEL files.</p>
<p><a class="anchor" id="quickstart"></a> </p>
<h2><a class="anchor" id="QuickStart"></a>
Quick Start</h2>
<p>For CytoScanHD the following library files are required to run apt-copynumber-cyto:</p>
<div class="fragment"><pre class="fragment">
CytoScanHD_Array.na32.1.annot.db:  NetAffx Annotation database file. It is a SQLite 3.x database.

CytoScanHD_Array.chrXprobes:       An ASCII file that contains probe_id (1-based) of 
                                   probes on chrX. Used for copy number probe chrX/Y ratio gender calling.

CytoScanHD_Array.chrYprobes:       An ASCII file that contains probe_id (1-based) of 
                                   probes on chrY as column 1 and probeset_id as column 2. Probe_id is used for 
                                   copy number probe chrX/Y ratio gender calling and probeset_id is used for
                                   hasY gender call and Y-Gender call. 

CytoScanHD_Array.r1.qca:           An ASCII file that defines the parameters for the calculations of qc-call-rate 
                                   and contrast-qc-rand that are used in the geno-qc analysis.
CytoScanHD_Array.r1.qcc:           An ASCII file that defines the probesets that are used by the geno-qc analysis

CytoScanHD_Array.cdf:              Chip definition file (binary file).

CytoScanHD_Array.probe_tab:        An ASCII file that contains probe sequence information. (Reference creation only).

CytoScanHD_Array.hg19.v3.refcovar: An hdf5 format file that contains covariate information for each probeset, where
                                   the first column is the probeset name and subsequent column(s) are covariates in float.
                                   (Reference creation only).

CytoScanHD_Array.snplist.txt:      An ASCII file that contains SNP probeset_id that are used to compute processed SNPQC and Raw SNPQC.
                                   The first row of the file must be "probeset_id". (Single sample mode processing only).

CytoScanHD_Array.na32.1.v3.REF_MODEL: Affymetrix copy number reference generated from a set of hapmaps and normal blood samples (hdf5 format file).
</pre></div><p>See <a href="VIGNETTE-27M-Cyto.html">2.7M cytogenetics vignette</a> for libraries and requirements for 2.7M analysis.</p>
<p>See <a href="VIGNETTE-750K-Cyto.html">CytoScan750K vignette</a> for libraries and requirements for analysis of CytoScan750K arrays.</p>
<p>apt-copynumber-cyto can be applied to multiple chip types, including Cyto 2.7M, CytoScan750K and CytoScanHD. Since different chip types require different analysis option and steps, the default parameter values may not always be appropriate to a chip type or the particular annotation files for that chip type. We strongly recommend all parameters be explicitly specified.</p>
<p>On unix systems, a command to build a reference for CytoScanHD data would look like this:</p>
<p>NOTE: This command needs to have the comments removed to be runnable, as they break the unix shell syntax. The comments and additional spaces are to clarify the options being used. The arguments which are joined with a "."s should be joined without spaces.</p>
<div class="fragment"><pre class="fragment">
apt-copynumber-cyto -v 4 \                #verbose output level in the log file ranging from 1 (least details) to 4 (most details)
	--cyto2 false \                           #true: Cytogenetics_Array; false: CytoScanHD_Array
	--force false \                           #true: Disable various checks including chip types; false: otherwise.
	--keep-intermediate-data false \          #true: dump out intermediate data during apt processing; false: otherwise.
	--doDualNormalization true \              #true: quantile normalize CN and SNP separately; false: do not quantile normalize CN and SNP separately
	--adapter-type-normalization false \      #true: perform adapter type normalization; false: otherwise. Should be always set to "false" for CytoScanHD_Array as the adjustment has been incorporated in "signal-adjustment-covariates"
	--probe-file CytoScanHD_Array.probe_tab \ #specify the probe sequence file
	--run-geno-qc true \                      #true: Run the GenoQC engine; false: otherwise
	--qca-file CytoScanHD_Array.r1.qca \      #if run-geno-qc is set to be true, it is required to provide this file
	--qcc-file CytoScanHD_Array.r1.qcc \      #if run-geno-qc is set to be true, it is required to provide this file
	--text-output false \                     #true: Output data in ASCII text format in addition to calvin format; false: otherwise
	--cnchp-output false \                    #true: report CNCHP file(s); false: otherwise. Either cnchp-output or cychp-output has to be true
	--cychp-output true \                     #true: report CYCHP file(s); false: otherwise. Either cnchp-output or cychp-output has to be true
	--reference-output CnNewRef.REF_MODEL \   #specify name of the new copy number reference file with file extension "REF_MODEL"
	--cdf-file CytoScanHD_Array.cdf \         #specify the chp definition file
	--chrX-probes CytoScanHD_Array.chrXprobes \ #specify the chrX probe file
	--chrY-probes CytoScanHD_Array.chrYprobes \ #specify the chrY probe file
	--annotation-file CytoScanHD_Array.na32.1.annot.db \ #specify the NetAffx database file
	--male-gender-ratio-cutoff 1.5 \          #male gender call threshold for chrX/Y ratio gender calling
	--female-gender-ratio-cutoff 0.9 \        #female gender call threshold for chrX/Y ratio gender calling
	--xx-cutoff 0.61 \                        #lower limit of the gender call threshold for hasXX gender call
	--xx-cutoff-high 0.95 \                   #upper limit of the gender call threshold for hasXX gender call
	--y-cutoff 0.58 \                         #gender call threshold for hasY gender call
	--out-dir output_dir \                    #specify the output directory
	--covariates-file  CytoScanHD_Array.hg19.v3.refcovar \      #specify the covariate file that contains probeset-based covariate(s) that doesn't exist in the annotation db file
	--signal-adjustment-covariates covariate-signal-adjuster.\  #covariate-signal adjuster
	        order=fragment-adapter-type,fragment-length.\       #the covariates to be used in the order in which they are to be applied
	        bin-type=discrete,equally-populated.\               #ordered list corresponding to each covariate controls whether the covariate is treated directly as discrete variable, or as a continuous variable (equally-populated or equally-spaced)
	        bin-count=0,100 \                                   #ordered list corresponding to each covariate controls how a continuous variable is discretized into bins. Must be 0 for discrete covariates
	--lr-adjustment-covariates covariate-lr-adjuster.\          #covariate log2 ratios adjuster
	        order=SuperGC,median-signal,marker-class.\          #the covariates to be used in the order in which they are to be applied
	        bin-type=discrete,equally-populated,discrete.\      #ordered list correspoinding to each covariate controls whether the covariate is treated directly as discrete variable, or as a continuous variable (equally-populated or equally-spaced)
	        bin-count=0,100,0.\                                 #ordered list corresponding to each covariate controls how a continuous variable is discretized into bins. Must be 0 for discrete covariates  
	        iqr-scaling=on,on,on.\                              #ordered list corresponding to each covariate on scales IQR for each bin to be equal 
	        subtract-from-XY=on,off,on \                        #ordered list corresponding to each covariate. “on” applies covariate adjustment to ChrX,ChrY markers. 
	--wave-correction-reference-method wave-correction-reference-method.\ #estimate waves using the reference samples 
	        trim=2.0.\                           #trim parameter for absolute adjusted log2 ratios after each wave
	        percentile=0.75.\                    #percentile to use for each probeset in finding wave. E.g., 0.75 means 75-th percentile.
	        wave-count=6.\                       #number of waves to calculate
	        demean=false \                       #true: demean prior to finding each wave; false: otherwise
	--allele-peaks-adjustment-covariates covariate-allele-peaks-adjuster.\ #covariate allelic difference adjuster
	        order=SuperGC.\                      #the covariates to be used in the order in which they are to be applied
	        bin-type=discrete.\                  #ordered list corresponding to each covariate controls whether the covariate is treated directly as discrete variable, or as a continuous variable (equally-populated or equally-spaced)
	        bin-count=0.\                        #ordered list corresponding to each covariate controls how a continuous variable is discretized into bins. Must be 0 for discrete covariates
	        coarse-allele-peak-adjustment=on.\   #if it is 'on' then turn on coarse allelic peak adjustment, and 'off' otherwise
	        coarse-allele-peak-adjustment-step=50.\             #step size for another density estimation across the genome
	        coarse-allele-peak-adjustment-window=200.\          #number of SNPs used by the kernel density estimator
	        coarse-allele-peak-adjustment-point-count=128.\     #number of output grids for the kernel density estimator
	        coarse-allele-peak-adjustment-bandwidth=0.25.\      #adjustment to the data-adaptive bandwidth for the kernel density estimator. Must be 0 &lt; bandwidth &lt;= 1.
	        coarse-allele-peak-adjustment-cutoff=0.05.\         #parameter to remove density peaks that are likely due to noise
	        coarse-allele-peak-adjustment-clean-threshold=0.35.\#controls with noisy markers are removed from the visualization because they are indeterminately far from adjacent peaks.
	        coarse-allele-peak-adjustment-outlier-trim=3.0.\    #trim input allelic differences to be 3 or -3 if they go beyond those limits.
	        master-peaks-point-count=513.\                      #number of output grids for the overall density estimator
	        master-peaks-bandwidth=0.45.\                       #adjustment to the data-adaptive bandwidth for the overall density estimator
	        covariate-bin-peaks-point-count=257.\               #number of output grids for per-covariate bin density estimator
	        covariate-bin-peaks-bandwidth=0.45.\                #adjustment to the data-adaptive bandwidth for the per-covariate bin density estimator. Must be 0 &lt; bandwidth &lt;= 1 .
	        kernel-function-selection=gaussian \                #The kernel choice for the density estimator. Currently only accept Gaussian kernel.
	--local-gc-background-correction-reference-method none \   #should be always set to "none" for CytoScanHD_Array
	--local-gc-background-intensity-adjustment-method none \   #should be always set to "none" for CytoScanHD_Array
	--image-correction-intensity-adjustment-method none \      #should be always set to "none" for CytoScanHD_Array
	--cel-files CELFileList.txt                                #an ASCII file that contains the full path of input CEL files for reference creation, where the first row of the file is "cel_files" and each subsequent row corresponds to each CEL file
</pre></div><p>A command to run analysis workflow using the reference built in the command above for CytoScanHD will look like the following below (note that this would have to be edited to remove all comments starting with #). If no analysis modules are specified in the command all will be run by default and if any are specified then only those specified are run. We strongly recommend that you explicitly specify all parameters in either a script or a xml-file that is in approved ChASParam format.</p>
<div class="fragment"><pre class="fragment">
apt-copynumber-cyto -v 4 \                                 #verbose output level in the log file ranging from 1 (least details) to 4 (most details)
    --cyto2 false \                                            #true: Cytogenetics_Array; false: CytoScanHD_Array
    --doDualNormalization true \                               #true: quantile normalize CN and SNP separately; false: do not quantile normalize CN and SNP separately
    --keep-intermediate-data false \                           #true: dump out intermediate data during apt processing; false: otherwise.
    --run-geno-qc true \                                       #true: Run the GenoQC engine; false: otherwise
    --qca-file CytoScanHD_Array.r1.qca \                       #if run-geno-qc is set to be true, it is required to provide this file
    --qcc-file CytoScanHD_Array.r1.qcc \                       #if run-geno-qc is set to be true, it is required to provide this file
    --snp-qc-use-contrast true \                               #only used by SNP6 or Cytogenetics_Array, will be ignored by CytoScanHD_Array
    --snp-qc-snp-list CytoScanHD_Array.snplist.txt \           #specify the snp list to compute the processed SNPQC and Raw SNPQC. If it is not specified, apt uses all SNPs.
    --force true \                                             #true: Disable various checks including chip types; false: otherwise.
    --adapter-type-normalization false \                       #true: perform adapter type normalization; false: otherwise. Should be always set to "false" for CytoScanHD_Array as the adjustment has been incorporated in "signal-adjustment-covariates"
    --text-output false \                                      #true: Output data in ASCII text format in addition to calvin format; false: otherwise
    --cnchp-output false \                                     #true: report CNCHP file(s); false: otherwise. Either cnchp-output or cychp-output has to be true
    --cychp-output true \                                      #true: report CYCHP file(s); false: otherwise. Either cnchp-output or cychp-output has to be true
    --set-analysis-name cyhd \                                 #Analysis name to use as prefix for output files. The official name for CytoScanHD_Array is cyhd.
    --cdf-file $LIBDIR/CytoScanHD_Array.cdf \                  #specify the chp definition file
    --chrX-probes CytoScanHD_Array.chrXprobes \                #specify the chrX probe file 
    --chrY-probes CytoScanHD_Array.chrYprobes \                #specify the chrY probe file
    --annotation-file CytoScanHD_Array.na32.1.annot.db \       #specify the NetAffx database file
    --reference-input CytoScanHD_Array.na32.1.v3.REF_MODEL \   #specify the copy number reference to use
    --out-dir output_dir \                                     #specify the output directory
    --male-gender-ratio-cutoff 1.5 \                           #male gender call threshold for chrX/Y ratio gender calling
    --female-gender-ratio-cutoff 0.9 \                         #female gender call threshold for chrX/Y ratio gender calling
    --xx-cutoff 0.61 \                                         #lower limit of the gender call threshold for hasXX gender call
    --xx-cutoff-high 0.95 \                                    #upper limit of the gender call threshold for hasXX gender call
    --y-cutoff 0.58 \                                          #gender call threshold for hasY gender call
    --local-gc-background-intensity-adjustment-method none \   #should be always set to "none" for CytoScanHD_Array
    --image-correction-intensity-adjustment-method none \      #should be always set to "none" for CytoScanHD_Array
    --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method\  #apply wave correction 
        .bandwidth=101.\                                   #used in non-parametric wave smoothing
        bin-count=25.\                                     #used in non-parametric wave smoothing
        wave-count=6.\                                     #must be less than or equal to the number of waves specified in wave-correction-reference-method
        wave-smooth=true \                                 #true: apply additional non-parametric wave smoothing; false: otherwise.
    --cn-calibrate-parameters calibrated-log2ratios.\      #specify parameters to compute calibrated log2 ratios 
        alpha-cn-calibrate=0.564278.\                      #calibrated alpha for autosomes
        alpha-X-cn-calibrate=0.619453.\                    #calibrated alpha for Chr X
        alpha-Y-cn-calibrate=0.494620.\                    #calibrated alpha for Chr Y
        beta-cn-calibrate=1.\                              #calibrated beta for autosomes
        beta-X-cn-calibrate=1.\                            #calibrated beta for Chr X
        beta-Y-cn-calibrate=1 \                            #calibrated beta for Chr Y
    --analysis genotype \                                  #generate genotype calls for genotypable SNPs and the genotype calls are used by LOH analysis module. Should always specify this for CytoScanHD_Array.
    --analysis log2-ratio.\                                #log2 ratio calculation analysis module
        gc-correction=false.\                              #true: apply gc-correction; false: otherwise. Should aways set it to be false for CytoScanHD_Array as it has done by covariate adjuster module.
        median-autosome-median-normalization=true.\        #true: apply median autosome median correction; false: otherwise.
        median-smooth-marker-count=5 \                     #running median marker count for weighted log2 ratio. It is ignored by CytoScanHD_Array, only for Cytogenetics_Array.
    --log2ratio-adjustment-method log2ratio-adjustment-method-high-pass-filter.\ #high pass filter log2 ratio adjustment analysis module
        use=true \                                         #true: use this module for the analysis; false: otherwise.
    --analysis allelic-difference-CytoScan.\               #allelic peak calculation module
        outlier-trim=3.0.\                                 #trim input allelic differences to be 3 or -3 if they go beyond those limits.
        step=20.\                                          #step size for another density estimation across the genome
        window=100.\                                       #number of SNPs used by the kernel density estimator
        point-count=128.\                                  #number of output grids for the kernel density estimator
        bandwidth=0.25.\                                   #adjustment to the data-adaptive bandwidth for the kernel density estimator. Must be 0 &lt; bandwidth &lt;= 1.
        cutoff=0.05.\                                      #parameter to remove density peaks that are likely due to noise
        clean-threshold=0.35.\                             #controls with noisy markers are removed from the visualization because they are indeterminately far from adjacent peaks.
        symmetry=true \                                    #true: the data is mirrored about the X axis before fitting the density curves; false: otherwise.
    --analysis kernel-smooth.\                             #apply kernel smooth to the log2 ratios
        sigma_span=50 \                                    #bandwidth parameter for Gaussian smooth
    --analysis cn-cyto2.\                                  #Hidden Markove Model (HMM) analysis module
        hmmCN_state=\'0,1,2,3,4\'.\                        #supported CN states for autosome
        hmmCN_mu=\'-2,-0.45,0,0.3,0.51\'\.\                #HMM state means for autosome
        hmmCN_sigma=\'0.35,0.35,0.25,0.25,0.25\'.\         #HMM state sigmas for autosome
        hmmCN_state-X=\'0,1,2,3,4\'.\                      #supported CN states for Chr X
        hmmCN_mu-X=\'-2,-0.47,0,0.33,0.53\'\.\             #HMM state means for Chr X
        hmmCN_sigma-X=\'0.35,0.35,0.25,0.25,0.25\'.\       #HMM state sigmas for Chr X
        hmmCN_state-Y=\'0,1,2,3,4\'.\                      #supported CN states for Chr Y
        hmmCN_mu-Y=\'-2,-0.45,0,0.3,0.51\'\.\              #HMM state means for Chr Y
        hmmCN_sigma-Y=\'0.35,0.35,0.25,0.25,0.25\'.\       #HMM state sigmas for Chr Y
        diagonal-weight-Y=0.995.\                          #transition probability matrix diagnal for Chr Y
        mapd-weight-Y=0.\                                  #per sample adjustment parameter to HMM state sigmas using MAPD for Chr Y
        min-segment-size-Y=5.\                             #require min segment size (number of markers) for Chr Y
        hmm-confidence-weight-Y=0.6.\                      #parameter for HMM confidence for Chr Y
        diagonal-weight=0.995.\                            #transition probability matrix diagnal for autosome
        mapd-weight=0.\                                    #per sample adjustment parameter to HMM state sigmas using MAPD for autosome
        min-segment-size=5.\                               #require min segment size (number of markers) for autosome
        hmm-confidence-weight=0.6.\                        #parameter for HMM confidence for autosome
        diagonal-weight-X=0.995.\                          #transition probability matrix diagnal for Chr X
        mapd-weight-X=0.\                                  #per sample adjustment parameter to HMM state sigmas using MAPD for Chr X
        min-segment-size-X=5.\                             #require min segment size (number of markers) for Chr X, except for the par regions
        hmm-confidence-weight-X=0.6.\                      #parameter for HMM confidence for Chr X
        shrink=true \                                      #true: apply wavelet shrinkage to the log2 ratio before HMM calls; false: otherwise
    --analysis cn-cyto2-gender.\                           #Y-Gender call analysis module
        cutoff=0.5 \                                       #cutoff threshold for Y-Gender call
    --analysis cn-segment \                                #generate HMM copy number call segment table. Should always specify this for CytoScanHD_Array.
    --analysis lohCytoScan.\                               #LOH analysis module for CytoScanHD_Array
        lohCS_errorrate=0.05.\                             #control error rate for LOH algorithm
        lohCS_beta=0.001.\                                 #parameter for LOH algorithm
        lohCS_alpha=0.01.\                                 #parameter for LOH algorithm
        lohCS_separation=1000000.\                         #LOH separation parameter for LOH algorithm
        lohCS_nMinMarkers=10.\                             #Minimum marker count for LOH algorithm
        lohCS_NoCallThreshold=0.05.\                       #No call threshold for LOH algorithm
        lohCS_minGenomicSpan=1000000 \                     #Minimum genomic span for LOH algorithm
    --analysis loh-segment \                               #generate LOH call segment table. Should always specify this for CytoScanHD_Array.
    --analysis cn-neutral-loh \                            #generate copy neutral LOH segment table. Should always specify this for CytoScanHD_Array.
    --analysis mosaicism.\                                 #Mosaicism Analysis Module
        gains-boundries=\'0.0820,0.1588,0.2318,0.3130\'.\      # approximate boundaries for 20%, 40%, 60%, 80% mosaic gain
        losses-boundries=\'-0.1049,-0.2263,-0.3477,-0.4691\'.\ # approximate boundaries for 20%, 40%, 60%, 80% mosaic loss
        marker-bandwidth=6000.\                                # window size for mosaicism determination
        confidence-window=251 \                                # mosaicism confidence parameter     
    --cel-files CELFileList.txt                            #an ASCII file that contains the full path of input CEL files for single sample mode processing, where the first row of the file is "cel_files" and each subsequent row corresponds to each CEL file
</pre></div><p>WARNING: apt-copynumber-cyto will overwrite any existing output files it finds. If you wish to keep existing results make sure to specify a different output directory name.</p>
<p>NOTE: On windows the DOS prompt does not support wildcard expansion and the preferred method is to supply a text file with the path to the cel files via the '--cel-files' option (see below for details of file format).</p>
<p>NOTE: The windows DOS prompt also does not allow a continuation of a command with the '\' character, unlike unix. So in the examples shown here the '\' character should be omitted and everything entered on a single line.</p>
<p><a class="anchor" id="runtimeperformance"></a> </p>
<h2><a class="anchor" id="RunTimePerformance"></a>
Runtime Performance</h2>
<p>As a performance estimate, running the 270 Hapmap samples on local disk on a 8 processor 2GHz Dual-Core AMD Opteron Processor 870 with 16G of RAM on a 64-bit linux OS took 801 minutes. RAM usage was 14 GB memory.</p>
<p><a class="anchor" id="options"></a> </p>
<h3><a class="anchor" id="manualOptions"></a>
Options:</h3>
<div class="fragment"><pre class="fragment">apt-copynumber-cyto - A program to compute copy number 
results from DNA analysis arrays.

sample usage for CytoScanHD:
	./apt-copynumber-cyto  --cyto2 false --run-geno-qc true &lt;OPTIONS&gt; 

sample usage for Cytogenetics_Array:
	./apt-copynumber-cyto --cyto2 true --run-geno-qc false &lt;OPTIONS&gt; 

where &lt;OPTIONS&gt; are as described bellow.
(See apt-copynumber-cyto.html manual and vignettes for recommended parameters.)

options:
 Common Options (not used by all programs)
   -h, --help                           Display program options and extra
                          documentation about possible analyses. See
                          -explain for information about a specific
                          operation. [default 'false'] 
   -v, --verbose How verbose to be with status messages 0 -
                          quiet, 1 - usual messages, 2 - more
                          messages. [default '1'] 
     --console-off Turn off the default messages to the 
                          console but not logging or sockets. 
                          [default 'false'] 
     --use-socket Host and port to print messages over in
                          localhost:port format [default ''] 
     --version Display version information. [default
                          'false'] 
   -f, --force Disable various checks including chip 
                          types. Consider using --chip-type option
                          rather than --force. [default 'false'] 
     --throw-exception Throw an exception rather than calling
                          exit() on error. Useful for debugging. This
                          option is intended for command line use
                          only. If you are wrapping an Engine and 
                          want exceptions thrown, then you should 
                          call Err::setThrowStatus(true) to ensure
                          that all Err::errAbort() calls result in an
                          exception. [default 'false'] 
     --analysis-files-path Search path for analysis library files. 
                          Will override AFFX_ANALYSIS_FILES_PATH
                          environment variable. [default ''] 
     --xml-file Input parameters in XML format (Will
                          override command line settings). [default
                          ''] 
     --temp-dir Directory for temporary files when working
                          off disk. Using network mounted drives is
                          not advised. When not set, the output 
                          folder will be used. The defaut is 
                          typically the output directory or the
                          current working directory. [default ''] 
   -o, --out-dir Directory for output files. Defaults to
                          current working directory. [default '.'] 
     --log-file The name of the log file. Generally 
                          defaults to the program name in the out-dir
                          folder. [default ''] 
 Engine Options (Not used on command line)
     --command-line The command line executed. [default ''] 
     --exec-guid The GUID for the process. [default ''] 
     --program-name The name of the program [default ''] 
     --program-company The company providing the program [default
                          ''] 
     --program-version The version of the program [default ''] 
     --program-cvs-id The CVS version of the program [default ''] 
     --version-to-report The version to report in the output files.
                          [default ''] 
     --free-mem-at-start How much physical memory was available when
                          the engine run started. [default '0'] 
     --meta-data-info Meta data in key=value pair that will be
                          output in headers. [default ''] 
 Input Options
     --chipstream String representing chipstream parameters.
                          This includes normal-diploid analysis
                          parameters . [default 'chipstream'] 
     --check-input-files Does an upfront check of the syntax and
                          content of the input files: 
                          gender-override, genotype-call-override,
                          snp-reference-input files. [default 'true'] 
     --doDualNormalization The usual default action is to normalize 
                          SNP and CN probeset separately. This option
                          allows a single normalization set for
                          CytoScanHD chips. [default 'true'] 
     --config-file The configuration file name as passed from
                          GTC or the Cyto Browser. [default ''] 
     --antigenomic-probe-file The probe indexes (ProbeID - 1) of the
                          antigenomic probes on the array. [default
                          ''] 
     --snp-reference-input-file Input SNP reference file name. [default ''] 
     --reference-input Input reference file name. [default ''] 
     --probe-file File defining probe sequences and 
                          locations. [default ''] 
     --cdf-file File defining probe sets. [default ''] 
     --spf-file spf format file defining probe sets.
                          [default ''] 
     --qcc-file File defining QC probesets. [default ''] 
     --qca-file File defining QC analysis methods. [default
                          ''] 
     --cel-files Text file specifying cel files to process,
                          one per line with the first line being
                          'cel_files'. [default ''] 
     --chrX-probes File containing probe_id (1-based) of 
                          probes on chrX. Used for copy number probe
                          chrX/Y ratio gender calling. [default ''] 
     --chrY-probes File containing probe_id (1-based) of 
                          probes on chrY. Used for copy number probe
                          chrX/Y ratio gender calling. [default ''] 
     --normal-diploid-files-file Text file specifying normal-diploid 
                          probeset files. First line must be
                          'normal_diploid_files'. [default ''] 
     --reference-cels 'Normal' CEL file(s) to process when doing
                          paired analysis of Cancer vs. Normal.
                          [default ''] 
     --annotation-file NetAffx Annotation database file. [default
                          ''] 
     --normal-diploid-files Normal Diploid probeset file names. 
                          [default ''] 
     --sketch-size The number number of point to be used for a
                          sketch. [default '50000'] 
     --covariates-file External covariates file. [default ''] 
     --minSegSeparation Value used to skip over centromere in LOH.
                          [default '1000000000'] 
 Output Options
     --snp-reference-output-file Output SNP reference file name. [default 
                          ''] 
     --reference-output Output reference file name. [default ''] 
     --file-name-prefix CYCHP file name prefix. [default ''] 
     --file-name-suffix CYCHP file name suffix. [default ''] 
     --file-name-ext CYCHP file name extension. [default 
                          'cychp'] 
 Analysis Options
     --run-geno-qc Run the GenoQC engine. [default 'false'] 
     --adapter-type-normalization Adapter Type Normalization option. true =
                          perform adapter type normalization. 
                          [default 'false'] 
   -a, --analysis String representing analysis pathway
                          desired. [default ''] 
     --wave-correction-reference-method String representing wave correction pathway
                          desired. [default ''] 
     --local-gc-background-correction-reference-method String representing local gc correction
                          background pathway desired. [default 
                          'none'] 
     --local-gc-background-intensity-adjustment-method String representing local gc correction
                          background intensity adjustment pathway
                          desired. [default ''] 
     --image-correction-intensity-adjustment-method String representing image correction
                          intensity adjustment pathway desired.
                          [default ''] 
     --log2ratio-adjustment-method String representing high pass filter
                          parameters for modification of log2ratios.
                          [default ''] 
     --wave-correction-log2ratio-adjustment-method String representing wave correction
                          log2ratio adjustment pathway desired.
                          [default ''] 
     --allele-peaks-reporter-method String representing allele peaks reporter
                          pathway desired. [default
                          'allele-peaks-reporter-method'] 
     --hmm-priors Read HMM priors from the reference file.
                          [default 'false'] 
     --hmm-means-check Check HMM priors in the reference file for
                          consistency. [default 'false'] 
     --brlmmp-parameters Parameters to use when running brlmmp.
                          [default ''] 
     --signal-adjustment-covariates String representing the covariate-based
                          signal adjustment pathway desired. [default
                          ''] 
     --lr-adjustment-covariates String representing the covariate-based 
                          log2 ratio adjustment pathway desired.
                          [default ''] 
     --allele-peaks-adjustment-covariates String representing the covariate-based
                          allele peaks adjustment pathway desired.
                          [default ''] 
 QC Options
     --snp-qc-use-contrast SNP QC use intensity contrast. [default
                          'false'] 
     --snp-qc-snp-list Input file containing a list of SNP Ids to
                          be used in calculating the SNPQC value.
                          [default ''] 
     --snp-qc-k SNP QC K value. [default '2.0'] 
     --snp-qc-em-threshold SNP QC EM Threshold. [default '0.05'] 
     --snp-qc-bin-size SNP QC bin size. [default '0.04'] 
 Misc Options
     --explain Explain a particular operation (i.e.
                          --explain cn-state or --explain loh).
                          [default ''] 
 Advanced Options
     --probeset-ids Tab delimited file with column 
                          'probeset_id' specifying probesets to
                          summarize. [default ''] 
     --global-parameter-override Global parameters in loh-cyto analysis 
                          input string will override those given in
                          the snp-reference-input-file. [default
                          'false'] 
     --keep-temp-reference-data Set to true, this option will keep the 
                          final signal and intensity values computed
                          while determining the CN reference file.
                          Warning: It may be large. [default 'false'] 
     --keep-intermediate-data Set to true, this option will keep all,
                          intensity values computed while invoking 
                          any intensity adjustment method. [default
                          'false'] 
     --keep-intermediate-data-local Set to true, this option will keep all,
                          intensity values computed while invoking 
                          any intensity adjustment method. This is a
                          duplicate option for local testing. 
                          [default 'false'] 
     --genotype-call-override-file Input file containing genotype calls to be
                          used in place of brlmmp-p calls. [default
                          ''] 
     --gender-override-file A file containing externally computed
                          genders. It is used in reference and
                          snp-reference generation [default ''] 
     --gc-content-override-file Input file used to override the GC content
                          read from the annotation files (Two columns
                          with header line, ProbeSetName/GCContent).
                          [default ''] 
     --gc-correction-bin-count The number of bins to use for GC content.
                          [default '25'] 
     --xChromosome X Chromosome [default '24'] 
     --yChromosome Y Chromosome [default '25'] 
     --male-gender-ratio-cutoff Male gender ratio cutoff [default '1.3'] 
     --female-gender-ratio-cutoff Female gender ratio cutoff [default '1.0'] 
     --reference-chromosome Reference chromosome [default '2'] 
     --xx-cutoff XX cutoff [default '0.8'] 
     --xx-cutoff-high XX cutoff high [default '1.07'] 
     --y-cutoff Y cutoff [default '0.65'] 
     --warning-message-limit Set the limit on the number of warning
                          message produced. [default '10'] 
     --waviness-block-size marker count [default '50'] 
     --waviness-genomic-span genomic segment length [default '0'] 
     --cn-calibrate-parameters SmoothSignal calibration parameters 
                          [default ''] 
     --use-old-kdensity-function For testing only, revert to the old slow
                          kdensity method for density estimation.
                          [default 'false'] 
 Engine Options (Not used on command line)
     --cels CEL files to process. [default ''] 
     --arrs ARR files to process. Must be paired with
                          cels. [default ''] 
     --result-files CYCHP files to output. Must be paired with
                          cels. [default ''] 
 Additional CNAnalysisEngine Options
     --geno-qc-file The file output from GenoQC. [default ''] 
     --cyto2 Processing CYTO2 chip. [default 'false'] 
     --array-name Array name or type to use. [default ''] 
     --set-analysis-name Analysis name to use as prefix for output
                          files. [default ''] 
     --text-output Output data in ASCII text format in 
                          addition to calvin format. [default 
                          'false'] 
     --cnchp-output Report CNCHP files [default 'true'] 
     --cychp-output Report CYCHP files [default 'false'] 
     --time-start The time the engine run was started 
                          [default ''] 
     --time-end The time the engine run ended [default ''] 
     --time-run-minutes The run time in minutes. [default ''] 
     --analysis-guid The GUID for the analysis run. [default ''] 

Data transformations:
   pdnn-reference-method CopyNumber PDNN 
   wave-correction-reference-methodCopyNumber WaveCorrection 
   additional-waves-reference-methodCopyNumber AdditionalWaves 
   pdnn-intensity-adjustment-methodCopynumber PDNN Intensity
                         Adjustment 
   high-pass-filter-intensity-adjustment-methodCopyNumber
                         HighPassFilter 
   wave-correction-log2ratio-adjustment-methodCopynumber Wave
                         Correction Log2Ratio Adjustment 
   log2ratio-adjustment-method-high-pass-filterCopyNumber
                         HighPassFilter Log2Ratio Adjustment 
   cn-state              CopyNumber CNState 
   cn-cyto2              CopyNumber CNCyto2 
   log2-ratio            Copynumber Log2Ratio 
   log2-ratio-cyto2      Copynumber Log2RatioCyto2 
   allelic-difference    Copynumber AllelicDifference 
   allelic-difference-CytoScanCopynumber AllelicDifference CytoScan 
   gaussian-smooth       CopyNumber GaussianSmooth 
   genotype              Genotype 
   kernel-smooth         CopyNumber KernelSmooth 
   loh                   CopyNumber LOH 
   loh-cyto2             CopyNumber LOH Cyto2 
   lohCytoScan           CopyNumber LOH CytoScan 
   cn-neutral-loh        Copynumber CNNeutralLOH 
   normal-diploid        Copynumber NormalDiploid 
   mosaicism             Copynumber Mosaicism 
   cn-gender             Copynumber CNGender 
   cn-cyto2-gender       Copynumber CNCyto2Gender 
   cn-segment            Copynumber SegmentCN 
   loh-segment           Copynumber SegmentLOH 
   allele-peaks          Copynumber AllelePeaks 
   chipstream            Copynumber Chipstream 
   covariate-signal-adjusterCovariate Signal Adjuster 
   covariate-lr-adjuster Covariate log2ratio Adjuster 
</pre></div><p><a class="anchor" id="faq"></a> </p>
<h2><a class="anchor" id="manFaq"></a>
Frequently Asked Questions</h2>
<p>Q. Does apt-copynumber-cyto support the whole genome SNP6 chips.</p>
<p>A. No. For SNP6 use apt-copynumber-workflow. </p>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:14 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
