<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/apt-copynumber-cyto/regression/apt-rt-copynumber-cyto.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
      <li><a class="el" href="dir_d61a7909814a46bce5ab2f5f0e583661.html">apt-copynumber-cyto</a>      </li>
      <li><a class="el" href="dir_295cb3590393dbdce00cd66c6df31eff.html">regression</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>apt-rt-copynumber-cyto.cpp</h1>  </div>
</div>
<div class="contents">
<div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2011 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment">// This has to be last. Otherwise windows compile fails</span>
<a name="l00022"></a>00022 <span class="preprocessor">#include &quot;<a class="code" href="Calvin_8h.html" title="This header contains the Calvin class definition.">calvin_files/utils/src/Calvin.h</a>&quot;</span>
<a name="l00023"></a>00023 <span class="preprocessor">#include &quot;file5/File5.h&quot;</span>
<a name="l00024"></a>00024 <span class="preprocessor">#include &quot;rtest/RT_Args.h&quot;</span>
<a name="l00025"></a>00025 <span class="preprocessor">#include &quot;rtest/RT_Check.h&quot;</span>
<a name="l00026"></a>00026 <span class="preprocessor">#include &quot;rtest/RT_Cmd.h&quot;</span>
<a name="l00027"></a>00027 <span class="preprocessor">#include &quot;rtest/RT_Main.h&quot;</span>
<a name="l00028"></a>00028 <span class="preprocessor">#include &quot;rtest/RT_Test.h&quot;</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;rtest/RT_Util.h&quot;</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="LogStream_8h.html" title="Output verbose messages to log file as we go along.">util/LogStream.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="Verbose_8h.html" title="Class for doing logging and some command line ui.">util/Verbose.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="comment">//</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &lt;cstring&gt;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00038"></a>00038 <span class="comment">//</span>
<a name="l00039"></a>00039 <span class="keyword">using namespace </span>std;
<a name="l00040"></a>00040 
<a name="l00041"></a>00041 RT_Test* qt_single();
<a name="l00042"></a>00042 RT_Test* qt_batch();
<a name="l00043"></a>00043 RT_Test* single();
<a name="l00044"></a>00044 RT_Test* batch();
<a name="l00045"></a>00045 
<a name="l00046"></a>00046 <span class="keyword">const</span> <span class="keywordtype">char</span>* cytogenetics_array_sixfiles[] = {<span class="stringliteral">&quot;HapMap-As_NA18547_A08_01_NN_20081218&quot;</span>, <span class="stringliteral">&quot;HapMap-As_NA18603_B03_01_NN_20081218&quot;</span>,
<a name="l00047"></a>00047                          <span class="stringliteral">&quot;HapMap-As_NA18943_C11_01_NN_20081218&quot;</span>, <span class="stringliteral">&quot;HapMap-As_NA18971_C02_01_NN_20081218&quot;</span>,
<a name="l00048"></a>00048                          <span class="stringliteral">&quot;HapMap-Cs_NA10857_A10_01_NN_20081218&quot;</span>, <span class="stringliteral">&quot;HapMap-Cs_NA12003_D04_01_NN_20081218&quot;</span>,
<a name="l00049"></a>00049                          NULL
<a name="l00050"></a>00050 };
<a name="l00051"></a>00051 
<a name="l00052"></a>00052 <span class="keyword">const</span> <span class="keywordtype">char</span>* qt_cytogenetics_array_sixfiles[] = {<span class="stringliteral">&quot;HapMap-As_NA18547_A08_01_NN_20081218&quot;</span>, <span class="stringliteral">&quot;HapMap-As_NA18603_B03_01_NN_20081218&quot;</span>,
<a name="l00053"></a>00053                         <span class="stringliteral">&quot;HapMap-As_NA18971_C02_01_NN_20081218&quot;</span>,
<a name="l00054"></a>00054                         <span class="stringliteral">&quot;HapMap-Cs_NA10857_A10_01_NN_20081218&quot;</span>, <span class="stringliteral">&quot;HapMap-Cs_NA12003_D04_01_NN_20081218&quot;</span>,
<a name="l00055"></a>00055                          NULL
<a name="l00056"></a>00056 };
<a name="l00057"></a>00057 
<a name="l00058"></a>00058 <span class="comment">// Data Set columns to ignore (Specify as Group.Set.Column)</span>
<a name="l00059"></a>00059 <span class="comment">//ignoreHeaders.insert(&quot;AlgorithmData.MarkerABSignal.SCAR&quot;);</span>
<a name="l00060"></a>00060 
<a name="l00061"></a>00061 <span class="comment">//Data Set column epsilon override (Specify as Group.Set.Column)</span>
<a name="l00062"></a>00062 <span class="comment">//mapEpsilon.insert(std::pair&lt;std::string, float&gt;(&quot;AlgorithmData.MarkerABSignal.SCAR&quot;, (float)0.0005));</span>
<a name="l00063"></a>00063 
<a name="l00064"></a>00064 <span class="comment"></span>
<a name="l00065"></a>00065 <span class="comment">///////////////////////////////////////////////////////////////////////////////////////////////////</span>
<a name="l00066"></a>00066 <span class="comment"></span><span class="keywordtype">int</span> <a class="code" href="canary-test_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="Everybody&amp;#39;s favorite function.">main</a>(<span class="keywordtype">int</span> argc, <span class="keywordtype">char</span>* argv[])
<a name="l00067"></a>00067 {
<a name="l00068"></a>00068   <span class="keywordflow">if</span> ( !Fs::dirExists(<span class="stringliteral">&quot;test-generated&quot;</span>) ) {
<a name="l00069"></a>00069     <a class="code" href="classFs.html#abf363360d88071c2f4a93a7e50d61c9f" title="Make one directory. ///.">Fs::mkdir</a>(<span class="stringliteral">&quot;test-generated&quot;</span>);
<a name="l00070"></a>00070   }
<a name="l00071"></a>00071 
<a name="l00072"></a>00072   RT_Main* rMain;
<a name="l00073"></a>00073   rMain = <span class="keyword">new</span> RT_Main(<span class="stringliteral">&quot;apt-rt-copynumber-cyto&quot;</span>);
<a name="l00074"></a>00074 
<a name="l00075"></a>00075   <span class="comment">//Store names of tests for Help Message</span>
<a name="l00076"></a>00076   <span class="comment">//Store it in our RT_Main Instance</span>
<a name="l00077"></a>00077   std::vector&lt;std::string&gt; testNameList;
<a name="l00078"></a>00078   testNameList.push_back(<span class="stringliteral">&quot;single&quot;</span>);
<a name="l00079"></a>00079   testNameList.push_back(<span class="stringliteral">&quot;batch&quot;</span>);
<a name="l00080"></a>00080   testNameList.push_back(<span class="stringliteral">&quot;qt_single&quot;</span>);
<a name="l00081"></a>00081   testNameList.push_back(<span class="stringliteral">&quot;qt_batch&quot;</span>);
<a name="l00082"></a>00082   rMain-&gt;setTestNameList(testNameList);
<a name="l00083"></a>00083 
<a name="l00084"></a>00084   <span class="comment">//Set Variables</span>
<a name="l00085"></a>00085   rMain-&gt;setVar(<span class="stringliteral">&quot;sdktop&quot;</span>, <span class="stringliteral">&quot;../../..&quot;</span>);
<a name="l00086"></a>00086   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_idata_copynumbercyto_testname&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/idata/copynumber-cyto/${testname}&quot;</span>);
<a name="l00087"></a>00087   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_idata_copynumbercyto_cytogeneticsarray_testname&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/idata/copynumber-cyto/Cytogenetics_Array/${testname}&quot;</span>);
<a name="l00088"></a>00088   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_idata_copynumbercyto_batch&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/idata/copynumber-cyto/batch&quot;</span>);
<a name="l00089"></a>00089   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_idata_copynumbercyto_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/idata/copynumber-cyto/Cytogenetics_Array&quot;</span>);
<a name="l00090"></a>00090   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_idata_lib_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/idata/lib/Cytogenetics_Array&quot;</span>);
<a name="l00091"></a>00091   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_idata_cel_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/idata/cel/Cytogenetics_Array&quot;</span>);
<a name="l00092"></a>00092   rMain-&gt;setVar(<span class="stringliteral">&quot;out_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${out_dir}/Cytogenetics_Array&quot;</span>);
<a name="l00093"></a>00093   rMain-&gt;setVar(<span class="stringliteral">&quot;out_testname&quot;</span>, <span class="stringliteral">&quot;${out_dir}/${testname}&quot;</span>);
<a name="l00094"></a>00094   rMain-&gt;setVar(<span class="stringliteral">&quot;out_batch&quot;</span>, <span class="stringliteral">&quot;${out_dir}/batch&quot;</span>);
<a name="l00095"></a>00095   rMain-&gt;setVar(<span class="stringliteral">&quot;out_cytogeneticsarray_testname&quot;</span>, <span class="stringliteral">&quot;${out_dir}/Cytogenetics_Array/${testname}&quot;</span>);
<a name="l00096"></a>00096 
<a name="l00097"></a>00097   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_testname&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/${testname}&quot;</span>);
<a name="l00098"></a>00098   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_lib_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/lib/Cytogenetics_Array&quot;</span>);
<a name="l00099"></a>00099   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_cel_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/cel/Cytogenetics_Array&quot;</span>);
<a name="l00100"></a>00100   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_copynumbercyto&quot;</span>, <span class="stringliteral">&quot;${gold_dir}/copynumber-cyto&quot;</span>);
<a name="l00101"></a>00101   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_copynumbercyto_testname&quot;</span>, <span class="stringliteral">&quot;${gold_copynumbercyto}/${testname}&quot;</span>);
<a name="l00102"></a>00102   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_copynumbercyto_cytogeneticsarray_testname&quot;</span>, <span class="stringliteral">&quot;${gold_copynumbercyto}/Cytogenetics_Array/${testname}&quot;</span>);
<a name="l00103"></a>00103   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_copynumbercyto_cytogeneticsarray&quot;</span>, <span class="stringliteral">&quot;${gold_copynumbercyto}/Cytogenetics_Array&quot;</span>);
<a name="l00104"></a>00104   rMain-&gt;setVar(<span class="stringliteral">&quot;gold_copynumbercyto_batch&quot;</span>, <span class="stringliteral">&quot;${gold_copynumbercyto}/batch&quot;</span>);
<a name="l00105"></a>00105   rMain-&gt;parseArgv(argc, argv);
<a name="l00106"></a>00106   <span class="keywordflow">if</span>(argc &lt; 1)
<a name="l00107"></a>00107     {
<a name="l00108"></a>00108       cout&lt;&lt;<span class="stringliteral">&quot;Error: Incorrect number of arguments given&quot;</span>&lt;&lt;endl;
<a name="l00109"></a>00109       exit(1);
<a name="l00110"></a>00110     }
<a name="l00111"></a>00111   <span class="keywordflow">else</span> <span class="keywordflow">if</span>(rMain-&gt;getVarVal(<span class="stringliteral">&quot;allIntegration&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00112"></a>00112     {
<a name="l00113"></a>00113       rMain-&gt;addTest(qt_single());
<a name="l00114"></a>00114       rMain-&gt;addTest(qt_batch());
<a name="l00115"></a>00115     }
<a name="l00116"></a>00116   <span class="keywordflow">else</span> <span class="keywordflow">if</span>(rMain-&gt;getVarVal(<span class="stringliteral">&quot;allNormal&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00117"></a>00117     {
<a name="l00118"></a>00118       rMain-&gt;addTest(single());
<a name="l00119"></a>00119       rMain-&gt;addTest(batch());
<a name="l00120"></a>00120     }
<a name="l00121"></a>00121   <span class="comment">//They want us to run all default tests</span>
<a name="l00122"></a>00122   <span class="keywordflow">else</span> <span class="keywordflow">if</span>(argc == 1 || rMain-&gt;getVarVal(<span class="stringliteral">&quot;all&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00123"></a>00123     {
<a name="l00124"></a>00124       <span class="comment">//RUN ALL TESTS OMG</span>
<a name="l00125"></a>00125       cout&lt;&lt;<span class="stringliteral">&quot;Running all default Regression Tests!&quot;</span>&lt;&lt;endl;
<a name="l00126"></a>00126       rMain-&gt;addTest(qt_single());
<a name="l00127"></a>00127       rMain-&gt;addTest(qt_batch());
<a name="l00128"></a>00128       rMain-&gt;addTest(single());
<a name="l00129"></a>00129       rMain-&gt;addTest(batch());
<a name="l00130"></a>00130     }
<a name="l00131"></a>00131   <span class="keywordflow">else</span>
<a name="l00132"></a>00132     {
<a name="l00133"></a>00133       cout&lt;&lt;<span class="stringliteral">&quot;Running Built-In Regression Tests&quot;</span>&lt;&lt;endl;
<a name="l00134"></a>00134 
<a name="l00135"></a>00135       <span class="keywordflow">if</span>(rMain-&gt;getVarVal(<span class="stringliteral">&quot;qt_single&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00136"></a>00136     {
<a name="l00137"></a>00137       rMain-&gt;addTest(qt_single());
<a name="l00138"></a>00138     }
<a name="l00139"></a>00139       <span class="keywordflow">if</span>(rMain-&gt;getVarVal(<span class="stringliteral">&quot;qt_batch&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00140"></a>00140     {
<a name="l00141"></a>00141       rMain-&gt;addTest(qt_batch());
<a name="l00142"></a>00142     }
<a name="l00143"></a>00143       <span class="keywordflow">if</span>(rMain-&gt;getVarVal(<span class="stringliteral">&quot;single&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00144"></a>00144     {
<a name="l00145"></a>00145       rMain-&gt;addTest(single());
<a name="l00146"></a>00146     }
<a name="l00147"></a>00147       <span class="keywordflow">if</span>(rMain-&gt;getVarVal(<span class="stringliteral">&quot;batch&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00148"></a>00148     {
<a name="l00149"></a>00149       rMain-&gt;addTest(batch());
<a name="l00150"></a>00150     }
<a name="l00151"></a>00151     }
<a name="l00152"></a>00152 
<a name="l00153"></a>00153   rMain-&gt;runTests();
<a name="l00154"></a>00154 }
<a name="l00155"></a>00155 <span class="comment"></span>
<a name="l00156"></a>00156 <span class="comment">///////////////////////////////////////////////////////////////////////////////////////////////////</span>
<a name="l00157"></a>00157 <span class="comment"></span>RT_Test* qt_single()
<a name="l00158"></a>00158 {
<a name="l00159"></a>00159   RT_Test* rt;
<a name="l00160"></a>00160   rt = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;single&quot;</span>);
<a name="l00161"></a>00161 
<a name="l00162"></a>00162   RT_Cmd* cmd = rt-&gt;newCmd();
<a name="l00163"></a>00163   cmd-&gt;setExe(<span class="stringliteral">&quot;${apt_copynumber_cyto}&quot;</span>);
<a name="l00164"></a>00164   cmd-&gt;addArg(<span class="stringliteral">&quot;-v&quot;</span>, <span class="stringliteral">&quot;3&quot;</span>);
<a name="l00165"></a>00165   cmd-&gt;addArg(<span class="stringliteral">&quot;--cyto2&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00166"></a>00166   cmd-&gt;addArg(<span class="stringliteral">&quot;--force&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00167"></a>00167   cmd-&gt;addArg(<span class="stringliteral">&quot;--text-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00168"></a>00168   cmd-&gt;addArg(<span class="stringliteral">&quot;--cnchp-output&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00169"></a>00169   cmd-&gt;addArg(<span class="stringliteral">&quot;--cychp-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00170"></a>00170   cmd-&gt;addArg(<span class="stringliteral">&quot;--set-analysis-name&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>);
<a name="l00171"></a>00171   cmd-&gt;addArg(<span class="stringliteral">&quot;--reference-input&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.na28.REF_MODEL&quot;</span>);
<a name="l00172"></a>00172   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-input-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.na28.snpref.a5&quot;</span>);
<a name="l00173"></a>00173   cmd-&gt;addArg(<span class="stringliteral">&quot;--spf-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.spf&quot;</span>);
<a name="l00174"></a>00174   cmd-&gt;addArg(<span class="stringliteral">&quot;--chrX-probes&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.chrXprobes&quot;</span>);
<a name="l00175"></a>00175   cmd-&gt;addArg(<span class="stringliteral">&quot;--chrY-probes&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.chrYprobes&quot;</span>);
<a name="l00176"></a>00176   cmd-&gt;addArg(<span class="stringliteral">&quot;--annotation-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.na28.annot.db&quot;</span>);
<a name="l00177"></a>00177   cmd-&gt;addArg(<span class="stringliteral">&quot;--probe-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.probe_tab&quot;</span>);
<a name="l00178"></a>00178   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-qc-snp-list&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.snplist.txt&quot;</span>);
<a name="l00179"></a>00179   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_idata_cel_cytogeneticsarray}/HapMap-As_NA18547_A08_01_NN_20081218.CEL&quot;</span>);
<a name="l00180"></a>00180   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_idata_cel_cytogeneticsarray}/HapMap-As_NA18603_B03_01_NN_20081218.CEL&quot;</span>);
<a name="l00181"></a>00181   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_idata_cel_cytogeneticsarray}/HapMap-As_NA18971_C02_01_NN_20081218.CEL&quot;</span>);
<a name="l00182"></a>00182   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_idata_cel_cytogeneticsarray}/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL&quot;</span>);
<a name="l00183"></a>00183   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_idata_cel_cytogeneticsarray}/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL&quot;</span>);
<a name="l00184"></a>00184   cmd-&gt;addArg(<span class="stringliteral">&quot;--out-dir&quot;</span>, <span class="stringliteral">&quot;${out_testname}&quot;</span>);
<a name="l00185"></a>00185   cmd-&gt;addArg(<span class="stringliteral">&quot;--male-gender-ratio-cutoff=1.3&quot;</span>);
<a name="l00186"></a>00186   cmd-&gt;addArg(<span class="stringliteral">&quot;--female-gender-ratio-cutoff=1.0&quot;</span>);
<a name="l00187"></a>00187   cmd-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff=0.8&quot;</span>);
<a name="l00188"></a>00188   cmd-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff-high=1.07&quot;</span>);
<a name="l00189"></a>00189   cmd-&gt;addArg(<span class="stringliteral">&quot;--y-cutoff=0.65&quot;</span>);
<a name="l00190"></a>00190   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span>);
<a name="l00191"></a>00191   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;kernel-smooth.sigma_span=50&quot;</span>);
<a name="l00192"></a>00192   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span>);
<a name="l00193"></a>00193   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2-gender.cutoff=0.5&quot;</span>);
<a name="l00194"></a>00194   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-segment&quot;</span>);
<a name="l00195"></a>00195   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span>);
<a name="l00196"></a>00196   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span>);
<a name="l00197"></a>00197   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span>);
<a name="l00198"></a>00198   cmd-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;pdnn-reference-method&quot;</span>);
<a name="l00199"></a>00199   cmd-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span>);
<a name="l00200"></a>00200   cmd-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span>);
<a name="l00201"></a>00201   cmd-&gt;addArg(<span class="stringliteral">&quot;--image-correction-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span>);
<a name="l00202"></a>00202   cmd-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-log2ratio-adjustment-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;</span>);
<a name="l00203"></a>00203   cmd-&gt;addArg(<span class="stringliteral">&quot;--reference-chromosome=22&quot;</span>);
<a name="l00204"></a>00204 
<a name="l00205"></a>00205   std::vector&lt;std::string&gt; gen, gold;
<a name="l00206"></a>00206   gen = Util::addPrefixSuffix(qt_cytogenetics_array_sixfiles, <span class="stringliteral">&quot;${out_testname}/&quot;</span>, <span class="stringliteral">&quot;.ca.cychp&quot;</span>);
<a name="l00207"></a>00207   gold = Util::addPrefixSuffix(qt_cytogenetics_array_sixfiles, <span class="stringliteral">&quot;${gold_idata_copynumbercyto_testname}/&quot;</span>, <span class="stringliteral">&quot;.ca.cychp&quot;</span>);
<a name="l00208"></a>00208 
<a name="l00209"></a>00209 RT_Check* check = rt-&gt;newCheck();
<a name="l00210"></a>00210   check-&gt;setExe(<span class="stringliteral">&quot;${apt_calvin_equivalent}&quot;</span>);
<a name="l00211"></a>00211   check-&gt;setGenFiles(gen);
<a name="l00212"></a>00212   check-&gt;setGoldFiles(gold);
<a name="l00213"></a>00213   check-&gt;setGenGoldFileIsArg(<span class="keyword">true</span>);
<a name="l00214"></a>00214   check-&gt;addArg(<span class="stringliteral">&quot;--ignore-params-file&quot;</span>, <span class="stringliteral">&quot;./ignoreParamFile.tsv&quot;</span>);
<a name="l00215"></a>00215   check-&gt;addArg(<span class="stringliteral">&quot;--epsilon-map-file&quot;</span>, <span class="stringliteral">&quot;./qt_mapEpsilonFile.tsv&quot;</span>);
<a name="l00216"></a>00216   check-&gt;addArg(<span class="stringliteral">&quot;--epsilon&quot;</span>, <span class="stringliteral">&quot;0.0001&quot;</span>);
<a name="l00217"></a>00217   check-&gt;addArg(<span class="stringliteral">&quot;--correlation&quot;</span>, <span class="stringliteral">&quot;0.9999989&quot;</span>);
<a name="l00218"></a>00218   check-&gt;addArg(<span class="stringliteral">&quot;--check-header&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00219"></a>00219 
<a name="l00220"></a>00220   <span class="keywordflow">return</span> rt;
<a name="l00221"></a>00221 }
<a name="l00222"></a>00222 <span class="comment"></span>
<a name="l00223"></a>00223 <span class="comment">///////////////////////////////////////////////////////////////////////////////////////////////////</span>
<a name="l00224"></a>00224 <span class="comment"></span>RT_Test* single()
<a name="l00225"></a>00225 {
<a name="l00226"></a>00226   RT_Test* rt;
<a name="l00227"></a>00227   rt = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;single&quot;</span>);
<a name="l00228"></a>00228 
<a name="l00229"></a>00229   RT_Cmd* cmd = rt-&gt;newCmd();
<a name="l00230"></a>00230   cmd-&gt;setExe(<span class="stringliteral">&quot;${apt_copynumber_cyto}&quot;</span>);
<a name="l00231"></a>00231   cmd-&gt;addArg(<span class="stringliteral">&quot;-v&quot;</span>, <span class="stringliteral">&quot;3&quot;</span>);
<a name="l00232"></a>00232   cmd-&gt;addArg(<span class="stringliteral">&quot;--cyto2&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00233"></a>00233   cmd-&gt;addArg(<span class="stringliteral">&quot;--force&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00234"></a>00234   cmd-&gt;addArg(<span class="stringliteral">&quot;--text-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00235"></a>00235   cmd-&gt;addArg(<span class="stringliteral">&quot;--cnchp-output&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00236"></a>00236   cmd-&gt;addArg(<span class="stringliteral">&quot;--cychp-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00237"></a>00237   cmd-&gt;addArg(<span class="stringliteral">&quot;--set-analysis-name&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>);
<a name="l00238"></a>00238   cmd-&gt;addArg(<span class="stringliteral">&quot;--reference-input&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.na28.REF_MODEL&quot;</span>);
<a name="l00239"></a>00239   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-input-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.na28.snpref.a5&quot;</span>);
<a name="l00240"></a>00240   cmd-&gt;addArg(<span class="stringliteral">&quot;--cdf-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.CDF&quot;</span>);
<a name="l00241"></a>00241   cmd-&gt;addArg(<span class="stringliteral">&quot;--chrX-probes&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.chrXprobes&quot;</span>);
<a name="l00242"></a>00242   cmd-&gt;addArg(<span class="stringliteral">&quot;--chrY-probes&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.chrYprobes&quot;</span>);
<a name="l00243"></a>00243   cmd-&gt;addArg(<span class="stringliteral">&quot;--annotation-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.na28.annot.db&quot;</span>);
<a name="l00244"></a>00244   cmd-&gt;addArg(<span class="stringliteral">&quot;--probe-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.probe_tab&quot;</span>);
<a name="l00245"></a>00245   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-qc-snp-list&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.snplist.txt&quot;</span>);
<a name="l00246"></a>00246   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18547_A08_01_NN_20081218.CEL&quot;</span>);
<a name="l00247"></a>00247   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18603_B03_01_NN_20081218.CEL&quot;</span>);
<a name="l00248"></a>00248   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18943_C11_01_NN_20081218.CEL&quot;</span>);
<a name="l00249"></a>00249   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18971_C02_01_NN_20081218.CEL&quot;</span>);
<a name="l00250"></a>00250   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL&quot;</span>);
<a name="l00251"></a>00251   cmd-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL&quot;</span>);
<a name="l00252"></a>00252   cmd-&gt;addArg(<span class="stringliteral">&quot;--out-dir&quot;</span>, <span class="stringliteral">&quot;${out_testname}&quot;</span>);
<a name="l00253"></a>00253   cmd-&gt;addArg(<span class="stringliteral">&quot;--male-gender-ratio-cutoff=1.3&quot;</span>);
<a name="l00254"></a>00254   cmd-&gt;addArg(<span class="stringliteral">&quot;--female-gender-ratio-cutoff=1.0&quot;</span>);
<a name="l00255"></a>00255   cmd-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff=0.8&quot;</span>);
<a name="l00256"></a>00256   cmd-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff-high=1.07&quot;</span>);
<a name="l00257"></a>00257   cmd-&gt;addArg(<span class="stringliteral">&quot;--y-cutoff=0.65&quot;</span>);
<a name="l00258"></a>00258   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span>);
<a name="l00259"></a>00259   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;kernel-smooth.sigma_span=50&quot;</span>);
<a name="l00260"></a>00260   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span>);
<a name="l00261"></a>00261   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2-gender.cutoff=0.5&quot;</span>);
<a name="l00262"></a>00262   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-segment&quot;</span>);
<a name="l00263"></a>00263   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span>);
<a name="l00264"></a>00264   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span>);
<a name="l00265"></a>00265   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span>);
<a name="l00266"></a>00266   cmd-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;pdnn-reference-method&quot;</span>);
<a name="l00267"></a>00267   cmd-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span>);
<a name="l00268"></a>00268   cmd-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span>);
<a name="l00269"></a>00269   cmd-&gt;addArg(<span class="stringliteral">&quot;--image-correction-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span>);
<a name="l00270"></a>00270   cmd-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-log2ratio-adjustment-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;</span>);
<a name="l00271"></a>00271 
<a name="l00272"></a>00272   std::vector&lt;std::string&gt; gen, gold;
<a name="l00273"></a>00273   gen = Util::addPrefixSuffix(cytogenetics_array_sixfiles, <span class="stringliteral">&quot;${out_testname}/&quot;</span>, <span class="stringliteral">&quot;.ca.cychp&quot;</span>);
<a name="l00274"></a>00274   gold = Util::addPrefixSuffix(cytogenetics_array_sixfiles, <span class="stringliteral">&quot;${gold_copynumbercyto_testname}/&quot;</span>, <span class="stringliteral">&quot;.ca.cychp&quot;</span>);
<a name="l00275"></a>00275 
<a name="l00276"></a>00276   RT_Check* check = rt-&gt;newCheck();
<a name="l00277"></a>00277   check-&gt;setExe(<span class="stringliteral">&quot;${apt_calvin_equivalent}&quot;</span>);
<a name="l00278"></a>00278   check-&gt;setGenFiles(gen);
<a name="l00279"></a>00279   check-&gt;setGoldFiles(gold);
<a name="l00280"></a>00280   check-&gt;setGenGoldFileIsArg(<span class="keyword">true</span>);
<a name="l00281"></a>00281   check-&gt;addArg(<span class="stringliteral">&quot;--ignore-params-file&quot;</span>, <span class="stringliteral">&quot;./ignoreParamFile.tsv&quot;</span>);
<a name="l00282"></a>00282   check-&gt;addArg(<span class="stringliteral">&quot;--epsilon&quot;</span>, <span class="stringliteral">&quot;0.0001&quot;</span>);
<a name="l00283"></a>00283   check-&gt;addArg(<span class="stringliteral">&quot;--correlation&quot;</span>, <span class="stringliteral">&quot;0.9999989&quot;</span>);
<a name="l00284"></a>00284   check-&gt;addArg(<span class="stringliteral">&quot;--check-header&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00285"></a>00285 
<a name="l00286"></a>00286   <span class="keywordflow">return</span> rt;
<a name="l00287"></a>00287 }
<a name="l00288"></a>00288 <span class="comment"></span>
<a name="l00289"></a>00289 <span class="comment">///////////////////////////////////////////////////////////////////////////////////////////////////</span>
<a name="l00290"></a>00290 <span class="comment"></span>RT_Test* qt_batch()
<a name="l00291"></a>00291 {
<a name="l00292"></a>00292   RT_Test* rt;
<a name="l00293"></a>00293   rt = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;batch&quot;</span>);
<a name="l00294"></a>00294   RT_Test* rt1;
<a name="l00295"></a>00295   rt1 = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;batch-part1&quot;</span>);
<a name="l00296"></a>00296   RT_Test* rt2;
<a name="l00297"></a>00297   rt2 = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;batch-part2&quot;</span>);
<a name="l00298"></a>00298 
<a name="l00299"></a>00299 
<a name="l00300"></a>00300   RT_Cmd* cmd1 = rt1-&gt;newCmd();
<a name="l00301"></a>00301   cmd1-&gt;setExe(<span class="stringliteral">&quot;${apt_copynumber_cyto}&quot;</span>);
<a name="l00302"></a>00302   cmd1-&gt;addArg(<span class="stringliteral">&quot;-v&quot;</span>, <span class="stringliteral">&quot;3&quot;</span>);
<a name="l00303"></a>00303   cmd1-&gt;addArg(<span class="stringliteral">&quot;--cyto2&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00304"></a>00304   cmd1-&gt;addArg(<span class="stringliteral">&quot;--force&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00305"></a>00305   cmd1-&gt;addArg(<span class="stringliteral">&quot;--text-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00306"></a>00306   cmd1-&gt;addArg(<span class="stringliteral">&quot;--cnchp-output&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00307"></a>00307   cmd1-&gt;addArg(<span class="stringliteral">&quot;--cychp-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00308"></a>00308   cmd1-&gt;addArg(<span class="stringliteral">&quot;--set-analysis-name&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>);
<a name="l00309"></a>00309   cmd1-&gt;addArg(<span class="stringliteral">&quot;--reference-output&quot;</span>, <span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00310"></a>00310   cmd1-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-input-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.na28.snpref.a5&quot;</span>);
<a name="l00311"></a>00311   cmd1-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-output-file&quot;</span>, <span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00312"></a>00312   cmd1-&gt;addArg(<span class="stringliteral">&quot;--spf-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.spf&quot;</span>);
<a name="l00313"></a>00313   cmd1-&gt;addArg(<span class="stringliteral">&quot;--chrX-probes&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.chrXprobes&quot;</span>);
<a name="l00314"></a>00314   cmd1-&gt;addArg(<span class="stringliteral">&quot;--chrY-probes&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.chrYprobes&quot;</span>);
<a name="l00315"></a>00315   cmd1-&gt;addArg(<span class="stringliteral">&quot;--annotation-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.na28.annot.db&quot;</span>);
<a name="l00316"></a>00316   cmd1-&gt;addArg(<span class="stringliteral">&quot;--probe-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.probe_tab&quot;</span>);
<a name="l00317"></a>00317   cmd1-&gt;addArg(<span class="stringliteral">&quot;--snp-qc-snp-list&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.snplist.txt&quot;</span>);
<a name="l00318"></a>00318   cmd1-&gt;addArg(<span class="stringliteral">&quot;--gender-override-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/HapMap-LOHReference-Full-Genders.txt&quot;</span>);
<a name="l00319"></a>00319   cmd1-&gt;addArg(<span class="stringliteral">&quot;--genotype-call-override-file&quot;</span>, <span class="stringliteral">&quot;${gold_idata_lib_cytogeneticsarray}/HapMapCalls_LOHReferenceSet-FullArray.txt&quot;</span>);
<a name="l00320"></a>00320   cmd1-&gt;addArg(<span class="stringliteral">&quot;--cel-files&quot;</span>, <span class="stringliteral">&quot;${gold_idata_cel_cytogeneticsarray}/CELFileList_5F5M.txt&quot;</span>);
<a name="l00321"></a>00321   cmd1-&gt;addArg(<span class="stringliteral">&quot;--out-dir&quot;</span>, <span class="stringliteral">&quot;${out_batch}&quot;</span>);
<a name="l00322"></a>00322   cmd1-&gt;addArg(<span class="stringliteral">&quot;--male-gender-ratio-cutoff=1.3&quot;</span>);
<a name="l00323"></a>00323   cmd1-&gt;addArg(<span class="stringliteral">&quot;--female-gender-ratio-cutoff=1.0&quot;</span>);
<a name="l00324"></a>00324   cmd1-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff=0.8&quot;</span>);
<a name="l00325"></a>00325   cmd1-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff-high=1.07&quot;</span>);
<a name="l00326"></a>00326   cmd1-&gt;addArg(<span class="stringliteral">&quot;--y-cutoff=0.65&quot;</span>);
<a name="l00327"></a>00327   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span>);
<a name="l00328"></a>00328   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;kernel-smooth.sigma_span=50&quot;</span>);
<a name="l00329"></a>00329   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span>);
<a name="l00330"></a>00330   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2-gender.cutoff=0.5&quot;</span>);
<a name="l00331"></a>00331   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-segment&quot;</span>);
<a name="l00332"></a>00332   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span>);
<a name="l00333"></a>00333   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span>);
<a name="l00334"></a>00334   cmd1-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span>);
<a name="l00335"></a>00335   cmd1-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;pdnn-reference-method&quot;</span>);
<a name="l00336"></a>00336   cmd1-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span>);
<a name="l00337"></a>00337   cmd1-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span>);
<a name="l00338"></a>00338   cmd1-&gt;addArg(<span class="stringliteral">&quot;--image-correction-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span>);
<a name="l00339"></a>00339   cmd1-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-log2ratio-adjustment-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;</span>);
<a name="l00340"></a>00340   cmd1-&gt;addArg(<span class="stringliteral">&quot;--reference-chromosome=22&quot;</span>);
<a name="l00341"></a>00341 
<a name="l00342"></a>00342   std::vector&lt;std::string&gt; gen;
<a name="l00343"></a>00343   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00344"></a>00344   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00345"></a>00345   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00346"></a>00346   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00347"></a>00347   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00348"></a>00348   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00349"></a>00349   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00350"></a>00350   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00351"></a>00351   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00352"></a>00352   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00353"></a>00353 
<a name="l00354"></a>00354   std::vector&lt;std::string&gt; gold;
<a name="l00355"></a>00355   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00356"></a>00356   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00357"></a>00357   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00358"></a>00358   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00359"></a>00359   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00360"></a>00360   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00361"></a>00361   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00362"></a>00362   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00363"></a>00363   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00364"></a>00364   gold.push_back(<span class="stringliteral">&quot;${gold_idata_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00365"></a>00365 
<a name="l00366"></a>00366   RT_Check* check1 = rt1-&gt;newCheck();
<a name="l00367"></a>00367   check1-&gt;setExe(<span class="stringliteral">&quot;${apt_file5_equivalent}&quot;</span>);
<a name="l00368"></a>00368   check1-&gt;setGenFiles(gen);
<a name="l00369"></a>00369   check1-&gt;setGoldFiles(gold);
<a name="l00370"></a>00370   check1-&gt;setGenGoldFileIsArg(<span class="keyword">true</span>);
<a name="l00371"></a>00371   check1-&gt;addArg(<span class="stringliteral">&quot;--ignore-columns-file&quot;</span>, <span class="stringliteral">&quot;./ignoreColumnFile.tsv&quot;</span>);
<a name="l00372"></a>00372   check1-&gt;addArg(<span class="stringliteral">&quot;--datasets-file&quot;</span>, <span class="stringliteral">&quot;./datasetFile.tsv&quot;</span>);
<a name="l00373"></a>00373   check1-&gt;addArg(<span class="stringliteral">&quot;--epsilon&quot;</span>, <span class="stringliteral">&quot;0.0001&quot;</span>);
<a name="l00374"></a>00374   check1-&gt;addArg(<span class="stringliteral">&quot;--correlation&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>);
<a name="l00375"></a>00375   <span class="comment">/*</span>
<a name="l00376"></a>00376 <span class="comment">  RT_Cmd* cmd2 = rt2-&gt;newCmd();</span>
<a name="l00377"></a>00377 <span class="comment">  cmd2-&gt;setExe(&quot;${apt_copynumber_cyto}&quot;);</span>
<a name="l00378"></a>00378 <span class="comment">  cmd2-&gt;addArg(&quot;-v&quot;, &quot;3&quot;);</span>
<a name="l00379"></a>00379 <span class="comment">  cmd2-&gt;addArg(&quot;--cyto2&quot;, &quot;true&quot;);</span>
<a name="l00380"></a>00380 <span class="comment">  cmd2-&gt;addArg(&quot;--force&quot;, &quot;true&quot;);</span>
<a name="l00381"></a>00381 <span class="comment">  cmd2-&gt;addArg(&quot;--text-output&quot;, &quot;true&quot;);</span>
<a name="l00382"></a>00382 <span class="comment">  cmd2-&gt;addArg(&quot;--cnchp-output&quot;, &quot;false&quot;);</span>
<a name="l00383"></a>00383 <span class="comment">  cmd2-&gt;addArg(&quot;--cychp-output&quot;, &quot;true&quot;);</span>
<a name="l00384"></a>00384 <span class="comment">  cmd2-&gt;addArg(&quot;--set-analysis-name&quot;, &quot;ca&quot;);</span>
<a name="l00385"></a>00385 <span class="comment">  cmd2-&gt;addArg(&quot;--reference-input&quot;, &quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;);</span>
<a name="l00386"></a>00386 <span class="comment">  cmd2-&gt;addArg(&quot;--snp-reference-input-file&quot;, &quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;);</span>
<a name="l00387"></a>00387 <span class="comment">  cmd2-&gt;addArg(&quot;--spf-file&quot;, &quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.spf&quot;);</span>
<a name="l00388"></a>00388 <span class="comment">  cmd2-&gt;addArg(&quot;--chrX-probes&quot;, &quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.chrXprobes&quot;);</span>
<a name="l00389"></a>00389 <span class="comment">  cmd2-&gt;addArg(&quot;--chrY-probes&quot;, &quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.chrYprobes&quot;);</span>
<a name="l00390"></a>00390 <span class="comment">  cmd2-&gt;addArg(&quot;--annotation-file&quot;, &quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.na28.annot.db&quot;);</span>
<a name="l00391"></a>00391 <span class="comment">  cmd2-&gt;addArg(&quot;--probe-file&quot;, &quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.probe_tab&quot;);</span>
<a name="l00392"></a>00392 <span class="comment">  cmd2-&gt;addArg(&quot;--snp-qc-snp-list&quot;, &quot;${gold_idata_lib_cytogeneticsarray}/Cytogenetics_Array.snplist.txt&quot;);</span>
<a name="l00393"></a>00393 <span class="comment">  cmd2-&gt;addArg(&quot;--cels&quot;, &quot;${gold_idata_cel_cytogeneticsarray}/HapMap-As_NA18547_A08_01_NN_20081218.CEL&quot;);</span>
<a name="l00394"></a>00394 <span class="comment">  cmd2-&gt;addArg(&quot;--cels&quot;, &quot;${gold_idata_cel_cytogeneticsarray}/HapMap-As_NA18603_B03_01_NN_20081218.CEL&quot;);</span>
<a name="l00395"></a>00395 <span class="comment">  cmd2-&gt;addArg(&quot;--cels&quot;, &quot;${gold_idata_cel_cytogeneticsarray}/HapMap-As_NA18971_C02_01_NN_20081218.CEL&quot;);</span>
<a name="l00396"></a>00396 <span class="comment">  cmd2-&gt;addArg(&quot;--cels&quot;, &quot;${gold_idata_cel_cytogeneticsarray}/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL&quot;);</span>
<a name="l00397"></a>00397 <span class="comment">  cmd2-&gt;addArg(&quot;--cels&quot;, &quot;${gold_idata_cel_cytogeneticsarray}/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL&quot;);</span>
<a name="l00398"></a>00398 <span class="comment">  cmd2-&gt;addArg(&quot;--out-dir&quot;, &quot;${out_batch}&quot;);</span>
<a name="l00399"></a>00399 <span class="comment">  cmd2-&gt;addArg(&quot;--male-gender-ratio-cutoff=1.3&quot;);</span>
<a name="l00400"></a>00400 <span class="comment">  cmd2-&gt;addArg(&quot;--female-gender-ratio-cutoff=1.0&quot;);</span>
<a name="l00401"></a>00401 <span class="comment">  cmd2-&gt;addArg(&quot;--xx-cutoff=0.8&quot;);</span>
<a name="l00402"></a>00402 <span class="comment">  cmd2-&gt;addArg(&quot;--xx-cutoff-high=1.07&quot;);</span>
<a name="l00403"></a>00403 <span class="comment">  cmd2-&gt;addArg(&quot;--y-cutoff=0.65&quot;);</span>
<a name="l00404"></a>00404 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;);</span>
<a name="l00405"></a>00405 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;kernel-smooth.sigma_span=50&quot;);</span>
<a name="l00406"></a>00406 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;);</span>
<a name="l00407"></a>00407 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;cn-cyto2-gender.cutoff=0.5&quot;);</span>
<a name="l00408"></a>00408 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;cn-segment&quot;);</span>
<a name="l00409"></a>00409 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;);</span>
<a name="l00410"></a>00410 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;mosaicism.marker-bandwidth=6000.confidence-window=251&quot;);</span>
<a name="l00411"></a>00411 <span class="comment">  cmd2-&gt;addArg(&quot;--analysis&quot;, &quot;allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;);</span>
<a name="l00412"></a>00412 <span class="comment">  cmd2-&gt;addArg(&quot;--local-gc-background-correction-reference-method&quot;, &quot;pdnn-reference-method&quot;);</span>
<a name="l00413"></a>00413 <span class="comment">  cmd2-&gt;addArg(&quot;--wave-correction-reference-method&quot;, &quot;wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;);</span>
<a name="l00414"></a>00414 <span class="comment">  cmd2-&gt;addArg(&quot;--local-gc-background-intensity-adjustment-method&quot;, &quot;pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;);</span>
<a name="l00415"></a>00415 <span class="comment">  cmd2-&gt;addArg(&quot;--image-correction-intensity-adjustment-method&quot;, &quot;high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;);</span>
<a name="l00416"></a>00416 <span class="comment">  cmd2-&gt;addArg(&quot;--wave-correction-log2ratio-adjustment-method&quot;, &quot;wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;);</span>
<a name="l00417"></a>00417 <span class="comment">  cmd2-&gt;addArg(&quot;--reference-chromosome=22&quot;);</span>
<a name="l00418"></a>00418 <span class="comment"></span>
<a name="l00419"></a>00419 <span class="comment">  std::vector&lt;std::string&gt; gen2,gold2;</span>
<a name="l00420"></a>00420 <span class="comment">  gen2 = Util::addPrefixSuffix(qt_cytogenetics_array_sixfiles, &quot;${out_batch}/&quot;, &quot;.ca.cychp&quot;);</span>
<a name="l00421"></a>00421 <span class="comment">  gold2 = Util::addPrefixSuffix(qt_cytogenetics_array_sixfiles, &quot;${gold_idata_copynumbercyto_batch}/&quot;, &quot;.ca.cychp&quot;);</span>
<a name="l00422"></a>00422 <span class="comment"></span>
<a name="l00423"></a>00423 <span class="comment">  RT_Check* check2 = rt2-&gt;newCheck();</span>
<a name="l00424"></a>00424 <span class="comment">  check2-&gt;setExe(&quot;${apt_calvin_equivalent}&quot;);</span>
<a name="l00425"></a>00425 <span class="comment">  check2-&gt;setGenFiles(gen2);</span>
<a name="l00426"></a>00426 <span class="comment">  check2-&gt;setGoldFiles(gold2);</span>
<a name="l00427"></a>00427 <span class="comment">  check2-&gt;setGenGoldFileIsArg(true);</span>
<a name="l00428"></a>00428 <span class="comment">  check2-&gt;addArg(&quot;--epsilon-map-file&quot;, &quot;./qt_mapEpsilonFile.tsv&quot;);</span>
<a name="l00429"></a>00429 <span class="comment">  check2-&gt;addArg(&quot;--ignore-params-file&quot;, &quot;./ignoreParamFile.tsv&quot;);</span>
<a name="l00430"></a>00430 <span class="comment">  check2-&gt;addArg(&quot;--epsilon&quot;, &quot;0.0001&quot;);</span>
<a name="l00431"></a>00431 <span class="comment">  check2-&gt;addArg(&quot;--correlation&quot;, &quot;0.9999989&quot;);</span>
<a name="l00432"></a>00432 <span class="comment">  check2-&gt;addArg(&quot;--check-header&quot;, &quot;false&quot;);</span>
<a name="l00433"></a>00433 <span class="comment">  */</span>
<a name="l00434"></a>00434   rt-&gt;addSubtest(rt1);
<a name="l00435"></a>00435   <span class="comment">//rt-&gt;addSubtest(rt2);</span>
<a name="l00436"></a>00436   <span class="keywordflow">return</span> rt;
<a name="l00437"></a>00437 }
<a name="l00438"></a>00438 <span class="comment"></span>
<a name="l00439"></a>00439 <span class="comment">///////////////////////////////////////////////////////////////////////////////////////////////////</span>
<a name="l00440"></a>00440 <span class="comment"></span>RT_Test* batch()
<a name="l00441"></a>00441 {
<a name="l00442"></a>00442   RT_Test* rt;
<a name="l00443"></a>00443   rt = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;batch&quot;</span>);
<a name="l00444"></a>00444   RT_Test* rt1;
<a name="l00445"></a>00445   rt1 = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;batch-part1&quot;</span>);
<a name="l00446"></a>00446   RT_Test* rt2;
<a name="l00447"></a>00447   rt2 = <span class="keyword">new</span> RT_Test(<span class="stringliteral">&quot;batch-part2&quot;</span>);
<a name="l00448"></a>00448 
<a name="l00449"></a>00449   RT_Cmd* cmd = rt1-&gt;newCmd();
<a name="l00450"></a>00450   cmd-&gt;setExe(<span class="stringliteral">&quot;${apt_copynumber_cyto}&quot;</span>);
<a name="l00451"></a>00451   cmd-&gt;addArg(<span class="stringliteral">&quot;-v&quot;</span>, <span class="stringliteral">&quot;3&quot;</span>);
<a name="l00452"></a>00452   cmd-&gt;addArg(<span class="stringliteral">&quot;--cyto2&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00453"></a>00453   cmd-&gt;addArg(<span class="stringliteral">&quot;--force&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00454"></a>00454   cmd-&gt;addArg(<span class="stringliteral">&quot;--text-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00455"></a>00455   cmd-&gt;addArg(<span class="stringliteral">&quot;--cnchp-output&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00456"></a>00456   cmd-&gt;addArg(<span class="stringliteral">&quot;--cychp-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00457"></a>00457   cmd-&gt;addArg(<span class="stringliteral">&quot;--set-analysis-name&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>);
<a name="l00458"></a>00458   cmd-&gt;addArg(<span class="stringliteral">&quot;--reference-output&quot;</span>, <span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00459"></a>00459   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-input-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.na28.snpref.a5&quot;</span>);
<a name="l00460"></a>00460   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-output-file&quot;</span>, <span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00461"></a>00461   cmd-&gt;addArg(<span class="stringliteral">&quot;--cdf-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.CDF&quot;</span>);
<a name="l00462"></a>00462   cmd-&gt;addArg(<span class="stringliteral">&quot;--chrX-probes&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.chrXprobes&quot;</span>);
<a name="l00463"></a>00463   cmd-&gt;addArg(<span class="stringliteral">&quot;--chrY-probes&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.chrYprobes&quot;</span>);
<a name="l00464"></a>00464   cmd-&gt;addArg(<span class="stringliteral">&quot;--annotation-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.na28.annot.db&quot;</span>);
<a name="l00465"></a>00465   cmd-&gt;addArg(<span class="stringliteral">&quot;--probe-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.probe_tab&quot;</span>);
<a name="l00466"></a>00466   cmd-&gt;addArg(<span class="stringliteral">&quot;--snp-qc-snp-list&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.snplist.txt&quot;</span>);
<a name="l00467"></a>00467   cmd-&gt;addArg(<span class="stringliteral">&quot;--gender-override-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/HapMap-LOHReference-Full-Genders.txt&quot;</span>);
<a name="l00468"></a>00468   cmd-&gt;addArg(<span class="stringliteral">&quot;--genotype-call-override-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/HapMapCalls_LOHReferenceSet-FullArray.txt&quot;</span>);
<a name="l00469"></a>00469   cmd-&gt;addArg(<span class="stringliteral">&quot;--cel-files&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/CELFileList_5F5M.txt&quot;</span>);
<a name="l00470"></a>00470   cmd-&gt;addArg(<span class="stringliteral">&quot;--out-dir&quot;</span>, <span class="stringliteral">&quot;${out_batch}&quot;</span>);
<a name="l00471"></a>00471   cmd-&gt;addArg(<span class="stringliteral">&quot;--male-gender-ratio-cutoff=1.3&quot;</span>);
<a name="l00472"></a>00472   cmd-&gt;addArg(<span class="stringliteral">&quot;--female-gender-ratio-cutoff=1.0&quot;</span>);
<a name="l00473"></a>00473   cmd-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff=0.8&quot;</span>);
<a name="l00474"></a>00474   cmd-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff-high=1.07&quot;</span>);
<a name="l00475"></a>00475   cmd-&gt;addArg(<span class="stringliteral">&quot;--y-cutoff=0.65&quot;</span>);
<a name="l00476"></a>00476   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span>);
<a name="l00477"></a>00477   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;kernel-smooth.sigma_span=50&quot;</span>);
<a name="l00478"></a>00478   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span>);
<a name="l00479"></a>00479   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2-gender.cutoff=0.5&quot;</span>);
<a name="l00480"></a>00480   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-segment&quot;</span>);
<a name="l00481"></a>00481   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span>);
<a name="l00482"></a>00482   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span>);
<a name="l00483"></a>00483   cmd-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span>);
<a name="l00484"></a>00484   cmd-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;pdnn-reference-method&quot;</span>);
<a name="l00485"></a>00485   cmd-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span>);
<a name="l00486"></a>00486   cmd-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span>);
<a name="l00487"></a>00487   cmd-&gt;addArg(<span class="stringliteral">&quot;--image-correction-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span>);
<a name="l00488"></a>00488   cmd-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-log2ratio-adjustment-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;</span>);
<a name="l00489"></a>00489 
<a name="l00490"></a>00490   std::vector&lt;std::string&gt; gen;
<a name="l00491"></a>00491   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00492"></a>00492   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00493"></a>00493   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00494"></a>00494   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00495"></a>00495   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00496"></a>00496   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00497"></a>00497   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00498"></a>00498   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00499"></a>00499   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00500"></a>00500   gen.push_back(<span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00501"></a>00501 
<a name="l00502"></a>00502   std::vector&lt;std::string&gt; gold;
<a name="l00503"></a>00503   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00504"></a>00504   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00505"></a>00505   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00506"></a>00506   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00507"></a>00507   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00508"></a>00508   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00509"></a>00509   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00510"></a>00510   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00511"></a>00511   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00512"></a>00512   gold.push_back(<span class="stringliteral">&quot;${gold_copynumbercyto_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00513"></a>00513 
<a name="l00514"></a>00514   RT_Check* check1 = rt1-&gt;newCheck();
<a name="l00515"></a>00515   check1-&gt;setExe(<span class="stringliteral">&quot;${apt_file5_equivalent}&quot;</span>);
<a name="l00516"></a>00516   check1-&gt;setGenFiles(gen);
<a name="l00517"></a>00517   check1-&gt;setGoldFiles(gold);
<a name="l00518"></a>00518   check1-&gt;setGenGoldFileIsArg(<span class="keyword">true</span>);
<a name="l00519"></a>00519   check1-&gt;addArg(<span class="stringliteral">&quot;--ignore-columns-file&quot;</span>, <span class="stringliteral">&quot;./ignoreColumnFile.tsv&quot;</span>);
<a name="l00520"></a>00520   check1-&gt;addArg(<span class="stringliteral">&quot;--datasets-file&quot;</span>, <span class="stringliteral">&quot;./datasetFile.tsv&quot;</span>);
<a name="l00521"></a>00521   check1-&gt;addArg(<span class="stringliteral">&quot;--epsilon&quot;</span>, <span class="stringliteral">&quot;0.0001&quot;</span>);
<a name="l00522"></a>00522   check1-&gt;addArg(<span class="stringliteral">&quot;--correlation&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>);
<a name="l00523"></a>00523 
<a name="l00524"></a>00524   RT_Cmd* cmd2 = rt2-&gt;newCmd();
<a name="l00525"></a>00525   cmd2-&gt;setExe(<span class="stringliteral">&quot;${apt_copynumber_cyto}&quot;</span>);
<a name="l00526"></a>00526   cmd2-&gt;addArg(<span class="stringliteral">&quot;-v&quot;</span>, <span class="stringliteral">&quot;3&quot;</span>);
<a name="l00527"></a>00527   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cyto2&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00528"></a>00528   cmd2-&gt;addArg(<span class="stringliteral">&quot;--force&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00529"></a>00529   cmd2-&gt;addArg(<span class="stringliteral">&quot;--text-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00530"></a>00530   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cnchp-output&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00531"></a>00531   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cychp-output&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>);
<a name="l00532"></a>00532   cmd2-&gt;addArg(<span class="stringliteral">&quot;--set-analysis-name&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>);
<a name="l00533"></a>00533   cmd2-&gt;addArg(<span class="stringliteral">&quot;--reference-input&quot;</span>, <span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>);
<a name="l00534"></a>00534   cmd2-&gt;addArg(<span class="stringliteral">&quot;--snp-reference-input-file&quot;</span>, <span class="stringliteral">&quot;${out_batch}/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>);
<a name="l00535"></a>00535   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cdf-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.CDF&quot;</span>);
<a name="l00536"></a>00536   cmd2-&gt;addArg(<span class="stringliteral">&quot;--chrX-probes&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.chrXprobes&quot;</span>);
<a name="l00537"></a>00537   cmd2-&gt;addArg(<span class="stringliteral">&quot;--chrY-probes&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.chrYprobes&quot;</span>);
<a name="l00538"></a>00538   cmd2-&gt;addArg(<span class="stringliteral">&quot;--annotation-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.na28.annot.db&quot;</span>);
<a name="l00539"></a>00539   cmd2-&gt;addArg(<span class="stringliteral">&quot;--probe-file&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.probe_tab&quot;</span>);
<a name="l00540"></a>00540   cmd2-&gt;addArg(<span class="stringliteral">&quot;--snp-qc-snp-list&quot;</span>, <span class="stringliteral">&quot;${gold_lib_cytogeneticsarray}/Cytogenetics_Array.snplist.txt&quot;</span>);
<a name="l00541"></a>00541   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18547_A08_01_NN_20081218.CEL&quot;</span>);
<a name="l00542"></a>00542   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18603_B03_01_NN_20081218.CEL&quot;</span>);
<a name="l00543"></a>00543   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18943_C11_01_NN_20081218.CEL&quot;</span>);
<a name="l00544"></a>00544   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-As_NA18971_C02_01_NN_20081218.CEL&quot;</span>);
<a name="l00545"></a>00545   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL&quot;</span>);
<a name="l00546"></a>00546   cmd2-&gt;addArg(<span class="stringliteral">&quot;--cels&quot;</span>, <span class="stringliteral">&quot;${gold_cel_cytogeneticsarray}/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL&quot;</span>);
<a name="l00547"></a>00547   cmd2-&gt;addArg(<span class="stringliteral">&quot;--out-dir&quot;</span>, <span class="stringliteral">&quot;${out_batch}&quot;</span>);
<a name="l00548"></a>00548   cmd2-&gt;addArg(<span class="stringliteral">&quot;--male-gender-ratio-cutoff=1.3&quot;</span>);
<a name="l00549"></a>00549   cmd2-&gt;addArg(<span class="stringliteral">&quot;--female-gender-ratio-cutoff=1.0&quot;</span>);
<a name="l00550"></a>00550   cmd2-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff=0.8&quot;</span>);
<a name="l00551"></a>00551   cmd2-&gt;addArg(<span class="stringliteral">&quot;--xx-cutoff-high=1.07&quot;</span>);
<a name="l00552"></a>00552   cmd2-&gt;addArg(<span class="stringliteral">&quot;--y-cutoff=0.65&quot;</span>);
<a name="l00553"></a>00553   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span>);
<a name="l00554"></a>00554   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;kernel-smooth.sigma_span=50&quot;</span>);
<a name="l00555"></a>00555   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span>);
<a name="l00556"></a>00556   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-cyto2-gender.cutoff=0.5&quot;</span>);
<a name="l00557"></a>00557   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;cn-segment&quot;</span>);
<a name="l00558"></a>00558   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span>);
<a name="l00559"></a>00559   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span>);
<a name="l00560"></a>00560   cmd2-&gt;addArg(<span class="stringliteral">&quot;--analysis&quot;</span>, <span class="stringliteral">&quot;allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span>);
<a name="l00561"></a>00561   cmd2-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;pdnn-reference-method&quot;</span>);
<a name="l00562"></a>00562   cmd2-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-reference-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span>);
<a name="l00563"></a>00563   cmd2-&gt;addArg(<span class="stringliteral">&quot;--local-gc-background-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span>);
<a name="l00564"></a>00564   cmd2-&gt;addArg(<span class="stringliteral">&quot;--image-correction-intensity-adjustment-method&quot;</span>, <span class="stringliteral">&quot;high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span>);
<a name="l00565"></a>00565   cmd2-&gt;addArg(<span class="stringliteral">&quot;--wave-correction-log2ratio-adjustment-method&quot;</span>, <span class="stringliteral">&quot;wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;</span>);
<a name="l00566"></a>00566 
<a name="l00567"></a>00567   std::vector&lt;std::string&gt; gen2,gold2;
<a name="l00568"></a>00568   gen2 = Util::addPrefixSuffix(cytogenetics_array_sixfiles, <span class="stringliteral">&quot;${out_batch}/&quot;</span>, <span class="stringliteral">&quot;.ca.cychp&quot;</span>);
<a name="l00569"></a>00569   gold2 = Util::addPrefixSuffix(cytogenetics_array_sixfiles, <span class="stringliteral">&quot;${gold_copynumbercyto_batch}/&quot;</span>, <span class="stringliteral">&quot;.ca.cychp&quot;</span>);
<a name="l00570"></a>00570 
<a name="l00571"></a>00571   RT_Check* check2 = rt2-&gt;newCheck();
<a name="l00572"></a>00572   check2-&gt;setExe(<span class="stringliteral">&quot;${apt_calvin_equivalent}&quot;</span>);
<a name="l00573"></a>00573   check2-&gt;setGenFiles(gen2);
<a name="l00574"></a>00574   check2-&gt;setGoldFiles(gold2);
<a name="l00575"></a>00575   check2-&gt;setGenGoldFileIsArg(<span class="keyword">true</span>);
<a name="l00576"></a>00576   check2-&gt;addArg(<span class="stringliteral">&quot;--ignore-params-file&quot;</span>, <span class="stringliteral">&quot;./ignoreParamFile.tsv&quot;</span>);
<a name="l00577"></a>00577   check2-&gt;addArg(<span class="stringliteral">&quot;--epsilon&quot;</span>, <span class="stringliteral">&quot;0.0001&quot;</span>);
<a name="l00578"></a>00578   check2-&gt;addArg(<span class="stringliteral">&quot;--correlation&quot;</span>, <span class="stringliteral">&quot;0.9999989&quot;</span>);
<a name="l00579"></a>00579   check2-&gt;addArg(<span class="stringliteral">&quot;--check-header&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00580"></a>00580 
<a name="l00581"></a>00581   rt-&gt;addSubtest(rt1);
<a name="l00582"></a>00582   rt-&gt;addSubtest(rt2);
<a name="l00583"></a>00583   <span class="keywordflow">return</span> rt;
<a name="l00584"></a>00584 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:54 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
