<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: CNLog2RatioData Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
      <li><a href="classes.html"><span>Class&nbsp;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a>  </div>
  <div class="headertitle">
<h1>CNLog2RatioData Class Reference</h1>  </div>
</div>
<div class="contents">
<!-- doxytag: class="CNLog2RatioData" -->
<p>A class to the log2 ratio data as used by the log2 ratio engine.  
<a href="#_details">More...</a></p>

<p><code>#include &lt;<a class="el" href="CNLog2RatioData_8h_source.html">CNLog2RatioData.h</a>&gt;</code></p>

<p><a href="classCNLog2RatioData-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3402cdc0a75ab758729d7a25b8f22359"></a><!-- doxytag: member="CNLog2RatioData::CNLog2RatioData" ref="a3402cdc0a75ab758729d7a25b8f22359" args="()" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a3402cdc0a75ab758729d7a25b8f22359">CNLog2RatioData</a> ()</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Constructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af207bb7cf8614bb1b33ac285334b8604"></a><!-- doxytag: member="CNLog2RatioData::~CNLog2RatioData" ref="af207bb7cf8614bb1b33ac285334b8604" args="()" -->
virtual&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#af207bb7cf8614bb1b33ac285334b8604">~CNLog2RatioData</a> ()</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Destructor. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a35afdbee8e95056161afd6bbc28d9baf"></a><!-- doxytag: member="CNLog2RatioData::clear" ref="a35afdbee8e95056161afd6bbc28d9baf" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a35afdbee8e95056161afd6bbc28d9baf">clear</a> ()</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Release the memory associated with the object. <br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2209f32a0196322602290d78aef97444"></a><!-- doxytag: member="CNLog2RatioData::clearProbeSets" ref="a2209f32a0196322602290d78aef97444" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>clearProbeSets</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad120feb1c150aa17eb3009e7c9e87011"></a><!-- doxytag: member="CNLog2RatioData::setLog2RatioEngine" ref="ad120feb1c150aa17eb3009e7c9e87011" args="(bool b)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>setLog2RatioEngine</b> (bool b)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8e38e2ff938caad33c7cc074ae079a77"></a><!-- doxytag: member="CNLog2RatioData::setEngine" ref="a8e38e2ff938caad33c7cc074ae079a77" args="(BaseEngine *pEngine)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>setEngine</b> (<a class="el" href="classBaseEngine.html">BaseEngine</a> *pEngine)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ace53fd2b52667b5e395a3510c9702729"></a><!-- doxytag: member="CNLog2RatioData::getExperiments" ref="ace53fd2b52667b5e395a3510c9702729" args="()" -->
<a class="el" href="classCNExperimentArray.html">CNExperimentArray</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getExperiments</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad13a5efbd168ca8d28e0d8f0d33592d5"></a><!-- doxytag: member="CNLog2RatioData::setExperiments" ref="ad13a5efbd168ca8d28e0d8f0d33592d5" args="(CNExperimentArray &amp;v)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>setExperiments</b> (<a class="el" href="classCNExperimentArray.html">CNExperimentArray</a> &amp;v)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac664a76fa71d45423ccf3fdba4606e65"></a><!-- doxytag: member="CNLog2RatioData::getProbeSets" ref="ac664a76fa71d45423ccf3fdba4606e65" args="()" -->
<a class="el" href="classCNProbeSetArray.html">CNProbeSetArray</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getProbeSets</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a25c9ae328ef07e03d63a35c99f66b6cd"></a><!-- doxytag: member="CNLog2RatioData::getProbeSetsAlternateSort" ref="a25c9ae328ef07e03d63a35c99f66b6cd" args="()" -->
<a class="el" href="classCNProbeSetArray.html">CNProbeSetArray</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getProbeSetsAlternateSort</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5f44fe65aca25d46efcda56fc080c4e4"></a><!-- doxytag: member="CNLog2RatioData::getAAlleleEstimates" ref="a5f44fe65aca25d46efcda56fc080c4e4" args="()" -->
<a class="el" href="classAffxMultiDimensionalArray.html">AffxMultiDimensionalArray</a><br class="typebreak"/>
&lt; float &gt; *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getAAlleleEstimates</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a24be512cf34977e599ff3b65011688a8"></a><!-- doxytag: member="CNLog2RatioData::getBAlleleEstimates" ref="a24be512cf34977e599ff3b65011688a8" args="()" -->
<a class="el" href="classAffxMultiDimensionalArray.html">AffxMultiDimensionalArray</a><br class="typebreak"/>
&lt; float &gt; *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getBAlleleEstimates</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaf7cb9cc3e5fd2d29d6e073b8b75c783"></a><!-- doxytag: member="CNLog2RatioData::getGenotypeCalls" ref="aaf7cb9cc3e5fd2d29d6e073b8b75c783" args="()" -->
<a class="el" href="classAffxMultiDimensionalArray.html">AffxMultiDimensionalArray</a>&lt; char &gt; *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getGenotypeCalls</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aee42b922d1475842780ba505d5cb4359"></a><!-- doxytag: member="CNLog2RatioData::getGenotypeConfidences" ref="aee42b922d1475842780ba505d5cb4359" args="()" -->
<a class="el" href="classAffxMultiDimensionalArray.html">AffxMultiDimensionalArray</a><br class="typebreak"/>
&lt; float &gt; *&nbsp;</td><td class="memItemRight" valign="bottom"><b>getGenotypeConfidences</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2be52968341f07239fc0fa267d0d8b2e"></a><!-- doxytag: member="CNLog2RatioData::getV1" ref="a2be52968341f07239fc0fa267d0d8b2e" args="()" -->
<a class="el" href="classAffxMultiDimensionalArray.html">AffxMultiDimensionalArray</a><br class="typebreak"/>
&lt; float &gt; &amp;&nbsp;</td><td class="memItemRight" valign="bottom"><b>getV1</b> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a817654e1d9b7e55f3c6c75faafa9c889"></a><!-- doxytag: member="CNLog2RatioData::loadAnnotation" ref="a817654e1d9b7e55f3c6c75faafa9c889" args="(bool bNewProbeSets)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>loadAnnotation</b> (bool bNewProbeSets)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#ac2d7caba60586cf34f587b645632eefb">loadModelFile</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the model file and set up the <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> Set array with the pre-calculated median values from the log2 ratio model.  <a href="#ac2d7caba60586cf34f587b645632eefb"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#ac81ed0f535099554a833785fbe32564d">writeModelFile</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Write the model file in text format.  <a href="#ac81ed0f535099554a833785fbe32564d"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a98563435a322545374c99293c7ff7c82">loadCyto2ModelFile</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the model file and set up the <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> Set array with the pre-calculated median values from the log2 ratio model.  <a href="#a98563435a322545374c99293c7ff7c82"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a336f2d8a1870247b2be8e0579b303b4c">writeCyto2ModelFile</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Write the model file in text format.  <a href="#a336f2d8a1870247b2be8e0579b303b4c"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a235e31fd7f73d9b75bb00257c85df4c6">loadSnpReferenceFile</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the SNP reference file and set up the <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> Set array with the pre-calculated median values from the log2 ratio model.  <a href="#a235e31fd7f73d9b75bb00257c85df4c6"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a9c95fc886b02b3a4bcdbb3a48d2cd867">loadGenotypeReport</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the QC metric from the genotype report file.  <a href="#a9c95fc886b02b3a4bcdbb3a48d2cd867"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#af82738d0c4c7943f08da2cb7ad038742">dumpGenotypeReport</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Dump the QC metrics into the copynumber report file.  <a href="#af82738d0c4c7943f08da2cb7ad038742"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#aa2ac281f90626048a718c0deccba50db">dumpCyto2Report</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Dump the QC metrics into the copynumber report file.  <a href="#aa2ac281f90626048a718c0deccba50db"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a80546473717d00e4a9d5a4e69b871485">loadExperiments</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the allele summary table file and set up the experiment array with experiment name.  <a href="#a80546473717d00e4a9d5a4e69b871485"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a781c14ec7d3a8d045c4ec139f73440a0">loadQCMetrics</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the QC metrics from the report file.  <a href="#a781c14ec7d3a8d045c4ec139f73440a0"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a1702c9c9e6478a46bacbc37bb1b35925">loadAlleleEstimatesByProbeSet</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the allele summary table file and set up the A Allele signal matrix and the B Allele signal matrix.  <a href="#a1702c9c9e6478a46bacbc37bb1b35925"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#ad6c180521113b00d369349723e3e47a2">loadAlleleEstimatesByProbeSetHdf5</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the allele summary table file and set up the A Allele signal matrix and the B Allele signal matrix.  <a href="#ad6c180521113b00d369349723e3e47a2"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a757da16cb39daaf4f0e2d414739eb43a">loadAlleleEstimatesByExperiment</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the allele summary table file and set up the A Allele signal matrix and the B Allele signal matrix.  <a href="#a757da16cb39daaf4f0e2d414739eb43a"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7c075590b19839175bcbf77f84691570"></a><!-- doxytag: member="CNLog2RatioData::loadAlleleEstimatesByExperimentHdf5" ref="a7c075590b19839175bcbf77f84691570" args="(const AffxString &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><b>loadAlleleEstimatesByExperimentHdf5</b> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a9a8ee2981aa3c6bb673393ee95cd19ca">loadGenotypeCallsByProbeSet</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strCallsFileName, int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the genotype call table file and set up the genotype call matrix.  <a href="#a9a8ee2981aa3c6bb673393ee95cd19ca"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a64cf0fb90ed523dafc453567ec79eeb0">loadGenotypeCallsByProbeSetHdf5</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the genotype call table file and set up the genotype call matrix.  <a href="#a64cf0fb90ed523dafc453567ec79eeb0"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#af8ab553c6f703e4011860c712f032a21">loadGenotypeCallsByExperiment</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strCallsFileName, int iExperimentIndex, int iExperimentsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the genotype call table file and set up the genotype call matrix.  <a href="#af8ab553c6f703e4011860c712f032a21"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a648b936befb1d9cad9cf48dab2bc6769">loadGenotypeCallsByExperimentHdf5</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the genotype call table file and set up the genotype call matrix.  <a href="#a648b936befb1d9cad9cf48dab2bc6769"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a2edd544bbba693d90514017e444672b7">loadGenotypeConfidencesByExperiment</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the genotype confidence table file and set up the genotype confidence matrix.  <a href="#a2edd544bbba693d90514017e444672b7"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#ad9ee3c93676ecac48ce2f5dff1c86bdb">loadGenotypeConfidencesByExperimentHdf5</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load the genotype confidence table file and set up the genotype confidence matrix.  <a href="#ad9ee3c93676ecac48ce2f5dff1c86bdb"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#add763f5c3430427312c6717c20f34547">loadGenderOverrideFile</a> (const <a class="el" href="classAffxString.html">AffxString</a> &amp;strFileName)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Load externally computed experiment genders.  <a href="#add763f5c3430427312c6717c20f34547"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#aadb1ae253facb3ba6bdf394e4ae2bdaa">calculateMedianSignal</a> (int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate the median singal for the log2 ratio model.  <a href="#aadb1ae253facb3ba6bdf394e4ae2bdaa"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#a6f09d55e72a487848b95feeb3ec99f8a">calculateMedianSignalsPerGenotype</a> (int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate the median singals for the log2 ratio model.  <a href="#a6f09d55e72a487848b95feeb3ec99f8a"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classCNLog2RatioData.html#ac1c03578abe571e213745a085a101815">calculateXYMedianSignal</a> (int iProbeSetIndex, int iProbeSetsToProcess)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate the median singals for the log2 ratio model.  <a href="#ac1c03578abe571e213745a085a101815"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aeaa2ef871c639e774ae2c3cc94108bbb"></a><!-- doxytag: member="CNLog2RatioData::loadProbeSets" ref="aeaa2ef871c639e774ae2c3cc94108bbb" args="()" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>loadProbeSets</b> ()</td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<p>A class to the log2 ratio data as used by the log2 ratio engine. </p>

<p>Definition at line <a class="el" href="CNLog2RatioData_8h_source.html#l00045">45</a> of file <a class="el" href="CNLog2RatioData_8h_source.html">CNLog2RatioData.h</a>.</p>
<hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="aadb1ae253facb3ba6bdf394e4ae2bdaa"></a><!-- doxytag: member="CNLog2RatioData::calculateMedianSignal" ref="aadb1ae253facb3ba6bdf394e4ae2bdaa" args="(int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::calculateMedianSignal </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculate the median singal for the log2 ratio model. </p>
<p>The data is processed by probes set meaning the all the experiment data is processed for a subset of the probe sets. Note that the median signal is A + B. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01824">1824</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, and <a class="el" href="AffxConv_8cpp_source.html#l00369">roundDouble()</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="a6f09d55e72a487848b95feeb3ec99f8a"></a><!-- doxytag: member="CNLog2RatioData::calculateMedianSignalsPerGenotype" ref="a6f09d55e72a487848b95feeb3ec99f8a" args="(int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::calculateMedianSignalsPerGenotype </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculate the median singals for the log2 ratio model. </p>
<p>The data is processed by probes set meaning the all the experiment data is processed for a subset of the probe sets. Note that the median signals broken out by genotype is A - B. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01846">1846</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Options_8cpp_source.html#l00172">Options::getOptInt()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, and <a class="el" href="AffxConv_8cpp_source.html#l00369">roundDouble()</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="ac1c03578abe571e213745a085a101815"></a><!-- doxytag: member="CNLog2RatioData::calculateXYMedianSignal" ref="ac1c03578abe571e213745a085a101815" args="(int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::calculateXYMedianSignal </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculate the median singals for the log2 ratio model. </p>
<p>The data is processed by probes set meaning the all the experiment data is processed for a subset of the probe sets. Note that the median signal is A + B. The hasXX and hasY is set in the calculateXY function. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01886">1886</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, and <a class="el" href="AffxConv_8cpp_source.html#l00369">roundDouble()</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="aa2ac281f90626048a718c0deccba50db"></a><!-- doxytag: member="CNLog2RatioData::dumpCyto2Report" ref="aa2ac281f90626048a718c0deccba50db" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::dumpCyto2Report </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Dump the QC metrics into the copynumber report file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to dump to </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00845">845</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02603">affx::TsvFile::cidx2cname()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="Guid_8cpp_source.html#l00134">affxutil::Guid::GenerateNewGuid()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="AffxConv_8cpp_source.html#l00079">getDouble()</a>, <a class="el" href="AffxConv_8cpp_source.html#l00051">getInt()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01683">affx::TsvFile::headersBegin()</a>, <a class="el" href="Fs_8cpp_source.html#l00326">Fs::join()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="TsvFile_8cpp_source.html#l01694">affx::TsvFile::nextHeaderPtr()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, and <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>.</p>

</div>
</div>
<a class="anchor" id="af82738d0c4c7943f08da2cb7ad038742"></a><!-- doxytag: member="CNLog2RatioData::dumpGenotypeReport" ref="af82738d0c4c7943f08da2cb7ad038742" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::dumpGenotypeReport </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Dump the QC metrics into the copynumber report file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to dump to </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00700">700</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02603">affx::TsvFile::cidx2cname()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="Guid_8cpp_source.html#l00134">affxutil::Guid::GenerateNewGuid()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="AffxConv_8cpp_source.html#l00079">getDouble()</a>, <a class="el" href="AffxConv_8cpp_source.html#l00051">getInt()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01683">affx::TsvFile::headersBegin()</a>, <a class="el" href="Fs_8cpp_source.html#l00326">Fs::join()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="TsvFile_8cpp_source.html#l01694">affx::TsvFile::nextHeaderPtr()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, and <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioEngine_8cpp_source.html#l00297">CNLog2RatioEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="a757da16cb39daaf4f0e2d414739eb43a"></a><!-- doxytag: member="CNLog2RatioData::loadAlleleEstimatesByExperiment" ref="a757da16cb39daaf4f0e2d414739eb43a" args="(const AffxString &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadAlleleEstimatesByExperiment </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the allele summary table file and set up the A Allele signal matrix and the B Allele signal matrix. </p>
<p>The data is load by experiment meaning the all the probe set data is loaded for a subset of the experiments. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The experiment index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of exeriments to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01185">1185</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="Options_8cpp_source.html#l00158">Options::getOptBool()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02737">affx::TsvFile::openTable()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>, and <a class="el" href="CNLog2RatioEngine_8cpp_source.html#l00297">CNLog2RatioEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="a1702c9c9e6478a46bacbc37bb1b35925"></a><!-- doxytag: member="CNLog2RatioData::loadAlleleEstimatesByProbeSet" ref="a1702c9c9e6478a46bacbc37bb1b35925" args="(const AffxString &amp;strFileName, int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadAlleleEstimatesByProbeSet </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the allele summary table file and set up the A Allele signal matrix and the B Allele signal matrix. </p>
<p>The data is load by <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> set meaning the all the experiment data is loaded for a subset of the probe sets. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00982">982</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="Options_8cpp_source.html#l00158">Options::getOptBool()</a>, <a class="el" href="CNLog2RatioData_8cpp_source.html#l01072">loadAlleleEstimatesByProbeSetHdf5()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02737">affx::TsvFile::openTable()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="ad6c180521113b00d369349723e3e47a2"></a><!-- doxytag: member="CNLog2RatioData::loadAlleleEstimatesByProbeSetHdf5" ref="ad6c180521113b00d369349723e3e47a2" args="(const AffxString &amp;strFileName, int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadAlleleEstimatesByProbeSetHdf5 </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the allele summary table file and set up the A Allele signal matrix and the B Allele signal matrix. </p>
<p>The data is load by <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> set meaning the all the experiment data is loaded for a subset of the probe sets. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01072">1072</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="Options_8cpp_source.html#l00158">Options::getOptBool()</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioData_8cpp_source.html#l00982">loadAlleleEstimatesByProbeSet()</a>.</p>

</div>
</div>
<a class="anchor" id="a98563435a322545374c99293c7ff7c82"></a><!-- doxytag: member="CNLog2RatioData::loadCyto2ModelFile" ref="a98563435a322545374c99293c7ff7c82" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::loadCyto2ModelFile </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the model file and set up the <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> Set array with the pre-calculated median values from the log2 ratio model. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00325">325</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>.</p>

</div>
</div>
<a class="anchor" id="a80546473717d00e4a9d5a4e69b871485"></a><!-- doxytag: member="CNLog2RatioData::loadExperiments" ref="a80546473717d00e4a9d5a4e69b871485" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int CNLog2RatioData::loadExperiments </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the allele summary table file and set up the experiment array with experiment name. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>int - The number of experiments loaded </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00438">438</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, and <a class="el" href="TsvFile_8cpp_source.html#l02737">affx::TsvFile::openTable()</a>.</p>

</div>
</div>
<a class="anchor" id="add763f5c3430427312c6717c20f34547"></a><!-- doxytag: member="CNLog2RatioData::loadGenderOverrideFile" ref="add763f5c3430427312c6717c20f34547" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenderOverrideFile </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load externally computed experiment genders. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load from </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00515">515</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, <a class="el" href="BioTypes_8h_source.html#l00072">affx::Female</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="BioTypes_8h_source.html#l00073">affx::Male</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>.</p>

</div>
</div>
<a class="anchor" id="af8ab553c6f703e4011860c712f032a21"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeCallsByExperiment" ref="af8ab553c6f703e4011860c712f032a21" args="(const AffxString &amp;strCallsFileName, int iExperimentIndex, int iExperimentsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenotypeCallsByExperiment </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the genotype call table file and set up the genotype call matrix. </p>
<p>The data is load by experiment meaning the all the probe set data is loaded for a subset of the experiments. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The experiment index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of exeriments to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01533">1533</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="CNLog2RatioData_8cpp_source.html#l01600">loadGenotypeCallsByExperimentHdf5()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02737">affx::TsvFile::openTable()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioEngine_8cpp_source.html#l00297">CNLog2RatioEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="a648b936befb1d9cad9cf48dab2bc6769"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeCallsByExperimentHdf5" ref="a648b936befb1d9cad9cf48dab2bc6769" args="(const AffxString &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenotypeCallsByExperimentHdf5 </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the genotype call table file and set up the genotype call matrix. </p>
<p>The data is load by experiment meaning the all the probe set data is loaded for a subset of the experiments. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The experiment index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of exeriments to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01600">1600</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="File5__Tsv_8cpp_source.html#l00418">affx::File5_Tsv::getColumnPtr()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioData_8cpp_source.html#l01533">loadGenotypeCallsByExperiment()</a>.</p>

</div>
</div>
<a class="anchor" id="a9a8ee2981aa3c6bb673393ee95cd19ca"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeCallsByProbeSet" ref="a9a8ee2981aa3c6bb673393ee95cd19ca" args="(const AffxString &amp;strCallsFileName, int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenotypeCallsByProbeSet </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the genotype call table file and set up the genotype call matrix. </p>
<p>The data is load by <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> set meaning the all the experiment data is loaded for a subset of the probe sets. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01380">1380</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="CNLog2RatioData_8cpp_source.html#l01448">loadGenotypeCallsByProbeSetHdf5()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02737">affx::TsvFile::openTable()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="a64cf0fb90ed523dafc453567ec79eeb0"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeCallsByProbeSetHdf5" ref="a64cf0fb90ed523dafc453567ec79eeb0" args="(const AffxString &amp;strFileName, int iProbeSetIndex, int iProbeSetsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenotypeCallsByProbeSetHdf5 </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iProbeSetsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the genotype call table file and set up the genotype call matrix. </p>
<p>The data is load by <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> set meaning the all the experiment data is loaded for a subset of the probe sets. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The probe set index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of probe sets to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01448">1448</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="File5__Tsv_8cpp_source.html#l00418">affx::File5_Tsv::getColumnPtr()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioData_8cpp_source.html#l01380">loadGenotypeCallsByProbeSet()</a>.</p>

</div>
</div>
<a class="anchor" id="a2edd544bbba693d90514017e444672b7"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeConfidencesByExperiment" ref="a2edd544bbba693d90514017e444672b7" args="(const AffxString &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenotypeConfidencesByExperiment </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the genotype confidence table file and set up the genotype confidence matrix. </p>
<p>The data is load by experiment meaning the all the probe set data is loaded for a subset of the experiments. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The experiment index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of exeriments to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01682">1682</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="CNLog2RatioData_8cpp_source.html#l01743">loadGenotypeConfidencesByExperimentHdf5()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02737">affx::TsvFile::openTable()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioEngine_8cpp_source.html#l00297">CNLog2RatioEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="ad9ee3c93676ecac48ce2f5dff1c86bdb"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeConfidencesByExperimentHdf5" ref="ad9ee3c93676ecac48ce2f5dff1c86bdb" args="(const AffxString &amp;strFileName, int iExperimentIndex, int iExperimentsToProcess)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadGenotypeConfidencesByExperimentHdf5 </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>iExperimentsToProcess</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the genotype confidence table file and set up the genotype confidence matrix. </p>
<p>The data is load by experiment meaning the all the probe set data is loaded for a subset of the experiments. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The experiment index to start from. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>int</em>&nbsp;</td><td>- The number of exeriments to process each time this function is called. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01743">1743</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="File5__Tsv_8cpp_source.html#l00418">affx::File5_Tsv::getColumnPtr()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

<p>Referenced by <a class="el" href="CNLog2RatioData_8cpp_source.html#l01682">loadGenotypeConfidencesByExperiment()</a>.</p>

</div>
</div>
<a class="anchor" id="a9c95fc886b02b3a4bcdbb3a48d2cd867"></a><!-- doxytag: member="CNLog2RatioData::loadGenotypeReport" ref="a9c95fc886b02b3a4bcdbb3a48d2cd867" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::loadGenotypeReport </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the QC metric from the genotype report file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load from </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00635">635</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02563">affx::TsvFile::cname2cidx()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="AffxConv_8cpp_source.html#l00051">getInt()</a>, <a class="el" href="Options_8cpp_source.html#l00158">Options::getOptBool()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, <a class="el" href="CNExperiment_8cpp_source.html#l00479">CNExperiment::setRawIntensityRatioGenderFromString()</a>, and <a class="el" href="TsvFile_8h_source.html#l00092">affx::TSV_OK</a>.</p>

<p>Referenced by <a class="el" href="CNReferenceEngine_8cpp_source.html#l00267">CNReferenceEngine::processData()</a>, and <a class="el" href="CNLog2RatioEngine_8cpp_source.html#l00297">CNLog2RatioEngine::processData()</a>.</p>

</div>
</div>
<a class="anchor" id="ac2d7caba60586cf34f587b645632eefb"></a><!-- doxytag: member="CNLog2RatioData::loadModelFile" ref="ac2d7caba60586cf34f587b645632eefb" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadModelFile </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the model file and set up the <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> Set array with the pre-calculated median values from the log2 ratio model. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p><dl class="todo"><dt><b><a class="el" href="todo.html#_todo000220">Todo:</a></b></dt><dd>AW: should this be "state" not "option"? Do we need to handle both for compatability reasons? </dd></dl>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000221">Todo:</a></b></dt><dd>AW: should this be "state" not "option"? Do we need to handle both for compatability reasons? </dd></dl>
</p>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00140">140</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02603">affx::TsvFile::cidx2cname()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="TsvFile_8cpp_source.html#l04891">affx::TsvFile::getLineCountInFile()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01683">affx::TsvFile::headersBegin()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01737">affx::TsvFile::headersNext()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="File5__Group_8cpp_source.html#l00433">affx::File5_Group::name_exists()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>, <a class="el" href="CNAnalysisMethodCovariateParams_8cpp_source.html#l00326">CovariateParams::restoreCovariateParams()</a>, <a class="el" href="ParameterNameValueType_8h_source.html#l00180">affymetrix_calvin_parameter::ParameterNameValueType::SetName()</a>, <a class="el" href="Options_8cpp_source.html#l00065">Options::setOpt()</a>, and <a class="el" href="ParameterNameValueType_8cpp_source.html#l00317">affymetrix_calvin_parameter::ParameterNameValueType::SetValueAscii()</a>.</p>

</div>
</div>
<a class="anchor" id="a781c14ec7d3a8d045c4ec139f73440a0"></a><!-- doxytag: member="CNLog2RatioData::loadQCMetrics" ref="a781c14ec7d3a8d045c4ec139f73440a0" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::loadQCMetrics </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the QC metrics from the report file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load from </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00600">600</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03576">affx::TsvFile::get()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01572">affx::TsvFile::getColumnCount()</a>, <a class="el" href="TsvFile_8h_source.html#l00431">affx::TsvFile::m_optAutoTrim</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, and <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>.</p>

</div>
</div>
<a class="anchor" id="a235e31fd7f73d9b75bb00257c85df4c6"></a><!-- doxytag: member="CNLog2RatioData::loadSnpReferenceFile" ref="a235e31fd7f73d9b75bb00257c85df4c6" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::loadSnpReferenceFile </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Load the SNP reference file and set up the <a class="el" href="classProbe.html" title="Represents an individual feature on a chip.">Probe</a> Set array with the pre-calculated median values from the log2 ratio model. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The name of the file to load. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l00370">370</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>.</p>

</div>
</div>
<a class="anchor" id="a336f2d8a1870247b2be8e0579b303b4c"></a><!-- doxytag: member="CNLog2RatioData::writeCyto2ModelFile" ref="a336f2d8a1870247b2be8e0579b303b4c" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void CNLog2RatioData::writeCyto2ModelFile </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Write the model file in text format. </p>
<p>All the median signals asssociated with the probe sets are written out. Values are rounded to insure that the stored values will match the calculated values when using the same input. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The file name to write to. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l02134">2134</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00430">Fs::dirname()</a>, <a class="el" href="Guid_8cpp_source.html#l00134">affxutil::Guid::GenerateNewGuid()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="Options_8cpp_source.html#l00192">Options::getOptVector()</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="Fs_8cpp_source.html#l01102">Fs::mkdirPath()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, and <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>.</p>

</div>
</div>
<a class="anchor" id="ac81ed0f535099554a833785fbe32564d"></a><!-- doxytag: member="CNLog2RatioData::writeModelFile" ref="ac81ed0f535099554a833785fbe32564d" args="(const AffxString &amp;strFileName)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool CNLog2RatioData::writeModelFile </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAffxString.html">AffxString</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>strFileName</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Write the model file in text format. </p>
<p>All the median signals asssociated with the probe sets are written out. Values are rounded to insure that the stored values will match the calculated values when using the same input. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td><a class="el" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; - The file name to write to. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>bool - true if successful </dd></dl>

<p>Definition at line <a class="el" href="CNLog2RatioData_8cpp_source.html#l01930">1930</a> of file <a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a>.</p>

<p>References <a class="el" href="TsvFile_8cpp_source.html#l01804">affx::TsvFile::addHeader()</a>, <a class="el" href="TsvFile_8cpp_source.html#l01436">affx::TsvFile::clear()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02765">affx::TsvFile::close()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02417">affx::TsvFile::defineColumn()</a>, <a class="el" href="Fs_8cpp_source.html#l00430">Fs::dirname()</a>, <a class="el" href="Guid_8cpp_source.html#l00134">affxutil::Guid::GenerateNewGuid()</a>, <a class="el" href="AffxConv_8cpp_source.html#l00079">getDouble()</a>, <a class="el" href="Options_8h_source.html#l00184">Options::getOpt()</a>, <a class="el" href="Options_8cpp_source.html#l00158">Options::getOptBool()</a>, <a class="el" href="Options_8cpp_source.html#l00192">Options::getOptVector()</a>, <a class="el" href="Options_8cpp_source.html#l00054">Options::isOptDefined()</a>, <a class="el" href="affy-base-types_8h_source.html#l00059">Max</a>, <a class="el" href="Fs_8cpp_source.html#l01102">Fs::mkdirPath()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, <a class="el" href="Verbose_8cpp_source.html#l00078">Verbose::progressStep()</a>, <a class="el" href="TsvFile_8cpp_source.html#l04673">affx::TsvFile::writeLevel()</a>, and <a class="el" href="TsvFile_8cpp_source.html#l04573">affx::TsvFile::writeTsv_v1()</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>copynumber/<a class="el" href="CNLog2RatioData_8h_source.html">CNLog2RatioData.h</a></li>
<li>copynumber/<a class="el" href="CNLog2RatioData_8cpp_source.html">CNLog2RatioData.cpp</a></li>
</ul>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:24 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
