<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/CNAnalysisMethodSegment.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>CNAnalysisMethodSegment.cpp</h1>  </div>
</div>
<div class="contents">
<a href="CNAnalysisMethodSegment_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file CNAnalysisMethodSegment.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> *</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief This file contains the CNAnalysisMethodSegment class members.</span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 <span class="preprocessor">#include &quot;<a class="code" href="CNAnalysisMethodSegment_8h.html" title="This header contains the CNAnalysisMethodSegment class definition.">copynumber/CNAnalysisMethodSegment.h</a>&quot;</span>
<a name="l00028"></a>00028 <span class="comment">//</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="StringUtils_8h.html">calvin_files/utils/src/StringUtils.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="AffxStatistics_8h.html" title="This header contains the AffxStatistics class definition.">util/AffxStatistics.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="Guid_8h.html" title="This file provides functions for creating globally unique identifiers.">util/Guid.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="comment">//</span>
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 <span class="comment"></span>
<a name="l00035"></a>00035 <span class="comment">/**</span>
<a name="l00036"></a>00036 <span class="comment"> * @brief Constructor</span>
<a name="l00037"></a>00037 <span class="comment"> */</span>
<a name="l00038"></a><a class="code" href="classCNAnalysisMethodSegment.html#a228ec2306c5893d2bd51dbb448b5612b">00038</a> <a class="code" href="classCNAnalysisMethodSegment.html#a228ec2306c5893d2bd51dbb448b5612b" title="Constructor.">CNAnalysisMethodSegment::CNAnalysisMethodSegment</a>()
<a name="l00039"></a>00039 {
<a name="l00040"></a>00040 }
<a name="l00041"></a>00041 <span class="comment"></span>
<a name="l00042"></a>00042 <span class="comment">/**</span>
<a name="l00043"></a>00043 <span class="comment"> * @brief Create new segments</span>
<a name="l00044"></a>00044 <span class="comment"> * @param int - The segment type to create</span>
<a name="l00045"></a>00045 <span class="comment"> * @param CNSegmentArray&amp; - The segment vector to load</span>
<a name="l00046"></a>00046 <span class="comment"> * @param int - The affx sex code</span>
<a name="l00047"></a>00047 <span class="comment"> * @param int - The X chromosome numeric value</span>
<a name="l00048"></a>00048 <span class="comment"> * @param int - The Y chromosome numeric value</span>
<a name="l00049"></a>00049 <span class="comment"> * @return int - The number of new segments created</span>
<a name="l00050"></a>00050 <span class="comment"> */</span>
<a name="l00051"></a><a class="code" href="classCNAnalysisMethodSegment.html#a4f9fed50b7bb5048f9835213b83995d3">00051</a> <span class="keywordtype">int</span> <a class="code" href="classCNAnalysisMethodSegment.html#a4f9fed50b7bb5048f9835213b83995d3" title="Create new segments.">CNAnalysisMethodSegment::newSegments</a>(    <span class="keywordtype">int</span> iSegmentType,
<a name="l00052"></a>00052                                               <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a>* vProbeSets,
<a name="l00053"></a>00053                                              <span class="keywordtype">int</span> iMinSegSeparation   )
<a name="l00054"></a>00054 {
<a name="l00055"></a>00055     <a class="code" href="classAffxMultiDimensionalArray.html" title="A class for managing one, two, or three dimensional arrays of primitive data types.">AffxMultiDimensionalArray&lt;int&gt;</a> vMarkerDistances(vProbeSets-&gt;getCount());
<a name="l00056"></a>00056     <span class="keywordtype">int</span> iPrevMarkerPosition = 0;
<a name="l00057"></a>00057     <span class="keywordflow">if</span> (vProbeSets-&gt;getCount() == 0) {<span class="keywordflow">return</span> 0;}
<a name="l00058"></a>00058     <span class="keywordtype">int</span> iSegmentCount = 0;
<a name="l00059"></a>00059     <span class="keywordtype">int</span> iMarkerCount = 0;
<a name="l00060"></a>00060     <span class="keywordtype">double</span> dSumConfidence = 0.0;
<a name="l00061"></a>00061     <span class="keywordtype">double</span> dConfidence=0.0;
<a name="l00062"></a>00062     <span class="keywordtype">int</span> iHetsInSegment=0;
<a name="l00063"></a>00063     <span class="keywordtype">int</span> iHomsInSegment=0;
<a name="l00064"></a>00064     <a class="code" href="classCNSegment.html" title="A class for storing Segment data.">CNSegment</a>* pobjSegment = NULL;
<a name="l00065"></a>00065     <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a>* pobjProbeSet = NULL;
<a name="l00066"></a>00066     <span class="keywordtype">int</span> iStartIndex = 0;
<a name="l00067"></a>00067     <span class="comment">// If Segment type loh-segment, loh-cyto2, cn-neutral-log, normal-diploid</span>
<a name="l00068"></a>00068     <span class="keywordflow">if</span> ((iSegmentType == 2) || (iSegmentType == 3) || (iSegmentType == 4) || (iSegmentType == 5))
<a name="l00069"></a>00069     {
<a name="l00070"></a>00070         <span class="keywordflow">for</span> (;(iStartIndex &lt; vProbeSets-&gt;getCount()); iStartIndex++)
<a name="l00071"></a>00071         {
<a name="l00072"></a>00072             pobjProbeSet = vProbeSets-&gt;getAt(iStartIndex);
<a name="l00073"></a>00073             <span class="keywordflow">if</span> ((<span class="keywordtype">double</span>)pobjProbeSet-&gt;getLoh() == (double)pobjProbeSet-&gt;getLoh()) <span class="comment">// Not = NaN</span>
<a name="l00074"></a>00074             {
<a name="l00075"></a>00075                 <span class="keywordflow">break</span>;
<a name="l00076"></a>00076             }
<a name="l00077"></a>00077         }
<a name="l00078"></a>00078     }
<a name="l00079"></a>00079     <span class="keywordtype">int</span> iCurrCNState = -1;
<a name="l00080"></a>00080     <span class="keywordtype">bool</span> cytoScanHD = m_pEngine-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;cytoscan-hd&quot;</span>);
<a name="l00081"></a>00081     <span class="keywordflow">if</span> (iStartIndex &lt; vProbeSets-&gt;getCount())
<a name="l00082"></a>00082     {
<a name="l00083"></a>00083         pobjProbeSet = vProbeSets-&gt;getAt(iStartIndex);
<a name="l00084"></a>00084         <span class="keywordflow">if</span> (cytoScanHD)
<a name="l00085"></a>00085         {
<a name="l00086"></a>00086             <span class="comment">// Case 1: SNP marker with CN=-1 sits at the beginning of the chromosome</span>
<a name="l00087"></a>00087             <span class="keywordtype">unsigned</span> <span class="keywordtype">char</span> currentChromosome = pobjProbeSet-&gt;getChromosome();
<a name="l00088"></a>00088             <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = iStartIndex; (iIndex &lt; vProbeSets-&gt;getCount()); iIndex++)
<a name="l00089"></a>00089             {
<a name="l00090"></a>00090                 <span class="keywordflow">if</span> (currentChromosome != vProbeSets-&gt;getAt(iIndex)-&gt;getChromosome()) <span class="keywordflow">break</span>;
<a name="l00091"></a>00091                 <span class="keywordflow">if</span> (vProbeSets-&gt;getAt(iIndex)-&gt;getCNState() != -1)
<a name="l00092"></a>00092                 {
<a name="l00093"></a>00093                     iCurrCNState = vProbeSets-&gt;getAt(iIndex)-&gt;getCNState();
<a name="l00094"></a>00094                     <span class="keywordflow">break</span>;
<a name="l00095"></a>00095                 }
<a name="l00096"></a>00096             }
<a name="l00097"></a>00097         }
<a name="l00098"></a>00098         <span class="keywordtype">char</span> cCall = pobjProbeSet-&gt;<a class="code" href="classCNProbeSet.html#a698aca1bcd60daeb33f60328027a2c81" title="Return the specified segment type call.">getSegmentTypeCall</a>(iSegmentType, getExperiment()-&gt;getCNCallGenderAsInt(), m_iXChromosome, m_iYChromosome);
<a name="l00099"></a>00099         <span class="keywordtype">float</span> fConfidence = pobjProbeSet-&gt;<a class="code" href="classCNProbeSet.html#a5789c3f228f5984e451222172f777a71" title="Return the specified segment type confidence.">getSegmentTypeConfidence</a>(iSegmentType);
<a name="l00100"></a>00100         <span class="keywordtype">bool</span> bPar = pobjProbeSet-&gt;isPseudoAutosomalRegion();
<a name="l00101"></a>00101         <span class="keywordflow">if</span> (cytoScanHD)
<a name="l00102"></a>00102         {
<a name="l00103"></a>00103             <span class="comment">// Override cCall in the two cases where CN state makes a difference.</span>
<a name="l00104"></a>00104                     <span class="keywordflow">if</span> (iSegmentType == 4) cCall = (((iCurrCNState == 2) &amp;&amp; (pobjProbeSet-&gt;getLoh() == 1)) ? (char)1 : (<span class="keywordtype">char</span>)0); <span class="comment">// CNNeutralLOH</span>
<a name="l00105"></a>00105                     <span class="keywordflow">if</span> (iSegmentType == 5) cCall = (((iCurrCNState == 2) &amp;&amp; (pobjProbeSet-&gt;getLoh() == 0)) ? (char)1 : (<span class="keywordtype">char</span>)0); <span class="comment">// Normal Diploid</span>
<a name="l00106"></a>00106         }
<a name="l00107"></a>00107 
<a name="l00108"></a>00108         pobjSegment = <span class="keyword">new</span> <a class="code" href="classCNSegment.html" title="A class for storing Segment data.">CNSegment</a>;
<a name="l00109"></a>00109         pobjSegment-&gt;setSegmentType(iSegmentType);
<a name="l00110"></a>00110         pobjSegment-&gt;setChromosome(pobjProbeSet-&gt;getChromosome());
<a name="l00111"></a>00111         pobjSegment-&gt;setStartPosition(pobjProbeSet-&gt;getPosition());
<a name="l00112"></a>00112         pobjSegment-&gt;setEndPosition(pobjProbeSet-&gt;getPosition());
<a name="l00113"></a>00113         pobjSegment-&gt;setCall(cCall);
<a name="l00114"></a>00114         pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>)0.5);
<a name="l00115"></a>00115         pobjSegment-&gt;setPseudoAutosomalRegion(bPar);
<a name="l00116"></a>00116         iMarkerCount = 0;
<a name="l00117"></a>00117         dSumConfidence = 0;
<a name="l00118"></a>00118         iPrevMarkerPosition = pobjProbeSet-&gt;getPosition();
<a name="l00119"></a>00119 
<a name="l00120"></a>00120         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = iStartIndex; (iIndex &lt; vProbeSets-&gt;getCount()); iIndex++)
<a name="l00121"></a>00121         {
<a name="l00122"></a>00122             pobjProbeSet = vProbeSets-&gt;getAt(iIndex);
<a name="l00123"></a>00123             cCall = pobjProbeSet-&gt;<a class="code" href="classCNProbeSet.html#a698aca1bcd60daeb33f60328027a2c81" title="Return the specified segment type call.">getSegmentTypeCall</a>(iSegmentType, getExperiment()-&gt;getCNCallGenderAsInt(), m_iXChromosome, m_iYChromosome);
<a name="l00124"></a>00124             fConfidence = pobjProbeSet-&gt;<a class="code" href="classCNProbeSet.html#a5789c3f228f5984e451222172f777a71" title="Return the specified segment type confidence.">getSegmentTypeConfidence</a>(iSegmentType);
<a name="l00125"></a>00125             bPar = pobjProbeSet-&gt;isPseudoAutosomalRegion();
<a name="l00126"></a>00126                         
<a name="l00127"></a>00127             <span class="comment">// Case 2: If a SNP marker falls within a CN segment (including CN or SNP), assign SNP marker with the same CN state</span>
<a name="l00128"></a>00128             <span class="comment">// Case 3: If a SNP marker falls outside any CN segment, assign CN state for this SNP marker based on the CN segment just right before it.</span>
<a name="l00129"></a>00129                         <span class="keywordflow">if</span> (pobjProbeSet-&gt;getCNState() != -1) {iCurrCNState = pobjProbeSet-&gt;getCNState();}
<a name="l00130"></a>00130 
<a name="l00131"></a>00131                         <span class="keywordflow">if</span> (cytoScanHD)
<a name="l00132"></a>00132                         {
<a name="l00133"></a>00133                 <span class="comment">// Override cCall in the two cases where CN state makes a difference.</span>
<a name="l00134"></a>00134                         <span class="keywordflow">if</span> (iSegmentType == 4) cCall = (((iCurrCNState == 2) &amp;&amp; (pobjProbeSet-&gt;getLoh() == 1)) ? (char)1 : (<span class="keywordtype">char</span>)0); <span class="comment">// CNNeutralLOH</span>
<a name="l00135"></a>00135                         <span class="keywordflow">if</span> (iSegmentType == 5) cCall = (((iCurrCNState == 2) &amp;&amp; (pobjProbeSet-&gt;getLoh() == 0)) ? (char)1 : (<span class="keywordtype">char</span>)0); <span class="comment">// Normal Diploid</span>
<a name="l00136"></a>00136                         }
<a name="l00137"></a>00137             <span class="keywordflow">if</span> ((iSegmentType == 2) || (iSegmentType == 3) || (iSegmentType == 4) || (iSegmentType == 5))
<a name="l00138"></a>00138             {
<a name="l00139"></a>00139                 <span class="keywordflow">if</span> ((<span class="keywordtype">double</span>)pobjProbeSet-&gt;getLoh() != (double)pobjProbeSet-&gt;getLoh()) <span class="comment">// NaN</span>
<a name="l00140"></a>00140                 {
<a name="l00141"></a>00141                     <span class="keywordflow">continue</span>;
<a name="l00142"></a>00142                 }
<a name="l00143"></a>00143             }
<a name="l00144"></a>00144             <span class="keywordtype">bool</span> newSegmentTestFlag = <span class="keyword">false</span>;
<a name="l00145"></a>00145             <span class="keywordflow">if</span>(iSegmentType == 2)
<a name="l00146"></a>00146             { 
<a name="l00147"></a>00147                 newSegmentTestFlag = (  
<a name="l00148"></a>00148                     (pobjProbeSet-&gt;getChromosome() != pobjSegment-&gt;getChromosome()) 
<a name="l00149"></a>00149                  || (cCall != pobjSegment-&gt;getCall()) 
<a name="l00150"></a>00150                  || ((getExperiment()-&gt;getCNCallGenderAsInt() == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669afdfe578a9e9044b7b274b736ed341a81" title="Has XY chromosomes.">affx::Male</a>) &amp;&amp; (iSegmentType == 1) &amp;&amp; (bPar != pobjSegment-&gt;isPseudoAutosomalRegion()))
<a name="l00151"></a>00151                  || (fabs((<span class="keywordtype">float</span>)iPrevMarkerPosition - (<span class="keywordtype">float</span>) (pobjProbeSet-&gt;getPosition())) &gt; (float) iMinSegSeparation) 
<a name="l00152"></a>00152                                      );
<a name="l00153"></a>00153             }
<a name="l00154"></a>00154             <span class="keywordflow">else</span>
<a name="l00155"></a>00155             {
<a name="l00156"></a>00156                 newSegmentTestFlag = (
<a name="l00157"></a>00157                     (pobjProbeSet-&gt;getChromosome() != pobjSegment-&gt;getChromosome()) 
<a name="l00158"></a>00158                  || (cCall != pobjSegment-&gt;getCall()) 
<a name="l00159"></a>00159                  || ((getExperiment()-&gt;getCNCallGenderAsInt() == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669afdfe578a9e9044b7b274b736ed341a81" title="Has XY chromosomes.">affx::Male</a>) &amp;&amp; (iSegmentType == 1) &amp;&amp; (bPar != pobjSegment-&gt;isPseudoAutosomalRegion()))
<a name="l00160"></a>00160                                      ); 
<a name="l00161"></a>00161             }
<a name="l00162"></a>00162 
<a name="l00163"></a>00163             <span class="keywordflow">if</span> (newSegmentTestFlag)
<a name="l00164"></a>00164             {
<a name="l00165"></a>00165                 pobjSegment-&gt;setMeanMarkerDistance((<span class="keywordtype">float</span>)vMarkerDistances.mean(iMarkerCount - 1));
<a name="l00166"></a>00166                 pobjSegment-&gt;setSegmentName(<a class="code" href="classaffxutil_1_1Guid.html#a2929c982cb9f3b5f6c5d7d818a8e6ac4">affxutil::Guid::GenerateNewGuid</a>());
<a name="l00167"></a>00167                 pobjSegment-&gt;setMarkerCount(iMarkerCount);
<a name="l00168"></a>00168                 pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>)(dSumConfidence / (<span class="keywordtype">double</span>)iMarkerCount));
<a name="l00169"></a>00169 
<a name="l00170"></a>00170                 <span class="keywordflow">if</span>(cytoScanHD)
<a name="l00171"></a>00171                 {
<a name="l00172"></a>00172                     <span class="keywordflow">if</span>(iSegmentType==2)
<a name="l00173"></a>00173                     {
<a name="l00174"></a>00174                         dConfidence = assignConfidenceSigmoidal(        iHetsInSegment,
<a name="l00175"></a>00175                                                                         iHetsInSegment + iHomsInSegment,
<a name="l00176"></a>00176                                                                         getExperiment()-&gt;getHetFrequency(),
<a name="l00177"></a>00177                                                                         getExperiment()-&gt;getPError(),
<a name="l00178"></a>00178                                                                         3,
<a name="l00179"></a>00179                                                                         10);
<a name="l00180"></a>00180                         pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>) dConfidence); 
<a name="l00181"></a>00181                     }
<a name="l00182"></a>00182                     <span class="keywordflow">if</span>(iSegmentType==4)
<a name="l00183"></a>00183                     {
<a name="l00184"></a>00184                         dConfidence = assignConfidenceSigmoidal(        iHetsInSegment,
<a name="l00185"></a>00185                                                                         iHetsInSegment + iHomsInSegment,
<a name="l00186"></a>00186                                                                         getExperiment()-&gt;getHetFrequency(),
<a name="l00187"></a>00187                                                                         getExperiment()-&gt;getPError(),
<a name="l00188"></a>00188                                                                         3,
<a name="l00189"></a>00189                                                                         10);
<a name="l00190"></a>00190                         pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>) (dConfidence/2.0));
<a name="l00191"></a>00191                     }
<a name="l00192"></a>00192 
<a name="l00193"></a>00193                }
<a name="l00194"></a>00194 
<a name="l00195"></a>00195                 m_vSegments.add(pobjSegment);
<a name="l00196"></a>00196                 iSegmentCount++;
<a name="l00197"></a>00197 
<a name="l00198"></a>00198                 <span class="keywordflow">if</span> (cytoScanHD &amp;&amp; pobjProbeSet-&gt;getChromosome() != pobjSegment-&gt;getChromosome())
<a name="l00199"></a>00199                 {
<a name="l00200"></a>00200                     <span class="comment">// Case 1: SNP marker with CN=-1 sits at the beginning of the chromosome</span>
<a name="l00201"></a>00201                     iCurrCNState = -1;
<a name="l00202"></a>00202                     <span class="keywordtype">unsigned</span> <span class="keywordtype">char</span> currentChromosome = pobjProbeSet-&gt;getChromosome();
<a name="l00203"></a>00203                     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iInner = iIndex; (iInner &lt; vProbeSets-&gt;getCount()); iInner++)
<a name="l00204"></a>00204                     {
<a name="l00205"></a>00205                         <span class="keywordflow">if</span> (currentChromosome != vProbeSets-&gt;getAt(iIndex)-&gt;getChromosome()) <span class="keywordflow">break</span>;
<a name="l00206"></a>00206                         <span class="keywordflow">if</span> (vProbeSets-&gt;getAt(iInner)-&gt;getCNState() != -1)
<a name="l00207"></a>00207                         {
<a name="l00208"></a>00208                             iCurrCNState = vProbeSets-&gt;getAt(iInner)-&gt;getCNState();
<a name="l00209"></a>00209                             <span class="keywordflow">break</span>;
<a name="l00210"></a>00210                         }
<a name="l00211"></a>00211                     }
<a name="l00212"></a>00212                      <span class="comment">// Override cCall in the two cases where CN state makes a difference.</span>
<a name="l00213"></a>00213                         <span class="keywordflow">if</span> (iSegmentType == 4) cCall = (((iCurrCNState == 2) &amp;&amp; (pobjProbeSet-&gt;getLoh() == 1)) ? (char)1 : (<span class="keywordtype">char</span>)0); <span class="comment">// CNNeutralLOH</span>
<a name="l00214"></a>00214                         <span class="keywordflow">if</span> (iSegmentType == 5) cCall = (((iCurrCNState == 2) &amp;&amp; (pobjProbeSet-&gt;getLoh() == 0)) ? (char)1 : (<span class="keywordtype">char</span>)0); <span class="comment">// Normal Diploid</span>
<a name="l00215"></a>00215                 }
<a name="l00216"></a>00216 
<a name="l00217"></a>00217                 pobjSegment = <span class="keyword">new</span> <a class="code" href="classCNSegment.html" title="A class for storing Segment data.">CNSegment</a>;
<a name="l00218"></a>00218                 pobjSegment-&gt;setSegmentType(iSegmentType);
<a name="l00219"></a>00219                 pobjSegment-&gt;setChromosome(pobjProbeSet-&gt;getChromosome());
<a name="l00220"></a>00220                 pobjSegment-&gt;setStartPosition(pobjProbeSet-&gt;getPosition());
<a name="l00221"></a>00221                 pobjSegment-&gt;setCall(cCall);
<a name="l00222"></a>00222                 pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>)0.5);
<a name="l00223"></a>00223                 pobjSegment-&gt;setPseudoAutosomalRegion(bPar);
<a name="l00224"></a>00224                 iHetsInSegment=0;
<a name="l00225"></a>00225                 iHomsInSegment=0;
<a name="l00226"></a>00226                 iMarkerCount = 0;
<a name="l00227"></a>00227                 dSumConfidence = 0;
<a name="l00228"></a>00228                 iPrevMarkerPosition = pobjProbeSet-&gt;getPosition();
<a name="l00229"></a>00229             }
<a name="l00230"></a>00230             iMarkerCount++;
<a name="l00231"></a>00231 
<a name="l00232"></a>00232             <span class="keywordflow">if</span>(pobjProbeSet-&gt;getGenotypeCall() == (char) 2 || pobjProbeSet-&gt;getGenotypeCall() == (char) 0)
<a name="l00233"></a>00233             {
<a name="l00234"></a>00234                 iHomsInSegment++;
<a name="l00235"></a>00235             }
<a name="l00236"></a>00236 
<a name="l00237"></a>00237             <span class="keywordflow">if</span>(pobjProbeSet-&gt;getGenotypeCall() == (char) 1 )
<a name="l00238"></a>00238             {
<a name="l00239"></a>00239                 iHetsInSegment++;
<a name="l00240"></a>00240             }
<a name="l00241"></a>00241 
<a name="l00242"></a>00242 
<a name="l00243"></a>00243             dSumConfidence += fConfidence;
<a name="l00244"></a>00244             pobjSegment-&gt;setEndPosition(pobjProbeSet-&gt;getPosition());
<a name="l00245"></a>00245             <span class="keywordflow">if</span> (iMarkerCount &gt; 1) {vMarkerDistances.set((iMarkerCount - 2), (pobjProbeSet-&gt;getPosition() - iPrevMarkerPosition));}
<a name="l00246"></a>00246             iPrevMarkerPosition = pobjProbeSet-&gt;getPosition();
<a name="l00247"></a>00247         }
<a name="l00248"></a>00248     }
<a name="l00249"></a>00249     <span class="keywordflow">if</span> (pobjSegment != NULL)
<a name="l00250"></a>00250     {
<a name="l00251"></a>00251         pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>)(dSumConfidence / (<span class="keywordtype">double</span>)iMarkerCount));
<a name="l00252"></a>00252 
<a name="l00253"></a>00253         <span class="keywordflow">if</span>(cytoScanHD)
<a name="l00254"></a>00254         { 
<a name="l00255"></a>00255             <span class="keywordflow">if</span>(iSegmentType==2)
<a name="l00256"></a>00256             {
<a name="l00257"></a>00257                 dConfidence = assignConfidenceSigmoidal(        iHetsInSegment,
<a name="l00258"></a>00258                                                                 iHetsInSegment + iHomsInSegment,
<a name="l00259"></a>00259                                                                 getExperiment()-&gt;getHetFrequency(),
<a name="l00260"></a>00260                                                                 getExperiment()-&gt;getPError(),
<a name="l00261"></a>00261                                                                 3,
<a name="l00262"></a>00262                                                                 10);
<a name="l00263"></a>00263                  pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>) dConfidence);
<a name="l00264"></a>00264             }
<a name="l00265"></a>00265             <span class="keywordflow">if</span>(iSegmentType==4)
<a name="l00266"></a>00266             {
<a name="l00267"></a>00267                 dConfidence = assignConfidenceSigmoidal(        iHetsInSegment,
<a name="l00268"></a>00268                                                                 iHetsInSegment + iHomsInSegment,
<a name="l00269"></a>00269                                                                 getExperiment()-&gt;getHetFrequency(),
<a name="l00270"></a>00270                                                                 getExperiment()-&gt;getPError(),
<a name="l00271"></a>00271                                                                 3,
<a name="l00272"></a>00272                                                                 10);
<a name="l00273"></a>00273                 pobjSegment-&gt;setConfidence((<span class="keywordtype">float</span>) (dConfidence/2.0));
<a name="l00274"></a>00274             }
<a name="l00275"></a>00275         }  
<a name="l00276"></a>00276 
<a name="l00277"></a>00277         pobjSegment-&gt;setMeanMarkerDistance((<span class="keywordtype">float</span>)vMarkerDistances.mean(iMarkerCount - 1));
<a name="l00278"></a>00278         pobjSegment-&gt;setMarkerCount(iMarkerCount);
<a name="l00279"></a>00279         pobjSegment-&gt;setSegmentName(<a class="code" href="classaffxutil_1_1Guid.html#a2929c982cb9f3b5f6c5d7d818a8e6ac4">affxutil::Guid::GenerateNewGuid</a>());
<a name="l00280"></a>00280         m_vSegments.add(pobjSegment); iSegmentCount++;
<a name="l00281"></a>00281     }
<a name="l00282"></a>00282     <span class="keywordflow">return</span> iSegmentCount;
<a name="l00283"></a>00283 }
<a name="l00284"></a>00284 
<a name="l00285"></a>00285 
<a name="l00286"></a>00286 
<a name="l00287"></a>00287 
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:56 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
