<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/CNAnalysisMethodGenotype.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>CNAnalysisMethodGenotype.cpp</h1>  </div>
</div>
<div class="contents">
<a href="CNAnalysisMethodGenotype_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">///////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file CNAnalysisMethodGenotype.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> *</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief This file contains the CNAnalysisMethodGenotype class members.</span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 <span class="preprocessor">#include &quot;<a class="code" href="CNAnalysisMethodGenotype_8h.html" title="This header contains the CNAnalysisMethodGenotype class definition.">copynumber/CNAnalysisMethodGenotype.h</a>&quot;</span>
<a name="l00028"></a>00028 <span class="comment">//</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="AffxStatistics_8h.html" title="This header contains the AffxStatistics class definition.">util/AffxStatistics.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="comment">//</span>
<a name="l00031"></a>00031 <span class="comment"></span>
<a name="l00032"></a>00032 <span class="comment">/**</span>
<a name="l00033"></a>00033 <span class="comment"> * @brief Supply a little how/what/why about the algorithms this</span>
<a name="l00034"></a>00034 <span class="comment"> * class performs and what parameters it takes.</span>
<a name="l00035"></a>00035 <span class="comment"> * @return SelfDoc</span>
<a name="l00036"></a>00036 <span class="comment"> */</span>
<a name="l00037"></a><a class="code" href="classCNAnalysisMethodGenotype.html#ae6eaf32064ed05c733feb4536a26aace">00037</a> <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> <a class="code" href="classCNAnalysisMethodGenotype.html#ae6eaf32064ed05c733feb4536a26aace" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">CNAnalysisMethodGenotype::explainSelf</a>()
<a name="l00038"></a>00038 {
<a name="l00039"></a>00039     <a class="code" href="classCNAnalysisMethodGenotype.html" title="The Genotype analysis method.">CNAnalysisMethodGenotype</a> obj;
<a name="l00040"></a>00040     <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc;
<a name="l00041"></a>00041     doc.<a class="code" href="classSelfDoc.html#a51db8ea9abfa958f92fb5bc601ad68dc" title="Setter method for name appearing in doc.">setDocName</a>(obj.getType());
<a name="l00042"></a>00042     doc.<a class="code" href="classSelfDoc.html#acd4638b447701d1a642effb6b8d2a396" title="Setter method for description appearing in doc.">setDocDescription</a>(obj.getDescription());
<a name="l00043"></a>00043     doc.<a class="code" href="classSelfDoc.html#aa43e6721e35792413711e07bda90861d" title="Setter method for parameters and their documentation.">setDocOptions</a>(obj.<a class="code" href="classCNAnalysisMethodGenotype.html#a1a983dc76469ae9ddf01e64e3bf1befe" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>());
<a name="l00044"></a>00044     <span class="keywordflow">return</span> doc;
<a name="l00045"></a>00045 }
<a name="l00046"></a>00046 <span class="comment"></span>
<a name="l00047"></a>00047 <span class="comment">/**</span>
<a name="l00048"></a>00048 <span class="comment"> * @brief Default Getter method for parameters and their documentation.</span>
<a name="l00049"></a>00049 <span class="comment"> * @return map of parameters and their descriptions.</span>
<a name="l00050"></a>00050 <span class="comment"> */</span>
<a name="l00051"></a><a class="code" href="classCNAnalysisMethodGenotype.html#a1a983dc76469ae9ddf01e64e3bf1befe">00051</a> std::vector&lt;SelfDoc::Opt&gt; <a class="code" href="classCNAnalysisMethodGenotype.html#a1a983dc76469ae9ddf01e64e3bf1befe" title="Default Getter method for parameters and their documentation.">CNAnalysisMethodGenotype::getDefaultDocOptions</a>()
<a name="l00052"></a>00052 {
<a name="l00053"></a>00053   std::vector&lt;SelfDoc::Opt&gt; opts;
<a name="l00054"></a>00054 
<a name="l00055"></a>00055   <span class="comment">// SelfDoc::Opt(name, type, value, default, min, max, description)</span>
<a name="l00056"></a>00056 <span class="comment">/*</span>
<a name="l00057"></a>00057 <span class="comment">  example: </span>
<a name="l00058"></a>00058 <span class="comment">  SelfDoc::Opt lohCN_errorrate = {&quot;lohCN_errorrate&quot;, SelfDoc::Opt::Double, &quot;0.05&quot;, &quot;0.05&quot;, &quot;NA&quot;, &quot;NA&quot;, &quot;LOH CN Error Rate&quot;};</span>
<a name="l00059"></a>00059 <span class="comment">  opts.push_back(lohCN_errorrate);</span>
<a name="l00060"></a>00060 <span class="comment"></span>
<a name="l00061"></a>00061 <span class="comment"></span>
<a name="l00062"></a>00062 <span class="comment">*/</span>
<a name="l00063"></a>00063   <span class="keywordflow">return</span> opts;
<a name="l00064"></a>00064 }
<a name="l00065"></a>00065 <span class="comment"></span>
<a name="l00066"></a>00066 <span class="comment">/**</span>
<a name="l00067"></a>00067 <span class="comment"> * @brief This static function should be overridden by child classes</span>
<a name="l00068"></a>00068 <span class="comment"> * to return an object of the correct type initialized correctly</span>
<a name="l00069"></a>00069 <span class="comment"> * with the parameters in the string, string map. All objects</span>
<a name="l00070"></a>00070 <span class="comment"> * created this way should be deleted when finished using.</span>
<a name="l00071"></a>00071 <span class="comment"> *</span>
<a name="l00072"></a>00072 <span class="comment"> * @param param - Map of key/value pairs to initialize the object.</span>
<a name="l00073"></a>00073 <span class="comment"> *</span>
<a name="l00074"></a>00074 <span class="comment"> * @return Pointer toCreate object, this should be sub casted as necessary.</span>
<a name="l00075"></a>00075 <span class="comment"> */</span>
<a name="l00076"></a><a class="code" href="classCNAnalysisMethodGenotype.html#aa5f268aeb794415e600c82d34f755de7">00076</a> <a class="code" href="classSelfCreate.html" title="Small interface for functions that know how to make an instance of themselves given a map of key...">SelfCreate</a>* <a class="code" href="classCNAnalysisMethodGenotype.html#aa5f268aeb794415e600c82d34f755de7" title="This static function should be overridden by child classes to return an object of the correct type in...">CNAnalysisMethodGenotype::newObject</a>(std::map&lt;std::string,std::string&gt;&amp; params)
<a name="l00077"></a>00077 {
<a name="l00078"></a>00078     <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc = <a class="code" href="classCNAnalysisMethodGenotype.html#ae6eaf32064ed05c733feb4536a26aace" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">explainSelf</a>();
<a name="l00079"></a>00079     std::vector&lt;SelfDoc::Opt&gt; opts = <a class="code" href="classCNAnalysisMethodGenotype.html#a1a983dc76469ae9ddf01e64e3bf1befe" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>();
<a name="l00080"></a>00080     <a class="code" href="classCNAnalysisMethodGenotype.html" title="The Genotype analysis method.">CNAnalysisMethodGenotype</a>* pMethod = <span class="keyword">new</span> <a class="code" href="classCNAnalysisMethodGenotype.html" title="The Genotype analysis method.">CNAnalysisMethodGenotype</a>();
<a name="l00081"></a>00081     std::string strPrefix = getPrefix();
<a name="l00082"></a>00082 <span class="comment">/* example</span>
<a name="l00083"></a>00083 <span class="comment">    pMethod-&gt;m_dLohCNErrorRate = setupDoubleParameter(&quot;lohCN_errorrate&quot;, strPrefix, params, doc, opts);</span>
<a name="l00084"></a>00084 <span class="comment">*/</span>
<a name="l00085"></a>00085     <span class="keywordflow">return</span> pMethod;
<a name="l00086"></a>00086 }
<a name="l00087"></a>00087 
<a name="l00088"></a>00088 <span class="comment">//Constructor</span>
<a name="l00089"></a>00089 CNAnalysisMethodGenotype::CNAnalysisMethodGenotype()
<a name="l00090"></a>00090 {
<a name="l00091"></a>00091     m_iPrevGenderCode = -1;
<a name="l00092"></a>00092     m_iGenderCode = -1;
<a name="l00093"></a>00093     m_bLocalProbeSetsDetermined = <span class="keyword">false</span>;
<a name="l00094"></a>00094     m_iSeparation=1000000;
<a name="l00095"></a>00095 }
<a name="l00096"></a>00096 
<a name="l00097"></a>00097 
<a name="l00098"></a>00098 <span class="comment"></span>
<a name="l00099"></a>00099 <span class="comment">/**</span>
<a name="l00100"></a>00100 <span class="comment"> * @brief run the analysis</span>
<a name="l00101"></a>00101 <span class="comment"> */</span>
<a name="l00102"></a><a class="code" href="classCNAnalysisMethodGenotype.html#ae7d692b8a323ce409e13ce390f7da2cf">00102</a> <span class="keywordtype">void</span> <a class="code" href="classCNAnalysisMethodGenotype.html#ae7d692b8a323ce409e13ce390f7da2cf" title="run the analysis">CNAnalysisMethodGenotype::run</a>()
<a name="l00103"></a>00103 {
<a name="l00104"></a>00104     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;MAJOR PROGRESS UPDATE: Calculating SNP Data.&quot;</span>);
<a name="l00105"></a>00105     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;CNAnalysisMethodGenotype::run(...) start&quot;</span>);
<a name="l00106"></a>00106     isSetup();
<a name="l00107"></a>00107 
<a name="l00108"></a>00108     imputeUncalledCNstateForSNPs(getProbeSets());  <span class="comment">// Assumes that probesets are Chromosome/Position ordered.</span>
<a name="l00109"></a>00109 
<a name="l00110"></a>00110     getProbeSets()-&gt;quickSort(0);   <span class="comment">// Sorted by ProbeSet Name.  Needed to load priors. </span>
<a name="l00111"></a>00111     calculateGenotypes();
<a name="l00112"></a>00112 
<a name="l00113"></a>00113     <span class="keywordflow">if</span> (m_pEngine-&gt;getEngineName() == <span class="stringliteral">&quot;CNCytoEngine&quot;</span> &amp;&amp; !m_pEngine-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;cyto2&quot;</span>))
<a name="l00114"></a>00114     {
<a name="l00115"></a>00115         getExperiment()-&gt;setCallRate(calculateCallRate());
<a name="l00116"></a>00116     }
<a name="l00117"></a>00117 
<a name="l00118"></a>00118     getProbeSets()-&gt;quickSort(1);   <span class="comment">// sorted by Chromosome then position</span>
<a name="l00119"></a>00119 
<a name="l00120"></a>00120     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;CNAnalysisMethodGenotype::run(...) end&quot;</span>);
<a name="l00121"></a>00121 }
<a name="l00122"></a>00122 
<a name="l00123"></a>00123 
<a name="l00124"></a>00124 <span class="keywordtype">float</span> CNAnalysisMethodGenotype::calculateCallRate()
<a name="l00125"></a>00125 {
<a name="l00126"></a>00126     <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> totalCount = 0;
<a name="l00127"></a>00127     <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> callCount = 0;
<a name="l00128"></a>00128     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iProbeSetIndex = 0; iProbeSetIndex &lt; getProbeSets()-&gt;getCount(); iProbeSetIndex++)
<a name="l00129"></a>00129     {
<a name="l00130"></a>00130         <span class="comment">// the three if clauses here (isProcess, SNP, confidence)</span>
<a name="l00131"></a>00131         <span class="comment">// should sync up with if clauses in CNReporterCychp::writeGenInfoDataSet()</span>
<a name="l00132"></a>00132         <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a>* p = (*getProbeSets())[iProbeSetIndex];
<a name="l00133"></a>00133         <span class="keywordflow">if</span> (!p-&gt;isProcess()) {
<a name="l00134"></a>00134             <span class="keywordflow">continue</span>;
<a name="l00135"></a>00135         }
<a name="l00136"></a>00136         <span class="keywordflow">if</span> (p-&gt;processAsSNP())
<a name="l00137"></a>00137         {
<a name="l00138"></a>00138             totalCount++;
<a name="l00139"></a>00139             <span class="keywordtype">float</span> fConfidenceThreshold = CNAnalysisMethod::getConfidenceThreshold(m_pEngine-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;brlmmp-parameters&quot;</span>));
<a name="l00140"></a>00140             <span class="keywordflow">if</span> (p-&gt;getGenotypeConfidence() &gt;= fConfidenceThreshold) {
<a name="l00141"></a>00141                 <span class="comment">// count as ALLELE_NO_CALL</span>
<a name="l00142"></a>00142                 <span class="keywordflow">continue</span>;
<a name="l00143"></a>00143             }
<a name="l00144"></a>00144             <span class="keywordflow">if</span> (p-&gt;<a class="code" href="classCNProbeSet.html#a1c3a1c94852692fd40f7b61f373b44fb" title="Return the affx genotype call code.">getGenotypeCallCode</a>() != (char)<a class="code" href="CHPFileData_8h.html#afd94016ae5feb88f6ef46eae61af879c">ALLELE_NO_CALL</a>)
<a name="l00145"></a>00145             {
<a name="l00146"></a>00146                 callCount++;
<a name="l00147"></a>00147             }
<a name="l00148"></a>00148         }
<a name="l00149"></a>00149     }
<a name="l00150"></a>00150     <span class="keywordflow">return</span> callCount/(float)totalCount;
<a name="l00151"></a>00151 }
<a name="l00152"></a>00152 
<a name="l00153"></a>00153 
<a name="l00154"></a>00154 <span class="keywordtype">void</span> CNAnalysisMethodGenotype::imputeUncalledCNstateForSNPs(<a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a>* pvProbeSets)
<a name="l00155"></a>00155 {
<a name="l00156"></a>00156 
<a name="l00157"></a>00157     <span class="comment">//  The algorithm is to advance up the genome until we hit the X chromosome.  </span>
<a name="l00158"></a>00158     <span class="comment">//  We assume that the PAR region is a contiguous initial segment of the X chromosome </span>
<a name="l00159"></a>00159     <span class="comment">//  and we impute a CN call of 2 for all markers in this region.</span>
<a name="l00160"></a>00160     <span class="comment">//</span>
<a name="l00161"></a>00161 
<a name="l00162"></a>00162 
<a name="l00163"></a>00163     <span class="comment">//  Here we advance up to the X chromosome.</span>
<a name="l00164"></a>00164     <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a>* pobjProbeSet = NULL;
<a name="l00165"></a>00165     <span class="keywordtype">int</span> iXChrBeginIndex=-1;
<a name="l00166"></a>00166 
<a name="l00167"></a>00167     <span class="keywordtype">int</span> iIndex=0;
<a name="l00168"></a>00168     <span class="keywordflow">for</span> (; iIndex &lt; pvProbeSets-&gt;getCount(); iIndex++)  
<a name="l00169"></a>00169     {
<a name="l00170"></a>00170         pobjProbeSet=pvProbeSets-&gt;getAt(iIndex);
<a name="l00171"></a>00171         <span class="keywordflow">if</span>(pobjProbeSet-&gt;getChromosome()==m_iXChromosome)
<a name="l00172"></a>00172         {
<a name="l00173"></a>00173             iXChrBeginIndex = iIndex;
<a name="l00174"></a>00174             <span class="keywordflow">break</span>;
<a name="l00175"></a>00175         }
<a name="l00176"></a>00176         
<a name="l00177"></a>00177     }
<a name="l00178"></a>00178    
<a name="l00179"></a>00179 
<a name="l00180"></a>00180     <span class="comment">// Here we set the imputed CN state to 2 for all markers in the PAR region. </span>
<a name="l00181"></a>00181     iIndex = iXChrBeginIndex;
<a name="l00182"></a>00182     pobjProbeSet = pvProbeSets-&gt;getAt(iIndex);
<a name="l00183"></a>00183 
<a name="l00184"></a>00184     <span class="keywordflow">if</span>(!(pobjProbeSet-&gt;isPseudoAutosomalRegion()))
<a name="l00185"></a>00185     {
<a name="l00186"></a>00186         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;PAR region is not at beginning of X chromosome as assumed by genotyping algorithm.&quot;</span>);
<a name="l00187"></a>00187     }
<a name="l00188"></a>00188 
<a name="l00189"></a>00189     <span class="keywordflow">while</span>(pobjProbeSet-&gt;isPseudoAutosomalRegion())
<a name="l00190"></a>00190     {
<a name="l00191"></a>00191         pobjProbeSet-&gt;setImputedCNState(2);
<a name="l00192"></a>00192         iIndex++;
<a name="l00193"></a>00193         pobjProbeSet = pvProbeSets-&gt;getAt(iIndex);
<a name="l00194"></a>00194     }
<a name="l00195"></a>00195 
<a name="l00196"></a>00196     <span class="comment">//  Here iBeginIndex points to the first marker outside of the PAR region.</span>
<a name="l00197"></a>00197     <span class="keywordtype">int</span> iBeginIndex = iIndex;
<a name="l00198"></a>00198 
<a name="l00199"></a>00199 
<a name="l00200"></a>00200     <span class="comment">// Here we advance up the genome and check whether the separation of adjacent markers is greater that the</span>
<a name="l00201"></a>00201     <span class="comment">// distance we expect across the centromere.  If it is we we pause and impute CN states to markers up to iIndex.</span>
<a name="l00202"></a>00202     <span class="comment">// Note that iIndex points to the last element in a block, so iBeginIndex is advanced one beyond this point.</span>
<a name="l00203"></a>00203     <span class="comment">// We hit an end when iIndex is advanced to one beyond the X chromosome.  We then impute this last block outside the loop. </span>
<a name="l00204"></a>00204     <span class="comment">// One important thing to note is that we impute the upper and lower bounds of each block.</span>
<a name="l00205"></a>00205     <span class="keywordflow">while</span>(pvProbeSets-&gt;getAt(iIndex)-&gt;getChromosome() == m_iXChromosome )
<a name="l00206"></a>00206     {
<a name="l00207"></a>00207         <span class="keywordflow">if</span> ( (pvProbeSets-&gt;getAt(iIndex+1)-&gt;getPosition() - pvProbeSets-&gt;getAt(iIndex)-&gt;getPosition()) &gt; m_iSeparation)
<a name="l00208"></a>00208         {
<a name="l00209"></a>00209               impute(iBeginIndex, iIndex, pvProbeSets); 
<a name="l00210"></a>00210               iBeginIndex = iIndex+1;
<a name="l00211"></a>00211         }    
<a name="l00212"></a>00212         iIndex++;
<a name="l00213"></a>00213     } 
<a name="l00214"></a>00214     impute(iBeginIndex,iIndex-1, pvProbeSets);
<a name="l00215"></a>00215 }
<a name="l00216"></a>00216 
<a name="l00217"></a>00217 <span class="keywordtype">void</span> CNAnalysisMethodGenotype::impute(<span class="keywordtype">int</span> iBeginIndex, <span class="keywordtype">int</span> iEndIndex, <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a>* pvProbeSets )
<a name="l00218"></a>00218 {
<a name="l00219"></a>00219 
<a name="l00220"></a>00220 
<a name="l00221"></a>00221 
<a name="l00222"></a>00222     <span class="comment">//  We first advance up to the marker with the first valid CN call.</span>
<a name="l00223"></a>00223     <span class="keywordtype">int</span> iIndex = iBeginIndex;
<a name="l00224"></a>00224     <span class="keywordflow">while</span>(iIndex &lt;= iEndIndex  &amp;&amp; pvProbeSets-&gt;getAt(iIndex)-&gt;getCNState()==-1)
<a name="l00225"></a>00225     {
<a name="l00226"></a>00226        iIndex++;
<a name="l00227"></a>00227     }
<a name="l00228"></a>00228     
<a name="l00229"></a>00229     <span class="comment">// iIndex now points to the first probeset that has a valid called CN state. </span>
<a name="l00230"></a>00230     <span class="keywordtype">int</span> iFirstValidCNCall=pvProbeSets-&gt;getAt(iIndex)-&gt;getCNState();
<a name="l00231"></a>00231 
<a name="l00232"></a>00232     <span class="comment">//  Here we set all the markers up to and including the above first probeset to have an imputed</span>
<a name="l00233"></a>00233     <span class="comment">//  call equal to the called state of that first marker </span>
<a name="l00234"></a>00234     <span class="keywordflow">for</span>(<span class="keywordtype">int</span> iNewIndex=iBeginIndex; iNewIndex &lt;= iIndex; iNewIndex++)
<a name="l00235"></a>00235     {
<a name="l00236"></a>00236         pvProbeSets-&gt;getAt(iNewIndex)-&gt;setImputedCNState(iFirstValidCNCall);
<a name="l00237"></a>00237     }
<a name="l00238"></a>00238    
<a name="l00239"></a>00239 
<a name="l00240"></a>00240 
<a name="l00241"></a>00241     <span class="comment">//  iPresentBeginIndex points to the marker with a valid CN call.</span>
<a name="l00242"></a>00242     <span class="keywordtype">int</span> iPresentCNState= iFirstValidCNCall;
<a name="l00243"></a>00243     <span class="keywordtype">int</span> iPresentBeginIndex= iIndex;
<a name="l00244"></a>00244         iIndex++;
<a name="l00245"></a>00245 
<a name="l00246"></a>00246     <span class="comment">//  The idea of the algorithm is to keep advancing until we find a marker with a valid CN call. This is the second &quot;if&quot; below.</span>
<a name="l00247"></a>00247     <span class="comment">//  We then impute a CN state to all markers up to this new marker with the last CN state. This is the &quot;else&quot; below.</span>
<a name="l00248"></a>00248     <span class="comment">//  Note that at that point iIndex points to the new marker with a valid CN call, so we don&#39;t impute a call to it.</span>
<a name="l00249"></a>00249     <span class="comment">//  At the end, we will advance until iIndex points to the last marker in the block we want to impute a CN call to.</span>
<a name="l00250"></a>00250     <span class="comment">//  There are two cases.  Either it has an valid CN state or not.  If it does it is imputed with this value.  If not it</span>
<a name="l00251"></a>00251     <span class="comment">//  get the imputed state equal to the previous valid CN state called.</span>
<a name="l00252"></a>00252     <span class="keywordflow">while</span>(iIndex&lt;=iEndIndex)
<a name="l00253"></a>00253     {
<a name="l00254"></a>00254         <span class="keywordflow">if</span>(iIndex==iEndIndex)
<a name="l00255"></a>00255         {
<a name="l00256"></a>00256             <span class="keywordflow">for</span>(<span class="keywordtype">int</span> iNewIndex=iPresentBeginIndex; iNewIndex&lt;= iEndIndex-1; iNewIndex++)
<a name="l00257"></a>00257             {
<a name="l00258"></a>00258                pvProbeSets-&gt;getAt(iNewIndex)-&gt;setImputedCNState(iPresentCNState);         
<a name="l00259"></a>00259             }
<a name="l00260"></a>00260             <span class="keywordflow">if</span>(pvProbeSets-&gt;getAt(iIndex)-&gt;getCNState() != -1)
<a name="l00261"></a>00261             {
<a name="l00262"></a>00262                 pvProbeSets-&gt;getAt(iIndex)-&gt;setImputedCNState(iPresentCNState);
<a name="l00263"></a>00263             }
<a name="l00264"></a>00264             <span class="keywordflow">break</span>;
<a name="l00265"></a>00265         } 
<a name="l00266"></a>00266            
<a name="l00267"></a>00267         <span class="keywordtype">int</span> iNewCNState=pvProbeSets-&gt;getAt(iIndex)-&gt;getCNState();
<a name="l00268"></a>00268         <span class="keywordflow">if</span>(iNewCNState==-1 )
<a name="l00269"></a>00269         {
<a name="l00270"></a>00270             iIndex++;
<a name="l00271"></a>00271             <span class="keywordflow">continue</span>;
<a name="l00272"></a>00272         }
<a name="l00273"></a>00273         <span class="keywordflow">else</span>
<a name="l00274"></a>00274         {
<a name="l00275"></a>00275             <span class="keywordflow">for</span>(<span class="keywordtype">int</span> iNewIndex=iPresentBeginIndex; iNewIndex&lt;=iIndex-1; iNewIndex++)
<a name="l00276"></a>00276             {
<a name="l00277"></a>00277                pvProbeSets-&gt;getAt(iNewIndex)-&gt;setImputedCNState(iPresentCNState);         
<a name="l00278"></a>00278             }
<a name="l00279"></a>00279             iPresentCNState=pvProbeSets-&gt;getAt(iIndex)-&gt;getCNState(); 
<a name="l00280"></a>00280             iPresentBeginIndex=iIndex;
<a name="l00281"></a>00281             iIndex++; 
<a name="l00282"></a>00282                
<a name="l00283"></a>00283         }
<a name="l00284"></a>00284     }
<a name="l00285"></a>00285 }
<a name="l00286"></a>00286 
<a name="l00287"></a>00287 
<a name="l00288"></a>00288 <span class="keywordtype">void</span> CNAnalysisMethodGenotype::calculateGenotypes()
<a name="l00289"></a>00289 {
<a name="l00290"></a>00290     <span class="comment">//set(&quot;brlmmp-parameters&quot;, &quot;CM=2.bins=100.mix=1.bic=2.HARD=3.SB=0.45.KX=1.KH=1.5.KXX=0.5.KAH=-0.6.KHB=-0.6.transform=MVA.AAM=2.0.BBM=-2.0.AAV=0.06.BBV=0.06.ABV=0.06.copyqc=0.wobble=0.05.MS=1&quot;)</span>
<a name="l00291"></a>00291 
<a name="l00292"></a>00292     <span class="keywordtype">bool</span> bCyto2 = ((m_pEngine-&gt;<a class="code" href="classOptions.html#a400b67224caf01af079f1709e0e868f4" title="Check to see if an option is defined.">isOptDefined</a>(<span class="stringliteral">&quot;cyto2&quot;</span>)) &amp;&amp; (m_pEngine-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;cyto2&quot;</span>)));
<a name="l00293"></a>00293     <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> sp;
<a name="l00294"></a>00294     initializeSnpParameters(sp, 2);
<a name="l00295"></a>00295     setBrlmmpParameters(sp);
<a name="l00296"></a>00296 
<a name="l00297"></a>00297     loadSnpDistributions();
<a name="l00298"></a>00298 
<a name="l00299"></a>00299     vector&lt;affx::GType&gt; tcall(1);
<a name="l00300"></a>00300     vector&lt;double&gt; tconf(1);
<a name="l00301"></a>00301     vector&lt;double&gt; tx(1);
<a name="l00302"></a>00302     vector&lt;double&gt; ty(1);
<a name="l00303"></a>00303     vector&lt;int&gt; GenoHint(1);
<a name="l00304"></a>00304     vector&lt;double&gt; SubsetInbredPenalty(1);
<a name="l00305"></a>00305     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iProbeSetIndex = 0; (iProbeSetIndex &lt; getProbeSets()-&gt;getCount()); iProbeSetIndex++)
<a name="l00306"></a>00306     {
<a name="l00307"></a>00307         <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a>* p = getProbeSets()-&gt;getAt(iProbeSetIndex);
<a name="l00308"></a>00308         <span class="keywordflow">if</span> ((bCyto2) &amp;&amp; (p-&gt;getGenotypeCall() != -1)) {
<a name="l00309"></a>00309             <span class="keywordflow">continue</span>;
<a name="l00310"></a>00310         }
<a name="l00311"></a>00311         p-&gt;setGenotypeCall(-1);
<a name="l00312"></a>00312         p-&gt;setGenotypeConfidence(0);
<a name="l00313"></a>00313         <span class="keywordflow">if</span> (p-&gt;getSnpDistribution() != NULL)
<a name="l00314"></a>00314         {
<a name="l00315"></a>00315             <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> tsp;
<a name="l00316"></a>00316             tsp.<a class="code" href="classsnp__param.html#a73c1d4ba7af3924f444007aaa2e802ce" title="copy parameters">copy</a>(sp); <span class="comment">// duplicate current parameters to keep commands</span>
<a name="l00317"></a>00317             tsp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a> = 2; <span class="comment">// what copy for this snp so know right model centers</span>
<a name="l00318"></a>00318             <span class="keywordflow">if</span> (getExperiment()-&gt;getRawIntensityRatioGenderAsInt() == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669afdfe578a9e9044b7b274b736ed341a81" title="Has XY chromosomes.">affx::Male</a>)
<a name="l00319"></a>00319             {
<a name="l00320"></a>00320                 <span class="keywordflow">if</span> ((p-&gt;getChromosome() == m_iXChromosome) &amp;&amp; (!p-&gt;isPseudoAutosomalRegion() &amp;&amp; p-&gt;getImputedCNState() == 1)) 
<a name="l00321"></a>00321                 {
<a name="l00322"></a>00322                     tsp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a> = 1;
<a name="l00323"></a>00323                 }
<a name="l00324"></a>00324                 <span class="keywordflow">if</span> ((p-&gt;getChromosome() == m_iXChromosome) &amp;&amp; (!p-&gt;isPseudoAutosomalRegion() &amp;&amp; p-&gt;getImputedCNState() &gt;= 2 )) 
<a name="l00325"></a>00325                 {
<a name="l00326"></a>00326                     tsp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a> = 2;
<a name="l00327"></a>00327                 }
<a name="l00328"></a>00328                 <span class="keywordflow">if</span> ((p-&gt;getChromosome() == m_iXChromosome) &amp;&amp; (!p-&gt;isPseudoAutosomalRegion() &amp;&amp; p-&gt;getImputedCNState() &lt;  1)) 
<a name="l00329"></a>00329                 {
<a name="l00330"></a>00330                     tsp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a> = 2;
<a name="l00331"></a>00331                 }
<a name="l00332"></a>00332                 <span class="keywordflow">else</span> <span class="comment">// Y Chromosome, Autosome or Par </span>
<a name="l00333"></a>00333                 {
<a name="l00334"></a>00334                         <span class="keywordflow">if</span> (p-&gt;getChromosome() == m_iYChromosome) 
<a name="l00335"></a>00335                         {
<a name="l00336"></a>00336                                 tsp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a> = 1;
<a name="l00337"></a>00337                         }
<a name="l00338"></a>00338                 }
<a name="l00339"></a>00339             }
<a name="l00340"></a>00340             <span class="keywordflow">else</span>  <span class="comment">// Female</span>
<a name="l00341"></a>00341             {
<a name="l00342"></a>00342                 <span class="keywordflow">if</span> (p-&gt;getChromosome() == m_iYChromosome) {
<a name="l00343"></a>00343                     <span class="keywordflow">continue</span>;
<a name="l00344"></a>00344                 }
<a name="l00345"></a>00345             }
<a name="l00346"></a>00346             tsp.prior.Copy(*p-&gt;getSnpDistribution());
<a name="l00347"></a>00347             <span class="comment">// note: using temporary copy of main parameters tsp</span>
<a name="l00348"></a>00348             tcall[0] = affx::NN;
<a name="l00349"></a>00349             tconf[0] = 0;
<a name="l00350"></a>00350             tx[0] = p-&gt;getAAlleleSignal();
<a name="l00351"></a>00351             ty[0] = p-&gt;getBAlleleSignal();
<a name="l00352"></a>00352             GenoHint[0] = -1;
<a name="l00353"></a>00353             SubsetInbredPenalty[0] = 0; <span class="comment">// null parameter</span>
<a name="l00354"></a>00354             <a class="code" href="GenoUtility_8cpp.html#a1a2531d6033d567af0446274e94f74a4" title="transform the data vector">GenoUtility_transformData</a>(tx, ty, <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1a6db9488b39205827082e765aa0e87297" title="Minus Vs Average.">MvA</a>, 1);
<a name="l00355"></a>00355             <a class="code" href="snp_8label_8cpp.html#a1050ceb8ac15c0c39cffa4dbf76709b3" title="The master routine: takes data and turns it into genotypes with no seeds needed.">bayes_label</a>(tcall,tconf,tx,ty,GenoHint,SubsetInbredPenalty,tsp);
<a name="l00356"></a>00356             p-&gt;setGenotypeCall(<a class="code" href="namespaceaffx.html#ac9e9039b462ebd2bec646ed4c5ff1122" title="Convert a GType enum to an external integer value ///.">affx::GType_to_int</a>(tcall[0]));
<a name="l00357"></a>00357             p-&gt;setGenotypeConfidence((<span class="keywordtype">float</span>)tconf[0]);
<a name="l00358"></a>00358         }
<a name="l00359"></a>00359     }
<a name="l00360"></a>00360 }
<a name="l00361"></a>00361 
<a name="l00362"></a>00362 
<a name="l00363"></a>00363 <span class="keywordtype">void</span> CNAnalysisMethodGenotype::initializeSnpParameters(<a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a>&amp; sp, <span class="keywordtype">int</span> iCallMethod)
<a name="l00364"></a>00364 {
<a name="l00365"></a>00365     sp.<a class="code" href="classsnp__param.html#a956b62e64130e43f407662c31f54b372" title="initialize to some useful defaults">Initialize</a>();
<a name="l00366"></a>00366     sp.<a class="code" href="classsnp__param.html#ac6c59acd942ff43d8aa6baa17ad6364b" title="do we use raw labels or posterior distributions?">callmethod</a> = iCallMethod;
<a name="l00367"></a>00367     sp.<a class="code" href="classsnp__param.html#ab106a36c2ad59a00fdad233b5577d0b6" title="use quick method">bins</a> = 100;
<a name="l00368"></a>00368     sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a> = 1;
<a name="l00369"></a>00369     sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a> = 2;
<a name="l00370"></a>00370     sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a> = 3;
<a name="l00371"></a>00371     sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a> = 0.45;
<a name="l00372"></a>00372 
<a name="l00373"></a>00373     sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = 1;
<a name="l00374"></a>00374     sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = 1.5;
<a name="l00375"></a>00375     sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = 1; <span class="comment">// bb.k always should = aa.k</span>
<a name="l00376"></a>00376 
<a name="l00377"></a>00377     sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a> = 2;
<a name="l00378"></a>00378     sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a> = 0;
<a name="l00379"></a>00379     sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a> = -2;
<a name="l00380"></a>00380 
<a name="l00381"></a>00381     sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = 0.06;
<a name="l00382"></a>00382     sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = 0.06;
<a name="l00383"></a>00383     sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = 0.06;
<a name="l00384"></a>00384 
<a name="l00385"></a>00385     sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a> = 10;
<a name="l00386"></a>00386     sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a> = 10;
<a name="l00387"></a>00387     sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a> = 10;
<a name="l00388"></a>00388 
<a name="l00389"></a>00389     sp.prior.xah = -0.6;
<a name="l00390"></a>00390     sp.prior.xab = 0.5;
<a name="l00391"></a>00391     sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a> = -0.6;
<a name="l00392"></a>00392 
<a name="l00393"></a>00393     sp.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a> = 0;
<a name="l00394"></a>00394     sp.<a class="code" href="classsnp__param.html#aabee7993a0cd2545731ff91f5d6cce4e" title="how much do we allow clusters to shift from prior">wobble</a> = 0.05;
<a name="l00395"></a>00395 }
<a name="l00396"></a>00396 
<a name="l00397"></a>00397 <span class="keywordtype">void</span> CNAnalysisMethodGenotype::setBrlmmpParameters(<a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a>&amp; sp)
<a name="l00398"></a>00398 {
<a name="l00399"></a>00399     <span class="keywordtype">string</span> spec = m_pEngine-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;brlmmp-parameters&quot;</span>);
<a name="l00400"></a>00400     <span class="keywordflow">if</span> (spec.empty()) {
<a name="l00401"></a>00401         <span class="keywordflow">return</span>;
<a name="l00402"></a>00402     }
<a name="l00403"></a>00403     <span class="keywordtype">string</span> name;
<a name="l00404"></a>00404     map&lt;string, string&gt; param;
<a name="l00405"></a>00405     <a class="code" href="classSelfCreate.html#afb6e01b234ac33a843cd92cd192e30de" title="Take a string of from &amp;quot;name.param1=value1.param2=value2&amp;quot; and populate the name and paramMap...">SelfCreate::fillInNameParam</a>(spec, name, param);
<a name="l00406"></a>00406 
<a name="l00407"></a>00407     map&lt;string, string&gt;::iterator iter;
<a name="l00408"></a>00408     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;bins&quot;</span>)) != param.end()) {
<a name="l00409"></a>00409         sp.<a class="code" href="classsnp__param.html#ab106a36c2ad59a00fdad233b5577d0b6" title="use quick method">bins</a> = <a class="code" href="AffxConv_8cpp.html#ab7d46fb769602a66d9e19f96e792498d" title="Converts a string with commas into an int.">getInt</a>(iter-&gt;second);
<a name="l00410"></a>00410     }
<a name="l00411"></a>00411     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;mix&quot;</span>)) != param.end()) {
<a name="l00412"></a>00412         sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a> = <a class="code" href="AffxConv_8cpp.html#ab7d46fb769602a66d9e19f96e792498d" title="Converts a string with commas into an int.">getInt</a>(iter-&gt;second);
<a name="l00413"></a>00413     }
<a name="l00414"></a>00414     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;bic&quot;</span>)) != param.end()) {
<a name="l00415"></a>00415         sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00416"></a>00416     }
<a name="l00417"></a>00417     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;HARD&quot;</span>)) != param.end()) {
<a name="l00418"></a>00418         sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a> = <a class="code" href="AffxConv_8cpp.html#ab7d46fb769602a66d9e19f96e792498d" title="Converts a string with commas into an int.">getInt</a>(iter-&gt;second);
<a name="l00419"></a>00419     }
<a name="l00420"></a>00420     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;SB&quot;</span>)) != param.end()) {
<a name="l00421"></a>00421         sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00422"></a>00422     }
<a name="l00423"></a>00423 
<a name="l00424"></a>00424     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;KX&quot;</span>)) != param.end()) {
<a name="l00425"></a>00425         sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00426"></a>00426     }
<a name="l00427"></a>00427     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;KH&quot;</span>)) != param.end()) {
<a name="l00428"></a>00428         sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00429"></a>00429     }
<a name="l00430"></a>00430     sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>; <span class="comment">// bb.k always should = aa.k</span>
<a name="l00431"></a>00431 
<a name="l00432"></a>00432     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;AAM&quot;</span>)) != param.end()) {
<a name="l00433"></a>00433         sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00434"></a>00434     }
<a name="l00435"></a>00435     <span class="comment">//sp.prior.ab.m = 0;    // set by initializeSnpParameters() already</span>
<a name="l00436"></a>00436     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;BBM&quot;</span>)) != param.end()) {
<a name="l00437"></a>00437         sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00438"></a>00438     }
<a name="l00439"></a>00439 
<a name="l00440"></a>00440     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;AAV&quot;</span>)) != param.end()) {
<a name="l00441"></a>00441         sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00442"></a>00442     }
<a name="l00443"></a>00443     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;ABV&quot;</span>)) != param.end()) {
<a name="l00444"></a>00444         sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00445"></a>00445     }
<a name="l00446"></a>00446     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;BBV&quot;</span>)) != param.end()) {
<a name="l00447"></a>00447         sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00448"></a>00448     }
<a name="l00449"></a>00449     <span class="comment">// done by initializeSnpParameters() already</span>
<a name="l00450"></a>00450     <span class="comment">//sp.prior.aa.v = 10.0;</span>
<a name="l00451"></a>00451     <span class="comment">//sp.prior.ab.v = 10.0;</span>
<a name="l00452"></a>00452     <span class="comment">//sp.prior.bb.v = 10.0;</span>
<a name="l00453"></a>00453 
<a name="l00454"></a>00454     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;KAH&quot;</span>)) != param.end()) {
<a name="l00455"></a>00455         sp.prior.xah = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00456"></a>00456     }
<a name="l00457"></a>00457     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;KXX&quot;</span>)) != param.end()) {
<a name="l00458"></a>00458         sp.prior.xab = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00459"></a>00459     }
<a name="l00460"></a>00460     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;KHB&quot;</span>)) != param.end()) {
<a name="l00461"></a>00461         sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00462"></a>00462     }
<a name="l00463"></a>00463 
<a name="l00464"></a>00464     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;copyqc&quot;</span>)) != param.end()) {
<a name="l00465"></a>00465         sp.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00466"></a>00466     }
<a name="l00467"></a>00467     <span class="keywordflow">if</span> ((iter = param.find(<span class="stringliteral">&quot;wobble&quot;</span>)) != param.end()) {
<a name="l00468"></a>00468         sp.<a class="code" href="classsnp__param.html#aabee7993a0cd2545731ff91f5d6cce4e" title="how much do we allow clusters to shift from prior">wobble</a> = <a class="code" href="AffxConv_8cpp.html#a4e40c31006c5b45e47ea0240921b98b2" title="Converts a string with commas into a double.">getDouble</a>(iter-&gt;second);
<a name="l00469"></a>00469     }
<a name="l00470"></a>00470 }
<a name="l00471"></a>00471 <span class="keywordtype">void</span> CNAnalysisMethodGenotype::loadSnpDistributions()
<a name="l00472"></a>00472 {
<a name="l00473"></a>00473         <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> strReferenceFileName = m_pEngine-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;reference-file&quot;</span>);
<a name="l00474"></a>00474         <a class="code" href="classaffx_1_1File5__File.html">affx::File5_File</a> file5;
<a name="l00475"></a>00475         <a class="code" href="classaffx_1_1File5__Group.html">affx::File5_Group</a>* group5 = NULL;
<a name="l00476"></a>00476         <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5 = NULL;
<a name="l00477"></a>00477 
<a name="l00478"></a>00478         file5.open(strReferenceFileName, affx::FILE5_OPEN_RO);
<a name="l00479"></a>00479 
<a name="l00480"></a>00480         <span class="keywordflow">if</span> ((m_pEngine-&gt;<a class="code" href="classOptions.html#a400b67224caf01af079f1709e0e868f4" title="Check to see if an option is defined.">isOptDefined</a>(<span class="stringliteral">&quot;cyto2&quot;</span>)) &amp;&amp; (m_pEngine-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;cyto2&quot;</span>)))
<a name="l00481"></a>00481         {
<a name="l00482"></a>00482             group5 = file5.openGroup(<span class="stringliteral">&quot;Cyto2&quot;</span>, affx::FILE5_OPEN);
<a name="l00483"></a>00483             tsv5 = group5-&gt;openTsv(<span class="stringliteral">&quot;SNPPosteriors&quot;</span>, affx::FILE5_OPEN);
<a name="l00484"></a>00484         }
<a name="l00485"></a>00485         <span class="keywordflow">else</span>
<a name="l00486"></a>00486         {
<a name="l00487"></a>00487             group5 = file5.openGroup(<span class="stringliteral">&quot;CN5&quot;</span>, affx::FILE5_OPEN);
<a name="l00488"></a>00488             tsv5 = group5-&gt;openTsv(<span class="stringliteral">&quot;CN5.snp-posteriors&quot;</span>, affx::FILE5_OPEN);
<a name="l00489"></a>00489         }
<a name="l00490"></a>00490         QuantLabelZ__readSnpPriorMap_tsv5_v2(m_iXChromosome, m_iYChromosome, *getExperiment(), *getProbeSets(), tsv5);
<a name="l00491"></a>00491         tsv5-&gt;close();
<a name="l00492"></a>00492         <span class="keyword">delete</span> tsv5;
<a name="l00493"></a>00493         group5-&gt;close();
<a name="l00494"></a>00494         <span class="keyword">delete</span> group5;
<a name="l00495"></a>00495         file5.close();
<a name="l00496"></a>00496 }
<a name="l00497"></a>00497 
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:56 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
