<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/CNAnalysisMethodAllelicDifferenceCytoScan.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>CNAnalysisMethodAllelicDifferenceCytoScan.cpp</h1>  </div>
</div>
<div class="contents">
<a href="CNAnalysisMethodAllelicDifferenceCytoScan_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file CNAnalysisMethodAllelicDifferenceCytoScan.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> *</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief This file contains the CNAnalysisMethodAllelicDifferenceCytoScan class members.</span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 <span class="preprocessor">#include &quot;<a class="code" href="CNAnalysisMethodAllelicDifferenceCytoScan_8h.html" title="This header contains the CNAnalysisMethodAllelicDifferenceCytoScan class definition.">copynumber/CNAnalysisMethodAllelicDifferenceCytoScan.h</a>&quot;</span>
<a name="l00028"></a>00028 <span class="preprocessor">#include &quot;<a class="code" href="CNAnalysisMethodCovariateParams_8h.html" title="Class containing common data for covariate-based signal and log2ratio adjustment.">copynumber/CNAnalysisMethodCovariateParams.h</a>&quot;</span>
<a name="l00029"></a>00029 <span class="comment">//</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;chipstream/HomHiLoCelListener.h&quot;</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="comment">//</span>
<a name="l00033"></a>00033 <span class="comment"></span>
<a name="l00034"></a>00034 <span class="comment">/**</span>
<a name="l00035"></a>00035 <span class="comment"> * @brief Supply a little how/what/why about the algorithms this</span>
<a name="l00036"></a>00036 <span class="comment"> * class performs and what parameters it takes.</span>
<a name="l00037"></a>00037 <span class="comment"> * @return SelfDoc</span>
<a name="l00038"></a>00038 <span class="comment"> */</span>
<a name="l00039"></a><a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aa7b6c1c06c274c47364e0595f25dd057">00039</a> <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aa7b6c1c06c274c47364e0595f25dd057" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">CNAnalysisMethodAllelicDifferenceCytoScan::explainSelf</a>()
<a name="l00040"></a>00040 {
<a name="l00041"></a>00041     <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html" title="The AllelicDifference analysis method - the CytoScan variation.">CNAnalysisMethodAllelicDifferenceCytoScan</a> obj;
<a name="l00042"></a>00042     <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc;
<a name="l00043"></a>00043     doc.<a class="code" href="classSelfDoc.html#a51db8ea9abfa958f92fb5bc601ad68dc" title="Setter method for name appearing in doc.">setDocName</a>(obj.getType());
<a name="l00044"></a>00044     doc.<a class="code" href="classSelfDoc.html#acd4638b447701d1a642effb6b8d2a396" title="Setter method for description appearing in doc.">setDocDescription</a>(obj.getDescription());
<a name="l00045"></a>00045     doc.<a class="code" href="classSelfDoc.html#aa43e6721e35792413711e07bda90861d" title="Setter method for parameters and their documentation.">setDocOptions</a>(obj.<a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aca9f8ed21acdf8b5b19008fc2a930e87" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>());
<a name="l00046"></a>00046     <span class="keywordflow">return</span> doc;
<a name="l00047"></a>00047 }
<a name="l00048"></a>00048 <span class="comment"></span>
<a name="l00049"></a>00049 <span class="comment">/**</span>
<a name="l00050"></a>00050 <span class="comment"> * @brief Default Getter method for parameters and their documentation.</span>
<a name="l00051"></a>00051 <span class="comment"> * @return map of parameters and their descriptions.</span>
<a name="l00052"></a>00052 <span class="comment"> */</span>
<a name="l00053"></a><a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aca9f8ed21acdf8b5b19008fc2a930e87">00053</a> std::vector&lt;SelfDoc::Opt&gt; <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aca9f8ed21acdf8b5b19008fc2a930e87" title="Default Getter method for parameters and their documentation.">CNAnalysisMethodAllelicDifferenceCytoScan::getDefaultDocOptions</a>()
<a name="l00054"></a>00054 {
<a name="l00055"></a>00055   std::vector&lt;SelfDoc::Opt&gt; opts;
<a name="l00056"></a>00056 
<a name="l00057"></a>00057   <span class="comment">// SelfDoc::Opt(name, type, value, default, min, max, description)</span>
<a name="l00058"></a>00058 
<a name="l00059"></a>00059   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> opt1 = {<span class="stringliteral">&quot;outlier-trim&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;3.0&quot;</span>, <span class="stringliteral">&quot;3.0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;AllelicDifference Outlier Trim&quot;</span>};
<a name="l00060"></a>00060   opts.push_back(opt1);
<a name="l00061"></a>00061 
<a name="l00062"></a>00062   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> Step = {<span class="stringliteral">&quot;step&quot;</span>, SelfDoc::Opt::Integer, <span class="stringliteral">&quot;20&quot;</span>, <span class="stringliteral">&quot;20&quot;</span>, <span class="stringliteral">&quot;1&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;AllelicDifference step size&quot;</span>};
<a name="l00063"></a>00063   opts.push_back(Step);
<a name="l00064"></a>00064 
<a name="l00065"></a>00065   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> Window = {<span class="stringliteral">&quot;window&quot;</span>, SelfDoc::Opt::Integer, <span class="stringliteral">&quot;100&quot;</span>, <span class="stringliteral">&quot;100&quot;</span>, <span class="stringliteral">&quot;30&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;AllelicDifference window&quot;</span>};
<a name="l00066"></a>00066   opts.push_back(Window);
<a name="l00067"></a>00067 
<a name="l00068"></a>00068   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> PointCount = {<span class="stringliteral">&quot;point-count&quot;</span>, SelfDoc::Opt::Integer, <span class="stringliteral">&quot;128&quot;</span>, <span class="stringliteral">&quot;128&quot;</span>, <span class="stringliteral">&quot;30&quot;</span>, <span class="stringliteral">&quot;1024&quot;</span>, <span class="stringliteral">&quot;AllelicDifference number of points&quot;</span>};
<a name="l00069"></a>00069   opts.push_back(PointCount);
<a name="l00070"></a>00070 
<a name="l00071"></a>00071   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> Bandwidth = {<span class="stringliteral">&quot;bandwidth&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;0.25&quot;</span>, <span class="stringliteral">&quot;0.25&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;1&quot;</span>, <span class="stringliteral">&quot;AllelicDifference bandwidth&quot;</span>};
<a name="l00072"></a>00072   opts.push_back(Bandwidth);
<a name="l00073"></a>00073 
<a name="l00074"></a>00074   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> Cutoff = {<span class="stringliteral">&quot;cutoff&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;0.05&quot;</span>, <span class="stringliteral">&quot;0.05&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;0.5&quot;</span>, <span class="stringliteral">&quot;AllelicDifference cutoff&quot;</span>};
<a name="l00075"></a>00075   opts.push_back(Cutoff);
<a name="l00076"></a>00076 
<a name="l00077"></a>00077   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> CleanThreshold = {<span class="stringliteral">&quot;clean-threshold&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;0.35&quot;</span>, <span class="stringliteral">&quot;0.35&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="stringliteral">&quot;AllelicDifference clean threshold&quot;</span>};
<a name="l00078"></a>00078   opts.push_back(CleanThreshold);
<a name="l00079"></a>00079 
<a name="l00080"></a>00080   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> Symmetry = {<span class="stringliteral">&quot;symmetry&quot;</span>, SelfDoc::Opt::Boolean, <span class="stringliteral">&quot;true&quot;</span>, <span class="stringliteral">&quot;true&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;AllelicDifference SCAR mirror flag&quot;</span>};
<a name="l00081"></a>00081   opts.push_back(Symmetry);
<a name="l00082"></a>00082 
<a name="l00083"></a>00083   <span class="keywordflow">return</span> opts;
<a name="l00084"></a>00084 }
<a name="l00085"></a>00085 <span class="comment"></span>
<a name="l00086"></a>00086 <span class="comment">/**</span>
<a name="l00087"></a>00087 <span class="comment"> * @brief This static function should be overridden by child classes</span>
<a name="l00088"></a>00088 <span class="comment"> * to return an object of the correct type initialized correctly</span>
<a name="l00089"></a>00089 <span class="comment"> * with the parameters in the string, string map. All objects</span>
<a name="l00090"></a>00090 <span class="comment"> * created this way should be deleted when finished using.</span>
<a name="l00091"></a>00091 <span class="comment"> *</span>
<a name="l00092"></a>00092 <span class="comment"> * @param param - Map of key/value pairs to initialize the object.</span>
<a name="l00093"></a>00093 <span class="comment"> *</span>
<a name="l00094"></a>00094 <span class="comment"> * @return Pointer toCreate object, this should be sub casted as necessary.</span>
<a name="l00095"></a>00095 <span class="comment"> */</span>
<a name="l00096"></a><a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#ad6bdedc7d4e805c44e4b91fe30a34fb8">00096</a> <a class="code" href="classSelfCreate.html" title="Small interface for functions that know how to make an instance of themselves given a map of key...">SelfCreate</a>* <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#ad6bdedc7d4e805c44e4b91fe30a34fb8" title="This static function should be overridden by child classes to return an object of the correct type in...">CNAnalysisMethodAllelicDifferenceCytoScan::newObject</a>(std::map&lt;std::string,std::string&gt;&amp; params)
<a name="l00097"></a>00097 {
<a name="l00098"></a>00098     <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc = <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aa7b6c1c06c274c47364e0595f25dd057" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">explainSelf</a>();
<a name="l00099"></a>00099     std::vector&lt;SelfDoc::Opt&gt; opts = <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#aca9f8ed21acdf8b5b19008fc2a930e87" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>();
<a name="l00100"></a>00100     <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html" title="The AllelicDifference analysis method - the CytoScan variation.">CNAnalysisMethodAllelicDifferenceCytoScan</a>* pMethod = <span class="keyword">new</span> <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html" title="The AllelicDifference analysis method - the CytoScan variation.">CNAnalysisMethodAllelicDifferenceCytoScan</a>();
<a name="l00101"></a>00101     std::string strPrefix = getPrefix();
<a name="l00102"></a>00102 
<a name="l00103"></a>00103     pMethod-&gt;m_dAllelicDifferenceOutlierTrim = <a class="code" href="classCNAnalysisMethod.html#a866c7f2fb9d785c9716277a9395f1701" title="Setup a double parameter for this analysis method, including min amd max validation.">setupDoubleParameter</a>(<span class="stringliteral">&quot;outlier-trim&quot;</span>, strPrefix, params, doc, opts);
<a name="l00104"></a>00104     pMethod-&gt;m_iStep = <a class="code" href="classCNAnalysisMethod.html#a72e009aa8390f61ca119d5204db6115c" title="Setup an int parameter for this analysis method, including min amd max validation.">setupIntParameter</a>(<span class="stringliteral">&quot;step&quot;</span>, strPrefix, params, doc, opts);
<a name="l00105"></a>00105     pMethod-&gt;m_iWindow = <a class="code" href="classCNAnalysisMethod.html#a72e009aa8390f61ca119d5204db6115c" title="Setup an int parameter for this analysis method, including min amd max validation.">setupIntParameter</a>(<span class="stringliteral">&quot;window&quot;</span>, strPrefix, params, doc, opts);
<a name="l00106"></a>00106     pMethod-&gt;m_iPointCount = <a class="code" href="classCNAnalysisMethod.html#a72e009aa8390f61ca119d5204db6115c" title="Setup an int parameter for this analysis method, including min amd max validation.">setupIntParameter</a>(<span class="stringliteral">&quot;point-count&quot;</span>, strPrefix, params, doc, opts);
<a name="l00107"></a>00107     pMethod-&gt;m_dBandwidth = <a class="code" href="classCNAnalysisMethod.html#a866c7f2fb9d785c9716277a9395f1701" title="Setup a double parameter for this analysis method, including min amd max validation.">setupDoubleParameter</a>(<span class="stringliteral">&quot;bandwidth&quot;</span>, strPrefix, params, doc, opts);
<a name="l00108"></a>00108     pMethod-&gt;m_dCutoff = <a class="code" href="classCNAnalysisMethod.html#a866c7f2fb9d785c9716277a9395f1701" title="Setup a double parameter for this analysis method, including min amd max validation.">setupDoubleParameter</a>(<span class="stringliteral">&quot;cutoff&quot;</span>, strPrefix, params, doc, opts);
<a name="l00109"></a>00109     pMethod-&gt;m_dCleanThreshold = <a class="code" href="classCNAnalysisMethod.html#a866c7f2fb9d785c9716277a9395f1701" title="Setup a double parameter for this analysis method, including min amd max validation.">setupDoubleParameter</a>(<span class="stringliteral">&quot;clean-threshold&quot;</span>, strPrefix, params, doc, opts);
<a name="l00110"></a>00110     pMethod-&gt;m_bSymmetry = <a class="code" href="classCNAnalysisMethod.html#ab0a31500922c8cb3c81dbd688ab0aadb" title="Setup a bool parameter for this analysis method.">setupBoolParameter</a>(<span class="stringliteral">&quot;symmetry&quot;</span>, strPrefix, params, doc, opts);
<a name="l00111"></a>00111 
<a name="l00112"></a>00112     <span class="keywordflow">return</span> pMethod;
<a name="l00113"></a>00113 }
<a name="l00114"></a>00114 <span class="comment"></span>
<a name="l00115"></a>00115 <span class="comment">/**</span>
<a name="l00116"></a>00116 <span class="comment"> * @brief Run the analysis.</span>
<a name="l00117"></a>00117 <span class="comment"> */</span>
<a name="l00118"></a><a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#af31d6cce4ae1c4b58d656b28c2bf4854">00118</a> <span class="keywordtype">void</span> <a class="code" href="classCNAnalysisMethodAllelicDifferenceCytoScan.html#af31d6cce4ae1c4b58d656b28c2bf4854" title="Run the analysis.">CNAnalysisMethodAllelicDifferenceCytoScan::run</a>()
<a name="l00119"></a>00119 {
<a name="l00120"></a>00120     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;CNAnalysisMethodAllelicDifferenceCytoScan::run(...) start&quot;</span>);
<a name="l00121"></a>00121     isSetup();
<a name="l00122"></a>00122 <span class="comment">//    Verbose::progressBegin(1, &quot;CNAnalysisMethodAllelicDifference::run(...) &quot;, 1, 1, 1);</span>
<a name="l00123"></a>00123 <span class="comment">//    Verbose::progressStep(1);</span>
<a name="l00124"></a>00124     <a class="code" href="classCNAnalysisMethodAllelicDifference.html#a5d67a62326185433f476a9f72d906e7b" title="Calcuate the Allelic Differences.">calculateAllelicDifferences</a>();
<a name="l00125"></a>00125     removeAllelicDifferencesModulation();
<a name="l00126"></a>00126     postprocessAllelicDifferences();
<a name="l00127"></a>00127 <span class="comment">//    Verbose::progressEnd(1, &quot;Done&quot;);</span>
<a name="l00128"></a>00128     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;CNAnalysisMethodAllelicDifferenceCytoScan::run(...) end&quot;</span>);
<a name="l00129"></a>00129 }
<a name="l00130"></a>00130 
<a name="l00131"></a>00131 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::removeAllelicDifferencesModulation()
<a name="l00132"></a>00132 {
<a name="l00133"></a>00133     getEngine()-&gt;<a class="code" href="classOptions.html#aaf215b45e7097fcedae68c15073eb0f8" title="Set option value.">setOpt</a>(<span class="stringliteral">&quot;AD-modulation-no-single-peak&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00134"></a>00134 
<a name="l00135"></a>00135     <span class="keywordflow">if</span> (CovariateParams::m_vAPCovariates.empty()) {
<a name="l00136"></a>00136         <span class="keywordflow">return</span>;
<a name="l00137"></a>00137     }
<a name="l00138"></a>00138 
<a name="l00139"></a>00139     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Running allelic differences modulation correction...&quot;</span>);
<a name="l00140"></a>00140 
<a name="l00141"></a>00141     <span class="comment">// for intermediate data</span>
<a name="l00142"></a>00142     <span class="keyword">const</span> <span class="keywordtype">bool</span> keepIntermediateData = m_pEngine-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;keep-intermediate-data&quot;</span>);
<a name="l00143"></a>00143     <span class="keyword">const</span> <span class="keywordtype">double</span> xiLimit = 2.5;
<a name="l00144"></a>00144 
<a name="l00145"></a>00145     <span class="keywordtype">bool</span> AD_modulation_no_single_peak_state = <span class="keyword">false</span>;
<a name="l00146"></a>00146 
<a name="l00147"></a>00147     determineLocalProbeSets();
<a name="l00148"></a>00148     fillChrBoundsImpl(getProbeSets(), m_chrBounds);
<a name="l00149"></a>00149 
<a name="l00150"></a>00150     std::vector&lt;double&gt; weights;
<a name="l00151"></a>00151     computeWeights(getProbeSets(), weights);
<a name="l00152"></a>00152 
<a name="l00153"></a>00153     <span class="keywordtype">int</span> numCovariates = CovariateParams::m_vAPCovariates.size();
<a name="l00154"></a>00154     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iCovariateIndex = 0; iCovariateIndex &lt; numCovariates; iCovariateIndex++)
<a name="l00155"></a>00155     {
<a name="l00156"></a>00156         m_vCoarseAllelePeaks.clear();
<a name="l00157"></a>00157 
<a name="l00158"></a>00158         std::vector&lt;double&gt; allelicDiffs;
<a name="l00159"></a>00159 
<a name="l00160"></a>00160         <span class="comment">// find master peaks</span>
<a name="l00161"></a>00161         std::vector&lt;double&gt; master_xi;
<a name="l00162"></a>00162         std::vector&lt;double&gt; master_density;
<a name="l00163"></a>00163         std::vector&lt;int&gt; master_valleyIndex;
<a name="l00164"></a>00164         std::vector&lt;int&gt; master_peakIndex;
<a name="l00165"></a>00165     
<a name="l00166"></a>00166         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; getProbeSets()-&gt;size(); i++)
<a name="l00167"></a>00167         {
<a name="l00168"></a>00168             <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a> *pset = (*getProbeSets())[i];
<a name="l00169"></a>00169             allelicDiffs.push_back(pset-&gt;getAllelicDifference());
<a name="l00170"></a>00170         }
<a name="l00171"></a>00171         findDensityAndPeaks( allelicDiffs,
<a name="l00172"></a>00172                              weights,
<a name="l00173"></a>00173                              master_xi,
<a name="l00174"></a>00174                              master_density,
<a name="l00175"></a>00175                              master_valleyIndex,
<a name="l00176"></a>00176                              master_peakIndex,
<a name="l00177"></a>00177                              CovariateParams::m_masterPeakPointCount,
<a name="l00178"></a>00178                              CovariateParams::m_masterPeakBandwidth
<a name="l00179"></a>00179                              );
<a name="l00180"></a>00180 
<a name="l00181"></a>00181         <span class="comment">// remove junk peaks and valleys at the extremities</span>
<a name="l00182"></a>00182         trimToRange(master_valleyIndex, master_xi, xiLimit);
<a name="l00183"></a>00183         trimToRange(master_peakIndex, master_xi, xiLimit);
<a name="l00184"></a>00184 
<a name="l00185"></a>00185         std::vector&lt;double&gt; intermediate;
<a name="l00186"></a>00186         <span class="keywordflow">if</span> (keepIntermediateData)
<a name="l00187"></a>00187         {
<a name="l00188"></a>00188             intermediate = allelicDiffs;      <span class="comment">// save for later</span>
<a name="l00189"></a>00189             intermDensitiesFileInit(iCovariateIndex);
<a name="l00190"></a>00190             intermDensitiesOutput( master_xi,
<a name="l00191"></a>00191                                    master_density,
<a name="l00192"></a>00192                                    master_peakIndex,
<a name="l00193"></a>00193                                    <span class="stringliteral">&quot;RawMasterDensity&quot;</span>,
<a name="l00194"></a>00194                                    <span class="stringliteral">&quot;RawMasterPeaks&quot;</span>
<a name="l00195"></a>00195                                    );
<a name="l00196"></a>00196         }
<a name="l00197"></a>00197 
<a name="l00198"></a>00198         <span class="comment">// nothing else to do if only one peak (data presumed too noisy)</span>
<a name="l00199"></a>00199         <span class="keywordflow">if</span> (master_peakIndex.size() &lt;= 1)
<a name="l00200"></a>00200         {
<a name="l00201"></a>00201             <span class="keywordflow">if</span> (keepIntermediateData) {
<a name="l00202"></a>00202                 outputIntermediateDataAD(intermediate, iCovariateIndex);
<a name="l00203"></a>00203                 intermDensitiesFileClose();
<a name="l00204"></a>00204             }
<a name="l00205"></a>00205             <span class="keywordflow">break</span>;
<a name="l00206"></a>00206         }
<a name="l00207"></a>00207 
<a name="l00208"></a>00208         <span class="comment">// coarse &quot;allele peaks&quot; processing if requested for this covariate</span>
<a name="l00209"></a>00209         <span class="keywordflow">if</span> (CovariateParams::m_vCoarseAPAdjust[iCovariateIndex] == <span class="stringliteral">&quot;on&quot;</span>)
<a name="l00210"></a>00210         {
<a name="l00211"></a>00211             allelicDiffs.clear();
<a name="l00212"></a>00212             weights.clear();
<a name="l00213"></a>00213             master_xi.clear();
<a name="l00214"></a>00214             master_density.clear();
<a name="l00215"></a>00215             master_valleyIndex.clear();
<a name="l00216"></a>00216             master_peakIndex.clear();
<a name="l00217"></a>00217 
<a name="l00218"></a>00218             <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a> coarseProbeSets;
<a name="l00219"></a>00219             coarseAPProcess(iCovariateIndex);
<a name="l00220"></a>00220 
<a name="l00221"></a>00221             <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; getProbeSets()-&gt;size(); i++)
<a name="l00222"></a>00222             {
<a name="l00223"></a>00223                 <span class="keywordtype">double</span> tmp = m_vCoarseAllelePeaks[i];
<a name="l00224"></a>00224                 <span class="keywordflow">if</span> (!isNaN(tmp))
<a name="l00225"></a>00225                 {
<a name="l00226"></a>00226                     coarseProbeSets.push_back((*getProbeSets())[i]);
<a name="l00227"></a>00227                     allelicDiffs.push_back(tmp);
<a name="l00228"></a>00228                 }
<a name="l00229"></a>00229             }
<a name="l00230"></a>00230             std::vector&lt;double&gt; coarseWeights;
<a name="l00231"></a>00231             computeWeights(&amp;coarseProbeSets, coarseWeights);
<a name="l00232"></a>00232             findDensityAndPeaks( allelicDiffs,
<a name="l00233"></a>00233                                  coarseWeights,
<a name="l00234"></a>00234                                  master_xi,
<a name="l00235"></a>00235                                  master_density,
<a name="l00236"></a>00236                                  master_valleyIndex,
<a name="l00237"></a>00237                                  master_peakIndex,
<a name="l00238"></a>00238                                  CovariateParams::m_masterPeakPointCount,
<a name="l00239"></a>00239                                  CovariateParams::m_masterPeakBandwidth
<a name="l00240"></a>00240                                  );
<a name="l00241"></a>00241 
<a name="l00242"></a>00242             <span class="comment">// remove junk peaks and valleys at the extremities</span>
<a name="l00243"></a>00243             trimToRange(master_valleyIndex, master_xi, xiLimit);
<a name="l00244"></a>00244             trimToRange(master_peakIndex, master_xi, xiLimit);
<a name="l00245"></a>00245 
<a name="l00246"></a>00246             <span class="keywordflow">if</span> (keepIntermediateData)
<a name="l00247"></a>00247             {
<a name="l00248"></a>00248                 intermDensitiesOutput( master_xi,
<a name="l00249"></a>00249                                        master_density,
<a name="l00250"></a>00250                                        master_peakIndex,
<a name="l00251"></a>00251                                        <span class="stringliteral">&quot;MasterDensity&quot;</span>,
<a name="l00252"></a>00252                                        <span class="stringliteral">&quot;MasterPeaks&quot;</span>
<a name="l00253"></a>00253                                        );
<a name="l00254"></a>00254             }
<a name="l00255"></a>00255         }
<a name="l00256"></a>00256  
<a name="l00257"></a>00257         <span class="comment">// gather the covariates for binning...</span>
<a name="l00258"></a>00258         std::vector&lt;float&gt; covariates;
<a name="l00259"></a>00259         std::vector&lt;std::pair&lt;int, int&gt; &gt; binInfo;
<a name="l00260"></a>00260         <span class="keywordtype">int</span> iTransIndex = CovariateParams::m_vAPCovariates[iCovariateIndex];
<a name="l00261"></a>00261         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = 0; iIndex &lt; getProbeSets()-&gt;size(); iIndex++)
<a name="l00262"></a>00262         {
<a name="l00263"></a>00263             <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a> *pset = (*getProbeSets())[iIndex];
<a name="l00264"></a>00264             <span class="keywordtype">float</span> covValue = pset-&gt;getCovariateValue(iTransIndex);
<a name="l00265"></a>00265             <span class="keywordflow">if</span> (!isNaN(covValue)) {
<a name="l00266"></a>00266                 covariates.push_back(covValue);
<a name="l00267"></a>00267                 binInfo.push_back(std::make_pair(0, iIndex));
<a name="l00268"></a>00268             }
<a name="l00269"></a>00269         }
<a name="l00270"></a>00270 
<a name="l00271"></a>00271         <span class="comment">// all covariates are NaN, don&#39;t bother</span>
<a name="l00272"></a>00272         <span class="keywordflow">if</span> (covariates.empty())
<a name="l00273"></a>00273         {
<a name="l00274"></a>00274             <span class="keywordflow">if</span> (keepIntermediateData) {
<a name="l00275"></a>00275                 intermDensitiesFileClose();
<a name="l00276"></a>00276             }
<a name="l00277"></a>00277             <span class="keywordflow">continue</span>;
<a name="l00278"></a>00278         }
<a name="l00279"></a>00279 
<a name="l00280"></a>00280         <span class="comment">// ...and bin them</span>
<a name="l00281"></a>00281         std::vector&lt;int&gt; binning(covariates.size());
<a name="l00282"></a>00282         <span class="keywordtype">int</span> numBins = determineBins(covariates, binning, iCovariateIndex);
<a name="l00283"></a>00283 
<a name="l00284"></a>00284         <span class="comment">// fill in the binning info</span>
<a name="l00285"></a>00285         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; binInfo.size(); i++) {
<a name="l00286"></a>00286             binInfo[i].first = binning[i];
<a name="l00287"></a>00287         }
<a name="l00288"></a>00288 
<a name="l00289"></a>00289         <span class="comment">// figure out bin membership all at once</span>
<a name="l00290"></a>00290         std::sort(binInfo.begin(), binInfo.end(), binCompare());
<a name="l00291"></a>00291 
<a name="l00292"></a>00292         std::vector&lt;int&gt; binBoundaries(numBins + 1);
<a name="l00293"></a>00293         binBoundaries[0] = 0;
<a name="l00294"></a>00294         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; numBins; i++)
<a name="l00295"></a>00295         {
<a name="l00296"></a>00296             std::pair&lt;int, int&gt; seek = std::make_pair(i, 0);
<a name="l00297"></a>00297             std::vector&lt;std::pair&lt;int, int&gt; &gt;::iterator bdry =
<a name="l00298"></a>00298                         std::lower_bound(binInfo.begin(), binInfo.end(), seek, binCompare());
<a name="l00299"></a>00299 
<a name="l00300"></a>00300             binBoundaries[i] = bdry - binInfo.begin();
<a name="l00301"></a>00301         }
<a name="l00302"></a>00302         binBoundaries[numBins] = binInfo.size();   <span class="comment">// sentinel</span>
<a name="l00303"></a>00303 
<a name="l00304"></a>00304         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iBinIndex = 0; iBinIndex &lt; numBins; iBinIndex++)
<a name="l00305"></a>00305         {
<a name="l00306"></a>00306             std::vector&lt;int&gt; probeSetsInBinIdx;
<a name="l00307"></a>00307             <span class="keywordflow">for</span> (
<a name="l00308"></a>00308                 <span class="keywordtype">int</span> iIndex = binBoundaries[iBinIndex];
<a name="l00309"></a>00309                 iIndex &lt; binBoundaries[iBinIndex + 1];
<a name="l00310"></a>00310                 iIndex++
<a name="l00311"></a>00311             ){
<a name="l00312"></a>00312                 probeSetsInBinIdx.push_back( binInfo[iIndex].second );
<a name="l00313"></a>00313             }
<a name="l00314"></a>00314 
<a name="l00315"></a>00315             <span class="comment">// in the coarse peaks case this is a subset of probeSetsInBinIdx[]</span>
<a name="l00316"></a>00316             <span class="comment">// corresponding to the probe sets after the no man&#39;s land filtering.</span>
<a name="l00317"></a>00317             <span class="comment">// Otherwise, it&#39;s the same as probeSetsInBinIdx[].</span>
<a name="l00318"></a>00318             <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a> probeSetsInBin;
<a name="l00319"></a>00319             allelicDiffs.clear();
<a name="l00320"></a>00320             <span class="keywordflow">if</span> (CovariateParams::m_vCoarseAPAdjust[iCovariateIndex] == <span class="stringliteral">&quot;on&quot;</span>)
<a name="l00321"></a>00321             {
<a name="l00322"></a>00322                 <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; probeSetsInBinIdx.size(); i++) {
<a name="l00323"></a>00323                     <span class="keywordtype">double</span> tmp = m_vCoarseAllelePeaks[probeSetsInBinIdx[i]];
<a name="l00324"></a>00324                     <span class="comment">// use only filtered values (no man&#39;s land in allele peaks)</span>
<a name="l00325"></a>00325                     <span class="keywordflow">if</span> (!isNaN(tmp)) {
<a name="l00326"></a>00326                         allelicDiffs.push_back(tmp); 
<a name="l00327"></a>00327                         probeSetsInBin.push_back((*getProbeSets())[probeSetsInBinIdx[i]]);
<a name="l00328"></a>00328                     }
<a name="l00329"></a>00329                 }
<a name="l00330"></a>00330             } <span class="keywordflow">else</span> {
<a name="l00331"></a>00331                 <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; probeSetsInBinIdx.size(); i++) {
<a name="l00332"></a>00332                     allelicDiffs.push_back( (*getProbeSets())[probeSetsInBinIdx[i]]-&gt;getAllelicDifference() );
<a name="l00333"></a>00333                     probeSetsInBin.push_back((*getProbeSets())[probeSetsInBinIdx[i]]);
<a name="l00334"></a>00334                 }
<a name="l00335"></a>00335             }
<a name="l00336"></a>00336 
<a name="l00337"></a>00337             <span class="comment">// bin too small, cannot calculate the bandwidth, don&#39;t bother</span>
<a name="l00338"></a>00338             <span class="keywordflow">if</span> (probeSetsInBin.size() &lt;= 1)
<a name="l00339"></a>00339             {
<a name="l00340"></a>00340                 <span class="keywordflow">continue</span>;
<a name="l00341"></a>00341             }
<a name="l00342"></a>00342 
<a name="l00343"></a>00343             std::vector&lt;double&gt; covBin_weights;
<a name="l00344"></a>00344             computeWeights(&amp;probeSetsInBin, covBin_weights);
<a name="l00345"></a>00345 
<a name="l00346"></a>00346             <span class="comment">// find covariate bin peaks</span>
<a name="l00347"></a>00347             std::vector&lt;double&gt; covBin_xi;
<a name="l00348"></a>00348             std::vector&lt;double&gt; covBin_density;
<a name="l00349"></a>00349             std::vector&lt;int&gt; covBin_valleyIndex;
<a name="l00350"></a>00350             std::vector&lt;int&gt; covBin_peakIndex;
<a name="l00351"></a>00351 
<a name="l00352"></a>00352             findDensityAndPeaks( allelicDiffs,
<a name="l00353"></a>00353                                  covBin_weights,
<a name="l00354"></a>00354                                  covBin_xi,
<a name="l00355"></a>00355                                  covBin_density,
<a name="l00356"></a>00356                                  covBin_valleyIndex,
<a name="l00357"></a>00357                                  covBin_peakIndex,
<a name="l00358"></a>00358                                  CovariateParams::m_covariatePeakPointCount,
<a name="l00359"></a>00359                                  CovariateParams::m_covariatePeakBandwidth
<a name="l00360"></a>00360                                  );
<a name="l00361"></a>00361 
<a name="l00362"></a>00362             <span class="comment">// remove junk peaks and valleys at the extremities</span>
<a name="l00363"></a>00363             trimToRange(covBin_valleyIndex, covBin_xi, xiLimit);
<a name="l00364"></a>00364             trimToRange(covBin_peakIndex, covBin_xi, xiLimit);
<a name="l00365"></a>00365 
<a name="l00366"></a>00366             <span class="comment">// the map is: key=cov. bin peak index, value=master peak index</span>
<a name="l00367"></a>00367             std::map&lt;int, int&gt; matchedPeaks;
<a name="l00368"></a>00368 
<a name="l00369"></a>00369             matchPeaks( master_xi,
<a name="l00370"></a>00370                         master_density,
<a name="l00371"></a>00371                         master_peakIndex,
<a name="l00372"></a>00372                         covBin_xi,
<a name="l00373"></a>00373                         covBin_peakIndex,
<a name="l00374"></a>00374                         matchedPeaks
<a name="l00375"></a>00375                         );
<a name="l00376"></a>00376 
<a name="l00377"></a>00377             <span class="keywordflow">if</span> (keepIntermediateData) {
<a name="l00378"></a>00378                 std::string binIndexStr = <a class="code" href="classConvert.html#ab3aa97a2e8dbdd6fae18517ff287d812" title="Make a string from an integer.">Convert::toString</a>(iBinIndex);
<a name="l00379"></a>00379                 intermDensitiesOutput( covBin_xi,
<a name="l00380"></a>00380                                        covBin_density,
<a name="l00381"></a>00381                                        covBin_peakIndex,
<a name="l00382"></a>00382                                        <span class="stringliteral">&quot;Bin_&quot;</span> + binIndexStr + <span class="stringliteral">&quot;_Density&quot;</span>,
<a name="l00383"></a>00383                                        <span class="stringliteral">&quot;Bin_&quot;</span> + binIndexStr + <span class="stringliteral">&quot;_Peaks&quot;</span>
<a name="l00384"></a>00384                                        );
<a name="l00385"></a>00385 
<a name="l00386"></a>00386                 intermMatchedPeaks( matchedPeaks,
<a name="l00387"></a>00387                                     covBin_xi,
<a name="l00388"></a>00388                                     master_xi,
<a name="l00389"></a>00389                                     <span class="stringliteral">&quot;Bin_&quot;</span> + binIndexStr + <span class="stringliteral">&quot;_MatchedPeaks&quot;</span>
<a name="l00390"></a>00390                                     );
<a name="l00391"></a>00391             }
<a name="l00392"></a>00392 
<a name="l00393"></a>00393             <span class="comment">// do not adjust this bin if cov. bin data is too noisy (one peak or less)</span>
<a name="l00394"></a>00394             <span class="keywordflow">if</span> (matchedPeaks.size() &lt;= 1)
<a name="l00395"></a>00395             {
<a name="l00396"></a>00396                 <span class="keywordflow">continue</span>;
<a name="l00397"></a>00397             }
<a name="l00398"></a>00398 
<a name="l00399"></a>00399             adjustByMatchedPeaks( getProbeSets(), probeSetsInBinIdx, master_xi, covBin_xi, matchedPeaks );
<a name="l00400"></a>00400 
<a name="l00401"></a>00401             <span class="comment">// mark the fact that ADs were altered</span>
<a name="l00402"></a>00402             AD_modulation_no_single_peak_state = <span class="keyword">true</span>;
<a name="l00403"></a>00403         }
<a name="l00404"></a>00404 
<a name="l00405"></a>00405         <span class="keywordflow">if</span> (keepIntermediateData) {
<a name="l00406"></a>00406             outputIntermediateDataAD(intermediate, iCovariateIndex);
<a name="l00407"></a>00407             intermDensitiesFileClose();
<a name="l00408"></a>00408         }
<a name="l00409"></a>00409     }
<a name="l00410"></a>00410 
<a name="l00411"></a>00411     <span class="comment">// flag the relevant cychp header parameter telling it whether the correction was applied</span>
<a name="l00412"></a>00412     getEngine()-&gt;<a class="code" href="classOptions.html#aaf215b45e7097fcedae68c15073eb0f8" title="Set option value.">setOpt</a>(<span class="stringliteral">&quot;AD-modulation-no-single-peak&quot;</span>, AD_modulation_no_single_peak_state ? <span class="stringliteral">&quot;true&quot;</span> : <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00413"></a>00413 
<a name="l00414"></a>00414     m_vCoarseAllelePeaks.clear();                       <span class="comment">// set size=0</span>
<a name="l00415"></a>00415     std::vector&lt;double&gt;().swap(m_vCoarseAllelePeaks);   <span class="comment">// set capacity=0</span>
<a name="l00416"></a>00416 
<a name="l00417"></a>00417     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Done&quot;</span>);
<a name="l00418"></a>00418 }
<a name="l00419"></a>00419 
<a name="l00420"></a>00420 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::trimToRange(
<a name="l00421"></a>00421                                                     std::vector&lt;int&gt;&amp; indices,
<a name="l00422"></a>00422                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; xi,
<a name="l00423"></a>00423                                                     <span class="keyword">const</span> <span class="keywordtype">double</span> xiLimit
<a name="l00424"></a>00424                                                     )
<a name="l00425"></a>00425 {
<a name="l00426"></a>00426     std::vector&lt;int&gt; indices_tmp(indices);
<a name="l00427"></a>00427     indices.clear();
<a name="l00428"></a>00428 
<a name="l00429"></a>00429     <span class="comment">// ignore indices whose xi values lie outside [-xiLimit, +xiLimit]</span>
<a name="l00430"></a>00430     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; indices_tmp.size(); i++) {
<a name="l00431"></a>00431         <span class="keywordflow">if</span> (std::fabs(xi[indices_tmp[i]]) &lt;= xiLimit) {
<a name="l00432"></a>00432             indices.push_back(indices_tmp[i]);
<a name="l00433"></a>00433         }
<a name="l00434"></a>00434     }
<a name="l00435"></a>00435 }
<a name="l00436"></a>00436 
<a name="l00437"></a>00437 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::findDensityAndPeaks(
<a name="l00438"></a>00438                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; allelicDiffs,
<a name="l00439"></a>00439                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; weights,
<a name="l00440"></a>00440                                                     std::vector&lt;double&gt;&amp; xi,
<a name="l00441"></a>00441                                                     std::vector&lt;double&gt;&amp; density,
<a name="l00442"></a>00442                                                     std::vector&lt;int&gt;&amp; valleyIndex,
<a name="l00443"></a>00443                                                     std::vector&lt;int&gt;&amp; peakIndex,
<a name="l00444"></a>00444                                                     <span class="keywordtype">int</span> numPoints,
<a name="l00445"></a>00445                                                     <span class="keywordtype">double</span> bandwidth
<a name="l00446"></a>00446                                                     )
<a name="l00447"></a>00447 {
<a name="l00448"></a>00448     <span class="keyword">const</span> <span class="keywordtype">bool</span> symmetrise = <span class="keyword">true</span>;
<a name="l00449"></a>00449 
<a name="l00450"></a>00450     <span class="keywordtype">double</span> bw;
<a name="l00451"></a>00451     <span class="keywordflow">if</span> (symmetrise)
<a name="l00452"></a>00452     {
<a name="l00453"></a>00453         <span class="keywordtype">int</span> N = allelicDiffs.size();
<a name="l00454"></a>00454         std::vector&lt;double&gt; symAllelicDiffs(allelicDiffs);
<a name="l00455"></a>00455         symAllelicDiffs.resize(2*N);
<a name="l00456"></a>00456         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = N; i &lt; 2*N; i++)
<a name="l00457"></a>00457         {
<a name="l00458"></a>00458             symAllelicDiffs[i] = -allelicDiffs[i - N];
<a name="l00459"></a>00459         }
<a name="l00460"></a>00460         bw = bwnrd(symAllelicDiffs, bandwidth);
<a name="l00461"></a>00461     }
<a name="l00462"></a>00462     <span class="keywordflow">else</span> {
<a name="l00463"></a>00463         bw = bwnrd(allelicDiffs, bandwidth);
<a name="l00464"></a>00464     }
<a name="l00465"></a>00465     fitDensityCurve(allelicDiffs, weights, xi, density, numPoints, bw, symmetrise);
<a name="l00466"></a>00466 
<a name="l00467"></a>00467     std::vector&lt;double&gt; sqDensity(density.size());
<a name="l00468"></a>00468     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; density.size(); i++) {
<a name="l00469"></a>00469         sqDensity[i] = density[i] * density[i];
<a name="l00470"></a>00470     }
<a name="l00471"></a>00471     <span class="keywordtype">double</span> overallSignal = trapzoid(xi, sqDensity);
<a name="l00472"></a>00472 
<a name="l00473"></a>00473     findpeaks(valleyIndex, peakIndex, density, 0.046*overallSignal, xi);
<a name="l00474"></a>00474 }
<a name="l00475"></a>00475 
<a name="l00476"></a>00476 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::computeWeights(
<a name="l00477"></a>00477                                                     <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a> *psets,
<a name="l00478"></a>00478                                                     std::vector&lt;double&gt;&amp; weights
<a name="l00479"></a>00479                                                     )
<a name="l00480"></a>00480 {
<a name="l00481"></a>00481     <span class="keywordtype">int</span> N = psets-&gt;size();
<a name="l00482"></a>00482     weights.assign(N, 1.0/N);
<a name="l00483"></a>00483 }
<a name="l00484"></a>00484 
<a name="l00485"></a>00485 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::adjustByMatchedPeaks(
<a name="l00486"></a>00486                                                         <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a> *psets,
<a name="l00487"></a>00487                                                         <span class="keyword">const</span> std::vector&lt;int&gt;&amp; probeSetsInBinIdx,
<a name="l00488"></a>00488                                                         <span class="keyword">const</span> std::vector&lt;double&gt;&amp; master_xi,
<a name="l00489"></a>00489                                                         <span class="keyword">const</span> std::vector&lt;double&gt;&amp; covBin_xi,
<a name="l00490"></a>00490                                                         <span class="keyword">const</span> std::map&lt;int, int&gt;&amp; matchedPeaks
<a name="l00491"></a>00491                                                         )
<a name="l00492"></a>00492 {
<a name="l00493"></a>00493     <span class="keywordflow">if</span> (matchedPeaks.empty())
<a name="l00494"></a>00494     {
<a name="l00495"></a>00495         <span class="keywordflow">return</span>;
<a name="l00496"></a>00496     }
<a name="l00497"></a>00497 
<a name="l00498"></a>00498     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> ii = 0; ii &lt; probeSetsInBinIdx.size(); ii++)
<a name="l00499"></a>00499     {
<a name="l00500"></a>00500         <span class="keywordtype">int</span> iIndex = probeSetsInBinIdx[ii];
<a name="l00501"></a>00501         <span class="keywordtype">float</span> oldAD = (*psets)[iIndex]-&gt;getAllelicDifference();
<a name="l00502"></a>00502 
<a name="l00503"></a>00503         <span class="comment">// note we rely on std::map keeping keys sorted at all times</span>
<a name="l00504"></a>00504         std::map&lt;int, int&gt;::const_iterator it, prev_it;
<a name="l00505"></a>00505         <span class="keywordflow">for</span> (it = matchedPeaks.begin(); it != matchedPeaks.end(); ++it) {
<a name="l00506"></a>00506             <span class="keywordtype">double</span> cov_xi = covBin_xi[(*it).first];
<a name="l00507"></a>00507             <span class="keywordflow">if</span> (oldAD - cov_xi &lt; 0) {
<a name="l00508"></a>00508                 <span class="keywordflow">break</span>;
<a name="l00509"></a>00509             }
<a name="l00510"></a>00510             prev_it = it;
<a name="l00511"></a>00511         }
<a name="l00512"></a>00512 
<a name="l00513"></a>00513         <span class="keywordtype">float</span> newAD;
<a name="l00514"></a>00514         <span class="keywordflow">if</span> (it == matchedPeaks.begin())
<a name="l00515"></a>00515         {
<a name="l00516"></a>00516             <span class="comment">// interpolation left of the leftmost peak</span>
<a name="l00517"></a>00517             newAD = oldAD * ( master_xi[(*it).second] / covBin_xi[(*it).first] );
<a name="l00518"></a>00518         }
<a name="l00519"></a>00519         <span class="keywordflow">else</span> <span class="keywordflow">if</span> (it != matchedPeaks.end())
<a name="l00520"></a>00520         {
<a name="l00521"></a>00521             <span class="comment">// linear interpolation between peaks</span>
<a name="l00522"></a>00522             <span class="keywordtype">double</span> tt = (oldAD - covBin_xi[(*prev_it).first]) / (covBin_xi[(*it).first] - covBin_xi[(*prev_it).first]);
<a name="l00523"></a>00523             newAD = master_xi[(*prev_it).second] + tt * (master_xi[(*it).second] - master_xi[(*prev_it).second]);
<a name="l00524"></a>00524         }
<a name="l00525"></a>00525         <span class="keywordflow">else</span>
<a name="l00526"></a>00526         {
<a name="l00527"></a>00527             <span class="comment">// interpolation right of the rightmost peak</span>
<a name="l00528"></a>00528             newAD = oldAD * ( master_xi[(*prev_it).second] / covBin_xi[(*prev_it).first] );
<a name="l00529"></a>00529         }
<a name="l00530"></a>00530 
<a name="l00531"></a>00531         (*psets)[iIndex]-&gt;setAllelicDifference(newAD);
<a name="l00532"></a>00532     }
<a name="l00533"></a>00533 }
<a name="l00534"></a>00534 
<a name="l00535"></a>00535 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::matchPeaks(
<a name="l00536"></a>00536                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; master_xi,
<a name="l00537"></a>00537                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; master_density,
<a name="l00538"></a>00538                                                     <span class="keyword">const</span> std::vector&lt;int&gt;&amp; master_peakIndex,
<a name="l00539"></a>00539                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; covBin_xi,
<a name="l00540"></a>00540                                                     <span class="keyword">const</span> std::vector&lt;int&gt;&amp; covBin_peakIndex,
<a name="l00541"></a>00541                                                     std::map&lt;int, int&gt;&amp; matchedPeaks
<a name="l00542"></a>00542                                                     )
<a name="l00543"></a>00543 {
<a name="l00544"></a>00544     <span class="keyword">const</span> <span class="keywordtype">float</span> peakTrim = 2.5;
<a name="l00545"></a>00545     <span class="keyword">const</span> <span class="keywordtype">float</span> significantlyCloserRatio = 0.5;
<a name="l00546"></a>00546 
<a name="l00547"></a>00547     <span class="comment">// Throw away peaks outside the [-peakTrim, +peakTrim] range and</span>
<a name="l00548"></a>00548     <span class="comment">// insert master peak index together with corresponding density into a pair.</span>
<a name="l00549"></a>00549     <span class="comment">// This is done to enable sorting master peaks by density (aka. ranking master peaks)</span>
<a name="l00550"></a>00550     <span class="comment">// which is needed for matching.</span>
<a name="l00551"></a>00551     std::vector&lt;std::pair&lt;int, double&gt; &gt; trimmed_master_peakIndex;
<a name="l00552"></a>00552     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; master_peakIndex.size(); i++) {
<a name="l00553"></a>00553         <span class="keywordtype">int</span> j = master_peakIndex[i];
<a name="l00554"></a>00554         <span class="keywordflow">if</span> (master_xi[j] &gt;= -peakTrim &amp;&amp; master_xi[j] &lt;= peakTrim) {
<a name="l00555"></a>00555             trimmed_master_peakIndex.push_back( std::make_pair(j, master_density[j]) );
<a name="l00556"></a>00556         }
<a name="l00557"></a>00557     }
<a name="l00558"></a>00558 
<a name="l00559"></a>00559     <span class="comment">// sort trimmed_master_peakIndex in the density height (rank) order</span>
<a name="l00560"></a>00560     std::sort( trimmed_master_peakIndex.begin(),
<a name="l00561"></a>00561                trimmed_master_peakIndex.end(),
<a name="l00562"></a>00562                comparePeakRank
<a name="l00563"></a>00563                );
<a name="l00564"></a>00564 
<a name="l00565"></a>00565     <span class="comment">// for covariate bin peaks collect just the peak indices inside [-peakTrim, +peakTrim]</span>
<a name="l00566"></a>00566     std::vector&lt;int&gt; trimmed_covBin_peakIndex;
<a name="l00567"></a>00567     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; covBin_peakIndex.size(); i++) {
<a name="l00568"></a>00568         <span class="keywordtype">int</span> j = covBin_peakIndex[i];
<a name="l00569"></a>00569         <span class="keywordflow">if</span> (covBin_xi[j] &gt;= -peakTrim &amp;&amp; covBin_xi[j] &lt;= peakTrim) {
<a name="l00570"></a>00570             trimmed_covBin_peakIndex.push_back(j);
<a name="l00571"></a>00571         }
<a name="l00572"></a>00572     }
<a name="l00573"></a>00573 
<a name="l00574"></a>00574     <span class="keywordflow">if</span> (trimmed_covBin_peakIndex.empty())
<a name="l00575"></a>00575     {
<a name="l00576"></a>00576         <span class="keywordflow">return</span>;
<a name="l00577"></a>00577     }
<a name="l00578"></a>00578 
<a name="l00579"></a>00579     <span class="comment">// peak matching</span>
<a name="l00580"></a>00580     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = 0; iIndex &lt; trimmed_master_peakIndex.size(); iIndex++) {
<a name="l00581"></a>00581         <span class="comment">// find cov. bin peak closest to this master peak</span>
<a name="l00582"></a>00582         <span class="keywordtype">int</span> closestIndex = 0;
<a name="l00583"></a>00583         <span class="keywordtype">double</span> closestDist = std::fabs( covBin_xi[trimmed_covBin_peakIndex[0]] - master_xi[trimmed_master_peakIndex[iIndex].first] );
<a name="l00584"></a>00584         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> jIndex = 1; jIndex &lt; trimmed_covBin_peakIndex.size(); jIndex++)
<a name="l00585"></a>00585         {
<a name="l00586"></a>00586             <span class="keywordtype">double</span> curDist = std::fabs( covBin_xi[trimmed_covBin_peakIndex[jIndex]] - master_xi[trimmed_master_peakIndex[iIndex].first] );
<a name="l00587"></a>00587             <span class="keywordflow">if</span> (closestDist &gt; curDist)
<a name="l00588"></a>00588             {
<a name="l00589"></a>00589                 closestDist = curDist;
<a name="l00590"></a>00590                 closestIndex = jIndex;
<a name="l00591"></a>00591             }
<a name="l00592"></a>00592         }
<a name="l00593"></a>00593 
<a name="l00594"></a>00594         <span class="comment">// see if it was already matched to a master peak</span>
<a name="l00595"></a>00595         std::map&lt;int, int&gt;::iterator it = matchedPeaks.find( trimmed_covBin_peakIndex[closestIndex] );
<a name="l00596"></a>00596         <span class="keywordtype">bool</span> matched = it != matchedPeaks.end();
<a name="l00597"></a>00597         <span class="keywordflow">if</span> (matched) {
<a name="l00598"></a>00598             <span class="comment">// now see if it is &quot;significantly closer&quot; to current master peak than</span>
<a name="l00599"></a>00599             <span class="comment">// to the master peak it&#39;s been associated with, and reassign if it is indeed &quot;significantly closer&quot;</span>
<a name="l00600"></a>00600             <span class="keywordtype">double</span> distAssociated = std::fabs( covBin_xi[trimmed_covBin_peakIndex[closestIndex]] - master_xi[(*it).second] );
<a name="l00601"></a>00601             <span class="keywordtype">double</span> distCurrent  = std::fabs( covBin_xi[trimmed_covBin_peakIndex[closestIndex]] - master_xi[trimmed_master_peakIndex[iIndex].first] );
<a name="l00602"></a>00602             <span class="keywordflow">if</span> (distCurrent &lt; significantlyCloserRatio * distAssociated)
<a name="l00603"></a>00603             {
<a name="l00604"></a>00604                 matchedPeaks[trimmed_covBin_peakIndex[closestIndex]] = trimmed_master_peakIndex[iIndex].first;
<a name="l00605"></a>00605             }
<a name="l00606"></a>00606         } <span class="keywordflow">else</span> {
<a name="l00607"></a>00607             <span class="comment">// match it to current master peak</span>
<a name="l00608"></a>00608             matchedPeaks[trimmed_covBin_peakIndex[closestIndex]] = trimmed_master_peakIndex[iIndex].first;
<a name="l00609"></a>00609         }
<a name="l00610"></a>00610     }
<a name="l00611"></a>00611 }
<a name="l00612"></a>00612 
<a name="l00613"></a>00613 <span class="keywordtype">bool</span> CNAnalysisMethodAllelicDifferenceCytoScan::comparePeakRank(<span class="keyword">const</span> std::pair&lt;int, double&gt;&amp; a, <span class="keyword">const</span> std::pair&lt;int, double&gt;&amp; b)
<a name="l00614"></a>00614 {
<a name="l00615"></a>00615     <span class="keywordflow">return</span> a.second &gt; b.second;
<a name="l00616"></a>00616 }
<a name="l00617"></a>00617 
<a name="l00618"></a>00618 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::outputIntermediateDataAD(<span class="keyword">const</span> std::vector&lt;double&gt;&amp; intermediate,
<a name="l00619"></a>00619                                                                          <span class="keywordtype">int</span> iCovariateIndex
<a name="l00620"></a>00620                                                                          )
<a name="l00621"></a>00621 {
<a name="l00622"></a>00622     <a class="code" href="classaffx_1_1File5__File.html">affx::File5_File</a> file5;
<a name="l00623"></a>00623     <a class="code" href="classaffx_1_1File5__Group.html">affx::File5_Group</a>* group5;
<a name="l00624"></a>00624     <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5 = NULL;
<a name="l00625"></a>00625 
<a name="l00626"></a>00626     std::string analysisString = <span class="stringliteral">&quot;analysis&quot;</span>;
<a name="l00627"></a>00627     std::string covarString = <span class="stringliteral">&quot;.covar&quot;</span> + <a class="code" href="classConvert.html#ab3aa97a2e8dbdd6fae18517ff287d812" title="Make a string from an integer.">Convert::toString</a>(iCovariateIndex);
<a name="l00628"></a>00628     std::string strExperimentName = getExperiment()-&gt;getExperimentName();
<a name="l00629"></a>00629     std::string fileName = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(getEngine()-&gt;getOpt(<span class="stringliteral">&quot;out-dir&quot;</span>), analysisString,
<a name="l00630"></a>00630                                     strExperimentName + covarString + <span class="stringliteral">&quot;.adjustedAD.a5&quot;</span>);
<a name="l00631"></a>00631 
<a name="l00632"></a>00632     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(3, <span class="stringliteral">&quot;Writing intermediate data for &quot;</span> + fileName);
<a name="l00633"></a>00633 
<a name="l00634"></a>00634     <span class="keywordflow">try</span> {
<a name="l00635"></a>00635         file5.open(fileName, affx::FILE5_CREATE | affx::FILE5_REPLACE);
<a name="l00636"></a>00636         group5 = file5.openGroup(<span class="stringliteral">&quot;AllelicDifferences&quot;</span>, affx::FILE5_REPLACE);
<a name="l00637"></a>00637 
<a name="l00638"></a>00638         tsv5 = group5-&gt;openTsv(<span class="stringliteral">&quot;AllelicDifferences&quot;</span>, affx::FILE5_REPLACE);
<a name="l00639"></a>00639         tsv5-&gt;defineColumn(0, 0, <span class="stringliteral">&quot;ProbeSetName&quot;</span>, affx::FILE5_DTYPE_STRING, 20);
<a name="l00640"></a>00640         tsv5-&gt;defineColumn(0, 1, <span class="stringliteral">&quot;Chromosome&quot;</span>, affx::FILE5_DTYPE_INT);
<a name="l00641"></a>00641         tsv5-&gt;defineColumn(0, 2, <span class="stringliteral">&quot;Position&quot;</span>, affx::FILE5_DTYPE_INT);
<a name="l00642"></a>00642         tsv5-&gt;defineColumn(0, 3, <span class="stringliteral">&quot;OldAllelicDifference&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00643"></a>00643         tsv5-&gt;defineColumn(0, 4, <span class="stringliteral">&quot;NewAllelicDifference&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00644"></a>00644 
<a name="l00645"></a>00645         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; getProbeSets()-&gt;size(); i++)
<a name="l00646"></a>00646         {
<a name="l00647"></a>00647             tsv5-&gt;set_string(0, 0, (*getProbeSets())[i]-&gt;getProbeSetName());
<a name="l00648"></a>00648             tsv5-&gt;set_i(0, 1, (*getProbeSets())[i]-&gt;getChromosome());
<a name="l00649"></a>00649             tsv5-&gt;set_i(0, 2, (*getProbeSets())[i]-&gt;getPosition());
<a name="l00650"></a>00650             tsv5-&gt;set_d(0, 3, intermediate[i]);
<a name="l00651"></a>00651             tsv5-&gt;set_d(0, 4, (*getProbeSets())[i]-&gt;getAllelicDifference());
<a name="l00652"></a>00652             tsv5-&gt;writeLevel(0);
<a name="l00653"></a>00653         }
<a name="l00654"></a>00654         tsv5-&gt;close();
<a name="l00655"></a>00655         <span class="keyword">delete</span> tsv5;
<a name="l00656"></a>00656 
<a name="l00657"></a>00657         <span class="keyword">delete</span> group5;
<a name="l00658"></a>00658         file5.close();
<a name="l00659"></a>00659     }
<a name="l00660"></a>00660     <span class="keywordflow">catch</span>(...){ <span class="keywordflow">throw</span>(<a class="code" href="classExcept.html" title="General purpose exception error for error handling.">Except</a>(<span class="stringliteral">&quot;Cannot open file &quot;</span> + fileName + <span class="stringliteral">&quot; while attempting to dump allelic difference data to file.&quot;</span>));}
<a name="l00661"></a>00661 }
<a name="l00662"></a>00662 
<a name="l00663"></a>00663 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::intermDensitiesFileInit(
<a name="l00664"></a>00664                                                     <span class="keyword">const</span> <span class="keywordtype">int</span> iCovariateIndex
<a name="l00665"></a>00665                                                     )
<a name="l00666"></a>00666 {
<a name="l00667"></a>00667     std::string analysisString = <span class="stringliteral">&quot;analysis&quot;</span>;
<a name="l00668"></a>00668     std::string covarString = <span class="stringliteral">&quot;.covar&quot;</span> + <a class="code" href="classConvert.html#ab3aa97a2e8dbdd6fae18517ff287d812" title="Make a string from an integer.">Convert::toString</a>(iCovariateIndex);
<a name="l00669"></a>00669     std::string strExperimentName = getExperiment()-&gt;getExperimentName();
<a name="l00670"></a>00670     std::string fileName = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(getEngine()-&gt;getOpt(<span class="stringliteral">&quot;out-dir&quot;</span>), analysisString,
<a name="l00671"></a>00671                                     strExperimentName + covarString + <span class="stringliteral">&quot;.densitiesPeaks.a5&quot;</span>);
<a name="l00672"></a>00672 
<a name="l00673"></a>00673     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(3, <span class="stringliteral">&quot;Writing intermediate data for &quot;</span> + fileName);
<a name="l00674"></a>00674 
<a name="l00675"></a>00675     <span class="keywordflow">try</span> {
<a name="l00676"></a>00676         m_file5.open(fileName, affx::FILE5_CREATE | affx::FILE5_REPLACE);
<a name="l00677"></a>00677         m_group5 = m_file5.openGroup(<span class="stringliteral">&quot;DensitiesAndPeaks&quot;</span>, affx::FILE5_REPLACE);
<a name="l00678"></a>00678     }
<a name="l00679"></a>00679     <span class="keywordflow">catch</span>(...){ <span class="keywordflow">throw</span>(<a class="code" href="classExcept.html" title="General purpose exception error for error handling.">Except</a>(<span class="stringliteral">&quot;Cannot open file &quot;</span> + fileName + <span class="stringliteral">&quot; while attempting to dump density curves data to file.&quot;</span>));}
<a name="l00680"></a>00680 }
<a name="l00681"></a>00681 
<a name="l00682"></a>00682 
<a name="l00683"></a>00683 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::intermDensitiesOutput(
<a name="l00684"></a>00684                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; xi,
<a name="l00685"></a>00685                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; density,
<a name="l00686"></a>00686                                                     <span class="keyword">const</span> std::vector&lt;int&gt;&amp; peakIndex,
<a name="l00687"></a>00687                                                     <span class="keyword">const</span> std::string&amp; densityName,
<a name="l00688"></a>00688                                                     <span class="keyword">const</span> std::string&amp; peaksName
<a name="l00689"></a>00689                                                     )
<a name="l00690"></a>00690 {
<a name="l00691"></a>00691     <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5 = NULL;
<a name="l00692"></a>00692 
<a name="l00693"></a>00693     tsv5 = m_group5-&gt;openTsv(densityName.c_str(), affx::FILE5_REPLACE);
<a name="l00694"></a>00694     tsv5-&gt;defineColumn(0, 0, <span class="stringliteral">&quot;xi&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00695"></a>00695     tsv5-&gt;defineColumn(0, 1, <span class="stringliteral">&quot;DensityValue&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00696"></a>00696 
<a name="l00697"></a>00697     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; xi.size(); i++)
<a name="l00698"></a>00698     {
<a name="l00699"></a>00699         tsv5-&gt;set_d(0, 0, xi[i]);
<a name="l00700"></a>00700         tsv5-&gt;set_d(0, 1, density[i]);
<a name="l00701"></a>00701         tsv5-&gt;writeLevel(0);
<a name="l00702"></a>00702     }
<a name="l00703"></a>00703     tsv5-&gt;close();
<a name="l00704"></a>00704     <span class="keyword">delete</span> tsv5;
<a name="l00705"></a>00705 
<a name="l00706"></a>00706     tsv5 = m_group5-&gt;openTsv(peaksName.c_str(), affx::FILE5_REPLACE);
<a name="l00707"></a>00707     tsv5-&gt;defineColumn(0, 0, <span class="stringliteral">&quot;PeakIndex&quot;</span>, affx::FILE5_DTYPE_INT);
<a name="l00708"></a>00708     tsv5-&gt;defineColumn(0, 1, <span class="stringliteral">&quot;PeakLocation&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00709"></a>00709 
<a name="l00710"></a>00710     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; peakIndex.size(); i++)
<a name="l00711"></a>00711     {
<a name="l00712"></a>00712         tsv5-&gt;set_i(0, 0, peakIndex[i]);
<a name="l00713"></a>00713         tsv5-&gt;set_d(0, 1, xi[ peakIndex[i] ]);
<a name="l00714"></a>00714         tsv5-&gt;writeLevel(0);
<a name="l00715"></a>00715     }
<a name="l00716"></a>00716     tsv5-&gt;close();
<a name="l00717"></a>00717     <span class="keyword">delete</span> tsv5;
<a name="l00718"></a>00718 }
<a name="l00719"></a>00719 
<a name="l00720"></a>00720 
<a name="l00721"></a>00721 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::intermMatchedPeaks(
<a name="l00722"></a>00722                                                     <span class="keyword">const</span> std::map&lt;int, int&gt;&amp; matchedPeaks,
<a name="l00723"></a>00723                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; covBin_xi,
<a name="l00724"></a>00724                                                     <span class="keyword">const</span> std::vector&lt;double&gt;&amp; master_xi,
<a name="l00725"></a>00725                                                     <span class="keyword">const</span> std::string&amp; matchedName
<a name="l00726"></a>00726                                                     )
<a name="l00727"></a>00727 {
<a name="l00728"></a>00728     <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5 = NULL;
<a name="l00729"></a>00729 
<a name="l00730"></a>00730     tsv5 = m_group5-&gt;openTsv(matchedName.c_str(), affx::FILE5_REPLACE);
<a name="l00731"></a>00731     tsv5-&gt;defineColumn(0, 0, <span class="stringliteral">&quot;MasterPeakIndex&quot;</span>, affx::FILE5_DTYPE_INT);
<a name="l00732"></a>00732     tsv5-&gt;defineColumn(0, 1, <span class="stringliteral">&quot;BinPeakIndex&quot;</span>, affx::FILE5_DTYPE_INT);
<a name="l00733"></a>00733     tsv5-&gt;defineColumn(0, 2, <span class="stringliteral">&quot;MasterPeakLocation&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00734"></a>00734     tsv5-&gt;defineColumn(0, 3, <span class="stringliteral">&quot;BinPeakLocation&quot;</span>, affx::FILE5_DTYPE_DOUBLE);
<a name="l00735"></a>00735 
<a name="l00736"></a>00736     std::map&lt;int, int&gt;::const_iterator it;
<a name="l00737"></a>00737     <span class="keywordflow">for</span> (it = matchedPeaks.begin(); it != matchedPeaks.end(); ++it)
<a name="l00738"></a>00738     {
<a name="l00739"></a>00739         tsv5-&gt;set_i(0, 0, (*it).second);
<a name="l00740"></a>00740         tsv5-&gt;set_i(0, 1, (*it).first);
<a name="l00741"></a>00741         tsv5-&gt;set_d(0, 2, master_xi[ (*it).second ]);
<a name="l00742"></a>00742         tsv5-&gt;set_d(0, 3, covBin_xi[ (*it).first ]);
<a name="l00743"></a>00743         tsv5-&gt;writeLevel(0);
<a name="l00744"></a>00744     }
<a name="l00745"></a>00745     tsv5-&gt;close();
<a name="l00746"></a>00746     <span class="keyword">delete</span> tsv5;
<a name="l00747"></a>00747 }
<a name="l00748"></a>00748                                                     
<a name="l00749"></a>00749 
<a name="l00750"></a>00750 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::intermDensitiesFileClose()
<a name="l00751"></a>00751 {
<a name="l00752"></a>00752     <span class="keyword">delete</span> m_group5;
<a name="l00753"></a>00753     m_file5.close();
<a name="l00754"></a>00754 }
<a name="l00755"></a>00755 
<a name="l00756"></a>00756 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::coarseAPProcess(<span class="keywordtype">int</span> iCovariateIndex)
<a name="l00757"></a>00757 {
<a name="l00758"></a>00758     <span class="comment">// same parameters as for standard allelic differences postprocessing for now</span>
<a name="l00759"></a>00759     <span class="keyword">const</span> <span class="keywordtype">float</span> dThreePeakFLD_X = 0.0;
<a name="l00760"></a>00760     <span class="keyword">const</span> <span class="keywordtype">float</span> dThreePeakFLD_Y = 0.0;
<a name="l00761"></a>00761 
<a name="l00762"></a>00762     <span class="keyword">const</span> <span class="keywordtype">float</span> dFourPeakFLD_X = 0.0;
<a name="l00763"></a>00763     <span class="keyword">const</span> <span class="keywordtype">float</span> dFourPeakFLD_Y = 0.0;
<a name="l00764"></a>00764 
<a name="l00765"></a>00765     <span class="keywordtype">float</span> threePeakShrink_x[] = { 0.0, 15.0, 20.0 };
<a name="l00766"></a>00766     <span class="keywordtype">float</span> threePeakShrink_y[] = { 0.4, 0.3, 0.2 };
<a name="l00767"></a>00767 
<a name="l00768"></a>00768     <span class="keywordtype">float</span> fourPeakShrink_x[] = { 0.0, 15.0, 20.0, 25.0 };
<a name="l00769"></a>00769     <span class="keywordtype">float</span> fourPeakShrink_y[] = { 0.6, 0.5, 0.45, 0.4 };
<a name="l00770"></a>00770 
<a name="l00771"></a>00771     m_vThreePeakFLD_X.push_back(dThreePeakFLD_X);
<a name="l00772"></a>00772     m_vThreePeakFLD_Y.push_back(dThreePeakFLD_Y);
<a name="l00773"></a>00773 
<a name="l00774"></a>00774     m_vFourPeakFLD_X.push_back(dFourPeakFLD_X);
<a name="l00775"></a>00775     m_vFourPeakFLD_Y.push_back(dFourPeakFLD_Y);
<a name="l00776"></a>00776 
<a name="l00777"></a>00777     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; 3; i++) {
<a name="l00778"></a>00778         m_vThreePeakShrink_X.push_back(threePeakShrink_x[i]);
<a name="l00779"></a>00779         m_vThreePeakShrink_Y.push_back(threePeakShrink_y[i]);
<a name="l00780"></a>00780     }
<a name="l00781"></a>00781     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; 4; i++) {
<a name="l00782"></a>00782         m_vFourPeakShrink_X.push_back(fourPeakShrink_x[i]);
<a name="l00783"></a>00783         m_vFourPeakShrink_Y.push_back(fourPeakShrink_y[i]);
<a name="l00784"></a>00784     }
<a name="l00785"></a>00785  
<a name="l00786"></a>00786     <a class="code" href="structCNAnalysisMethod_1_1PeakShrinkOverride.html">CNAnalysisMethod::PeakShrinkOverride</a> paramOverride;
<a name="l00787"></a>00787 
<a name="l00788"></a>00788     paramOverride.m_iStep_override           = CovariateParams::m_vCoarseAPAdjustStep          [iCovariateIndex];
<a name="l00789"></a>00789     paramOverride.m_iWindow_override         = CovariateParams::m_vCoarseAPAdjustWindow        [iCovariateIndex];
<a name="l00790"></a>00790     paramOverride.m_iPointCount_override     = CovariateParams::m_vCoarseAPAdjustPointCount    [iCovariateIndex];
<a name="l00791"></a>00791     paramOverride.m_dBandwidth_override      = CovariateParams::m_vCoarseAPAdjustBandwidth     [iCovariateIndex];
<a name="l00792"></a>00792     paramOverride.m_dCutoff_override         = CovariateParams::m_vCoarseAPAdjustCutoff        [iCovariateIndex];
<a name="l00793"></a>00793     paramOverride.m_dCleanThreshold_override = CovariateParams::m_vCoarseAPAdjustCleanthreshold[iCovariateIndex];
<a name="l00794"></a>00794     paramOverride.m_iCovariateIndex          = iCovariateIndex;
<a name="l00795"></a>00795     paramOverride.m_bSymmetry_override       = <span class="keyword">true</span>;
<a name="l00796"></a>00796 
<a name="l00797"></a>00797     shrinkToPeaks(getProbeSets(), &amp;paramOverride);
<a name="l00798"></a>00798 }
<a name="l00799"></a>00799 
<a name="l00800"></a>00800 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::postprocessAllelicDifferences()
<a name="l00801"></a>00801 {
<a name="l00802"></a>00802     resetAllelePeakInitialValues();
<a name="l00803"></a>00803     determineLocalProbeSets();
<a name="l00804"></a>00804     fillChrBoundsImpl(getProbeSets(), m_chrBounds);
<a name="l00805"></a>00805     calculateSNPQC(getProbeSets());
<a name="l00806"></a>00806 
<a name="l00807"></a>00807     <span class="keyword">const</span> <span class="keywordtype">float</span> dThreePeakFLD_X = 0.0;
<a name="l00808"></a>00808     <span class="keyword">const</span> <span class="keywordtype">float</span> dThreePeakFLD_Y = 0.0;
<a name="l00809"></a>00809 
<a name="l00810"></a>00810     <span class="keyword">const</span> <span class="keywordtype">float</span> dFourPeakFLD_X = 0.0;
<a name="l00811"></a>00811     <span class="keyword">const</span> <span class="keywordtype">float</span> dFourPeakFLD_Y = 0.0;
<a name="l00812"></a>00812 
<a name="l00813"></a>00813     <span class="keywordtype">float</span> threePeakShrink_x[] = { 0.0, 15.0, 20.0 };
<a name="l00814"></a>00814     <span class="keywordtype">float</span> threePeakShrink_y[] = { 0.4, 0.3, 0.2 };
<a name="l00815"></a>00815 
<a name="l00816"></a>00816     <span class="keywordtype">float</span> fourPeakShrink_x[] = { 0.0, 15.0, 20.0, 25.0 };
<a name="l00817"></a>00817     <span class="keywordtype">float</span> fourPeakShrink_y[] = { 0.6, 0.5, 0.45, 0.4 };
<a name="l00818"></a>00818 
<a name="l00819"></a>00819     m_vThreePeakFLD_X.push_back(dThreePeakFLD_X);
<a name="l00820"></a>00820     m_vThreePeakFLD_Y.push_back(dThreePeakFLD_Y);
<a name="l00821"></a>00821 
<a name="l00822"></a>00822     m_vFourPeakFLD_X.push_back(dFourPeakFLD_X);
<a name="l00823"></a>00823     m_vFourPeakFLD_Y.push_back(dFourPeakFLD_Y);
<a name="l00824"></a>00824 
<a name="l00825"></a>00825     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; 3; i++) {
<a name="l00826"></a>00826         m_vThreePeakShrink_X.push_back(threePeakShrink_x[i]);
<a name="l00827"></a>00827         m_vThreePeakShrink_Y.push_back(threePeakShrink_y[i]);
<a name="l00828"></a>00828     }
<a name="l00829"></a>00829     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; 4; i++) {
<a name="l00830"></a>00830         m_vFourPeakShrink_X.push_back(fourPeakShrink_x[i]);
<a name="l00831"></a>00831         m_vFourPeakShrink_Y.push_back(fourPeakShrink_y[i]);
<a name="l00832"></a>00832     }
<a name="l00833"></a>00833 
<a name="l00834"></a>00834     shrinkToPeaks(getProbeSets());
<a name="l00835"></a>00835 }
<a name="l00836"></a>00836 
<a name="l00837"></a>00837 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::calculateSNPQC(<a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a>* vProbeSets)
<a name="l00838"></a>00838 {
<a name="l00839"></a>00839     <span class="keyword">const</span> <span class="keywordtype">double</span> <a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a> = log(2.0);
<a name="l00840"></a>00840 
<a name="l00841"></a>00841         <span class="comment">// Make a copy of pointers to probe sets sorted by name</span>
<a name="l00842"></a>00842         <a class="code" href="classCNProbeSetArray.html" title="A vector of ProbeSet Pointers.">CNProbeSetArray</a> resortedProbeSets = *vProbeSets;
<a name="l00843"></a>00843         resortedProbeSets.quickSort(0);
<a name="l00844"></a>00844 
<a name="l00845"></a>00845     <a class="code" href="classAffxArray.html">AffxArray&lt;AffxString&gt;</a> arSNPIds;
<a name="l00846"></a>00846     <span class="keywordflow">if</span> ((m_pEngine-&gt;<a class="code" href="classOptions.html#a400b67224caf01af079f1709e0e868f4" title="Check to see if an option is defined.">isOptDefined</a>(<span class="stringliteral">&quot;snp-qc-snp-list&quot;</span>)) &amp;&amp; (m_pEngine-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;snp-qc-snp-list&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>))
<a name="l00847"></a>00847     {
<a name="l00848"></a>00848       <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00849"></a>00849       tsv.<a class="code" href="classaffx_1_1TsvFile.html#a0b08c1eea21d4b6a893f57df16521a46" title="remove whitespace from value?">m_optAutoTrim</a> = <span class="keyword">true</span>;
<a name="l00850"></a>00850       tsv.<a class="code" href="classaffx_1_1TsvFile.html#a28ad46dc5cadab63edc56cb5f36a0efe" title="Opens a file as a table.">openTable</a>(m_pEngine-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;snp-qc-snp-list&quot;</span>));
<a name="l00851"></a>00851       std::string strSnp;
<a name="l00852"></a>00852       <span class="keywordflow">while</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == affx::TSV_OK) {
<a name="l00853"></a>00853         tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0,0,strSnp);
<a name="l00854"></a>00854         arSNPIds.add(<span class="keyword">new</span> <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>(strSnp));
<a name="l00855"></a>00855       }
<a name="l00856"></a>00856       tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00857"></a>00857 
<a name="l00858"></a>00858     }
<a name="l00859"></a>00859     arSNPIds.quickSort(0);
<a name="l00860"></a>00860 
<a name="l00861"></a>00861     std::vector&lt;double&gt; vData;
<a name="l00862"></a>00862 
<a name="l00863"></a>00863         <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> strReferenceFileName = m_pEngine-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;reference-file&quot;</span>);
<a name="l00864"></a>00864         <a class="code" href="classaffx_1_1File5__File.html">affx::File5_File</a> file5;
<a name="l00865"></a>00865         <a class="code" href="classaffx_1_1File5__Group.html">affx::File5_Group</a>* group5 = NULL;
<a name="l00866"></a>00866         <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5 = NULL;
<a name="l00867"></a>00867 
<a name="l00868"></a>00868         file5.open(strReferenceFileName, affx::FILE5_OPEN_RO);
<a name="l00869"></a>00869         group5 = file5.openGroup(<span class="stringliteral">&quot;CN5&quot;</span>, affx::FILE5_OPEN);
<a name="l00870"></a>00870         tsv5 = group5-&gt;openTsv(<span class="stringliteral">&quot;CN5.snp-posteriors&quot;</span>, affx::FILE5_OPEN);
<a name="l00871"></a>00871 
<a name="l00872"></a>00872         <span class="keywordflow">while</span> (tsv5-&gt;nextLine() == affx::FILE5_OK) {
<a name="l00873"></a>00873             <span class="keywordtype">string</span> probeSetName;
<a name="l00874"></a>00874             <span class="keywordtype">double</span> clusterAAmean;
<a name="l00875"></a>00875             <span class="keywordtype">double</span> clusterABmean;
<a name="l00876"></a>00876             <span class="keywordtype">double</span> clusterBBmean;
<a name="l00877"></a>00877 
<a name="l00878"></a>00878             tsv5-&gt;get(0, 0, &amp;probeSetName);
<a name="l00879"></a>00879             tsv5-&gt;get(0, 2, &amp;clusterAAmean);
<a name="l00880"></a>00880             tsv5-&gt;get(0, 6, &amp;clusterABmean);
<a name="l00881"></a>00881             tsv5-&gt;get(0, 10, &amp;clusterBBmean);
<a name="l00882"></a>00882 
<a name="l00883"></a>00883             <span class="keywordflow">if</span> (!arSNPIds.empty()) {
<a name="l00884"></a>00884                 <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> str(probeSetName);
<a name="l00885"></a>00885                 <span class="keywordflow">if</span> (arSNPIds.binarySearch(str, 0) == -1) {
<a name="l00886"></a>00886                     <span class="keywordflow">continue</span>;
<a name="l00887"></a>00887                 }
<a name="l00888"></a>00888             }
<a name="l00889"></a>00889             <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a> searchProbeSet;
<a name="l00890"></a>00890             searchProbeSet.setProbeSetName(probeSetName);
<a name="l00891"></a>00891             pair&lt;CNProbeSetArray::iterator, CNProbeSetArray::iterator&gt; range = equal_range(
<a name="l00892"></a>00892                                                                                     resortedProbeSets.begin(),
<a name="l00893"></a>00893                                                                                     resortedProbeSets.end(),
<a name="l00894"></a>00894                                                                                     &amp;searchProbeSet,
<a name="l00895"></a>00895                                                                                     <a class="code" href="structNameCompare.html">NameCompare</a>()
<a name="l00896"></a>00896                                                                                     );
<a name="l00897"></a>00897             <span class="keywordflow">if</span> (range.first == range.second) {
<a name="l00898"></a>00898                 <span class="keywordflow">continue</span>;
<a name="l00899"></a>00899             }
<a name="l00900"></a>00900 
<a name="l00901"></a>00901             <a class="code" href="classCNProbeSet.html" title="A class to store probe set data.">CNProbeSet</a>* foundProbeSet = *range.first;
<a name="l00902"></a>00902             <span class="keywordflow">if</span> (foundProbeSet-&gt;processAsVisualization())
<a name="l00903"></a>00903             {
<a name="l00904"></a>00904                 <span class="keywordtype">double</span> log2ratio = log(foundProbeSet-&gt;getAAlleleSignal()/foundProbeSet-&gt;getBAlleleSignal())/log2;
<a name="l00905"></a>00905                 <span class="keywordtype">double</span> scaledLog2Ratio = log2ratio - clusterABmean;
<a name="l00906"></a>00906                 scaledLog2Ratio /= 0.5*(abs(clusterAAmean - clusterABmean) + abs(clusterBBmean - clusterABmean));
<a name="l00907"></a>00907                 vData.push_back(scaledLog2Ratio);
<a name="l00908"></a>00908             }
<a name="l00909"></a>00909         }
<a name="l00910"></a>00910         tsv5-&gt;close();
<a name="l00911"></a>00911         <span class="keyword">delete</span> tsv5;
<a name="l00912"></a>00912 
<a name="l00913"></a>00913         group5-&gt;close();
<a name="l00914"></a>00914         <span class="keyword">delete</span> group5;
<a name="l00915"></a>00915         file5.close();
<a name="l00916"></a>00916 
<a name="l00917"></a>00917     arSNPIds.deleteAll();
<a name="l00918"></a>00918 
<a name="l00919"></a>00919     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; (i &lt; vData.size()); i++)
<a name="l00920"></a>00920     {
<a name="l00921"></a>00921         <span class="keywordflow">if</span> (vData[i] != vData[i])
<a name="l00922"></a>00922         {
<a name="l00923"></a>00923             vData[i] = 0;
<a name="l00924"></a>00924         }
<a name="l00925"></a>00925     }
<a name="l00926"></a>00926 
<a name="l00927"></a>00927     m_pobjExperiment-&gt;setSNPQC(<a class="code" href="classHomHiLoCelListener.html#abd84bc7a9c7f6fc88d98145dde03be03" title="Calculate snpqc metric.">HomHiLoCelListener::computeSNPQC</a>(vData));
<a name="l00928"></a>00928     <span class="comment">// If this is run now, it shouldn&#39;t be rerun in CNAnalysisEngine::postProcessing()</span>
<a name="l00929"></a>00929     m_pobjExperiment-&gt;setIsSNPQCset(<span class="keyword">true</span>);
<a name="l00930"></a>00930 }
<a name="l00931"></a>00931 
<a name="l00932"></a>00932 <span class="keywordtype">void</span> CNAnalysisMethodAllelicDifferenceCytoScan::ADOutlierTrim(<span class="keywordtype">double</span>&amp; ad)
<a name="l00933"></a>00933 {
<a name="l00934"></a>00934     <span class="keywordflow">if</span> (ad &gt; m_dAllelicDifferenceOutlierTrim) {
<a name="l00935"></a>00935         ad = m_dAllelicDifferenceOutlierTrim;
<a name="l00936"></a>00936     }
<a name="l00937"></a>00937     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (ad &lt; -m_dAllelicDifferenceOutlierTrim) {
<a name="l00938"></a>00938         ad = -m_dAllelicDifferenceOutlierTrim;
<a name="l00939"></a>00939     }
<a name="l00940"></a>00940 }
<a name="l00941"></a>00941 
<a name="l00942"></a>00942 <span class="keywordtype">int</span> CNAnalysisMethodAllelicDifferenceCytoScan::determineBins(vector&lt;float&gt;&amp; covValues, vector&lt;int&gt;&amp; binning, <span class="keywordtype">int</span> covariateIndex)
<a name="l00943"></a>00943 {
<a name="l00944"></a>00944     <span class="keywordflow">if</span> (CovariateParams::useEquallyPopulatedAPBins(covariateIndex))
<a name="l00945"></a>00945     {   
<a name="l00946"></a>00946         <span class="keywordtype">int</span> numBins = CovariateParams::getNumAPBins(covariateIndex);
<a name="l00947"></a>00947         <a class="code" href="classCNAnalysisMethod.html#a81d12d272d79fb2457c36fdde7de1e0c" title="Bin vValues[] into iBinCount bins containing equal number of elements.">CNAnalysisMethod::binEqualNumber</a>(covValues, binning, numBins);
<a name="l00948"></a>00948         <span class="keywordflow">return</span> numBins;
<a name="l00949"></a>00949     }       
<a name="l00950"></a>00950     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (CovariateParams::useEquallySpacedAPBins(covariateIndex))
<a name="l00951"></a>00951     {       
<a name="l00952"></a>00952         <span class="keywordtype">int</span> numBins = CovariateParams::getNumAPBins(covariateIndex);
<a name="l00953"></a>00953         <a class="code" href="classCNAnalysisMethod.html#ab54b0868e549930db0b759b60443bb87" title="Bin vValues[] into iBinCount equally spaced bins.">CNAnalysisMethod::binEqualSpacing</a>(covValues, binning, numBins);
<a name="l00954"></a>00954         <span class="keywordflow">return</span> numBins;
<a name="l00955"></a>00955     }           
<a name="l00956"></a>00956     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (CovariateParams::isAPCovariateDiscrete(covariateIndex))
<a name="l00957"></a>00957     {       
<a name="l00958"></a>00958         <span class="keywordflow">return</span> <a class="code" href="classCNAnalysisMethod.html#a66c165a2261561adeae660bef0fc1b47" title="Bin vValues[] (presumed discrete) into bins containing equal values.">CNAnalysisMethod::covariateIsBinAssignment</a>(covValues, binning);
<a name="l00959"></a>00959     }           
<a name="l00960"></a>00960     <span class="keywordflow">else</span> <span class="comment">// unknown binning</span>
<a name="l00961"></a>00961     {       
<a name="l00962"></a>00962         <span class="comment">// This is a late error.</span>
<a name="l00963"></a>00963         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Unknown allelic difference covariate binning type&quot;</span>);
<a name="l00964"></a>00964     }
<a name="l00965"></a>00965     <span class="keywordflow">return</span> 0;
<a name="l00966"></a>00966 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:55 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
