<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/CNFamilialAnalysisMethodLOD.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>CNFamilialAnalysisMethodLOD.cpp</h1>  </div>
</div>
<div class="contents">
<a href="CNFamilialAnalysisMethodLOD_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file CNFamilialAnalysisMethodLOD.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> *</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief This file contains the CNFamilialAnalysisMethodLOD class members.</span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 <span class="preprocessor">#include &quot;<a class="code" href="CNFamilialAnalysisMethodLOD_8h.html" title="This header contains the CNFamilialAnalysisMethodLOD class definition.">copynumber/CNFamilialAnalysisMethodLOD.h</a>&quot;</span>
<a name="l00028"></a>00028 <span class="comment">//</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="CytogeneticsTrioAnalysis_8h.html" title="This header contains some Trio Analysis functions for Cytogenetics.">copynumber/CytogeneticsTrioAnalysis.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="comment">//</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="PrimeEM_8h.html">algorithm/em/PrimeEM.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="StringUtils_8h.html">calvin_files/utils/src/StringUtils.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="TsvFile_8h.html" title="Headers for the TsvFile classes.">file/TsvFile/TsvFile.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="statfun_8h.html">stats/statfun.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="AffxStatistics_8h.html" title="This header contains the AffxStatistics class definition.">util/AffxStatistics.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;util/AptErrno.h&quot;</span>
<a name="l00037"></a>00037 <span class="comment">//</span>
<a name="l00038"></a>00038 <span class="comment"></span>
<a name="l00039"></a>00039 <span class="comment">/**</span>
<a name="l00040"></a>00040 <span class="comment"> * @brief Supply a little how/what/why about the algorithms this</span>
<a name="l00041"></a>00041 <span class="comment"> * class performs and what parameters it takes.</span>
<a name="l00042"></a>00042 <span class="comment"> * @return SelfDoc</span>
<a name="l00043"></a>00043 <span class="comment"> */</span>
<a name="l00044"></a><a class="code" href="classCNFamilialAnalysisMethodLOD.html#a9e1dff9d0fe3eb244bcfa2daca8054a7">00044</a> <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> <a class="code" href="classCNFamilialAnalysisMethodLOD.html#a9e1dff9d0fe3eb244bcfa2daca8054a7" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">CNFamilialAnalysisMethodLOD::explainSelf</a>()
<a name="l00045"></a>00045 {
<a name="l00046"></a>00046     <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc;
<a name="l00047"></a>00047     doc.<a class="code" href="classSelfDoc.html#a51db8ea9abfa958f92fb5bc601ad68dc" title="Setter method for name appearing in doc.">setDocName</a>(CNFamilialAnalysisMethodLOD::getType());
<a name="l00048"></a>00048     doc.<a class="code" href="classSelfDoc.html#acd4638b447701d1a642effb6b8d2a396" title="Setter method for description appearing in doc.">setDocDescription</a>(CNFamilialAnalysisMethodLOD::getDescription());
<a name="l00049"></a>00049     doc.<a class="code" href="classSelfDoc.html#aa43e6721e35792413711e07bda90861d" title="Setter method for parameters and their documentation.">setDocOptions</a>(<a class="code" href="classCNFamilialAnalysisMethodLOD.html#a1345cd2a3e696d77c62bcecd2daf743f" title="Default Getter method for parameters and their documentation.">CNFamilialAnalysisMethodLOD::getDefaultDocOptions</a>());
<a name="l00050"></a>00050     <span class="keywordflow">return</span> doc;
<a name="l00051"></a>00051 }
<a name="l00052"></a>00052 <span class="comment"></span>
<a name="l00053"></a>00053 <span class="comment">/**</span>
<a name="l00054"></a>00054 <span class="comment"> * @brief Default Getter method for parameters and their documentation.</span>
<a name="l00055"></a>00055 <span class="comment"> * @return map of parameters and their descriptions.</span>
<a name="l00056"></a>00056 <span class="comment"> */</span>
<a name="l00057"></a><a class="code" href="classCNFamilialAnalysisMethodLOD.html#a1345cd2a3e696d77c62bcecd2daf743f">00057</a> std::vector&lt;SelfDoc::Opt&gt; <a class="code" href="classCNFamilialAnalysisMethodLOD.html#a1345cd2a3e696d77c62bcecd2daf743f" title="Default Getter method for parameters and their documentation.">CNFamilialAnalysisMethodLOD::getDefaultDocOptions</a>()
<a name="l00058"></a>00058 {
<a name="l00059"></a>00059   std::vector&lt;SelfDoc::Opt&gt; opts;
<a name="l00060"></a>00060 
<a name="l00061"></a>00061   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> validityThreshold = {<span class="stringliteral">&quot;role-validity-threshold&quot;</span>, SelfDoc::Opt::Float, <span class="stringliteral">&quot;1000.0&quot;</span>, <span class="stringliteral">&quot;1000.0&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot; Paternity Role Validity Threshold&quot;</span>};
<a name="l00062"></a>00062   opts.push_back(validityThreshold);
<a name="l00063"></a>00063 
<a name="l00064"></a>00064   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> errorRate = {<span class="stringliteral">&quot;error-rate&quot;</span>, SelfDoc::Opt::Float, <span class="stringliteral">&quot;0.01&quot;</span>, <span class="stringliteral">&quot;0.01&quot;</span>, <span class="stringliteral">&quot;0.0&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="stringliteral">&quot; Paternity Error Rate&quot;</span>};
<a name="l00065"></a>00065   opts.push_back(errorRate);
<a name="l00066"></a>00066 
<a name="l00067"></a>00067   <span class="keywordflow">return</span> opts;
<a name="l00068"></a>00068 }
<a name="l00069"></a>00069 <span class="comment"></span>
<a name="l00070"></a>00070 <span class="comment">/**</span>
<a name="l00071"></a>00071 <span class="comment"> * @brief This static function should be overridden by child classes</span>
<a name="l00072"></a>00072 <span class="comment"> * to return an object of the correct type initialized correctly</span>
<a name="l00073"></a>00073 <span class="comment"> * with the parameters in the string, string map. All objects</span>
<a name="l00074"></a>00074 <span class="comment"> * created this way should be deleted when finished using.</span>
<a name="l00075"></a>00075 <span class="comment"> *</span>
<a name="l00076"></a>00076 <span class="comment"> * @param param - Map of key/value pairs to initialize the object.</span>
<a name="l00077"></a>00077 <span class="comment"> *</span>
<a name="l00078"></a>00078 <span class="comment"> * @return Pointer toCreate object, this should be sub casted as necessary.</span>
<a name="l00079"></a>00079 <span class="comment"> */</span>
<a name="l00080"></a><a class="code" href="classCNFamilialAnalysisMethodLOD.html#ae2cf4537a2e54bc2f5b3eecbda3e0a89">00080</a> <a class="code" href="classSelfCreate.html" title="Small interface for functions that know how to make an instance of themselves given a map of key...">SelfCreate</a>* <a class="code" href="classCNFamilialAnalysisMethodLOD.html#ae2cf4537a2e54bc2f5b3eecbda3e0a89" title="This static function should be overridden by child classes to return an object of the correct type in...">CNFamilialAnalysisMethodLOD::newObject</a>(std::map&lt;std::string,std::string&gt;&amp; params)
<a name="l00081"></a>00081 {
<a name="l00082"></a>00082     <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc = <a class="code" href="classCNFamilialAnalysisMethodLOD.html#a9e1dff9d0fe3eb244bcfa2daca8054a7" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">explainSelf</a>();
<a name="l00083"></a>00083     std::vector&lt;SelfDoc::Opt&gt; opts = <a class="code" href="classCNFamilialAnalysisMethodLOD.html#a1345cd2a3e696d77c62bcecd2daf743f" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>();
<a name="l00084"></a>00084     <a class="code" href="classCNFamilialAnalysisMethodLOD.html">CNFamilialAnalysisMethodLOD</a>* pMethod = <span class="keyword">new</span> <a class="code" href="classCNFamilialAnalysisMethodLOD.html#af689166bc76e639124d5657bffb45254" title="Constructor.">CNFamilialAnalysisMethodLOD</a>();
<a name="l00085"></a>00085     std::string strPrefix = getPrefix();
<a name="l00086"></a>00086     pMethod-&gt;m_fRoleValidityThreshold = <a class="code" href="classCNFamilialAnalysisMethod.html#a46374131a756ff2df8921c12d609530d" title="Setup a float parameter for this analysis method, including min amd max validation.">setupFloatParameter</a>(<span class="stringliteral">&quot;role-validity-threshold&quot;</span>, strPrefix, params, doc, opts);
<a name="l00087"></a>00087     pMethod-&gt;m_fErrorRate = <a class="code" href="classCNFamilialAnalysisMethod.html#a46374131a756ff2df8921c12d609530d" title="Setup a float parameter for this analysis method, including min amd max validation.">setupFloatParameter</a>(<span class="stringliteral">&quot;error-rate&quot;</span>, strPrefix, params, doc, opts);
<a name="l00088"></a>00088     pMethod-&gt;<a class="code" href="classSelfDoc.html#a51db8ea9abfa958f92fb5bc601ad68dc" title="Setter method for name appearing in doc.">setDocName</a>(getType());
<a name="l00089"></a>00089     pMethod-&gt;<a class="code" href="classSelfDoc.html#acd4638b447701d1a642effb6b8d2a396" title="Setter method for description appearing in doc.">setDocDescription</a>(getDescription());
<a name="l00090"></a>00090     <span class="comment">// specify Calvin groups and datasets this analysis needs to read from cychp file</span>
<a name="l00091"></a>00091     pMethod-&gt;addGroupDataset(<span class="stringliteral">&quot;ProbeSets&quot;</span>, <span class="stringliteral">&quot;CopyNumber&quot;</span>);
<a name="l00092"></a>00092     pMethod-&gt;addGroupDataset(<span class="stringliteral">&quot;AlgorithmData&quot;</span>, <span class="stringliteral">&quot;MarkerABSignal&quot;</span>);
<a name="l00093"></a>00093     pMethod-&gt;addGroupDataset(<span class="stringliteral">&quot;Genotyping&quot;</span>, <span class="stringliteral">&quot;Calls&quot;</span>);
<a name="l00094"></a>00094     <span class="keywordflow">return</span> pMethod;
<a name="l00095"></a>00095 }
<a name="l00096"></a>00096 <span class="comment"></span>
<a name="l00097"></a>00097 <span class="comment">/**</span>
<a name="l00098"></a>00098 <span class="comment"> * @brief Constructor</span>
<a name="l00099"></a>00099 <span class="comment"> */</span>
<a name="l00100"></a><a class="code" href="classCNFamilialAnalysisMethodLOD.html#af689166bc76e639124d5657bffb45254">00100</a> <a class="code" href="classCNFamilialAnalysisMethodLOD.html#af689166bc76e639124d5657bffb45254" title="Constructor.">CNFamilialAnalysisMethodLOD::CNFamilialAnalysisMethodLOD</a>()
<a name="l00101"></a>00101 {
<a name="l00102"></a>00102     m_fRoleValidityThreshold = 0.0;
<a name="l00103"></a>00103     m_fErrorRate = 0.0;
<a name="l00104"></a>00104     m_strAlleleFrequencyFileName = <span class="stringliteral">&quot;&quot;</span>;
<a name="l00105"></a>00105 }
<a name="l00106"></a>00106 
<a name="l00107"></a>00107 <span class="keywordtype">bool</span> CNFamilialAnalysisMethodLOD::readFile(<span class="keyword">const</span> <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a>&amp; strFileName) {
<a name="l00108"></a>00108   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00109"></a>00109   tsv.<a class="code" href="classaffx_1_1TsvFile.html#a0b08c1eea21d4b6a893f57df16521a46" title="remove whitespace from value?">m_optAutoTrim</a> = <span class="keyword">true</span>;
<a name="l00110"></a>00110   tsv.<a class="code" href="classaffx_1_1TsvFile.html#afc7f5801dcb18812c1929ebc461a19af" title="Set the option abortonerror ///.">setAbortOnError</a>(<span class="keyword">false</span>);
<a name="l00111"></a>00111   
<a name="l00112"></a>00112   m_vAlleleFrequencies.reserve(<a class="code" href="classaffx_1_1TsvFile.html#a1cff7b1fb49cd84b32501a8c5dfcea08" title="Count the number of total lines, assuming a text file. ///.">affx::TsvFile::getLineCountInFile</a>(strFileName, <span class="keyword">false</span>));
<a name="l00113"></a>00113 
<a name="l00114"></a>00114   <span class="keywordflow">if</span> ( tsv.<a class="code" href="classaffx_1_1TsvFile.html#a28ad46dc5cadab63edc56cb5f36a0efe" title="Opens a file as a table.">openTable</a>(strFileName) != affx::TSV_OK ) {
<a name="l00115"></a>00115     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00116"></a>00116   }
<a name="l00117"></a>00117   std::string strSnpId;
<a name="l00118"></a>00118   std::string strAAllele;
<a name="l00119"></a>00119   std::string strBAllele;
<a name="l00120"></a>00120   
<a name="l00121"></a>00121   <span class="keywordflow">while</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0)==affx::TSV_OK) {
<a name="l00122"></a>00122     <span class="keywordflow">if</span>( tsv.<a class="code" href="classaffx_1_1TsvFile.html#a94ad901d12ab05cce0f050f99665970f" title="The number of columns for this level ///.">getColumnCount</a>(0) == 3 ) {
<a name="l00123"></a>00123       tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0,0,strSnpId);
<a name="l00124"></a>00124       tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0,1,strAAllele);
<a name="l00125"></a>00125       tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0,2,strBAllele);
<a name="l00126"></a>00126       <a class="code" href="classAlleleFrequency.html" title="Class to store allele frequency data.">AlleleFrequency</a>* p = <span class="keyword">new</span> <a class="code" href="classAlleleFrequency.html" title="Class to store allele frequency data.">AlleleFrequency</a>;
<a name="l00127"></a>00127       p-&gt;SNPID = strSnpId;
<a name="l00128"></a>00128       p-&gt;AAlleleFrequency = <a class="code" href="classAffxByteArray.html" title="A class for managing large byte arrays.">AffxByteArray</a>(strAAllele).parseDouble();
<a name="l00129"></a>00129       p-&gt;BAlleleFrequency = <a class="code" href="classAffxByteArray.html" title="A class for managing large byte arrays.">AffxByteArray</a>(strBAllele).parseDouble();
<a name="l00130"></a>00130       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(4, std::string(<span class="stringliteral">&quot;strSnpId, strAAllele, strBAllele: &quot;</span>) + strSnpId + <span class="stringliteral">&quot;, &quot;</span> +  strAAllele + <span class="stringliteral">&quot;, &quot;</span> + strBAllele);
<a name="l00131"></a>00131       m_vAlleleFrequencies.push_back(p);
<a name="l00132"></a>00132     }
<a name="l00133"></a>00133   }
<a name="l00134"></a>00134   tsv.<a class="code" href="classaffx_1_1TsvFile.html#a6a397412f48d477d0e87b00f914d06f7" title="Closes the file and clears bindings and other info.">clear</a>();
<a name="l00135"></a>00135   std::sort(m_vAlleleFrequencies.begin(), m_vAlleleFrequencies.end(), <a class="code" href="structAlleleFreqComp.html" title="Comparison predicate for AlleleFrequency pointers.">AlleleFreqComp</a>());
<a name="l00136"></a>00136 
<a name="l00137"></a>00137   <span class="keywordflow">return</span> <span class="keyword">true</span>;
<a name="l00138"></a>00138 }
<a name="l00139"></a>00139 
<a name="l00140"></a>00140 <span class="keywordtype">bool</span> CNFamilialAnalysisMethodLOD::genotypeCallsAvailable(<a class="code" href="classCNCychp.html" title="A class for accessing and storing Cychp data.">CNCychp</a>&amp; cychp)
<a name="l00141"></a>00141 {
<a name="l00142"></a>00142     <span class="comment">// Elaborate as needed for future array types</span>
<a name="l00143"></a>00143     <span class="keywordflow">return</span> cychp.getCychpHeader().getArrayType() != <span class="stringliteral">&quot;Cytogenetics_Array&quot;</span>;
<a name="l00144"></a>00144 }
<a name="l00145"></a>00145 <span class="comment"></span>
<a name="l00146"></a>00146 <span class="comment">/**</span>
<a name="l00147"></a>00147 <span class="comment"> * @brief Run the analysis.</span>
<a name="l00148"></a>00148 <span class="comment"> */</span>
<a name="l00149"></a><a class="code" href="classCNFamilialAnalysisMethodLOD.html#a9543ebe3ba949d12e761f9789d54cd0a">00149</a> <span class="keywordtype">void</span> <a class="code" href="classCNFamilialAnalysisMethodLOD.html#a9543ebe3ba949d12e761f9789d54cd0a" title="Run the analysis.">CNFamilialAnalysisMethodLOD::run</a>()
<a name="l00150"></a>00150 {
<a name="l00151"></a>00151     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;CNFamilialAnalysisMethodLOD::run(...) start&quot;</span>);
<a name="l00152"></a>00152     Verbose::progressBegin(1, <span class="stringliteral">&quot;CNFamilialAnalysisMethodLOD::run(...) &quot;</span>, 5, 1, 5);
<a name="l00153"></a>00153     <a class="code" href="classCNCychp.html" title="A class for accessing and storing Cychp data.">CNCychp</a>&amp; cychpIndex  = getCychpIndex();
<a name="l00154"></a>00154     <a class="code" href="classCNCychp.html" title="A class for accessing and storing Cychp data.">CNCychp</a>&amp; cychpMother = getCychpMother();
<a name="l00155"></a>00155     <a class="code" href="classCNCychp.html" title="A class for accessing and storing Cychp data.">CNCychp</a>&amp; cychpFather = getCychpFather();
<a name="l00156"></a>00156 
<a name="l00157"></a>00157     <span class="keywordflow">if</span> (!readFile(m_strAlleleFrequencyFileName))
<a name="l00158"></a>00158     {
<a name="l00159"></a>00159         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Cannot open allele frequency file: &quot;</span> + m_strAlleleFrequencyFileName);
<a name="l00160"></a>00160     }
<a name="l00161"></a>00161     <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00162"></a>00162 
<a name="l00163"></a>00163     <span class="keyword">const</span> <span class="keywordtype">bool</span> motherPresent = cychpMother.getFileName() != <span class="stringliteral">&quot;&quot;</span>;
<a name="l00164"></a>00164     <span class="keyword">const</span> <span class="keywordtype">bool</span> fatherPresent = cychpFather.getFileName() != <span class="stringliteral">&quot;&quot;</span>;
<a name="l00165"></a>00165 
<a name="l00166"></a>00166     <span class="keywordtype">bool</span> useGenotypeCalls = genotypeCallsAvailable(cychpIndex);
<a name="l00167"></a>00167     <span class="comment">//bool useGenotypeCalls = true;</span>
<a name="l00168"></a>00168     <span class="keywordtype">double</span> errorRate;
<a name="l00169"></a>00169 
<a name="l00170"></a>00170     std::vector&lt;char&gt; vIndexGenotypeCalls;
<a name="l00171"></a>00171     std::vector&lt;char&gt; vMotherGenotypeCalls;
<a name="l00172"></a>00172     std::vector&lt;char&gt; vFatherGenotypeCalls;
<a name="l00173"></a>00173     std::vector&lt;float&gt; vAAlleleFrequencies;
<a name="l00174"></a>00174     std::vector&lt;float&gt; vBAlleleFrequencies;
<a name="l00175"></a>00175     std::vector&lt;unsigned char&gt; vChromosomes;
<a name="l00176"></a>00176 
<a name="l00177"></a>00177     std::vector&lt;char&gt; vIndexAllGenotypeCalls;
<a name="l00178"></a>00178     std::vector&lt;char&gt; vMotherAllGenotypeCalls;
<a name="l00179"></a>00179     std::vector&lt;char&gt; vFatherAllGenotypeCalls;
<a name="l00180"></a>00180     std::vector&lt;unsigned char&gt; vAllChromosomes;
<a name="l00181"></a>00181 
<a name="l00182"></a>00182     <span class="keywordflow">if</span> (!useGenotypeCalls)
<a name="l00183"></a>00183     {
<a name="l00184"></a>00184         <span class="comment">// Calls not available - try to deduce them from SCAR distributions</span>
<a name="l00185"></a>00185         std::vector&lt;double&gt; vIndexSCAR;
<a name="l00186"></a>00186         std::vector&lt;double&gt; vMotherSCAR;
<a name="l00187"></a>00187         std::vector&lt;double&gt; vFatherSCAR;
<a name="l00188"></a>00188         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iMarkerIndex = 0; (iMarkerIndex &lt; cychpIndex.getCychpProbeSetsCopyNumbers().getCount()); iMarkerIndex++)
<a name="l00189"></a>00189         {
<a name="l00190"></a>00190             <a class="code" href="classCNCychpProbeSetsCopyNumber.html" title="A class for storing Cychp ProbeSets CopyNumber data.">CNCychpProbeSetsCopyNumber</a>* pMarker = cychpIndex.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex);
<a name="l00191"></a>00191             <a class="code" href="classAlleleFrequency.html" title="Class to store allele frequency data.">AlleleFrequency</a> objSearch;
<a name="l00192"></a>00192             objSearch.SNPID = pMarker-&gt;getProbeSetName();
<a name="l00193"></a>00193             aIterPair_t range = std::equal_range(m_vAlleleFrequencies.begin(), m_vAlleleFrequencies.end(), &amp;objSearch, <a class="code" href="structAlleleFreqComp.html" title="Comparison predicate for AlleleFrequency pointers.">AlleleFreqComp</a>());
<a name="l00194"></a>00194             <span class="keywordflow">if</span> (range.first != range.second)
<a name="l00195"></a>00195             {
<a name="l00196"></a>00196                 vIndexSCAR.push_back(pMarker-&gt;getSCAR());
<a name="l00197"></a>00197                 <span class="keywordflow">if</span> (motherPresent)
<a name="l00198"></a>00198                 {
<a name="l00199"></a>00199                     vMotherSCAR.push_back(cychpMother.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex)-&gt;getSCAR());
<a name="l00200"></a>00200                 }
<a name="l00201"></a>00201                 <span class="keywordflow">if</span> (fatherPresent)
<a name="l00202"></a>00202                 {
<a name="l00203"></a>00203                     vFatherSCAR.push_back(cychpFather.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex)-&gt;getSCAR());
<a name="l00204"></a>00204                 }
<a name="l00205"></a>00205                 vAAlleleFrequencies.push_back((*range.first)-&gt;AAlleleFrequency);
<a name="l00206"></a>00206                 vBAlleleFrequencies.push_back((*range.first)-&gt;BAlleleFrequency);
<a name="l00207"></a>00207                 vChromosomes.push_back(pMarker-&gt;getChromosome());
<a name="l00208"></a>00208             }
<a name="l00209"></a>00209         }
<a name="l00210"></a>00210         <span class="keywordtype">int</span> iSCARCount = vIndexSCAR.size();
<a name="l00211"></a>00211         <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00212"></a>00212 
<a name="l00213"></a>00213         <a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a> indexEM;
<a name="l00214"></a>00214         <a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a> motherEM;
<a name="l00215"></a>00215         <a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a> fatherEM;
<a name="l00216"></a>00216 
<a name="l00217"></a>00217         <span class="comment">// 3 component mixture model fit to the SCAR data</span>
<a name="l00218"></a>00218         fit3CompModel(indexEM, vIndexSCAR);
<a name="l00219"></a>00219         vIndexSCAR.clear();
<a name="l00220"></a>00220         std::vector&lt;double&gt;().swap(vIndexSCAR);
<a name="l00221"></a>00221 
<a name="l00222"></a>00222         <span class="keywordflow">if</span> (motherPresent)
<a name="l00223"></a>00223         {
<a name="l00224"></a>00224             fit3CompModel(motherEM, vMotherSCAR);
<a name="l00225"></a>00225             vMotherSCAR.clear();
<a name="l00226"></a>00226             std::vector&lt;double&gt;().swap(vMotherSCAR);
<a name="l00227"></a>00227         }
<a name="l00228"></a>00228         <span class="keywordflow">if</span> (fatherPresent)
<a name="l00229"></a>00229         {
<a name="l00230"></a>00230             fit3CompModel(fatherEM, vFatherSCAR);
<a name="l00231"></a>00231             vFatherSCAR.clear();
<a name="l00232"></a>00232             std::vector&lt;double&gt;().swap(vFatherSCAR);
<a name="l00233"></a>00233         }
<a name="l00234"></a>00234 
<a name="l00235"></a>00235         <span class="keywordtype">int</span> gLen = indexEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.size();
<a name="l00236"></a>00236         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iSnpIndex = 0; iSnpIndex &lt; gLen; iSnpIndex++)
<a name="l00237"></a>00237         {
<a name="l00238"></a>00238             vIndexGenotypeCalls.push_back(indexEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>[iSnpIndex]);
<a name="l00239"></a>00239             <span class="keywordflow">if</span> (motherPresent)
<a name="l00240"></a>00240             {
<a name="l00241"></a>00241                 vMotherGenotypeCalls.push_back(motherEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>[iSnpIndex]);
<a name="l00242"></a>00242             }
<a name="l00243"></a>00243             <span class="keywordflow">if</span> (fatherPresent)
<a name="l00244"></a>00244             {
<a name="l00245"></a>00245                 vFatherGenotypeCalls.push_back(fatherEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>[iSnpIndex]);
<a name="l00246"></a>00246             }
<a name="l00247"></a>00247         }
<a name="l00248"></a>00248         errorRate = estMaxErrorRate(indexEM, motherEM, fatherEM);
<a name="l00249"></a>00249     }
<a name="l00250"></a>00250     <span class="keywordflow">else</span> {
<a name="l00251"></a>00251         <span class="comment">// Genotype calls available - use them</span>
<a name="l00252"></a>00252         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iMarkerIndex = 0; iMarkerIndex &lt; cychpIndex.getCychpProbeSetsCopyNumbers().getCount(); iMarkerIndex++)
<a name="l00253"></a>00253         {
<a name="l00254"></a>00254             <a class="code" href="classCNCychpProbeSetsCopyNumber.html" title="A class for storing Cychp ProbeSets CopyNumber data.">CNCychpProbeSetsCopyNumber</a>* pMarker = cychpIndex.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex);
<a name="l00255"></a>00255             <span class="keywordflow">if</span> (!pMarker-&gt;getIsGenotypeCallSet())
<a name="l00256"></a>00256             {
<a name="l00257"></a>00257                 <span class="keywordflow">continue</span>;
<a name="l00258"></a>00258             }
<a name="l00259"></a>00259             vIndexAllGenotypeCalls.push_back(pMarker-&gt;getGenotypeCall());
<a name="l00260"></a>00260             <span class="keywordflow">if</span> (motherPresent)
<a name="l00261"></a>00261             {
<a name="l00262"></a>00262                 vMotherAllGenotypeCalls.push_back(cychpMother.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex)-&gt;getGenotypeCall());
<a name="l00263"></a>00263             }
<a name="l00264"></a>00264             <span class="keywordflow">if</span> (fatherPresent)
<a name="l00265"></a>00265             {
<a name="l00266"></a>00266                 vFatherAllGenotypeCalls.push_back(cychpFather.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex)-&gt;getGenotypeCall());
<a name="l00267"></a>00267             }
<a name="l00268"></a>00268             vAllChromosomes.push_back(pMarker-&gt;getChromosome());
<a name="l00269"></a>00269 
<a name="l00270"></a>00270             <a class="code" href="classAlleleFrequency.html" title="Class to store allele frequency data.">AlleleFrequency</a> objSearch;
<a name="l00271"></a>00271             objSearch.SNPID = pMarker-&gt;getProbeSetName();
<a name="l00272"></a>00272             aIterPair_t range = std::equal_range(m_vAlleleFrequencies.begin(), m_vAlleleFrequencies.end(), &amp;objSearch, <a class="code" href="structAlleleFreqComp.html" title="Comparison predicate for AlleleFrequency pointers.">AlleleFreqComp</a>());
<a name="l00273"></a>00273             <span class="keywordflow">if</span> (range.first != range.second)
<a name="l00274"></a>00274             {
<a name="l00275"></a>00275                 vIndexGenotypeCalls.push_back(pMarker-&gt;getGenotypeCall());
<a name="l00276"></a>00276                 <span class="keywordflow">if</span> (motherPresent)
<a name="l00277"></a>00277                 {
<a name="l00278"></a>00278                     vMotherGenotypeCalls.push_back(cychpMother.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex)-&gt;getGenotypeCall());
<a name="l00279"></a>00279                 }
<a name="l00280"></a>00280                 <span class="keywordflow">if</span> (fatherPresent)
<a name="l00281"></a>00281                 {
<a name="l00282"></a>00282                     vFatherGenotypeCalls.push_back(cychpFather.getCychpProbeSetsCopyNumbers().getAt(iMarkerIndex)-&gt;getGenotypeCall());
<a name="l00283"></a>00283                 }
<a name="l00284"></a>00284                 vAAlleleFrequencies.push_back((*range.first)-&gt;AAlleleFrequency);
<a name="l00285"></a>00285                 vBAlleleFrequencies.push_back((*range.first)-&gt;BAlleleFrequency);
<a name="l00286"></a>00286                 vChromosomes.push_back(pMarker-&gt;getChromosome());
<a name="l00287"></a>00287             }
<a name="l00288"></a>00288         }
<a name="l00289"></a>00289         errorRate = m_fErrorRate;
<a name="l00290"></a>00290     }
<a name="l00291"></a>00291 
<a name="l00292"></a>00292     std::for_each(m_vAlleleFrequencies.begin(), m_vAlleleFrequencies.end(), <a class="code" href="structDeleteAlleleFreq.html" title="Functor to delete objects from vectors of pointers.">DeleteAlleleFreq</a>());
<a name="l00293"></a>00293     <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00294"></a>00294     <a class="code" href="classCytogeneticsTrioAnalysis.html" title="A class for doing trio analysis.">CytogeneticsTrioAnalysis</a> objTrioAnalysis;
<a name="l00295"></a>00295     <span class="keywordtype">unsigned</span> <span class="keywordtype">char</span> ucMotherCall    = 0;
<a name="l00296"></a>00296     <span class="keywordtype">unsigned</span> <span class="keywordtype">char</span> ucFatherCall    = 0;
<a name="l00297"></a>00297     <span class="keywordtype">unsigned</span> <span class="keywordtype">char</span> ucFatherDuoCall = 0;
<a name="l00298"></a>00298     <span class="keywordtype">double</span> dFatherConfidence    = 0;
<a name="l00299"></a>00299     <span class="keywordtype">double</span> dMotherConfidence    = 0;
<a name="l00300"></a>00300     <span class="keywordtype">double</span> dFatherDuoConfidence = 0;
<a name="l00301"></a>00301 
<a name="l00302"></a>00302     <span class="keywordflow">if</span> (motherPresent &amp;&amp; fatherPresent)
<a name="l00303"></a>00303     {
<a name="l00304"></a>00304         dMotherConfidence    = objTrioAnalysis.<a class="code" href="classCytogeneticsTrioAnalysis.html#a979051a2c14470a8f43a3c7bcd772c1f" title="Some Trio Analysis functions for Cytogenetics.">calculateLODPaternity</a>(vIndexGenotypeCalls, vMotherGenotypeCalls, vAAlleleFrequencies, vBAlleleFrequencies, errorRate);
<a name="l00305"></a>00305         dFatherConfidence    = objTrioAnalysis.<a class="code" href="classCytogeneticsTrioAnalysis.html#a979051a2c14470a8f43a3c7bcd772c1f" title="Some Trio Analysis functions for Cytogenetics.">calculateLODPaternity</a>(vIndexGenotypeCalls, vMotherGenotypeCalls, vFatherGenotypeCalls, vAAlleleFrequencies, vBAlleleFrequencies, errorRate);
<a name="l00306"></a>00306         dFatherDuoConfidence = objTrioAnalysis.<a class="code" href="classCytogeneticsTrioAnalysis.html#a979051a2c14470a8f43a3c7bcd772c1f" title="Some Trio Analysis functions for Cytogenetics.">calculateLODPaternity</a>(vIndexGenotypeCalls, vFatherGenotypeCalls, vAAlleleFrequencies, vBAlleleFrequencies, errorRate);
<a name="l00307"></a>00307         Verbose::progressStep(1);
<a name="l00308"></a>00308         <span class="keywordflow">if</span> (dMotherConfidence    &gt; m_fRoleValidityThreshold) {ucMotherCall    = (<span class="keywordtype">unsigned</span> char)1;}
<a name="l00309"></a>00309         <span class="keywordflow">if</span> (dFatherConfidence    &gt; m_fRoleValidityThreshold) {ucFatherCall    = (<span class="keywordtype">unsigned</span> char)1;}
<a name="l00310"></a>00310         <span class="keywordflow">if</span> (dFatherDuoConfidence &gt; m_fRoleValidityThreshold) {ucFatherDuoCall = (<span class="keywordtype">unsigned</span> char)1;}
<a name="l00311"></a>00311     }
<a name="l00312"></a>00312     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (motherPresent)
<a name="l00313"></a>00313     {
<a name="l00314"></a>00314         dMotherConfidence = objTrioAnalysis.<a class="code" href="classCytogeneticsTrioAnalysis.html#a979051a2c14470a8f43a3c7bcd772c1f" title="Some Trio Analysis functions for Cytogenetics.">calculateLODPaternity</a>(vIndexGenotypeCalls, vMotherGenotypeCalls, vAAlleleFrequencies, vBAlleleFrequencies, errorRate);
<a name="l00315"></a>00315         Verbose::progressStep(1);
<a name="l00316"></a>00316         <span class="keywordflow">if</span> (dMotherConfidence &gt; m_fRoleValidityThreshold) {ucMotherCall = (<span class="keywordtype">unsigned</span> char)1;}
<a name="l00317"></a>00317     }
<a name="l00318"></a>00318     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (fatherPresent)
<a name="l00319"></a>00319     {
<a name="l00320"></a>00320         dFatherDuoConfidence = objTrioAnalysis.<a class="code" href="classCytogeneticsTrioAnalysis.html#a979051a2c14470a8f43a3c7bcd772c1f" title="Some Trio Analysis functions for Cytogenetics.">calculateLODPaternity</a>(vIndexGenotypeCalls, vFatherGenotypeCalls, vAAlleleFrequencies, vBAlleleFrequencies, errorRate);
<a name="l00321"></a>00321         Verbose::progressStep(1);
<a name="l00322"></a>00322         <span class="keywordflow">if</span> (dFatherDuoConfidence &gt; m_fRoleValidityThreshold) {ucFatherDuoCall = (<span class="keywordtype">unsigned</span> char)1;}
<a name="l00323"></a>00323     }
<a name="l00324"></a>00324     cychpIndex.setFamilialCall((<span class="keywordtype">unsigned</span> <span class="keywordtype">char</span>)1);
<a name="l00325"></a>00325     cychpIndex.setFamilialConfidence((<span class="keywordtype">float</span>)1);
<a name="l00326"></a>00326 
<a name="l00327"></a>00327     cychpMother.setFamilialCall(ucMotherCall);
<a name="l00328"></a>00328     cychpMother.setFamilialConfidence((<span class="keywordtype">float</span>)dMotherConfidence);
<a name="l00329"></a>00329 
<a name="l00330"></a>00330     cychpFather.setFamilialCall(ucFatherCall);
<a name="l00331"></a>00331     cychpFather.setFamilialConfidence((<span class="keywordtype">float</span>)dFatherConfidence);
<a name="l00332"></a>00332 
<a name="l00333"></a>00333     cychpFather.setFamilialDuoCall(ucFatherDuoCall);
<a name="l00334"></a>00334     cychpFather.setFamilialDuoConfidence((<span class="keywordtype">float</span>)dFatherDuoConfidence);
<a name="l00335"></a>00335 
<a name="l00336"></a>00336     <span class="comment">// Mendelian inheritance errors</span>
<a name="l00337"></a>00337     <span class="keywordflow">if</span> (useGenotypeCalls)
<a name="l00338"></a>00338     {
<a name="l00339"></a>00339         countMIE(vIndexAllGenotypeCalls, vMotherAllGenotypeCalls, vFatherAllGenotypeCalls, vAllChromosomes);
<a name="l00340"></a>00340     }
<a name="l00341"></a>00341     <span class="keywordflow">else</span> {
<a name="l00342"></a>00342         countMIE(vIndexGenotypeCalls, vMotherGenotypeCalls, vFatherGenotypeCalls, vChromosomes);
<a name="l00343"></a>00343     }
<a name="l00344"></a>00344 
<a name="l00345"></a>00345     Verbose::progressEnd(1, <span class="stringliteral">&quot;Done&quot;</span>);
<a name="l00346"></a>00346     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;CNFamilialAnalysisMethodLOD::run(...) end&quot;</span>);
<a name="l00347"></a>00347 }
<a name="l00348"></a>00348 
<a name="l00349"></a>00349 <span class="keywordtype">void</span> CNFamilialAnalysisMethodLOD::fit3CompModel(<a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a>&amp; em, <span class="keyword">const</span> std::vector&lt;double&gt;&amp; vSCAR)
<a name="l00350"></a>00350 {
<a name="l00351"></a>00351     <a class="code" href="classCEMSeed.html" title="Object to hold all prior information to constitute an EM seed.">CEMSeed</a> seed;
<a name="l00352"></a>00352     setSeed(seed);
<a name="l00353"></a>00353     em.<a class="code" href="classCPrimeEM.html#ae6633a6897c30d1c94bbb9d284c4e672" title="set the data to be classified">setData</a>(vSCAR);
<a name="l00354"></a>00354     em.<a class="code" href="classCPrimeEM.html#a6169a816ab26a0752f81a916f10029b9" title="confidence threshold set function, for backward compatibility">setThreshold</a>(1.1);    <span class="comment">// force no -1 calls</span>
<a name="l00355"></a>00355     em.<a class="code" href="classCPrimeEM.html#a9c2454ee67f79c7465be66233ac48994" title="the function to get EM&amp;#39;s estimate with a specific initial seed">EMEstimate</a>(seed);
<a name="l00356"></a>00356 }
<a name="l00357"></a>00357 
<a name="l00358"></a>00358 <span class="keywordtype">double</span> CNFamilialAnalysisMethodLOD::estMaxErrorRate(<a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a>&amp; indexEM,
<a name="l00359"></a>00359                                                    <a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a>&amp; motherEM,
<a name="l00360"></a>00360                                                    <a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a>&amp; fatherEM)
<a name="l00361"></a>00361 {
<a name="l00362"></a>00362     std::vector&lt;double&gt; errorRates;
<a name="l00363"></a>00363     errorRates.push_back(estErrorRate(indexEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a0c5fe09b5823515a7125c18f522cea83" title="cluster centers">m_mu</a>, indexEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a25ef544dc666cbad66131245c2d3c1c0" title="cluster sigmas">m_sigma</a>));
<a name="l00364"></a>00364     <span class="keywordflow">if</span> (getCychpMother().getFileName() != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00365"></a>00365     {
<a name="l00366"></a>00366         errorRates.push_back(estErrorRate(motherEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a0c5fe09b5823515a7125c18f522cea83" title="cluster centers">m_mu</a>, motherEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a25ef544dc666cbad66131245c2d3c1c0" title="cluster sigmas">m_sigma</a>));
<a name="l00367"></a>00367     }
<a name="l00368"></a>00368     <span class="keywordflow">if</span> (getCychpFather().getFileName() != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00369"></a>00369     {
<a name="l00370"></a>00370         errorRates.push_back(estErrorRate(fatherEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a0c5fe09b5823515a7125c18f522cea83" title="cluster centers">m_mu</a>, fatherEM.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>()-&gt;<a class="code" href="classCEMEst.html#a25ef544dc666cbad66131245c2d3c1c0" title="cluster sigmas">m_sigma</a>));
<a name="l00371"></a>00371     }
<a name="l00372"></a>00372     <span class="keywordflow">return</span> *std::max_element(errorRates.begin(), errorRates.end());
<a name="l00373"></a>00373 }
<a name="l00374"></a>00374 
<a name="l00375"></a>00375 <span class="keywordtype">double</span> CNFamilialAnalysisMethodLOD::estErrorRate(<span class="keyword">const</span> std::vector&lt;double&gt;&amp; mean, <span class="keyword">const</span> std::vector&lt;double&gt;&amp; sigma)
<a name="l00376"></a>00376 {
<a name="l00377"></a>00377     <span class="keyword">const</span> <span class="keywordtype">double</span> sqrt2 = 1.414213562373;
<a name="l00378"></a>00378 
<a name="l00379"></a>00379     <span class="keywordtype">double</span> intersect1 = findIntersections(mean[1], mean[2], sigma[1], sigma[2]).first;
<a name="l00380"></a>00380     <span class="keywordtype">double</span> intersect2 = findIntersections(mean[1], mean[0], sigma[1], sigma[0]).second;
<a name="l00381"></a>00381 
<a name="l00382"></a>00382     <span class="keywordtype">double</span> erfArg1 = (intersect1 - mean[2])/(sigma[2]*sqrt2);
<a name="l00383"></a>00383     <span class="keywordtype">double</span> erfArg2 = (intersect2 - mean[0])/(sigma[0]*sqrt2);
<a name="l00384"></a>00384 
<a name="l00385"></a>00385     <span class="keywordtype">double</span> errorRate1 = 0.5 + 0.5*<a class="code" href="namespaceaffxstat.html#a01c7688a08b5c80031c8425ac33b8150">affxstat::erf</a>(erfArg1);
<a name="l00386"></a>00386     <span class="keywordtype">double</span> errorRate2 = 0.5 - 0.5*<a class="code" href="namespaceaffxstat.html#a01c7688a08b5c80031c8425ac33b8150">affxstat::erf</a>(erfArg2);
<a name="l00387"></a>00387 
<a name="l00388"></a>00388     <span class="keywordflow">return</span> errorRate1 &gt; errorRate2 ? errorRate1 : errorRate2;
<a name="l00389"></a>00389 }
<a name="l00390"></a>00390 
<a name="l00391"></a>00391 <span class="keywordtype">void</span> CNFamilialAnalysisMethodLOD::setSeed(<a class="code" href="classCEMSeed.html" title="Object to hold all prior information to constitute an EM seed.">CEMSeed</a>&amp; seed)
<a name="l00392"></a>00392 {
<a name="l00393"></a>00393     seed.<a class="code" href="classCEMSeed.html#a1d85b2d7578a68250caa978a06bf65fc" title="Set the centers of clusters assuming upto three clusters in total.">setMu</a>(-1.0, 0, 1.0);
<a name="l00394"></a>00394     seed.<a class="code" href="classCEMSeed.html#af7f9a507dcdfb399eab17dec2650bad0" title="Set the sigmas of clusters assuming upto three clusters in total.">setSigma</a>(0.1f, 0.1f, 0.1f);
<a name="l00395"></a>00395     seed.<a class="code" href="classCEMSeed.html#ae014524f6c4329ca8b18364b316361c1" title="Set the weight of clusters assuming upto three clusters in total.">setWeight</a>(0.33f, 0.34f, 0.33f);
<a name="l00396"></a>00396     seed.<a class="code" href="classCEMSeed.html#ab53cf92fa9b34fdae1394e5bb33b9498" title="Set the minimum of centers of clusters assuming upto three clusters in total.">setMinMu</a>(-2, -.05, .25);
<a name="l00397"></a>00397     seed.<a class="code" href="classCEMSeed.html#abb9d2912972d1c6ca13b40a91793a74c" title="Set the minimum of sigmas of clusters assuming upto three clusters in total.">setMinSigma</a>(0.02f, 0.02f, 0.02f);
<a name="l00398"></a>00398     seed.<a class="code" href="classCEMSeed.html#abf4e379c205f693d233e7e6154157a7f" title="Set the maximum of centers of clusters assuming upto three clusters in total.">setMaxMu</a>(-.25, .05, 2);
<a name="l00399"></a>00399     seed.<a class="code" href="classCEMSeed.html#a7507e7757be1c92ce717ed86ec1c2aa2" title="Set the maximum of sigmas of clusters assuming upto three clusters in total.">setMaxSigma</a>(.3, .3, .3);
<a name="l00400"></a>00400 }
<a name="l00401"></a>00401 
<a name="l00402"></a>00402 <span class="comment">// Find intersection points of given gaussians</span>
<a name="l00403"></a>00403 std::pair&lt;double, double&gt; CNFamilialAnalysisMethodLOD::findIntersections(<span class="keywordtype">double</span> mean1, <span class="keywordtype">double</span> mean2, <span class="keywordtype">double</span> sigma1, <span class="keywordtype">double</span> sigma2)
<a name="l00404"></a>00404 {
<a name="l00405"></a>00405     <span class="keyword">using</span> std::abs;
<a name="l00406"></a>00406     <span class="keyword">using</span> std::log;
<a name="l00407"></a>00407     <span class="keyword">using</span> std::sqrt;
<a name="l00408"></a>00408 
<a name="l00409"></a>00409     <span class="keyword">const</span> <span class="keywordtype">double</span> epsilon = 1.0e-10;
<a name="l00410"></a>00410 
<a name="l00411"></a>00411     <span class="keywordtype">double</span> intersect1;
<a name="l00412"></a>00412     <span class="keywordtype">double</span> intersect2;
<a name="l00413"></a>00413 
<a name="l00414"></a>00414     <span class="keywordflow">if</span> (abs(sigma1 - sigma2) &lt; epsilon)
<a name="l00415"></a>00415     {
<a name="l00416"></a>00416         intersect1 = intersect2 = 0.5*(mean1 + mean2);
<a name="l00417"></a>00417     }
<a name="l00418"></a>00418     <span class="keywordflow">else</span> {
<a name="l00419"></a>00419         <span class="keywordtype">double</span> A = sigma1*sigma1 - sigma2*sigma2;
<a name="l00420"></a>00420         <span class="keywordtype">double</span> B = -2.0*(mean2*sigma1*sigma1 - mean1*sigma2*sigma2);
<a name="l00421"></a>00421         <span class="keywordtype">double</span> C = mean2*mean2*sigma1*sigma1 - mean1*mean1*sigma2*sigma2 - 2.0*sigma2*sigma2*sigma1*sigma1*(log(sigma2) - log(sigma1));
<a name="l00422"></a>00422         intersect1 = (-B - sqrt(B*B - 4.0*A*C))/(2.0*A);
<a name="l00423"></a>00423         intersect2 = (-B + sqrt(B*B - 4.0*A*C))/(2.0*A);
<a name="l00424"></a>00424     }
<a name="l00425"></a>00425     <span class="keywordflow">return</span> std::make_pair(intersect1, intersect2);
<a name="l00426"></a>00426 }
<a name="l00427"></a>00427 
<a name="l00428"></a>00428 <span class="keywordtype">void</span> CNFamilialAnalysisMethodLOD::countMIE(
<a name="l00429"></a>00429                                     <span class="keyword">const</span> std::vector&lt;char&gt;&amp; vIndexGenotypeCalls,
<a name="l00430"></a>00430                                     <span class="keyword">const</span> std::vector&lt;char&gt;&amp; vMotherGenotypeCalls,
<a name="l00431"></a>00431                                     <span class="keyword">const</span> std::vector&lt;char&gt;&amp; vFatherGenotypeCalls,
<a name="l00432"></a>00432                                     <span class="keyword">const</span> std::vector&lt;unsigned char&gt;&amp; vChromosomes
<a name="l00433"></a>00433                                     )
<a name="l00434"></a>00434 {
<a name="l00435"></a>00435     <span class="keyword">const</span> <span class="keywordtype">bool</span> trioCase = !vMotherGenotypeCalls.empty() &amp;&amp; !vFatherGenotypeCalls.empty();
<a name="l00436"></a>00436     <span class="keyword">const</span> <span class="keywordtype">bool</span> duoMat   = !vMotherGenotypeCalls.empty();
<a name="l00437"></a>00437     <span class="keyword">const</span> <span class="keywordtype">bool</span> duoPat   = !vFatherGenotypeCalls.empty();
<a name="l00438"></a>00438 
<a name="l00439"></a>00439     <span class="keywordflow">if</span> (trioCase)
<a name="l00440"></a>00440     {
<a name="l00441"></a>00441         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = 0; iIndex &lt; vIndexGenotypeCalls.size(); iIndex++)
<a name="l00442"></a>00442         {
<a name="l00443"></a>00443             <a class="code" href="classCNCychp.html#af56dcf7f5ebb89f2ffb2256e5cfdca6b" title="For Mendelian Inheritance Error (MIE) results.">CNCychp::m_mMarkerCount</a>[vChromosomes[iIndex]]++;
<a name="l00444"></a>00444             <span class="keywordflow">if</span> (isMIE(vIndexGenotypeCalls[iIndex], vMotherGenotypeCalls[iIndex], vFatherGenotypeCalls[iIndex]))
<a name="l00445"></a>00445             {
<a name="l00446"></a>00446                 CNCychp::m_mMIE_Trio[vChromosomes[iIndex]]++;
<a name="l00447"></a>00447             }
<a name="l00448"></a>00448             <span class="keywordflow">if</span> (isMIE(vIndexGenotypeCalls[iIndex], vMotherGenotypeCalls[iIndex]))
<a name="l00449"></a>00449             {
<a name="l00450"></a>00450                 CNCychp::m_mMIE_Mat[vChromosomes[iIndex]]++;
<a name="l00451"></a>00451             }
<a name="l00452"></a>00452             <span class="keywordflow">if</span> (isMIE(vIndexGenotypeCalls[iIndex], vFatherGenotypeCalls[iIndex]))
<a name="l00453"></a>00453             {
<a name="l00454"></a>00454                 CNCychp::m_mMIE_Pat[vChromosomes[iIndex]]++;
<a name="l00455"></a>00455             }
<a name="l00456"></a>00456         }
<a name="l00457"></a>00457     }
<a name="l00458"></a>00458     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (duoMat)
<a name="l00459"></a>00459     {
<a name="l00460"></a>00460         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = 0; iIndex &lt; vIndexGenotypeCalls.size(); iIndex++)
<a name="l00461"></a>00461         {
<a name="l00462"></a>00462             <a class="code" href="classCNCychp.html#af56dcf7f5ebb89f2ffb2256e5cfdca6b" title="For Mendelian Inheritance Error (MIE) results.">CNCychp::m_mMarkerCount</a>[vChromosomes[iIndex]]++;
<a name="l00463"></a>00463             <span class="keywordflow">if</span> (isMIE(vIndexGenotypeCalls[iIndex], vMotherGenotypeCalls[iIndex]))
<a name="l00464"></a>00464             {
<a name="l00465"></a>00465                 CNCychp::m_mMIE_Mat[vChromosomes[iIndex]]++;
<a name="l00466"></a>00466             }
<a name="l00467"></a>00467         }
<a name="l00468"></a>00468     }
<a name="l00469"></a>00469     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (duoPat)
<a name="l00470"></a>00470     {
<a name="l00471"></a>00471         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> iIndex = 0; iIndex &lt; vIndexGenotypeCalls.size(); iIndex++)
<a name="l00472"></a>00472         {
<a name="l00473"></a>00473             <a class="code" href="classCNCychp.html#af56dcf7f5ebb89f2ffb2256e5cfdca6b" title="For Mendelian Inheritance Error (MIE) results.">CNCychp::m_mMarkerCount</a>[vChromosomes[iIndex]]++;
<a name="l00474"></a>00474             <span class="keywordflow">if</span> (isMIE(vIndexGenotypeCalls[iIndex], vFatherGenotypeCalls[iIndex]))
<a name="l00475"></a>00475             {
<a name="l00476"></a>00476                 CNCychp::m_mMIE_Pat[vChromosomes[iIndex]]++;
<a name="l00477"></a>00477             }
<a name="l00478"></a>00478         }
<a name="l00479"></a>00479     }
<a name="l00480"></a>00480 }
<a name="l00481"></a>00481 
<a name="l00482"></a>00482 <span class="keyword">inline</span> <span class="keywordtype">bool</span> CNFamilialAnalysisMethodLOD::isMIE(<span class="keywordtype">char</span> genIndex, <span class="keywordtype">char</span> genMother, <span class="keywordtype">char</span> genFather)
<a name="l00483"></a>00483 {
<a name="l00484"></a>00484     <span class="comment">// Mendelian error if either:</span>
<a name="l00485"></a>00485     <span class="comment">//  Mother  Father  Child</span>
<a name="l00486"></a>00486     <span class="comment">//  AA      AA      AB|BB</span>
<a name="l00487"></a>00487     <span class="comment">//  AA      BB      AA|BB</span>
<a name="l00488"></a>00488     <span class="comment">//  AA      AB      BB</span>
<a name="l00489"></a>00489     <span class="comment">//  AB      AA      BB</span>
<a name="l00490"></a>00490     <span class="comment">//  AB      BB      AA</span>
<a name="l00491"></a>00491     <span class="comment">//  BB      AB      AA</span>
<a name="l00492"></a>00492     <span class="comment">//  BB      AA      AA|BB</span>
<a name="l00493"></a>00493     <span class="comment">//  BB      BB      AA|AB</span>
<a name="l00494"></a>00494     <span class="comment">//</span>
<a name="l00495"></a>00495     <span class="keywordflow">return</span>
<a name="l00496"></a>00496         genMother == 0 &amp;&amp; genFather == 0 &amp;&amp; (genIndex == 1 || genIndex == 2)  ||
<a name="l00497"></a>00497         genMother == 0 &amp;&amp; genFather == 2 &amp;&amp; (genIndex == 0 || genIndex == 2)  ||
<a name="l00498"></a>00498         genMother == 0 &amp;&amp; genFather == 1 &amp;&amp; genIndex == 2                     ||
<a name="l00499"></a>00499         genMother == 1 &amp;&amp; genFather == 0 &amp;&amp; genIndex == 2                     ||
<a name="l00500"></a>00500         genMother == 1 &amp;&amp; genFather == 2 &amp;&amp; genIndex == 0                     ||
<a name="l00501"></a>00501         genMother == 2 &amp;&amp; genFather == 1 &amp;&amp; genIndex == 0                     ||
<a name="l00502"></a>00502         genMother == 2 &amp;&amp; genFather == 0 &amp;&amp; (genIndex == 0 || genIndex == 2)  ||
<a name="l00503"></a>00503         genMother == 2 &amp;&amp; genFather == 2 &amp;&amp; (genIndex == 0 || genIndex == 1);
<a name="l00504"></a>00504 }
<a name="l00505"></a>00505 
<a name="l00506"></a>00506 <span class="keyword">inline</span> <span class="keywordtype">bool</span> CNFamilialAnalysisMethodLOD::isMIE(<span class="keywordtype">char</span> genIndex, <span class="keywordtype">char</span> genParent)
<a name="l00507"></a>00507 {
<a name="l00508"></a>00508     <span class="comment">// Mendelian error if either:</span>
<a name="l00509"></a>00509     <span class="comment">//  Parent  Child</span>
<a name="l00510"></a>00510     <span class="comment">//  AA      BB</span>
<a name="l00511"></a>00511     <span class="comment">//  BB      AA</span>
<a name="l00512"></a>00512     <span class="comment">//</span>
<a name="l00513"></a>00513     <span class="keywordflow">return</span>
<a name="l00514"></a>00514         genParent == 0 &amp;&amp; genIndex == 2  ||
<a name="l00515"></a>00515         genParent == 2 &amp;&amp; genIndex == 0;
<a name="l00516"></a>00516 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:57 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
