<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/DmListener.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>DmListener.cpp</h1>  </div>
</div>
<div class="contents">
<a href="DmListener_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   DmListener.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Chuck Sugnet</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Sun Mar 12 11:55:49 2006</span>
<a name="l00026"></a>00026 <span class="comment"> *</span>
<a name="l00027"></a>00027 <span class="comment"> * @brief Class for using DM calls as cell files become available.</span>
<a name="l00028"></a>00028 <span class="comment"> */</span>
<a name="l00029"></a>00029 
<a name="l00030"></a>00030 <span class="comment">//</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="DmListener_8h.html" title="Class for using DM calls as cell files become available.">chipstream/DmListener.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="comment">//</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &lt;cmath&gt;</span>
<a name="l00034"></a>00034 <span class="comment">//</span>
<a name="l00035"></a>00035 <span class="keyword">using namespace </span>std;
<a name="l00036"></a>00036 <span class="keyword">using namespace </span>affx;
<a name="l00037"></a>00037 
<a name="l00038"></a>00038 DmListener::DmCall::DmCall()
<a name="l00039"></a>00039 {
<a name="l00040"></a>00040   good = <span class="keyword">false</span>;
<a name="l00041"></a>00041   call = affx::NN; <span class="comment">// unknown</span>
<a name="l00042"></a>00042   conf = 0;
<a name="l00043"></a>00043 }
<a name="l00044"></a>00044 <span class="comment"></span>
<a name="l00045"></a>00045 <span class="comment">/**</span>
<a name="l00046"></a>00046 <span class="comment"> * Constructor takes a vector of ProbeSets for which calls will be</span>
<a name="l00047"></a>00047 <span class="comment"> * done as cel files become available.</span>
<a name="l00048"></a>00048 <span class="comment"> *</span>
<a name="l00049"></a>00049 <span class="comment"> * @param psVec - Vector of probesets to be calculated.  @param</span>
<a name="l00050"></a>00050 <span class="comment"> * hetMult - Factor to add to log likelihood to balance het/hom calls</span>
<a name="l00051"></a>00051 <span class="comment"> * 0 for no effect.</span>
<a name="l00052"></a>00052 <span class="comment"> *</span>
<a name="l00053"></a>00053 <span class="comment"> * @return</span>
<a name="l00054"></a>00054 <span class="comment"> */</span>
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 
<a name="l00057"></a><a class="code" href="classDmListener.html#a097e78ab5914e3e2ed150e747e6397d0">00057</a> <a class="code" href="classDmListener.html#a097e78ab5914e3e2ed150e747e6397d0" title="Constructor takes a vector of ProbeSets for which calls will be done as cel files become available...">DmListener::DmListener</a>(std::vector&lt;ProbeListPacked&gt; &amp;probesets, <span class="keywordtype">float</span> hetMult,
<a name="l00058"></a>00058                        <span class="keywordtype">int</span> maxChips, std::vector&lt;float&gt; &amp;thresholds,
<a name="l00059"></a>00059                        std::map&lt;string, bool&gt; &amp;chrXSnps, std::string callRateName)
<a name="l00060"></a>00060 {
<a name="l00061"></a>00061   std::vector&lt;ProbeSet *&gt;::iterator iter;
<a name="l00062"></a>00062   <a class="code" href="classErr.html#ad1a633f602fe60689cdc12d9ca1842e9" title="errAbort based version of assert with an error message.">Err::check</a>(!thresholds.empty(), <span class="stringliteral">&quot;DmListener::DmListener() - Can&#39;t supply no thresholds.&quot;</span>);
<a name="l00063"></a>00063   m_HetMultiplier = hetMult;
<a name="l00064"></a>00064   m_ChipCount = 0;
<a name="l00065"></a>00065   m_MaxChips = maxChips;
<a name="l00066"></a>00066   m_Thresholds = thresholds;
<a name="l00067"></a>00067   m_TotalCalls.resize(m_MaxChips);
<a name="l00068"></a>00068   m_PassCalls.resize(m_Thresholds.size());
<a name="l00069"></a>00069   m_HetChrXCalls.resize(m_Thresholds.size());
<a name="l00070"></a>00070   m_FirstPassGender.resize(maxChips, <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669aa259d114b82ef05d3f1f9d3c81c84719" title="Don&amp;#39;t know genotype, use XX model.">affx::UnknownGender</a>);
<a name="l00071"></a>00071   m_FirstPassHetRate.resize(maxChips, 0.0);
<a name="l00072"></a>00072   m_ChrXCalls.resize(maxChips);
<a name="l00073"></a>00073   m_ChrXSnps = chrXSnps;
<a name="l00074"></a>00074   m_CallIndex = 0;
<a name="l00075"></a>00075   m_ChrXCallIndex = 0;
<a name="l00076"></a>00076   <span class="keywordflow">for</span> (uint32_t i = 0; i &lt; m_PassCalls.size(); i++) {
<a name="l00077"></a>00077     m_PassCalls[i].resize(m_MaxChips);
<a name="l00078"></a>00078     m_HetChrXCalls[i].resize(m_MaxChips);
<a name="l00079"></a>00079   }
<a name="l00080"></a>00080   <span class="keywordflow">for</span> (uint32_t i = 0; i &lt; probesets.size(); i++) {
<a name="l00081"></a>00081     m_PlVec.push_back(probesets[i]);
<a name="l00082"></a>00082   }
<a name="l00083"></a>00083 
<a name="l00084"></a>00084   m_KeepGenotypes = <span class="keyword">true</span>;
<a name="l00085"></a>00085 
<a name="l00086"></a>00086 
<a name="l00087"></a>00087   m_GenderName = <span class="stringliteral">&quot;dm-chrX-het-rate&quot;</span>;
<a name="l00088"></a>00088   m_GenderDescription = <span class="stringliteral">&quot;DM chrX het rate&quot;</span>;
<a name="l00089"></a>00089   setPrefix(<span class="stringliteral">&quot;dm-listener&quot;</span>);
<a name="l00090"></a>00090 
<a name="l00091"></a>00091   <span class="keywordflow">if</span> (callRateName != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00092"></a>00092     setCallRateName(callRateName);
<a name="l00093"></a>00093 
<a name="l00094"></a>00094   <span class="keywordflow">if</span> (m_ChrXSnps.size() &gt; 0) {
<a name="l00095"></a>00095     <a class="code" href="classChipSummary.html#a8c573ca48c79d016640b01b9fa632294">declareMetric</a>(m_GenderName + <span class="stringliteral">&quot;_gender&quot;</span>, ChipSummary::Metric::String);
<a name="l00096"></a>00096     <a class="code" href="classChipSummary.html#a8c573ca48c79d016640b01b9fa632294">declareMetric</a>(m_GenderName + <span class="stringliteral">&quot;_gender_chrX_het_rate&quot;</span>, ChipSummary::Metric::Double);
<a name="l00097"></a>00097   }
<a name="l00098"></a>00098   <a class="code" href="classChipSummary.html#a8c573ca48c79d016640b01b9fa632294">declareMetric</a>(m_CallRateName, ChipSummary::Metric::Double);
<a name="l00099"></a>00099 }
<a name="l00100"></a>00100 
<a name="l00101"></a><a class="code" href="classDmListener.html#aea713ca7a0b3c9c7e550e3ecfede9d32">00101</a> <a class="code" href="classDmListener.html#aea713ca7a0b3c9c7e550e3ecfede9d32" title="Virtual destructor.">DmListener::~DmListener</a>()
<a name="l00102"></a>00102 {
<a name="l00103"></a>00103 }
<a name="l00104"></a>00104 
<a name="l00105"></a>00105 <span class="comment"></span>
<a name="l00106"></a>00106 <span class="comment">/**</span>
<a name="l00107"></a>00107 <span class="comment"> * Clear genotype results. Gender results are retained.</span>
<a name="l00108"></a>00108 <span class="comment"> * Call rates (total and chrX) are retained and accumulated</span>
<a name="l00109"></a>00109 <span class="comment"> * over calls to this. The only thing this does is it</span>
<a name="l00110"></a>00110 <span class="comment"> * drops the genotype calls.</span>
<a name="l00111"></a>00111 <span class="comment"> * This allows us to save memory by not keeping around</span>
<a name="l00112"></a>00112 <span class="comment"> * genotypes that are no longer needed.</span>
<a name="l00113"></a>00113 <span class="comment"> */</span>
<a name="l00114"></a>00114 
<a name="l00115"></a><a class="code" href="classDmListener.html#ac1dc39a515bf55390cbe2fe8dda10c46">00115</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#ac1dc39a515bf55390cbe2fe8dda10c46" title="Clear genotype results.">DmListener::clearForIteration</a>()
<a name="l00116"></a>00116 {
<a name="l00117"></a>00117   m_KnownCalls.clear();
<a name="l00118"></a>00118   m_ChipCount = 0;
<a name="l00119"></a>00119 }
<a name="l00120"></a>00120 <span class="comment"></span>
<a name="l00121"></a>00121 <span class="comment">/**</span>
<a name="l00122"></a>00122 <span class="comment"> * Do we need to keep genotypes around for providing</span>
<a name="l00123"></a>00123 <span class="comment"> * seed calls? Default yes. If set to no, then</span>
<a name="l00124"></a>00124 <span class="comment"> * getGenoCalls will fail if called.</span>
<a name="l00125"></a>00125 <span class="comment"> * @param s - true or false</span>
<a name="l00126"></a>00126 <span class="comment"> */</span>
<a name="l00127"></a><a class="code" href="classDmListener.html#ab9252a76286593db30c81a7d38575c58">00127</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#ab9252a76286593db30c81a7d38575c58" title="Do we need to keep genotypes around for providing seed calls? Default yes.">DmListener::setKeepGenotypes</a>(<span class="keywordtype">bool</span> s)
<a name="l00128"></a>00128 {
<a name="l00129"></a>00129   m_KeepGenotypes = s;
<a name="l00130"></a>00130 }
<a name="l00131"></a>00131 
<a name="l00132"></a>00132 
<a name="l00133"></a><a class="code" href="classDmListener.html#a50189d7294c1e7de78ccd63116217900">00133</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#a50189d7294c1e7de78ccd63116217900" title="Set a new layout -- allows us to save some memory and only deal with SNPs in the current iteration...">DmListener::setLayout</a>(<a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> &amp;layout)
<a name="l00134"></a>00134 {
<a name="l00135"></a>00135   m_PlVec.clear();
<a name="l00136"></a>00136   <span class="keywordflow">for</span> (uint32_t i = 0; i &lt; layout.<a class="code" href="classChipLayout.html#a71df054a01d62ca33d6b8de8c8c5c3ca" title="How many probe sets in Pgf file?">getProbeSetCount</a>(); i++) {
<a name="l00137"></a>00137     m_PlVec.push_back(layout.<a class="code" href="classChipLayout.html#a3127c7baedda03477f28da70bbfb686e" title="Get a probelist by the index in the layout.">getProbeListAtIndex</a>(i));
<a name="l00138"></a>00138   }
<a name="l00139"></a>00139 }
<a name="l00140"></a>00140 <span class="comment"></span>
<a name="l00141"></a>00141 <span class="comment">/**</span>
<a name="l00142"></a>00142 <span class="comment"> * @brief Check if a call exists for a given probeset name</span>
<a name="l00143"></a>00143 <span class="comment"> * @param name - name of probeset to get genotype calls for.</span>
<a name="l00144"></a>00144 <span class="comment"> * @return boolean result</span>
<a name="l00145"></a>00145 <span class="comment"> */</span>
<a name="l00146"></a><a class="code" href="classDmListener.html#af43fe1bf75be4d7feb026ec8e23c3ddd">00146</a> <span class="keywordtype">bool</span> <a class="code" href="classDmListener.html#af43fe1bf75be4d7feb026ec8e23c3ddd" title="Check if a call exists for a given probeset name.">DmListener::checkGenoCallsName</a>(<span class="keyword">const</span> std::string &amp;name)
<a name="l00147"></a>00147 {
<a name="l00148"></a>00148   <span class="keywordflow">return</span> m_KnownCalls.end() != m_KnownCalls.find(name);
<a name="l00149"></a>00149 }
<a name="l00150"></a>00150 <span class="comment"></span>
<a name="l00151"></a>00151 <span class="comment">/** Accessor function. */</span>
<a name="l00152"></a><a class="code" href="classDmListener.html#a376601933d0745a012d80bae65547079">00152</a> <span class="keywordtype">int</span> <a class="code" href="classDmListener.html#a376601933d0745a012d80bae65547079" title="Accessor function.">DmListener::getChipCount</a>()
<a name="l00153"></a>00153 {
<a name="l00154"></a>00154   <span class="keywordflow">return</span> m_ChipCount;
<a name="l00155"></a>00155 }
<a name="l00156"></a>00156 <span class="comment"></span>
<a name="l00157"></a>00157 <span class="comment">/**</span>
<a name="l00158"></a>00158 <span class="comment"> * Get the call rate at a particular threshold. Note that we can keep</span>
<a name="l00159"></a>00159 <span class="comment"> * track of the rate at multiple thresholds, but only keep the actual</span>
<a name="l00160"></a>00160 <span class="comment"> * calls for a particular threshold due to space (RAM) concerns.</span>
<a name="l00161"></a>00161 <span class="comment"> *</span>
<a name="l00162"></a>00162 <span class="comment"> * @param callRates - Vector to be filled in with call rates, one for each chip.</span>
<a name="l00163"></a>00163 <span class="comment"> * @param thresholdIndex - Which threshold to use (as defined in constructor).</span>
<a name="l00164"></a>00164 <span class="comment"> */</span>
<a name="l00165"></a><a class="code" href="classDmListener.html#a8d767a0f6220dc2c95e67e99586963ec">00165</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#a8d767a0f6220dc2c95e67e99586963ec" title="Get the call rate at a particular threshold.">DmListener::getCallRates</a>(std::vector&lt;float&gt; &amp;callRates, <span class="keywordtype">int</span> thresholdIndex)
<a name="l00166"></a>00166 {
<a name="l00167"></a>00167   callRates.clear();
<a name="l00168"></a>00168   callRates.resize(m_ChipCount);
<a name="l00169"></a>00169   assert(thresholdIndex &lt; m_PassCalls.size());
<a name="l00170"></a>00170   assert(m_PassCalls[thresholdIndex].size() == m_ChipCount);
<a name="l00171"></a>00171   <span class="keywordflow">for</span> (uint32_t i = 0; i &lt; m_ChipCount; i++) {
<a name="l00172"></a>00172     assert(m_TotalCalls[i] &gt; 0);
<a name="l00173"></a>00173     callRates[i] = (float)m_PassCalls[thresholdIndex][i] / m_TotalCalls[i];
<a name="l00174"></a>00174   }
<a name="l00175"></a>00175 }
<a name="l00176"></a>00176 <span class="comment"></span>
<a name="l00177"></a>00177 <span class="comment">/**</span>
<a name="l00178"></a>00178 <span class="comment"> * Get the call rate for a specific chip</span>
<a name="l00179"></a>00179 <span class="comment"> * @param chip - the chip index to pull</span>
<a name="l00180"></a>00180 <span class="comment"> * @param thresholdIndx - which threshold should be applied</span>
<a name="l00181"></a>00181 <span class="comment"> * @return The call rate for that chip</span>
<a name="l00182"></a>00182 <span class="comment"> */</span>
<a name="l00183"></a><a class="code" href="classDmListener.html#a31271c403cb75acf6cff9a1773c2f7e4">00183</a> <span class="keywordtype">float</span> <a class="code" href="classDmListener.html#a31271c403cb75acf6cff9a1773c2f7e4" title="Get the call rate for a specific chip.">DmListener::getCallRate</a>(<span class="keywordtype">int</span> chip, <span class="keywordtype">int</span> thresholdIndex)
<a name="l00184"></a>00184 {
<a name="l00185"></a>00185   <span class="keywordflow">if</span> (m_TotalCalls[chip] == 0) {
<a name="l00186"></a>00186     <span class="keywordflow">return</span> -1;
<a name="l00187"></a>00187   } <span class="keywordflow">else</span> {
<a name="l00188"></a>00188     <span class="comment">//Verbose::out(1,&quot;Chip &quot; + ToStr(chip) + &quot; has &quot; + ToStr(m_PassCalls[thresholdIndex][chip]) + &quot; passing calls out of &quot; + ToStr(m_TotalCalls[chip]));</span>
<a name="l00189"></a>00189     <span class="keywordflow">return</span> (<span class="keywordtype">float</span>)m_PassCalls[thresholdIndex][chip] / m_TotalCalls[chip];
<a name="l00190"></a>00190   }
<a name="l00191"></a>00191 }
<a name="l00192"></a>00192 
<a name="l00193"></a>00193 <span class="comment"></span>
<a name="l00194"></a>00194 <span class="comment">/**</span>
<a name="l00195"></a>00195 <span class="comment"> * Get the call rates for chrX snps at a particular threshold. Note that we</span>
<a name="l00196"></a>00196 <span class="comment"> * can keep track of the rate at multiple thresholds, but only keep the actual</span>
<a name="l00197"></a>00197 <span class="comment"> * calls for a particular threshold due to space (RAM) concerns.</span>
<a name="l00198"></a>00198 <span class="comment"> *</span>
<a name="l00199"></a>00199 <span class="comment"> * @param callRates - Vector to be filled in with call rates for chrx snps,</span>
<a name="l00200"></a>00200 <span class="comment"> *                    one for each chip.</span>
<a name="l00201"></a>00201 <span class="comment"> * @param thresholdIndex - Which threshold to use (as defined in constructor.</span>
<a name="l00202"></a>00202 <span class="comment"> */</span>
<a name="l00203"></a><a class="code" href="classDmListener.html#af0af728c30461f3f002f14dfb318d958">00203</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#af0af728c30461f3f002f14dfb318d958" title="Get the call rates for chrX snps at a particular threshold.">DmListener::getHetChrXRates</a>(std::vector&lt;float&gt; &amp;hetRates, <span class="keywordtype">int</span> thresholdIndex)
<a name="l00204"></a>00204 {
<a name="l00205"></a>00205   hetRates.clear();
<a name="l00206"></a>00206   hetRates.resize(m_ChipCount);
<a name="l00207"></a>00207   assert(thresholdIndex &lt; m_HetChrXCalls.size());
<a name="l00208"></a>00208   assert(m_HetChrXCalls[thresholdIndex].size() == m_ChipCount);
<a name="l00209"></a>00209   <span class="keywordflow">for</span> (uint32_t i = 0; i &lt; m_ChipCount; i++) {
<a name="l00210"></a>00210     <span class="keywordflow">if</span> (m_ChrXCalls[i] == 0) {
<a name="l00211"></a>00211       hetRates[i] = -1;
<a name="l00212"></a>00212     } <span class="keywordflow">else</span> {
<a name="l00213"></a>00213       hetRates[i] = (float)m_HetChrXCalls[thresholdIndex][i] / m_ChrXCalls[i];
<a name="l00214"></a>00214     }
<a name="l00215"></a>00215   }
<a name="l00216"></a>00216 }
<a name="l00217"></a>00217 <span class="comment"></span>
<a name="l00218"></a>00218 <span class="comment">/**</span>
<a name="l00219"></a>00219 <span class="comment"> * Get the chrX het rate for a specific chip</span>
<a name="l00220"></a>00220 <span class="comment"> * @param chip - the chip index to pull</span>
<a name="l00221"></a>00221 <span class="comment"> * @param thresholdIndex - which threshold should be applied</span>
<a name="l00222"></a>00222 <span class="comment"> * @return the chrX het rate for that chip</span>
<a name="l00223"></a>00223 <span class="comment"> */</span>
<a name="l00224"></a><a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308">00224</a> <span class="keywordtype">float</span> <a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">DmListener::getHetChrXRate</a>(<span class="keywordtype">int</span> chip, <span class="keywordtype">int</span> thresholdIndex)
<a name="l00225"></a>00225 {
<a name="l00226"></a>00226   <span class="keywordflow">if</span> (m_ChrXCalls[chip] == 0) {
<a name="l00227"></a>00227     <span class="keywordflow">return</span> -1;
<a name="l00228"></a>00228   } <span class="keywordflow">else</span> {
<a name="l00229"></a>00229     <span class="keywordflow">return</span> (<span class="keywordtype">float</span>)m_HetChrXCalls[thresholdIndex][chip] / m_ChrXCalls[chip];
<a name="l00230"></a>00230   }
<a name="l00231"></a>00231 }
<a name="l00232"></a>00232 
<a name="l00233"></a>00233 <span class="comment"></span>
<a name="l00234"></a>00234 <span class="comment">/**</span>
<a name="l00235"></a>00235 <span class="comment"> * Set the index (into the thresholds vector in constructor) that we will</span>
<a name="l00236"></a>00236 <span class="comment"> * use by default when queried for call rates.</span>
<a name="l00237"></a>00237 <span class="comment"> *</span>
<a name="l00238"></a>00238 <span class="comment"> * @param index - index into the call rates at different thresholds</span>
<a name="l00239"></a>00239 <span class="comment"> * we keep track of as defined by the thresholds vector in</span>
<a name="l00240"></a>00240 <span class="comment"> * constructor.</span>
<a name="l00241"></a>00241 <span class="comment"> */</span>
<a name="l00242"></a><a class="code" href="classDmListener.html#aa15c9b8f93ce57f8cb6c0075aa55afcd">00242</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#aa15c9b8f93ce57f8cb6c0075aa55afcd" title="Set the index (into the thresholds vector in constructor) that we will use by default when queried fo...">DmListener::setDefaultCallIndex</a>(<span class="keywordtype">int</span> index)
<a name="l00243"></a>00243 {
<a name="l00244"></a>00244   <span class="keywordflow">if</span> ((<span class="keywordtype">size_t</span>)index &gt;= m_PassCalls.size()) {
<a name="l00245"></a>00245     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;DmListener::setDefaultCallIndex() - Can&#39;t set default to: &quot;</span> +
<a name="l00246"></a>00246                   <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(index) + <span class="stringliteral">&quot; when there are at most: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_PassCalls.size()));
<a name="l00247"></a>00247   }
<a name="l00248"></a>00248   m_CallIndex = index;
<a name="l00249"></a>00249 }
<a name="l00250"></a>00250 <span class="comment"></span>
<a name="l00251"></a>00251 <span class="comment">/**</span>
<a name="l00252"></a>00252 <span class="comment"> * Set the index (into the thresholds vector in constructor) that we will</span>
<a name="l00253"></a>00253 <span class="comment"> * use by default when queried for call rates.</span>
<a name="l00254"></a>00254 <span class="comment"> *</span>
<a name="l00255"></a>00255 <span class="comment"> * @param index - index into the call rates at different thresholds</span>
<a name="l00256"></a>00256 <span class="comment"> * we keep track of as defined by the thresholds vector in</span>
<a name="l00257"></a>00257 <span class="comment"> * constructor.</span>
<a name="l00258"></a>00258 <span class="comment"> */</span>
<a name="l00259"></a><a class="code" href="classDmListener.html#a156c100d4c36fcf95495f37bf942ed6f">00259</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#a156c100d4c36fcf95495f37bf942ed6f" title="Set the index (into the thresholds vector in constructor) that we will use by default when queried fo...">DmListener::setDefaultChrXCallIndex</a>(<span class="keywordtype">int</span> index)
<a name="l00260"></a>00260 {
<a name="l00261"></a>00261   <span class="keywordflow">if</span> ((<span class="keywordtype">size_t</span>)index &gt;= m_HetChrXCalls.size()) {
<a name="l00262"></a>00262     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;DmListener::setDefaultChrXCallIndex() - Can&#39;t set default to: &quot;</span> +
<a name="l00263"></a>00263                   <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(index) + <span class="stringliteral">&quot; when there are at most: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_HetChrXCalls.size()));
<a name="l00264"></a>00264   }
<a name="l00265"></a>00265   m_ChrXCallIndex = index;
<a name="l00266"></a>00266 }
<a name="l00267"></a>00267 <span class="comment"></span>
<a name="l00268"></a>00268 <span class="comment">/**</span>
<a name="l00269"></a>00269 <span class="comment"> * Get the call rates for snps for each chip.</span>
<a name="l00270"></a>00270 <span class="comment"> * @param callRates - Vector to be filled in with call rates for chrx snps,</span>
<a name="l00271"></a>00271 <span class="comment"> * one for each chip.</span>
<a name="l00272"></a>00272 <span class="comment"> */</span>
<a name="l00273"></a><a class="code" href="classDmListener.html#aea66e6ec64eb07aa5fa5662f9afe020d">00273</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#a8d767a0f6220dc2c95e67e99586963ec" title="Get the call rate at a particular threshold.">DmListener::getCallRates</a>(std::vector&lt;float&gt; &amp;callRates)
<a name="l00274"></a>00274 {
<a name="l00275"></a>00275   <a class="code" href="classDmListener.html#a8d767a0f6220dc2c95e67e99586963ec" title="Get the call rate at a particular threshold.">getCallRates</a>(callRates, m_CallIndex);
<a name="l00276"></a>00276 }
<a name="l00277"></a>00277 <span class="comment"></span>
<a name="l00278"></a>00278 <span class="comment">/**</span>
<a name="l00279"></a>00279 <span class="comment"> * Get the call rate for a specific chip</span>
<a name="l00280"></a>00280 <span class="comment"> * @param chip - chip index to get call rate for</span>
<a name="l00281"></a>00281 <span class="comment"> * @return The call rate</span>
<a name="l00282"></a>00282 <span class="comment"> */</span>
<a name="l00283"></a><a class="code" href="classDmListener.html#acf5e0a66464e88b59681b709c5ada9ff">00283</a> <span class="keywordtype">float</span> <a class="code" href="classDmListener.html#a31271c403cb75acf6cff9a1773c2f7e4" title="Get the call rate for a specific chip.">DmListener::getCallRate</a>(<span class="keywordtype">int</span> chip)
<a name="l00284"></a>00284 {
<a name="l00285"></a>00285   <span class="keywordflow">return</span> <a class="code" href="classDmListener.html#a31271c403cb75acf6cff9a1773c2f7e4" title="Get the call rate for a specific chip.">getCallRate</a>(chip, m_CallIndex);
<a name="l00286"></a>00286 }
<a name="l00287"></a>00287 <span class="comment"></span>
<a name="l00288"></a>00288 <span class="comment">/**</span>
<a name="l00289"></a>00289 <span class="comment"> * Get the call rates for chrX snps for each chip.</span>
<a name="l00290"></a>00290 <span class="comment"> * @param callRates - Vector to be filled in with call rates for chrx snps,</span>
<a name="l00291"></a>00291 <span class="comment"> * one for each chip.</span>
<a name="l00292"></a>00292 <span class="comment"> */</span>
<a name="l00293"></a><a class="code" href="classDmListener.html#adafe0c0b162809dd12e0c6215b1a99c3">00293</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#af0af728c30461f3f002f14dfb318d958" title="Get the call rates for chrX snps at a particular threshold.">DmListener::getHetChrXRates</a>(std::vector&lt;float&gt; &amp;hetRates)
<a name="l00294"></a>00294 {
<a name="l00295"></a>00295   <a class="code" href="classDmListener.html#af0af728c30461f3f002f14dfb318d958" title="Get the call rates for chrX snps at a particular threshold.">getHetChrXRates</a>(hetRates, m_ChrXCallIndex);
<a name="l00296"></a>00296 }
<a name="l00297"></a>00297 <span class="comment"></span>
<a name="l00298"></a>00298 <span class="comment">/**</span>
<a name="l00299"></a>00299 <span class="comment"> * Get the chrX het rate for a specific chip</span>
<a name="l00300"></a>00300 <span class="comment"> * @param chip - the chip index to get call rate for</span>
<a name="l00301"></a>00301 <span class="comment"> * @return the chrC het rate</span>
<a name="l00302"></a>00302 <span class="comment"> */</span>
<a name="l00303"></a><a class="code" href="classDmListener.html#a37913e2040972bd96e946ef9f7aa023b">00303</a> <span class="keywordtype">float</span> <a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">DmListener::getHetChrXRate</a>(<span class="keywordtype">int</span> chip)
<a name="l00304"></a>00304 {
<a name="l00305"></a>00305   <span class="keywordflow">return</span> <a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">getHetChrXRate</a>(chip, m_ChrXCallIndex);
<a name="l00306"></a>00306 }
<a name="l00307"></a>00307 <span class="comment"></span>
<a name="l00308"></a>00308 <span class="comment">/**</span>
<a name="l00309"></a>00309 <span class="comment"> * Will this probeset work for DM? That is does it have MM probes and</span>
<a name="l00310"></a>00310 <span class="comment"> * matched quartets of probes on A and B allele?</span>
<a name="l00311"></a>00311 <span class="comment"> * @param ps - Probeset to check.</span>
<a name="l00312"></a>00312 <span class="comment"> * @return - true if ok for DM, false otherwise.</span>
<a name="l00313"></a>00313 <span class="comment"> */</span>
<a name="l00314"></a><a class="code" href="classDmListener.html#a7610d6a3dae4f64dbfa6a991994181b1">00314</a> <span class="keywordtype">bool</span> <a class="code" href="classDmListener.html#a7610d6a3dae4f64dbfa6a991994181b1" title="Will this probeset work for DM? That is does it have MM probes and matched quartets of probes on A an...">DmListener::okForDm</a>(<span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *ps)
<a name="l00315"></a>00315 {
<a name="l00316"></a>00316   <span class="keywordtype">bool</span> ok = <span class="keyword">true</span>;
<a name="l00317"></a>00317   <span class="keywordflow">if</span> (ps-&gt;<a class="code" href="classProbeSet.html#aa3e4170184169cc3bee6d958d74940ba" title="What type of probeset is this?">psType</a> != ProbeSet::GenoType) {
<a name="l00318"></a>00318     ok = <span class="keyword">false</span>;
<a name="l00319"></a>00319   } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (ps-&gt;<a class="code" href="classProbeSet.html#a6d8ed948634c42033bd0c35d2640b274" title="How many groups (or blocks) are there? Used extensively in genotyping chips.">numGroups</a> != 2 &amp;&amp; ps-&gt;<a class="code" href="classProbeSet.html#a6d8ed948634c42033bd0c35d2640b274" title="How many groups (or blocks) are there? Used extensively in genotyping chips.">numGroups</a> != 4) {
<a name="l00320"></a>00320     ok = <span class="keyword">false</span>;
<a name="l00321"></a>00321   } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (ps-&gt;<a class="code" href="classProbeSet.html#a23b9df97be8aacc01ad21642b8c966b9" title="Number of atoms in each group, used to get the offsets into the atoms vector.">atomsPerGroup</a>.empty()) {
<a name="l00322"></a>00322     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;DmListener::okForDm() - Can&#39;t have multiple groups, but atomsPerGroup be emtpy in probeset: &#39;&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot;&#39;&quot;</span>);
<a name="l00323"></a>00323     ok = <span class="keyword">false</span>;
<a name="l00324"></a>00324   } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (!ps-&gt;hasMM()) {
<a name="l00325"></a>00325     ok = <span class="keyword">false</span>;
<a name="l00326"></a>00326   } <span class="keywordflow">else</span> {
<a name="l00327"></a>00327     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> psIx = 0; psIx &lt; ps-&gt;<a class="code" href="classProbeSet.html#a6d8ed948634c42033bd0c35d2640b274" title="How many groups (or blocks) are there? Used extensively in genotyping chips.">numGroups</a> - 1; psIx += 2) {
<a name="l00328"></a>00328       <span class="keywordflow">if</span> (ps-&gt;<a class="code" href="classProbeSet.html#a23b9df97be8aacc01ad21642b8c966b9" title="Number of atoms in each group, used to get the offsets into the atoms vector.">atomsPerGroup</a>[psIx] != ps-&gt;<a class="code" href="classProbeSet.html#a23b9df97be8aacc01ad21642b8c966b9" title="Number of atoms in each group, used to get the offsets into the atoms vector.">atomsPerGroup</a>[psIx+1])
<a name="l00329"></a>00329         ok = <span class="keyword">false</span>;
<a name="l00330"></a>00330     }
<a name="l00331"></a>00331   }
<a name="l00332"></a>00332   <span class="keywordflow">return</span> ok;
<a name="l00333"></a>00333 }
<a name="l00334"></a>00334 
<a name="l00335"></a><a class="code" href="classDmListener.html#aa7d44b1b47aa3608f935c50522423c3e">00335</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#aa7d44b1b47aa3608f935c50522423c3e" title="Fill in a vector of CQuartet containing the cel file info for the given probeset.">DmListener::fillInQuartet</a>(vector&lt;CQuartet&gt; &amp;qVec, <span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *ps, <a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel)
<a name="l00336"></a>00336 {
<a name="l00337"></a>00337   assert(ps);
<a name="l00338"></a>00338   <span class="keywordflow">if</span> (!<a class="code" href="classDmListener.html#a7610d6a3dae4f64dbfa6a991994181b1" title="Will this probeset work for DM? That is does it have MM probes and matched quartets of probes on A an...">okForDm</a>(ps))
<a name="l00339"></a>00339     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Probeset &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot; not suitable for DM.&quot;</span>);
<a name="l00340"></a>00340 
<a name="l00341"></a>00341   <span class="comment">/* Genotype probe sets can have either 2 or 4 groups, for each</span>
<a name="l00342"></a>00342 <span class="comment">     type the even (0,2) are A alleles and odd (1,3) are B allele.s */</span>
<a name="l00343"></a>00343   <span class="keywordtype">int</span> atomIndex = 0;
<a name="l00344"></a>00344   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> psIx = 0; psIx &lt; ps-&gt;<a class="code" href="classProbeSet.html#a6d8ed948634c42033bd0c35d2640b274" title="How many groups (or blocks) are there? Used extensively in genotyping chips.">numGroups</a> - 1; psIx += 2) {
<a name="l00345"></a>00345     <a class="code" href="classErr.html#ad1a633f602fe60689cdc12d9ca1842e9" title="errAbort based version of assert with an error message.">Err::check</a>(ps-&gt;<a class="code" href="classProbeSet.html#a23b9df97be8aacc01ad21642b8c966b9" title="Number of atoms in each group, used to get the offsets into the atoms vector.">atomsPerGroup</a>[psIx] == ps-&gt;<a class="code" href="classProbeSet.html#a23b9df97be8aacc01ad21642b8c966b9" title="Number of atoms in each group, used to get the offsets into the atoms vector.">atomsPerGroup</a>[psIx+1],
<a name="l00346"></a>00346                <span class="stringliteral">&quot;DmListener::makeCall() - Though groups were symmetrical.&quot;</span>);
<a name="l00347"></a>00347     <span class="keywordtype">int</span> atomsPer = ps-&gt;<a class="code" href="classProbeSet.html#a23b9df97be8aacc01ad21642b8c966b9" title="Number of atoms in each group, used to get the offsets into the atoms vector.">atomsPerGroup</a>[psIx];
<a name="l00348"></a>00348     <span class="comment">/* Each block will have atomsPer atoms in there. */</span>
<a name="l00349"></a>00349     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> atomIx = 0; atomIx &lt; atomsPer; atomIx++) {
<a name="l00350"></a>00350       <span class="keywordtype">int</span> aIndex = atomIx + (atomIndex);
<a name="l00351"></a>00351       <span class="keywordtype">int</span> bIndex = aIndex + atomsPer;
<a name="l00352"></a>00352       <a class="code" href="classAtom.html" title="Represents a collection of probes that make up a unit for an intensity measurement of a particular se...">Atom</a> *A = ps-&gt;<a class="code" href="classProbeSet.html#ab92eb06095441bf193a6fce3f758540b" title="Atoms that make up this probe set.">atoms</a>[aIndex];
<a name="l00353"></a>00353       <a class="code" href="classAtom.html" title="Represents a collection of probes that make up a unit for an intensity measurement of a particular se...">Atom</a> *B = ps-&gt;<a class="code" href="classProbeSet.html#ab92eb06095441bf193a6fce3f758540b" title="Atoms that make up this probe set.">atoms</a>[bIndex];
<a name="l00354"></a>00354 
<a name="l00355"></a>00355       <span class="comment">// Fill in PM/MM pair</span>
<a name="l00356"></a>00356       <span class="keywordflow">if</span> (ps-&gt;hasMM()) {
<a name="l00357"></a>00357         <span class="keywordtype">int</span> aPM = -1, aMM = -1;
<a name="l00358"></a>00358         assert(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>.size() % 2 == 0);
<a name="l00359"></a>00359         assert(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>.size() % 2 == 0);
<a name="l00360"></a>00360         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> probeIx = 0; probeIx + 1 &lt; A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>.size(); probeIx += 2) {
<a name="l00361"></a>00361           <a class="code" href="classCQuartet.html" title="An object to probe quartet, each quartet contains 4 cells each cell contains intensity, variance and number of pixels, store variance instead of stdev.">CQuartet</a> q;
<a name="l00362"></a>00362           <span class="keywordtype">int</span> first = probeIx;
<a name="l00363"></a>00363           <span class="keywordtype">int</span> next = probeIx + 1;
<a name="l00364"></a>00364           <span class="comment">/* this is a little akward, need to pull out the a &amp; b allele info,</span>
<a name="l00365"></a>00365 <span class="comment">             but I don&#39;t think there is a hard and fast rule about the order</span>
<a name="l00366"></a>00366 <span class="comment">             of the pm and mm probes, so check both for both a &amp; b. */</span>
<a name="l00367"></a>00367 
<a name="l00368"></a>00368           <span class="comment">/* A allele */</span>
<a name="l00369"></a>00369           <span class="keywordflow">if</span> (Probe::isPm(*A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[first]) &amp;&amp; Probe::isMm(*A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[next])) {
<a name="l00370"></a>00370             aPM = first;
<a name="l00371"></a>00371             aMM = next;
<a name="l00372"></a>00372           } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (Probe::isPm(*A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[next]) &amp;&amp; Probe::isMm(*A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[first])) {
<a name="l00373"></a>00373             aPM = next;
<a name="l00374"></a>00374             aMM = first;
<a name="l00375"></a>00375           } <span class="keywordflow">else</span> {
<a name="l00376"></a>00376             <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Need a pm and mm probe in probeset: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot; atom: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(atomIx));
<a name="l00377"></a>00377           }
<a name="l00378"></a>00378           q.<a class="code" href="classCQuartet.html#acd1657016ba2fd98a5e7fa931adbfd1b" title="set intensity for allele type=allele and match type = matchtype">SetIntensity</a>(CQuartet::A_Allele, CQuartet::PM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a76d2432b4075dbb9f00a1cf306a3a614">GetIntensity</a>(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aPM]-&gt;id));
<a name="l00379"></a>00379           q.<a class="code" href="classCQuartet.html#acd1657016ba2fd98a5e7fa931adbfd1b" title="set intensity for allele type=allele and match type = matchtype">SetIntensity</a>(CQuartet::A_Allele, CQuartet::MM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a76d2432b4075dbb9f00a1cf306a3a614">GetIntensity</a>(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aMM]-&gt;id));
<a name="l00380"></a>00380           <span class="keywordtype">double</span> pm_stdev = cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a98f22aa88b61307c5340b7de102d70c8">GetStdv</a>(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aPM]-&gt;id);
<a name="l00381"></a>00381           <span class="keywordtype">double</span> mm_stdev = cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a98f22aa88b61307c5340b7de102d70c8">GetStdv</a>(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aMM]-&gt;id);
<a name="l00382"></a>00382           
<a name="l00383"></a>00383           q.<a class="code" href="classCQuartet.html#a2439587511c4089094a73879d00c43d6" title="set variance for allele type=allele and match type = matchtype">SetVariance</a>(CQuartet::A_Allele, CQuartet::PM, pm_stdev * pm_stdev);
<a name="l00384"></a>00384           q.<a class="code" href="classCQuartet.html#a2439587511c4089094a73879d00c43d6" title="set variance for allele type=allele and match type = matchtype">SetVariance</a>(CQuartet::A_Allele, CQuartet::MM, mm_stdev * mm_stdev);
<a name="l00385"></a>00385           q.<a class="code" href="classCQuartet.html#af834903290a6faa4bc2d362bd0009871" title="set number of pixels for allele type=allele and match type = matchtype">SetPixels</a>(CQuartet::A_Allele, CQuartet::PM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a9dfaab112224b9a4c52fb8e81a3e6843">GetPixels</a>(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aPM]-&gt;id));
<a name="l00386"></a>00386           q.<a class="code" href="classCQuartet.html#af834903290a6faa4bc2d362bd0009871" title="set number of pixels for allele type=allele and match type = matchtype">SetPixels</a>(CQuartet::A_Allele, CQuartet::MM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a9dfaab112224b9a4c52fb8e81a3e6843">GetPixels</a>(A-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aMM]-&gt;id));
<a name="l00387"></a>00387 
<a name="l00388"></a>00388           <span class="comment">/* B Allele */</span>
<a name="l00389"></a>00389           <span class="keywordflow">if</span> (Probe::isPm(*B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[first]) &amp;&amp; Probe::isMm(*B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[next])) {
<a name="l00390"></a>00390             aPM = first;
<a name="l00391"></a>00391             aMM = next;
<a name="l00392"></a>00392           } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (Probe::isPm(*B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[next]) &amp;&amp; Probe::isMm(*B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[first])) {
<a name="l00393"></a>00393             aPM = next;
<a name="l00394"></a>00394             aMM = first;
<a name="l00395"></a>00395           } <span class="keywordflow">else</span> {
<a name="l00396"></a>00396             <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Need a pm and mm probe in probeset: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot; atom: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(atomIx));
<a name="l00397"></a>00397           }
<a name="l00398"></a>00398           q.<a class="code" href="classCQuartet.html#acd1657016ba2fd98a5e7fa931adbfd1b" title="set intensity for allele type=allele and match type = matchtype">SetIntensity</a>(CQuartet::B_Allele, CQuartet::PM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a76d2432b4075dbb9f00a1cf306a3a614">GetIntensity</a>(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aPM]-&gt;id));
<a name="l00399"></a>00399           q.<a class="code" href="classCQuartet.html#acd1657016ba2fd98a5e7fa931adbfd1b" title="set intensity for allele type=allele and match type = matchtype">SetIntensity</a>(CQuartet::B_Allele, CQuartet::MM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a76d2432b4075dbb9f00a1cf306a3a614">GetIntensity</a>(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aMM]-&gt;id));
<a name="l00400"></a>00400           pm_stdev = cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a98f22aa88b61307c5340b7de102d70c8">GetStdv</a>(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aPM]-&gt;id);
<a name="l00401"></a>00401           mm_stdev = cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a98f22aa88b61307c5340b7de102d70c8">GetStdv</a>(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aMM]-&gt;id);
<a name="l00402"></a>00402           q.<a class="code" href="classCQuartet.html#a2439587511c4089094a73879d00c43d6" title="set variance for allele type=allele and match type = matchtype">SetVariance</a>(CQuartet::B_Allele, CQuartet::PM, pm_stdev * pm_stdev);
<a name="l00403"></a>00403           q.<a class="code" href="classCQuartet.html#a2439587511c4089094a73879d00c43d6" title="set variance for allele type=allele and match type = matchtype">SetVariance</a>(CQuartet::B_Allele, CQuartet::MM, mm_stdev * mm_stdev);
<a name="l00404"></a>00404           q.<a class="code" href="classCQuartet.html#af834903290a6faa4bc2d362bd0009871" title="set number of pixels for allele type=allele and match type = matchtype">SetPixels</a>(CQuartet::B_Allele, CQuartet::PM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a9dfaab112224b9a4c52fb8e81a3e6843">GetPixels</a>(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aPM]-&gt;id));
<a name="l00405"></a>00405           q.<a class="code" href="classCQuartet.html#af834903290a6faa4bc2d362bd0009871" title="set number of pixels for allele type=allele and match type = matchtype">SetPixels</a>(CQuartet::B_Allele, CQuartet::MM, cel-&gt;<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html#a9dfaab112224b9a4c52fb8e81a3e6843">GetPixels</a>(B-&gt;<a class="code" href="classAtom.html#a5953a4944a1e24b3f51a20ce92d8def2" title="Vector of probes that make up this atom. Order is important.">probes</a>[aMM]-&gt;id));
<a name="l00406"></a>00406           qVec.push_back(q);
<a name="l00407"></a>00407         }
<a name="l00408"></a>00408       } <span class="keywordflow">else</span> {
<a name="l00409"></a>00409         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Need a pm and mm probe in probeset: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot; atom: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(atomIx));
<a name="l00410"></a>00410       }
<a name="l00411"></a>00411     } <span class="comment">// for each atom</span>
<a name="l00412"></a>00412     atomIndex += (2 * atomsPer);
<a name="l00413"></a>00413   } <span class="comment">// for each group</span>
<a name="l00414"></a>00414 }
<a name="l00415"></a>00415 <span class="comment"></span>
<a name="l00416"></a>00416 <span class="comment">/**</span>
<a name="l00417"></a>00417 <span class="comment"> * Make DM genotyping call on a particular probeset given the data in the</span>
<a name="l00418"></a>00418 <span class="comment"> * cel file.</span>
<a name="l00419"></a>00419 <span class="comment"> *</span>
<a name="l00420"></a>00420 <span class="comment"> * @param ps - probeset to do genotyping call.</span>
<a name="l00421"></a>00421 <span class="comment"> * @param cel - cel file to use for data.</span>
<a name="l00422"></a>00422 <span class="comment"> *</span>
<a name="l00423"></a>00423 <span class="comment"> * @return Genotype call and confidence.</span>
<a name="l00424"></a>00424 <span class="comment"> */</span>
<a name="l00425"></a><a class="code" href="classDmListener.html#afe06545c789850e2eb2820e0a7bb6b99">00425</a> <a class="code" href="classDmListener_1_1DmCall.html" title="Utility class for result of DM algorithm.">DmListener::DmCall</a> <a class="code" href="classDmListener.html#afe06545c789850e2eb2820e0a7bb6b99" title="Make DM genotyping call on a particular probeset given the data in the cel file.">DmListener::makeCall</a>(<span class="keywordtype">float</span> hetMult, <span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *ps, <a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel)
<a name="l00426"></a>00426 {
<a name="l00427"></a>00427   assert(ps);
<a name="l00428"></a>00428   <span class="comment">// set up the result.</span>
<a name="l00429"></a>00429   <a class="code" href="classDmListener_1_1DmCall.html" title="Utility class for result of DM algorithm.">DmCall</a> result;
<a name="l00430"></a>00430   result.<a class="code" href="classDmListener_1_1DmCall.html#af21fd177519156d8dfd34bfa6dfe3855" title="The genotype call: NN(-1),AA(0),AB(1),BB(2).">call</a> = NN; <span class="comment">// -1; unknown</span>
<a name="l00431"></a>00431   result.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> = 0;
<a name="l00432"></a>00432   <span class="comment">// if this probeset isn&#39;t compatible with DM then return NN call.</span>
<a name="l00433"></a>00433   <span class="keywordflow">if</span> (!<a class="code" href="classDmListener.html#a7610d6a3dae4f64dbfa6a991994181b1" title="Will this probeset work for DM? That is does it have MM probes and matched quartets of probes on A an...">okForDm</a>(ps)) {
<a name="l00434"></a>00434     <span class="keywordflow">return</span> result;
<a name="l00435"></a>00435   }
<a name="l00436"></a>00436 
<a name="l00437"></a>00437   vector&lt;CQuartet&gt; qVec;
<a name="l00438"></a>00438   <a class="code" href="classDmListener.html#aa7d44b1b47aa3608f935c50522423c3e" title="Fill in a vector of CQuartet containing the cel file info for the given probeset.">fillInQuartet</a>(qVec, ps, cel);
<a name="l00439"></a>00439 
<a name="l00440"></a>00440   std::pair&lt;float, int&gt; call;
<a name="l00441"></a>00441   <span class="comment">// Here is the actual work.</span>
<a name="l00442"></a>00442   <span class="keywordflow">if</span> (<a class="code" href="DM_8cpp.html#aac00f1af0cf7f35503620cb04e4d8cfe" title="the main entry function of DM algorithm, which starts with a vector of probe quartets and some parame...">CallDM</a>(qVec, call, hetMult)) {
<a name="l00443"></a>00443     result.<a class="code" href="classDmListener_1_1DmCall.html#ad03436cdf482a423781e48f51a4515a8" title="Was the resulting call successful?">good</a> = <span class="keyword">true</span>;
<a name="l00444"></a>00444     result.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> = call.first;
<a name="l00445"></a>00445     result.<a class="code" href="classDmListener_1_1DmCall.html#af21fd177519156d8dfd34bfa6dfe3855" title="The genotype call: NN(-1),AA(0),AB(1),BB(2).">call</a> = <a class="code" href="namespaceaffx.html#a8368f9d25c1badaa812816c14dc29ec0" title="Convert an integer from a file to an internal GType ///.">GType_from_int</a>(call.second);
<a name="l00446"></a>00446   } <span class="keywordflow">else</span> {
<a name="l00447"></a>00447     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Dm call for probeset: &#39;&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot;&#39; failed.&quot;</span>);
<a name="l00448"></a>00448   }
<a name="l00449"></a>00449 
<a name="l00450"></a>00450   <span class="comment">// sanity check...</span>
<a name="l00451"></a>00451   <span class="comment">// assert(result.call &gt;= -1 &amp;&amp; result.call &lt;= 2);</span>
<a name="l00452"></a>00452   assert(<a class="code" href="namespaceaffx.html#a1511273f4d17d3a0f2744cef7aaa412b" title="Check that the gtype is AA,AB,BB,NN ///.">GType_valid</a>(result.<a class="code" href="classDmListener_1_1DmCall.html#af21fd177519156d8dfd34bfa6dfe3855" title="The genotype call: NN(-1),AA(0),AB(1),BB(2).">call</a>));
<a name="l00453"></a>00453   <span class="keywordflow">if</span> (result.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> &lt; 0 || result.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> &gt; 1) {
<a name="l00454"></a>00454     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;DmListener::makeCall() - Expecting p-value between 0 and 1, got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(result.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a>));
<a name="l00455"></a>00455   }
<a name="l00456"></a>00456   <span class="keywordflow">return</span> result;
<a name="l00457"></a>00457 }
<a name="l00458"></a>00458 
<a name="l00459"></a><a class="code" href="classDmListener.html#a9f3ed342294216d3e8f88a69f3a3e175">00459</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#a9f3ed342294216d3e8f88a69f3a3e175" title="Threshold the call based on the default threshold value.">DmListener::assignCall</a>(std::string &amp;name, <a class="code" href="classDmListener_1_1DmCall.html" title="Utility class for result of DM algorithm.">DmListener::DmCall</a> &amp;call, std::vector&lt;affx::GType&gt; &amp;genoCall)
<a name="l00460"></a>00460 {
<a name="l00461"></a>00461   <span class="comment">// keep statistics for all call rates.</span>
<a name="l00462"></a>00462   <span class="comment">// some SNPs we might see on multiple iterations (ie chrX snps)</span>
<a name="l00463"></a>00463   map&lt;string, vector&lt;bool&gt; &gt;::iterator mapIter;
<a name="l00464"></a>00464 
<a name="l00465"></a>00465   <span class="comment">// Initialize SNP to not seen, if not in map</span>
<a name="l00466"></a>00466   mapIter = m_SeenSNPs.find(name);
<a name="l00467"></a>00467   <span class="keywordflow">if</span> (mapIter == m_SeenSNPs.end()) {
<a name="l00468"></a>00468     m_SeenSNPs[name].resize(m_MaxChips, <span class="keyword">false</span>);
<a name="l00469"></a>00469   }
<a name="l00470"></a>00470 
<a name="l00471"></a>00471   <span class="comment">// If not seen, the compute stats</span>
<a name="l00472"></a>00472   <span class="keywordflow">if</span> (m_SeenSNPs[name][m_ChipCount] == <span class="keyword">false</span>) {
<a name="l00473"></a>00473     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> threshIx = 0; threshIx &lt; m_Thresholds.size(); threshIx++) {
<a name="l00474"></a>00474       <span class="keywordflow">if</span> (call.<a class="code" href="classDmListener_1_1DmCall.html#ad03436cdf482a423781e48f51a4515a8" title="Was the resulting call successful?">good</a> &amp;&amp; call.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> &lt; m_Thresholds[threshIx]) {
<a name="l00475"></a>00475         m_PassCalls[threshIx][m_ChipCount]++;
<a name="l00476"></a>00476       }
<a name="l00477"></a>00477     }
<a name="l00478"></a>00478     <span class="keywordflow">if</span> (call.<a class="code" href="classDmListener_1_1DmCall.html#ad03436cdf482a423781e48f51a4515a8" title="Was the resulting call successful?">good</a>) {
<a name="l00479"></a>00479       m_TotalCalls[m_ChipCount]++;
<a name="l00480"></a>00480     }
<a name="l00481"></a>00481     <span class="comment">//Verbose::out(1,&quot;Adding Call: &quot; + ToStr(name) + &quot;, &quot; + ToStr(call.conf) + &quot;, &quot; + ToStr(call.good));</span>
<a name="l00482"></a>00482     <span class="keywordflow">if</span> (m_ChrXSnps.find(name) != m_ChrXSnps.end()) {
<a name="l00483"></a>00483       <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> threshIx = 0; threshIx &lt; m_Thresholds.size(); threshIx++) {
<a name="l00484"></a>00484         <span class="keywordflow">if</span> (call.<a class="code" href="classDmListener_1_1DmCall.html#ad03436cdf482a423781e48f51a4515a8" title="Was the resulting call successful?">good</a> &amp;&amp; call.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> &lt; m_Thresholds[threshIx] &amp;&amp; call.<a class="code" href="classDmListener_1_1DmCall.html#af21fd177519156d8dfd34bfa6dfe3855" title="The genotype call: NN(-1),AA(0),AB(1),BB(2).">call</a> == <a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a83b51b7ed2e4a1b4e23378773929465b" title="Heterozygous, one copy of both A and B. (1).">AB</a>) {
<a name="l00485"></a>00485           m_HetChrXCalls[threshIx][m_ChipCount]++;
<a name="l00486"></a>00486         }
<a name="l00487"></a>00487       }
<a name="l00488"></a>00488       m_ChrXCalls[m_ChipCount]++;
<a name="l00489"></a>00489     }
<a name="l00490"></a>00490     m_SeenSNPs[name][m_ChipCount] = <span class="keyword">true</span>;
<a name="l00491"></a>00491   }
<a name="l00492"></a>00492 
<a name="l00493"></a>00493   <span class="comment">// use index 0 to make actual call, when DM fails we set to NN.</span>
<a name="l00494"></a>00494   <span class="keywordflow">if</span> (!call.<a class="code" href="classDmListener_1_1DmCall.html#ad03436cdf482a423781e48f51a4515a8" title="Was the resulting call successful?">good</a> || call.<a class="code" href="classDmListener_1_1DmCall.html#a898f7312a362291ee486b72921807561" title="p-value for DM call.">conf</a> &gt;= m_Thresholds[0]) {
<a name="l00495"></a>00495     call.<a class="code" href="classDmListener_1_1DmCall.html#af21fd177519156d8dfd34bfa6dfe3855" title="The genotype call: NN(-1),AA(0),AB(1),BB(2).">call</a> = NN;
<a name="l00496"></a>00496   }
<a name="l00497"></a>00497   genoCall[m_ChipCount] = call.<a class="code" href="classDmListener_1_1DmCall.html#af21fd177519156d8dfd34bfa6dfe3855" title="The genotype call: NN(-1),AA(0),AB(1),BB(2).">call</a>;
<a name="l00498"></a>00498 }
<a name="l00499"></a>00499 <span class="comment"></span>
<a name="l00500"></a>00500 <span class="comment">/**</span>
<a name="l00501"></a>00501 <span class="comment"> * Process another cel files worth of data.</span>
<a name="l00502"></a>00502 <span class="comment"> *</span>
<a name="l00503"></a>00503 <span class="comment"> * @param cel - Filehandle to open cel file.</span>
<a name="l00504"></a>00504 <span class="comment"> */</span>
<a name="l00505"></a><a class="code" href="classDmListener.html#ac0c8a922e606e6c9d3815aee7a4577cc">00505</a> <span class="keywordtype">void</span> <a class="code" href="classDmListener.html#ac0c8a922e606e6c9d3815aee7a4577cc" title="Process another cel files worth of data.">DmListener::newChip</a>(<a class="code" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel)
<a name="l00506"></a>00506 {
<a name="l00507"></a>00507   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> psIx = 0; psIx &lt; m_PlVec.size(); psIx++) {
<a name="l00508"></a>00508     <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *ps = <a class="code" href="classProbeListFactory.html#aaed559bbfe723ed4fe225f805a2411e8">ProbeListFactory::asProbeSet</a>(m_PlVec[psIx]);
<a name="l00509"></a>00509     std::map&lt;string, std::vector&lt;GType&gt; &gt;::iterator mapIter;
<a name="l00510"></a>00510     <a class="code" href="classDmListener_1_1DmCall.html" title="Utility class for result of DM algorithm.">DmCall</a> call = <a class="code" href="classDmListener.html#afe06545c789850e2eb2820e0a7bb6b99" title="Make DM genotyping call on a particular probeset given the data in the cel file.">makeCall</a>(m_HetMultiplier, ps, cel);
<a name="l00511"></a>00511 
<a name="l00512"></a>00512     std::string name = ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>;
<a name="l00513"></a>00513     mapIter = m_KnownCalls.find(name);
<a name="l00514"></a>00514     <span class="keywordflow">if</span> (mapIter == m_KnownCalls.end()) {
<a name="l00515"></a>00515       <span class="comment">// First time seen.</span>
<a name="l00516"></a>00516       std::vector&lt;GType&gt; genoCall(m_MaxChips);
<a name="l00517"></a>00517       <a class="code" href="classDmListener.html#a9f3ed342294216d3e8f88a69f3a3e175" title="Threshold the call based on the default threshold value.">assignCall</a>(name, call, genoCall);
<a name="l00518"></a>00518       <span class="keywordflow">if</span> (m_KeepGenotypes)
<a name="l00519"></a>00519         m_KnownCalls[name] = genoCall;
<a name="l00520"></a>00520     } <span class="keywordflow">else</span> {
<a name="l00521"></a>00521       <span class="comment">// More have been seen</span>
<a name="l00522"></a>00522       <a class="code" href="classDmListener.html#a9f3ed342294216d3e8f88a69f3a3e175" title="Threshold the call based on the default threshold value.">assignCall</a>(name, call, mapIter-&gt;second);
<a name="l00523"></a>00523     }
<a name="l00524"></a>00524     <span class="keyword">delete</span> ps;
<a name="l00525"></a>00525   }
<a name="l00526"></a>00526   <span class="keywordflow">if</span> (m_ChipCount == 0) {
<a name="l00527"></a>00527     <span class="comment">// trim excess capacity.</span>
<a name="l00528"></a>00528     std::map&lt;string, std::vector&lt;affx::GType&gt; &gt; tmp(m_KnownCalls);
<a name="l00529"></a>00529     m_KnownCalls.swap(tmp);
<a name="l00530"></a>00530   }
<a name="l00531"></a>00531 
<a name="l00532"></a>00532   <span class="comment">// track the gender info over multiple iterations</span>
<a name="l00533"></a>00533   <span class="keywordflow">if</span> (m_ChipCount &gt;= m_Seen.size())
<a name="l00534"></a>00534     m_Seen.resize(m_ChipCount + 1, <span class="keyword">false</span>);
<a name="l00535"></a>00535   <span class="keywordflow">if</span> (!m_Seen[m_ChipCount]) {
<a name="l00536"></a>00536     std::vector&lt;ChipSummary::Metric&gt; metrics;
<a name="l00537"></a>00537     <a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>.push_back(metrics);
<a name="l00538"></a>00538     <span class="keywordflow">if</span> (m_ChrXSnps.size() &gt; 0) {
<a name="l00539"></a>00539       <a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>[m_ChipCount].push_back(<a class="code" href="classChipSummary_1_1Metric.html" title="Class for storing metrics.">ChipSummary::Metric</a>(m_GenderName + <span class="stringliteral">&quot;_gender&quot;</span>,
<a name="l00540"></a>00540                                             <a class="code" href="namespaceaffx.html#a91c5e36f4ed8c88d90f3aa3086ca95df" title="convert the BioType Gender to a string ///">affx::getGenderString</a>(getGender(m_ChipCount))));
<a name="l00541"></a>00541       <a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>[m_ChipCount].push_back(<a class="code" href="classChipSummary_1_1Metric.html" title="Class for storing metrics.">ChipSummary::Metric</a>(m_GenderName + <span class="stringliteral">&quot;_gender_chrX_het_rate&quot;</span>,
<a name="l00542"></a>00542                                             <a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">getHetChrXRate</a>(m_ChipCount)));
<a name="l00543"></a>00543       m_FirstPassGender[m_ChipCount] = getGender(m_ChipCount);
<a name="l00544"></a>00544       m_FirstPassHetRate[m_ChipCount] = <a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">getHetChrXRate</a>(m_ChipCount);
<a name="l00545"></a>00545     }
<a name="l00546"></a>00546     <a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>[m_ChipCount].push_back(<a class="code" href="classChipSummary_1_1Metric.html" title="Class for storing metrics.">ChipSummary::Metric</a>(m_CallRateName,
<a name="l00547"></a>00547                                           <a class="code" href="classDmListener.html#a31271c403cb75acf6cff9a1773c2f7e4" title="Get the call rate for a specific chip.">getCallRate</a>(m_ChipCount)));
<a name="l00548"></a>00548     m_Seen[m_ChipCount] = <span class="keyword">true</span>;
<a name="l00549"></a>00549   } <span class="keywordflow">else</span> {
<a name="l00550"></a>00550     <span class="comment">// Update call rate</span>
<a name="l00551"></a>00551     <span class="keywordtype">bool</span> found = <span class="keyword">false</span>;
<a name="l00552"></a>00552     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; <a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>[m_ChipCount].size(); i++) {
<a name="l00553"></a>00553       <span class="keywordflow">if</span> (<a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>[m_ChipCount][i].m_Name == m_CallRateName) {
<a name="l00554"></a>00554         <span class="keywordtype">float</span> callRate = <a class="code" href="classDmListener.html#a31271c403cb75acf6cff9a1773c2f7e4" title="Get the call rate for a specific chip.">getCallRate</a>(m_ChipCount);
<a name="l00555"></a>00555         <a class="code" href="classChipSummary.html#a9dfa0a4049fef93ff9ec0d45f548c6dc" title="Stats for ChipSummary.">m_SummaryStats</a>[m_ChipCount][i].m_Double = callRate;
<a name="l00556"></a>00556         <span class="comment">//Verbose::out(1,&quot;Call rate update on chip &quot; + ToStr(m_ChipCount) + &quot; to &quot; + ToStr(callRate));</span>
<a name="l00557"></a>00557         found = <span class="keyword">true</span>;
<a name="l00558"></a>00558       }
<a name="l00559"></a>00559     }
<a name="l00560"></a>00560     <span class="keywordflow">if</span> (!found) {
<a name="l00561"></a>00561       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Unable to update dm-listener call rate for chip &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_ChipCount));
<a name="l00562"></a>00562     }
<a name="l00563"></a>00563 
<a name="l00564"></a>00564     <span class="comment">// We expect no change in gender and chrX het rate -- ie all</span>
<a name="l00565"></a>00565     <span class="comment">// the chrX snps must be seen on the first pass</span>
<a name="l00566"></a>00566     <span class="keywordflow">if</span> (m_ChrXSnps.size() &gt; 0) {
<a name="l00567"></a>00567       <span class="keywordflow">if</span> (fabs(m_FirstPassHetRate[m_ChipCount] - <a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">getHetChrXRate</a>(m_ChipCount)) &gt; 0.00001)
<a name="l00568"></a>00568         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<span class="stringliteral">&quot;DmListener chrX het rate changed between iterations!&quot;</span>) +
<a name="l00569"></a>00569                       <span class="stringliteral">&quot; It was &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_FirstPassHetRate[m_ChipCount]) +
<a name="l00570"></a>00570                       <span class="stringliteral">&quot; but is now &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classDmListener.html#a036a9ee28deac748a2bb93ff9ef38308" title="Get the chrX het rate for a specific chip.">getHetChrXRate</a>(m_ChipCount)));
<a name="l00571"></a>00571       <span class="keywordflow">if</span> (m_FirstPassGender[m_ChipCount] != getGender(m_ChipCount))
<a name="l00572"></a>00572         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<span class="stringliteral">&quot;DmListener computed gender changed between iterations!&quot;</span>) +
<a name="l00573"></a>00573                       <span class="stringliteral">&quot; It was &quot;</span> + <a class="code" href="namespaceaffx.html#a91c5e36f4ed8c88d90f3aa3086ca95df" title="convert the BioType Gender to a string ///">affx::getGenderString</a>(m_FirstPassGender[m_ChipCount]) +
<a name="l00574"></a>00574                       <span class="stringliteral">&quot; but is now &quot;</span> + <a class="code" href="namespaceaffx.html#a91c5e36f4ed8c88d90f3aa3086ca95df" title="convert the BioType Gender to a string ///">affx::getGenderString</a>(getGender(m_ChipCount)));
<a name="l00575"></a>00575     }
<a name="l00576"></a>00576   }
<a name="l00577"></a>00577 
<a name="l00578"></a>00578   <a class="code" href="classChipSummary.html#a81304a00c3b53c6ae1feb65d07d883fb" title="Set Valid State.">setValid</a>(<span class="keyword">true</span>);
<a name="l00579"></a>00579   m_ChipCount++;
<a name="l00580"></a>00580 }
<a name="l00581"></a>00581 <span class="comment"></span>
<a name="l00582"></a>00582 <span class="comment">/**</span>
<a name="l00583"></a>00583 <span class="comment"> * Get the genotype calls for a particular probeset.</span>
<a name="l00584"></a>00584 <span class="comment"> * @param name - name of probeset to get genotype calls for.</span>
<a name="l00585"></a>00585 <span class="comment"> * @return - A vector of the calls from dm algorithm.</span>
<a name="l00586"></a>00586 <span class="comment"> */</span>
<a name="l00587"></a><a class="code" href="classDmListener.html#a5558d30d7f436d257fbae915d29874fd">00587</a> std::vector&lt;affx::GType&gt; <a class="code" href="classDmListener.html#a5558d30d7f436d257fbae915d29874fd" title="Get the genotype calls for a particular probeset.">DmListener::getGenoCalls</a>(<span class="keyword">const</span> std::string &amp;name)
<a name="l00588"></a>00588 {
<a name="l00589"></a>00589   <span class="keywordflow">if</span> (!m_KeepGenotypes)
<a name="l00590"></a>00590     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Genotypes requested from DmListener bug flag to keep them was set to false!&quot;</span>);
<a name="l00591"></a>00591   std::map&lt;std::string, std::vector&lt;affx::GType&gt; &gt;::iterator mapIter;
<a name="l00592"></a>00592   mapIter = m_KnownCalls.find(name);
<a name="l00593"></a>00593   <span class="keywordflow">if</span> (mapIter == m_KnownCalls.end()) {
<a name="l00594"></a>00594     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;DmListener::getGenoCall() - Can&#39;t find genotypes for name: &quot;</span> + name);
<a name="l00595"></a>00595   }
<a name="l00596"></a>00596   <span class="keywordflow">return</span> mapIter-&gt;second;
<a name="l00597"></a>00597 }
<a name="l00598"></a>00598 <span class="comment"></span>
<a name="l00599"></a>00599 <span class="comment">/**</span>
<a name="l00600"></a>00600 <span class="comment"> * Set the full name to use for the call rate metric reported</span>
<a name="l00601"></a>00601 <span class="comment"> * via the chip summary interface.</span>
<a name="l00602"></a>00602 <span class="comment"> * @param label - the full name to use for the call rate metric</span>
<a name="l00603"></a>00603 <span class="comment"> */</span>
<a name="l00604"></a>00604 <span class="keywordtype">void</span> DmListener::setCallRateName(std::string label) {
<a name="l00605"></a>00605   m_CallRateName = label;
<a name="l00606"></a>00606 }
<a name="l00607"></a>00607 <span class="comment"></span>
<a name="l00608"></a>00608 <span class="comment">/**</span>
<a name="l00609"></a>00609 <span class="comment"> * Set the prefix to use on chip summary metric labels. Defaults</span>
<a name="l00610"></a>00610 <span class="comment"> * to &quot;dm-listener&quot;.</span>
<a name="l00611"></a>00611 <span class="comment"> * @param prefix - the prefix to use</span>
<a name="l00612"></a>00612 <span class="comment"> */</span>
<a name="l00613"></a>00613 <span class="keywordtype">void</span> DmListener::setPrefix(std::string prefix) {
<a name="l00614"></a>00614   m_Prefix = prefix;
<a name="l00615"></a>00615   m_CallRateName = m_Prefix + <span class="stringliteral">&quot;_call_rate&quot;</span>;
<a name="l00616"></a>00616 }
<a name="l00617"></a>00617 
<a name="l00618"></a>00618 <span class="comment">/* Quick and dirty. */</span>
<a name="l00619"></a>00619 <span class="keywordtype">float</span> DmListener::square(<span class="keywordtype">float</span> x) {
<a name="l00620"></a>00620   <span class="keywordflow">return</span> x * x;
<a name="l00621"></a>00621 }
<a name="l00622"></a>00622 
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:51 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
