<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: translation/GeneCall.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_b6d6d4a2e58a0f89851f849ab50b1d34.html">translation</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>GeneCall.cpp</h1>  </div>
</div>
<div class="contents">
<a href="GeneCall_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2008 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment">/**</span>
<a name="l00022"></a>00022 <span class="comment"> * @file   GeneCall.cpp</span>
<a name="l00023"></a>00023 <span class="comment"> * @author Mybrid Spalding</span>
<a name="l00024"></a>00024 <span class="comment"> * @date   Tue Mar 18 09:04:13 PDT 2008</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief  Business object that is the heart of the matter, class with the algorithm that translates experiment data.</span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="preprocessor">#include &quot;<a class="code" href="GeneCall_8h.html" title="Business object that is the heart of the matter, class with the algorithm that translates experiment ...">translation/GeneCall.h</a>&quot;</span>
<a name="l00029"></a>00029 <span class="comment">//</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="CallResults_8h.html" title="Binary representation of allele translation call.">translation/CallResults.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="GenotypeTableModel_8h.html" title="Genotype Data (CHP) class.">translation/GenotypeTableModel.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="TranslationTableModel_8h.html" title="Single instance data model class that wraps the translation file in the TranslationInputTsvModel.">translation/TranslationTableModel.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="comment">//</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="Err_8h.html" title="Some utility error functions, currently prints message and aborts().">util/Err.h</a>&quot;</span> <span class="comment">// includes &quot;util/Verbose.h&quot;</span>
<a name="l00035"></a>00035 <span class="comment">//</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &lt;sstream&gt;</span>
<a name="l00037"></a>00037 <span class="comment">//</span>
<a name="l00038"></a>00038 
<a name="l00039"></a>00039 
<a name="l00040"></a>00040 <span class="keyword">using namespace </span>std;
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00044"></a>00044 <span class="comment">/**</span>
<a name="l00045"></a>00045 <span class="comment"> * GeneCall::GeneCall</span>
<a name="l00046"></a>00046 <span class="comment"> * Synopsis:</span>
<a name="l00047"></a>00047 <span class="comment"> *</span>
<a name="l00048"></a>00048 <span class="comment"> * Translation Table constructer. The translation table row order is</span>
<a name="l00049"></a>00049 <span class="comment"> * per marker. The column order starting with Reference, Variant</span>
<a name="l00050"></a>00050 <span class="comment"> * and then A1 - ANN is what is being defined here as a call.</span>
<a name="l00051"></a>00051 <span class="comment"> * The translation table is rows of markes with columns of calls.</span>
<a name="l00052"></a>00052 <span class="comment"> *</span>
<a name="l00053"></a>00053 <span class="comment"> * Each gene has a number of calls designated. The GeneCall object</span>
<a name="l00054"></a>00054 <span class="comment"> * represents the entire set of calls for a gene. The GeneCall is</span>
<a name="l00055"></a>00055 <span class="comment"> * a set of sets, where the sets are CallSets. The CallSets are sets</span>
<a name="l00056"></a>00056 <span class="comment"> * of CallElements where the CallElements represent the probset id</span>
<a name="l00057"></a>00057 <span class="comment"> * with reference and variant alleles.</span>
<a name="l00058"></a>00058 <span class="comment"> *</span>
<a name="l00059"></a>00059 <span class="comment"> * A GeneCall then represents the set of columns A1 - A50 (ANN) of the</span>
<a name="l00060"></a>00060 <span class="comment"> * translation table for a particular gene.</span>
<a name="l00061"></a>00061 <span class="comment"> *</span>
<a name="l00062"></a>00062 <span class="comment"> * The columns are chopped up at the Gene level. Within</span>
<a name="l00063"></a>00063 <span class="comment"> * a gene, the column is subdivided by haplotype or marker.</span>
<a name="l00064"></a>00064 <span class="comment"> * A column, like A1, then will have one GeneCall per marker where</span>
<a name="l00065"></a>00065 <span class="comment"> * the &quot;Haplotype&quot; column is &quot;N&quot;. For those markers of the gene</span>
<a name="l00066"></a>00066 <span class="comment"> * where &quot;Haplotype&quot; column is &quot;Y&quot; then exactly one GeneCall is created.</span>
<a name="l00067"></a>00067 <span class="comment"> *</span>
<a name="l00068"></a>00068 <span class="comment"> * NOTE: While in general the &quot;Haplotype&quot; column is &quot;Y&quot; records are grouped</span>
<a name="l00069"></a>00069 <span class="comment"> * together for a particular gene in the translation table, this is not</span>
<a name="l00070"></a>00070 <span class="comment"> * always the case. Probe sets with &quot;Haplotype&quot; column of &quot;Y&quot; can be</span>
<a name="l00071"></a>00071 <span class="comment"> * interpersed with the &quot;Haplotype&quot; column &quot;N&quot; probesets.</span>
<a name="l00072"></a>00072 <span class="comment"> *</span>
<a name="l00073"></a>00073 <span class="comment"> * For a particular gene there is ever only one haplotype group.</span>
<a name="l00074"></a>00074 <span class="comment"> * That group is designated by the set of markers for the gene with</span>
<a name="l00075"></a>00075 <span class="comment"> * &quot;Haplotype&quot; column indicator of &quot;Y&quot;.</span>
<a name="l00076"></a>00076 <span class="comment"> * A gene haplotype call (GeneCall) then corresponds to a column for</span>
<a name="l00077"></a>00077 <span class="comment"> * all those rows of a gene in the translation table within a haplotype group.</span>
<a name="l00078"></a>00078 <span class="comment"> *</span>
<a name="l00079"></a>00079 <span class="comment"> * This constructor takes in the header row for a particular gene</span>
<a name="l00080"></a>00080 <span class="comment"> * and then instantiates all the CallSets for all the columns.</span>
<a name="l00081"></a>00081 <span class="comment"> *</span>
<a name="l00082"></a>00082 <span class="comment"> * The CallSets for each call is subsequently filled with the probe set ids,</span>
<a name="l00083"></a>00083 <span class="comment"> * reference and variant alleles when the translation table is scanned.</span>
<a name="l00084"></a>00084 <span class="comment"> *</span>
<a name="l00085"></a>00085 <span class="comment"> * In other words, multiple data structures are being created when the</span>
<a name="l00086"></a>00086 <span class="comment"> * translation table is being scanned. To prevent scanning the table</span>
<a name="l00087"></a>00087 <span class="comment"> * multiple times this constructor simply instantiates the empty CallSet</span>
<a name="l00088"></a>00088 <span class="comment"> * per call and the callee (TranslationTable) is responsible for filling</span>
<a name="l00089"></a>00089 <span class="comment"> * in the CallSets per call as the subsequent rows are being scanned.</span>
<a name="l00090"></a>00090 <span class="comment"> *</span>
<a name="l00091"></a>00091 <span class="comment"> *</span>
<a name="l00092"></a>00092 <span class="comment"> * @param rte - RunTimeEnvironment which contains options</span>
<a name="l00093"></a>00093 <span class="comment"> * @param ttm - TranslationTableModel with the meta data about the call.</span>
<a name="l00094"></a>00094 <span class="comment"> * @param headerRow - the row in the ttm that has the meta data.</span>
<a name="l00095"></a>00095 <span class="comment"> * @param isHaplotypeCall - if the data row has been predetermined to</span>
<a name="l00096"></a>00096 <span class="comment"> *   have &quot;Y&quot; in the &quot;Haplotype&quot; column.</span>
<a name="l00097"></a>00097 <span class="comment"> *</span>
<a name="l00098"></a>00098 <span class="comment"> */</span>
<a name="l00099"></a>00099 <span class="comment">/*****************************************************************************/</span>
<a name="l00100"></a><a class="code" href="classGeneCall.html#acbd850ffa3532b1f6afc0712a48f0dde">00100</a> <a class="code" href="classGeneCall.html#acbd850ffa3532b1f6afc0712a48f0dde" title="GeneCall::GeneCall Synopsis:">GeneCall::GeneCall</a>(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00101"></a>00101                    <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm,
<a name="l00102"></a>00102                    <span class="keyword">const</span> <span class="keywordtype">int</span> headerRow,
<a name="l00103"></a>00103                    <span class="keywordtype">bool</span> isHaplotypeCall)
<a name="l00104"></a>00104 
<a name="l00105"></a>00105 {
<a name="l00106"></a>00106 
<a name="l00107"></a>00107   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>((headerRow &gt;= 0) &amp;&amp; (headerRow &lt; ttm.size()), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00108"></a>00108 
<a name="l00109"></a>00109   m_gene = ttm.m_rows[headerRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_GENE)];
<a name="l00110"></a>00110 
<a name="l00111"></a>00111   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <span class="stringliteral">&quot;GeneCall::GeneCall initializing gene: &quot;</span> + m_gene);
<a name="l00112"></a>00112 
<a name="l00113"></a>00113   <span class="comment">// Initialize the m_alleleSet for each possible call.</span>
<a name="l00114"></a>00114 
<a name="l00115"></a>00115   std::stringstream alleleNamesSStr;
<a name="l00116"></a>00116 
<a name="l00117"></a>00117   pcrecpp::RE reCommentedOut(<span class="stringliteral">&quot;^\\s*#&quot;</span>);
<a name="l00118"></a>00118   <span class="comment">// Skip commented out column names.</span>
<a name="l00119"></a>00119   std::map&lt;int, int&gt; columnToAlleleIndex;
<a name="l00120"></a>00120   std::map&lt;std::string, int&gt; numHaplotypeMarkers;
<a name="l00121"></a>00121 
<a name="l00122"></a>00122 
<a name="l00123"></a>00123   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_REFERENCE); i &lt; ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ae33ba37f59967a7bd0caefe58de785ed" title="TranslationInputTsvTableModel::columnCount Synopsis:">columnCount</a>(); i++) {
<a name="l00124"></a>00124 
<a name="l00125"></a>00125 
<a name="l00126"></a>00126     <span class="comment">// For markers, only add the allele if the header is copy number 0.</span>
<a name="l00127"></a>00127     <span class="comment">// We can only tell if the header is copy number zero if the data</span>
<a name="l00128"></a>00128     <span class="comment">// in the following row equals &quot;0&quot;;</span>
<a name="l00129"></a>00129     <span class="keywordflow">if</span> (!isHaplotypeCall &amp;&amp; (i &gt;= ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_ALLELE_START)) &amp;&amp; (((headerRow + 1) &gt;= ttm.size()) || (ttm.m_rows[headerRow +1][i] != <span class="stringliteral">&quot;0&quot;</span>))) {
<a name="l00130"></a>00130       <span class="keywordflow">continue</span>;
<a name="l00131"></a>00131     }
<a name="l00132"></a>00132 
<a name="l00133"></a>00133     std::string alleleName = ttm.<a class="code" href="classTranslationTableModel.html#a9569cee6d0e635259f07dbd8cbd99596" title="TranslationTableModel::getHeaderRowColumnName Synopsis: A header row repeats the descriptive headers ...">getHeaderRowColumnName</a>(headerRow, i);
<a name="l00134"></a>00134 
<a name="l00135"></a>00135     <span class="keywordflow">if</span> (<a class="code" href="classGeneCall.html#ac9276bb5f94a5916eefc26aacdbbc027" title="GeneCall::existsAlleleCallSet Synopsis:">existsAlleleCallSet</a>(alleleName)) {
<a name="l00136"></a>00136       <span class="comment">// Copy Number 0 applies to all markers, ignore it.</span>
<a name="l00137"></a>00137       <span class="keywordflow">if</span> (((headerRow + 1) &lt; ttm.size()) &amp;&amp; (ttm.m_rows[headerRow +1][i] == <span class="stringliteral">&quot;0&quot;</span>)) {
<a name="l00138"></a>00138         <span class="keywordflow">continue</span>;
<a name="l00139"></a>00139       }
<a name="l00140"></a>00140 
<a name="l00141"></a>00141       <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(ADT_VERBOSE_NORMAL, <span class="stringliteral">&quot;Duplicate allele name found: &quot;</span> + alleleName + <span class="stringliteral">&quot; for translation file row: &quot;</span> + ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ac209f9165791ba0d30aa905965eba1f6" title="TranslationInputTsvTableModel::getRowAsString Synopsis:">getRowAsString</a>(headerRow));
<a name="l00142"></a>00142     }
<a name="l00143"></a>00143 
<a name="l00144"></a>00144     <span class="comment">// All allele name columns will have a name until we get to the last one.</span>
<a name="l00145"></a>00145     <span class="keywordflow">if</span> (!alleleName.empty() &amp;&amp; !reCommentedOut.PartialMatch(alleleName)) {
<a name="l00146"></a>00146       columnToAlleleIndex[i] = m_alleleSet.size();
<a name="l00147"></a>00147       m_alleleSet.push_back(<a class="code" href="classCallSet.html">CallSet</a>(alleleName, i));
<a name="l00148"></a>00148       alleleNamesSStr &lt;&lt; alleleName &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
<a name="l00149"></a>00149       numHaplotypeMarkers[alleleName] = 0;
<a name="l00150"></a>00150     }
<a name="l00151"></a>00151   }
<a name="l00152"></a>00152 
<a name="l00153"></a>00153   <span class="keywordflow">if</span> (isHaplotypeCall) {
<a name="l00154"></a>00154     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = headerRow; (i &lt; ttm.size()) &amp;&amp; (ttm.m_rows[i][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_GENE)] == m_gene); i++) {
<a name="l00155"></a>00155       <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_REFERENCE);
<a name="l00156"></a>00156            j &lt; ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ae33ba37f59967a7bd0caefe58de785ed" title="TranslationInputTsvTableModel::columnCount Synopsis:">columnCount</a>(); j++) {
<a name="l00157"></a>00157 
<a name="l00158"></a>00158         <span class="keywordflow">if</span> (!ttm.m_rows[i][j].empty() &amp;&amp;
<a name="l00159"></a>00159             (ttm.m_rows[i][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_HAPLOTYPE)] == <span class="stringliteral">&quot;Y&quot;</span>)) {
<a name="l00160"></a>00160           numHaplotypeMarkers[m_alleleSet[columnToAlleleIndex[j]].m_name]++;
<a name="l00161"></a>00161         }
<a name="l00162"></a>00162       }
<a name="l00163"></a>00163     }
<a name="l00164"></a>00164     <span class="comment">// If an Allele doesn&#39;t belong to a haplotype group, then just remove it.</span>
<a name="l00165"></a>00165     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = m_alleleSet.size() - 1; i &gt;= 0 ; i--) {
<a name="l00166"></a>00166       <span class="keywordflow">if</span> ((m_alleleSet[i].m_columnInTranslationTable &gt;
<a name="l00167"></a>00167            ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_ALLELE_START)) &amp;&amp;
<a name="l00168"></a>00168           (numHaplotypeMarkers[m_alleleSet[i].m_name] == 0)) {
<a name="l00169"></a>00169         m_alleleSet.erase(m_alleleSet.begin() + i);
<a name="l00170"></a>00170       }
<a name="l00171"></a>00171     }
<a name="l00172"></a>00172   }
<a name="l00173"></a>00173 
<a name="l00174"></a>00174   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <span class="stringliteral">&quot;GeneCall::GeneCall gene: &quot;</span> + m_gene + <span class="stringliteral">&quot;: initializing allele names: &quot;</span> + alleleNamesSStr.str());
<a name="l00175"></a>00175   <span class="comment">//describeVerbose(rte, ADT_VERBOSE_NORMAL );</span>
<a name="l00176"></a>00176 
<a name="l00177"></a>00177   <span class="keywordflow">return</span>;
<a name="l00178"></a>00178 
<a name="l00179"></a>00179 }
<a name="l00180"></a>00180 <span class="comment">// end GeneCall::GeneCall</span>
<a name="l00181"></a>00181 <span class="comment">/*****************************************************************************/</span>
<a name="l00182"></a>00182 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00183"></a>00183 <span class="comment">/**</span>
<a name="l00184"></a>00184 <span class="comment"> * GeneCall::addCallElementToAlleleCallSet</span>
<a name="l00185"></a>00185 <span class="comment"> * Synopsis:</span>
<a name="l00186"></a>00186 <span class="comment"> *</span>
<a name="l00187"></a>00187 <span class="comment"> * Initialize a CallElement from the translation table (and not</span>
<a name="l00188"></a>00188 <span class="comment"> * an experiment data in the Genotype data (CHP)).</span>
<a name="l00189"></a>00189 <span class="comment"> * To do this we need to identify the ProbeSet and</span>
<a name="l00190"></a>00190 <span class="comment"> * the reference and variant base. The variant base will be either in the</span>
<a name="l00191"></a>00191 <span class="comment"> * column that resides between A1-A29 or in the &quot;Variant&quot; column itself</span>
<a name="l00192"></a>00192 <span class="comment"> * if no such animal exists.</span>
<a name="l00193"></a>00193 <span class="comment"> *</span>
<a name="l00194"></a>00194 <span class="comment"> *</span>
<a name="l00195"></a>00195 <span class="comment"> * @param ttm - TranslationTableModel with the meta data about the call.</span>
<a name="l00196"></a>00196 <span class="comment"> * @param headerRow - the row in the ttm that has the meta data.</span>
<a name="l00197"></a>00197 <span class="comment"> * @param row - the row in the TranslationTableModel in question.</span>
<a name="l00198"></a>00198 <span class="comment"> * @param column - the column in the TranslationTableModel in question.</span>
<a name="l00199"></a>00199 <span class="comment"> * @param allowMultiAllelic - probeSet may appear multiple times.</span>
<a name="l00200"></a>00200 <span class="comment"> * @param rowIsHaplotype - &quot;Haplotype&quot; column has &quot;Y&quot;</span>
<a name="l00201"></a>00201 <span class="comment"> *</span>
<a name="l00202"></a>00202 <span class="comment"> * @return - true upon successfull add.</span>
<a name="l00203"></a>00203 <span class="comment"> */</span>
<a name="l00204"></a>00204 <span class="comment">/*****************************************************************************/</span>
<a name="l00205"></a><a class="code" href="classGeneCall.html#afdad68791a3d92cb1ccd78bcebff0c60">00205</a> <span class="keywordtype">bool</span> <a class="code" href="classGeneCall.html#afdad68791a3d92cb1ccd78bcebff0c60" title="GeneCall::addCallElementToAlleleCallSet Synopsis:">GeneCall::addCallElementToAlleleCallSet</a>(<a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm,
<a name="l00206"></a>00206     <span class="keyword">const</span> <span class="keywordtype">int</span> headerRow,
<a name="l00207"></a>00207     <span class="keyword">const</span> <span class="keywordtype">int</span> row,
<a name="l00208"></a>00208     <span class="keyword">const</span> <span class="keywordtype">int</span> column,
<a name="l00209"></a>00209     <span class="keyword">const</span> <span class="keywordtype">bool</span> allowMultiAllelic,
<a name="l00210"></a>00210     <span class="keyword">const</span> <span class="keywordtype">bool</span> rowIsHaplotype)
<a name="l00211"></a>00211 
<a name="l00212"></a>00212 {
<a name="l00213"></a>00213 
<a name="l00214"></a>00214   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>((headerRow &gt;= 0) &amp;&amp; (headerRow &lt; ttm.size()) &amp;&amp;
<a name="l00215"></a>00215                  (row &gt;= 0) &amp;&amp; (row &lt; ttm.size()) &amp;&amp;
<a name="l00216"></a>00216                  (headerRow &lt; row), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00217"></a>00217   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>((column &gt;= ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_REFERENCE)) &amp;&amp;
<a name="l00218"></a>00218                  (column &lt; ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ae33ba37f59967a7bd0caefe58de785ed" title="TranslationInputTsvTableModel::columnCount Synopsis:">columnCount</a>()), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00219"></a>00219 
<a name="l00220"></a>00220 
<a name="l00221"></a>00221   std::string alleleName = ttm.<a class="code" href="classTranslationTableModel.html#a9569cee6d0e635259f07dbd8cbd99596" title="TranslationTableModel::getHeaderRowColumnName Synopsis: A header row repeats the descriptive headers ...">getHeaderRowColumnName</a>(headerRow, column);
<a name="l00222"></a>00222 
<a name="l00223"></a>00223   std::vector&lt; class CallSet &gt;::iterator itCS;
<a name="l00224"></a>00224 
<a name="l00225"></a>00225   <span class="keywordflow">for</span> (itCS = m_alleleSet.begin();
<a name="l00226"></a>00226        itCS != m_alleleSet.end() ; ++itCS) {
<a name="l00227"></a>00227 
<a name="l00228"></a>00228     <span class="keywordflow">if</span> (itCS-&gt;m_name == alleleName) {
<a name="l00229"></a>00229       itCS-&gt;m_type = rowIsHaplotype ? ADT_CALL_TYPE_HAPLOTYPE_GROUP :
<a name="l00230"></a>00230                      ADT_CALL_TYPE_MARKER;
<a name="l00231"></a>00231       itCS-&gt;m_isDescriptive = <span class="keyword">true</span>;
<a name="l00232"></a>00232       <span class="keywordflow">return</span> itCS-&gt;addCallElement(ttm, row, column, allowMultiAllelic, rowIsHaplotype);
<a name="l00233"></a>00233     }
<a name="l00234"></a>00234   }
<a name="l00235"></a>00235 
<a name="l00236"></a>00236   std::stringstream msgSStr;
<a name="l00237"></a>00237 
<a name="l00238"></a>00238   msgSStr &lt;&lt; <span class="stringliteral">&quot;addCallElementToAlleleCallSet: can&#39;t find allele name: &quot;</span> &lt;&lt;  alleleName &lt;&lt;  <span class="stringliteral">&quot; from the following possibilities: &quot;</span>;
<a name="l00239"></a>00239   <span class="keywordflow">for</span> (itCS = m_alleleSet.begin();
<a name="l00240"></a>00240        itCS != m_alleleSet.end() ; ++itCS) {
<a name="l00241"></a>00241     msgSStr &lt;&lt; itCS-&gt;m_name &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
<a name="l00242"></a>00242   }
<a name="l00243"></a>00243 
<a name="l00244"></a>00244   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, msgSStr.str());
<a name="l00245"></a>00245 
<a name="l00246"></a>00246   <span class="keywordflow">for</span> (itCS = m_alleleSet.begin();
<a name="l00247"></a>00247        itCS != m_alleleSet.end() ; ++itCS) {
<a name="l00248"></a>00248     <span class="comment">//itCS-&gt;describeVerbose(ADT_VERBOSE_NORMAL );</span>
<a name="l00249"></a>00249   }
<a name="l00250"></a>00250 
<a name="l00251"></a>00251   <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00252"></a>00252 
<a name="l00253"></a>00253 }
<a name="l00254"></a>00254 <span class="comment">// end GeneCall::addCallElementToAlleleCallSet</span>
<a name="l00255"></a>00255 <span class="comment">/*****************************************************************************/</span>
<a name="l00256"></a>00256 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00257"></a>00257 <span class="comment">/**</span>
<a name="l00258"></a>00258 <span class="comment"> * _getNumDescriptive</span>
<a name="l00259"></a>00259 <span class="comment"> * Synopsis:</span>
<a name="l00260"></a>00260 <span class="comment"> *</span>
<a name="l00261"></a>00261 <span class="comment"> * Helper function for describeVerbose only.</span>
<a name="l00262"></a>00262 <span class="comment"> * Return the number of CallSets in the m_alleleSet which are descriptive.</span>
<a name="l00263"></a>00263 <span class="comment"> *</span>
<a name="l00264"></a>00264 <span class="comment"> *</span>
<a name="l00265"></a>00265 <span class="comment"> * @return int - the number of descriptive</span>
<a name="l00266"></a>00266 <span class="comment"> */</span>
<a name="l00267"></a>00267 <span class="comment">/*****************************************************************************/</span>
<a name="l00268"></a>00268 <span class="keyword">static</span> <span class="keywordtype">int</span> _getNumDescriptive(<span class="keyword">const</span> std::vector&lt;CallSet&gt; &amp; alleleSet)
<a name="l00269"></a>00269 {
<a name="l00270"></a>00270 
<a name="l00271"></a>00271   <span class="keywordtype">int</span> i = 0;
<a name="l00272"></a>00272   <span class="keywordtype">int</span> numDescriptive = 0;
<a name="l00273"></a>00273 
<a name="l00274"></a>00274   <span class="keywordflow">for</span> (i = 0, numDescriptive = 0 ; i &lt; alleleSet.size(); i++) {
<a name="l00275"></a>00275     <span class="keywordflow">if</span> (alleleSet[i].m_isDescriptive)
<a name="l00276"></a>00276       numDescriptive++;
<a name="l00277"></a>00277   }
<a name="l00278"></a>00278 
<a name="l00279"></a>00279   <span class="keywordflow">return</span> numDescriptive;
<a name="l00280"></a>00280 
<a name="l00281"></a>00281 }
<a name="l00282"></a>00282 <span class="comment">// end _getNumDescriptive</span>
<a name="l00283"></a>00283 <span class="comment">/*****************************************************************************/</span>
<a name="l00284"></a>00284 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00285"></a>00285 <span class="comment">/**</span>
<a name="l00286"></a>00286 <span class="comment"> * _getNumNonDescriptive</span>
<a name="l00287"></a>00287 <span class="comment"> * Synopsis:</span>
<a name="l00288"></a>00288 <span class="comment"> *</span>
<a name="l00289"></a>00289 <span class="comment"> * Helper function for describeVerbose only.</span>
<a name="l00290"></a>00290 <span class="comment"> * Return the number of CallSets in the m_alleleSet which are not descriptive.</span>
<a name="l00291"></a>00291 <span class="comment"> *</span>
<a name="l00292"></a>00292 <span class="comment"> *</span>
<a name="l00293"></a>00293 <span class="comment"> * @return int - the number of non descriptive</span>
<a name="l00294"></a>00294 <span class="comment"> *</span>
<a name="l00295"></a>00295 <span class="comment"> */</span>
<a name="l00296"></a>00296 <span class="comment">/*****************************************************************************/</span>
<a name="l00297"></a>00297 <span class="keyword">static</span> <span class="keywordtype">int</span> _getNumNonDescriptive(<span class="keyword">const</span> std::vector&lt;CallSet&gt; &amp; alleleSet)
<a name="l00298"></a>00298 {
<a name="l00299"></a>00299   <span class="keywordflow">return</span> (alleleSet.size() - _getNumDescriptive(alleleSet));
<a name="l00300"></a>00300 }
<a name="l00301"></a>00301 <span class="comment">// end _getNumNonDescriptive</span>
<a name="l00302"></a>00302 <span class="comment">/*****************************************************************************/</span>
<a name="l00303"></a>00303 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00304"></a>00304 <span class="comment">/**</span>
<a name="l00305"></a>00305 <span class="comment"> * GeneCall::describeVerbose</span>
<a name="l00306"></a>00306 <span class="comment"> * Synopsis:</span>
<a name="l00307"></a>00307 <span class="comment"> *</span>
<a name="l00308"></a>00308 <span class="comment"> * Run time debug routine that can be invoked with various verbose levels.</span>
<a name="l00309"></a>00309 <span class="comment"> * ADT_VERBOSE_INPUT_FILES is the default level.</span>
<a name="l00310"></a>00310 <span class="comment"> *</span>
<a name="l00311"></a>00311 <span class="comment"> * @param rte       - run time environment</span>
<a name="l00312"></a>00312 <span class="comment"> * @param level     - ADT_VERBOSE_INPUT_FILES by default,</span>
<a name="l00313"></a>00313 <span class="comment"> *                    otherwise the ADT_VERBOSE_ENUM level to output</span>
<a name="l00314"></a>00314 <span class="comment"> *</span>
<a name="l00315"></a>00315 <span class="comment"> */</span>
<a name="l00316"></a>00316 <span class="comment">/*****************************************************************************/</span>
<a name="l00317"></a><a class="code" href="classGeneCall.html#a804c2c06e1c5717f669c7d2dee212214">00317</a> <span class="keywordtype">void</span> <a class="code" href="classGeneCall.html#a804c2c06e1c5717f669c7d2dee212214" title="GeneCall::describeVerbose Synopsis:">GeneCall::describeVerbose</a>(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, ADT_VERBOSE_ENUM level)
<a name="l00318"></a>00318 {
<a name="l00319"></a>00319 
<a name="l00320"></a>00320   <span class="keywordflow">if</span> (!level &amp;&amp; (rte.m_currentVerbosity &lt; ADT_VERBOSE_INPUT_FILES))
<a name="l00321"></a>00321     <span class="keywordflow">return</span>;
<a name="l00322"></a>00322 
<a name="l00323"></a>00323   ADT_VERBOSE_ENUM outLevel = level ? level : ADT_VERBOSE_INPUT_FILES;
<a name="l00324"></a>00324 
<a name="l00325"></a>00325   std::stringstream describe1SStr;
<a name="l00326"></a>00326 
<a name="l00327"></a>00327   describe1SStr &lt;&lt; <span class="stringliteral">&quot;GeneCall::describeVerbose gene: &quot;</span> &lt;&lt; m_gene &lt;&lt; <span class="stringliteral">&quot;:&quot;</span>;
<a name="l00328"></a>00328   describe1SStr &lt;&lt; <span class="stringliteral">&quot; alleleSet total count : &quot;</span>;
<a name="l00329"></a>00329   describe1SStr &lt;&lt; m_alleleSet.size() &lt;&lt; endl;
<a name="l00330"></a>00330   describe1SStr &lt;&lt; <span class="stringliteral">&quot;GeneCall::describeVerbose gene: &quot;</span> &lt;&lt; m_gene &lt;&lt; <span class="stringliteral">&quot;:&quot;</span>;
<a name="l00331"></a>00331   describe1SStr &lt;&lt; <span class="stringliteral">&quot; alleleSet descriptive count : &quot;</span>;
<a name="l00332"></a>00332   describe1SStr &lt;&lt; _getNumDescriptive(m_alleleSet) &lt;&lt; endl;
<a name="l00333"></a>00333   describe1SStr &lt;&lt; <span class="stringliteral">&quot;GeneCall::describeVerbose gene: &quot;</span> &lt;&lt; m_gene &lt;&lt; <span class="stringliteral">&quot;:&quot;</span>;
<a name="l00334"></a>00334   describe1SStr &lt;&lt; <span class="stringliteral">&quot; alleleSet non-descriptive count : &quot;</span>;
<a name="l00335"></a>00335   describe1SStr &lt;&lt; _getNumNonDescriptive(m_alleleSet);
<a name="l00336"></a>00336   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(outLevel, describe1SStr.str());
<a name="l00337"></a>00337 
<a name="l00338"></a>00338   std::stringstream describe4SStr;
<a name="l00339"></a>00339   describe4SStr &lt;&lt; <span class="stringliteral">&quot;GeneCall::describeVerbose gene: &quot;</span> &lt;&lt; m_gene &lt;&lt; <span class="stringliteral">&quot;: &quot;</span>;
<a name="l00340"></a>00340   describe4SStr &lt;&lt; <span class="stringliteral">&quot;vector&lt;CallSet&gt; alleleSet follows: &quot;</span> ;
<a name="l00341"></a>00341 
<a name="l00342"></a>00342   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(outLevel, describe4SStr.str());
<a name="l00343"></a>00343 
<a name="l00344"></a>00344   <span class="keywordtype">int</span> i;
<a name="l00345"></a>00345   <span class="keywordflow">for</span> (i = 0; i &lt; m_alleleSet.size(); i++) {
<a name="l00346"></a>00346     m_alleleSet[i].describeVerbose(rte, level);
<a name="l00347"></a>00347   }
<a name="l00348"></a>00348 
<a name="l00349"></a>00349   <span class="keywordflow">return</span>;
<a name="l00350"></a>00350 
<a name="l00351"></a>00351 }
<a name="l00352"></a>00352 <span class="comment">// end GeneCall::describeVerbose</span>
<a name="l00353"></a>00353 <span class="comment">/*****************************************************************************/</span>
<a name="l00354"></a>00354 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00355"></a>00355 <span class="comment">/**</span>
<a name="l00356"></a>00356 <span class="comment"> * GeneCall::existsAlleleCallSet</span>
<a name="l00357"></a>00357 <span class="comment"> * Synopsis:</span>
<a name="l00358"></a>00358 <span class="comment"> *</span>
<a name="l00359"></a>00359 <span class="comment"> * Returns true if a CallSet in m_alleleSet has the passed in allele name.</span>
<a name="l00360"></a>00360 <span class="comment"> *</span>
<a name="l00361"></a>00361 <span class="comment"> * @params alleleName - the call name to check (i.e. &quot;Ref&quot;, &quot;*4&quot;)</span>
<a name="l00362"></a>00362 <span class="comment"> *</span>
<a name="l00363"></a>00363 <span class="comment"> */</span>
<a name="l00364"></a>00364 <span class="comment">/*****************************************************************************/</span>
<a name="l00365"></a><a class="code" href="classGeneCall.html#ac9276bb5f94a5916eefc26aacdbbc027">00365</a> <span class="keywordtype">bool</span> <a class="code" href="classGeneCall.html#ac9276bb5f94a5916eefc26aacdbbc027" title="GeneCall::existsAlleleCallSet Synopsis:">GeneCall::existsAlleleCallSet</a>(<span class="keyword">const</span> std::string &amp; alleleName)
<a name="l00366"></a>00366 {
<a name="l00367"></a>00367 
<a name="l00368"></a>00368   std::vector&lt;CallSet&gt;::iterator itCS;
<a name="l00369"></a>00369 
<a name="l00370"></a>00370   <span class="keywordflow">for</span> (itCS = m_alleleSet.begin(); itCS != m_alleleSet.end(); itCS++) {
<a name="l00371"></a>00371     <span class="keywordflow">if</span> (itCS-&gt;m_name == alleleName) {
<a name="l00372"></a>00372       <span class="keywordflow">return</span> <span class="keyword">true</span>;
<a name="l00373"></a>00373     }
<a name="l00374"></a>00374   }
<a name="l00375"></a>00375 
<a name="l00376"></a>00376   <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00377"></a>00377 }
<a name="l00378"></a>00378 <span class="comment">// end GeneCall::existsAlleleCallSet</span>
<a name="l00379"></a>00379 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00380"></a>00380 <span class="comment">/**</span>
<a name="l00381"></a>00381 <span class="comment"> * GeneCall::getAlleleCallSet</span>
<a name="l00382"></a>00382 <span class="comment"> * Synopsis:</span>
<a name="l00383"></a>00383 <span class="comment"> *</span>
<a name="l00384"></a>00384 <span class="comment"> * Convienance function to return either the Reference or Variant allele set.</span>
<a name="l00385"></a>00385 <span class="comment"> *</span>
<a name="l00386"></a>00386 <span class="comment"> * @param index - between 0 and  num m_alleleSet.size()</span>
<a name="l00387"></a>00387 <span class="comment"> *</span>
<a name="l00388"></a>00388 <span class="comment"> * @return CallSet reference corresponding to the index.</span>
<a name="l00389"></a>00389 <span class="comment"> */</span>
<a name="l00390"></a>00390 <span class="comment">/*****************************************************************************/</span>
<a name="l00391"></a><a class="code" href="classGeneCall.html#a30757e06c17f4315d3686a5341f4b547">00391</a> <a class="code" href="classCallSet.html">CallSet</a> &amp;  <a class="code" href="classGeneCall.html#a30757e06c17f4315d3686a5341f4b547" title="GeneCall::getAlleleCallSet Synopsis:">GeneCall::getAlleleCallSet</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index)
<a name="l00392"></a>00392 {
<a name="l00393"></a>00393 
<a name="l00394"></a>00394   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(m_alleleSet.size() &gt; 0, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00395"></a>00395 
<a name="l00396"></a>00396   <span class="keywordflow">return</span> m_alleleSet[index];
<a name="l00397"></a>00397 
<a name="l00398"></a>00398 }
<a name="l00399"></a>00399 <span class="comment">// end GeneCall::getAlleleCallSet</span>
<a name="l00400"></a>00400 <span class="comment">/*****************************************************************************/</span>
<a name="l00401"></a>00401 <span class="comment"></span>
<a name="l00402"></a>00402 <span class="comment">//////////////////////////////////////////////////////////////////////////////</span>
<a name="l00403"></a>00403 <span class="comment"></span><span class="comment">//                      TRANSLATION ALGORITHM                               //</span><span class="comment"></span>
<a name="l00404"></a>00404 <span class="comment">//////////////////////////////////////////////////////////////////////////////</span>
<a name="l00405"></a>00405 <span class="comment"></span>
<a name="l00406"></a>00406 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00407"></a>00407 <span class="comment">/**</span>
<a name="l00408"></a>00408 <span class="comment"> * _getProbeSetBaseCount</span>
<a name="l00409"></a>00409 <span class="comment"> * Synopsis:</span>
<a name="l00410"></a>00410 <span class="comment"> *</span>
<a name="l00411"></a>00411 <span class="comment"> * Helper function to be called only once by _calculateTHPR, broken out</span>
<a name="l00412"></a>00412 <span class="comment"> * for clarity.</span>
<a name="l00413"></a>00413 <span class="comment"> *</span>
<a name="l00414"></a>00414 <span class="comment"> * Any ProbeSet Id in the translation table will have two possible bases</span>
<a name="l00415"></a>00415 <span class="comment"> * for the first row, and one additional for every row thereafter.</span>
<a name="l00416"></a>00416 <span class="comment"> * This is represented in the completeSet passed in as one CallElement</span>
<a name="l00417"></a>00417 <span class="comment"> * per ProbeSet Id and a list of bases associated therein.</span>
<a name="l00418"></a>00418 <span class="comment"> *</span>
<a name="l00419"></a>00419 <span class="comment"> *</span>
<a name="l00420"></a>00420 <span class="comment"> *</span>
<a name="l00421"></a>00421 <span class="comment"> * @params chromatid1 - CallSet generated from ALLELE1 column from Genotype data.</span>
<a name="l00422"></a>00422 <span class="comment"> * @params completeSet - CallSet generated from the translation table</span>
<a name="l00423"></a>00423 <span class="comment"> *                       with possible multi-allelic bases.</span>
<a name="l00424"></a>00424 <span class="comment"> * @params ch1Index - CallElement index to compare. Note that CallElements</span>
<a name="l00425"></a>00425 <span class="comment"> *                   are keyed by ProbeSet. The i&#39;th element refers to the</span>
<a name="l00426"></a>00426 <span class="comment"> *                   i&#39;th element in chromatid1 and the same probe set in</span>
<a name="l00427"></a>00427 <span class="comment"> *                   completeSet that corresponds to the i&#39;th element in</span>
<a name="l00428"></a>00428 <span class="comment"> *                   chromatid1.</span>
<a name="l00429"></a>00429 <span class="comment"> *</span>
<a name="l00430"></a>00430 <span class="comment"> * @returns count - the number of bases in completeSet.</span>
<a name="l00431"></a>00431 <span class="comment"> */</span>
<a name="l00432"></a>00432 <span class="comment">/*****************************************************************************/</span>
<a name="l00433"></a>00433 <span class="keyword">static</span> <span class="keywordtype">int</span> _getProbeSetBaseCount(<span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid1,
<a name="l00434"></a>00434                                  <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; completeSet,
<a name="l00435"></a>00435                                  <span class="keyword">const</span> <span class="keywordtype">int</span> ch1Index)
<a name="l00436"></a>00436 {
<a name="l00437"></a>00437 
<a name="l00438"></a>00438   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(ch1Index &lt; chromatid1.<a class="code" href="classCallSet.html#adc8616dcd78472498079dbd5784d102f" title="inline">size</a>(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00439"></a>00439 
<a name="l00440"></a>00440   std::map&lt; std::string, CallElement&gt;::const_iterator iCEit
<a name="l00441"></a>00441   = chromatid1.m_ceSet.begin();
<a name="l00442"></a>00442 
<a name="l00443"></a>00443   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; ch1Index; i++, iCEit++) {};
<a name="l00444"></a>00444 
<a name="l00445"></a>00445   <span class="keyword">const</span> <a class="code" href="classCallElement.html">CallElement</a> &amp; ceCompleteSet
<a name="l00446"></a>00446   = completeSet.<a class="code" href="classCallSet.html#a6af4574eed8d3b390a5882ea9f279d8d" title="CallSet::getProbeSetCallElement Synopsis:">getProbeSetCallElement</a>(iCEit-&gt;second.m_probeSet);
<a name="l00447"></a>00447 
<a name="l00448"></a>00448   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(ceCompleteSet.getCompleteSetSize() &gt; 1, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00449"></a>00449 
<a name="l00450"></a>00450   <span class="keywordflow">return</span> (ceCompleteSet.getCompleteSetSize());
<a name="l00451"></a>00451 
<a name="l00452"></a>00452 }
<a name="l00453"></a>00453 <span class="comment">// end GeneCall::_getProbeSetBaseCount</span>
<a name="l00454"></a>00454 <span class="comment">/*****************************************************************************/</span>
<a name="l00455"></a>00455 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00456"></a>00456 <span class="comment">/**</span>
<a name="l00457"></a>00457 <span class="comment"> * _isHeteroZygousPair</span>
<a name="l00458"></a>00458 <span class="comment"> * Synopsis:</span>
<a name="l00459"></a>00459 <span class="comment"> *</span>
<a name="l00460"></a>00460 <span class="comment"> * Helper function to be called only once by _calculateTHPR, broken out</span>
<a name="l00461"></a>00461 <span class="comment"> * for clarity.</span>
<a name="l00462"></a>00462 <span class="comment"> *</span>
<a name="l00463"></a>00463 <span class="comment"> * If the CallElements between the two CallSets for the i&#39;th element have</span>
<a name="l00464"></a>00464 <span class="comment"> * the same base then they are consider homozygous, else heterozygous.</span>
<a name="l00465"></a>00465 <span class="comment"> *</span>
<a name="l00466"></a>00466 <span class="comment"> *</span>
<a name="l00467"></a>00467 <span class="comment"> * @params chromatid1 - CallSet generated from ALLELE1 column from Genotype data.</span>
<a name="l00468"></a>00468 <span class="comment"> * @params chromatid2 - CallSet generated from ALLELE2 column from Genotype data.</span>
<a name="l00469"></a>00469 <span class="comment"> * @params ch1Index - CallElement index to compare. Note that CallElements</span>
<a name="l00470"></a>00470 <span class="comment"> *                   are keyed by ProbeSet. The i&#39;th element refers to the</span>
<a name="l00471"></a>00471 <span class="comment"> *                   i&#39;th element in chromatid1 and the same probe set in</span>
<a name="l00472"></a>00472 <span class="comment"> *                   chromatid2 that corresponds to the i&#39;th element in</span>
<a name="l00473"></a>00473 <span class="comment"> *                   chromatid1.</span>
<a name="l00474"></a>00474 <span class="comment"> *</span>
<a name="l00475"></a>00475 <span class="comment"> * @returns true - if the bases for the CallElements compared are different.</span>
<a name="l00476"></a>00476 <span class="comment"> */</span>
<a name="l00477"></a>00477 <span class="comment">/*****************************************************************************/</span>
<a name="l00478"></a>00478 <span class="keyword">static</span> <span class="keywordtype">bool</span> _isHeteroZygousPair(<span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid1,
<a name="l00479"></a>00479                                 <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid2,
<a name="l00480"></a>00480                                 <span class="keyword">const</span> <span class="keywordtype">int</span> ch1Index)
<a name="l00481"></a>00481 
<a name="l00482"></a>00482 {
<a name="l00483"></a>00483   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(ch1Index &lt; chromatid1.<a class="code" href="classCallSet.html#adc8616dcd78472498079dbd5784d102f" title="inline">size</a>(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00484"></a>00484   std::map&lt; std::string, CallElement&gt;::const_iterator iCEit =
<a name="l00485"></a>00485     chromatid1.m_ceSet.begin();
<a name="l00486"></a>00486 
<a name="l00487"></a>00487   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; ch1Index; i++, iCEit++) {}
<a name="l00488"></a>00488 
<a name="l00489"></a>00489   <span class="keyword">const</span> <a class="code" href="classCallElement.html">CallElement</a> &amp; ceChromatid1 = chromatid1.<a class="code" href="classCallSet.html#a6af4574eed8d3b390a5882ea9f279d8d" title="CallSet::getProbeSetCallElement Synopsis:">getProbeSetCallElement</a>(iCEit-&gt;second.m_probeSet);
<a name="l00490"></a>00490   <span class="keyword">const</span> <a class="code" href="classCallElement.html">CallElement</a> &amp; ceChromatid2 = chromatid2.<a class="code" href="classCallSet.html#a6af4574eed8d3b390a5882ea9f279d8d" title="CallSet::getProbeSetCallElement Synopsis:">getProbeSetCallElement</a>(iCEit-&gt;second.m_probeSet);
<a name="l00491"></a>00491 
<a name="l00492"></a>00492   <span class="comment">// Ok, even though there can be multiple bases in a CallElement, that</span>
<a name="l00493"></a>00493   <span class="comment">// only applies to CallElements constructed from the Translation Table.</span>
<a name="l00494"></a>00494   <span class="comment">// For CallElements constructed from the Genotype data there is always</span>
<a name="l00495"></a>00495   <span class="comment">// ever exactly one base.</span>
<a name="l00496"></a>00496 
<a name="l00497"></a>00497   std::string base1 = ceChromatid1.m_bases[0];
<a name="l00498"></a>00498   std::string base2 = ceChromatid2.m_bases[0];
<a name="l00499"></a>00499 
<a name="l00500"></a>00500   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(!base1.empty() &amp;&amp; !base2.empty(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00501"></a>00501 
<a name="l00502"></a>00502 
<a name="l00503"></a>00503   <span class="keywordflow">return</span> (base1 != base2);
<a name="l00504"></a>00504 }
<a name="l00505"></a>00505 <span class="comment">// end GeneCall::_isHeteroZygousPair</span>
<a name="l00506"></a>00506 <span class="comment">/*****************************************************************************/</span>
<a name="l00507"></a>00507 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00508"></a>00508 <span class="comment">/**</span>
<a name="l00509"></a>00509 <span class="comment"> * _calculateTHPR</span>
<a name="l00510"></a>00510 <span class="comment"> * Synopsis:</span>
<a name="l00511"></a>00511 <span class="comment"> *</span>
<a name="l00512"></a>00512 <span class="comment"> * Helper function to be called only once by GeneCall::translateExperimentCall,</span>
<a name="l00513"></a>00513 <span class="comment"> * broken out for clarity.</span>
<a name="l00514"></a>00514 <span class="comment"> *</span>
<a name="l00515"></a>00515 <span class="comment"> * Calculate the requred number of haplotype matches given tri-allelic and</span>
<a name="l00516"></a>00516 <span class="comment"> * wild-cards are in the mix. Note that MARKERs are considered HAPLOTYPE GROUP</span>
<a name="l00517"></a>00517 <span class="comment"> * of size 1 for the purposes of this API.</span>
<a name="l00518"></a>00518 <span class="comment"> *</span>
<a name="l00519"></a>00519 <span class="comment"> *</span>
<a name="l00520"></a>00520 <span class="comment"> * Key:</span>
<a name="l00521"></a>00521 <span class="comment"> *</span>
<a name="l00522"></a>00522 <span class="comment"> * 0 : false, no heterozygous base pairs have been detected.</span>
<a name="l00523"></a>00523 <span class="comment"> * 1 : true, at least one heterozygous base pair has been detected.</span>
<a name="l00524"></a>00524 <span class="comment"> *</span>
<a name="l00525"></a>00525 <span class="comment"> * THP0 : Total Haplotype Marker Pairs, no heterozygous base pairs have been detected.</span>
<a name="l00526"></a>00526 <span class="comment"> * THP1 : Total Haplotype Marker Pairs, at least one heterozygous base pair has been detected.</span>
<a name="l00527"></a>00527 <span class="comment"> * i : iteration count of the number of haplotype marker pairs analyzed so far.</span>
<a name="l00528"></a>00528 <span class="comment"> * Ni : for a Wildcard, iteration translation table possible allele bases</span>
<a name="l00529"></a>00529 <span class="comment"> *      corresponding to the</span>
<a name="l00530"></a>00530 <span class="comment"> *      probe set Haplotype marker pairs with the wildcard.</span>
<a name="l00531"></a>00531 <span class="comment"> *      For bi-allelic this is two bases,</span>
<a name="l00532"></a>00532 <span class="comment"> *      the reference and the variant. For tri-allelic this number is 3, for</span>
<a name="l00533"></a>00533 <span class="comment"> *      the reference and two variants.</span>
<a name="l00534"></a>00534 <span class="comment"> *</span>
<a name="l00535"></a>00535 <span class="comment"> *  * : wildcard pair of Haplotype Markers.</span>
<a name="l00536"></a>00536 <span class="comment"> *</span>
<a name="l00537"></a>00537 <span class="comment"> * Formuala:</span>
<a name="l00538"></a>00538 <span class="comment"> *</span>
<a name="l00539"></a>00539 <span class="comment"> *</span>
<a name="l00540"></a>00540 <span class="comment"> * THP0(0) = 1</span>
<a name="l00541"></a>00541 <span class="comment"> * THP1(0) = 0</span>
<a name="l00542"></a>00542 <span class="comment"> *</span>
<a name="l00543"></a>00543 <span class="comment"> *             { i+1: homozygous   =&gt; THP0(i)</span>
<a name="l00544"></a>00544 <span class="comment"> * THP0(i+1) = | i+1: heterozygous =&gt; 0</span>
<a name="l00545"></a>00545 <span class="comment"> *             { i+1: *            =&gt; THP0(i) * Ni</span>
<a name="l00546"></a>00546 <span class="comment"> *</span>
<a name="l00547"></a>00547 <span class="comment"> *             { i+1: homozygous   =&gt; THP1(i)</span>
<a name="l00548"></a>00548 <span class="comment"> * THP1(i+1) = | i+1: heterozygous =&gt; 2*THP1(i) + THP0(i)</span>
<a name="l00549"></a>00549 <span class="comment"> *             { i+1: *            =&gt; Ni * THP1(i) + ( Ni(Ni-1)/2 ) *</span>
<a name="l00550"></a>00550 <span class="comment"> *                                    2*THP1(i) + ( Ni(Ni - 1) )/2) * THP0(i)</span>
<a name="l00551"></a>00551 <span class="comment"> *</span>
<a name="l00552"></a>00552 <span class="comment"> * THPR = THP1(total) + THP0(total)</span>
<a name="l00553"></a>00553 <span class="comment"> *</span>
<a name="l00554"></a>00554 <span class="comment"> * @param chromatid1 - ALLELE1 column in the genotype file.</span>
<a name="l00555"></a>00555 <span class="comment"> * @param chromatid2 - ALLELE2 column in the genotype file.</span>
<a name="l00556"></a>00556 <span class="comment"> * @param completeSet - from the translation table, the complete set</span>
<a name="l00557"></a>00557 <span class="comment"> *                       of possible probe sets, possible multi-allelic.</span>
<a name="l00558"></a>00558 <span class="comment"> * @param HPPA      - passed in for return.</span>
<a name="l00559"></a>00559 <span class="comment"> *</span>
<a name="l00560"></a>00560 <span class="comment"> * @return HPPA - the number of haplotype pairs possible per allele.</span>
<a name="l00561"></a>00561 <span class="comment"> * @return THPR - the total number of required haplotype paris.</span>
<a name="l00562"></a>00562 <span class="comment"> *                 If this total is not found then UNK should be used</span>
<a name="l00563"></a>00563 <span class="comment"> *                 to generate the required total.</span>
<a name="l00564"></a>00564 <span class="comment"> */</span>
<a name="l00565"></a>00565 <span class="comment">/*****************************************************************************/</span>
<a name="l00566"></a>00566 <span class="keyword">static</span> <span class="keywordtype">int</span> _calculateTHPR(<span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid1,
<a name="l00567"></a>00567                           <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid2,
<a name="l00568"></a>00568                           <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; completeSet,
<a name="l00569"></a>00569                           <span class="keywordtype">int</span> &amp; HPPA)
<a name="l00570"></a>00570 {
<a name="l00571"></a>00571 
<a name="l00572"></a>00572 
<a name="l00573"></a>00573 
<a name="l00574"></a>00574   <span class="comment">// NON-HAPLOTYPE GROUP MARKERS and HAPLOTYPE GROUPS of size 1.</span>
<a name="l00575"></a>00575   <span class="keywordflow">if</span> (chromatid1.m_copyNumber == 0) {
<a name="l00576"></a>00576     <span class="keywordflow">return</span> 1;
<a name="l00577"></a>00577   }
<a name="l00578"></a>00578 
<a name="l00579"></a>00579   <span class="comment">// HAPLOTYPE GROUP MARKERS</span>
<a name="l00580"></a>00580   <span class="keywordtype">int</span> THPR = 0;
<a name="l00581"></a>00581 
<a name="l00582"></a>00582   <span class="keywordtype">int</span> THP0 = 1;
<a name="l00583"></a>00583   <span class="keywordtype">int</span> THP1 = 0;
<a name="l00584"></a>00584 
<a name="l00585"></a>00585   HPPA = 1;
<a name="l00586"></a>00586   std::map&lt; std::string, CallElement&gt;::const_iterator iCEit = chromatid1.m_ceSet.begin();
<a name="l00587"></a>00587 
<a name="l00588"></a>00588   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; chromatid1.<a class="code" href="classCallSet.html#adc8616dcd78472498079dbd5784d102f" title="inline">size</a>(); i++, iCEit++) {
<a name="l00589"></a>00589 
<a name="l00590"></a>00590     <span class="keywordtype">bool</span> isHet = _isHeteroZygousPair(chromatid1, chromatid2, i);
<a name="l00591"></a>00591     <span class="keywordtype">int</span>  Ni    = _getProbeSetBaseCount(chromatid1, completeSet, i);
<a name="l00592"></a>00592 
<a name="l00593"></a>00593 
<a name="l00594"></a>00594     <span class="keywordflow">if</span> (iCEit-&gt;second.hasWildCards()) {
<a name="l00595"></a>00595       <span class="comment">// HPPA</span>
<a name="l00596"></a>00596       HPPA *= Ni;
<a name="l00597"></a>00597       <span class="comment">// THP1</span>
<a name="l00598"></a>00598       THP1 = Ni * THP1 + (Ni) * (Ni - 1) * THP1 + ((Ni) * (Ni - 1) / 2) * THP0;
<a name="l00599"></a>00599       <span class="comment">// THP0</span>
<a name="l00600"></a>00600       THP0 = THP0 * Ni;
<a name="l00601"></a>00601     } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (isHet) {
<a name="l00602"></a>00602       <span class="comment">// THP1</span>
<a name="l00603"></a>00603       THP1 = 2 * THP1 + THP0;
<a name="l00604"></a>00604       <span class="comment">// THP0</span>
<a name="l00605"></a>00605       THP0 = 0;
<a name="l00606"></a>00606     } <span class="keywordflow">else</span> { <span class="comment">// is homozygous</span>
<a name="l00607"></a>00607       <span class="comment">// THP1 = THP1;</span>
<a name="l00608"></a>00608       <span class="comment">// THP0 = THP0;</span>
<a name="l00609"></a>00609     }
<a name="l00610"></a>00610 
<a name="l00611"></a>00611   }
<a name="l00612"></a>00612 
<a name="l00613"></a>00613   THPR = THP0 + THP1;
<a name="l00614"></a>00614 
<a name="l00615"></a>00615   <span class="keywordflow">return</span> THPR;
<a name="l00616"></a>00616 
<a name="l00617"></a>00617 }
<a name="l00618"></a>00618 <span class="comment">// end GeneCall::_calculateTHPR</span>
<a name="l00619"></a>00619 <span class="comment">/*****************************************************************************/</span>
<a name="l00620"></a>00620 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00621"></a>00621 <span class="comment">/**</span>
<a name="l00622"></a>00622 <span class="comment"> * _translateExperimentCallCheck</span>
<a name="l00623"></a>00623 <span class="comment"> * Synopsis:</span>
<a name="l00624"></a>00624 <span class="comment"> *</span>
<a name="l00625"></a>00625 <span class="comment"> * Helper function to be called only once by GeneCall::translateExperimentCall,</span>
<a name="l00626"></a>00626 <span class="comment"> * broken out for clarity.</span>
<a name="l00627"></a>00627 <span class="comment"> *</span>
<a name="l00628"></a>00628 <span class="comment"> * Does all the upfront validation and exception throwing.</span>
<a name="l00629"></a>00629 <span class="comment"> * Note the multiple catches. This means all exceptions will be logged</span>
<a name="l00630"></a>00630 <span class="comment"> * if multiple exist and not just the first exception.</span>
<a name="l00631"></a>00631 <span class="comment"> *</span>
<a name="l00632"></a>00632 <span class="comment"> * @param rte - RunTimeEnvironment which contains options</span>
<a name="l00633"></a>00633 <span class="comment"> * @param completeSet - the known universe of accepted probeSets/bases.</span>
<a name="l00634"></a>00634 <span class="comment"> * @param chromatid1 - ALLELE1 from the Genotype Short Report.</span>
<a name="l00635"></a>00635 <span class="comment"> * @param chromatid2 - ALLELE2 from the Genotype Short Report.</span>
<a name="l00636"></a>00636 <span class="comment"> * @param alleleSet  - the CallSets of calls.</span>
<a name="l00637"></a>00637 <span class="comment"> *</span>
<a name="l00638"></a>00638 <span class="comment"> *</span>
<a name="l00639"></a>00639 <span class="comment"> * @returns true - if no exceptions were logged and its ok to continue</span>
<a name="l00640"></a>00640 <span class="comment"> *                 call translation.</span>
<a name="l00641"></a>00641 <span class="comment"> */</span>
<a name="l00642"></a>00642 <span class="comment">/*****************************************************************************/</span>
<a name="l00643"></a>00643 <span class="keyword">static</span> <span class="keywordtype">bool</span> _translateExperimentCallCheck(<a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00644"></a>00644     <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; completeSet,
<a name="l00645"></a>00645     <span class="keyword">const</span> std::string &amp; geneName,
<a name="l00646"></a>00646     <span class="keyword">const</span> std::string &amp; experimentName,
<a name="l00647"></a>00647     <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid1,
<a name="l00648"></a>00648     <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid2,
<a name="l00649"></a>00649     <span class="keyword">const</span> std::vector&lt;CallSet&gt; &amp; alleleSet)
<a name="l00650"></a>00650 {
<a name="l00651"></a>00651 
<a name="l00652"></a>00652   <span class="keywordtype">bool</span> okToCall = <span class="keyword">true</span>;
<a name="l00653"></a>00653 
<a name="l00654"></a>00654   std::vector&lt;class CallSet&gt;::const_iterator testCallSetIt = alleleSet.begin();
<a name="l00655"></a>00655 
<a name="l00656"></a>00656   <span class="comment">// All experiments should have a minimum of 1 CallSet, the reference set.</span>
<a name="l00657"></a>00657   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(testCallSetIt != alleleSet.end(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00658"></a>00658 
<a name="l00659"></a>00659   <span class="keywordflow">if</span> (!completeSet.<a class="code" href="classCallSet.html#adca18c7e8d6f731f063dffac6fa95831" title="CallSet::contains( CallSet ) Synopsis.">contains</a>(chromatid1)) {
<a name="l00660"></a>00660     std::stringstream msgSStr;
<a name="l00661"></a>00661     completeSet.<a class="code" href="classCallSet.html#a154f8707d153c4644d55f6adb68c8f8e" title="CallSet::describeVerbose.">describeVerbose</a>(rte, ADT_VERBOSE_NORMAL);
<a name="l00662"></a>00662     chromatid1.<a class="code" href="classCallSet.html#a154f8707d153c4644d55f6adb68c8f8e" title="CallSet::describeVerbose.">describeVerbose</a>(rte, ADT_VERBOSE_NORMAL);
<a name="l00663"></a>00663     msgSStr &lt;&lt; experimentName &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; geneName &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; chromatid1.m_name &lt;&lt;  <span class="stringliteral">&quot;: invalid probe set / base detected.&quot;</span>;
<a name="l00664"></a>00664     <span class="keywordflow">if</span> (rte.m_adtOpts.m_ignoreUnknownAlleles) {
<a name="l00665"></a>00665       okToCall = <span class="keyword">false</span>;
<a name="l00666"></a>00666     } <span class="keywordflow">else</span> {
<a name="l00667"></a>00667       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(msgSStr.str());
<a name="l00668"></a>00668     }
<a name="l00669"></a>00669 
<a name="l00670"></a>00670   }
<a name="l00671"></a>00671 
<a name="l00672"></a>00672   <span class="keywordflow">if</span> (!completeSet.<a class="code" href="classCallSet.html#adca18c7e8d6f731f063dffac6fa95831" title="CallSet::contains( CallSet ) Synopsis.">contains</a>(chromatid2)) {
<a name="l00673"></a>00673     std::stringstream msgSStr;
<a name="l00674"></a>00674     msgSStr &lt;&lt; experimentName &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; geneName &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; chromatid2.m_name &lt;&lt;  <span class="stringliteral">&quot;: invalid probe set / base detected.&quot;</span>;
<a name="l00675"></a>00675     <span class="keywordflow">if</span> (rte.m_adtOpts.m_ignoreUnknownAlleles) {
<a name="l00676"></a>00676       okToCall = <span class="keyword">false</span>;
<a name="l00677"></a>00677     } <span class="keywordflow">else</span> {
<a name="l00678"></a>00678       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(msgSStr.str());
<a name="l00679"></a>00679     }
<a name="l00680"></a>00680   }
<a name="l00681"></a>00681 
<a name="l00682"></a>00682   std::stringstream completeSStr;
<a name="l00683"></a>00683   completeSStr &lt;&lt;  experimentName &lt;&lt; <span class="stringliteral">&quot;, &quot;</span> &lt;&lt;  geneName &lt;&lt;  <span class="stringliteral">&quot;, &quot;</span> &lt;&lt; testCallSetIt-&gt;m_name &lt;&lt; <span class="stringliteral">&quot;: incomplete probe set for haplotype group detected.&quot;</span>;
<a name="l00684"></a>00684 
<a name="l00685"></a>00685   <span class="keywordflow">if</span> (chromatid1.<a class="code" href="classCallSet.html#adc8616dcd78472498079dbd5784d102f" title="inline">size</a>() != testCallSetIt-&gt;size()) {
<a name="l00686"></a>00686 
<a name="l00687"></a>00687     <span class="keywordflow">if</span> (rte.m_adtOpts.m_enforceCompleteHaplotypeGroup) {
<a name="l00688"></a>00688       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(completeSStr.str());
<a name="l00689"></a>00689     }
<a name="l00690"></a>00690     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_EXCEPTION, completeSStr.str());
<a name="l00691"></a>00691     okToCall = <span class="keyword">false</span>;
<a name="l00692"></a>00692   }
<a name="l00693"></a>00693 
<a name="l00694"></a>00694   <span class="keywordflow">if</span> (chromatid2.m_ceSet.<a class="code" href="classCallSet.html#adc8616dcd78472498079dbd5784d102f" title="inline">size</a>() != testCallSetIt-&gt;m_ceSet.size()) {
<a name="l00695"></a>00695 
<a name="l00696"></a>00696     <span class="keywordflow">if</span> (rte.m_adtOpts.m_enforceCompleteHaplotypeGroup) {
<a name="l00697"></a>00697       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(completeSStr.str());
<a name="l00698"></a>00698     }
<a name="l00699"></a>00699     okToCall = <span class="keyword">false</span>;
<a name="l00700"></a>00700     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_EXCEPTION, completeSStr.str());
<a name="l00701"></a>00701   }
<a name="l00702"></a>00702 
<a name="l00703"></a>00703   <span class="comment">// Ok, this is a HAPLOTYPE GROUP or MARKER and all items are NC</span>
<a name="l00704"></a>00704   <span class="comment">// then just return an empty call.</span>
<a name="l00705"></a>00705   <span class="keywordflow">if</span> (chromatid1.<a class="code" href="classCallSet.html#a15ac498f0337d1b516d11b2ca6a6c338" title="inline">hasWildCards</a>() &amp;&amp;
<a name="l00706"></a>00706       (chromatid1.<a class="code" href="classCallSet.html#a963436256d733e47dd7ea15263323354" title="CallSet::numWildCars Synopsis:">numWildCards</a>() == chromatid1.<a class="code" href="classCallSet.html#adc8616dcd78472498079dbd5784d102f" title="inline">size</a>())) {
<a name="l00707"></a>00707 
<a name="l00708"></a>00708     okToCall = <span class="keyword">false</span>;
<a name="l00709"></a>00709   }
<a name="l00710"></a>00710 
<a name="l00711"></a>00711   <span class="keywordflow">return</span> okToCall;
<a name="l00712"></a>00712 }
<a name="l00713"></a>00713 <span class="comment">// end _translateExperimentCallCheck</span>
<a name="l00714"></a>00714 <span class="comment">/*****************************************************************************/</span>
<a name="l00715"></a>00715 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00716"></a>00716 <span class="comment">/**</span>
<a name="l00717"></a>00717 <span class="comment"> * _translateExperimentCallCopyNumberOne</span>
<a name="l00718"></a>00718 <span class="comment"> * Synopsis:</span>
<a name="l00719"></a>00719 <span class="comment"> *</span>
<a name="l00720"></a>00720 <span class="comment"> *  Helper function for translateExperimentCall broken out for brevity.</span>
<a name="l00721"></a>00721 <span class="comment"> *  Specifically translates to a copy number 1 call result.</span>
<a name="l00722"></a>00722 <span class="comment"> *</span>
<a name="l00723"></a>00723 <span class="comment"> * @param rte         - single instance of the run time environment</span>
<a name="l00724"></a>00724 <span class="comment"> * @param results     - for return</span>
<a name="l00725"></a>00725 <span class="comment"> * @param chromatid1  - from the experiment data for the single chromatid</span>
<a name="l00726"></a>00726 <span class="comment"> * @param geneCopyNumberZeroCallSet - for the specific gene</span>
<a name="l00727"></a>00727 <span class="comment"> *</span>
<a name="l00728"></a>00728 <span class="comment"> * @return results - updated with copy number &quot;data&quot; for pairing with</span>
<a name="l00729"></a>00729 <span class="comment"> * the chromatid1 passed in.</span>
<a name="l00730"></a>00730 <span class="comment"> */</span>
<a name="l00731"></a>00731 <span class="comment">/*****************************************************************************/</span>
<a name="l00732"></a>00732 <span class="keyword">static</span> <span class="keywordtype">void</span> _translateExperimentCallCopyNumberOne(
<a name="l00733"></a>00733   <span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00734"></a>00734   <a class="code" href="classCallResults.html">CallResults</a> &amp; results,
<a name="l00735"></a>00735   <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid1,
<a name="l00736"></a>00736   <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; geneCopyNumberZeroCallSet)
<a name="l00737"></a>00737 {
<a name="l00738"></a>00738 
<a name="l00739"></a>00739   <span class="comment">// COPY NUMBER 1</span>
<a name="l00740"></a>00740   <span class="keywordflow">if</span> (chromatid1.m_copyNumber != 1) {
<a name="l00741"></a>00741     <span class="keywordflow">return</span>;
<a name="l00742"></a>00742   }
<a name="l00743"></a>00743   std::string copy1Name;
<a name="l00744"></a>00744 
<a name="l00745"></a>00745   <span class="keywordflow">if</span> (geneCopyNumberZeroCallSet.<a class="code" href="classCallSet.html#adfd65845f5730a7e1ad09cf9823786d6" title="inline">getCallType</a>() != ADT_CALL_TYPE_NULL) {
<a name="l00746"></a>00746     copy1Name = geneCopyNumberZeroCallSet.m_name;
<a name="l00747"></a>00747   } <span class="keywordflow">else</span> {
<a name="l00748"></a>00748     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(results.m_experimentName + <span class="stringliteral">&quot; &quot;</span> + results.m_geneName + <span class="stringliteral">&quot;: copy number 1 values detected with no copy number 0 designation in the translation file.&quot;</span>);
<a name="l00749"></a>00749   }
<a name="l00750"></a>00750 
<a name="l00751"></a>00751   <span class="keywordflow">if</span> (results.size() == 0) {
<a name="l00752"></a>00752     results.<a class="code" href="classCallResults.html#ae6d30bdb376fc53fb8195bce055cf63e" title="CallResults::appendAlleleCall Synopsis:">appendAlleleCall</a>(copy1Name , 1, geneCopyNumberZeroCallSet);
<a name="l00753"></a>00753     <span class="keywordflow">return</span>;
<a name="l00754"></a>00754   }
<a name="l00755"></a>00755 
<a name="l00756"></a>00756   std::map&lt; std::string, bool&gt; callCount;
<a name="l00757"></a>00757   std::vector&lt;CallResults::AlleleCall&gt; newResults;
<a name="l00758"></a>00758 
<a name="l00759"></a>00759 
<a name="l00760"></a>00760   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = results.m_alleleCalls.size() - 1; i &gt;= 0 ; i--) {
<a name="l00761"></a>00761 
<a name="l00762"></a>00762     std::string alleleCall = copy1Name +  <span class="stringliteral">&quot;/&quot;</span> +  results.m_alleleCalls[i].m_allele;
<a name="l00763"></a>00763     <span class="comment">//cerr &lt;&lt; experimentName &lt;&lt; &quot; &quot; &lt;&lt;  geneName &lt;&lt; &quot; : copy number 0 call&quot;  &lt;&lt; alleleCall &lt;&lt; endl;</span>
<a name="l00764"></a>00764     alleleCall = alleleCall.substr(0, alleleCall.rfind(<span class="charliteral">&#39;/&#39;</span>));
<a name="l00765"></a>00765 
<a name="l00766"></a>00766     <span class="keywordflow">if</span> (callCount[alleleCall]) {
<a name="l00767"></a>00767       results.m_alleleCalls.erase(results.m_alleleCalls.begin() + 1);
<a name="l00768"></a>00768       results.m_markerCallCount--;
<a name="l00769"></a>00769     } <span class="keywordflow">else</span> {
<a name="l00770"></a>00770       callCount[alleleCall] = <span class="keyword">true</span>;
<a name="l00771"></a>00771       results.m_alleleCalls[i].m_allele = alleleCall;
<a name="l00772"></a>00772     }
<a name="l00773"></a>00773   }
<a name="l00774"></a>00774 
<a name="l00775"></a>00775   <span class="keywordflow">return</span>;
<a name="l00776"></a>00776 
<a name="l00777"></a>00777 }
<a name="l00778"></a>00778 <span class="comment">// end _translateExperimentCallCopyNumberOne</span>
<a name="l00779"></a>00779 <span class="comment">/*****************************************************************************/</span>
<a name="l00780"></a>00780 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00781"></a>00781 <span class="comment">/**</span>
<a name="l00782"></a>00782 <span class="comment"> * Translating an experiment call means checking for allele names that</span>
<a name="l00783"></a>00783 <span class="comment"> * correspond in the DMET3 translation table. For reporting purposes all</span>
<a name="l00784"></a>00784 <span class="comment"> * input experiment records from the Genotype CHIP or Short Report files</span>
<a name="l00785"></a>00785 <span class="comment"> * are to have a corresponding record somewhere in the reporting, i.e.</span>
<a name="l00786"></a>00786 <span class="comment"> * be included in the result set. There only exception are for unrecognized</span>
<a name="l00787"></a>00787 <span class="comment"> * genes. An unrecognized gene is simply skipped over. Records for unrecognized</span>
<a name="l00788"></a>00788 <span class="comment"> * genes will not be in the experiment results.</span>
<a name="l00789"></a>00789 <span class="comment"> *</span>
<a name="l00790"></a>00790 <span class="comment"> *</span>
<a name="l00791"></a>00791 <span class="comment"> * @param rte - RunTimeEnvironment which contains options</span>
<a name="l00792"></a>00792 <span class="comment"> * @param ttm - TranslationTableModel with the meta data about the call.</span>
<a name="l00793"></a>00793 <span class="comment"> * @param headerRow - the row in the ttm that has the meta data.</span>
<a name="l00794"></a>00794 <span class="comment"> *</span>
<a name="l00795"></a>00795 <span class="comment"> */</span>
<a name="l00796"></a>00796 <span class="comment">/*****************************************************************************/</span>
<a name="l00797"></a><a class="code" href="classGeneCall.html#a70871515298b6d8a54199225120c986a">00797</a> <a class="code" href="classCallResults.html">CallResults</a> <a class="code" href="classGeneCall.html#a70871515298b6d8a54199225120c986a" title="Translating an experiment call means checking for allele names that correspond in the DMET3 translati...">GeneCall::translateExperimentCall</a>(<a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00798"></a>00798     <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; completeSet,
<a name="l00799"></a>00799     <a class="code" href="classGenotypeTableModel.html">GenotypeTableModel</a> &amp; gtm,
<a name="l00800"></a>00800     <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid1,
<a name="l00801"></a>00801     <a class="code" href="classCallSet.html">CallSet</a> &amp; chromatid2,
<a name="l00802"></a>00802     <span class="keyword">const</span> std::string &amp; dmet2CopyNumberCall,
<a name="l00803"></a>00803     <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; geneCopyNumberZeroCallSet)
<a name="l00804"></a>00804 {
<a name="l00805"></a>00805 
<a name="l00806"></a>00806 
<a name="l00807"></a>00807   <span class="keywordtype">bool</span> isHaplotypeGroup
<a name="l00808"></a>00808   = (chromatid1.<a class="code" href="classCallSet.html#adfd65845f5730a7e1ad09cf9823786d6" title="inline">getCallType</a>() == ADT_CALL_TYPE_HAPLOTYPE_GROUP);
<a name="l00809"></a>00809 
<a name="l00810"></a>00810   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> geneCopyNumber = 2;
<a name="l00811"></a>00811 
<a name="l00812"></a>00812   <a class="code" href="classCallResults.html">CallResults</a> results(gtm.m_geneName, gtm.m_experimentName, chromatid1, chromatid2);
<a name="l00813"></a>00813 
<a name="l00814"></a>00814 
<a name="l00815"></a>00815   <span class="comment">// The callee passes in possibly invalid CallSets for the</span>
<a name="l00816"></a>00816   <span class="comment">// 2 chromosomes derived from the</span>
<a name="l00817"></a>00817   <span class="comment">// experiment records in the Genotype Short Report file.</span>
<a name="l00818"></a>00818   <span class="keywordflow">if</span> (!  _translateExperimentCallCheck(rte, completeSet, gtm.m_geneName,
<a name="l00819"></a>00819                                        gtm.m_experimentName, chromatid1,
<a name="l00820"></a>00820                                        chromatid2, m_alleleSet)) {
<a name="l00821"></a>00821     <span class="keywordflow">return</span> results;
<a name="l00822"></a>00822   }
<a name="l00823"></a>00823 
<a name="l00824"></a>00824   <span class="comment">//DMET2 COPY NUMBER 0</span>
<a name="l00825"></a>00825   <span class="keywordflow">if</span> (rte.m_adtOpts.m_dmet2Calling) {
<a name="l00826"></a>00826     <span class="keywordflow">if</span> (((rte.m_adtOpts.m_inputTTableType == ADT_TRANSLATION_TABLE_TYPE_DMET3) || isHaplotypeGroup) &amp;&amp; !dmet2CopyNumberCall.empty()) {
<a name="l00827"></a>00827       results.<a class="code" href="classCallResults.html#ae6d30bdb376fc53fb8195bce055cf63e" title="CallResults::appendAlleleCall Synopsis:">appendAlleleCall</a>(dmet2CopyNumberCall, 0);
<a name="l00828"></a>00828 
<a name="l00829"></a>00829       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, gtm.m_geneName + <span class="stringliteral">&quot;: &quot;</span> + gtm.m_experimentName + <span class="stringliteral">&quot;: copy number call: &quot;</span> + dmet2CopyNumberCall);
<a name="l00830"></a>00830       <span class="keywordflow">return</span> results;
<a name="l00831"></a>00831     }
<a name="l00832"></a>00832   } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (chromatid1.m_copyNumber == 0) {
<a name="l00833"></a>00833     <span class="keywordflow">if</span> (geneCopyNumberZeroCallSet.<a class="code" href="classCallSet.html#adfd65845f5730a7e1ad09cf9823786d6" title="inline">getCallType</a>() != ADT_CALL_TYPE_NULL) {
<a name="l00834"></a>00834       std::string alleleCall = geneCopyNumberZeroCallSet.m_name +  <span class="stringliteral">&quot;/&quot;</span> +  geneCopyNumberZeroCallSet.m_name;
<a name="l00835"></a>00835       results.<a class="code" href="classCallResults.html#ae6d30bdb376fc53fb8195bce055cf63e" title="CallResults::appendAlleleCall Synopsis:">appendAlleleCall</a>(alleleCall, 0, geneCopyNumberZeroCallSet, geneCopyNumberZeroCallSet);
<a name="l00836"></a>00836       <span class="keywordflow">return</span> results;
<a name="l00837"></a>00837     }
<a name="l00838"></a>00838 
<a name="l00839"></a>00839     <span class="keywordflow">else</span> {
<a name="l00840"></a>00840       geneCopyNumberZeroCallSet.<a class="code" href="classCallSet.html#a154f8707d153c4644d55f6adb68c8f8e" title="CallSet::describeVerbose.">describeVerbose</a>(rte, ADT_VERBOSE_NORMAL);
<a name="l00841"></a>00841       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(gtm.m_experimentName + <span class="stringliteral">&quot; &quot;</span> + gtm.m_geneName + <span class="stringliteral">&quot;: copy number 0 values detected with no copy number 0 designation in the translation file.&quot;</span>);
<a name="l00842"></a>00842     }
<a name="l00843"></a>00843   }
<a name="l00844"></a>00844 
<a name="l00845"></a>00845   geneCopyNumber = chromatid1.m_copyNumber;
<a name="l00846"></a>00846 
<a name="l00847"></a>00847   <span class="comment">// THPS = HPPA * AHPS - (AHPS * ( AHPS-1) )/ 2;</span>
<a name="l00848"></a>00848   <span class="keywordtype">int</span>                HPPA  = 1; <span class="comment">// Haplotype Pairs per allele</span>
<a name="l00849"></a>00849   <span class="keywordtype">int</span>                THPS  = 0; <span class="comment">// Total Haplotype pairs seen</span>
<a name="l00850"></a>00850   std::map&lt; std::string, int &gt; AHPS;      <span class="comment">// Allele haplotype pairs seen.</span>
<a name="l00851"></a>00851 
<a name="l00852"></a>00852   <span class="keywordtype">int</span>                THPR  = 0; <span class="comment">// Total Haplotype pairs required.</span>
<a name="l00853"></a>00853   <span class="comment">// THPR is a complex calculation broken out into another routine.</span>
<a name="l00854"></a>00854   THPR = _calculateTHPR(chromatid1, chromatid2, completeSet, HPPA);
<a name="l00855"></a>00855 
<a name="l00856"></a>00856   <span class="keywordflow">for</span> (std::vector&lt;CallSet&gt;::const_iterator iCSit = m_alleleSet.begin();
<a name="l00857"></a>00857        iCSit != m_alleleSet.end(); iCSit++) {
<a name="l00858"></a>00858     AHPS[iCSit-&gt;m_name] = 0;
<a name="l00859"></a>00859   }
<a name="l00860"></a>00860 
<a name="l00861"></a>00861 
<a name="l00862"></a>00862   <span class="keywordflow">if</span> (rte.m_currentVerbosity == ADT_VERBOSE_CALL) {
<a name="l00863"></a>00863     std::stringstream msgSStr;
<a name="l00864"></a>00864     msgSStr &lt;&lt; <span class="stringliteral">&quot;Total Haplotype Pairs Possible/Haplotype Pairs Per Allele: &quot;</span>;
<a name="l00865"></a>00865     msgSStr &lt;&lt; THPR &lt;&lt; <span class="stringliteral">&quot;/&quot;</span> &lt;&lt; HPPA &lt;&lt; endl;
<a name="l00866"></a>00866     msgSStr &lt;&lt; <span class="stringliteral">&quot;chromatid1: &quot;</span> &lt;&lt; (*chromatid1.m_ceSet.begin()).first &lt;&lt; <span class="stringliteral">&quot;, &quot;</span>;
<a name="l00867"></a>00867     msgSStr &lt;&lt; (*chromatid1.m_ceSet.begin()).second.basesToString() &lt;&lt; endl;
<a name="l00868"></a>00868     msgSStr &lt;&lt; <span class="stringliteral">&quot;chromatid2: &quot;</span> &lt;&lt; (*chromatid2.m_ceSet.begin()).first &lt;&lt; <span class="stringliteral">&quot;, &quot;</span>;
<a name="l00869"></a>00869     msgSStr &lt;&lt; (*chromatid2.m_ceSet.begin()).second.basesToString() &lt;&lt; endl;
<a name="l00870"></a>00870     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, msgSStr.str());
<a name="l00871"></a>00871   }
<a name="l00872"></a>00872 
<a name="l00873"></a>00873   <span class="comment">// The marker reference and variant sets are always the first two</span>
<a name="l00874"></a>00874   <span class="comment">// sets, skip over them for haplotype groups;</span>
<a name="l00875"></a>00875   <span class="keywordtype">int</span> offset = isHaplotypeGroup ? 2 : 0;
<a name="l00876"></a>00876 
<a name="l00877"></a>00877   <span class="keywordflow">for</span> (std::vector&lt;CallSet&gt;::const_iterator iCSit = m_alleleSet.begin() + offset;
<a name="l00878"></a>00878        iCSit != m_alleleSet.end() &amp;&amp; (THPS &lt; THPR);  ++iCSit) {
<a name="l00879"></a>00879 
<a name="l00880"></a>00880     <span class="comment">// The following condition is possible if the allele name was</span>
<a name="l00881"></a>00881     <span class="comment">// added as the second half of the call. If the number possible pairs</span>
<a name="l00882"></a>00882     <span class="comment">// has already been seen for this allele then just optimize</span>
<a name="l00883"></a>00883     <span class="comment">// by continuing.</span>
<a name="l00884"></a>00884     <span class="keywordflow">if</span> (!(AHPS[iCSit-&gt;m_name] &lt; HPPA))  <span class="keywordflow">continue</span>;
<a name="l00885"></a>00885 
<a name="l00886"></a>00886     <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; firstSet = *iCSit;
<a name="l00887"></a>00887 
<a name="l00888"></a>00888     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot;First CallSet: &quot;</span> + firstSet.m_name, <span class="keyword">false</span>);
<a name="l00889"></a>00889 
<a name="l00890"></a>00890     <span class="comment">// Check if any combination of the two chromsomes match each</span>
<a name="l00891"></a>00891     <span class="comment">// element. If so we have a match and the second match</span>
<a name="l00892"></a>00892     <span class="comment">// will be a result of the remaining bases not used to make the first</span>
<a name="l00893"></a>00893     <span class="comment">// match.</span>
<a name="l00894"></a>00894 
<a name="l00895"></a>00895     <span class="keywordflow">if</span> (! firstSet.<a class="code" href="classCallSet.html#ad9775c6b5351d62d7b04c8ef919f1285" title="CallSet::match Synopsis:">match</a>(chromatid1, chromatid2)) {
<a name="l00896"></a>00896       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot; not found!&quot;</span>);
<a name="l00897"></a>00897       <span class="keywordflow">continue</span>;
<a name="l00898"></a>00898     }
<a name="l00899"></a>00899 
<a name="l00900"></a>00900     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot; match!&quot;</span>);
<a name="l00901"></a>00901 
<a name="l00902"></a>00902     <span class="comment">// Build the second set from the difference between the</span>
<a name="l00903"></a>00903     <span class="comment">// the test set and the inputs.</span>
<a name="l00904"></a>00904 
<a name="l00905"></a>00905     <a class="code" href="classCallSet.html">CallSet</a> secondSet = firstSet.<a class="code" href="classCallSet.html#a12b6cc7556a97d7d3da3db9fd3d8c10d" title="CallSet::buildSecondSet Synopsis:">buildSecondSet</a>(rte, chromatid1, chromatid2);
<a name="l00906"></a>00906 
<a name="l00907"></a>00907     <span class="keywordflow">for</span> (std::vector&lt;CallSet&gt;::const_iterator jCSit = iCSit;
<a name="l00908"></a>00908          (jCSit != m_alleleSet.end()) &amp;&amp; (results.size() &lt; THPR) &amp;&amp;
<a name="l00909"></a>00909          (AHPS[firstSet.m_name] &lt; HPPA);   ++jCSit) {
<a name="l00910"></a>00910 
<a name="l00911"></a>00911       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot;Second CallSet: &quot;</span> + jCSit-&gt;m_name, <span class="keyword">false</span>);
<a name="l00912"></a>00912 
<a name="l00913"></a>00913       <span class="keywordflow">if</span> (secondSet == *jCSit) {
<a name="l00914"></a>00914 
<a name="l00915"></a>00915         <span class="keywordflow">if</span> (jCSit-&gt;m_name != firstSet.m_name) {
<a name="l00916"></a>00916           AHPS[jCSit-&gt;m_name]++;
<a name="l00917"></a>00917         }
<a name="l00918"></a>00918         AHPS[firstSet.m_name]++;
<a name="l00919"></a>00919 
<a name="l00920"></a>00920         std::string alleleCall = firstSet.m_name +  <span class="stringliteral">&quot;/&quot;</span> +  jCSit-&gt;m_name;
<a name="l00921"></a>00921         results.<a class="code" href="classCallResults.html#ae6d30bdb376fc53fb8195bce055cf63e" title="CallResults::appendAlleleCall Synopsis:">appendAlleleCall</a>(alleleCall, geneCopyNumber, firstSet, *jCSit);
<a name="l00922"></a>00922 
<a name="l00923"></a>00923         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot; match (&quot;</span> + alleleCall + <span class="stringliteral">&quot;)!&quot;</span>);
<a name="l00924"></a>00924       } <span class="keywordflow">else</span> {
<a name="l00925"></a>00925         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot; not found!&quot;</span>);
<a name="l00926"></a>00926       }
<a name="l00927"></a>00927 
<a name="l00928"></a>00928     } <span class="comment">// for each allele in the translation table get a second match.</span>
<a name="l00929"></a>00929 
<a name="l00930"></a>00930     <span class="keywordflow">if</span> (AHPS[firstSet.m_name] &lt; HPPA) {
<a name="l00931"></a>00931       std::string alleleCall = firstSet.m_name + <span class="stringliteral">&quot;/UNK&quot;</span> ;
<a name="l00932"></a>00932       results.<a class="code" href="classCallResults.html#ae6d30bdb376fc53fb8195bce055cf63e" title="CallResults::appendAlleleCall Synopsis:">appendAlleleCall</a>(alleleCall, geneCopyNumber, firstSet);
<a name="l00933"></a>00933     }
<a name="l00934"></a>00934     THPS += HPPA;
<a name="l00935"></a>00935 
<a name="l00936"></a>00936   } <span class="comment">// For each allele in the translation table, check for a match.</span>
<a name="l00937"></a>00937 
<a name="l00938"></a>00938   <span class="comment">// If the number of allele pairs seen is less than the total required</span>
<a name="l00939"></a>00939   <span class="comment">// then we add &quot;UNK/UNK&quot; to represent the difference.</span>
<a name="l00940"></a>00940   <span class="keywordflow">if</span> (THPS &lt; THPR) {
<a name="l00941"></a>00941     results.<a class="code" href="classCallResults.html#ae6d30bdb376fc53fb8195bce055cf63e" title="CallResults::appendAlleleCall Synopsis:">appendAlleleCall</a>(<span class="stringliteral">&quot;UNK/UNK&quot;</span>, geneCopyNumber);
<a name="l00942"></a>00942   }
<a name="l00943"></a>00943 
<a name="l00944"></a>00944   _translateExperimentCallCopyNumberOne(rte, results, chromatid1, geneCopyNumberZeroCallSet);
<a name="l00945"></a>00945 
<a name="l00946"></a>00946   results.<a class="code" href="classCallResults.html#ae0c2988cb11fffff4aa77896295f8870" title="CallResults::orderCalls Synopsis:">orderCalls</a>();
<a name="l00947"></a>00947 
<a name="l00948"></a>00948   <span class="keywordflow">return</span> results;
<a name="l00949"></a>00949 
<a name="l00950"></a>00950 }
<a name="l00951"></a>00951 <span class="comment">// end GeneCall::translateExperimentCall</span>
<a name="l00952"></a>00952 <span class="comment">/*****************************************************************************/</span>
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:03 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
