<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/GenoUtility.cpp File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<h1>GenoUtility.cpp File Reference</h1>  </div>
</div>
<div class="contents">
<code>#include &quot;<a class="el" href="GenoUtility_8h_source.html">chipstream/GenoUtility.h</a>&quot;</code><br/>
<code>#include &quot;<a class="el" href="QuantMethodFactory_8h_source.html">chipstream/QuantMethodFactory.h</a>&quot;</code><br/>
<code>#include &quot;<a class="el" href="TsvFile_8h_source.html">file/TsvFile/TsvFile.h</a>&quot;</code><br/>
<code>#include &quot;<a class="el" href="Convert_8h_source.html">util/Convert.h</a>&quot;</code><br/>
<code>#include &quot;<a class="el" href="Fs_8h_source.html">util/Fs.h</a>&quot;</code><br/>
<code>#include &quot;<a class="el" href="Util_8h_source.html">util/Util.h</a>&quot;</code><br/>
<code>#include &quot;<a class="el" href="Verbose_8h_source.html">util/Verbose.h</a>&quot;</code><br/>

<p><a href="GenoUtility_8cpp_source.html">Go to the source code of this file.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="func-members"></a>
Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#aacfe80c3e8bd468402502f56a7704b15">valueForTransformation</a> (double aVal, double bVal, enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> trans, double K, double &amp;xVal, double &amp;yVal)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">This is the engine that takes data and transforms it into clustering spaces.  <a href="#aacfe80c3e8bd468402502f56a7704b15"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a>&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#ab5c8b5877412aebfe6eb80db02bbc096">transformationForString</a> (const std::string &amp;str)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="classConvert.html" title="Convert.">Convert</a> a (text insensitive) text repesentation of a transformation into the enumration.  <a href="#ab5c8b5877412aebfe6eb80db02bbc096"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">std::string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a119c3e97f150b0b9b91dea503b19c424">stringForTransformation</a> (enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> t)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="classConvert.html" title="Convert.">Convert</a> a transformation enum to string representation.  <a href="#a119c3e97f150b0b9b91dea503b19c424"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">std::pair&lt; std::string, <br class="typebreak"/>
std::string &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a8686c3ca8eabf2bb104f81c970d2ea2d">columnNamesForTransformation</a> (enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> t)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Return the calumn names for the values generated by a given transformation.  <a href="#a8686c3ca8eabf2bb104f81c970d2ea2d"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a1a2531d6033d567af0446274e94f74a4">GenoUtility_transformData</a> (std::vector&lt; double &gt; &amp;aVals, std::vector&lt; double &gt; &amp;bVals, <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> m_Transform, double m_K)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">transform the data vector  <a href="#a1a2531d6033d567af0446274e94f74a4"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a1e7d2233042fe7b5a9fc83b4f79b545c">ContrastCentersStretch</a> (double K, double a, double b, double &amp;x, double &amp;y)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Contrast Centers Stretch.  <a href="#a1e7d2233042fe7b5a9fc83b4f79b545c"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#ac842841ae05672e63157ac80b8a016f6">ContrastExtremesStretch</a> (double K, double a, double b, double &amp;x, double &amp;y)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Contrast Extremes Stretch.  <a href="#ac842841ae05672e63157ac80b8a016f6"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#ac3236f3bb0d22355aaa716877d56f881">alleleMedians</a> (<a class="el" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel, const <a class="el" href="classProbeSet.html">ProbeSet</a> *ps, float &amp;Amedian, float &amp;Bmedian)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate the median PM values for the A and B alleles given data in cel file.  <a href="#ac3236f3bb0d22355aaa716877d56f881"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af7c858fd83f4b5c2bf8e27289c88ba30"></a><!-- doxytag: member="GenoUtility.cpp::probesetMean" ref="af7c858fd83f4b5c2bf8e27289c88ba30" args="(affymetrix_fusion_io::FusionCELData *cel, const ProbeSet *ps, float &amp;mean)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><b>probesetMean</b> (<a class="el" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel, const <a class="el" href="classProbeSet.html">ProbeSet</a> *ps, float &amp;mean)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5448c1a443cdced19df9217f8943fc51"></a><!-- doxytag: member="GenoUtility.cpp::probesetStd" ref="a5448c1a443cdced19df9217f8943fc51" args="(affymetrix_fusion_io::FusionCELData *cel, const ProbeSet *ps, float &amp;std)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><b>probesetStd</b> (<a class="el" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel, const <a class="el" href="classProbeSet.html">ProbeSet</a> *ps, float &amp;std)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aced747e28939fe3f56e4ab48558bc297"></a><!-- doxytag: member="GenoUtility.cpp::probesetMeanStd" ref="aced747e28939fe3f56e4ab48558bc297" args="(affymetrix_fusion_io::FusionCELData *cel, const ProbeSet *ps, float &amp;mean, float &amp;std)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><b>probesetMeanStd</b> (<a class="el" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel, const <a class="el" href="classProbeSet.html">ProbeSet</a> *ps, float &amp;mean, float &amp;std)</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a631afdfd53685cf4cc3f0cf46b448c5c">logRatio</a> (double sigA, double sigB)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Calculate the natural log ratio for two probeset signal.  <a href="#a631afdfd53685cf4cc3f0cf46b448c5c"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a208943f147d29a97c3aab8015aa707dc">loadTargetSketchFromFile</a> (const char *fileName, std::vector&lt; float &gt; &amp;sketch)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Open and read a target distribution from a text file.  <a href="#a208943f147d29a97c3aab8015aa707dc"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="GenoUtility_8cpp.html#a0267cd994e2dec10ca3a6f1e4c16113c">loadSNPHints</a> (std::string fileName, std::vector&lt; std::string &gt; &amp;colNames, std::map&lt; std::string, std::vector&lt; affx::GType &gt; &gt; &amp;knownCalls)</td></tr>
<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Read in the precomputed genotypes from a tab delimited file.  <a href="#a0267cd994e2dec10ca3a6f1e4c16113c"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5bb05df24f53389aa00c2fc8f05210b6"></a><!-- doxytag: member="GenoUtility.cpp::probesetMinMax" ref="a5bb05df24f53389aa00c2fc8f05210b6" args="(affymetrix_fusion_io::FusionCELData *cel, const ProbeSet *ps, float &amp;Min, float &amp;Max)" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><b>probesetMinMax</b> (<a class="el" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *cel, const <a class="el" href="classProbeSet.html">ProbeSet</a> *ps, float &amp;Min, float &amp;Max)</td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<dl class="author"><dt><b>Author:</b></dt><dd>Earl Hubbell </dd></dl>
<dl class="date"><dt><b>Date:</b></dt><dd>Aug 2 2006 </dd></dl>

<p>Definition in file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>
<hr/><h2>Function Documentation</h2>
<a class="anchor" id="ac3236f3bb0d22355aaa716877d56f881"></a><!-- doxytag: member="GenoUtility.cpp::alleleMedians" ref="ac3236f3bb0d22355aaa716877d56f881" args="(affymetrix_fusion_io::FusionCELData *cel, const ProbeSet *ps, float &amp;Amedian, float &amp;Bmedian)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool alleleMedians </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classaffymetrix__fusion__io_1_1FusionCELData.html">affymetrix_fusion_io::FusionCELData</a> *&nbsp;</td>
          <td class="paramname"> <em>cel</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classProbeSet.html">ProbeSet</a> *&nbsp;</td>
          <td class="paramname"> <em>ps</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float &amp;&nbsp;</td>
          <td class="paramname"> <em>Amedian</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float &amp;&nbsp;</td>
          <td class="paramname"> <em>Bmedian</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculate the median PM values for the A and B alleles given data in cel file. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>cel</em>&nbsp;</td><td>- Cel file of data. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ps</em>&nbsp;</td><td>- <a class="el" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> for determining A and B alleles. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>Amedian</em>&nbsp;</td><td>- Filled in with median PM A value. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>Bmedian</em>&nbsp;</td><td>- Filled in with median PM B value.</td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd></dd></dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00230">230</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="ProbeSet_8h_source.html#l00600">ProbeSet::atoms</a>, <a class="el" href="ProbeSet_8h_source.html#l00606">ProbeSet::atomsPerGroup</a>, <a class="el" href="FusionCELData_8cpp_source.html#l00379">affymetrix_fusion_io::FusionCELData::GetIntensity()</a>, <a class="el" href="ProbeSet_8h_source.html#l00215">Probe::id</a>, <a class="el" href="ProbeSet_8h_source.html#l00604">ProbeSet::numGroups</a>, <a class="el" href="ProbeSet_8h_source.html#l00310">Atom::probes</a>, and <a class="el" href="ProbeSet_8h_source.html#l00602">ProbeSet::psType</a>.</p>

<p>Referenced by <a class="el" href="HomHiLoCelListener_8cpp_source.html#l00161">HomHiLoCelListener::CalculateEMContrast()</a>, and <a class="el" href="EmGenderCelListener_8cpp_source.html#l00088">EmGenderCelListener::CalculateEMContrast()</a>.</p>

</div>
</div>
<a class="anchor" id="a8686c3ca8eabf2bb104f81c970d2ea2d"></a><!-- doxytag: member="GenoUtility.cpp::columnNamesForTransformation" ref="a8686c3ca8eabf2bb104f81c970d2ea2d" args="(enum Transformation t)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::pair&lt;std::string,std::string&gt; columnNamesForTransformation </td>
          <td>(</td>
          <td class="paramtype">enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a>&nbsp;</td>
          <td class="paramname"> <em>t</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Return the calumn names for the values generated by a given transformation. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>t</em>&nbsp;</td><td>- tranformation enum value. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>- a pair of column names </dd></dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00137">137</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="GenoUtility_8h_source.html#l00103">CCS</a>, <a class="el" href="GenoUtility_8h_source.html#l00102">CES</a>, <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, <a class="el" href="GenoUtility_8h_source.html#l00100">MvA</a>, <a class="el" href="GenoUtility_8h_source.html#l00101">RvT</a>, and <a class="el" href="Convert_8h_source.html#l00248">ToStr()</a>.</p>

<p>Referenced by <a class="el" href="QuantLabelZ_8h_source.html#l00391">QuantLabelZ::getAlleleValueNames()</a>.</p>

</div>
</div>
<a class="anchor" id="a1e7d2233042fe7b5a9fc83b4f79b545c"></a><!-- doxytag: member="GenoUtility.cpp::ContrastCentersStretch" ref="a1e7d2233042fe7b5a9fc83b4f79b545c" args="(double K, double a, double b, double &amp;x, double &amp;y)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void ContrastCentersStretch </td>
          <td>(</td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>K</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double &amp;&nbsp;</td>
          <td class="paramname"> <em>x</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double &amp;&nbsp;</td>
          <td class="paramname"> <em>y</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Contrast Centers Stretch. </p>
<p>In 2007 this is the current favorite for algorithms, scales values between -1 and 1. <img class="formulaInl" alt="$ X = asinh(K * (A-B)/(A+B)) / asinh(K) $" src="form_7.png"/> <br/>
 <img class="formulaInl" alt="$ Y = log_2(A + B) $" src="form_8.png"/> <br/>
</p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>K</em>&nbsp;</td><td>- </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>- Value (usually summarized intensity) from A allele. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>b</em>&nbsp;</td><td>- Value (usually summarized intensity) from B allele. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>x</em>&nbsp;</td><td>- </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>y</em>&nbsp;</td><td>- </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00185">185</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, and <a class="el" href="Util_8h_source.html#l00066">log2</a>.</p>

<p>Referenced by <a class="el" href="EmGenderCelListener_8cpp_source.html#l00088">EmGenderCelListener::CalculateEMContrast()</a>, and <a class="el" href="GenoUtility_8cpp_source.html#l00052">valueForTransformation()</a>.</p>

</div>
</div>
<a class="anchor" id="ac842841ae05672e63157ac80b8a016f6"></a><!-- doxytag: member="GenoUtility.cpp::ContrastExtremesStretch" ref="ac842841ae05672e63157ac80b8a016f6" args="(double K, double a, double b, double &amp;x, double &amp;y)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void ContrastExtremesStretch </td>
          <td>(</td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>K</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double &amp;&nbsp;</td>
          <td class="paramname"> <em>x</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double &amp;&nbsp;</td>
          <td class="paramname"> <em>y</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Contrast Extremes Stretch. </p>
<p><img class="formulaInl" alt="$ X = sinh(K * (A-B)/(A+B)) / sinh(K) $" src="form_9.png"/> <br/>
 <img class="formulaInl" alt="$ Y = log_2(A + B) $" src="form_8.png"/> <br/>
</p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>K</em>&nbsp;</td><td>- </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>- Value (usually summarized intensity) from A allele. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>b</em>&nbsp;</td><td>- Value (usually summarized intensity) from B allele. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>x</em>&nbsp;</td><td>- </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>y</em>&nbsp;</td><td>- </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00209">209</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, and <a class="el" href="Util_8h_source.html#l00066">log2</a>.</p>

<p>Referenced by <a class="el" href="HomHiLoCelListener_8cpp_source.html#l00161">HomHiLoCelListener::CalculateEMContrast()</a>, <a class="el" href="MultiChannelNonOverlapCelListener_8cpp_source.html#l00403">MultiChannelNonOverlapCelListener::GimmeContrast()</a>, <a class="el" href="MultiChannelHomHiLoCelListener_8cpp_source.html#l00276">MultiChannelHomHiLoCelListener::GimmeContrast()</a>, and <a class="el" href="GenoUtility_8cpp_source.html#l00052">valueForTransformation()</a>.</p>

</div>
</div>
<a class="anchor" id="a1a2531d6033d567af0446274e94f74a4"></a><!-- doxytag: member="GenoUtility.cpp::GenoUtility_transformData" ref="a1a2531d6033d567af0446274e94f74a4" args="(std::vector&lt; double &gt; &amp;aVals, std::vector&lt; double &gt; &amp;bVals, Transformation m_Transform, double m_K)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenoUtility_transformData </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; double &gt; &amp;&nbsp;</td>
          <td class="paramname"> <em>aVals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double &gt; &amp;&nbsp;</td>
          <td class="paramname"> <em>bVals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a>&nbsp;</td>
          <td class="paramname"> <em>m_Transform</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>m_K</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>transform the data vector </p>
<p>transforms a vector of data </p>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00159">159</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="GenoUtility_8cpp_source.html#l00052">valueForTransformation()</a>.</p>

<p>Referenced by <a class="el" href="QuantLabelZMulti_8cpp_source.html#l00820">QuantLabelZMulti::computeEstimate()</a>, <a class="el" href="QuantLabelZ_8cpp_source.html#l00319">QuantLabelZ::SelectGoodProbes()</a>, and <a class="el" href="QuantLabelZ_8h_source.html#l00331">QuantLabelZ::transform()</a>.</p>

</div>
</div>
<a class="anchor" id="a0267cd994e2dec10ca3a6f1e4c16113c"></a><!-- doxytag: member="GenoUtility.cpp::loadSNPHints" ref="a0267cd994e2dec10ca3a6f1e4c16113c" args="(std::string fileName, std::vector&lt; std::string &gt; &amp;colNames, std::map&lt; std::string, std::vector&lt; affx::GType &gt; &gt; &amp;knownCalls)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void loadSNPHints </td>
          <td>(</td>
          <td class="paramtype">std::string&nbsp;</td>
          <td class="paramname"> <em>fileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; std::string &gt; &amp;&nbsp;</td>
          <td class="paramname"> <em>colNames</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::map&lt; std::string, std::vector&lt; affx::GType &gt; &gt; &amp;&nbsp;</td>
          <td class="paramname"> <em>knownCalls</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Read in the precomputed genotypes from a tab delimited file. </p>
<p>Must be a column for each Celfile with same header name as celfile and an row for every snp to be analyzed. There can be extra columns and order of columns is not important as the resulting order will be the same as the colNames vector. Steal from <a class="el" href="classGenoSeedTxt.html" title="Implements interface for a genotyping oracle used for genotyping seeds in brlmm.">GenoSeedTxt</a>. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>fileName</em>&nbsp;</td><td>- text file containing the known genotype calls  colNames - the name of samples.  knownCalls -- the map that store knownCalls infomation for each SNP. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00438">438</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="Fs_8cpp_source.html#l00395">Fs::basename()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02765">affx::TsvFile::close()</a>, <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, <a class="el" href="BioTypes_8cpp_source.html#l00088">affx::GType_from_int()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>, <a class="el" href="Verbose_8cpp_source.html#l00259">Verbose::out()</a>, <a class="el" href="Convert_8h_source.html#l00248">ToStr()</a>, and <a class="el" href="TsvFile_8h_source.html#l00149">affx::TSV_BIND_REQUIRED</a>.</p>

</div>
</div>
<a class="anchor" id="a208943f147d29a97c3aab8015aa707dc"></a><!-- doxytag: member="GenoUtility.cpp::loadTargetSketchFromFile" ref="a208943f147d29a97c3aab8015aa707dc" args="(const char *fileName, std::vector&lt; float &gt; &amp;sketch)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void loadTargetSketchFromFile </td>
          <td>(</td>
          <td class="paramtype">const char *&nbsp;</td>
          <td class="paramname"> <em>fileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; float &gt; &amp;&nbsp;</td>
          <td class="paramname"> <em>sketch</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Open and read a target distribution from a text file. </p>
<p>Must have a column called 'intensities' with distribution quantiles sorted from highest to lowest.</p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>fileName</em>&nbsp;</td><td>- text file containing column of quantiles.  sketch - vector that stores the sketch. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00410">410</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="TsvFile_8cpp_source.html#l02765">affx::TsvFile::close()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02563">affx::TsvFile::cname2cidx()</a>, <a class="el" href="TsvFile_8cpp_source.html#l03344">affx::TsvFile::nextLevel()</a>, <a class="el" href="TsvFile_8cpp_source.html#l02643">affx::TsvFile::open()</a>, and <a class="el" href="TsvFile_8h_source.html#l00092">affx::TSV_OK</a>.</p>

</div>
</div>
<a class="anchor" id="a631afdfd53685cf4cc3f0cf46b448c5c"></a><!-- doxytag: member="GenoUtility.cpp::logRatio" ref="a631afdfd53685cf4cc3f0cf46b448c5c" args="(double sigA, double sigB)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double logRatio </td>
          <td>(</td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>sigA</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>sigB</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calculate the natural log ratio for two probeset signal. </p>
<p>signalA -- e.x. signal channel  signalB -- e.x. background channel </p>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00395">395</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a119c3e97f150b0b9b91dea503b19c424"></a><!-- doxytag: member="GenoUtility.cpp::stringForTransformation" ref="a119c3e97f150b0b9b91dea503b19c424" args="(enum Transformation t)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string stringForTransformation </td>
          <td>(</td>
          <td class="paramtype">enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a>&nbsp;</td>
          <td class="paramname"> <em>t</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p><a class="el" href="classConvert.html" title="Convert.">Convert</a> a transformation enum to string representation. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>t</em>&nbsp;</td><td>- tranformation enum value. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>- text description of enumeration. </dd></dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00111">111</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="GenoUtility_8h_source.html#l00103">CCS</a>, <a class="el" href="GenoUtility_8h_source.html#l00102">CES</a>, <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, <a class="el" href="GenoUtility_8h_source.html#l00100">MvA</a>, <a class="el" href="GenoUtility_8h_source.html#l00101">RvT</a>, and <a class="el" href="Convert_8h_source.html#l00248">ToStr()</a>.</p>

<p>Referenced by <a class="el" href="QuantLabelZ_8cpp_source.html#l00087">QuantLabelZ::QuantLabelZ()</a>.</p>

</div>
</div>
<a class="anchor" id="ab5c8b5877412aebfe6eb80db02bbc096"></a><!-- doxytag: member="GenoUtility.cpp::transformationForString" ref="ab5c8b5877412aebfe6eb80db02bbc096" args="(const std::string &amp;str)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> transformationForString </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&nbsp;</td>
          <td class="paramname"> <em>str</em></td>
          <td>&nbsp;)&nbsp;</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p><a class="el" href="classConvert.html" title="Convert.">Convert</a> a (text insensitive) text repesentation of a transformation into the enumration. </p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>s</em>&nbsp;</td><td>- string specifying transformation. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>- Enumeration version of transformation. </dd></dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00090">90</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="Util_8h_source.html#l00114">Util::downcaseString()</a>, and <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>.</p>

<p>Referenced by <a class="el" href="QuantLabelZMulti_8cpp_source.html#l00163">QuantLabelZMulti::newObject()</a>, and <a class="el" href="QuantLabelZ_8cpp_source.html#l01232">QuantLabelZ::newObject()</a>.</p>

</div>
</div>
<a class="anchor" id="aacfe80c3e8bd468402502f56a7704b15"></a><!-- doxytag: member="GenoUtility.cpp::valueForTransformation" ref="aacfe80c3e8bd468402502f56a7704b15" args="(double aVal, double bVal, enum Transformation trans, double K, double &amp;xVal, double &amp;yVal)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void valueForTransformation </td>
          <td>(</td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>aVal</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>bVal</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">enum <a class="el" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a>&nbsp;</td>
          <td class="paramname"> <em>trans</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&nbsp;</td>
          <td class="paramname"> <em>K</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double &amp;&nbsp;</td>
          <td class="paramname"> <em>xVal</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double &amp;&nbsp;</td>
          <td class="paramname"> <em>yVal</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>This is the engine that takes data and transforms it into clustering spaces. </p>
<p>the engine that transforms the data</p>
<dl><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>aVal</em>&nbsp;</td><td>- a allele probe summary </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>bVal</em>&nbsp;</td><td>- b allele probe summary </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>trans</em>&nbsp;</td><td>- transformation of data </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>K</em>&nbsp;</td><td>- stretching parameter </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>xVal</em>&nbsp;</td><td>- output difference of a vs b, transformed </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>yVal</em>&nbsp;</td><td>- output total amount of stuff, transformed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="GenoUtility_8cpp_source.html#l00052">52</a> of file <a class="el" href="GenoUtility_8cpp_source.html">GenoUtility.cpp</a>.</p>

<p>References <a class="el" href="GenoUtility_8h_source.html#l00103">CCS</a>, <a class="el" href="GenoUtility_8h_source.html#l00102">CES</a>, <a class="el" href="GenoUtility_8cpp_source.html#l00185">ContrastCentersStretch()</a>, <a class="el" href="GenoUtility_8cpp_source.html#l00209">ContrastExtremesStretch()</a>, <a class="el" href="Err_8cpp_source.html#l00068">Err::errAbort()</a>, <a class="el" href="Util_8h_source.html#l00066">log2</a>, <a class="el" href="GenoUtility_8h_source.html#l00100">MvA</a>, <a class="el" href="GenoUtility_8h_source.html#l00101">RvT</a>, and <a class="el" href="Convert_8h_source.html#l00248">ToStr()</a>.</p>

<p>Referenced by <a class="el" href="GenoUtility_8cpp_source.html#l00159">GenoUtility_transformData()</a>.</p>

</div>
</div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:07 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
