<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/GenoSeed.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>GenoSeed.h</h1>  </div>
</div>
<div class="contents">
<a href="GenoSeed_8h.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment">/**</span>
<a name="l00022"></a>00022 <span class="comment"> * @file   GenoSeed.h</span>
<a name="l00023"></a>00023 <span class="comment"> * @author Chuck Sugnet</span>
<a name="l00024"></a>00024 <span class="comment"> * @date   Fri Jun 30 14:44:22 2006</span>
<a name="l00025"></a>00025 <span class="comment"> * </span>
<a name="l00026"></a>00026 <span class="comment"> * @brief Abstract interface for a genotyping oracle used for</span>
<a name="l00027"></a>00027 <span class="comment"> * genotyping seeds in brlmm. Originally seed calls came from DM, but</span>
<a name="l00028"></a>00028 <span class="comment"> * now they are a subject of much research.</span>
<a name="l00029"></a>00029 <span class="comment"> * </span>
<a name="l00030"></a>00030 <span class="comment"> */</span>
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 <span class="preprocessor">#ifndef _GENOSEED_H_</span>
<a name="l00033"></a>00033 <span class="preprocessor"></span><span class="preprocessor">#define _GENOSEED_H_</span>
<a name="l00034"></a>00034 <span class="preprocessor"></span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="BioTypes_8h.html" title="Central location enumerations and other types that aren&amp;#39;t particular to a single class or applica...">chipstream/BioTypes.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;chipstream/GenderCalls.h&quot;</span>
<a name="l00037"></a>00037 <span class="comment"></span>
<a name="l00038"></a>00038 <span class="comment">/** Abstract interface for a genotyping oracle used for genotyping</span>
<a name="l00039"></a>00039 <span class="comment">    seeds in brlmm. */</span>
<a name="l00040"></a><a class="code" href="classGenoSeed.html">00040</a> <span class="keyword">class </span><a class="code" href="classGenoSeed.html" title="Abstract interface for a genotyping oracle used for genotyping seeds in brlmm.">GenoSeed</a> : <span class="keyword">public</span> <a class="code" href="classGenderCalls.html" title="Class for determining gender from a cel set of cel files.">GenderCalls</a> {
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 <span class="keyword">public</span>:
<a name="l00043"></a>00043 <span class="comment"></span>
<a name="l00044"></a>00044 <span class="comment">  /** How many chips does this object know about. */</span>
<a name="l00045"></a>00045   <span class="keyword">virtual</span> <span class="keywordtype">int</span> <a class="code" href="classGenoSeed.html#a13f8c4af007a0a50998d3275ae117214" title="How many chips does this object know about.">getChipCount</a>() = 0;
<a name="l00046"></a>00046 <span class="comment"></span>
<a name="l00047"></a>00047 <span class="comment">  /** </span>
<a name="l00048"></a>00048 <span class="comment">   * Get the call rates for snps for each chip.</span>
<a name="l00049"></a>00049 <span class="comment">   * @param callRates - Vector to be filled in with call rates for chrx snps,</span>
<a name="l00050"></a>00050 <span class="comment">   * one for each chip.</span>
<a name="l00051"></a>00051 <span class="comment">   */</span>
<a name="l00052"></a>00052   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenoSeed.html#a4d31c7d9a36963eee970be1c3a8ebc91" title="Get the call rates for snps for each chip.">getCallRates</a>(std::vector&lt;float&gt; &amp;callRates) = 0;
<a name="l00053"></a>00053 <span class="comment"></span>
<a name="l00054"></a>00054 <span class="comment">  /** </span>
<a name="l00055"></a>00055 <span class="comment">   * Get the call rates for chrX snps for each chip.</span>
<a name="l00056"></a>00056 <span class="comment">   * @param callRates - Vector to be filled in with call rates for chrx snps,</span>
<a name="l00057"></a>00057 <span class="comment">   * one for each chip.</span>
<a name="l00058"></a>00058 <span class="comment">   */</span>
<a name="l00059"></a>00059   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classGenoSeed.html#a393f1b7d87769d90e2b2cb1e6ce846a7" title="Get the call rates for chrX snps for each chip.">getHetChrXRates</a>(std::vector&lt;float&gt; &amp;hetRates) = 0;
<a name="l00060"></a>00060 
<a name="l00061"></a>00061   <span class="keyword">virtual</span> <span class="keywordtype">float</span> getHetChrXRate(<span class="keywordtype">int</span> chip) = 0;
<a name="l00062"></a>00062 <span class="comment"></span>
<a name="l00063"></a>00063 <span class="comment">  /** </span>
<a name="l00064"></a>00064 <span class="comment">   * Get the genotype calls for a particular probeset.</span>
<a name="l00065"></a>00065 <span class="comment">   * @param name - name of probeset to get genotype calls for.</span>
<a name="l00066"></a>00066 <span class="comment">   * @return - A vector of the calls from dm algorithm.</span>
<a name="l00067"></a>00067 <span class="comment">   */</span>
<a name="l00068"></a>00068   <span class="keyword">virtual</span> std::vector&lt;affx::GType&gt; <a class="code" href="classGenoSeed.html#ab3c2932ea7190946962b670a5a5ed259" title="Get the genotype calls for a particular probeset.">getGenoCalls</a>(<span class="keyword">const</span> std::string &amp;name) = 0;
<a name="l00069"></a>00069 <span class="comment"></span>
<a name="l00070"></a>00070 <span class="comment">  /**</span>
<a name="l00071"></a>00071 <span class="comment">   * @brief Check if a call exists for a given probeset</span>
<a name="l00072"></a>00072 <span class="comment">   * @param name - name of probeset to get genotype calls for.</span>
<a name="l00073"></a>00073 <span class="comment">   * @return boolean result</span>
<a name="l00074"></a>00074 <span class="comment">   */</span>
<a name="l00075"></a>00075   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> <a class="code" href="classGenoSeed.html#a07a49a96380e19af763958c6f9a5fbdb" title="Check if a call exists for a given probeset.">checkGenoCallsName</a>(<span class="keyword">const</span> std::string &amp;name) = 0;
<a name="l00076"></a>00076 <span class="comment"></span>
<a name="l00077"></a>00077 <span class="comment">  /** Virtual destructor for virtual class. */</span>
<a name="l00078"></a>00078   <span class="keyword">virtual</span> <a class="code" href="classGenoSeed.html#a45e00fa3d7778ebe91320dacccd4e3fd" title="Virtual destructor for virtual class.">~GenoSeed</a>();
<a name="l00079"></a>00079 <span class="comment"></span>
<a name="l00080"></a>00080 <span class="comment">  /**</span>
<a name="l00081"></a>00081 <span class="comment">   * Figure out the gender of samples based on the number of het calls on</span>
<a name="l00082"></a>00082 <span class="comment">   * chromosome X. This is all very human centric...</span>
<a name="l00083"></a>00083 <span class="comment">   */</span>
<a name="l00084"></a>00084   <span class="keyword">virtual</span> std::vector&lt;affx::Gender&gt; <a class="code" href="classGenoSeed.html#afd738e20e2c27c358e59ba02defa8c64" title="Figure out the gender of samples based on the number of het calls on chromosome X.">getGenders</a>();
<a name="l00085"></a>00085 
<a name="l00086"></a>00086   <span class="keyword">virtual</span> <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669" title="Those pesky XY people need a different prior for all the SNPs on X that aren&amp;#39;t on the psuedoautos...">affx::Gender</a> getGender(<span class="keywordtype">int</span> chip);
<a name="l00087"></a>00087 
<a name="l00088"></a>00088 };
<a name="l00089"></a>00089 
<a name="l00090"></a>00090 <span class="preprocessor">#endif </span><span class="comment">/* _GENOSEED_H_ */</span>
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:51 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
