<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: translation/GenotypeOverrideTableModel.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_b6d6d4a2e58a0f89851f849ab50b1d34.html">translation</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>GenotypeOverrideTableModel.cpp</h1>  </div>
</div>
<div class="contents">
<a href="GenotypeOverrideTableModel_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2008 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   GenotypeOverrideTableModel.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Mybrid Spalding</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Tue Jul 22 08:33:25 PDT 2008</span>
<a name="l00026"></a>00026 <span class="comment"> * @brief  Class to supplant Genotype data (CHP) with user supplied data.</span>
<a name="l00027"></a>00027 <span class="comment"> */</span>
<a name="l00028"></a>00028 
<a name="l00029"></a>00029 <span class="comment">//</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="GenotypeOverrideTableModel_8h.html" title="Class to supplant Genotype data (CHP) with user supplied data.">translation/GenotypeOverrideTableModel.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="comment">//</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="CallElement_8h.html" title="Atomic class element for translation.">translation/CallElement.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="RunTimeEnvironment_8h.html" title="A single instance class to hold the global runtime environment, similar shell enviorment variables...">translation/RunTimeEnvironment.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="TranslationTableModel_8h.html" title="Single instance data model class that wraps the translation file in the TranslationInputTsvModel.">translation/TranslationTableModel.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="comment">//</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;<a class="code" href="Err_8h.html" title="Some utility error functions, currently prints message and aborts().">util/Err.h</a>&quot;</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00038"></a>00038 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00039"></a>00039 <span class="comment">//</span>
<a name="l00040"></a>00040 <span class="preprocessor">#include &quot;pcrecpp.h&quot;</span>
<a name="l00041"></a>00041 <span class="comment">//</span>
<a name="l00042"></a>00042 <span class="preprocessor">#include &lt;sstream&gt;</span>
<a name="l00043"></a>00043 
<a name="l00044"></a>00044 <span class="keyword">using namespace </span>affx;
<a name="l00045"></a>00045 <span class="keyword">using namespace </span>std;
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 <span class="keyword">const</span> <a class="code" href="classTittmColumnDefinition.html">TittmColumnDefinition</a> GOTM_COLUMN_DEFINITIONS[] = { <span class="comment">// column_name, empty_ok, index, valid_regex</span>
<a name="l00048"></a>00048   { std::string(<span class="stringliteral">&quot;CHP Filename&quot;</span>),      0, 0, std::string(<span class="stringliteral">&quot;.*&quot;</span>), NULL },
<a name="l00049"></a>00049   { std::string(<span class="stringliteral">&quot;Gene&quot;</span>),              0, 1,
<a name="l00050"></a>00050     std::string(<span class="stringliteral">&quot;^[A-Z\\d]+$&quot;</span>), NULL},
<a name="l00051"></a>00051   { std::string(<span class="stringliteral">&quot;Common Name&quot;</span>),       0, 2, std::string(<span class="stringliteral">&quot;.*&quot;</span>), NULL  },
<a name="l00052"></a>00052   { std::string(<span class="stringliteral">&quot;Probe Set ID&quot;</span>),      0, 3,
<a name="l00053"></a>00053     std::string(<span class="stringliteral">&quot;^(?:[\\w\\d]+)$&quot;</span>), NULL },
<a name="l00054"></a>00054   { std::string(<span class="stringliteral">&quot;Basecall&quot;</span>),          1, 4,
<a name="l00055"></a>00055     std::string(<span class="stringliteral">&quot;^(?:0|ZeroCopyNumber|PossibleRareAllele|NoCall|NotAvailable|PRA|NC|-)|(?:(?:[ACGT0]+|INS|Ins|DEL|Del|-)/(?:[ACGT]+|INS|Ins|DEL|Del|-))$&quot;</span>), NULL },
<a name="l00056"></a>00056   { std::string(<span class="stringliteral">&quot;Override Comment&quot;</span>),  1, 5, std::string(<span class="stringliteral">&quot;.*&quot;</span>), NULL },
<a name="l00057"></a>00057   { std::string(<span class="stringliteral">&quot;Reference Allele&quot;</span>),  0, 6,
<a name="l00058"></a>00058     std::string(<span class="stringliteral">&quot;^(?:[ACGT0]+)|(?:INS)|(?:DEL)|(?:-)$&quot;</span>), NULL },
<a name="l00059"></a>00059   { std::string(<span class="stringliteral">&quot;Variant Allele&quot;</span>),    0, 7,
<a name="l00060"></a>00060     std::string(<span class="stringliteral">&quot;^(?:[ACGT]+|INS|Ins|DEL|Del|-)(?:[,](?:[ACGT]+|INS|Ins|DEL|Del|-))*$&quot;</span>), NULL },
<a name="l00061"></a>00061 };
<a name="l00062"></a>00062 
<a name="l00063"></a>00063 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00064"></a>00064 <span class="comment">/**</span>
<a name="l00065"></a>00065 <span class="comment"> * filterRowCallback:</span>
<a name="l00066"></a>00066 <span class="comment"> * Synopsis:</span>
<a name="l00067"></a>00067 <span class="comment"> *</span>
<a name="l00068"></a>00068 <span class="comment"> *  A callback function passed to TranslationInputTsvTableModel.</span>
<a name="l00069"></a>00069 <span class="comment"> *</span>
<a name="l00070"></a>00070 <span class="comment"> *  DEPRECATED: As of 10/08/2008 filtering is deprecated. If someone</span>
<a name="l00071"></a>00071 <span class="comment"> *  wants to override a specific call with a NoCall they can now do so.</span>
<a name="l00072"></a>00072 <span class="comment"> *  This function filters out NoCall values such that they would never</span>
<a name="l00073"></a>00073 <span class="comment"> *  be used from the override file.</span>
<a name="l00074"></a>00074 <span class="comment"> *</span>
<a name="l00075"></a>00075 <span class="comment"> *</span>
<a name="l00076"></a>00076 <span class="comment"> *</span>
<a name="l00077"></a>00077 <span class="comment"> * @param name1 - description</span>
<a name="l00078"></a>00078 <span class="comment"> * @param name1 - description</span>
<a name="l00079"></a>00079 <span class="comment"> * @return - description</span>
<a name="l00080"></a>00080 <span class="comment"> */</span>
<a name="l00081"></a>00081 <span class="comment">/*****************************************************************************/</span>
<a name="l00082"></a>00082 <span class="comment">/*</span>
<a name="l00083"></a>00083 <span class="comment">static bool filterRowCallback(std::vector&lt; std::string &gt; &amp; row)</span>
<a name="l00084"></a>00084 <span class="comment">{</span>
<a name="l00085"></a>00085 <span class="comment"></span>
<a name="l00086"></a>00086 <span class="comment">  pcrecpp::RE filter(&quot;No[cC]all|PossibleRareAllele|NC|PRA&quot;);</span>
<a name="l00087"></a>00087 <span class="comment"></span>
<a name="l00088"></a>00088 <span class="comment">  APT_ERR_ASSERT(row.size() == GenotypeOverrideTableModel::END_GOTM_FIELD_INDEX, &quot;&quot;);</span>
<a name="l00089"></a>00089 <span class="comment"></span>
<a name="l00090"></a>00090 <span class="comment">  if (row[GenotypeOverrideTableModel::BASECALL_INDEX].empty() ||</span>
<a name="l00091"></a>00091 <span class="comment">      filter.PartialMatch(row[GenotypeOverrideTableModel::BASECALL_INDEX])) {</span>
<a name="l00092"></a>00092 <span class="comment">    return true;</span>
<a name="l00093"></a>00093 <span class="comment">  }</span>
<a name="l00094"></a>00094 <span class="comment"></span>
<a name="l00095"></a>00095 <span class="comment"></span>
<a name="l00096"></a>00096 <span class="comment">  return false;</span>
<a name="l00097"></a>00097 <span class="comment"></span>
<a name="l00098"></a>00098 <span class="comment">}</span>
<a name="l00099"></a>00099 <span class="comment">*/</span>
<a name="l00100"></a>00100 <span class="comment">// end filterRowCallback</span>
<a name="l00101"></a>00101 <span class="comment">/*****************************************************************************/</span>
<a name="l00102"></a>00102 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00103"></a>00103 <span class="comment">/**</span>
<a name="l00104"></a>00104 <span class="comment"> * GenotypeOverrideTableModel::GenotypeOverrideTableModel</span>
<a name="l00105"></a>00105 <span class="comment"> * Synopsis:</span>
<a name="l00106"></a>00106 <span class="comment"> *</span>
<a name="l00107"></a>00107 <span class="comment"> * This is the TransferObject to the TsvFile DataAccessObject design</span>
<a name="l00108"></a>00108 <span class="comment"> * pattern. There is only ever one of these in existence, i.e. this</span>
<a name="l00109"></a>00109 <span class="comment"> * is a single instance class.</span>
<a name="l00110"></a>00110 <span class="comment"> *</span>
<a name="l00111"></a>00111 <span class="comment"> * genotype override file -&gt; genotype override table model</span>
<a name="l00112"></a>00112 <span class="comment"> *</span>
<a name="l00113"></a>00113 <span class="comment"> * Default TSV constructor for the GenotypeOverrideTableModel.</span>
<a name="l00114"></a>00114 <span class="comment"> *</span>
<a name="l00115"></a>00115 <span class="comment"> * This constructor is beefy for one reason: to validate</span>
<a name="l00116"></a>00116 <span class="comment"> * rows and then to swap the row information with error messaging</span>
<a name="l00117"></a>00117 <span class="comment"> * in the event that the row is invalid.</span>
<a name="l00118"></a>00118 <span class="comment"> *</span>
<a name="l00119"></a>00119 <span class="comment"> * Unlike most other input files for the translation algorithm,</span>
<a name="l00120"></a>00120 <span class="comment"> * the override file is permitted to have bad data and translation</span>
<a name="l00121"></a>00121 <span class="comment"> * still runs. When bad rows are detected, an error message is stuffed</span>
<a name="l00122"></a>00122 <span class="comment"> * into the override comment field and the override row is otherwise</span>
<a name="l00123"></a>00123 <span class="comment"> * ignored.</span>
<a name="l00124"></a>00124 <span class="comment"> *</span>
<a name="l00125"></a>00125 <span class="comment"> * Validation occurs per allele or probe set with the exception of</span>
<a name="l00126"></a>00126 <span class="comment"> * Zero Copy Number. With Zero Copy Number all probe sets for a gene</span>
<a name="l00127"></a>00127 <span class="comment"> * must be present or the override is rejected.</span>
<a name="l00128"></a>00128 <span class="comment"> * </span>
<a name="l00129"></a>00129 <span class="comment"> * @param rte - RunTimeEnvironment which contains options</span>
<a name="l00130"></a>00130 <span class="comment"> * @param genoOverrideFileName - the input file to read in.</span>
<a name="l00131"></a>00131 <span class="comment"> * @param ttm - The translation table instance used for validation.</span>
<a name="l00132"></a>00132 <span class="comment"> */</span>
<a name="l00133"></a>00133 <span class="comment">/*****************************************************************************/</span>
<a name="l00134"></a><a class="code" href="classGenotypeOverrideTableModel.html#a8ae358c8e4de5a882bb06c59be229318">00134</a> <a class="code" href="classGenotypeOverrideTableModel.html#a8ae358c8e4de5a882bb06c59be229318" title="filterRowCallback: Synopsis:">GenotypeOverrideTableModel::GenotypeOverrideTableModel</a>(<span class="keyword">const</span> <span class="keyword">class</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp;rte,  <span class="keyword">const</span> std::string &amp;genoOverrideFileName, <span class="keyword">class</span> <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm) :
<a name="l00135"></a>00135     <a class="code" href="classTranslationInputTsvTableModel.html">TranslationInputTsvTableModel</a>(rte, genoOverrideFileName, GOTM_COLUMN_DEFINITIONS, (sizeof(GOTM_COLUMN_DEFINITIONS) / sizeof(GOTM_COLUMN_DEFINITIONS[0])), true, NULL, NULL, false)
<a name="l00136"></a>00136 {
<a name="l00137"></a>00137 
<a name="l00138"></a>00138   std::vector&lt; std::string &gt; alleles;
<a name="l00139"></a>00139   std::set&lt; std::string &gt;  copyNumberZeroGenes;
<a name="l00140"></a>00140   std::map&lt; std::string, std::set&lt; std::string &gt; &gt; geneMarkers;
<a name="l00141"></a>00141   std::stringstream        invalidSStr;
<a name="l00142"></a>00142   pcrecpp::RE              zeroCopyRE(<span class="stringliteral">&quot;ZeroCopyNumber|0&quot;</span>);
<a name="l00143"></a>00143   pcrecpp::RE              zeroCopyValidSetRE(<span class="stringliteral">&quot;ZeroCopyNumber|0|NotAvailable&quot;</span>);
<a name="l00144"></a>00144   
<a name="l00145"></a>00145   <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> row = 0; row &lt; m_rows.size(); row++) {
<a name="l00146"></a>00146 
<a name="l00147"></a>00147     <span class="comment">// VALIDATION</span>
<a name="l00148"></a>00148     <span class="comment">// We requested that validation not be done by the constructor so</span>
<a name="l00149"></a>00149     <span class="comment">// that special action can be taken here with respect to validation.</span>
<a name="l00150"></a>00150     <span class="comment">// Specifically, invalid rows are appended to a reporting structure</span>
<a name="l00151"></a>00151     <span class="comment">// and does not cause the translation to stop.</span>
<a name="l00152"></a>00152 
<a name="l00153"></a>00153     <span class="keywordtype">bool</span> okRow         = <span class="keyword">true</span>;
<a name="l00154"></a>00154     <span class="keywordtype">bool</span> okRowProbeSet = <span class="keyword">true</span>;
<a name="l00155"></a>00155     std::string probeSet    = m_rows[row][PROBE_SET_INDEX];
<a name="l00156"></a>00156     std::string gene        = m_rows[row][GENE_INDEX];
<a name="l00157"></a>00157     std::string chpFile     = m_rows[row][CHP_FILE_INDEX] ;
<a name="l00158"></a>00158     std::string key         = chpFile + <span class="stringliteral">&quot;::&quot;</span> + probeSet;
<a name="l00159"></a>00159     std::string geneKey     = chpFile + <span class="stringliteral">&quot;::&quot;</span> + gene;
<a name="l00160"></a>00160     alleles.clear();
<a name="l00161"></a>00161     
<a name="l00162"></a>00162     invalidSStr.str(<span class="stringliteral">&quot;&quot;</span>);
<a name="l00163"></a>00163       
<a name="l00164"></a>00164     invalidSStr &lt;&lt; <span class="stringliteral">&quot;Invalid basecall override ignored&quot;</span>;
<a name="l00165"></a>00165 
<a name="l00166"></a>00166     <span class="keywordflow">if</span> ( (geneMarkers.count( geneKey )  == 0 ) || ( !geneMarkers[geneKey].count(probeSet)) )  {
<a name="l00167"></a>00167       geneMarkers[geneKey].insert(probeSet);
<a name="l00168"></a>00168     }
<a name="l00169"></a>00169     <span class="keywordflow">else</span> {
<a name="l00170"></a>00170       okRow = <span class="keyword">false</span>;
<a name="l00171"></a>00171       invalidSStr &lt;&lt; <span class="stringliteral">&quot; | &quot;</span> &lt;&lt; chpFile &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; gene &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <span class="stringliteral">&quot; duplicate  probe set &quot;</span> &lt;&lt; probeSet &lt;&lt; <span class="stringliteral">&quot; detected.&quot;</span> &lt;&lt;  endl;
<a name="l00172"></a>00172     }
<a name="l00173"></a>00173 
<a name="l00174"></a>00174     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> col = 0; (col &lt; <a class="code" href="classTranslationInputTsvTableModel.html#ae33ba37f59967a7bd0caefe58de785ed" title="TranslationInputTsvTableModel::columnCount Synopsis:">columnCount</a>()) &amp;&amp; okRow; col++) {
<a name="l00175"></a>00175 
<a name="l00176"></a>00176       std::string columnValue = m_rows[row][col];
<a name="l00177"></a>00177         
<a name="l00178"></a>00178       <span class="keywordflow">if</span> (! GOTM_COLUMN_DEFINITIONS[col].m_validRE.FullMatch(columnValue)) {
<a name="l00179"></a>00179         okRow = <span class="keyword">false</span>;
<a name="l00180"></a>00180 
<a name="l00181"></a>00181         invalidSStr &lt;&lt; <span class="stringliteral">&quot; | Unrecognized &quot;</span> &lt;&lt; GOTM_COLUMN_DEFINITIONS[col].m_columnName &lt;&lt; <span class="stringliteral">&quot; (&quot;</span> &lt;&lt; columnValue &lt;&lt; <span class="stringliteral">&quot;) &quot;</span> ;
<a name="l00182"></a>00182 
<a name="l00183"></a>00183         <span class="keywordflow">if</span> (col == PROBE_SET_INDEX) {
<a name="l00184"></a>00184           okRowProbeSet = <span class="keyword">false</span>;
<a name="l00185"></a>00185         }
<a name="l00186"></a>00186 
<a name="l00187"></a>00187       } <span class="comment">// if column is valid</span>
<a name="l00188"></a>00188 
<a name="l00189"></a>00189     } <span class="comment">// for each column</span>
<a name="l00190"></a>00190 
<a name="l00191"></a>00191 
<a name="l00192"></a>00192     <span class="comment">// BASECALL VALIDATION</span>
<a name="l00193"></a>00193     <span class="comment">// Make sure the base call is comprised of values found in the</span>
<a name="l00194"></a>00194     <span class="comment">// reference and variant columns.</span>
<a name="l00195"></a>00195 
<a name="l00196"></a>00196     <span class="keywordflow">if</span> (okRow) {
<a name="l00197"></a>00197 
<a name="l00198"></a>00198       alleles = <a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">getOverrideAlleles</a>(m_rows[row][CHP_FILE_INDEX], m_rows[row][PROBE_SET_INDEX], row);
<a name="l00199"></a>00199 
<a name="l00200"></a>00200       <span class="keywordflow">if</span> ((alleles.size() != 2) &amp;&amp; !zeroCopyValidSetRE.FullMatch(alleles[0] ) ) {
<a name="l00201"></a>00201         okRow = <span class="keyword">false</span>;
<a name="l00202"></a>00202         invalidSStr &lt;&lt; <span class="stringliteral">&quot; | basecall format must be a copy number two designation  allele/allele &quot;</span> &lt;&lt; m_rows[row][BASECALL_INDEX];
<a name="l00203"></a>00203       }
<a name="l00204"></a>00204  
<a name="l00205"></a>00205       <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; okRow &amp;&amp; (i &lt; alleles.size()); i++) {
<a name="l00206"></a>00206 
<a name="l00207"></a>00207         <span class="keywordflow">if</span> (!<a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(alleles[i]) &amp;&amp;
<a name="l00208"></a>00208             !zeroCopyRE.FullMatch(alleles[i]) &amp;&amp;
<a name="l00209"></a>00209             !(alleles[i] == m_rows[row][REFERENCE_INDEX]) &amp;&amp;
<a name="l00210"></a>00210             !(pcrecpp::RE(alleles[i]).PartialMatch(<span class="stringliteral">&quot;((?:^)|(?:,))&quot;</span> + m_rows[row][VARIANT_INDEX] + <span class="stringliteral">&quot;(?:(?:,)|(?:$))&quot;</span>))) {
<a name="l00211"></a>00211           okRow = <span class="keyword">false</span>;
<a name="l00212"></a>00212           invalidSStr &lt;&lt; <span class="stringliteral">&quot; | basecall &quot;</span> &lt;&lt; m_rows[row][BASECALL_INDEX] &lt;&lt; <span class="stringliteral">&quot; has alleles not found in either the reference {&quot;</span> &lt;&lt; m_rows[row][REFERENCE_INDEX] &lt;&lt;  <span class="stringliteral">&quot;} or variant list {&quot;</span> &lt;&lt;  m_rows[row][VARIANT_INDEX] &lt;&lt; <span class="stringliteral">&quot;} &quot;</span>;
<a name="l00213"></a>00213         }
<a name="l00214"></a>00214       }
<a name="l00215"></a>00215     }
<a name="l00216"></a>00216 
<a name="l00217"></a>00217 
<a name="l00218"></a>00218     <span class="comment">// TRANSLATION TABLE VALIDATION</span>
<a name="l00219"></a>00219     <span class="keywordflow">if</span> (okRow) {
<a name="l00220"></a>00220       <span class="keywordtype">bool</span> geneValid     = <span class="keyword">false</span>;
<a name="l00221"></a>00221       <span class="keywordtype">bool</span> variantValid  = <span class="keyword">false</span>;
<a name="l00222"></a>00222       <span class="keywordtype">bool</span> referenceValid = <span class="keyword">false</span>;
<a name="l00223"></a>00223       std::string rowGene;
<a name="l00224"></a>00224       std::string rowReference;
<a name="l00225"></a>00225       std::string rowVariants;
<a name="l00226"></a>00226 
<a name="l00227"></a>00227       <span class="keywordflow">if</span> (ttm.m_probeSetRowIndex.count(probeSet) == 0) {
<a name="l00228"></a>00228         okRow = <span class="keyword">false</span>;
<a name="l00229"></a>00229         okRowProbeSet = <span class="keyword">false</span>;
<a name="l00230"></a>00230         invalidSStr &lt;&lt; <span class="stringliteral">&quot; | &quot;</span>  &lt;&lt; rte.m_adtOpts.m_inputTTableFile &lt;&lt; <span class="stringliteral">&quot; does not contain probe set(&quot;</span> &lt;&lt; probeSet &lt;&lt; <span class="stringliteral">&quot;) &quot;</span>;
<a name="l00231"></a>00231       } <span class="keywordflow">else</span> {
<a name="l00232"></a>00232         <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0 ; i &lt; ttm.m_probeSetRowIndex[probeSet].size(); i++) {
<a name="l00233"></a>00233 
<a name="l00234"></a>00234           <span class="keywordtype">int</span> ttRow = ttm.m_probeSetRowIndex[probeSet][i];
<a name="l00235"></a>00235           <span class="keywordflow">if</span> (m_rows[row][GENE_INDEX] == ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_GENE)]) {
<a name="l00236"></a>00236             geneValid = <span class="keyword">true</span>;
<a name="l00237"></a>00237           }
<a name="l00238"></a>00238           rowGene = ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_GENE)];
<a name="l00239"></a>00239           <span class="keywordflow">if</span> (m_rows[row][REFERENCE_INDEX] == ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_REFERENCE)]) {
<a name="l00240"></a>00240             referenceValid = <span class="keyword">true</span>;
<a name="l00241"></a>00241           }
<a name="l00242"></a>00242           rowReference = ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_REFERENCE)];
<a name="l00243"></a>00243           <span class="keywordflow">if</span> (pcrecpp::RE(ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_VARIANT)]).PartialMatch(<span class="stringliteral">&quot;((?:^)|(?:,))&quot;</span> + m_rows[row][VARIANT_INDEX] + <span class="stringliteral">&quot;(?:(?:,)|(?:$))&quot;</span>)) {
<a name="l00244"></a>00244             variantValid = <span class="keyword">true</span>;
<a name="l00245"></a>00245           }
<a name="l00246"></a>00246           <span class="keywordflow">if</span> (rowVariants.empty()) {
<a name="l00247"></a>00247             rowVariants = ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_VARIANT)];
<a name="l00248"></a>00248           } <span class="keywordflow">else</span> {
<a name="l00249"></a>00249             rowVariants = rowVariants + <span class="stringliteral">&quot;,&quot;</span> + ttm.m_rows[ttRow][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_VARIANT)];
<a name="l00250"></a>00250           }
<a name="l00251"></a>00251         } <span class="comment">// for each allele of a possible multiallelic probe set.</span>
<a name="l00252"></a>00252 
<a name="l00253"></a>00253         <span class="keywordflow">if</span> (!geneValid) {
<a name="l00254"></a>00254           okRow = <span class="keyword">false</span>;
<a name="l00255"></a>00255           invalidSStr &lt;&lt; <span class="stringliteral">&quot; | &quot;</span>  &lt;&lt; rte.m_adtOpts.m_inputTTableFile &lt;&lt; <span class="stringliteral">&quot; probe set &quot;</span> &lt;&lt; probeSet &lt;&lt; <span class="stringliteral">&quot; gene &quot;</span> &lt;&lt; rowGene &lt;&lt; <span class="stringliteral">&quot; does not match (&quot;</span> &lt;&lt; m_rows[row][GENE_INDEX] &lt;&lt; <span class="stringliteral">&quot;) &quot;</span>;
<a name="l00256"></a>00256 
<a name="l00257"></a>00257         }
<a name="l00258"></a>00258         <span class="keywordflow">if</span> (!referenceValid) {
<a name="l00259"></a>00259           okRow = <span class="keyword">false</span>;
<a name="l00260"></a>00260           invalidSStr &lt;&lt; <span class="stringliteral">&quot; | &quot;</span>  &lt;&lt; rte.m_adtOpts.m_inputTTableFile &lt;&lt; <span class="stringliteral">&quot; probe set &quot;</span> &lt;&lt; probeSet &lt;&lt; <span class="stringliteral">&quot; reference &quot;</span> &lt;&lt; rowReference &lt;&lt; <span class="stringliteral">&quot; does not match (&quot;</span> &lt;&lt; m_rows[row][REFERENCE_INDEX] &lt;&lt; <span class="stringliteral">&quot;) &quot;</span>;
<a name="l00261"></a>00261 
<a name="l00262"></a>00262         }
<a name="l00263"></a>00263         <span class="keywordflow">if</span> (!variantValid) {
<a name="l00264"></a>00264           okRow = <span class="keyword">false</span>;
<a name="l00265"></a>00265           invalidSStr &lt;&lt; <span class="stringliteral">&quot; | &quot;</span>  &lt;&lt; rte.m_adtOpts.m_inputTTableFile &lt;&lt; <span class="stringliteral">&quot; probe set &quot;</span> &lt;&lt; probeSet &lt;&lt; <span class="stringliteral">&quot; variant set {&quot;</span> &lt;&lt; rowVariants &lt;&lt; <span class="stringliteral">&quot;} does not contain (&quot;</span> &lt;&lt; m_rows[row][REFERENCE_INDEX] &lt;&lt; <span class="stringliteral">&quot;) &quot;</span>;
<a name="l00266"></a>00266 
<a name="l00267"></a>00267         }
<a name="l00268"></a>00268 
<a name="l00269"></a>00269         <span class="keywordflow">if</span> ( okRow &amp;&amp; alleles.size() &amp;&amp; zeroCopyRE.FullMatch( alleles[0] ) &amp;&amp;
<a name="l00270"></a>00270              !(ttm.<a class="code" href="classTranslationTableModel.html#a9748ec04f92938fdd4e04bf685d86186" title="TranslationTableModel::getCopyNumberColumn: Synopsis:">getCopyNumberColumn</a>(ttm.m_probeSetRowIndex[probeSet][0]) &gt; 0) ) {
<a name="l00271"></a>00271             okRow = <span class="keyword">false</span>;
<a name="l00272"></a>00272             invalidSStr &lt;&lt; <span class="stringliteral">&quot; | &quot;</span>  &lt;&lt; <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(chpFile)
<a name="l00273"></a>00273                         &lt;&lt; <span class="stringliteral">&quot; gene probe set &quot;</span> &lt;&lt; gene &lt;&lt; <span class="stringliteral">&quot;, &quot;</span> &lt;&lt; probeSet 
<a name="l00274"></a>00274                         &lt;&lt; <span class="stringliteral">&quot; given ZeroCopyNumber override but has no such copy number zero designation in the translation table.&quot;</span>;
<a name="l00275"></a>00275         }
<a name="l00276"></a>00276       }
<a name="l00277"></a>00277     }
<a name="l00278"></a>00278 
<a name="l00279"></a>00279     <span class="keywordflow">if</span> (okRow) {
<a name="l00280"></a>00280       m_chpFileProbeSetToRowIndex[ key ] = row;
<a name="l00281"></a>00281       <span class="keywordflow">if</span> ( zeroCopyRE.FullMatch( alleles[0] ) &amp;&amp; !copyNumberZeroGenes.count(geneKey) ) {
<a name="l00282"></a>00282         copyNumberZeroGenes.insert(geneKey);
<a name="l00283"></a>00283       }
<a name="l00284"></a>00284     } <span class="keywordflow">else</span> {
<a name="l00285"></a>00285       <span class="keywordflow">if</span> (okRowProbeSet) {
<a name="l00286"></a>00286 
<a name="l00287"></a>00287         m_chpFileProbeSetInvalidBasecallComment[key].push_back(m_rows[row][BASECALL_INDEX]);
<a name="l00288"></a>00288         m_chpFileProbeSetInvalidBasecallComment[key].push_back(invalidSStr.str());
<a name="l00289"></a>00289         
<a name="l00290"></a>00290       }
<a name="l00291"></a>00291 
<a name="l00292"></a>00292       invalidSStr &lt;&lt; <span class="stringliteral">&quot; ROW {&quot;</span> &lt;&lt; <a class="code" href="classTranslationInputTsvTableModel.html#ac209f9165791ba0d30aa905965eba1f6" title="TranslationInputTsvTableModel::getRowAsString Synopsis:">getRowAsString</a>(row) &lt;&lt; <span class="stringliteral">&quot;}&quot;</span>;
<a name="l00293"></a>00293 
<a name="l00294"></a>00294       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, invalidSStr.str());
<a name="l00295"></a>00295     }
<a name="l00296"></a>00296 
<a name="l00297"></a>00297   } <span class="comment">// for each row</span>
<a name="l00298"></a>00298 
<a name="l00299"></a>00299   <span class="keywordflow">if</span> ( copyNumberZeroGenes.size() == 0 ) {
<a name="l00300"></a>00300     <span class="keywordflow">return</span>;
<a name="l00301"></a>00301   }
<a name="l00302"></a>00302 
<a name="l00303"></a>00303   <span class="comment">// ZERO COPY NUMBER VALIDATION</span>
<a name="l00304"></a>00304   
<a name="l00305"></a>00305   <span class="comment">// Validate that all copy number zero genes have the same set of markers</span>
<a name="l00306"></a>00306   <span class="comment">// as the translation table.</span>
<a name="l00307"></a>00307   <span class="comment">// Note there is an important order of occurence here.</span>
<a name="l00308"></a>00308   <span class="comment">// The marker list which filters the translation table needs to be applied</span>
<a name="l00309"></a>00309   <span class="comment">// to the translation table before this constructor is called.</span>
<a name="l00310"></a>00310   <span class="comment">// The m_probeSetIndex is maintained to only have the filtered probe sets</span>
<a name="l00311"></a>00311   <span class="comment">// and not the complete set of probe sets. The translation table contains</span>
<a name="l00312"></a>00312   <span class="comment">// all probe sets where the filtered probe sets have been ignored or</span>
<a name="l00313"></a>00313   <span class="comment">// commented out.</span>
<a name="l00314"></a>00314 
<a name="l00315"></a>00315   std::map&lt; std::string, std::string &gt; validCopyNumberZeroCheck;
<a name="l00316"></a>00316   <span class="keywordtype">bool</span> invalidCopyNumberZeroGenesExist = <span class="keyword">false</span>;
<a name="l00317"></a>00317 
<a name="l00318"></a>00318   <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> row = 0; row &lt; m_rows.size(); row++) {
<a name="l00319"></a>00319 
<a name="l00320"></a>00320     std::string probeSet    = m_rows[row][PROBE_SET_INDEX];
<a name="l00321"></a>00321     std::string gene        = m_rows[row][GENE_INDEX];
<a name="l00322"></a>00322     std::string chpFile     = m_rows[row][CHP_FILE_INDEX] ;
<a name="l00323"></a>00323     std::string key         = chpFile + <span class="stringliteral">&quot;::&quot;</span> + probeSet;
<a name="l00324"></a>00324     std::string geneKey     = chpFile + <span class="stringliteral">&quot;::&quot;</span> + gene;
<a name="l00325"></a>00325 
<a name="l00326"></a>00326     <span class="keywordflow">if</span> ( ! copyNumberZeroGenes.count(geneKey) || ! m_chpFileProbeSetToRowIndex.count( key ) || validCopyNumberZeroCheck.count(geneKey) ) {
<a name="l00327"></a>00327       <span class="keywordflow">continue</span>;
<a name="l00328"></a>00328     }
<a name="l00329"></a>00329 
<a name="l00330"></a>00330     alleles = <a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">getOverrideAlleles</a>(m_rows[row][CHP_FILE_INDEX], m_rows[row][PROBE_SET_INDEX], row);
<a name="l00331"></a>00331 
<a name="l00332"></a>00332     <span class="keywordflow">if</span> ( !zeroCopyValidSetRE.FullMatch(alleles[0]) ) {
<a name="l00333"></a>00333         invalidSStr.str(<span class="stringliteral">&quot;&quot;</span>);
<a name="l00334"></a>00334         invalidSStr &lt;&lt; <span class="stringliteral">&quot;Invalid basecall override ignored | &quot;</span>;
<a name="l00335"></a>00335         invalidSStr &lt;&lt; <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(chpFile) &lt;&lt; <span class="stringliteral">&quot; gene &quot;</span> &lt;&lt; gene &lt;&lt; <span class="stringliteral">&quot; probe set &quot;</span> &lt;&lt; probeSet &lt;&lt; <span class="stringliteral">&quot; is not ZeroCopyNumber, but &quot;</span> &lt;&lt; m_rows[row][BASECALL_INDEX] &lt;&lt; <span class="stringliteral">&quot;. ZeroCopyNumber override requires all probe sets to be present and ZeroCopyNumber basecall.&quot;</span> ;
<a name="l00336"></a>00336         validCopyNumberZeroCheck[geneKey] = invalidSStr.str();
<a name="l00337"></a>00337         invalidCopyNumberZeroGenesExist = <span class="keyword">true</span>;
<a name="l00338"></a>00338         <span class="keywordflow">break</span>;
<a name="l00339"></a>00339 
<a name="l00340"></a>00340     }
<a name="l00341"></a>00341 
<a name="l00342"></a>00342     <span class="keywordflow">if</span> ( geneMarkers[geneKey].size() &lt; ttm.m_geneMarkers[gene].size() ) {
<a name="l00343"></a>00343       invalidSStr.str(<span class="stringliteral">&quot;&quot;</span>);
<a name="l00344"></a>00344       invalidSStr &lt;&lt; <span class="stringliteral">&quot;Invalid basecall override ignored | &quot;</span>;
<a name="l00345"></a>00345       invalidSStr &lt;&lt; <span class="stringliteral">&quot; Gene &quot;</span> &lt;&lt; gene &lt;&lt; <span class="stringliteral">&quot; ZeroCopyNumber requires &quot;</span> &lt;&lt; ttm.m_geneMarkers[gene].size() &lt;&lt; <span class="stringliteral">&quot; probe set markers but only &quot;</span> &lt;&lt; geneMarkers[geneKey].size() &lt;&lt; <span class="stringliteral">&quot; probe set markers were provided.&quot;</span>;
<a name="l00346"></a>00346       validCopyNumberZeroCheck[geneKey] = invalidSStr.str();
<a name="l00347"></a>00347       invalidCopyNumberZeroGenesExist = <span class="keyword">true</span>;
<a name="l00348"></a>00348       <span class="keywordflow">continue</span>;
<a name="l00349"></a>00349     }      
<a name="l00350"></a>00350 
<a name="l00351"></a>00351     
<a name="l00352"></a>00352     <span class="keywordflow">if</span> ( ! validCopyNumberZeroCheck.count(geneKey) ) {
<a name="l00353"></a>00353       <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; ttm.m_geneMarkers[gene].size(); i++ ) {
<a name="l00354"></a>00354         <span class="keywordflow">if</span> ( ! geneMarkers[geneKey].count( ttm.m_geneMarkers[gene][i] ) ) {
<a name="l00355"></a>00355           invalidSStr.str(<span class="stringliteral">&quot;&quot;</span>);
<a name="l00356"></a>00356           invalidSStr &lt;&lt; <span class="stringliteral">&quot;Invalid basecall override ignored | &quot;</span>;
<a name="l00357"></a>00357           invalidSStr &lt;&lt; <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(chpFile) &lt;&lt; <span class="stringliteral">&quot; gene &quot;</span> &lt;&lt; gene &lt;&lt; <span class="stringliteral">&quot; ZeroCopyNumber markers provided differ from the translation table where &quot;</span> &lt;&lt; ttm.m_geneMarkers[gene][i] &lt;&lt; <span class="stringliteral">&quot; was not found.&quot;</span>;
<a name="l00358"></a>00358           validCopyNumberZeroCheck[geneKey] = invalidSStr.str();
<a name="l00359"></a>00359           invalidCopyNumberZeroGenesExist = <span class="keyword">true</span>;
<a name="l00360"></a>00360         }
<a name="l00361"></a>00361       }
<a name="l00362"></a>00362     }
<a name="l00363"></a>00363 
<a name="l00364"></a>00364   }
<a name="l00365"></a>00365 
<a name="l00366"></a>00366   <span class="comment">// Remove the overrides from the valid list and set up the invalid</span>
<a name="l00367"></a>00367   <span class="comment">// message.</span>
<a name="l00368"></a>00368   
<a name="l00369"></a>00369   <span class="keywordflow">if</span> ( invalidCopyNumberZeroGenesExist ) {
<a name="l00370"></a>00370 
<a name="l00371"></a>00371     <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> row = 0; row &lt; m_rows.size(); row++) {
<a name="l00372"></a>00372 
<a name="l00373"></a>00373       std::string probeSet    = m_rows[row][PROBE_SET_INDEX];
<a name="l00374"></a>00374       std::string gene        = m_rows[row][GENE_INDEX];
<a name="l00375"></a>00375       std::string chpFile     = m_rows[row][CHP_FILE_INDEX] ;
<a name="l00376"></a>00376       std::string key         = chpFile + <span class="stringliteral">&quot;::&quot;</span> + probeSet;
<a name="l00377"></a>00377       std::string geneKey     = chpFile + <span class="stringliteral">&quot;::&quot;</span> + gene;
<a name="l00378"></a>00378 
<a name="l00379"></a>00379       <span class="keywordflow">if</span> ( validCopyNumberZeroCheck.count(geneKey) &amp;&amp;
<a name="l00380"></a>00380            (validCopyNumberZeroCheck[geneKey] != std::string(<span class="stringliteral">&quot;VALID&quot;</span>)) ) {
<a name="l00381"></a>00381 
<a name="l00382"></a>00382         <span class="keywordflow">if</span> ( m_chpFileProbeSetToRowIndex.count(key)) {
<a name="l00383"></a>00383           m_chpFileProbeSetToRowIndex.erase( key );
<a name="l00384"></a>00384           m_chpFileProbeSetInvalidBasecallComment[key].push_back(m_rows[row][BASECALL_INDEX]);
<a name="l00385"></a>00385           m_chpFileProbeSetInvalidBasecallComment[key].push_back(validCopyNumberZeroCheck[geneKey]);
<a name="l00386"></a>00386         }
<a name="l00387"></a>00387       }
<a name="l00388"></a>00388     }
<a name="l00389"></a>00389 
<a name="l00390"></a>00390   }
<a name="l00391"></a>00391 
<a name="l00392"></a>00392   <span class="keywordflow">return</span>;
<a name="l00393"></a>00393 
<a name="l00394"></a>00394 }
<a name="l00395"></a>00395 <span class="comment">// end GenotypeOverrideTableModel::GenotypeOverrideTableModel</span>
<a name="l00396"></a>00396 <span class="comment">/*****************************************************************************/</span>
<a name="l00397"></a>00397 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00398"></a>00398 <span class="comment">/**</span>
<a name="l00399"></a>00399 <span class="comment"> * GenotypeOverrideTableModel::describeVerbose:</span>
<a name="l00400"></a>00400 <span class="comment"> * Synopsis:</span>
<a name="l00401"></a>00401 <span class="comment"> *</span>
<a name="l00402"></a>00402 <span class="comment"> * Run time debug routine that can be invoked with various verbose levels.</span>
<a name="l00403"></a>00403 <span class="comment"> * ADT_VERBOSE_INPUT_FILES is the default level.</span>
<a name="l00404"></a>00404 <span class="comment"> *</span>
<a name="l00405"></a>00405 <span class="comment"> * @param rte       - run time environment</span>
<a name="l00406"></a>00406 <span class="comment"> * @param override  - ADT_VERBOSE_INPUT_FILES by default,</span>
<a name="l00407"></a>00407 <span class="comment"> *                    otherwise the ADT_VERBOSE_ENUM level to output</span>
<a name="l00408"></a>00408 <span class="comment"> *</span>
<a name="l00409"></a>00409 <span class="comment"> */</span>
<a name="l00410"></a>00410 
<a name="l00411"></a>00411 <span class="comment">/*****************************************************************************/</span>
<a name="l00412"></a><a class="code" href="classGenotypeOverrideTableModel.html#a398fdbc859f10a53f23a51adb7cf5882">00412</a> <span class="keywordtype">void</span> <a class="code" href="classGenotypeOverrideTableModel.html#a398fdbc859f10a53f23a51adb7cf5882" title="GenotypeOverrideTableModel::describeVerbose: Synopsis:">GenotypeOverrideTableModel::describeVerbose</a>(ADT_VERBOSE_ENUM level)
<a name="l00413"></a>00413 {
<a name="l00414"></a>00414 
<a name="l00415"></a>00415   std::stringstream msgSStr;
<a name="l00416"></a>00416 
<a name="l00417"></a>00417   msgSStr &lt;&lt; <span class="stringliteral">&quot;GenotypeOverrideTableModel::describeVerbose&quot;</span> &lt;&lt; endl;
<a name="l00418"></a>00418 
<a name="l00419"></a>00419   msgSStr &lt;&lt; <span class="stringliteral">&quot;m_chpFileProbeSetToRowIndex:&quot;</span> &lt;&lt; endl;
<a name="l00420"></a>00420 
<a name="l00421"></a>00421   std::map&lt; std::string, int &gt;::iterator itSI = m_chpFileProbeSetToRowIndex.begin();
<a name="l00422"></a>00422 
<a name="l00423"></a>00423   <span class="keywordflow">for</span> (; itSI != m_chpFileProbeSetToRowIndex.end(); itSI++) {
<a name="l00424"></a>00424 
<a name="l00425"></a>00425     msgSStr &lt;&lt; itSI-&gt;first &lt;&lt; <span class="stringliteral">&quot; row [&quot;</span> &lt;&lt; itSI-&gt;second &lt;&lt; <span class="stringliteral">&quot;]&quot;</span> &lt;&lt; endl;
<a name="l00426"></a>00426 
<a name="l00427"></a>00427   }
<a name="l00428"></a>00428 
<a name="l00429"></a>00429   std::map&lt; std::string, std::string &gt;::iterator itSS = m_chpFileProbeSetOriginalBasecall.begin();
<a name="l00430"></a>00430 
<a name="l00431"></a>00431   msgSStr &lt;&lt; <span class="stringliteral">&quot;m_chpFileProbeSetOriginalBasecall:&quot;</span> &lt;&lt; endl;
<a name="l00432"></a>00432   <span class="keywordflow">for</span> (; itSS != m_chpFileProbeSetOriginalBasecall.end(); itSS++) {
<a name="l00433"></a>00433 
<a name="l00434"></a>00434     msgSStr &lt;&lt; itSS-&gt;first &lt;&lt; <span class="stringliteral">&quot; : &quot;</span> &lt;&lt; itSS-&gt;second &lt;&lt; endl;
<a name="l00435"></a>00435 
<a name="l00436"></a>00436   }
<a name="l00437"></a>00437 
<a name="l00438"></a>00438 
<a name="l00439"></a>00439   msgSStr &lt;&lt; <span class="stringliteral">&quot;m_chpFileProbeSetInvalidBasecallComment:&quot;</span> &lt;&lt; endl;
<a name="l00440"></a>00440 
<a name="l00441"></a>00441   std::map&lt; std::string,  std::vector&lt; std::string &gt; &gt;::iterator itSVS
<a name="l00442"></a>00442   = m_chpFileProbeSetInvalidBasecallComment.begin();
<a name="l00443"></a>00443 
<a name="l00444"></a>00444   <span class="keywordflow">for</span> (; itSVS != m_chpFileProbeSetInvalidBasecallComment.end(); itSVS++) {
<a name="l00445"></a>00445 
<a name="l00446"></a>00446     msgSStr &lt;&lt; itSVS-&gt;first &lt;&lt; <span class="stringliteral">&quot; : &quot;</span>;
<a name="l00447"></a>00447     <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; itSVS-&gt;second.size(); i++) {
<a name="l00448"></a>00448       msgSStr &lt;&lt; itSVS-&gt;second[i] &lt;&lt; <span class="stringliteral">&quot;, &quot;</span>;
<a name="l00449"></a>00449     }
<a name="l00450"></a>00450 
<a name="l00451"></a>00451     msgSStr &lt;&lt; endl;
<a name="l00452"></a>00452   }
<a name="l00453"></a>00453 
<a name="l00454"></a>00454   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(level, msgSStr.str());
<a name="l00455"></a>00455 
<a name="l00456"></a>00456 }
<a name="l00457"></a>00457 <span class="comment">// end GenotypeOverrideTableModel::describeVerbose</span>
<a name="l00458"></a>00458 <span class="comment">/*****************************************************************************/</span>
<a name="l00459"></a>00459 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00460"></a>00460 <span class="comment">/**</span>
<a name="l00461"></a>00461 <span class="comment"> * GenotypeOverrideTableModel::getOverrideAlleles:</span>
<a name="l00462"></a>00462 <span class="comment"> * Synopsis:</span>
<a name="l00463"></a>00463 <span class="comment"> *</span>
<a name="l00464"></a>00464 <span class="comment"> *    If an override exists, returns a non-empty std::vector pair of std::strings</span>
<a name="l00465"></a>00465 <span class="comment"> * corresponding to the override basecall alleles.</span>
<a name="l00466"></a>00466 <span class="comment"> *</span>
<a name="l00467"></a>00467 <span class="comment"> * NOTE: Override files only have &quot;NoCall&quot;, &quot;PossibleRareAllele&quot; or &quot;</span>
<a name="l00468"></a>00468 <span class="comment"> * &quot;NotAvailable&quot; and not the format of &quot;NoCall/NoCall&quot; as</span>
<a name="l00469"></a>00469 <span class="comment"> * would be returned by the GenoCallCoder. Therefore the return result</span>
<a name="l00470"></a>00470 <span class="comment"> * for wildcard overrides are reformated to the GenoCallCoder format</span>
<a name="l00471"></a>00471 <span class="comment"> * (&quot;NoCall/NoCall&quot;).</span>
<a name="l00472"></a>00472 <span class="comment"> *</span>
<a name="l00473"></a>00473 <span class="comment"> *</span>
<a name="l00474"></a>00474 <span class="comment"> *</span>
<a name="l00475"></a>00475 <span class="comment"> * @param chpFileName - experiment chp file name (minus .CHP)</span>
<a name="l00476"></a>00476 <span class="comment"> * @param probeSetId  - probeset id in question.</span>
<a name="l00477"></a>00477 <span class="comment"> * @param row         - the row in &quot;m_rows&quot;.</span>
<a name="l00478"></a>00478 <span class="comment"> *</span>
<a name="l00479"></a>00479 <span class="comment"> * @return - alleles as a std::string pair inside a std::vector.</span>
<a name="l00480"></a>00480 <span class="comment"> */</span>
<a name="l00481"></a>00481 <span class="comment">/*****************************************************************************/</span>
<a name="l00482"></a><a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5">00482</a> std::vector&lt; std::string &gt; <a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">GenotypeOverrideTableModel::getOverrideAlleles</a>(<span class="keyword">const</span> std::string &amp; chpFileName, <span class="keyword">const</span> std::string &amp; probeSetId, <span class="keywordtype">int</span> row)
<a name="l00483"></a>00483 {
<a name="l00484"></a>00484 
<a name="l00485"></a>00485   std::vector&lt; std::string &gt; results;
<a name="l00486"></a>00486 
<a name="l00487"></a>00487   <span class="keywordflow">if</span> (chpFileName.empty() || probeSetId.empty()) {
<a name="l00488"></a>00488     <span class="keywordflow">return</span> results;
<a name="l00489"></a>00489   }
<a name="l00490"></a>00490   <span class="keywordtype">bool</span> printResults = <span class="keyword">false</span>;
<a name="l00491"></a>00491   
<a name="l00492"></a>00492   <span class="keywordflow">if</span> (row &lt; 0) {
<a name="l00493"></a>00493 
<a name="l00494"></a>00494     printResults = <span class="keyword">true</span>;
<a name="l00495"></a>00495     std::string key = chpFileName + <span class="stringliteral">&quot;::&quot;</span> + probeSetId;
<a name="l00496"></a>00496 
<a name="l00497"></a>00497     std::map&lt; std::string , int &gt;::const_iterator itSI;
<a name="l00498"></a>00498     itSI = m_chpFileProbeSetToRowIndex.find(key) ;
<a name="l00499"></a>00499 
<a name="l00500"></a>00500     <span class="keywordflow">if</span> (itSI == m_chpFileProbeSetToRowIndex.end()) {
<a name="l00501"></a>00501       <span class="keywordflow">return</span> results;
<a name="l00502"></a>00502     }
<a name="l00503"></a>00503     row = itSI-&gt;second;
<a name="l00504"></a>00504   }
<a name="l00505"></a>00505 
<a name="l00506"></a>00506   std::string basecall = m_rows[row][BASECALL_INDEX];
<a name="l00507"></a>00507 
<a name="l00508"></a>00508   pcrecpp::RE split(<span class="stringliteral">&quot;([^/]+)/([^/]+)&quot;</span>);
<a name="l00509"></a>00509 
<a name="l00510"></a>00510   std::string allele1;
<a name="l00511"></a>00511   std::string allele2;
<a name="l00512"></a>00512 
<a name="l00513"></a>00513   <span class="keywordflow">if</span> (split.FullMatch(basecall, &amp;allele1, &amp;allele2)) {
<a name="l00514"></a>00514     results.push_back(allele1);
<a name="l00515"></a>00515     results.push_back(allele2);
<a name="l00516"></a>00516   }
<a name="l00517"></a>00517   
<a name="l00518"></a>00518   <span class="keywordflow">if</span> (pcrecpp::RE(<span class="stringliteral">&quot;0|ZeroCopyNumber|PossibleRareAllele|NoCall|NotAvailable|NC|PCRA|NA&quot;</span>).FullMatch(basecall)) {
<a name="l00519"></a>00519     <span class="keywordflow">if</span> ( basecall  == <span class="stringliteral">&quot;ZeroCopyNumber&quot;</span> ) {
<a name="l00520"></a>00520       basecall = <span class="stringliteral">&quot;0&quot;</span>;
<a name="l00521"></a>00521     }
<a name="l00522"></a>00522     results.push_back(basecall);
<a name="l00523"></a>00523     <span class="keywordflow">if</span> ( basecall != <span class="stringliteral">&quot;0&quot;</span> ) {
<a name="l00524"></a>00524       results.push_back(basecall);
<a name="l00525"></a>00525     }
<a name="l00526"></a>00526   }
<a name="l00527"></a>00527 
<a name="l00528"></a>00528 
<a name="l00529"></a>00529   <span class="keywordflow">return</span> results;
<a name="l00530"></a>00530 
<a name="l00531"></a>00531 }
<a name="l00532"></a>00532 <span class="comment">// end GenotypeOverrideTableModel::getOverrideAlleles</span>
<a name="l00533"></a>00533 <span class="comment">/*****************************************************************************/</span>
<a name="l00534"></a>00534 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00535"></a>00535 <span class="comment">/**</span>
<a name="l00536"></a>00536 <span class="comment"> * GenotypeOverrideTableModel::getOverrideInfo:</span>
<a name="l00537"></a>00537 <span class="comment"> * Synopsis:</span>
<a name="l00538"></a>00538 <span class="comment"> *</span>
<a name="l00539"></a>00539 <span class="comment"> *    If an override exists, returns a non-empty std::vector PAIR of std::strings</span>
<a name="l00540"></a>00540 <span class="comment"> * corresponding to the override basecall original basecall and the</span>
<a name="l00541"></a>00541 <span class="comment"> * override comment (override basecall, override comment).</span>
<a name="l00542"></a>00542 <span class="comment"> *</span>
<a name="l00543"></a>00543 <span class="comment"> *</span>
<a name="l00544"></a>00544 <span class="comment"> *</span>
<a name="l00545"></a>00545 <span class="comment"> * @param chpFileName - experiment chp file name (minus .CHP)</span>
<a name="l00546"></a>00546 <span class="comment"> * @param probeSetId  - assay id in question.</span>
<a name="l00547"></a>00547 <span class="comment"> * @param includeInvalidOriginalBaseCall - boolean that returns that includes the original base call, even for invalid calls. </span>
<a name="l00548"></a>00548 <span class="comment"> *</span>
<a name="l00549"></a>00549 <span class="comment"> * @return - (original basecall, commment) as a std::vector pair of std::strings.</span>
<a name="l00550"></a>00550 <span class="comment"> */</span>
<a name="l00551"></a>00551 <span class="comment">/*****************************************************************************/</span>
<a name="l00552"></a><a class="code" href="classGenotypeOverrideTableModel.html#aa9a3cafd9eca1946d0fae5d666a7b590">00552</a> std::vector&lt; std::string &gt; <a class="code" href="classGenotypeOverrideTableModel.html#aa9a3cafd9eca1946d0fae5d666a7b590" title="GenotypeOverrideTableModel::getOverrideInfo: Synopsis:">GenotypeOverrideTableModel::getOverrideInfo</a>(<span class="keyword">const</span> std::string &amp; chpFileName, <span class="keyword">const</span> std::string &amp; probeSetId, <span class="keywordtype">bool</span> includeInvalidOriginalBaseCall )
<a name="l00553"></a>00553 {
<a name="l00554"></a>00554 
<a name="l00555"></a>00555   std::vector&lt; std::string &gt; results;
<a name="l00556"></a>00556 
<a name="l00557"></a>00557   <span class="keywordflow">if</span> (chpFileName.empty() || probeSetId.empty()) {
<a name="l00558"></a>00558     <span class="keywordflow">return</span> results;
<a name="l00559"></a>00559   }
<a name="l00560"></a>00560 
<a name="l00561"></a>00561   std::string key = chpFileName + <span class="stringliteral">&quot;::&quot;</span> + probeSetId;
<a name="l00562"></a>00562 
<a name="l00563"></a>00563   std::map&lt; std::string , int &gt;::const_iterator itSI;
<a name="l00564"></a>00564 
<a name="l00565"></a>00565   itSI = m_chpFileProbeSetToRowIndex.find(key) ;
<a name="l00566"></a>00566 
<a name="l00567"></a>00567 
<a name="l00568"></a>00568   <span class="keywordflow">if</span> (itSI == m_chpFileProbeSetToRowIndex.end()) {
<a name="l00569"></a>00569 
<a name="l00570"></a>00570     <span class="comment">// INVALID uncalled file row check.</span>
<a name="l00571"></a>00571     std::map&lt; std::string , std::vector&lt; std::string &gt;  &gt;::const_iterator jtSVS;
<a name="l00572"></a>00572     jtSVS =  m_chpFileProbeSetInvalidBasecallComment.find(key);
<a name="l00573"></a>00573 
<a name="l00574"></a>00574     <span class="keywordflow">if</span> (jtSVS == m_chpFileProbeSetInvalidBasecallComment.end()) {
<a name="l00575"></a>00575       <span class="comment">// COMMON CASE, most probe set rows will not be in the uncalled file.</span>
<a name="l00576"></a>00576       <span class="keywordflow">return</span> results;
<a name="l00577"></a>00577     }
<a name="l00578"></a>00578 
<a name="l00579"></a>00579     <span class="comment">// ROW was INVALID</span>
<a name="l00580"></a>00580     <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(jtSVS-&gt;second.size() == 2, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00581"></a>00581 
<a name="l00582"></a>00582     <span class="keywordflow">if</span> ( includeInvalidOriginalBaseCall ) {
<a name="l00583"></a>00583       results.push_back(jtSVS-&gt;second[0]);
<a name="l00584"></a>00584     }
<a name="l00585"></a>00585     <span class="keywordflow">else</span> {
<a name="l00586"></a>00586       results.push_back(<span class="stringliteral">&quot;&quot;</span>);
<a name="l00587"></a>00587     }
<a name="l00588"></a>00588 
<a name="l00589"></a>00589     results.push_back(jtSVS-&gt;second[1]);
<a name="l00590"></a>00590 
<a name="l00591"></a>00591     <span class="keywordflow">return</span> results;
<a name="l00592"></a>00592   }
<a name="l00593"></a>00593 
<a name="l00594"></a>00594   std::string comment = m_rows[itSI-&gt;second][OVERRIDE_COMMENT_INDEX];
<a name="l00595"></a>00595 
<a name="l00596"></a>00596   std::map&lt; std::string, std::string &gt;::const_iterator itSS;
<a name="l00597"></a>00597 
<a name="l00598"></a>00598   itSS = m_chpFileProbeSetOriginalBasecall.find(key);
<a name="l00599"></a>00599 
<a name="l00600"></a>00600   <span class="keywordflow">if</span> (itSS == m_chpFileProbeSetOriginalBasecall.end()) {
<a name="l00601"></a>00601     <span class="keywordflow">return</span> results;
<a name="l00602"></a>00602   }
<a name="l00603"></a>00603 
<a name="l00604"></a>00604   std::string basecall = itSS-&gt;second;
<a name="l00605"></a>00605 
<a name="l00606"></a>00606   results.push_back(basecall);
<a name="l00607"></a>00607   results.push_back(comment);
<a name="l00608"></a>00608 
<a name="l00609"></a>00609   <span class="keywordflow">return</span> results;
<a name="l00610"></a>00610 
<a name="l00611"></a>00611 }
<a name="l00612"></a>00612 <span class="comment">// end GenotypeOverrideTableModel::getOverrideInfo</span>
<a name="l00613"></a>00613 <span class="comment">/*****************************************************************************/</span>
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:03 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
