<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/example/multiChannelExample.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
      <li><a class="el" href="dir_40aa2dab55bdb3efe170886e0a4c54fb.html">example</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>multiChannelExample.cpp</h1>  </div>
</div>
<div class="contents">
<a href="multiChannelExample_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   multiChannelExample.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Alan Williams</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Fri Jan  4 13:53:19 PST 2008</span>
<a name="l00026"></a>00026 <span class="comment"> * </span>
<a name="l00027"></a>00027 <span class="comment"> * @brief A simple example of multi channel analysis</span>
<a name="l00028"></a>00028 <span class="comment"> * Please do not use this example for real life data</span>
<a name="l00029"></a>00029 <span class="comment"> * analysis.</span>
<a name="l00030"></a>00030 <span class="comment"> */</span>
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 <span class="comment">//</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="AnalysisStreamFactory_8h.html" title="Object for making new analysis streams.">chipstream/AnalysisStreamFactory.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="BioTypes_8h.html" title="Central location enumerations and other types that aren&amp;#39;t particular to a single class or applica...">chipstream/BioTypes.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="CelReader_8h.html" title="Class for reading cel files and passing the data to memory representation and analysis objects...">chipstream/CelReader.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;<a class="code" href="DiskIntensityMart_8h.html" title="Class for encapsulating microarray data and storing it on disk in a reasonable way to reduce memory i...">chipstream/DiskIntensityMart.h</a>&quot;</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &quot;chipstream/FileGenders.h&quot;</span>
<a name="l00038"></a>00038 <span class="preprocessor">#include &quot;<a class="code" href="GenoSeedTxt_8h.html" title="Implements interface for a genotyping oracle used for genotyping seeds in brlmm. In this case the gen...">chipstream/GenoSeedTxt.h</a>&quot;</span>
<a name="l00039"></a>00039 <span class="preprocessor">#include &quot;<a class="code" href="MultiQuantMethodListener_8h.html" title="Class for reporting results from multiple quantification methods.">chipstream/MultiQuantMethodListener.h</a>&quot;</span>
<a name="l00040"></a>00040 <span class="preprocessor">#include &quot;<a class="code" href="PmOnlyAdjust_8h.html" title="Pm adjustment class that does no modification.">chipstream/PmOnlyAdjust.h</a>&quot;</span>
<a name="l00041"></a>00041 <span class="preprocessor">#include &quot;<a class="code" href="QuantLabelZ_8h.html">chipstream/QuantLabelZ.h</a>&quot;</span>
<a name="l00042"></a>00042 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodFactory_8h.html" title="Factory class for making chip streams based on a string representation.">chipstream/QuantMethodFactory.h</a>&quot;</span>
<a name="l00043"></a>00043 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodGTypeReport_8h.html" title="Reporter for outputting genotype results.">chipstream/QuantMethodGTypeReport.h</a>&quot;</span>
<a name="l00044"></a>00044 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodReportListener_8h.html" title="Listens for reporting signals from QuantMethod and then passes them off to another coordinating objec...">chipstream/QuantMethodReportListener.h</a>&quot;</span>
<a name="l00045"></a>00045 <span class="preprocessor">#include &quot;<a class="code" href="SpecialSnps_8h.html">chipstream/SpecialSnps.h</a>&quot;</span>
<a name="l00046"></a>00046 <span class="preprocessor">#include &quot;<a class="code" href="TsvFile_8h.html" title="Headers for the TsvFile classes.">file/TsvFile/TsvFile.h</a>&quot;</span>
<a name="l00047"></a>00047 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00048"></a>00048 <span class="preprocessor">#include &quot;<a class="code" href="LogStream_8h.html" title="Output verbose messages to log file as we go along.">util/LogStream.h</a>&quot;</span>
<a name="l00049"></a>00049 <span class="preprocessor">#include &quot;<a class="code" href="PgOptions_8h.html" title="Classes for reading program command line options.">util/PgOptions.h</a>&quot;</span>
<a name="l00050"></a>00050 <span class="comment">//</span>
<a name="l00051"></a>00051 <span class="preprocessor">#include &lt;cstring&gt;</span>
<a name="l00052"></a>00052 <span class="preprocessor">#include &lt;iostream&gt;</span>
<a name="l00053"></a>00053 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00054"></a>00054 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 <span class="keyword">using namespace </span>std;
<a name="l00057"></a>00057 <span class="keyword">using namespace </span>affx;
<a name="l00058"></a>00058 
<a name="l00059"></a>00059 <span class="keyword">using namespace </span>std;
<a name="l00060"></a>00060 
<a name="l00061"></a><a class="code" href="classAOptions.html">00061</a> <span class="keyword">class </span><a class="code" href="classAOptions.html">AOptions</a> {
<a name="l00062"></a>00062 <span class="keyword">public</span>:
<a name="l00063"></a>00063   <span class="keywordtype">bool</span> help;
<a name="l00064"></a>00064   <span class="keywordtype">int</span> verbose;
<a name="l00065"></a>00065   <span class="keywordtype">string</span> algoParams;
<a name="l00066"></a>00066   <span class="keywordtype">string</span> outDir;
<a name="l00067"></a>00067   <span class="keywordtype">string</span> exprAnalysisString;
<a name="l00068"></a>00068   <span class="keywordtype">string</span> chrXSnps;
<a name="l00069"></a>00069   <span class="keywordtype">string</span> specialSnps;
<a name="l00070"></a>00070   <span class="keywordtype">string</span> readModelsBrlmmp;
<a name="l00071"></a>00071   <span class="keywordtype">bool</span> writeModels;
<a name="l00072"></a>00072   <span class="keywordtype">bool</span> summaries;
<a name="l00073"></a>00073   <span class="keywordtype">bool</span> doFeatureEffects;
<a name="l00074"></a>00074   <span class="keywordtype">bool</span> doResiduals;
<a name="l00075"></a>00075   <span class="keywordtype">bool</span> selectProbes;
<a name="l00076"></a>00076   <span class="keywordtype">string</span> genderFileIn;
<a name="l00077"></a>00077   <span class="keywordtype">string</span> quantMethod;
<a name="l00078"></a><a class="code" href="classAOptions.html#a5c0b549af05cc88f1562f334db0a709a">00078</a>   std::string genoTypesFile;     <span class="comment">///&lt; File to input seed genotypes</span>
<a name="l00079"></a>00079 <span class="comment"></span>  vector&lt;string&gt; sketchInFile;
<a name="l00080"></a>00080   <span class="keywordtype">bool</span> writeSketch;
<a name="l00081"></a>00081 
<a name="l00082"></a>00082   <span class="keywordtype">string</span> spfFile;
<a name="l00083"></a>00083   <span class="keywordtype">string</span> cdfFile;
<a name="l00084"></a>00084 
<a name="l00085"></a>00085   <span class="keywordtype">int</span> channelCount;
<a name="l00086"></a>00086   vector&lt;string&gt; sampleNames;
<a name="l00087"></a><a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9">00087</a>   vector&lt;vector&lt;string&gt; &gt; celFiles; <span class="comment">///&lt; first dim channel, second is samples</span>
<a name="l00088"></a>00088 <span class="comment"></span>
<a name="l00089"></a>00089   <span class="keywordtype">string</span> channelFile;
<a name="l00090"></a>00090 };
<a name="l00091"></a>00091 
<a name="l00092"></a>00092 <span class="keywordtype">void</span> define_options(<a class="code" href="classPgOptions.html">PgOptions</a>* opts)
<a name="l00093"></a>00093 {
<a name="l00094"></a>00094   opts-&gt;<a class="code" href="classPgOptions.html#aa40409c034ecf67ad32694103782e97e" title="set the usage info for this program ///">setUsage</a>(<span class="stringliteral">&quot;multiChannelExample - example of feeding multiple channels into a\n&quot;</span>
<a name="l00095"></a>00095                  <span class="stringliteral">&quot;ChipStream for genotype calling.\n&quot;</span>
<a name="l00096"></a>00096                  <span class="stringliteral">&quot;\n&quot;</span>
<a name="l00097"></a>00097                  <span class="stringliteral">&quot;usage:\n&quot;</span>
<a name="l00098"></a>00098                  <span class="stringliteral">&quot;   ./multiChannelExample -p CM=1.bins=100.K=2.SB=0.003.MS=0.05 \\\n&quot;</span>
<a name="l00099"></a>00099                  <span class="stringliteral">&quot;       -o results \\\n&quot;</span>
<a name="l00100"></a>00100                  <span class="stringliteral">&quot;       --read-models-brlmmp GenomeWideSNP_5.models \\\n&quot;</span>
<a name="l00101"></a>00101                  <span class="stringliteral">&quot;       --chrX-snps GenomeWideSNP_5.chrx \\\n&quot;</span>
<a name="l00102"></a>00102                  <span class="stringliteral">&quot;       --read-genders genders.txt\n&quot;</span>);
<a name="l00103"></a>00103                  
<a name="l00104"></a>00104   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;h&quot;</span>, <span class="stringliteral">&quot;help&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00105"></a>00105                      <span class="stringliteral">&quot;This message.&quot;</span>,
<a name="l00106"></a>00106                      <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00107"></a>00107   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;explain&quot;</span>, PgOpt::STRING_OPT,
<a name="l00108"></a>00108                      <span class="stringliteral">&quot;Display documentation for algo method parameters.&quot;</span>,
<a name="l00109"></a>00109                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00110"></a>00110   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;v&quot;</span>, <span class="stringliteral">&quot;verbose&quot;</span>, PgOpt::INT_OPT,
<a name="l00111"></a>00111                      <span class="stringliteral">&quot;How verbose to be with status messages 0 - quiet, 1 - usual messages, 2 - more messages.&quot;</span>,
<a name="l00112"></a>00112                      <span class="stringliteral">&quot;1&quot;</span>);
<a name="l00113"></a>00113   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;cel-files&quot;</span>, PgOpt::STRING_OPT,
<a name="l00114"></a>00114                      <span class="stringliteral">&quot;File listing CEL files to process. There must be a &#39;sample_name&#39; column, and 2 or more &#39;channel_*&#39;. Current channels are ordered (starting with 0) by the order in this file.&quot;</span>,
<a name="l00115"></a>00115                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00116"></a>00116   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;d&quot;</span>, <span class="stringliteral">&quot;cdf-file&quot;</span>, PgOpt::STRING_OPT,
<a name="l00117"></a>00117                      <span class="stringliteral">&quot;CDF file providing chip layout information.&quot;</span>,
<a name="l00118"></a>00118                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00119"></a>00119   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;spf-file&quot;</span>, PgOpt::STRING_OPT,
<a name="l00120"></a>00120                      <span class="stringliteral">&quot;SPF file providing chip layout information.&quot;</span>,
<a name="l00121"></a>00121                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00122"></a>00122   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;channel-file&quot;</span>, PgOpt::STRING_OPT,
<a name="l00123"></a>00123                      <span class="stringliteral">&quot;Channel file providing info on what alleles are in what channel for each SNP. There should be a row for each SNP. There must be a &#39;probeset_id&#39;, &#39;allele_A_channel&#39;, and an &#39;allele_B_channel&#39; columns. The value corresponds to the channels in the --cel-files specified file. Valid values are integers &gt;= 0.&quot;</span>,
<a name="l00124"></a>00124                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00125"></a>00125   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;p&quot;</span>, <span class="stringliteral">&quot;algo-params&quot;</span>, PgOpt::STRING_OPT,
<a name="l00126"></a>00126                      <span class="stringliteral">&quot;A string specifying algo parameters. See --explain option for acceptable parameters.&quot;</span>,
<a name="l00127"></a>00127                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00128"></a>00128   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;a&quot;</span>, <span class="stringliteral">&quot;expr-analysis&quot;</span>, PgOpt::STRING_OPT,
<a name="l00129"></a>00129                      <span class="stringliteral">&quot;Expression analysis string.&quot;</span>,
<a name="l00130"></a>00130                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00131"></a>00131   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;o&quot;</span>, <span class="stringliteral">&quot;out-dir&quot;</span>, PgOpt::STRING_OPT,
<a name="l00132"></a>00132                      <span class="stringliteral">&quot;Directory to write result files into.&quot;</span>,
<a name="l00133"></a>00133                      <span class="stringliteral">&quot;.&quot;</span>);
<a name="l00134"></a>00134   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;version&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00135"></a>00135                      <span class="stringliteral">&quot;Display version information.&quot;</span>,
<a name="l00136"></a>00136                      <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00137"></a>00137   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;chrX-snps&quot;</span>, PgOpt::STRING_OPT,
<a name="l00138"></a>00138                      <span class="stringliteral">&quot;File containing snps on chrX (non-pseudoautosomal region).&quot;</span>,
<a name="l00139"></a>00139                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00140"></a>00140   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;special-snps&quot;</span>,PgOpt::STRING_OPT,
<a name="l00141"></a>00141                      <span class="stringliteral">&quot;File containing all snps of unusual copy (chrX,mito,Y)&quot;</span>,
<a name="l00142"></a>00142                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00143"></a>00143   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;write-models&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00144"></a>00144                      <span class="stringliteral">&quot;Should we write snp specific models out for analysis? [experimental]&quot;</span>,
<a name="l00145"></a>00145                      <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00146"></a>00146   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;summaries&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00147"></a>00147                      <span class="stringliteral">&quot;Should we write out the allele summaries&quot;</span>,
<a name="l00148"></a>00148                      <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00149"></a>00149   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;feature-effects&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00150"></a>00150                      <span class="stringliteral">&quot;Output feature effects when available.&quot;</span>,
<a name="l00151"></a>00151                      <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00152"></a>00152   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;feature-details&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00153"></a>00153                      <span class="stringliteral">&quot;Output probe by chip specific details (often residuals) when available.&quot;</span>,
<a name="l00154"></a>00154                      <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00155"></a>00155   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;read-models-brlmmp&quot;</span>, PgOpt::STRING_OPT,
<a name="l00156"></a>00156                      <span class="stringliteral">&quot;File to read precomputed BRLMM-P snp specific models from.&quot;</span>,
<a name="l00157"></a>00157                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00158"></a>00158   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;read-genders&quot;</span>, PgOpt::STRING_OPT,
<a name="l00159"></a>00159                      <span class="stringliteral">&quot;Explicitly read genders from a file.&quot;</span>,
<a name="l00160"></a>00160                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00161"></a>00161 
<a name="l00162"></a>00162   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;genotypes&quot;</span>, PgOpt::STRING_OPT,
<a name="l00163"></a>00163                      <span class="stringliteral">&quot;File to read seed genotypes from.&quot;</span>,
<a name="l00164"></a>00164                      <span class="stringliteral">&quot;&quot;</span>);
<a name="l00165"></a>00165   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>,<span class="stringliteral">&quot;select-probes&quot;</span>,PgOpt::BOOL_OPT,
<a name="l00166"></a>00166                                           <span class="stringliteral">&quot;evaluate probe/probe-set quality&quot;</span>,
<a name="l00167"></a>00167                         <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00168"></a>00168   opts-&gt;<a class="code" href="classPgOptions.html#a6b593f1d35c010fbe6ae6ec1780e5b26" title="Short for to define an option with multiple values. ///.">defOptMult</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;target-sketch&quot;</span>, PgOpt::STRING_OPT,
<a name="l00169"></a>00169                     <span class="stringliteral">&quot;File specifying a target distribution to use for quantile normalization.&quot;</span>,
<a name="l00170"></a>00170                     <span class="stringliteral">&quot;&quot;</span>);
<a name="l00171"></a>00171   opts-&gt;<a class="code" href="classPgOptions.html#a7b3fbfff8c4496985f80bcfd49a59d61" title="define an option to be handled by this option parser. ///">defineOption</a>(<span class="stringliteral">&quot;&quot;</span>, <span class="stringliteral">&quot;write-sketch&quot;</span>, PgOpt::BOOL_OPT,
<a name="l00172"></a>00172                     <span class="stringliteral">&quot;Write the quantile normalization distribution (or sketch) to a file for reuse with target-sketch option.&quot;</span>,
<a name="l00173"></a>00173                     <span class="stringliteral">&quot;false&quot;</span>);
<a name="l00174"></a>00174 }
<a name="l00175"></a>00175 
<a name="l00176"></a><a class="code" href="multiChannelExample_8cpp.html#aeb252dd3507d78854a59942bcd6f5c52">00176</a> <span class="keywordtype">void</span> <a class="code" href="apt-cel-extract_8cpp.html#a1518275cea55ffab3a9dceea1dbd00f4" title="Fill in options given command line args.">fillInOptions</a>(<a class="code" href="classPgOptions.html">PgOptions</a> *opts, <a class="code" href="classAOptions.html">AOptions</a> &amp;o) {
<a name="l00177"></a>00177   <a class="code" href="classPgOpt.html" title="///">PgOpt</a> *asOpt = NULL;
<a name="l00178"></a>00178   o.help = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;help&quot;</span>);
<a name="l00179"></a>00179   o.verbose = opts-&gt;<a class="code" href="classPgOptions.html#aad90f35336ffd02498d5487d40919641" title="Get the value of the option as a int. ///.">getInt</a>(<span class="stringliteral">&quot;verbose&quot;</span>);
<a name="l00180"></a>00180   o.outDir = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;out-dir&quot;</span>);
<a name="l00181"></a>00181   <a class="code" href="classUtil.html#a21a08a463a4b0b321b58979b0404e017" title="Chop off the last character if it is a path separator.">Util::chompLastIfSep</a>(o.outDir);
<a name="l00182"></a>00182 
<a name="l00183"></a>00183   o.spfFile = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;spf-file&quot;</span>);
<a name="l00184"></a>00184   o.cdfFile = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;cdf-file&quot;</span>);
<a name="l00185"></a>00185 
<a name="l00186"></a>00186   o.channelFile = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;channel-file&quot;</span>);
<a name="l00187"></a>00187 
<a name="l00188"></a>00188   o.quantMethod = <span class="stringliteral">&quot;brlmm-p&quot;</span>;
<a name="l00189"></a>00189   o.algoParams = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;algo-params&quot;</span>);
<a name="l00190"></a>00190   <span class="keywordflow">if</span>(o.algoParams == <span class="stringliteral">&quot;&quot;</span>)
<a name="l00191"></a>00191     o.algoParams = <span class="stringliteral">&quot;CM=1.bins=100.mix=1.bic=2.HARD=3.SB=0.45.KX=1.KH=1.5.KXX=0.5.KAH=-0.6.KHB=-0.6.transform=MVA.AAM=2.0.BBM=-2.0.AAV=0.06.BBV=0.06.ABV=0.06.copyqc=0.000001.wobble=0.05.MS=0.05&quot;</span>;
<a name="l00192"></a>00192   o.exprAnalysisString = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;expr-analysis&quot;</span>);
<a name="l00193"></a>00193   <span class="keywordflow">if</span>(o.exprAnalysisString == <span class="stringliteral">&quot;&quot;</span>)
<a name="l00194"></a>00194     o.exprAnalysisString = <span class="stringliteral">&quot;med-norm.target=1000,pm-only,med-polish.expon=true&quot;</span>;
<a name="l00195"></a>00195 
<a name="l00196"></a>00196   o.specialSnps = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;special-snps&quot;</span>);
<a name="l00197"></a>00197   o.chrXSnps = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;chrX-snps&quot;</span>);
<a name="l00198"></a>00198   o.readModelsBrlmmp=opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;read-models-brlmmp&quot;</span>);
<a name="l00199"></a>00199   o.writeModels = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;write-models&quot;</span>);
<a name="l00200"></a>00200   o.summaries = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;summaries&quot;</span>);
<a name="l00201"></a>00201   o.doFeatureEffects = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;feature-effects&quot;</span>);
<a name="l00202"></a>00202   o.doResiduals = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;feature-details&quot;</span>);
<a name="l00203"></a>00203   o.genderFileIn = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;read-genders&quot;</span>);
<a name="l00204"></a>00204   o.<a class="code" href="classAOptions.html#a5c0b549af05cc88f1562f334db0a709a" title="File to input seed genotypes.">genoTypesFile</a> = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;genotypes&quot;</span>);
<a name="l00205"></a>00205   o.selectProbes = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;select-probes&quot;</span>);
<a name="l00206"></a>00206   o.<a class="code" href="classAOptions.html#a5c0b549af05cc88f1562f334db0a709a" title="File to input seed genotypes.">genoTypesFile</a> = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;genotypes&quot;</span>);
<a name="l00207"></a>00207   o.writeSketch = opts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;write-sketch&quot;</span>);
<a name="l00208"></a>00208   asOpt = opts-&gt;<a class="code" href="classPgOptions.html#a637bef24edc9520ea2e871283e338724" title="Finds the option which is bound to this name.">findOpt</a>(<span class="stringliteral">&quot;target-sketch&quot;</span>);
<a name="l00209"></a>00209   <span class="keywordflow">if</span> (asOpt!=NULL) {
<a name="l00210"></a>00210     asOpt-&gt;<a class="code" href="classPgOpt.html#a09d3945f1bfd6edda948cf50d330ad95" title="Push all our user supplied values into another vector. ///.">push_user_values_into</a>(o.sketchInFile);
<a name="l00211"></a>00211   }
<a name="l00212"></a>00212 
<a name="l00213"></a>00213   <span class="comment">/* Sanity checks... */</span>
<a name="l00214"></a>00214   <span class="keywordflow">if</span>(o.outDir == <span class="stringliteral">&quot;&quot;</span>) 
<a name="l00215"></a>00215     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Must specify an output directory.&quot;</span>);
<a name="l00216"></a>00216 
<a name="l00217"></a>00217   <span class="comment">/* Make our directory. */</span>
<a name="l00218"></a>00218   <span class="keywordflow">if</span>(!Fs::isWriteableDir(o.outDir.c_str())) {
<a name="l00219"></a>00219     <span class="keywordflow">if</span>(<a class="code" href="classFs.html#aea474e8df7665f070c5334fa322bd283" title="make all path of directories ///">Fs::mkdirPath</a>(o.outDir, <span class="keyword">false</span>) != APT_OK) {
<a name="l00220"></a>00220       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Can&#39;t make or write to directory: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(o.outDir));
<a name="l00221"></a>00221     }
<a name="l00222"></a>00222   }
<a name="l00223"></a>00223 
<a name="l00224"></a>00224   <span class="comment">/* load up cel file info */</span>
<a name="l00225"></a>00225   <span class="keywordflow">if</span> (opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;cel-files&quot;</span>)!=<span class="stringliteral">&quot;&quot;</span>) {
<a name="l00226"></a>00226     <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00227"></a>00227 <span class="preprocessor">#ifdef WIN32</span>
<a name="l00228"></a>00228 <span class="preprocessor"></span>    tsv.<a class="code" href="classaffx_1_1TsvFile.html#a3104f733113db8733c335bf425c89a7f" title="Obey the escapechar?">m_optEscapeOk</a> = <span class="keyword">false</span>;
<a name="l00229"></a>00229 <span class="preprocessor">#endif</span>
<a name="l00230"></a>00230 <span class="preprocessor"></span>    std::string celFiles = opts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;cel-files&quot;</span>);
<a name="l00231"></a>00231     std::string sample;
<a name="l00232"></a>00232     tsv.bind(0, <span class="stringliteral">&quot;sample_name&quot;</span>, &amp;sample, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">TSV_BIND_REQUIRED</a>);
<a name="l00233"></a>00233     <span class="keywordflow">if</span>(tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(celFiles) != TSV_OK) {
<a name="l00234"></a>00234       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Couldn&#39;t open cell-files file: &#39;&quot;</span> + celFiles + <span class="stringliteral">&quot;&#39;&quot;</span>);
<a name="l00235"></a>00235     }
<a name="l00236"></a>00236     <span class="keywordtype">int</span> numCol = tsv.<a class="code" href="classaffx_1_1TsvFile.html#a94ad901d12ab05cce0f050f99665970f" title="The number of columns for this level ///.">getColumnCount</a>(0);
<a name="l00237"></a>00237     vector&lt;int&gt; colIndexes;
<a name="l00238"></a>00238     <span class="keywordflow">for</span>(<span class="keywordtype">int</span> cidx=0; cidx &lt; numCol; cidx++) {
<a name="l00239"></a>00239         <span class="keywordtype">string</span> name;
<a name="l00240"></a>00240         tsv.<a class="code" href="classaffx_1_1TsvFile.html#a0958f996be8dea8c5e5d2a0bac36819b" title="Get the name of a column by its level and index.">cidx2cname</a>(0,cidx,name);
<a name="l00241"></a>00241         <span class="keywordflow">if</span>(name.find(<span class="stringliteral">&quot;channel_&quot;</span>) == 0) {<span class="comment"></span>
<a name="l00242"></a>00242 <span class="comment">            ///@todo we order the channels by order seen, not the label</span>
<a name="l00243"></a>00243 <span class="comment">            ///      would be nice to track the label and associate via</span>
<a name="l00244"></a>00244 <span class="comment">            ///      label in lib/panel file</span>
<a name="l00245"></a>00245 <span class="comment"></span>            colIndexes.push_back(cidx);
<a name="l00246"></a>00246         }
<a name="l00247"></a>00247     }
<a name="l00248"></a>00248     o.channelCount = colIndexes.size();
<a name="l00249"></a>00249     o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>.resize(o.channelCount);
<a name="l00250"></a>00250     tsv.<a class="code" href="classaffx_1_1TsvFile.html#a05d594ceb54634fb1f39ca768276800e" title="Rewind back to the start of the data and clear the fields ///.">rewind</a>();
<a name="l00251"></a>00251     <span class="keywordflow">while</span>(tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == TSV_OK) {
<a name="l00252"></a>00252         o.sampleNames.push_back(sample);
<a name="l00253"></a>00253         <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;colIndexes.size(); i++) {
<a name="l00254"></a>00254             <span class="keywordtype">string</span> val;
<a name="l00255"></a>00255             <span class="keywordflow">if</span>(tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0,i+1,val)==TSV_OK) {
<a name="l00256"></a>00256                 o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>[i].push_back(val);
<a name="l00257"></a>00257             } <span class="keywordflow">else</span>
<a name="l00258"></a>00258                 <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Unable to read column &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(i+1) + <span class="stringliteral">&quot; from &quot;</span> + celFiles);
<a name="l00259"></a>00259         }
<a name="l00260"></a>00260     }
<a name="l00261"></a>00261     tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00262"></a>00262     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Read &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>[0].size()) + <span class="stringliteral">&quot; sample(s) and &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>.size()) + <span class="stringliteral">&quot; channel(s) from: &quot;</span> + <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(celFiles));
<a name="l00263"></a>00263   } <span class="keywordflow">else</span> {
<a name="l00264"></a>00264       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Must provide a list of cel files for each channel.&quot;</span>);
<a name="l00265"></a>00265   }
<a name="l00266"></a>00266 }
<a name="l00267"></a>00267 
<a name="l00268"></a><a class="code" href="multiChannelExample_8cpp.html#ad561ebe590e97e4549466bd1d3333f51">00268</a> <span class="keywordtype">void</span> <a class="code" href="multiChannelExample_8cpp.html#ad561ebe590e97e4549466bd1d3333f51">callGenotypes</a>(<a class="code" href="classAOptions.html">AOptions</a> &amp;o) {
<a name="l00269"></a>00269 
<a name="l00270"></a>00270   <span class="comment">/* Get the layout for the chip. */</span>
<a name="l00271"></a>00271   <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> layout;
<a name="l00272"></a>00272   <span class="keywordflow">if</span>(o.cdfFile != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l00273"></a>00273     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Opening cdf file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(o.cdfFile));
<a name="l00274"></a>00274     layout.<a class="code" href="classChipLayout.html#a13aad6c63375822f6a407f1328e500de" title="Open a file and parse the probe sets.">openCdfAll</a>(o.cdfFile);
<a name="l00275"></a>00275   } <span class="keywordflow">else</span> <span class="keywordflow">if</span>(o.spfFile != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l00276"></a>00276     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Opening spf file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(o.spfFile));
<a name="l00277"></a>00277     layout.openSpfAll(o.spfFile);
<a name="l00278"></a>00278   } <span class="keywordflow">else</span> {
<a name="l00279"></a>00279     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Must provide a CDF or SPF file.&quot;</span>);
<a name="l00280"></a>00280   }
<a name="l00281"></a>00281 
<a name="l00282"></a>00282   <span class="comment">/* Get the channel file */</span>
<a name="l00283"></a>00283   map&lt;string,vector&lt;int&gt; &gt; sampleAlleleMap; <span class="comment">///@todo not particularly space efficient</span>
<a name="l00284"></a>00284 <span class="comment"></span>  <span class="keywordflow">if</span> (o.channelFile!=<span class="stringliteral">&quot;&quot;</span>) {
<a name="l00285"></a>00285     <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00286"></a>00286 <span class="preprocessor">#ifdef WIN32</span>
<a name="l00287"></a>00287 <span class="preprocessor"></span>    tsv.<a class="code" href="classaffx_1_1TsvFile.html#a3104f733113db8733c335bf425c89a7f" title="Obey the escapechar?">m_optEscapeOk</a> = <span class="keyword">false</span>;
<a name="l00288"></a>00288 <span class="preprocessor">#endif</span>
<a name="l00289"></a>00289 <span class="preprocessor"></span>    std::string probeset;
<a name="l00290"></a>00290     <span class="keywordtype">int</span> alleleA, alleleB;
<a name="l00291"></a>00291     tsv.bind(0, <span class="stringliteral">&quot;probeset_name&quot;</span>, &amp;probeset, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">TSV_BIND_REQUIRED</a>);
<a name="l00292"></a>00292     tsv.bind(0, <span class="stringliteral">&quot;allele_A_channel&quot;</span>, &amp;alleleA, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">TSV_BIND_REQUIRED</a>);
<a name="l00293"></a>00293     tsv.bind(0, <span class="stringliteral">&quot;allele_B_channel&quot;</span>, &amp;alleleB, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">TSV_BIND_REQUIRED</a>);
<a name="l00294"></a>00294     <span class="keywordflow">if</span>(tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(o.channelFile) != TSV_OK) {
<a name="l00295"></a>00295       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Couldn&#39;t open channel file: &#39;&quot;</span> + o.channelFile + <span class="stringliteral">&quot;&#39;&quot;</span>);
<a name="l00296"></a>00296     }
<a name="l00297"></a>00297     tsv.<a class="code" href="classaffx_1_1TsvFile.html#a05d594ceb54634fb1f39ca768276800e" title="Rewind back to the start of the data and clear the fields ///.">rewind</a>();
<a name="l00298"></a>00298     <span class="keywordflow">while</span>(tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == TSV_OK) {
<a name="l00299"></a>00299         sampleAlleleMap[probeset].push_back(alleleA);
<a name="l00300"></a>00300         sampleAlleleMap[probeset].push_back(alleleB);
<a name="l00301"></a>00301     }
<a name="l00302"></a>00302     tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00303"></a>00303   } <span class="keywordflow">else</span> {
<a name="l00304"></a>00304       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;No channel file. Assuming allele A on channel A and allele B on channel B.&quot;</span>);
<a name="l00305"></a>00305   }
<a name="l00306"></a>00306 
<a name="l00307"></a>00307   <span class="comment">/* Construct the analysis streams for each factory */</span>
<a name="l00308"></a>00308   <a class="code" href="classAnalysisStreamFactory.html" title="Object for making new analysis streams from a text specification.">AnalysisStreamFactory</a> asFactory; 
<a name="l00309"></a>00309   vector&lt;AnalysisStream *&gt; streams;
<a name="l00310"></a>00310   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Allele estimates computed using: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(o.exprAnalysisString));
<a name="l00311"></a>00311   Verbose::progressBegin(1,<span class="stringliteral">&quot;Construction ChipStreams for computing allele estimates on each channel&quot;</span>,
<a name="l00312"></a>00312     o.channelCount, 0, o.channelCount);
<a name="l00313"></a>00313   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> channel = 0; channel &lt; o.channelCount; channel++) {
<a name="l00314"></a>00314     <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00315"></a>00315     <span class="comment">// do we need to write out the sketch</span>
<a name="l00316"></a>00316     <span class="keywordflow">if</span>(o.writeSketch) {
<a name="l00317"></a>00317       <span class="keywordtype">string</span> prefix = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(o.outDir,<span class="stringliteral">&quot;ch&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(channel) + <span class="stringliteral">&quot;.&quot;</span>);
<a name="l00318"></a>00318       asFactory.<a class="code" href="classAnalysisStreamFactory.html#a0e119d04b6da39e12a1c399c36f07628" title="Toggle whether or not we&amp;#39;re writing out the normalization distribution.">setWriteSketchDir</a>(prefix);
<a name="l00319"></a>00319     }
<a name="l00320"></a>00320     <span class="comment">// do we have a sketch to read in</span>
<a name="l00321"></a>00321     <span class="keywordflow">if</span>(o.sketchInFile.size() == 1) {
<a name="l00322"></a>00322       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Opening target normalization file for channel &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(channel) + 
<a name="l00323"></a>00323               <span class="stringliteral">&quot;: &quot;</span> + <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(o.sketchInFile[0].c_str()));
<a name="l00324"></a>00324       asFactory.<a class="code" href="classAnalysisStreamFactory.html#ae5f39c90584457e47920727bd3d031c1" title="Read in the target sketch from a.">readTargetSketchFromFile</a>(o.sketchInFile[0].c_str());
<a name="l00325"></a>00325     }<span class="keywordflow">else</span> <span class="keywordflow">if</span>(o.sketchInFile.size() &gt; 1) {
<a name="l00326"></a>00326       <span class="keywordflow">if</span>(o.sketchInFile.size() != o.channelCount) {
<a name="l00327"></a>00327           <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;When specifying more than 1 sketch, number of sketches must match number of channels&quot;</span>);
<a name="l00328"></a>00328       }
<a name="l00329"></a>00329       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Opening target normalization file for channel &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(channel) + 
<a name="l00330"></a>00330               <span class="stringliteral">&quot;: &quot;</span> + <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(o.sketchInFile[channel].c_str()));
<a name="l00331"></a>00331       asFactory.<a class="code" href="classAnalysisStreamFactory.html#ae5f39c90584457e47920727bd3d031c1" title="Read in the target sketch from a.">readTargetSketchFromFile</a>(o.sketchInFile[channel].c_str());
<a name="l00332"></a>00332     }
<a name="l00333"></a>00333     streams.push_back(asFactory.<a class="code" href="classAnalysisStreamFactory.html#a45a3731c9f0015ba0cbadf529cc87ccf" title="Factory for created AnalysisStreams from a string description.">constructAnalysisStream</a>(o.exprAnalysisString.c_str(),layout));
<a name="l00334"></a>00334   }
<a name="l00335"></a>00335   Verbose::progressEnd(1, <span class="stringliteral">&quot;Done.&quot;</span>);
<a name="l00336"></a>00336   
<a name="l00337"></a>00337   <span class="comment">/* Setup Intensity Mart for each channel */</span>
<a name="l00338"></a>00338   Verbose::progressBegin(1,<span class="stringliteral">&quot;Setting up intensity marts for each channel&quot;</span>,
<a name="l00339"></a>00339     streams.size(), 0, streams.size());
<a name="l00340"></a>00340   vector&lt;DiskIntensityMart *&gt; iMarts;
<a name="l00341"></a>00341   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;streams.size(); i++) {
<a name="l00342"></a>00342       <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00343"></a>00343       vector&lt;int&gt; desiredOrder;
<a name="l00344"></a>00344       <span class="keywordflow">for</span>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> psIx = 0; psIx &lt; layout.<a class="code" href="classChipLayout.html#a71df054a01d62ca33d6b8de8c8c5c3ca" title="How many probe sets in Pgf file?">getProbeSetCount</a>(); psIx++) {
<a name="l00345"></a>00345           <a class="code" href="classProbeListPacked.html">ProbeListPacked</a> pl = layout.<a class="code" href="classChipLayout.html#a3127c7baedda03477f28da70bbfb686e" title="Get a probelist by the index in the layout.">getProbeListAtIndex</a>(psIx);
<a name="l00346"></a>00346           <span class="keywordflow">for</span>(<span class="keywordtype">int</span> pIx = 0; pIx &lt; pl.probe_cnt(); pIx++) {
<a name="l00347"></a>00347               desiredOrder.push_back(pl.get_probeId(pIx));
<a name="l00348"></a>00348           }
<a name="l00349"></a>00349       }
<a name="l00350"></a>00350       <a class="code" href="classDiskIntensityMart.html" title="Very simple version of an IntensityMart based directly on cel files for comparison and troubleshootin...">DiskIntensityMart</a> * ptr = <span class="keyword">new</span> <a class="code" href="classDiskIntensityMart.html" title="Very simple version of an IntensityMart based directly on cel files for comparison and troubleshootin...">DiskIntensityMart</a>(
<a name="l00351"></a>00351                 desiredOrder,
<a name="l00352"></a>00352                 o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>[i], 
<a name="l00353"></a>00353                 50 * 1048576, 
<a name="l00354"></a>00354                 o.outDir, <span class="stringliteral">&quot;multiChannelExample.tmp.ch&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(i), <span class="keyword">true</span>);
<a name="l00355"></a>00355       iMarts.push_back(ptr);
<a name="l00356"></a>00356   }
<a name="l00357"></a>00357   Verbose::progressEnd(1, <span class="stringliteral">&quot;Done.&quot;</span>);
<a name="l00358"></a>00358 
<a name="l00359"></a>00359   <span class="comment">/* Setup CelReaders for each Channel */</span>
<a name="l00360"></a>00360   Verbose::progressBegin(1,<span class="stringliteral">&quot;Setting up CEL readers for each channel&quot;</span>,
<a name="l00361"></a>00361     streams.size(), 0, streams.size());
<a name="l00362"></a>00362   vector&lt;CelReader&gt; cReaders(streams.size());
<a name="l00363"></a>00363   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;streams.size(); i++) {
<a name="l00364"></a>00364     <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00365"></a>00365     <span class="comment">/* As cel files are read the chipstream wants to see them and learn parameters. */</span>
<a name="l00366"></a>00366     cReaders[i].registerStream(streams[i]-&gt;getChipStreamHead());
<a name="l00367"></a>00367 
<a name="l00368"></a>00368     <span class="comment">/* As cel files are read the data cache needs to store intensities. */</span>
<a name="l00369"></a>00369     cReaders[i].registerIntensityMart(iMarts[i]);
<a name="l00370"></a>00370   }
<a name="l00371"></a>00371   Verbose::progressEnd(1, <span class="stringliteral">&quot;Done.&quot;</span>);
<a name="l00372"></a>00372 
<a name="l00373"></a>00373   <span class="comment">/* Setup the MultiChannelQuantMethod */</span>
<a name="l00374"></a>00374   <a class="code" href="classMultiQuantMethodListener.html" title="Class for reporting results of quantification methods.">MultiQuantMethodListener</a> multiQuant;
<a name="l00375"></a>00375   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;streams.size(); i++) {
<a name="l00376"></a>00376       <span class="comment">// Because the stream owns the reporter we</span>
<a name="l00377"></a>00377       <span class="comment">// use the listener rather than adding the multi quant directly</span>
<a name="l00378"></a>00378       <a class="code" href="classQuantMethodReportListener.html">QuantMethodReportListener</a> *listener = <span class="keyword">new</span> <a class="code" href="classQuantMethodReportListener.html">QuantMethodReportListener</a>();
<a name="l00379"></a>00379       listener-&gt;registerListener(&amp;multiQuant);
<a name="l00380"></a>00380       streams[i]-&gt;addReporter(listener);
<a name="l00381"></a>00381   }
<a name="l00382"></a>00382 
<a name="l00383"></a>00383   <span class="comment">// Create Quant Method</span>
<a name="l00384"></a>00384   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Creating QuantMethod for genotyping from: &quot;</span> + o.quantMethod + <span class="stringliteral">&quot;.&quot;</span> + o.algoParams);
<a name="l00385"></a>00385   <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> quantMethodFactory(QuantMethodFactory::GenoType);
<a name="l00386"></a>00386   <span class="keywordtype">string</span> callingString = o.quantMethod + <span class="stringliteral">&quot;.&quot;</span> + o.algoParams;
<a name="l00387"></a>00387   <a class="code" href="classQuantGTypeMethod.html" title="Quantification methods used for making genotyping calls implement (currently just brlmm) this interfa...">QuantGTypeMethod</a> *qMethod = 
<a name="l00388"></a>00388       quantMethodFactory.<a class="code" href="classQuantMethodFactory.html#a6fcc5486c4245085dd61854f5471451b" title="Factory for creating genotyping QuantMethod from string representation.">quantGTypeMethodForString</a>(callingString, layout, QuantMethodFactory::GenoType);
<a name="l00389"></a>00389   <a class="code" href="classQuantLabelZ.html" title="This is the primary class wrapping the bayes_label_routine it handles bookkeeping, gender determination, etc.">QuantLabelZ</a> *qBrlmmp = <span class="keyword">dynamic_cast&lt;</span><a class="code" href="classQuantLabelZ.html" title="This is the primary class wrapping the bayes_label_routine it handles bookkeeping, gender determination, etc.">QuantLabelZ</a> *<span class="keyword">&gt;</span>(qMethod);
<a name="l00390"></a>00390 
<a name="l00391"></a>00391   <span class="comment">// setup genders</span>
<a name="l00392"></a>00392   vector&lt;Gender&gt; genders;
<a name="l00393"></a>00393   <span class="keywordflow">if</span>(o.genderFileIn != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l00394"></a>00394       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Reading genders from &quot;</span> + o.genderFileIn);
<a name="l00395"></a>00395       <a class="code" href="classFileGenders.html" title="Read gender assignments from file.">FileGenders</a> genderReader(o.genderFileIn,o.sampleNames);
<a name="l00396"></a>00396       genders = genderReader.<a class="code" href="classFileGenders.html#aae27b5d97d0527f3b8108016d824cd63" title="Get the genders for the cel files that have been seen.">getGenders</a>();
<a name="l00397"></a>00397   } <span class="keywordflow">else</span> {
<a name="l00398"></a>00398       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;No gender file. Assuming all samples are unknown.&quot;</span>);
<a name="l00399"></a>00399       <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i = 0; i &lt; o.sampleNames.size(); i++)
<a name="l00400"></a>00400           genders.push_back(<a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669aa259d114b82ef05d3f1f9d3c81c84719" title="Don&amp;#39;t know genotype, use XX model.">UnknownGender</a>);
<a name="l00401"></a>00401   }
<a name="l00402"></a>00402   qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#a75d2851d0da897d336e42010653fbb4e" title="sets the genders for all samples">setGenders</a>(genders);
<a name="l00403"></a>00403 
<a name="l00404"></a>00404   <span class="comment">// setup special SNPs</span>
<a name="l00405"></a>00405   SpecialSnpMap specialSnps;
<a name="l00406"></a>00406   map&lt;string,bool&gt; haploidSnps;
<a name="l00407"></a>00407   <span class="keywordflow">if</span>(o.specialSnps != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l00408"></a>00408       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Reading in special SNPs from o.specialSnps&quot;</span>);
<a name="l00409"></a>00409       <span class="keywordtype">string</span> chip, specialType;
<a name="l00410"></a>00410       readSpecialSnpsFile(&amp;specialSnps, &amp;chip, &amp;specialType, o.specialSnps);
<a name="l00411"></a>00411       haploidFromSpecial(specialSnps,haploidSnps);
<a name="l00412"></a>00412   } <span class="keywordflow">else</span> <span class="keywordflow">if</span>(o.chrXSnps != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l00413"></a>00413       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Reading in special SNPs from o.chrXSnps&quot;</span>);
<a name="l00414"></a>00414       <a class="code" href="SpecialSnps_8cpp.html#a24da823ec3912983e63d79ccdd48e510" title="Determine which version of a chrx snp have and open appropriately.">fillInHaploidSnps</a>(o.chrXSnps, haploidSnps, NULL, <span class="keyword">true</span>);
<a name="l00415"></a>00415       specialFromHaploid(specialSnps, haploidSnps);
<a name="l00416"></a>00416   } <span class="keywordflow">else</span> {
<a name="l00417"></a>00417       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;No special SNPs file provided. Assuming everything is bi-allelic.&quot;</span>);
<a name="l00418"></a>00418   }
<a name="l00419"></a>00419 
<a name="l00420"></a>00420   <span class="keywordflow">if</span>(!specialSnps.empty())
<a name="l00421"></a>00421       qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#ac14724ad579f92836d1a6c779b9b396b" title="which snps are special">setSpecialSnps</a>(specialSnps);
<a name="l00422"></a>00422   <span class="keywordflow">else</span> <span class="keywordflow">if</span>(!haploidSnps.empty())
<a name="l00423"></a>00423       qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#a5afceb726f32320451aeb0c5a334a92c" title="which snps are haploid snps">setHaploidSnps</a>(haploidSnps);
<a name="l00424"></a>00424 
<a name="l00425"></a>00425   <span class="comment">// read in SNP models</span>
<a name="l00426"></a>00426   <span class="keywordflow">if</span>(o.readModelsBrlmmp!=<span class="stringliteral">&quot;&quot;</span>) {
<a name="l00427"></a>00427       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Reading in SNP models from &quot;</span> + o.readModelsBrlmmp);
<a name="l00428"></a>00428       qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#ab4a3970a1d381a0885d82ebc66ddb3ca" title="read in the snp specific priors from a file">readSnpPriorMap</a>(o.readModelsBrlmmp.c_str());
<a name="l00429"></a>00429   } <span class="keywordflow">else</span> {
<a name="l00430"></a>00430       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;No SNP models provided. Assuming generic model.&quot;</span>);
<a name="l00431"></a>00431   }
<a name="l00432"></a>00432 
<a name="l00433"></a>00433   <span class="comment">// write SNP models?</span>
<a name="l00434"></a>00434   <span class="keywordflow">if</span>(o.writeModels) {
<a name="l00435"></a>00435     <span class="keywordtype">string</span> outfile = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(o.outDir,o.quantMethod + <span class="stringliteral">&quot;.snp-posteriors&quot;</span>);
<a name="l00436"></a>00436     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">writeSnpPosteriorTsv</a>(outfile,affx::TsvReport::FMT_TSV);
<a name="l00437"></a>00437   }
<a name="l00438"></a>00438   <span class="keywordflow">if</span> (o.selectProbes){
<a name="l00439"></a>00439           <span class="keywordtype">string</span> outfile = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(o.outDir,o.quantMethod + <span class="stringliteral">&quot;.select-probes&quot;</span>);
<a name="l00440"></a>00440     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#ad7cae32c6830efa7475fb948f2442682" title="Write header to the selection file so people don&amp;#39;t have to look up everything.">writeSnpProbeTsv</a>(outfile,affx::TsvReport::FMT_TSV);
<a name="l00441"></a>00441   }
<a name="l00442"></a>00442   
<a name="l00443"></a>00443   <span class="comment">// read in seed genotypes</span>
<a name="l00444"></a>00444   std::map&lt;const char *,std::vector&lt;GType&gt;, <a class="code" href="structUtil_1_1ltstr.html" title="Comparison object for use in map for char *.">Util::ltstr</a> &gt; knownGenotypes;
<a name="l00445"></a>00445   <span class="keywordflow">if</span>(!o.<a class="code" href="classAOptions.html#a5c0b549af05cc88f1562f334db0a709a" title="File to input seed genotypes.">genoTypesFile</a>.empty()) {
<a name="l00446"></a>00446     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Reading in SNP genotype calls from &quot;</span> + o.<a class="code" href="classAOptions.html#a5c0b549af05cc88f1562f334db0a709a" title="File to input seed genotypes.">genoTypesFile</a>);
<a name="l00447"></a>00447     std::vector&lt;std::string&gt; probesets;
<a name="l00448"></a>00448     <a class="code" href="classGenoSeedTxt.html" title="Implements interface for a genotyping oracle used for genotyping seeds in brlmm.">GenoSeedTxt</a> seeds(o.<a class="code" href="classAOptions.html#a5c0b549af05cc88f1562f334db0a709a" title="File to input seed genotypes.">genoTypesFile</a>.c_str(), layout, o.sampleNames, haploidSnps);
<a name="l00449"></a>00449     seeds.fillProbesets(probesets);
<a name="l00450"></a>00450 
<a name="l00451"></a>00451     <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;probesets.size(); i++) {
<a name="l00452"></a>00452       <span class="keyword">const</span> <span class="keywordtype">char</span> *name = <a class="code" href="classUtil.html#a508b7aff9569ed929945d0d37eca384e" title="Create a copy of a string.">Util::cloneString</a>(probesets[i].c_str());
<a name="l00453"></a>00453       vector&lt;GType&gt; callVec = seeds.getGenoCalls(name);
<a name="l00454"></a>00454       knownGenotypes[name] = callVec;
<a name="l00455"></a>00455       <span class="comment">//Verbose::out(1,&quot;Setting known genotypes for: &quot; + ToStr(name));</span>
<a name="l00456"></a>00456     }
<a name="l00457"></a>00457     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#a22b847bd3dff1573d18dc4f248ddd509" title="Set the initial guesses of genotypes, usually from DM predictions.">setKnownGenoTypes</a>(&amp;knownGenotypes);
<a name="l00458"></a>00458   }
<a name="l00459"></a>00459 
<a name="l00460"></a>00460   <span class="comment">// Dummy Data Mart for Reporters</span>
<a name="l00461"></a>00461   <span class="comment">// Used to spoof sample names</span>
<a name="l00462"></a>00462   <a class="code" href="classDiskIntensityMart.html" title="Very simple version of an IntensityMart based directly on cel files for comparison and troubleshootin...">DiskIntensityMart</a> *dummyMart = iMarts[0];
<a name="l00463"></a>00463     
<a name="l00464"></a>00464   <span class="comment">// Create Report</span>
<a name="l00465"></a>00465   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Creating main genotype reporter.&quot;</span>);
<a name="l00466"></a>00466   <a class="code" href="classQuantMethodGTypeReport.html">QuantMethodGTypeReport</a> *report = <span class="keyword">new</span> <a class="code" href="classQuantMethodGTypeReport.html">QuantMethodGTypeReport</a>();
<a name="l00467"></a>00467   report-&gt;setDirPath(o.outDir);
<a name="l00468"></a>00468   report-&gt;setFileprefix(o.quantMethod);
<a name="l00469"></a>00469   report-&gt;setFormat(affx::TsvReport::FMT_TSV);
<a name="l00470"></a>00470   report-&gt;setFilename(<span class="stringliteral">&quot;QuantMethodGTypeReport-DEBUG&quot;</span>); <span class="comment">// filename not used -- for debugging</span>
<a name="l00471"></a>00471   report-&gt;setPrecision(5);
<a name="l00472"></a>00472   
<a name="l00473"></a>00473   <span class="keywordtype">string</span> qMethodGuid = <a class="code" href="classaffxutil_1_1Guid.html#a2929c982cb9f3b5f6c5d7d818a8e6ac4">affxutil::Guid::GenerateNewGuid</a>();
<a name="l00474"></a>00474   <span class="keywordtype">string</span> execGuid = <a class="code" href="classaffxutil_1_1Guid.html#a2929c982cb9f3b5f6c5d7d818a8e6ac4">affxutil::Guid::GenerateNewGuid</a>();
<a name="l00475"></a>00475   <span class="keywordtype">string</span> timeStr = <span class="stringliteral">&quot;&quot;</span>; <span class="comment">///@todo set time string</span>
<a name="l00476"></a>00476 <span class="comment"></span>  <span class="keywordtype">string</span> commandLine = <span class="stringliteral">&quot;&quot;</span>; <span class="comment">///@todo set command line string</span>
<a name="l00477"></a>00477 <span class="comment"></span>  <span class="keywordtype">string</span> execVersion = <span class="stringliteral">&quot;&quot;</span>; <span class="comment">///@todo set command line string</span>
<a name="l00478"></a>00478 <span class="comment"></span>  <a class="code" href="classAnalysisInfo.html" title="Information about a particular analysis run.">AnalysisInfo</a> info; <span class="comment">///@todo populate analysis info object</span>
<a name="l00479"></a>00479 <span class="comment"></span>  
<a name="l00480"></a>00480   report-&gt;<a class="code" href="classQuantMethodReport.html#acc296969ce5774f4479fc7275de753f0" title="Print a message out to the stream.">addStdHeaders</a>((<a class="code" href="classQuantMethodReport.html" title="Class for reporting results of quantification methods.">QuantMethodReport</a> *)report, 
<a name="l00481"></a>00481                         execGuid,
<a name="l00482"></a>00482                         qMethodGuid,
<a name="l00483"></a>00483                         timeStr,
<a name="l00484"></a>00484                         commandLine,
<a name="l00485"></a>00485                         <span class="comment">//o.sampleNames,</span>
<a name="l00486"></a>00486                         execVersion, 
<a name="l00487"></a>00487                         info);
<a name="l00488"></a>00488   <span class="comment">//</span>
<a name="l00489"></a>00489   report-&gt;<a class="code" href="classQuantMethodReport.html#a4defc7fcba241175db8c7c37d42f790c" title="Get set up for a run of reporting probesets.">prepare</a>(*qMethod, *dummyMart);
<a name="l00490"></a>00490 
<a name="l00491"></a>00491   <span class="comment">// Create allele summaries report</span>
<a name="l00492"></a>00492   <a class="code" href="classQuantMethodExprReport.html" title="Class for reporting results of quantification methods.">QuantMethodExprReport</a> *summaryReport = NULL;
<a name="l00493"></a>00493   <a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *eMethod = NULL;
<a name="l00494"></a>00494   <span class="keywordflow">if</span>(o.summaries) {
<a name="l00495"></a>00495         eMethod = <span class="keyword">dynamic_cast&lt;</span><a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *<span class="keyword">&gt;</span>(streams[0]-&gt;getQuantMethod());
<a name="l00496"></a>00496         <span class="keywordflow">if</span>(eMethod == NULL)
<a name="l00497"></a>00497             <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Must use QuantExprMethod to report summaries&quot;</span>);
<a name="l00498"></a>00498 <span class="comment">//        string reportPrefix = o.outDir + PATH_SEPARATOR + o.quantMethod + &quot;.&quot; + eMethod-&gt;getType();</span>
<a name="l00499"></a>00499 <span class="comment">//        summaryReport = new QuantMethodExprReport(reportPrefix.c_str(),</span>
<a name="l00500"></a>00500 <span class="comment">//                                                  *eMethod, </span>
<a name="l00501"></a>00501 <span class="comment">//                                                  true,</span>
<a name="l00502"></a>00502 <span class="comment">//                                                  o.doFeatureEffects,</span>
<a name="l00503"></a>00503 <span class="comment">//                                                  o.doResiduals,</span>
<a name="l00504"></a>00504 <span class="comment">//                                                  5);</span>
<a name="l00505"></a>00505         summaryReport = <span class="keyword">new</span> <a class="code" href="classQuantMethodExprReport.html" title="Class for reporting results of quantification methods.">QuantMethodExprReport</a>(info.getNumCols());
<a name="l00506"></a>00506         summaryReport-&gt;setIsHeaderBuffer(1);
<a name="l00507"></a>00507         summaryReport-&gt;setDirPath(o.outDir);
<a name="l00508"></a>00508         summaryReport-&gt;setFileprefix(o.quantMethod + <span class="stringliteral">&quot;.&quot;</span> + eMethod-&gt;<a class="code" href="classQuantExprMethod.html#a467749213636536eeab20b0100b2cadd" title="What is the name of the quantification method?">getType</a>());
<a name="l00509"></a>00509         summaryReport-&gt;setFilename(<span class="stringliteral">&quot;QuantMethodExprReport-DEBUG&quot;</span>); <span class="comment">// filename not used -- for debugging</span>
<a name="l00510"></a>00510         summaryReport-&gt;setFormat(affx::TsvReport::FMT_TSV);
<a name="l00511"></a>00511         <span class="comment">//</span>
<a name="l00512"></a>00512         summaryReport-&gt;m_qMethod=eMethod;
<a name="l00513"></a>00513         summaryReport-&gt;setPrecision(5);
<a name="l00514"></a>00514         summaryReport-&gt;m_DoSummary=<span class="keyword">true</span>;
<a name="l00515"></a>00515         summaryReport-&gt;m_DoFeatureEffects=<span class="keyword">true</span>;
<a name="l00516"></a>00516         summaryReport-&gt;m_DoResiduals=<span class="keyword">true</span>;
<a name="l00517"></a>00517 
<a name="l00518"></a>00518         <span class="comment">//summaryReport-&gt;writeHeader(summaryReport, execGuid.c_str(), qMethodGuid.c_str(),</span>
<a name="l00519"></a>00519         <span class="comment">//                          timeStr.c_str(), commandLine.c_str(),</span>
<a name="l00520"></a>00520         <span class="comment">//                          o.sampleNames, execVersion.c_str(), info);</span>
<a name="l00521"></a>00521 
<a name="l00522"></a>00522         report-&gt;<a class="code" href="classQuantMethodReport.html#acc296969ce5774f4479fc7275de753f0" title="Print a message out to the stream.">addStdHeaders</a>(report,
<a name="l00523"></a>00523                               execGuid,
<a name="l00524"></a>00524                               qMethodGuid,
<a name="l00525"></a>00525                               timeStr,
<a name="l00526"></a>00526                               commandLine,
<a name="l00527"></a>00527                               execVersion,
<a name="l00528"></a>00528                               info);
<a name="l00529"></a>00529 
<a name="l00530"></a>00530         summaryReport-&gt;<a class="code" href="classQuantMethodExprReport.html#a8cf2fd8394b93f50a4dc21735430e04b" title="Get set up for a run of reporting probesets.">prepare</a>(*eMethod, *dummyMart);
<a name="l00531"></a>00531   }
<a name="l00532"></a>00532 
<a name="l00533"></a>00533   <span class="comment">// Dummies needed by the reporter</span>
<a name="l00534"></a>00534   <a class="code" href="classPmOnlyAdjust.html" title="Pm adjustment class that does no modification.">PmOnlyAdjust</a> pmAdjust;
<a name="l00535"></a>00535   vector&lt;ChipStream *&gt; iTrans;
<a name="l00536"></a>00536 
<a name="l00537"></a>00537   <span class="comment">/* register the cel files to open. */</span>
<a name="l00538"></a>00538   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Registering CEL files for each channel.&quot;</span>);
<a name="l00539"></a>00539   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> chIx=0; chIx&lt;streams.size(); chIx++) {
<a name="l00540"></a>00540     vector&lt;string&gt; fileNames;
<a name="l00541"></a>00541     <span class="keywordflow">for</span>(<span class="keywordtype">int</span> celIx = 0; celIx &lt; o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>[chIx].size(); celIx++)
<a name="l00542"></a>00542         fileNames.push_back(o.<a class="code" href="classAOptions.html#abcbd67993800d7e2f87bad10a4ef41b9" title="first dim channel, second is samples">celFiles</a>[chIx][celIx]);
<a name="l00543"></a>00543     cReaders[chIx].setFiles(fileNames);
<a name="l00544"></a>00544   }
<a name="l00545"></a>00545 
<a name="l00546"></a>00546   <span class="comment">// Read in cel files</span>
<a name="l00547"></a>00547   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;cReaders.size(); i++) {
<a name="l00548"></a>00548       cReaders[i].readFiles();
<a name="l00549"></a>00549   }
<a name="l00550"></a>00550 
<a name="l00551"></a>00551 
<a name="l00552"></a>00552   <span class="comment">/* Loop through all of our probe sets and do an expression quantification for each one. */</span>
<a name="l00553"></a>00553   Verbose::progressBegin(1,<span class="stringliteral">&quot;Processing probesets&quot;</span>,
<a name="l00554"></a>00554     40, (<span class="keywordtype">int</span>)(layout.<a class="code" href="classChipLayout.html#a71df054a01d62ca33d6b8de8c8c5c3ca" title="How many probe sets in Pgf file?">getProbeSetCount</a>()/40), layout.<a class="code" href="classChipLayout.html#a71df054a01d62ca33d6b8de8c8c5c3ca" title="How many probe sets in Pgf file?">getProbeSetCount</a>());
<a name="l00555"></a>00555   <span class="keywordflow">for</span>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> psIx = 0; psIx &lt; layout.<a class="code" href="classChipLayout.html#a71df054a01d62ca33d6b8de8c8c5c3ca" title="How many probe sets in Pgf file?">getProbeSetCount</a>(); psIx++) {
<a name="l00556"></a>00556     <a class="code" href="classVerbose.html#ab28a28741a3ab2086099e2d66e920eb5" title="Print a dot out to let the user know we are still alive and making progress.">Verbose::progressStep</a>(1);
<a name="l00557"></a>00557     <a class="code" href="classProbeListPacked.html">ProbeListPacked</a> pList = layout.<a class="code" href="classChipLayout.html#a3127c7baedda03477f28da70bbfb686e" title="Get a probelist by the index in the layout.">getProbeListAtIndex</a>(psIx);
<a name="l00558"></a>00558     <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *ps = <a class="code" href="classProbeListFactory.html#aaed559bbfe723ed4fe225f805a2411e8">ProbeListFactory::asProbeSet</a>(pList);
<a name="l00559"></a>00559     <a class="code" href="classProbeSetGroup.html" title="Group of probe sets that should be processed as a single large probe set.">ProbeSetGroup</a> group(ps);     <span class="comment">// psGroup should delete the memory for ps...</span>
<a name="l00560"></a>00560     <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;streams.size(); i++){
<a name="l00561"></a>00561         streams[i]-&gt;doAnalysis(group, *iMarts[i], <span class="keyword">true</span>);
<a name="l00562"></a>00562     }
<a name="l00563"></a>00563 <span class="comment"></span>
<a name="l00564"></a>00564 <span class="comment">    ///@todo much of this should probably be in a MultiQuantLabelZ</span>
<a name="l00565"></a>00565 <span class="comment">    ///      rather than using QuantLabelZ directly. Accessing</span>
<a name="l00566"></a>00566 <span class="comment">    ///      m_Results directly is ugly as well.</span>
<a name="l00567"></a>00567 <span class="comment"></span>
<a name="l00568"></a>00568     <span class="comment">// set probeset name for report</span>
<a name="l00569"></a>00569     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#a4b257a271e340b468e41a22865b738e6" title="Set the probeset name.">setProbeSetName</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>);
<a name="l00570"></a>00570 
<a name="l00571"></a>00571     <span class="comment">// Figure out the channels for this SNP</span><span class="comment"></span>
<a name="l00572"></a>00572 <span class="comment">    ///@todo default should be global (ie if we have a channel file, then all probesets must be found in it. only default to 0/1 when there is no channel file)</span>
<a name="l00573"></a>00573 <span class="comment"></span>    <span class="keywordtype">int</span> alleleAIndex = 0;
<a name="l00574"></a>00574     <span class="keywordtype">int</span> alleleBIndex = 1;
<a name="l00575"></a>00575     <span class="keywordflow">if</span>(sampleAlleleMap.find(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) != sampleAlleleMap.end()) {
<a name="l00576"></a>00576         alleleAIndex = sampleAlleleMap[ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>][0];
<a name="l00577"></a>00577         alleleBIndex = sampleAlleleMap[ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>][1];
<a name="l00578"></a>00578     } <span class="keywordflow">else</span> {
<a name="l00579"></a>00579         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Unable to find probeset &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot; in channel file. Assuming channels 1 and 2.&quot;</span>);
<a name="l00580"></a>00580     }
<a name="l00581"></a>00581 
<a name="l00582"></a>00582     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#adde341ea0e1f3e3e18ff6e26b105da1e" title="Set allele values.">setAlleleValues</a>(multiQuant.m_Results[alleleAIndex], multiQuant.m_Results[alleleBIndex]);
<a name="l00583"></a>00583 
<a name="l00584"></a>00584     <span class="comment">// dump allele summaries</span>
<a name="l00585"></a>00585     <span class="keywordflow">if</span>(o.summaries) {
<a name="l00586"></a>00586         <span class="comment">// Allele A</span>
<a name="l00587"></a>00587         group.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a> = <a class="code" href="classUtil.html#a508b7aff9569ed929945d0d37eca384e" title="Create a copy of a string.">Util::cloneString</a>((<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot;-A&quot;</span>).c_str());
<a name="l00588"></a>00588         summaryReport-&gt;<a class="code" href="classQuantMethodExprReport.html#a5b88eafda34c24751a4baee0de35ac4f" title="After every probeset computation this function is called an is an opportunity to query the quantifica...">report</a>(group, *(streams[alleleAIndex]-&gt;getQuantMethod()), *(iMarts[alleleAIndex]), iTrans, pmAdjust);
<a name="l00589"></a>00589         <span class="comment">// Allele B</span>
<a name="l00590"></a>00590         group.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a> = <a class="code" href="classUtil.html#a508b7aff9569ed929945d0d37eca384e" title="Create a copy of a string.">Util::cloneString</a>((<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot;-B&quot;</span>).c_str());
<a name="l00591"></a>00591         summaryReport-&gt;<a class="code" href="classQuantMethodExprReport.html#a5b88eafda34c24751a4baee0de35ac4f" title="After every probeset computation this function is called an is an opportunity to query the quantifica...">report</a>(group, *(streams[alleleBIndex]-&gt;getQuantMethod()), *(iMarts[alleleBIndex]), iTrans, pmAdjust);
<a name="l00592"></a>00592     }
<a name="l00593"></a>00593 
<a name="l00594"></a>00594     <span class="comment">// transform the data as specified</span>
<a name="l00595"></a>00595     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909" title="Transform allele estimates.">transform</a>();
<a name="l00596"></a>00596    
<a name="l00597"></a>00597     <span class="comment">// load up seed genotypes</span>
<a name="l00598"></a>00598     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#aa0bb7cf31d9f77d47947cf91d2fa75e0" title="Load up the initial call for the requested SNP.">loadInitialCallsForSnp</a>(ps-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>);
<a name="l00599"></a>00599 
<a name="l00600"></a>00600     <span class="comment">// make the call</span>
<a name="l00601"></a>00601     qBrlmmp-&gt;<a class="code" href="classQuantLabelZ.html#ad83c392dc97c1557e25bdd8b9b901ec9" title="Do the heavy lifting of estimation.">computeEstimate</a>();
<a name="l00602"></a>00602     
<a name="l00603"></a>00603     <span class="comment">// report results</span>
<a name="l00604"></a>00604     report-&gt;<a class="code" href="classQuantMethodReport.html#a6045850dff05e101ad1afed3c752e250" title="After every probeset computation this function is called an is an opportunity to query the quantifica...">report</a>(group, *qMethod, *iMarts[0], iTrans, pmAdjust);
<a name="l00605"></a>00605 
<a name="l00606"></a>00606     <span class="comment">// clear out results for the current probeset</span>
<a name="l00607"></a>00607     multiQuant.m_Results.clear();
<a name="l00608"></a>00608   }
<a name="l00609"></a>00609   Verbose::progressEnd(1, <span class="stringliteral">&quot;Done.&quot;</span>);
<a name="l00610"></a>00610 
<a name="l00611"></a>00611   <span class="comment">// close the report</span>
<a name="l00612"></a>00612   report-&gt;<a class="code" href="classQuantMethodReport.html#a07a2e186a7bff8b6618f160d496dc0e4" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">finish</a>(*qMethod);
<a name="l00613"></a>00613   <span class="keywordflow">if</span>(summaryReport != NULL) 
<a name="l00614"></a>00614     summaryReport-&gt;<a class="code" href="classQuantMethodExprReport.html#abdc173d827ddbe276e12a9e461a781c2" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">finish</a>(*eMethod);
<a name="l00615"></a>00615 
<a name="l00616"></a>00616   <span class="comment">// delete the QuantMethod</span>
<a name="l00617"></a>00617   <span class="keyword">delete</span> qBrlmmp;
<a name="l00618"></a>00618   
<a name="l00619"></a>00619   <span class="comment">// Free up memory</span>
<a name="l00620"></a>00620   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i = 0; i &lt; streams.size(); i++)
<a name="l00621"></a>00621     <span class="keyword">delete</span> streams[i];
<a name="l00622"></a>00622   <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i = 0; i &lt; iMarts.size(); i++)
<a name="l00623"></a>00623     <span class="keyword">delete</span> iMarts[i];
<a name="l00624"></a>00624 }
<a name="l00625"></a>00625 
<a name="l00626"></a>00626 <span class="keywordtype">void</span> printParams(<span class="keywordtype">string</span> algo) {
<a name="l00627"></a>00627     cout &lt;&lt; endl;
<a name="l00628"></a>00628     cout &lt;&lt; <span class="stringliteral">&quot;+--------------------------------------------------------+&quot;</span> &lt;&lt; endl;
<a name="l00629"></a>00629     cout &lt;&lt; <span class="stringliteral">&quot;| Algo Parameters                                        |&quot;</span> &lt;&lt; endl;
<a name="l00630"></a>00630     cout &lt;&lt; <span class="stringliteral">&quot;| Not all are relevant when feeding in allele summaries  |&quot;</span> &lt;&lt; endl;
<a name="l00631"></a>00631     cout &lt;&lt; <span class="stringliteral">&quot;+--------------------------------------------------------+&quot;</span> &lt;&lt; endl;
<a name="l00632"></a>00632     cout &lt;&lt; endl;
<a name="l00633"></a>00633 
<a name="l00634"></a>00634     <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> gFactory(QuantMethodFactory::GenoType);
<a name="l00635"></a>00635     vector&lt;vector&lt;SelfDoc&gt; &gt; docs;
<a name="l00636"></a>00636     docs.push_back(gFactory.getDocs());
<a name="l00637"></a>00637 
<a name="l00638"></a>00638     <span class="keywordflow">for</span>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; docs.size(); i++) {
<a name="l00639"></a>00639       <span class="keywordflow">for</span>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> j = 0; j &lt; docs[i].size(); j++) {
<a name="l00640"></a>00640         <span class="keywordflow">if</span>(docs[i][j].getDocName() == algo) {
<a name="l00641"></a>00641             <a class="code" href="classSelfDoc.html#a99b47e52dc1c81e34a79ed5f71ac55ef" title="Write out a self documenter name, description, and all of its parameters.">SelfDoc::printExplanation</a>(docs[i][j], cout);
<a name="l00642"></a>00642         }
<a name="l00643"></a>00643       }
<a name="l00644"></a>00644     }
<a name="l00645"></a>00645 }
<a name="l00646"></a>00646 <span class="comment"></span>
<a name="l00647"></a>00647 <span class="comment">/** Everbody&#39;s favorite function... */</span>
<a name="l00648"></a><a class="code" href="multiChannelExample_8cpp.html#ac0f2228420376f4db7e1274f2b41667c">00648</a> <span class="keywordtype">int</span> <a class="code" href="canary-test_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="Everybody&amp;#39;s favorite function.">main</a>(<span class="keywordtype">int</span> argc, <span class="keyword">const</span> <span class="keywordtype">char</span> *argv[]) {
<a name="l00649"></a>00649   <span class="keyword">const</span> <span class="keywordtype">string</span> version (<span class="stringliteral">&quot;NON-OFFICIAL-RELEASE&quot;</span>);
<a name="l00650"></a>00650   <span class="keyword">const</span> <span class="keywordtype">string</span> cvsId (<span class="stringliteral">&quot;$Id: multiChannelExample.cpp,v 1.34 2009-10-30 07:20:38 csugne Exp $&quot;</span>);
<a name="l00651"></a>00651   ofstream logOut;
<a name="l00652"></a>00652   <span class="keywordtype">string</span> logName;
<a name="l00653"></a>00653 
<a name="l00654"></a>00654   <a class="code" href="classAOptions.html">AOptions</a> o;
<a name="l00655"></a>00655   <a class="code" href="classPgOptions.html">PgOptions</a> *pgOpts = NULL;
<a name="l00656"></a>00656   pgOpts = <span class="keyword">new</span> <a class="code" href="classPgOptions.html">PgOptions</a>();
<a name="l00657"></a>00657   define_options(pgOpts);
<a name="l00658"></a>00658   pgOpts-&gt;<a class="code" href="classPgOptions.html#afb2acbc339ff978eee109450c7b4bc07" title="Match the command line arguments from ///.">parseArgv</a>(argv);
<a name="l00659"></a>00659 
<a name="l00660"></a>00660   <span class="keywordflow">if</span>(pgOpts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;help&quot;</span>) == <span class="keyword">true</span> || argc == 1)
<a name="l00661"></a>00661   {
<a name="l00662"></a>00662     pgOpts-&gt;<a class="code" href="classPgOptions.html#ada18bd8d5b64cf07665e679cc85875c0" title="Print out a litte ditty about program and its usage. ///.">usage</a>();
<a name="l00663"></a>00663     cout &lt;&lt; <span class="stringliteral">&quot;version:&quot;</span> &lt;&lt; endl;
<a name="l00664"></a>00664     cout &lt;&lt; <span class="stringliteral">&quot;   &quot;</span> &lt;&lt; version &lt;&lt; endl;
<a name="l00665"></a>00665     cout &lt;&lt; <span class="stringliteral">&quot;   &quot;</span> &lt;&lt; cvsId &lt;&lt; endl;
<a name="l00666"></a>00666 
<a name="l00667"></a>00667   }
<a name="l00668"></a>00668   <span class="keywordflow">else</span> <span class="keywordflow">if</span>(pgOpts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;explain&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>)
<a name="l00669"></a>00669     printParams(pgOpts-&gt;<a class="code" href="classPgOptions.html#a43eb777dd82aabdee1bc766b5b98ca5b" title="Get the value of the option as a string. ///.">get</a>(<span class="stringliteral">&quot;explain&quot;</span>));
<a name="l00670"></a>00670   <span class="keywordflow">else</span> <span class="keywordflow">if</span>(pgOpts-&gt;<a class="code" href="classPgOptions.html#a2558435fb2edc0b54c22e320f637e812" title="Get the value of the option as a boolean. ///.">getBool</a>(<span class="stringliteral">&quot;version&quot;</span>))
<a name="l00671"></a>00671     cout &lt;&lt; <span class="stringliteral">&quot;version: &quot;</span> &lt;&lt; version &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; cvsId &lt;&lt; endl;
<a name="l00672"></a>00672   <span class="keywordflow">else</span> {
<a name="l00673"></a>00673     <a class="code" href="apt-cel-extract_8cpp.html#a1518275cea55ffab3a9dceea1dbd00f4" title="Fill in options given command line args.">fillInOptions</a>(pgOpts, o);
<a name="l00674"></a>00674     logName = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(o.outDir,<span class="stringliteral">&quot;multiChannelExample.log&quot;</span>);
<a name="l00675"></a>00675     <a class="code" href="classFs.html#a1cc6e82074c84dd92bd7cdcf7d374b7c" title="Open an ofstream for writing to.">Fs::mustOpenToWrite</a>(logOut, logName);
<a name="l00676"></a>00676     <a class="code" href="classLogStream.html" title="Log messages with timestamps.">LogStream</a> log(3, &amp;logOut);
<a name="l00677"></a>00677     Verbose::pushMsgHandler(&amp;log);
<a name="l00678"></a>00678     <a class="code" href="classVerbose.html#af20e052bc19f95d3a949d1daa1647fa8" title="Functions to add and remove handlers for communcation functions.">Verbose::pushProgressHandler</a>(&amp;log);
<a name="l00679"></a>00679     Verbose::pushWarnHandler(&amp;log);
<a name="l00680"></a>00680     <a class="code" href="classVerbose.html#a416deece7d5098dc2e8759ef5fb134e3" title="Set the level of verbosity desired.">Verbose::setLevel</a>(o.verbose);
<a name="l00681"></a>00681     <a class="code" href="multiChannelExample_8cpp.html#ad561ebe590e97e4549466bd1d3333f51">callGenotypes</a>(o);
<a name="l00682"></a>00682   }
<a name="l00683"></a>00683   <span class="keywordflow">return</span> 0;
<a name="l00684"></a>00684 }
<a name="l00685"></a>00685     
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:51 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
