<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: MANUAL: multiChannelExample (1.18.2)</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>MANUAL: multiChannelExample (1.18.2) </h1>  </div>
</div>
<div class="contents">
<p><a class="anchor" id="options"></a></p>
<div class="fragment"><pre class="fragment">multiChannelExample - example of feeding multiple channels into a
ChipStream for genotype calling.

usage:
   ./multiChannelExample -p CM=1.bins=100.K=2.SB=0.003.MS=0.05 \
       -o results \
       --read-models-brlmmp GenomeWideSNP_5.models \
       --chrX-snps GenomeWideSNP_5.chrx \
       --read-genders genders.txt


options:
   -h, --help                           This message. [default 'false'] 
     --explain Display documentation for algo method
                          parameters. [default ''] 
   -v, --verbose How verbose to be with status messages 0 -
                          quiet, 1 - usual messages, 2 - more
                          messages. [default '1'] 
     --cel-files File listing CEL files to process. There
                          must be a 'sample_name' column, and 2 or
                          more 'channel_*'. Current channels are
                          ordered (starting with 0) by the order in
                          this file. [default ''] 
   -d, --cdf-file CDF file providing chip layout information.
                          [default ''] 
     --spf-file SPF file providing chip layout information.
                          [default ''] 
     --channel-file Channel file providing info on what alleles
                          are in what channel for each SNP. There
                          should be a row for each SNP. There must be
                          a 'probeset_id', 'allele_A_channel', and an
                          'allele_B_channel' columns. The value
                          corresponds to the channels in the
                          --cel-files specified file. Valid values 
                          are integers &gt;= 0. [default ''] 
   -p, --algo-params A string specifying algo parameters. See
                          --explain option for acceptable parameters.
                          [default ''] 
   -a, --expr-analysis Expression analysis string. [default ''] 
   -o, --out-dir Directory to write result files into.
                          [default '.'] 
     --version Display version information. [default
                          'false'] 
     --chrX-snps File containing snps on chrX
                          (non-pseudoautosomal region). [default ''] 
     --special-snps File containing all snps of unusual copy
                          (chrX,mito,Y) [default ''] 
     --write-models Should we write snp specific models out for
                          analysis? [experimental] [default 'false'] 
     --summaries Should we write out the allele summaries
                          [default 'false'] 
     --feature-effects Output feature effects when available.
                          [default 'false'] 
     --feature-details Output probe by chip specific details 
                          (often residuals) when available. [default
                          'false'] 
     --read-models-brlmmp File to read precomputed BRLMM-P snp
                          specific models from. [default ''] 
     --read-genders Explicitly read genders from a file.
                          [default ''] 
     --genotypes File to read seed genotypes from. [default
                          ''] 
     --select-probes evaluate probe/probe-set quality [default
                          'false'] 
     --target-sketch File specifying a target distribution to 
                          use for quantile normalization. [default 
                          ''] 
     --write-sketch Write the quantile normalization
                          distribution (or sketch) to a file for 
                          reuse with target-sketch option. [default
                          'false'] 
   NON-OFFICIAL-RELEASE
   $Id: multiChannelExample.cpp,v 1.34 2009-10-30 07:20:38 csugne Exp $
</pre></div><p>Q. Do you have a question?</p>
<p>A. No, I do not. </p>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:13 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
