<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/QuantLabelZ.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>QuantLabelZ.cpp</h1>  </div>
</div>
<div class="contents">
<a href="QuantLabelZ_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   QuantLabelZ.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Earl Hubbell</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Fri Feb 24 11:02:09 2006</span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="comment">//</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="ArtifactReduction_8h.html" title="Class for doing normalization. Can do sketch and full quantile (just set sketch to chip size) and sup...">chipstream/ArtifactReduction.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="QuantLabelZ_8h.html">chipstream/QuantLabelZ.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="comment">//</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="BioTypes_8h.html" title="Central location enumerations and other types that aren&amp;#39;t particular to a single class or applica...">chipstream/BioTypes.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;chipstream/MatrixUtil.h&quot;</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;chipstream/QuantLabelZIO.h&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodFactory_8h.html" title="Factory class for making chip streams based on a string representation.">chipstream/QuantMethodFactory.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;chipstream/GenotypeInfo.h&quot;</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00038"></a>00038 
<a name="l00039"></a>00039 <span class="keyword">using namespace </span>affx;
<a name="l00040"></a>00040 <span class="comment"></span>
<a name="l00041"></a>00041 <span class="comment">//////////</span>
<a name="l00042"></a>00042 <span class="comment"></span><span class="comment"></span>
<a name="l00043"></a>00043 <span class="comment">/**</span>
<a name="l00044"></a>00044 <span class="comment"> *  Utility function for returning subset values to place in whole sample list</span>
<a name="l00045"></a>00045 <span class="comment"> *</span>
<a name="l00046"></a>00046 <span class="comment"> * @param Target - destination for values</span>
<a name="l00047"></a>00047 <span class="comment"> * @param src - values defined for subset</span>
<a name="l00048"></a>00048 <span class="comment"> * @param subset - definition of subset of samples</span>
<a name="l00049"></a>00049 <span class="comment"> */</span>
<a name="l00050"></a>00050 <span class="keyword">template</span> &lt;<span class="keyword">typename</span> T1&gt;
<a name="l00051"></a><a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb">00051</a> <span class="keywordtype">void</span> <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(vector&lt;T1&gt; &amp;Target, vector&lt;T1&gt; &amp;src, vector&lt;int&gt; &amp;subset)
<a name="l00052"></a>00052 {
<a name="l00053"></a>00053   <span class="keywordtype">int</span> ti,i;
<a name="l00054"></a>00054   ti=0;
<a name="l00055"></a>00055   <span class="keywordflow">for</span> (i=0; i&lt;subset.size(); i++)
<a name="l00056"></a>00056   <span class="keywordflow">if</span> (subset[i]&gt;0) {
<a name="l00057"></a>00057     Target[i] = src[ti];
<a name="l00058"></a>00058     ti++;
<a name="l00059"></a>00059   }
<a name="l00060"></a>00060 }
<a name="l00061"></a>00061 <span class="comment"></span>
<a name="l00062"></a>00062 <span class="comment">/**</span>
<a name="l00063"></a>00063 <span class="comment"> *  Utility function to grab a set of samples (like male/female)</span>
<a name="l00064"></a>00064 <span class="comment"> *</span>
<a name="l00065"></a>00065 <span class="comment"> * @param target - destination for values</span>
<a name="l00066"></a>00066 <span class="comment"> * @param source - values defined for subset</span>
<a name="l00067"></a>00067 <span class="comment"> * @param subset - definition of subset of samples</span>
<a name="l00068"></a>00068 <span class="comment"> */</span>
<a name="l00069"></a>00069 <span class="keyword">template</span> &lt;<span class="keyword">typename</span> T1&gt;
<a name="l00070"></a><a class="code" href="QuantLabelZ_8cpp.html#a29242d59cf441197811d945aea09720d">00070</a> <span class="keywordtype">void</span> <a class="code" href="QuantLabelZ_8cpp.html#a29242d59cf441197811d945aea09720d" title="Utility function to grab a set of samples (like male/female).">t_extract_subset</a>(std::vector&lt;T1&gt; &amp;target, std::vector&lt;T1&gt; &amp;source, std::vector&lt;int&gt; &amp;subset)
<a name="l00071"></a>00071 {
<a name="l00072"></a>00072   <span class="keywordtype">int</span> i,ti;
<a name="l00073"></a>00073 
<a name="l00074"></a>00074   ti=0;
<a name="l00075"></a>00075   <span class="keywordflow">for</span> (i=0; i&lt;source.size(); i++) {
<a name="l00076"></a>00076     <span class="keywordflow">if</span> (subset[i]&gt;0) {
<a name="l00077"></a>00077       target[ti] = source[i];
<a name="l00078"></a>00078       ti++;
<a name="l00079"></a>00079     }
<a name="l00080"></a>00080   }
<a name="l00081"></a>00081 }
<a name="l00082"></a>00082 <span class="comment"></span>
<a name="l00083"></a>00083 <span class="comment">//////////</span>
<a name="l00084"></a>00084 <span class="comment"></span><span class="comment"></span>
<a name="l00085"></a>00085 <span class="comment">//////////</span>
<a name="l00086"></a>00086 <span class="comment"></span>
<a name="l00087"></a><a class="code" href="classQuantLabelZ.html#a2cea6fc92475386eeca1511393bc6af4">00087</a> <a class="code" href="classQuantLabelZ.html#a2cea6fc92475386eeca1511393bc6af4" title="probability of BB call">QuantLabelZ::QuantLabelZ</a>(<span class="keyword">enum</span> <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> transform,
<a name="l00088"></a>00088                          <span class="keywordtype">double</span> K,
<a name="l00089"></a>00089                          <span class="keywordtype">bool</span> lowPrecision) 
<a name="l00090"></a>00090 {
<a name="l00091"></a>00091   <span class="keywordflow">if</span> (transform &gt; <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1a0af5537cdc9498e453464b9ac2e4f97f" title="Contrast Centers Stretch (nee alternative south san francisco).">CCS</a> || transform &lt; <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1a6db9488b39205827082e765aa0e87297" title="Minus Vs Average.">MvA</a>) {
<a name="l00092"></a>00092     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantLabelZ() - Expecting transform to be between: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1a6db9488b39205827082e765aa0e87297" title="Minus Vs Average.">MvA</a>) + <span class="stringliteral">&quot; (MvA) and: &quot;</span> + 
<a name="l00093"></a>00093                   <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1a0af5537cdc9498e453464b9ac2e4f97f" title="Contrast Centers Stretch (nee alternative south san francisco).">CCS</a>) + <span class="stringliteral">&quot; (CCS), but got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(transform));
<a name="l00094"></a>00094   }
<a name="l00095"></a>00095   m_bCopyNumber = <span class="keyword">false</span>;
<a name="l00096"></a>00096   <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a> = lowPrecision;
<a name="l00097"></a>00097   <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a> = transform;
<a name="l00098"></a>00098   <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e" title="Scaling Parameter to use in transformations.">m_K</a> = K;
<a name="l00099"></a>00099   em_thresh = 0.05;
<a name="l00100"></a>00100   em_cutoff = 0.5;
<a name="l00101"></a>00101   gender_cutoff = 0.1;
<a name="l00102"></a>00102   <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a> = NULL;
<a name="l00103"></a>00103   m_SelectionOverride = 0; <span class="comment">// do not override values for selection</span>
<a name="l00104"></a>00104   <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>.<a class="code" href="classsnp__param.html#a956b62e64130e43f407662c31f54b372" title="initialize to some useful defaults">Initialize</a>();
<a name="l00105"></a>00105   m_KnownGenoTypes = NULL;
<a name="l00106"></a>00106   m_SnpPosteriorTsv_ver=-1;
<a name="l00107"></a>00107   m_ProbeSetsToReport = NULL;
<a name="l00108"></a>00108   m_ZWGenderCalling = <span class="keyword">false</span>;
<a name="l00109"></a>00109 
<a name="l00110"></a>00110   <a class="code" href="classQuantLabelZ.html#acfd3b0ca4567d0d580602e974bd8de7a" title="Fill in our self documentation from current state.">setupSelfDoc</a>(*<span class="keyword">this</span>);
<a name="l00111"></a>00111   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;transform&quot;</span>, <a class="code" href="GenoUtility_8cpp.html#a119c3e97f150b0b9b91dea503b19c424" title="Convert a transformation enum to string representation.">stringForTransformation</a>(<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>));
<a name="l00112"></a>00112   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;K&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e" title="Scaling Parameter to use in transformations.">m_K</a>));
<a name="l00113"></a>00113   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;lowprecision&quot;</span>, <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a>);
<a name="l00114"></a>00114   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;copynumber&quot;</span>, m_bCopyNumber);
<a name="l00115"></a>00115   <span class="comment">// set some bogus filenames for debugging.</span>
<a name="l00116"></a>00116   m_NormSummaryTsv.setFilename(<span class="stringliteral">&quot;m_NormSummaryTsv-DEBUG&quot;</span>);
<a name="l00117"></a>00117   m_SnpPosteriorTsv.setFilename(<span class="stringliteral">&quot;m_SnpPosteriorTsv-DEBUG&quot;</span>);
<a name="l00118"></a>00118 }
<a name="l00119"></a>00119   <span class="comment"></span>
<a name="l00120"></a>00120 <span class="comment">/** Destructor. */</span>
<a name="l00121"></a><a class="code" href="classQuantLabelZ.html#a67181557019fd8cd7600c57f8b989bac">00121</a> <a class="code" href="classQuantLabelZ.html#a67181557019fd8cd7600c57f8b989bac" title="Destructor.">QuantLabelZ::~QuantLabelZ</a>()
<a name="l00122"></a>00122 {
<a name="l00123"></a>00123   <span class="keywordflow">if</span> (m_SequentialModelTsv.is_open()) 
<a name="l00124"></a>00124     m_SequentialModelTsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00125"></a>00125   <span class="keywordflow">if</span> (m_NormSummaryTsv.is_open())
<a name="l00126"></a>00126     m_NormSummaryTsv.close();
<a name="l00127"></a>00127   <span class="keywordflow">if</span> (m_SnpPosteriorTsv.is_open())
<a name="l00128"></a>00128     m_SnpPosteriorTsv.close();
<a name="l00129"></a>00129   <span class="keywordflow">if</span> (m_SnpProbeTsv.is_open())
<a name="l00130"></a>00130     m_SnpProbeTsv.close();
<a name="l00131"></a>00131   <span class="comment">//</span>
<a name="l00132"></a>00132   <span class="keywordflow">for</span> (std::vector&lt;QuantMethodReport *&gt;::iterator it = <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>.begin(); it != <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>.end(); it++)
<a name="l00133"></a>00133     {
<a name="l00134"></a>00134       <span class="keyword">delete</span> *it;
<a name="l00135"></a>00135     }
<a name="l00136"></a>00136   <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>.clear();
<a name="l00137"></a>00137   m_vectSnpPriors.deleteAll();
<a name="l00138"></a>00138   <span class="keywordflow">if</span> (<a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a> != NULL) {<span class="keyword">delete</span> <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>; <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a> = NULL;}
<a name="l00139"></a>00139 }
<a name="l00140"></a>00140 
<a name="l00141"></a><a class="code" href="classQuantLabelZ.html#a1c8b961309e6485648634d5884e6bd60">00141</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a1c8b961309e6485648634d5884e6bd60" title="Get set up for a run of reporting probesets.">QuantLabelZ::prepare</a>(<span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart)
<a name="l00142"></a>00142 {
<a name="l00143"></a>00143   <span class="keywordtype">bool</span> result = <span class="keyword">true</span>;
<a name="l00144"></a>00144   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;<a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>.size(); i++)
<a name="l00145"></a>00145     result = result &amp;&amp; <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>[i]-&gt;<a class="code" href="classQuantLabelZ.html#a1c8b961309e6485648634d5884e6bd60" title="Get set up for a run of reporting probesets.">prepare</a>(*<span class="keyword">this</span>, iMart);
<a name="l00146"></a>00146 
<a name="l00147"></a>00147   <span class="comment">//</span>
<a name="l00148"></a>00148   <span class="keywordflow">if</span> (!m_SnpPosteriorName.empty()) {
<a name="l00149"></a>00149     std::vector&lt;std::string&gt;::const_iterator keyIx, paramIx;
<a name="l00150"></a>00150     <span class="comment">//</span>
<a name="l00151"></a>00151     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Got &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Info.m_ParamValues.size()) + <span class="stringliteral">&quot; headers to write.&quot;</span>);
<a name="l00152"></a>00152     m_SnpPosteriorTsv.addHeader(<span class="stringliteral">&quot;affymetrix-algorithm-param-probeset_count&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Info.<a class="code" href="classAnalysisInfo.html#ac85efbbb6618cb87c39511af32c311a7" title="How many probesets are on the chip and in cdf file. This is /// critical as probesets must be the exa...">m_NumProbeSets</a>));
<a name="l00153"></a>00153     <span class="keywordflow">for</span> (keyIx = m_Info.<a class="code" href="classAnalysisInfo.html#ad48864dcd9e2e22d9d74dfbbb00fff04" title="These next two vectors are matched pairs, one contains the /// keys and the other the values...">m_ParamNames</a>.begin(), paramIx = m_Info.m_ParamValues.begin();
<a name="l00154"></a>00154         keyIx != m_Info.<a class="code" href="classAnalysisInfo.html#ad48864dcd9e2e22d9d74dfbbb00fff04" title="These next two vectors are matched pairs, one contains the /// keys and the other the values...">m_ParamNames</a>.end() &amp;&amp; paramIx != m_Info.m_ParamValues.end();
<a name="l00155"></a>00155         ++keyIx, ++paramIx) {
<a name="l00156"></a>00156       <span class="comment">// @todo should we using the AffymetrixParameterConsts.h #defined values?</span>
<a name="l00157"></a>00157       m_SnpPosteriorTsv.addHeader(<span class="stringliteral">&quot;affymetrix-algorithm-param-&quot;</span> + *keyIx, *paramIx);
<a name="l00158"></a>00158     }
<a name="l00159"></a>00159     <span class="keywordflow">for</span> (keyIx = m_Info.<a class="code" href="classAnalysisInfo.html#ac8ac9d2dfbbcfc1f4f94abba513bcc94" title="Two vectors for client supplied meta-data. Matched pairs of vectors, one contains the /// keys and th...">m_ClientInfoNames</a>.begin(), paramIx = m_Info.m_ClientInfoValues.begin();
<a name="l00160"></a>00160         keyIx != m_Info.<a class="code" href="classAnalysisInfo.html#ac8ac9d2dfbbcfc1f4f94abba513bcc94" title="Two vectors for client supplied meta-data. Matched pairs of vectors, one contains the /// keys and th...">m_ClientInfoNames</a>.end() &amp;&amp; paramIx != m_Info.m_ClientInfoValues.end();
<a name="l00161"></a>00161       ++keyIx, ++paramIx) {
<a name="l00162"></a>00162       <span class="comment">// @todo should we using the AffymetrixParameterConsts.h #defined values?</span>
<a name="l00163"></a>00163       m_SnpPosteriorTsv.addHeader(<span class="stringliteral">&quot;affymetrix-application-meta-data-info-&quot;</span> + *keyIx, *paramIx);
<a name="l00164"></a>00164     }
<a name="l00165"></a>00165     QuantLabelZ__writeSnpPosteriorTsv(m_SnpPosteriorTsv,m_SnpPosteriorName,m_PosteriorFormat,m_SnpPosteriorTsv_ver); 
<a name="l00166"></a>00166   }
<a name="l00167"></a>00167   <span class="keywordflow">return</span> result;
<a name="l00168"></a>00168 }
<a name="l00169"></a>00169 
<a name="l00170"></a><a class="code" href="classQuantLabelZ.html#a259ef2cb352e7e9911bff1e0efbf3c13">00170</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a259ef2cb352e7e9911bff1e0efbf3c13" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">QuantLabelZ::finish</a>() 
<a name="l00171"></a>00171 {
<a name="l00172"></a>00172   <span class="keywordtype">bool</span> result = <span class="keyword">true</span>;
<a name="l00173"></a>00173   m_SnpPosteriorTsv.close();
<a name="l00174"></a>00174   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;<a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>.size(); i++)
<a name="l00175"></a>00175     result = result &amp;&amp; <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>[i]-&gt;<a class="code" href="classQuantLabelZ.html#a259ef2cb352e7e9911bff1e0efbf3c13" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">finish</a>(*<span class="keyword">this</span>);
<a name="l00176"></a>00176   <span class="keywordflow">return</span> result;
<a name="l00177"></a>00177 }
<a name="l00178"></a>00178 <span class="comment"></span>
<a name="l00179"></a>00179 <span class="comment">/** Fill in our self documentation from current state. */</span>
<a name="l00180"></a><a class="code" href="classQuantLabelZ.html#acfd3b0ca4567d0d580602e974bd8de7a">00180</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#acfd3b0ca4567d0d580602e974bd8de7a" title="Fill in our self documentation from current state.">QuantLabelZ::setupSelfDoc</a>(<a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> &amp;doc) {
<a name="l00181"></a>00181   doc.<a class="code" href="classSelfDoc.html#a51db8ea9abfa958f92fb5bc601ad68dc" title="Setter method for name appearing in doc.">setDocName</a>(<a class="code" href="QuantLabelZ_8h.html#a7225f2c577370e0fcaa1de762b340288" title="String describing quant method.">QUANTBRLMMP</a>);
<a name="l00182"></a>00182   doc.<a class="code" href="classSelfDoc.html#acd4638b447701d1a642effb6b8d2a396" title="Setter method for description appearing in doc.">setDocDescription</a>(<span class="stringliteral">&quot;Do genotyping calls with BRLMM-P (perfect match) algorithm.&quot;</span>);
<a name="l00183"></a>00183   doc.<a class="code" href="classSelfDoc.html#aa43e6721e35792413711e07bda90861d" title="Setter method for parameters and their documentation.">setDocOptions</a>(<a class="code" href="classQuantLabelZ.html#a0da80e2dbed413e783338e1ae66c8b43" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>());
<a name="l00184"></a>00184 }
<a name="l00185"></a>00185 
<a name="l00186"></a><a class="code" href="classQuantLabelZ.html#a23d0f3fb96579a159ad8d088fd7747eb">00186</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a23d0f3fb96579a159ad8d088fd7747eb" title="Clear out all of our results and values.">QuantLabelZ::blankSelf</a>() 
<a name="l00187"></a>00187 {
<a name="l00188"></a>00188   <a class="code" href="classQuantGTypeMethod.html#aff29fdf8763f00f5e1eb451162cb38c2" title="clear a probe set">clearProbeSet</a>(<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>);
<a name="l00189"></a>00189   <a class="code" href="classQuantGTypeMethod.html#aff29fdf8763f00f5e1eb451162cb38c2" title="clear a probe set">clearProbeSet</a>(<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00190"></a>00190   <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.clear();
<a name="l00191"></a>00191   <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.clear();
<a name="l00192"></a>00192   m_Calls.clear();
<a name="l00193"></a>00193   <a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>.clear();
<a name="l00194"></a>00194   <a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>.clear();
<a name="l00195"></a>00195   m_InitialCalls.clear();
<a name="l00196"></a>00196 }
<a name="l00197"></a>00197 <span class="comment"></span>
<a name="l00198"></a>00198 <span class="comment">//////////</span>
<a name="l00199"></a>00199 <span class="comment"></span><span class="comment"></span>
<a name="l00200"></a>00200 <span class="comment">/**</span>
<a name="l00201"></a>00201 <span class="comment"> * Sets up the intensities and summaries for a probe set</span>
<a name="l00202"></a>00202 <span class="comment"> * Separate function, because shared between several routines (gender, normalization, labeling)</span>
<a name="l00203"></a>00203 <span class="comment"> * Ideally, would be inherited from global genotyping class</span>
<a name="l00204"></a>00204 <span class="comment"> *</span>
<a name="l00205"></a>00205 <span class="comment"> * @param gtPs - probe set to be set up</span>
<a name="l00206"></a>00206 <span class="comment"> * @param layout - probes located on chip</span>
<a name="l00207"></a>00207 <span class="comment"> * @param iMart - serves up raw intensities</span>
<a name="l00208"></a>00208 <span class="comment"> * @param iTrans - transforms intensities</span>
<a name="l00209"></a>00209 <span class="comment"> * @param pmAdjust - background adjust</span>
<a name="l00210"></a>00210 <span class="comment"> * @param doReport - throw summaries out to reporters or not</span>
<a name="l00211"></a>00211 <span class="comment"> */</span>
<a name="l00212"></a><a class="code" href="classQuantLabelZ.html#a6c51a62f1bcd9d8d9b357bc9db118620">00212</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a6c51a62f1bcd9d8d9b357bc9db118620" title="Extracts intensities, etc for a probe set, so further analysis can proceed.">QuantLabelZ::SetUpProbeSet</a>(<span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs,
<a name="l00213"></a>00213                                 <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, std::vector&lt;ChipStream *&gt; &amp;iTrans,
<a name="l00214"></a>00214                                 <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust, <span class="keywordtype">bool</span> doReport){
<a name="l00215"></a>00215 
<a name="l00216"></a>00216   <span class="keywordtype">bool</span> success=<span class="keyword">true</span>;
<a name="l00217"></a>00217 
<a name="l00218"></a>00218   <span class="keywordflow">if</span> (!<a class="code" href="classQuantGTypeMethod.html#a71d6dfe5d15e5d77354f770913c23824" title="Can the ProbeSet set up the intensities and summaries.">canSetUpProbeSet</a>(gtPs)) {<span class="keywordflow">return</span> <span class="keyword">false</span>;}
<a name="l00219"></a>00219   <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> = gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>;
<a name="l00220"></a>00220   <a class="code" href="classQuantGTypeMethod.html#a447dc22d0ca6a8c52c522f1f7eddfa41" title="Genotyping probeset currently using.">m_GtProbeSet</a> = gtPs;
<a name="l00221"></a>00221   <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.clear();
<a name="l00222"></a>00222   <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.clear();
<a name="l00223"></a>00223 
<a name="l00224"></a>00224   <span class="keywordtype">bool</span> report = doReport;
<a name="l00225"></a>00225   <span class="keywordflow">if</span> (m_ProbeSetsToReport != NULL) {
<a name="l00226"></a>00226     <span class="keywordflow">if</span> (m_ProbeSetsToReport-&gt;find(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>.c_str())==m_ProbeSetsToReport-&gt;end()) {
<a name="l00227"></a>00227       report = <span class="keyword">false</span>;
<a name="l00228"></a>00228     }
<a name="l00229"></a>00229   }
<a name="l00230"></a>00230   <a class="code" href="classQuantGTypeMethod.html#a1a853c2d2821e9c4930fb5c3807f36f9" title="fill in the probe set with the allele values">fillInAlleleProbeSets</a>(*gtPs, <a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00231"></a>00231   <span class="comment">// if by some mischance we have failed to specify method</span>
<a name="l00232"></a>00232   <span class="keywordflow">if</span> (<a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a> == NULL) {
<a name="l00233"></a>00233     <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> factory(QuantMethodFactory::Expression);
<a name="l00234"></a>00234     <span class="keywordtype">string</span> spec = <span class="stringliteral">&quot;plier.optmethod=1&quot;</span>;
<a name="l00235"></a>00235     <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> layout;
<a name="l00236"></a>00236     <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a> = factory.<a class="code" href="classQuantMethodFactory.html#abfb5fe4b06ae6e00419c35d55885feed" title="Factory for creating QuantMethod from string representation.">quantExprMethodForString</a>(spec, layout, QuantMethodFactory::Expression);
<a name="l00237"></a>00237   }
<a name="l00238"></a>00238   <span class="comment">// load up the data</span>
<a name="l00239"></a>00239   success &amp;= <a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, iMart, iTrans, pmAdjust, <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>, <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a>, report,<a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>);
<a name="l00240"></a>00240   <span class="keywordflow">if</span> (gtPs-&gt;<a class="code" href="classProbeSet.html#aa3e4170184169cc3bee6d958d74940ba" title="What type of probeset is this?">psType</a> == ProbeSet::Copynumber) {<span class="keywordflow">return</span> <span class="keyword">false</span>;}
<a name="l00241"></a>00241   <span class="keywordflow">if</span> (<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>.<a class="code" href="classProbeSet.html#ab92eb06095441bf193a6fce3f758540b" title="Atoms that make up this probe set.">atoms</a>.size() &lt; 1) {
<a name="l00242"></a>00242     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00243"></a>00243   }
<a name="l00244"></a>00244   success &amp;= <a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, iMart, iTrans, pmAdjust, <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>, <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a>, report,<a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>);
<a name="l00245"></a>00245 
<a name="l00246"></a>00246   <span class="comment">// grab probe values if doing probe selection</span>
<a name="l00247"></a>00247   <span class="keywordflow">if</span> (m_SnpProbeTsv.is_open()){
<a name="l00248"></a>00248     <span class="comment">//do this again, because summarizeAllele clears the probeset m_Aallele and m_Ballele</span>
<a name="l00249"></a>00249     <a class="code" href="classQuantGTypeMethod.html#a1a853c2d2821e9c4930fb5c3807f36f9" title="fill in the probe set with the allele values">fillInAlleleProbeSets</a>(*gtPs, <a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00250"></a>00250     <a class="code" href="classQuantGTypeMethod.html#ab62a74a7fc8feaaff51d96496b041be9" title="extract the probes for an allele specified">extractProbes</a>(m_PM_A, &amp;<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, m_A_probeId, iMart,iTrans,pmAdjust,<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a>,<a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>);
<a name="l00251"></a>00251     <a class="code" href="classQuantGTypeMethod.html#ab62a74a7fc8feaaff51d96496b041be9" title="extract the probes for an allele specified">extractProbes</a>(m_PM_B, &amp;<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>, m_B_probeId, iMart,iTrans,pmAdjust,<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a>,<a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>);
<a name="l00252"></a>00252   }
<a name="l00253"></a>00253 
<a name="l00254"></a>00254   <span class="comment">// transform the data as specified</span>
<a name="l00255"></a>00255   <a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909" title="Transform allele estimates.">transform</a>();
<a name="l00256"></a>00256   <a class="code" href="classQuantLabelZ.html#aa0bb7cf31d9f77d47947cf91d2fa75e0" title="Load up the initial call for the requested SNP.">loadInitialCallsForSnp</a>(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>);
<a name="l00257"></a>00257   <span class="keywordflow">if</span> ( m_bProbeSetTrustOpt &amp;&amp; m_ProbeSetTrust.empty() ) {
<a name="l00258"></a>00258     loadProbeSetTrustFromTmpFile();
<a name="l00259"></a>00259   }
<a name="l00260"></a>00260   <span class="keywordflow">return</span>(success);
<a name="l00261"></a>00261 }
<a name="l00262"></a>00262 <span class="comment"></span>
<a name="l00263"></a>00263 <span class="comment">/**</span>
<a name="l00264"></a>00264 <span class="comment"> * Load up the initial calls for SNP</span>
<a name="l00265"></a>00265 <span class="comment"> */</span>
<a name="l00266"></a><a class="code" href="classQuantLabelZ.html#aa0bb7cf31d9f77d47947cf91d2fa75e0">00266</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aa0bb7cf31d9f77d47947cf91d2fa75e0" title="Load up the initial call for the requested SNP.">QuantLabelZ::loadInitialCallsForSnp</a>(<span class="keyword">const</span> <span class="keywordtype">string</span> &amp;probeset) {
<a name="l00267"></a>00267   <span class="comment">// setup hints for genotype</span>
<a name="l00268"></a>00268   m_InitialCalls.clear();
<a name="l00269"></a>00269   <span class="keywordflow">if</span> (m_KnownGenoTypes != NULL) {
<a name="l00270"></a>00270     KnownGenoTypes_iter_t mapIter = m_KnownGenoTypes-&gt;find(probeset.c_str());
<a name="l00271"></a>00271     <span class="comment">//  std::map&lt;string, vector&lt;affx::GType&gt; &gt;::iterator mapIter = m_KnownGenoTypes.find(gtPs-&gt;name);</span>
<a name="l00272"></a>00272     <span class="keywordflow">if</span> (mapIter == m_KnownGenoTypes-&gt;end()) {
<a name="l00273"></a>00273         <span class="comment">// unlike BRLMM, this is not &gt;necessary&lt; for analysis</span>
<a name="l00274"></a>00274         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(4,<span class="stringliteral">&quot;Did not find known genotype for probeset &quot;</span> + probeset);
<a name="l00275"></a>00275     }
<a name="l00276"></a>00276     <span class="keywordflow">else</span>{
<a name="l00277"></a>00277         vector&lt;affx::GType&gt; &amp;gType = mapIter-&gt;second;
<a name="l00278"></a>00278         <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; gType.size(); i++) {
<a name="l00279"></a>00279         m_InitialCalls.push_back(gType[i]);
<a name="l00280"></a>00280         }
<a name="l00281"></a>00281     }
<a name="l00282"></a>00282   }
<a name="l00283"></a>00283 }
<a name="l00284"></a>00284 
<a name="l00285"></a>00285 <span class="keywordtype">void</span> QuantLabelZ::loadProbeSetTrustFromTmpFile(){
<a name="l00286"></a>00286   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00287"></a>00287   std::string chpId;
<a name="l00288"></a>00288   std::string probesetId;
<a name="l00289"></a>00289   <span class="keywordtype">int</span> trustedCount;
<a name="l00290"></a>00290   
<a name="l00291"></a>00291   <span class="comment">// Aborts on error. </span>
<a name="l00292"></a>00292   <span class="keywordflow">if</span> ( tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(m_ProbeSetTrustTmpFileName) != affx::TSV_OK ) {
<a name="l00293"></a>00293     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_ProbeSetTrustTmpFileName + <span class="stringliteral">&quot; file not found or permission denied&quot;</span>);
<a name="l00294"></a>00294   }
<a name="l00295"></a>00295 
<a name="l00296"></a>00296   tsv.bind(0, <span class="stringliteral">&quot;chp_id&quot;</span>, &amp;chpId,   <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00297"></a>00297   tsv.bind(0, <span class="stringliteral">&quot;probeset_id&quot;</span>, &amp;probesetId,   <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00298"></a>00298   tsv.bind(0, <span class="stringliteral">&quot;trust_count&quot;</span>, &amp;trustedCount, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00299"></a>00299 
<a name="l00300"></a>00300   m_ProbeSetTrust.clear();
<a name="l00301"></a>00301   
<a name="l00302"></a>00302   <span class="keywordflow">while</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0)==TSV_OK) {
<a name="l00303"></a>00303     std::string trustKey = chpId + <span class="stringliteral">&quot;:&quot;</span> + probesetId;
<a name="l00304"></a>00304     <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>( m_ProbeSetTrust.count(trustKey) == 0,
<a name="l00305"></a>00305                     <span class="stringliteral">&quot;Invalid artifact reduction trusted probeset file, &quot;</span>
<a name="l00306"></a>00306                     + m_ProbeSetTrustTmpFileName
<a name="l00307"></a>00307                     + <span class="stringliteral">&quot;, has duplicate id: &quot;</span>
<a name="l00308"></a>00308                     + trustKey);
<a name="l00309"></a>00309     
<a name="l00310"></a>00310     m_ProbeSetTrust[trustKey] = trustedCount;
<a name="l00311"></a>00311   }
<a name="l00312"></a>00312 
<a name="l00313"></a>00313   tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00314"></a>00314 }
<a name="l00315"></a>00315 <span class="comment"></span>
<a name="l00316"></a>00316 <span class="comment">/**</span>
<a name="l00317"></a>00317 <span class="comment"> * SelectGoodProbes selects probe pairs that contribute most to classification</span>
<a name="l00318"></a>00318 <span class="comment"> */</span>
<a name="l00319"></a><a class="code" href="classQuantLabelZ.html#aee903e230d96abad58c24118b312b100">00319</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aee903e230d96abad58c24118b312b100" title="Selects good probes.">QuantLabelZ::SelectGoodProbes</a>() {
<a name="l00320"></a>00320   <span class="comment">// rescan the data to find the probes that contributed most to the output</span>
<a name="l00321"></a>00321   <span class="comment">// compare genotypes to input data</span>
<a name="l00322"></a>00322 
<a name="l00323"></a>00323   <span class="comment">// step 1: obtain all the raw probe input data from ExprMethod</span>
<a name="l00324"></a>00324   <span class="comment">// step 1 is actually done in SetUpProbeSet</span>
<a name="l00325"></a>00325   <span class="comment">// step 2: Iterate over all pairs of probes</span>
<a name="l00326"></a>00326   <span class="comment">// step 2a: construct transformed data for each pair of probes</span>
<a name="l00327"></a>00327   <span class="comment">// setp 2b: compute genotyping regression/classification against data</span>
<a name="l00328"></a>00328   <span class="comment">// step 2c: evaluate goodness of this probe pair</span>
<a name="l00329"></a>00329   <span class="comment">// step 3: dump best probe(s) to summary file with evaluation</span>
<a name="l00330"></a>00330   <span class="comment">// step 3a: dump all probes to summary file with evaluation</span>
<a name="l00331"></a>00331 
<a name="l00332"></a>00332   vector&lt;double&gt; DummyA, DummyB,Predictor;
<a name="l00333"></a>00333   vector&lt; vector&lt;double&gt; &gt; Covariate;
<a name="l00334"></a>00334   vector&lt;int&gt; RealCalls,TmpCalls;
<a name="l00335"></a>00335   vector&lt;double&gt; Summaries;
<a name="l00336"></a>00336   <span class="keywordtype">double</span> AIC;
<a name="l00337"></a>00337   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i,j,k,ValidCallNum;
<a name="l00338"></a>00338   vector&lt;double&gt; PriorM;
<a name="l00339"></a>00339   vector&lt;double&gt; PriorW;
<a name="l00340"></a>00340 
<a name="l00341"></a>00341   PriorM.resize(3);
<a name="l00342"></a>00342   PriorW.resize(3);
<a name="l00343"></a>00343   <span class="comment">// use the fitted posterior</span>
<a name="l00344"></a>00344   PriorM[0] = <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l00345"></a>00345   PriorM[1] = <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l00346"></a>00346   PriorM[2] = <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l00347"></a>00347   PriorW[0]=PriorW[1]=PriorW[2] = 0.01; <span class="comment">// 1 percent data point if no obs</span>
<a name="l00348"></a>00348 
<a name="l00349"></a>00349   RealCalls.resize(m_Calls.size());
<a name="l00350"></a>00350   <span class="keywordflow">for</span> (i=0; i&lt;RealCalls.size(); i++) {
<a name="l00351"></a>00351      <span class="comment">// make sure this is &quot;int&quot; worthy and 0-2 range</span>
<a name="l00352"></a>00352       RealCalls[i] = <a class="code" href="namespaceaffx.html#ac9e9039b462ebd2bec646ed4c5ff1122" title="Convert a GType enum to an external integer value ///.">affx::GType_to_int</a>(m_Calls[i]);
<a name="l00353"></a>00353    }
<a name="l00354"></a>00354   <span class="keywordflow">for</span> (i=0; i&lt;m_InitialCalls.size(); i++) {
<a name="l00355"></a>00355       <span class="keywordflow">if</span> (m_InitialCalls[i]!=<a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a1cef2fc4697bd5fc67116fbd1ee2027b" title="No call, unknown genotype. The internal is value is &amp;#39;3&amp;#39;; external is &amp;#39;-1&amp;#39;,.">affx::NN</a> &amp;&amp; m_SelectionOverride)
<a name="l00356"></a>00356         <span class="comment">// make sure this is &quot;int&quot; worthy and 0-2 range</span>
<a name="l00357"></a>00357         RealCalls[i] = <a class="code" href="namespaceaffx.html#ac9e9039b462ebd2bec646ed4c5ff1122" title="Convert a GType enum to an external integer value ///.">affx::GType_to_int</a>(m_InitialCalls[i]);
<a name="l00358"></a>00358    }
<a name="l00359"></a>00359 
<a name="l00360"></a>00360   <span class="comment">// now I have a set of calls, some of which may be nonexistent</span>
<a name="l00361"></a>00361   <span class="comment">// filter the data by valid calls</span>
<a name="l00362"></a>00362   ValidCallNum=0;
<a name="l00363"></a>00363   <span class="keywordflow">for</span> (i=0; i&lt;RealCalls.size(); i++) {
<a name="l00364"></a>00364     <span class="keywordflow">if</span> (RealCalls[i]&gt;-1) {
<a name="l00365"></a>00365         ValidCallNum++;
<a name="l00366"></a>00366     }
<a name="l00367"></a>00367   }
<a name="l00368"></a>00368   <span class="keywordflow">if</span> (ValidCallNum &lt; 2) {
<a name="l00369"></a>00369     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;At least two valid calls are needed for each probeset in order to use --select-probes.  Probeset &quot;</span> + <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot; has one or zero valid calls.&quot;</span>);
<a name="l00370"></a>00370   }
<a name="l00371"></a>00371   TmpCalls.resize(ValidCallNum); <span class="comment">// how many calls Valid</span>
<a name="l00372"></a>00372   k=0;
<a name="l00373"></a>00373   <span class="keywordflow">for</span> (i=0; i&lt;RealCalls.size(); i++)
<a name="l00374"></a>00374     <span class="keywordflow">if</span> (RealCalls[i]&gt;-1){
<a name="l00375"></a>00375         TmpCalls[k] = RealCalls[i];
<a name="l00376"></a>00376         k++;
<a name="l00377"></a>00377     }
<a name="l00378"></a>00378 
<a name="l00379"></a>00379   DummyA.resize(ValidCallNum);
<a name="l00380"></a>00380   DummyB.resize(ValidCallNum);
<a name="l00381"></a>00381 
<a name="l00382"></a>00382   <span class="comment">// loop over all probes</span>
<a name="l00383"></a>00383   <span class="keywordflow">for</span> (j=0; j&lt;m_PM_A.size() &amp;&amp; (ValidCallNum&gt;0); j++) {
<a name="l00384"></a>00384     k=0;
<a name="l00385"></a>00385     <span class="keywordflow">for</span> (i=0; i&lt;m_PM_A[j].size(); i++) {
<a name="l00386"></a>00386         <span class="comment">// filter the data by valid calls</span>
<a name="l00387"></a>00387         <span class="keywordflow">if</span> (RealCalls[i]&gt;-1) {
<a name="l00388"></a>00388             DummyA[k] = m_PM_A[j][i];
<a name="l00389"></a>00389             DummyB[k] = m_PM_B[j][i];
<a name="l00390"></a>00390             k++;
<a name="l00391"></a>00391         }
<a name="l00392"></a>00392     }
<a name="l00393"></a>00393     <a class="code" href="GenoUtility_8cpp.html#a1a2531d6033d567af0446274e94f74a4" title="transform the data vector">GenoUtility_transformData</a>(DummyA, DummyB, <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>,<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e" title="Scaling Parameter to use in transformations.">m_K</a>);
<a name="l00394"></a>00394     Covariate.clear();
<a name="l00395"></a>00395     <span class="comment">//</span>
<a name="l00396"></a>00396     <span class="comment">// take out &#39;no calls&#39;</span>
<a name="l00397"></a>00397     Covariate.push_back(DummyA);
<a name="l00398"></a>00398     Summaries.clear();
<a name="l00399"></a>00399     <span class="comment">//cout &lt;&lt; endl;</span>
<a name="l00400"></a>00400     <span class="comment">//cout &lt;&lt; m_ProbesetName &lt;&lt; &quot;\t&quot;; // dummy test</span>
<a name="l00401"></a>00401     AIC = LogisticRegression(Summaries,Predictor,TmpCalls,Covariate,PriorM,PriorW,100,.001);
<a name="l00402"></a>00402 
<a name="l00403"></a>00403     <a class="code" href="classQuantLabelZ.html#a16617ac9ce4b9229c84c6f26af40cc55" title="Writeprobe writes the summaries to the file.">writeSnpProbeValue</a>(m_SnpProbeTsv,j,m_A_probeId[j],m_B_probeId[j],
<a name="l00404"></a>00404                         Summaries,Predictor,Predictor,<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>);
<a name="l00405"></a>00405   }
<a name="l00406"></a>00406 }
<a name="l00407"></a>00407 
<a name="l00408"></a>00408 <span class="comment">/*</span>
<a name="l00409"></a>00409 <span class="comment"> * @brief normalizes the data or other modifications</span>
<a name="l00410"></a>00410 <span class="comment"> */</span>
<a name="l00411"></a><a class="code" href="classQuantLabelZ.html#afa5e98cbfea742735a11e38542ed72ba">00411</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#afa5e98cbfea742735a11e38542ed72ba" title="preprocess and normalize values">QuantLabelZ::PreprocessValues</a>() {
<a name="l00412"></a>00412   <span class="comment">// transform data, do any normalization</span>
<a name="l00413"></a>00413   <span class="comment">// transform the data</span>
<a name="l00414"></a>00414   <span class="comment">// transform()</span>
<a name="l00415"></a>00415   <span class="comment">// generate calls for tranformed data</span>
<a name="l00416"></a>00416 
<a name="l00417"></a>00417   <span class="comment">// here, we apply normalization prediction across everything</span>
<a name="l00418"></a>00418 
<a name="l00419"></a>00419   <span class="comment">// AValues = Contrast</span>
<a name="l00420"></a>00420   <span class="comment">// BValues = Strength</span>
<a name="l00421"></a>00421   vector&lt;double&gt; Cov; <span class="comment">// to be covariates, offset/cubed as required</span>
<a name="l00422"></a>00422   <span class="keywordtype">double</span> tmpf, strength, contrast;
<a name="l00423"></a>00423   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> sIx; <span class="comment">// sample iterator</span>
<a name="l00424"></a>00424   <span class="keywordflow">for</span> (sIx=0; sIx&lt;normMap.size(); sIx++) {
<a name="l00425"></a>00425     <span class="comment">// isolate this, because not all transforms alike</span>
<a name="l00426"></a>00426     strength = <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>[sIx];
<a name="l00427"></a>00427     contrast = <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>[sIx];
<a name="l00428"></a>00428     <span class="comment">// make offset/cubed covariates</span>
<a name="l00429"></a>00429     normMap[sIx].MakeCov(Cov,strength);
<a name="l00430"></a>00430     <span class="comment">// generate prediction</span>
<a name="l00431"></a>00431     tmpf = normMap[sIx].MakePrediction(contrast,Cov);
<a name="l00432"></a>00432     <span class="comment">// place back in appropriate register</span>
<a name="l00433"></a>00433     <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>[sIx]=tmpf;
<a name="l00434"></a>00434   }
<a name="l00435"></a>00435 
<a name="l00436"></a>00436   <span class="comment">// write out processed values</span>
<a name="l00437"></a>00437   <span class="keywordflow">if</span> (m_NormSummaryTsv.is_open()) {
<a name="l00438"></a>00438     <a class="code" href="classQuantLabelZ.html#a4f185569daf922ace4d429c9f5c6752b" title="Writes a summary of the transformed values to a file - i.e as set up for genotyping, not expression.">writeNormSummaryValue</a>(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>,<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>,<a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>);
<a name="l00439"></a>00439   }
<a name="l00440"></a>00440 
<a name="l00441"></a>00441 }
<a name="l00442"></a>00442 
<a name="l00443"></a>00443 <span class="keywordtype">void</span> QuantLabelZ::EliminateUntrustedSamples(vector&lt;int&gt; &amp; SnpSampleCopyNumber){
<a name="l00444"></a>00444   <span class="comment">// remove any sample who doesn&#39;t have trustworthy probes</span>
<a name="l00445"></a>00445   <span class="comment">// by setting &quot;copynumber&quot; =0 in such cases</span>
<a name="l00446"></a>00446   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> sIx=0; sIx&lt;<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size(); sIx++){
<a name="l00447"></a>00447     <span class="keywordflow">if</span> (!isTrusted(sIx)){
<a name="l00448"></a>00448       SnpSampleCopyNumber[sIx]=0;
<a name="l00449"></a>00449     }
<a name="l00450"></a>00450   }
<a name="l00451"></a>00451 }
<a name="l00452"></a>00452 
<a name="l00453"></a>00453 <span class="comment">// just checks to see if this sample has any trusted probes</span>
<a name="l00454"></a>00454 <span class="keywordtype">bool</span> QuantLabelZ::isTrusted(<span class="keywordtype">int</span> cIx){
<a name="l00455"></a>00455   <span class="keywordflow">if</span> ( m_ProbeSetTrust.empty() ) {
<a name="l00456"></a>00456     <span class="keywordflow">return</span> <span class="keyword">true</span>;
<a name="l00457"></a>00457   }
<a name="l00458"></a>00458   map&lt;string,int&gt;::iterator iter;
<a name="l00459"></a>00459   <span class="keywordtype">bool</span> retval=<span class="keyword">true</span>;  <span class="comment">// return trusted if not found</span>
<a name="l00460"></a>00460   
<a name="l00461"></a>00461   <span class="comment">// only get zero block for testing</span>
<a name="l00462"></a>00462   std::string trustKey = <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(cIx)+<span class="stringliteral">&quot;:&quot;</span>+<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>; 
<a name="l00463"></a>00463   iter = m_ProbeSetTrust.find(trustKey);
<a name="l00464"></a>00464   <span class="keywordflow">if</span> (iter!=m_ProbeSetTrust.end() &amp;&amp;  (iter-&gt;second&lt;1) ) {
<a name="l00465"></a>00465     retval=<span class="keyword">false</span>;
<a name="l00466"></a>00466   }
<a name="l00467"></a>00467   <span class="keywordflow">return</span>(retval);
<a name="l00468"></a>00468 }
<a name="l00469"></a>00469 
<a name="l00470"></a>00470 <span class="comment">// void QuantLabelZ::getPrior(string &amp;TmpName, int copyIx, snp_labeled_distribution &amp;objSearch, snp_param &amp;tsp) {</span>
<a name="l00471"></a>00471 <span class="comment">//   // first search for name+copy number</span>
<a name="l00472"></a>00472 <span class="comment">//   objSearch.probeset_id = TmpName;</span>
<a name="l00473"></a>00473 <span class="comment">//   int iSearchIndex = m_vectSnpPriors.binarySearch(objSearch, 0);</span>
<a name="l00474"></a>00474 <span class="comment">//   if (iSearchIndex == -1)</span>
<a name="l00475"></a>00475 <span class="comment">//     {</span>
<a name="l00476"></a>00476 <span class="comment">//       // first fail to generic - avoid command line needed!</span>
<a name="l00477"></a>00477 <span class="comment">//       if (copyIx&lt;2)</span>
<a name="l00478"></a>00478 <span class="comment">//         {</span>
<a name="l00479"></a>00479 <span class="comment">//           objSearch.probeset_id = &quot;GENERIC:&quot; + ToStr(copyIx);</span>
<a name="l00480"></a>00480 <span class="comment">//               }</span>
<a name="l00481"></a>00481 <span class="comment">//       else</span>
<a name="l00482"></a>00482 <span class="comment">//         objSearch.probeset_id = &quot;GENERIC&quot;;</span>
<a name="l00483"></a>00483 <span class="comment">//       iSearchIndex = m_vectSnpPriors.binarySearch(objSearch,0);</span>
<a name="l00484"></a>00484 <span class="comment">//       if (iSearchIndex==-1)</span>
<a name="l00485"></a>00485 <span class="comment">//         {</span>
<a name="l00486"></a>00486           
<a name="l00487"></a>00487 <span class="comment">//           // Tell the user we tried to search for GENERIC as well but didn&#39;t find either</span>
<a name="l00488"></a>00488 <span class="comment">//           Err::errAbort(&quot;Can&#39;t find model for SNP: &quot; + TmpName + &quot; or &quot; + objSearch.probeset_id);</span>
<a name="l00489"></a>00489 <span class="comment">//         }</span>
<a name="l00490"></a>00490 <span class="comment">//     }</span>
<a name="l00491"></a>00491 <span class="comment">//   // if success, grab &amp; copy</span>
<a name="l00492"></a>00492 <span class="comment">//   if (iSearchIndex&gt;-1)</span>
<a name="l00493"></a>00493 <span class="comment">//     {</span>
<a name="l00494"></a>00494 <span class="comment">//       snp_labeled_distribution* p = m_vectSnpPriors.getAt(iSearchIndex);</span>
<a name="l00495"></a>00495 <span class="comment">//       // this is really one copy operation!!!</span>
<a name="l00496"></a>00496 <span class="comment">//       tsp.prior.Copy(p-&gt;Dist);</span>
<a name="l00497"></a>00497 <span class="comment">//       /*</span>
<a name="l00498"></a>00498 <span class="comment">//         tsp.prior.aa = p-&gt;Dist.aa;</span>
<a name="l00499"></a>00499 <span class="comment">//         tsp.prior.ab = p-&gt;Dist.ab;</span>
<a name="l00500"></a>00500 <span class="comment">//         tsp.prior.bb = p-&gt;Dist.bb;</span>
<a name="l00501"></a>00501 <span class="comment">//         tsp.prior.xah = p-&gt;Dist.xah;</span>
<a name="l00502"></a>00502 <span class="comment">//         tsp.prior.xab = p-&gt;Dist.xab;</span>
<a name="l00503"></a>00503 <span class="comment">//         tsp.prior.xhb = p-&gt;Dist.xhb;</span>
<a name="l00504"></a>00504 <span class="comment">//         tsp.prior.yah = p-&gt;Dist.yah;</span>
<a name="l00505"></a>00505 <span class="comment">//         tsp.prior.yab = p-&gt;Dist.yab;</span>
<a name="l00506"></a>00506 <span class="comment">//         tsp.prior.yhb = p-&gt;Dist.yhb;</span>
<a name="l00507"></a>00507 <span class="comment">//         tsp.prior.xyah = p-&gt;Dist.xyah;</span>
<a name="l00508"></a>00508 <span class="comment">//         tsp.prior.xyab = p-&gt;Dist.xyab;</span>
<a name="l00509"></a>00509 <span class="comment">//         tsp.prior.xyhb = p-&gt;Dist.xyhb;</span>
<a name="l00510"></a>00510 <span class="comment">//         tsp.prior.yxah = p-&gt;Dist.yxah;</span>
<a name="l00511"></a>00511 <span class="comment">//         tsp.prior.yxab = p-&gt;Dist.yxab;</span>
<a name="l00512"></a>00512 <span class="comment">//         tsp.prior.yxhb = p-&gt;Dist.yxhb;</span>
<a name="l00513"></a>00513 <span class="comment">//       */</span>
<a name="l00514"></a>00514 <span class="comment">//     }</span>
<a name="l00515"></a>00515 <span class="comment">// }</span>
<a name="l00516"></a>00516 
<a name="l00517"></a>00517 <span class="keywordtype">void</span> QuantLabelZ::getSequentialPrior(std::string &amp;name, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;tsp, <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> &amp;sDist) {
<a name="l00518"></a>00518   <span class="keywordflow">if</span> (m_SequentialModelTsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == affx::TSV_OK) {
<a name="l00519"></a>00519     QuantLabelZ__SnpPriorFromStrings(sDist, m_BB, m_AB, m_AA, m_CV);
<a name="l00520"></a>00520     sDist.probeset_id = m_Id;
<a name="l00521"></a>00521     tsp.prior.Copy(sDist.Dist);
<a name="l00522"></a>00522     <span class="keywordflow">if</span> (name != m_Id) {
<a name="l00523"></a>00523       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Expecting: &quot;</span> + name + <span class="stringliteral">&quot; but got: &quot;</span> + m_Id);
<a name="l00524"></a>00524     }
<a name="l00525"></a>00525   }
<a name="l00526"></a>00526   <span class="keywordflow">else</span> {
<a name="l00527"></a>00527     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Expecting model for snp: &quot;</span> + name);
<a name="l00528"></a>00528   }
<a name="l00529"></a>00529 }
<a name="l00530"></a>00530 
<a name="l00531"></a>00531 <span class="keywordtype">void</span> QuantLabelZ::getPrior(std::string &amp;TmpName, <span class="keywordtype">int</span> presentCopyNumber, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;tsp, <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> &amp;sDist) {
<a name="l00532"></a>00532   <span class="keywordflow">if</span> (m_SequentialModelTsv.is_open()) {
<a name="l00533"></a>00533     getSequentialPrior(TmpName, tsp, sDist);
<a name="l00534"></a>00534   }
<a name="l00535"></a>00535   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (m_vectSnpPriors.getCount() &gt; 0) {
<a name="l00536"></a>00536     <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> objSearch;
<a name="l00537"></a>00537     <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>* p=NULL; 
<a name="l00538"></a>00538     objSearch.probeset_id = TmpName;
<a name="l00539"></a>00539     <span class="keywordtype">int</span> iSearchIndex = m_vectSnpPriors.binarySearch(objSearch, 0);
<a name="l00540"></a>00540     <span class="keywordflow">if</span> (iSearchIndex == -1){
<a name="l00541"></a>00541       <span class="keywordflow">if</span> (presentCopyNumber&lt;2){
<a name="l00542"></a>00542         objSearch.probeset_id = <span class="stringliteral">&quot;GENERIC:&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(presentCopyNumber);
<a name="l00543"></a>00543       } <span class="keywordflow">else</span> {
<a name="l00544"></a>00544         objSearch.probeset_id = <span class="stringliteral">&quot;GENERIC&quot;</span>;
<a name="l00545"></a>00545       }
<a name="l00546"></a>00546       iSearchIndex = m_vectSnpPriors.binarySearch(objSearch,0);
<a name="l00547"></a>00547     } 
<a name="l00548"></a>00548     <span class="keywordflow">if</span> (iSearchIndex==-1) {
<a name="l00549"></a>00549       <span class="comment">// should this fail to global prior?</span>
<a name="l00550"></a>00550       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t find model for SNP: &quot;</span> + TmpName);
<a name="l00551"></a>00551     } <span class="keywordflow">else</span> {
<a name="l00552"></a>00552       p = m_vectSnpPriors.getAt(iSearchIndex);
<a name="l00553"></a>00553       sDist = *p;
<a name="l00554"></a>00554       <span class="comment">// this is really one copy operation!!!</span>
<a name="l00555"></a>00555       tsp.prior.Copy(p-&gt;Dist);
<a name="l00556"></a>00556       <span class="comment">/*</span>
<a name="l00557"></a>00557 <span class="comment">        tsp.prior.aa = p-&gt;Dist.aa;</span>
<a name="l00558"></a>00558 <span class="comment">        tsp.prior.ab = p-&gt;Dist.ab;</span>
<a name="l00559"></a>00559 <span class="comment">        tsp.prior.bb = p-&gt;Dist.bb;</span>
<a name="l00560"></a>00560 <span class="comment">        tsp.prior.xah = p-&gt;Dist.xah;</span>
<a name="l00561"></a>00561 <span class="comment">        tsp.prior.xab = p-&gt;Dist.xab;</span>
<a name="l00562"></a>00562 <span class="comment">        tsp.prior.xhb = p-&gt;Dist.xhb;</span>
<a name="l00563"></a>00563 <span class="comment">        tsp.prior.yah = p-&gt;Dist.yah;</span>
<a name="l00564"></a>00564 <span class="comment">        tsp.prior.yab = p-&gt;Dist.yab;</span>
<a name="l00565"></a>00565 <span class="comment">        tsp.prior.yhb = p-&gt;Dist.yhb;</span>
<a name="l00566"></a>00566 <span class="comment">        tsp.prior.xyah = p-&gt;Dist.xyah;</span>
<a name="l00567"></a>00567 <span class="comment">        tsp.prior.xyab = p-&gt;Dist.xyab;</span>
<a name="l00568"></a>00568 <span class="comment">        tsp.prior.xyhb = p-&gt;Dist.xyhb;</span>
<a name="l00569"></a>00569 <span class="comment">        tsp.prior.yxah = p-&gt;Dist.yxah;</span>
<a name="l00570"></a>00570 <span class="comment">        tsp.prior.yxab = p-&gt;Dist.yxab;</span>
<a name="l00571"></a>00571 <span class="comment">        tsp.prior.yxhb = p-&gt;Dist.yxhb;</span>
<a name="l00572"></a>00572 <span class="comment">      */</span>
<a name="l00573"></a>00573     }
<a name="l00574"></a>00574   }
<a name="l00575"></a>00575 }
<a name="l00576"></a>00576 
<a name="l00577"></a><a class="code" href="classQuantLabelZ.html#ac3a27a5f1f6ed05985950ae124193b09">00577</a> std::string <a class="code" href="classQuantLabelZ.html#ac3a27a5f1f6ed05985950ae124193b09" title="return comma-joined string of call probabilites: BB,AB,AA,Ocean">QuantLabelZ::getCallProbabilities</a>(<span class="keywordtype">int</span> idx) {
<a name="l00578"></a>00578   <span class="keywordflow">return</span> <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantLabelZ.html#a9aac165608acb81536b632f5da67910a" title="probability of het call">m_Txb</a>[idx])+<span class="stringliteral">&quot;,&quot;</span>+<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantLabelZ.html#a7a880d38f86a87b124407c7939bc2900" title="probability of AA call">m_Txh</a>[idx])+<span class="stringliteral">&quot;,&quot;</span>+<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Txa[idx])+<span class="stringliteral">&quot;,&quot;</span>+<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(1 - <a class="code" href="classQuantLabelZ.html#a9aac165608acb81536b632f5da67910a" title="probability of het call">m_Txb</a>[idx] - <a class="code" href="classQuantLabelZ.html#a7a880d38f86a87b124407c7939bc2900" title="probability of AA call">m_Txh</a>[idx] - m_Txa[idx]);
<a name="l00579"></a>00579 }
<a name="l00580"></a>00580 
<a name="l00581"></a>00581 <span class="keywordtype">void</span> QuantLabelZ::SubsetSetup(vector&lt;int&gt; &amp;samplesubset, <span class="keywordtype">int</span> &amp;subsetlength, <span class="keywordtype">int</span> copyIx, vector&lt;int&gt;&amp; SnpSampleCopyNumber) {
<a name="l00582"></a>00582   <span class="keywordtype">int</span> sIx;
<a name="l00583"></a>00583 
<a name="l00584"></a>00584    <span class="keywordflow">for</span> (sIx=0; sIx&lt;samplesubset.size(); sIx++) {
<a name="l00585"></a>00585       <span class="comment">//logic</span>
<a name="l00586"></a>00586       <span class="keywordflow">if</span> (SnpSampleCopyNumber[sIx]==copyIx)
<a name="l00587"></a>00587         samplesubset[sIx]=1;
<a name="l00588"></a>00588       <span class="keywordflow">else</span>
<a name="l00589"></a>00589         samplesubset[sIx]=0;
<a name="l00590"></a>00590     }
<a name="l00591"></a>00591 
<a name="l00592"></a>00592     <span class="comment">// setup input data</span>
<a name="l00593"></a>00593     subsetlength = 0;
<a name="l00594"></a>00594     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;samplesubset.size(); i++)
<a name="l00595"></a>00595       subsetlength += samplesubset[i];
<a name="l00596"></a>00596 }
<a name="l00597"></a>00597 <span class="comment"></span>
<a name="l00598"></a>00598 <span class="comment">/**</span>
<a name="l00599"></a>00599 <span class="comment"> * @brief Do the heavy lifting of estimation.</span>
<a name="l00600"></a>00600 <span class="comment"> */</span>
<a name="l00601"></a><a class="code" href="classQuantLabelZ.html#ad83c392dc97c1557e25bdd8b9b901ec9">00601</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ad83c392dc97c1557e25bdd8b9b901ec9" title="Do the heavy lifting of estimation.">QuantLabelZ::computeEstimate</a>() {
<a name="l00602"></a>00602 
<a name="l00603"></a>00603 
<a name="l00604"></a>00604   <span class="keyword">static</span> <span class="keywordtype">int</span> call_count = 0;
<a name="l00605"></a>00605   call_count++;
<a name="l00606"></a>00606   
<a name="l00607"></a>00607   <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> objSearch;
<a name="l00608"></a>00608   <span class="comment">// first, do anything necessary</span>
<a name="l00609"></a>00609   <a class="code" href="classQuantLabelZ.html#afa5e98cbfea742735a11e38542ed72ba" title="preprocess and normalize values">PreprocessValues</a>();
<a name="l00610"></a>00610 
<a name="l00611"></a>00611   vector&lt;affx::GType&gt; tcall;
<a name="l00612"></a>00612   vector&lt;double&gt; tconf, tx, ty;
<a name="l00613"></a>00613   vector&lt;double&gt; txa, txh, txb;
<a name="l00614"></a>00614 
<a name="l00615"></a>00615   <span class="comment">// set up snp specific prior if available</span>
<a name="l00616"></a>00616   <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> tsp;
<a name="l00617"></a>00617 
<a name="l00618"></a>00618   <span class="comment">// numeric method does its magic here</span>
<a name="l00619"></a>00619   <span class="comment">// call the numerics</span>
<a name="l00620"></a>00620   <span class="keywordtype">int</span> length;
<a name="l00621"></a>00621   length = <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size();
<a name="l00622"></a>00622   m_Calls.assign(length,<a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a1cef2fc4697bd5fc67116fbd1ee2027b" title="No call, unknown genotype. The internal is value is &amp;#39;3&amp;#39;; external is &amp;#39;-1&amp;#39;,.">affx::NN</a>);
<a name="l00623"></a>00623   <a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>.assign(length,0);
<a name="l00624"></a>00624   <span class="keywordflow">if</span> (m_OutputProbabilities) {
<a name="l00625"></a>00625     m_Txa.assign(length, 0);
<a name="l00626"></a>00626     <a class="code" href="classQuantLabelZ.html#a7a880d38f86a87b124407c7939bc2900" title="probability of AA call">m_Txh</a>.assign(length, 0);
<a name="l00627"></a>00627     <a class="code" href="classQuantLabelZ.html#a9aac165608acb81536b632f5da67910a" title="probability of het call">m_Txb</a>.assign(length, 0);
<a name="l00628"></a>00628   }
<a name="l00629"></a>00629 
<a name="l00630"></a>00630   <span class="comment">// standardized distance to each cluster</span>
<a name="l00631"></a>00631   <a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>.resize(3);
<a name="l00632"></a>00632   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> m_Di=0; m_Di&lt;3; m_Di++)
<a name="l00633"></a>00633     <a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>[m_Di].assign(length,0);
<a name="l00634"></a>00634 
<a name="l00635"></a>00635   vector&lt;int&gt; samplesubset;
<a name="l00636"></a>00636   vector&lt;int&gt; GenoHint, GenoRef;
<a name="l00637"></a>00637   vector&lt;double&gt; SubsetInbredPenalty; 
<a name="l00638"></a>00638   
<a name="l00639"></a>00639   samplesubset.assign(length,1);
<a name="l00640"></a>00640   <span class="keywordtype">int</span> subsetlength;
<a name="l00641"></a>00641   <span class="keywordtype">int</span> copyIx,sIx;
<a name="l00642"></a>00642   <span class="keywordtype">string</span> TmpName;
<a name="l00643"></a>00643 
<a name="l00644"></a>00644   <span class="comment">// put together the reference vector of seeds, if it exists</span>
<a name="l00645"></a>00645   GenoRef.assign(length,<a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a1cef2fc4697bd5fc67116fbd1ee2027b" title="No call, unknown genotype. The internal is value is &amp;#39;3&amp;#39;; external is &amp;#39;-1&amp;#39;,.">affx::NN</a>);
<a name="l00646"></a>00646   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> qIx=0; qIx&lt;m_InitialCalls.size(); qIx++)
<a name="l00647"></a>00647     GenoRef[qIx] = <a class="code" href="namespaceaffx.html#ac9e9039b462ebd2bec646ed4c5ff1122" title="Convert a GType enum to an external integer value ///.">affx::GType_to_int</a>(m_InitialCalls[qIx]);
<a name="l00648"></a>00648 
<a name="l00649"></a>00649   <span class="comment">// look up the right copy number for this snp</span>
<a name="l00650"></a>00650   <span class="keywordtype">int</span> MaleCopy,FemaleCopy;
<a name="l00651"></a>00651   map&lt;string,pair&lt;int, int&gt; &gt;::iterator test;
<a name="l00652"></a>00652   test=<a class="code" href="classQuantLabelZ.html#acc7830f7121d9a30d4ff7f38acb04ddb" title="index of special snps and their copy numbers">m_SpecialSnps</a>.find(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>);
<a name="l00653"></a>00653   <span class="keywordflow">if</span> (test==<a class="code" href="classQuantLabelZ.html#acc7830f7121d9a30d4ff7f38acb04ddb" title="index of special snps and their copy numbers">m_SpecialSnps</a>.end()) {
<a name="l00654"></a>00654     MaleCopy=2;
<a name="l00655"></a>00655     FemaleCopy=2;
<a name="l00656"></a>00656   }
<a name="l00657"></a>00657   <span class="keywordflow">else</span> {
<a name="l00658"></a>00658     MaleCopy=test-&gt;second.first;
<a name="l00659"></a>00659     FemaleCopy=test-&gt;second.second;
<a name="l00660"></a>00660   }
<a name="l00661"></a>00661 
<a name="l00662"></a>00662   <span class="comment">// assign sample copy numbers for this snp</span>
<a name="l00663"></a>00663   vector&lt;int&gt; SnpSampleCopyNumber;
<a name="l00664"></a>00664   SnpSampleCopyNumber.assign(length,2); <span class="comment">// default for all regular snps</span>
<a name="l00665"></a>00665   <span class="keywordflow">for</span> (sIx=0; sIx&lt;length; sIx++) {
<a name="l00666"></a>00666     <span class="keywordflow">if</span> (<a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983" title="What gender is each sample?">m_Genders</a>[sIx] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669ad5afd116ccd5bf71e27bab31a23fac97" title="Has XX chromosomes.">affx::Female</a>)
<a name="l00667"></a>00667       SnpSampleCopyNumber[sIx] = FemaleCopy;
<a name="l00668"></a>00668     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (<a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983" title="What gender is each sample?">m_Genders</a>[sIx] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669afdfe578a9e9044b7b274b736ed341a81" title="Has XY chromosomes.">affx::Male</a>)
<a name="l00669"></a>00669       SnpSampleCopyNumber[sIx] = MaleCopy;
<a name="l00670"></a>00670     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (<a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983" title="What gender is each sample?">m_Genders</a>[sIx] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669aa259d114b82ef05d3f1f9d3c81c84719" title="Don&amp;#39;t know genotype, use XX model.">affx::UnknownGender</a>) {
<a name="l00671"></a>00671       <span class="keywordflow">if</span> (m_ZWGenderCalling) {
<a name="l00672"></a>00672         SnpSampleCopyNumber[sIx] = MaleCopy;
<a name="l00673"></a>00673       }
<a name="l00674"></a>00674       <span class="keywordflow">else</span> {
<a name="l00675"></a>00675         SnpSampleCopyNumber[sIx] = FemaleCopy;
<a name="l00676"></a>00676       }
<a name="l00677"></a>00677     }
<a name="l00678"></a>00678     <span class="keywordflow">else</span>
<a name="l00679"></a>00679       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Unknown gender seen by genotype caller: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983" title="What gender is each sample?">m_Genders</a>[sIx]));
<a name="l00680"></a>00680   }
<a name="l00681"></a>00681 
<a name="l00682"></a>00682   map&lt;string,vector&lt;int&gt; &gt;::iterator test2;
<a name="l00683"></a>00683   test2=m_SpecialSampleSnps.find(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>);
<a name="l00684"></a>00684   <span class="keywordflow">if</span> (test2!=m_SpecialSampleSnps.end()) {
<a name="l00685"></a>00685     <span class="keywordflow">for</span> (sIx=0; sIx&lt;length; sIx++) {
<a name="l00686"></a>00686       SnpSampleCopyNumber[sIx] = test2-&gt;second[sIx];
<a name="l00687"></a>00687     }
<a name="l00688"></a>00688   }
<a name="l00689"></a>00689 
<a name="l00690"></a>00690   <span class="keywordflow">if</span> (m_bProbeSetTrustOpt)
<a name="l00691"></a>00691     EliminateUntrustedSamples(SnpSampleCopyNumber);
<a name="l00692"></a>00692 
<a name="l00693"></a>00693   <span class="comment">// Do subsets by copy number</span>
<a name="l00694"></a>00694   <span class="keywordflow">for</span> (copyIx=2; copyIx&gt;0; copyIx--) {
<a name="l00695"></a>00695     
<a name="l00696"></a>00696     <span class="comment">// assign appropriate subset to current copy number</span>
<a name="l00697"></a>00697     SubsetSetup(samplesubset,subsetlength, copyIx, SnpSampleCopyNumber);
<a name="l00698"></a>00698 
<a name="l00699"></a>00699     <span class="comment">// if there&#39;s anything to do</span>
<a name="l00700"></a>00700     <span class="keywordflow">if</span> (subsetlength&gt;0) {
<a name="l00701"></a>00701       <span class="comment">// if there&#39;s something to do, find the SNP specific prior</span>
<a name="l00702"></a>00702 
<a name="l00703"></a>00703       tsp.<a class="code" href="classsnp__param.html#a73c1d4ba7af3924f444007aaa2e802ce" title="copy parameters">copy</a>(<a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>); <span class="comment">// duplicate current parameters to keep commands</span>
<a name="l00704"></a>00704       tsp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a>=copyIx; <span class="comment">// what copy for this snp so know right model centers</span>
<a name="l00705"></a>00705 
<a name="l00706"></a>00706       <span class="comment">// look up this snp and/or prepare for output</span>
<a name="l00707"></a>00707       TmpName = <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>;
<a name="l00708"></a>00708       <span class="keywordflow">if</span> (copyIx&lt;2)
<a name="l00709"></a>00709         TmpName = <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot;:&quot;</span> +<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(copyIx);
<a name="l00710"></a>00710           <span class="comment">// if any to look up</span>
<a name="l00711"></a>00711       <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> sDist;
<a name="l00712"></a>00712       <span class="keywordflow">if</span> (m_vectSnpPriors.getCount() &gt; 0 || m_SequentialModelTsv.is_open()) {
<a name="l00713"></a>00713         getPrior(TmpName, copyIx, tsp, sDist);
<a name="l00714"></a>00714       }
<a name="l00715"></a>00715 
<a name="l00716"></a>00716       tcall.resize(subsetlength);
<a name="l00717"></a>00717       tconf.resize(subsetlength);
<a name="l00718"></a>00718       tx.resize(subsetlength);
<a name="l00719"></a>00719       ty.resize(subsetlength);
<a name="l00720"></a>00720       GenoHint.assign(subsetlength,-1);
<a name="l00721"></a>00721       SubsetInbredPenalty.assign(subsetlength,0); <span class="comment">// Null penalty</span>
<a name="l00722"></a>00722 
<a name="l00723"></a>00723       <a class="code" href="QuantLabelZ_8cpp.html#a29242d59cf441197811d945aea09720d" title="Utility function to grab a set of samples (like male/female).">t_extract_subset</a>(tx,<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>,samplesubset);
<a name="l00724"></a>00724       <a class="code" href="QuantLabelZ_8cpp.html#a29242d59cf441197811d945aea09720d" title="Utility function to grab a set of samples (like male/female).">t_extract_subset</a>(ty,<a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>,samplesubset);
<a name="l00725"></a>00725       <a class="code" href="QuantLabelZ_8cpp.html#a29242d59cf441197811d945aea09720d" title="Utility function to grab a set of samples (like male/female).">t_extract_subset</a>(GenoHint,GenoRef,samplesubset);
<a name="l00726"></a>00726       <a class="code" href="QuantLabelZ_8cpp.html#a29242d59cf441197811d945aea09720d" title="Utility function to grab a set of samples (like male/female).">t_extract_subset</a>(SubsetInbredPenalty,m_InbredHetPenalty,samplesubset);
<a name="l00727"></a>00727       <span class="comment">// am I getting anything useful here</span>
<a name="l00728"></a>00728      
<a name="l00729"></a>00729       <span class="comment">// note: using temporary copy of main parameters tsp</span>
<a name="l00730"></a>00730       <a class="code" href="snp_8label_8cpp.html#a1050ceb8ac15c0c39cffa4dbf76709b3" title="The master routine: takes data and turns it into genotypes with no seeds needed.">bayes_label</a>(tcall,tconf,tx,ty,GenoHint,SubsetInbredPenalty,tsp,txa,txh,txb,m_OutputProbabilities);
<a name="l00731"></a>00731 
<a name="l00732"></a>00732       <span class="comment">// transfer data to calls</span>
<a name="l00733"></a>00733       <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(m_Calls,tcall,samplesubset);
<a name="l00734"></a>00734       <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(<a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>,tconf,samplesubset);
<a name="l00735"></a>00735       <span class="keywordflow">if</span> (m_OutputProbabilities) {
<a name="l00736"></a>00736         <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(m_Txa,txa,samplesubset);
<a name="l00737"></a>00737         <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(<a class="code" href="classQuantLabelZ.html#a7a880d38f86a87b124407c7939bc2900" title="probability of AA call">m_Txh</a>,txh,samplesubset);
<a name="l00738"></a>00738         <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(<a class="code" href="classQuantLabelZ.html#a9aac165608acb81536b632f5da67910a" title="probability of het call">m_Txb</a>,txb,samplesubset);
<a name="l00739"></a>00739       }
<a name="l00740"></a>00740 
<a name="l00741"></a>00741       <span class="comment">// note: bayes_label uses probability, not distance</span>
<a name="l00742"></a>00742       <span class="comment">// computing standardized distance to be compatible QC</span>
<a name="l00743"></a>00743 
<a name="l00744"></a>00744       vector&lt;double&gt; tsd;
<a name="l00745"></a>00745       <span class="keywordtype">int</span> tsdi;
<a name="l00746"></a>00746       tsd.assign(subsetlength,0);
<a name="l00747"></a>00747       <span class="keywordflow">for</span> (tsdi=0; tsdi&lt;subsetlength; tsdi++) {
<a name="l00748"></a>00748         tsd[tsdi] = abs((tsp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-tx[tsdi])/sqrt(tsp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>));
<a name="l00749"></a>00749       }
<a name="l00750"></a>00750       <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(<a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>[0],tsd,samplesubset);
<a name="l00751"></a>00751       <span class="keywordflow">for</span> (tsdi=0; tsdi&lt;subsetlength; tsdi++) {
<a name="l00752"></a>00752         tsd[tsdi] = abs((tsp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-tx[tsdi])/sqrt(tsp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>));
<a name="l00753"></a>00753       }
<a name="l00754"></a>00754       <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(<a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>[1],tsd,samplesubset);
<a name="l00755"></a>00755       <span class="keywordflow">for</span> (tsdi=0; tsdi&lt;subsetlength; tsdi++) {
<a name="l00756"></a>00756         tsd[tsdi] = abs((tsp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-tx[tsdi])/sqrt(tsp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>));
<a name="l00757"></a>00757       }
<a name="l00758"></a>00758       <a class="code" href="QuantLabelZ_8cpp.html#a6db6b12c67b41b802a43bad2b5ad61fb" title="Utility function for returning subset values to place in whole sample list.">t_place_subset</a>(<a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>[2],tsd,samplesubset);
<a name="l00759"></a>00759 
<a name="l00760"></a>00760       <span class="comment">// now possible output of snp-specific posterior</span>
<a name="l00761"></a>00761       <span class="comment">// that is, now we&#39;ve updated the prior with the data</span>
<a name="l00762"></a>00762       <span class="comment">// and we want to save what we learned</span>
<a name="l00763"></a>00763       <span class="keywordflow">if</span> (m_SnpPosteriorTsv.is_open()) {
<a name="l00764"></a>00764         <a class="code" href="classQuantLabelZ.html#a35793acf2ae3d85c79930cadb62beb7a" title="Writes one specific posterior to a file along with snp name.">writeSnpPosteriorValue</a>(TmpName, tsp);
<a name="l00765"></a>00765       }
<a name="l00766"></a>00766 
<a name="l00767"></a>00767     } <span class="comment">// done processing this copy number</span>
<a name="l00768"></a>00768   } <span class="comment">// back into the loop for more copy number tries</span>
<a name="l00769"></a>00769 
<a name="l00770"></a>00770   <span class="comment">// made calls for all samples</span>
<a name="l00771"></a>00771   <span class="comment">// select probe that best mimics</span>
<a name="l00772"></a>00772   <span class="comment">// hack: chrX not yet handled differently</span>
<a name="l00773"></a>00773   <span class="comment">// but should work in regression framework?</span>
<a name="l00774"></a>00774 
<a name="l00775"></a>00775   <span class="keywordflow">if</span> (m_SnpProbeTsv.is_open()){
<a name="l00776"></a>00776     <a class="code" href="classQuantLabelZ.html#aee903e230d96abad58c24118b312b100" title="Selects good probes.">SelectGoodProbes</a>();
<a name="l00777"></a>00777   }
<a name="l00778"></a>00778   
<a name="l00779"></a>00779 }
<a name="l00780"></a>00780 <span class="comment"></span>
<a name="l00781"></a>00781 <span class="comment">/**</span>
<a name="l00782"></a>00782 <span class="comment"> * @brief Set up the quantification method given all the data about the probe</span>
<a name="l00783"></a>00783 <span class="comment"> * set, chip layout and data.</span>
<a name="l00784"></a>00784 <span class="comment"> *</span>
<a name="l00785"></a>00785 <span class="comment"> * @param psGroup - Probes to be used for final estimate.</span>
<a name="l00786"></a>00786 <span class="comment"> * @param layout - Chip layout annotation.</span>
<a name="l00787"></a>00787 <span class="comment"> * @param iMart - Raw data from chips.</span>
<a name="l00788"></a>00788 <span class="comment"> * @param iTrans - Transformations to be applied to data before use.</span>
<a name="l00789"></a>00789 <span class="comment"> * @param pmAdjust - How to estimate background, or MM probe.</span>
<a name="l00790"></a>00790 <span class="comment"> * @return True if setup sucessful, false otherwise.</span>
<a name="l00791"></a>00791 <span class="comment"> */</span>
<a name="l00792"></a><a class="code" href="classQuantLabelZ.html#a6881e990126c3e628f72a99a94452cbe">00792</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a6881e990126c3e628f72a99a94452cbe" title="Set up the quantification method given all the data about the probe set, chip layout and data...">QuantLabelZ::setUp</a>(<a class="code" href="classProbeSetGroup.html" title="Group of probe sets that should be processed as a single large probe set.">ProbeSetGroup</a> &amp;psGroup, <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, 
<a name="l00793"></a>00793                         std::vector&lt;ChipStream *&gt; &amp;iTrans, <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust) {
<a name="l00794"></a>00794         <span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs = NULL; <span class="comment">//         Genotyping probeset</span>
<a name="l00795"></a>00795   <a class="code" href="classQuantLabelZ.html#a23d0f3fb96579a159ad8d088fd7747eb" title="Clear out all of our results and values.">blankSelf</a>(); <span class="comment">// Make sure we clear out the past analysis before starting this one.</span>
<a name="l00796"></a>00796   <span class="keywordtype">bool</span> success = <span class="keyword">true</span>;
<a name="l00797"></a>00797   <span class="comment">/* Sanity checks about probesets. */</span>
<a name="l00798"></a>00798   <span class="keywordflow">if</span> (psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>.empty())
<a name="l00799"></a>00799     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Zero probesets in ProbeSetGroup (group: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(psGroup.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00800"></a>00800   <span class="keywordflow">if</span> (psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>.size() &gt; 1)
<a name="l00801"></a>00801     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t have multiple probesets in a genotyping ProbeSetGroup (group: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(psGroup.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00802"></a>00802 
<a name="l00803"></a>00803   <span class="comment">/* Remember this probeset. */</span>
<a name="l00804"></a>00804   gtPs = psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>[0];
<a name="l00805"></a>00805   success &amp;= <a class="code" href="classQuantLabelZ.html#a6c51a62f1bcd9d8d9b357bc9db118620" title="Extracts intensities, etc for a probe set, so further analysis can proceed.">SetUpProbeSet</a>(gtPs, iMart,iTrans,pmAdjust, <span class="keyword">true</span>);
<a name="l00806"></a>00806   <span class="keywordflow">if</span> (!success)
<a name="l00807"></a>00807     <a class="code" href="classQuantLabelZ.html#a23d0f3fb96579a159ad8d088fd7747eb" title="Clear out all of our results and values.">blankSelf</a>();
<a name="l00808"></a>00808   <span class="keywordflow">return</span> success;
<a name="l00809"></a>00809 }
<a name="l00810"></a>00810 <span class="comment"></span>
<a name="l00811"></a>00811 <span class="comment">/**</span>
<a name="l00812"></a>00812 <span class="comment"> * Construct a contrast normalization function from a collection of probesets. *</span>
<a name="l00813"></a>00813 <span class="comment"> * @param probeSets - ProbeSets to be used in computing normalizer.</span>
<a name="l00814"></a>00814 <span class="comment"> * @param layout - description and chip and probesets.</span>
<a name="l00815"></a>00815 <span class="comment"> * @param iMart - Cel file data for probesets.</span>
<a name="l00816"></a>00816 <span class="comment"> * @param iTrans - Transformers to operate on raw data (i.e. normalization and bg subtraction).</span>
<a name="l00817"></a>00817 <span class="comment"> * @param pmAdjust - Individual probe adjustment like pm-mm or pm-only.</span>
<a name="l00818"></a>00818 <span class="comment"> *</span>
<a name="l00819"></a>00819 <span class="comment"> * @return - returns a normalization function set.</span>
<a name="l00820"></a>00820 <span class="comment"> */</span>
<a name="l00821"></a><a class="code" href="classQuantLabelZ.html#a8e0d8c1489dd615f81feec9f094d2aeb">00821</a> vector&lt;NormalizationPredictor&gt; <a class="code" href="classQuantLabelZ.html#a8e0d8c1489dd615f81feec9f094d2aeb" title="makes the normalization functions for all samples">QuantLabelZ::makeNorm</a>(std::vector&lt;std::string&gt; probeSets, <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> &amp;layout,
<a name="l00822"></a>00822                                    <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, std::vector&lt;ChipStream *&gt; &amp;iTrans,
<a name="l00823"></a>00823                                    <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust) {
<a name="l00824"></a>00824   <span class="comment">//normMap Qmap;</span>
<a name="l00825"></a>00825   vector&lt;vector&lt;double&gt; &gt; SampleContrast;
<a name="l00826"></a>00826   vector&lt;vector&lt;double&gt; &gt; SampleStrength;
<a name="l00827"></a>00827   <span class="keywordtype">bool</span> success=<span class="keyword">true</span>;
<a name="l00828"></a>00828   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i;
<a name="l00829"></a>00829   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> nSample; <span class="comment">// how do I detect this again?</span>
<a name="l00830"></a>00830   nSample=0;
<a name="l00831"></a>00831 
<a name="l00832"></a>00832   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;making normalization map&quot;</span>);
<a name="l00833"></a>00833   <span class="keywordflow">if</span> (probeSets.empty()) {
<a name="l00834"></a>00834      <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;bogus snp sample&quot;</span>);
<a name="l00835"></a>00835   }
<a name="l00836"></a>00836 
<a name="l00837"></a>00837   <span class="comment">// for each probeset in my sample</span>
<a name="l00838"></a>00838   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> psIx = 0; psIx &lt; probeSets.size(); psIx++) {
<a name="l00839"></a>00839     <span class="comment">// Have to skip probesets that don&#39;t have 2 or 4 groups as it</span>
<a name="l00840"></a>00840     <span class="comment">// appears that some are mislabelled in the cdf file.</span>
<a name="l00841"></a>00841     <a class="code" href="classProbeListPacked.html">ProbeListPacked</a> pList = layout.<a class="code" href="classChipLayout.html#ae465f92816b1c9983a45d1f73293cc7b" title="Find a probe set using the name as a key.">getProbeListByName</a>(probeSets[psIx]);<span class="comment"></span>
<a name="l00842"></a>00842 <span class="comment">    ///@todo deal with missing probes from kill list</span>
<a name="l00843"></a>00843 <span class="comment"></span>    <span class="keywordflow">if</span> (pList.isNull()) {
<a name="l00844"></a>00844       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantLabelZ::makeNorm(): Cannot deal with missing probes from kill list&quot;</span>);
<a name="l00845"></a>00845     }
<a name="l00846"></a>00846     <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs = <a class="code" href="classProbeListFactory.html#aaed559bbfe723ed4fe225f805a2411e8">ProbeListFactory::asProbeSet</a>(pList);
<a name="l00847"></a>00847     success = <a class="code" href="classQuantLabelZ.html#a6c51a62f1bcd9d8d9b357bc9db118620" title="Extracts intensities, etc for a probe set, so further analysis can proceed.">SetUpProbeSet</a>(gtPs, iMart,iTrans,pmAdjust, <span class="keyword">false</span>);
<a name="l00848"></a>00848     <span class="keywordflow">if</span> (!success) {
<a name="l00849"></a>00849       <span class="keyword">delete</span> gtPs;
<a name="l00850"></a>00850       <span class="keywordflow">continue</span>;
<a name="l00851"></a>00851     }
<a name="l00852"></a>00852     <span class="comment">// add snpStats values</span>
<a name="l00853"></a>00853     <span class="comment">// currently contrast/strength</span>
<a name="l00854"></a>00854     <span class="comment">// push into the slots by sample</span>
<a name="l00855"></a>00855     <span class="comment">// set up the slots to put everything in</span>
<a name="l00856"></a>00856     <span class="keywordflow">if</span> (nSample&lt;1) {
<a name="l00857"></a>00857       nSample=<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size();
<a name="l00858"></a>00858       SampleContrast.resize(nSample);
<a name="l00859"></a>00859       SampleStrength.resize(nSample);
<a name="l00860"></a>00860     }
<a name="l00861"></a>00861     <span class="keywordflow">for</span> (i=0; i&lt;nSample; i++) {
<a name="l00862"></a>00862       SampleContrast[i].push_back(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>[i]);
<a name="l00863"></a>00863       SampleStrength[i].push_back(<a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>[i]);
<a name="l00864"></a>00864     }
<a name="l00865"></a>00865     <span class="comment">// look up parameters</span>
<a name="l00866"></a>00866     <span class="comment">//cout &lt;&lt; &quot;SNP:\t&quot; &lt;&lt; psIx &lt;&lt; endl;</span>
<a name="l00867"></a>00867     <span class="keyword">delete</span> gtPs;
<a name="l00868"></a>00868   }
<a name="l00869"></a>00869   <a class="code" href="classNormalizationFrame.html" title="Holds some useful values in relation to each other.">NormalizationFrame</a> TmpFrame;
<a name="l00870"></a>00870   vector&lt;NormalizationPredictor&gt; Qmap;
<a name="l00871"></a>00871   Qmap.resize(nSample);
<a name="l00872"></a>00872   <span class="keywordflow">for</span> (i=0; i&lt;nSample; i++) {
<a name="l00873"></a>00873     <span class="comment">// for each sample</span>
<a name="l00874"></a>00874     TmpFrame.<a class="code" href="classNormalizationFrame.html#aeaf88ff39f1c1e6acb95d40652e9fd86" title="just the dependent variable">FillInContrast</a>(SampleContrast[i]);
<a name="l00875"></a>00875     TmpFrame.<a class="code" href="classNormalizationFrame.html#a13be975a971604fa8159d4d6cd788c27" title="just the one covariate">FillInCovariates</a>(SampleStrength[i]);
<a name="l00876"></a>00876     TmpFrame.<a class="code" href="classNormalizationFrame.html#a3fffe184d5187c6707947ede68223959" title="wrapper for numeric function">FitEM</a>();
<a name="l00877"></a>00877     <span class="comment">//store in Qmap</span>
<a name="l00878"></a>00878     Qmap[i].Copy(TmpFrame.<a class="code" href="classNormalizationFrame.html#a9b5a4f8d837e07c0ad9d3701cf0cd543" title="functions for predicting">ThisFit</a>);
<a name="l00879"></a>00879     <span class="comment">/*cout &lt;&lt; &quot;Sample:\t&quot; &lt;&lt; i &lt;&lt; endl;</span>
<a name="l00880"></a>00880 <span class="comment">    td(Qmap[i].fitAA);</span>
<a name="l00881"></a>00881 <span class="comment">    td(Qmap[i].fitAB);</span>
<a name="l00882"></a>00882 <span class="comment">    td(Qmap[i].fitBB);</span>
<a name="l00883"></a>00883 <span class="comment">    td(Qmap[i].sigma);*/</span>
<a name="l00884"></a>00884   }
<a name="l00885"></a>00885   <span class="comment">// compute normalization function here</span>
<a name="l00886"></a>00886   <span class="comment">// fun for the whole family</span>
<a name="l00887"></a>00887   <span class="keywordflow">return</span> Qmap;
<a name="l00888"></a>00888 }
<a name="l00889"></a>00889 <span class="comment"></span>
<a name="l00890"></a>00890 <span class="comment">/**</span>
<a name="l00891"></a>00891 <span class="comment"> * Calls gender using EM.</span>
<a name="l00892"></a>00892 <span class="comment"> * This should not be anywhere near LabelZ, but code structure &amp; time make this easiest.</span>
<a name="l00893"></a>00893 <span class="comment"> *</span>
<a name="l00894"></a>00894 <span class="comment"> * @param contrast - list of contrast values for chrX snps in one sample</span>
<a name="l00895"></a>00895 <span class="comment"> * @param em_thresh -  when are we done?</span>
<a name="l00896"></a>00896 <span class="comment"> * @param em_cutoff - enough data to make decision?</span>
<a name="l00897"></a>00897 <span class="comment"> * @param gender_cutoff - make decision as to gender</span>
<a name="l00898"></a>00898 <span class="comment"> */</span>
<a name="l00899"></a><a class="code" href="QuantLabelZ_8cpp.html#aac52f95f7c541c7b78dec8b769a6bbd1">00899</a> <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669" title="Those pesky XY people need a different prior for all the SNPs on X that aren&amp;#39;t on the psuedoautos...">Gender</a> <a class="code" href="QuantLabelZ_8cpp.html#aac52f95f7c541c7b78dec8b769a6bbd1" title="Calls gender using EM.">t_CallEMGender</a>(<span class="keyword">const</span> std::vector&lt;double&gt;&amp; contrast,
<a name="l00900"></a>00900                       <span class="keywordtype">double</span>&amp; em_thresh,
<a name="l00901"></a>00901                       <span class="keywordtype">double</span>&amp; em_cutoff,
<a name="l00902"></a>00902                       <span class="keywordtype">double</span>&amp; gender_cutoff) {
<a name="l00903"></a>00903   <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669" title="Those pesky XY people need a different prior for all the SNPs on X that aren&amp;#39;t on the psuedoautos...">Gender</a> retval;
<a name="l00904"></a>00904   <a class="code" href="classCEMSeed.html" title="Object to hold all prior information to constitute an EM seed.">CEMSeed</a> seed;
<a name="l00905"></a>00905   seed.<a class="code" href="classCEMSeed.html#a1d85b2d7578a68250caa978a06bf65fc" title="Set the centers of clusters assuming upto three clusters in total.">setMu</a>(-0.66, 0, 0.66);
<a name="l00906"></a>00906   seed.<a class="code" href="classCEMSeed.html#af7f9a507dcdfb399eab17dec2650bad0" title="Set the sigmas of clusters assuming upto three clusters in total.">setSigma</a>(0.1f, 0.1f, 0.1f);
<a name="l00907"></a>00907   seed.<a class="code" href="classCEMSeed.html#ae014524f6c4329ca8b18364b316361c1" title="Set the weight of clusters assuming upto three clusters in total.">setWeight</a>(0.33f, 0.34f, 0.33f);
<a name="l00908"></a>00908   seed.<a class="code" href="classCEMSeed.html#ab53cf92fa9b34fdae1394e5bb33b9498" title="Set the minimum of centers of clusters assuming upto three clusters in total.">setMinMu</a>(-2, -.05, .25);
<a name="l00909"></a>00909   seed.<a class="code" href="classCEMSeed.html#abb9d2912972d1c6ca13b40a91793a74c" title="Set the minimum of sigmas of clusters assuming upto three clusters in total.">setMinSigma</a>(0.02f, 0.02f, 0.02f);
<a name="l00910"></a>00910   seed.<a class="code" href="classCEMSeed.html#abf4e379c205f693d233e7e6154157a7f" title="Set the maximum of centers of clusters assuming upto three clusters in total.">setMaxMu</a>(-.25, .05, 2);
<a name="l00911"></a>00911   seed.<a class="code" href="classCEMSeed.html#a7507e7757be1c92ce717ed86ec1c2aa2" title="Set the maximum of sigmas of clusters assuming upto three clusters in total.">setMaxSigma</a>(.3, .3, .3);
<a name="l00912"></a>00912 
<a name="l00913"></a>00913   <a class="code" href="classCPrimeEM.html" title="The primary class to hold all EM operations.">CPrimeEM</a> em;
<a name="l00914"></a>00914   em.<a class="code" href="classCPrimeEM.html#ae6633a6897c30d1c94bbb9d284c4e672" title="set the data to be classified">setData</a>(contrast);
<a name="l00915"></a>00915   em.<a class="code" href="classCPrimeEM.html#a6169a816ab26a0752f81a916f10029b9" title="confidence threshold set function, for backward compatibility">setThreshold</a>(em_thresh);
<a name="l00916"></a>00916   em.<a class="code" href="classCPrimeEM.html#a9c2454ee67f79c7465be66233ac48994" title="the function to get EM&amp;#39;s estimate with a specific initial seed">EMEstimate</a>(seed);
<a name="l00917"></a>00917   <a class="code" href="classCEMEst.html" title="Object to hold all EM posterior information including some classification quality stats...">CEMEst</a>* pEst = em.<a class="code" href="classCPrimeEM.html#a66f8c305c57d8751e588fdee92c1ec94" title="get to posterior">getEMEstimates</a>();
<a name="l00918"></a>00918   <span class="comment">//int nNC = count_if (pEst-&gt;m_class.begin(),pEst-&gt;m_class.end(),bind2nd(equal_to&lt;int&gt;(),-1));</span>
<a name="l00919"></a>00919   <span class="keywordtype">int</span> nAA = count_if (pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.begin(),pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.end(),bind2nd(equal_to&lt;int&gt;(),0));
<a name="l00920"></a>00920   <span class="keywordtype">int</span> nBB = count_if (pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.begin(),pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.end(),bind2nd(equal_to&lt;int&gt;(),2));
<a name="l00921"></a>00921   <span class="keywordtype">int</span> nAB = count_if (pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.begin(),pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.end(),bind2nd(equal_to&lt;int&gt;(),1));
<a name="l00922"></a>00922   <span class="keywordtype">double</span> HCR=-1;
<a name="l00923"></a>00923   <span class="keywordflow">if</span> (nAA+nBB+nBB &gt; 0 &amp;&amp; <span class="keywordtype">double</span>(nAA+nAB+nBB)/pEst-&gt;<a class="code" href="classCEMEst.html#a76437dcd2f2e5a2b93429124946fda8d" title="EM cluster member assignment.">m_class</a>.size() &gt; em_cutoff) {
<a name="l00924"></a>00924       HCR = double(nAB)/double(nAA+nAB+nBB);
<a name="l00925"></a>00925       <span class="keywordflow">if</span> (HCR &lt; gender_cutoff)
<a name="l00926"></a>00926         retval = affx::Male; <span class="comment">// male</span>
<a name="l00927"></a>00927       <span class="keywordflow">else</span>
<a name="l00928"></a>00928         retval = affx::Female; <span class="comment">// female</span>
<a name="l00929"></a>00929   }
<a name="l00930"></a>00930   <span class="keywordflow">else</span>
<a name="l00931"></a>00931     retval = affx::UnknownGender; <span class="comment">// unknown</span>
<a name="l00932"></a>00932   <span class="keywordflow">return</span> retval;
<a name="l00933"></a>00933 }
<a name="l00934"></a>00934 <span class="comment"></span>
<a name="l00935"></a>00935 <span class="comment">/**</span>
<a name="l00936"></a>00936 <span class="comment"> * Find genders of sample - uses EM, not labeling as you might think!</span>
<a name="l00937"></a>00937 <span class="comment"> * Again, this function would be elsewhere if we had a good structure</span>
<a name="l00938"></a>00938 <span class="comment"> *</span>
<a name="l00939"></a>00939 <span class="comment"> * @param genders - to be determined - what are the genders of the sample</span>
<a name="l00940"></a>00940 <span class="comment"> * @param probeSets - use these snps to find genders</span>
<a name="l00941"></a>00941 <span class="comment"> * @param layout - where are probes on array</span>
<a name="l00942"></a>00942 <span class="comment"> * @param iMart - serve up raw intensities</span>
<a name="l00943"></a>00943 <span class="comment"> * @param iTrans - adjust intensities (normalize, etc)</span>
<a name="l00944"></a>00944 <span class="comment"> * @param pmAdjust - adjust intensities for background</span>
<a name="l00945"></a>00945 <span class="comment"> */</span>
<a name="l00946"></a><a class="code" href="classQuantLabelZ.html#a384472516ce59e10b120bf84eff95345">00946</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a384472516ce59e10b120bf84eff95345" title="Determines the gender of samples from chrX snps by EM.">QuantLabelZ::FindGender</a>(std::vector&lt;affx::Gender&gt; &amp;genders,
<a name="l00947"></a>00947                              std::vector&lt;ProbeSet *&gt; probeSets,
<a name="l00948"></a>00948                              <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> &amp;layout,
<a name="l00949"></a>00949                              <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart,
<a name="l00950"></a>00950                              std::vector&lt;ChipStream *&gt; &amp;iTrans,
<a name="l00951"></a>00951                              <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust) {
<a name="l00952"></a>00952 
<a name="l00953"></a>00953   <span class="comment">// probeSets contains X-chromosome snps</span>
<a name="l00954"></a>00954   <span class="comment">// so we&#39;re going to grab all the summary values</span>
<a name="l00955"></a>00955   <span class="comment">// and call some genders</span>
<a name="l00956"></a>00956   vector&lt;vector&lt;double&gt; &gt; SampleContrast;
<a name="l00957"></a>00957   <span class="comment">// dead code?</span>
<a name="l00958"></a>00958   <span class="comment">//vector&lt;vector&lt;double&gt; &gt; SampleStrength;</span>
<a name="l00959"></a>00959   <span class="keywordtype">bool</span> success=<span class="keyword">true</span>;
<a name="l00960"></a>00960   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i;
<a name="l00961"></a>00961   <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> nSample; <span class="comment">// how do I detect this again?</span>
<a name="l00962"></a>00962   nSample=0;
<a name="l00963"></a>00963 
<a name="l00964"></a>00964   <span class="comment">//cout &lt;&lt; &quot;Preparing to call gender&quot; &lt;&lt; endl;</span>
<a name="l00965"></a>00965   <span class="keywordflow">if</span> (!probeSets[0]) {
<a name="l00966"></a>00966     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;bogus snp sample&quot;</span>);
<a name="l00967"></a>00967   }
<a name="l00968"></a>00968 
<a name="l00969"></a>00969   <span class="comment">// for each probeset in my sample</span>
<a name="l00970"></a>00970   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> psIx = 0; psIx &lt; probeSets.size(); psIx++) {
<a name="l00971"></a>00971     <span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs = probeSets[psIx];
<a name="l00972"></a>00972     <span class="comment">// Have to skip probesets that don&#39;t have 2 or 4 groups as it</span>
<a name="l00973"></a>00973     <span class="comment">// appears that some are mislabelled in the cdf file.</span>
<a name="l00974"></a>00974 
<a name="l00975"></a>00975     success = <a class="code" href="classQuantLabelZ.html#a6c51a62f1bcd9d8d9b357bc9db118620" title="Extracts intensities, etc for a probe set, so further analysis can proceed.">SetUpProbeSet</a>(gtPs, iMart,iTrans,pmAdjust, <span class="keyword">false</span>);
<a name="l00976"></a>00976     <span class="keywordflow">if</span> (!success)
<a name="l00977"></a>00977       <span class="keywordflow">continue</span>;
<a name="l00978"></a>00978     <span class="comment">// add snpStats values</span>
<a name="l00979"></a>00979     <span class="comment">// currently contrast/strength</span>
<a name="l00980"></a>00980     <span class="comment">// push into the slots by sample</span>
<a name="l00981"></a>00981     <span class="comment">// set up the slots to put everything in</span>
<a name="l00982"></a>00982     <span class="comment">//cout &lt;&lt; &quot;Set up pset: &quot; &lt;&lt; psIx &lt;&lt; endl;</span>
<a name="l00983"></a>00983     <span class="keywordflow">if</span> (nSample&lt;1) {
<a name="l00984"></a>00984       nSample=<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size();
<a name="l00985"></a>00985       SampleContrast.resize(nSample);
<a name="l00986"></a>00986       <span class="comment">// dead code?</span>
<a name="l00987"></a>00987       <span class="comment">// SampleStrength.resize(nSample);</span>
<a name="l00988"></a>00988     }
<a name="l00989"></a>00989     <span class="keywordflow">for</span> (i=0; i&lt;nSample; i++) {
<a name="l00990"></a>00990       SampleContrast[i].push_back(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>[i]);
<a name="l00991"></a>00991       <span class="comment">// dead code? -jhg</span>
<a name="l00992"></a>00992       <span class="comment">//SampleStrength[i].push_back(m_BValues[i]);</span>
<a name="l00993"></a>00993     }
<a name="l00994"></a>00994   }
<a name="l00995"></a>00995 
<a name="l00996"></a>00996   <span class="comment">// same here as normalization, now call gender</span>
<a name="l00997"></a>00997   <span class="comment">// okay have setup the big matrix of contrasts</span>
<a name="l00998"></a>00998   <span class="comment">// for each sample, call the gender</span>
<a name="l00999"></a>00999   <span class="comment">//cout &lt;&lt; &quot;setup contrast: call gender &quot; &lt;&lt; nSample &lt;&lt; &quot;\t&quot; &lt;&lt; genders.size() &lt;&lt; endl;</span>
<a name="l01000"></a>01000   std::vector&lt;affx::Gender&gt; Gmap;
<a name="l01001"></a>01001   Gmap.resize(nSample);
<a name="l01002"></a>01002   <span class="keywordflow">for</span> (i=0; i&lt;nSample; i++) {
<a name="l01003"></a>01003     <span class="keywordflow">if</span> (<span class="keyword">true</span>) {
<a name="l01004"></a>01004       Gmap[i] = <a class="code" href="QuantLabelZ_8cpp.html#aac52f95f7c541c7b78dec8b769a6bbd1" title="Calls gender using EM.">t_CallEMGender</a>(SampleContrast[i],em_thresh,em_cutoff,gender_cutoff);
<a name="l01005"></a>01005       <span class="comment">//cout &lt;&lt; i &lt;&lt; &quot;\t&quot;&lt;&lt; Gmap[i] &lt;&lt; &quot;\t&quot; &lt;&lt; genders[i] &lt;&lt; endl;</span>
<a name="l01006"></a>01006       genders[i] = Gmap[i];
<a name="l01007"></a>01007     }
<a name="l01008"></a>01008   }
<a name="l01009"></a>01009   <span class="comment">//cout &lt;&lt; endl;</span>
<a name="l01010"></a>01010   <span class="comment">//return Gmap;</span>
<a name="l01011"></a>01011 }
<a name="l01012"></a>01012 <span class="comment"></span>
<a name="l01013"></a>01013 <span class="comment">//////////</span>
<a name="l01014"></a>01014 <span class="comment"></span><span class="comment"></span>
<a name="l01015"></a>01015 <span class="comment">/**</span>
<a name="l01016"></a>01016 <span class="comment"> * @brief Default Getter method for parameters and their documentation.</span>
<a name="l01017"></a>01017 <span class="comment"> * @return map of parameters and their descriptions.</span>
<a name="l01018"></a>01018 <span class="comment"> */</span>
<a name="l01019"></a><a class="code" href="classQuantLabelZ.html#a0da80e2dbed413e783338e1ae66c8b43">01019</a> std::vector&lt;SelfDoc::Opt&gt; <a class="code" href="classQuantLabelZ.html#a0da80e2dbed413e783338e1ae66c8b43" title="Default Getter method for parameters and their documentation.">QuantLabelZ::getDefaultDocOptions</a>() {
<a name="l01020"></a>01020   std::vector&lt;SelfDoc::Opt&gt; opts;
<a name="l01021"></a>01021 
<a name="l01022"></a>01022   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> k = {<span class="stringliteral">&quot;K&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;4.0&quot;</span>, <span class="stringliteral">&quot;4.0&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01023"></a>01023                     <span class="stringliteral">&quot;Scale parameter used used in CCS and CES transformations.&quot;</span>};
<a name="l01024"></a>01024   opts.push_back(k);
<a name="l01025"></a>01025   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> <a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909" title="Transform allele estimates.">transform</a> = {<span class="stringliteral">&quot;transform&quot;</span>, SelfDoc::Opt::String, <span class="stringliteral">&quot;CCS&quot;</span>, <span class="stringliteral">&quot;CCS&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01026"></a>01026                             <span class="stringliteral">&quot;Transformation of initial data are we feeding into the classifier? {&#39;CCS&#39;, &#39;CES&#39;, &#39;MvA&#39;,&#39;RvT&#39;}&quot;</span>};
<a name="l01027"></a>01027   opts.push_back(transform);
<a name="l01028"></a>01028   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> lowprecision = {<span class="stringliteral">&quot;lowprecision&quot;</span>, SelfDoc::Opt::Boolean, <span class="stringliteral">&quot;false&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01029"></a>01029                                <span class="stringliteral">&quot;R prototype uses summary values rounded to first decimal place. Use this flag to be simulate behavior.&quot;</span>};
<a name="l01030"></a>01030   opts.push_back(lowprecision);
<a name="l01031"></a>01031 
<a name="l01032"></a>01032   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kx = {<span class="stringliteral">&quot;KX&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;4.0&quot;</span>,<span class="stringliteral">&quot;4.0&quot;</span>,<span class="stringliteral">&quot;0.0001&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01033"></a>01033                      <span class="stringliteral">&quot;Prior strength for Homs&quot;</span>};
<a name="l01034"></a>01034   opts.push_back(kx);
<a name="l01035"></a>01035   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kh = {<span class="stringliteral">&quot;KH&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.2&quot;</span>,<span class="stringliteral">&quot;0.2&quot;</span>,<span class="stringliteral">&quot;0.0001&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01036"></a>01036                      <span class="stringliteral">&quot;Prior strength for Hets&quot;</span>};
<a name="l01037"></a>01037   opts.push_back(kh);
<a name="l01038"></a>01038 
<a name="l01039"></a>01039   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> v = {<span class="stringliteral">&quot;V&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;10&quot;</span>,<span class="stringliteral">&quot;10&quot;</span>,<span class="stringliteral">&quot;0.0001&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01040"></a>01040                     <span class="stringliteral">&quot;Prior strength for variances&quot;</span>};
<a name="l01041"></a>01041   opts.push_back(v);
<a name="l01042"></a>01042 
<a name="l01043"></a>01043   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> aam = {<span class="stringliteral">&quot;AAM&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.66&quot;</span>,<span class="stringliteral">&quot;0.66&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01044"></a>01044                       <span class="stringliteral">&quot;Prior location of AA mean&quot;</span>};
<a name="l01045"></a>01045   opts.push_back(aam);
<a name="l01046"></a>01046   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> abm = {<span class="stringliteral">&quot;ABM&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01047"></a>01047                       <span class="stringliteral">&quot;Prior location of AB mean&quot;</span>};
<a name="l01048"></a>01048   opts.push_back(abm);
<a name="l01049"></a>01049   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bbm = {<span class="stringliteral">&quot;BBM&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;-.66&quot;</span>,<span class="stringliteral">&quot;-.66&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01050"></a>01050                       <span class="stringliteral">&quot;Prior location of BB mean&quot;</span>};
<a name="l01051"></a>01051   opts.push_back(bbm);
<a name="l01052"></a>01052   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bby = {<span class="stringliteral">&quot;BBY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;10.0&quot;</span>,<span class="stringliteral">&quot;10.0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01053"></a>01053                       <span class="stringliteral">&quot;Prior location of BB Y mean&quot;</span>};
<a name="l01054"></a>01054   opts.push_back(bby);
<a name="l01055"></a>01055   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> aby = {<span class="stringliteral">&quot;ABY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;10.0&quot;</span>,<span class="stringliteral">&quot;10.0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01056"></a>01056                       <span class="stringliteral">&quot;Prior location of AB Y mean&quot;</span>};
<a name="l01057"></a>01057   opts.push_back(aby);
<a name="l01058"></a>01058   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> aay = {<span class="stringliteral">&quot;AAY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;10.0&quot;</span>,<span class="stringliteral">&quot;10.0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01059"></a>01059                       <span class="stringliteral">&quot;Prior location of AA Y mean&quot;</span>};
<a name="l01060"></a>01060   opts.push_back(aay);
<a name="l01061"></a>01061 
<a name="l01062"></a>01062   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> aav = {<span class="stringliteral">&quot;AAV&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;.005&quot;</span>,<span class="stringliteral">&quot;.005&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01063"></a>01063                       <span class="stringliteral">&quot;Prior variance of AA&quot;</span>};
<a name="l01064"></a>01064   opts.push_back(aav);
<a name="l01065"></a>01065   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> abv = {<span class="stringliteral">&quot;ABV&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;.010&quot;</span>,<span class="stringliteral">&quot;.010&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01066"></a>01066                       <span class="stringliteral">&quot;Prior variance of AB&quot;</span>};
<a name="l01067"></a>01067   opts.push_back(abv);
<a name="l01068"></a>01068   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bbv = {<span class="stringliteral">&quot;BBV&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;.005&quot;</span>,<span class="stringliteral">&quot;.005&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01069"></a>01069                       <span class="stringliteral">&quot;Prior variance of BB&quot;</span>};
<a name="l01070"></a>01070   opts.push_back(bbv);
<a name="l01071"></a>01071 
<a name="l01072"></a>01072   <span class="comment">// within cluster variances/covariances</span>
<a name="l01073"></a>01073  <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> aayv = {<span class="stringliteral">&quot;AAYV&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01074"></a>01074                       <span class="stringliteral">&quot;Prior Variance of AA, Y coordinate&quot;</span>};
<a name="l01075"></a>01075   opts.push_back(aayv);
<a name="l01076"></a>01076   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> abyv = {<span class="stringliteral">&quot;ABYV&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01077"></a>01077                        <span class="stringliteral">&quot;Prior Variance of AB, Y coordinate&quot;</span>};
<a name="l01078"></a>01078   opts.push_back(abyv);
<a name="l01079"></a>01079   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bbyv = {<span class="stringliteral">&quot;BBYV&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01080"></a>01080                        <span class="stringliteral">&quot;Prior Variance of BB, Y coordinate&quot;</span>};
<a name="l01081"></a>01081   opts.push_back(bbyv);
<a name="l01082"></a>01082   <span class="comment">// within cluster covariances</span>
<a name="l01083"></a>01083   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> aaxy = {<span class="stringliteral">&quot;AAXY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01084"></a>01084                        <span class="stringliteral">&quot;Prior CoVariance of AA, XY&quot;</span>};
<a name="l01085"></a>01085   opts.push_back(aaxy);
<a name="l01086"></a>01086   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> abxy = {<span class="stringliteral">&quot;ABXY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01087"></a>01087                        <span class="stringliteral">&quot;Prior CoVariance of AB, XY&quot;</span>};
<a name="l01088"></a>01088   opts.push_back(abxy);
<a name="l01089"></a>01089   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bbxy = {<span class="stringliteral">&quot;BBXY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01090"></a>01090                        <span class="stringliteral">&quot;Prior CoVariance of BB, XY &quot;</span>};
<a name="l01091"></a>01091   opts.push_back(bbxy);
<a name="l01092"></a>01092 
<a name="l01093"></a>01093   <span class="comment">// between cluster covariances X to X</span>
<a name="l01094"></a>01094   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kxx = {<span class="stringliteral">&quot;KXX&quot;</span>, SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01095"></a>01095                       <span class="stringliteral">&quot;Prior strength for hom covariance&quot;</span>};
<a name="l01096"></a>01096   opts.push_back(kxx);
<a name="l01097"></a>01097   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kah = {<span class="stringliteral">&quot;KAH&quot;</span>, SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01098"></a>01098                       <span class="stringliteral">&quot;Prior strength for A-H covariance&quot;</span>};
<a name="l01099"></a>01099   opts.push_back(kah);
<a name="l01100"></a>01100   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> khb = {<span class="stringliteral">&quot;KHB&quot;</span>, SelfDoc::Opt::Float, <span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01101"></a>01101                       <span class="stringliteral">&quot;Prior strength for H-B covariance&quot;</span>};
<a name="l01102"></a>01102   opts.push_back(khb);
<a name="l01103"></a>01103   <span class="comment">// between cluster covariances Y to Y</span>
<a name="l01104"></a>01104   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kyah = {<span class="stringliteral">&quot;KYAH&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01105"></a>01105                        <span class="stringliteral">&quot;Prior strength for Y-Y A-H cluster&quot;</span>};
<a name="l01106"></a>01106   opts.push_back(kyah);
<a name="l01107"></a>01107   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kyab = {<span class="stringliteral">&quot;KYAB&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01108"></a>01108                        <span class="stringliteral">&quot;Prior Strength for Y-Y A-B cluster&quot;</span>};
<a name="l01109"></a>01109   opts.push_back(kyab);
<a name="l01110"></a>01110   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> kyhb = {<span class="stringliteral">&quot;KYHB&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01111"></a>01111                        <span class="stringliteral">&quot;Prior Strength for Y-Y H-B cluster&quot;</span>};
<a name="l01112"></a>01112   opts.push_back(kyhb);
<a name="l01113"></a>01113 
<a name="l01114"></a>01114   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> comvar = {<span class="stringliteral">&quot;COMVAR&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;1&quot;</span>,<span class="stringliteral">&quot;1&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01115"></a>01115                          <span class="stringliteral">&quot;Flag: common variance&quot;</span>};
<a name="l01116"></a>01116   opts.push_back(comvar);
<a name="l01117"></a>01117   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> hard = {<span class="stringliteral">&quot;HARD&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;2&quot;</span>,<span class="stringliteral">&quot;2&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01118"></a>01118                        <span class="stringliteral">&quot;Flag: type of hard shell&quot;</span>};
<a name="l01119"></a>01119   opts.push_back(hard);
<a name="l01120"></a>01120   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> sb   = {<span class="stringliteral">&quot;SB&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;.05&quot;</span>,<span class="stringliteral">&quot;.05&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01121"></a>01121                        <span class="stringliteral">&quot;Size of shell barrier&quot;</span>};
<a name="l01122"></a>01122   opts.push_back(sb);
<a name="l01123"></a>01123   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> cm   = {<span class="stringliteral">&quot;CM&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01124"></a>01124                        <span class="stringliteral">&quot;Type of call method, CM=1 for posterior, CM=2 for single-sample&quot;</span>};
<a name="l01125"></a>01125   opts.push_back(cm);
<a name="l01126"></a>01126   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> maxscore = {<span class="stringliteral">&quot;MS&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;.2&quot;</span>,<span class="stringliteral">&quot;.2&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;2&quot;</span>,
<a name="l01127"></a>01127                            <span class="stringliteral">&quot;Threshold for no-calls&quot;</span>};
<a name="l01128"></a>01128   opts.push_back(maxscore);
<a name="l01129"></a>01129   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bins = {<span class="stringliteral">&quot;bins&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01130"></a>01130                        <span class="stringliteral">&quot;Use efficient binning to speed up labeling&quot;</span>};
<a name="l01131"></a>01131   opts.push_back(bins);
<a name="l01132"></a>01132   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> hints = {<span class="stringliteral">&quot;hints&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01133"></a>01133                         <span class="stringliteral">&quot;Use reference genotype data to indicate clusters&quot;</span>};
<a name="l01134"></a>01134   opts.push_back(hints);
<a name="l01135"></a>01135 
<a name="l01136"></a>01136   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> <span class="keyword">override</span> = {<span class="stringliteral">&quot;override&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01137"></a>01137                            <span class="stringliteral">&quot;Use reference genotype data to select probes&quot;</span>};
<a name="l01138"></a>01138   opts.push_back(<span class="keyword">override</span>);
<a name="l01139"></a>01139 
<a name="l01140"></a>01140 
<a name="l01141"></a>01141   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> mix = {<span class="stringliteral">&quot;mix&quot;</span>, SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01142"></a>01142                       <span class="stringliteral">&quot;Apply mixture frequency penalty to clusters&quot;</span>};
<a name="l01143"></a>01143   opts.push_back(mix);
<a name="l01144"></a>01144   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> bic = {<span class="stringliteral">&quot;bic&quot;</span>, SelfDoc::Opt::Float, <span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01145"></a>01145                       <span class="stringliteral">&quot;BIC penalty for clusters&quot;</span>};
<a name="l01146"></a>01146   opts.push_back(bic);
<a name="l01147"></a>01147 
<a name="l01148"></a>01148   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> CSepPen = {<span class="stringliteral">&quot;CSepPen&quot;</span>, SelfDoc::Opt::Float, <span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01149"></a>01149                       <span class="stringliteral">&quot;Penalty for ClusterSep/Stdev too small&quot;</span>};
<a name="l01150"></a>01150   opts.push_back(CSepPen);
<a name="l01151"></a>01151   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> CSepThr = {<span class="stringliteral">&quot;CSepThr&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;16&quot;</span>,<span class="stringliteral">&quot;16&quot;</span>,<span class="stringliteral">&quot;0.000000001&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01152"></a>01152                          <span class="stringliteral">&quot;ClusterSep/Stdev threshold discounting by Geman-McClure&quot;</span>};
<a name="l01153"></a>01153   opts.push_back(CSepThr);
<a name="l01154"></a>01154 
<a name="l01155"></a>01155   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> IsoHetY = {<span class="stringliteral">&quot;IsoHetY&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01156"></a>01156                                 <span class="stringliteral">&quot;Het mean must lie above the line connecting hom means&quot;</span>};
<a name="l01157"></a>01157   opts.push_back(IsoHetY);
<a name="l01158"></a>01158 
<a name="l01159"></a>01159 
<a name="l01160"></a>01160   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> lambda = {<span class="stringliteral">&quot;lambda&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;1&quot;</span>,<span class="stringliteral">&quot;1&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;1&quot;</span>,
<a name="l01161"></a>01161                          <span class="stringliteral">&quot;Controls mixing of common variances&quot;</span>};
<a name="l01162"></a>01162   opts.push_back(lambda);
<a name="l01163"></a>01163   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> wobble = {<span class="stringliteral">&quot;wobble&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;.0001&quot;</span>,<span class="stringliteral">&quot;.0001&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01164"></a>01164                          <span class="stringliteral">&quot;Limits prior pseudo-observations to 1/wobble&quot;</span>};
<a name="l01165"></a>01165   opts.push_back(wobble);
<a name="l01166"></a>01166   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> copyqc = {<span class="stringliteral">&quot;copyqc&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;1&quot;</span>,
<a name="l01167"></a>01167                          <span class="stringliteral">&quot;Test for outlier size values (CNV or errors)&quot;</span>};
<a name="l01168"></a>01168   opts.push_back(copyqc);
<a name="l01169"></a>01169   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> copytype = {<span class="stringliteral">&quot;copytype&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01170"></a>01170                            <span class="stringliteral">&quot;Flag: Method for handling outlier data points(CNV/errors)&quot;</span>};
<a name="l01171"></a>01171   opts.push_back(copytype);
<a name="l01172"></a>01172 
<a name="l01173"></a>01173   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> ocean = {<span class="stringliteral">&quot;ocean&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;1&quot;</span>,
<a name="l01174"></a>01174                         <span class="stringliteral">&quot;Test datapoints against uniform ocean probability&quot;</span>};
<a name="l01175"></a>01175   opts.push_back(ocean);
<a name="l01176"></a>01176 
<a name="l01177"></a>01177   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> inflatePRA = {<span class="stringliteral">&quot;inflatePRA&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01178"></a>01178                              <span class="stringliteral">&quot;Make calls adding uncertainty in mean to observed variance&quot;</span>};
<a name="l01179"></a>01179   opts.push_back(inflatePRA);
<a name="l01180"></a>01180 
<a name="l01181"></a>01181   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> clustertype = {<span class="stringliteral">&quot;clustertype&quot;</span>,SelfDoc::Opt::Integer,<span class="stringliteral">&quot;1&quot;</span>,<span class="stringliteral">&quot;1&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01182"></a>01182                               <span class="stringliteral">&quot;Flag: type of cluster (1-d, etc)&quot;</span>};
<a name="l01183"></a>01183   opts.push_back(clustertype);
<a name="l01184"></a>01184 
<a name="l01185"></a>01185   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> cp = {<span class="stringliteral">&quot;CP&quot;</span>, SelfDoc::Opt::Float,<span class="stringliteral">&quot;16&quot;</span>,<span class="stringliteral">&quot;16&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01186"></a>01186                      <span class="stringliteral">&quot;Penalty for contradicting reference genotype&quot;</span>};
<a name="l01187"></a>01187   opts.push_back(cp);
<a name="l01188"></a>01188   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> hok = {<span class="stringliteral">&quot;Hok&quot;</span>, SelfDoc::Opt::Integer,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01189"></a>01189                       <span class="stringliteral">&quot;Allow Hints to be flipped in genotype&quot;</span>};
<a name="l01190"></a>01190   opts.push_back(hok);
<a name="l01191"></a>01191   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> x_em_thresh = {<span class="stringliteral">&quot;em_thresh&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.05&quot;</span>,<span class="stringliteral">&quot;0.05&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01192"></a>01192                               <span class="stringliteral">&quot;set threshold for em gender routine&quot;</span>};
<a name="l01193"></a>01193   opts.push_back(x_em_thresh);
<a name="l01194"></a>01194   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> x_em_cutoff = {<span class="stringliteral">&quot;em_cutoff&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.5&quot;</span>,<span class="stringliteral">&quot;0.5&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01195"></a>01195                               <span class="stringliteral">&quot;set cutoff for em gender routine&quot;</span>};
<a name="l01196"></a>01196   opts.push_back(x_em_cutoff);
<a name="l01197"></a>01197   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> x_gender_cutoff = {<span class="stringliteral">&quot;gender_cutoff&quot;</span>,SelfDoc::Opt::Float,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0.1&quot;</span>,<span class="stringliteral">&quot;0&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01198"></a>01198                                   <span class="stringliteral">&quot;set cutoff for which gender in em gender call&quot;</span>};
<a name="l01199"></a>01199   opts.push_back(x_gender_cutoff);
<a name="l01200"></a>01200   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> copynumber = {<span class="stringliteral">&quot;copynumber&quot;</span>,SelfDoc::Opt::Boolean,<span class="stringliteral">&quot;false&quot;</span>,<span class="stringliteral">&quot;false&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01201"></a>01201                                   <span class="stringliteral">&quot;Is this method begin called as part of a copynumber run?&quot;</span>};
<a name="l01202"></a>01202   opts.push_back(copynumber);
<a name="l01203"></a>01203   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> trustcheck = {<span class="stringliteral">&quot;TrustCheck&quot;</span>,SelfDoc::Opt::Boolean,<span class="stringliteral">&quot;false&quot;</span>,<span class="stringliteral">&quot;false&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,<span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01204"></a>01204             <span class="stringliteral">&quot;Completely untrusted, blimeshed probesets in samples are no-calls&quot;</span>};
<a name="l01205"></a>01205   opts.push_back(trustcheck);
<a name="l01206"></a>01206 
<a name="l01207"></a>01207   <span class="keywordflow">return</span> opts;
<a name="l01208"></a>01208 }
<a name="l01209"></a>01209 <span class="comment"></span>
<a name="l01210"></a>01210 <span class="comment">/**</span>
<a name="l01211"></a>01211 <span class="comment"> * @brief Supply a little how/what/why about the algorithms this</span>
<a name="l01212"></a>01212 <span class="comment"> * class performs and what parameters it takes.</span>
<a name="l01213"></a>01213 <span class="comment"> * @return SelfDoc</span>
<a name="l01214"></a>01214 <span class="comment"> */</span>
<a name="l01215"></a><a class="code" href="classQuantLabelZ.html#ad4dbeec9c331299c65e6859e16b9e9d8">01215</a> <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> <a class="code" href="classQuantLabelZ.html#ad4dbeec9c331299c65e6859e16b9e9d8" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">QuantLabelZ::explainSelf</a>() {
<a name="l01216"></a>01216   <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc;
<a name="l01217"></a>01217   <a class="code" href="classQuantLabelZ.html#acfd3b0ca4567d0d580602e974bd8de7a" title="Fill in our self documentation from current state.">setupSelfDoc</a>(doc);
<a name="l01218"></a>01218   <span class="keywordflow">return</span> doc;
<a name="l01219"></a>01219 }
<a name="l01220"></a>01220 
<a name="l01221"></a>01221 <span class="comment"></span>
<a name="l01222"></a>01222 <span class="comment">/**</span>
<a name="l01223"></a>01223 <span class="comment"> * @brief This static function should be overridden by child classes</span>
<a name="l01224"></a>01224 <span class="comment"> * to return an object of the correct type initialized correctly</span>
<a name="l01225"></a>01225 <span class="comment"> * with the parameters in the string, string map. All objects</span>
<a name="l01226"></a>01226 <span class="comment"> * created this way should be deleted when finished using.</span>
<a name="l01227"></a>01227 <span class="comment"> *</span>
<a name="l01228"></a>01228 <span class="comment"> * @param param - Map of key/value pairs to initialize the object.</span>
<a name="l01229"></a>01229 <span class="comment"> *</span>
<a name="l01230"></a>01230 <span class="comment"> * @return Pointer toCreate object, this should be sub casted as necessary.</span>
<a name="l01231"></a>01231 <span class="comment"> */</span>
<a name="l01232"></a><a class="code" href="classQuantLabelZ.html#a3063c55509d69c79f1ad5833f9f62691">01232</a> <a class="code" href="classSelfCreate.html" title="Small interface for functions that know how to make an instance of themselves given a map of key...">SelfCreate</a> *<a class="code" href="classQuantLabelZ.html#a3063c55509d69c79f1ad5833f9f62691" title="This static function should be overridden by child classes to return an object of the correct type in...">QuantLabelZ::newObject</a>(std::map&lt;std::string,std::string&gt; &amp;param) {
<a name="l01233"></a>01233   <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc = <a class="code" href="classQuantLabelZ.html#ad4dbeec9c331299c65e6859e16b9e9d8" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">explainSelf</a>();
<a name="l01234"></a>01234   <span class="keywordtype">bool</span> lowprecision = <span class="keyword">false</span>;
<a name="l01235"></a>01235   <span class="keyword">enum</span> <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> <a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909" title="Transform allele estimates.">transform</a> = CCS;
<a name="l01236"></a>01236   <span class="keywordtype">double</span> K=4;
<a name="l01237"></a>01237   <span class="keywordtype">string</span> transformStr;
<a name="l01238"></a>01238   map&lt;string,string&gt;::iterator iter;
<a name="l01239"></a>01239 
<a name="l01240"></a>01240   fillInValue(K, <span class="stringliteral">&quot;K&quot;</span>, param, doc);
<a name="l01241"></a>01241 
<a name="l01242"></a>01242   fillInValue(transformStr, <span class="stringliteral">&quot;transform&quot;</span>, param, doc);
<a name="l01243"></a>01243   transform = <a class="code" href="GenoUtility_8cpp.html#ab5c8b5877412aebfe6eb80db02bbc096" title="Convert a (text insensitive) text repesentation of a transformation into the enumration.">transformationForString</a>(transformStr);
<a name="l01244"></a>01244   <a class="code" href="classQuantLabelZ.html" title="This is the primary class wrapping the bayes_label_routine it handles bookkeeping, gender determination, etc.">QuantLabelZ</a> *labelz = <span class="keyword">new</span> <a class="code" href="classQuantLabelZ.html#a2cea6fc92475386eeca1511393bc6af4" title="probability of BB call">QuantLabelZ</a>(transform, K, lowprecision);
<a name="l01245"></a>01245 
<a name="l01246"></a>01246   <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> ts;
<a name="l01247"></a>01247   ts.<a class="code" href="classsnp__param.html#a956b62e64130e43f407662c31f54b372" title="initialize to some useful defaults">Initialize</a>();
<a name="l01248"></a>01248   <span class="comment">// prior strength</span>
<a name="l01249"></a>01249   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,<span class="stringliteral">&quot;KX&quot;</span>,param,doc);
<a name="l01250"></a>01250   ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>=ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01251"></a>01251   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,<span class="stringliteral">&quot;KH&quot;</span>,param,doc);
<a name="l01252"></a>01252 
<a name="l01253"></a>01253   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,<span class="stringliteral">&quot;V&quot;</span>,param,doc);
<a name="l01254"></a>01254   ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>=ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01255"></a>01255   ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>=ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01256"></a>01256   <span class="comment">// prior centers</span>
<a name="l01257"></a>01257   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,<span class="stringliteral">&quot;BBM&quot;</span>,param,doc);
<a name="l01258"></a>01258   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,<span class="stringliteral">&quot;ABM&quot;</span>,param,doc);
<a name="l01259"></a>01259   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,<span class="stringliteral">&quot;AAM&quot;</span>,param,doc);
<a name="l01260"></a>01260   <span class="comment">// prior y coordinates</span>
<a name="l01261"></a>01261   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>,<span class="stringliteral">&quot;BBY&quot;</span>,param,doc);
<a name="l01262"></a>01262   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>,<span class="stringliteral">&quot;ABY&quot;</span>,param,doc);
<a name="l01263"></a>01263   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>,<span class="stringliteral">&quot;AAY&quot;</span>,param,doc);
<a name="l01264"></a>01264   <span class="comment">// prior variances</span>
<a name="l01265"></a>01265   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,<span class="stringliteral">&quot;AAV&quot;</span>,param,doc);
<a name="l01266"></a>01266   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,<span class="stringliteral">&quot;BBV&quot;</span>,param,doc);
<a name="l01267"></a>01267   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,<span class="stringliteral">&quot;ABV&quot;</span>,param,doc);
<a name="l01268"></a>01268   <span class="comment">// prior y variances</span>
<a name="l01269"></a>01269   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>,<span class="stringliteral">&quot;AAYV&quot;</span>,param,doc);
<a name="l01270"></a>01270   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>,<span class="stringliteral">&quot;ABYV&quot;</span>,param,doc);
<a name="l01271"></a>01271   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>,<span class="stringliteral">&quot;BBYV&quot;</span>,param,doc);
<a name="l01272"></a>01272   <span class="comment">// prior xy covariances</span>
<a name="l01273"></a>01273   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>,<span class="stringliteral">&quot;AAXY&quot;</span>,param,doc);
<a name="l01274"></a>01274   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>,<span class="stringliteral">&quot;ABXY&quot;</span>,param,doc);
<a name="l01275"></a>01275   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>,<span class="stringliteral">&quot;BBXY&quot;</span>,param,doc);
<a name="l01276"></a>01276   <span class="comment">// X-X covariances between clusters</span>
<a name="l01277"></a>01277    fillInValue(ts.prior.xab,<span class="stringliteral">&quot;KXX&quot;</span>,param,doc);
<a name="l01278"></a>01278   fillInValue(ts.prior.xah,<span class="stringliteral">&quot;KAH&quot;</span>,param,doc);
<a name="l01279"></a>01279   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a>,<span class="stringliteral">&quot;KHB&quot;</span>,param,doc);
<a name="l01280"></a>01280   <span class="comment">// Y-Y covariances between clusters</span>
<a name="l01281"></a>01281   fillInValue(ts.prior.yah,<span class="stringliteral">&quot;KYAH&quot;</span>,param,doc);
<a name="l01282"></a>01282   fillInValue(ts.prior.yab,<span class="stringliteral">&quot;KYAB&quot;</span>,param,doc);
<a name="l01283"></a>01283   fillInValue(ts.prior.<a class="code" href="classsnp__distribution.html#a2bd112107fcbccf699b612f410aaa73c" title="cross-correlation terms between clusters y">yhb</a>,<span class="stringliteral">&quot;KYHB&quot;</span>,param,doc);
<a name="l01284"></a>01284   <span class="comment">// wacky parameters</span>
<a name="l01285"></a>01285   fillInValue(ts.<a class="code" href="classsnp__param.html#a5996056f7436f5974c6e6b6bd7925a27" title="do we use common variances for all clusters?">comvar</a>,<span class="stringliteral">&quot;COMVAR&quot;</span>,param,doc);
<a name="l01286"></a>01286   fillInValue(ts.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a>,<span class="stringliteral">&quot;HARD&quot;</span>,param,doc);
<a name="l01287"></a>01287   fillInValue(ts.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a>,<span class="stringliteral">&quot;SB&quot;</span>,param,doc);
<a name="l01288"></a>01288   fillInValue(ts.<a class="code" href="classsnp__param.html#ac6c59acd942ff43d8aa6baa17ad6364b" title="do we use raw labels or posterior distributions?">callmethod</a>,<span class="stringliteral">&quot;CM&quot;</span>,param,doc);
<a name="l01289"></a>01289   fillInValue(ts.<a class="code" href="classsnp__param.html#ab106a36c2ad59a00fdad233b5577d0b6" title="use quick method">bins</a>,<span class="stringliteral">&quot;bins&quot;</span>,param,doc);
<a name="l01290"></a>01290   fillInValue(ts.<a class="code" href="classsnp__param.html#ab10a7ca6260c50da3d78db608ad5faa3" title="use hinted genotypes as reference">hints</a>,<span class="stringliteral">&quot;hints&quot;</span>,param,doc);
<a name="l01291"></a>01291   fillInValue(ts.<a class="code" href="classsnp__param.html#a5bb07d912ac1deddb4be58a2e35f68cf" title="penalty for contradicting a hint">contradictionpenalty</a>, <span class="stringliteral">&quot;CP&quot;</span>,param,doc);
<a name="l01292"></a>01292   fillInValue(ts.<a class="code" href="classsnp__param.html#afc17331895d50682f60ee3912e733c8c" title="allow hints to be flipped in genotype 0,1,2&amp;lt;-&amp;gt;2,1,0">hok</a>,<span class="stringliteral">&quot;Hok&quot;</span>,param,doc);
<a name="l01293"></a>01293   fillInValue(ts.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a>,<span class="stringliteral">&quot;mix&quot;</span>,param,doc);
<a name="l01294"></a>01294   fillInValue(ts.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a>,<span class="stringliteral">&quot;bic&quot;</span>,param,doc);
<a name="l01295"></a>01295   fillInValue(ts.<a class="code" href="classsnp__param.html#ab30f88b4fe7b2476acf77727745ae6ab" title="penalty for FLD separation too low">CSepPen</a>,<span class="stringliteral">&quot;CSepPen&quot;</span>,param,doc);
<a name="l01296"></a>01296   fillInValue(ts.<a class="code" href="classsnp__param.html#a3fad707f15cabcac4f8c7097781fd9b8" title="stop penalizing using Geman-McClure">CSepThr</a>,<span class="stringliteral">&quot;CSepThr&quot;</span>,param,doc);
<a name="l01297"></a>01297   fillInValue(ts.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>,<span class="stringliteral">&quot;lambda&quot;</span>,param,doc);
<a name="l01298"></a>01298   fillInValue(ts.<a class="code" href="classsnp__param.html#aabee7993a0cd2545731ff91f5d6cce4e" title="how much do we allow clusters to shift from prior">wobble</a>,<span class="stringliteral">&quot;wobble&quot;</span>,param,doc);
<a name="l01299"></a>01299   fillInValue(ts.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a>,<span class="stringliteral">&quot;copyqc&quot;</span>,param,doc);
<a name="l01300"></a>01300   fillInValue(ts.<a class="code" href="classsnp__param.html#a1095b75124fb8658e9fafec4140d854d" title="how to handle copy number outlier detection">copytype</a>,<span class="stringliteral">&quot;copytype&quot;</span>,param,doc);
<a name="l01301"></a>01301   fillInValue(ts.<a class="code" href="classsnp__param.html#a5137d727c65f534313add6d70f3be81b" title="do I handle multiple cluster dimensions">clustertype</a>,<span class="stringliteral">&quot;clustertype&quot;</span>,param,doc);
<a name="l01302"></a>01302   fillInValue(ts.<a class="code" href="classsnp__param.html#ad0561d07114f9c66a173dd3b904492a4" title="uniform density to compare weird data points against">ocean</a>,<span class="stringliteral">&quot;ocean&quot;</span>,param,doc);
<a name="l01303"></a>01303   fillInValue(ts.<a class="code" href="classsnp__param.html#a379a6b8e6a44cb1b2f519f74932d8af3" title="increase variance by uncertainty in mean location when making calls">inflatePRA</a>,<span class="stringliteral">&quot;inflatePRA&quot;</span>,param,doc);
<a name="l01304"></a>01304   fillInValue(ts.<a class="code" href="classsnp__param.html#a01576510bc234b4084eb61d2da4d9f00" title="ensure Het &amp;gt; average Hom Y by at least epsilon">IsoHetY</a>,<span class="stringliteral">&quot;IsoHetY&quot;</span>,param,doc);
<a name="l01305"></a>01305   <span class="comment">//printf(&quot;BBM: %f AAM: %f CM: %d\n&quot;,ts.bb.m,ts.aa.m,ts.callmethod);</span>
<a name="l01306"></a>01306   <span class="comment">//ts.aa.Dump();</span>
<a name="l01307"></a>01307   <span class="comment">//ts.ab.Dump();</span>
<a name="l01308"></a>01308   <span class="comment">//ts.bb.Dump();</span>
<a name="l01309"></a>01309   labelz-&gt;<a class="code" href="classQuantLabelZ.html#aac7e6719fad7743dd468cad018f621c2" title="Set the snp-parameters for this run.">setLabelParam</a>(ts);
<a name="l01310"></a>01310   <span class="keywordtype">double</span> t_em_cutoff,t_em_thresh,t_gender_cutoff;
<a name="l01311"></a>01311   fillInValue(t_em_cutoff,<span class="stringliteral">&quot;em_cutoff&quot;</span>,param,doc);
<a name="l01312"></a>01312   fillInValue(t_em_thresh,<span class="stringliteral">&quot;em_thresh&quot;</span>,param,doc);
<a name="l01313"></a>01313   fillInValue(t_gender_cutoff,<span class="stringliteral">&quot;gender_cutoff&quot;</span>,param,doc);
<a name="l01314"></a>01314   labelz-&gt;<a class="code" href="classQuantLabelZ.html#acaef334285a1263ca32c6070f620a184" title="Sets the EM specific parameters for calling genders from chrX snps.">setGenderParam</a>(t_em_thresh,t_em_cutoff,t_gender_cutoff);
<a name="l01315"></a>01315   <span class="keywordtype">double</span> score;
<a name="l01316"></a>01316   fillInValue(score,<span class="stringliteral">&quot;MS&quot;</span>,param,doc);
<a name="l01317"></a>01317   labelz-&gt;<a class="code" href="classQuantLabelZ.html#ac97957acc212aebea23d9f2fbdeb1702" title="sets the max-score used for this analysis method">setMaxScore</a>(score);
<a name="l01318"></a>01318   <span class="keywordtype">int</span> overrideflag;
<a name="l01319"></a>01319   fillInValue(overrideflag,<span class="stringliteral">&quot;override&quot;</span>,param,doc);
<a name="l01320"></a>01320   labelz-&gt;<a class="code" href="classQuantLabelZ.html#acd9b42625edf51edc84489627065f649" title="Set override flag.">setOverride</a>(overrideflag);
<a name="l01321"></a>01321   <span class="keywordtype">bool</span> bCopyNumber = <span class="keyword">false</span>;
<a name="l01322"></a>01322   fillInValue(bCopyNumber,<span class="stringliteral">&quot;copynumber&quot;</span>,param,doc);
<a name="l01323"></a>01323   labelz-&gt;setCopyNumber(bCopyNumber);
<a name="l01324"></a>01324   <span class="keywordtype">bool</span> trust_check = <span class="keyword">false</span>;
<a name="l01325"></a>01325   fillInValue(trust_check,<span class="stringliteral">&quot;TrustCheck&quot;</span>,param,doc);
<a name="l01326"></a>01326   labelz-&gt;<a class="code" href="classQuantLabelZ.html#aad45cb95c07349cdfb3e80c88b068679" title="Set TrustCheck flag.">setTrustCheck</a>(trust_check);
<a name="l01327"></a>01327   <span class="keywordflow">return</span> labelz;
<a name="l01328"></a>01328 }
<a name="l01329"></a>01329 <span class="comment"></span>
<a name="l01330"></a>01330 <span class="comment">//////////////////////////////</span>
<a name="l01331"></a>01331 <span class="comment">//////////////////////////////</span>
<a name="l01332"></a>01332 <span class="comment"></span><span class="comment">//</span>
<a name="l01333"></a>01333 <span class="comment">//</span>
<a name="l01334"></a>01334 <span class="comment">// QuantLableZ IO functions.</span>
<a name="l01335"></a>01335 <span class="comment">//</span>
<a name="l01336"></a>01336 
<a name="l01337"></a>01337 <span class="comment"></span>
<a name="l01338"></a>01338 <span class="comment">//////////</span>
<a name="l01339"></a>01339 <span class="comment"></span><span class="comment"></span>
<a name="l01340"></a>01340 <span class="comment">/**</span>
<a name="l01341"></a>01341 <span class="comment"> * Writes normalization functions to a file</span>
<a name="l01342"></a>01342 <span class="comment"> *</span>
<a name="l01343"></a>01343 <span class="comment"> * @param nMap - normalization functions</span>
<a name="l01344"></a>01344 <span class="comment"> * @param fileName - the file to write</span>
<a name="l01345"></a>01345 <span class="comment"> */</span>
<a name="l01346"></a><a class="code" href="classQuantLabelZ.html#a12c40ec12ddbaabcd129750af201e5e8">01346</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a12c40ec12ddbaabcd129750af201e5e8" title="writes out the contrast normalization functions">QuantLabelZ::writeNormMap</a>(std::vector&lt;NormalizationPredictor&gt; &amp;nMap,
<a name="l01347"></a>01347                                <span class="keyword">const</span> std::string&amp; fileName)  {
<a name="l01348"></a>01348   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l01349"></a>01349   <span class="comment">//</span>
<a name="l01350"></a>01350   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,0,<span class="stringliteral">&quot;Sample&quot;</span>);
<a name="l01351"></a>01351   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,1,<span class="stringliteral">&quot;FitBB&quot;</span>);
<a name="l01352"></a>01352   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,2,<span class="stringliteral">&quot;FitAB&quot;</span>);
<a name="l01353"></a>01353   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,3,<span class="stringliteral">&quot;FitAA&quot;</span>);
<a name="l01354"></a>01354   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,4,<span class="stringliteral">&quot;sigma&quot;</span>);
<a name="l01355"></a>01355   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,5,<span class="stringliteral">&quot;Cov&quot;</span>);
<a name="l01356"></a>01356   tsv.<a class="code" href="classaffx_1_1TsvFile.html#aadc22863a63b9c6588b0f2dc61875809" title="User methods.">defineColumn</a>(0,6,<span class="stringliteral">&quot;Targ&quot;</span>);
<a name="l01357"></a>01357   <span class="comment">//</span>
<a name="l01358"></a>01358   tsv.<a class="code" href="classaffx_1_1TsvFile.html#a1903489904b4e2f8f2a262b8f10cdb0e" title="Set the options up for &amp;quot;tsv&amp;quot; format and write the tsv-v1 header ///.">writeTsv_v1</a>(fileName);
<a name="l01359"></a>01359 
<a name="l01360"></a>01360   <span class="comment">//</span>
<a name="l01361"></a>01361   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0;i&lt;nMap.size(); i++) {
<a name="l01362"></a>01362     <span class="comment">//</span>
<a name="l01363"></a>01363     tsv.set(0,0,i);
<a name="l01364"></a>01364     <span class="comment">// these are all vectors values.</span>
<a name="l01365"></a>01365     tsv.set(0,1,nMap[i].fitBB);
<a name="l01366"></a>01366     tsv.set(0,2,nMap[i].fitAB);
<a name="l01367"></a>01367     tsv.set(0,3,nMap[i].fitAA);
<a name="l01368"></a>01368     tsv.set(0,4,nMap[i].sigma);
<a name="l01369"></a>01369     tsv.set(0,5,nMap[i].CenterCovariates);
<a name="l01370"></a>01370     tsv.set(0,6,nMap[i].TargetCenter);
<a name="l01371"></a>01371     <span class="comment">//</span>
<a name="l01372"></a>01372     tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2a459be42bea9ea3f180773370ec0de7" title="Write a level of data to the file. ///.">writeLevel</a>(0);
<a name="l01373"></a>01373   }
<a name="l01374"></a>01374   tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l01375"></a>01375 }
<a name="l01376"></a>01376 <span class="comment"></span>
<a name="l01377"></a>01377 <span class="comment">//////////</span>
<a name="l01378"></a>01378 <span class="comment"></span>
<a name="l01379"></a>01379 <span class="keywordtype">void</span> QuantLabelZ::writeNormSummaryTsv(<span class="keyword">const</span> std::string&amp; fileName, 
<a name="l01380"></a>01380                                       std::vector&lt;std::string&gt; &amp;headers,
<a name="l01381"></a>01381                                       affx::TsvReport::TsvReportFmt_t format) {
<a name="l01382"></a>01382   m_NormSummaryTsv.setFilename(fileName);
<a name="l01383"></a>01383   m_NormSummaryTsv.setFormat(format);
<a name="l01384"></a>01384   <span class="comment">//</span>
<a name="l01385"></a>01385   <span class="comment">// @todo: Is this the best format for the report? -jhg</span>
<a name="l01386"></a>01386   <span class="comment">// we could skip the &quot;Delta&quot; or &quot;Size&quot; by using more columns or seperate reports.</span>
<a name="l01387"></a>01387   m_NormSummaryTsv.addHeaderComment(<span class="stringliteral">&quot;minimal file for normalized values&quot;</span>);
<a name="l01388"></a>01388   <span class="comment">// the name as a string</span>
<a name="l01389"></a>01389   m_NormSummaryTsv.defineStringColumn(0,0,<span class="stringliteral">&quot;probeset_id&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l01390"></a>01390   <span class="comment">// &quot;Delta&quot; or &quot;Size&quot;</span>
<a name="l01391"></a>01391   m_NormSummaryTsv.defineStringColumn(0,1,<span class="stringliteral">&quot;datatype&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l01392"></a>01392   <span class="comment">// </span>
<a name="l01393"></a>01393   m_NormSummaryTsv.defineColumns(<a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(headers),affx::FILE5_DTYPE_DOUBLE,0);
<a name="l01394"></a>01394   <span class="comment">//</span>
<a name="l01395"></a>01395   m_NormSummaryTsv.writeTsv_v1();
<a name="l01396"></a>01396 }
<a name="l01397"></a>01397 <span class="comment"></span>
<a name="l01398"></a>01398 <span class="comment">/**</span>
<a name="l01399"></a>01399 <span class="comment"> * Writes a summary of the transformed values to a file - i.e as set up for genotyping, not expression</span>
<a name="l01400"></a>01400 <span class="comment"> *</span>
<a name="l01401"></a>01401 <span class="comment"> * @param m_NormSummaryOut - output stream</span>
<a name="l01402"></a>01402 <span class="comment"> * @param m_AValues - first coordinate of transformed data - always &quot;contrast&quot;-like values</span>
<a name="l01403"></a>01403 <span class="comment"> * @param m_BValues - second coordinate of transformed data - always &quot;strength&quot;-like values</span>
<a name="l01404"></a>01404 <span class="comment"> * @param m_ProbeSetName - what is our SNP?</span>
<a name="l01405"></a>01405 <span class="comment"> */</span>
<a name="l01406"></a><a class="code" href="classQuantLabelZ.html#a4f185569daf922ace4d429c9f5c6752b">01406</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a4f185569daf922ace4d429c9f5c6752b" title="Writes a summary of the transformed values to a file - i.e as set up for genotyping, not expression.">QuantLabelZ::writeNormSummaryValue</a>(<span class="keyword">const</span> std::string&amp; m_ProbeSetName,
<a name="l01407"></a>01407                                         <span class="keyword">const</span> vector&lt;double&gt;&amp; m_AValues,
<a name="l01408"></a>01408                                         <span class="keyword">const</span> vector&lt;double&gt;&amp; m_BValues) {
<a name="l01409"></a>01409   <span class="comment">// delta line first.</span>
<a name="l01410"></a>01410   m_NormSummaryTsv.set_string(0,0,m_ProbeSetName);
<a name="l01411"></a>01411   m_NormSummaryTsv.set_string(0,1,<span class="stringliteral">&quot;Delta&quot;</span>);
<a name="l01412"></a>01412   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;m_AValues.size(); i++) {
<a name="l01413"></a>01413     m_NormSummaryTsv.set_d(0,2+i,m_AValues[i]);
<a name="l01414"></a>01414   }
<a name="l01415"></a>01415   m_NormSummaryTsv.writeLevel(0);
<a name="l01416"></a>01416 
<a name="l01417"></a>01417   <span class="comment">// Then the &quot;Size&quot; line.</span>
<a name="l01418"></a>01418   m_NormSummaryTsv.set_string(0,0,m_ProbeSetName);
<a name="l01419"></a>01419   m_NormSummaryTsv.set_string(0,1,<span class="stringliteral">&quot;Size&quot;</span>);
<a name="l01420"></a>01420   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;m_BValues.size(); i++) {
<a name="l01421"></a>01421     m_NormSummaryTsv.set_d(0,2+i,m_BValues[i]);
<a name="l01422"></a>01422   }
<a name="l01423"></a>01423   m_NormSummaryTsv.writeLevel(0);
<a name="l01424"></a>01424 }
<a name="l01425"></a>01425 <span class="comment"></span>
<a name="l01426"></a>01426 <span class="comment">//////////</span>
<a name="l01427"></a>01427 <span class="comment"></span><span class="comment"></span>
<a name="l01428"></a>01428 <span class="comment">/**</span>
<a name="l01429"></a>01429 <span class="comment"> * Write header to the selection file so people don&#39;t have to look up everything</span>
<a name="l01430"></a>01430 <span class="comment"> *</span>
<a name="l01431"></a>01431 <span class="comment"> * @param m_Out - the file for a header</span>
<a name="l01432"></a>01432 <span class="comment"> */</span>
<a name="l01433"></a><a class="code" href="classQuantLabelZ.html#ad7cae32c6830efa7475fb948f2442682">01433</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ad7cae32c6830efa7475fb948f2442682" title="Write header to the selection file so people don&amp;#39;t have to look up everything.">QuantLabelZ::writeSnpProbeTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l01434"></a>01434                                    affx::TsvReport::TsvReportFmt_t format) {
<a name="l01435"></a>01435   <span class="comment">//</span>
<a name="l01436"></a>01436   m_SnpProbeTsv.setFilename(fileName);
<a name="l01437"></a>01437   m_SnpProbeTsv.setFormat(format);
<a name="l01438"></a>01438   <span class="comment">// write out a header in the file</span>
<a name="l01439"></a>01439   <span class="comment">// describing what we actually are doing</span>
<a name="l01440"></a>01440   <span class="keywordtype">int</span> c=0;
<a name="l01441"></a>01441   m_SnpProbeTsv.defineStringColumn(0,c++,<span class="stringliteral">&quot;ProbeSet&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l01442"></a>01442   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;A&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01443"></a>01443   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;ProbeNum&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01444"></a>01444   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;ProbeAindex&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01445"></a>01445   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;ProbeBindex&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01446"></a>01446   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;HomLogisticCorrect&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01447"></a>01447   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;HetLogisticCorrect&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01448"></a>01448   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;LogisticCorrect&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01449"></a>01449   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;MeanAA&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01450"></a>01450   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;MeanAB&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01451"></a>01451   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;MeanBB&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01452"></a>01452   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;nAA&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01453"></a>01453   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;nAB&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01454"></a>01454   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;nBB&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01455"></a>01455   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;Stdev&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01456"></a>01456   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;FLDAH&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01457"></a>01457   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;FLDHB&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01458"></a>01458   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;FLDAB&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01459"></a>01459   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;Entropy&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01460"></a>01460   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;AIC&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01461"></a>01461   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;B0&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01462"></a>01462   m_SnpProbeTsv.defineColumn(0,c++,<span class="stringliteral">&quot;B1&quot;</span>,affx::FILE5_DTYPE_DOUBLE);
<a name="l01463"></a>01463   <span class="comment">//</span>
<a name="l01464"></a>01464   m_SnpProbeTsv.writeTsv_v1();
<a name="l01465"></a>01465 }
<a name="l01466"></a>01466 <span class="comment"></span>
<a name="l01467"></a>01467 <span class="comment">/**</span>
<a name="l01468"></a>01468 <span class="comment"> * Writeprobe writes the summaries to the file</span>
<a name="l01469"></a>01469 <span class="comment"> *</span>
<a name="l01470"></a>01470 <span class="comment"> * @param m_Out - stream required</span>
<a name="l01471"></a>01471 <span class="comment"> * @param probenum - what probe we are</span>
<a name="l01472"></a>01472 <span class="comment"> * @param Aid - probe identity for A</span>
<a name="l01473"></a>01473 <span class="comment"> * @param Bid - probe identity for B</span>
<a name="l01474"></a>01474 <span class="comment"> * @param Summaries - summary information for probe fit</span>
<a name="l01475"></a>01475 <span class="comment"> * @param m_Ax - A data (fitted)</span>
<a name="l01476"></a>01476 <span class="comment"> * @param m_Bx - B data (fitted - not used)</span>
<a name="l01477"></a>01477 <span class="comment"> * @param m_ProbeSetName</span>
<a name="l01478"></a>01478 <span class="comment"> */</span>
<a name="l01479"></a><a class="code" href="classQuantLabelZ.html#a16617ac9ce4b9229c84c6f26af40cc55">01479</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a16617ac9ce4b9229c84c6f26af40cc55" title="Writeprobe writes the summaries to the file.">QuantLabelZ::writeSnpProbeValue</a>(<a class="code" href="classaffx_1_1TsvReport.html" title="Unifies the presentation of file/TsvFile and file ///.">affx::TsvReport</a>&amp; tsv,
<a name="l01480"></a>01480                                      <span class="keywordtype">int</span> probenum,
<a name="l01481"></a>01481                                      <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> Aid,
<a name="l01482"></a>01482                                      <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> Bid,
<a name="l01483"></a>01483                                      <span class="keyword">const</span> vector&lt;double&gt; &amp;Summaries,
<a name="l01484"></a>01484                                      <span class="keyword">const</span> vector&lt;double&gt; &amp;m_Ax,
<a name="l01485"></a>01485                                      <span class="keyword">const</span> vector&lt;double&gt; &amp;m_Bx,
<a name="l01486"></a>01486                                      <span class="keywordtype">string</span>&amp; m_ProbeSetName) {
<a name="l01487"></a>01487   <span class="comment">// todo add tsvdata here.</span>
<a name="l01488"></a>01488   <span class="keywordtype">int</span> c=0;
<a name="l01489"></a>01489   tsv.set_string(0,c++,m_ProbeSetName);
<a name="l01490"></a>01490   tsv.set_string(0,c++,<span class="stringliteral">&quot;A&quot;</span>); <span class="comment">// This is defined as double above.</span>
<a name="l01491"></a>01491   tsv.set_d(0,c++,probenum);
<a name="l01492"></a>01492   tsv.set_d(0,c++,Aid);
<a name="l01493"></a>01493   tsv.set_d(0,c++,Bid);
<a name="l01494"></a>01494   <span class="comment">//</span>
<a name="l01495"></a>01495   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;Summaries.size(); i++) {
<a name="l01496"></a>01496     tsv.set_d(0,c++,Summaries[i]);
<a name="l01497"></a>01497   }
<a name="l01498"></a>01498   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0;i&lt;m_Ax.size();i++) {
<a name="l01499"></a>01499     tsv.set_d(0,c++,m_Ax[i]);
<a name="l01500"></a>01500   }
<a name="l01501"></a>01501   <span class="comment">//</span>
<a name="l01502"></a>01502   tsv.writeLevel(0);
<a name="l01503"></a>01503 }
<a name="l01504"></a>01504 <span class="comment"></span>
<a name="l01505"></a>01505 <span class="comment">//////////</span>
<a name="l01506"></a>01506 <span class="comment"></span><span class="comment"></span>
<a name="l01507"></a>01507 <span class="comment">/// @brief Pick the output version based on the file format.  (txt=&gt;v1,a5=&gt;v2)</span>
<a name="l01508"></a><a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9">01508</a> <span class="comment"></span><span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">QuantLabelZ::writeSnpPosteriorTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l01509"></a>01509                                        affx::TsvReport::TsvReportFmt_t format_file) {
<a name="l01510"></a>01510   <span class="keywordtype">int</span> format_ver = 1;
<a name="l01511"></a>01511   <span class="keywordflow">if</span> (format_file==TsvReport::FMT_TSV) {
<a name="l01512"></a>01512     format_ver=1;
<a name="l01513"></a>01513   }
<a name="l01514"></a>01514   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (format_file==TsvReport::FMT_A5) {
<a name="l01515"></a>01515           <span class="keywordflow">if</span> (m_bCopyNumber) {
<a name="l01516"></a>01516       format_ver = 2;
<a name="l01517"></a>01517     }
<a name="l01518"></a>01518           <span class="keywordflow">else</span> {
<a name="l01519"></a>01519       format_ver = 3;
<a name="l01520"></a>01520     }
<a name="l01521"></a>01521   }
<a name="l01522"></a>01522   <span class="keywordflow">else</span> {
<a name="l01523"></a>01523     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;writeSnpPosteriorTsv&quot;</span>);
<a name="l01524"></a>01524   }
<a name="l01525"></a>01525   <span class="comment">//</span>
<a name="l01526"></a>01526   <a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">writeSnpPosteriorTsv</a>(fileName,format_file,format_ver);
<a name="l01527"></a>01527 }
<a name="l01528"></a>01528 
<a name="l01529"></a>01529 <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">QuantLabelZ::writeSnpPosteriorTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l01530"></a>01530                                        affx::TsvReport::TsvReportFmt_t format_file,
<a name="l01531"></a>01531                                        <span class="keywordtype">int</span> format_ver) {
<a name="l01532"></a>01532   <span class="comment">// remember the version being written (1 or 2)</span>
<a name="l01533"></a>01533   m_SnpPosteriorTsv_ver=format_ver;
<a name="l01534"></a>01534   m_SnpPosteriorName = fileName;
<a name="l01535"></a>01535   m_PosteriorFormat = format_file;
<a name="l01536"></a>01536 }
<a name="l01537"></a>01537 <span class="comment">//   //</span>
<a name="l01538"></a>01538 <span class="comment">//     std::vector&lt;std::string&gt;::const_iterator keyIx, paramIx;</span>
<a name="l01539"></a>01539 <span class="comment">//   //</span>
<a name="l01540"></a>01540 <span class="comment">//     Verbose::out(1, &quot;Got &quot; + ToStr(m_Info.m_ParamValues.size()) + &quot; headers to write.&quot;);</span>
<a name="l01541"></a>01541 <span class="comment">//   for (keyIx = m_Info.m_ParamNames.begin(), paramIx = m_Info.m_ParamValues.begin();</span>
<a name="l01542"></a>01542 <span class="comment">//       keyIx != m_Info.m_ParamNames.end() &amp;&amp; paramIx != m_Info.m_ParamValues.end();</span>
<a name="l01543"></a>01543 <span class="comment">//       ++keyIx, ++paramIx) {</span>
<a name="l01544"></a>01544 <span class="comment">//     // @todo should we using the AffymetrixParameterConsts.h #defined values?</span>
<a name="l01545"></a>01545 <span class="comment">//     m_SnpPosteriorTsv.addHeader(&quot;affymetrix-algorithm-param-&quot; + *keyIx, *paramIx);</span>
<a name="l01546"></a>01546 <span class="comment">//   }</span>
<a name="l01547"></a>01547 <span class="comment">//   for (keyIx = m_Info.m_ClientInfoNames.begin(), paramIx = m_Info.m_ClientInfoValues.begin();</span>
<a name="l01548"></a>01548 <span class="comment">//       keyIx != m_Info.m_ClientInfoNames.end() &amp;&amp; paramIx != m_Info.m_ClientInfoValues.end();</span>
<a name="l01549"></a>01549 <span class="comment">//       ++keyIx, ++paramIx) {</span>
<a name="l01550"></a>01550 <span class="comment">//     // @todo should we using the AffymetrixParameterConsts.h #defined values?</span>
<a name="l01551"></a>01551 <span class="comment">//     m_SnpPosteriorTsv.addHeader(&quot;affymetrix-application-meta-data-info-&quot; + *keyIx, *paramIx);</span>
<a name="l01552"></a>01552 <span class="comment">//   }</span>
<a name="l01553"></a>01553 <span class="comment">//   QuantLabelZ__writeSnpPosteriorTsv(m_SnpPosteriorTsv,m_SnpPosteriorName,format_file,format_ver);</span>
<a name="l01554"></a>01554 <span class="comment">//}</span>
<a name="l01555"></a>01555 <span class="comment"></span>
<a name="l01556"></a>01556 <span class="comment">/**</span>
<a name="l01557"></a>01557 <span class="comment"> *  Writes one specific posterior to a file along with snp name</span>
<a name="l01558"></a>01558 <span class="comment"> *</span>
<a name="l01559"></a>01559 <span class="comment"> * @param tsp - snp specific parameters, including posterior distribution of cluster centers</span>
<a name="l01560"></a>01560 <span class="comment"> * @param out - output stream</span>
<a name="l01561"></a>01561 <span class="comment"> * @param id - snp id</span>
<a name="l01562"></a>01562 <span class="comment"> * @param fieldDelim - always tab</span>
<a name="l01563"></a>01563 <span class="comment"> * @param sep - always comma</span>
<a name="l01564"></a>01564 <span class="comment"> */</span>
<a name="l01565"></a><a class="code" href="classQuantLabelZ.html#a35793acf2ae3d85c79930cadb62beb7a">01565</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a35793acf2ae3d85c79930cadb62beb7a" title="Writes one specific posterior to a file along with snp name.">QuantLabelZ::writeSnpPosteriorValue</a>(<span class="keyword">const</span> std::string&amp; probeset_id,
<a name="l01566"></a>01566                                          <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;tsp) {
<a name="l01567"></a>01567   QuantLabelZ__writeSnpPosteriorValue(m_SnpPosteriorTsv,m_SnpPosteriorTsv_ver,
<a name="l01568"></a>01568                                       probeset_id,tsp);
<a name="l01569"></a>01569 }
<a name="l01570"></a>01570 
<a name="l01571"></a><a class="code" href="classQuantLabelZ.html#ab4a3970a1d381a0885d82ebc66ddb3ca">01571</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ab4a3970a1d381a0885d82ebc66ddb3ca" title="read in the snp specific priors from a file">QuantLabelZ::readSnpPriorMap</a>(<span class="keyword">const</span> std::string&amp; fileName) {
<a name="l01572"></a>01572   QuantLabelZ__readSnpPriorMap(m_vectSnpPriors,fileName);
<a name="l01573"></a>01573 }
<a name="l01574"></a>01574 <span class="keywordtype">void</span> QuantLabelZ::readSnpPriorMap_tsv5(<a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5) {
<a name="l01575"></a>01575   QuantLabelZ__readSnpPriorMap_tsv5(m_vectSnpPriors,tsv5);
<a name="l01576"></a>01576 }
<a name="l01577"></a>01577 
<a name="l01578"></a><a class="code" href="classQuantLabelZ.html#a9bac1828828da05da22eacadab7a322b">01578</a> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a9bac1828828da05da22eacadab7a322b" title="Custom configuration for this QuantMethod.">QuantLabelZ::setParameters</a>(<a class="code" href="classPsBoard.html">PsBoard</a> &amp;board) {
<a name="l01579"></a>01579   <a class="code" href="classOptions.html" title="base class for tracking class options">Options</a> *o = board.getOptions();
<a name="l01580"></a>01580   vector&lt;string&gt; celFiles = o-&gt;<a class="code" href="classOptions.html#ab20d0083424d9d9a8b94da46f487f57a" title="Return a vector of values for a particular option.">getOptVector</a>(<span class="stringliteral">&quot;cels&quot;</span>);
<a name="l01581"></a>01581   <span class="keywordtype">string</span> outDir = o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;out-dir&quot;</span>);
<a name="l01582"></a>01582   <span class="keywordflow">if</span> (o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;read-models-brlmmp&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l01583"></a>01583       <span class="keywordflow">if</span> (!o-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;brlmmp-models-non-sequential&quot;</span>)) {
<a name="l01584"></a>01584           m_SequentialModelFile = o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;read-models-brlmmp&quot;</span>);
<a name="l01585"></a>01585           m_SequentialModelTsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(m_SequentialModelFile);
<a name="l01586"></a>01586           m_SequentialModelTsv.bind(0,<span class="stringliteral">&quot;id&quot;</span>, &amp;m_Id, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l01587"></a>01587           m_SequentialModelTsv.bind(0,<span class="stringliteral">&quot;BB&quot;</span>, &amp;m_BB, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l01588"></a>01588           m_SequentialModelTsv.bind(0,<span class="stringliteral">&quot;AB&quot;</span>, &amp;m_AB, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l01589"></a>01589           m_SequentialModelTsv.bind(0,<span class="stringliteral">&quot;AA&quot;</span>, &amp;m_AA, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l01590"></a>01590           
<a name="l01591"></a>01591           <span class="keywordtype">int</span> fc_cidx =  m_SequentialModelTsv.<a class="code" href="classaffx_1_1TsvFile.html#a2dbd33415a815e221ae61cf3feac5500" title="Covert a column index to a column index ///.">cname2cidx</a>(0,<span class="stringliteral">&quot;CV&quot;</span>);
<a name="l01592"></a>01592           <span class="keywordflow">if</span> (fc_cidx!=<a class="code" href="namespaceaffx.html#ad00847b6f358e7269f493808ee6bd395a8de9b9f4b38e163987bfe315dba4c63e" title="The value was not found.">affx::TSV_ERR_NOTFOUND</a>) {
<a name="l01593"></a>01593               m_SequentialModelTsv.bind(0, <span class="stringliteral">&quot;CV&quot;</span>, &amp;m_CV, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l01594"></a>01594           }
<a name="l01595"></a>01595           <span class="keywordflow">else</span> {
<a name="l01596"></a>01596               m_CV = <span class="stringliteral">&quot;0,0,0&quot;</span>;
<a name="l01597"></a>01597           }
<a name="l01598"></a>01598           
<a name="l01599"></a>01599       }
<a name="l01600"></a>01600       <span class="keywordflow">else</span> {
<a name="l01601"></a>01601           <a class="code" href="classQuantLabelZ.html#ab4a3970a1d381a0885d82ebc66ddb3ca" title="read in the snp specific priors from a file">readSnpPriorMap</a>(o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;read-models-brlmmp&quot;</span>));
<a name="l01602"></a>01602       }
<a name="l01603"></a>01603   }
<a name="l01604"></a>01604   
<a name="l01605"></a>01605   <span class="comment">// @todo refactor - put current analysis path name on bboard</span>
<a name="l01606"></a>01606   <span class="keywordtype">string</span> outfile = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(outDir,<a class="code" href="classQuantLabelZ.html#a43cabd662be310c394c4317afcc68286" title="What is the name of the quantification method?">getType</a>() + <span class="stringliteral">&quot;.snp-posteriors&quot;</span>);
<a name="l01607"></a>01607   <span class="keywordtype">bool</span> writeModels = o-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;write-models&quot;</span>);
<a name="l01608"></a>01608   <span class="keywordflow">if</span> (writeModels) {
<a name="l01609"></a>01609     <a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">writeSnpPosteriorTsv</a>(outfile,affx::TsvReport::FMT_TSV);
<a name="l01610"></a>01610     <span class="keywordtype">string</span> modelFile = outfile + <span class="stringliteral">&quot;.txt&quot;</span>;
<a name="l01611"></a>01611     board.set(<span class="stringliteral">&quot;model-file-written&quot;</span>, modelFile);
<a name="l01612"></a>01612   }
<a name="l01613"></a>01613   <span class="comment">// @todo refactor - put current analysis path name on bboard</span>
<a name="l01614"></a>01614   outfile =<a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(outDir,<a class="code" href="classQuantLabelZ.html#a43cabd662be310c394c4317afcc68286" title="What is the name of the quantification method?">getType</a>() + <span class="stringliteral">&quot;.normalized-summary&quot;</span>);
<a name="l01615"></a>01615   <span class="keywordflow">if</span> (o-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;summaries&quot;</span>)) {
<a name="l01616"></a>01616     writeNormSummaryTsv(outfile, celFiles,affx::TsvReport::FMT_TSV);
<a name="l01617"></a>01617   }
<a name="l01618"></a>01618   <span class="keywordflow">if</span> (o-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;select-probes&quot;</span>)) {
<a name="l01619"></a>01619     <span class="comment">// @todo refactor - put current analysis path name on bboard</span>
<a name="l01620"></a>01620     outfile =<a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(outDir,<a class="code" href="classQuantLabelZ.html#a43cabd662be310c394c4317afcc68286" title="What is the name of the quantification method?">getType</a>() + <span class="stringliteral">&quot;.select-probes&quot;</span>);
<a name="l01621"></a>01621     <a class="code" href="classQuantLabelZ.html#ad7cae32c6830efa7475fb948f2442682" title="Write header to the selection file so people don&amp;#39;t have to look up everything.">writeSnpProbeTsv</a>(outfile,affx::TsvReport::FMT_TSV);
<a name="l01622"></a>01622   }
<a name="l01623"></a>01623 
<a name="l01624"></a>01624   <span class="comment">// Add quantification method.</span>
<a name="l01625"></a>01625   <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> qFactory;
<a name="l01626"></a>01626   <span class="keywordtype">string</span> qMethodSpec = o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;qmethod-spec&quot;</span>);
<a name="l01627"></a>01627   <a class="code" href="classQuantMethod.html" title="QuantMethod - Interface for computing quantification summaries from PM intensities grouped into probe...">QuantMethod</a> *qMethod = qFactory.<a class="code" href="classQuantMethodFactory.html#ab0f3db6b9aec42ff32ae69b9791329c8" title="Create a pointer to a new QuantMethod object as described in the string specification.">quantMethodForString</a>(qMethodSpec, board);
<a name="l01628"></a>01628   <a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *eMethod = NULL;
<a name="l01629"></a>01629   eMethod= <span class="keyword">static_cast&lt;</span><a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *<span class="keyword">&gt;</span>(qMethod);
<a name="l01630"></a>01630   <span class="keywordflow">if</span> (eMethod == NULL) {
<a name="l01631"></a>01631     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Couldn&#39;t make a QuantExprMethod for specification: &quot;</span> + qMethodSpec);
<a name="l01632"></a>01632   }
<a name="l01633"></a>01633   <a class="code" href="classQuantLabelZ.html#ac8c2cc4ab6dc319d4c0bff9b3f299642" title="Set the quantification method to use for summarizing individual alleles.">setQuantExprMethod</a>(eMethod);
<a name="l01634"></a>01634   
<a name="l01635"></a>01635   <a class="code" href="classGenotypeInfo.html">GenotypeInfo</a> *gInfo = board.getGenotypeInfo();
<a name="l01636"></a>01636   <span class="keywordflow">if</span> (gInfo == NULL) {
<a name="l01637"></a>01637     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantLabelZ::setParameters() - GenotypeInfo is NULL&quot;</span>);
<a name="l01638"></a>01638   }
<a name="l01639"></a>01639   <span class="keywordflow">if</span> (!gInfo-&gt;m_SpecialSnps-&gt;empty())
<a name="l01640"></a>01640     <a class="code" href="classQuantLabelZ.html#ac14724ad579f92836d1a6c779b9b396b" title="which snps are special">setSpecialSnps</a>(*(gInfo-&gt;m_SpecialSnps));
<a name="l01641"></a>01641   <span class="keywordflow">if</span> (!gInfo-&gt;m_HaploidSnps-&gt;empty())
<a name="l01642"></a>01642     <a class="code" href="classQuantLabelZ.html#a5afceb726f32320451aeb0c5a334a92c" title="which snps are haploid snps">setHaploidSnps</a>(*(gInfo-&gt;m_HaploidSnps));
<a name="l01643"></a>01643   <a class="code" href="classQuantLabelZ.html#a75d2851d0da897d336e42010653fbb4e" title="sets the genders for all samples">setGenders</a>(gInfo-&gt;m_Genders-&gt;<a class="code" href="classGenderCalls.html#abb3474c5f29e2959ce730f539b73b006" title="Get the genders for the cel files that have been seen.">getGenders</a>());
<a name="l01644"></a>01644   <span class="comment">// add sample covariate inbreeding penalty here</span>
<a name="l01645"></a>01645   <span class="comment">// this is highly indirect at this point, but perhaps there&#39;s a general model for sample covariates</span>
<a name="l01646"></a>01646   setInbredHetPenalty(gInfo-&gt;m_Inbred-&gt;<a class="code" href="classInbredStatus.html#acbcac0e963a7c38b4da2fcf706ca9493" title="Get the inbreeding status for the cel files that have been seen.">getInbredStatus</a>());
<a name="l01647"></a>01647   <span class="keywordflow">if</span> (o-&gt;<a class="code" href="classOptions.html#a400b67224caf01af079f1709e0e868f4" title="Check to see if an option is defined.">isOptDefined</a>(<span class="stringliteral">&quot;output-probabilities&quot;</span>)) {
<a name="l01648"></a>01648     m_OutputProbabilities = (o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;output-probabilities&quot;</span>)!=<span class="stringliteral">&quot;&quot;</span>);
<a name="l01649"></a>01649   }
<a name="l01650"></a>01650   <span class="keywordflow">try</span> {
<a name="l01651"></a>01651     board.get(ArtifactReduction::getProbesetTrustFileKey(), &amp;m_ProbeSetTrustTmpFileName);
<a name="l01652"></a>01652   }
<a name="l01653"></a>01653   <span class="keywordflow">catch</span>(...) {
<a name="l01654"></a>01654     <span class="comment">// HACK: board.get takes abortOnError as a parameter but that parameter is ignored. </span>
<a name="l01655"></a>01655     <span class="comment">// just ignore it.</span>
<a name="l01656"></a>01656   }
<a name="l01657"></a>01657   
<a name="l01658"></a>01658 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:53 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
