<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/QuantBRLMM.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>QuantBRLMM.cpp</h1>  </div>
</div>
<div class="contents">
<a href="QuantBRLMM_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   QuantBRLMM.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Chuck Sugnet</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Fri Feb 24 11:02:09 2006</span>
<a name="l00026"></a>00026 <span class="comment"> *</span>
<a name="l00027"></a>00027 <span class="comment"> * Implements a SNP genotype calling method motivated by Nusrat Rabbee&#39;s</span>
<a name="l00028"></a>00028 <span class="comment"> * RLMM (Robust Linear Model with Mahalanobis Distance Classifier) that utilizes</span>
<a name="l00029"></a>00029 <span class="comment"> * a Baysian prior about the centers of SNP clusters. Idea is to produce a</span>
<a name="l00030"></a>00030 <span class="comment"> * summary value for all probes that measure same genotype (default using RMA to</span>
<a name="l00031"></a>00031 <span class="comment"> * summarize) and then classify the genotype based on the relative intensity of</span>
<a name="l00032"></a>00032 <span class="comment"> * those estimates. For example in genotype AA the A probesets should be</span>
<a name="l00033"></a>00033 <span class="comment"> * relatively bright and B probesets should be relatively dim. For genotype AB</span>
<a name="l00034"></a>00034 <span class="comment"> * both probesets should be lighting up. Exactly how much intensity is expected</span>
<a name="l00035"></a>00035 <span class="comment"> * is learned from a random set of SNPs in the data using high confidence DM</span>
<a name="l00036"></a>00036 <span class="comment"> * calls as reference.</span>
<a name="l00037"></a>00037 <span class="comment"> *</span>
<a name="l00038"></a>00038 <span class="comment"> * For further information about RLMM see:</span>
<a name="l00039"></a>00039 <span class="comment"> * http://www.stat.berkeley.edu/users/nrabbee/RLMM/</span>
<a name="l00040"></a>00040 <span class="comment"> *</span>
<a name="l00041"></a>00041 <span class="comment"> */</span>
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 <span class="comment">//</span>
<a name="l00044"></a>00044 <span class="preprocessor">#include &quot;<a class="code" href="QuantBRLMM_8h.html">chipstream/QuantBRLMM.h</a>&quot;</span>
<a name="l00045"></a>00045 <span class="comment">//</span>
<a name="l00046"></a>00046 <span class="preprocessor">#include &quot;<a class="code" href="BioTypes_8h.html" title="Central location enumerations and other types that aren&amp;#39;t particular to a single class or applica...">chipstream/BioTypes.h</a>&quot;</span>
<a name="l00047"></a>00047 <span class="preprocessor">#include &quot;chipstream/MatrixUtil.h&quot;</span>
<a name="l00048"></a>00048 <span class="preprocessor">#include &quot;chipstream/ProbeListFactory.h&quot;</span>
<a name="l00049"></a>00049 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodFactory_8h.html" title="Factory class for making chip streams based on a string representation.">chipstream/QuantMethodFactory.h</a>&quot;</span>
<a name="l00050"></a>00050 <span class="comment">//</span>
<a name="l00051"></a>00051 <span class="preprocessor">#include &quot;<a class="code" href="TsvFile_8h.html" title="Headers for the TsvFile classes.">file/TsvFile/TsvFile.h</a>&quot;</span>
<a name="l00052"></a>00052 <span class="preprocessor">#include &quot;<a class="code" href="Convert_8h.html" title="Utility functions for converting between types.">util/Convert.h</a>&quot;</span>
<a name="l00053"></a>00053 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00054"></a>00054 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00055"></a>00055 <span class="preprocessor">#include &quot;<a class="code" href="Verbose_8h.html" title="Class for doing logging and some command line ui.">util/Verbose.h</a>&quot;</span>
<a name="l00056"></a>00056 <span class="comment">//</span>
<a name="l00057"></a>00057 <span class="preprocessor">#include &lt;sstream&gt;</span>
<a name="l00058"></a>00058 
<a name="l00059"></a>00059 <span class="keyword">using namespace </span>affx;
<a name="l00060"></a>00060 
<a name="l00061"></a>00061 QuantBRLMM::BrlmmParam::BrlmmParam() {
<a name="l00062"></a>00062   m_HetMultiplier = 1;
<a name="l00063"></a>00063   m_Iterations = 0;
<a name="l00064"></a>00064   m_IterationThreshold = 0.6;
<a name="l00065"></a>00065   m_K = 1;
<a name="l00066"></a>00066   m_PriorPseudoCount = 20;
<a name="l00067"></a>00067   m_Transform = MvA;
<a name="l00068"></a>00068   m_MinCluster = 2;
<a name="l00069"></a>00069   m_LowPrecision = <span class="keyword">false</span>;
<a name="l00070"></a>00070 }
<a name="l00071"></a>00071 <span class="comment"></span>
<a name="l00072"></a>00072 <span class="comment">/** Constructor, currently creates prior estimates from pre-calculated data from R. */</span>
<a name="l00073"></a><a class="code" href="classQuantBRLMM.html#a703d4e7e27b3cb69e48518618b7c650a">00073</a> <a class="code" href="classQuantBRLMM.html#a703d4e7e27b3cb69e48518618b7c650a" title="Constructor, currently creates prior estimates from pre-calculated data from R.">QuantBRLMM::QuantBRLMM</a>(<span class="keywordtype">double</span> hetMultiplier, <span class="keywordtype">int</span> iterations, <span class="keywordtype">double</span> iterationThreshold,
<a name="l00074"></a>00074                        <span class="keywordtype">double</span> K, <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> transform,
<a name="l00075"></a>00075                        <span class="keywordtype">int</span> priorWeight, <span class="keywordtype">int</span> priorMinCall, <span class="keywordtype">bool</span> lowPrecision) {
<a name="l00076"></a>00076   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a474238d1f98a341c21dc9c426e79036d" title="Number to balance het calls with to balance performance on het/hom calls.">m_HetMultiplier</a> = hetMultiplier;
<a name="l00077"></a>00077   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#adf7bcd06c19218d14fe7d49b79b3c7b2" title="How many iterations to do of BRLMM, i.e. boostrapping from initial calls.">m_Iterations</a> = iterations;
<a name="l00078"></a>00078   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a42c05834641c32011c0f5b6eac7ffe97" title="Scaling Parameter to use in transformations.">m_K</a> = K;
<a name="l00079"></a>00079   <span class="keywordflow">if</span> (iterationThreshold &lt; 0 || iterationThreshold &gt; 1) {
<a name="l00080"></a>00080     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM() - iter-thresh must be &gt;= 0 and &lt;= 1&quot;</span>);
<a name="l00081"></a>00081   }
<a name="l00082"></a>00082   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#ad9919e718c6d119d94f2b824abce8d53" title="What is our threshold (maximum value) for confidences befor /// calling &amp;#39;NN&amp;#39; in iterative mod...">m_IterationThreshold</a> = iterationThreshold;
<a name="l00083"></a>00083   <span class="keywordflow">if</span> (transform &gt; <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2dea69397eccbeb17925a9d8361f9fb8" title="Contrast Centers Stretch (previously known as alternative south san francisco).">CCS</a> || transform &lt; <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aaf0dc6d5b83fda75ca175ce28504710c" title="Minus Vs Average.">MvA</a>) {
<a name="l00084"></a>00084     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM() - Expecting transform to be between: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aaf0dc6d5b83fda75ca175ce28504710c" title="Minus Vs Average.">MvA</a>) + <span class="stringliteral">&quot; (MvA) and: &quot;</span> +
<a name="l00085"></a>00085                   <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2dea69397eccbeb17925a9d8361f9fb8" title="Contrast Centers Stretch (previously known as alternative south san francisco).">CCS</a>) + <span class="stringliteral">&quot; (CCS), but got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(transform));
<a name="l00086"></a>00086   }
<a name="l00087"></a>00087   <span class="keywordflow">if</span> (priorWeight &lt; 0) {
<a name="l00088"></a>00088     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM() - Expecting a non-negative prior weight.&quot;</span>);
<a name="l00089"></a>00089   }
<a name="l00090"></a>00090   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a3d250342e8ec26cdfe1f162274d7bfab" title="How much weight should the prior get?">m_PriorPseudoCount</a> = priorWeight;
<a name="l00091"></a>00091   <span class="keywordflow">if</span> (priorMinCall &lt; 2) {
<a name="l00092"></a>00092     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM() - Must have at least 2 examples per genotype for &quot;</span>
<a name="l00093"></a>00093                   <span class="stringliteral">&quot;use in prior generation. Current priorMinCall is: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(priorMinCall));
<a name="l00094"></a>00094   }
<a name="l00095"></a>00095   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa94a4cd0558a115b819dadc3501bff80" title="Minimum examples for each genotype to be included in prior estimation.">m_MinCluster</a> = priorMinCall;
<a name="l00096"></a>00096   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa9ebe7bbc8cb7de369bf2df9e795690c" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression testss).">m_LowPrecision</a> = lowPrecision;
<a name="l00097"></a>00097   m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a928f99fd333fed0c34013ce8d23acb49" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a> = transform;
<a name="l00098"></a>00098   m_QuantMethod = NULL;
<a name="l00099"></a>00099   m_Prior = NULL;
<a name="l00100"></a>00100   m_CurrentPrior = NULL;
<a name="l00101"></a>00101   m_BRLMMMaxScore = -1;
<a name="l00102"></a>00102 
<a name="l00103"></a>00103   <a class="code" href="classQuantBRLMM.html#a53c159ab51bc8a716b63001410146e10" title="Fill in our self documentation from current state.">setupSelfDoc</a>(*<span class="keyword">this</span>);
<a name="l00104"></a>00104 
<a name="l00105"></a>00105   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;het-mult&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a474238d1f98a341c21dc9c426e79036d" title="Number to balance het calls with to balance performance on het/hom calls.">m_HetMultiplier</a>));
<a name="l00106"></a>00106   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;iterations&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#adf7bcd06c19218d14fe7d49b79b3c7b2" title="How many iterations to do of BRLMM, i.e. boostrapping from initial calls.">m_Iterations</a>));
<a name="l00107"></a>00107   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;iter-thresh&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#ad9919e718c6d119d94f2b824abce8d53" title="What is our threshold (maximum value) for confidences befor /// calling &amp;#39;NN&amp;#39; in iterative mod...">m_IterationThreshold</a>));
<a name="l00108"></a>00108   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;K&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a42c05834641c32011c0f5b6eac7ffe97" title="Scaling Parameter to use in transformations.">m_K</a>));
<a name="l00109"></a>00109   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;transform&quot;</span>, <a class="code" href="classQuantBRLMM.html#a03d17022504aa06d976b555fbd6e8ea7" title="Convert a transformation enum to string representation.">stringForTransformation</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a928f99fd333fed0c34013ce8d23acb49" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>));
<a name="l00110"></a>00110   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;prior-weight&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a3d250342e8ec26cdfe1f162274d7bfab" title="How much weight should the prior get?">m_PriorPseudoCount</a>));
<a name="l00111"></a>00111   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;prior-mincall&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa94a4cd0558a115b819dadc3501bff80" title="Minimum examples for each genotype to be included in prior estimation.">m_MinCluster</a>));
<a name="l00112"></a>00112   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;lowprecision&quot;</span>, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa9ebe7bbc8cb7de369bf2df9e795690c" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression testss).">m_LowPrecision</a>);
<a name="l00113"></a>00113 }
<a name="l00114"></a>00114 <span class="comment"></span>
<a name="l00115"></a>00115 <span class="comment">/** Fill in our self documentation from current state. */</span>
<a name="l00116"></a><a class="code" href="classQuantBRLMM.html#a53c159ab51bc8a716b63001410146e10">00116</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a53c159ab51bc8a716b63001410146e10" title="Fill in our self documentation from current state.">QuantBRLMM::setupSelfDoc</a>(<a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> &amp;doc) {
<a name="l00117"></a>00117   doc.<a class="code" href="classSelfDoc.html#a51db8ea9abfa958f92fb5bc601ad68dc" title="Setter method for name appearing in doc.">setDocName</a>(<a class="code" href="QuantBRLMM_8h.html#aa9aa554ff92b91869db108754ee24023" title="String describing quant method.">QUANTBRLMM</a>);
<a name="l00118"></a>00118   doc.<a class="code" href="classSelfDoc.html#acd4638b447701d1a642effb6b8d2a396" title="Setter method for description appearing in doc.">setDocDescription</a>(<span class="stringliteral">&quot;Do genotyping calls using the BRLMM (Bayesian RLMM) algorithm.&quot;</span>);
<a name="l00119"></a>00119   doc.<a class="code" href="classSelfDoc.html#aa43e6721e35792413711e07bda90861d" title="Setter method for parameters and their documentation.">setDocOptions</a>(<a class="code" href="classQuantBRLMM.html#af4303a14cab2be76711893a4082e8c19" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>());
<a name="l00120"></a>00120 }
<a name="l00121"></a>00121 <span class="comment"></span>
<a name="l00122"></a>00122 <span class="comment">/** Destructor. */</span>
<a name="l00123"></a><a class="code" href="classQuantBRLMM.html#ad9a33be6eb00a6823576eac29302f5d1">00123</a> <a class="code" href="classQuantBRLMM.html#ad9a33be6eb00a6823576eac29302f5d1" title="Destructor.">QuantBRLMM::~QuantBRLMM</a>() {
<a name="l00124"></a>00124   <span class="keywordflow">if</span> (<a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.is_open()) {
<a name="l00125"></a>00125     <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.close();
<a name="l00126"></a>00126   }
<a name="l00127"></a>00127   <span class="keywordflow">if</span> (<a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.is_open()) {
<a name="l00128"></a>00128     <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.close();
<a name="l00129"></a>00129   }
<a name="l00130"></a>00130   <span class="keywordflow">for</span> (std::vector&lt;QuantMethodReport *&gt;::iterator it = m_Reporters.begin(); it != m_Reporters.end(); it++)
<a name="l00131"></a>00131     {
<a name="l00132"></a>00132       <span class="keyword">delete</span> *it;
<a name="l00133"></a>00133     }
<a name="l00134"></a>00134   m_Reporters.clear();
<a name="l00135"></a>00135   <a class="code" href="Util_8h.html#ad0787eda9709999bbd42dc11b21deb0d" title="delete function that deletes the pointer and sets it to NULL.">Freez</a>(m_QuantMethod);
<a name="l00136"></a>00136 }
<a name="l00137"></a>00137 <span class="comment"></span>
<a name="l00138"></a>00138 <span class="comment">/**</span>
<a name="l00139"></a>00139 <span class="comment"> * Load up the snp models from a text file.</span>
<a name="l00140"></a>00140 <span class="comment"> * @param fileName - text file to read models from.</span>
<a name="l00141"></a>00141 <span class="comment"> */</span>
<a name="l00142"></a><a class="code" href="classQuantBRLMM.html#a873e8c5e7d97ea648f749d1c410706f0">00142</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a873e8c5e7d97ea648f749d1c410706f0" title="Load up the snp models from a text file.">QuantBRLMM::readSnpModels</a>(<span class="keyword">const</span> std::string&amp; fileName) {
<a name="l00143"></a>00143   <a class="code" href="classQuantBRLMM.html#a5f6db5e8a6818003a7e4629e555366a3" title="Load in a number of snp specific models.">loadSnpModels</a>(m_SnpModelMap, fileName);
<a name="l00144"></a>00144 }
<a name="l00145"></a>00145 
<a name="l00146"></a>00146 <span class="keywordtype">void</span> QuantBRLMM::transformData(std::vector&lt;double&gt; &amp;aVals,
<a name="l00147"></a>00147                                std::vector&lt;double&gt; &amp;bVals,
<a name="l00148"></a>00148                                <a class="code" href="classQuantBRLMM_1_1BrlmmParam.html">QuantBRLMM::BrlmmParam</a> &amp;param) {
<a name="l00149"></a>00149   assert(aVals.size() == bVals.size());
<a name="l00150"></a>00150   <span class="keywordtype">double</span> x = 0, y = 0;
<a name="l00151"></a>00151   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; aVals.size(); i++) {
<a name="l00152"></a>00152     valueForTransformation(aVals[i], bVals[i], param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a928f99fd333fed0c34013ce8d23acb49" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>, param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a42c05834641c32011c0f5b6eac7ffe97" title="Scaling Parameter to use in transformations.">m_K</a>, x, y);
<a name="l00153"></a>00153     aVals[i] = x;
<a name="l00154"></a>00154     bVals[i] = y;
<a name="l00155"></a>00155   }
<a name="l00156"></a>00156 }
<a name="l00157"></a>00157 <span class="comment"></span>
<a name="l00158"></a>00158 <span class="comment">/**</span>
<a name="l00159"></a>00159 <span class="comment"> * Specify the prior that will be used for predictions.</span>
<a name="l00160"></a>00160 <span class="comment"> *</span>
<a name="l00161"></a>00161 <span class="comment"> * @param prior - Prior data to use.</span>
<a name="l00162"></a>00162 <span class="comment"> */</span>
<a name="l00163"></a><a class="code" href="classQuantBRLMM.html#ad6cb1cdd30699b802c7d81248da58da0">00163</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#ad6cb1cdd30699b802c7d81248da58da0" title="Specify the prior that will be used for predictions.">QuantBRLMM::setPrior</a>(<span class="keyword">const</span> <a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior) {
<a name="l00164"></a>00164   <span class="keyword">delete</span> m_Prior;
<a name="l00165"></a>00165   m_Prior = <span class="keyword">new</span> <a class="code" href="classClusterPrior.html">ClusterPrior</a>();
<a name="l00166"></a>00166   m_Prior-&gt;id = prior.id;
<a name="l00167"></a>00167   m_Prior-&gt;covars = prior.covars;
<a name="l00168"></a>00168   m_Prior-&gt;centers = prior.centers;
<a name="l00169"></a>00169   m_Prior-&gt;centerVars = prior.centerVars;
<a name="l00170"></a>00170 }
<a name="l00171"></a>00171 <span class="comment"></span>
<a name="l00172"></a>00172 <span class="comment">/**</span>
<a name="l00173"></a>00173 <span class="comment"> * @brief What is the name of the quantification method?</span>
<a name="l00174"></a>00174 <span class="comment"> * @return name of adjuster.</span>
<a name="l00175"></a>00175 <span class="comment"> */</span>
<a name="l00176"></a><a class="code" href="classQuantBRLMM.html#ac61886478a0d57b3c464015fa0ebc196">00176</a> std::string <a class="code" href="classQuantBRLMM.html#ac61886478a0d57b3c464015fa0ebc196" title="What is the name of the quantification method?">QuantBRLMM::getType</a>() {
<a name="l00177"></a>00177   <span class="keywordflow">return</span> std::string(<a class="code" href="QuantBRLMM_8h.html#aa9aa554ff92b91869db108754ee24023" title="String describing quant method.">QUANTBRLMM</a>);
<a name="l00178"></a>00178 }
<a name="l00179"></a>00179 
<a name="l00180"></a>00180 <span class="comment">/* Set the quantification method to use for summarizing individual alleles */</span>
<a name="l00181"></a>00181 <span class="keywordtype">void</span> QuantBRLMM::setQuantExprMethod(<a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *qMethod) {
<a name="l00182"></a>00182   <a class="code" href="Util_8h.html#ad0787eda9709999bbd42dc11b21deb0d" title="delete function that deletes the pointer and sets it to NULL.">Freez</a>(m_QuantMethod);
<a name="l00183"></a>00183   m_QuantMethod = qMethod;
<a name="l00184"></a>00184 }
<a name="l00185"></a>00185 <span class="comment">/* Get the quantification method to use for summarizing individual alleles */</span>
<a name="l00186"></a>00186 <a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *QuantBRLMM::getQuantExprMethod() {
<a name="l00187"></a>00187   <span class="keywordflow">return</span> m_QuantMethod;
<a name="l00188"></a>00188 }
<a name="l00189"></a>00189 <span class="comment"></span>
<a name="l00190"></a>00190 <span class="comment">/**</span>
<a name="l00191"></a>00191 <span class="comment"> * What version of the BRLMM algorithm are we implementing?</span>
<a name="l00192"></a>00192 <span class="comment"> * @return version string.</span>
<a name="l00193"></a>00193 <span class="comment"> */</span>
<a name="l00194"></a><a class="code" href="classQuantBRLMM.html#a2110ed6b324c0399cc2467ea5da113a4">00194</a> std::string <a class="code" href="classQuantBRLMM.html#a2110ed6b324c0399cc2467ea5da113a4" title="What version of the BRLMM algorithm are we implementing?">QuantBRLMM::getVersion</a>() {
<a name="l00195"></a>00195   <span class="keywordflow">return</span> QUANTBRLMM_VERSION;
<a name="l00196"></a>00196 }
<a name="l00197"></a>00197 <span class="comment"></span>
<a name="l00198"></a>00198 <span class="comment">/** </span>
<a name="l00199"></a>00199 <span class="comment"> * Set the initial guesses of genotypes, usually from DM predictions.</span>
<a name="l00200"></a>00200 <span class="comment"> * </span>
<a name="l00201"></a>00201 <span class="comment"> * @param genotypes vector of genotypes AA=0, AB=1, AB=2, NN=3,</span>
<a name="l00202"></a>00202 <span class="comment"> */</span>
<a name="l00203"></a><a class="code" href="classQuantBRLMM.html#a72de5e66b4d6f1d6fc63426846d20950">00203</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a72de5e66b4d6f1d6fc63426846d20950" title="Set the initial guesses of genotypes, usually from DM predictions.">QuantBRLMM::setKnownGenoTypes</a>(std::map&lt;<span class="keyword">const</span> <span class="keywordtype">char</span> *,std::vector&lt;affx::GType&gt;, <a class="code" href="structUtil_1_1ltstr.html" title="Comparison object for use in map for char *.">Util::ltstr</a>&gt; &amp;genotypes) {
<a name="l00204"></a>00204   m_KnownGenoTypes = genotypes;
<a name="l00205"></a>00205 }
<a name="l00206"></a>00206 <span class="comment"></span>
<a name="l00207"></a>00207 <span class="comment">/**</span>
<a name="l00208"></a>00208 <span class="comment"> * Clear out all of our results and values.</span>
<a name="l00209"></a>00209 <span class="comment"> */</span>
<a name="l00210"></a><a class="code" href="classQuantBRLMM.html#a680eb2ad5f77c33e5f91a11d01b460cf">00210</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a680eb2ad5f77c33e5f91a11d01b460cf" title="Clear out all of our results and values.">QuantBRLMM::blankSelf</a>() {
<a name="l00211"></a>00211   <a class="code" href="classQuantGTypeMethod.html#aff29fdf8763f00f5e1eb451162cb38c2" title="clear a probe set">clearProbeSet</a>(<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>);
<a name="l00212"></a>00212   <a class="code" href="classQuantGTypeMethod.html#aff29fdf8763f00f5e1eb451162cb38c2" title="clear a probe set">clearProbeSet</a>(<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00213"></a>00213   <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.clear();
<a name="l00214"></a>00214   <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.clear();
<a name="l00215"></a>00215   m_InitialCalls.clear();
<a name="l00216"></a>00216   m_Calls.clear();
<a name="l00217"></a>00217   m_Confidences.clear();
<a name="l00218"></a>00218   m_Distances.clear();
<a name="l00219"></a>00219 };
<a name="l00220"></a>00220 <span class="comment"></span>
<a name="l00221"></a>00221 <span class="comment">/**</span>
<a name="l00222"></a>00222 <span class="comment"> * How many genotyping calls do we have? Also indicates how many</span>
<a name="l00223"></a>00223 <span class="comment"> * samples there are.</span>
<a name="l00224"></a>00224 <span class="comment"> *</span>
<a name="l00225"></a>00225 <span class="comment"> * @return Number of genotyping calls made.</span>
<a name="l00226"></a>00226 <span class="comment"> */</span>
<a name="l00227"></a><a class="code" href="classQuantBRLMM.html#ada9804ca9a3a11558d594813ad603e4b">00227</a> <span class="keywordtype">size_t</span> <a class="code" href="classQuantBRLMM.html#ada9804ca9a3a11558d594813ad603e4b" title="How many genotyping calls do we have? Also indicates how many samples there are.">QuantBRLMM::getNumCalls</a>() { <span class="keywordflow">return</span> m_Calls.size(); }
<a name="l00228"></a>00228 <span class="comment"></span>
<a name="l00229"></a>00229 <span class="comment">/**</span>
<a name="l00230"></a>00230 <span class="comment"> * Get our confidence value for a particular call in a particular sample.</span>
<a name="l00231"></a>00231 <span class="comment"> * @param index - sample of interest.</span>
<a name="l00232"></a>00232 <span class="comment"> * @return - Confidence in the predicted call.</span>
<a name="l00233"></a>00233 <span class="comment"> */</span>
<a name="l00234"></a><a class="code" href="classQuantBRLMM.html#a1b285d4ac1220d6a1e6b7dce90428b62">00234</a> <span class="keywordtype">double</span> <a class="code" href="classQuantBRLMM.html#a1b285d4ac1220d6a1e6b7dce90428b62" title="Get our confidence value for a particular call in a particular sample.">QuantBRLMM::getConfidence</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index) {
<a name="l00235"></a>00235   <span class="keywordflow">if</span> (index &gt;= m_Confidences.size()) {
<a name="l00236"></a>00236     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Asking for call at index &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(index) + <span class="stringliteral">&quot; when Probeset &quot;</span> +
<a name="l00237"></a>00237                   <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot; has only &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Confidences.size()) + <span class="stringliteral">&quot; calls.&quot;</span>);
<a name="l00238"></a>00238   }
<a name="l00239"></a>00239   <span class="keywordflow">return</span> m_Confidences[index];
<a name="l00240"></a>00240 }
<a name="l00241"></a>00241 <span class="comment"></span>
<a name="l00242"></a>00242 <span class="comment">/**</span>
<a name="l00243"></a>00243 <span class="comment"> * Get our confidence value for a particular call in a particular sample.</span>
<a name="l00244"></a>00244 <span class="comment"> * @param genoType -  Genotype of interest.</span>
<a name="l00245"></a>00245 <span class="comment"> * @param index - sample of interest.</span>
<a name="l00246"></a>00246 <span class="comment"> * @return - Confidence in the predicted call.</span>
<a name="l00247"></a>00247 <span class="comment"> */</span>
<a name="l00248"></a><a class="code" href="classQuantBRLMM.html#a8e5df36010d56b0782f0df8b4d6a1603">00248</a> <span class="keywordtype">double</span> <a class="code" href="classQuantBRLMM.html#a8e5df36010d56b0782f0df8b4d6a1603" title="Get our confidence value for a particular call in a particular sample.">QuantBRLMM::getConfGenotype</a>(affx::GType  genoType, <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index) {
<a name="l00249"></a>00249   assert(index &lt; m_Distances.size());
<a name="l00250"></a>00250   <span class="keywordflow">if</span> (genoType == <a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a1cef2fc4697bd5fc67116fbd1ee2027b" title="No call, unknown genotype. The internal is value is &amp;#39;3&amp;#39;; external is &amp;#39;-1&amp;#39;,.">affx::NN</a>)
<a name="l00251"></a>00251     <span class="keywordflow">return</span> FLT_MAX;
<a name="l00252"></a>00252   <span class="keywordflow">return</span> m_Distances[index][genoType];
<a name="l00253"></a>00253 }
<a name="l00254"></a>00254 <span class="comment"></span>
<a name="l00255"></a>00255 <span class="comment">/**</span>
<a name="l00256"></a>00256 <span class="comment"> * Get the summary value after transformation for the A and B alleles respectively.</span>
<a name="l00257"></a>00257 <span class="comment"> *</span>
<a name="l00258"></a>00258 <span class="comment"> * @param index - Which chip to get values for.</span>
<a name="l00259"></a>00259 <span class="comment"> * @param aValue - Filled in with transformed a value.</span>
<a name="l00260"></a>00260 <span class="comment"> * @param bValue - Filled in with transformed b value.</span>
<a name="l00261"></a>00261 <span class="comment"> */</span>
<a name="l00262"></a><a class="code" href="classQuantBRLMM.html#ac0cd5a76fb8f17ab31f19ca35b3174f3">00262</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#ac0cd5a76fb8f17ab31f19ca35b3174f3" title="Get the summary value after transformation for the A and B alleles respectively.">QuantBRLMM::getAlleleValues</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index, <span class="keywordtype">double</span> &amp;aValue, <span class="keywordtype">double</span> &amp;bValue) {
<a name="l00263"></a>00263   assert(index &lt; <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size());
<a name="l00264"></a>00264   assert(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size() == <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.size());
<a name="l00265"></a>00265   aValue = <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>[index];
<a name="l00266"></a>00266   bValue = <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>[index];
<a name="l00267"></a>00267 }
<a name="l00268"></a>00268 <span class="comment"></span>
<a name="l00269"></a>00269 <span class="comment">/**</span>
<a name="l00270"></a>00270 <span class="comment"> * Get the summary value names for the A and B alleles respectively.</span>
<a name="l00271"></a>00271 <span class="comment"> *</span>
<a name="l00272"></a>00272 <span class="comment"> * @param aName - Filled in with a value name.</span>
<a name="l00273"></a>00273 <span class="comment"> * @param bName - Filled in with b value name.</span>
<a name="l00274"></a>00274 <span class="comment"> */</span>
<a name="l00275"></a><a class="code" href="classQuantBRLMM.html#a91dab0e7e2526025e0ade4628dd2c88b">00275</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a91dab0e7e2526025e0ade4628dd2c88b" title="Get the summary value names for the A and B alleles respectively.">QuantBRLMM::getAlleleValueNames</a>(std::string &amp;aName, std::string &amp;bName)
<a name="l00276"></a>00276 {
<a name="l00277"></a>00277   aName = <span class="stringliteral">&quot;Contrast&quot;</span>;
<a name="l00278"></a>00278   bName = <span class="stringliteral">&quot;Strength&quot;</span>;
<a name="l00279"></a>00279 }
<a name="l00280"></a>00280 <span class="comment"></span>
<a name="l00281"></a>00281 <span class="comment">/**</span>
<a name="l00282"></a>00282 <span class="comment"> * Get the name of the probeset that these calls are being made for.</span>
<a name="l00283"></a>00283 <span class="comment"> * @return name of probeset.</span>
<a name="l00284"></a>00284 <span class="comment"> */</span>
<a name="l00285"></a><a class="code" href="classQuantBRLMM.html#a35841cae2003df29b6968efb52e3a3a4">00285</a> <span class="keyword">const</span> std::string &amp;<a class="code" href="classQuantBRLMM.html#a35841cae2003df29b6968efb52e3a3a4" title="Get the name of the probeset that these calls are being made for.">QuantBRLMM::getProbeSetName</a>() { <span class="keywordflow">return</span> <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>; }
<a name="l00286"></a>00286 <span class="comment"></span>
<a name="l00287"></a>00287 <span class="comment">/**</span>
<a name="l00288"></a>00288 <span class="comment"> * Add an expression reporter to output the summary values for each</span>
<a name="l00289"></a>00289 <span class="comment"> * allele, residuals, etc.</span>
<a name="l00290"></a>00290 <span class="comment"> *</span>
<a name="l00291"></a>00291 <span class="comment"> * @param reporter - the object that will output all summary values.</span>
<a name="l00292"></a>00292 <span class="comment"> */</span>
<a name="l00293"></a><a class="code" href="classQuantBRLMM.html#a7ed6bc24b8df6012fb41b8be487acdc7">00293</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a7ed6bc24b8df6012fb41b8be487acdc7" title="Add an expression reporter to output the summary values for each allele, residuals, etc.">QuantBRLMM::addExprReporter</a>(<a class="code" href="classQuantMethodReport.html" title="Class for reporting results of quantification methods.">QuantMethodReport</a> *reporter) {
<a name="l00294"></a>00294   m_Reporters.push_back(reporter);
<a name="l00295"></a>00295 }
<a name="l00296"></a>00296 <span class="comment"></span>
<a name="l00297"></a>00297 <span class="comment">/**</span>
<a name="l00298"></a>00298 <span class="comment"> * Get the parameters used for this brlmm run.</span>
<a name="l00299"></a>00299 <span class="comment"> */</span>
<a name="l00300"></a><a class="code" href="classQuantBRLMM.html#a6a3ae7a02c823637903a2d82d158b5b7">00300</a> <a class="code" href="classQuantBRLMM_1_1BrlmmParam.html">QuantBRLMM::BrlmmParam</a> <a class="code" href="classQuantBRLMM.html#a6a3ae7a02c823637903a2d82d158b5b7" title="Get the parameters used for this brlmm run.">QuantBRLMM::getParam</a>() {
<a name="l00301"></a>00301   <span class="keywordflow">return</span> m_Param;
<a name="l00302"></a>00302 }
<a name="l00303"></a>00303 
<a name="l00304"></a>00304 <span class="keywordtype">void</span> QuantBRLMM::setHaploidSnps(std::map&lt;std::string,bool&gt; &amp;haploidSnps) {
<a name="l00305"></a>00305   m_HaploidSnps = haploidSnps;
<a name="l00306"></a>00306 }
<a name="l00307"></a>00307 
<a name="l00308"></a>00308 <span class="keywordtype">void</span> QuantBRLMM::setSnpSpecificPriors(std::map&lt;std::string,ClusterPrior&gt; &amp;snpPriorMap) {
<a name="l00309"></a>00309   m_SnpPriorMap = snpPriorMap;
<a name="l00310"></a>00310 }
<a name="l00311"></a>00311 
<a name="l00312"></a>00312 <span class="keywordtype">void</span> QuantBRLMM::setGenders(std::vector&lt;affx::Gender&gt; &amp;genders) {
<a name="l00313"></a>00313   m_Genders = genders;
<a name="l00314"></a>00314 }
<a name="l00315"></a>00315 
<a name="l00316"></a>00316 <span class="keywordtype">double</span> QuantBRLMM::getMaxScore() { <span class="keywordflow">return</span>(m_BRLMMMaxScore); };
<a name="l00317"></a>00317 <span class="keywordtype">void</span> QuantBRLMM::setMaxScore(<span class="keywordtype">float</span> v){
<a name="l00318"></a>00318   m_BRLMMMaxScore=v;
<a name="l00319"></a>00319 
<a name="l00320"></a>00320   <span class="comment">// keep our options in sync so that the state is</span>
<a name="l00321"></a>00321   <span class="comment">// correctly reported</span>
<a name="l00322"></a>00322   <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;MS&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_BRLMMMaxScore));
<a name="l00323"></a>00323 };
<a name="l00324"></a>00324 
<a name="l00325"></a>00325 
<a name="l00326"></a><a class="code" href="classQuantBRLMM.html#a370cff001d055600645091956a385a5b">00326</a> <span class="keywordtype">double</span> <a class="code" href="classQuantBRLMM.html#a370cff001d055600645091956a385a5b" title="Get the min value for a call.">QuantBRLMM::getMinThresh</a>() { <span class="keywordflow">return</span> -DBL_MAX;}
<a name="l00327"></a><a class="code" href="classQuantBRLMM.html#a736c04fa1b48ef371480394af9a478c3">00327</a> <span class="keywordtype">double</span> <a class="code" href="classQuantBRLMM.html#a736c04fa1b48ef371480394af9a478c3" title="Get the max value for a call.">QuantBRLMM::getMaxThresh</a>() { <span class="keywordflow">return</span> m_BRLMMMaxScore;}
<a name="l00328"></a>00328 
<a name="l00329"></a>00329 <span class="keywordtype">void</span> QuantBRLMM::setDmCallsOut(<span class="keyword">const</span> std::string&amp; fileName, std::vector&lt;std::string&gt; &amp;col_names) {
<a name="l00330"></a>00330   <span class="comment">//    m_DmCallsFileName = fileName;</span>
<a name="l00331"></a>00331   <span class="comment">//    if (m_DmCallsOut.is_open())</span>
<a name="l00332"></a>00332   <span class="comment">//      m_DmCallsOut.close();</span>
<a name="l00333"></a>00333   <span class="comment">//    Util::mustOpenToWrite(m_DmCallsOut, m_DmCallsFileName.c_str());</span>
<a name="l00334"></a>00334   <span class="comment">//    m_DmCallsOut &lt;&lt; &quot;#Calls: -1=NN, 0=AA, 1=AB, 2=BB&quot; &lt;&lt; endl;</span>
<a name="l00335"></a>00335   <span class="comment">//    m_DmCallsOut &lt;&lt; &quot;probeset_id&quot;;</span>
<a name="l00336"></a>00336   <span class="comment">//    for (unsigned int i = 0; i &lt; headers.size(); i++) {</span>
<a name="l00337"></a>00337   <span class="comment">//      std::string header = Util::fileRoot(headers[i]);</span>
<a name="l00338"></a>00338   <span class="comment">//      m_DmCallsOut &lt;&lt; &quot;\t&quot; &lt;&lt; header;</span>
<a name="l00339"></a>00339   <span class="comment">//    }</span>
<a name="l00340"></a>00340   <span class="comment">//    m_DmCallsOut &lt;&lt; endl;</span>
<a name="l00341"></a>00341   <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.setFilename(fileName);
<a name="l00342"></a>00342   <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.setFormat(affx::TsvReport::FMT_TSV);
<a name="l00343"></a>00343   <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.addHeaderComment(<span class="stringliteral">&quot;Calls: -1=NN, 0=AA, 1=AB, 2=BB&quot;</span>);
<a name="l00344"></a>00344   <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.defineStringColumn(0,0,<span class="stringliteral">&quot;probeset_id&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l00345"></a>00345   <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.defineColumns(<a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(col_names),affx::FILE5_DTYPE_INT,0);
<a name="l00346"></a>00346   <span class="comment">//</span>
<a name="l00347"></a>00347   <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.writeTsv_v1();
<a name="l00348"></a>00348 }
<a name="l00349"></a>00349 
<a name="l00350"></a><a class="code" href="classQuantBRLMM.html#ac6c6f4e8ffe27ec21902476b3d663fcc">00350</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantBRLMM.html#ac6c6f4e8ffe27ec21902476b3d663fcc" title="Get set up for a run of reporting probesets.">QuantBRLMM::prepare</a>(<span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart) {
<a name="l00351"></a>00351   <span class="keywordtype">bool</span> result = <span class="keyword">true</span>;
<a name="l00352"></a>00352   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;m_Reporters.size(); i++)
<a name="l00353"></a>00353     result = result &amp;&amp; m_Reporters[i]-&gt;<a class="code" href="classQuantBRLMM.html#ac6c6f4e8ffe27ec21902476b3d663fcc" title="Get set up for a run of reporting probesets.">prepare</a>(*<span class="keyword">this</span>, iMart);
<a name="l00354"></a>00354   <span class="keywordflow">return</span> result;
<a name="l00355"></a>00355 }
<a name="l00356"></a>00356 
<a name="l00357"></a><a class="code" href="classQuantBRLMM.html#a4eea52e73038237e10b4328deb8aede1">00357</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantBRLMM.html#a4eea52e73038237e10b4328deb8aede1" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">QuantBRLMM::finish</a>() {
<a name="l00358"></a>00358   <span class="keywordtype">bool</span> result = <span class="keyword">true</span>;
<a name="l00359"></a>00359   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;m_Reporters.size(); i++)
<a name="l00360"></a>00360     result = result &amp;&amp; m_Reporters[i]-&gt;<a class="code" href="classQuantBRLMM.html#a4eea52e73038237e10b4328deb8aede1" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">finish</a>(*<span class="keyword">this</span>);
<a name="l00361"></a>00361   <span class="keywordflow">return</span> result;
<a name="l00362"></a>00362 }
<a name="l00363"></a>00363 <span class="comment"></span>
<a name="l00364"></a>00364 <span class="comment">/**</span>
<a name="l00365"></a>00365 <span class="comment"> * @brief Do the heavy lifting of estimation.</span>
<a name="l00366"></a>00366 <span class="comment"> */</span>
<a name="l00367"></a><a class="code" href="classQuantBRLMM.html#a14f9cb0971989fc0ee0d37f5242bdc2d">00367</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a14f9cb0971989fc0ee0d37f5242bdc2d" title="Do the heavy lifting of estimation.">QuantBRLMM::computeEstimate</a>() {
<a name="l00368"></a>00368   transformData(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, m_Param);
<a name="l00369"></a>00369   <a class="code" href="classClusterPrior.html">ClusterPrior</a> *prior = priorForSnp();
<a name="l00370"></a>00370   <span class="keywordflow">if</span> (m_HaploidSnps.find(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>) == m_HaploidSnps.end()) {
<a name="l00371"></a>00371     <a class="code" href="classClusterModel.html">ClusterModel</a> model;
<a name="l00372"></a>00372     <span class="comment">/* Check to see if there is a precomputed model. */</span>
<a name="l00373"></a>00373     <span class="keywordflow">if</span> (!m_SnpModelMap.empty()) {
<a name="l00374"></a>00374       std::map&lt;std::string,ClusterModel&gt;::iterator iter;
<a name="l00375"></a>00375       iter = m_SnpModelMap.find(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot;-2&quot;</span>);
<a name="l00376"></a>00376       <span class="keywordflow">if</span> (iter == m_SnpModelMap.end()) {
<a name="l00377"></a>00377         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t find model for SNP: &quot;</span> + <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>);
<a name="l00378"></a>00378       }
<a name="l00379"></a>00379       model = iter-&gt;second;
<a name="l00380"></a>00380     }
<a name="l00381"></a>00381     <span class="keywordflow">else</span> {
<a name="l00382"></a>00382       model = <a class="code" href="classQuantBRLMM.html#af1acb3c6affa478feb02b2443d7d4962" title="Wrapper function for getting bayesian adjusted parameters given data.">fitModel</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, m_Distances,
<a name="l00383"></a>00383                        m_InitialCalls, *prior, <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>, m_Param);
<a name="l00384"></a>00384     }
<a name="l00385"></a>00385     <span class="comment">/* If we are saving models write this one out. */</span>
<a name="l00386"></a>00386     <span class="keywordflow">if</span> (<a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.is_open()) {
<a name="l00387"></a>00387       <a class="code" href="classQuantBRLMM.html#a008b914483156a796f217fae336ea644" title="Write out a text representation of the model to the file specified.">writeModelOut</a>(model,<a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>, <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>, 2);
<a name="l00388"></a>00388     }
<a name="l00389"></a>00389     <a class="code" href="classQuantBRLMM.html#ad0808c8bfeef3f05b4733ea08a91cd0d">brlmmCall</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, model, m_Calls, m_Confidences, m_Distances, m_Param);
<a name="l00390"></a>00390   }
<a name="l00391"></a>00391   <span class="keywordflow">else</span> {
<a name="l00392"></a>00392     <a class="code" href="classQuantBRLMM.html#a448772f9128f2987c4d8f4e3656961b2" title="This is the fix for the snps on chromosome X that don&amp;#39;t appear diploid in men (i...">callHaploidSnp</a>(*prior);
<a name="l00393"></a>00393   }
<a name="l00394"></a>00394 }
<a name="l00395"></a>00395 <span class="comment"></span>
<a name="l00396"></a>00396 <span class="comment">/**</span>
<a name="l00397"></a>00397 <span class="comment"> * Remove the het cluster from a prior by setting center to be very</span>
<a name="l00398"></a>00398 <span class="comment"> * large negative number and removing covariance with other</span>
<a name="l00399"></a>00399 <span class="comment"> * clusters.</span>
<a name="l00400"></a>00400 <span class="comment"> * @param prior - Prior to remove het cluster from.</span>
<a name="l00401"></a>00401 <span class="comment"> * @param trans - transformation used for clustering.</span>
<a name="l00402"></a>00402 <span class="comment"> */</span>
<a name="l00403"></a><a class="code" href="classQuantBRLMM.html#a761e1e30d718599d1ce3aad6853b8ffb">00403</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a761e1e30d718599d1ce3aad6853b8ffb" title="Remove the het cluster from a prior by setting center to be very large negative number and removing c...">QuantBRLMM::removeHetFromPrior</a>(<a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior, <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> trans) {
<a name="l00404"></a>00404 
<a name="l00405"></a>00405   <span class="comment">/* Adjust het center to very large negative number. */</span>
<a name="l00406"></a>00406   <span class="keywordflow">if</span> (trans == <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aaf0dc6d5b83fda75ca175ce28504710c" title="Minus Vs Average.">MvA</a>) {
<a name="l00407"></a>00407     prior.centers.element(2) = -1 * FLT_MAX; <span class="comment">// -5000 in R code.</span>
<a name="l00408"></a>00408     prior.centers.element(3) = 0;
<a name="l00409"></a>00409   }
<a name="l00410"></a>00410   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (trans == <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aea687c012cf0ba203c45713fd99a9673" title="R vs Theta.">RvT</a> || trans == <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2dea69397eccbeb17925a9d8361f9fb8" title="Contrast Centers Stretch (previously known as alternative south san francisco).">CCS</a> || trans == <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2729e1dbbb52465172e689eada41e72a" title="Contrast Extremes Stretch (previously known as south san francisco).">CES</a>) {
<a name="l00411"></a>00411     prior.centers.element(2) = 0;
<a name="l00412"></a>00412     prior.centers.element(3) = -1 * FLT_MAX; <span class="comment">// -5000 in R code</span>
<a name="l00413"></a>00413   }
<a name="l00414"></a>00414   <span class="keywordflow">else</span> {
<a name="l00415"></a>00415     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::removeHetFromPrior() - Don&#39;t recognize transform type: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(trans));
<a name="l00416"></a>00416   }
<a name="l00417"></a>00417 
<a name="l00418"></a>00418   <span class="comment">/* cancel het covariance with other centers. */</span>
<a name="l00419"></a>00419   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; prior.covars.Nrows(); i++) {
<a name="l00420"></a>00420     prior.covars.element(i,2) = 0;
<a name="l00421"></a>00421     prior.covars.element(2,i) = 0;
<a name="l00422"></a>00422     prior.covars.element(i,3) = 0;
<a name="l00423"></a>00423     prior.covars.element(3,i) = 0;
<a name="l00424"></a>00424   }
<a name="l00425"></a>00425 
<a name="l00426"></a>00426   <span class="comment">/* make hets not vary particularly much, .01 comes directly from R code. */</span>
<a name="l00427"></a>00427   prior.covars.element(2,2) = .01;
<a name="l00428"></a>00428   prior.covars.element(3,3) = .01;
<a name="l00429"></a>00429 }
<a name="l00430"></a>00430 <span class="comment"></span>
<a name="l00431"></a>00431 <span class="comment">/**</span>
<a name="l00432"></a>00432 <span class="comment"> * This is the fix for the snps on chromosome X that don&#39;t appear</span>
<a name="l00433"></a>00433 <span class="comment"> * diploid in men (i.e. aren&#39;t in the pseudoautosomal region). The</span>
<a name="l00434"></a>00434 <span class="comment"> * concept is to fit and call the model separately for men and</span>
<a name="l00435"></a>00435 <span class="comment"> * women. Implementation is to set all the male initial calls to NN</span>
<a name="l00436"></a>00436 <span class="comment"> * when fitting for women. For fitting and calling men the female</span>
<a name="l00437"></a>00437 <span class="comment"> * samples are set to NN and further the prior is ajusted such that</span>
<a name="l00438"></a>00438 <span class="comment"> * het calls are impossible to get.</span>
<a name="l00439"></a>00439 <span class="comment"> */</span>
<a name="l00440"></a><a class="code" href="classQuantBRLMM.html#a448772f9128f2987c4d8f4e3656961b2">00440</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a448772f9128f2987c4d8f4e3656961b2" title="This is the fix for the snps on chromosome X that don&amp;#39;t appear diploid in men (i...">QuantBRLMM::callHaploidSnp</a>(<a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior) {
<a name="l00441"></a>00441   vector&lt;affx::GType&gt; xxCalls, xyCalls;
<a name="l00442"></a>00442   vector&lt;double&gt; xxConfidences, xyConfidences;
<a name="l00443"></a>00443   vector&lt;vector&lt;double&gt; &gt; xxDistances, xyDistances;
<a name="l00444"></a>00444   vector&lt;affx::GType&gt; xxInitCalls(m_InitialCalls), xyInitCalls(m_InitialCalls);
<a name="l00445"></a>00445   <a class="code" href="classClusterModel.html">ClusterModel</a> model;
<a name="l00446"></a>00446   std::map&lt;std::string,ClusterModel&gt;::iterator iter;
<a name="l00447"></a>00447   <span class="comment">/* Sanity check. */</span>
<a name="l00448"></a>00448   <a class="code" href="classErr.html#ad1a633f602fe60689cdc12d9ca1842e9" title="errAbort based version of assert with an error message.">Err::check</a>(m_Genders.size() == m_InitialCalls.size(),
<a name="l00449"></a>00449              <span class="stringliteral">&quot;Error: QuantBRLMM::callHaploidSnp() - m_Gender and m_InitialCalls are different sizes.&quot;</span>);
<a name="l00450"></a>00450 
<a name="l00451"></a>00451   <span class="comment">/* Do the female fitting and calls. */</span>
<a name="l00452"></a>00452   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; m_InitialCalls.size(); i++) {
<a name="l00453"></a>00453     <span class="keywordflow">if</span> (m_Genders[i] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669afdfe578a9e9044b7b274b736ed341a81" title="Has XY chromosomes.">affx::Male</a>) {
<a name="l00454"></a>00454       xxInitCalls[i] = NN;
<a name="l00455"></a>00455     }
<a name="l00456"></a>00456   }
<a name="l00457"></a>00457   <a class="code" href="classClusterModel.html">ClusterModel</a> xxModel;
<a name="l00458"></a>00458   <span class="comment">/* Check to see if there is a precomputed model. */</span>
<a name="l00459"></a>00459   <span class="keywordflow">if</span> (!m_SnpModelMap.empty()) {
<a name="l00460"></a>00460     <span class="keywordtype">string</span> name = <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot;-2&quot;</span>;
<a name="l00461"></a>00461     iter = m_SnpModelMap.find(name);
<a name="l00462"></a>00462     <span class="keywordflow">if</span> (iter == m_SnpModelMap.end()) {
<a name="l00463"></a>00463       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t find model for SNP: &quot;</span> + <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot;-2&quot;</span>);
<a name="l00464"></a>00464     }
<a name="l00465"></a>00465     xxModel = iter-&gt;second;
<a name="l00466"></a>00466   }
<a name="l00467"></a>00467   <span class="keywordflow">else</span> {
<a name="l00468"></a>00468     xxModel = <a class="code" href="classQuantBRLMM.html#af1acb3c6affa478feb02b2443d7d4962" title="Wrapper function for getting bayesian adjusted parameters given data.">fitModel</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, xxDistances,
<a name="l00469"></a>00469                  xxInitCalls, prior, <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>, m_Param);
<a name="l00470"></a>00470 
<a name="l00471"></a>00471   }
<a name="l00472"></a>00472   <span class="comment">/* If we are saving models write this one out. */</span>
<a name="l00473"></a>00473   <span class="keywordflow">if</span> (<a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.is_open()) {
<a name="l00474"></a>00474     <a class="code" href="classQuantBRLMM.html#a008b914483156a796f217fae336ea644" title="Write out a text representation of the model to the file specified.">writeModelOut</a>(xxModel, <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>, <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>, 2);
<a name="l00475"></a>00475   }
<a name="l00476"></a>00476   <a class="code" href="classQuantBRLMM.html#ad0808c8bfeef3f05b4733ea08a91cd0d">brlmmCall</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, xxModel, xxCalls, xxConfidences, xxDistances, m_Param);
<a name="l00477"></a>00477 
<a name="l00478"></a>00478   <span class="comment">/* Do the male fitting and calls. */</span>
<a name="l00479"></a>00479   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; m_InitialCalls.size(); i++) {
<a name="l00480"></a>00480       <span class="keywordflow">if</span> (m_Genders[i] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669ad5afd116ccd5bf71e27bab31a23fac97" title="Has XX chromosomes.">affx::Female</a> || m_Genders[i] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669aa259d114b82ef05d3f1f9d3c81c84719" title="Don&amp;#39;t know genotype, use XX model.">affx::UnknownGender</a>) {
<a name="l00481"></a>00481       xyInitCalls[i] = NN;
<a name="l00482"></a>00482     }
<a name="l00483"></a>00483   }
<a name="l00484"></a>00484   <a class="code" href="classClusterPrior.html">ClusterPrior</a> xyPrior = prior;
<a name="l00485"></a>00485   <span class="comment">/* This is where we adjust the prior to remove het cluster. */</span>
<a name="l00486"></a>00486   <a class="code" href="classQuantBRLMM.html#a761e1e30d718599d1ce3aad6853b8ffb" title="Remove the het cluster from a prior by setting center to be very large negative number and removing c...">removeHetFromPrior</a>(xyPrior, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a928f99fd333fed0c34013ce8d23acb49" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>);
<a name="l00487"></a>00487   <span class="comment">/* Then fit as we usually would. */</span>
<a name="l00488"></a>00488   <a class="code" href="classClusterModel.html">ClusterModel</a> xyModel;
<a name="l00489"></a>00489   <span class="comment">/* Check to see if there is a precomputed model. */</span>
<a name="l00490"></a>00490   <span class="keywordflow">if</span> (!m_SnpModelMap.empty()) {
<a name="l00491"></a>00491     <span class="keywordtype">string</span> name = <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot;-1&quot;</span>;
<a name="l00492"></a>00492     iter = m_SnpModelMap.find(name);
<a name="l00493"></a>00493     <span class="keywordflow">if</span> (iter == m_SnpModelMap.end()) {
<a name="l00494"></a>00494       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t find model for SNP: &quot;</span> + <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot;-1&quot;</span>);
<a name="l00495"></a>00495     }
<a name="l00496"></a>00496     xyModel = iter-&gt;second;
<a name="l00497"></a>00497   }
<a name="l00498"></a>00498   <span class="keywordflow">else</span> {
<a name="l00499"></a>00499     xyModel = <a class="code" href="classQuantBRLMM.html#af1acb3c6affa478feb02b2443d7d4962" title="Wrapper function for getting bayesian adjusted parameters given data.">fitModel</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, xyDistances,
<a name="l00500"></a>00500                        xyInitCalls, xyPrior, <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>, m_Param);
<a name="l00501"></a>00501   }
<a name="l00502"></a>00502   <span class="comment">/* If we are saving models write this one out. */</span>
<a name="l00503"></a>00503   <span class="keywordflow">if</span> (<a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.is_open()) {
<a name="l00504"></a>00504     <a class="code" href="classQuantBRLMM.html#a008b914483156a796f217fae336ea644" title="Write out a text representation of the model to the file specified.">writeModelOut</a>(xyModel, <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>, <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>, 1);
<a name="l00505"></a>00505   }
<a name="l00506"></a>00506   <a class="code" href="classQuantBRLMM.html#ad0808c8bfeef3f05b4733ea08a91cd0d">brlmmCall</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, xyModel, xyCalls, xyConfidences, xyDistances, m_Param);
<a name="l00507"></a>00507 
<a name="l00508"></a>00508   m_Calls.resize(xyCalls.size());
<a name="l00509"></a>00509   m_Confidences.resize(xyConfidences.size());
<a name="l00510"></a>00510   m_Distances.resize(xyDistances.size());
<a name="l00511"></a>00511   <span class="comment">/* fill in the results appropriate for each gender. */</span>
<a name="l00512"></a>00512   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; m_Genders.size(); i++) {
<a name="l00513"></a>00513       <span class="keywordflow">if</span> (m_Genders[i] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669ad5afd116ccd5bf71e27bab31a23fac97" title="Has XX chromosomes.">affx::Female</a> || m_Genders[i] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669aa259d114b82ef05d3f1f9d3c81c84719" title="Don&amp;#39;t know genotype, use XX model.">affx::UnknownGender</a>) {
<a name="l00514"></a>00514       m_Calls[i] = xxCalls[i];
<a name="l00515"></a>00515       m_Confidences[i] = xxConfidences[i];
<a name="l00516"></a>00516       m_Distances[i] = xxDistances[i];
<a name="l00517"></a>00517     }
<a name="l00518"></a>00518     <span class="keywordflow">else</span> <span class="keywordflow">if</span> (m_Genders[i] == <a class="code" href="namespaceaffx.html#a1c064510a1a9f30814f840025d73d669afdfe578a9e9044b7b274b736ed341a81" title="Has XY chromosomes.">affx::Male</a>) {
<a name="l00519"></a>00519       m_Calls[i] = xyCalls[i];
<a name="l00520"></a>00520       m_Confidences[i] = xyConfidences[i];
<a name="l00521"></a>00521       m_Distances[i] = xyDistances[i];
<a name="l00522"></a>00522     }
<a name="l00523"></a>00523     <span class="keywordflow">else</span> {
<a name="l00524"></a>00524       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::callHaploidSnp() - Don&#39;t recognize gender: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_Genders[i]));
<a name="l00525"></a>00525     }
<a name="l00526"></a>00526   }
<a name="l00527"></a>00527 }
<a name="l00528"></a>00528 <span class="comment"></span>
<a name="l00529"></a>00529 <span class="comment">/**</span>
<a name="l00530"></a>00530 <span class="comment"> * Write out a text representation of the prior to the file specified.</span>
<a name="l00531"></a>00531 <span class="comment"> * @param prior - Prior to be written.</span>
<a name="l00532"></a>00532 <span class="comment"> * @param out - ostream to write to.</span>
<a name="l00533"></a>00533 <span class="comment"> */</span>
<a name="l00534"></a><a class="code" href="classQuantBRLMM.html#abf28bce895fc18fbd73d5744262486d7">00534</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#abf28bce895fc18fbd73d5744262486d7" title="Write out a text representation of the prior to the file specified.">QuantBRLMM::writePriorOut</a>(<a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior, std::ostream &amp;out,
<a name="l00535"></a>00535                                <span class="keyword">const</span> std::string&amp; fieldDelim, <span class="keyword">const</span> std::string&amp; sep) {
<a name="l00536"></a>00536   printColumnVector(prior.centers, &amp;out, sep);
<a name="l00537"></a>00537   out &lt;&lt; fieldDelim;
<a name="l00538"></a>00538   printColumnVector(prior.centerVars, &amp;out, sep);
<a name="l00539"></a>00539   out &lt;&lt; fieldDelim;
<a name="l00540"></a>00540   printMatrix(prior.covars, &amp;out, sep);
<a name="l00541"></a>00541   out &lt;&lt; endl;
<a name="l00542"></a>00542 }
<a name="l00543"></a>00543 
<a name="l00544"></a><a class="code" href="classQuantBRLMM.html#ad5e489c5edabd5fe6c0f7d1e34f777be">00544</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#ad5e489c5edabd5fe6c0f7d1e34f777be" title="Tell brlmm to output snp models to file.">QuantBRLMM::setSnpModelOut</a>(<span class="keyword">const</span> std::string&amp; fileName,affx::TsvReport::TsvReportFmt_t format)
<a name="l00545"></a>00545 {
<a name="l00546"></a>00546   <span class="comment">//</span>
<a name="l00547"></a>00547   <span class="comment">// printf(&quot;### QuantBRLMM::setSnpModelOut(&#39;%s&#39;,%d)\n&quot;,fileName.c_str(),format);</span>
<a name="l00548"></a>00548   <span class="comment">//</span>
<a name="l00549"></a>00549   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.setFilename(fileName);
<a name="l00550"></a>00550   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.setFormat(format);
<a name="l00551"></a>00551   <span class="comment">//</span>
<a name="l00552"></a>00552   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.defineStringColumn(0,0,<span class="stringliteral">&quot;probeset_id&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l00553"></a>00553   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.defineColumn(0,1,<span class="stringliteral">&quot;copy_number&quot;</span>,affx::FILE5_DTYPE_INT);
<a name="l00554"></a>00554   <span class="comment">//</span>
<a name="l00555"></a>00555   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.defineStringColumn(0,2,<span class="stringliteral">&quot;center&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l00556"></a>00556   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.defineStringColumn(0,3,<span class="stringliteral">&quot;var&quot;</span>,<a class="code" href="TsvReport_8h.html#acca8549e7b33eebca1f62ee22fd83e95" title="*sigh* So this isnt a &amp;quot;report&amp;quot; at all. /// Reports should have the signature &amp;quot;prepare/...">TSVREPORT_PROBESET_STRLEN</a>);
<a name="l00557"></a>00557   <span class="comment">// Add headers here.</span>
<a name="l00558"></a>00558   <a class="code" href="classQuantBRLMM.html#ac85a70f66b43e43943f3da174d0378de" title="Output stream for printing snp specific models.">m_SnpModelsTsv</a>.writeTsv_v1();
<a name="l00559"></a>00559 }
<a name="l00560"></a>00560 <span class="comment"></span>
<a name="l00561"></a>00561 <span class="comment">/**</span>
<a name="l00562"></a>00562 <span class="comment"> * Write out a text representation of the model to the file specified.</span>
<a name="l00563"></a>00563 <span class="comment"> * @param model - Model to be written.</span>
<a name="l00564"></a>00564 <span class="comment"> * @param out - ostream to write to.</span>
<a name="l00565"></a>00565 <span class="comment"> */</span>
<a name="l00566"></a><a class="code" href="classQuantBRLMM.html#a008b914483156a796f217fae336ea644">00566</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a008b914483156a796f217fae336ea644" title="Write out a text representation of the model to the file specified.">QuantBRLMM::writeModelOut</a>(<a class="code" href="classClusterModel.html">ClusterModel</a>&amp; model,<a class="code" href="classaffx_1_1TsvReport.html" title="Unifies the presentation of file/TsvFile and file ///.">affx::TsvReport</a>&amp; tsv, <span class="keyword">const</span> std::string&amp; name,<span class="keywordtype">int</span> copies)
<a name="l00567"></a>00567 {
<a name="l00568"></a>00568   tsv.set_string(0,0,name); <span class="comment">// &quot;probeset_id&quot;</span>
<a name="l00569"></a>00569   tsv.set_i(0,1,copies);    <span class="comment">// &quot;copy_number&quot;</span>
<a name="l00570"></a>00570   <span class="comment">//</span>
<a name="l00571"></a>00571   tsv.set_string(0,2,stringMatrix(model.clusterMeans,<span class="stringliteral">&quot;,&quot;</span>)); <span class="comment">// &quot;center&quot;</span>
<a name="l00572"></a>00572   tsv.set_string(0,3,stringMatrix(model.clusterVars, <span class="stringliteral">&quot;,&quot;</span>)); <span class="comment">// &quot;var&quot;</span>
<a name="l00573"></a>00573   <span class="comment">//</span>
<a name="l00574"></a>00574   tsv.writeLevel(0);
<a name="l00575"></a>00575 }
<a name="l00576"></a>00576 <span class="comment"></span>
<a name="l00577"></a>00577 <span class="comment">/**</span>
<a name="l00578"></a>00578 <span class="comment"> * Write out a text representation of the prior to the file specified.</span>
<a name="l00579"></a>00579 <span class="comment"> * @param prior</span>
<a name="l00580"></a>00580 <span class="comment"> * @param fileName</span>
<a name="l00581"></a>00581 <span class="comment"> */</span>
<a name="l00582"></a>00582 
<a name="l00583"></a>00583 <span class="comment">// Why is this singular</span>
<a name="l00584"></a><a class="code" href="classQuantBRLMM.html#a1a7e2033f6184a97ecef944e46c0aed5">00584</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a1a7e2033f6184a97ecef944e46c0aed5" title="Write out a text representation of the prior to the file specified.">QuantBRLMM::writePrior</a>(<a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior, <span class="keyword">const</span> std::string&amp; fileName) {
<a name="l00585"></a>00585   std::string fieldDelim = <span class="stringliteral">&quot;\t&quot;</span>;
<a name="l00586"></a>00586   std::string sep=<span class="stringliteral">&quot;,&quot;</span>;
<a name="l00587"></a>00587   ofstream out;
<a name="l00588"></a>00588   <a class="code" href="classFs.html#a1cc6e82074c84dd92bd7cdcf7d374b7c" title="Open an ofstream for writing to.">Fs::mustOpenToWrite</a>(out, fileName);
<a name="l00589"></a>00589   out.setf(ios::fixed, ios::floatfield);
<a name="l00590"></a>00590   out.precision(10);
<a name="l00591"></a>00591   out &lt;&lt; <span class="stringliteral">&quot;center&quot;</span> &lt;&lt; fieldDelim &lt;&lt; <span class="stringliteral">&quot;var&quot;</span> &lt;&lt; fieldDelim &lt;&lt; <span class="stringliteral">&quot;center.var&quot;</span> &lt;&lt; endl;
<a name="l00592"></a>00592   <a class="code" href="classQuantBRLMM.html#abf28bce895fc18fbd73d5744262486d7" title="Write out a text representation of the prior to the file specified.">QuantBRLMM::writePriorOut</a>(prior, out, fieldDelim, sep);
<a name="l00593"></a>00593 }
<a name="l00594"></a>00594 <span class="comment"></span>
<a name="l00595"></a>00595 <span class="comment">/**</span>
<a name="l00596"></a>00596 <span class="comment"> * Covert string representations into ClusterPrior</span>
<a name="l00597"></a>00597 <span class="comment"> *</span>
<a name="l00598"></a>00598 <span class="comment"> * @param id - identifying string.</span>
<a name="l00599"></a>00599 <span class="comment"> * @param center - centers.</span>
<a name="l00600"></a>00600 <span class="comment"> * @param var - variance of center variances.</span>
<a name="l00601"></a>00601 <span class="comment"> * @param covar - covariance of centers.</span>
<a name="l00602"></a>00602 <span class="comment"> *</span>
<a name="l00603"></a>00603 <span class="comment"> * @return - ClusterPrior.</span>
<a name="l00604"></a>00604 <span class="comment"> */</span>
<a name="l00605"></a><a class="code" href="classQuantBRLMM.html#a7b90f77e1f46e0f348ae56a83f03236d">00605</a> <a class="code" href="classClusterPrior.html">ClusterPrior</a> <a class="code" href="classQuantBRLMM.html#a7b90f77e1f46e0f348ae56a83f03236d" title="Covert string representations into ClusterPrior.">QuantBRLMM::clusterPriorFromStrings</a>(<span class="keyword">const</span> std::string&amp; fileName, std::string &amp;<span class="keywordtype">id</span>, std::string &amp;center,
<a name="l00606"></a>00606                                                  std::string &amp;var, std::string &amp;covar) {
<a name="l00607"></a>00607   <a class="code" href="classClusterPrior.html">ClusterPrior</a> prior;
<a name="l00608"></a>00608   vector&lt;double&gt; vec;
<a name="l00609"></a>00609   vector&lt;string&gt; words;
<a name="l00610"></a>00610 
<a name="l00611"></a>00611   prior.id = id;
<a name="l00612"></a>00612 
<a name="l00613"></a>00613   <span class="comment">/* Read in the centers. */</span>
<a name="l00614"></a>00614   <a class="code" href="classUtil.html#ac0aefd5694432916da36f4f2cfe8bd76" title="Chop up a string into a vector of words.">Util::chopString</a>(center, <span class="charliteral">&#39;,&#39;</span>, words);
<a name="l00615"></a>00615   vec.clear();
<a name="l00616"></a>00616   <span class="keywordflow">if</span> (words.size() != 6)
<a name="l00617"></a>00617     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Expecting 6 entires in &#39;center&#39; entry in file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(fileName) + <span class="stringliteral">&quot; got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(words.size()));
<a name="l00618"></a>00618   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; words.size(); i++) {
<a name="l00619"></a>00619     vec.push_back(<a class="code" href="classConvert.html#ac0adf99ce0708f0e714ffefe33286616" title="Make an double from a c string.">Convert::toDouble</a>(words[i].c_str()));
<a name="l00620"></a>00620   }
<a name="l00621"></a>00621   prior.centers = <a class="code" href="classQuantBRLMM.html#a1ad81844f338fefacfeee8e12677363a" title="Conversion utility for transforming arrays to column vectors.">columnVectorFromArray</a>(&amp;vec[0], (<span class="keywordtype">int</span>)vec.size());
<a name="l00622"></a>00622 
<a name="l00623"></a>00623   <span class="comment">/* Read in the variances. */</span>
<a name="l00624"></a>00624   <a class="code" href="classUtil.html#ac0aefd5694432916da36f4f2cfe8bd76" title="Chop up a string into a vector of words.">Util::chopString</a>(var, <span class="charliteral">&#39;,&#39;</span>, words);
<a name="l00625"></a>00625   vec.clear();
<a name="l00626"></a>00626   <span class="keywordflow">if</span> (words.size() != 9)
<a name="l00627"></a>00627     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Expecting 9 entires in &#39;var&#39; entry in file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(fileName) + <span class="stringliteral">&quot; got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(words.size()));
<a name="l00628"></a>00628   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; words.size(); i++) {
<a name="l00629"></a>00629     vec.push_back(<a class="code" href="classConvert.html#ac0adf99ce0708f0e714ffefe33286616" title="Make an double from a c string.">Convert::toDouble</a>(words[i].c_str()));
<a name="l00630"></a>00630   }
<a name="l00631"></a>00631   prior.centerVars = <a class="code" href="classQuantBRLMM.html#a1ad81844f338fefacfeee8e12677363a" title="Conversion utility for transforming arrays to column vectors.">columnVectorFromArray</a>(&amp;vec[0], (<span class="keywordtype">int</span>)vec.size());
<a name="l00632"></a>00632 
<a name="l00633"></a>00633   <span class="comment">/* Do the covariacne matrix. */</span>
<a name="l00634"></a>00634   Util::chopString(covar, <span class="charliteral">&#39;,&#39;</span>, words);
<a name="l00635"></a>00635   vec.clear();
<a name="l00636"></a>00636   <span class="keywordflow">if</span> (words.size() != 36)
<a name="l00637"></a>00637     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Expecting 36 entires in &#39;centers.var&#39; entry in file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(fileName) + <span class="stringliteral">&quot; got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(words.size()));
<a name="l00638"></a>00638   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; words.size(); i++) {
<a name="l00639"></a>00639     vec.push_back(<a class="code" href="classConvert.html#ac0adf99ce0708f0e714ffefe33286616" title="Make an double from a c string.">Convert::toDouble</a>(words[i].c_str()));
<a name="l00640"></a>00640   }
<a name="l00641"></a>00641   prior.covars = <a class="code" href="classQuantBRLMM.html#a1cb1a160c110e08d99999bf00d313d94" title="Conversion utility for transforming arrays to matrices.">matrixFromArray</a>(&amp;vec[0], (<span class="keywordtype">int</span>)vec.size(), 6, 6);
<a name="l00642"></a>00642   <span class="keywordflow">return</span> prior;
<a name="l00643"></a>00643 
<a name="l00644"></a>00644 }
<a name="l00645"></a>00645 <span class="comment"></span>
<a name="l00646"></a>00646 <span class="comment">/**</span>
<a name="l00647"></a>00647 <span class="comment"> * Covert string representations into ClusterModel</span>
<a name="l00648"></a>00648 <span class="comment"> *</span>
<a name="l00649"></a>00649 <span class="comment"> * @param id - identifying string.</span>
<a name="l00650"></a>00650 <span class="comment"> * @param center - centers.</span>
<a name="l00651"></a>00651 <span class="comment"> * @param var - variance of center variances.</span>
<a name="l00652"></a>00652 <span class="comment"> *</span>
<a name="l00653"></a>00653 <span class="comment"> * @return - ClusterModel</span>
<a name="l00654"></a>00654 <span class="comment"> */</span>
<a name="l00655"></a><a class="code" href="classQuantBRLMM.html#a38aebf72436d249fef32c09cde6ed027">00655</a> <a class="code" href="classClusterModel.html">ClusterModel</a> <a class="code" href="classQuantBRLMM.html#a38aebf72436d249fef32c09cde6ed027" title="Covert string representations into ClusterModel.">QuantBRLMM::clusterModelFromStrings</a>(<span class="keyword">const</span> std::string&amp; fileName, std::string &amp;<span class="keywordtype">id</span>,
<a name="l00656"></a>00656                                                  std::string &amp;center, std::string &amp;var) {
<a name="l00657"></a>00657   <a class="code" href="classClusterModel.html">ClusterModel</a> model;
<a name="l00658"></a>00658   vector&lt;double&gt; vec(9);
<a name="l00659"></a>00659   vector&lt;string&gt; words;
<a name="l00660"></a>00660   <span class="comment">/* Read in the centers. */</span>
<a name="l00661"></a>00661   <a class="code" href="classUtil.html#ac0aefd5694432916da36f4f2cfe8bd76" title="Chop up a string into a vector of words.">Util::chopString</a>(center, <span class="charliteral">&#39;,&#39;</span>, words);
<a name="l00662"></a>00662   vec.clear();
<a name="l00663"></a>00663   <span class="keywordflow">if</span> (words.size() != 6)
<a name="l00664"></a>00664     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Expecting 6 entires in &#39;center&#39; entry in file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(fileName) +
<a name="l00665"></a>00665                   <span class="stringliteral">&quot; got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(words.size()) + <span class="stringliteral">&quot; for probest: &quot;</span> + <span class="keywordtype">id</span>);
<a name="l00666"></a>00666   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; words.size(); i++) {
<a name="l00667"></a>00667     vec.push_back(<a class="code" href="classConvert.html#ac0adf99ce0708f0e714ffefe33286616" title="Make an double from a c string.">Convert::toDouble</a>(words[i].c_str()));
<a name="l00668"></a>00668   }
<a name="l00669"></a>00669   model.clusterMeans = <a class="code" href="classQuantBRLMM.html#a1cb1a160c110e08d99999bf00d313d94" title="Conversion utility for transforming arrays to matrices.">matrixFromArray</a>(&amp;vec[0], (<span class="keywordtype">int</span>)vec.size(), 6, 1);
<a name="l00670"></a>00670 
<a name="l00671"></a>00671   <span class="comment">/* Read in the variances. */</span>
<a name="l00672"></a>00672   <a class="code" href="classUtil.html#ac0aefd5694432916da36f4f2cfe8bd76" title="Chop up a string into a vector of words.">Util::chopString</a>(var, <span class="charliteral">&#39;,&#39;</span>, words);
<a name="l00673"></a>00673   vec.clear();
<a name="l00674"></a>00674   <span class="keywordflow">if</span> (words.size() != 9)
<a name="l00675"></a>00675     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Expecting 9 entires in &#39;var&#39; entry in file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(fileName) +
<a name="l00676"></a>00676                   <span class="stringliteral">&quot; got: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(words.size()) + <span class="stringliteral">&quot; for probest: &quot;</span> + <span class="keywordtype">id</span>);
<a name="l00677"></a>00677   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; words.size(); i++) {
<a name="l00678"></a>00678     vec.push_back(<a class="code" href="classConvert.html#ac0adf99ce0708f0e714ffefe33286616" title="Make an double from a c string.">Convert::toDouble</a>(words[i].c_str()));
<a name="l00679"></a>00679   }
<a name="l00680"></a>00680   model.clusterVars = <a class="code" href="classQuantBRLMM.html#a1cb1a160c110e08d99999bf00d313d94" title="Conversion utility for transforming arrays to matrices.">matrixFromArray</a>(&amp;vec[0], (<span class="keywordtype">int</span>)vec.size(), 9, 1);
<a name="l00681"></a>00681   <span class="keywordflow">return</span> model;
<a name="l00682"></a>00682 }
<a name="l00683"></a>00683 
<a name="l00684"></a>00684 <span class="comment"></span>
<a name="l00685"></a>00685 <span class="comment">/**</span>
<a name="l00686"></a>00686 <span class="comment"> * Read the prior from the test file specified.</span>
<a name="l00687"></a>00687 <span class="comment"> * @param fileName - Name of file to read from.</span>
<a name="l00688"></a>00688 <span class="comment"> * @return - The prior read in from file.</span>
<a name="l00689"></a>00689 <span class="comment"> */</span>
<a name="l00690"></a><a class="code" href="classQuantBRLMM.html#a71f90338aa453da6b7c453a4fe44c1ab">00690</a> <a class="code" href="classClusterPrior.html">ClusterPrior</a> <a class="code" href="classQuantBRLMM.html#a71f90338aa453da6b7c453a4fe44c1ab" title="Read the prior from the test file specified.">QuantBRLMM::loadPrior</a>(<span class="keyword">const</span> std::string&amp; fileName) {
<a name="l00691"></a>00691   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00692"></a>00692   <span class="keywordtype">string</span> center, centerVar, var, <span class="keywordtype">id</span>=<span class="stringliteral">&quot;unknown&quot;</span>;
<a name="l00693"></a>00693   <a class="code" href="classClusterPrior.html">ClusterPrior</a> prior;
<a name="l00694"></a>00694   vector&lt;double&gt; vec;
<a name="l00695"></a>00695   tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(fileName);
<a name="l00696"></a>00696   <span class="comment">//tsv.bind(0,&quot;id&quot;, &amp;id, affx::TSV_BIND_REQUIRED);</span>
<a name="l00697"></a>00697   tsv.bind(0,<span class="stringliteral">&quot;center&quot;</span>, &amp;center, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00698"></a>00698   tsv.bind(0,<span class="stringliteral">&quot;var&quot;</span>, &amp;var, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00699"></a>00699   tsv.bind(0,<span class="stringliteral">&quot;center.var&quot;</span>, &amp;centerVar, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00700"></a>00700   <span class="keywordflow">if</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) != affx::TSV_OK) {
<a name="l00701"></a>00701     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Didnt&#39; get an entry in file: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(fileName));
<a name="l00702"></a>00702   }
<a name="l00703"></a>00703   prior = <a class="code" href="classQuantBRLMM.html#a7b90f77e1f46e0f348ae56a83f03236d" title="Covert string representations into ClusterPrior.">QuantBRLMM::clusterPriorFromStrings</a>(fileName, <span class="keywordtype">id</span>, center, var, centerVar);
<a name="l00704"></a>00704   tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00705"></a>00705   <span class="keywordflow">return</span> prior;
<a name="l00706"></a>00706 }
<a name="l00707"></a>00707 <span class="comment"></span>
<a name="l00708"></a>00708 <span class="comment">/**</span>
<a name="l00709"></a>00709 <span class="comment"> * Generate a string representation of a snp model.</span>
<a name="l00710"></a>00710 <span class="comment"> * @return string - text description of snp clusters.</span>
<a name="l00711"></a>00711 <span class="comment"> */</span>
<a name="l00712"></a><a class="code" href="classQuantBRLMM.html#a29ae28aefa68a869921c8f3a42c39110">00712</a> std::string <a class="code" href="classQuantBRLMM.html#a29ae28aefa68a869921c8f3a42c39110" title="Generate a string representation of a snp model.">QuantBRLMM::getModelString</a>() {
<a name="l00713"></a>00713   <span class="keywordtype">int</span> i = 0, j = 0;
<a name="l00714"></a>00714   std::string s = <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>;
<a name="l00715"></a>00715   s += <span class="stringliteral">&quot;\t&quot;</span>;
<a name="l00716"></a>00716   <span class="keywordflow">for</span> (i = 0; i &lt; m_CurrentPrior-&gt;centers.Nrows() - 1; i++) {
<a name="l00717"></a>00717     s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;centers.element(i));
<a name="l00718"></a>00718     s += <span class="stringliteral">&quot;,&quot;</span>;
<a name="l00719"></a>00719   }
<a name="l00720"></a>00720   s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;centers.element(i));
<a name="l00721"></a>00721   s += <span class="stringliteral">&quot;\t&quot;</span>;
<a name="l00722"></a>00722   <span class="keywordflow">for</span> (i = 0; i &lt; m_CurrentPrior-&gt;centerVars.Nrows() - 1; i++) {
<a name="l00723"></a>00723     s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;centerVars.element(i));
<a name="l00724"></a>00724     s += <span class="stringliteral">&quot;,&quot;</span>;
<a name="l00725"></a>00725   }
<a name="l00726"></a>00726   s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;centerVars.element(i));
<a name="l00727"></a>00727   s += <span class="stringliteral">&quot;\t&quot;</span>;
<a name="l00728"></a>00728   <span class="keywordflow">for</span> (i = 0; i &lt; m_CurrentPrior-&gt;covars.Nrows()-1; i++) {
<a name="l00729"></a>00729     <span class="keywordflow">for</span> (j = 0; j &lt; m_CurrentPrior-&gt;covars.Ncols(); j++) {
<a name="l00730"></a>00730       s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;covars.element(i,j));
<a name="l00731"></a>00731       s += <span class="stringliteral">&quot;,&quot;</span>;
<a name="l00732"></a>00732     }
<a name="l00733"></a>00733   }
<a name="l00734"></a>00734   <span class="keywordflow">for</span> (j = 0; j &lt; m_CurrentPrior-&gt;covars.Ncols() - 1; j++) {
<a name="l00735"></a>00735     s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;covars.element(i,j));
<a name="l00736"></a>00736     s += <span class="stringliteral">&quot;,&quot;</span>;
<a name="l00737"></a>00737   }
<a name="l00738"></a>00738   s += <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(m_CurrentPrior-&gt;covars.element(i,j));
<a name="l00739"></a>00739   <span class="keywordflow">return</span> s;
<a name="l00740"></a>00740 }
<a name="l00741"></a>00741 <span class="comment"></span>
<a name="l00742"></a>00742 <span class="comment">/**</span>
<a name="l00743"></a>00743 <span class="comment"> * Load in a number of snp specific models.</span>
<a name="l00744"></a>00744 <span class="comment"> * @param snpModels - map of snp probeset names to models.</span>
<a name="l00745"></a>00745 <span class="comment"> * @param fileName - filename to read models from.</span>
<a name="l00746"></a>00746 <span class="comment"> */</span>
<a name="l00747"></a><a class="code" href="classQuantBRLMM.html#a5f6db5e8a6818003a7e4629e555366a3">00747</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a5f6db5e8a6818003a7e4629e555366a3" title="Load in a number of snp specific models.">QuantBRLMM::loadSnpModels</a>(std::map&lt;std::string, ClusterModel&gt; &amp;snpModels,<span class="keyword">const</span> std::string&amp; fileName) {
<a name="l00748"></a>00748   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00749"></a>00749   <span class="keywordtype">string</span> center, var, id, copies;
<a name="l00750"></a>00750   <a class="code" href="classClusterModel.html">ClusterModel</a> model;
<a name="l00751"></a>00751   snpModels.clear();
<a name="l00752"></a>00752   tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(fileName);
<a name="l00753"></a>00753 
<a name="l00754"></a>00754   <span class="comment">// we want to allow either &quot;probeset_id&quot; or &quot;id&quot;...</span>
<a name="l00755"></a>00755   <span class="keywordtype">int</span> probeset_id_cidx;
<a name="l00756"></a>00756   probeset_id_cidx=tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2dbd33415a815e221ae61cf3feac5500" title="Covert a column index to a column index ///.">cname2cidx</a>(0,<span class="stringliteral">&quot;probeset_id&quot;</span>,<span class="stringliteral">&quot;id&quot;</span>);
<a name="l00757"></a>00757   <span class="keywordflow">if</span> (probeset_id_cidx&lt;0) {
<a name="l00758"></a>00758     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::loadSnpModels: no &#39;probeset_id&#39; or &#39;id&#39; column&quot;</span>);
<a name="l00759"></a>00759   }
<a name="l00760"></a>00760   <span class="keywordtype">int</span> copy_number_cidx;
<a name="l00761"></a>00761   copy_number_cidx=tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2dbd33415a815e221ae61cf3feac5500" title="Covert a column index to a column index ///.">cname2cidx</a>(0,<span class="stringliteral">&quot;copy_number&quot;</span>,<span class="stringliteral">&quot;copies&quot;</span>);
<a name="l00762"></a>00762   <span class="keywordflow">if</span> (copy_number_cidx&lt;0) {
<a name="l00763"></a>00763     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::loadSnpModels: no &#39;copy_number&#39; or &#39;copies&#39; column&quot;</span>);
<a name="l00764"></a>00764   }
<a name="l00765"></a>00765   <span class="comment">//</span>
<a name="l00766"></a>00766   tsv.bind(0,probeset_id_cidx, &amp;<span class="keywordtype">id</span>    , <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00767"></a>00767   tsv.bind(0,copy_number_cidx, &amp;copies, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00768"></a>00768   tsv.bind(0,<span class="stringliteral">&quot;center&quot;</span>, &amp;center, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00769"></a>00769   tsv.bind(0,<span class="stringliteral">&quot;var&quot;</span>, &amp;var, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00770"></a>00770   <span class="comment">//</span>
<a name="l00771"></a>00771   <span class="keywordflow">while</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == affx::TSV_OK) {
<a name="l00772"></a>00772     model = <a class="code" href="classQuantBRLMM.html#a38aebf72436d249fef32c09cde6ed027" title="Covert string representations into ClusterModel.">clusterModelFromStrings</a>(fileName, <span class="keywordtype">id</span>, center, var);
<a name="l00773"></a>00773     <span class="keywordflow">if</span> (snpModels.find(<span class="keywordtype">id</span>) != snpModels.end()) {
<a name="l00774"></a>00774       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::loadSnpModels() - Id: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<span class="keywordtype">id</span>) + <span class="stringliteral">&quot; seen multiple times.&quot;</span>);
<a name="l00775"></a>00775     }
<a name="l00776"></a>00776     snpModels[<span class="keywordtype">id</span> + <span class="stringliteral">&quot;-&quot;</span> + copies] = model;
<a name="l00777"></a>00777   }
<a name="l00778"></a>00778   tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00779"></a>00779 }
<a name="l00780"></a>00780 <span class="comment"></span>
<a name="l00781"></a>00781 <span class="comment">/**</span>
<a name="l00782"></a>00782 <span class="comment"> * Load in a number of snp specific priors.</span>
<a name="l00783"></a>00783 <span class="comment"> * @param snpPriors - map of snp probeset names to priors.</span>
<a name="l00784"></a>00784 <span class="comment"> * @param fileName - filename to read priors from.</span>
<a name="l00785"></a>00785 <span class="comment"> */</span>
<a name="l00786"></a><a class="code" href="classQuantBRLMM.html#a90060a28781e3d3d92bba20fba16c81f">00786</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#a90060a28781e3d3d92bba20fba16c81f" title="Load in a number of snp specific priors.">QuantBRLMM::loadSnpPriors</a>(std::map&lt;std::string,ClusterPrior&gt; &amp;snpPriors,
<a name="l00787"></a>00787                                <span class="keyword">const</span> std::string&amp; fileName) {
<a name="l00788"></a>00788   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> tsv;
<a name="l00789"></a>00789   <span class="keywordtype">string</span> center, centerVar, var, <span class="keywordtype">id</span>=<span class="stringliteral">&quot;unknown&quot;</span>;
<a name="l00790"></a>00790   <a class="code" href="classClusterPrior.html">ClusterPrior</a> prior;
<a name="l00791"></a>00791   snpPriors.clear();
<a name="l00792"></a>00792   tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(fileName);
<a name="l00793"></a>00793   tsv.bind(0,<span class="stringliteral">&quot;id&quot;</span>, &amp;<span class="keywordtype">id</span>, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00794"></a>00794   tsv.bind(0,<span class="stringliteral">&quot;center&quot;</span>, &amp;center, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00795"></a>00795   tsv.bind(0,<span class="stringliteral">&quot;var&quot;</span>, &amp;var, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00796"></a>00796   tsv.bind(0,<span class="stringliteral">&quot;center.var&quot;</span>, &amp;centerVar, <a class="code" href="namespaceaffx.html#a8f36853f1d3b6cfb79edebd82760691daf605fbe0e244c4a604d176991d8df2c6" title="warn if not bound">affx::TSV_BIND_REQUIRED</a>);
<a name="l00797"></a>00797   <span class="keywordflow">while</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == affx::TSV_OK) {
<a name="l00798"></a>00798     prior = <a class="code" href="classQuantBRLMM.html#a7b90f77e1f46e0f348ae56a83f03236d" title="Covert string representations into ClusterPrior.">clusterPriorFromStrings</a>(fileName, <span class="keywordtype">id</span>, center, var, centerVar);
<a name="l00799"></a>00799     <span class="keywordflow">if</span> (snpPriors.find(prior.id) != snpPriors.end()) {
<a name="l00800"></a>00800       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::loadSnpPriors() - Id: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(prior.id) + <span class="stringliteral">&quot; seen multiple times.&quot;</span>);
<a name="l00801"></a>00801     }
<a name="l00802"></a>00802     snpPriors[prior.id] = prior;
<a name="l00803"></a>00803   }
<a name="l00804"></a>00804   tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00805"></a>00805 }
<a name="l00806"></a>00806 
<a name="l00807"></a>00807 <span class="comment"></span>
<a name="l00808"></a>00808 <span class="comment">/**</span>
<a name="l00809"></a>00809 <span class="comment"> * Compute the prior from a collection of snps and fill in the prior.</span>
<a name="l00810"></a>00810 <span class="comment"> *</span>
<a name="l00811"></a>00811 <span class="comment"> * @param snpStats - Stats for a relatively large number of snps</span>
<a name="l00812"></a>00812 <span class="comment"> * @param prior - Fill in the prior.</span>
<a name="l00813"></a>00813 <span class="comment"> */</span>
<a name="l00814"></a><a class="code" href="classQuantBRLMM.html#afec1a0f13ad6d4586fae2c084d2e2e33">00814</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#afec1a0f13ad6d4586fae2c084d2e2e33" title="Compute the prior from a collection of snps and fill in the prior.">QuantBRLMM::computeClusterPrior</a>(<span class="keyword">const</span> std::vector&lt;ClusterStats&gt; &amp;snpStats, <a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior) {
<a name="l00815"></a>00815   vector&lt;vector&lt;double&gt; &gt; means(6);
<a name="l00816"></a>00816   vector&lt;vector&lt;double&gt; &gt; vars(9);
<a name="l00817"></a>00817   prior.snpCount = snpStats.size();
<a name="l00818"></a>00818   <span class="comment">/* Save up vectors of the values for calculating stats. */</span>
<a name="l00819"></a>00819   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> snpIx = 0; snpIx &lt; snpStats.size(); snpIx++) {
<a name="l00820"></a>00820     <span class="keyword">const</span> <a class="code" href="classClusterStats.html">ClusterStats</a> &amp;stats = snpStats[snpIx];
<a name="l00821"></a>00821     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> meanIx = 0; meanIx &lt; stats.<a class="code" href="classClusterStats.html#aa4a1a793033967c11e9abaa3ecd0cc3e" title="Simon style means object, not sure why indexed like this but it /// goes: /// mMeans[0],aMeans[0],mMeans[1],aMeans[1],etc..">means</a>.size(); meanIx++) {
<a name="l00822"></a>00822       means[meanIx].push_back(stats.<a class="code" href="classClusterStats.html#aa4a1a793033967c11e9abaa3ecd0cc3e" title="Simon style means object, not sure why indexed like this but it /// goes: /// mMeans[0],aMeans[0],mMeans[1],aMeans[1],etc..">means</a>[meanIx]);
<a name="l00823"></a>00823     }
<a name="l00824"></a>00824     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> varIx = 0; varIx &lt; stats.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>.size(); varIx++) {
<a name="l00825"></a>00825       vars[varIx].push_back(stats.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>[varIx]);
<a name="l00826"></a>00826     }
<a name="l00827"></a>00827   }
<a name="l00828"></a>00828   prior.centers.ReSize(6);
<a name="l00829"></a>00829   prior.centerVars.ReSize(9);
<a name="l00830"></a>00830   prior.covars.ReSize(6,6);
<a name="l00831"></a>00831   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> meanIx = 0; meanIx &lt; means.size(); meanIx++) {
<a name="l00832"></a>00832     prior.centers.element(meanIx) = <a class="code" href="classAverage.html">Average&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(means[meanIx].begin(),
<a name="l00833"></a>00833                                                                                means[meanIx].end());
<a name="l00834"></a>00834     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> nIx = 0; nIx &lt; means.size(); nIx++) {
<a name="l00835"></a>00835       prior.covars.element(meanIx, nIx) = <a class="code" href="classUnbiasedCovariance.html">UnbiasedCovariance&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(means[meanIx].begin(),
<a name="l00836"></a>00836                                                                                                 means[meanIx].end(),
<a name="l00837"></a>00837                                                                                                 means[nIx].begin());
<a name="l00838"></a>00838     }
<a name="l00839"></a>00839   }
<a name="l00840"></a>00840   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> varIx = 0; varIx &lt; vars.size(); varIx++)
<a name="l00841"></a>00841     prior.centerVars.element(varIx) = <a class="code" href="classAverage.html">Average&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(vars[varIx].begin(), vars[varIx].end());
<a name="l00842"></a>00842 }
<a name="l00843"></a>00843 <span class="comment"></span>
<a name="l00844"></a>00844 <span class="comment">/**</span>
<a name="l00845"></a>00845 <span class="comment"> * Construct a prior from a collection of probesets. Note that to be part of</span>
<a name="l00846"></a>00846 <span class="comment"> * the prior a SNP must have at least 2 examples per AA, AB, and BB otherwise</span>
<a name="l00847"></a>00847 <span class="comment"> * the variance will be zero.</span>
<a name="l00848"></a>00848 <span class="comment"> *</span>
<a name="l00849"></a>00849 <span class="comment"> * @param probeSets - ProbeSets to be used in computing prior.</span>
<a name="l00850"></a>00850 <span class="comment"> * @param layout - description and chip and probesets.</span>
<a name="l00851"></a>00851 <span class="comment"> * @param iMart - Cel file data for probesets.</span>
<a name="l00852"></a>00852 <span class="comment"> * @param iTrans - Transformers to operate on raw data (i.e. normalization and bg subtraction).</span>
<a name="l00853"></a>00853 <span class="comment"> * @param pmAdjust - Individual probe adjustment like pm-mm or pm-only.</span>
<a name="l00854"></a>00854 <span class="comment"> *</span>
<a name="l00855"></a>00855 <span class="comment"> * @return - Prior constructed from summary of all the probesets.</span>
<a name="l00856"></a>00856 <span class="comment"> */</span>
<a name="l00857"></a><a class="code" href="classQuantBRLMM.html#afafae26ed3cd67f3e51ca8fd77cc7669">00857</a> <a class="code" href="classClusterPrior.html">ClusterPrior</a> <a class="code" href="classQuantBRLMM.html#afafae26ed3cd67f3e51ca8fd77cc7669" title="Construct a prior from a collection of probesets.">QuantBRLMM::makePrior</a>(std::vector&lt;string&gt;&amp; probeSetNames,
<a name="l00858"></a>00858                                    <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a>&amp; layout,
<a name="l00859"></a>00859                                    <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a>&amp; iMart,
<a name="l00860"></a>00860                                    std::vector&lt;ChipStream *&gt;&amp; iTrans,
<a name="l00861"></a>00861                                    <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust)
<a name="l00862"></a>00862 {
<a name="l00863"></a>00863   <a class="code" href="classClusterPrior.html">ClusterPrior</a> prior;
<a name="l00864"></a>00864   vector&lt;ClusterStats&gt; snpStats;
<a name="l00865"></a>00865   <span class="keywordtype">int</span> minCalls = m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa94a4cd0558a115b819dadc3501bff80" title="Minimum examples for each genotype to be included in prior estimation.">m_MinCluster</a>;
<a name="l00866"></a>00866   <a class="code" href="classErr.html#ad1a633f602fe60689cdc12d9ca1842e9" title="errAbort based version of assert with an error message.">Err::check</a>(minCalls &gt;= 2, <span class="stringliteral">&quot;QuantBRLMM::makePrior() - Threshold must have at least two calls per genotype.&quot;</span>);
<a name="l00867"></a>00867   <span class="keywordflow">try</span> {
<a name="l00868"></a>00868     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> psIx = 0; psIx &lt; probeSetNames.size(); psIx++) {
<a name="l00869"></a>00869       <a class="code" href="classProbeListPacked.html">ProbeListPacked</a> pList = layout.<a class="code" href="classChipLayout.html#ae465f92816b1c9983a45d1f73293cc7b" title="Find a probe set using the name as a key.">getProbeListByName</a>(probeSetNames[psIx]);<span class="comment"></span>
<a name="l00870"></a>00870 <span class="comment">      ///@todo deal with missing probes from kill list</span>
<a name="l00871"></a>00871 <span class="comment"></span>      <span class="keywordflow">if</span> (pList.isNull()) {
<a name="l00872"></a>00872          <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::makePrior(): Cannot deal with missing probes from kill list&quot;</span>);
<a name="l00873"></a>00873       }
<a name="l00874"></a>00874       <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs = <a class="code" href="classProbeListFactory.html#aaed559bbfe723ed4fe225f805a2411e8">ProbeListFactory::asProbeSet</a>(pList);
<a name="l00875"></a>00875       <span class="comment">// Have to skip probesets that don&#39;t have 2 or 4 groups as it</span>
<a name="l00876"></a>00876       <span class="comment">// appears that some are mislabelled in the cdf file.</span>
<a name="l00877"></a>00877       <span class="keywordflow">if</span> (gtPs-&gt;<a class="code" href="classProbeSet.html#aa3e4170184169cc3bee6d958d74940ba" title="What type of probeset is this?">psType</a> != ProbeSet::GenoType || (gtPs-&gt;<a class="code" href="classProbeSet.html#a6d8ed948634c42033bd0c35d2640b274" title="How many groups (or blocks) are there? Used extensively in genotyping chips.">numGroups</a> != 2 &amp;&amp; gtPs-&gt;<a class="code" href="classProbeSet.html#a6d8ed948634c42033bd0c35d2640b274" title="How many groups (or blocks) are there? Used extensively in genotyping chips.">numGroups</a> != 4))
<a name="l00878"></a>00878         <span class="keywordflow">continue</span>;
<a name="l00879"></a>00879       <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.clear();
<a name="l00880"></a>00880       <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.clear();
<a name="l00881"></a>00881       <span class="comment">/* Load up the postulated genotypes for this snp. */</span>
<a name="l00882"></a>00882       <span class="comment">//map&lt;string, vector&lt;char&gt; &gt;::iterator mapIter = m_KnownGenoTypes.find(gtPs-&gt;name);</span>
<a name="l00883"></a>00883       KnownGenoTypes_iter_t mapIter = m_KnownGenoTypes.find(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>);
<a name="l00884"></a>00884       <span class="keywordflow">if</span> (mapIter == m_KnownGenoTypes.end()) {
<a name="l00885"></a>00885         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::makePrior() - Can&#39;t find genotype data for SNP: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>));
<a name="l00886"></a>00886       }
<a name="l00887"></a>00887       vector&lt;GType&gt; &amp;gType = mapIter-&gt;second;
<a name="l00888"></a>00888       m_InitialCalls.clear();
<a name="l00889"></a>00889       <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; gType.size(); i++) {
<a name="l00890"></a>00890         m_InitialCalls.push_back(gType[i]);
<a name="l00891"></a>00891       }
<a name="l00892"></a>00892       <a class="code" href="classQuantGTypeMethod.html#a1a853c2d2821e9c4930fb5c3807f36f9" title="fill in the probe set with the allele values">fillInAlleleProbeSets</a>(*gtPs, <a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00893"></a>00893       <span class="keywordflow">if</span> (!<a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, iMart, iTrans, pmAdjust, m_QuantMethod, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa9ebe7bbc8cb7de369bf2df9e795690c" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression testss).">m_LowPrecision</a>))
<a name="l00894"></a>00894         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Couldn&#39;t calculate A summary values for SNP: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>));
<a name="l00895"></a>00895       <span class="keywordflow">if</span> (!<a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, iMart, iTrans, pmAdjust, m_QuantMethod, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa9ebe7bbc8cb7de369bf2df9e795690c" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression testss).">m_LowPrecision</a>))
<a name="l00896"></a>00896         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Couldn&#39;t calculate B summary values for SNP: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>));
<a name="l00897"></a>00897       transformData(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, m_Param);
<a name="l00898"></a>00898       <a class="code" href="classClusterStats.html">ClusterStats</a> stats = <a class="code" href="classQuantBRLMM.html#ad80f38b0695ffc08aef491b000deacb3" title="Count up some elementary statistics about a particular snp over multiple experiments.">computeClusterStats</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, m_InitialCalls,
<a name="l00899"></a>00899                                                m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a928f99fd333fed0c34013ce8d23acb49" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a42c05834641c32011c0f5b6eac7ffe97" title="Scaling Parameter to use in transformations.">m_K</a>);
<a name="l00900"></a>00900       <span class="comment">/* Only use these stats if there at least two data points per snp. */</span>
<a name="l00901"></a>00901       <span class="keywordflow">if</span> (stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>.size() == 3 &amp;&amp; stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[0] &gt;= minCalls &amp;&amp;
<a name="l00902"></a>00902          stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[1] &gt;= minCalls &amp;&amp; stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[2] &gt;= minCalls) {
<a name="l00903"></a>00903         snpStats.push_back(stats);
<a name="l00904"></a>00904       }
<a name="l00905"></a>00905       <span class="keyword">delete</span> gtPs;
<a name="l00906"></a>00906     }
<a name="l00907"></a>00907     <a class="code" href="classQuantBRLMM.html#afec1a0f13ad6d4586fae2c084d2e2e33" title="Compute the prior from a collection of snps and fill in the prior.">computeClusterPrior</a>(snpStats, prior);
<a name="l00908"></a>00908   }
<a name="l00909"></a>00909   <span class="keywordflow">catch</span>(...) {
<a name="l00910"></a>00910     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;exception while calculating prior.&quot;</span>);
<a name="l00911"></a>00911   }
<a name="l00912"></a>00912   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(snpStats.size()) + <span class="stringliteral">&quot; of &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(probeSetNames.size()) + <span class="stringliteral">&quot; (&quot;</span> +
<a name="l00913"></a>00913                <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>((<span class="keywordtype">float</span>)snpStats.size()/probeSetNames.size()*100.0) +
<a name="l00914"></a>00914                <span class="stringliteral">&quot;%) of SNPs had at least 2 observations per genotype.&quot;</span>);
<a name="l00915"></a>00915   <span class="keywordflow">return</span> prior;
<a name="l00916"></a>00916 }
<a name="l00917"></a>00917 <span class="comment"></span>
<a name="l00918"></a>00918 <span class="comment">/**</span>
<a name="l00919"></a>00919 <span class="comment"> * @brief Set up the quantification method given all the data about the probe</span>
<a name="l00920"></a>00920 <span class="comment"> * set, chip layout and data.</span>
<a name="l00921"></a>00921 <span class="comment"> *</span>
<a name="l00922"></a>00922 <span class="comment"> * @param psGroup - Probes to be used for final estimate.</span>
<a name="l00923"></a>00923 <span class="comment"> * @param layout - Chip layout annotation.</span>
<a name="l00924"></a>00924 <span class="comment"> * @param iMart - Raw data from chips.</span>
<a name="l00925"></a>00925 <span class="comment"> * @param iTrans - Transformations to be applied to data before use.</span>
<a name="l00926"></a>00926 <span class="comment"> * @param pmAdjust - How to estimate background, or MM probe.</span>
<a name="l00927"></a>00927 <span class="comment"> * @return True if setup sucessful, false otherwise.</span>
<a name="l00928"></a>00928 <span class="comment"> */</span>
<a name="l00929"></a><a class="code" href="classQuantBRLMM.html#a98c9763caf138fe3a960c0702e6dc723">00929</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantBRLMM.html#a98c9763caf138fe3a960c0702e6dc723" title="Set up the quantification method given all the data about the probe set, chip layout and data...">QuantBRLMM::setUp</a>(<a class="code" href="classProbeSetGroup.html" title="Group of probe sets that should be processed as a single large probe set.">ProbeSetGroup</a> &amp;psGroup, <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, 
<a name="l00930"></a>00930                        std::vector&lt;ChipStream *&gt; &amp;iTrans, <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust) {
<a name="l00931"></a>00931   <span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs = NULL; <span class="comment">// Genotyping probeset</span>
<a name="l00932"></a>00932   <a class="code" href="classQuantBRLMM.html#a680eb2ad5f77c33e5f91a11d01b460cf" title="Clear out all of our results and values.">blankSelf</a>(); <span class="comment">// Make sure we clear out the past analysis before starting this one.</span>
<a name="l00933"></a>00933   <span class="keywordtype">bool</span> success = <span class="keyword">true</span>;
<a name="l00934"></a>00934   <span class="comment">/* Sanity checks about probesets. */</span>
<a name="l00935"></a>00935   <span class="keywordflow">if</span> (psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>.empty())
<a name="l00936"></a>00936     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Zero probesets in ProbeSetGroup (group: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(psGroup.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00937"></a>00937   <span class="comment">/* Remember this probeset. */</span>
<a name="l00938"></a>00938   gtPs = psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>[0];
<a name="l00939"></a>00939   <a class="code" href="classQuantGTypeMethod.html#a447dc22d0ca6a8c52c522f1f7eddfa41" title="Genotyping probeset currently using.">m_GtProbeSet</a> = gtPs;
<a name="l00940"></a>00940   <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> = gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>;
<a name="l00941"></a>00941   <span class="keywordflow">if</span> (!<a class="code" href="classQuantGTypeMethod.html#a71d6dfe5d15e5d77354f770913c23824" title="Can the ProbeSet set up the intensities and summaries.">canSetUpProbeSet</a>(gtPs)) {<span class="keywordflow">return</span> <span class="keyword">false</span>;}
<a name="l00942"></a>00942   <span class="keywordflow">if</span> (psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>.size() &gt; 1)
<a name="l00943"></a>00943     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t have multiple probesets in a genotyping ProbeSetGroup (group: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(psGroup.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00944"></a>00944 
<a name="l00945"></a>00945   <span class="comment">/* Load up the postulated genotypes for this snp. */</span>
<a name="l00946"></a>00946   KnownGenoTypes_t::iterator mapIter = m_KnownGenoTypes.find(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>);
<a name="l00947"></a>00947   <span class="keywordflow">if</span> (mapIter == m_KnownGenoTypes.end()) {
<a name="l00948"></a>00948     <a class="code" href="classQuantBRLMM.html#a680eb2ad5f77c33e5f91a11d01b460cf" title="Clear out all of our results and values.">blankSelf</a>();
<a name="l00949"></a>00949     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(2, <span class="stringliteral">&quot;QuantBRLMM::setUp() - Can&#39;t find genotype data for SNP: (probeset: name=&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>) + <span class="stringliteral">&quot;, type=&quot;</span> + <a class="code" href="classProbeSet.html#a24255185fa5ead036e3791ad41298b55" title="Generate probeset_type string from enum Type.">ProbeSet::typeEnumToString</a>(gtPs-&gt;<a class="code" href="classProbeSet.html#aa3e4170184169cc3bee6d958d74940ba" title="What type of probeset is this?">psType</a>) + <span class="stringliteral">&quot;)&quot;</span>);
<a name="l00950"></a>00950     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00951"></a>00951   }
<a name="l00952"></a>00952   <span class="keywordflow">if</span> (m_QuantMethod == NULL) {
<a name="l00953"></a>00953     <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> factory(QuantMethodFactory::Expression);
<a name="l00954"></a>00954     <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> layout;
<a name="l00955"></a>00955     <span class="keywordtype">string</span> spec = <span class="stringliteral">&quot;plier.optmethod=1&quot;</span>;
<a name="l00956"></a>00956     m_QuantMethod = factory.<a class="code" href="classQuantMethodFactory.html#abfb5fe4b06ae6e00419c35d55885feed" title="Factory for creating QuantMethod from string representation.">quantExprMethodForString</a>(spec, layout, QuantMethodFactory::Expression);
<a name="l00957"></a>00957   }
<a name="l00958"></a>00958 
<a name="l00959"></a>00959   vector&lt;affx::GType&gt; &amp;gType = mapIter-&gt;second;
<a name="l00960"></a>00960   m_InitialCalls.clear();
<a name="l00961"></a>00961   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; gType.size(); i++) {
<a name="l00962"></a>00962      m_InitialCalls.push_back(gType[i]);
<a name="l00963"></a>00963    }
<a name="l00964"></a>00964   <a class="code" href="classQuantGTypeMethod.html#a1a853c2d2821e9c4930fb5c3807f36f9" title="fill in the probe set with the allele values">fillInAlleleProbeSets</a>(*gtPs, <a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00965"></a>00965 
<a name="l00966"></a>00966   <span class="comment">/* Get summaries for each allele. */</span>
<a name="l00967"></a>00967   success &amp;= <a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, iMart, iTrans, pmAdjust,
<a name="l00968"></a>00968                              m_QuantMethod, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa9ebe7bbc8cb7de369bf2df9e795690c" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression testss).">m_LowPrecision</a>, <span class="keyword">true</span>, m_Reporters);
<a name="l00969"></a>00969   <span class="keywordflow">if</span> (gtPs-&gt;<a class="code" href="classProbeSet.html#aa3e4170184169cc3bee6d958d74940ba" title="What type of probeset is this?">psType</a> == ProbeSet::Copynumber) {
<a name="l00970"></a>00970     <a class="code" href="classQuantBRLMM.html#a680eb2ad5f77c33e5f91a11d01b460cf" title="Clear out all of our results and values.">blankSelf</a>();
<a name="l00971"></a>00971     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00972"></a>00972   }
<a name="l00973"></a>00973   success &amp;= <a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, iMart, iTrans, pmAdjust,
<a name="l00974"></a>00974                              m_QuantMethod, m_Param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#aa9ebe7bbc8cb7de369bf2df9e795690c" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression testss).">m_LowPrecision</a>, <span class="keyword">true</span>, m_Reporters);
<a name="l00975"></a>00975   <span class="comment">//</span>
<a name="l00976"></a>00976   <span class="keywordflow">if</span> (<a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.is_open()) {
<a name="l00977"></a>00977     <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.set_string(0,0,<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>);
<a name="l00978"></a>00978     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; m_InitialCalls.size(); i++) {
<a name="l00979"></a>00979       <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.set_i(0,1+i,<a class="code" href="namespaceaffx.html#ac9e9039b462ebd2bec646ed4c5ff1122" title="Convert a GType enum to an external integer value ///.">affx::GType_to_int</a>(m_InitialCalls[i]));
<a name="l00980"></a>00980     }
<a name="l00981"></a>00981     <a class="code" href="classQuantBRLMM.html#a36e248207d3a69d8925365df52491fb9" title="Output stream for printing snp specific models.">m_DmCallsTsv</a>.writeLevel(0);
<a name="l00982"></a>00982   }
<a name="l00983"></a>00983 
<a name="l00984"></a>00984   <span class="keywordflow">if</span> (!success)
<a name="l00985"></a>00985     <a class="code" href="classQuantBRLMM.html#a680eb2ad5f77c33e5f91a11d01b460cf" title="Clear out all of our results and values.">blankSelf</a>();
<a name="l00986"></a>00986   <span class="keywordflow">return</span> success;
<a name="l00987"></a>00987 }
<a name="l00988"></a>00988 
<a name="l00989"></a>00989 <a class="code" href="classClusterPrior.html">ClusterPrior</a> *QuantBRLMM::priorForSnp() {
<a name="l00990"></a>00990   <span class="keywordflow">if</span> (m_SnpPriorMap.empty()) {
<a name="l00991"></a>00991     m_CurrentPrior = m_Prior;
<a name="l00992"></a>00992     <span class="keywordflow">return</span> m_Prior;
<a name="l00993"></a>00993   }
<a name="l00994"></a>00994   <span class="keywordflow">else</span> {
<a name="l00995"></a>00995     std::map&lt;std::string,ClusterPrior&gt;::iterator mapIter = m_SnpPriorMap.find(<a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>);
<a name="l00996"></a>00996     <span class="keywordflow">if</span> (mapIter == m_SnpPriorMap.end()) {
<a name="l00997"></a>00997       m_CurrentPrior = m_Prior;
<a name="l00998"></a>00998       <span class="keywordflow">return</span> m_Prior;
<a name="l00999"></a>00999       <span class="comment">//      Err::errAbort(&quot;QuantBRLMM::priorForSnp() - Can&#39;t find prior for SNP: &quot; + m_ProbesetName);</span>
<a name="l01000"></a>01000     }
<a name="l01001"></a>01001     m_CurrentPrior = &amp;mapIter-&gt;second;
<a name="l01002"></a>01002     <span class="keywordflow">return</span> &amp;mapIter-&gt;second;
<a name="l01003"></a>01003   }
<a name="l01004"></a>01004 }
<a name="l01005"></a>01005 <span class="comment"></span>
<a name="l01006"></a>01006 <span class="comment">/** Conversion utility for transforming vectors to column vectors. */</span>
<a name="l01007"></a><a class="code" href="classQuantBRLMM.html#a5205c4974d362c640198f75c0547982d">01007</a> ColumnVector <a class="code" href="classQuantBRLMM.html#a5205c4974d362c640198f75c0547982d" title="Conversion utility for transforming vectors to column vectors.">QuantBRLMM::columnVectorFromVector</a>(std::vector&lt;double&gt; &amp;v) {
<a name="l01008"></a>01008   ColumnVector x;
<a name="l01009"></a>01009   x.ReSize(v.size());
<a name="l01010"></a>01010   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; v.size(); i++)
<a name="l01011"></a>01011     x.element(i) = v[i];
<a name="l01012"></a>01012   <span class="keywordflow">return</span> x;
<a name="l01013"></a>01013 }
<a name="l01014"></a>01014 <span class="comment"></span>
<a name="l01015"></a>01015 <span class="comment">/** Conversion utility for transforming arrays to column vectors. */</span>
<a name="l01016"></a><a class="code" href="classQuantBRLMM.html#a1ad81844f338fefacfeee8e12677363a">01016</a> ColumnVector <a class="code" href="classQuantBRLMM.html#a1ad81844f338fefacfeee8e12677363a" title="Conversion utility for transforming arrays to column vectors.">QuantBRLMM::columnVectorFromArray</a>(<span class="keywordtype">double</span> *array, <span class="keywordtype">int</span> count) {
<a name="l01017"></a>01017   ColumnVector x;
<a name="l01018"></a>01018   x.ReSize(count);
<a name="l01019"></a>01019   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; count; i++)
<a name="l01020"></a>01020     x.element(i) = array[i];
<a name="l01021"></a>01021   <span class="keywordflow">return</span> x;
<a name="l01022"></a>01022 }
<a name="l01023"></a>01023 <span class="comment"></span>
<a name="l01024"></a>01024 <span class="comment">/** Conversion utility for transforming arrays to matrices. */</span>
<a name="l01025"></a><a class="code" href="classQuantBRLMM.html#a1cb1a160c110e08d99999bf00d313d94">01025</a> Matrix <a class="code" href="classQuantBRLMM.html#a1cb1a160c110e08d99999bf00d313d94" title="Conversion utility for transforming arrays to matrices.">QuantBRLMM::matrixFromArray</a>(<span class="keywordtype">double</span> *array, <span class="keywordtype">int</span> count, <span class="keywordtype">int</span> nRow, <span class="keywordtype">int</span> nCol) {
<a name="l01026"></a>01026   Matrix m(nRow, nCol);
<a name="l01027"></a>01027   assert(nRow * nCol == count);
<a name="l01028"></a>01028   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; count; i++) {
<a name="l01029"></a>01029     <span class="keywordtype">int</span> col = i / nRow;
<a name="l01030"></a>01030     <span class="keywordtype">int</span> row = i - col * nRow;
<a name="l01031"></a>01031     m.element(row, col) = array[i];
<a name="l01032"></a>01032   }
<a name="l01033"></a>01033   <span class="keywordflow">return</span> m;
<a name="l01034"></a>01034 }
<a name="l01035"></a>01035 
<a name="l01036"></a>01036 <span class="comment"></span>
<a name="l01037"></a>01037 <span class="comment">/** Make banded count matrix. */</span>
<a name="l01038"></a><a class="code" href="classQuantBRLMM.html#a76dda9af539b10df3326cf43c113b147">01038</a> Matrix <a class="code" href="classQuantBRLMM.html#a76dda9af539b10df3326cf43c113b147" title="Make banded count matrix.">QuantBRLMM::makeCountMat</a>(std::vector&lt;int&gt; &amp;counts) {
<a name="l01039"></a>01039   Matrix m(6,6);
<a name="l01040"></a>01040   m = 0;
<a name="l01041"></a>01041   assert(counts.size() == 3);
<a name="l01042"></a>01042   m.element(0,0) = counts[0];
<a name="l01043"></a>01043   m.element(0,1) = counts[0];
<a name="l01044"></a>01044   m.element(1,0) = counts[0];
<a name="l01045"></a>01045   m.element(1,1) = counts[0];
<a name="l01046"></a>01046 
<a name="l01047"></a>01047   m.element(2,2) = counts[1];
<a name="l01048"></a>01048   m.element(2,3) = counts[1];
<a name="l01049"></a>01049   m.element(3,2) = counts[1];
<a name="l01050"></a>01050   m.element(3,3) = counts[1];
<a name="l01051"></a>01051 
<a name="l01052"></a>01052   m.element(4,4) = counts[2];
<a name="l01053"></a>01053   m.element(4,5) = counts[2];
<a name="l01054"></a>01054   m.element(5,4) = counts[2];
<a name="l01055"></a>01055   m.element(5,5) = counts[2];
<a name="l01056"></a>01056   <span class="keywordflow">return</span> m;
<a name="l01057"></a>01057 }
<a name="l01058"></a>01058 <span class="comment"></span>
<a name="l01059"></a>01059 <span class="comment">/** Reformat variance/covariance vector as a banded matrix. */</span>
<a name="l01060"></a><a class="code" href="classQuantBRLMM.html#a6a5480a106596136e420508b49b3b370">01060</a> Matrix <a class="code" href="classQuantBRLMM.html#a6a5480a106596136e420508b49b3b370" title="Reformat variance/covariance vector as a banded matrix.">QuantBRLMM::formatAsBigVarMat</a>(ColumnVector v) {
<a name="l01061"></a>01061   Matrix m(6, 6);
<a name="l01062"></a>01062   m = 0;
<a name="l01063"></a>01063   <span class="keywordtype">int</span> nRows = v.Nrows();
<a name="l01064"></a>01064   assert(nRows == 9);
<a name="l01065"></a>01065   <span class="comment">// Diagonal are m and a variances.</span>
<a name="l01066"></a>01066   m.element(0,0) = v.element(0);
<a name="l01067"></a>01067   m.element(1,1) = v.element(1);
<a name="l01068"></a>01068   m.element(2,2) = v.element(3);
<a name="l01069"></a>01069   m.element(3,3) = v.element(4);
<a name="l01070"></a>01070   m.element(4,4) = v.element(6);
<a name="l01071"></a>01071   m.element(5,5) = v.element(7);
<a name="l01072"></a>01072   <span class="comment">// Fill in the covariances appropriately</span>
<a name="l01073"></a>01073   m.element(0,1) = m.element(1,0) = v.element(2);
<a name="l01074"></a>01074   m.element(2,3) = m.element(3,2) = v.element(5);
<a name="l01075"></a>01075   m.element(4,5) = m.element(5,4) = v.element(8);<span class="comment"></span>
<a name="l01076"></a>01076 <span class="comment">  /// @todo what about other points like 1,3 or 1,5?</span>
<a name="l01077"></a>01077 <span class="comment"></span>  <span class="keywordflow">return</span> m;
<a name="l01078"></a>01078 }
<a name="l01079"></a>01079 <span class="comment"></span>
<a name="l01080"></a>01080 <span class="comment">/** Bayesian update of the mean vector, this is where the voodoo</span>
<a name="l01081"></a>01081 <span class="comment">    happens. Have to query Simon more about the motivation. */</span>
<a name="l01082"></a><a class="code" href="classQuantBRLMM.html#acf88e5de22303b2cfc1a929433875f14">01082</a> Matrix <a class="code" href="classQuantBRLMM.html#acf88e5de22303b2cfc1a929433875f14" title="Bayesian update of the mean vector, this is where the voodoo happens.">QuantBRLMM::bayesianMeanUpdate</a>(ColumnVector &amp;newVar, std::vector&lt;int&gt; &amp;counts, ColumnVector &amp;means,
<a name="l01083"></a>01083                                              Matrix &amp;priorCenterVar, ColumnVector &amp;priorCenter) {
<a name="l01084"></a>01084   Matrix sigmaInv = <a class="code" href="classQuantBRLMM.html#a6a5480a106596136e420508b49b3b370" title="Reformat variance/covariance vector as a banded matrix.">QuantBRLMM::formatAsBigVarMat</a>(newVar);
<a name="l01085"></a>01085   sigmaInv = sigmaInv.i();
<a name="l01086"></a>01086   Matrix countMatrix = <a class="code" href="classQuantBRLMM.html#a76dda9af539b10df3326cf43c113b147" title="Make banded count matrix.">QuantBRLMM::makeCountMat</a>(counts);
<a name="l01087"></a>01087   Matrix centerVarInv = priorCenterVar.i();
<a name="l01088"></a>01088   Matrix elementMult = elementMultiplication(countMatrix, sigmaInv);
<a name="l01089"></a>01089   Matrix t1 = centerVarInv + elementMult;
<a name="l01090"></a>01090   t1 = t1.i();
<a name="l01091"></a>01091   Matrix t2 = centerVarInv * priorCenter;
<a name="l01092"></a>01092   Matrix temp = elementMult * means;
<a name="l01093"></a>01093   t2 = t2 + temp;
<a name="l01094"></a>01094   Matrix newMean = t1 * t2;
<a name="l01095"></a>01095   <span class="keywordflow">return</span> newMean;
<a name="l01096"></a>01096 }
<a name="l01097"></a>01097 <span class="comment"></span>
<a name="l01098"></a>01098 <span class="comment">/// @todo should degFree be weight or counts rather than degrees of freedom?</span>
<a name="l01099"></a><a class="code" href="classQuantBRLMM.html#a7ea1bd4cc10263c529ca483800fb3e09">01099</a> <span class="comment"></span>ColumnVector <a class="code" href="classQuantBRLMM.html#a7ea1bd4cc10263c529ca483800fb3e09">QuantBRLMM::shrinkVariance</a>(<a class="code" href="classClusterStats.html">ClusterStats</a> &amp;empirical, <a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior, <span class="keywordtype">int</span> degFree) {
<a name="l01100"></a>01100   ColumnVector newVar;
<a name="l01101"></a>01101   newVar.ReSize(empirical.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>.size());<span class="comment"></span>
<a name="l01102"></a>01102 <span class="comment">  /// @todo improve handling of NA rather than setting to zero</span>
<a name="l01103"></a>01103 <span class="comment"></span>  <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; empirical.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>.size(); i++) {
<a name="l01104"></a>01104     <span class="keywordtype">int</span> counts = empirical.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[i/3];
<a name="l01105"></a>01105     <span class="comment">/* empirical variance gets set to zero if there weren&#39;t enough</span>
<a name="l01106"></a>01106 <span class="comment">       observations for variance estimations (i.e. less than 2). If we</span>
<a name="l01107"></a>01107 <span class="comment">       don&#39;t have enough observations to compute a variance set the counts</span>
<a name="l01108"></a>01108 <span class="comment">       to zero so that we just use the prior estimate of variance. */</span>
<a name="l01109"></a>01109     <span class="keywordflow">if</span> (counts &lt; 2) {
<a name="l01110"></a>01110       counts = 0;
<a name="l01111"></a>01111     }
<a name="l01112"></a>01112     <span class="keywordtype">int</span> totalCount = counts + degFree;
<a name="l01113"></a>01113     <span class="keywordtype">double</span> result = ((counts * empirical.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>[i]) +
<a name="l01114"></a>01114                     (degFree * prior.centerVars.element(i)))/totalCount;
<a name="l01115"></a>01115     newVar.element(i) = result;
<a name="l01116"></a>01116   }
<a name="l01117"></a>01117   <span class="keywordflow">return</span> newVar;
<a name="l01118"></a>01118 }
<a name="l01119"></a>01119 <span class="comment"></span>
<a name="l01120"></a>01120 <span class="comment">/** Wrapper function for getting bayesian adjusted parameters given data. */</span>
<a name="l01121"></a><a class="code" href="classQuantBRLMM.html#af1acb3c6affa478feb02b2443d7d4962">01121</a> <a class="code" href="classClusterModel.html">ClusterModel</a> <a class="code" href="classQuantBRLMM.html#af1acb3c6affa478feb02b2443d7d4962" title="Wrapper function for getting bayesian adjusted parameters given data.">QuantBRLMM::fitModel</a>(std::vector&lt;double&gt; &amp;aValues,
<a name="l01122"></a>01122                                   std::vector&lt;double&gt; &amp;bValues,
<a name="l01123"></a>01123                                   std::vector&lt;std::vector&lt;double&gt; &gt; &amp;genoDist,
<a name="l01124"></a>01124                                   std::vector&lt;affx::GType&gt; &amp;genoTypes,
<a name="l01125"></a>01125                                   <a class="code" href="classClusterPrior.html">ClusterPrior</a> &amp;prior, <span class="keyword">const</span> std::string &amp;name,
<a name="l01126"></a>01126                                   <a class="code" href="classQuantBRLMM_1_1BrlmmParam.html">BrlmmParam</a> &amp;param) {
<a name="l01127"></a>01127 
<a name="l01128"></a>01128   <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">Transformation</a> transform = param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a928f99fd333fed0c34013ce8d23acb49" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>;
<a name="l01129"></a>01129   <span class="keywordtype">int</span> iterations = param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#adf7bcd06c19218d14fe7d49b79b3c7b2" title="How many iterations to do of BRLMM, i.e. boostrapping from initial calls.">m_Iterations</a>;
<a name="l01130"></a>01130   <span class="keywordtype">double</span> iterationThreshold = param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#ad9919e718c6d119d94f2b824abce8d53" title="What is our threshold (maximum value) for confidences befor /// calling &amp;#39;NN&amp;#39; in iterative mod...">m_IterationThreshold</a>;
<a name="l01131"></a>01131   <span class="keywordtype">double</span> K = param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a42c05834641c32011c0f5b6eac7ffe97" title="Scaling Parameter to use in transformations.">m_K</a>;
<a name="l01132"></a>01132 
<a name="l01133"></a>01133   <a class="code" href="classClusterStats.html">ClusterStats</a> stats = <a class="code" href="classQuantBRLMM.html#ad80f38b0695ffc08aef491b000deacb3" title="Count up some elementary statistics about a particular snp over multiple experiments.">QuantBRLMM::computeClusterStats</a>(aValues, bValues, genoTypes, transform, K );
<a name="l01134"></a>01134 
<a name="l01135"></a>01135   <a class="code" href="classClusterModel.html">ClusterModel</a> model;
<a name="l01136"></a>01136   ColumnVector newVar = <a class="code" href="classQuantBRLMM.html#a7ea1bd4cc10263c529ca483800fb3e09">shrinkVariance</a>(stats, prior, param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a3d250342e8ec26cdfe1f162274d7bfab" title="How much weight should the prior get?">m_PriorPseudoCount</a>);
<a name="l01137"></a>01137   model.clusterVars = newVar;
<a name="l01138"></a>01138   ColumnVector means = <a class="code" href="classQuantBRLMM.html#a5205c4974d362c640198f75c0547982d" title="Conversion utility for transforming vectors to column vectors.">columnVectorFromVector</a>(stats.<a class="code" href="classClusterStats.html#aa4a1a793033967c11e9abaa3ecd0cc3e" title="Simon style means object, not sure why indexed like this but it /// goes: /// mMeans[0],aMeans[0],mMeans[1],aMeans[1],etc..">means</a>);
<a name="l01139"></a>01139   model.clusterMeans = <a class="code" href="classQuantBRLMM.html#acf88e5de22303b2cfc1a929433875f14" title="Bayesian update of the mean vector, this is where the voodoo happens.">QuantBRLMM::bayesianMeanUpdate</a>(newVar, stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>, means, prior.covars, prior.centers);
<a name="l01140"></a>01140   <span class="comment">/* If iterations are greater than zero then we are going to recall</span>
<a name="l01141"></a>01141 <span class="comment">     using current model and relearn a new model using those calls</span>
<a name="l01142"></a>01142 <span class="comment">     instead of original calls. */</span>
<a name="l01143"></a>01143   <span class="keywordflow">while</span> (iterations &gt; 0) {
<a name="l01144"></a>01144     iterations--;
<a name="l01145"></a>01145     vector&lt;affx::GType&gt; newCalls;  <span class="comment">// brlmm calls for current iteration.</span>
<a name="l01146"></a>01146     vector&lt;double&gt; newConfidences; <span class="comment">// brlmm confidences for current iteration.</span>
<a name="l01147"></a>01147     <span class="comment">/* Make initial call using current model. */</span>
<a name="l01148"></a>01148     <a class="code" href="classQuantBRLMM.html#ad0808c8bfeef3f05b4733ea08a91cd0d">brlmmCall</a>(aValues, bValues, model, newCalls, newConfidences, genoDist, param);
<a name="l01149"></a>01149 
<a name="l01150"></a>01150     <span class="comment">/* Set calls that are less than our threshold to NN */</span>
<a name="l01151"></a>01151     assert(newCalls.size() == newConfidences.size());
<a name="l01152"></a>01152     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; newCalls.size(); i++) {
<a name="l01153"></a>01153       <span class="keywordflow">if</span> (newConfidences[i] &gt; iterationThreshold) {
<a name="l01154"></a>01154         newCalls[i] = NN;
<a name="l01155"></a>01155       }
<a name="l01156"></a>01156     }
<a name="l01157"></a>01157     <span class="comment">/* Compute a new model with updated calls. */</span>
<a name="l01158"></a>01158     <a class="code" href="classClusterStats.html">ClusterStats</a> stats = <a class="code" href="classQuantBRLMM.html#ad80f38b0695ffc08aef491b000deacb3" title="Count up some elementary statistics about a particular snp over multiple experiments.">computeClusterStats</a>(aValues, bValues, newCalls, transform, K);
<a name="l01159"></a>01159     ColumnVector newVar = <a class="code" href="classQuantBRLMM.html#a7ea1bd4cc10263c529ca483800fb3e09">shrinkVariance</a>(stats, prior, param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a3d250342e8ec26cdfe1f162274d7bfab" title="How much weight should the prior get?">m_PriorPseudoCount</a>);
<a name="l01160"></a>01160     model.clusterVars = newVar;
<a name="l01161"></a>01161     ColumnVector means = <a class="code" href="classQuantBRLMM.html#a5205c4974d362c640198f75c0547982d" title="Conversion utility for transforming vectors to column vectors.">columnVectorFromVector</a>(stats.<a class="code" href="classClusterStats.html#aa4a1a793033967c11e9abaa3ecd0cc3e" title="Simon style means object, not sure why indexed like this but it /// goes: /// mMeans[0],aMeans[0],mMeans[1],aMeans[1],etc..">means</a>);
<a name="l01162"></a>01162     model.clusterMeans = <a class="code" href="classQuantBRLMM.html#acf88e5de22303b2cfc1a929433875f14" title="Bayesian update of the mean vector, this is where the voodoo happens.">QuantBRLMM::bayesianMeanUpdate</a>(newVar, stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>, means, prior.covars, prior.centers);
<a name="l01163"></a>01163   }
<a name="l01164"></a>01164   <span class="keywordflow">return</span> model;
<a name="l01165"></a>01165 }
<a name="l01166"></a>01166 
<a name="l01167"></a>01167 <span class="keywordtype">void</span> QuantBRLMM::valueForTransformation(<span class="keywordtype">double</span> aVal, <span class="keywordtype">double</span> bVal, <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> trans,
<a name="l01168"></a>01168                                          <span class="keywordtype">double</span> K, <span class="keywordtype">double</span> &amp;xVal, <span class="keywordtype">double</span> &amp;yVal) {
<a name="l01169"></a>01169   <span class="keywordtype">double</span> c = 0.001;
<a name="l01170"></a>01170   <span class="keywordtype">double</span> denom = 0;
<a name="l01171"></a>01171   <span class="keywordtype">double</span> aV = 0.0;
<a name="l01172"></a>01172   <span class="keywordtype">double</span> bV = 0.0;
<a name="l01173"></a>01173   <span class="keywordtype">double</span> r = 0.0;
<a name="l01174"></a>01174   <span class="keywordflow">switch</span> (trans) {
<a name="l01175"></a>01175   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aaf0dc6d5b83fda75ca175ce28504710c" title="Minus Vs Average.">MvA</a>:
<a name="l01176"></a>01176     <span class="keywordflow">if</span> (aVal == 0) aVal = c;
<a name="l01177"></a>01177     <span class="keywordflow">if</span> (bVal == 0) bVal = c;
<a name="l01178"></a>01178     xVal = (<a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a>(bVal) + <a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a>(aVal)) / 2;
<a name="l01179"></a>01179     yVal = <a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a>(bVal) - <a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a>(aVal);
<a name="l01180"></a>01180     <span class="keywordflow">break</span>;
<a name="l01181"></a>01181   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aea687c012cf0ba203c45713fd99a9673" title="R vs Theta.">RvT</a>:
<a name="l01182"></a>01182     <span class="comment">// Check to make sure we&#39;re not got to take log of 0</span>
<a name="l01183"></a>01183     denom = bVal;
<a name="l01184"></a>01184     <span class="keywordflow">if</span> (bVal == 0) bVal = c;
<a name="l01185"></a>01185     xVal = atan(aVal/ denom);
<a name="l01186"></a>01186     yVal = log( sqrt((aVal * aVal) + (bVal * bVal) + c) );
<a name="l01187"></a>01187     <span class="keywordflow">break</span>;
<a name="l01188"></a>01188   <span class="keywordflow">case</span>  <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2729e1dbbb52465172e689eada41e72a" title="Contrast Extremes Stretch (previously known as south san francisco).">CES</a>:
<a name="l01189"></a>01189     <span class="comment">// Check to make sure we&#39;re not got to take log of 0</span>
<a name="l01190"></a>01190     denom = bVal + aVal;
<a name="l01191"></a>01191     <span class="keywordflow">if</span> (denom == 0) denom = c;
<a name="l01192"></a>01192     xVal = sinh( K * (aVal - bVal)/denom) / sinh(K);
<a name="l01193"></a>01193     yVal = <a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a>(denom);
<a name="l01194"></a>01194     <span class="keywordflow">break</span>;
<a name="l01195"></a>01195   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2dea69397eccbeb17925a9d8361f9fb8" title="Contrast Centers Stretch (previously known as alternative south san francisco).">CCS</a>:
<a name="l01196"></a>01196     denom = bVal + aVal;
<a name="l01197"></a>01197     <span class="keywordflow">if</span> (denom == 0) denom = c;
<a name="l01198"></a>01198     aV = aVal;
<a name="l01199"></a>01199     <span class="keywordflow">if</span> (aV == 0) aV = c;
<a name="l01200"></a>01200     bV = bVal;
<a name="l01201"></a>01201     <span class="keywordflow">if</span> (bV == 0) bV = c;
<a name="l01202"></a>01202     r = K * (aV - bV)/denom;
<a name="l01203"></a>01203     r = log(r + sqrt(r*r+1));
<a name="l01204"></a>01204     xVal = r / log(K + sqrt(K*K+1));
<a name="l01205"></a>01205     <span class="comment">//r = K * (aV - bV)/denom;</span>
<a name="l01206"></a>01206     <span class="comment">// r = log(r + sqrt(r*r+1));</span>
<a name="l01207"></a>01207     <span class="comment">//    xVal = (log(r + sqrt(r*r+1))) / log(K + sqrt(K*K+1));</span>
<a name="l01208"></a>01208     <span class="comment">// Note that we are using natural logarithm here (log_e), not log2</span>
<a name="l01209"></a>01209     <span class="comment">// xVal = asinh(K * tanh( .5 * (log(aV) - log(bV)))) / asinh(K);</span>
<a name="l01210"></a>01210     yVal = <a class="code" href="Util_8h.html#a18af743c2cec4baeee9ffb27999ddaad" title="Log base two.">log2</a>(denom);
<a name="l01211"></a>01211     <span class="keywordflow">break</span>;
<a name="l01212"></a>01212   <span class="keywordflow">default</span>:
<a name="l01213"></a>01213     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::fillInTransformedData() - Don&#39;t recognize transformation type: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(trans));
<a name="l01214"></a>01214   }
<a name="l01215"></a>01215 }
<a name="l01216"></a>01216 <span class="comment"></span>
<a name="l01217"></a>01217 <span class="comment">/**</span>
<a name="l01218"></a>01218 <span class="comment"> * Transform the data in aValues and bValues using the method specified by trans.</span>
<a name="l01219"></a>01219 <span class="comment"> *</span>
<a name="l01220"></a>01220 <span class="comment"> * @param aValues - Summary values for A allele.</span>
<a name="l01221"></a>01221 <span class="comment"> * @param bValues - Summary values for B allele.</span>
<a name="l01222"></a>01222 <span class="comment"> * @param genoTypes - Genotypes for each sample.</span>
<a name="l01223"></a>01223 <span class="comment"> * @param Transformation - Type of transformation to use.</span>
<a name="l01224"></a>01224 <span class="comment"> * @param xDim - two dimensional matrix for x dimension with rows</span>
<a name="l01225"></a>01225 <span class="comment"> * (index 0) being genotype (0,1,2) and columns being values observed.</span>
<a name="l01226"></a>01226 <span class="comment"> * @param yDim - two dimensional matrix for y dimension with rows</span>
<a name="l01227"></a>01227 <span class="comment"> * (index 0) being genotype (0,1,2) and columns being values observed.</span>
<a name="l01228"></a>01228 <span class="comment"> */</span>
<a name="l01229"></a><a class="code" href="classQuantBRLMM.html#aae609c657d28f62ae7b923c73a93fc46">01229</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#aae609c657d28f62ae7b923c73a93fc46" title="Transform the data in aValues and bValues using the method specified by trans.">QuantBRLMM::fillInTransformedData</a>(std::vector&lt;double&gt; &amp;aValues,
<a name="l01230"></a>01230                                        std::vector&lt;double&gt; &amp;bValues,
<a name="l01231"></a>01231                                        std::vector&lt;affx::GType&gt; &amp;genoTypes,
<a name="l01232"></a>01232                                        <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> trans,
<a name="l01233"></a>01233                                        <span class="keywordtype">double</span> K,
<a name="l01234"></a>01234                                        std::vector&lt;std::vector&lt;double&gt; &gt; &amp;xDim,
<a name="l01235"></a>01235                                        std::vector&lt;std::vector&lt;double&gt; &gt; &amp;yDim) {
<a name="l01236"></a>01236   <span class="comment">/* Sanity check. */</span>
<a name="l01237"></a>01237   assert(aValues.size() == bValues.size());
<a name="l01238"></a>01238   assert(aValues.size() == genoTypes.size());
<a name="l01239"></a>01239 
<a name="l01240"></a>01240   <span class="comment">/* Clean out any previous results. */</span>
<a name="l01241"></a>01241   xDim.clear();
<a name="l01242"></a>01242   yDim.clear();
<a name="l01243"></a>01243 
<a name="l01244"></a>01244   xDim.resize(3); <span class="comment">// Number of genotypes 0, 1, 2</span>
<a name="l01245"></a>01245   yDim.resize(3); <span class="comment">// Number of genotypes 0, 1, 2</span>
<a name="l01246"></a>01246   <span class="comment">//  double xVal = 0, yVal = 0;</span>
<a name="l01247"></a>01247   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; aValues.size(); i++) {
<a name="l01248"></a>01248     <span class="keywordflow">if</span> (genoTypes[i] == <a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a1cef2fc4697bd5fc67116fbd1ee2027b" title="No call, unknown genotype. The internal is value is &amp;#39;3&amp;#39;; external is &amp;#39;-1&amp;#39;,.">NN</a>)
<a name="l01249"></a>01249       <span class="keywordflow">continue</span>;
<a name="l01250"></a>01250     <span class="comment">//    valueForTransformation(aValues[i], bValues[i], trans, K, xVal, yVal);</span>
<a name="l01251"></a>01251     xDim[genoTypes[i]].push_back( aValues[i] );
<a name="l01252"></a>01252     yDim[genoTypes[i]].push_back( bValues[i] );
<a name="l01253"></a>01253   }
<a name="l01254"></a>01254 }
<a name="l01255"></a>01255 <span class="comment"></span>
<a name="l01256"></a>01256 <span class="comment">/** Count up some elementary statistics about a particular snp over</span>
<a name="l01257"></a>01257 <span class="comment">    multiple experiments. */</span>
<a name="l01258"></a><a class="code" href="classQuantBRLMM.html#ad80f38b0695ffc08aef491b000deacb3">01258</a> <a class="code" href="classClusterStats.html">ClusterStats</a> <a class="code" href="classQuantBRLMM.html#ad80f38b0695ffc08aef491b000deacb3" title="Count up some elementary statistics about a particular snp over multiple experiments.">QuantBRLMM::computeClusterStats</a>(std::vector&lt;double&gt; &amp;aValues,
<a name="l01259"></a>01259                                              std::vector&lt;double&gt; &amp;bValues,
<a name="l01260"></a>01260                                              std::vector&lt;affx::GType&gt; &amp;genoTypes,
<a name="l01261"></a>01261                                              <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> trans,
<a name="l01262"></a>01262                                              <span class="keywordtype">double</span> K) {
<a name="l01263"></a>01263   <span class="comment">// Sanity checks</span>
<a name="l01264"></a>01264   assert(aValues.size() &gt; 0);
<a name="l01265"></a>01265   <a class="code" href="classErr.html#ad1a633f602fe60689cdc12d9ca1842e9" title="errAbort based version of assert with an error message.">Err::check</a>(aValues.size() == bValues.size(), <span class="stringliteral">&quot;computClusterStats() - Expression should be same size.&quot;</span>);
<a name="l01266"></a>01266   <a class="code" href="classErr.html#ad1a633f602fe60689cdc12d9ca1842e9" title="errAbort based version of assert with an error message.">Err::check</a>(bValues.size() == genoTypes.size(), <span class="stringliteral">&quot;computClusterStats() - Expression should be same size as genotyping.&quot;</span>);
<a name="l01267"></a>01267   <span class="comment">// Indexes are AA = 0, AB=1, BB=2 dType = &quot;difference&quot; bType =</span>
<a name="l01268"></a>01268   <span class="comment">// &quot;intensity&quot; Normally these would be M &amp; A (like MA plots), but</span>
<a name="l01269"></a>01269   <span class="comment">// A gets confusing with genotype &#39;A&#39;</span>
<a name="l01270"></a>01270   vector&lt;vector &lt;double&gt; &gt; dTypeValues(3);
<a name="l01271"></a>01271   vector&lt;vector &lt;double&gt; &gt; iTypeValues(3);
<a name="l01272"></a>01272   <span class="comment">//  bool allGood = true;</span>
<a name="l01273"></a>01273   <a class="code" href="classClusterStats.html">ClusterStats</a> stats;
<a name="l01274"></a>01274   <span class="comment">// Loop through and group into genotype specific data.</span>
<a name="l01275"></a>01275   <a class="code" href="classQuantBRLMM.html#aae609c657d28f62ae7b923c73a93fc46" title="Transform the data in aValues and bValues using the method specified by trans.">fillInTransformedData</a>(aValues, bValues, genoTypes, trans, K, iTypeValues, dTypeValues);
<a name="l01276"></a>01276 
<a name="l01277"></a>01277   <span class="comment">// Compute the means of the clusters for each genotype. Note that if</span>
<a name="l01278"></a>01278   <span class="comment">// we don&#39;t have enough samples to compute the variance it is set to</span>
<a name="l01279"></a>01279   <span class="comment">// zero as a placeholder, but it is necessary to check the counts</span>
<a name="l01280"></a>01280   <span class="comment">// later to make sure we&#39;re not using an invalid number.</span>
<a name="l01281"></a>01281   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; 3; i++) {
<a name="l01282"></a>01282     stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[i] = iTypeValues[i].size();
<a name="l01283"></a>01283     <span class="keywordflow">if</span> (stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[i] &gt; 0) {
<a name="l01284"></a>01284       stats.<a class="code" href="classClusterStats.html#aeb095e22b8004e7aecec5491ee795eb3" title="Mean M (intensity) of each genotype (center of the clusters);.">mMeans</a>[i] = <a class="code" href="classAverage.html">Average&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(iTypeValues[i].begin(), iTypeValues[i].end());
<a name="l01285"></a>01285       stats.<a class="code" href="classClusterStats.html#a4cc2e8d89e81745f3904277d685f48b5" title="Mean A (difference) of each genotype (center of the clusters);.">aMeans</a>[i] = <a class="code" href="classAverage.html">Average&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(dTypeValues[i].begin(), dTypeValues[i].end());
<a name="l01286"></a>01286       <span class="keywordflow">if</span> (stats.<a class="code" href="classClusterStats.html#a0630d56a98160ba8ae9d27606f9c0cd3" title="How many of each type of SNP genotype were observed.">counts</a>[i] &gt; 1) {
<a name="l01287"></a>01287         stats.mVars[i] = <a class="code" href="classUnbiasedVariance.html">UnbiasedVariance&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(iTypeValues[i].begin(), iTypeValues[i].end());
<a name="l01288"></a>01288         stats.aVars[i] = <a class="code" href="classUnbiasedVariance.html">UnbiasedVariance&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(dTypeValues[i].begin(), dTypeValues[i].end());
<a name="l01289"></a>01289         stats.covars[i] = <a class="code" href="classUnbiasedCovariance.html">UnbiasedCovariance&lt;vector&lt;double&gt;::iterator</a>, <span class="keywordtype">double</span>&gt;()(dTypeValues[i].begin(), dTypeValues[i].end(), iTypeValues[i].begin());
<a name="l01290"></a>01290       }
<a name="l01291"></a>01291       <span class="keywordflow">else</span> {
<a name="l01292"></a>01292         stats.mVars[i] = 0;
<a name="l01293"></a>01293         stats.aVars[i] = 0;
<a name="l01294"></a>01294         stats.covars[i] = 0;
<a name="l01295"></a>01295       }
<a name="l01296"></a>01296     }
<a name="l01297"></a>01297     <span class="keywordflow">else</span> {
<a name="l01298"></a>01298       stats.<a class="code" href="classClusterStats.html#aeb095e22b8004e7aecec5491ee795eb3" title="Mean M (intensity) of each genotype (center of the clusters);.">mMeans</a>[i] = 0;
<a name="l01299"></a>01299       stats.mVars[i] = 0;
<a name="l01300"></a>01300       stats.<a class="code" href="classClusterStats.html#a4cc2e8d89e81745f3904277d685f48b5" title="Mean A (difference) of each genotype (center of the clusters);.">aMeans</a>[i] = 0;
<a name="l01301"></a>01301       stats.aVars[i] = 0;
<a name="l01302"></a>01302       stats.covars[i] = 0;
<a name="l01303"></a>01303     }
<a name="l01304"></a>01304     stats.<a class="code" href="classClusterStats.html#aa4a1a793033967c11e9abaa3ecd0cc3e" title="Simon style means object, not sure why indexed like this but it /// goes: /// mMeans[0],aMeans[0],mMeans[1],aMeans[1],etc..">means</a>.push_back(stats.<a class="code" href="classClusterStats.html#aeb095e22b8004e7aecec5491ee795eb3" title="Mean M (intensity) of each genotype (center of the clusters);.">mMeans</a>[i]);
<a name="l01305"></a>01305     stats.<a class="code" href="classClusterStats.html#aa4a1a793033967c11e9abaa3ecd0cc3e" title="Simon style means object, not sure why indexed like this but it /// goes: /// mMeans[0],aMeans[0],mMeans[1],aMeans[1],etc..">means</a>.push_back(stats.<a class="code" href="classClusterStats.html#a4cc2e8d89e81745f3904277d685f48b5" title="Mean A (difference) of each genotype (center of the clusters);.">aMeans</a>[i]);
<a name="l01306"></a>01306     stats.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>.push_back(stats.mVars[i]);
<a name="l01307"></a>01307     stats.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>.push_back(stats.aVars[i]);
<a name="l01308"></a>01308     stats.<a class="code" href="classClusterStats.html#a396289aa98315a86b16e8b1088a61f82" title="Simon style vars object, not sure why indexed like this but it /// goes: /// mVars[0],aVars[0],covars[0],mVars[1],aVars[1],covars[1],etc..">vars</a>.push_back(stats.covars[i]);
<a name="l01309"></a>01309   }
<a name="l01310"></a>01310   <span class="keywordflow">return</span> stats;
<a name="l01311"></a>01311 }
<a name="l01312"></a>01312 
<a name="l01313"></a>01313 <span class="comment">/* Make the inverse matrix from the variance and covariances. */</span>
<a name="l01314"></a>01314 Matrix QuantBRLMM::makeVarInv(Matrix &amp;clusterVars, <span class="keywordtype">int</span> index) {
<a name="l01315"></a>01315   Matrix var(2,2);
<a name="l01316"></a>01316   index = 3 * index;
<a name="l01317"></a>01317   var.element(0,0) = clusterVars.element(index,0);
<a name="l01318"></a>01318   var.element(1,1) = clusterVars.element(index+1,0);
<a name="l01319"></a>01319   var.element(0,1) = clusterVars.element(index+2,0);
<a name="l01320"></a>01320   var.element(1,0) = clusterVars.element(index+2,0);
<a name="l01321"></a>01321   <span class="keywordflow">return</span> var.i();
<a name="l01322"></a>01322 }
<a name="l01323"></a>01323 
<a name="l01324"></a>01324 <span class="comment">/* Take the model derived in fitModel and use it to do predictions. */</span>
<a name="l01325"></a><a class="code" href="classQuantBRLMM.html#ad0808c8bfeef3f05b4733ea08a91cd0d">01325</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBRLMM.html#ad0808c8bfeef3f05b4733ea08a91cd0d">QuantBRLMM::brlmmCall</a>(<span class="keyword">const</span> std::vector&lt;double&gt; &amp;aValues,
<a name="l01326"></a>01326                            <span class="keyword">const</span> std::vector&lt;double&gt; &amp;bValues,
<a name="l01327"></a>01327                            <a class="code" href="classClusterModel.html">ClusterModel</a> &amp;model,
<a name="l01328"></a>01328                            std::vector&lt;affx::GType&gt; &amp;calls,
<a name="l01329"></a>01329                            std::vector&lt;double&gt; &amp;confidences,
<a name="l01330"></a>01330                            std::vector&lt;std::vector&lt;double&gt; &gt; &amp;genoDist,
<a name="l01331"></a>01331                            <a class="code" href="classQuantBRLMM_1_1BrlmmParam.html">BrlmmParam</a> &amp;param)
<a name="l01332"></a>01332 {
<a name="l01333"></a>01333   <span class="comment">//</span>
<a name="l01334"></a>01334   assert(aValues.size() &gt; 0);
<a name="l01335"></a>01335   assert(aValues.size() == bValues.size());
<a name="l01336"></a>01336   <span class="comment">// JHG - debugging</span>
<a name="l01337"></a>01337   <span class="comment">// printVec(&quot;aValues&quot;,aValues);</span>
<a name="l01338"></a>01338   <span class="comment">// printVec(&quot;aValues&quot;,bValues);</span>
<a name="l01339"></a>01339   <span class="comment">//</span>
<a name="l01340"></a>01340   <span class="keywordtype">double</span> hetMult = param.<a class="code" href="classQuantBRLMM_1_1BrlmmParam.html#a474238d1f98a341c21dc9c426e79036d" title="Number to balance het calls with to balance performance on het/hom calls.">m_HetMultiplier</a>;
<a name="l01341"></a>01341   Matrix obs(aValues.size(),2); <span class="comment">// Our observed means in both M and A dimensions.</span>
<a name="l01342"></a>01342   Matrix distance(aValues.size(),3); <span class="comment">// Our distances to clusters.</span>
<a name="l01343"></a>01343   calls.clear();
<a name="l01344"></a>01344   confidences.clear();
<a name="l01345"></a>01345 
<a name="l01346"></a>01346   <span class="comment">/* Load up the observed means into matrix. */</span>
<a name="l01347"></a>01347   <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; i &lt; aValues.size(); i++) {
<a name="l01348"></a>01348     obs.element(i, 0) = aValues[i];
<a name="l01349"></a>01349     obs.element(i, 1) = bValues[i];
<a name="l01350"></a>01350   }
<a name="l01351"></a>01351 
<a name="l01352"></a>01352   <span class="comment">/* Calculate the distances for each genotype. */</span>
<a name="l01353"></a>01353   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> colIx = 0; colIx &lt; distance.Ncols(); colIx++) {
<a name="l01354"></a>01354     Matrix varInv = QuantBRLMM::makeVarInv(model.clusterVars, colIx);
<a name="l01355"></a>01355     <span class="comment">// I&#39;m thinking these are the covariances for the mahalanobis distance</span>
<a name="l01356"></a>01356     <span class="keywordtype">double</span> t1 = varInv.element(0,0);
<a name="l01357"></a>01357     <span class="keywordtype">double</span> t2 = varInv.element(0,1) + varInv.element(1,0); <span class="comment">/// @todo why is this additive?</span>
<a name="l01358"></a>01358 <span class="comment"></span>    <span class="keywordtype">double</span> t3 = varInv.element(1,1);
<a name="l01359"></a>01359 
<a name="l01360"></a>01360     <span class="comment">/* Calculate the distance for different samples for this genotype. */</span>
<a name="l01361"></a>01361     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> rowIx = 0; rowIx &lt; distance.Nrows(); rowIx++) {
<a name="l01362"></a>01362       <span class="keywordtype">double</span> iDist =  obs.element(rowIx,0) - model.clusterMeans.element(colIx * 2, 0);
<a name="l01363"></a>01363       <span class="keywordtype">double</span> dDist =  obs.element(rowIx,1) - model.clusterMeans.element(colIx * 2 + 1, 0);
<a name="l01364"></a>01364       <span class="keywordtype">double</span> dist = iDist*iDist*t1 + iDist*dDist*t2 + dDist*dDist*t3;
<a name="l01365"></a>01365       <span class="comment">/* If we are calculating distance for AB genotype (column 1) then multiply by our</span>
<a name="l01366"></a>01366 <span class="comment">         het multiplier as a way to balance het/hom accurancy. */</span>
<a name="l01367"></a>01367       <span class="keywordflow">if</span> (colIx == 1) {
<a name="l01368"></a>01368         dist = dist * hetMult;
<a name="l01369"></a>01369       }
<a name="l01370"></a>01370       distance.element(rowIx, colIx) = dist;
<a name="l01371"></a>01371     }
<a name="l01372"></a>01372   }
<a name="l01373"></a>01373   genoDist.resize(distance.Nrows());
<a name="l01374"></a>01374 
<a name="l01375"></a>01375   <span class="comment">/* for each sample predict based on which cluster is closest. */</span>
<a name="l01376"></a>01376   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> rowIx = 0; rowIx &lt; distance.Nrows(); rowIx++) {
<a name="l01377"></a>01377     <span class="keywordtype">double</span> minDistance = FLT_MAX;
<a name="l01378"></a>01378     <span class="keywordtype">int</span> minDistIx = -1;
<a name="l01379"></a>01379     <span class="keywordtype">double</span> secondMin = FLT_MAX;
<a name="l01380"></a>01380     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> colIx = 0; colIx &lt; distance.Ncols(); colIx++) {
<a name="l01381"></a>01381       <span class="keywordtype">double</span> dist = distance.element(rowIx, colIx);
<a name="l01382"></a>01382       genoDist[rowIx].push_back(dist);
<a name="l01383"></a>01383       <span class="comment">// JHG - debug</span>
<a name="l01384"></a>01384       <span class="comment">// printf(&quot;dist=%.10f\n&quot;,dist);</span>
<a name="l01385"></a>01385       <span class="keywordflow">if</span> (dist &lt; minDistance) {
<a name="l01386"></a>01386         secondMin = minDistance;
<a name="l01387"></a>01387         minDistance = dist;
<a name="l01388"></a>01388         minDistIx = colIx;
<a name="l01389"></a>01389       }
<a name="l01390"></a>01390       <span class="keywordflow">else</span> <span class="keywordflow">if</span> (dist &lt; secondMin) {
<a name="l01391"></a>01391         secondMin = dist;
<a name="l01392"></a>01392       }
<a name="l01393"></a>01393     }
<a name="l01394"></a>01394     calls.push_back((affx::GType)minDistIx);
<a name="l01395"></a>01395     confidences.push_back(minDistance/secondMin);
<a name="l01396"></a>01396   }
<a name="l01397"></a>01397 }
<a name="l01398"></a>01398 <span class="comment"></span>
<a name="l01399"></a>01399 <span class="comment">/**</span>
<a name="l01400"></a>01400 <span class="comment"> * @brief Default Getter method for parameters and their documentation.</span>
<a name="l01401"></a>01401 <span class="comment"> * @return map of parameters and their descriptions.</span>
<a name="l01402"></a>01402 <span class="comment"> */</span>
<a name="l01403"></a><a class="code" href="classQuantBRLMM.html#af4303a14cab2be76711893a4082e8c19">01403</a> std::vector&lt;SelfDoc::Opt&gt; <a class="code" href="classQuantBRLMM.html#af4303a14cab2be76711893a4082e8c19" title="Default Getter method for parameters and their documentation.">QuantBRLMM::getDefaultDocOptions</a>() {
<a name="l01404"></a>01404   std::vector&lt;SelfDoc::Opt&gt; opts;
<a name="l01405"></a>01405   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> hetMult = {<span class="stringliteral">&quot;het-mult&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>,
<a name="l01406"></a>01406                           <span class="stringliteral">&quot;Number to balance het calls with to balance performance on het/hom calls.&quot;</span>};
<a name="l01407"></a>01407   opts.push_back(hetMult);
<a name="l01408"></a>01408   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> iterations = {<span class="stringliteral">&quot;iterations&quot;</span>, SelfDoc::Opt::Integer, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01409"></a>01409                              <span class="stringliteral">&quot;Number of times to iterate BRLMM classifier, feeding in new calls from previous iteration.&quot;</span>};
<a name="l01410"></a>01410   opts.push_back(iterations);
<a name="l01411"></a>01411   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> iterThresh = {<span class="stringliteral">&quot;iter-thresh&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;0.3&quot;</span>, <span class="stringliteral">&quot;0.3&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;1&quot;</span>,
<a name="l01412"></a>01412                              <span class="stringliteral">&quot;Maximum confidence score to use when doing iterations internally [0,1].&quot;</span>};
<a name="l01413"></a>01413   opts.push_back(iterThresh);
<a name="l01414"></a>01414   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> k = {<span class="stringliteral">&quot;K&quot;</span>, SelfDoc::Opt::Double, <span class="stringliteral">&quot;4.0&quot;</span>, <span class="stringliteral">&quot;4.0&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01415"></a>01415                     <span class="stringliteral">&quot;Scale parameter used used in CCS and CES transformations.&quot;</span>};
<a name="l01416"></a>01416   opts.push_back(k);
<a name="l01417"></a>01417   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> transform = {<span class="stringliteral">&quot;transform&quot;</span>, SelfDoc::Opt::String, <span class="stringliteral">&quot;CCS&quot;</span>, <span class="stringliteral">&quot;CCS&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01418"></a>01418                             <span class="stringliteral">&quot;Transformation of initial data are we feeding into the classifier? {&#39;CCS&#39;, &#39;CES&#39;, &#39;MvA&#39;,&#39;RvT&#39;}&quot;</span>};
<a name="l01419"></a>01419   opts.push_back(transform);
<a name="l01420"></a>01420   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> priorWeight = {<span class="stringliteral">&quot;prior-weight&quot;</span>, SelfDoc::Opt::Integer, <span class="stringliteral">&quot;40&quot;</span>, <span class="stringliteral">&quot;40&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01421"></a>01421                               <span class="stringliteral">&quot;Psuedocount weight should the prior have? Also known as &#39;degrees of freedom&#39; in R code.&quot;</span>};
<a name="l01422"></a>01422   opts.push_back(priorWeight);
<a name="l01423"></a>01423   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> priorMinCall = {<span class="stringliteral">&quot;prior-mincall&quot;</span>, SelfDoc::Opt::Integer, <span class="stringliteral">&quot;2&quot;</span>, <span class="stringliteral">&quot;2&quot;</span>, <span class="stringliteral">&quot;2&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01424"></a>01424                                <span class="stringliteral">&quot;Minimum number of genotypes per cluster for inclusion in prior estimation, must be &gt;= 2.&quot;</span>};
<a name="l01425"></a>01425   opts.push_back(priorMinCall);
<a name="l01426"></a>01426   <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> lowprecision = {<span class="stringliteral">&quot;lowprecision&quot;</span>, SelfDoc::Opt::Boolean, <span class="stringliteral">&quot;false&quot;</span>, <span class="stringliteral">&quot;false&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>, <span class="stringliteral">&quot;NA&quot;</span>,
<a name="l01427"></a>01427                                <span class="stringliteral">&quot;R prototype uses summary values rounded to first decimal place. Use this flag to be simulate behavior.&quot;</span>};
<a name="l01428"></a>01428   opts.push_back(lowprecision);
<a name="l01429"></a>01429  <a class="code" href="classSelfDoc_1_1Opt.html" title="Description of one possible option/parameter for a self describing class.">SelfDoc::Opt</a> maxscore = {<span class="stringliteral">&quot;MS&quot;</span>,SelfDoc::Opt::Float, <span class="stringliteral">&quot;0.5&quot;</span>,<span class="stringliteral">&quot;0.5&quot;</span>, <span class="stringliteral">&quot;0&quot;</span>, <span class="stringliteral">&quot;2&quot;</span>,
<a name="l01430"></a>01430                                 <span class="stringliteral">&quot;Threshold for making no-calls&quot;</span>};
<a name="l01431"></a>01431         opts.push_back(maxscore);
<a name="l01432"></a>01432  <span class="keywordflow">return</span> opts;
<a name="l01433"></a>01433 }
<a name="l01434"></a>01434 <span class="comment"></span>
<a name="l01435"></a>01435 <span class="comment">/**</span>
<a name="l01436"></a>01436 <span class="comment"> * @brief Supply a little how/what/why about the algorithms this</span>
<a name="l01437"></a>01437 <span class="comment"> * class performs and what parameters it takes.</span>
<a name="l01438"></a>01438 <span class="comment"> * @return SelfDoc</span>
<a name="l01439"></a>01439 <span class="comment"> */</span>
<a name="l01440"></a><a class="code" href="classQuantBRLMM.html#a3632279e1824737d1d37242d2d25191e">01440</a> <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> <a class="code" href="classQuantBRLMM.html#a3632279e1824737d1d37242d2d25191e" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">QuantBRLMM::explainSelf</a>() {
<a name="l01441"></a>01441   <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc;
<a name="l01442"></a>01442   <a class="code" href="classQuantBRLMM.html#a53c159ab51bc8a716b63001410146e10" title="Fill in our self documentation from current state.">setupSelfDoc</a>(doc);
<a name="l01443"></a>01443   <span class="keywordflow">return</span> doc;
<a name="l01444"></a>01444 }
<a name="l01445"></a>01445 <span class="comment"></span>
<a name="l01446"></a>01446 <span class="comment">/**</span>
<a name="l01447"></a>01447 <span class="comment"> * Convert a (text insensitive) text repesentation of a transformation into</span>
<a name="l01448"></a>01448 <span class="comment"> * the enumration.</span>
<a name="l01449"></a>01449 <span class="comment"> * @param s - string specifying transformation.</span>
<a name="l01450"></a>01450 <span class="comment"> * @return - Enumeration version of transformation.</span>
<a name="l01451"></a>01451 <span class="comment"> */</span>
<a name="l01452"></a><a class="code" href="classQuantBRLMM.html#ac5880b4177534c9516f3d7243cb9aed7">01452</a> <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> <a class="code" href="classQuantBRLMM.html#ac5880b4177534c9516f3d7243cb9aed7" title="Convert a (text insensitive) text repesentation of a transformation into the enumration.">QuantBRLMM::transformationForString</a>(<span class="keyword">const</span> std::string&amp; str) {
<a name="l01453"></a>01453   std::string lowerS = <a class="code" href="classUtil.html#ab73e26f5a353284899216123a494e3bf" title="Copy and downcase a 8bit string. ///.">Util::downcaseString</a>(str);
<a name="l01454"></a>01454 
<a name="l01455"></a>01455   <span class="keywordflow">if</span> (lowerS == <span class="stringliteral">&quot;mva&quot;</span>)
<a name="l01456"></a>01456     <span class="keywordflow">return</span> MvA;
<a name="l01457"></a>01457   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (lowerS == <span class="stringliteral">&quot;rvt&quot;</span>)
<a name="l01458"></a>01458     <span class="keywordflow">return</span> RvT;
<a name="l01459"></a>01459   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (lowerS == <span class="stringliteral">&quot;ssf&quot;</span> || lowerS == <span class="stringliteral">&quot;ces&quot;</span>)
<a name="l01460"></a>01460     <span class="keywordflow">return</span> CES;
<a name="l01461"></a>01461   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (lowerS == <span class="stringliteral">&quot;assf&quot;</span> || lowerS == <span class="stringliteral">&quot;ccs&quot;</span>)
<a name="l01462"></a>01462     <span class="keywordflow">return</span> CCS;
<a name="l01463"></a>01463   <span class="keywordflow">else</span>
<a name="l01464"></a>01464     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::transformationForString() - Don&#39;t recognize transformation: &quot;</span> + str);
<a name="l01465"></a>01465   <span class="keywordflow">return</span> MvA;
<a name="l01466"></a>01466 }
<a name="l01467"></a>01467 <span class="comment"></span>
<a name="l01468"></a>01468 <span class="comment">/**</span>
<a name="l01469"></a>01469 <span class="comment"> * Convert a transformation enum to string representation</span>
<a name="l01470"></a>01470 <span class="comment"> * @param t - tranformation enum value.</span>
<a name="l01471"></a>01471 <span class="comment"> * @return - text description of enumeration.</span>
<a name="l01472"></a>01472 <span class="comment"> */</span>
<a name="l01473"></a><a class="code" href="classQuantBRLMM.html#a03d17022504aa06d976b555fbd6e8ea7">01473</a> std::string <a class="code" href="classQuantBRLMM.html#a03d17022504aa06d976b555fbd6e8ea7" title="Convert a transformation enum to string representation.">QuantBRLMM::stringForTransformation</a>(<span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> t) {
<a name="l01474"></a>01474   <span class="keywordflow">switch</span>(t) {
<a name="l01475"></a>01475   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aaf0dc6d5b83fda75ca175ce28504710c" title="Minus Vs Average.">MvA</a> :
<a name="l01476"></a>01476     <span class="keywordflow">return</span> <span class="stringliteral">&quot;MvA&quot;</span>;
<a name="l01477"></a>01477     <span class="keywordflow">break</span>;
<a name="l01478"></a>01478   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638aea687c012cf0ba203c45713fd99a9673" title="R vs Theta.">RvT</a> :
<a name="l01479"></a>01479     <span class="keywordflow">return</span> <span class="stringliteral">&quot;RvT&quot;</span>;
<a name="l01480"></a>01480     <span class="keywordflow">break</span>;
<a name="l01481"></a>01481   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2729e1dbbb52465172e689eada41e72a" title="Contrast Extremes Stretch (previously known as south san francisco).">CES</a> :
<a name="l01482"></a>01482     <span class="keywordflow">return</span> <span class="stringliteral">&quot;CES&quot;</span>;
<a name="l01483"></a>01483     <span class="keywordflow">break</span>;
<a name="l01484"></a>01484   <span class="keywordflow">case</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638a2dea69397eccbeb17925a9d8361f9fb8" title="Contrast Centers Stretch (previously known as alternative south san francisco).">CCS</a>:
<a name="l01485"></a>01485     <span class="keywordflow">return</span> <span class="stringliteral">&quot;CCS&quot;</span>;
<a name="l01486"></a>01486     <span class="keywordflow">break</span>;
<a name="l01487"></a>01487   <span class="keywordflow">default</span>:
<a name="l01488"></a>01488     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::stringForTransformation() - Don&#39;t recognize type: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(t));
<a name="l01489"></a>01489   }
<a name="l01490"></a>01490   <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBRLMM::stringForTransformation() - Should never reach this point.&quot;</span>);
<a name="l01491"></a>01491   <span class="keywordflow">return</span> <span class="stringliteral">&quot;&quot;</span>;
<a name="l01492"></a>01492 }
<a name="l01493"></a>01493 
<a name="l01494"></a>01494 <span class="comment"></span>
<a name="l01495"></a>01495 <span class="comment">/**</span>
<a name="l01496"></a>01496 <span class="comment"> * @brief This static function should be overridden by child classes</span>
<a name="l01497"></a>01497 <span class="comment"> * to return an object of the correct type initialized correctly</span>
<a name="l01498"></a>01498 <span class="comment"> * with the parameters in the string, string map. All objects</span>
<a name="l01499"></a>01499 <span class="comment"> * created this way should be deleted when finished using.</span>
<a name="l01500"></a>01500 <span class="comment"> *</span>
<a name="l01501"></a>01501 <span class="comment"> * @param param - Map of key/value pairs to initialize the object.</span>
<a name="l01502"></a>01502 <span class="comment"> *</span>
<a name="l01503"></a>01503 <span class="comment"> * @return Pointer toCreate object, this should be sub casted as necessary.</span>
<a name="l01504"></a>01504 <span class="comment"> */</span>
<a name="l01505"></a><a class="code" href="classQuantBRLMM.html#a26b9afc9130094403ebf4baff03a8a07">01505</a> <a class="code" href="classSelfCreate.html" title="Small interface for functions that know how to make an instance of themselves given a map of key...">SelfCreate</a> *<a class="code" href="classQuantBRLMM.html#a26b9afc9130094403ebf4baff03a8a07" title="This static function should be overridden by child classes to return an object of the correct type in...">QuantBRLMM::newObject</a>(std::map&lt;std::string,std::string&gt; &amp;param) {
<a name="l01506"></a>01506   <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> doc = <a class="code" href="classQuantBRLMM.html#a3632279e1824737d1d37242d2d25191e" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">explainSelf</a>();
<a name="l01507"></a>01507   <span class="keywordtype">double</span> hetMult = 1.0, K = 4;
<a name="l01508"></a>01508   <span class="keywordtype">double</span> iterationThreshold = 0.3;
<a name="l01509"></a>01509   <span class="keywordtype">int</span> iterations = 0;
<a name="l01510"></a>01510   <span class="keywordtype">bool</span> lowprecision = <span class="keyword">false</span>;
<a name="l01511"></a>01511   <span class="keyword">enum</span> <a class="code" href="classQuantBRLMM.html#aa41baee052c062022a0ab6a999c21638" title="What transformation of intensity estimates to feed into classifier.">QuantBRLMM::Transformation</a> transform = CCS;
<a name="l01512"></a>01512   <span class="keywordtype">string</span> transformStr;
<a name="l01513"></a>01513   map&lt;string,string&gt;::iterator iter;
<a name="l01514"></a>01514   <span class="keywordtype">int</span> priorWeight = 40;
<a name="l01515"></a>01515   <span class="keywordtype">int</span> priorMinCall = 2;
<a name="l01516"></a>01516   fillInValue(hetMult, <span class="stringliteral">&quot;het-mult&quot;</span>, param, doc);
<a name="l01517"></a>01517   fillInValue(iterations, <span class="stringliteral">&quot;iterations&quot;</span>, param, doc);
<a name="l01518"></a>01518   fillInValue(iterationThreshold, <span class="stringliteral">&quot;iter-thresh&quot;</span>, param, doc);
<a name="l01519"></a>01519   fillInValue(K, <span class="stringliteral">&quot;K&quot;</span>, param, doc);
<a name="l01520"></a>01520   fillInValue(transformStr, <span class="stringliteral">&quot;transform&quot;</span>, param, doc);
<a name="l01521"></a>01521   transform = <a class="code" href="classQuantBRLMM.html#ac5880b4177534c9516f3d7243cb9aed7" title="Convert a (text insensitive) text repesentation of a transformation into the enumration.">transformationForString</a>(transformStr);
<a name="l01522"></a>01522   fillInValue(priorWeight, <span class="stringliteral">&quot;prior-weight&quot;</span>, param, doc);
<a name="l01523"></a>01523   fillInValue(priorMinCall, <span class="stringliteral">&quot;prior-mincall&quot;</span>, param, doc);
<a name="l01524"></a>01524   fillInValue(lowprecision, <span class="stringliteral">&quot;lowprecision&quot;</span>, param, doc);
<a name="l01525"></a>01525   <a class="code" href="classQuantBRLMM.html" title="Implements a SNP genotype calling method motivated by Nusrat Rabbee&amp;#39;s RLMM (Robust Linear Model w...">QuantBRLMM</a> *brlmm = <span class="keyword">new</span> <a class="code" href="classQuantBRLMM.html#a703d4e7e27b3cb69e48518618b7c650a" title="Constructor, currently creates prior estimates from pre-calculated data from R.">QuantBRLMM</a>(hetMult, iterations, iterationThreshold, K, transform,
<a name="l01526"></a>01526                                      priorWeight, priorMinCall, lowprecision);
<a name="l01527"></a>01527         <span class="keywordtype">double</span> score;
<a name="l01528"></a>01528         fillInValue(score,<span class="stringliteral">&quot;MS&quot;</span>,param,doc);
<a name="l01529"></a>01529         brlmm-&gt;setMaxScore(score);
<a name="l01530"></a>01530   <span class="keywordflow">return</span> brlmm;
<a name="l01531"></a>01531 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:52 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
