<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/QuantLabelZ.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>QuantLabelZ.h</h1>  </div>
</div>
<div class="contents">
<a href="QuantLabelZ_8h.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2007 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00023"></a>00023 <span class="comment"></span><span class="comment">//</span>
<a name="l00024"></a>00024 <span class="comment">// Copyright (C) 2007 Affymetrix, Inc.</span>
<a name="l00025"></a>00025 <span class="comment">//</span>
<a name="l00026"></a>00026 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00027"></a>00027 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00028"></a>00028 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00029"></a>00029 <span class="comment">// </span>
<a name="l00030"></a>00030 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00031"></a>00031 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00032"></a>00032 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00033"></a>00033 <span class="comment">// General Public License for more details.</span>
<a name="l00034"></a>00034 <span class="comment">// </span>
<a name="l00035"></a>00035 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00036"></a>00036 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00037"></a>00037 <span class="comment">// </span>
<a name="l00038"></a>00038 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00039"></a>00039 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00040"></a>00040 <span class="comment">//</span><span class="comment"></span>
<a name="l00041"></a>00041 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00042"></a>00042 <span class="comment"></span><span class="comment"></span>
<a name="l00043"></a>00043 <span class="comment">/**</span>
<a name="l00044"></a>00044 <span class="comment"> * @file   QuantLabelZ.h</span>
<a name="l00045"></a>00045 <span class="comment"> * @author Earl Hubbell </span>
<a name="l00046"></a>00046 <span class="comment"> * @date   Oct 30 2006</span>
<a name="l00047"></a>00047 <span class="comment"> */</span>
<a name="l00048"></a>00048 
<a name="l00049"></a>00049 <span class="preprocessor">#ifndef _QUANTLABELZ_H_</span>
<a name="l00050"></a>00050 <span class="preprocessor"></span><span class="preprocessor">#define _QUANTLABELZ_H_</span>
<a name="l00051"></a>00051 <span class="preprocessor"></span>
<a name="l00052"></a>00052 <span class="comment">//</span>
<a name="l00053"></a>00053 <span class="preprocessor">#include &quot;<a class="code" href="BioTypes_8h.html" title="Central location enumerations and other types that aren&amp;#39;t particular to a single class or applica...">chipstream/BioTypes.h</a>&quot;</span>
<a name="l00054"></a>00054 <span class="preprocessor">#include &quot;<a class="code" href="ClusterZ_8h.html">chipstream/ClusterZ.h</a>&quot;</span>
<a name="l00055"></a>00055 <span class="preprocessor">#include &quot;<a class="code" href="GenoUtility_8h.html">chipstream/GenoUtility.h</a>&quot;</span>
<a name="l00056"></a>00056 <span class="preprocessor">#include &quot;<a class="code" href="ProbeSet_8h.html" title="Container class for probe sets.">chipstream/ProbeSet.h</a>&quot;</span>
<a name="l00057"></a>00057 <span class="preprocessor">#include &quot;<a class="code" href="QuantExprMethod_8h.html" title="Interface for computing quantification summaries from PM intensities grouped into probe set groups...">chipstream/QuantExprMethod.h</a>&quot;</span>
<a name="l00058"></a>00058 <span class="preprocessor">#include &quot;<a class="code" href="QuantGTypeMethod_8h.html" title="Interface for methods computing SNP genotyping calls.">chipstream/QuantGTypeMethod.h</a>&quot;</span>
<a name="l00059"></a>00059 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethod_8h.html" title="Abstract interface for a quantification method.">chipstream/QuantMethod.h</a>&quot;</span>
<a name="l00060"></a>00060 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodExprReport_8h.html" title="Class for reporting results of quantification methods.">chipstream/QuantMethodExprReport.h</a>&quot;</span>
<a name="l00061"></a>00061 <span class="comment">//</span>
<a name="l00062"></a>00062 <span class="preprocessor">#include &quot;<a class="code" href="covarnorm_8h.html">algorithm/covarnorm/covarnorm.h</a>&quot;</span>
<a name="l00063"></a>00063 <span class="preprocessor">#include &quot;<a class="code" href="PrimeEM_8h.html">algorithm/em/PrimeEM.h</a>&quot;</span>
<a name="l00064"></a>00064 <span class="preprocessor">#include &quot;<a class="code" href="ProbeSelector_8h.html">algorithm/selector/ProbeSelector.h</a>&quot;</span>
<a name="l00065"></a>00065 <span class="preprocessor">#include &quot;file5/File5.h&quot;</span>
<a name="l00066"></a>00066 <span class="preprocessor">#include &quot;<a class="code" href="snp_8label_8h.html">label/snp.label.h</a>&quot;</span>
<a name="l00067"></a>00067 <span class="preprocessor">#include &quot;<a class="code" href="stats_8h.html" title="///">stats/stats.h</a>&quot;</span>
<a name="l00068"></a>00068 <span class="preprocessor">#include &quot;<a class="code" href="AffxArray_8h.html" title="This header contains the AffxArray class definition.">util/AffxArray.h</a>&quot;</span>
<a name="l00069"></a>00069 <span class="comment">//</span>
<a name="l00070"></a>00070 <span class="preprocessor">#include &quot;newmat.h&quot;</span>
<a name="l00071"></a>00071 <span class="comment">//</span>
<a name="l00072"></a>00072 <span class="preprocessor">#include &lt;cfloat&gt;</span>
<a name="l00073"></a>00073 <span class="preprocessor">#include &lt;cstring&gt;</span>
<a name="l00074"></a>00074 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00075"></a>00075 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00076"></a>00076 <span class="comment">//</span>
<a name="l00077"></a>00077 
<a name="l00078"></a>00078 <span class="preprocessor">#if defined (WIN32)</span>
<a name="l00079"></a>00079 <span class="preprocessor"></span><span class="preprocessor">#define asinh(x) (log(x+sqrt(x*x+1))) </span>
<a name="l00080"></a>00080 <span class="preprocessor"></span><span class="preprocessor">#endif</span>
<a name="l00081"></a>00081 <span class="preprocessor"></span><span class="comment"></span>
<a name="l00082"></a>00082 <span class="comment">/// String describing quant method</span>
<a name="l00083"></a><a class="code" href="QuantLabelZ_8h.html#a7225f2c577370e0fcaa1de762b340288">00083</a> <span class="comment"></span><span class="preprocessor">#define QUANTBRLMMP &quot;brlmm-p&quot;</span>
<a name="l00084"></a>00084 <span class="preprocessor"></span><span class="comment"></span>
<a name="l00085"></a>00085 <span class="comment">//////////</span>
<a name="l00086"></a>00086 <span class="comment"></span>
<a name="l00087"></a><a class="code" href="classsnp__labeled__distribution.html">00087</a> <span class="keyword">class </span><a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>
<a name="l00088"></a>00088 {
<a name="l00089"></a>00089 <span class="keyword">public</span>:
<a name="l00090"></a>00090     std::string probeset_id;
<a name="l00091"></a>00091 
<a name="l00092"></a>00092     <a class="code" href="classsnp__distribution.html" title="this class holds all three genotype clusters">snp_distribution</a> Dist;
<a name="l00093"></a>00093     <span class="comment">//cluster_data bb;</span>
<a name="l00094"></a>00094     <span class="comment">//cluster_data ab;</span>
<a name="l00095"></a>00095     <span class="comment">//cluster_data aa;</span>
<a name="l00096"></a>00096     <span class="comment">//double cv1; = xah</span>
<a name="l00097"></a>00097     <span class="comment">//double cv2; = xab</span>
<a name="l00098"></a>00098     <span class="comment">//double cv3; = xhb</span>
<a name="l00099"></a>00099 
<a name="l00100"></a>00100     <span class="keywordtype">int</span> compareTo(<a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>&amp; that, <span class="keywordtype">int</span> iCompareCode)
<a name="l00101"></a>00101     {
<a name="l00102"></a>00102         <span class="keywordtype">int</span> iCompareResult = 0;
<a name="l00103"></a>00103         <span class="keywordflow">switch</span> (iCompareCode)
<a name="l00104"></a>00104         {
<a name="l00105"></a>00105         <span class="keywordflow">case</span> 0:
<a name="l00106"></a>00106             iCompareResult = probeset_id.compare(that.probeset_id);
<a name="l00107"></a>00107             <span class="keywordflow">break</span>;
<a name="l00108"></a>00108         }
<a name="l00109"></a>00109         <span class="keywordflow">return</span> iCompareResult;
<a name="l00110"></a>00110     }
<a name="l00111"></a>00111 
<a name="l00112"></a><a class="code" href="structsnp__labeled__distribution_1_1ComparePred.html">00112</a>     <span class="keyword">template</span>&lt;<span class="keywordtype">int</span> k&gt; <span class="keyword">struct </span><a class="code" href="structsnp__labeled__distribution_1_1ComparePred.html">ComparePred</a> {
<a name="l00113"></a>00113         <span class="keywordtype">bool</span> operator()(<span class="keyword">const</span> <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>* lhs, <span class="keyword">const</span> <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>* rhs)<span class="keyword"> const </span>{
<a name="l00114"></a>00114             <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;snp_labeled_distribution: ComparePred instantiated with an invalid compare code = &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(k));
<a name="l00115"></a>00115             <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00116"></a>00116         }
<a name="l00117"></a>00117     };
<a name="l00118"></a>00118 };
<a name="l00119"></a>00119 
<a name="l00120"></a><a class="code" href="structsnp__labeled__distribution_1_1ComparePred_3_010_01_4.html">00120</a> <span class="keyword">template</span>&lt;&gt; <span class="keyword">struct </span><a class="code" href="structsnp__labeled__distribution_1_1ComparePred.html">snp_labeled_distribution::ComparePred</a>&lt;0&gt; {
<a name="l00121"></a>00121     <span class="keywordtype">bool</span> operator()(<span class="keyword">const</span> <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>* lhs, <span class="keyword">const</span> <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a>* rhs)<span class="keyword"> const </span>{
<a name="l00122"></a>00122         <span class="keywordflow">return</span> lhs-&gt;probeset_id.compare(rhs-&gt;probeset_id) &lt; 0;
<a name="l00123"></a>00123     }
<a name="l00124"></a>00124 };
<a name="l00125"></a>00125 <span class="comment"></span>
<a name="l00126"></a>00126 <span class="comment">//////////</span>
<a name="l00127"></a>00127 <span class="comment"></span><span class="comment"></span>
<a name="l00128"></a>00128 <span class="comment">/**</span>
<a name="l00129"></a>00129 <span class="comment"> * This is the primary class wrapping the bayes_label_routine</span>
<a name="l00130"></a>00130 <span class="comment"> * it handles bookkeeping, gender determination, etc</span>
<a name="l00131"></a>00131 <span class="comment"> */</span>
<a name="l00132"></a><a class="code" href="classQuantLabelZ.html">00132</a> <span class="keyword">class </span><a class="code" href="classQuantLabelZ.html" title="This is the primary class wrapping the bayes_label_routine it handles bookkeeping, gender determination, etc.">QuantLabelZ</a> : <span class="keyword">public</span> <a class="code" href="classQuantGTypeMethod.html" title="Quantification methods used for making genotyping calls implement (currently just brlmm) this interfa...">QuantGTypeMethod</a> {
<a name="l00133"></a>00133 
<a name="l00134"></a>00134 <span class="keyword">public</span>:
<a name="l00135"></a>00135 <span class="comment"></span>
<a name="l00136"></a>00136 <span class="comment">  /// What transformation of intensity estimates to feed into classifier.</span>
<a name="l00137"></a>00137 <span class="comment"></span>  
<a name="l00138"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html">00138</a>   <span class="keyword">class </span><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html" title="What transformation of intensity estimates to feed into classifier.">BrlmmParam</a> {
<a name="l00139"></a>00139     
<a name="l00140"></a>00140   <span class="keyword">public</span>:
<a name="l00141"></a>00141 
<a name="l00142"></a>00142     <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html" title="What transformation of intensity estimates to feed into classifier.">BrlmmParam</a>() {
<a name="l00143"></a>00143       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ab8a874357cd7cf86a7f8fba0acf53508" title="Number to balance het calls with to balance performance on het/hom calls.">m_HetMultiplier</a> = 1;
<a name="l00144"></a>00144       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a40daffe79582c804947cd65ca165cb2d" title="How many iterations to do of LabelZ, i.e. boostrapping from initial calls.">m_Iterations</a> = 0;
<a name="l00145"></a>00145       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ab4a7d2ab81cf781f5214976298befe56" title="What is our threshold (maximum value) for confidences befor /// calling &amp;#39;NN&amp;#39; in iterative mod...">m_IterationThreshold</a> = 0.6;
<a name="l00146"></a>00146       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e" title="Scaling Parameter to use in transformations.">m_K</a> = 1;
<a name="l00147"></a>00147       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a3fbc36e45e6b0b402cf99181d54b9bba" title="How much weight should the prior get?">m_PriorPseudoCount</a> = 20;
<a name="l00148"></a>00148       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a> = MvA;
<a name="l00149"></a>00149       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ad5fcdd569b8c67f3481be1ad83c83f6d" title="Minimum examples for each genotype to be included in prior estimation.">m_MinCluster</a> = 2;
<a name="l00150"></a>00150       <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a> = <span class="keyword">false</span>;
<a name="l00151"></a>00151     }
<a name="l00152"></a>00152     
<a name="l00153"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ab8a874357cd7cf86a7f8fba0acf53508">00153</a>     <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ab8a874357cd7cf86a7f8fba0acf53508" title="Number to balance het calls with to balance performance on het/hom calls.">m_HetMultiplier</a>;          <span class="comment">///&lt; Number to balance het calls with to balance performance on het/hom calls.</span>
<a name="l00154"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a40daffe79582c804947cd65ca165cb2d">00154</a> <span class="comment"></span>    <span class="keywordtype">int</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a40daffe79582c804947cd65ca165cb2d" title="How many iterations to do of LabelZ, i.e. boostrapping from initial calls.">m_Iterations</a>;                <span class="comment">///&lt; How many iterations to do of LabelZ, i.e. boostrapping from initial calls.</span>
<a name="l00155"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ab4a7d2ab81cf781f5214976298befe56">00155</a> <span class="comment"></span>    <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ab4a7d2ab81cf781f5214976298befe56" title="What is our threshold (maximum value) for confidences befor /// calling &amp;#39;NN&amp;#39; in iterative mod...">m_IterationThreshold</a>;      <span class="comment">///&lt; What is our threshold (maximum value) for confidences befor </span>
<a name="l00156"></a>00156 <span class="comment"></span><span class="comment">                                     ///  calling &#39;NN&#39; in iterative mode?</span>
<a name="l00157"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e">00157</a> <span class="comment"></span>    <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e" title="Scaling Parameter to use in transformations.">m_K</a>;                      <span class="comment">///&lt; Scaling Parameter to use in transformations.</span>
<a name="l00158"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d">00158</a> <span class="comment"></span>    <span class="keyword">enum</span> <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>; <span class="comment">///&lt; What transformation of initial data are we feeding into the classifier?</span>
<a name="l00159"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a3fbc36e45e6b0b402cf99181d54b9bba">00159</a> <span class="comment"></span>    <span class="keywordtype">int</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a3fbc36e45e6b0b402cf99181d54b9bba" title="How much weight should the prior get?">m_PriorPseudoCount</a>;          <span class="comment">///&lt; How much weight should the prior get?</span>
<a name="l00160"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ad5fcdd569b8c67f3481be1ad83c83f6d">00160</a> <span class="comment"></span>    <span class="keywordtype">int</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#ad5fcdd569b8c67f3481be1ad83c83f6d" title="Minimum examples for each genotype to be included in prior estimation.">m_MinCluster</a>;                <span class="comment">///&lt; Minimum examples for each genotype to be included in prior estimation.</span>
<a name="l00161"></a><a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960">00161</a> <span class="comment"></span>    <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#a42a668a034fcd91cdaf0b4dc7a411960" title="Summary values fed into R are usually truncated, should we do /// that to be compatible? (used for regression tests).">m_LowPrecision</a>;             <span class="comment">///&lt; Summary values fed into R are usually truncated, should we do </span>
<a name="l00162"></a>00162 <span class="comment"></span><span class="comment">                                     /// that to be compatible? (used for regression tests)</span>
<a name="l00163"></a>00163 <span class="comment"></span>  };
<a name="l00164"></a>00164 
<a name="l00165"></a>00165   <span class="keywordtype">bool</span> m_OutputProbabilities;
<a name="l00166"></a>00166   std::vector&lt;double&gt; m_Txa;         <span class="comment">/// probability of AA call</span>
<a name="l00167"></a><a class="code" href="classQuantLabelZ.html#a7a880d38f86a87b124407c7939bc2900">00167</a> <span class="comment"></span>  std::vector&lt;double&gt; <a class="code" href="classQuantLabelZ.html#a7a880d38f86a87b124407c7939bc2900" title="probability of AA call">m_Txh</a>;         <span class="comment">/// probability of het call</span>
<a name="l00168"></a><a class="code" href="classQuantLabelZ.html#a9aac165608acb81536b632f5da67910a">00168</a> <span class="comment"></span>  std::vector&lt;double&gt; <a class="code" href="classQuantLabelZ.html#a9aac165608acb81536b632f5da67910a" title="probability of het call">m_Txb</a>;         <span class="comment">/// probability of BB call</span>
<a name="l00169"></a>00169 <span class="comment"></span>
<a name="l00170"></a>00170 <span class="comment"></span>
<a name="l00171"></a>00171 <span class="comment">  /** Constructor, currently creates prior estimates from pre-calculated data from R. */</span>
<a name="l00172"></a>00172   <a class="code" href="classQuantLabelZ.html#a2cea6fc92475386eeca1511393bc6af4" title="probability of BB call">QuantLabelZ</a>( <span class="keyword">enum</span> <a class="code" href="GenoUtility_8h.html#a0176e2f61ce5ea7a3495a511cf46b2c1">Transformation</a> <a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909" title="Transform allele estimates.">transform</a>,
<a name="l00173"></a>00173                <span class="keywordtype">double</span> K,
<a name="l00174"></a>00174                <span class="keywordtype">bool</span> lowPrecision);
<a name="l00175"></a>00175 <span class="comment"></span>
<a name="l00176"></a>00176 <span class="comment">  /** Fill in our self documentation from current state. */</span>
<a name="l00177"></a>00177   <span class="keyword">static</span> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#acfd3b0ca4567d0d580602e974bd8de7a" title="Fill in our self documentation from current state.">setupSelfDoc</a>(<a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> &amp;doc);<span class="comment"></span>
<a name="l00178"></a>00178 <span class="comment">  /** Destructor. */</span>
<a name="l00179"></a>00179   <span class="keyword">virtual</span> <a class="code" href="classQuantLabelZ.html#a67181557019fd8cd7600c57f8b989bac" title="Destructor.">~QuantLabelZ</a>();
<a name="l00180"></a>00180 
<a name="l00181"></a>00181 <span class="comment">//  virtual void getPrior(string &amp;TmpName, int copyIx, snp_labeled_distribution &amp;objSearch, snp_param &amp;tsp);</span>
<a name="l00182"></a>00182 <span class="comment"></span>
<a name="l00183"></a>00183 <span class="comment">  /** </span>
<a name="l00184"></a>00184 <span class="comment">   * @brief Set up the quantification method given all the data about the probe</span>
<a name="l00185"></a>00185 <span class="comment">   * set, chip layout and data.</span>
<a name="l00186"></a>00186 <span class="comment">   * </span>
<a name="l00187"></a>00187 <span class="comment">   * @param psGroup - Probes to be used for final estimate.</span>
<a name="l00188"></a>00188 <span class="comment">   * @param layout - Chip layout annotation.</span>
<a name="l00189"></a>00189 <span class="comment">   * @param iMart - Raw data from chips.</span>
<a name="l00190"></a>00190 <span class="comment">   * @param iTrans - Transformations to be applied to data before use.</span>
<a name="l00191"></a>00191 <span class="comment">   * @param pmAdjust - How to estimate background, or MM probe.</span>
<a name="l00192"></a>00192 <span class="comment">   * @return True if setup sucessful, false otherwise.</span>
<a name="l00193"></a>00193 <span class="comment">   */</span>
<a name="l00194"></a>00194   <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a6881e990126c3e628f72a99a94452cbe" title="Set up the quantification method given all the data about the probe set, chip layout and data...">setUp</a>(<a class="code" href="classProbeSetGroup.html" title="Group of probe sets that should be processed as a single large probe set.">ProbeSetGroup</a> &amp;psGroup, <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, 
<a name="l00195"></a>00195              std::vector&lt;ChipStream *&gt; &amp;iTrans, <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust);
<a name="l00196"></a>00196 
<a name="l00197"></a>00197   <span class="keywordtype">void</span> getSequentialPrior(std::string &amp;name, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;tsp, <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> &amp;sDist);
<a name="l00198"></a>00198   
<a name="l00199"></a>00199   <span class="keywordtype">void</span> getPrior(std::string &amp;TmpName, <span class="keywordtype">int</span> presentCopyNumber, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;tsp, <a class="code" href="classsnp__labeled__distribution.html">snp_labeled_distribution</a> &amp;sDist);
<a name="l00200"></a>00200 <span class="comment"></span>
<a name="l00201"></a>00201 <span class="comment">  /**</span>
<a name="l00202"></a>00202 <span class="comment">   * @brief return comma-joined string of call probabilites: BB,AB,AA,Ocean</span>
<a name="l00203"></a>00203 <span class="comment">   */</span>
<a name="l00204"></a>00204   std::string <a class="code" href="classQuantLabelZ.html#ac3a27a5f1f6ed05985950ae124193b09" title="return comma-joined string of call probabilites: BB,AB,AA,Ocean">getCallProbabilities</a>(<span class="keywordtype">int</span> idx);
<a name="l00205"></a>00205 
<a name="l00206"></a>00206 <span class="comment"></span>
<a name="l00207"></a>00207 <span class="comment">  /** </span>
<a name="l00208"></a>00208 <span class="comment">   * @brief Do the heavy lifting of estimation.</span>
<a name="l00209"></a>00209 <span class="comment">   */</span>
<a name="l00210"></a>00210   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ad83c392dc97c1557e25bdd8b9b901ec9" title="Do the heavy lifting of estimation.">computeEstimate</a>();
<a name="l00211"></a>00211 <span class="comment"></span>
<a name="l00212"></a>00212 <span class="comment">  /**</span>
<a name="l00213"></a>00213 <span class="comment">   * @brief preprocess and normalize values</span>
<a name="l00214"></a>00214 <span class="comment">   */</span>
<a name="l00215"></a>00215   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#afa5e98cbfea742735a11e38542ed72ba" title="preprocess and normalize values">PreprocessValues</a>();
<a name="l00216"></a>00216  <span class="comment"></span>
<a name="l00217"></a>00217 <span class="comment">  /** </span>
<a name="l00218"></a>00218 <span class="comment">   * @brief What is the name of the quantification method?</span>
<a name="l00219"></a>00219 <span class="comment">   * @return name of adjuster.</span>
<a name="l00220"></a>00220 <span class="comment">   */</span>  
<a name="l00221"></a><a class="code" href="classQuantLabelZ.html#a43cabd662be310c394c4317afcc68286">00221</a>   std::string <a class="code" href="classQuantLabelZ.html#a43cabd662be310c394c4317afcc68286" title="What is the name of the quantification method?">getType</a>() {
<a name="l00222"></a>00222     <span class="keywordflow">return</span> std::string(<a class="code" href="QuantLabelZ_8h.html#a7225f2c577370e0fcaa1de762b340288" title="String describing quant method.">QUANTBRLMMP</a>);
<a name="l00223"></a>00223   }
<a name="l00224"></a>00224 <span class="comment"></span>
<a name="l00225"></a>00225 <span class="comment">  /** </span>
<a name="l00226"></a>00226 <span class="comment">   * What version of the BRLMM-p algorithm are we implementing?</span>
<a name="l00227"></a>00227 <span class="comment">   * @return version string.</span>
<a name="l00228"></a>00228 <span class="comment">   */</span>
<a name="l00229"></a><a class="code" href="classQuantLabelZ.html#a04287367c0d1fc91a5a43c986841e88f">00229</a>   std::string <a class="code" href="classQuantLabelZ.html#a04287367c0d1fc91a5a43c986841e88f" title="What version of the BRLMM-p algorithm are we implementing?">getVersion</a>() {
<a name="l00230"></a>00230     <span class="keywordflow">return</span> <span class="stringliteral">&quot;1.0&quot;</span>;
<a name="l00231"></a>00231   }
<a name="l00232"></a>00232 <span class="comment"></span>
<a name="l00233"></a>00233 <span class="comment">  /** Set the quantification method to use for summarizing individual alleles */</span>
<a name="l00234"></a><a class="code" href="classQuantLabelZ.html#ac8c2cc4ab6dc319d4c0bff9b3f299642">00234</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ac8c2cc4ab6dc319d4c0bff9b3f299642" title="Set the quantification method to use for summarizing individual alleles.">setQuantExprMethod</a>(<a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *qMethod) {
<a name="l00235"></a>00235     <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a> = qMethod;
<a name="l00236"></a>00236   }
<a name="l00237"></a>00237 <span class="comment"></span>
<a name="l00238"></a>00238 <span class="comment">  /** Get the quantification method to use for summarizing individual alleles */</span>
<a name="l00239"></a><a class="code" href="classQuantLabelZ.html#a7b96f626c861143496927f87282e8914">00239</a>   <a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *<a class="code" href="classQuantLabelZ.html#a7b96f626c861143496927f87282e8914" title="Get the quantification method to use for summarizing individual alleles.">getQuantExprMethod</a>() {
<a name="l00240"></a>00240     <span class="keywordflow">return</span> <a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>;
<a name="l00241"></a>00241   }
<a name="l00242"></a>00242 <span class="comment"></span>
<a name="l00243"></a>00243 <span class="comment">  /** </span>
<a name="l00244"></a>00244 <span class="comment">   * @brief Default Getter method for parameters and their documentation.</span>
<a name="l00245"></a>00245 <span class="comment">   * @return map of parameters and their descriptions.</span>
<a name="l00246"></a>00246 <span class="comment">   */</span>
<a name="l00247"></a>00247   <span class="keyword">static</span> std::vector&lt;SelfDoc::Opt&gt; <a class="code" href="classQuantLabelZ.html#a0da80e2dbed413e783338e1ae66c8b43" title="Default Getter method for parameters and their documentation.">getDefaultDocOptions</a>();
<a name="l00248"></a>00248 <span class="comment"></span>
<a name="l00249"></a>00249 <span class="comment">  /** </span>
<a name="l00250"></a>00250 <span class="comment">   * @brief Supply a little how/what/why about the algorithms this</span>
<a name="l00251"></a>00251 <span class="comment">   * class performs and what parameters it takes.</span>
<a name="l00252"></a>00252 <span class="comment">   * @return SelfDoc</span>
<a name="l00253"></a>00253 <span class="comment">   */</span>
<a name="l00254"></a>00254   <span class="keyword">static</span> <a class="code" href="classSelfDoc.html" title="Small interface to for algorithmic classes that can explain themselves.">SelfDoc</a> <a class="code" href="classQuantLabelZ.html#ad4dbeec9c331299c65e6859e16b9e9d8" title="Supply a little how/what/why about the algorithms this class performs and what parameters it takes...">explainSelf</a>();
<a name="l00255"></a>00255  <span class="comment"></span>
<a name="l00256"></a>00256 <span class="comment">  /** </span>
<a name="l00257"></a>00257 <span class="comment">   * @brief This static function should be overridden by child classes</span>
<a name="l00258"></a>00258 <span class="comment">   * to return an object of the correct type initialized correctly</span>
<a name="l00259"></a>00259 <span class="comment">   * with the parameters in the string, string map. All objects</span>
<a name="l00260"></a>00260 <span class="comment">   * created this way should be deleted when finished using.</span>
<a name="l00261"></a>00261 <span class="comment">   * </span>
<a name="l00262"></a>00262 <span class="comment">   * @param param - Map of key/value pairs to initialize the object.</span>
<a name="l00263"></a>00263 <span class="comment">   * </span>
<a name="l00264"></a>00264 <span class="comment">   * @return Pointer toCreate object, this should be sub casted as necessary.</span>
<a name="l00265"></a>00265 <span class="comment">   */</span>
<a name="l00266"></a>00266   <span class="keyword">static</span> <a class="code" href="classSelfCreate.html" title="Small interface for functions that know how to make an instance of themselves given a map of key...">SelfCreate</a> *<a class="code" href="classQuantLabelZ.html#a3063c55509d69c79f1ad5833f9f62691" title="This static function should be overridden by child classes to return an object of the correct type in...">newObject</a>(std::map&lt;std::string,std::string&gt; &amp;param);
<a name="l00267"></a>00267   <span class="comment"></span>
<a name="l00268"></a>00268 <span class="comment">  /** </span>
<a name="l00269"></a>00269 <span class="comment">   * Clear out all of our results and values.</span>
<a name="l00270"></a>00270 <span class="comment">   */</span>
<a name="l00271"></a>00271   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a23d0f3fb96579a159ad8d088fd7747eb" title="Clear out all of our results and values.">blankSelf</a>();
<a name="l00272"></a>00272 <span class="comment"></span>
<a name="l00273"></a>00273 <span class="comment">  /** </span>
<a name="l00274"></a>00274 <span class="comment">   * How many genotyping calls do we have? Also indicates how many</span>
<a name="l00275"></a>00275 <span class="comment">   * samples there are.</span>
<a name="l00276"></a>00276 <span class="comment">   *</span>
<a name="l00277"></a>00277 <span class="comment">   * @return Number of genotyping calls made.</span>
<a name="l00278"></a>00278 <span class="comment">   */</span>
<a name="l00279"></a><a class="code" href="classQuantLabelZ.html#a8f80929384930a6dba286e6d59d86e2c">00279</a>   <span class="keyword">virtual</span> <span class="keyword">inline</span> <span class="keywordtype">size_t</span> <a class="code" href="classQuantLabelZ.html#a8f80929384930a6dba286e6d59d86e2c" title="How many genotyping calls do we have? Also indicates how many samples there are.">getNumCalls</a>() { <span class="keywordflow">return</span> m_Calls.size(); }
<a name="l00280"></a>00280 <span class="comment"></span>
<a name="l00281"></a>00281 <span class="comment">  /** </span>
<a name="l00282"></a>00282 <span class="comment">   * Get our confidence value for a particular call in a particular sample.</span>
<a name="l00283"></a>00283 <span class="comment">   * @param index - sample of interest.</span>
<a name="l00284"></a>00284 <span class="comment">   * @return - Confidence in the predicted call.</span>
<a name="l00285"></a>00285 <span class="comment">   */</span>
<a name="l00286"></a><a class="code" href="classQuantLabelZ.html#a91aac61561a7fc442acb8292cbd2063d">00286</a>   <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ.html#a91aac61561a7fc442acb8292cbd2063d" title="Get our confidence value for a particular call in a particular sample.">getConfidence</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index) { 
<a name="l00287"></a>00287     <span class="keywordflow">if</span>(index &gt;= <a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>.size()) {
<a name="l00288"></a>00288       <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Asking for call at index &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(index) + <span class="stringliteral">&quot; when Probeset &quot;</span> + 
<a name="l00289"></a>00289                     <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> + <span class="stringliteral">&quot; has only &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>.size()) + <span class="stringliteral">&quot; calls.&quot;</span>);
<a name="l00290"></a>00290     }
<a name="l00291"></a>00291     <span class="keywordflow">return</span> <a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>[index];
<a name="l00292"></a>00292   }
<a name="l00293"></a>00293 <span class="comment"></span>
<a name="l00294"></a>00294 <span class="comment">  /** </span>
<a name="l00295"></a>00295 <span class="comment">   * Returns standardized distances to cluster centers for given genotype.</span>
<a name="l00296"></a>00296 <span class="comment">   * @param genoType -  Genotype of interest.</span>
<a name="l00297"></a>00297 <span class="comment">   * @param index - sample of interest.</span>
<a name="l00298"></a>00298 <span class="comment">   * @return - Standardized distance to cluster center.</span>
<a name="l00299"></a>00299 <span class="comment">   */</span>
<a name="l00300"></a><a class="code" href="classQuantLabelZ.html#aa36c2924228ba5dd36dd52e4e2930b7d">00300</a>   <span class="keyword">virtual</span> <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ.html#aa36c2924228ba5dd36dd52e4e2930b7d" title="Returns standardized distances to cluster centers for given genotype.">getConfGenotype</a>(affx::GType genoType, <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index) {<span class="comment"></span>
<a name="l00301"></a>00301 <span class="comment">      ///@todo we should either implement this, or not output </span>
<a name="l00302"></a>00302 <span class="comment">      ///      it in QuantMethodReportGTypeChipsummary</span>
<a name="l00303"></a>00303 <span class="comment"></span>    assert(index &lt; <a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>[0].size());
<a name="l00304"></a>00304     <span class="keywordflow">if</span>(genoType == <a class="code" href="namespaceaffx.html#a3886a33ccf4ba70052a915d878634c30a1cef2fc4697bd5fc67116fbd1ee2027b" title="No call, unknown genotype. The internal is value is &amp;#39;3&amp;#39;; external is &amp;#39;-1&amp;#39;,.">affx::NN</a>)
<a name="l00305"></a>00305       <span class="keywordflow">return</span> FLT_MAX;
<a name="l00306"></a>00306    <span class="comment">// this matrix is reversed from BRLMM for convenience</span>
<a name="l00307"></a>00307    <span class="comment">// in computing the distances vs subsets of the data</span>
<a name="l00308"></a>00308    <span class="comment">// which occurs in chrX/Y cases in males/females</span>
<a name="l00309"></a>00309     <span class="keywordflow">return</span> <a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>[genoType][index];
<a name="l00310"></a>00310 
<a name="l00311"></a>00311     <span class="comment">//return 0;</span>
<a name="l00312"></a>00312   }
<a name="l00313"></a>00313  <span class="comment"></span>
<a name="l00314"></a>00314 <span class="comment">  /** </span>
<a name="l00315"></a>00315 <span class="comment">   * Get the summary value after transformation for the A and B alleles respectively.</span>
<a name="l00316"></a>00316 <span class="comment">   * </span>
<a name="l00317"></a>00317 <span class="comment">   * @param index - Which chip to get values for.</span>
<a name="l00318"></a>00318 <span class="comment">   * @param aValue - Filled in with transformed a value.</span>
<a name="l00319"></a>00319 <span class="comment">   * @param bValue - Filled in with transformed b value.</span>
<a name="l00320"></a>00320 <span class="comment">   */</span>
<a name="l00321"></a><a class="code" href="classQuantLabelZ.html#afa689649ce7b7c3aa0baebad17807a47">00321</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#afa689649ce7b7c3aa0baebad17807a47" title="Get the summary value after transformation for the A and B alleles respectively.">getAlleleValues</a>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index, <span class="keywordtype">double</span> &amp;aValue, <span class="keywordtype">double</span> &amp;bValue) {
<a name="l00322"></a>00322     assert(index &lt; <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size());
<a name="l00323"></a>00323     assert(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.size() == <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.size());
<a name="l00324"></a>00324     aValue = <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>[index];
<a name="l00325"></a>00325     bValue = <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>[index];
<a name="l00326"></a>00326   }
<a name="l00327"></a>00327 <span class="comment"></span>
<a name="l00328"></a>00328 <span class="comment">  /**</span>
<a name="l00329"></a>00329 <span class="comment">   * Transform allele estimates</span>
<a name="l00330"></a>00330 <span class="comment">   */</span>
<a name="l00331"></a><a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909">00331</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a0ec9313e121ec2cc1304350a7bb4a909" title="Transform allele estimates.">transform</a>() {
<a name="l00332"></a>00332       <a class="code" href="GenoUtility_8cpp.html#a1a2531d6033d567af0446274e94f74a4" title="transform the data vector">GenoUtility_transformData</a>(<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>,<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aeefcaa5a378125462d8586f2ac95796e" title="Scaling Parameter to use in transformations.">m_K</a>);
<a name="l00333"></a>00333   }<span class="comment"></span>
<a name="l00334"></a>00334 <span class="comment"> /**</span>
<a name="l00335"></a>00335 <span class="comment">   * Set probeset selection data as &#39;probes&#39;</span>
<a name="l00336"></a>00336 <span class="comment">   *</span>
<a name="l00337"></a>00337 <span class="comment">   * @param aVals - vector of allele A estimates</span>
<a name="l00338"></a>00338 <span class="comment">   * @param bVals - vector of allele B estimates</span>
<a name="l00339"></a>00339 <span class="comment">   */</span>
<a name="l00340"></a><a class="code" href="classQuantLabelZ.html#a68455ce9cb3a0fe17578913a5b49a1a7">00340</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a68455ce9cb3a0fe17578913a5b49a1a7" title="Set probeset selection data as &amp;#39;probes&amp;#39;.">setProbeSetSelect</a>(std::vector&lt;double&gt; &amp;aVals, std::vector&lt;double&gt; &amp;bVals){
<a name="l00341"></a>00341     <span class="keyword">const</span> <span class="keywordtype">size_t</span> NTarg = aVals.size();
<a name="l00342"></a>00342     <span class="keywordtype">int</span> NFeat = 1; <span class="comment">// just the probeset</span>
<a name="l00343"></a>00343         m_PM_A.clear();
<a name="l00344"></a>00344         m_PM_A.push_back(vector&lt;double&gt;(NTarg));
<a name="l00345"></a>00345      
<a name="l00346"></a>00346         m_PM_B.clear();
<a name="l00347"></a>00347         m_PM_B.push_back(vector&lt;double&gt;(NTarg));
<a name="l00348"></a>00348 
<a name="l00349"></a>00349         <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i=0; i&lt; NTarg; i++){
<a name="l00350"></a>00350             m_PM_A[0][i] = aVals[i];
<a name="l00351"></a>00351             m_PM_B[0][i] = bVals[i];
<a name="l00352"></a>00352         }
<a name="l00353"></a>00353         m_A_probeId.assign(NFeat,0);
<a name="l00354"></a>00354         m_B_probeId.assign(NFeat,0);
<a name="l00355"></a>00355   }<span class="comment"></span>
<a name="l00356"></a>00356 <span class="comment">  /**</span>
<a name="l00357"></a>00357 <span class="comment">   * Set allele values</span>
<a name="l00358"></a>00358 <span class="comment">   * </span>
<a name="l00359"></a>00359 <span class="comment">   * @param aVals - vector of allele A estimates</span>
<a name="l00360"></a>00360 <span class="comment">   * @param bVals - vector of allele B estimates</span>
<a name="l00361"></a>00361 <span class="comment">   */</span>
<a name="l00362"></a><a class="code" href="classQuantLabelZ.html#adde341ea0e1f3e3e18ff6e26b105da1e">00362</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#adde341ea0e1f3e3e18ff6e26b105da1e" title="Set allele values.">setAlleleValues</a>(std::vector&lt;double&gt; &amp;aVals, std::vector&lt;double&gt; &amp;bVals) {
<a name="l00363"></a>00363       <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a> = aVals;
<a name="l00364"></a>00364       <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a> = bVals;
<a name="l00365"></a>00365       <span class="keywordflow">if</span> (m_SnpProbeTsv.is_open())
<a name="l00366"></a>00366         <a class="code" href="classQuantLabelZ.html#a68455ce9cb3a0fe17578913a5b49a1a7" title="Set probeset selection data as &amp;#39;probes&amp;#39;.">setProbeSetSelect</a>(aVals,bVals);
<a name="l00367"></a>00367   }
<a name="l00368"></a>00368 <span class="comment"></span>
<a name="l00369"></a>00369 <span class="comment">  /**</span>
<a name="l00370"></a>00370 <span class="comment">   * Set override flag</span>
<a name="l00371"></a>00371 <span class="comment">   * </span>
<a name="l00372"></a>00372 <span class="comment">   * @param oflag - select an override option for hints</span>
<a name="l00373"></a>00373 <span class="comment">   *</span>
<a name="l00374"></a>00374 <span class="comment">   */</span>
<a name="l00375"></a><a class="code" href="classQuantLabelZ.html#acd9b42625edf51edc84489627065f649">00375</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#acd9b42625edf51edc84489627065f649" title="Set override flag.">setOverride</a>(<span class="keywordtype">int</span> oflag) {
<a name="l00376"></a>00376       m_SelectionOverride = oflag;
<a name="l00377"></a>00377       <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;override&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(oflag));
<a name="l00378"></a>00378   }
<a name="l00379"></a>00379 
<a name="l00380"></a>00380 
<a name="l00381"></a>00381   <span class="keyword">inline</span> <span class="keywordtype">void</span> setOutputProbabilities(<span class="keywordtype">bool</span> outFlag) {
<a name="l00382"></a>00382       m_OutputProbabilities = outFlag;
<a name="l00383"></a>00383   }
<a name="l00384"></a>00384  <span class="comment"></span>
<a name="l00385"></a>00385 <span class="comment">  /** </span>
<a name="l00386"></a>00386 <span class="comment">   * Get the summary value names for the A and B alleles respectively.</span>
<a name="l00387"></a>00387 <span class="comment">   * </span>
<a name="l00388"></a>00388 <span class="comment">   * @param aName - Filled in with a value name.</span>
<a name="l00389"></a>00389 <span class="comment">   * @param bName - Filled in with b value name.</span>
<a name="l00390"></a>00390 <span class="comment">   */</span>
<a name="l00391"></a><a class="code" href="classQuantLabelZ.html#a39723c8a88b8deadb321bbe100307292">00391</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a39723c8a88b8deadb321bbe100307292" title="Get the summary value names for the A and B alleles respectively.">getAlleleValueNames</a>(std::string &amp;aName, std::string &amp;bName)
<a name="l00392"></a>00392   {
<a name="l00393"></a>00393       pair&lt;string,string&gt; labels = <a class="code" href="GenoUtility_8cpp.html#a8686c3ca8eabf2bb104f81c970d2ea2d" title="Return the calumn names for the values generated by a given transformation.">columnNamesForTransformation</a>(<a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>.<a class="code" href="classQuantLabelZ_1_1BrlmmParam.html#aca36e3ccdb2991480dd150165de4083d" title="What transformation of initial data are we feeding into the classifier?">m_Transform</a>);
<a name="l00394"></a>00394       aName = labels.first;
<a name="l00395"></a>00395       bName = labels.second;
<a name="l00396"></a>00396   }
<a name="l00397"></a>00397 <span class="comment"></span>
<a name="l00398"></a>00398 <span class="comment">  /** </span>
<a name="l00399"></a>00399 <span class="comment">   * Get the name of the probeset that these calls are being made for. </span>
<a name="l00400"></a>00400 <span class="comment">   * @return name of probeset.</span>
<a name="l00401"></a>00401 <span class="comment">   */</span>
<a name="l00402"></a><a class="code" href="classQuantLabelZ.html#abf83857e95efa7b646fb6122888c4aaa">00402</a>   <span class="keyword">const</span> std::string &amp;<a class="code" href="classQuantLabelZ.html#abf83857e95efa7b646fb6122888c4aaa" title="Get the name of the probeset that these calls are being made for.">getProbeSetName</a>() { <span class="keywordflow">return</span> <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a>; }
<a name="l00403"></a>00403 <span class="comment"></span>
<a name="l00404"></a>00404 <span class="comment">  /** </span>
<a name="l00405"></a>00405 <span class="comment">   * Set the probeset name</span>
<a name="l00406"></a>00406 <span class="comment">   * @param probeset name</span>
<a name="l00407"></a>00407 <span class="comment">   */</span>
<a name="l00408"></a><a class="code" href="classQuantLabelZ.html#a4b257a271e340b468e41a22865b738e6">00408</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a4b257a271e340b468e41a22865b738e6" title="Set the probeset name.">setProbeSetName</a>(<span class="keyword">const</span> std::string &amp;probesetName) {
<a name="l00409"></a>00409       <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> = probesetName;
<a name="l00410"></a>00410   }
<a name="l00411"></a>00411 <span class="comment"></span>
<a name="l00412"></a>00412 <span class="comment">  /** </span>
<a name="l00413"></a>00413 <span class="comment">   * Add an expression reporter to output the summary values for each</span>
<a name="l00414"></a>00414 <span class="comment">   * allele, residuals, etc.</span>
<a name="l00415"></a>00415 <span class="comment">   * </span>
<a name="l00416"></a>00416 <span class="comment">   * @param reporter - the object that will output all summary values.</span>
<a name="l00417"></a>00417 <span class="comment">   */</span>
<a name="l00418"></a><a class="code" href="classQuantLabelZ.html#a81cb59bbc6e9b6ba66c18f67fa2d20e4">00418</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a81cb59bbc6e9b6ba66c18f67fa2d20e4" title="Add an expression reporter to output the summary values for each allele, residuals, etc.">addExprReporter</a>(<a class="code" href="classQuantMethodReport.html" title="Class for reporting results of quantification methods.">QuantMethodReport</a> *reporter) {
<a name="l00419"></a>00419     <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>.push_back(reporter);
<a name="l00420"></a>00420   }
<a name="l00421"></a>00421 <span class="comment"></span>
<a name="l00422"></a>00422 <span class="comment">  /** </span>
<a name="l00423"></a>00423 <span class="comment">   * Get the BRLMM-like parameters used for this run.</span>
<a name="l00424"></a>00424 <span class="comment">   */</span> 
<a name="l00425"></a><a class="code" href="classQuantLabelZ.html#a3f29598c97ae984284fc9b4dfd0b5df6">00425</a>   <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html" title="What transformation of intensity estimates to feed into classifier.">BrlmmParam</a> <a class="code" href="classQuantLabelZ.html#a3f29598c97ae984284fc9b4dfd0b5df6" title="Get the BRLMM-like parameters used for this run.">getParam</a>() {
<a name="l00426"></a>00426     <span class="keywordflow">return</span> <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>;
<a name="l00427"></a>00427   }<span class="comment"></span>
<a name="l00428"></a>00428 <span class="comment">  /**</span>
<a name="l00429"></a>00429 <span class="comment">   * Set the snp-parameters for this run.</span>
<a name="l00430"></a>00430 <span class="comment">   */</span>
<a name="l00431"></a><a class="code" href="classQuantLabelZ.html#aac7e6719fad7743dd468cad018f621c2">00431</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aac7e6719fad7743dd468cad018f621c2" title="Set the snp-parameters for this run.">setLabelParam</a>(<a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> s){
<a name="l00432"></a>00432     <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>.<a class="code" href="classsnp__param.html#a73c1d4ba7af3924f444007aaa2e802ce" title="copy parameters">copy</a>(s);
<a name="l00433"></a>00433 
<a name="l00434"></a>00434     <span class="comment">// keep our options in sync so that the state is</span>
<a name="l00435"></a>00435     <span class="comment">// correctly reported</span>
<a name="l00436"></a>00436     s.setDocValues(*<span class="keyword">this</span>);
<a name="l00437"></a>00437   }
<a name="l00438"></a>00438 <span class="comment"></span>
<a name="l00439"></a>00439 <span class="comment">  /**</span>
<a name="l00440"></a>00440 <span class="comment">   * Get the parameters for this run of brlmm-p.</span>
<a name="l00441"></a>00441 <span class="comment">   */</span>
<a name="l00442"></a><a class="code" href="classQuantLabelZ.html#a8b7b8a2ea154e2e3169408e0500d5be4">00442</a>   <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> <a class="code" href="classQuantLabelZ.html#a8b7b8a2ea154e2e3169408e0500d5be4" title="Get the parameters for this run of brlmm-p.">getLabelParam</a>(){
<a name="l00443"></a>00443     <span class="keywordflow">return</span> <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>;
<a name="l00444"></a>00444   }
<a name="l00445"></a>00445 <span class="comment"></span>
<a name="l00446"></a>00446 <span class="comment">  /**</span>
<a name="l00447"></a>00447 <span class="comment">   * makes the normalization functions for all samples</span>
<a name="l00448"></a>00448 <span class="comment">   *</span>
<a name="l00449"></a>00449 <span class="comment">   * @param probeSets - my sample of snps to determine normalization function by intensity</span>
<a name="l00450"></a>00450 <span class="comment">   * @param layout - finding those probes on the chip</span>
<a name="l00451"></a>00451 <span class="comment">   * @param iMart - raw intensities</span>
<a name="l00452"></a>00452 <span class="comment">   * @param iTrans - transformations</span>
<a name="l00453"></a>00453 <span class="comment">   * @param pmAdjust - adjustments, such as background subtraction</span>
<a name="l00454"></a>00454 <span class="comment">   */</span>
<a name="l00455"></a>00455   vector&lt;NormalizationPredictor&gt; <a class="code" href="classQuantLabelZ.html#a8e0d8c1489dd615f81feec9f094d2aeb" title="makes the normalization functions for all samples">makeNorm</a>(std::vector&lt;std::string&gt; probeSets, <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> &amp;layout,
<a name="l00456"></a>00456                                           <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, std::vector&lt;ChipStream *&gt; &amp;iTrans,
<a name="l00457"></a>00457                                           <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust); 
<a name="l00458"></a>00458 <span class="comment"></span>
<a name="l00459"></a>00459 <span class="comment">  /**</span>
<a name="l00460"></a>00460 <span class="comment">   * sets the normalization functions for all samples</span>
<a name="l00461"></a>00461 <span class="comment">   * </span>
<a name="l00462"></a>00462 <span class="comment">   * @param Qmap - normalization functions</span>
<a name="l00463"></a>00463 <span class="comment">   */</span>
<a name="l00464"></a><a class="code" href="classQuantLabelZ.html#a3b8e9682bc1b43ef80676f1eb9556d87">00464</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a3b8e9682bc1b43ef80676f1eb9556d87" title="sets the normalization functions for all samples">setNorm</a>(vector&lt;NormalizationPredictor&gt; &amp;Qmap)
<a name="l00465"></a>00465   {
<a name="l00466"></a>00466     normMap.resize(Qmap.size());
<a name="l00467"></a>00467     <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i=0; i&lt;normMap.size(); i++)
<a name="l00468"></a>00468     {
<a name="l00469"></a>00469       normMap[i].Copy(Qmap[i]);
<a name="l00470"></a>00470     }
<a name="l00471"></a>00471   }
<a name="l00472"></a>00472 <span class="comment"></span>
<a name="l00473"></a>00473 <span class="comment">  /** returns the max-score used for this analysis method */</span>
<a name="l00474"></a><a class="code" href="classQuantLabelZ.html#ae00d3a3a41c2f6d35b3bff848c892116">00474</a>   <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ.html#ae00d3a3a41c2f6d35b3bff848c892116" title="returns the max-score used for this analysis method">getMaxScore</a>(){ <span class="keywordflow">return</span>(LabelzMaxScore);};<span class="comment"></span>
<a name="l00475"></a>00475 <span class="comment">  /** sets the max-score used for this analysis method */</span>
<a name="l00476"></a><a class="code" href="classQuantLabelZ.html#ac97957acc212aebea23d9f2fbdeb1702">00476</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ac97957acc212aebea23d9f2fbdeb1702" title="sets the max-score used for this analysis method">setMaxScore</a>(<span class="keywordtype">double</span> v){
<a name="l00477"></a>00477     LabelzMaxScore=v;
<a name="l00478"></a>00478 
<a name="l00479"></a>00479     <span class="comment">// keep our options in sync so that the state is</span>
<a name="l00480"></a>00480     <span class="comment">// correctly reported</span>
<a name="l00481"></a>00481     <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;MS&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(LabelzMaxScore));
<a name="l00482"></a>00482   };
<a name="l00483"></a>00483 <span class="comment"></span>
<a name="l00484"></a>00484 <span class="comment">  /**</span>
<a name="l00485"></a>00485 <span class="comment">   * which snps are haploid snps</span>
<a name="l00486"></a>00486 <span class="comment">   *</span>
<a name="l00487"></a>00487 <span class="comment">   * @param haploidSnps - map of string to chrX status </span>
<a name="l00488"></a>00488 <span class="comment">   */</span>
<a name="l00489"></a><a class="code" href="classQuantLabelZ.html#a5afceb726f32320451aeb0c5a334a92c">00489</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a5afceb726f32320451aeb0c5a334a92c" title="which snps are haploid snps">setHaploidSnps</a>(std::map&lt;std::string,bool&gt; &amp;haploidSnps) {
<a name="l00490"></a>00490     m_HaploidSnps = haploidSnps;
<a name="l00491"></a>00491   }
<a name="l00492"></a>00492 <span class="comment"></span>
<a name="l00493"></a>00493 <span class="comment">  /**</span>
<a name="l00494"></a>00494 <span class="comment">   * which snps are special</span>
<a name="l00495"></a>00495 <span class="comment">   *</span>
<a name="l00496"></a>00496 <span class="comment">   * @ param SpecialSnps - map of string to copy status</span>
<a name="l00497"></a>00497 <span class="comment">   */</span>
<a name="l00498"></a><a class="code" href="classQuantLabelZ.html#ac14724ad579f92836d1a6c779b9b396b">00498</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ac14724ad579f92836d1a6c779b9b396b" title="which snps are special">setSpecialSnps</a>(std::map&lt;std::string,std::pair&lt;int, int&gt; &gt; &amp;SpecialSnps){
<a name="l00499"></a>00499     <a class="code" href="classQuantLabelZ.html#acc7830f7121d9a30d4ff7f38acb04ddb" title="index of special snps and their copy numbers">m_SpecialSnps</a> = SpecialSnps;
<a name="l00500"></a>00500   }
<a name="l00501"></a>00501 <span class="comment"></span>
<a name="l00502"></a>00502 <span class="comment">  /**</span>
<a name="l00503"></a>00503 <span class="comment">   * which snps are special and sample-specific</span>
<a name="l00504"></a>00504 <span class="comment">   *</span>
<a name="l00505"></a>00505 <span class="comment">   * @ param SpecialSampleSnps - map of string to vector of copy status by sample</span>
<a name="l00506"></a>00506 <span class="comment">   */</span>
<a name="l00507"></a><a class="code" href="classQuantLabelZ.html#a7e75ab7ac471019f5aebde93a21d58b1">00507</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a7e75ab7ac471019f5aebde93a21d58b1" title="which snps are special and sample-specific">setSpecialSampleSnps</a>(std::map&lt;std::string,std::vector&lt;int&gt; &gt; &amp;SpecialSampleSnps,
<a name="l00508"></a>00508                             std::vector&lt;std::string&gt; &amp;SpecialSampleNames){
<a name="l00509"></a>00509     m_SpecialSampleSnps = SpecialSampleSnps;
<a name="l00510"></a>00510     m_SpecialSampleNames = SpecialSampleNames;
<a name="l00511"></a>00511   }
<a name="l00512"></a>00512 <span class="comment"></span>
<a name="l00513"></a>00513 <span class="comment">  /**</span>
<a name="l00514"></a>00514 <span class="comment">   * sets the genders for all samples</span>
<a name="l00515"></a>00515 <span class="comment">   *</span>
<a name="l00516"></a>00516 <span class="comment">   * @param genders - gender vector describing samples</span>
<a name="l00517"></a>00517 <span class="comment">   */</span>
<a name="l00518"></a><a class="code" href="classQuantLabelZ.html#a75d2851d0da897d336e42010653fbb4e">00518</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a75d2851d0da897d336e42010653fbb4e" title="sets the genders for all samples">setGenders</a>(<span class="keyword">const</span> std::vector&lt;affx::Gender&gt; &amp;genders) {
<a name="l00519"></a>00519     <a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983" title="What gender is each sample?">m_Genders</a> = genders;
<a name="l00520"></a>00520   }
<a name="l00521"></a>00521 
<a name="l00522"></a>00522   <span class="keywordtype">void</span> setInbredHetPenalty(<span class="keyword">const</span> std::vector&lt;double&gt; &amp;InbredPenalty){
<a name="l00523"></a>00523           m_InbredHetPenalty = InbredPenalty;
<a name="l00524"></a>00524    }
<a name="l00525"></a>00525 <span class="comment"></span>
<a name="l00526"></a>00526 <span class="comment">  /**</span>
<a name="l00527"></a>00527 <span class="comment">   *  Extracts intensities, etc for a probe set, so further analysis can proceed</span>
<a name="l00528"></a>00528 <span class="comment">   *</span>
<a name="l00529"></a>00529 <span class="comment">   *  @param gtPs - pointer to probe set</span>
<a name="l00530"></a>00530 <span class="comment">   * @param layout - layout of probesets on array</span>
<a name="l00531"></a>00531 <span class="comment">   * @param iMart - method to grab raw intensities and serve them</span>
<a name="l00532"></a>00532 <span class="comment">   * @param iTrans - transformations of intensities</span>
<a name="l00533"></a>00533 <span class="comment">   * @param pmAdjust - background subtraction, etc</span>
<a name="l00534"></a>00534 <span class="comment">   * @param doReport - do we invoke the output reporters for summary values</span>
<a name="l00535"></a>00535 <span class="comment">   */</span>
<a name="l00536"></a>00536   <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a6c51a62f1bcd9d8d9b357bc9db118620" title="Extracts intensities, etc for a probe set, so further analysis can proceed.">SetUpProbeSet</a>(<span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs,
<a name="l00537"></a>00537                      <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, std::vector&lt;ChipStream *&gt; &amp;iTrans, 
<a name="l00538"></a>00538                      <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust,<span class="keywordtype">bool</span> doReport);
<a name="l00539"></a>00539 
<a name="l00540"></a>00540   <span class="keywordtype">void</span> SubsetSetup(vector&lt;int&gt; &amp;samplesubset, <span class="keywordtype">int</span> &amp;subsetlength, <span class="keywordtype">int</span> copyIx, vector&lt;int&gt;&amp; SnpSampleCopyNumber);
<a name="l00541"></a>00541 <span class="comment"></span>
<a name="l00542"></a>00542 <span class="comment">  /**</span>
<a name="l00543"></a>00543 <span class="comment">   * Determines the gender of samples from chrX snps by EM.</span>
<a name="l00544"></a>00544 <span class="comment">   * </span>
<a name="l00545"></a>00545 <span class="comment">   * @param genders - vector of genders for samples that this routine fills in</span>
<a name="l00546"></a>00546 <span class="comment">   * @param probeSets - sample of snps that happen to be chrX</span>
<a name="l00547"></a>00547 <span class="comment">   * @param layout - layout of probe sets on array</span>
<a name="l00548"></a>00548 <span class="comment">   * @param iMart - serves up raw intensities</span>
<a name="l00549"></a>00549 <span class="comment">   * @param iTrans - modifies the raw intensities</span>
<a name="l00550"></a>00550 <span class="comment">   * @param pmAdjuster - subtracts background</span>
<a name="l00551"></a>00551 <span class="comment">   */</span>
<a name="l00552"></a>00552   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a384472516ce59e10b120bf84eff95345" title="Determines the gender of samples from chrX snps by EM.">FindGender</a>(std::vector&lt;affx::Gender&gt; &amp;genders, std::vector&lt;ProbeSet *&gt; probeSets, 
<a name="l00553"></a>00553                     <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> &amp;layout, <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, std::vector&lt;ChipStream *&gt; &amp;iTrans, 
<a name="l00554"></a>00554                     <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust);
<a name="l00555"></a>00555 <span class="comment"></span>
<a name="l00556"></a>00556 <span class="comment">  /** </span>
<a name="l00557"></a>00557 <span class="comment">   * Set the initial guesses of genotypes, usually from DM predictions.</span>
<a name="l00558"></a>00558 <span class="comment">   * </span>
<a name="l00559"></a>00559 <span class="comment">   * @param genotypes vector of genotypes AA=0, AB=1, AB=2, NN=3,</span>
<a name="l00560"></a>00560 <span class="comment">   */</span>
<a name="l00561"></a><a class="code" href="classQuantLabelZ.html#a22b847bd3dff1573d18dc4f248ddd509">00561</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a22b847bd3dff1573d18dc4f248ddd509" title="Set the initial guesses of genotypes, usually from DM predictions.">setKnownGenoTypes</a>(std::map&lt;<span class="keyword">const</span> <span class="keywordtype">char</span> *,std::vector&lt;affx::GType&gt;, <a class="code" href="structUtil_1_1ltstr.html" title="Comparison object for use in map for char *.">Util::ltstr</a>&gt; *genotypes) {
<a name="l00562"></a>00562     m_KnownGenoTypes = genotypes;
<a name="l00563"></a>00563   }
<a name="l00564"></a>00564 <span class="comment"></span>
<a name="l00565"></a>00565 <span class="comment">  /**</span>
<a name="l00566"></a>00566 <span class="comment">   * Sets the EM specific parameters for calling genders from chrX snps</span>
<a name="l00567"></a>00567 <span class="comment">   *</span>
<a name="l00568"></a>00568 <span class="comment">   * @param em_t - threshold for em</span>
<a name="l00569"></a>00569 <span class="comment">   * @param em_c - cutoff for em</span>
<a name="l00570"></a>00570 <span class="comment">   * @param g_c - cutoff for choosing genders</span>
<a name="l00571"></a>00571 <span class="comment">   */</span>
<a name="l00572"></a><a class="code" href="classQuantLabelZ.html#acaef334285a1263ca32c6070f620a184">00572</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#acaef334285a1263ca32c6070f620a184" title="Sets the EM specific parameters for calling genders from chrX snps.">setGenderParam</a>(<span class="keywordtype">double</span> em_t,<span class="keywordtype">double</span> em_c,<span class="keywordtype">double</span> g_c)
<a name="l00573"></a>00573   {
<a name="l00574"></a>00574     em_thresh = em_t;
<a name="l00575"></a>00575     em_cutoff = em_c;
<a name="l00576"></a>00576     gender_cutoff = g_c;
<a name="l00577"></a>00577 
<a name="l00578"></a>00578     <span class="comment">// keep our options in sync so that the state is</span>
<a name="l00579"></a>00579     <span class="comment">// correctly reported</span>
<a name="l00580"></a>00580     <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;em_thresh&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(em_thresh));
<a name="l00581"></a>00581     <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;em_cutoff&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(em_cutoff));
<a name="l00582"></a>00582     <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;gender_cutoff&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(gender_cutoff));
<a name="l00583"></a>00583   }
<a name="l00584"></a>00584 <span class="comment"></span>
<a name="l00585"></a>00585 <span class="comment">  /**</span>
<a name="l00586"></a>00586 <span class="comment">   * Selects good probes</span>
<a name="l00587"></a>00587 <span class="comment">   * </span>
<a name="l00588"></a>00588 <span class="comment">   */</span>
<a name="l00589"></a>00589   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aee903e230d96abad58c24118b312b100" title="Selects good probes.">SelectGoodProbes</a>();
<a name="l00590"></a>00590 <span class="comment"></span>
<a name="l00591"></a>00591 <span class="comment">  /**</span>
<a name="l00592"></a>00592 <span class="comment">   * Load up the initial call for the requested SNP</span>
<a name="l00593"></a>00593 <span class="comment">   * </span>
<a name="l00594"></a>00594 <span class="comment">   * @param probeset - name of the probeset for which to load known genotypes</span>
<a name="l00595"></a>00595 <span class="comment">   */</span>
<a name="l00596"></a>00596   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aa0bb7cf31d9f77d47947cf91d2fa75e0" title="Load up the initial call for the requested SNP.">loadInitialCallsForSnp</a>(<span class="keyword">const</span> <span class="keywordtype">string</span> &amp;probeset);
<a name="l00597"></a>00597 
<a name="l00598"></a>00598   <span class="keyword">typedef</span> std::map&lt;const char *,std::vector&lt;affx::GType&gt;, <a class="code" href="structUtil_1_1ltstr.html" title="Comparison object for use in map for char *.">Util::ltstr</a> &gt;  KnownGenoTypes_t;
<a name="l00599"></a>00599   <span class="keyword">typedef</span> KnownGenoTypes_t::iterator KnownGenoTypes_iter_t;
<a name="l00600"></a>00600 
<a name="l00601"></a><a class="code" href="classQuantLabelZ.html#af1fedebd9a757f82da747b717154121f">00601</a>   <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ.html#af1fedebd9a757f82da747b717154121f" title="Get the min value for a call.">getMinThresh</a>() { <span class="keywordflow">return</span> -DBL_MAX;}
<a name="l00602"></a><a class="code" href="classQuantLabelZ.html#a15e614024267b86c0ad74281a6d851e1">00602</a>   <span class="keywordtype">double</span> <a class="code" href="classQuantLabelZ.html#a15e614024267b86c0ad74281a6d851e1" title="Get the max value for a call.">getMaxThresh</a>() { <span class="keywordflow">return</span> LabelzMaxScore;}
<a name="l00603"></a>00603   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a1c8b961309e6485648634d5884e6bd60" title="Get set up for a run of reporting probesets.">prepare</a>(<span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart);
<a name="l00604"></a>00604   <span class="keyword">virtual</span> <span class="keywordtype">bool</span> <a class="code" href="classQuantLabelZ.html#a259ef2cb352e7e9911bff1e0efbf3c13" title="No more probesets will be processed, this is a chance to finish outputting results and clean up...">finish</a>();
<a name="l00605"></a>00605 
<a name="l00606"></a>00606   <span class="keywordtype">void</span> setCopyNumber(<span class="keywordtype">bool</span> b) {
<a name="l00607"></a>00607     m_bCopyNumber = b;
<a name="l00608"></a>00608     <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;copynumber&quot;</span>,<a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(b));
<a name="l00609"></a>00609   }
<a name="l00610"></a>00610 
<a name="l00611"></a>00611   <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a9bac1828828da05da22eacadab7a322b" title="Custom configuration for this QuantMethod.">setParameters</a>(<a class="code" href="classPsBoard.html">PsBoard</a> &amp;board);
<a name="l00612"></a>00612 
<a name="l00613"></a>00613   <span class="comment">// ArtifactReduction</span>
<a name="l00614"></a>00614   <span class="comment">//do a dummy trick which will explode memory painfully</span>
<a name="l00615"></a>00615   <span class="comment">//avert by loading in only probeset-id&#39;s that are relevant?</span>
<a name="l00616"></a>00616   <span class="comment">//</span>
<a name="l00617"></a>00617   <span class="comment">// given a sample:probeset:block trio, tell me the confidence</span>
<a name="l00618"></a>00618   std::map&lt;string, int&gt; m_ProbeSetTrust;  
<a name="l00619"></a>00619   <span class="keywordtype">void</span> EliminateUntrustedSamples(vector&lt; int &gt;&amp;SnpSampleCopyNumber);
<a name="l00620"></a>00620   <span class="keywordtype">void</span> loadProbeSetTrustFromTmpFile();
<a name="l00621"></a>00621 <span class="comment"></span>
<a name="l00622"></a>00622 <span class="comment">  /**</span>
<a name="l00623"></a>00623 <span class="comment">   * Set TrustCheck flag</span>
<a name="l00624"></a>00624 <span class="comment">   * </span>
<a name="l00625"></a>00625 <span class="comment">   * @param tc_flag - indicates whether to look for trusted probe file from</span>
<a name="l00626"></a>00626 <span class="comment">   * ArtifactReduction</span>
<a name="l00627"></a>00627 <span class="comment">   */</span>
<a name="l00628"></a><a class="code" href="classQuantLabelZ.html#aad45cb95c07349cdfb3e80c88b068679">00628</a>   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aad45cb95c07349cdfb3e80c88b068679" title="Set TrustCheck flag.">setTrustCheck</a>(<span class="keywordtype">bool</span> tc_flag) {
<a name="l00629"></a>00629       m_bProbeSetTrustOpt = tc_flag;
<a name="l00630"></a>00630       <a class="code" href="classSelfDoc.html#a00cf96accbd1ca4a3eec8ea947cde382" title="Look up option with name and set it to value.">setOptValue</a>(<span class="stringliteral">&quot;TrustCheck&quot;</span>, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(tc_flag));
<a name="l00631"></a>00631   }
<a name="l00632"></a>00632 
<a name="l00633"></a>00633 <span class="keyword">protected</span>:
<a name="l00634"></a>00634   std::string m_SnpPosteriorName;
<a name="l00635"></a>00635   affx::TsvReport::TsvReportFmt_t m_PosteriorFormat;
<a name="l00636"></a>00636   std::string m_SequentialModelFile;
<a name="l00637"></a>00637   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> m_SequentialModelTsv;
<a name="l00638"></a>00638   std::string m_Id, m_BB, m_AB, m_AA, m_CV;
<a name="l00639"></a>00639   <span class="keywordtype">bool</span> m_bCopyNumber; <span class="comment">/// This meethod is begin called as part of a copynumber run.</span>
<a name="l00640"></a><a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750">00640</a> <span class="comment"></span>  <a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *<a class="code" href="classQuantLabelZ.html#a80bc774890dba758e37221c24c6f0750" title="This meethod is begin called as part of a copynumber run.">m_QuantMethod</a>; <span class="comment">///&lt; Quantification method used for summarizing snps.</span>
<a name="l00641"></a><a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412">00641</a> <span class="comment"></span>  <a class="code" href="classQuantLabelZ_1_1BrlmmParam.html" title="What transformation of intensity estimates to feed into classifier.">BrlmmParam</a> <a class="code" href="classQuantLabelZ.html#a7d599ef51c95cf54041f258872ce6412" title="Various parameters used by the brlmm algorithm.">m_Param</a>; <span class="comment">///&lt; Various parameters used by the brlmm algorithm.</span>
<a name="l00642"></a><a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92">00642</a> <span class="comment"></span>  std::vector&lt;QuantMethodReport *&gt; <a class="code" href="classQuantLabelZ.html#a159af6fa4e27c546d5f8fe0c1f8d9b92" title="Report object for outputting results.">m_Reporters</a>;   <span class="comment">///&lt; Report object for outputting results</span>
<a name="l00643"></a>00643 <span class="comment"></span>  <span class="comment">//  std::map&lt;std::string, snp_param&gt; m_SnpPriorMap; ///&lt; customized priors indexed by probeset name</span>
<a name="l00644"></a>00644   <span class="comment">//AffxArray&lt;snp_posterior&gt; m_vectSnpPriors;</span>
<a name="l00645"></a><a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be">00645</a>   <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> <a class="code" href="classQuantLabelZ.html#ac7ffc86985526812b186af719bbc02be" title="parameters used by labeling">sp</a>; <span class="comment">///&lt; parameters used by labeling</span>
<a name="l00646"></a><a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a">00646</a> <span class="comment"></span>  std::vector&lt;double&gt; <a class="code" href="classQuantLabelZ.html#a62471aecfb8fb5b0afb172c3413f795a" title="Our resulting confidences in those calls.">m_Confidences</a>; <span class="comment">///&lt; Our resulting confidences in those calls.</span>
<a name="l00647"></a>00647 <span class="comment"></span>  <a class="code" href="classAffxArray.html">AffxArray&lt;snp_labeled_distribution&gt;</a> m_vectSnpPriors;
<a name="l00648"></a><a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6">00648</a>   std::vector&lt;std::vector&lt;double&gt; &gt; <a class="code" href="classQuantLabelZ.html#a7e5e595fb3d2bfe39f6e8973ab56a3e6" title="standardized distances from AA, AB, and BB cluster centers">m_Distances</a>; <span class="comment">///&lt; standardized distances from AA, AB, and BB cluster centers</span>
<a name="l00649"></a><a class="code" href="classQuantLabelZ.html#acc7830f7121d9a30d4ff7f38acb04ddb">00649</a> <span class="comment"></span>  std::map&lt;std::string,std::pair&lt;int, int&gt; &gt; <a class="code" href="classQuantLabelZ.html#acc7830f7121d9a30d4ff7f38acb04ddb" title="index of special snps and their copy numbers">m_SpecialSnps</a>; <span class="comment">///&lt; index of special snps and their copy numbers</span>
<a name="l00650"></a><a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983">00650</a> <span class="comment"></span>  std::vector&lt;affx::Gender&gt; <a class="code" href="classQuantLabelZ.html#affaf5b9a057a6c13d946d21ab95cf983" title="What gender is each sample?">m_Genders</a>;     <span class="comment">///&lt; What gender is each sample?</span>
<a name="l00651"></a>00651 <span class="comment"></span>  std::set&lt;const char *, Util::ltstr&gt; *m_ProbeSetsToReport;
<a name="l00652"></a>00652 
<a name="l00653"></a>00653 <span class="keyword">private</span>:
<a name="l00654"></a>00654 
<a name="l00655"></a>00655   <span class="keywordtype">double</span> LabelzMaxScore; <span class="comment">///&lt; parameter used for making no-calls when doing LabelZ</span>
<a name="l00656"></a>00656 <span class="comment"></span>
<a name="l00657"></a>00657   <span class="comment">// goofy I/O stuff</span>
<a name="l00658"></a>00658   <span class="comment">// read &gt;in&lt; prior</span>
<a name="l00659"></a>00659   <span class="comment">// process data </span>
<a name="l00660"></a>00660   <span class="comment">// &gt;output&lt; posterior which contains SNP specific tuning</span>
<a name="l00661"></a>00661   <span class="comment">// when you read it in, it&#39;s a prior</span>
<a name="l00662"></a>00662   <span class="comment">// when you output it, it&#39;s a posterior</span>
<a name="l00663"></a>00663 
<a name="l00664"></a>00664   <span class="comment">// normalization map</span>
<a name="l00665"></a>00665   <span class="comment">// vector of normalization predictors</span>
<a name="l00666"></a>00666   <span class="comment">// one per sample</span>
<a name="l00667"></a>00667   vector&lt; NormalizationPredictor&gt; normMap; <span class="comment">///&lt; vector of normalization functions per sample</span>
<a name="l00668"></a>00668 <span class="comment"></span>
<a name="l00669"></a>00669   std::map&lt;std::string, vector&lt;double&gt; &gt; m_SnpCovarMap; <span class="comment">///&lt; customized covariates indexed by probeset name</span>
<a name="l00670"></a>00670 <span class="comment"></span>
<a name="l00671"></a>00671   <span class="comment">// handle haploid Snps and gender issues</span>
<a name="l00672"></a>00672   std::map&lt;std::string,bool&gt; m_HaploidSnps; <span class="comment">///&lt; Index of probesets that are in a haploid regions (usually X SNPs).</span>
<a name="l00673"></a>00673 <span class="comment"></span>  std::map&lt;std::string,std::vector&lt;int&gt; &gt; m_SpecialSampleSnps; <span class="comment">///&lt; index of special sample-specific snps and their copy numbers</span>
<a name="l00674"></a>00674 <span class="comment"></span>  std::vector&lt;std::string&gt; m_SpecialSampleNames; <span class="comment">///&lt; index of sample names for special sample-specific snps</span>
<a name="l00675"></a>00675 <span class="comment"></span>  <span class="keywordtype">double</span> em_thresh, em_cutoff,gender_cutoff; <span class="comment">///&lt; three parameters for calling genders</span>
<a name="l00676"></a>00676 <span class="comment"></span>
<a name="l00677"></a>00677   <span class="comment">// this is used as &quot;hints&quot; for generating the likelihood</span>
<a name="l00678"></a>00678   <span class="comment">// these are not &gt;necessary&lt; as for BRLMM</span>
<a name="l00679"></a>00679   <span class="comment">// but can be &gt;helpful&lt; in evaluating likelihood or generating specific priors</span>
<a name="l00680"></a>00680   <span class="comment">// i.e. work from partial data to classify full data</span>
<a name="l00681"></a>00681   KnownGenoTypes_t *m_KnownGenoTypes; <span class="comment">///&lt; Map from probeset names to the appopriate m_InitialCalls;</span>
<a name="l00682"></a>00682 <span class="comment"></span>  std::vector&lt;affx::GType&gt; m_InitialCalls; <span class="comment">///&lt; Genotype calls we are supplied with for learning.</span>
<a name="l00683"></a>00683 <span class="comment"></span>
<a name="l00684"></a>00684   <span class="comment">// Fix up for genotyping variably inbred samples</span>
<a name="l00685"></a>00685   <span class="comment">// likelihood penalty for deviating from normal sample behavior</span>
<a name="l00686"></a>00686   <span class="comment">// positive for penalizing hets, negative for increasing hets</span>
<a name="l00687"></a>00687   std::vector&lt;double&gt; m_InbredHetPenalty; <span class="comment">///&lt; Sample specific covariate like gender, but affecting all SNPs</span>
<a name="l00688"></a>00688 <span class="comment"></span>
<a name="l00689"></a>00689   <span class="comment">// store the individual probe data for selection</span>
<a name="l00690"></a>00690   std::vector&lt;std::vector&lt;double&gt; &gt; m_PM_A; <span class="comment">///&lt; individual intensities A</span>
<a name="l00691"></a>00691 <span class="comment"></span>  std::vector&lt;std::vector&lt;double&gt; &gt; m_PM_B; <span class="comment">///&lt; individual intensities B</span>
<a name="l00692"></a>00692 <span class="comment"></span>  std::vector&lt; unsigned int &gt; m_A_probeId; <span class="comment">///&lt; individual probe indexes A</span>
<a name="l00693"></a>00693 <span class="comment"></span>  std::vector&lt; unsigned int &gt; m_B_probeId; <span class="comment">///&lt; individual probe indexes B</span>
<a name="l00694"></a>00694 <span class="comment"></span>  <span class="keywordtype">int</span> m_SelectionOverride; <span class="comment">///&lt; override calls with references for selection</span>
<a name="l00695"></a>00695 <span class="comment"></span>
<a name="l00696"></a>00696 <span class="keyword">public</span>:
<a name="l00697"></a>00697 
<a name="l00698"></a>00698   <span class="comment">// Artifact Reduction</span>
<a name="l00699"></a>00699   <span class="keywordtype">bool</span> isTrusted(<span class="keywordtype">int</span>);
<a name="l00700"></a>00700   <span class="keywordtype">void</span> EliminateUntrustedSamples();
<a name="l00701"></a>00701   <span class="keywordtype">void</span> readProbesetTrust(<span class="keyword">const</span> std::string &amp;fileName);
<a name="l00702"></a>00702   std::string m_ProbeSetTrustTmpFileName;
<a name="l00703"></a>00703   <span class="keywordtype">bool</span> m_bProbeSetTrustRead;
<a name="l00704"></a>00704   <span class="keywordtype">bool</span> m_bProbeSetTrustOpt;
<a name="l00705"></a>00705   <span class="keywordtype">bool</span> m_ZWGenderCalling;<span class="comment"></span>
<a name="l00706"></a>00706 <span class="comment">  ///&lt; </span>
<a name="l00707"></a>00707 <span class="comment"></span>  <a class="code" href="classaffx_1_1TsvReport.html" title="Unifies the presentation of file/TsvFile and file ///.">affx::TsvReport</a> m_SnpPosteriorTsv;
<a name="l00708"></a>00708 
<a name="l00709"></a>00709   <span class="keywordtype">int</span> m_SnpPosteriorTsv_ver;<span class="comment"></span>
<a name="l00710"></a>00710 <span class="comment">  ///&lt;</span>
<a name="l00711"></a>00711 <span class="comment"></span>  <a class="code" href="classaffx_1_1TsvReport.html" title="Unifies the presentation of file/TsvFile and file ///.">affx::TsvReport</a> m_NormSummaryTsv;<span class="comment"></span>
<a name="l00712"></a>00712 <span class="comment">  ///&lt; </span>
<a name="l00713"></a>00713 <span class="comment"></span>  <a class="code" href="classaffx_1_1TsvReport.html" title="Unifies the presentation of file/TsvFile and file ///.">affx::TsvReport</a> m_SnpProbeTsv;
<a name="l00714"></a>00714 <span class="comment"></span>
<a name="l00715"></a>00715 <span class="comment">  /**</span>
<a name="l00716"></a>00716 <span class="comment">   * writes out the contrast normalization functions</span>
<a name="l00717"></a>00717 <span class="comment">   *</span>
<a name="l00718"></a>00718 <span class="comment">   * @param nMap - function for normalization</span>
<a name="l00719"></a>00719 <span class="comment">   * @param fileName - which file to write to</span>
<a name="l00720"></a>00720 <span class="comment">   */</span>
<a name="l00721"></a>00721   <span class="keyword">static</span> <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a12c40ec12ddbaabcd129750af201e5e8" title="writes out the contrast normalization functions">writeNormMap</a>(vector&lt;NormalizationPredictor&gt; &amp;nMap, 
<a name="l00722"></a>00722                     <span class="keyword">const</span> std::string&amp; fileName);
<a name="l00723"></a>00723 <span class="comment"></span>
<a name="l00724"></a>00724 <span class="comment">  ////</span>
<a name="l00725"></a>00725 <span class="comment"></span>  <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">writeSnpPosteriorTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l00726"></a>00726                             affx::TsvReport::TsvReportFmt_t format);<span class="comment"></span>
<a name="l00727"></a>00727 <span class="comment">  ///</span>
<a name="l00728"></a>00728 <span class="comment"></span>  <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a37af5d01565df9a913e6f93033a836d9" title="Pick the output version based on the file format. (txt=&amp;gt;v1,a5=&amp;gt;v2).">writeSnpPosteriorTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l00729"></a>00729                             affx::TsvReport::TsvReportFmt_t format,
<a name="l00730"></a>00730                             <span class="keywordtype">int</span> ver);<span class="comment"></span>
<a name="l00731"></a>00731 <span class="comment">  ///</span>
<a name="l00732"></a>00732 <span class="comment"></span>  <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a35793acf2ae3d85c79930cadb62beb7a" title="Writes one specific posterior to a file along with snp name.">writeSnpPosteriorValue</a>(<span class="keyword">const</span> std::string&amp; probeset_id, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;tsp);
<a name="l00733"></a>00733  
<a name="l00734"></a>00734   std::string getModelFile() {
<a name="l00735"></a>00735     <span class="keywordflow">return</span> m_SnpPosteriorTsv.getFilePath();
<a name="l00736"></a>00736   }
<a name="l00737"></a>00737 <span class="comment"></span>
<a name="l00738"></a>00738 <span class="comment">  /**</span>
<a name="l00739"></a>00739 <span class="comment">   * open the stream for snp-specific probe selection</span>
<a name="l00740"></a>00740 <span class="comment">   */</span>
<a name="l00741"></a>00741   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#aa30f8acb1fb4b2f0e07774461d31cdf0" title="open the stream for snp-specific probe selection">readSnpProbeTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l00742"></a>00742                        affx::TsvReport::TsvReportFmt_t format);
<a name="l00743"></a>00743   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ad7cae32c6830efa7475fb948f2442682" title="Write header to the selection file so people don&amp;#39;t have to look up everything.">writeSnpProbeTsv</a>(<span class="keyword">const</span> std::string&amp; fileName,
<a name="l00744"></a>00744                         affx::TsvReport::TsvReportFmt_t format);
<a name="l00745"></a>00745 
<a name="l00746"></a>00746   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a16617ac9ce4b9229c84c6f26af40cc55" title="Writeprobe writes the summaries to the file.">writeSnpProbeValue</a>(<a class="code" href="classaffx_1_1TsvReport.html" title="Unifies the presentation of file/TsvFile and file ///.">affx::TsvReport</a>&amp; tsv,
<a name="l00747"></a>00747                           <span class="keywordtype">int</span> probenum,
<a name="l00748"></a>00748                           <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> Aid,
<a name="l00749"></a>00749                           <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> Bid,
<a name="l00750"></a>00750                           <span class="keyword">const</span> vector&lt;double&gt; &amp;Summaries,
<a name="l00751"></a>00751                           <span class="keyword">const</span> vector&lt;double&gt; &amp;m_Ax,
<a name="l00752"></a>00752                           <span class="keyword">const</span> vector&lt;double&gt; &amp;m_Bx,
<a name="l00753"></a>00753                           <span class="keywordtype">string</span>&amp; m_ProbeSetName);
<a name="l00754"></a>00754 <span class="comment"></span>
<a name="l00755"></a>00755 <span class="comment">  /**</span>
<a name="l00756"></a>00756 <span class="comment">   * open file for normalized values</span>
<a name="l00757"></a>00757 <span class="comment">   */</span>
<a name="l00758"></a>00758   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a4ae4ada8dc508759ea8af9e83ee91841" title="open file for normalized values">readNormSummaryTsv</a>(<span class="keyword">const</span> std::string&amp; fileName, 
<a name="l00759"></a>00759                           std::vector&lt;std::string&gt; &amp;headers,
<a name="l00760"></a>00760                           affx::TsvReport::TsvReportFmt_t format);
<a name="l00761"></a>00761   <span class="keywordtype">void</span> writeNormSummaryTsv(<span class="keyword">const</span> std::string&amp; fileName, 
<a name="l00762"></a>00762                            std::vector&lt;std::string&gt; &amp;headers,
<a name="l00763"></a>00763                            affx::TsvReport::TsvReportFmt_t format);
<a name="l00764"></a>00764   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#a4f185569daf922ace4d429c9f5c6752b" title="Writes a summary of the transformed values to a file - i.e as set up for genotyping, not expression.">writeNormSummaryValue</a>(<span class="keyword">const</span> std::string&amp; m_ProbeSetName,
<a name="l00765"></a>00765                              <span class="keyword">const</span> std::vector&lt;double&gt; &amp;<a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>,
<a name="l00766"></a>00766                              <span class="keyword">const</span> std::vector&lt;double&gt; &amp;<a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>);<span class="comment"></span>
<a name="l00767"></a>00767 <span class="comment">  /**</span>
<a name="l00768"></a>00768 <span class="comment">   * read in the snp specific priors from a file</span>
<a name="l00769"></a>00769 <span class="comment">   */</span>
<a name="l00770"></a>00770   <span class="keywordtype">void</span> <a class="code" href="classQuantLabelZ.html#ab4a3970a1d381a0885d82ebc66ddb3ca" title="read in the snp specific priors from a file">readSnpPriorMap</a>(<span class="keyword">const</span> std::string&amp; fileName);
<a name="l00771"></a>00771   <span class="keywordtype">void</span> readSnpPriorMap_tsv5(<a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a>* tsv5);
<a name="l00772"></a>00772 
<a name="l00773"></a>00773   <span class="keyword">virtual</span> <span class="keywordtype">void</span> registerProbeSetsToReport(std::set&lt;const char *, Util::ltstr&gt; *probeSetsToReport) {
<a name="l00774"></a>00774     m_ProbeSetsToReport = probeSetsToReport;
<a name="l00775"></a>00775   }
<a name="l00776"></a>00776 
<a name="l00777"></a>00777 
<a name="l00778"></a>00778 };
<a name="l00779"></a>00779 
<a name="l00780"></a>00780 <span class="preprocessor">#endif </span><span class="comment">/* _QUANTLABELZ_H_ */</span>
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:53 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
