<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/QuantBirdseed.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>QuantBirdseed.cpp</h1>  </div>
</div>
<div class="contents">
<div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2007 The Broad Institute and Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">//</span>
<a name="l00023"></a>00023 <span class="preprocessor">#include &quot;chipstream/QuantBirdseed.h&quot;</span>
<a name="l00024"></a>00024 <span class="comment">//</span>
<a name="l00025"></a>00025 <span class="preprocessor">#include &quot;birdseed-dev/PriorsReader.h&quot;</span>
<a name="l00026"></a>00026 <span class="preprocessor">#include &quot;broadutil/BroadUtil.h&quot;</span>
<a name="l00027"></a>00027 <span class="comment">//</span>
<a name="l00028"></a>00028 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodFactory_8h.html" title="Factory class for making chip streams based on a string representation.">chipstream/QuantMethodFactory.h</a>&quot;</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="QuantMethodReport_8h.html" title="Class for reporting results of quantification methods.">chipstream/QuantMethodReport.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;chipstream/GenotypeInfo.h&quot;</span>
<a name="l00031"></a>00031 <span class="comment">//</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 <span class="comment"></span>
<a name="l00035"></a>00035 <span class="comment">/** Constructor, currently creates prior estimates from pre-calculated data from R. */</span>
<a name="l00036"></a><a class="code" href="classQuantBirdseed.html#a461fed3d1ffc2b4791165f6c17deb586">00036</a> <a class="code" href="classQuantBirdseed.html#a461fed3d1ffc2b4791165f6c17deb586" title="Constructor, currently creates prior estimates from pre-calculated data from R.">QuantBirdseed::QuantBirdseed</a>(<span class="keywordtype">double</span> confidenceThreshold, <span class="keywordtype">double</span> correctionFactor):
<a name="l00037"></a>00037   m_ConfidenceThreshold(confidenceThreshold),
<a name="l00038"></a>00038   m_CorrectionFactor(correctionFactor),
<a name="l00039"></a>00039   verbosity(0)
<a name="l00040"></a>00040 {
<a name="l00041"></a>00041   <a class="code" href="classQuantBirdseed.html#a1dab8ffb3062eefd56689f8f0ee23efb" title="Quantification method used for summarizing snps.">m_QuantMethod</a> = NULL;
<a name="l00042"></a>00042 }
<a name="l00043"></a>00043 
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 QuantBirdseed::~QuantBirdseed()
<a name="l00046"></a>00046 {
<a name="l00047"></a>00047         <span class="keywordflow">for</span> (std::vector&lt;QuantMethodReport *&gt;::iterator it = <a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>.begin(); it != <a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>.end(); it++)
<a name="l00048"></a>00048     {
<a name="l00049"></a>00049                 <span class="keyword">delete</span> *it;
<a name="l00050"></a>00050         }
<a name="l00051"></a>00051         <a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>.clear();
<a name="l00052"></a>00052     <span class="keywordflow">if</span> (m_ClusterOstrm.get() != NULL) {
<a name="l00053"></a>00053         m_ClusterOstrm-&gt;close();
<a name="l00054"></a>00054     }
<a name="l00055"></a>00055 }
<a name="l00056"></a>00056 
<a name="l00057"></a>00057 
<a name="l00058"></a>00058 <span class="comment">// Copied from QuantBRLMM</span>
<a name="l00059"></a>00059 <span class="keywordtype">void</span> QuantBirdseed::blankSelf() {
<a name="l00060"></a>00060     <a class="code" href="classQuantGTypeMethod.html#aff29fdf8763f00f5e1eb451162cb38c2" title="clear a probe set">clearProbeSet</a>(<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>);
<a name="l00061"></a>00061     <a class="code" href="classQuantGTypeMethod.html#aff29fdf8763f00f5e1eb451162cb38c2" title="clear a probe set">clearProbeSet</a>(<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00062"></a>00062     <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>.clear();
<a name="l00063"></a>00063     <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>.clear();
<a name="l00064"></a>00064     <span class="comment">//m_InitialCalls.clear();</span>
<a name="l00065"></a>00065     m_Calls.clear();
<a name="l00066"></a>00066     <a class="code" href="classQuantBirdseed.html#ae3975acac1f57ce585a126421aceb5dd" title="Our resulting confidences in those calls.">m_Confidences</a>.clear();
<a name="l00067"></a>00067     <a class="code" href="classQuantBirdseed.html#aa8d1cca089a1b962a26d35cac5ba4df8" title="Our resulting distances for AA, AB, and BB genotypes.">m_Distances</a>.clear();
<a name="l00068"></a>00068 };
<a name="l00069"></a>00069 
<a name="l00070"></a>00070 <span class="keywordtype">void</span> QuantBirdseed::setClusterOutFile(<span class="keyword">const</span> std::string&amp; clusterOutPath) {
<a name="l00071"></a>00071   m_ClusterOutFile = clusterOutPath;
<a name="l00072"></a>00072   m_ClusterOstrm.reset(<span class="keyword">new</span> ofstream(clusterOutPath.c_str()));
<a name="l00073"></a>00073 
<a name="l00074"></a>00074 }
<a name="l00075"></a>00075 
<a name="l00076"></a><a class="code" href="classQuantBirdseed.html#a2160acc32f503ed638de0d6aa789e4ae">00076</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantBirdseed.html#a2160acc32f503ed638de0d6aa789e4ae" title="Get set up for a run of reporting probesets.">QuantBirdseed::prepare</a>(<span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart) {
<a name="l00077"></a>00077     <span class="keywordtype">bool</span> result = <span class="keyword">true</span>;
<a name="l00078"></a>00078     <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;<a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>.size(); i++)
<a name="l00079"></a>00079         result = result &amp;&amp; <a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>[i]-&gt;<a class="code" href="classQuantBirdseed.html#a2160acc32f503ed638de0d6aa789e4ae" title="Get set up for a run of reporting probesets.">prepare</a>(*<span class="keyword">this</span>, iMart);
<a name="l00080"></a>00080 
<a name="l00081"></a>00081   <span class="keywordflow">if</span> (m_ClusterOstrm.get() != NULL) {
<a name="l00082"></a>00082     (*m_ClusterOstrm) &lt;&lt; <span class="stringliteral">&quot;#%affymetrix-algorithm-param-&quot;</span> &lt;&lt; <span class="stringliteral">&quot;probeset_count&quot;</span> &lt;&lt; <span class="stringliteral">&quot;=&quot;</span> &lt;&lt; m_Info.<a class="code" href="classAnalysisInfo.html#ac85efbbb6618cb87c39511af32c311a7" title="How many probesets are on the chip and in cdf file. This is /// critical as probesets must be the exa...">m_NumProbeSets</a> &lt;&lt; endl;
<a name="l00083"></a>00083     std::vector&lt;std::string&gt;::const_iterator keyIx, paramIx;
<a name="l00084"></a>00084     <span class="comment">// The birdseed code from broad doesn&#39;t know about our headers so just put them at top of file</span>
<a name="l00085"></a>00085     <span class="keywordflow">for</span>(keyIx = m_Info.<a class="code" href="classAnalysisInfo.html#ad48864dcd9e2e22d9d74dfbbb00fff04" title="These next two vectors are matched pairs, one contains the /// keys and the other the values...">m_ParamNames</a>.begin(), paramIx = m_Info.m_ParamValues.begin();
<a name="l00086"></a>00086       keyIx != m_Info.<a class="code" href="classAnalysisInfo.html#ad48864dcd9e2e22d9d74dfbbb00fff04" title="These next two vectors are matched pairs, one contains the /// keys and the other the values...">m_ParamNames</a>.end() &amp;&amp; paramIx != m_Info.m_ParamValues.end();
<a name="l00087"></a>00087         ++keyIx, ++paramIx) {
<a name="l00088"></a>00088       <span class="comment">// @todo should we using the AffymetrixParameterConsts.h #defined values?</span>
<a name="l00089"></a>00089       (*m_ClusterOstrm) &lt;&lt; <span class="stringliteral">&quot;#%affymetrix-algorithm-param-&quot;</span> &lt;&lt; *keyIx &lt;&lt; <span class="stringliteral">&quot;=&quot;</span> &lt;&lt; *paramIx &lt;&lt; endl;
<a name="l00090"></a>00090     }
<a name="l00091"></a>00091     <span class="keywordflow">for</span>(keyIx = m_Info.<a class="code" href="classAnalysisInfo.html#ac8ac9d2dfbbcfc1f4f94abba513bcc94" title="Two vectors for client supplied meta-data. Matched pairs of vectors, one contains the /// keys and th...">m_ClientInfoNames</a>.begin(), paramIx = m_Info.m_ClientInfoValues.begin();
<a name="l00092"></a>00092         keyIx != m_Info.<a class="code" href="classAnalysisInfo.html#ac8ac9d2dfbbcfc1f4f94abba513bcc94" title="Two vectors for client supplied meta-data. Matched pairs of vectors, one contains the /// keys and th...">m_ClientInfoNames</a>.end() &amp;&amp; paramIx != m_Info.m_ClientInfoValues.end();
<a name="l00093"></a>00093       ++keyIx, ++paramIx) {
<a name="l00094"></a>00094       <span class="comment">// @todo should we using the AffymetrixParameterConsts.h #defined values?</span>
<a name="l00095"></a>00095       (*m_ClusterOstrm) &lt;&lt; <span class="stringliteral">&quot;#%affymetrix-application-meta-data-info-&quot;</span> &lt;&lt; *keyIx &lt;&lt; <span class="stringliteral">&quot;=&quot;</span> &lt;&lt; *paramIx &lt;&lt; endl;
<a name="l00096"></a>00096       
<a name="l00097"></a>00097     }
<a name="l00098"></a>00098   }
<a name="l00099"></a>00099   <span class="keywordflow">return</span> result;
<a name="l00100"></a>00100 }
<a name="l00101"></a>00101 
<a name="l00102"></a>00102 <span class="comment">// Copied from QuantBRLMM</span><span class="comment"></span>
<a name="l00103"></a>00103 <span class="comment">/**</span>
<a name="l00104"></a>00104 <span class="comment"> * @brief Set up the quantification method given all the data about the probe</span>
<a name="l00105"></a>00105 <span class="comment"> * set, chip layout and data.</span>
<a name="l00106"></a>00106 <span class="comment"> *</span>
<a name="l00107"></a>00107 <span class="comment"> * @param psGroup - Probes to be used for final estimate.</span>
<a name="l00108"></a>00108 <span class="comment"> * @param layout - Chip layout annotation.</span>
<a name="l00109"></a>00109 <span class="comment"> * @param iMart - Raw data from chips.</span>
<a name="l00110"></a>00110 <span class="comment"> * @param iTrans - Transformations to be applied to data before use.</span>
<a name="l00111"></a>00111 <span class="comment"> * @param pmAdjust - How to estimate background, or MM probe.</span>
<a name="l00112"></a>00112 <span class="comment"> * @return True if setup sucessful, false otherwise.</span>
<a name="l00113"></a>00113 <span class="comment"> */</span>
<a name="l00114"></a><a class="code" href="classQuantBirdseed.html#a9d67b068499f0eb1c75f62ff399eb50a">00114</a> <span class="keywordtype">bool</span> <a class="code" href="classQuantBirdseed.html#a9d67b068499f0eb1c75f62ff399eb50a" title="Set up the quantification method given all the data about the probe set, chip layout and data...">QuantBirdseed::setUp</a>(<a class="code" href="classProbeSetGroup.html" title="Group of probe sets that should be processed as a single large probe set.">ProbeSetGroup</a> &amp;psGroup, <span class="keyword">const</span> <a class="code" href="classIntensityMart.html" title="IntensityMart.">IntensityMart</a> &amp;iMart, 
<a name="l00115"></a>00115                           std::vector&lt;ChipStream *&gt; &amp;iTrans, <a class="code" href="classPmAdjuster.html" title="Interface for determining a change based on intensity of perfect match.">PmAdjuster</a> &amp;pmAdjust)
<a name="l00116"></a>00116 {
<a name="l00117"></a>00117 <span class="preprocessor">#if 0</span>
<a name="l00118"></a>00118 <span class="preprocessor"></span>    <span class="comment">// Not doing this anymore.  If correction factor is negative, a SNP-specific correction</span>
<a name="l00119"></a>00119     <span class="comment">// factor is calculated by birdseed.</span>
<a name="l00120"></a>00120     <span class="keywordflow">if</span> (m_CorrectionFactor &lt;= 0) {
<a name="l00121"></a>00121         m_CorrectionFactor = averageIntensities(iMart);
<a name="l00122"></a>00122     }
<a name="l00123"></a>00123 <span class="preprocessor">#endif</span>
<a name="l00124"></a>00124 <span class="preprocessor"></span>    <span class="keyword">const</span> <a class="code" href="classProbeSet.html" title="Collection of atoms that measure target that should be absent or present at the same time...">ProbeSet</a> *gtPs = NULL; <span class="comment">// Genotyping probeset</span>
<a name="l00125"></a>00125     blankSelf(); <span class="comment">// Make sure we clear out the past analysis before starting this one.</span>
<a name="l00126"></a>00126     <span class="keywordtype">bool</span> success = <span class="keyword">true</span>;
<a name="l00127"></a>00127     <span class="comment">/* Sanity checks about probesets. */</span>
<a name="l00128"></a>00128     <span class="keywordflow">if</span>(psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>.empty())
<a name="l00129"></a>00129         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Zero probesets in ProbeSetGroup (group: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(psGroup.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00130"></a>00130     <span class="comment">/* Remember this probeset. */</span>
<a name="l00131"></a>00131     gtPs = psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>[0];
<a name="l00132"></a>00132     <a class="code" href="classQuantGTypeMethod.html#a447dc22d0ca6a8c52c522f1f7eddfa41" title="Genotyping probeset currently using.">m_GtProbeSet</a> = gtPs;
<a name="l00133"></a>00133     <a class="code" href="classQuantGTypeMethod.html#ab4ce64f2ef8bc3ec6d554fa0c1f13212" title="Name of current probeset.">m_ProbesetName</a> = gtPs-&gt;<a class="code" href="classProbeSet.html#af10b2dac78a597ec736691302687c3f3" title="Name of probe set.">name</a>;
<a name="l00134"></a>00134     <span class="keywordflow">if</span> (!<a class="code" href="classQuantGTypeMethod.html#a71d6dfe5d15e5d77354f770913c23824" title="Can the ProbeSet set up the intensities and summaries.">canSetUpProbeSet</a>(gtPs)) {<span class="keywordflow">return</span> <span class="keyword">false</span>;}
<a name="l00135"></a>00135     <span class="keywordflow">if</span>(psGroup.<a class="code" href="classProbeSetGroup.html#aa197d2e48f3c2e8ae67cc8f5d54735ac" title="Vector of probe set pointers.">probeSets</a>.size() &gt; 1)
<a name="l00136"></a>00136         <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Can&#39;t have multiple probesets in a genotyping ProbeSetGroup (group: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(psGroup.<a class="code" href="classProbeSetGroup.html#a0eb8ea2dfda45864880e0495539753db" title="Name of probe set group.">name</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00137"></a>00137  
<a name="l00138"></a>00138     <span class="keywordflow">if</span>(<a class="code" href="classQuantBirdseed.html#a1dab8ffb3062eefd56689f8f0ee23efb" title="Quantification method used for summarizing snps.">m_QuantMethod</a> == NULL) {
<a name="l00139"></a>00139         <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> factory(QuantMethodFactory::Expression);
<a name="l00140"></a>00140         <span class="keywordtype">string</span> spec = <span class="stringliteral">&quot;plier.optmethod=1&quot;</span>;
<a name="l00141"></a>00141         <a class="code" href="classChipLayout.html" title="ChipLayout - Data structure to represent the probesets encoded in the CDF, SPF or PGF files which def...">ChipLayout</a> layout;
<a name="l00142"></a>00142         <a class="code" href="classQuantBirdseed.html#a1dab8ffb3062eefd56689f8f0ee23efb" title="Quantification method used for summarizing snps.">m_QuantMethod</a> = factory.<a class="code" href="classQuantMethodFactory.html#abfb5fe4b06ae6e00419c35d55885feed" title="Factory for creating QuantMethod from string representation.">quantExprMethodForString</a>(spec, layout, QuantMethodFactory::Expression);
<a name="l00143"></a>00143     }
<a name="l00144"></a>00144 
<a name="l00145"></a>00145     <a class="code" href="classQuantGTypeMethod.html#a1a853c2d2821e9c4930fb5c3807f36f9" title="fill in the probe set with the allele values">fillInAlleleProbeSets</a>(*gtPs, <a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>);
<a name="l00146"></a>00146 
<a name="l00147"></a>00147     <span class="comment">/* Get summaries for each allele. */</span>
<a name="l00148"></a>00148     success &amp;= <a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#a027c241e4b65ebdcb30edddd4c8bebef" title="Probeset to detect the A allele.">m_Aallele</a>, <a class="code" href="classQuantGTypeMethod.html#ac52d4ab62a8b33973ff30e7502259138" title="Our summarized values for A allele intensity.">m_AValues</a>, iMart, iTrans, pmAdjust, <a class="code" href="classQuantBirdseed.html#a1dab8ffb3062eefd56689f8f0ee23efb" title="Quantification method used for summarizing snps.">m_QuantMethod</a>, <span class="keyword">false</span>, <span class="keyword">true</span>, <a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>);
<a name="l00149"></a>00149         <span class="keywordflow">if</span> (gtPs-&gt;<a class="code" href="classProbeSet.html#aa3e4170184169cc3bee6d958d74940ba" title="What type of probeset is this?">psType</a> == ProbeSet::Copynumber) {blankSelf(); <span class="keywordflow">return</span> <span class="keyword">false</span>;}
<a name="l00150"></a>00150     success &amp;= <a class="code" href="classQuantGTypeMethod.html#a6561033a69acb59e1b6ac76976f80561" title="summarize this allele">summarizeAllele</a>(&amp;<a class="code" href="classQuantGTypeMethod.html#aff97b047d5b025e7487355df0307ac9e" title="Probeset to detect the B allele.">m_Ballele</a>, <a class="code" href="classQuantGTypeMethod.html#af8ff23fd09ff5899b4497caf1ba7b7b9" title="Our summarized values for B allele intensity.">m_BValues</a>, iMart, iTrans, pmAdjust, <a class="code" href="classQuantBirdseed.html#a1dab8ffb3062eefd56689f8f0ee23efb" title="Quantification method used for summarizing snps.">m_QuantMethod</a>, <span class="keyword">false</span>, <span class="keyword">true</span>, <a class="code" href="classQuantBirdseed.html#a2d8ece68df62b221b0a621cfa52697fd" title="Report object for outputting results.">m_Reporters</a>);
<a name="l00151"></a>00151     <span class="keywordflow">if</span>(!success)
<a name="l00152"></a>00152         blankSelf();
<a name="l00153"></a>00153     <span class="keywordflow">return</span> success;
<a name="l00154"></a>00154 }
<a name="l00155"></a>00155 
<a name="l00156"></a><a class="code" href="classQuantBirdseed.html#a646dedbfcd698523d82423014c09bce7">00156</a> <span class="keywordtype">void</span> <a class="code" href="classQuantBirdseed.html#a646dedbfcd698523d82423014c09bce7" title="Custom configuration for this QuantMethod.">QuantBirdseed::setParameters</a>(<a class="code" href="classPsBoard.html">PsBoard</a> &amp;board) {
<a name="l00157"></a>00157   <a class="code" href="classOptions.html" title="base class for tracking class options">Options</a> *o = board.getOptions();
<a name="l00158"></a>00158   vector&lt;string&gt; celFiles = o-&gt;<a class="code" href="classOptions.html#ab20d0083424d9d9a8b94da46f487f57a" title="Return a vector of values for a particular option.">getOptVector</a>(<span class="stringliteral">&quot;cels&quot;</span>);
<a name="l00159"></a>00159   <span class="keywordtype">string</span> outDir = o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;out-dir&quot;</span>);
<a name="l00160"></a>00160   <a class="code" href="classGenotypeInfo.html">GenotypeInfo</a> *gInfo = board.getGenotypeInfo();
<a name="l00161"></a>00161   <span class="keywordflow">if</span>(gInfo == NULL) {
<a name="l00162"></a>00162     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;QuantBirdseed::setParameters() - GenotypeInfo is NULL&quot;</span>);
<a name="l00163"></a>00163   }
<a name="l00164"></a>00164   setGenders(gInfo-&gt;m_Genders-&gt;<a class="code" href="classGenderCalls.html#abb3474c5f29e2959ce730f539b73b006" title="Get the genders for the cel files that have been seen.">getGenders</a>());
<a name="l00165"></a>00165   <span class="keywordflow">if</span> (o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;read-models-birdseed&quot;</span>) != <span class="stringliteral">&quot;&quot;</span>) {
<a name="l00166"></a>00166       <span class="keywordtype">string</span> modelFile = o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;read-models-birdseed&quot;</span>);
<a name="l00167"></a>00167       m_PriorsReader.reset(<span class="keyword">new</span> <a class="code" href="classbirdseed_1_1dev_1_1PriorsReaderTemplate.html">birdseed::dev::PriorsReader</a>(*gInfo-&gt;m_SpecialSnps,
<a name="l00168"></a>00168                                                            <span class="keyword">new</span> <a class="code" href="classbirdseed_1_1dev_1_1TsvPriorsReader.html" title="Used to read the Affy Tsv Priors format.">birdseed::dev::TsvPriorsReader</a>(modelFile, NULL)));
<a name="l00169"></a>00169   }
<a name="l00170"></a>00170 
<a name="l00171"></a>00171   <span class="comment">// @todo refactor - put current analysis path name on bboard</span>
<a name="l00172"></a>00172   <span class="keywordtype">string</span> outfile = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(outDir,<a class="code" href="classQuantMethod.html#ab1b819505891590c9f3e57d9cc67d4de" title="What is the name of the quantification method?">getType</a>() + <span class="stringliteral">&quot;.snp-posteriors&quot;</span>);
<a name="l00173"></a>00173   <span class="keywordtype">bool</span> writeModels = o-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;write-models&quot;</span>);
<a name="l00174"></a>00174   <span class="keywordflow">if</span>(writeModels) {
<a name="l00175"></a>00175       setClusterOutFile(outfile);
<a name="l00176"></a>00176       <span class="keywordtype">string</span> modelFile = outfile + <span class="stringliteral">&quot;.txt&quot;</span>;
<a name="l00177"></a>00177       board.set(<span class="stringliteral">&quot;model-file-written&quot;</span>, modelFile);
<a name="l00178"></a>00178   }
<a name="l00179"></a>00179   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (o-&gt;<a class="code" href="classOptions.html#a3c98038375042f9b230b3e4e2151fe00" title="Get the boolean value of an option.">getOptBool</a>(<span class="stringliteral">&quot;a5-write-models&quot;</span>)) {<span class="comment"></span>
<a name="l00180"></a>00180 <span class="comment">      ///@todo implement models in A5 for birdseed</span>
<a name="l00181"></a>00181 <span class="comment"></span>      <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;--a5-write-models for birdseed not implemented&quot;</span>);
<a name="l00182"></a>00182   }
<a name="l00183"></a>00183   <span class="comment">// Add quantification method.</span>
<a name="l00184"></a>00184   <a class="code" href="classQuantMethodFactory.html" title="Factory class for making chip streams based on a string representation.">QuantMethodFactory</a> qFactory;
<a name="l00185"></a>00185   <span class="keywordtype">string</span> qMethodSpec = o-&gt;<a class="code" href="classOptions.html#a0efc7e2a4a55297f0f642d1d8bc08b82" title="Get the value of an option.">getOpt</a>(<span class="stringliteral">&quot;qmethod-spec&quot;</span>);
<a name="l00186"></a>00186   <a class="code" href="classQuantMethod.html" title="QuantMethod - Interface for computing quantification summaries from PM intensities grouped into probe...">QuantMethod</a> *qMethod = qFactory.<a class="code" href="classQuantMethodFactory.html#ab0f3db6b9aec42ff32ae69b9791329c8" title="Create a pointer to a new QuantMethod object as described in the string specification.">quantMethodForString</a>(qMethodSpec, board);
<a name="l00187"></a>00187 
<a name="l00188"></a>00188   <span class="comment">// @todo - handle feature effects!</span>
<a name="l00189"></a>00189   <a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *eMethod = NULL;
<a name="l00190"></a>00190   eMethod= <span class="keyword">static_cast&lt;</span><a class="code" href="classQuantExprMethod.html" title="QuantExprMethod - Interface for computing quantification summaries from PM intensities grouped into p...">QuantExprMethod</a> *<span class="keyword">&gt;</span>(qMethod);
<a name="l00191"></a>00191   <span class="keywordflow">if</span> (eMethod == NULL) {
<a name="l00192"></a>00192     <a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Couldn&#39;t make a QuantExprMethod for specification: &quot;</span> + qMethodSpec);
<a name="l00193"></a>00193   }
<a name="l00194"></a>00194   setQuantExprMethod(eMethod);
<a name="l00195"></a>00195 
<a name="l00196"></a>00196 
<a name="l00197"></a>00197 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:52 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
