<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: label/snp.label.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_e1acf30fd02690554687a3cc62e9f05c.html">label</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>snp.label.cpp</h1>  </div>
</div>
<div class="contents">
<a href="snp_8label_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2005 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   snp.label.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Earl Hubbell</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Jan 22 2007</span>
<a name="l00026"></a>00026 <span class="comment"> * </span>
<a name="l00027"></a>00027 <span class="comment"> * Does the labeling algorithmics.</span>
<a name="l00028"></a>00028 <span class="comment"> * </span>
<a name="l00029"></a>00029 <span class="comment"> */</span>
<a name="l00030"></a>00030 
<a name="l00031"></a>00031 <span class="comment">//</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="snp_8label_8h.html">label/snp.label.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="comment">//</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="BioTypes_8h.html" title="Central location enumerations and other types that aren&amp;#39;t particular to a single class or applica...">chipstream/BioTypes.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;chipstream/MatrixUtil.h&quot;</span>
<a name="l00036"></a>00036 <span class="comment">//</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &quot;newmat.h&quot;</span>
<a name="l00038"></a>00038 <span class="comment">//</span>
<a name="l00039"></a>00039 <span class="preprocessor">#include &lt;algorithm&gt;</span>
<a name="l00040"></a>00040 <span class="preprocessor">#include &lt;cmath&gt;</span>
<a name="l00041"></a>00041 <span class="preprocessor">#include &lt;cstdio&gt;</span>
<a name="l00042"></a>00042 <span class="preprocessor">#include &lt;cstdlib&gt;</span>
<a name="l00043"></a>00043 <span class="preprocessor">#include &lt;functional&gt;</span>
<a name="l00044"></a>00044 <span class="preprocessor">#include &lt;memory&gt;</span>
<a name="l00045"></a>00045 <span class="preprocessor">#include &lt;new&gt;</span>
<a name="l00046"></a>00046 <span class="preprocessor">#include &lt;utility&gt;</span>
<a name="l00047"></a>00047 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00048"></a>00048 <span class="comment">//</span>
<a name="l00049"></a>00049 
<a name="l00050"></a>00050 <span class="comment">// draft of plain-vanilla code for doing labeling numerics</span>
<a name="l00051"></a>00051 <span class="comment">// just match the code blindly</span>
<a name="l00052"></a>00052 
<a name="l00053"></a>00053 <span class="keyword">using namespace </span>std;
<a name="l00054"></a>00054 <span class="comment"></span>
<a name="l00055"></a>00055 <span class="comment">/**</span>
<a name="l00056"></a>00056 <span class="comment"> * hash indexer</span>
<a name="l00057"></a>00057 <span class="comment"> * </span>
<a name="l00058"></a>00058 <span class="comment"> * @param i - first index</span>
<a name="l00059"></a>00059 <span class="comment"> * @param j - second index</span>
<a name="l00060"></a>00060 <span class="comment"> * @param length - size of side of array</span>
<a name="l00061"></a>00061 <span class="comment"> */</span>
<a name="l00062"></a><a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c">00062</a> <span class="keyword">inline</span> <span class="keywordtype">int</span> <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(<span class="keywordtype">int</span> i, <span class="keywordtype">int</span> j, <span class="keywordtype">int</span> length)
<a name="l00063"></a>00063 {
<a name="l00064"></a>00064     <span class="keywordflow">return</span>(i*length+j);
<a name="l00065"></a>00065 }
<a name="l00066"></a>00066 <span class="comment"></span>
<a name="l00067"></a>00067 <span class="comment">/**</span>
<a name="l00068"></a>00068 <span class="comment"> * copies one float vector to another</span>
<a name="l00069"></a>00069 <span class="comment"> * </span>
<a name="l00070"></a>00070 <span class="comment"> * @param final - destination vector</span>
<a name="l00071"></a>00071 <span class="comment"> * @param source - source vector</span>
<a name="l00072"></a>00072 <span class="comment"> * @param length - length of data</span>
<a name="l00073"></a>00073 <span class="comment"> */</span>
<a name="l00074"></a><a class="code" href="snp_8label_8cpp.html#ad32e3aff38d25c6b491ca6ed6e7fbc0d">00074</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#ad32e3aff38d25c6b491ca6ed6e7fbc0d" title="copies one float vector to another">transfer</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>, <span class="keywordtype">double</span> *source, <span class="keywordtype">int</span> length)
<a name="l00075"></a>00075 {
<a name="l00076"></a>00076     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;length; i++)
<a name="l00077"></a>00077         <span class="keyword">final</span>[i] = source[i];
<a name="l00078"></a>00078 }
<a name="l00079"></a>00079 <span class="comment"></span>
<a name="l00080"></a>00080 <span class="comment">/**</span>
<a name="l00081"></a>00081 <span class="comment"> * cumulative sum of source vector</span>
<a name="l00082"></a>00082 <span class="comment"> *</span>
<a name="l00083"></a>00083 <span class="comment"> * @param final - destination for cumulative sum</span>
<a name="l00084"></a>00084 <span class="comment"> * @param source - source of individual values</span>
<a name="l00085"></a>00085 <span class="comment"> * @param length - length of data</span>
<a name="l00086"></a>00086 <span class="comment"> */</span>
<a name="l00087"></a><a class="code" href="snp_8label_8cpp.html#a089fb51f8383b6eea013f35531b091a9">00087</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a089fb51f8383b6eea013f35531b091a9" title="cumulative sum of source vector">cumsum</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>, <span class="keywordtype">double</span> *source, <span class="keywordtype">int</span> length)
<a name="l00088"></a>00088 {
<a name="l00089"></a>00089     <span class="keyword">final</span>[0] = source[0];
<a name="l00090"></a>00090     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=1; i&lt;length; i++)
<a name="l00091"></a>00091         <span class="keyword">final</span>[i] = source[i]+<span class="keyword">final</span>[i-1];
<a name="l00092"></a>00092 }
<a name="l00093"></a>00093 <span class="comment"></span>
<a name="l00094"></a>00094 <span class="comment">/**</span>
<a name="l00095"></a>00095 <span class="comment"> * cumulative sum of source vector using STL</span>
<a name="l00096"></a>00096 <span class="comment"> * </span>
<a name="l00097"></a>00097 <span class="comment"> * @param final - destination for cumulative sum</span>
<a name="l00098"></a>00098 <span class="comment"> * @param source - source of individual values</span>
<a name="l00099"></a>00099 <span class="comment"> * @param length - length of data to sum over [carried over]</span>
<a name="l00100"></a>00100 <span class="comment"> */</span>
<a name="l00101"></a><a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20">00101</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(vector&lt;double&gt; &amp;<span class="keyword">final</span>,<span class="keyword">const</span> vector&lt;double&gt; &amp;source, <span class="keywordtype">int</span> length)
<a name="l00102"></a>00102 {
<a name="l00103"></a>00103     <span class="keyword">final</span>.resize(length);
<a name="l00104"></a>00104     <span class="keyword">final</span>[0] = source[0];
<a name="l00105"></a>00105     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=1; i&lt;length; i++)
<a name="l00106"></a>00106         <span class="keyword">final</span>[i] = source[i] + <span class="keyword">final</span>[i-1];
<a name="l00107"></a>00107 }
<a name="l00108"></a>00108 <span class="comment"></span>
<a name="l00109"></a>00109 <span class="comment">/**</span>
<a name="l00110"></a>00110 <span class="comment"> * difference in cumulative sums: i.e. sum from b to a</span>
<a name="l00111"></a>00111 <span class="comment"> *</span>
<a name="l00112"></a>00112 <span class="comment"> * @param source - vector of cumulative sums</span>
<a name="l00113"></a>00113 <span class="comment"> * @param length - length of data</span>
<a name="l00114"></a>00114 <span class="comment"> * @param a - larger cumulative point</span>
<a name="l00115"></a>00115 <span class="comment"> * @param b - smaller cumulative point</span>
<a name="l00116"></a>00116 <span class="comment"> */</span>
<a name="l00117"></a><a class="code" href="snp_8label_8cpp.html#ad748ef26c031cf6086e3a490afd3a47f">00117</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#ad748ef26c031cf6086e3a490afd3a47f" title="difference in cumulative sums: i.e.">delta</a>(<span class="keywordtype">double</span> *source,<span class="keywordtype">int</span> length, <span class="keywordtype">int</span> a, <span class="keywordtype">int</span> b)
<a name="l00118"></a>00118 {
<a name="l00119"></a>00119     <span class="keywordtype">double</span> tmp;
<a name="l00120"></a>00120     <span class="comment">// difference in cumuldtive sums</span>
<a name="l00121"></a>00121     tmp = source[a]-source[b];
<a name="l00122"></a>00122     <span class="keywordflow">if</span> (a&lt;b)
<a name="l00123"></a>00123         tmp = 0;
<a name="l00124"></a>00124     <span class="keywordflow">return</span>(tmp);
<a name="l00125"></a>00125 }
<a name="l00126"></a>00126 <span class="comment"></span>
<a name="l00127"></a>00127 <span class="comment">/**</span>
<a name="l00128"></a>00128 <span class="comment"> * sum up the source array along one axis</span>
<a name="l00129"></a>00129 <span class="comment"> *</span>
<a name="l00130"></a>00130 <span class="comment"> * @param final - output, marginal sums of array</span>
<a name="l00131"></a>00131 <span class="comment"> * @param source - input, all the individual values</span>
<a name="l00132"></a>00132 <span class="comment"> * @param length - dimension of array</span>
<a name="l00133"></a>00133 <span class="comment"> * @param flag - which margin to sum to</span>
<a name="l00134"></a>00134 <span class="comment"> */</span>
<a name="l00135"></a><a class="code" href="snp_8label_8cpp.html#a476e5d98fb03d9efaccb100f40cd672f">00135</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a476e5d98fb03d9efaccb100f40cd672f" title="sum up the source array along one axis">apply_sum</a>(vector&lt;double&gt; &amp;<span class="keyword">final</span>,<span class="keyword">const</span> vector&lt;double&gt; &amp;source, <span class="keywordtype">int</span> length, <span class="keywordtype">int</span> flag)
<a name="l00136"></a>00136 {
<a name="l00137"></a>00137     <span class="keywordtype">int</span> i,j;
<a name="l00138"></a>00138     <span class="keywordtype">double</span> sum;
<a name="l00139"></a>00139 
<a name="l00140"></a>00140     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00141"></a>00141     {
<a name="l00142"></a>00142         sum=0;
<a name="l00143"></a>00143         <span class="keywordflow">for</span> (j=0; j&lt;length; j++)
<a name="l00144"></a>00144         {
<a name="l00145"></a>00145             <span class="keywordflow">if</span> (flag)
<a name="l00146"></a>00146                 sum += source[<a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,length)];
<a name="l00147"></a>00147             <span class="keywordflow">else</span>
<a name="l00148"></a>00148                 sum += source[<a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(j,i,length)];
<a name="l00149"></a>00149         }
<a name="l00150"></a>00150         <span class="keyword">final</span>[i] = sum;
<a name="l00151"></a>00151     }
<a name="l00152"></a>00152 }
<a name="l00153"></a>00153 <span class="comment"></span>
<a name="l00154"></a>00154 <span class="comment">/**</span>
<a name="l00155"></a>00155 <span class="comment"> * takes a cumulative sum and makes it cumulate from the other end</span>
<a name="l00156"></a>00156 <span class="comment"> *</span>
<a name="l00157"></a>00157 <span class="comment"> * @param final - output, reverse cumulative sum</span>
<a name="l00158"></a>00158 <span class="comment"> * @param source - input, forward direction cumulative sum</span>
<a name="l00159"></a>00159 <span class="comment"> * @param length - length of data</span>
<a name="l00160"></a>00160 <span class="comment"> */</span>
<a name="l00161"></a><a class="code" href="snp_8label_8cpp.html#a4ad3f76517de13f099989b46c01d6c50">00161</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a4ad3f76517de13f099989b46c01d6c50" title="takes a cumulative sum and makes it cumulate from the other end">reversecumsum</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>,<span class="keyword">const</span> <span class="keywordtype">double</span> *source, <span class="keywordtype">int</span> length)
<a name="l00162"></a>00162 {
<a name="l00163"></a>00163     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;length; i++)
<a name="l00164"></a>00164         <span class="keyword">final</span>[i] = source[length-1]-source[i];
<a name="l00165"></a>00165 }
<a name="l00166"></a>00166 <span class="comment"></span>
<a name="l00167"></a>00167 <span class="comment">/**</span>
<a name="l00168"></a>00168 <span class="comment"> * takes a cumulative sum and makes it cumulate from the other end</span>
<a name="l00169"></a>00169 <span class="comment"> *</span>
<a name="l00170"></a>00170 <span class="comment"> * @param final - output, reverse cumulative sum</span>
<a name="l00171"></a>00171 <span class="comment"> * @param source - input, forward direction cumulative sum</span>
<a name="l00172"></a>00172 <span class="comment"> * @param length - length of data (not needed for STL)</span>
<a name="l00173"></a>00173 <span class="comment"> */</span>
<a name="l00174"></a><a class="code" href="snp_8label_8cpp.html#a71b11dd6854f20d84e0f3f8993a6cdbe">00174</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a71b11dd6854f20d84e0f3f8993a6cdbe" title="takes a cumulative sum and makes it cumulate from the other end">vreversecumsum</a>(vector&lt;double&gt; &amp;<span class="keyword">final</span>, <span class="keyword">const</span> vector&lt;double&gt; &amp;source, <span class="keywordtype">int</span> length)
<a name="l00175"></a>00175 {
<a name="l00176"></a>00176     <span class="keywordtype">double</span> tmp;
<a name="l00177"></a>00177     <span class="keyword">final</span>.resize(length);
<a name="l00178"></a>00178     tmp = source[length-1];
<a name="l00179"></a>00179     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;length; i++)
<a name="l00180"></a>00180         <span class="keyword">final</span>[i] = tmp-source[i];
<a name="l00181"></a>00181 }
<a name="l00182"></a>00182 <span class="comment"></span>
<a name="l00183"></a>00183 <span class="comment">/**</span>
<a name="l00184"></a>00184 <span class="comment"> * takes outer difference of two vectors</span>
<a name="l00185"></a>00185 <span class="comment"> * </span>
<a name="l00186"></a>00186 <span class="comment"> * @param final - output matrix</span>
<a name="l00187"></a>00187 <span class="comment"> * @param a - input vector</span>
<a name="l00188"></a>00188 <span class="comment"> * @param b - input vector</span>
<a name="l00189"></a>00189 <span class="comment"> * @param length - length of data</span>
<a name="l00190"></a>00190 <span class="comment"> */</span>
<a name="l00191"></a><a class="code" href="snp_8label_8cpp.html#aa07c57aa95a5e1df7ed8e7d7c5154780">00191</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#aa07c57aa95a5e1df7ed8e7d7c5154780" title="takes outer difference of two vectors">outer_minus</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>, <span class="keyword">const</span> <span class="keywordtype">double</span> *a,<span class="keyword">const</span> <span class="keywordtype">double</span> *b, <span class="keywordtype">int</span> length)
<a name="l00192"></a>00192 {
<a name="l00193"></a>00193     <span class="keywordtype">int</span> i,j;
<a name="l00194"></a>00194 
<a name="l00195"></a>00195     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00196"></a>00196         <span class="keywordflow">for</span> (j=i; j&lt;length; j++)
<a name="l00197"></a>00197             <span class="keyword">final</span>[<a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,length)] = a[j]-b[i];
<a name="l00198"></a>00198 }
<a name="l00199"></a>00199 <span class="comment"></span>
<a name="l00200"></a>00200 <span class="comment">/**</span>
<a name="l00201"></a>00201 <span class="comment"> * mask out a portion of an array</span>
<a name="l00202"></a>00202 <span class="comment"> *</span>
<a name="l00203"></a>00203 <span class="comment"> * @param final - array to be masked</span>
<a name="l00204"></a>00204 <span class="comment"> * @param length - size of margin of array</span>
<a name="l00205"></a>00205 <span class="comment"> * @param val - value to be placed in masked area</span>
<a name="l00206"></a>00206 <span class="comment"> */</span>
<a name="l00207"></a><a class="code" href="snp_8label_8cpp.html#a553a86714dd371ce3f5833b6f8b3f447">00207</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a553a86714dd371ce3f5833b6f8b3f447" title="mask out a portion of an array">mask</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>, <span class="keywordtype">int</span> length, <span class="keywordtype">double</span> val)
<a name="l00208"></a>00208 {
<a name="l00209"></a>00209     <span class="keywordtype">int</span> i,j;
<a name="l00210"></a>00210     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00211"></a>00211         <span class="keywordflow">for</span> (j=0; j&lt;i; j++)
<a name="l00212"></a>00212             <span class="keyword">final</span>[<a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,length)] = val;
<a name="l00213"></a>00213 }
<a name="l00214"></a>00214 <span class="comment"></span>
<a name="l00215"></a>00215 <span class="comment">/**</span>
<a name="l00216"></a>00216 <span class="comment"> * mask out a portion of an array</span>
<a name="l00217"></a>00217 <span class="comment"> *</span>
<a name="l00218"></a>00218 <span class="comment"> * @param final - array to be masked</span>
<a name="l00219"></a>00219 <span class="comment"> * @param length - size of margin of array</span>
<a name="l00220"></a>00220 <span class="comment"> * @param val - value to be placed in masked area</span>
<a name="l00221"></a>00221 <span class="comment"> */</span>
<a name="l00222"></a><a class="code" href="snp_8label_8cpp.html#a0753c4704b4a8316c7aec8ad57f1e59f">00222</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a0753c4704b4a8316c7aec8ad57f1e59f" title="mask out a portion of an array">vmask</a>(vector&lt;double&gt; &amp;<span class="keyword">final</span>, <span class="keywordtype">int</span> length, <span class="keywordtype">double</span> val)
<a name="l00223"></a>00223 {
<a name="l00224"></a>00224     <span class="keywordtype">int</span> i,j;
<a name="l00225"></a>00225     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00226"></a>00226         <span class="keywordflow">for</span> (j=0; j&lt;i; j++)
<a name="l00227"></a>00227             <span class="keyword">final</span>[<a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,length)] = val;
<a name="l00228"></a>00228     
<a name="l00229"></a>00229 }
<a name="l00230"></a>00230 <span class="comment"></span>
<a name="l00231"></a>00231 <span class="comment">/**</span>
<a name="l00232"></a>00232 <span class="comment"> * sum two vectors</span>
<a name="l00233"></a>00233 <span class="comment"> *</span>
<a name="l00234"></a>00234 <span class="comment"> * @param final - output</span>
<a name="l00235"></a>00235 <span class="comment"> * @param a - input vector</span>
<a name="l00236"></a>00236 <span class="comment"> * @param b - input vector</span>
<a name="l00237"></a>00237 <span class="comment"> * @param length - length of input vectors</span>
<a name="l00238"></a>00238 <span class="comment"> */</span>
<a name="l00239"></a><a class="code" href="snp_8label_8cpp.html#a8de4a2f660ae0703b6f1f922b0b29478">00239</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a8de4a2f660ae0703b6f1f922b0b29478" title="sum two vectors">vsum</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>,<span class="keyword">const</span> <span class="keywordtype">double</span> *a,<span class="keyword">const</span> <span class="keywordtype">double</span> *b, <span class="keywordtype">int</span> length)
<a name="l00240"></a>00240 {
<a name="l00241"></a>00241     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;length; i++)
<a name="l00242"></a>00242         <span class="keyword">final</span>[i] = a[i]+b[i];
<a name="l00243"></a>00243 }
<a name="l00244"></a>00244 <span class="comment"></span>
<a name="l00245"></a>00245 <span class="comment">/**</span>
<a name="l00246"></a>00246 <span class="comment"> * add a constant to each element of a vector</span>
<a name="l00247"></a>00247 <span class="comment"> * </span>
<a name="l00248"></a>00248 <span class="comment"> * @param final - output vector</span>
<a name="l00249"></a>00249 <span class="comment"> * @param a - input vector</span>
<a name="l00250"></a>00250 <span class="comment"> * @param k - value to be added</span>
<a name="l00251"></a>00251 <span class="comment"> * @param length - length of data</span>
<a name="l00252"></a>00252 <span class="comment"> */</span>
<a name="l00253"></a><a class="code" href="snp_8label_8cpp.html#a36c1a215b939f32b274bc2e7bf086e71">00253</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a36c1a215b939f32b274bc2e7bf086e71" title="add a constant to each element of a vector">vplusscaldr</a>(<span class="keywordtype">double</span> *<span class="keyword">final</span>, <span class="keywordtype">double</span> *a, <span class="keywordtype">double</span> k, <span class="keywordtype">int</span> length)
<a name="l00254"></a>00254 {
<a name="l00255"></a>00255     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;length; i++)
<a name="l00256"></a>00256         <span class="keyword">final</span>[i] = a[i] + k;
<a name="l00257"></a>00257 }
<a name="l00258"></a>00258 <span class="comment"></span>
<a name="l00259"></a>00259 <span class="comment">/**</span>
<a name="l00260"></a>00260 <span class="comment"> * multiply each element of a vector by a value</span>
<a name="l00261"></a>00261 <span class="comment"> * </span>
<a name="l00262"></a>00262 <span class="comment"> * @param X - vector</span>
<a name="l00263"></a>00263 <span class="comment"> * @param factor - factor to multiply each element by</span>
<a name="l00264"></a>00264 <span class="comment"> */</span>
<a name="l00265"></a><a class="code" href="snp_8label_8cpp.html#a8edf9aba80a02747bdb9108abc2f1e27">00265</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a8edf9aba80a02747bdb9108abc2f1e27" title="multiply each element of a vector by a value">vmult</a>(vector&lt;double&gt; &amp;X, <span class="keywordtype">double</span> factor)
<a name="l00266"></a>00266 {
<a name="l00267"></a>00267     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;X.size(); i++)
<a name="l00268"></a>00268         X[i] *= factor;
<a name="l00269"></a>00269 }
<a name="l00270"></a>00270 <span class="comment"></span>
<a name="l00271"></a>00271 <span class="comment">/**</span>
<a name="l00272"></a>00272 <span class="comment"> * inverts 3x3 matrix</span>
<a name="l00273"></a>00273 <span class="comment"> * specialty code, avoid calling external routines</span>
<a name="l00274"></a>00274 <span class="comment"> *</span>
<a name="l00275"></a>00275 <span class="comment"> * @param w - output inverse of matrix</span>
<a name="l00276"></a>00276 <span class="comment"> * @param v - input matrix</span>
<a name="l00277"></a>00277 <span class="comment"> */</span>
<a name="l00278"></a><a class="code" href="snp_8label_8cpp.html#a57e554714be3edbd1b2393e5824b03b4">00278</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a57e554714be3edbd1b2393e5824b03b4" title="inverts 3x3 matrix specialty code, avoid calling external routines">inversethree</a>(vector&lt;double&gt; &amp;w, vector&lt;double&gt; &amp;v)
<a name="l00279"></a>00279 {
<a name="l00280"></a>00280     w.resize(9);
<a name="l00281"></a>00281     <span class="keywordtype">double</span> det;
<a name="l00282"></a>00282     <span class="comment">// v = 9 element matrix 11,12,...33</span>
<a name="l00283"></a>00283     w[0] = v[4]*v[8]-v[5]*v[7];
<a name="l00284"></a>00284     w[1] = v[2]*v[7]-v[1]*v[8];
<a name="l00285"></a>00285     w[2] = v[1]*v[5]-v[2]*v[4];
<a name="l00286"></a>00286     w[3] = v[5]*v[6]-v[3]*v[8];
<a name="l00287"></a>00287     w[4] = v[0]*v[8]-v[2]*v[6];
<a name="l00288"></a>00288     w[5] = v[2]*v[3]-v[0]*v[5];
<a name="l00289"></a>00289     w[6] = v[3]*v[7]-v[4]*v[6];
<a name="l00290"></a>00290     w[7] = v[1]*v[6]-v[0]*v[7];
<a name="l00291"></a>00291     w[8] = v[0]*v[4]-v[1]*v[3];
<a name="l00292"></a>00292     
<a name="l00293"></a>00293     det = v[0]*w[0]+v[3]*w[1]+v[6]*w[2];
<a name="l00294"></a>00294     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;9; i++)
<a name="l00295"></a>00295         w[i]/=det;
<a name="l00296"></a>00296 }
<a name="l00297"></a>00297 <span class="comment"></span>
<a name="l00298"></a>00298 <span class="comment">/**</span>
<a name="l00299"></a>00299 <span class="comment"> * sum two vectors</span>
<a name="l00300"></a>00300 <span class="comment"> *</span>
<a name="l00301"></a>00301 <span class="comment"> * @param z - output</span>
<a name="l00302"></a>00302 <span class="comment"> * @param a - input</span>
<a name="l00303"></a>00303 <span class="comment"> * @param b - input</span>
<a name="l00304"></a>00304 <span class="comment"> */</span>
<a name="l00305"></a><a class="code" href="snp_8label_8cpp.html#a1f2ff039971eee5c78e94867f1cc9369">00305</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a1f2ff039971eee5c78e94867f1cc9369" title="sum two vectors">sumthree</a>(vector&lt;double&gt; &amp;z, vector&lt;double&gt; &amp;a, vector&lt;double&gt; &amp;b)
<a name="l00306"></a>00306 {
<a name="l00307"></a>00307     z.resize(a.size());
<a name="l00308"></a>00308     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;a.size(); i++)
<a name="l00309"></a>00309         z[i] = a[i]+b[i];
<a name="l00310"></a>00310 }
<a name="l00311"></a>00311 <span class="comment"></span>
<a name="l00312"></a>00312 <span class="comment">/**</span>
<a name="l00313"></a>00313 <span class="comment"> * multiply two 3x3 matrices</span>
<a name="l00314"></a>00314 <span class="comment"> *</span>
<a name="l00315"></a>00315 <span class="comment"> * @param z - output</span>
<a name="l00316"></a>00316 <span class="comment"> * @param a - input 3x3 matrix (as 9 length vector)</span>
<a name="l00317"></a>00317 <span class="comment"> * @param b - input 3x3 matrix (as 9 length vector)</span>
<a name="l00318"></a>00318 <span class="comment"> */</span>
<a name="l00319"></a><a class="code" href="snp_8label_8cpp.html#afc25b20145576567e02ac9bfa60efaa0">00319</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#afc25b20145576567e02ac9bfa60efaa0" title="multiply two 3x3 matrices">timesthree</a>(vector&lt;double&gt; &amp;z, vector&lt;double&gt; &amp;a, vector&lt;double&gt; &amp;b)
<a name="l00320"></a>00320 {
<a name="l00321"></a>00321     z.resize(9);
<a name="l00322"></a>00322     z[0] = a[0]*b[0]+a[1]*b[3]+a[2]*b[6];
<a name="l00323"></a>00323     z[1] = a[0]*b[1]+a[1]*b[4]+a[2]*b[7];
<a name="l00324"></a>00324     z[2] = a[0]*b[2]+a[1]*b[5]+a[2]*b[8];
<a name="l00325"></a>00325     z[3] = a[3]*b[0]+a[4]*b[3]+a[5]*b[6];
<a name="l00326"></a>00326     z[4] = a[3]*b[1]+a[4]*b[4]+a[5]*b[7];
<a name="l00327"></a>00327     z[5] = a[3]*b[2]+a[4]*b[5]+a[5]*b[8];
<a name="l00328"></a>00328     z[6] = a[6]*b[0]+a[7]*b[3]+a[8]*b[6];
<a name="l00329"></a>00329     z[7] = a[6]*b[1]+a[7]*b[4]+a[8]*b[7];
<a name="l00330"></a>00330     z[8] = a[6]*b[2]+a[7]*b[5]+a[8]*b[8];
<a name="l00331"></a>00331 }
<a name="l00332"></a>00332 <span class="comment"></span>
<a name="l00333"></a>00333 <span class="comment">/**</span>
<a name="l00334"></a>00334 <span class="comment"> * multiply a 1x3 vector by a 3x3 matrix</span>
<a name="l00335"></a>00335 <span class="comment"> *</span>
<a name="l00336"></a>00336 <span class="comment"> * @param z - output</span>
<a name="l00337"></a>00337 <span class="comment"> * @param a - 3x3 matrix (as 9 length vector)</span>
<a name="l00338"></a>00338 <span class="comment"> * @param b - 1x3 vector (as 3 length vector)</span>
<a name="l00339"></a>00339 <span class="comment"> */</span>
<a name="l00340"></a><a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289">00340</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(vector&lt;double&gt; &amp;z, vector&lt;double&gt; &amp;a, vector&lt;double&gt; &amp;b)
<a name="l00341"></a>00341 {
<a name="l00342"></a>00342     z.resize(3);
<a name="l00343"></a>00343     z[0] = a[0]*b[0]+a[1]*b[1]+a[2]*b[2];
<a name="l00344"></a>00344     z[1] = a[3]*b[0]+a[4]*b[1]+a[5]*b[2];
<a name="l00345"></a>00345     z[2] = a[6]*b[0]+a[7]*b[1]+a[8]*b[2];
<a name="l00346"></a>00346 }
<a name="l00347"></a>00347 
<a name="l00348"></a>00348 <span class="comment"></span>
<a name="l00349"></a>00349 <span class="comment">/**</span>
<a name="l00350"></a>00350 <span class="comment"> * bayesian mean - prior + observed</span>
<a name="l00351"></a>00351 <span class="comment"> * </span>
<a name="l00352"></a>00352 <span class="comment"> * @param k - prior pseudo-observations</span>
<a name="l00353"></a>00353 <span class="comment"> * @param m - prior mean</span>
<a name="l00354"></a>00354 <span class="comment"> * @param mm - observed sum of data</span>
<a name="l00355"></a>00355 <span class="comment"> * @param n - number of observations</span>
<a name="l00356"></a>00356 <span class="comment"> */</span>
<a name="l00357"></a><a class="code" href="snp_8label_8cpp.html#ad1202229394b0f93827ae5a67abd951f">00357</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#ad1202229394b0f93827ae5a67abd951f" title="bayesian mean - prior + observed">b_mean</a>(<span class="keywordtype">double</span> k, <span class="keywordtype">double</span> m,<span class="keywordtype">double</span> mm, <span class="keywordtype">double</span> n)
<a name="l00358"></a>00358 {
<a name="l00359"></a>00359     <span class="comment">// mean is prior + observed</span>
<a name="l00360"></a>00360     <span class="keywordflow">return</span> (k*m+mm)/(k+n);
<a name="l00361"></a>00361 }
<a name="l00362"></a>00362 <span class="comment"></span>
<a name="l00363"></a>00363 <span class="comment">/**</span>
<a name="l00364"></a>00364 <span class="comment"> * bayesian variance - prior + observed + shift in means</span>
<a name="l00365"></a>00365 <span class="comment"> *</span>
<a name="l00366"></a>00366 <span class="comment"> * @param v - prior pseudo-observations</span>
<a name="l00367"></a>00367 <span class="comment"> * @param pv - prior variance</span>
<a name="l00368"></a>00368 <span class="comment"> * @param ddx - sum of squares of observations</span>
<a name="l00369"></a>00369 <span class="comment"> * @param cc - sum of observations</span>
<a name="l00370"></a>00370 <span class="comment"> * @param nc - number of observations</span>
<a name="l00371"></a>00371 <span class="comment"> * @param k - prior pseudo-observations for mean</span>
<a name="l00372"></a>00372 <span class="comment"> * @param mcc - observed mean</span>
<a name="l00373"></a>00373 <span class="comment"> * @param pm - prior location for mean</span>
<a name="l00374"></a>00374 <span class="comment"> */</span>
<a name="l00375"></a><a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799">00375</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(<span class="keywordtype">double</span> v, <span class="keywordtype">double</span> pv, <span class="keywordtype">double</span> ddx, <span class="keywordtype">double</span> cc, <span class="keywordtype">double</span> nc, <span class="keywordtype">double</span> k, <span class="keywordtype">double</span> mcc, <span class="keywordtype">double</span> pm)
<a name="l00376"></a>00376 {
<a name="l00377"></a>00377     <span class="keywordtype">double</span> tmp;
<a name="l00378"></a>00378     <span class="comment">// variance is prior</span>
<a name="l00379"></a>00379     tmp = v*pv;
<a name="l00380"></a>00380     <span class="comment">// plus observed ssq sum(x^2)-sum(x)^2/n</span>
<a name="l00381"></a>00381     tmp += ddx-cc*cc/(nc+.0001);
<a name="l00382"></a>00382     <span class="comment">// plus change in means squared</span>
<a name="l00383"></a>00383     tmp += (k/(k+nc))*nc*(mcc-pm)*(mcc-pm);
<a name="l00384"></a>00384     <span class="comment">// divided by effective number of observations</span>
<a name="l00385"></a>00385     tmp = tmp/(v+nc);
<a name="l00386"></a>00386     <span class="keywordflow">return</span> (tmp);
<a name="l00387"></a>00387 }
<a name="l00388"></a>00388 <span class="comment"></span>
<a name="l00389"></a>00389 <span class="comment">/**</span>
<a name="l00390"></a>00390 <span class="comment"> * gaussian likelihood, in convenient running sums form</span>
<a name="l00391"></a>00391 <span class="comment"> *</span>
<a name="l00392"></a>00392 <span class="comment"> * @param xcx - sum of squares of observations</span>
<a name="l00393"></a>00393 <span class="comment"> * @param lc  - posterior mean</span>
<a name="l00394"></a>00394 <span class="comment"> * @param xc - sum of observations</span>
<a name="l00395"></a>00395 <span class="comment"> * @param yc - number of observations</span>
<a name="l00396"></a>00396 <span class="comment"> * @param dvc - posterior variance</span>
<a name="l00397"></a>00397 <span class="comment"> */</span>
<a name="l00398"></a><a class="code" href="snp_8label_8cpp.html#a2ca380c421378333fbafecc2304e27a6">00398</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#a2ca380c421378333fbafecc2304e27a6" title="gaussian likelihood, in convenient running sums form">l_posterior</a>(<span class="keywordtype">double</span> xcx, <span class="keywordtype">double</span> lc, <span class="keywordtype">double</span> xc, <span class="keywordtype">double</span> yc, <span class="keywordtype">double</span> dvc)
<a name="l00399"></a>00399 {
<a name="l00400"></a>00400     <span class="keywordtype">double</span> tmp;
<a name="l00401"></a>00401     tmp = (xcx-2*lc*xc+lc*lc*yc)/dvc + yc*log(dvc);
<a name="l00402"></a>00402     <span class="keywordflow">return</span>(tmp);
<a name="l00403"></a>00403 }
<a name="l00404"></a>00404 <span class="comment"></span>
<a name="l00405"></a>00405 <span class="comment">/**</span>
<a name="l00406"></a>00406 <span class="comment"> * gaussian likelihood, for just one piece of data</span>
<a name="l00407"></a>00407 <span class="comment"> * </span>
<a name="l00408"></a>00408 <span class="comment"> * @param m - mean</span>
<a name="l00409"></a>00409 <span class="comment"> * @param x - observed value</span>
<a name="l00410"></a>00410 <span class="comment"> * @param v - variance</span>
<a name="l00411"></a>00411 <span class="comment"> */</span>
<a name="l00412"></a><a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21">00412</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(<span class="keywordtype">double</span> m, <span class="keywordtype">double</span> x, <span class="keywordtype">double</span> v)
<a name="l00413"></a>00413 {
<a name="l00414"></a>00414     <span class="keywordflow">return</span> ((m-x)*(m-x)/v + log(v));
<a name="l00415"></a>00415 }
<a name="l00416"></a>00416 
<a name="l00417"></a>00417 <span class="keywordtype">double</span> l_two_normal(<span class="keywordtype">double</span> x, <span class="keywordtype">double</span> y, <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;cl, <span class="keywordtype">double</span> inflatePRA)
<a name="l00418"></a>00418 {
<a name="l00419"></a>00419     <span class="keywordtype">double</span> dx,dy,cs,r,z;
<a name="l00420"></a>00420     <span class="keywordtype">double</span> TWOPI = 6.28318531;
<a name="l00421"></a>00421     <span class="keywordtype">double</span> up;
<a name="l00422"></a>00422 
<a name="l00423"></a>00423     <span class="comment">// increase variance for calls</span>
<a name="l00424"></a>00424     up = 1 + inflatePRA/cl.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l00425"></a>00425     
<a name="l00426"></a>00426     dx = x-cl.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l00427"></a>00427     dy = y-cl.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l00428"></a>00428 
<a name="l00429"></a>00429     <span class="comment">// compute correlation</span>
<a name="l00430"></a>00430     cs = sqrt(cl.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*cl.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>);
<a name="l00431"></a>00431     r = cl.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>/cs;
<a name="l00432"></a>00432     
<a name="l00433"></a>00433     <span class="comment">// log-likelihood</span>
<a name="l00434"></a>00434     z = dx*dx/(up*cl.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>) - 2*r*dx*dy/(up*cs) + dy*dy/(up*cl.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>);
<a name="l00435"></a>00435     z /= 2*(1-r*r);
<a name="l00436"></a>00436     z += log(TWOPI*up*cs*sqrt(1-r*r));    
<a name="l00437"></a>00437     <span class="keywordflow">return</span>(z);
<a name="l00438"></a>00438 }
<a name="l00439"></a>00439 <span class="comment"></span>
<a name="l00440"></a>00440 <span class="comment">/**</span>
<a name="l00441"></a>00441 <span class="comment"> * cheap inverse-gamma likelihood, for one piece of data</span>
<a name="l00442"></a>00442 <span class="comment"> * </span>
<a name="l00443"></a>00443 <span class="comment"> * @param av - prior expected variance</span>
<a name="l00444"></a>00444 <span class="comment"> * @param dv - observed variance</span>
<a name="l00445"></a>00445 <span class="comment"> * @param v - pseudo-observations for prior</span>
<a name="l00446"></a>00446 <span class="comment"> */</span>
<a name="l00447"></a><a class="code" href="snp_8label_8cpp.html#ad9a6d2941c7fb47be8cd441cbb60c2d9">00447</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#ad9a6d2941c7fb47be8cd441cbb60c2d9" title="cheap inverse-gamma likelihood, for one piece of data">l_inverse</a>(<span class="keywordtype">double</span> av, <span class="keywordtype">double</span> dv, <span class="keywordtype">double</span> v)
<a name="l00448"></a>00448 {
<a name="l00449"></a>00449     <span class="keywordflow">return</span>(av/dv + (v+1)*log(dv));
<a name="l00450"></a>00450 }
<a name="l00451"></a>00451 <span class="comment"></span>
<a name="l00452"></a>00452 <span class="comment">/**</span>
<a name="l00453"></a>00453 <span class="comment"> * check if centers too close, return bignum if true</span>
<a name="l00454"></a>00454 <span class="comment"> *</span>
<a name="l00455"></a>00455 <span class="comment"> * @param a - cluster aa</span>
<a name="l00456"></a>00456 <span class="comment"> * @param b - cluster bb</span>
<a name="l00457"></a>00457 <span class="comment"> * @param h - cluster ab</span>
<a name="l00458"></a>00458 <span class="comment"> * @param shell - minimum distance squared for penalty</span>
<a name="l00459"></a>00459 <span class="comment"> */</span>
<a name="l00460"></a><a class="code" href="snp_8label_8cpp.html#a448b37d74fcf36fe83e37c69e5c74b2b">00460</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#a448b37d74fcf36fe83e37c69e5c74b2b" title="check if centers too close, return bignum if true">fix_shell</a>(<span class="keywordtype">double</span> a, <span class="keywordtype">double</span> b, <span class="keywordtype">double</span> h, <span class="keywordtype">double</span> shell)
<a name="l00461"></a>00461 {
<a name="l00462"></a>00462     <span class="comment">// hard shell between entries</span>
<a name="l00463"></a>00463     <span class="keywordflow">if</span> ((a-h)*(a-h)&lt;shell)
<a name="l00464"></a>00464         <span class="keywordflow">return</span>(100);
<a name="l00465"></a>00465     <span class="keywordflow">if</span> ((h-b)*(h-b)&lt;shell)
<a name="l00466"></a>00466         <span class="keywordflow">return</span>(100);
<a name="l00467"></a>00467     <span class="keywordflow">if</span> ((a-b)*(a-b)&lt;6*shell)
<a name="l00468"></a>00468         <span class="keywordflow">return</span>(100);
<a name="l00469"></a>00469     <span class="keywordflow">return</span>(0);
<a name="l00470"></a>00470 }
<a name="l00471"></a>00471 <span class="comment"></span>
<a name="l00472"></a>00472 <span class="comment">/**</span>
<a name="l00473"></a>00473 <span class="comment"> * prevents hom clusters from approaching midline</span>
<a name="l00474"></a>00474 <span class="comment"> * not useful on B chip due to asymmetric probe choice</span>
<a name="l00475"></a>00475 <span class="comment"> *</span>
<a name="l00476"></a>00476 <span class="comment"> * @param a - cluster aa mean</span>
<a name="l00477"></a>00477 <span class="comment"> * @param b - cluster bb mean</span>
<a name="l00478"></a>00478 <span class="comment"> * @param h - cluater ab mean</span>
<a name="l00479"></a>00479 <span class="comment"> * @param shell - homs can&#39;t be closer to center than this or penalty</span>
<a name="l00480"></a>00480 <span class="comment"> */</span>
<a name="l00481"></a><a class="code" href="snp_8label_8cpp.html#a6d67d855bb034343920e82fbaf1bdb2b">00481</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#a6d67d855bb034343920e82fbaf1bdb2b" title="prevents hom clusters from approaching midline not useful on B chip due to asymmetric probe choice...">fix_hom_shell</a>(<span class="keywordtype">double</span> a, <span class="keywordtype">double</span> b, <span class="keywordtype">double</span> h, <span class="keywordtype">double</span> shell)
<a name="l00482"></a>00482 {
<a name="l00483"></a>00483     <span class="comment">// hard shell between homs and midline</span>
<a name="l00484"></a>00484     <span class="keywordtype">double</span> tmp;
<a name="l00485"></a>00485     tmp = sqrt(shell);
<a name="l00486"></a>00486     <span class="keywordflow">if</span> (a&lt; tmp)
<a name="l00487"></a>00487         <span class="keywordflow">return</span>(100);
<a name="l00488"></a>00488     <span class="keywordflow">if</span> (b&gt; -1*tmp)
<a name="l00489"></a>00489         <span class="keywordflow">return</span>(100);
<a name="l00490"></a>00490     <span class="keywordflow">return</span>(0);
<a name="l00491"></a>00491 }
<a name="l00492"></a>00492 <span class="comment"></span>
<a name="l00493"></a>00493 <span class="comment">/**</span>
<a name="l00494"></a>00494 <span class="comment"> * Given 3 genotype probabilities, make a call/confidence output</span>
<a name="l00495"></a>00495 <span class="comment"> * </span>
<a name="l00496"></a>00496 <span class="comment"> * @call - call of final genotype 2,1,0</span>
<a name="l00497"></a>00497 <span class="comment"> * @conf - confidence of final genotypes 1-likelihood</span>
<a name="l00498"></a>00498 <span class="comment"> * @txd  - relative probability of d = b allele dominant</span>
<a name="l00499"></a>00499 <span class="comment"> * @txh  - relative probability of a+b allele</span>
<a name="l00500"></a>00500 <span class="comment"> * @txc  - relative probability of c = a allele dominant</span>
<a name="l00501"></a>00501 <span class="comment"> */</span>
<a name="l00502"></a><a class="code" href="snp_8label_8cpp.html#a2a658dc31870f851cdbdf18aa2d3d39e">00502</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a2a658dc31870f851cdbdf18aa2d3d39e" title="Given 3 genotype probabilities, make a call/confidence output.">make_a_call</a>(affx::GType &amp;call, <span class="keywordtype">double</span> &amp;conf, <span class="keywordtype">double</span> txd, <span class="keywordtype">double</span> txh, <span class="keywordtype">double</span> txc)
<a name="l00503"></a>00503 {
<a name="l00504"></a>00504     <span class="keywordtype">double</span> tmp;
<a name="l00505"></a>00505 
<a name="l00506"></a>00506   call = affx::BB;  <span class="comment">// 2;</span>
<a name="l00507"></a>00507   tmp = txc;
<a name="l00508"></a>00508   <span class="keywordflow">if</span> (txh&gt;tmp) {
<a name="l00509"></a>00509     call = affx::AB; <span class="comment">// 1</span>
<a name="l00510"></a>00510     tmp = txh;
<a name="l00511"></a>00511   }
<a name="l00512"></a>00512   <span class="keywordflow">if</span> (txd&gt;tmp) {
<a name="l00513"></a>00513     call = affx::AA; <span class="comment">// 0</span>
<a name="l00514"></a>00514     tmp =txd;    
<a name="l00515"></a>00515   }
<a name="l00516"></a>00516   conf = 1-tmp;
<a name="l00517"></a>00517 }
<a name="l00518"></a>00518     <span class="comment"></span>
<a name="l00519"></a>00519 <span class="comment">/**</span>
<a name="l00520"></a>00520 <span class="comment"> *  Given cluster parameters for each genotype, call data as genotypes</span>
<a name="l00521"></a>00521 <span class="comment"> *</span>
<a name="l00522"></a>00522 <span class="comment"> * @param call - output called genotypes</span>
<a name="l00523"></a>00523 <span class="comment"> * @param conf - output confidences for each call</span>
<a name="l00524"></a>00524 <span class="comment"> * @param x    - input data in contrast space</span>
<a name="l00525"></a>00525 <span class="comment"> * @param length - number of data points</span>
<a name="l00526"></a>00526 <span class="comment"> * @param rbma - aa cluster mean</span>
<a name="l00527"></a>00527 <span class="comment"> * @param rbva - aa cluster variance</span>
<a name="l00528"></a>00528 <span class="comment"> * @param rbmb - bb cluster mean</span>
<a name="l00529"></a>00529 <span class="comment"> * @param rbvb - bb cluster variance</span>
<a name="l00530"></a>00530 <span class="comment"> * @param rbmh - ab cluster mean</span>
<a name="l00531"></a>00531 <span class="comment"> * @param rbvh - ab cluster variance</span>
<a name="l00532"></a>00532 <span class="comment"> * @param cn - copy number - are we 1 copy (all homs) or two allowing hets</span>
<a name="l00533"></a>00533 <span class="comment"> */</span>
<a name="l00534"></a><a class="code" href="snp_8label_8cpp.html#a86cb22d116534c54810fb8ea68b1fb47">00534</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a86cb22d116534c54810fb8ea68b1fb47" title="Given cluster parameters for each genotype, call data as genotypes.">make_calls</a>(vector&lt;affx::GType&gt; &amp;call,
<a name="l00535"></a>00535                 vector&lt;double&gt; &amp;conf,
<a name="l00536"></a>00536                 vector&lt;double&gt; &amp;x,
<a name="l00537"></a>00537                 <span class="keywordtype">int</span> length,
<a name="l00538"></a>00538                 <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;pa,
<a name="l00539"></a>00539                 <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;ph,
<a name="l00540"></a>00540                 <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;pb,
<a name="l00541"></a>00541                 <span class="keywordtype">int</span> cn,
<a name="l00542"></a>00542                 <span class="keywordtype">int</span> freqflag,
<a name="l00543"></a>00543                 vector&lt;double&gt; &amp;vTxa,
<a name="l00544"></a>00544                 vector&lt;double&gt; &amp;vTxh,
<a name="l00545"></a>00545                 vector&lt;double&gt; &amp;vTxb,
<a name="l00546"></a>00546                 <span class="keywordtype">bool</span> store_probabilities
<a name="l00547"></a>00547     )
<a name="l00548"></a>00548 {
<a name="l00549"></a>00549     <span class="keywordtype">int</span> i;
<a name="l00550"></a>00550     <span class="keywordtype">double</span> tmp,sm;
<a name="l00551"></a>00551     <span class="keywordtype">double</span> txa,txb,txh;
<a name="l00552"></a>00552     <span class="keywordtype">double</span> lfreqa, lfreqb,lfreqh;
<a name="l00553"></a>00553 
<a name="l00554"></a>00554     <span class="comment">// precompute modification for frequency of call</span>
<a name="l00555"></a>00555     <span class="comment">// i.e. rare alleles less likely than high frequency alleles</span>
<a name="l00556"></a>00556     <span class="comment">// change this to use frequency instead of precision of mean</span>
<a name="l00557"></a>00557     lfreqa=0;
<a name="l00558"></a>00558     lfreqb=0;
<a name="l00559"></a>00559     lfreqh=0;
<a name="l00560"></a>00560     <span class="keywordflow">if</span> (freqflag)
<a name="l00561"></a>00561     {
<a name="l00562"></a>00562         <span class="comment">// log frequency</span>
<a name="l00563"></a>00563         <span class="comment">// will be normalized out when computing relative probability</span>
<a name="l00564"></a>00564         <span class="comment">// so no need to adjust here</span>
<a name="l00565"></a>00565         <span class="comment">// also note posterior is always &gt;0 </span>
<a name="l00566"></a>00566         <span class="comment">// note negation later</span>
<a name="l00567"></a>00567         lfreqa = -log(pa.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00568"></a>00568         lfreqb = -log(pb.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00569"></a>00569         lfreqh = -log(ph.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00570"></a>00570     }
<a name="l00571"></a>00571 
<a name="l00572"></a>00572     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00573"></a>00573     {
<a name="l00574"></a>00574         txa=<a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(x[i],pa.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,pa.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>)/2 + lfreqa;
<a name="l00575"></a>00575         txb=<a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(x[i],pb.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,pb.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>)/2 + lfreqb;
<a name="l00576"></a>00576         txh=<a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(x[i],ph.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,ph.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>)/2 + lfreqh;
<a name="l00577"></a>00577         tmp = txa;
<a name="l00578"></a>00578         <span class="keywordflow">if</span> (txh&lt;tmp &amp;&amp; cn&gt;1)
<a name="l00579"></a>00579             tmp=txh;
<a name="l00580"></a>00580         <span class="keywordflow">if</span> (txb&lt;tmp)
<a name="l00581"></a>00581             tmp =txb;
<a name="l00582"></a>00582         txa = txa-tmp;
<a name="l00583"></a>00583         txb = txb-tmp;
<a name="l00584"></a>00584         txh = txh-tmp;
<a name="l00585"></a>00585         txa = exp(-txa);
<a name="l00586"></a>00586         txb = exp(-txb);
<a name="l00587"></a>00587         txh = exp(-txh);
<a name="l00588"></a>00588         <span class="keywordflow">if</span> (cn&lt;2)
<a name="l00589"></a>00589             txh = 0; <span class="comment">// no chance of hets if 1 copy</span>
<a name="l00590"></a>00590         sm = txa+txb+txh;
<a name="l00591"></a>00591         txa /= sm;
<a name="l00592"></a>00592         txb /= sm;
<a name="l00593"></a>00593         txh /= sm;
<a name="l00594"></a>00594         <a class="code" href="snp_8label_8cpp.html#a2a658dc31870f851cdbdf18aa2d3d39e" title="Given 3 genotype probabilities, make a call/confidence output.">make_a_call</a>(call[i],conf[i],txa,txh,txb);
<a name="l00595"></a>00595         <span class="keywordflow">if</span> (store_probabilities) {
<a name="l00596"></a>00596             vTxa[i]=txa;
<a name="l00597"></a>00597             vTxh[i]=txh;
<a name="l00598"></a>00598             vTxb[i]=txb;            
<a name="l00599"></a>00599         }
<a name="l00600"></a>00600     }
<a name="l00601"></a>00601 
<a name="l00602"></a>00602   <span class="comment">// JHG -debug</span>
<a name="l00603"></a>00603   <span class="comment">// printVec(&quot;conf&quot;,conf);</span>
<a name="l00604"></a>00604   <span class="comment">// printVec(&quot;x&quot;,x);</span>
<a name="l00605"></a>00605 }
<a name="l00606"></a>00606 
<a name="l00607"></a>00607 <span class="comment">//make_two_calls(call,conf,x,y,x.size(),sp.posterior, sp.copynumber,sp.mix);</span>
<a name="l00608"></a>00608 <span class="comment"></span>
<a name="l00609"></a>00609 <span class="comment">/**</span>
<a name="l00610"></a>00610 <span class="comment"> *  Given cluster parameters for each genotype, call data as genotypes</span>
<a name="l00611"></a>00611 <span class="comment"> *</span>
<a name="l00612"></a>00612 <span class="comment"> * @param call - output called genotypes</span>
<a name="l00613"></a>00613 <span class="comment"> * @param conf - output confidences for each call</span>
<a name="l00614"></a>00614 <span class="comment"> * @param x    - input data in contrast space</span>
<a name="l00615"></a>00615 <span class="comment"> * @param length - number of data points</span>
<a name="l00616"></a>00616 <span class="comment"> * @param rbma - aa cluster mean</span>
<a name="l00617"></a>00617 <span class="comment"> * @param rbva - aa cluster variance</span>
<a name="l00618"></a>00618 <span class="comment"> * @param rbmb - bb cluster mean</span>
<a name="l00619"></a>00619 <span class="comment"> * @param rbvb - bb cluster variance</span>
<a name="l00620"></a>00620 <span class="comment"> * @param rbmh - ab cluster mean</span>
<a name="l00621"></a>00621 <span class="comment"> * @param rbvh - ab cluster variance</span>
<a name="l00622"></a>00622 <span class="comment"> * @param cn - copy number - are we 1 copy (all homs) or two allowing hets</span>
<a name="l00623"></a>00623 <span class="comment"> */</span>
<a name="l00624"></a><a class="code" href="snp_8label_8cpp.html#a64093c4de47f32c8b4821a5b3c37dab3">00624</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a64093c4de47f32c8b4821a5b3c37dab3" title="Given cluster parameters for each genotype, call data as genotypes.">make_two_calls</a>(vector&lt;affx::GType&gt; &amp;call,
<a name="l00625"></a>00625                     vector&lt;double&gt; &amp;conf,
<a name="l00626"></a>00626                     vector&lt;double&gt; &amp;x,
<a name="l00627"></a>00627                     vector&lt;double&gt; &amp;y,
<a name="l00628"></a>00628                     <span class="keywordtype">int</span> length,
<a name="l00629"></a>00629                     <a class="code" href="classsnp__distribution.html" title="this class holds all three genotype clusters">snp_distribution</a> &amp;p,
<a name="l00630"></a>00630                     <span class="keywordtype">int</span> cn,
<a name="l00631"></a>00631                     <span class="keywordtype">int</span> freqflag,
<a name="l00632"></a>00632                     <span class="keywordtype">double</span> ocean,
<a name="l00633"></a>00633                     <span class="keywordtype">double</span> inflatePRA,
<a name="l00634"></a>00634                     vector&lt;double&gt; &amp;vTxa,
<a name="l00635"></a>00635                     vector&lt;double&gt; &amp;vTxh,
<a name="l00636"></a>00636                     vector&lt;double&gt; &amp;vTxb,
<a name="l00637"></a>00637                     <span class="keywordtype">bool</span> store_probabilities
<a name="l00638"></a>00638     )
<a name="l00639"></a>00639 {
<a name="l00640"></a>00640     <span class="keywordtype">int</span> i;
<a name="l00641"></a>00641     <span class="keywordtype">double</span> tmp,sm;
<a name="l00642"></a>00642     <span class="keywordtype">double</span> txa,txb,txh,txt;
<a name="l00643"></a>00643     <span class="keywordtype">double</span> lfreqa, lfreqb,lfreqh, ltot;
<a name="l00644"></a>00644 
<a name="l00645"></a>00645 
<a name="l00646"></a>00646     <span class="comment">// precompute modification for frequency of call</span>
<a name="l00647"></a>00647     <span class="comment">// i.e. rare alleles less likely than outliers to high frequency alleles</span>
<a name="l00648"></a>00648     <span class="comment">// Use frequency here, no longer reuse posterior precision of mean</span>
<a name="l00649"></a>00649     lfreqa=0;
<a name="l00650"></a>00650     lfreqb=0;
<a name="l00651"></a>00651     lfreqh=0;
<a name="l00652"></a>00652     <span class="keywordflow">if</span> (freqflag)
<a name="l00653"></a>00653     {
<a name="l00654"></a>00654         <span class="comment">// log frequency</span>
<a name="l00655"></a>00655         <span class="comment">// also note posterior is always &gt;0 </span>
<a name="l00656"></a>00656         <span class="comment">// f not working?</span>
<a name="l00657"></a>00657         <span class="comment">// note negation later</span>
<a name="l00658"></a>00658         lfreqa = -log(p.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00659"></a>00659         lfreqb = -log(p.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00660"></a>00660         lfreqh = -log(p.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00661"></a>00661         ltot = log(p.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>+p.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>+p.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>);
<a name="l00662"></a>00662         lfreqa += ltot;
<a name="l00663"></a>00663         lfreqb += ltot;
<a name="l00664"></a>00664         lfreqh += ltot;
<a name="l00665"></a>00665     }
<a name="l00666"></a>00666 
<a name="l00667"></a>00667     <span class="comment">// want to maintain absolute scale in case comparing to &#39;absolute&#39; issues</span>
<a name="l00668"></a>00668     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00669"></a>00669     {
<a name="l00670"></a>00670         txa=l_two_normal(x[i],y[i],p.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>,inflatePRA) + lfreqa;
<a name="l00671"></a>00671         txb=l_two_normal(x[i],y[i],p.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>,inflatePRA) + lfreqb;
<a name="l00672"></a>00672         txh=l_two_normal(x[i],y[i],p.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>,inflatePRA) + lfreqh;
<a name="l00673"></a>00673 
<a name="l00674"></a>00674         <span class="comment">//cout &lt;&lt; txa &lt;&lt; &quot;\t&quot; &lt;&lt; txh &lt;&lt; &quot;\t&quot; &lt;&lt; txb &lt;&lt; endl;</span>
<a name="l00675"></a>00675         <span class="comment">// find the best of three</span>
<a name="l00676"></a>00676         tmp = txa;
<a name="l00677"></a>00677         <span class="keywordflow">if</span> (txh&lt;tmp &amp;&amp; cn&gt;1)
<a name="l00678"></a>00678             tmp=txh;
<a name="l00679"></a>00679         <span class="keywordflow">if</span> (txb&lt;tmp)
<a name="l00680"></a>00680             tmp =txb;
<a name="l00681"></a>00681         txa = txa-tmp;
<a name="l00682"></a>00682         txb = txb-tmp;
<a name="l00683"></a>00683         txh = txh-tmp;
<a name="l00684"></a>00684         txa = exp(-txa);
<a name="l00685"></a>00685         txb = exp(-txb);
<a name="l00686"></a>00686         txh = exp(-txh);
<a name="l00687"></a>00687         txt = ocean*exp(tmp); <span class="comment">// how deep is the uniform ocean relative to these guys</span>
<a name="l00688"></a>00688         <span class="comment">//cout &lt;&lt; txa &lt;&lt; &quot;\t&quot; &lt;&lt; txh &lt;&lt; &quot;\t&quot; &lt;&lt; txb &lt;&lt; endl;</span>
<a name="l00689"></a>00689         <span class="keywordflow">if</span> (cn&lt;2)
<a name="l00690"></a>00690             txh = 0; <span class="comment">// no chance of hets if 1 copy</span>
<a name="l00691"></a>00691         <span class="comment">// rescale by ocean to reduce confidence</span>
<a name="l00692"></a>00692         sm = txa+txb+txh+txt;
<a name="l00693"></a>00693         txa /= sm;
<a name="l00694"></a>00694         txb /= sm;
<a name="l00695"></a>00695         txh /= sm;
<a name="l00696"></a>00696         
<a name="l00697"></a>00697         <span class="comment">//cout &lt;&lt; txa &lt;&lt; &quot;\t&quot; &lt;&lt; txh &lt;&lt; &quot;\t&quot; &lt;&lt; txb &lt;&lt; endl;</span>
<a name="l00698"></a>00698         <a class="code" href="snp_8label_8cpp.html#a2a658dc31870f851cdbdf18aa2d3d39e" title="Given 3 genotype probabilities, make a call/confidence output.">make_a_call</a>(call[i],conf[i],txa,txh,txb);
<a name="l00699"></a>00699         <span class="keywordflow">if</span> (store_probabilities) {
<a name="l00700"></a>00700             vTxa[i]=txa;
<a name="l00701"></a>00701             vTxh[i]=txh;
<a name="l00702"></a>00702             vTxb[i]=txb;
<a name="l00703"></a>00703         }
<a name="l00704"></a>00704     }
<a name="l00705"></a>00705 
<a name="l00706"></a>00706   <span class="comment">// JHG -debug</span>
<a name="l00707"></a>00707   <span class="comment">// printVec(&quot;conf&quot;,conf);</span>
<a name="l00708"></a>00708   <span class="comment">// printVec(&quot;x&quot;,x);</span>
<a name="l00709"></a>00709 }
<a name="l00710"></a>00710 <span class="comment"></span>
<a name="l00711"></a>00711 <span class="comment">/** make utility for comparing pairs */</span>
<a name="l00712"></a><a class="code" href="structfirst__less.html">00712</a> <span class="keyword">struct </span><a class="code" href="structfirst__less.html" title="make utility for comparing pairs">first_less</a> : <span class="keyword">public</span> binary_function&lt;pair&lt;int,double&gt;,pair&lt;int,double&gt;,bool&gt;
<a name="l00713"></a>00713 {
<a name="l00714"></a>00714     <span class="keywordtype">bool</span> operator() (<span class="keyword">const</span> pair&lt;int,double&gt;&amp; x, <span class="keyword">const</span> pair&lt;int,double&gt;&amp;y)<span class="keyword"> const</span>
<a name="l00715"></a>00715 <span class="keyword">    </span>{
<a name="l00716"></a>00716         <span class="keywordflow">return</span>(x.first &lt;y.first);
<a name="l00717"></a>00717     }
<a name="l00718"></a>00718 };
<a name="l00719"></a>00719 <span class="comment"></span>
<a name="l00720"></a>00720 <span class="comment">/** utility for comparing pairs */</span>
<a name="l00721"></a><a class="code" href="structsecond__less.html">00721</a> <span class="keyword">struct </span><a class="code" href="structsecond__less.html" title="utility for comparing pairs">second_less</a> : <span class="keyword">public</span> binary_function&lt;pair&lt;int,double&gt;,pair&lt;int,double&gt;,bool&gt;
<a name="l00722"></a>00722 {
<a name="l00723"></a>00723     <span class="keywordtype">bool</span> operator()(<span class="keyword">const</span> pair&lt;int,double&gt;&amp; x,<span class="keyword">const</span> pair&lt;int,double&gt;&amp;y)<span class="keyword"> const</span>
<a name="l00724"></a>00724 <span class="keyword">    </span>{
<a name="l00725"></a>00725         <span class="keywordflow">return</span>(x.second&lt;y.second);
<a name="l00726"></a>00726     }
<a name="l00727"></a>00727 };    
<a name="l00728"></a>00728 <span class="comment"></span>
<a name="l00729"></a>00729 <span class="comment">/**</span>
<a name="l00730"></a>00730 <span class="comment"> * make calls using the final labelings - primarily a matter of unsorting the contrast values</span>
<a name="l00731"></a>00731 <span class="comment"> *</span>
<a name="l00732"></a>00732 <span class="comment"> * @param call - output called genotypes</span>
<a name="l00733"></a>00733 <span class="comment"> * @param conf - confidence of call</span>
<a name="l00734"></a>00734 <span class="comment"> * @param x - raw contrast values (unsorted)</span>
<a name="l00735"></a>00735 <span class="comment"> * @param length - number of data points</span>
<a name="l00736"></a>00736 <span class="comment"> * @param ed - estimated relative probability for d = bb genotype (sorted)</span>
<a name="l00737"></a>00737 <span class="comment"> * @param ec - estimated relative probability for c = aa genotype (sorted)</span>
<a name="l00738"></a>00738 <span class="comment"> * @param eh - estimated relative probability for h = ab genotype (sorted)</span>
<a name="l00739"></a>00739 <span class="comment"> * @param cn - copynumber 1 = homs only, copynumber 2 = hets allowed</span>
<a name="l00740"></a>00740 <span class="comment"> */</span>
<a name="l00741"></a><a class="code" href="snp_8label_8cpp.html#a10bb8cda395efca9a0885ab042920921">00741</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a10bb8cda395efca9a0885ab042920921" title="make calls using the final labelings - primarily a matter of unsorting the contrast values...">make_label_calls</a>(vector&lt;affx::GType&gt; &amp;call, vector&lt;double&gt; &amp;conf, vector&lt;double&gt; &amp;x, 
<a name="l00742"></a>00742                       <span class="keywordtype">int</span> length, vector&lt;double&gt; &amp;ed, vector&lt;double&gt; &amp;ec, vector&lt;double&gt; &amp;eh, <span class="keywordtype">int</span> cn)
<a name="l00743"></a>00743 {
<a name="l00744"></a>00744     <span class="comment">// z = sorted values</span>
<a name="l00745"></a>00745     <span class="comment">// ed, ec, eh = cluster memberships</span>
<a name="l00746"></a>00746     <span class="comment">// x = values to be matched with z</span>
<a name="l00747"></a>00747     <span class="comment">// note: for future work, this should be &quot;bin&quot; compatible</span>
<a name="l00748"></a>00748 
<a name="l00749"></a>00749     <span class="keywordtype">int</span> i,j;
<a name="l00750"></a>00750     vector&lt; pair&lt;int,double&gt; &gt; data;
<a name="l00751"></a>00751 
<a name="l00752"></a>00752     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00753"></a>00753     {
<a name="l00754"></a>00754         pair&lt;int,double&gt; p;
<a name="l00755"></a>00755         p.first = i;
<a name="l00756"></a>00756         p.second = x[i];
<a name="l00757"></a>00757         data.push_back(p);
<a name="l00758"></a>00758     }
<a name="l00759"></a>00759 
<a name="l00760"></a>00760     sort(data.begin(),data.end(),<a class="code" href="structsecond__less.html" title="utility for comparing pairs">second_less</a>());
<a name="l00761"></a>00761 
<a name="l00762"></a>00762     <span class="keywordflow">for</span> (i=0; i&lt;length; i++)
<a name="l00763"></a>00763     {
<a name="l00764"></a>00764         <span class="comment">// for each value in data (sorted x)</span>
<a name="l00765"></a>00765         <span class="comment">// find the corresponding z value</span>
<a name="l00766"></a>00766         <span class="comment">// place the corresponding call/conf for x</span>
<a name="l00767"></a>00767         j = data[i].first;
<a name="l00768"></a>00768         <span class="keywordflow">if</span> (cn&gt;1)
<a name="l00769"></a>00769             <a class="code" href="snp_8label_8cpp.html#a2a658dc31870f851cdbdf18aa2d3d39e" title="Given 3 genotype probabilities, make a call/confidence output.">make_a_call</a>(call[j],conf[j],ec[i],eh[i],ed[i]);    
<a name="l00770"></a>00770         <span class="keywordflow">else</span>
<a name="l00771"></a>00771             <a class="code" href="snp_8label_8cpp.html#a2a658dc31870f851cdbdf18aa2d3d39e" title="Given 3 genotype probabilities, make a call/confidence output.">make_a_call</a>(call[j],conf[j],ec[i],0,ed[i]);
<a name="l00772"></a>00772     }
<a name="l00773"></a>00773 }
<a name="l00774"></a>00774 <span class="comment"></span>
<a name="l00775"></a>00775 <span class="comment">/**</span>
<a name="l00776"></a>00776 <span class="comment"> * assigns the sorted constrast data into &quot;bins&quot; for fast processing</span>
<a name="l00777"></a>00777 <span class="comment"> *</span>
<a name="l00778"></a>00778 <span class="comment"> * @param zall - the bins of the data, number per bin, sums, sum-squares, hints</span>
<a name="l00779"></a>00779 <span class="comment"> * @param z - the raw data, sorted by constrast</span>
<a name="l00780"></a>00780 <span class="comment"> * @param zerohints - what are the reference genotypes supplied</span>
<a name="l00781"></a>00781 <span class="comment"> * @param zlength - how many values in the data</span>
<a name="l00782"></a>00782 <span class="comment"> * @param bins - how many bins at most for the data</span>
<a name="l00783"></a>00783 <span class="comment"> * @param hok - can homs be flipped in hints?</span>
<a name="l00784"></a>00784 <span class="comment"> */</span>
<a name="l00785"></a><a class="code" href="snp_8label_8cpp.html#a1a85fb40b35261384899759ca5a83b66">00785</a> <span class="keywordtype">int</span> <a class="code" href="snp_8label_8cpp.html#a1a85fb40b35261384899759ca5a83b66" title="assigns the sorted constrast data into &amp;quot;bins&amp;quot; for fast processing">setup_bins</a>(<a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall, vector&lt;double&gt; &amp;z, vector&lt;double&gt; &amp;w, vector&lt;int&gt; &amp;zerohints, vector&lt;double&gt; &amp;zeroInbred, <span class="keywordtype">int</span> zlength, <span class="keywordtype">int</span> bins, <span class="keywordtype">int</span> hok, <span class="keywordtype">double</span> Cpenalty)
<a name="l00786"></a>00786 {
<a name="l00787"></a>00787 
<a name="l00788"></a>00788     <span class="keywordtype">double</span> delta,boundary;
<a name="l00789"></a>00789     <span class="keywordtype">int</span> ti,i;
<a name="l00790"></a>00790 
<a name="l00791"></a>00791     delta = (z[zlength-1]-z[0])/(bins+1);
<a name="l00792"></a>00792 
<a name="l00793"></a>00793     zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>[0] = 0;
<a name="l00794"></a>00794     zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>[0]= 0;
<a name="l00795"></a>00795     zall.<a class="code" href="classbinned__data.html#ac35086a3cf907ea147acd035c1f6ad4d" title="sum of y values in each bin">zy</a>[0] = 0;
<a name="l00796"></a>00796     zall.<a class="code" href="classbinned__data.html#af76af2898cfd28e3ef8f9722144ef60f" title="sum of squares of y values in each bin">zzy</a>[0] = 0;
<a name="l00797"></a>00797     zall.<a class="code" href="classbinned__data.html#a6d45afa8716fc45698c70b5862bbb118" title="sum of x*y values in each bin">zxy</a>[0] = 0;
<a name="l00798"></a>00798     zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>[0] = 0;
<a name="l00799"></a>00799     zall.<a class="code" href="classbinned__data.html#a60ff501908d011c168b7621678d1dacf" title="penalty*number of hint genotypes not d">nod</a>[0] = 0;
<a name="l00800"></a>00800     zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>[0] = 0;    
<a name="l00801"></a>00801     zall.<a class="code" href="classbinned__data.html#aa06f7ef273167f48d09163cfa0bca595" title="penalty*number of hint genotypes not c">noc</a>[0] = 0;
<a name="l00802"></a>00802     ti = 0;
<a name="l00803"></a>00803     boundary = z[0]-1;
<a name="l00804"></a>00804     <span class="keywordflow">for</span> (i=0; i&lt;zlength; i++)
<a name="l00805"></a>00805     {
<a name="l00806"></a>00806         <span class="comment">// when do we go to the next thing?</span>
<a name="l00807"></a>00807         <span class="comment">// when a) bins is turned off</span>
<a name="l00808"></a>00808         <span class="comment">// b) when fewer points than bins</span>
<a name="l00809"></a>00809         <span class="comment">// c) when point outside domain</span>
<a name="l00810"></a>00810         <span class="keywordflow">if</span> (z[i]&gt;boundary || bins==0 || zlength&lt;bins)
<a name="l00811"></a>00811         {
<a name="l00812"></a>00812             ti++;
<a name="l00813"></a>00813             boundary = z[i]+delta;
<a name="l00814"></a>00814             <span class="comment">// initialize the new bin</span>
<a name="l00815"></a>00815             zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>[ti] = 0;
<a name="l00816"></a>00816             zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>[ti] = 0;
<a name="l00817"></a>00817             zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>[ti] = 0;
<a name="l00818"></a>00818             zall.<a class="code" href="classbinned__data.html#ac35086a3cf907ea147acd035c1f6ad4d" title="sum of y values in each bin">zy</a>[ti] = 0;
<a name="l00819"></a>00819             zall.<a class="code" href="classbinned__data.html#af76af2898cfd28e3ef8f9722144ef60f" title="sum of squares of y values in each bin">zzy</a>[ti] = 0;
<a name="l00820"></a>00820             zall.<a class="code" href="classbinned__data.html#a6d45afa8716fc45698c70b5862bbb118" title="sum of x*y values in each bin">zxy</a>[ti] = 0;
<a name="l00821"></a>00821             <span class="comment">// hints</span>
<a name="l00822"></a>00822             zall.<a class="code" href="classbinned__data.html#a60ff501908d011c168b7621678d1dacf" title="penalty*number of hint genotypes not d">nod</a>[ti] = 0;    
<a name="l00823"></a>00823             zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>[ti] = 0;
<a name="l00824"></a>00824             zall.<a class="code" href="classbinned__data.html#aa06f7ef273167f48d09163cfa0bca595" title="penalty*number of hint genotypes not c">noc</a>[ti] = 0;
<a name="l00825"></a>00825         }
<a name="l00826"></a>00826         <span class="comment">// add current point to bins</span>
<a name="l00827"></a>00827         zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>[ti]  += 1;
<a name="l00828"></a>00828         zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>[ti]  += z[i];
<a name="l00829"></a>00829         zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>[ti] += z[i]*z[i];
<a name="l00830"></a>00830         zall.<a class="code" href="classbinned__data.html#ac35086a3cf907ea147acd035c1f6ad4d" title="sum of y values in each bin">zy</a>[ti]  += w[i];
<a name="l00831"></a>00831         zall.<a class="code" href="classbinned__data.html#af76af2898cfd28e3ef8f9722144ef60f" title="sum of squares of y values in each bin">zzy</a>[ti] += w[i]*w[i];
<a name="l00832"></a>00832         zall.<a class="code" href="classbinned__data.html#a6d45afa8716fc45698c70b5862bbb118" title="sum of x*y values in each bin">zxy</a>[ti] += z[i]*w[i];
<a name="l00833"></a>00833         <span class="comment">// count contradictions</span>
<a name="l00834"></a>00834         <span class="comment">// if no information, no contradictions</span>
<a name="l00835"></a>00835         <span class="keywordflow">if</span> (zerohints[i]==0)
<a name="l00836"></a>00836         {
<a name="l00837"></a>00837             <span class="keywordflow">if</span> (hok&lt;1)
<a name="l00838"></a>00838                 zall.<a class="code" href="classbinned__data.html#a60ff501908d011c168b7621678d1dacf" title="penalty*number of hint genotypes not d">nod</a>[ti] +=Cpenalty;
<a name="l00839"></a>00839             zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>[ti] +=Cpenalty;
<a name="l00840"></a>00840         }
<a name="l00841"></a>00841         <span class="keywordflow">if</span> (zerohints[i]==1)
<a name="l00842"></a>00842         {
<a name="l00843"></a>00843             zall.<a class="code" href="classbinned__data.html#a60ff501908d011c168b7621678d1dacf" title="penalty*number of hint genotypes not d">nod</a>[ti] +=Cpenalty;
<a name="l00844"></a>00844             zall.<a class="code" href="classbinned__data.html#aa06f7ef273167f48d09163cfa0bca595" title="penalty*number of hint genotypes not c">noc</a>[ti] +=Cpenalty;
<a name="l00845"></a>00845         }
<a name="l00846"></a>00846         <span class="keywordflow">if</span> (zerohints[i]==2)
<a name="l00847"></a>00847         {
<a name="l00848"></a>00848             zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>[ti] +=Cpenalty;
<a name="l00849"></a>00849             <span class="keywordflow">if</span> (hok&lt;1)
<a name="l00850"></a>00850                 zall.<a class="code" href="classbinned__data.html#aa06f7ef273167f48d09163cfa0bca595" title="penalty*number of hint genotypes not c">noc</a>[ti] +=Cpenalty;
<a name="l00851"></a>00851         }
<a name="l00852"></a>00852         <span class="comment">// bin ti contains item [i] which has a bias for or against hets by being inbred.</span>
<a name="l00853"></a>00853         zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>[ti] += zeroInbred[i];
<a name="l00854"></a>00854     }
<a name="l00855"></a>00855     
<a name="l00856"></a>00856      <span class="comment">// ti indicates last point here</span>
<a name="l00857"></a>00857     <span class="comment">// how much did we actually use</span>
<a name="l00858"></a>00858     <span class="keywordflow">return</span>(ti+1);
<a name="l00859"></a>00859 }
<a name="l00860"></a>00860 <span class="comment"></span>
<a name="l00861"></a>00861 <span class="comment">/**</span>
<a name="l00862"></a>00862 <span class="comment"> * generate sums of data based on weighted cluster membership</span>
<a name="l00863"></a>00863 <span class="comment"> * </span>
<a name="l00864"></a>00864 <span class="comment"> * @param n - effective number of data points </span>
<a name="l00865"></a>00865 <span class="comment"> * @param m - effective sum of data points</span>
<a name="l00866"></a>00866 <span class="comment"> * @param s - effective sum of squares of data points</span>
<a name="l00867"></a>00867 <span class="comment"> * @param zall - binned contrast data for points</span>
<a name="l00868"></a>00868 <span class="comment"> * @param wi - weight of bins</span>
<a name="l00869"></a>00869 <span class="comment"> * @param ZS - number of bins</span>
<a name="l00870"></a>00870 <span class="comment"> */</span>
<a name="l00871"></a><a class="code" href="snp_8label_8cpp.html#a2cf95e5f494d77911161becc1518ef95">00871</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a2cf95e5f494d77911161becc1518ef95" title="generate sums of data based on weighted cluster membership">compute_binned_cluster</a>(<span class="keywordtype">double</span> &amp;n, <span class="keywordtype">double</span> &amp;m, <span class="keywordtype">double</span> &amp;s, <a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall,
<a name="l00872"></a>00872          vector&lt;double&gt; &amp;wi, <span class="keywordtype">int</span> ZS)
<a name="l00873"></a>00873 {
<a name="l00874"></a>00874     n=0;
<a name="l00875"></a>00875     m=0;
<a name="l00876"></a>00876     s=0;
<a name="l00877"></a>00877 
<a name="l00878"></a>00878     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;ZS-1; i++)
<a name="l00879"></a>00879     {
<a name="l00880"></a>00880         n +=  zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>[i+1]*wi[i];
<a name="l00881"></a>00881         m +=  zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>[i+1]*wi[i];
<a name="l00882"></a>00882         s +=  zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>[i+1]*wi[i];
<a name="l00883"></a>00883     }
<a name="l00884"></a>00884 }
<a name="l00885"></a>00885 <span class="comment"></span>
<a name="l00886"></a>00886 <span class="comment">/**</span>
<a name="l00887"></a>00887 <span class="comment"> * generate sums of data based on weighted cluster membership</span>
<a name="l00888"></a>00888 <span class="comment"> * </span>
<a name="l00889"></a>00889 <span class="comment"> * @param n - effective number of data points </span>
<a name="l00890"></a>00890 <span class="comment"> * @param m - effective sum of data points</span>
<a name="l00891"></a>00891 <span class="comment"> * @param s - effective sum of squares of data points</span>
<a name="l00892"></a>00892 <span class="comment"> * @param zall - binned contrast data for points</span>
<a name="l00893"></a>00893 <span class="comment"> * @param wi - weight of bins</span>
<a name="l00894"></a>00894 <span class="comment"> * @param ZS - number of bins</span>
<a name="l00895"></a>00895 <span class="comment"> */</span>
<a name="l00896"></a><a class="code" href="snp_8label_8cpp.html#a35346d828a53f326d0e58596c6067c07">00896</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a35346d828a53f326d0e58596c6067c07" title="generate sums of data based on weighted cluster membership">compute_two_binned_cluster</a>(vector&lt;double&gt; &amp;six, <a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall,
<a name="l00897"></a>00897          vector&lt;double&gt; &amp;wi, <span class="keywordtype">int</span> ZS)
<a name="l00898"></a>00898 {
<a name="l00899"></a>00899     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j=0; j&lt;6; j++)
<a name="l00900"></a>00900         six[j] = 0;
<a name="l00901"></a>00901     <span class="comment">/*n=0;</span>
<a name="l00902"></a>00902 <span class="comment">    m=0;</span>
<a name="l00903"></a>00903 <span class="comment">    s=0;</span>
<a name="l00904"></a>00904 <span class="comment">    my=0;</span>
<a name="l00905"></a>00905 <span class="comment">    sy=0;</span>
<a name="l00906"></a>00906 <span class="comment">    xy=0;*/</span>
<a name="l00907"></a>00907     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;ZS-1; i++)
<a name="l00908"></a>00908     {
<a name="l00909"></a>00909         six[0] +=  zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>[i+1]*wi[i];
<a name="l00910"></a>00910         six[1] +=  zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>[i+1]*wi[i];
<a name="l00911"></a>00911         six[2] +=  zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>[i+1]*wi[i];
<a name="l00912"></a>00912         six[3] +=  zall.<a class="code" href="classbinned__data.html#ac35086a3cf907ea147acd035c1f6ad4d" title="sum of y values in each bin">zy</a>[i+1]*wi[i];
<a name="l00913"></a>00913         six[4] +=  zall.<a class="code" href="classbinned__data.html#af76af2898cfd28e3ef8f9722144ef60f" title="sum of squares of y values in each bin">zzy</a>[i+1]*wi[i];
<a name="l00914"></a>00914         six[5] +=  zall.<a class="code" href="classbinned__data.html#a6d45afa8716fc45698c70b5862bbb118" title="sum of x*y values in each bin">zxy</a>[i+1]*wi[i];
<a name="l00915"></a>00915     }
<a name="l00916"></a>00916 }
<a name="l00917"></a>00917 <span class="comment"></span>
<a name="l00918"></a>00918 <span class="comment">/**</span>
<a name="l00919"></a>00919 <span class="comment"> * take probabilities of bins and turn them into probabilities of individual points</span>
<a name="l00920"></a>00920 <span class="comment"> *</span>
<a name="l00921"></a>00921 <span class="comment"> * @param tex - output per point</span>
<a name="l00922"></a>00922 <span class="comment"> * @param length - number of real data points</span>
<a name="l00923"></a>00923 <span class="comment"> * @param ex - binned data</span>
<a name="l00924"></a>00924 <span class="comment"> * @param nz - number of points per bin</span>
<a name="l00925"></a>00925 <span class="comment"> * @param ZS - number of bins</span>
<a name="l00926"></a>00926 <span class="comment"> */</span>
<a name="l00927"></a><a class="code" href="snp_8label_8cpp.html#a0290a1a25d31fc26c81588daca77a8f2">00927</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a0290a1a25d31fc26c81588daca77a8f2" title="take probabilities of bins and turn them into probabilities of individual points">expand_data</a>(vector&lt;double&gt; &amp;tex, <span class="keywordtype">int</span> length,vector&lt;double&gt; &amp;ex,vector&lt;double&gt; &amp;nz,<span class="keywordtype">int</span> ZS){
<a name="l00928"></a>00928     <span class="keywordtype">int</span> ti,tj,i;
<a name="l00929"></a>00929     i=0;
<a name="l00930"></a>00930     <span class="comment">// for each bin</span>
<a name="l00931"></a>00931     <span class="keywordflow">for</span> (ti=0; ti&lt;ZS-1; ti++)
<a name="l00932"></a>00932     {
<a name="l00933"></a>00933         <span class="comment">// for each data point in bin</span>
<a name="l00934"></a>00934         <span class="keywordflow">for</span> (tj=0; tj&lt;floor(nz[ti+1]+.0001); tj++)
<a name="l00935"></a>00935         {
<a name="l00936"></a>00936             <span class="comment">// expanded value is value for bin</span>
<a name="l00937"></a>00937             tex[i] = ex[ti];
<a name="l00938"></a>00938             i++;
<a name="l00939"></a>00939         }
<a name="l00940"></a>00940     }
<a name="l00941"></a>00941 }
<a name="l00942"></a>00942 <span class="comment"></span>
<a name="l00943"></a>00943 <span class="comment">/**</span>
<a name="l00944"></a>00944 <span class="comment"> *  convert log-likelihood of all labelings into relative probability</span>
<a name="l00945"></a>00945 <span class="comment"> * </span>
<a name="l00946"></a>00946 <span class="comment"> * @param q - matrix to be converted</span>
<a name="l00947"></a>00947 <span class="comment"> * @param ZS - edge size of matrix</span>
<a name="l00948"></a>00948 <span class="comment"> */</span>
<a name="l00949"></a><a class="code" href="snp_8label_8cpp.html#a8e097ae837d7e8b8ff40b974a07570aa">00949</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a8e097ae837d7e8b8ff40b974a07570aa" title="convert log-likelihood of all labelings into relative probability">qtoeq</a>(vector&lt;double&gt; &amp;q, <span class="keywordtype">int</span> ZS)
<a name="l00950"></a>00950 {
<a name="l00951"></a>00951     <span class="keywordtype">double</span> qmin;
<a name="l00952"></a>00952     <span class="keywordtype">int</span> tin;
<a name="l00953"></a>00953     <span class="keywordtype">int</span> i,j;
<a name="l00954"></a>00954 
<a name="l00955"></a>00955 <span class="comment">// compute minimum of useful values</span>
<a name="l00956"></a>00956     <span class="comment">// so as to rescale log-likelihood to relative levels</span>
<a name="l00957"></a>00957     qmin = q[0];
<a name="l00958"></a>00958     <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l00959"></a>00959         <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l00960"></a>00960         {
<a name="l00961"></a>00961             tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l00962"></a>00962             <span class="keywordflow">if</span> (qmin&gt;q[tin])
<a name="l00963"></a>00963                 qmin=q[tin];
<a name="l00964"></a>00964         }
<a name="l00965"></a>00965     <span class="comment">// q, eq</span>
<a name="l00966"></a>00966     <span class="keywordflow">for</span> (i=0; i&lt;q.size(); i++)
<a name="l00967"></a>00967     {
<a name="l00968"></a>00968         q[i] = qmin-q[i];
<a name="l00969"></a>00969         q[i] = exp(q[i]);
<a name="l00970"></a>00970     }
<a name="l00971"></a>00971     <span class="comment">// mask out useless values</span>
<a name="l00972"></a>00972     <a class="code" href="snp_8label_8cpp.html#a0753c4704b4a8316c7aec8ad57f1e59f" title="mask out a portion of an array">vmask</a>(q,ZS,0);
<a name="l00973"></a>00973 }
<a name="l00974"></a>00974 <span class="comment"></span>
<a name="l00975"></a>00975 <span class="comment">/**</span>
<a name="l00976"></a>00976 <span class="comment"> * turn the overall relative probability of labelings into relative probability of individual genotypes</span>
<a name="l00977"></a>00977 <span class="comment"> *</span>
<a name="l00978"></a>00978 <span class="comment"> * @param ec - relative probability of c = a allele dominant</span>
<a name="l00979"></a>00979 <span class="comment"> * @param eh - relative probability of h = ab allele about equal</span>
<a name="l00980"></a>00980 <span class="comment"> * @param ed - relative probability of d = b allele dominant</span>
<a name="l00981"></a>00981 <span class="comment"> * @param ZS - marginal size of matrix</span>
<a name="l00982"></a>00982 <span class="comment"> * @param q  - matrix of relative probability of labelings</span>
<a name="l00983"></a>00983 <span class="comment"> */</span>
<a name="l00984"></a><a class="code" href="snp_8label_8cpp.html#a4e0eeb85252ceaca83c545aef7ba4c4e">00984</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a4e0eeb85252ceaca83c545aef7ba4c4e" title="turn the overall relative probability of labelings into relative probability of individual genotypes...">eqtoproblabel</a>(vector&lt;double&gt; &amp;ec, vector&lt;double&gt; &amp;eh, vector&lt;double&gt; &amp;ed, <span class="keywordtype">int</span> ZS, vector&lt;double&gt; q)
<a name="l00985"></a>00985 {
<a name="l00986"></a>00986     <span class="keywordtype">int</span> i;
<a name="l00987"></a>00987     <span class="keywordtype">double</span> total;
<a name="l00988"></a>00988 
<a name="l00989"></a>00989     total =0; 
<a name="l00990"></a>00990     <span class="keywordflow">for</span> (i=0; i&lt;q.size(); i++)
<a name="l00991"></a>00991         total += q[i];
<a name="l00992"></a>00992     
<a name="l00993"></a>00993     <span class="comment">// get cluster memberships</span>
<a name="l00994"></a>00994     <span class="comment">//  note 0 here is entry &quot;0&quot; in the original list of length l</span>
<a name="l00995"></a>00995     <span class="comment">// entry j is in ec[l] iff j&lt;=l</span>
<a name="l00996"></a>00996     <span class="comment">// so straight cumulative sum</span>
<a name="l00997"></a>00997     vector&lt;double&gt; eqt;
<a name="l00998"></a>00998     eqt.resize(ZS);
<a name="l00999"></a>00999     <a class="code" href="snp_8label_8cpp.html#a476e5d98fb03d9efaccb100f40cd672f" title="sum up the source array along one axis">apply_sum</a>(eqt,q,ZS,0);
<a name="l01000"></a>01000     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(ec,eqt,ZS);
<a name="l01001"></a>01001     <span class="comment">// entry i is in ed[l] iff i&gt;l</span>
<a name="l01002"></a>01002     <span class="comment">// do cumsum to get i&lt;=l, then subtract</span>
<a name="l01003"></a>01003     <a class="code" href="snp_8label_8cpp.html#a476e5d98fb03d9efaccb100f40cd672f" title="sum up the source array along one axis">apply_sum</a>(eqt,q,ZS,1); 
<a name="l01004"></a>01004     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(ed,eqt,ZS);
<a name="l01005"></a>01005     
<a name="l01006"></a>01006     <span class="keywordflow">for</span> (i=0; i&lt;ZS-1; i++)
<a name="l01007"></a>01007     {
<a name="l01008"></a>01008         ed[i] = total-ed[i];
<a name="l01009"></a>01009         <span class="comment">// entry i is in eh[l] iff not in a or b</span>
<a name="l01010"></a>01010         eh[i] = total-ed[i]-ec[i];
<a name="l01011"></a>01011         <span class="comment">// and therefore the sum is always total</span>
<a name="l01012"></a>01012         <span class="comment">//sm     = total;</span>
<a name="l01013"></a>01013         ed[i] /= total;
<a name="l01014"></a>01014         ec[i] /= total;
<a name="l01015"></a>01015         eh[i] /= total;
<a name="l01016"></a>01016     }
<a name="l01017"></a>01017 }
<a name="l01018"></a>01018 <span class="comment"></span>
<a name="l01019"></a>01019 <span class="comment">/**</span>
<a name="l01020"></a>01020 <span class="comment"> *  setup the matrixes for doing individual BRLMM steps</span>
<a name="l01021"></a>01021 <span class="comment"> *</span>
<a name="l01022"></a>01022 <span class="comment"> * @param m - prior mean vector</span>
<a name="l01023"></a>01023 <span class="comment"> * @param Minv - inverse matrix of covariances (pseudo-observations) for means</span>
<a name="l01024"></a>01024 <span class="comment"> * @param Sinv - inverse matrix of variances of observations within genotypes</span>
<a name="l01025"></a>01025 <span class="comment"> * @param snp_param - parameters of algorithm</span>
<a name="l01026"></a>01026 <span class="comment"> */</span>
<a name="l01027"></a><a class="code" href="snp_8label_8cpp.html#ac2829bc689e49ff601dd7a2cabb6eec2">01027</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#ac2829bc689e49ff601dd7a2cabb6eec2" title="setup the matrixes for doing individual BRLMM steps">setupBRLMM</a>(vector&lt;double&gt; &amp;m, vector&lt;double&gt; &amp;Minv, vector&lt;double&gt; &amp;Sinv, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l01028"></a>01028 {
<a name="l01029"></a>01029             m.assign(3,0);
<a name="l01030"></a>01030             m[0] = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01031"></a>01031             m[1] = sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01032"></a>01032             m[2] = sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01033"></a>01033             <span class="comment">// only work with Minv</span>
<a name="l01034"></a>01034             <span class="comment">//inversethree(Minv,M);</span>
<a name="l01035"></a>01035             Minv.assign(9,0);
<a name="l01036"></a>01036             <span class="comment">// set up allowed variation in cluster center</span>
<a name="l01037"></a>01037             Minv[0] = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>/sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;
<a name="l01038"></a>01038             Minv[4] = sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>/sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;
<a name="l01039"></a>01039             Minv[8] = sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>/sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;
<a name="l01040"></a>01040             <span class="comment">// handle cross-terms here</span>
<a name="l01041"></a>01041             Minv[1] = sp.prior.xah/sqrt(sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>);
<a name="l01042"></a>01042             Minv[2] = sp.prior.xab/sqrt(sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>);
<a name="l01043"></a>01043             Minv[3] = Minv[1];
<a name="l01044"></a>01044             Minv[5] = sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a>/sqrt(sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>);
<a name="l01045"></a>01045             Minv[6] = Minv[2];
<a name="l01046"></a>01046             Minv[7] = Minv[5];
<a name="l01047"></a>01047             <span class="comment">// only work with Sinv;</span>
<a name="l01048"></a>01048             <span class="comment">//inversethree(Sinv,S);</span>
<a name="l01049"></a>01049             Sinv.assign(9,0);
<a name="l01050"></a>01050             Sinv[0] = 1/sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;
<a name="l01051"></a>01051             Sinv[4] = 1/sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;
<a name="l01052"></a>01052             Sinv[8] = 1/sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;
<a name="l01053"></a>01053 }
<a name="l01054"></a>01054 
<a name="l01055"></a>01055 
<a name="l01056"></a>01056 <span class="keywordtype">void</span> force_isotonic(<span class="keywordtype">double</span> &amp;mb, <span class="keywordtype">double</span> &amp;mh, <span class="keywordtype">double</span> &amp;ma, <span class="keywordtype">double</span> wb, <span class="keywordtype">double</span> wh, <span class="keywordtype">double</span> wa,<span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#ad748ef26c031cf6086e3a490afd3a47f" title="difference in cumulative sums: i.e.">delta</a>)
<a name="l01057"></a>01057 {
<a name="l01058"></a>01058     <span class="keywordtype">double</span> gamma, tmp;
<a name="l01059"></a>01059     <span class="comment">// forces means to be isotonic, separated by delta</span>
<a name="l01060"></a>01060     <span class="comment">// moves low weight centers most</span>
<a name="l01061"></a>01061     gamma = delta*(wb-wa)/(wb+wh+wa);
<a name="l01062"></a>01062     <span class="comment">// shift means towards center of mass by appropriate values</span>
<a name="l01063"></a>01063     mb += delta-gamma;
<a name="l01064"></a>01064     mh += 0-gamma;
<a name="l01065"></a>01065     ma += -1*delta-gamma;
<a name="l01066"></a>01066     <span class="comment">// Pool Adjacent Violators</span>
<a name="l01067"></a>01067     <span class="comment">// possible that mb&lt;=mh&lt;=ma not true</span>
<a name="l01068"></a>01068     <span class="keywordflow">if</span> (mb&gt;mh)
<a name="l01069"></a>01069     {
<a name="l01070"></a>01070         tmp = (wb*mb+wh*mh)/(wb+wh);
<a name="l01071"></a>01071         mb = tmp;
<a name="l01072"></a>01072         mh = tmp;
<a name="l01073"></a>01073     }
<a name="l01074"></a>01074     <span class="keywordflow">if</span> (mh&gt;ma)
<a name="l01075"></a>01075     {
<a name="l01076"></a>01076         tmp = (wh*mh+wa*ma)/(wh+wa);
<a name="l01077"></a>01077         mh = tmp;
<a name="l01078"></a>01078         ma = tmp;
<a name="l01079"></a>01079         <span class="keywordflow">if</span> (mb&gt;mh)
<a name="l01080"></a>01080         {
<a name="l01081"></a>01081             tmp = (wb*mb+wh*mh+wa*ma)/(wb+wh+wa);
<a name="l01082"></a>01082             mb = tmp;
<a name="l01083"></a>01083             mh = tmp;
<a name="l01084"></a>01084             ma = tmp;
<a name="l01085"></a>01085         }
<a name="l01086"></a>01086     }
<a name="l01087"></a>01087     <span class="comment">// now mb&lt;=mh&lt;=ma is true and average data hasn&#39;t changed</span>
<a name="l01088"></a>01088     <span class="comment">// now we make sure that they are separated by delta</span>
<a name="l01089"></a>01089     mb -= delta-gamma;
<a name="l01090"></a>01090     mh -= 0-gamma;
<a name="l01091"></a>01091     ma -= -1*delta-gamma;
<a name="l01092"></a>01092     <span class="comment">// now mb+delta&lt;=mh, mh+delta&lt;ma</span>
<a name="l01093"></a>01093 }
<a name="l01094"></a>01094 
<a name="l01095"></a>01095 <span class="keywordtype">void</span> force_IsoHetY(<a class="code" href="classsnp__distribution.html" title="this class holds all three genotype clusters">snp_distribution</a> &amp;post, <span class="keywordtype">double</span> IsoHetWeight)
<a name="l01096"></a>01096 {
<a name="l01097"></a>01097     <span class="comment">// ensure that the Y means of clusters are in the right order</span>
<a name="l01098"></a>01098     <span class="comment">// Het mean &gt; Hom mean average</span>
<a name="l01099"></a>01099     <span class="keywordtype">double</span> ystarab, ystar;
<a name="l01100"></a>01100     <span class="keywordtype">double</span> wab,wh,wa,wb;
<a name="l01101"></a>01101     <span class="keywordtype">double</span> delta,deltab,deltaa;
<a name="l01102"></a>01102     <span class="keywordtype">double</span> dxah,dxhb,dxab;
<a name="l01103"></a>01103     
<a name="l01104"></a>01104 
<a name="l01105"></a>01105     <span class="comment">// interpolated Y value from hom clusters</span>
<a name="l01106"></a>01106     dxhb = post.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-post.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01107"></a>01107     dxah = post.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-post.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01108"></a>01108     dxab = post.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-post.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01109"></a>01109     
<a name="l01110"></a>01110     ystarab = (dxhb*post.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> + dxah*post.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>)/dxab;
<a name="l01111"></a>01111     <span class="keywordflow">if</span> (ystarab&gt;post.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>)
<a name="l01112"></a>01112     {
<a name="l01113"></a>01113         wh = post.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>*IsoHetWeight;
<a name="l01114"></a>01114         wa = post.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01115"></a>01115         wb = post.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01116"></a>01116         wab = wa+wb;
<a name="l01117"></a>01117         <span class="comment">// violation of constraint</span>
<a name="l01118"></a>01118         ystar = (ystarab*wab+post.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>*wh)/(wab+wh);
<a name="l01119"></a>01119         post.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = ystar; <span class="comment">// here&#39;s where I&#39;ve moved the het</span>
<a name="l01120"></a>01120 
<a name="l01121"></a>01121         <span class="comment">// now move homs to make their center work</span>
<a name="l01122"></a>01122         <span class="comment">// on average must move this much</span>
<a name="l01123"></a>01123         delta = (ystar-ystarab);
<a name="l01124"></a>01124         <span class="comment">// move each by weight of corresponding entity</span>
<a name="l01125"></a>01125         deltab = delta/((wb*dxhb+wa*dxah)/(wa*dxab));
<a name="l01126"></a>01126         deltaa = delta/((wb*dxhb+wa*dxah)/(wb*dxab));
<a name="l01127"></a>01127         post.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> += deltab;
<a name="l01128"></a>01128         post.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> += deltaa;
<a name="l01129"></a>01129     }
<a name="l01130"></a>01130 }
<a name="l01131"></a>01131 
<a name="l01132"></a>01132 <span class="keywordtype">void</span> bayes_two_variance(<a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;post, vector&lt;double&gt; &amp;sixc, <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;prior)
<a name="l01133"></a>01133 {
<a name="l01134"></a>01134     post.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a> = prior.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a> + sixc[0];
<a name="l01135"></a>01135 
<a name="l01136"></a>01136     post.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = prior.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>*prior.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>;  <span class="comment">// prior</span>
<a name="l01137"></a>01137         post.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> += sixc[2]-sixc[1]*sixc[1]/(sixc[0]+0.001); <span class="comment">// observed ssq, avoid dividebyzero</span>
<a name="l01138"></a>01138         post.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> += ((prior.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>*sixc[0])/(prior.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>+sixc[0])) * (post.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-prior.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>)*(post.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-prior.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);  
<a name="l01139"></a>01139         post.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> /= post.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01140"></a>01140 
<a name="l01141"></a>01141     post.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> = prior.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>*prior.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>; <span class="comment">// prior</span>
<a name="l01142"></a>01142     post.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> += sixc[4]-sixc[3]*sixc[3]/(sixc[0]+0.001);
<a name="l01143"></a>01143     post.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> += ((prior.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>*sixc[0])/(prior.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>+sixc[0])) * (post.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>-prior.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>)*(post.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>-prior.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>);  
<a name="l01144"></a>01144     post.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> /= post.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01145"></a>01145 
<a name="l01146"></a>01146     <span class="comment">//cout &lt;&lt; prior.v*prior.yss &lt;&lt; &quot;\t&quot; &lt;&lt; sixc[4]-sixc[3]*sixc[3]/(sixc[0]+0.001) &lt;&lt; &quot;\t&quot; &lt;&lt; ((prior.k*sixc[0])/(prior.k+sixc[0])) * (post.ym-prior.ym)*(post.ym-prior.ym) &lt;&lt; endl;</span>
<a name="l01147"></a>01147 
<a name="l01148"></a>01148     post.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a> = prior.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>*prior.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>; <span class="comment">// prior</span>
<a name="l01149"></a>01149     post.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a> += sixc[5]-sixc[1]*sixc[3]/(sixc[0]+0.001);
<a name="l01150"></a>01150     post.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a> += ((prior.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>*sixc[0])/(prior.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>+sixc[0])) * (post.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>-prior.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>)*(post.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>-prior.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01151"></a>01151     post.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a> /= post.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01152"></a>01152 }
<a name="l01153"></a>01153 
<a name="l01154"></a>01154 
<a name="l01155"></a>01155 
<a name="l01156"></a>01156 <span class="keywordtype">void</span> shrink_cluster_variance(<a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;out, <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;a, <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;b, <a class="code" href="classcluster__data.html" title="this class summarizes a single genotype">cluster_data</a> &amp;c, <span class="keywordtype">double</span> lambda)
<a name="l01157"></a>01157 {
<a name="l01158"></a>01158     <span class="keywordtype">double</span> main,cross;
<a name="l01159"></a>01159     
<a name="l01160"></a>01160     main = 3-2*lambda;
<a name="l01161"></a>01161     cross = lambda;
<a name="l01162"></a>01162 
<a name="l01163"></a>01163     out.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a> = (main*a.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*a.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*b.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*b.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*c.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*c.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>)/(main*a.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*b.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*c.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>);
<a name="l01164"></a>01164     out.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> = (main*a.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>*a.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*b.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>*b.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*c.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>*c.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>)/(main*a.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*b.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>+cross*c.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>);
<a name="l01165"></a>01165     out.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a> = sqrt(out.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*out.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>)*a.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>/(sqrt(a.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>*a.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>));  <span class="comment">// covariance shrinkage, same correlation</span>
<a name="l01166"></a>01166     <span class="comment">// what should degree of confidence in variance be???</span>
<a name="l01167"></a>01167 }
<a name="l01168"></a>01168 
<a name="l01169"></a>01169 <span class="keywordtype">void</span> bayes_two_shrink_variance(<a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l01170"></a>01170 {
<a name="l01171"></a>01171     <span class="comment">// keep the correlation the same within each cluster</span>
<a name="l01172"></a>01172     <span class="comment">// move the variances towards each other?</span>
<a name="l01173"></a>01173     <span class="comment">// analogue of 1-d common variance</span>
<a name="l01174"></a>01174     
<a name="l01175"></a>01175     <a class="code" href="classsnp__distribution.html" title="this class holds all three genotype clusters">snp_distribution</a> tmp;
<a name="l01176"></a>01176     tmp.Copy(sp.posterior);
<a name="l01177"></a>01177 
<a name="l01178"></a>01178     shrink_cluster_variance(tmp.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>,sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>);
<a name="l01179"></a>01179     shrink_cluster_variance(tmp.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>,sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>);
<a name="l01180"></a>01180     shrink_cluster_variance(tmp.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>,sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>);
<a name="l01181"></a>01181     
<a name="l01182"></a>01182     sp.posterior.Copy(tmp); <span class="comment">// copy shrunk values</span>
<a name="l01183"></a>01183 }
<a name="l01184"></a>01184 
<a name="l01185"></a>01185 <span class="keywordtype">void</span> Dump_Matrix(Matrix &amp;M)
<a name="l01186"></a>01186 {
<a name="l01187"></a>01187     cout &lt;&lt; endl;
<a name="l01188"></a>01188     
<a name="l01189"></a>01189     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;M.Nrows(); i++)
<a name="l01190"></a>01190     {
<a name="l01191"></a>01191         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j=0; j&lt;M.Ncols(); j++)
<a name="l01192"></a>01192         {
<a name="l01193"></a>01193             cout &lt;&lt; M.element(i,j) &lt;&lt; <span class="stringliteral">&quot;\t&quot;</span>;
<a name="l01194"></a>01194         }
<a name="l01195"></a>01195         cout &lt;&lt; endl;
<a name="l01196"></a>01196     }
<a name="l01197"></a>01197 }
<a name="l01198"></a>01198 
<a name="l01199"></a>01199 <span class="keywordtype">void</span> Dump_Column(ColumnVector &amp;V)
<a name="l01200"></a>01200 {
<a name="l01201"></a>01201     cout &lt;&lt; endl;
<a name="l01202"></a>01202     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i=0; i&lt;V.Nrows(); i++)
<a name="l01203"></a>01203             cout &lt;&lt; V.element(i) &lt;&lt; <span class="stringliteral">&quot;\t&quot;</span>;
<a name="l01204"></a>01204     cout &lt;&lt; endl;
<a name="l01205"></a>01205 }
<a name="l01206"></a>01206 
<a name="l01207"></a>01207 <span class="keywordtype">void</span> labels_two_posterior(vector&lt;double&gt; &amp;ec, vector&lt;double&gt; &amp;eh, vector&lt;double&gt; &amp;ed, <span class="keywordtype">int</span> ZS, <a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l01208"></a>01208 {
<a name="l01209"></a>01209     <span class="comment">// Means = (K+N)^-1 (K*u+N*X)</span>
<a name="l01210"></a>01210     <span class="comment">// Fix isotonic X dimension means</span>
<a name="l01211"></a>01211     <span class="comment">// K&lt;- K+N</span>
<a name="l01212"></a>01212     <span class="comment">// V1&lt;-V + SSQ + [kn/(k+n)] * (u-Means)^2</span>
<a name="l01213"></a>01213     <span class="comment">// Vshrink = common var    </span>
<a name="l01214"></a>01214     
<a name="l01215"></a>01215     vector&lt;double&gt; sixd,sixc,sixh;
<a name="l01216"></a>01216     sixd.resize(6);
<a name="l01217"></a>01217     sixc.resize(6);
<a name="l01218"></a>01218     sixh.resize(6);
<a name="l01219"></a>01219     <span class="comment">// compute basic stats, raw mean/sum-square/number of entries</span>
<a name="l01220"></a>01220     <span class="comment">// for each genotype assignment mixed fraction</span>
<a name="l01221"></a>01221     <a class="code" href="snp_8label_8cpp.html#a35346d828a53f326d0e58596c6067c07" title="generate sums of data based on weighted cluster membership">compute_two_binned_cluster</a>(sixd,zall,ed,ZS);
<a name="l01222"></a>01222     <a class="code" href="snp_8label_8cpp.html#a35346d828a53f326d0e58596c6067c07" title="generate sums of data based on weighted cluster membership">compute_two_binned_cluster</a>(sixc,zall,ec,ZS);
<a name="l01223"></a>01223     <a class="code" href="snp_8label_8cpp.html#a35346d828a53f326d0e58596c6067c07" title="generate sums of data based on weighted cluster membership">compute_two_binned_cluster</a>(sixh,zall,eh,ZS);
<a name="l01224"></a>01224 
<a name="l01225"></a>01225     <span class="comment">// now we set up the N matrix</span>
<a name="l01226"></a>01226     <span class="comment">// why is this banded in BRLMM?</span>
<a name="l01227"></a>01227     Matrix N(6,6);
<a name="l01228"></a>01228       N = 0;
<a name="l01229"></a>01229       N.element(0,0) = sixc[0];
<a name="l01230"></a>01230       N.element(1,1) = sixc[0];
<a name="l01231"></a>01231      N.element(2,2) = sixh[0];
<a name="l01232"></a>01232       N.element(3,3) = sixh[0];
<a name="l01233"></a>01233       N.element(4,4) = sixd[0];
<a name="l01234"></a>01234       N.element(5,5) = sixd[0];
<a name="l01235"></a>01235 
<a name="l01236"></a>01236     Matrix K(6,6);
<a name="l01237"></a>01237     K=0;
<a name="l01238"></a>01238     K.element(0,0) = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01239"></a>01239     K.element(1,1) = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01240"></a>01240     K.element(2,2) = sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01241"></a>01241     K.element(3,3) = sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01242"></a>01242     K.element(4,4) = sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01243"></a>01243     K.element(5,5) = sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01244"></a>01244     <span class="comment">// fill in off-diagonal elements for x to x</span>
<a name="l01245"></a>01245     K.element(0,2) = sp.prior.xah;
<a name="l01246"></a>01246     K.element(2,0) = sp.prior.xah;
<a name="l01247"></a>01247     K.element(2,4) = sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a>;
<a name="l01248"></a>01248     K.element(4,2) = sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a>;
<a name="l01249"></a>01249     K.element(0,4) = sp.prior.xab;
<a name="l01250"></a>01250     K.element(4,0) = sp.prior.xab;
<a name="l01251"></a>01251     <span class="comment">// fill in off diagonal elements for y to y</span>
<a name="l01252"></a>01252     K.element(1,3) = sp.prior.yah;
<a name="l01253"></a>01253     K.element(3,1) = sp.prior.yah;
<a name="l01254"></a>01254     K.element(3,5) = sp.prior.<a class="code" href="classsnp__distribution.html#a2bd112107fcbccf699b612f410aaa73c" title="cross-correlation terms between clusters y">yhb</a>;
<a name="l01255"></a>01255     K.element(5,3) = sp.prior.<a class="code" href="classsnp__distribution.html#a2bd112107fcbccf699b612f410aaa73c" title="cross-correlation terms between clusters y">yhb</a>;
<a name="l01256"></a>01256     K.element(1,5) = sp.prior.yab;
<a name="l01257"></a>01257     K.element(5,1) = sp.prior.yab;
<a name="l01258"></a>01258 
<a name="l01259"></a>01259     <span class="comment">// temporary!!!</span>
<a name="l01260"></a>01260     <span class="comment">// Need 4x3 covariance pairs</span>
<a name="l01261"></a>01261 
<a name="l01262"></a>01262     <span class="comment">// left hand side</span>
<a name="l01263"></a>01263     Matrix Left = K + N;
<a name="l01264"></a>01264     <span class="comment">//Dump_Matrix(Left);</span>
<a name="l01265"></a>01265     <span class="comment">//Dump_Matrix(K);</span>
<a name="l01266"></a>01266     <span class="comment">//Dump_Matrix(N);</span>
<a name="l01267"></a>01267     Left = Left.i();
<a name="l01268"></a>01268     <span class="comment">//Dump_Matrix(Left);    </span>
<a name="l01269"></a>01269     
<a name="l01270"></a>01270     
<a name="l01271"></a>01271     ColumnVector Mu;
<a name="l01272"></a>01272     Mu.ReSize(6);
<a name="l01273"></a>01273     Mu.element(0) = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01274"></a>01274     Mu.element(1) = sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l01275"></a>01275     Mu.element(2) = sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01276"></a>01276     Mu.element(3) = sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l01277"></a>01277     Mu.element(4) = sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>;
<a name="l01278"></a>01278     Mu.element(5) = sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l01279"></a>01279     <span class="comment">//Dump_Column(Mu);</span>
<a name="l01280"></a>01280 
<a name="l01281"></a>01281 
<a name="l01282"></a>01282     Matrix KMu = K*Mu;
<a name="l01283"></a>01283     <span class="comment">//Dump_Matrix(KMu);</span>
<a name="l01284"></a>01284     
<a name="l01285"></a>01285     <span class="comment">// premultipy, prevent div/0 in computing mean</span>
<a name="l01286"></a>01286     
<a name="l01287"></a>01287     ColumnVector Nv;
<a name="l01288"></a>01288     Nv.ReSize(6);
<a name="l01289"></a>01289     Nv.element(0) = sixc[1];
<a name="l01290"></a>01290     Nv.element(1) = sixc[3];
<a name="l01291"></a>01291     Nv.element(2) = sixh[1];
<a name="l01292"></a>01292     Nv.element(3) = sixh[3];
<a name="l01293"></a>01293     Nv.element(4) = sixd[1];
<a name="l01294"></a>01294     Nv.element(5) = sixd[3];
<a name="l01295"></a>01295 
<a name="l01296"></a>01296     <span class="comment">//Dump_Column(Nv);</span>
<a name="l01297"></a>01297     
<a name="l01298"></a>01298     Matrix Right = KMu + Nv;
<a name="l01299"></a>01299 
<a name="l01300"></a>01300     <span class="comment">//Dump_Matrix(Right);</span>
<a name="l01301"></a>01301 
<a name="l01302"></a>01302     Matrix Final = Left*Right;
<a name="l01303"></a>01303 
<a name="l01304"></a>01304     <span class="comment">//Dump_Matrix(Final);</span>
<a name="l01305"></a>01305     <span class="comment">// update means</span>
<a name="l01306"></a>01306     sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>  = Final.element(0,0);
<a name="l01307"></a>01307     sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = Final.element(1,0);
<a name="l01308"></a>01308     sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>  = Final.element(2,0);
<a name="l01309"></a>01309     sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = Final.element(3,0);
<a name="l01310"></a>01310     sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>  = Final.element(4,0);
<a name="l01311"></a>01311     sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = Final.element(5,0); 
<a name="l01312"></a>01312 
<a name="l01313"></a>01313     <span class="comment">//cout &lt;&lt; &quot;Executed through&quot; &lt;&lt; endl;</span>
<a name="l01314"></a>01314 
<a name="l01315"></a>01315     <span class="comment">// update posterior strength</span>
<a name="l01316"></a>01316     sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = sixc[0] + sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01317"></a>01317     sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = sixh[0] + sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01318"></a>01318     sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = sixd[0] + sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>;
<a name="l01319"></a>01319 
<a name="l01320"></a>01320     <span class="comment">// update posterior frequency</span>
<a name="l01321"></a>01321     <span class="comment">// Omit currently</span>
<a name="l01322"></a>01322     
<a name="l01323"></a>01323     <span class="comment">// fix up positions of posterior in contrast dimension</span>
<a name="l01324"></a>01324     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a>==3)
<a name="l01325"></a>01325         force_isotonic(sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>, sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a>);
<a name="l01326"></a>01326 
<a name="l01327"></a>01327     <span class="comment">// put het above hom average</span>
<a name="l01328"></a>01328     <span class="comment">// this is to avoid unrealistic situations in the face of copy number</span>
<a name="l01329"></a>01329     <span class="comment">// the numerical control is the relative weight of Hets: near zero means move het, near infinity means move hom</span>
<a name="l01330"></a>01330     <span class="comment">// 1.0 means weight hets and homs equally for the final line.</span>
<a name="l01331"></a>01331     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a01576510bc234b4084eb61d2da4d9f00" title="ensure Het &amp;gt; average Hom Y by at least epsilon">IsoHetY</a>&gt;0)
<a name="l01332"></a>01332         force_IsoHetY(sp.posterior,sp.<a class="code" href="classsnp__param.html#a01576510bc234b4084eb61d2da4d9f00" title="ensure Het &amp;gt; average Hom Y by at least epsilon">IsoHetY</a>);
<a name="l01333"></a>01333 
<a name="l01334"></a>01334     <span class="comment">// now I have the means computed!!!</span>
<a name="l01335"></a>01335     <span class="comment">// on to compute variances</span>
<a name="l01336"></a>01336     <span class="comment">// Technically, these are variances computed from the non-isotonic means</span>
<a name="l01337"></a>01337     bayes_two_variance(sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>,sixc,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>);
<a name="l01338"></a>01338     bayes_two_variance(sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>,sixh,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>);
<a name="l01339"></a>01339     bayes_two_variance(sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>,sixd,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>);
<a name="l01340"></a>01340     <span class="comment">// how do I shrink variances in 2-D correctly?</span>
<a name="l01341"></a>01341     bayes_two_shrink_variance(sp);
<a name="l01342"></a>01342     
<a name="l01343"></a>01343     <span class="comment">// cluster covariances in pseudo-obs</span>
<a name="l01344"></a>01344     sp.posterior.xah = sp.prior.xah;
<a name="l01345"></a>01345     sp.posterior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a> = sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a>;
<a name="l01346"></a>01346     sp.posterior.xab = sp.prior.xab;
<a name="l01347"></a>01347     sp.posterior.yah = sp.prior.yah;
<a name="l01348"></a>01348     sp.posterior.<a class="code" href="classsnp__distribution.html#a2bd112107fcbccf699b612f410aaa73c" title="cross-correlation terms between clusters y">yhb</a> = sp.prior.<a class="code" href="classsnp__distribution.html#a2bd112107fcbccf699b612f410aaa73c" title="cross-correlation terms between clusters y">yhb</a>;
<a name="l01349"></a>01349     sp.posterior.yab = sp.prior.yab;
<a name="l01350"></a>01350 }
<a name="l01351"></a>01351 
<a name="l01352"></a>01352 <span class="comment"></span>
<a name="l01353"></a>01353 <span class="comment">/**</span>
<a name="l01354"></a>01354 <span class="comment"> * convert binned label relative probabilities into cluster parameters</span>
<a name="l01355"></a>01355 <span class="comment"> *</span>
<a name="l01356"></a>01356 <span class="comment"> * @param ec - relative probability of c = a allele dominates</span>
<a name="l01357"></a>01357 <span class="comment"> * @param eh - relative probability of h = ab about equal</span>
<a name="l01358"></a>01358 <span class="comment"> * @param ed - relative probability of d = b allele dominates</span>
<a name="l01359"></a>01359 <span class="comment"> * @param ZS - length of bins</span>
<a name="l01360"></a>01360 <span class="comment"> * @param zall - all binned data, sums, sums of squares</span>
<a name="l01361"></a>01361 <span class="comment"> * @param sp - snp parameters</span>
<a name="l01362"></a>01362 <span class="comment"> */</span>
<a name="l01363"></a><a class="code" href="snp_8label_8cpp.html#aa6c1f401c915be4f6db927f0b0b627c8">01363</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#aa6c1f401c915be4f6db927f0b0b627c8" title="convert binned label relative probabilities into cluster parameters">labelstoposterior</a>(vector&lt;double&gt; &amp;ec, vector&lt;double&gt; &amp;eh, vector&lt;double&gt; &amp;ed, <span class="keywordtype">int</span> ZS, <a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall, <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l01364"></a>01364 {
<a name="l01365"></a>01365     <span class="comment">// compute posterior mean/variance based on model average labels</span>
<a name="l01366"></a>01366     <span class="keywordtype">double</span> rnd,rnc,rnh,rmd,rmc,rmh,rsd,rsc,rsh;
<a name="l01367"></a>01367     <span class="keywordtype">double</span> rbma,rbmb,rbmh,rbva,rbvb,rbvh;
<a name="l01368"></a>01368     <span class="comment">//double tmp;</span>
<a name="l01369"></a>01369     
<a name="l01370"></a>01370     <span class="comment">// compute basic stats, raw mean/sum-square/number of entries</span>
<a name="l01371"></a>01371     <span class="comment">// for each genotype assignment mixed fraction</span>
<a name="l01372"></a>01372     <a class="code" href="snp_8label_8cpp.html#a2cf95e5f494d77911161becc1518ef95" title="generate sums of data based on weighted cluster membership">compute_binned_cluster</a>(rnd,rmd,rsd,zall,ed,ZS);
<a name="l01373"></a>01373     <a class="code" href="snp_8label_8cpp.html#a2cf95e5f494d77911161becc1518ef95" title="generate sums of data based on weighted cluster membership">compute_binned_cluster</a>(rnc,rmc,rsc,zall,ec,ZS);
<a name="l01374"></a>01374     <a class="code" href="snp_8label_8cpp.html#a2cf95e5f494d77911161becc1518ef95" title="generate sums of data based on weighted cluster membership">compute_binned_cluster</a>(rnh,rmh,rsh,zall,eh,ZS);
<a name="l01375"></a>01375 
<a name="l01376"></a>01376     <span class="comment">// Compute the posterior mean/variance</span>
<a name="l01377"></a>01377     <span class="comment">// i.e. do BRLMM step here</span>
<a name="l01378"></a>01378     <span class="comment">//rbmb = b_mean(sp.bb.k,sp.bb.m,rmd,rnd);</span>
<a name="l01379"></a>01379     <span class="comment">//rbma = b_mean(sp.aa.k,sp.aa.m,rmc,rnc);</span>
<a name="l01380"></a>01380     <span class="comment">//rbmh = b_mean(sp.ab.k,sp.ab.m,rmh,rnh);</span>
<a name="l01381"></a>01381     
<a name="l01382"></a>01382 <span class="comment">// BRLMM</span>
<a name="l01383"></a>01383 <span class="comment">// setup BRLMM</span>
<a name="l01384"></a>01384             vector&lt;double&gt; m;
<a name="l01385"></a>01385             vector&lt;double&gt; Minv;
<a name="l01386"></a>01386             vector&lt;double&gt; Sinv;
<a name="l01387"></a>01387             <a class="code" href="snp_8label_8cpp.html#ac2829bc689e49ff601dd7a2cabb6eec2" title="setup the matrixes for doing individual BRLMM steps">setupBRLMM</a>(m,Minv,Sinv,sp);
<a name="l01388"></a>01388             <span class="comment">// only work with Sinv;</span>
<a name="l01389"></a>01389             <span class="comment">// scratch vectors</span>
<a name="l01390"></a>01390             vector&lt;double&gt; Tz;
<a name="l01391"></a>01391             vector&lt;double&gt; Zt;
<a name="l01392"></a>01392             vector&lt;double&gt; Xinv;
<a name="l01393"></a>01393             vector&lt;double&gt; Xt;
<a name="l01394"></a>01394             vector&lt;double&gt; Mz;
<a name="l01395"></a>01395             <span class="comment">// vectors that change</span>
<a name="l01396"></a>01396             <span class="comment">// sum of x</span>
<a name="l01397"></a>01397             vector&lt;double&gt; Nv;
<a name="l01398"></a>01398             <span class="comment">// number of data points</span>
<a name="l01399"></a>01399             vector&lt;double&gt; N;
<a name="l01400"></a>01400     <span class="comment">// brlmm here</span>
<a name="l01401"></a>01401             Nv.resize(3);
<a name="l01402"></a>01402             Nv[0] = rmc;
<a name="l01403"></a>01403             Nv[1] = rmh;
<a name="l01404"></a>01404             Nv[2] = rmd;
<a name="l01405"></a>01405             N.assign(9,0);
<a name="l01406"></a>01406             N[0] = rnc;
<a name="l01407"></a>01407             N[4] = rnh;
<a name="l01408"></a>01408             N[8] = rnd;
<a name="l01409"></a>01409             <a class="code" href="snp_8label_8cpp.html#afc25b20145576567e02ac9bfa60efaa0" title="multiply two 3x3 matrices">timesthree</a>(Tz,N,Sinv);
<a name="l01410"></a>01410             <a class="code" href="snp_8label_8cpp.html#a1f2ff039971eee5c78e94867f1cc9369" title="sum two vectors">sumthree</a>(Zt, Minv,Tz);
<a name="l01411"></a>01411             <a class="code" href="snp_8label_8cpp.html#a57e554714be3edbd1b2393e5824b03b4" title="inverts 3x3 matrix specialty code, avoid calling external routines">inversethree</a>(Xinv,Zt);
<a name="l01412"></a>01412             <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(Tz,Sinv, Nv);
<a name="l01413"></a>01413             <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(Zt,Minv, m);
<a name="l01414"></a>01414             <a class="code" href="snp_8label_8cpp.html#a1f2ff039971eee5c78e94867f1cc9369" title="sum two vectors">sumthree</a>(Xt,Tz,Zt);
<a name="l01415"></a>01415             <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(Mz,Xinv,Xt);
<a name="l01416"></a>01416             rbma = Mz[0];
<a name="l01417"></a>01417             rbmh = Mz[1];
<a name="l01418"></a>01418             rbmb = Mz[2];
<a name="l01419"></a>01419     <span class="comment">// fix up positions of posterior</span>
<a name="l01420"></a>01420     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a>==3)
<a name="l01421"></a>01421         force_isotonic(rbmb,rbmh,rbma,rnd+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rnh+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rnd+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>, sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a>);
<a name="l01422"></a>01422 
<a name="l01423"></a>01423     <span class="comment">// do within-cluster variance as before</span>
<a name="l01424"></a>01424     rbvb = <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,rsd,rmd,rnd,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rbmb,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01425"></a>01425     rbva = <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,rsc,rmc,rnc,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rbma,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01426"></a>01426     rbvh = <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,rsh,rmh,rnh,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rbmh,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01427"></a>01427     
<a name="l01428"></a>01428     <span class="comment">// fix up variances if needed</span>
<a name="l01429"></a>01429     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a5996056f7436f5974c6e6b6bd7925a27" title="do we use common variances for all clusters?">comvar</a>==1 )
<a name="l01430"></a>01430     {
<a name="l01431"></a>01431             <span class="keywordtype">double</span> tc,td,th;
<a name="l01432"></a>01432             <span class="keywordtype">double</span> tnc,tnd,tnh;
<a name="l01433"></a>01433             <span class="keywordtype">double</span> tl,tlm;
<a name="l01434"></a>01434             tnd = rnd+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01435"></a>01435             tnc = rnc+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01436"></a>01436             tnh = rnh+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01437"></a>01437             td = rbvb*tnd;
<a name="l01438"></a>01438             tc = rbva*tnc;
<a name="l01439"></a>01439             th = rbvh*tnh;
<a name="l01440"></a>01440             <span class="comment">// cross-talk between variances 0 = independent, 1 = common variance</span>
<a name="l01441"></a>01441             tl = sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>;
<a name="l01442"></a>01442             tlm = 3-2*sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>;
<a name="l01443"></a>01443             rbvb = (tlm*td+tl*tc+tl*th)/(tlm*tnd+tl*tnc+tl*tnh);
<a name="l01444"></a>01444             rbva = (tl*td+tlm*tc+tl*th)/(tl*tnd+tlm*tnc+tl*tnh);
<a name="l01445"></a>01445             rbvh = (tl*td+tl*tc+tlm*th)/(tl*tnd+tl*tnc+tlm*tnh);
<a name="l01446"></a>01446     }
<a name="l01447"></a>01447 
<a name="l01448"></a>01448     <span class="comment">// output posterior + estimated fraction of data used for each cluster</span>
<a name="l01449"></a>01449     sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a79a9adc273730da3464a0ad808d419cb" title="set the values">Set</a>(rbma,rbva,rnc+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rnc+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>);
<a name="l01450"></a>01450     sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a79a9adc273730da3464a0ad808d419cb" title="set the values">Set</a>(rbmh,rbvh,rnh+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rnh+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>);
<a name="l01451"></a>01451     sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a79a9adc273730da3464a0ad808d419cb" title="set the values">Set</a>(rbmb,rbvb,rnd+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,rnd+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab37cdb91afcdf8b3c0924a150b31b4ed" title="covariance of cluster in both directions">xyss</a>);
<a name="l01452"></a>01452     <span class="comment">// set up posterior covariances to be prior covariances</span>
<a name="l01453"></a>01453     <span class="comment">// since working with inverse, not actually changed by increased observations</span>
<a name="l01454"></a>01454     sp.posterior.xah = sp.prior.xah;
<a name="l01455"></a>01455     sp.posterior.xab = sp.prior.xab;
<a name="l01456"></a>01456     sp.posterior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a> = sp.prior.<a class="code" href="classsnp__distribution.html#a7daecf444a60233a0de51e4c32fe7091" title="cross-correlation terms between clusters x">xhb</a>;
<a name="l01457"></a>01457 }
<a name="l01458"></a>01458 
<a name="l01459"></a>01459 <span class="comment"></span>
<a name="l01460"></a>01460 <span class="comment">/**</span>
<a name="l01461"></a>01461 <span class="comment"> * This penalizes unusual frequency distributions</span>
<a name="l01462"></a>01462 <span class="comment"> * </span>
<a name="l01463"></a>01463 <span class="comment"> * @param a - first cluster freq</span>
<a name="l01464"></a>01464 <span class="comment"> * @param b - second cluster freq</span>
<a name="l01465"></a>01465 <span class="comment"> * @param c - third cluster freq</span>
<a name="l01466"></a>01466 <span class="comment"> * @param oa - number of observations</span>
<a name="l01467"></a>01467 <span class="comment"> * @param ob - number of observations</span>
<a name="l01468"></a>01468 <span class="comment"> * @param oc - number of observations</span>
<a name="l01469"></a>01469 <span class="comment"> * @param lambda - parameterized safety value to avoid zero frequency</span>
<a name="l01470"></a>01470 <span class="comment"> */</span>
<a name="l01471"></a><a class="code" href="snp_8label_8cpp.html#a13b6fb2fd4777966e2fc309ddac76b38">01471</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#a13b6fb2fd4777966e2fc309ddac76b38" title="This penalizes unusual frequency distributions.">mixture_penalty</a>(<span class="keywordtype">double</span> a,<span class="keywordtype">double</span> b, <span class="keywordtype">double</span> c,<span class="keywordtype">double</span> oa, <span class="keywordtype">double</span> ob, <span class="keywordtype">double</span> oc,<span class="keywordtype">double</span> lambda)
<a name="l01472"></a>01472 {
<a name="l01473"></a>01473     <span class="keywordtype">double</span> total, entropy;
<a name="l01474"></a>01474 
<a name="l01475"></a>01475     <span class="comment">// penalty for mixing fraction</span>
<a name="l01476"></a>01476     <span class="comment">// want to avoid splitting clusters</span>
<a name="l01477"></a>01477     <span class="comment">// so 20/0/0 is good, 7/7/7 less good</span>
<a name="l01478"></a>01478     <span class="comment">// i.e. need more evidence to provoke a new cluster</span>
<a name="l01479"></a>01479     total = a+b+c+3*lambda;
<a name="l01480"></a>01480     entropy = oa * log((a+lambda)/total)+ob*log((b+lambda)/total)+oc*log((c+lambda)/total);
<a name="l01481"></a>01481     entropy = -1*entropy;
<a name="l01482"></a>01482     <span class="keywordflow">return</span>(entropy);
<a name="l01483"></a>01483 }
<a name="l01484"></a>01484 <span class="comment"></span>
<a name="l01485"></a>01485 <span class="comment">/**</span>
<a name="l01486"></a>01486 <span class="comment"> * This penalizes unusual cluster distributions (violating HW extremely)</span>
<a name="l01487"></a>01487 <span class="comment"> * Expected to be useful for screening/finding initial priors</span>
<a name="l01488"></a>01488 <span class="comment"> *</span>
<a name="l01489"></a>01489 <span class="comment"> * @param a - hom cluster freq</span>
<a name="l01490"></a>01490 <span class="comment"> * @param b - hom cluster freq</span>
<a name="l01491"></a>01491 <span class="comment"> * @param c - het cluster freq</span>
<a name="l01492"></a>01492 <span class="comment"> */</span>
<a name="l01493"></a><a class="code" href="snp_8label_8cpp.html#a188ea3959b9c819926f6a1ea3207f4d6">01493</a> <span class="keywordtype">double</span> <a class="code" href="snp_8label_8cpp.html#a188ea3959b9c819926f6a1ea3207f4d6" title="This penalizes unusual cluster distributions (violating HW extremely) Expected to be useful for scree...">HW_penalty</a>(<span class="keywordtype">double</span> a, <span class="keywordtype">double</span> b, <span class="keywordtype">double</span> c, <span class="keywordtype">double</span> lambda)
<a name="l01494"></a>01494 {
<a name="l01495"></a>01495     <span class="keywordtype">double</span> total, penalty, p, q;
<a name="l01496"></a>01496 
<a name="l01497"></a>01497     total = 2*a+2*b+2*c+2*lambda; <span class="comment">// total number of chromosomes +1 for always potential errors</span>
<a name="l01498"></a>01498     p = (2*a+c+lambda)/total; <span class="comment">// hom chromosomes + het chromosomes belonging to the allele</span>
<a name="l01499"></a>01499     q = 1-p;              <span class="comment">// other hom</span>
<a name="l01500"></a>01500     <span class="comment">// a* log(p*p)+c*log(2*p*q) + b*log(q*q) = (2a+c)*log(p)+(2b+c)*log(q)+c*log(2)</span>
<a name="l01501"></a>01501     penalty = (2*a+c)*log(p) + (2*b+c)*log(q)+c*log((<span class="keywordtype">double</span>)2);
<a name="l01502"></a>01502     penalty = -1*penalty;
<a name="l01503"></a>01503     <span class="keywordflow">return</span>(penalty);
<a name="l01504"></a>01504 }
<a name="l01505"></a>01505 
<a name="l01506"></a>01506 <span class="comment"></span>
<a name="l01507"></a>01507 <span class="comment">/**</span>
<a name="l01508"></a>01508 <span class="comment"> * This is the &quot;labeling&quot; routine that tries out all plausible genotypes for best likelihood</span>
<a name="l01509"></a>01509 <span class="comment"> * Basic goal: efficiently evaluate likelihood of three clusters based on seed</span>
<a name="l01510"></a>01510 <span class="comment"> * Loops over BB,...BB (i times), AB,...AB (j-i), AA,...AA (n-j+1)</span>
<a name="l01511"></a>01511 <span class="comment"> *</span>
<a name="l01512"></a>01512 <span class="comment"> * @param ec - relative probability of c = a allele dominates</span>
<a name="l01513"></a>01513 <span class="comment"> * @param eh - relative probability of h = ab allele similar</span>
<a name="l01514"></a>01514 <span class="comment"> * @param ec - relative probability of d = b allele dominates</span>
<a name="l01515"></a>01515 <span class="comment"> * @param ZS - number of bins + 1 = size of data</span>
<a name="l01516"></a>01516 <span class="comment"> * @param zall - data in bins, sum, sum of squares</span>
<a name="l01517"></a>01517 <span class="comment"> * @param sp - snp parameters</span>
<a name="l01518"></a>01518 <span class="comment"> */</span>
<a name="l01519"></a><a class="code" href="snp_8label_8cpp.html#af8085a9844832cf7a94ad62a95704b67">01519</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#af8085a9844832cf7a94ad62a95704b67" title="This is the &amp;quot;labeling&amp;quot; routine that tries out all plausible genotypes for best likelihood B...">integratebrlmmoverlabelings</a>(vector&lt;double&gt; &amp;ec, vector&lt;double&gt; &amp;eh, vector&lt;double&gt; &amp;ed, <span class="keywordtype">int</span> ZS, <a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall,<a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l01520"></a>01520 {
<a name="l01521"></a>01521     <span class="keywordtype">int</span> ZSS;
<a name="l01522"></a>01522     <span class="keywordtype">int</span> NS;
<a name="l01523"></a>01523     <span class="keywordtype">int</span> i,j;
<a name="l01524"></a>01524     <span class="comment">//double tmp;</span>
<a name="l01525"></a>01525     <span class="comment">// allocate memory for these puppies</span>
<a name="l01526"></a>01526     <span class="comment">// over allocate for most of these, but temporary</span>
<a name="l01527"></a>01527     ZSS = ZS*ZS;
<a name="l01528"></a>01528 
<a name="l01529"></a>01529     <span class="comment">// likelihood of data, given </span>
<a name="l01530"></a>01530     vector&lt;double&gt; q;
<a name="l01531"></a>01531     q.resize(ZSS);
<a name="l01532"></a>01532     q.assign(ZSS,0);
<a name="l01533"></a>01533     vector&lt;double&gt; ddx;
<a name="l01534"></a>01534     vector&lt;double&gt; ccx;
<a name="l01535"></a>01535     vector&lt;double&gt; nd;
<a name="l01536"></a>01536     vector&lt;double&gt; nc;
<a name="l01537"></a>01537     vector&lt;double&gt; dd;
<a name="l01538"></a>01538     vector&lt;double&gt; cc;
<a name="l01539"></a>01539     ddx.resize(ZS);
<a name="l01540"></a>01540     ccx.resize(ZS);
<a name="l01541"></a>01541     nd.resize(ZS);
<a name="l01542"></a>01542     nc.resize(ZS);
<a name="l01543"></a>01543     dd.resize(ZS);
<a name="l01544"></a>01544     cc.resize(ZS);
<a name="l01545"></a>01545 
<a name="l01546"></a>01546     vector&lt;double&gt; dnod;
<a name="l01547"></a>01547     vector&lt;double&gt; dnoh;
<a name="l01548"></a>01548     vector&lt;double&gt; dnoc,cnoc;
<a name="l01549"></a>01549     dnod.resize(ZS);
<a name="l01550"></a>01550     <span class="comment">//hnod.resize(ZS);</span>
<a name="l01551"></a>01551     <span class="comment">//cnod.resize(ZS);</span>
<a name="l01552"></a>01552     dnoh.resize(ZS);
<a name="l01553"></a>01553     <span class="comment">//hnoh.resize(ZS);</span>
<a name="l01554"></a>01554     <span class="comment">//cnoh.resize(ZS);</span>
<a name="l01555"></a>01555     dnoc.resize(ZS);
<a name="l01556"></a>01556     <span class="comment">//hnoc.resize(ZS);</span>
<a name="l01557"></a>01557     cnoc.resize(ZS);
<a name="l01558"></a>01558 
<a name="l01559"></a>01559     <span class="comment">// count up the number of contradictions</span>
<a name="l01560"></a>01560     <span class="comment">// cumulative contradictions</span>
<a name="l01561"></a>01561     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(dnod,zall.<a class="code" href="classbinned__data.html#a60ff501908d011c168b7621678d1dacf" title="penalty*number of hint genotypes not d">nod</a>,ZS);
<a name="l01562"></a>01562     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(dnoh,zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>,ZS);
<a name="l01563"></a>01563     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(dnoc,zall.<a class="code" href="classbinned__data.html#aa06f7ef273167f48d09163cfa0bca595" title="penalty*number of hint genotypes not c">noc</a>,ZS);
<a name="l01564"></a>01564     <span class="comment">//vreversecumsum(cnod,dnod,ZS);</span>
<a name="l01565"></a>01565     <span class="comment">//vreversecumsum(cnoh,dnoh,ZS);</span>
<a name="l01566"></a>01566     <a class="code" href="snp_8label_8cpp.html#a71b11dd6854f20d84e0f3f8993a6cdbe" title="takes a cumulative sum and makes it cumulate from the other end">vreversecumsum</a>(cnoc,dnoc,ZS);
<a name="l01567"></a>01567     <span class="comment">// multiply by the penalty to reduce computational load</span>
<a name="l01568"></a>01568     <span class="comment">// obsolete: contradictions are directly penalized in the bins</span>
<a name="l01569"></a>01569     <span class="comment">//vmult(cnoc,sp.contradictionpenalty);</span>
<a name="l01570"></a>01570     <span class="comment">//vmult(dnod,sp.contradictionpenalty);</span>
<a name="l01571"></a>01571     <span class="comment">//vmult(dnoh,sp.contradictionpenalty);</span>
<a name="l01572"></a>01572 
<a name="l01573"></a>01573     
<a name="l01574"></a>01574 
<a name="l01575"></a>01575     <span class="comment">// cumulative number - weight for bins</span>
<a name="l01576"></a>01576     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(nd,zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>,ZS);
<a name="l01577"></a>01577     NS = (int)floor(nd[ZS-1]+.00001); <span class="comment">// number of points total</span>
<a name="l01578"></a>01578     <a class="code" href="snp_8label_8cpp.html#a71b11dd6854f20d84e0f3f8993a6cdbe" title="takes a cumulative sum and makes it cumulate from the other end">vreversecumsum</a>(nc,nd,ZS);
<a name="l01579"></a>01579     <span class="comment">//cumulative sum of values</span>
<a name="l01580"></a>01580     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(dd,zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>,ZS);
<a name="l01581"></a>01581     <a class="code" href="snp_8label_8cpp.html#a71b11dd6854f20d84e0f3f8993a6cdbe" title="takes a cumulative sum and makes it cumulate from the other end">vreversecumsum</a>(cc,dd,ZS);
<a name="l01582"></a>01582     <span class="comment">// cumulative sum squares</span>
<a name="l01583"></a>01583     <a class="code" href="snp_8label_8cpp.html#ae947c9f1d5e7e344a40a3672ce12da20" title="cumulative sum of source vector using STL">vcumsum</a>(ddx,zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>,ZS);
<a name="l01584"></a>01584     <a class="code" href="snp_8label_8cpp.html#a71b11dd6854f20d84e0f3f8993a6cdbe" title="takes a cumulative sum and makes it cumulate from the other end">vreversecumsum</a>(ccx,ddx,ZS);
<a name="l01585"></a>01585 
<a name="l01586"></a>01586     <span class="keywordtype">int</span> tin;
<a name="l01587"></a>01587 
<a name="l01588"></a>01588     <span class="comment">// temporaries</span>
<a name="l01589"></a>01589     <span class="keywordtype">double</span> ycij,xcij,xcxij;
<a name="l01590"></a>01590     <span class="keywordtype">double</span> ydij,xdij,xdxij;
<a name="l01591"></a>01591     <span class="keywordtype">double</span> yhij,xhij,xhxij;
<a name="l01592"></a>01592     <span class="keywordtype">double</span> ldij,lcij,lhij;
<a name="l01593"></a>01593     <span class="keywordtype">double</span> dvdij,dvcij,dvhij;
<a name="l01594"></a>01594 
<a name="l01595"></a>01595     <span class="comment">// setup brlmm bits that don&#39;t change</span>
<a name="l01596"></a>01596     <span class="comment">// speed</span>
<a name="l01597"></a>01597             vector&lt;double&gt; m;
<a name="l01598"></a>01598             vector&lt;double&gt; Minv;
<a name="l01599"></a>01599             vector&lt;double&gt; Sinv;
<a name="l01600"></a>01600             <a class="code" href="snp_8label_8cpp.html#ac2829bc689e49ff601dd7a2cabb6eec2" title="setup the matrixes for doing individual BRLMM steps">setupBRLMM</a>(m,Minv,Sinv,sp);
<a name="l01601"></a>01601             <span class="comment">// only work with Sinv;</span>
<a name="l01602"></a>01602             <span class="comment">// scratch vectors</span>
<a name="l01603"></a>01603             vector&lt;double&gt; Tz;
<a name="l01604"></a>01604             vector&lt;double&gt; Zt;
<a name="l01605"></a>01605             vector&lt;double&gt; Xinv;
<a name="l01606"></a>01606             vector&lt;double&gt; Xt;
<a name="l01607"></a>01607             vector&lt;double&gt; Mz;
<a name="l01608"></a>01608             <span class="comment">// vectors that change</span>
<a name="l01609"></a>01609             <span class="comment">// sum of x</span>
<a name="l01610"></a>01610             vector&lt;double&gt; Nv;
<a name="l01611"></a>01611             <span class="comment">// number of data points</span>
<a name="l01612"></a>01612             vector&lt;double&gt; N;
<a name="l01613"></a>01613             Nv.resize(3);
<a name="l01614"></a>01614             N.assign(9,0);
<a name="l01615"></a>01615 
<a name="l01616"></a>01616     <span class="comment">// do some more setup</span>
<a name="l01617"></a>01617     <span class="comment">// so that things that aren&#39;t executed don&#39;t get repeated conditionals</span>
<a name="l01618"></a>01618     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a>&lt;2)
<a name="l01619"></a>01619     {
<a name="l01620"></a>01620         <span class="comment">// no hets</span>
<a name="l01621"></a>01621         <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01622"></a>01622             <span class="keywordflow">for</span> (j=i+1; j&lt;ZS; j++)
<a name="l01623"></a>01623             {
<a name="l01624"></a>01624                 tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01625"></a>01625                 q[tin] += NS*1000;
<a name="l01626"></a>01626             }
<a name="l01627"></a>01627     }
<a name="l01628"></a>01628 
<a name="l01629"></a>01629     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ab10a7ca6260c50da3d78db608ad5faa3" title="use hinted genotypes as reference">hints</a>==1)
<a name="l01630"></a>01630     {
<a name="l01631"></a>01631         <span class="comment">// penalty for contradiction of reference</span>
<a name="l01632"></a>01632         <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01633"></a>01633             <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l01634"></a>01634             {
<a name="l01635"></a>01635                 tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01636"></a>01636                 q[tin] += (dnod[i]+cnoc[j]+(dnoh[j]-dnoh[i]));
<a name="l01637"></a>01637             }
<a name="l01638"></a>01638 
<a name="l01639"></a>01639     }
<a name="l01640"></a>01640     
<a name="l01641"></a>01641     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a>==1)
<a name="l01642"></a>01642     {
<a name="l01643"></a>01643         <span class="comment">// mixture penalty under observed data</span>
<a name="l01644"></a>01644         <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01645"></a>01645             <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l01646"></a>01646             {
<a name="l01647"></a>01647                 tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01648"></a>01648                 q[tin] += <a class="code" href="snp_8label_8cpp.html#a13b6fb2fd4777966e2fc309ddac76b38" title="This penalizes unusual frequency distributions.">mixture_penalty</a>(nc[j],nd[i],nd[j]-nd[i],nc[j],nd[i],nd[j]-nd[i],sp.<a class="code" href="classsnp__param.html#a32739397c0d48abb7bf3b94ef3a11912" title="ensure avoid taking log of zero frequencies">SafetyFrequency</a>);
<a name="l01649"></a>01649             }
<a name="l01650"></a>01650     }
<a name="l01651"></a>01651     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a>==2)
<a name="l01652"></a>01652     {
<a name="l01653"></a>01653         <span class="comment">// mixture penalty under observed+prior number of observations</span>
<a name="l01654"></a>01654         <span class="comment">// change to use frequency not &quot;prior precision&quot;</span>
<a name="l01655"></a>01655         <span class="comment">// make room for better code</span>
<a name="l01656"></a>01656         <span class="comment">// i.e. posterior should be grouped correctly</span>
<a name="l01657"></a>01657         <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01658"></a>01658             <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l01659"></a>01659             {
<a name="l01660"></a>01660                 tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01661"></a>01661                 q[tin] += <a class="code" href="snp_8label_8cpp.html#a13b6fb2fd4777966e2fc309ddac76b38" title="This penalizes unusual frequency distributions.">mixture_penalty</a>(nc[j]+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,nd[i]+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,nd[j]-nd[i]+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,nc[j],nd[i],nd[j]-nd[i],sp.<a class="code" href="classsnp__param.html#a32739397c0d48abb7bf3b94ef3a11912" title="ensure avoid taking log of zero frequencies">SafetyFrequency</a>);
<a name="l01662"></a>01662             }
<a name="l01663"></a>01663     }
<a name="l01664"></a>01664     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a>==3)
<a name="l01665"></a>01665     {
<a name="l01666"></a>01666         <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01667"></a>01667             <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l01668"></a>01668             {
<a name="l01669"></a>01669                 tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01670"></a>01670                 q[tin] += <a class="code" href="snp_8label_8cpp.html#a188ea3959b9c819926f6a1ea3207f4d6" title="This penalizes unusual cluster distributions (violating HW extremely) Expected to be useful for scree...">HW_penalty</a>(nc[j],nd[i],nd[j]-nd[i], 0.5*sp.<a class="code" href="classsnp__param.html#a32739397c0d48abb7bf3b94ef3a11912" title="ensure avoid taking log of zero frequencies">SafetyFrequency</a>);
<a name="l01671"></a>01671             }
<a name="l01672"></a>01672     }
<a name="l01673"></a>01673     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a>&gt;0)
<a name="l01674"></a>01674     {
<a name="l01675"></a>01675         <span class="comment">// if useful to do some bic correction</span>
<a name="l01676"></a>01676         <span class="comment">// bic * log(n) - empirical correction</span>
<a name="l01677"></a>01677         <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01678"></a>01678         {
<a name="l01679"></a>01679             <span class="comment">// subtract off once for 2 cluster</span>
<a name="l01680"></a>01680             <span class="comment">// [0,0],[0,ZS-1],[ZS-1,ZS-1] occur twice</span>
<a name="l01681"></a>01681             <span class="comment">// so subtract off 2 times for 1 cluster</span>
<a name="l01682"></a>01682             tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,i,ZS);
<a name="l01683"></a>01683                     q[tin] -= sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a>*log((<span class="keywordtype">double</span>)NS);
<a name="l01684"></a>01684             tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(0,i,ZS);
<a name="l01685"></a>01685             q[tin] -= sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a>*log((<span class="keywordtype">double</span>)NS);
<a name="l01686"></a>01686             tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,ZS-1,ZS);
<a name="l01687"></a>01687             q[tin] -= sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a>*log((<span class="keywordtype">double</span>)NS);
<a name="l01688"></a>01688             <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l01689"></a>01689             {
<a name="l01690"></a>01690                 tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01691"></a>01691                 <span class="comment">// add 3 times for everyone</span>
<a name="l01692"></a>01692                 q[tin] += 3*sp.<a class="code" href="classsnp__param.html#adfe46fd4d9582396882e89eaf3592c5a" title="bic penalty level bic*nparam*log(n)">bic</a>*log((<span class="keywordtype">double</span>)NS);
<a name="l01693"></a>01693             }
<a name="l01694"></a>01694         }
<a name="l01695"></a>01695     }
<a name="l01696"></a>01696     <span class="comment">// likelihood of data under posterior</span>
<a name="l01697"></a>01697     <span class="comment">// (x-m)^2 = x^2-2*x*m+m^2</span>
<a name="l01698"></a>01698     <span class="comment">// running sums sum(x^2)-2*(sum(x)*m+sum(1)*m^2</span>
<a name="l01699"></a>01699     <span class="comment">// big loop: try all reasonable labelings</span>
<a name="l01700"></a>01700     <span class="comment">// bb^i-&gt;ab^(j-i)-&gt;aa^(n-j+1)</span>
<a name="l01701"></a>01701     <span class="keywordflow">for</span> (i=0; i&lt;ZS; i++)
<a name="l01702"></a>01702         <span class="keywordflow">for</span> (j=i; j&lt;ZS; j++)
<a name="l01703"></a>01703         {
<a name="l01704"></a>01704             tin = <a class="code" href="snp_8label_8cpp.html#a05c9fe8994d405ac2e71e60d84a41a3c" title="hash indexer">hax</a>(i,j,ZS);
<a name="l01705"></a>01705             <span class="comment">// initialize global matrix</span>
<a name="l01706"></a>01706             <span class="comment">// c</span>
<a name="l01707"></a>01707             ycij = nc[j];
<a name="l01708"></a>01708             xcij = cc[j];
<a name="l01709"></a>01709             xcxij = ccx[j];
<a name="l01710"></a>01710             <span class="comment">// d</span>
<a name="l01711"></a>01711             ydij = nd[i];
<a name="l01712"></a>01712             xdij = dd[i];
<a name="l01713"></a>01713             xdxij = ddx[i];
<a name="l01714"></a>01714             <span class="comment">// h</span>
<a name="l01715"></a>01715             yhij = nd[j]-nd[i];
<a name="l01716"></a>01716             xhij = dd[j]-dd[i];
<a name="l01717"></a>01717             xhxij = ddx[j]-ddx[i];
<a name="l01718"></a>01718 
<a name="l01719"></a>01719 
<a name="l01720"></a>01720             <span class="comment">// brlmm here</span>
<a name="l01721"></a>01721             <span class="comment">// setup centers</span>
<a name="l01722"></a>01722             Nv[0] = xcij;
<a name="l01723"></a>01723             Nv[1] = xhij;
<a name="l01724"></a>01724             Nv[2] = xdij;
<a name="l01725"></a>01725             <span class="comment">// number of data points</span>
<a name="l01726"></a>01726             N[0] = ycij;
<a name="l01727"></a>01727             N[4] = yhij;
<a name="l01728"></a>01728             N[8] = ydij;
<a name="l01729"></a>01729             <span class="comment">// do some inversions</span>
<a name="l01730"></a>01730             <a class="code" href="snp_8label_8cpp.html#afc25b20145576567e02ac9bfa60efaa0" title="multiply two 3x3 matrices">timesthree</a>(Tz,N,Sinv);
<a name="l01731"></a>01731             <a class="code" href="snp_8label_8cpp.html#a1f2ff039971eee5c78e94867f1cc9369" title="sum two vectors">sumthree</a>(Zt, Minv,Tz);
<a name="l01732"></a>01732             <a class="code" href="snp_8label_8cpp.html#a57e554714be3edbd1b2393e5824b03b4" title="inverts 3x3 matrix specialty code, avoid calling external routines">inversethree</a>(Xinv,Zt);
<a name="l01733"></a>01733             <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(Tz,Sinv, Nv);
<a name="l01734"></a>01734             <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(Zt,Minv, m);
<a name="l01735"></a>01735             <a class="code" href="snp_8label_8cpp.html#a1f2ff039971eee5c78e94867f1cc9369" title="sum two vectors">sumthree</a>(Xt,Tz,Zt);
<a name="l01736"></a>01736             <a class="code" href="snp_8label_8cpp.html#a0119ec6004b03769d03578f1a2c7e289" title="multiply a 1x3 vector by a 3x3 matrix">timesone</a>(Mz,Xinv,Xt);
<a name="l01737"></a>01737             <span class="comment">// Tz = three means</span>
<a name="l01738"></a>01738             <span class="comment">//printf(&quot;%d %f %f %f\n&quot;, tin, Mz[0],Mz[1],Mz[2]);</span>
<a name="l01739"></a>01739             
<a name="l01740"></a>01740             
<a name="l01741"></a>01741             <span class="comment">// figure out the triple mean/variance combo</span>
<a name="l01742"></a>01742             ldij = Mz[2];
<a name="l01743"></a>01743             lhij = Mz[1];
<a name="l01744"></a>01744             lcij = Mz[0];
<a name="l01745"></a>01745             <span class="comment">// force means apart by shellbarrier</span>
<a name="l01746"></a>01746             <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a>==3)
<a name="l01747"></a>01747                 force_isotonic(ldij,lhij,lcij,ydij+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,yhij+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,ydij+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>, sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a>);
<a name="l01748"></a>01748             <span class="comment">// compute variance given means</span>
<a name="l01749"></a>01749             <span class="comment">// within-cluster variance + prior + distance shifted</span>
<a name="l01750"></a>01750             dvcij = <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,xcxij,xcij,ycij,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,lcij,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01751"></a>01751             dvdij = <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,xdxij,xdij,ydij,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,ldij,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01752"></a>01752             dvhij = <a class="code" href="snp_8label_8cpp.html#aafff55244ded61e2583fe23532e19799" title="bayesian variance - prior + observed + shift in means">b_var</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,xhxij,xhij,yhij,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>,lhij,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>);
<a name="l01753"></a>01753 
<a name="l01754"></a>01754             <span class="comment">// fix up common variance</span>
<a name="l01755"></a>01755             <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a5996056f7436f5974c6e6b6bd7925a27" title="do we use common variances for all clusters?">comvar</a>==1)
<a name="l01756"></a>01756             {
<a name="l01757"></a>01757                 <span class="keywordtype">double</span> tc,td,th;
<a name="l01758"></a>01758                 <span class="keywordtype">double</span> tnc,tnd,tnh;
<a name="l01759"></a>01759                 <span class="keywordtype">double</span> tl,tlm;
<a name="l01760"></a>01760                 tnd = ydij+sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01761"></a>01761                 tnc = ycij+sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01762"></a>01762                 tnh = yhij+sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>;
<a name="l01763"></a>01763                 td = dvdij*tnd;
<a name="l01764"></a>01764                 tc = dvcij*tnc;
<a name="l01765"></a>01765                 th = dvhij*tnh;
<a name="l01766"></a>01766                 <span class="comment">// cross-talk between variances 0 = independent, 1 = common variance</span>
<a name="l01767"></a>01767                 tl = sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>;
<a name="l01768"></a>01768                 tlm = 3-2*sp.<a class="code" href="classsnp__param.html#adf74b0bee8088a3039b721b11abc585e" title="control mixing of variances if common">lambda</a>;
<a name="l01769"></a>01769                 dvdij = (tlm*td+tl*tc+tl*th)/(tlm*tnd+tl*tnc+tl*tnh);
<a name="l01770"></a>01770                 dvcij = (tl*td+tlm*tc+tl*th)/(tl*tnd+tlm*tnc+tl*tnh);
<a name="l01771"></a>01771                 dvhij = (tl*td+tl*tc+tlm*th)/(tl*tnd+tl*tnc+tlm*tnh);
<a name="l01772"></a>01772             }<span class="comment"></span>
<a name="l01773"></a>01773 <span class="comment">        ////</span>
<a name="l01774"></a>01774 <span class="comment">        ////</span>
<a name="l01775"></a>01775 <span class="comment">        //// now do the global likelihood with calculated values</span>
<a name="l01776"></a>01776 <span class="comment"></span>            <span class="comment">// likelihood of the data gvien mean/variance    </span>
<a name="l01777"></a>01777             <span class="comment">//q[tin] = 0;</span>
<a name="l01778"></a>01778             q[tin] += <a class="code" href="snp_8label_8cpp.html#a2ca380c421378333fbafecc2304e27a6" title="gaussian likelihood, in convenient running sums form">l_posterior</a>(xdxij,ldij,xdij,ydij,dvdij);
<a name="l01779"></a>01779             q[tin] += <a class="code" href="snp_8label_8cpp.html#a2ca380c421378333fbafecc2304e27a6" title="gaussian likelihood, in convenient running sums form">l_posterior</a>(xhxij,lhij,xhij,yhij,dvhij);
<a name="l01780"></a>01780             q[tin] += <a class="code" href="snp_8label_8cpp.html#a2ca380c421378333fbafecc2304e27a6" title="gaussian likelihood, in convenient running sums form">l_posterior</a>(xcxij,lcij,xcij,ycij,dvcij);    
<a name="l01781"></a>01781             <span class="comment">//printf(&quot;%d %lf\n&quot;, i, q[i]);    </span>
<a name="l01782"></a>01782     
<a name="l01783"></a>01783             <span class="comment">// likelihood of mean under prior</span>
<a name="l01784"></a>01784             <span class="comment">//q[tin] += l_normal(sp.bb.m,ldij,sp.bb.ss/sp.bb.k);</span>
<a name="l01785"></a>01785             <span class="comment">//q[tin] += l_normal(sp.ab.m,lhij,sp.ab.ss/sp.ab.k);</span>
<a name="l01786"></a>01786             <span class="comment">//q[tin] += l_normal(sp.aa.m,lcij,sp.aa.ss/sp.aa.k);</span>
<a name="l01787"></a>01787             q[tin] += <a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,ldij,1/Minv[8]);
<a name="l01788"></a>01788             q[tin] += <a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,lhij,1/Minv[4]);
<a name="l01789"></a>01789             q[tin] += <a class="code" href="snp_8label_8cpp.html#a7280a531de940ab662d2c20605be5d21" title="gaussian likelihood, for just one piece of data">l_normal</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a6d6d0888d2b43c15e866d2beaaf63643" title="mean of cluster">m</a>,lcij,1/Minv[0]);
<a name="l01790"></a>01790             <span class="comment">// likelihood of variance under prior</span>
<a name="l01791"></a>01791             q[tin] += <a class="code" href="snp_8label_8cpp.html#ad9a6d2941c7fb47be8cd441cbb60c2d9" title="cheap inverse-gamma likelihood, for one piece of data">l_inverse</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,dvdij,sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>);
<a name="l01792"></a>01792             q[tin] += <a class="code" href="snp_8label_8cpp.html#ad9a6d2941c7fb47be8cd441cbb60c2d9" title="cheap inverse-gamma likelihood, for one piece of data">l_inverse</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,dvhij,sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>);
<a name="l01793"></a>01793             q[tin] += <a class="code" href="snp_8label_8cpp.html#ad9a6d2941c7fb47be8cd441cbb60c2d9" title="cheap inverse-gamma likelihood, for one piece of data">l_inverse</a>(sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#ab5687aae0f32c286b69bd0cf5e0d9d37" title="variance of cluster">ss</a>,dvcij,sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#aaaf1dec521d21ec061d6d9e7c3b1e78c" title="strength of variance (pseudo-observations)">v</a>);
<a name="l01794"></a>01794             <span class="comment">// half</span>
<a name="l01795"></a>01795             q[tin] /= 2;
<a name="l01796"></a>01796             
<a name="l01797"></a>01797             <span class="comment">// penalize assorted peculiar features of the data</span>
<a name="l01798"></a>01798             <span class="comment">// i.e. too close, too strange</span>
<a name="l01799"></a>01799             <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a>==2)
<a name="l01800"></a>01800                 q[tin] += NS*<a class="code" href="snp_8label_8cpp.html#a448b37d74fcf36fe83e37c69e5c74b2b" title="check if centers too close, return bignum if true">fix_shell</a>(ldij,lcij,lhij,sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a>);
<a name="l01801"></a>01801             <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1900bf669c78de37107156241c6adcd4" title="do we stop cluster centers from being too close?">hardshell</a>==1)
<a name="l01802"></a>01802                 q[tin] += NS*<a class="code" href="snp_8label_8cpp.html#a6d67d855bb034343920e82fbaf1bdb2b" title="prevents hom clusters from approaching midline not useful on B chip due to asymmetric probe choice...">fix_hom_shell</a>(ldij,lcij,lhij,sp.<a class="code" href="classsnp__param.html#aea03fd76d67d7a0ebf46827c26f684e9" title="how close can cluster centers be?">shellbarrier</a>);
<a name="l01803"></a>01803             <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ab30f88b4fe7b2476acf77727745ae6ab" title="penalty for FLD separation too low">CSepPen</a>&gt;0)
<a name="l01804"></a>01804             {
<a name="l01805"></a>01805                 <span class="comment">// avoid cluster splitting by ad-hoc penalty for FLD too small</span>
<a name="l01806"></a>01806                 <span class="comment">// rather ad-hoc favor for having FLD large!</span>
<a name="l01807"></a>01807                 <span class="keywordtype">double</span> flddh,fldhc,flddc;
<a name="l01808"></a>01808                 <span class="comment">// compute FLD for cluster separation</span>
<a name="l01809"></a>01809                 flddh = (ldij-lhij)*(ldij-lhij)/(dvdij+dvhij);
<a name="l01810"></a>01810                 fldhc = (lhij-lcij)*(lhij-lcij)/(dvhij+dvcij);
<a name="l01811"></a>01811                 flddc = (ldij-lcij)*(ldij-lcij)/(dvdij+dvcij);
<a name="l01812"></a>01812                 <span class="comment">// threshold FLD to avoid too much repulsion</span>
<a name="l01813"></a>01813                 flddh = flddh/(1+flddh/sp.<a class="code" href="classsnp__param.html#a3fad707f15cabcac4f8c7097781fd9b8" title="stop penalizing using Geman-McClure">CSepThr</a>);
<a name="l01814"></a>01814                 fldhc = fldhc/(1+fldhc/sp.<a class="code" href="classsnp__param.html#a3fad707f15cabcac4f8c7097781fd9b8" title="stop penalizing using Geman-McClure">CSepThr</a>);
<a name="l01815"></a>01815                 <span class="comment">// might have only homs, so keep this</span>
<a name="l01816"></a>01816                 flddc = flddc/(1+flddc/(2*sp.<a class="code" href="classsnp__param.html#a3fad707f15cabcac4f8c7097781fd9b8" title="stop penalizing using Geman-McClure">CSepThr</a>));
<a name="l01817"></a>01817                 <span class="comment">// apply data points at risk</span>
<a name="l01818"></a>01818                 flddh *= (ydij+yhij);
<a name="l01819"></a>01819                 fldhc *= (yhij+ycij);
<a name="l01820"></a>01820                 flddc *= (ydij+ycij);
<a name="l01821"></a>01821                 <span class="comment">// favor! [larger FLD better, up to a point]</span>
<a name="l01822"></a>01822                 q[tin] -= sp.<a class="code" href="classsnp__param.html#ab30f88b4fe7b2476acf77727745ae6ab" title="penalty for FLD separation too low">CSepPen</a> * (flddh+fldhc+flddc);
<a name="l01823"></a>01823             }
<a name="l01824"></a>01824         }
<a name="l01825"></a>01825 
<a name="l01826"></a>01826     
<a name="l01827"></a>01827     <a class="code" href="snp_8label_8cpp.html#a8e097ae837d7e8b8ff40b974a07570aa" title="convert log-likelihood of all labelings into relative probability">qtoeq</a>(q,ZS);
<a name="l01828"></a>01828     <span class="comment">// convert q to label</span>
<a name="l01829"></a>01829     <a class="code" href="snp_8label_8cpp.html#a4e0eeb85252ceaca83c545aef7ba4c4e" title="turn the overall relative probability of labelings into relative probability of individual genotypes...">eqtoproblabel</a>(ec,eh,ed,ZS,q);
<a name="l01830"></a>01830 }
<a name="l01831"></a>01831 <span class="comment"></span>
<a name="l01832"></a>01832 <span class="comment">/**</span>
<a name="l01833"></a>01833 <span class="comment"> * tunes up the prior to allow for cluster shifts</span>
<a name="l01834"></a>01834 <span class="comment"> *</span>
<a name="l01835"></a>01835 <span class="comment"> */</span>
<a name="l01836"></a><a class="code" href="snp_8label_8cpp.html#a8f4d65cf5118d9aad3b3c4e8c40e613f">01836</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a8f4d65cf5118d9aad3b3c4e8c40e613f" title="tunes up the prior to allow for cluster shifts">apply_wobble</a>(<a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l01837"></a>01837 {
<a name="l01838"></a>01838     <span class="comment">// apply &quot;wobble&quot; to account for possible cluster shifts</span>
<a name="l01839"></a>01839     <span class="comment">// i.e. even if infinite amount of prior data</span>
<a name="l01840"></a>01840     <span class="comment">// this experimental set could be different</span>
<a name="l01841"></a>01841     <span class="comment">//</span>
<a name="l01842"></a>01842     <span class="comment">// // change to &quot;f&quot; here to preserve prior relative frequency after changing k</span>
<a name="l01843"></a>01843     <span class="comment">//</span>
<a name="l01844"></a>01844     <span class="comment">// note that this is currently &quot;undone&quot; in the actual 1-D seeding routine to preserve continuity of behavior</span>
<a name="l01845"></a>01845     <span class="comment">// flag will be installed to allow better behavior</span>
<a name="l01846"></a>01846     
<a name="l01847"></a>01847     <span class="keywordtype">double</span> tmp;
<a name="l01848"></a>01848     tmp = 1/sp.<a class="code" href="classsnp__param.html#aabee7993a0cd2545731ff91f5d6cce4e" title="how much do we allow clusters to shift from prior">wobble</a>;
<a name="l01849"></a>01849     <span class="keywordflow">if</span> (tmp&lt;sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>)
<a name="l01850"></a>01850         sp.prior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = tmp;
<a name="l01851"></a>01851     <span class="keywordflow">if</span> (tmp&lt;sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>)
<a name="l01852"></a>01852         sp.prior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = tmp;
<a name="l01853"></a>01853     <span class="keywordflow">if</span> (tmp&lt;sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a>)
<a name="l01854"></a>01854         sp.prior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a09bcd6068f14afa7d1f52c270e3a2be7" title="strength of mean (pseudo-observations)">k</a> = tmp;
<a name="l01855"></a>01855     <span class="comment">// note that the posterior will reflect the &quot;lower&quot; preknowledge</span>
<a name="l01856"></a>01856     <span class="comment">// of the prior data</span>
<a name="l01857"></a>01857     
<a name="l01858"></a>01858 }
<a name="l01859"></a>01859 
<a name="l01860"></a>01860 <span class="comment">/*</span>
<a name="l01861"></a>01861 <span class="comment"> * Do QC check to see if values are weird, indicative of error</span>
<a name="l01862"></a>01862 <span class="comment"> *</span>
<a name="l01863"></a>01863 <span class="comment"> * @param call - genotype per data point</span>
<a name="l01864"></a>01864 <span class="comment"> * @param conf - confidence per data point</span>
<a name="l01865"></a>01865 <span class="comment"> * @param y - strength per data point</span>
<a name="l01866"></a>01866 <span class="comment"> */</span>
<a name="l01867"></a>01867 <span class="keywordtype">void</span> qc_modify_confidences(vector&lt;affx::GType&gt; &amp;call, vector&lt;double&gt; &amp;conf, vector&lt;double&gt; &amp;y, <span class="keywordtype">double</span> perr)
<a name="l01868"></a>01868 {
<a name="l01869"></a>01869     <span class="comment">// use scatter in y direction to find outlier data points</span>
<a name="l01870"></a>01870     <span class="comment">// outlier = pr(under model for y)&lt;pr(error)</span>
<a name="l01871"></a>01871     <span class="comment">// assume normal distribution of residuals</span>
<a name="l01872"></a>01872     <span class="comment">// assume common variance (improve power for minor cluster)</span>
<a name="l01873"></a>01873     <span class="comment">// don&#39;t assume common mean</span>
<a name="l01874"></a>01874     <span class="keywordtype">int</span> cIx, i;
<a name="l01875"></a>01875     <span class="keywordtype">double</span> tmp,num,ss;
<a name="l01876"></a>01876     vector&lt;double&gt; residual;
<a name="l01877"></a>01877     <span class="keywordtype">double</span> wgt;
<a name="l01878"></a>01878 
<a name="l01879"></a>01879 
<a name="l01880"></a>01880     residual.resize(y.size());
<a name="l01881"></a>01881     <span class="keywordflow">for</span> (cIx=0; cIx&lt;3; cIx++) <span class="comment">// valid calls</span>
<a name="l01882"></a>01882     {
<a name="l01883"></a>01883         tmp = 0;
<a name="l01884"></a>01884         num = 0.00001; <span class="comment">// tiny</span>
<a name="l01885"></a>01885         <span class="keywordflow">for</span> (i=0; i&lt;call.size(); i++)
<a name="l01886"></a>01886             <span class="keywordflow">if</span> (call[i]==cIx)
<a name="l01887"></a>01887             {
<a name="l01888"></a>01888                 tmp += y[i];
<a name="l01889"></a>01889                 num +=1;
<a name="l01890"></a>01890             }
<a name="l01891"></a>01891         tmp/=num; <span class="comment">// now have mean</span>
<a name="l01892"></a>01892         <span class="keywordflow">for</span> (i=0; i&lt;call.size(); i++)
<a name="l01893"></a>01893             <span class="keywordflow">if</span> (call[i]==cIx)
<a name="l01894"></a>01894             {
<a name="l01895"></a>01895                 residual[i] = y[i]-tmp;
<a name="l01896"></a>01896             }
<a name="l01897"></a>01897     }
<a name="l01898"></a>01898     ss = 0;
<a name="l01899"></a>01899     <span class="keywordflow">for</span> (i=0; i&lt;residual.size(); i++)
<a name="l01900"></a>01900         ss+= residual[i]*residual[i];
<a name="l01901"></a>01901     ss /= residual.size(); <span class="comment">// degrees of freedom close enough</span>
<a name="l01902"></a>01902     ss += 0.001; <span class="comment">// avoid zero</span>
<a name="l01903"></a>01903     <span class="keywordflow">for</span> (i=0; i&lt;residual.size(); i++)
<a name="l01904"></a>01904     {
<a name="l01905"></a>01905         <span class="comment">// &quot;standardized&quot; probability (don&#39;t care about variance)</span>
<a name="l01906"></a>01906         tmp = exp(-1*residual[i]*residual[i]/(2*ss))/(2.51);
<a name="l01907"></a>01907         <span class="comment">// generate a weight</span>
<a name="l01908"></a>01908         <span class="keywordflow">if</span> (tmp&gt;0)
<a name="l01909"></a>01909             wgt = perr/tmp;
<a name="l01910"></a>01910         <span class="keywordflow">else</span>
<a name="l01911"></a>01911             wgt = perr/(0.000001);
<a name="l01912"></a>01912         conf[i] = (conf[i] + wgt)/(1+wgt); <span class="comment">// perr = &quot;relative likelihood of error&quot;</span>
<a name="l01913"></a>01913     }
<a name="l01914"></a>01914 }
<a name="l01915"></a>01915 
<a name="l01916"></a>01916 <span class="keywordtype">void</span> qc_compute_copy_posterior(vector&lt;affx::GType&gt; call, vector&lt;double&gt; &amp;y, <a class="code" href="classsnp__distribution.html" title="this class holds all three genotype clusters">snp_distribution</a> &amp;posterior)
<a name="l01917"></a>01917 {
<a name="l01918"></a>01918     vector&lt;double&gt; cmean;
<a name="l01919"></a>01919     vector&lt;double&gt; cnum;
<a name="l01920"></a>01920     vector&lt;double&gt; residual;
<a name="l01921"></a>01921     <span class="keywordtype">double</span> shrinkage = 0.00001;
<a name="l01922"></a>01922     <span class="keywordtype">double</span> vshrinkage = 0.1;
<a name="l01923"></a>01923     <span class="keywordtype">double</span> minvar = 0.1;
<a name="l01924"></a>01924     <span class="keywordtype">int</span> cIx, i;
<a name="l01925"></a>01925     <span class="keywordtype">double</span> cy,ss;
<a name="l01926"></a>01926     
<a name="l01927"></a>01927     <span class="comment">// old_style computation</span>
<a name="l01928"></a>01928     residual.resize(call.size());
<a name="l01929"></a>01929     cmean.resize(3);
<a name="l01930"></a>01930     cnum.resize(3);
<a name="l01931"></a>01931     
<a name="l01932"></a>01932     <span class="keywordflow">for</span> (cIx=0; cIx&lt;3; cIx++)
<a name="l01933"></a>01933     {
<a name="l01934"></a>01934         cmean[cIx] = 0;
<a name="l01935"></a>01935         cnum[cIx] = 0;
<a name="l01936"></a>01936         <span class="keywordflow">for</span> (i=0; i&lt;call.size(); i++)
<a name="l01937"></a>01937         {
<a name="l01938"></a>01938             <span class="keywordflow">if</span> (call[i]==cIx)
<a name="l01939"></a>01939             {
<a name="l01940"></a>01940                 cmean[cIx] += y[i];
<a name="l01941"></a>01941                 cnum[cIx] +=1;
<a name="l01942"></a>01942             }
<a name="l01943"></a>01943         }
<a name="l01944"></a>01944     }
<a name="l01945"></a>01945     cy = (cmean[0] + cmean[1] + cmean[2])/(cnum[0]+cnum[1]+cnum[2]);
<a name="l01946"></a>01946     
<a name="l01947"></a>01947     <span class="comment">// shrink towards common mean to fill in nonexistent calls</span>
<a name="l01948"></a>01948     cmean[0] = (cy*shrinkage+cmean[0])/(cnum[0]+shrinkage);
<a name="l01949"></a>01949     cmean[1] = (cy*shrinkage+cmean[1])/(cnum[1]+shrinkage);
<a name="l01950"></a>01950     cmean[2] = (cy*shrinkage+cmean[2])/(cnum[2]+shrinkage);
<a name="l01951"></a>01951     <span class="comment">// all three exist</span>
<a name="l01952"></a>01952     <span class="keywordflow">for</span> (i=0; i&lt;call.size(); i++)
<a name="l01953"></a>01953     {
<a name="l01954"></a>01954         <span class="comment">// if (call[i]&gt;-1 &amp;&amp; call[i]&lt;3) {</span>
<a name="l01955"></a>01955     <span class="keywordflow">if</span> (<a class="code" href="namespaceaffx.html#a540bbcd077173d8d2349e402dbbe5f0e" title="Check that the gtype is AA,AB,BB (not NN) ///.">affx::GType_called</a>(call[i])) { <span class="comment">// aa,ab,bb</span>
<a name="l01956"></a>01956             residual[i] = y[i] - cmean[call[i]];
<a name="l01957"></a>01957         }
<a name="l01958"></a>01958         <span class="keywordflow">else</span> {
<a name="l01959"></a>01959             residual[i] = 0.0;
<a name="l01960"></a>01960     }
<a name="l01961"></a>01961     }
<a name="l01962"></a>01962     ss = 0;
<a name="l01963"></a>01963     <span class="keywordflow">for</span> (i=0; i&lt;residual.size(); i++)
<a name="l01964"></a>01964         ss+= residual[i]*residual[i];
<a name="l01965"></a>01965     ss = (ss + minvar*vshrinkage)/(residual.size()+vshrinkage); <span class="comment">// degrees of freedom close enough for large data sets</span>
<a name="l01966"></a>01966 
<a name="l01967"></a>01967     <span class="comment">// output means (shrunk towards center)</span>
<a name="l01968"></a>01968     posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = cmean[0];
<a name="l01969"></a>01969     posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = cmean[1];
<a name="l01970"></a>01970     posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a> = cmean[2];
<a name="l01971"></a>01971     <span class="comment">// output common variance</span>
<a name="l01972"></a>01972     posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> = ss;
<a name="l01973"></a>01973     posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> = ss;
<a name="l01974"></a>01974     posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a> = ss;
<a name="l01975"></a>01975 }
<a name="l01976"></a>01976 
<a name="l01977"></a>01977 
<a name="l01978"></a>01978 <span class="comment">/*</span>
<a name="l01979"></a>01979 <span class="comment"> * Do QC check to see if values are weird, indicative of error</span>
<a name="l01980"></a>01980 <span class="comment"> *</span>
<a name="l01981"></a>01981 <span class="comment"> * @param call - genotype per data point</span>
<a name="l01982"></a>01982 <span class="comment"> * @param conf - confidence per data point</span>
<a name="l01983"></a>01983 <span class="comment"> * @param y - strength per data point</span>
<a name="l01984"></a>01984 <span class="comment"> */</span>
<a name="l01985"></a>01985 <span class="keywordtype">void</span> qc_posterior_modify_confidences(vector&lt;affx::GType&gt; &amp;call, vector&lt;double&gt; &amp;conf, vector&lt;double&gt; &amp;y, <span class="keywordtype">double</span> perr, <a class="code" href="classsnp__distribution.html" title="this class holds all three genotype clusters">snp_distribution</a> dist)
<a name="l01986"></a>01986 {
<a name="l01987"></a>01987     <span class="comment">// use scatter in y direction to find outlier data points</span>
<a name="l01988"></a>01988     <span class="comment">// outlier = pr(under model for y)&lt;pr(error)</span>
<a name="l01989"></a>01989     <span class="comment">// assume normal distribution of residuals</span>
<a name="l01990"></a>01990     <span class="comment">// assume common variance (improve power for minor cluster)</span>
<a name="l01991"></a>01991     <span class="comment">// don&#39;t assume common mean</span>
<a name="l01992"></a>01992     <span class="keywordtype">int</span> cIx, i;
<a name="l01993"></a>01993     <span class="keywordtype">double</span> tmp;
<a name="l01994"></a>01994   <span class="comment">// double num,ss;</span>
<a name="l01995"></a>01995     <span class="keywordtype">double</span> residual;
<a name="l01996"></a>01996     vector&lt;double&gt; cmean,cvar;
<a name="l01997"></a>01997     <span class="keywordtype">double</span> wgt;
<a name="l01998"></a>01998 
<a name="l01999"></a>01999     cmean.resize(3);
<a name="l02000"></a>02000     cvar.resize(3);
<a name="l02001"></a>02001     cmean[0] = dist.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l02002"></a>02002     cmean[1] = dist.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l02003"></a>02003     cmean[2] = dist.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a574a5134fb9a0d351afb6c08421f74c4" title="mean of cluster in other dimension">ym</a>;
<a name="l02004"></a>02004 
<a name="l02005"></a>02005     cvar[0] = dist.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>;
<a name="l02006"></a>02006     cvar[1] = dist.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>;
<a name="l02007"></a>02007     cvar[2] = dist.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>.<a class="code" href="classcluster__data.html#a1ef1f56306029c073109d5fcc9922e1d" title="variance of cluster in other dimension">yss</a>;
<a name="l02008"></a>02008     
<a name="l02009"></a>02009     <span class="keywordflow">for</span> (cIx=0; cIx&lt;3; cIx++) <span class="comment">// valid calls</span>
<a name="l02010"></a>02010     {
<a name="l02011"></a>02011         <span class="keywordflow">for</span> (i=0; i&lt;call.size(); i++)
<a name="l02012"></a>02012             <span class="keywordflow">if</span> (call[i]==cIx)
<a name="l02013"></a>02013             {
<a name="l02014"></a>02014                 residual = y[i]-cmean[cIx];
<a name="l02015"></a>02015                 tmp = exp(-1*residual*residual/(2*cvar[cIx]))/(2.51);
<a name="l02016"></a>02016                 <span class="comment">// generate a weight</span>
<a name="l02017"></a>02017                 <span class="keywordflow">if</span> (tmp&gt;0)
<a name="l02018"></a>02018                 {
<a name="l02019"></a>02019                     wgt = perr/tmp;
<a name="l02020"></a>02020                     conf[i] = (conf[i] + wgt)/(1+wgt); <span class="comment">// perr = &quot;relative likelihood of error&quot;</span>
<a name="l02021"></a>02021                 }
<a name="l02022"></a>02022                 <span class="keywordflow">else</span>
<a name="l02023"></a>02023                     conf[i] = 1;            
<a name="l02024"></a>02024             }
<a name="l02025"></a>02025     }
<a name="l02026"></a>02026 }
<a name="l02027"></a>02027 <span class="comment"></span>
<a name="l02028"></a>02028 <span class="comment">/**</span>
<a name="l02029"></a>02029 <span class="comment"> * Set up the binned data from the regular data</span>
<a name="l02030"></a>02030 <span class="comment"> * compute running sums, allocate sorted bins</span>
<a name="l02031"></a>02031 <span class="comment"> *</span>
<a name="l02032"></a>02032 <span class="comment"> * @param zall - overall structure to keep track of data + soft genotype labels by bin</span>
<a name="l02033"></a>02033 <span class="comment"> * @param x    - initial data points (unsorted)</span>
<a name="l02034"></a>02034 <span class="comment"> * @param genohints - hints associated with data points</span>
<a name="l02035"></a>02035 <span class="comment"> * @param numbins - how many bins to make, if bins&gt;points, go with points</span>
<a name="l02036"></a>02036 <span class="comment"> *</span>
<a name="l02037"></a>02037 <span class="comment"> */</span>
<a name="l02038"></a><a class="code" href="snp_8label_8cpp.html#a2979dac75c6e93d694bae23f061b1be8">02038</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a2979dac75c6e93d694bae23f061b1be8" title="Set up the binned data from the regular data compute running sums, allocate sorted bins...">initialize_bins</a>(<a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> &amp;zall, vector&lt;double&gt; &amp;x, vector&lt;double&gt; &amp;y, vector&lt;int&gt; &amp;genohints, vector&lt;double&gt; &amp;SubsetInbred, <span class="keywordtype">int</span> numbins, <span class="keywordtype">int</span> hok, <span class="keywordtype">double</span> CP)
<a name="l02039"></a>02039 {
<a name="l02040"></a>02040     <span class="keywordtype">int</span> i;
<a name="l02041"></a>02041 
<a name="l02042"></a>02042     <span class="comment">// takes a sorted vector of length z and labels it</span>
<a name="l02043"></a>02043     <span class="comment">// this will be different when using bins</span>
<a name="l02044"></a>02044 
<a name="l02045"></a>02045 
<a name="l02046"></a>02046 
<a name="l02047"></a>02047     <span class="comment">// call sides &quot;d&quot; and &quot;c&quot; to avoid genotype confusion</span>
<a name="l02048"></a>02048     <span class="comment">// negative genotypes are &quot;d&quot;, positive are &quot;c&quot;, middle are &quot;h&quot;</span>
<a name="l02049"></a>02049 
<a name="l02050"></a>02050     <span class="comment">// divide sorted points into bins by z</span>
<a name="l02051"></a>02051     <span class="comment">// division by range(points)/(bins+1)</span>
<a name="l02052"></a>02052     <span class="comment">// if bins==0, no change in behavior, exactly 1 point/bin</span>
<a name="l02053"></a>02053     <span class="comment">// min 1 point per bin (nz)</span>
<a name="l02054"></a>02054     <span class="comment">// sum(z) in zz</span>
<a name="l02055"></a>02055     <span class="comment">// sum(z^2) in zzx</span>
<a name="l02056"></a>02056     <span class="comment">// count the bins</span>
<a name="l02057"></a>02057     <span class="comment">//     </span>
<a name="l02058"></a>02058     zall.length = x.size();
<a name="l02059"></a>02059     zall.NS = zall.length+1;
<a name="l02060"></a>02060     
<a name="l02061"></a>02061     vector&lt;double&gt; z,w;
<a name="l02062"></a>02062     vector&lt;int&gt; zerohints;
<a name="l02063"></a>02063     vector&lt;double&gt; zeroInbred;
<a name="l02064"></a>02064     <span class="comment">// local hints always null unless assigned otherwise</span>
<a name="l02065"></a>02065     zerohints.assign(zall.length,-1);
<a name="l02066"></a>02066     zeroInbred.assign(zall.length,0);<span class="comment"></span>
<a name="l02067"></a>02067 <span class="comment">    /// store sorted data</span>
<a name="l02068"></a>02068 <span class="comment"></span>    z.resize(zall.length);
<a name="l02069"></a>02069     w.resize(zall.length); <span class="comment">// y values stored in sorted x order</span><span class="comment"></span>
<a name="l02070"></a>02070 <span class="comment">    ///</span>
<a name="l02071"></a>02071 <span class="comment"></span>    zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>.resize(zall.NS);
<a name="l02072"></a>02072     zall.<a class="code" href="classbinned__data.html#a219d5a70962af7bc6a277330371effd7" title="sum of things in each bin">zz</a>.resize(zall.NS);
<a name="l02073"></a>02073     zall.<a class="code" href="classbinned__data.html#a28c7393a6cba0a818139f4c940663d5c" title="sum of squares in each bin">zzx</a>.resize(zall.NS);
<a name="l02074"></a>02074     <span class="comment">// y terms</span>
<a name="l02075"></a>02075     zall.<a class="code" href="classbinned__data.html#ac35086a3cf907ea147acd035c1f6ad4d" title="sum of y values in each bin">zy</a>.resize(zall.NS);
<a name="l02076"></a>02076     zall.<a class="code" href="classbinned__data.html#af76af2898cfd28e3ef8f9722144ef60f" title="sum of squares of y values in each bin">zzy</a>.resize(zall.NS);
<a name="l02077"></a>02077     zall.<a class="code" href="classbinned__data.html#a6d45afa8716fc45698c70b5862bbb118" title="sum of x*y values in each bin">zxy</a>.resize(zall.NS);
<a name="l02078"></a>02078     <span class="comment">// hints</span>
<a name="l02079"></a>02079     zall.<a class="code" href="classbinned__data.html#a60ff501908d011c168b7621678d1dacf" title="penalty*number of hint genotypes not d">nod</a>.resize(zall.NS);
<a name="l02080"></a>02080     zall.<a class="code" href="classbinned__data.html#ae348fc731eaf4b54f411fefd01452503" title="penalty*number of hint genotypes not h">noh</a>.resize(zall.NS);
<a name="l02081"></a>02081     zall.<a class="code" href="classbinned__data.html#aa06f7ef273167f48d09163cfa0bca595" title="penalty*number of hint genotypes not c">noc</a>.resize(zall.NS);
<a name="l02082"></a>02082 
<a name="l02083"></a>02083     vector&lt; pair&lt;int,double&gt; &gt; data;
<a name="l02084"></a>02084 
<a name="l02085"></a>02085     <span class="keywordflow">for</span> (i=0; i&lt;zall.length; i++)
<a name="l02086"></a>02086     {
<a name="l02087"></a>02087         pair&lt;int,double&gt; p;
<a name="l02088"></a>02088         <span class="comment">/*</span>
<a name="l02089"></a>02089 <span class="comment">        if (genohints.size()&gt;0)</span>
<a name="l02090"></a>02090 <span class="comment">            p.first = genohints[i];</span>
<a name="l02091"></a>02091 <span class="comment">        else</span>
<a name="l02092"></a>02092 <span class="comment">            p.first = -1; // no hint here */</span>
<a name="l02093"></a>02093         p.first = i; <span class="comment">// keep index</span>
<a name="l02094"></a>02094         p.second = x[i];
<a name="l02095"></a>02095         data.push_back(p);
<a name="l02096"></a>02096     }
<a name="l02097"></a>02097 
<a name="l02098"></a>02098     <span class="comment">// sort the data keeping the index to refer to others</span>
<a name="l02099"></a>02099     sort(data.begin(),data.end(),<a class="code" href="structsecond__less.html" title="utility for comparing pairs">second_less</a>());
<a name="l02100"></a>02100 
<a name="l02101"></a>02101     <span class="keywordflow">for</span> (i=0; i&lt;zall.length; i++)
<a name="l02102"></a>02102     {
<a name="l02103"></a>02103         z[i] = x[data[i].first];
<a name="l02104"></a>02104         w[i] = y[data[i].first];
<a name="l02105"></a>02105         <span class="keywordflow">if</span> (genohints.size()&gt;0)
<a name="l02106"></a>02106             zerohints[i] = genohints.at(data[i].first);
<a name="l02107"></a>02107         <span class="keywordflow">else</span>
<a name="l02108"></a>02108             zerohints[i] = -1; <span class="comment">//no hint here</span>
<a name="l02109"></a>02109         <span class="keywordflow">if</span> (SubsetInbred.size()&gt;0)
<a name="l02110"></a>02110             zeroInbred[i] = SubsetInbred.at(data[i].first);
<a name="l02111"></a>02111     }
<a name="l02112"></a>02112     zall.ZS = <a class="code" href="snp_8label_8cpp.html#a1a85fb40b35261384899759ca5a83b66" title="assigns the sorted constrast data into &amp;quot;bins&amp;quot; for fast processing">setup_bins</a>(zall,z,w,zerohints, zeroInbred, zall.length,numbins,hok,CP);
<a name="l02113"></a>02113     zall.ec.resize(zall.ZS);
<a name="l02114"></a>02114     zall.ed.resize(zall.ZS);
<a name="l02115"></a>02115     zall.eh.resize(zall.ZS);    
<a name="l02116"></a>02116 }
<a name="l02117"></a>02117 
<a name="l02118"></a>02118 <span class="comment"></span>
<a name="l02119"></a>02119 <span class="comment">/**</span>
<a name="l02120"></a>02120 <span class="comment"> * The master routine: takes data and turns it into genotypes with no seeds needed</span>
<a name="l02121"></a>02121 <span class="comment"> *</span>
<a name="l02122"></a>02122 <span class="comment"> * @param call - output of genotypes per data point</span>
<a name="l02123"></a>02123 <span class="comment"> * @param conf - output of confidences (1-relative probability)</span>
<a name="l02124"></a>02124 <span class="comment"> * @param x    - contrast values per data point</span>
<a name="l02125"></a>02125 <span class="comment"> * @param y     - size values per data point - used as QC check for copy number</span>
<a name="l02126"></a>02126 <span class="comment"> * @param genohints - any reference genotypes you have lying around as hints</span>
<a name="l02127"></a>02127 <span class="comment"> * @param sp - snp parameters: prior information, algorithm parameters, etc.</span>
<a name="l02128"></a>02128 <span class="comment"> */</span>
<a name="l02129"></a><a class="code" href="snp_8label_8h.html#a981b49e12e6a8c3d84148ab1fd75cfc2">02129</a> <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a1050ceb8ac15c0c39cffa4dbf76709b3" title="The master routine: takes data and turns it into genotypes with no seeds needed.">bayes_label</a>(vector&lt;affx::GType&gt; &amp;call, 
<a name="l02130"></a>02130                  vector&lt;double&gt; &amp;conf, 
<a name="l02131"></a>02131                  vector&lt;double&gt; &amp;x, 
<a name="l02132"></a>02132                  vector&lt;double&gt; &amp;y, 
<a name="l02133"></a>02133                  vector&lt;int&gt; &amp;genohints, 
<a name="l02134"></a>02134                  vector&lt;double&gt; &amp;SubsetInbred, 
<a name="l02135"></a>02135                  <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp)
<a name="l02136"></a>02136 {
<a name="l02137"></a>02137   vector&lt;double&gt; ted,teh,tec;
<a name="l02138"></a>02138   <a class="code" href="snp_8label_8cpp.html#a1050ceb8ac15c0c39cffa4dbf76709b3" title="The master routine: takes data and turns it into genotypes with no seeds needed.">bayes_label</a>(call, conf, x, y, genohints, SubsetInbred, sp, ted, teh, tec, <span class="keyword">false</span>);
<a name="l02139"></a>02139 }
<a name="l02140"></a>02140 <span class="keywordtype">void</span> <a class="code" href="snp_8label_8cpp.html#a1050ceb8ac15c0c39cffa4dbf76709b3" title="The master routine: takes data and turns it into genotypes with no seeds needed.">bayes_label</a>(vector&lt;affx::GType&gt; &amp;call, 
<a name="l02141"></a>02141                  vector&lt;double&gt; &amp;conf, 
<a name="l02142"></a>02142                  vector&lt;double&gt; &amp;x, 
<a name="l02143"></a>02143                  vector&lt;double&gt; &amp;y, 
<a name="l02144"></a>02144                  vector&lt;int&gt; &amp;genohints, 
<a name="l02145"></a>02145                  vector&lt;double&gt; &amp;SubsetInbred, 
<a name="l02146"></a>02146                  <a class="code" href="classsnp__param.html" title="this holds all the parameters for a snp, including alg choices">snp_param</a> &amp;sp, 
<a name="l02147"></a>02147                  vector&lt;double&gt; &amp;ted,
<a name="l02148"></a>02148                  vector&lt;double&gt; &amp;teh,
<a name="l02149"></a>02149                  vector&lt;double&gt; &amp;tec, 
<a name="l02150"></a>02150                  <span class="keywordtype">bool</span> store_probabilities)
<a name="l02151"></a>02151 {
<a name="l02152"></a>02152     <a class="code" href="classbinned__data.html" title="this class summarizes binned data">binned_data</a> zall;
<a name="l02153"></a>02153 
<a name="l02154"></a>02154     <span class="comment">// set up posterior in case no processing of data (&quot;single sample mode&quot;)</span>
<a name="l02155"></a>02155     sp.<a class="code" href="classsnp__param.html#a0f49b07efe5f9072c622f9813e3e1ed0" title="transfer prior directly to posterior">priortoposterior</a>();
<a name="l02156"></a>02156     <span class="comment">// weaken prior if cluster shifting expected</span>
<a name="l02157"></a>02157     <a class="code" href="snp_8label_8cpp.html#a8f4d65cf5118d9aad3b3c4e8c40e613f" title="tunes up the prior to allow for cluster shifts">apply_wobble</a>(sp);
<a name="l02158"></a>02158 
<a name="l02159"></a>02159     <span class="comment">// if we&#39;re not running single-sample mode</span>
<a name="l02160"></a>02160     <span class="comment">// we want to fit the binned data</span>
<a name="l02161"></a>02161     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ac6c59acd942ff43d8aa6baa17ad6364b" title="do we use raw labels or posterior distributions?">callmethod</a>&lt;2)
<a name="l02162"></a>02162     {
<a name="l02163"></a>02163         <span class="comment">// format the data &amp; hints into bins</span>
<a name="l02164"></a>02164         <a class="code" href="snp_8label_8cpp.html#a2979dac75c6e93d694bae23f061b1be8" title="Set up the binned data from the regular data compute running sums, allocate sorted bins...">initialize_bins</a>(zall, x, y, genohints, SubsetInbred, sp.<a class="code" href="classsnp__param.html#ab106a36c2ad59a00fdad233b5577d0b6" title="use quick method">bins</a>,sp.<a class="code" href="classsnp__param.html#afc17331895d50682f60ee3912e733c8c" title="allow hints to be flipped in genotype 0,1,2&amp;lt;-&amp;gt;2,1,0">hok</a>,sp.<a class="code" href="classsnp__param.html#a5bb07d912ac1deddb4be58a2e35f68cf" title="penalty for contradicting a hint">contradictionpenalty</a>);
<a name="l02165"></a>02165 
<a name="l02166"></a>02166         <span class="comment">// take the setup-bins, integrate over all possible labelings</span>
<a name="l02167"></a>02167         <span class="comment">// computing the relative likelihood of the data    </span>
<a name="l02168"></a>02168         <a class="code" href="snp_8label_8cpp.html#af8085a9844832cf7a94ad62a95704b67" title="This is the &amp;quot;labeling&amp;quot; routine that tries out all plausible genotypes for best likelihood B...">integratebrlmmoverlabelings</a>(zall.ec,zall.eh,zall.ed,zall.ZS,zall,sp);
<a name="l02169"></a>02169 
<a name="l02170"></a>02170         <span class="comment">// take the estimated mixture fraction of labels</span>
<a name="l02171"></a>02171         <span class="comment">// turn into approximate &quot;posterior&quot; distribution </span>
<a name="l02172"></a>02172         <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a5137d727c65f534313add6d70f3be81b" title="do I handle multiple cluster dimensions">clustertype</a>==1)
<a name="l02173"></a>02173             <a class="code" href="snp_8label_8cpp.html#aa6c1f401c915be4f6db927f0b0b627c8" title="convert binned label relative probabilities into cluster parameters">labelstoposterior</a>(zall.ec,zall.eh,zall.ed,zall.ZS,zall,sp);
<a name="l02174"></a>02174         <span class="keywordflow">else</span>
<a name="l02175"></a>02175             labels_two_posterior(zall.ec,zall.eh,zall.ed,zall.ZS,zall,sp);
<a name="l02176"></a>02176     }
<a name="l02177"></a>02177     <span class="comment">// make calls based on mean/variance/confidence</span>
<a name="l02178"></a>02178     <span class="comment">// x &gt;unsorted&lt; matrix of calls</span>
<a name="l02179"></a>02179     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ac6c59acd942ff43d8aa6baa17ad6364b" title="do we use raw labels or posterior distributions?">callmethod</a>&gt;0)
<a name="l02180"></a>02180     {
<a name="l02181"></a>02181         <span class="keywordflow">if</span> (store_probabilities) {
<a name="l02182"></a>02182             ted.resize(x.size());
<a name="l02183"></a>02183             teh.resize(x.size());
<a name="l02184"></a>02184             tec.resize(x.size());
<a name="l02185"></a>02185         }
<a name="l02186"></a>02186         <span class="comment">// make calls only by posterior distribution</span>
<a name="l02187"></a>02187         <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a5137d727c65f534313add6d70f3be81b" title="do I handle multiple cluster dimensions">clustertype</a>==1)
<a name="l02188"></a>02188             <a class="code" href="snp_8label_8cpp.html#a86cb22d116534c54810fb8ea68b1fb47" title="Given cluster parameters for each genotype, call data as genotypes.">make_calls</a>(call,conf,x,x.size(),sp.posterior.<a class="code" href="classsnp__distribution.html#a7325d44177912f9851253b1af5893ee3" title="aa allele cluster">aa</a>, sp.posterior.<a class="code" href="classsnp__distribution.html#a1cabc31d59fa37b51256698c5cae3329" title="ab genotype cluster">ab</a>, sp.posterior.<a class="code" href="classsnp__distribution.html#a65361a16c9c09a1870492f441ab9e299" title="bb genotype cluster">bb</a>, sp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a>,sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a>,ted,teh,tec,store_probabilities);
<a name="l02189"></a>02189         <span class="keywordflow">else</span>
<a name="l02190"></a>02190             <a class="code" href="snp_8label_8cpp.html#a64093c4de47f32c8b4821a5b3c37dab3" title="Given cluster parameters for each genotype, call data as genotypes.">make_two_calls</a>(call,conf,x,y,x.size(),sp.posterior, sp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a>,sp.<a class="code" href="classsnp__param.html#ae4dba9978ef175bdadc5a3379289c0ef" title="turn on mixture distribution penalty">mix</a>,sp.<a class="code" href="classsnp__param.html#ad0561d07114f9c66a173dd3b904492a4" title="uniform density to compare weird data points against">ocean</a>,sp.<a class="code" href="classsnp__param.html#a379a6b8e6a44cb1b2f519f74932d8af3" title="increase variance by uncertainty in mean location when making calls">inflatePRA</a>,ted,teh,tec, store_probabilities);
<a name="l02191"></a>02191     }
<a name="l02192"></a>02192     <span class="keywordflow">else</span>
<a name="l02193"></a>02193     {
<a name="l02194"></a>02194         <span class="comment">// output the raw labels matched to data points</span>
<a name="l02195"></a>02195         ted.resize(zall.length);
<a name="l02196"></a>02196         tec.resize(zall.length);
<a name="l02197"></a>02197         teh.resize(zall.length);
<a name="l02198"></a>02198         <span class="comment">// expand data from bins into 1-1 data (still sorted)</span>
<a name="l02199"></a>02199         <a class="code" href="snp_8label_8cpp.html#a0290a1a25d31fc26c81588daca77a8f2" title="take probabilities of bins and turn them into probabilities of individual points">expand_data</a>(ted,zall.length,zall.ed,zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>,zall.ZS);
<a name="l02200"></a>02200         <a class="code" href="snp_8label_8cpp.html#a0290a1a25d31fc26c81588daca77a8f2" title="take probabilities of bins and turn them into probabilities of individual points">expand_data</a>(tec,zall.length,zall.ec,zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>,zall.ZS);
<a name="l02201"></a>02201         <a class="code" href="snp_8label_8cpp.html#a0290a1a25d31fc26c81588daca77a8f2" title="take probabilities of bins and turn them into probabilities of individual points">expand_data</a>(teh,zall.length,zall.eh,zall.<a class="code" href="classbinned__data.html#a9fb44065f655f9420c69c91cca1072c2" title="number of things in each bin">nz</a>,zall.ZS);
<a name="l02202"></a>02202         <span class="comment">// make calls by expanded data (unsorting)</span>
<a name="l02203"></a>02203         <a class="code" href="snp_8label_8cpp.html#a10bb8cda395efca9a0885ab042920921" title="make calls using the final labelings - primarily a matter of unsorting the contrast values...">make_label_calls</a>(call,conf,x,x.size(),ted,tec,teh,sp.<a class="code" href="classsnp__param.html#a5e4855e2b4b0dbd944e007f71f81a362" title="only one copy, for example, XY snps">copynumber</a>);
<a name="l02204"></a>02204         <span class="keywordflow">if</span> (!store_probabilities) {
<a name="l02205"></a>02205             ted.clear();
<a name="l02206"></a>02206             teh.clear();
<a name="l02207"></a>02207             tec.clear();
<a name="l02208"></a>02208         }
<a name="l02209"></a>02209     }
<a name="l02210"></a>02210     <span class="comment">// do copy number QC test</span>
<a name="l02211"></a>02211     <span class="comment">// only if not running in single-sample mode</span>
<a name="l02212"></a>02212     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1095b75124fb8658e9fafec4140d854d" title="how to handle copy number outlier detection">copytype</a>==0)
<a name="l02213"></a>02213     {
<a name="l02214"></a>02214         <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a>&gt;0 &amp;&amp; sp.<a class="code" href="classsnp__param.html#ac6c59acd942ff43d8aa6baa17ad6364b" title="do we use raw labels or posterior distributions?">callmethod</a>&lt;2)
<a name="l02215"></a>02215         {
<a name="l02216"></a>02216             qc_modify_confidences(call,conf,y,sp.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a>);
<a name="l02217"></a>02217         }
<a name="l02218"></a>02218     }
<a name="l02219"></a>02219     <span class="comment">// new style copyqc</span>
<a name="l02220"></a>02220     <span class="comment">// uses prior information if single-sample mode to modify confidence</span>
<a name="l02221"></a>02221     <span class="comment">// Note possible low confidence for unobserved clusters!!!</span>
<a name="l02222"></a>02222     <span class="comment">// must have 2-dimensional clusters stored/allowed</span>
<a name="l02223"></a>02223     <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a1095b75124fb8658e9fafec4140d854d" title="how to handle copy number outlier detection">copytype</a>==1)
<a name="l02224"></a>02224     {
<a name="l02225"></a>02225         <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a>&gt;0)
<a name="l02226"></a>02226         {
<a name="l02227"></a>02227             <span class="comment">// do not compute posterior if single-sample mode</span>
<a name="l02228"></a>02228             <span class="keywordflow">if</span> (sp.<a class="code" href="classsnp__param.html#ac6c59acd942ff43d8aa6baa17ad6364b" title="do we use raw labels or posterior distributions?">callmethod</a>&lt;2)
<a name="l02229"></a>02229                 qc_compute_copy_posterior(call,y,sp.posterior);
<a name="l02230"></a>02230             <span class="comment">// modify confidences if either mode</span>
<a name="l02231"></a>02231             qc_posterior_modify_confidences(call,conf,y,sp.<a class="code" href="classsnp__param.html#a46bef7e8276bb252b9787ebaf8f90f9b" title="check for possible outlier &amp;quot;size&amp;quot; values">copyqc</a>,sp.posterior);
<a name="l02232"></a>02232         }
<a name="l02233"></a>02233     }
<a name="l02234"></a>02234 }    
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:01 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
