<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: chipstream/apt-summary-genotype/SummaryGenotypeEngine.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_fa2ccd765111060c049013d9614ef4eb.html">chipstream</a>      </li>
      <li><a class="el" href="dir_ce02c23fc41304c673cecc0931b4979a.html">apt-summary-genotype</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>SummaryGenotypeEngine.h</h1>  </div>
</div>
<div class="contents">
<a href="SummaryGenotypeEngine_8h.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2009 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   SummaryGenotypeEngine.h</span>
<a name="l00024"></a>00024 <span class="comment"> * @author Luis Jevons</span>
<a name="l00025"></a>00025 <span class="comment"> * @date   Wed April 17 16:01:09 2009</span>
<a name="l00026"></a>00026 <span class="comment"> * </span>
<a name="l00027"></a>00027 <span class="comment"> * @brief Core routines for summary-genotype binaries. By separating the</span>
<a name="l00028"></a>00028 <span class="comment"> * command line parsing form the computation we allow a GUI application to share</span>
<a name="l00029"></a>00029 <span class="comment"> * the core computation once setting up the option class. </span>
<a name="l00030"></a>00030 <span class="comment"> */</span>
<a name="l00031"></a>00031 <span class="preprocessor">#ifndef _SUMMARYGENOTYPEENGINE_H_</span>
<a name="l00032"></a>00032 <span class="preprocessor"></span><span class="preprocessor">#define _SUMMARYGENOTYPEENGINE_H_</span>
<a name="l00033"></a>00033 <span class="preprocessor"></span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="TsvFile_8h.html" title="Headers for the TsvFile classes.">file/TsvFile/TsvFile.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="BaseEngine_8h.html" title="base engine base class">util/BaseEngine.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;file5/File5_Tsv.h&quot;</span>
<a name="l00037"></a>00037 <span class="comment">//</span>
<a name="l00038"></a>00038 <span class="preprocessor">#include &lt;cstring&gt;</span>
<a name="l00039"></a>00039 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00040"></a>00040 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00041"></a>00041 <span class="comment">//</span>
<a name="l00042"></a>00042 <span class="comment"></span>
<a name="l00043"></a>00043 <span class="comment">/**</span>
<a name="l00044"></a>00044 <span class="comment"> * Class to wrap up the complexity of the different possible input</span>
<a name="l00045"></a>00045 <span class="comment"> * files to SummaryGenotypeEngine</span>
<a name="l00046"></a>00046 <span class="comment"> */</span>
<a name="l00047"></a><a class="code" href="classSgIoHelper.html">00047</a> <span class="keyword">class </span><a class="code" href="classSgIoHelper.html" title="Class to wrap up the complexity of the different possible input files to SummaryGenotypeEngine.">SgIoHelper</a> {
<a name="l00048"></a>00048   
<a name="l00049"></a>00049 <span class="keyword">public</span>: 
<a name="l00050"></a>00050   <span class="comment"></span>
<a name="l00051"></a>00051 <span class="comment">  /** </span>
<a name="l00052"></a>00052 <span class="comment">   * Constructor. Take a number of files and figure out which one(s) has</span>
<a name="l00053"></a>00053 <span class="comment">   * been set to read from.</span>
<a name="l00054"></a>00054 <span class="comment">   * </span>
<a name="l00055"></a>00055 <span class="comment">   * @param summariesFile   Text summary file with results interleved.</span>
<a name="l00056"></a>00056 <span class="comment">   * @param summariesFileA  Text summary file for A alleles</span>
<a name="l00057"></a>00057 <span class="comment">   * @param summariesFileB  Text summary file for B alleles</span>
<a name="l00058"></a>00058 <span class="comment">   * @param file5           File5 version of the summary file above</span>
<a name="l00059"></a>00059 <span class="comment">   * @param file5Group      Which group the summary file5 is in with name &quot;summaries&quot;</span>
<a name="l00060"></a>00060 <span class="comment">   */</span>
<a name="l00061"></a>00061   <a class="code" href="classSgIoHelper.html#ada8461b373fce9553e01f5f946dfdd54" title="Constructor.">SgIoHelper</a>(<span class="keyword">const</span> std::string &amp;summariesFile,
<a name="l00062"></a>00062            <span class="keyword">const</span> std::string &amp;summariesFileA, <span class="keyword">const</span> std::string &amp;summariesFileB,
<a name="l00063"></a>00063            <span class="keyword">const</span> std::string &amp;file5,
<a name="l00064"></a>00064            <span class="keyword">const</span> std::string &amp;file5Group);  
<a name="l00065"></a>00065 
<a name="l00066"></a>00066 <span class="comment"></span>
<a name="l00067"></a>00067 <span class="comment">  /** Destructor */</span>
<a name="l00068"></a>00068   <a class="code" href="classSgIoHelper.html#a5d5c5282731c23dd363325f764bbd9ad" title="Destructor.">~SgIoHelper</a>();
<a name="l00069"></a>00069 <span class="comment"></span>
<a name="l00070"></a>00070 <span class="comment">  /** </span>
<a name="l00071"></a>00071 <span class="comment">   * Read in a single line from out summaries file and fill label and values.</span>
<a name="l00072"></a>00072 <span class="comment">   * </span>
<a name="l00073"></a>00073 <span class="comment">   * @param label      Probesetname to be filled in</span>
<a name="l00074"></a>00074 <span class="comment">   * @param vals       Values to be filled in</span>
<a name="l00075"></a>00075 <span class="comment">   * @param summaries  File to read next row from</span>
<a name="l00076"></a>00076 <span class="comment">   * @param colNames   Names of the column in summaries file</span>
<a name="l00077"></a>00077 <span class="comment">   */</span>
<a name="l00078"></a>00078   <span class="keywordtype">void</span> <a class="code" href="classSgIoHelper.html#a9ba39a7cfde128d845c2fdad58631be8" title="Read in a single line from out summaries file and fill label and values.">getLine</a>(std::string &amp;label, std::vector&lt;double&gt; &amp;vals, 
<a name="l00079"></a>00079                <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> &amp;summaries, std::vector&lt;std::string&gt; &amp;colNames);
<a name="l00080"></a>00080 <span class="comment"></span>
<a name="l00081"></a>00081 <span class="comment">  /** </span>
<a name="l00082"></a>00082 <span class="comment">   * Read in a single line from out summaries file5 and fill label and values.</span>
<a name="l00083"></a>00083 <span class="comment">   * </span>
<a name="l00084"></a>00084 <span class="comment">   * @param label      Probesetname to be filled in</span>
<a name="l00085"></a>00085 <span class="comment">   * @param vals       Values to be filled in</span>
<a name="l00086"></a>00086 <span class="comment">   * @param summaries  File5 to read next row from</span>
<a name="l00087"></a>00087 <span class="comment">   * @param colNames   Names of the column in summaries file</span>
<a name="l00088"></a>00088 <span class="comment">   */</span>
<a name="l00089"></a>00089   <span class="keywordtype">void</span> <a class="code" href="classSgIoHelper.html#a47053a0e6dbbe864186cd720d597ec79" title="Read in a single line from out summaries file5 and fill label and values.">getFile5Line</a>(std::string &amp;label, std::vector&lt;double&gt; &amp;vals, 
<a name="l00090"></a>00090                     <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a> *summaries, std::vector&lt;std::string&gt; &amp;colNames);
<a name="l00091"></a>00091 <span class="comment"></span>
<a name="l00092"></a>00092 <span class="comment">  /** </span>
<a name="l00093"></a>00093 <span class="comment">   * Read the next snp marker from the single interleved file5 summary file</span>
<a name="l00094"></a>00094 <span class="comment">   *  </span>
<a name="l00095"></a>00095 <span class="comment">   * @param colNames Names of the columns in summary file.</span>
<a name="l00096"></a>00096 <span class="comment">   * @param aVals Summary values for the A allele</span>
<a name="l00097"></a>00097 <span class="comment">   * @param bVals Summary values for the B allele</span>
<a name="l00098"></a>00098 <span class="comment">   * @param probesetName Name of probeset</span>
<a name="l00099"></a>00099 <span class="comment">   * </span>
<a name="l00100"></a>00100 <span class="comment">   * @return true if successful, false otherwise</span>
<a name="l00101"></a>00101 <span class="comment">   */</span>
<a name="l00102"></a>00102   <span class="keywordtype">bool</span> <a class="code" href="classSgIoHelper.html#aa6f292bd628e59ca57868943a0a3ca62" title="Read the next snp marker from the single interleved file5 summary file.">getNextMarkerTsvSingle</a>(std::vector&lt;std::string&gt; &amp;colNames, std::vector&lt;double&gt; &amp;aVals, 
<a name="l00103"></a>00103                               std::vector&lt;double&gt; &amp;bVals, std::string &amp;probesetName);
<a name="l00104"></a>00104 <span class="comment"></span>
<a name="l00105"></a>00105 <span class="comment">  /** </span>
<a name="l00106"></a>00106 <span class="comment">   * Read the next snp marker from the file5 </span>
<a name="l00107"></a>00107 <span class="comment">   *  </span>
<a name="l00108"></a>00108 <span class="comment">   * @param colNames Names of the columns in summary file.</span>
<a name="l00109"></a>00109 <span class="comment">   * @param aVals Summary values for the A allele</span>
<a name="l00110"></a>00110 <span class="comment">   * @param bVals Summary values for the B allele</span>
<a name="l00111"></a>00111 <span class="comment">   * @param probesetName Name of probeset</span>
<a name="l00112"></a>00112 <span class="comment">   * </span>
<a name="l00113"></a>00113 <span class="comment">   * @return true if successful, false otherwise</span>
<a name="l00114"></a>00114 <span class="comment">   */</span>
<a name="l00115"></a>00115   <span class="keywordtype">bool</span> <a class="code" href="classSgIoHelper.html#a48083af830fe8c536bcb37838bd4e52b" title="Read the next snp marker from the file5.">getNextMarkerFile5</a>(std::vector&lt;std::string&gt; &amp;colNames, std::vector&lt;double&gt; &amp;aVals, 
<a name="l00116"></a>00116                           std::vector&lt;double&gt; &amp;bVals, std::string &amp;probesetName);<span class="comment"></span>
<a name="l00117"></a>00117 <span class="comment">  /** </span>
<a name="l00118"></a>00118 <span class="comment">   * Read the next snp marker from the paired summary A &amp; B files</span>
<a name="l00119"></a>00119 <span class="comment">   *  </span>
<a name="l00120"></a>00120 <span class="comment">   * @param colNames Names of the columns in summary file.</span>
<a name="l00121"></a>00121 <span class="comment">   * @param aVals Summary values for the A allele</span>
<a name="l00122"></a>00122 <span class="comment">   * @param bVals Summary values for the B allele</span>
<a name="l00123"></a>00123 <span class="comment">   * @param probesetName Name of probeset</span>
<a name="l00124"></a>00124 <span class="comment">   * </span>
<a name="l00125"></a>00125 <span class="comment">   * @return true if successful, false otherwise</span>
<a name="l00126"></a>00126 <span class="comment">   */</span>
<a name="l00127"></a>00127   <span class="keywordtype">bool</span> <a class="code" href="classSgIoHelper.html#a28cbaf1a92be7f3fb6ddfa90c5551e69" title="Read the next snp marker from the paired summary A &amp;amp; B files.">getNextMarkerTwoTsv</a>(std::vector&lt;std::string&gt; &amp;colNames, std::vector&lt;double&gt; &amp;aVals, 
<a name="l00128"></a>00128                            std::vector&lt;double&gt; &amp;bVals, std::string &amp;probesetName);<span class="comment"></span>
<a name="l00129"></a>00129 <span class="comment">  /** </span>
<a name="l00130"></a>00130 <span class="comment">   * Read the next snp marker.</span>
<a name="l00131"></a>00131 <span class="comment">   *  </span>
<a name="l00132"></a>00132 <span class="comment">   * @param colNames Names of the columns in summary file.</span>
<a name="l00133"></a>00133 <span class="comment">   * @param aVals Summary values for the A allele</span>
<a name="l00134"></a>00134 <span class="comment">   * @param bVals Summary values for the B allele</span>
<a name="l00135"></a>00135 <span class="comment">   * @param probesetName Name of probeset</span>
<a name="l00136"></a>00136 <span class="comment">   * </span>
<a name="l00137"></a>00137 <span class="comment">   * @return true if successful, false otherwise</span>
<a name="l00138"></a>00138 <span class="comment">   */</span>
<a name="l00139"></a>00139   <span class="keywordtype">bool</span> <a class="code" href="classSgIoHelper.html#a6e91bab82eb0d030f0343d8b81c9e948" title="Read the next snp marker.">getNextMarker</a>(std::vector&lt;std::string&gt; &amp;colNames, std::vector&lt;double&gt; &amp;aVals, 
<a name="l00140"></a>00140                      std::vector&lt;double&gt; &amp;bVals, std::string &amp;probesetName);
<a name="l00141"></a>00141 <span class="comment"></span>
<a name="l00142"></a>00142 <span class="comment">  /** </span>
<a name="l00143"></a>00143 <span class="comment">   * Get the names of the columns in the summary files.</span>
<a name="l00144"></a>00144 <span class="comment">   * </span>
<a name="l00145"></a>00145 <span class="comment">   * @return names in order of the columns</span>
<a name="l00146"></a>00146 <span class="comment">   */</span>
<a name="l00147"></a><a class="code" href="classSgIoHelper.html#a27be6a9294919349a554f477caeedc52">00147</a>   std::vector&lt;std::string&gt; <a class="code" href="classSgIoHelper.html#a27be6a9294919349a554f477caeedc52" title="Get the names of the columns in the summary files.">getColNames</a>() {
<a name="l00148"></a>00148     <span class="keywordflow">return</span> m_ColNames;
<a name="l00149"></a>00149   }
<a name="l00150"></a>00150 
<a name="l00151"></a>00151 <span class="keyword">private</span>:
<a name="l00152"></a>00152   <span class="comment"></span>
<a name="l00153"></a>00153 <span class="comment">  /// Column names from the summary files not including the probeset_id column</span>
<a name="l00154"></a>00154 <span class="comment"></span>  std::vector&lt;std::string&gt; m_ColNames;<span class="comment"></span>
<a name="l00155"></a>00155 <span class="comment">  /// All the summaries as a single interleved TSV file</span>
<a name="l00156"></a>00156 <span class="comment"></span>  <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> summaries;<span class="comment"></span>
<a name="l00157"></a>00157 <span class="comment">  /// All the A allele summaries as a TSV File</span>
<a name="l00158"></a>00158 <span class="comment"></span>  <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> summariesA;<span class="comment"></span>
<a name="l00159"></a>00159 <span class="comment">  /// All the B allele summaries as a TSV File</span>
<a name="l00160"></a>00160 <span class="comment"></span>  <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> summariesB;<span class="comment"></span>
<a name="l00161"></a>00161 <span class="comment">  /// Summaries in a file5 format</span>
<a name="l00162"></a>00162 <span class="comment"></span>  <a class="code" href="classaffx_1_1File5__File.html">affx::File5_File</a> f5Summaries;<span class="comment"></span>
<a name="l00163"></a>00163 <span class="comment">  /// Within the file5 format, the group that the intensities are in.</span>
<a name="l00164"></a>00164 <span class="comment"></span>  <a class="code" href="classaffx_1_1File5__Group.html">affx::File5_Group</a> *f5SumGroup;<span class="comment"></span>
<a name="l00165"></a>00165 <span class="comment">  /// The tsv file within the File5 that contains the &quot;summaries&quot;</span>
<a name="l00166"></a>00166 <span class="comment"></span>  <a class="code" href="classaffx_1_1File5__Tsv.html">affx::File5_Tsv</a> *f5SumTsv;
<a name="l00167"></a>00167 };
<a name="l00168"></a>00168 
<a name="l00169"></a><a class="code" href="classSummaryGenotypeEngine.html">00169</a> <span class="keyword">class </span><a class="code" href="classSummaryGenotypeEngine.html">SummaryGenotypeEngine</a> : <span class="keyword">public</span> <a class="code" href="classBaseEngine.html" title="Base class for analysis engines.">BaseEngine</a> {
<a name="l00170"></a>00170 
<a name="l00171"></a>00171 <span class="keyword">public</span>:
<a name="l00172"></a>00172 
<a name="l00173"></a>00173   <span class="keyword">virtual</span> std::string getEngineName() { <span class="keywordflow">return</span> SummaryGenotypeEngine::EngineName(); }
<a name="l00174"></a>00174   <span class="keyword">static</span> <span class="keyword">const</span> std::string EngineName() { <span class="keywordflow">return</span> <span class="stringliteral">&quot;SummaryGenotypeEngine&quot;</span>; }
<a name="l00175"></a>00175 <span class="comment"></span>
<a name="l00176"></a>00176 <span class="comment">  /**</span>
<a name="l00177"></a>00177 <span class="comment">   * Constructor</span>
<a name="l00178"></a>00178 <span class="comment">   */</span>
<a name="l00179"></a>00179   <a class="code" href="classSummaryGenotypeEngine.html#a0aa8c15fb7f3c1fe92a8b04fd245a4b9" title="Constructor.">SummaryGenotypeEngine</a>();
<a name="l00180"></a>00180 <span class="comment"></span>
<a name="l00181"></a>00181 <span class="comment">  /**</span>
<a name="l00182"></a>00182 <span class="comment">   * Destructor</span>
<a name="l00183"></a>00183 <span class="comment">   */</span>
<a name="l00184"></a>00184   <a class="code" href="classSummaryGenotypeEngine.html#ab8b78f7da6b728a1ae5a3f626d8700b3" title="Destructor.">~SummaryGenotypeEngine</a>();
<a name="l00185"></a>00185 <span class="comment"></span>
<a name="l00186"></a>00186 <span class="comment">  /*! A class to register the summary engine. */</span>
<a name="l00187"></a><a class="code" href="classSummaryGenotypeEngine_1_1Reg.html">00187</a>   <span class="keyword">class </span><a class="code" href="classSummaryGenotypeEngine_1_1Reg.html">Reg</a> : <span class="keyword">public</span> <a class="code" href="classEngineReg.html">EngineReg</a>
<a name="l00188"></a>00188   {
<a name="l00189"></a>00189   <span class="keyword">public</span>:<span class="comment"></span>
<a name="l00190"></a>00190 <span class="comment">    /*! Constructor - register the summary engine. */</span>
<a name="l00191"></a><a class="code" href="classSummaryGenotypeEngine_1_1Reg.html#a2b161e8991adf35011d8858cb39b26d4">00191</a>     <a class="code" href="classSummaryGenotypeEngine_1_1Reg.html#a2b161e8991adf35011d8858cb39b26d4">Reg</a>() : <a class="code" href="classEngineReg.html">EngineReg</a>(<a class="code" href="classSummaryGenotypeEngine.html">SummaryGenotypeEngine</a>::EngineName())
<a name="l00192"></a>00192     {
<a name="l00193"></a>00193     }
<a name="l00194"></a>00194 <span class="comment"></span>
<a name="l00195"></a>00195 <span class="comment">    /*! Creates an object.</span>
<a name="l00196"></a>00196 <span class="comment">     * @return The object.</span>
<a name="l00197"></a>00197 <span class="comment">     */</span>
<a name="l00198"></a><a class="code" href="classSummaryGenotypeEngine_1_1Reg.html#ac949fe999a21869d4d8904efcd2aa804">00198</a>     <a class="code" href="classBaseEngine.html" title="Base class for analysis engines.">BaseEngine</a> *<a class="code" href="classSummaryGenotypeEngine_1_1Reg.html#ac949fe999a21869d4d8904efcd2aa804">MakeObject</a>() { <span class="keywordflow">return</span> <span class="keyword">new</span> <a class="code" href="classSummaryGenotypeEngine.html">SummaryGenotypeEngine</a>; }
<a name="l00199"></a>00199   };
<a name="l00200"></a>00200 <span class="comment"></span>
<a name="l00201"></a>00201 <span class="comment">  /*! The one and only registration object. */</span>
<a name="l00202"></a><a class="code" href="classSummaryGenotypeEngine.html#a9be8a8609608dfd04a2bdf21ee8dfcfb">00202</a>   <span class="keyword">static</span> <a class="code" href="classSummaryGenotypeEngine_1_1Reg.html">Reg</a> <a class="code" href="classSummaryGenotypeEngine.html#a9be8a8609608dfd04a2bdf21ee8dfcfb">reg</a>;
<a name="l00203"></a>00203 <span class="comment"></span>
<a name="l00204"></a>00204 <span class="comment">  /*! Converts the type to the summary engine type.</span>
<a name="l00205"></a>00205 <span class="comment">   * @param chip The pointer to the base engine object.</span>
<a name="l00206"></a>00206 <span class="comment">   * @return The summary engine type or NULL if not compatible.</span>
<a name="l00207"></a>00207 <span class="comment">   */</span>
<a name="l00208"></a>00208   <span class="keyword">static</span> <a class="code" href="classSummaryGenotypeEngine.html">SummaryGenotypeEngine</a> * <a class="code" href="classSummaryGenotypeEngine.html#adb579494c2ccfed76de0995f1c28f940">FromBase</a>(<a class="code" href="classBaseEngine.html" title="Base class for analysis engines.">BaseEngine</a> *engine);
<a name="l00209"></a>00209 
<a name="l00210"></a>00210 <span class="keyword">private</span>:
<a name="l00211"></a>00211 
<a name="l00212"></a>00212   <span class="keyword">virtual</span> <span class="keywordtype">void</span> defineOptions();
<a name="l00213"></a>00213   <span class="keyword">virtual</span> <span class="keywordtype">void</span> checkOptionsImp();
<a name="l00214"></a>00214   <span class="keyword">virtual</span> <span class="keywordtype">void</span> defineStates();
<a name="l00215"></a>00215   <span class="keyword">virtual</span> <span class="keywordtype">void</span> runImp();
<a name="l00216"></a>00216 };
<a name="l00217"></a>00217 
<a name="l00218"></a>00218 <span class="preprocessor">#endif</span>
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:50 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
