<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/apt-copynumber-cyto/regression/test-regression-copynumber-cyto-quick.valgrind.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
      <li><a class="el" href="dir_d61a7909814a46bce5ab2f5f0e583661.html">apt-copynumber-cyto</a>      </li>
      <li><a class="el" href="dir_295cb3590393dbdce00cd66c6df31eff.html">regression</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>test-regression-copynumber-cyto-quick.valgrind.cpp</h1>  </div>
</div>
<div class="contents">
<div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2009 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   test-regression-copynumber-cyto-quick.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @brief  Program for doing regression tests on cyto data.</span>
<a name="l00025"></a>00025 <span class="comment"> */</span>
<a name="l00026"></a>00026 
<a name="l00027"></a>00027 <span class="preprocessor">#include &quot;<a class="code" href="Calvin_8h.html" title="This header contains the Calvin class definition.">calvin_files/utils/src/Calvin.h</a>&quot;</span>
<a name="l00028"></a>00028 <span class="preprocessor">#include &quot;file5/File5.h&quot;</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="FsTestDir_8h.html" title="///">util/FsTestDir.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="LogStream_8h.html" title="Output verbose messages to log file as we go along.">util/LogStream.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="Verbose_8h.html" title="Class for doing logging and some command line ui.">util/Verbose.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="comment">//</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &lt;cstring&gt;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00037"></a>00037 <span class="comment">//</span>
<a name="l00038"></a>00038 <span class="keyword">using namespace </span>std;
<a name="l00039"></a>00039 
<a name="l00040"></a>00040 <span class="keywordtype">string</span> valgrindString1 = <span class="stringliteral">&quot;valgrind --leak-check=yes --log-file=&quot;</span>;
<a name="l00041"></a>00041 <span class="keywordtype">string</span> valgrindString2 = <span class="stringliteral">&quot;.valgrind.txt &quot;</span>;
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 
<a name="l00046"></a>00046 <span class="keyword">class </span><a class="code" href="classtest__regression__copynumber__cyto__quick.html">test_regression_copynumber_cyto_quick</a>
<a name="l00047"></a>00047 {
<a name="l00048"></a>00048 <span class="keyword">public</span>:
<a name="l00049"></a>00049     <span class="keywordtype">int</span> numPassed;
<a name="l00050"></a>00050     <span class="keywordtype">int</span> numFailed;
<a name="l00051"></a>00051     std::string testDir;
<a name="l00052"></a>00052     std::set&lt;std::string&gt; setIgnore;
<a name="l00053"></a>00053     std::map&lt;std::string, float&gt; mapEpsilon;
<a name="l00054"></a>00054     std::vector&lt;std::string&gt; vCytogenetics_Array_SixFiles;
<a name="l00055"></a>00055 
<a name="l00056"></a>00056     <a class="code" href="classtest__regression__copynumber__cyto__quick.html">test_regression_copynumber_cyto_quick</a>()
<a name="l00057"></a>00057     {
<a name="l00058"></a>00058         numPassed = 0;
<a name="l00059"></a>00059         numFailed = 0;
<a name="l00060"></a>00060 
<a name="l00061"></a>00061         <span class="comment">// Header paramters to ignore</span>
<a name="l00062"></a>00062         setIgnore.insert(<span class="stringliteral">&quot;FileCreationTime&quot;</span>);
<a name="l00063"></a>00063         setIgnore.insert(<span class="stringliteral">&quot;FileIdentifier&quot;</span>);
<a name="l00064"></a>00064         setIgnore.insert(<span class="stringliteral">&quot;program-version&quot;</span>);
<a name="l00065"></a>00065         setIgnore.insert(<span class="stringliteral">&quot;create_date&quot;</span>);
<a name="l00066"></a>00066         setIgnore.insert(<span class="stringliteral">&quot;create-date&quot;</span>);
<a name="l00067"></a>00067         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-verbose&quot;</span>);
<a name="l00068"></a>00068         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-exec-guid&quot;</span>);
<a name="l00069"></a>00069         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-program-cvs-id&quot;</span>);
<a name="l00070"></a>00070         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-version-to-report&quot;</span>);
<a name="l00071"></a>00071         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-command-line&quot;</span>);
<a name="l00072"></a>00072         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-mem-usage&quot;</span>);
<a name="l00073"></a>00073         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-run-probeset-genotype&quot;</span>);
<a name="l00074"></a>00074         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-cels&quot;</span>);
<a name="l00075"></a>00075         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-out-dir&quot;</span>);
<a name="l00076"></a>00076         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-state-time-start&quot;</span>);
<a name="l00077"></a>00077         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-state-free-mem-at-start&quot;</span>);
<a name="l00078"></a>00078         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-temp-dir&quot;</span>);
<a name="l00079"></a>00079 
<a name="l00080"></a>00080         <span class="comment">// Data Set columns to ignore (Specify as Group.Set.Column)</span>
<a name="l00081"></a>00081 <span class="comment">//        setIgnore.insert(&quot;AlgorithmData.MarkerABSignal.SCAR&quot;);</span>
<a name="l00082"></a>00082 
<a name="l00083"></a>00083         <span class="comment">// Data Set column epsilon override (Specify as Group.Set.Column)</span>
<a name="l00084"></a>00084 <span class="comment">//        mapEpsilon.insert(std::pair&lt;std::string, float&gt;(&quot;AlgorithmData.MarkerABSignal.SCAR&quot;, (float)0.0005));</span>
<a name="l00085"></a>00085         mapEpsilon.insert(std::pair&lt;string, float&gt;(<span class="stringliteral">&quot;Segments.CN.MeanMarkerDistance&quot;</span>, 376.0f));
<a name="l00086"></a>00086         <span class="comment">// File tags to check for equivalency</span>
<a name="l00087"></a>00087         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-As_NA18547_A08_01_NN_20081218&quot;</span>);
<a name="l00088"></a>00088         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-As_NA18603_B03_01_NN_20081218&quot;</span>);
<a name="l00089"></a>00089         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-As_NA18971_C02_01_NN_20081218&quot;</span>);
<a name="l00090"></a>00090         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-Cs_NA10857_A10_01_NN_20081218&quot;</span>);
<a name="l00091"></a>00091         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-Cs_NA12003_D04_01_NN_20081218&quot;</span>);
<a name="l00092"></a>00092     }
<a name="l00093"></a>00093 
<a name="l00094"></a>00094     <span class="keywordtype">void</span> execute(<span class="keyword">const</span> std::string&amp; str) {
<a name="l00095"></a>00095       std::string cmd = <a class="code" href="classFs.html#af8ce27acb118594a0e073ce6d2d70da4" title="convert only the first component of the command to unc. / return the command string.">Fs::convertCommandToUnc</a>(str);
<a name="l00096"></a>00096       <span class="keywordflow">if</span> (system(cmd.c_str()) != 0) {<a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Execution failed for command: &quot;</span> + cmd);}
<a name="l00097"></a>00097     }
<a name="l00098"></a>00098 
<a name="l00099"></a>00099     <span class="keywordtype">void</span> equivalency(<span class="keyword">const</span> std::string&amp; name, <span class="keyword">const</span> std::string&amp; modifier, std::vector&lt;string&gt;&amp; vFileTags)
<a name="l00100"></a>00100     {
<a name="l00101"></a>00101         <span class="keywordtype">bool</span> bPassed = <span class="keyword">true</span>;
<a name="l00102"></a>00102         <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> strFileName1;
<a name="l00103"></a>00103         <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> strFileName2;
<a name="l00104"></a>00104         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;*&quot;</span>);
<a name="l00105"></a>00105         <span class="keywordflow">for</span>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; (i &lt; vFileTags.size()); i++)
<a name="l00106"></a>00106         {
<a name="l00107"></a>00107 
<a name="l00108"></a>00108             strFileName1 = <span class="stringliteral">&quot;../../../regression-data/data/idata/copynumber-cyto/&quot;</span> + name + <span class="stringliteral">&quot;/&quot;</span> + vFileTags[i] + <span class="stringliteral">&quot;.&quot;</span> + modifier + <span class="stringliteral">&quot;.cychp&quot;</span>;
<a name="l00109"></a>00109             strFileName2 = testDir + <span class="stringliteral">&quot;/&quot;</span> + name + <span class="stringliteral">&quot;/&quot;</span> + vFileTags[i] + <span class="stringliteral">&quot;.&quot;</span> + modifier + <span class="stringliteral">&quot;.cychp&quot;</span>;
<a name="l00110"></a>00110             <span class="keywordflow">if</span> (!Calvin::equivalent(strFileName1, strFileName2, setIgnore, mapEpsilon, 0.0001, 0.9999989, <span class="keyword">false</span>)) {bPassed = <span class="keyword">false</span>;}
<a name="l00111"></a>00111         }
<a name="l00112"></a>00112         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;*&quot;</span>);
<a name="l00113"></a>00113         <span class="keywordflow">if</span> (bPassed) {
<a name="l00114"></a>00114                   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;***Passed*** &quot;</span> + name + <span class="stringliteral">&quot; vs &quot;</span> + modifier);
<a name="l00115"></a>00115                   numPassed++;
<a name="l00116"></a>00116                 }
<a name="l00117"></a>00117                 <span class="keywordflow">else</span> {
<a name="l00118"></a>00118                   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;***Failed*** &quot;</span> + name + <span class="stringliteral">&quot; vs &quot;</span> + modifier);
<a name="l00119"></a>00119                   numFailed++;
<a name="l00120"></a>00120                 }
<a name="l00121"></a>00121     }
<a name="l00122"></a>00122 
<a name="l00123"></a>00123     <span class="keywordtype">void</span> run()
<a name="l00124"></a>00124     {
<a name="l00125"></a>00125         single();
<a name="l00126"></a>00126         batch();
<a name="l00127"></a>00127     }
<a name="l00128"></a>00128 
<a name="l00129"></a>00129     <span class="keywordtype">void</span> single()
<a name="l00130"></a>00130     {
<a name="l00131"></a>00131           <span class="keywordtype">string</span> name = <span class="stringliteral">&quot;qt-single&quot;</span>;
<a name="l00132"></a>00132           <a class="code" href="classFs.html#a1206f6cf8f445461466f224e6cd1bc52" title="///">Fs::rmdirPath</a>(testDir + <span class="stringliteral">&quot;/single/&quot;</span>, <span class="keyword">false</span>);
<a name="l00133"></a>00133 
<a name="l00134"></a>00134                  execute( valgrindString1 + name + valgrindString2 + <span class="stringliteral">&quot; ./apt-copynumber-cyto \</span>
<a name="l00135"></a>00135 <span class="stringliteral">                    -v 3 --cyto2 true \</span>
<a name="l00136"></a>00136 <span class="stringliteral">                    --force true \</span>
<a name="l00137"></a>00137 <span class="stringliteral">                    --text-output true \</span>
<a name="l00138"></a>00138 <span class="stringliteral">                    --cnchp-output false \</span>
<a name="l00139"></a>00139 <span class="stringliteral">                    --cychp-output true \</span>
<a name="l00140"></a>00140 <span class="stringliteral">                    --set-analysis-name ca \</span>
<a name="l00141"></a>00141 <span class="stringliteral">                    --reference-input ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.na28.REF_MODEL \</span>
<a name="l00142"></a>00142 <span class="stringliteral">                    --snp-reference-input-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.na28.snpref.a5 \</span>
<a name="l00143"></a>00143 <span class="stringliteral">                    --spf-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.spf \</span>
<a name="l00144"></a>00144 <span class="stringliteral">                    --chrX-probes ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.chrXprobes \</span>
<a name="l00145"></a>00145 <span class="stringliteral">                    --chrY-probes ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.chrYprobes \</span>
<a name="l00146"></a>00146 <span class="stringliteral">                    --annotation-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.na28.annot.db \</span>
<a name="l00147"></a>00147 <span class="stringliteral">                    --probe-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.probe_tab \</span>
<a name="l00148"></a>00148 <span class="stringliteral">                    --snp-qc-snp-list ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.snplist.txt \</span>
<a name="l00149"></a>00149 <span class="stringliteral">                    --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-As_NA18547_A08_01_NN_20081218.CEL \</span>
<a name="l00150"></a>00150 <span class="stringliteral">                    --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-As_NA18603_B03_01_NN_20081218.CEL \</span>
<a name="l00151"></a>00151 <span class="stringliteral">                    --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-As_NA18971_C02_01_NN_20081218.CEL \</span>
<a name="l00152"></a>00152 <span class="stringliteral">                    --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL \</span>
<a name="l00153"></a>00153 <span class="stringliteral">                    --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL \</span>
<a name="l00154"></a>00154 <span class="stringliteral">                    --out-dir &quot;</span> + testDir + <span class="stringliteral">&quot;/single \</span>
<a name="l00155"></a>00155 <span class="stringliteral">                    --male-gender-ratio-cutoff=1.3 \</span>
<a name="l00156"></a>00156 <span class="stringliteral">                    --female-gender-ratio-cutoff=1.0 \</span>
<a name="l00157"></a>00157 <span class="stringliteral">                    --xx-cutoff=0.8 \</span>
<a name="l00158"></a>00158 <span class="stringliteral">                    --xx-cutoff-high=1.07 \</span>
<a name="l00159"></a>00159 <span class="stringliteral">                    --y-cutoff=0.65 \</span>
<a name="l00160"></a>00160 <span class="stringliteral">                    --analysis log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5 \</span>
<a name="l00161"></a>00161 <span class="stringliteral">                    --analysis kernel-smooth.sigma_span=50 \</span>
<a name="l00162"></a>00162 <span class="stringliteral">                    --analysis cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6 \</span>
<a name="l00163"></a>00163 <span class="stringliteral">                    --analysis cn-cyto2-gender.cutoff=0.5 \</span>
<a name="l00164"></a>00164 <span class="stringliteral">                    --analysis cn-segment \</span>
<a name="l00165"></a>00165 <span class="stringliteral">                    --analysis loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0 \</span>
<a name="l00166"></a>00166 <span class="stringliteral">                    --analysis mosaicism.marker-bandwidth=6000.confidence-window=251 \</span>
<a name="l00167"></a>00167 <span class="stringliteral">                    --analysis allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05 \</span>
<a name="l00168"></a>00168 <span class="stringliteral">                    --local-gc-background-correction-reference-method pdnn-reference-method \</span>
<a name="l00169"></a>00169 <span class="stringliteral">                    --wave-correction-reference-method wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false \</span>
<a name="l00170"></a>00170 <span class="stringliteral">                    --local-gc-background-intensity-adjustment-method pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0 \</span>
<a name="l00171"></a>00171 <span class="stringliteral">                    --image-correction-intensity-adjustment-method high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001 \</span>
<a name="l00172"></a>00172 <span class="stringliteral">                    --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3 --reference-chromosome=22 \</span>
<a name="l00173"></a>00173 <span class="stringliteral">            &quot;</span>);
<a name="l00174"></a>00174         equivalency(<span class="stringliteral">&quot;single&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>, vCytogenetics_Array_SixFiles);
<a name="l00175"></a>00175     }
<a name="l00176"></a>00176 
<a name="l00177"></a>00177     <span class="keywordtype">void</span> batch()
<a name="l00178"></a>00178     {
<a name="l00179"></a>00179           <span class="keywordtype">string</span> name1 = <span class="stringliteral">&quot;qt-batch-part1&quot;</span>;
<a name="l00180"></a>00180           <span class="keywordtype">string</span> name2 = <span class="stringliteral">&quot;qt-batch-part2&quot;</span>;
<a name="l00181"></a>00181           <a class="code" href="classFs.html#a1206f6cf8f445461466f224e6cd1bc52" title="///">Fs::rmdirPath</a>(testDir + <span class="stringliteral">&quot;/batch/&quot;</span>, <span class="keyword">false</span>);
<a name="l00182"></a>00182 
<a name="l00183"></a>00183                  execute( valgrindString1 + name1 + valgrindString2 + <span class="stringliteral">&quot; ./apt-copynumber-cyto \</span>
<a name="l00184"></a>00184 <span class="stringliteral">                    -v 3 --cyto2 true \</span>
<a name="l00185"></a>00185 <span class="stringliteral">                    --force true \</span>
<a name="l00186"></a>00186 <span class="stringliteral">                    --text-output true \</span>
<a name="l00187"></a>00187 <span class="stringliteral">                    --cnchp-output false \</span>
<a name="l00188"></a>00188 <span class="stringliteral">                    --cychp-output true \</span>
<a name="l00189"></a>00189 <span class="stringliteral">                    --set-analysis-name ca \</span>
<a name="l00190"></a>00190 <span class="stringliteral">                    --reference-output &quot;</span> + testDir + <span class="stringliteral">&quot;/batch/Cytogenetics_Array.10_cels.REF_MODEL \</span>
<a name="l00191"></a>00191 <span class="stringliteral">                    --snp-reference-input-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.na28.snpref.a5 \</span>
<a name="l00192"></a>00192 <span class="stringliteral">                    --snp-reference-output-file &quot;</span> + testDir + <span class="stringliteral">&quot;/batch/Cytogenetics_Array.10_cels.snpref.a5 \</span>
<a name="l00193"></a>00193 <span class="stringliteral">                    --spf-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.spf \</span>
<a name="l00194"></a>00194 <span class="stringliteral">                    --chrX-probes ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.chrXprobes \</span>
<a name="l00195"></a>00195 <span class="stringliteral">                    --chrY-probes ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.chrYprobes \</span>
<a name="l00196"></a>00196 <span class="stringliteral">                    --annotation-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.na28.annot.db \</span>
<a name="l00197"></a>00197 <span class="stringliteral">                    --probe-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.probe_tab \</span>
<a name="l00198"></a>00198 <span class="stringliteral">                    --snp-qc-snp-list ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.snplist.txt \</span>
<a name="l00199"></a>00199 <span class="stringliteral">                    --gender-override-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/HapMap-LOHReference-Full-Genders.txt \</span>
<a name="l00200"></a>00200 <span class="stringliteral">                    --genotype-call-override-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/HapMapCalls_LOHReferenceSet-FullArray.txt \</span>
<a name="l00201"></a>00201 <span class="stringliteral">                    --cel-files ../../../regression-data/data/idata/cel/Cytogenetics_Array/CELFileList_5F5M.txt \</span>
<a name="l00202"></a>00202 <span class="stringliteral">                    --out-dir &quot;</span> + testDir + <span class="stringliteral">&quot;/batch \</span>
<a name="l00203"></a>00203 <span class="stringliteral">                    --male-gender-ratio-cutoff=1.3 \</span>
<a name="l00204"></a>00204 <span class="stringliteral">                    --female-gender-ratio-cutoff=1.0 \</span>
<a name="l00205"></a>00205 <span class="stringliteral">                    --xx-cutoff=0.8 \</span>
<a name="l00206"></a>00206 <span class="stringliteral">                    --xx-cutoff-high=1.07 \</span>
<a name="l00207"></a>00207 <span class="stringliteral">                    --y-cutoff=0.65 \</span>
<a name="l00208"></a>00208 <span class="stringliteral">                    --analysis log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5 \</span>
<a name="l00209"></a>00209 <span class="stringliteral">                    --analysis kernel-smooth.sigma_span=50 \</span>
<a name="l00210"></a>00210 <span class="stringliteral">                    --analysis cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6 \</span>
<a name="l00211"></a>00211 <span class="stringliteral">                    --analysis cn-cyto2-gender.cutoff=0.5 \</span>
<a name="l00212"></a>00212 <span class="stringliteral">                    --analysis cn-segment \</span>
<a name="l00213"></a>00213 <span class="stringliteral">                    --analysis loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0 \</span>
<a name="l00214"></a>00214 <span class="stringliteral">                    --analysis mosaicism.marker-bandwidth=6000.confidence-window=251 \</span>
<a name="l00215"></a>00215 <span class="stringliteral">                    --analysis allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05 \</span>
<a name="l00216"></a>00216 <span class="stringliteral">                    --local-gc-background-correction-reference-method pdnn-reference-method \</span>
<a name="l00217"></a>00217 <span class="stringliteral">                    --wave-correction-reference-method wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false \</span>
<a name="l00218"></a>00218 <span class="stringliteral">                    --local-gc-background-intensity-adjustment-method pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0 \</span>
<a name="l00219"></a>00219 <span class="stringliteral">                    --image-correction-intensity-adjustment-method high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001 \</span>
<a name="l00220"></a>00220 <span class="stringliteral">                    --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3 --reference-chromosome=22 \</span>
<a name="l00221"></a>00221 <span class="stringliteral">            &quot;</span>);
<a name="l00222"></a>00222 
<a name="l00223"></a>00223         <span class="keywordtype">bool</span> bPassed = <span class="keyword">true</span>;
<a name="l00224"></a>00224         std::set&lt;std::string&gt; setIgnore;
<a name="l00225"></a>00225         std::string strFileName1 = <span class="stringliteral">&quot;../../../regression-data/data/idata/copynumber-cyto/batch/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>;
<a name="l00226"></a>00226         std::string strFileName2 = testDir + <span class="stringliteral">&quot;/batch/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>;
<a name="l00227"></a>00227 <span class="comment">//        if (!affx::File5_File::equivalent(strFileName1, strFileName2, &quot;SNPReference&quot;, &quot;Parameters&quot;, setIgnore, 0.0001, 1.0)) {bPassed = false;}</span>
<a name="l00228"></a>00228         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;SNPReference&quot;</span>, <span class="stringliteral">&quot;EMSNParameters&quot;</span>, setIgnore, 0.0001, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00229"></a>00229         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;SNPReference&quot;</span>, <span class="stringliteral">&quot;EMSNArray&quot;</span>, setIgnore, 0.0001, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00230"></a>00230         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;SNPReference&quot;</span>, <span class="stringliteral">&quot;SNPReference&quot;</span>, setIgnore, 0.0003, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00231"></a>00231 
<a name="l00232"></a>00232         strFileName1 = <span class="stringliteral">&quot;../../../regression-data/data/idata/copynumber-cyto/batch/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>;
<a name="l00233"></a>00233         strFileName2 = testDir + <span class="stringliteral">&quot;/batch/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>;
<a name="l00234"></a>00234 <span class="comment">//        if (!affx::File5_File::equivalent(strFileName1, strFileName2, &quot;Cyto2&quot;, &quot;Parameters&quot;, setIgnore, 0.0001, 1.0)) {bPassed = false;}</span>
<a name="l00235"></a>00235         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;AntigenomicProbes&quot;</span>, setIgnore, 0.0001, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00236"></a>00236         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;MedianSignals&quot;</span>, setIgnore, 0.5, 0.999999)) {bPassed = <span class="keyword">false</span>;}
<a name="l00237"></a>00237         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;ProbeEffects&quot;</span>, setIgnore, 0.0001, 0.999999)) {bPassed = <span class="keyword">false</span>;}
<a name="l00238"></a>00238         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;SNPPosteriors&quot;</span>, setIgnore, 0.0001, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00239"></a>00239         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;SketchCN&quot;</span>, setIgnore, 0.0001, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00240"></a>00240         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;SketchSNP&quot;</span>, setIgnore, 0.0001, 1.0)) {bPassed = <span class="keyword">false</span>;}
<a name="l00241"></a>00241         <span class="keywordflow">if</span> (!<a class="code" href="classaffx_1_1File5__File.html#a0a44055c962238474b5974d89d7bcc92">affx::File5_File::equivalent</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;WaveCorrection&quot;</span>, setIgnore, 0.0001, 0.999999)) {bPassed = <span class="keyword">false</span>;}
<a name="l00242"></a>00242 
<a name="l00243"></a>00243         <span class="keywordflow">if</span> (bPassed)
<a name="l00244"></a>00244         {
<a name="l00245"></a>00245                  execute( valgrindString1 + name2 + valgrindString2 + <span class="stringliteral">&quot; ./apt-copynumber-cyto \</span>
<a name="l00246"></a>00246 <span class="stringliteral">                        -v 3 --cyto2 true \</span>
<a name="l00247"></a>00247 <span class="stringliteral">                        --force true \</span>
<a name="l00248"></a>00248 <span class="stringliteral">                        --text-output true \</span>
<a name="l00249"></a>00249 <span class="stringliteral">                        --cnchp-output false \</span>
<a name="l00250"></a>00250 <span class="stringliteral">                        --cychp-output true \</span>
<a name="l00251"></a>00251 <span class="stringliteral">                        --set-analysis-name ca \</span>
<a name="l00252"></a>00252 <span class="stringliteral">                        --reference-input &quot;</span> + testDir + <span class="stringliteral">&quot;/batch/Cytogenetics_Array.10_cels.REF_MODEL \</span>
<a name="l00253"></a>00253 <span class="stringliteral">                        --snp-reference-input-file &quot;</span> + testDir + <span class="stringliteral">&quot;/batch/Cytogenetics_Array.10_cels.snpref.a5 \</span>
<a name="l00254"></a>00254 <span class="stringliteral">                        --spf-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.spf \</span>
<a name="l00255"></a>00255 <span class="stringliteral">                        --chrX-probes ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.chrXprobes \</span>
<a name="l00256"></a>00256 <span class="stringliteral">                        --chrY-probes ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.chrYprobes \</span>
<a name="l00257"></a>00257 <span class="stringliteral">                        --annotation-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.na28.annot.db \</span>
<a name="l00258"></a>00258 <span class="stringliteral">                        --probe-file ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.probe_tab \</span>
<a name="l00259"></a>00259 <span class="stringliteral">                        --snp-qc-snp-list ../../../regression-data/data/idata/lib/Cytogenetics_Array/Cytogenetics_Array.snplist.txt \</span>
<a name="l00260"></a>00260 <span class="stringliteral">                        --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-As_NA18547_A08_01_NN_20081218.CEL \</span>
<a name="l00261"></a>00261 <span class="stringliteral">                        --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-As_NA18603_B03_01_NN_20081218.CEL \</span>
<a name="l00262"></a>00262 <span class="stringliteral">                        --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-As_NA18971_C02_01_NN_20081218.CEL \</span>
<a name="l00263"></a>00263 <span class="stringliteral">                        --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL \</span>
<a name="l00264"></a>00264 <span class="stringliteral">                        --cels ../../../regression-data/data/idata/cel/Cytogenetics_Array/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL \</span>
<a name="l00265"></a>00265 <span class="stringliteral">                        --out-dir &quot;</span> + testDir + <span class="stringliteral">&quot;/batch \</span>
<a name="l00266"></a>00266 <span class="stringliteral">                        --male-gender-ratio-cutoff=1.3 \</span>
<a name="l00267"></a>00267 <span class="stringliteral">                        --female-gender-ratio-cutoff=1.0 \</span>
<a name="l00268"></a>00268 <span class="stringliteral">                        --xx-cutoff=0.8 \</span>
<a name="l00269"></a>00269 <span class="stringliteral">                        --xx-cutoff-high=1.07 \</span>
<a name="l00270"></a>00270 <span class="stringliteral">                        --y-cutoff=0.65 \</span>
<a name="l00271"></a>00271 <span class="stringliteral">                        --analysis log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5 \</span>
<a name="l00272"></a>00272 <span class="stringliteral">                        --analysis kernel-smooth.sigma_span=50 \</span>
<a name="l00273"></a>00273 <span class="stringliteral">                        --analysis cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6 \</span>
<a name="l00274"></a>00274 <span class="stringliteral">                        --analysis cn-cyto2-gender.cutoff=0.5 \</span>
<a name="l00275"></a>00275 <span class="stringliteral">                        --analysis cn-segment \</span>
<a name="l00276"></a>00276 <span class="stringliteral">                        --analysis loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0 \</span>
<a name="l00277"></a>00277 <span class="stringliteral">                        --analysis mosaicism.marker-bandwidth=6000.confidence-window=251 \</span>
<a name="l00278"></a>00278 <span class="stringliteral">                        --analysis allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05 \</span>
<a name="l00279"></a>00279 <span class="stringliteral">                        --local-gc-background-correction-reference-method pdnn-reference-method \</span>
<a name="l00280"></a>00280 <span class="stringliteral">                        --wave-correction-reference-method wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false \</span>
<a name="l00281"></a>00281 <span class="stringliteral">                        --local-gc-background-intensity-adjustment-method pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0 \</span>
<a name="l00282"></a>00282 <span class="stringliteral">                        --image-correction-intensity-adjustment-method high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001 \</span>
<a name="l00283"></a>00283 <span class="stringliteral">                        --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3 --reference-chromosome=22 \</span>
<a name="l00284"></a>00284 <span class="stringliteral">                &quot;</span>);
<a name="l00285"></a>00285 
<a name="l00286"></a>00286             equivalency(<span class="stringliteral">&quot;batch&quot;</span>, <span class="stringliteral">&quot;ca&quot;</span>, vCytogenetics_Array_SixFiles);
<a name="l00287"></a>00287         }
<a name="l00288"></a>00288         <span class="keywordflow">else</span> {numFailed++;}
<a name="l00289"></a>00289     }
<a name="l00290"></a>00290 };
<a name="l00291"></a>00291 
<a name="l00292"></a>00292 <span class="keywordtype">int</span> <a class="code" href="canary-test_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="Everybody&amp;#39;s favorite function.">main</a>(<span class="keywordtype">int</span> argc, <span class="keywordtype">char</span>* argv[]) {
<a name="l00293"></a>00293   <span class="keywordflow">try</span> {
<a name="l00294"></a>00294     <a class="code" href="classFsTestDir.html">FsTestDir</a> testDir;
<a name="l00295"></a>00295     testDir.<a class="code" href="classFsTestDir.html#ac14fb44eb179d1094ef5c9d127a39bbb">setTestDir</a>(<span class="stringliteral">&quot;copynumber/copynumber-cyto-valgrind&quot;</span>,<span class="keyword">true</span>);
<a name="l00296"></a>00296     
<a name="l00297"></a>00297     <span class="comment">/* Set up the logging and message handlers. */</span>
<a name="l00298"></a>00298 
<a name="l00299"></a>00299     ofstream logOut;
<a name="l00300"></a>00300     <span class="keywordtype">string</span> logName;
<a name="l00301"></a>00301     logName = testDir.asString() + <span class="stringliteral">&quot;/test-regression-copynumber-cyto-quick.log&quot;</span>;
<a name="l00302"></a>00302     valgrindString1 = valgrindString1 + testDir.asString() + <span class="stringliteral">&quot;/&quot;</span>;
<a name="l00303"></a>00303     
<a name="l00304"></a>00304     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Log file: &quot;</span> + logName);
<a name="l00305"></a>00305     <a class="code" href="classFs.html#a1cc6e82074c84dd92bd7cdcf7d374b7c" title="Open an ofstream for writing to.">Fs::mustOpenToWrite</a>(logOut, logName.c_str());
<a name="l00306"></a>00306     <a class="code" href="classLogStream.html" title="Log messages with timestamps.">LogStream</a> log(3, &amp;logOut);
<a name="l00307"></a>00307 
<a name="l00308"></a>00308     Verbose::pushMsgHandler(&amp;log);
<a name="l00309"></a>00309     <a class="code" href="classVerbose.html#af20e052bc19f95d3a949d1daa1647fa8" title="Functions to add and remove handlers for communcation functions.">Verbose::pushProgressHandler</a>(&amp;log);
<a name="l00310"></a>00310     Verbose::pushWarnHandler(&amp;log);
<a name="l00311"></a>00311 
<a name="l00312"></a>00312     <a class="code" href="classVerbose.html#a416deece7d5098dc2e8759ef5fb134e3" title="Set the level of verbosity desired.">Verbose::setLevel</a>(3);
<a name="l00313"></a>00313     <a class="code" href="classtest__regression__copynumber__cyto__quick.html">test_regression_copynumber_cyto_quick</a> test;
<a name="l00314"></a>00314     test.testDir = testDir.asString();
<a name="l00315"></a>00315 
<a name="l00316"></a>00316     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Execute regression test: test-regression-copynumber-cyto-valgrind&quot;</span>);
<a name="l00317"></a>00317 
<a name="l00318"></a>00318     test.run();
<a name="l00319"></a>00319 
<a name="l00320"></a>00320     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;NumPassed: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(test.numPassed) + <span class="stringliteral">&quot; NumFailed: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(test.numFailed) + <span class="stringliteral">&quot; for test-regression-copynumber-cyto-valgrind&quot;</span>);
<a name="l00321"></a>00321     logOut.close();
<a name="l00322"></a>00322     <span class="keywordflow">return</span> test.numFailed != 0;
<a name="l00323"></a>00323   }
<a name="l00324"></a>00324   <span class="keywordflow">catch</span>(...) {
<a name="l00325"></a>00325       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Unexpected Error: uncaught exception.&quot;</span>);
<a name="l00326"></a>00326       <span class="keywordflow">return</span> 1;
<a name="l00327"></a>00327   }
<a name="l00328"></a>00328   <span class="keywordflow">return</span> 1;
<a name="l00329"></a>00329 }
<a name="l00330"></a>00330 
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:54 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
