<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: copynumber/apt-copynumber-cyto/regression/test-regression-copynumber-cyto-quick.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_41dda80f44caddc01a4a55d902223772.html">copynumber</a>      </li>
      <li><a class="el" href="dir_d61a7909814a46bce5ab2f5f0e583661.html">apt-copynumber-cyto</a>      </li>
      <li><a class="el" href="dir_295cb3590393dbdce00cd66c6df31eff.html">regression</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>test-regression-copynumber-cyto-quick.cpp</h1>  </div>
</div>
<div class="contents">
<a href="test-regression-copynumber-cyto-quick_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2009 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify</span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as</span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">//</span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful,</span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU</span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">//</span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License</span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the</span>
<a name="l00016"></a>00016 <span class="comment">//</span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc.,</span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment"></span>
<a name="l00022"></a>00022 <span class="comment">/**</span>
<a name="l00023"></a>00023 <span class="comment"> * @file   test-regression-copynumber-cyto-quick.cpp</span>
<a name="l00024"></a>00024 <span class="comment"> * @brief  Program for doing regression tests on cyto data.</span>
<a name="l00025"></a>00025 <span class="comment"> */</span>
<a name="l00026"></a>00026 
<a name="l00027"></a>00027 <span class="preprocessor">#include &lt;string&gt;</span>
<a name="l00028"></a>00028 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="FsTestDir_8h.html" title="///">util/FsTestDir.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="Verbose_8h.html" title="Class for doing logging and some command line ui.">util/Verbose.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="Calvin_8h.html" title="This header contains the Calvin class definition.">calvin_files/utils/src/Calvin.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="RegressionSuite_8h.html" title="Base class for holding basic options for regression data.">util/RegressionSuite.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="RegressionTest_8h.html" title="Utility class for helping to do regression testing.">util/RegressionTest.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="RegressionCheck_8h.html" title="Virtual class for checks to be done after a regression test.">util/RegressionCheck.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;copynumber/apt-copynumber-cyto/regression/File5EquivalentCheck.h&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;<a class="code" href="LogStream_8h.html" title="Output verbose messages to log file as we go along.">util/LogStream.h</a>&quot;</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00038"></a>00038 <span class="preprocessor">#include &quot;file5/File5.h&quot;</span>
<a name="l00039"></a>00039 
<a name="l00040"></a>00040 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00041"></a>00041 <span class="keyword">using namespace </span>std;
<a name="l00042"></a>00042 
<a name="l00043"></a><a class="code" href="classtest__regression__copynumber__cyto__quick.html">00043</a> <span class="keyword">class </span><a class="code" href="classtest__regression__copynumber__cyto__quick.html">test_regression_copynumber_cyto_quick</a> : <span class="keyword">public</span> <a class="code" href="classRegressionSuite.html">RegressionSuite</a>, <span class="keyword">public</span> <a class="code" href="classRegressionCheck.html" title="Abstract base class (i.e.">RegressionCheck</a>
<a name="l00044"></a>00044 {
<a name="l00045"></a>00045 <span class="keyword">public</span>:
<a name="l00046"></a>00046     <span class="keywordtype">int</span> numPassed;
<a name="l00047"></a>00047     <span class="keywordtype">int</span> numFailed;
<a name="l00048"></a>00048     std::string m_testDir; 
<a name="l00049"></a>00049     std::set&lt;std::string&gt; setIgnore;
<a name="l00050"></a>00050     std::map&lt;std::string, float&gt; mapEpsilon;
<a name="l00051"></a>00051     std::vector&lt;std::string&gt; vCytogenetics_Array_SixFiles;
<a name="l00052"></a>00052 
<a name="l00053"></a>00053   std::string m_OutPrefix;
<a name="l00054"></a>00054   std::string m_OutModifier;
<a name="l00055"></a>00055   std::set&lt;std::string&gt; m_SetSetIgnore;
<a name="l00056"></a>00056   std::vector&lt;std::string&gt; m_ToCompare;
<a name="l00057"></a>00057 
<a name="l00058"></a><a class="code" href="classtest__regression__copynumber__cyto__quick.html#a8ec63efc2c3b2d2727f9e19a363c1848">00058</a>   <span class="keywordtype">bool</span> <a class="code" href="classtest__regression__copynumber__cyto__quick.html#a8ec63efc2c3b2d2727f9e19a363c1848" title="Check the condition expected.">check</a>(std::string &amp;msg) {
<a name="l00059"></a>00059     <span class="keywordflow">return</span> equivalency(m_OutPrefix, m_OutModifier, m_ToCompare, m_SetSetIgnore);
<a name="l00060"></a>00060   }
<a name="l00061"></a>00061 
<a name="l00062"></a>00062     <a class="code" href="classtest__regression__copynumber__cyto__quick.html">test_regression_copynumber_cyto_quick</a>()
<a name="l00063"></a>00063     {
<a name="l00064"></a>00064         numPassed = 0;
<a name="l00065"></a>00065         numFailed = 0;
<a name="l00066"></a>00066 
<a name="l00067"></a>00067         <span class="comment">// Header paramters to ignore</span>
<a name="l00068"></a>00068         setIgnore.insert(<span class="stringliteral">&quot;FileCreationTime&quot;</span>);
<a name="l00069"></a>00069         setIgnore.insert(<span class="stringliteral">&quot;FileIdentifier&quot;</span>);
<a name="l00070"></a>00070         setIgnore.insert(<span class="stringliteral">&quot;program-version&quot;</span>);
<a name="l00071"></a>00071         setIgnore.insert(<span class="stringliteral">&quot;create_date&quot;</span>);
<a name="l00072"></a>00072         setIgnore.insert(<span class="stringliteral">&quot;create-date&quot;</span>);
<a name="l00073"></a>00073         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-verbose&quot;</span>);
<a name="l00074"></a>00074         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-exec-guid&quot;</span>);
<a name="l00075"></a>00075         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-program-cvs-id&quot;</span>);
<a name="l00076"></a>00076         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-version-to-report&quot;</span>);
<a name="l00077"></a>00077         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-command-line&quot;</span>);
<a name="l00078"></a>00078         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-mem-usage&quot;</span>);
<a name="l00079"></a>00079         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-run-probeset-genotype&quot;</span>);
<a name="l00080"></a>00080         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-cels&quot;</span>);
<a name="l00081"></a>00081         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-out-dir&quot;</span>);
<a name="l00082"></a>00082         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-state-time-start&quot;</span>);
<a name="l00083"></a>00083         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-state-free-mem-at-start&quot;</span>);
<a name="l00084"></a>00084         setIgnore.insert(<span class="stringliteral">&quot;affymetrix-algorithm-param-option-temp-dir&quot;</span>);
<a name="l00085"></a>00085 
<a name="l00086"></a>00086         <span class="comment">// Data Set columns to ignore (Specify as Group.Set.Column)</span>
<a name="l00087"></a>00087 <span class="comment">//        setIgnore.insert(&quot;AlgorithmData.MarkerABSignal.SCAR&quot;);</span>
<a name="l00088"></a>00088 
<a name="l00089"></a>00089         <span class="comment">// Data Set column epsilon override (Specify as Group.Set.Column)</span>
<a name="l00090"></a>00090 <span class="comment">//        mapEpsilon.insert(std::pair&lt;std::string, float&gt;(&quot;AlgorithmData.MarkerABSignal.SCAR&quot;, (float)0.0005));</span>
<a name="l00091"></a>00091         mapEpsilon.insert(std::pair&lt;string, float&gt;(<span class="stringliteral">&quot;Segments.CN.MeanMarkerDistance&quot;</span>, 376.0f));
<a name="l00092"></a>00092         <span class="comment">// File tags to check for equivalency</span>
<a name="l00093"></a>00093         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-As_NA18603_B03_01_NN_20081218&quot;</span>);
<a name="l00094"></a>00094         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-Cs_NA10857_A10_01_NN_20081218&quot;</span>);
<a name="l00095"></a>00095 <span class="preprocessor">                #ifdef __LP64__</span>
<a name="l00096"></a>00096 <span class="preprocessor"></span><span class="preprocessor">                #ifdef __linux__</span>
<a name="l00097"></a>00097 <span class="preprocessor"></span>        vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-As_NA18547_A08_01_NN_20081218&quot;</span>);
<a name="l00098"></a>00098         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-Cs_NA12003_D04_01_NN_20081218&quot;</span>);
<a name="l00099"></a>00099         vCytogenetics_Array_SixFiles.push_back(<span class="stringliteral">&quot;HapMap-As_NA18971_C02_01_NN_20081218&quot;</span>);
<a name="l00100"></a>00100 <span class="preprocessor">                #endif</span>
<a name="l00101"></a>00101 <span class="preprocessor"></span><span class="preprocessor">                #endif</span>
<a name="l00102"></a>00102 <span class="preprocessor"></span>    }
<a name="l00103"></a>00103 
<a name="l00104"></a>00104     <span class="keywordtype">void</span> execute(<span class="keyword">const</span> std::string&amp; str)  {
<a name="l00105"></a>00105       std::string cmd = <a class="code" href="classFs.html#af8ce27acb118594a0e073ce6d2d70da4" title="convert only the first component of the command to unc. / return the command string.">Fs::convertCommandToUnc</a>(str);
<a name="l00106"></a>00106       <span class="keywordflow">if</span> (system(cmd.c_str()) != 0) {<a class="code" href="classErr.html#a121c0cbb1687c93444a1b5f7f6ea56af" title="Print the message supplied and abort program.">Err::errAbort</a>(<span class="stringliteral">&quot;Execution failed for command: &quot;</span> + cmd);}
<a name="l00107"></a>00107     }
<a name="l00108"></a>00108 
<a name="l00109"></a>00109         <span class="keywordtype">void</span> equivalency(   <span class="keyword">const</span> std::string&amp; name,
<a name="l00110"></a>00110                             <span class="keyword">const</span> std::string&amp; modifier,
<a name="l00111"></a>00111                             std::vector&lt;string&gt;&amp; vFileTags)
<a name="l00112"></a>00112         {
<a name="l00113"></a>00113                 std::set&lt;std::string&gt; setSetIgnore;
<a name="l00114"></a>00114 
<a name="l00115"></a>00115                  equivalency(    name,
<a name="l00116"></a>00116                                 modifier,
<a name="l00117"></a>00117                                 vFileTags,
<a name="l00118"></a>00118                                 setSetIgnore);
<a name="l00119"></a>00119 
<a name="l00120"></a>00120         }
<a name="l00121"></a>00121 
<a name="l00122"></a>00122         <span class="keywordtype">bool</span> equivalency(   <span class="keyword">const</span> std::string&amp; name,
<a name="l00123"></a>00123                             <span class="keyword">const</span> std::string&amp; modifier,
<a name="l00124"></a>00124                             std::vector&lt;string&gt;&amp; vFileTags,
<a name="l00125"></a>00125                             std::set&lt;std::string&gt;&amp; setSetIgnore)
<a name="l00126"></a>00126     {
<a name="l00127"></a>00127         <span class="keywordtype">bool</span> bPassed = <span class="keyword">true</span>;
<a name="l00128"></a>00128         <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> strFileName1;
<a name="l00129"></a>00129         <a class="code" href="classAffxString.html" title="A class derived from std::string.">AffxString</a> strFileName2;
<a name="l00130"></a>00130         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;*&quot;</span>);
<a name="l00131"></a>00131         <span class="keywordflow">for</span>(<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 0; (i &lt; vFileTags.size()); i++)
<a name="l00132"></a>00132         {
<a name="l00133"></a>00133 
<a name="l00134"></a>00134             strFileName1 = <span class="stringliteral">&quot;../../../regression-data/data/idata/copynumber-cyto/&quot;</span> + name + <span class="stringliteral">&quot;/&quot;</span> + vFileTags[i] + <span class="stringliteral">&quot;.&quot;</span> + modifier + <span class="stringliteral">&quot;.cychp&quot;</span>;
<a name="l00135"></a>00135             strFileName2 = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(m_testDir , name,  vFileTags[i] ) + <span class="stringliteral">&quot;.&quot;</span> + modifier + <span class="stringliteral">&quot;.cychp&quot;</span>;
<a name="l00136"></a>00136             <span class="keywordflow">if</span> (!Calvin::equivalent(strFileName1, strFileName2, setIgnore,setSetIgnore, mapEpsilon, 0.0001, 0.9999989, <span class="keyword">false</span>)) {bPassed = <span class="keyword">false</span>;}
<a name="l00137"></a>00137         }
<a name="l00138"></a>00138         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;*&quot;</span>);
<a name="l00139"></a>00139         <span class="keywordflow">if</span> (bPassed) {
<a name="l00140"></a>00140                   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;***Passed*** &quot;</span> + name + <span class="stringliteral">&quot; vs &quot;</span> + modifier);
<a name="l00141"></a>00141                   numPassed++;
<a name="l00142"></a>00142                 }
<a name="l00143"></a>00143                 <span class="keywordflow">else</span> {
<a name="l00144"></a>00144                   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;***Failed*** &quot;</span> + name + <span class="stringliteral">&quot; vs &quot;</span> + modifier);
<a name="l00145"></a>00145                   numFailed++;
<a name="l00146"></a>00146                 }
<a name="l00147"></a>00147                 <span class="keywordflow">return</span> bPassed;
<a name="l00148"></a>00148     }
<a name="l00149"></a>00149 
<a name="l00150"></a>00150     <span class="keywordtype">void</span> run()
<a name="l00151"></a>00151     {
<a name="l00152"></a>00152       <span class="comment">//single();</span>
<a name="l00153"></a>00153       batch();
<a name="l00154"></a>00154     }
<a name="l00155"></a>00155 
<a name="l00156"></a>00156     <span class="keywordtype">void</span> single()
<a name="l00157"></a>00157     {
<a name="l00158"></a>00158           <span class="comment">//Fs::rm_rf(&quot;test-generated/qt-single/&quot;, false);</span>
<a name="l00159"></a>00159           std::set&lt;std::string&gt; setSetIgnore;
<a name="l00160"></a>00160 <span class="preprocessor">#ifndef __LP64__</span>
<a name="l00161"></a>00161 <span class="preprocessor"></span>          {
<a name="l00162"></a>00162                   std::string ignoreString = <span class="stringliteral">&quot;Segments.Mosaicism&quot;</span>;
<a name="l00163"></a>00163                   std::string ignoreString2 = <span class="stringliteral">&quot;ProbeSets.AllelePeaks&quot;</span>;
<a name="l00164"></a>00164                   setSetIgnore.insert(ignoreString);
<a name="l00165"></a>00165                   setSetIgnore.insert(ignoreString2);
<a name="l00166"></a>00166           }
<a name="l00167"></a>00167 <span class="preprocessor">#endif</span>
<a name="l00168"></a>00168 <span class="preprocessor"></span><span class="preprocessor">#ifdef __APPLE__</span>
<a name="l00169"></a>00169 <span class="preprocessor"></span>          {
<a name="l00170"></a>00170                   std::string ignoreString = <span class="stringliteral">&quot;Segments.Mosaicism&quot;</span>;
<a name="l00171"></a>00171                   std::string ignoreString2 = <span class="stringliteral">&quot;ProbeSets.AllelePeaks&quot;</span>;
<a name="l00172"></a>00172                   setSetIgnore.insert(ignoreString);
<a name="l00173"></a>00173                   setSetIgnore.insert(ignoreString2);
<a name="l00174"></a>00174           }
<a name="l00175"></a>00175 <span class="preprocessor">#endif</span>
<a name="l00176"></a>00176 <span class="preprocessor"></span>          std::string ignoreString = <span class="stringliteral">&quot;ProbeSets.AllelePeaks&quot;</span>;
<a name="l00177"></a>00177           setSetIgnore.insert(ignoreString);
<a name="l00178"></a>00178 
<a name="l00179"></a>00179 
<a name="l00180"></a>00180           <span class="keywordtype">string</span> idata=<span class="stringliteral">&quot;../../../regression-data/data/idata&quot;</span>;
<a name="l00181"></a>00181           <span class="keywordtype">string</span> outdir = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(m_testDir , <span class="stringliteral">&quot;qt-single&quot;</span>);
<a name="l00182"></a>00182           vector&lt;RegressionCheck *&gt; checks;
<a name="l00183"></a>00183           m_OutPrefix = <span class="stringliteral">&quot;qt-single&quot;</span>;
<a name="l00184"></a>00184           m_OutModifier = <span class="stringliteral">&quot;ca&quot;</span>;
<a name="l00185"></a>00185           m_ToCompare =  vCytogenetics_Array_SixFiles;
<a name="l00186"></a>00186           m_SetSetIgnore.clear();
<a name="l00187"></a>00187           m_SetSetIgnore = setSetIgnore;
<a name="l00188"></a>00188           checks.push_back(<span class="keyword">this</span>);
<a name="l00189"></a>00189           <span class="keywordtype">string</span> name = <span class="stringliteral">&quot;qt-single&quot;</span>;
<a name="l00190"></a>00190           <span class="keywordtype">string</span> command =
<a name="l00191"></a>00191             string(<span class="stringliteral">&quot;./apt-copynumber-cyto&quot;</span>) +
<a name="l00192"></a>00192             <span class="stringliteral">&quot; -v 1&quot;</span> +
<a name="l00193"></a>00193             <span class="stringliteral">&quot; --cyto2 true&quot;</span> +
<a name="l00194"></a>00194             <span class="stringliteral">&quot; --force true&quot;</span> +
<a name="l00195"></a>00195             <span class="stringliteral">&quot; --text-output true&quot;</span> +
<a name="l00196"></a>00196             <span class="stringliteral">&quot; --cnchp-output false&quot;</span> +
<a name="l00197"></a>00197             <span class="stringliteral">&quot; --cychp-output true&quot;</span> +
<a name="l00198"></a>00198             <span class="stringliteral">&quot; --set-analysis-name ca&quot;</span> +
<a name="l00199"></a>00199             <span class="stringliteral">&quot; --reference-input &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.na28.REF_MODEL&quot;</span> +
<a name="l00200"></a>00200             <span class="stringliteral">&quot; --snp-reference-input-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.na28.snpref.a5&quot;</span> +
<a name="l00201"></a>00201             <span class="stringliteral">&quot; --spf-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.spf&quot;</span> +
<a name="l00202"></a>00202             <span class="stringliteral">&quot; --chrX-probes &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.chrXprobes&quot;</span> +
<a name="l00203"></a>00203             <span class="stringliteral">&quot; --chrY-probes &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.chrYprobes&quot;</span> +
<a name="l00204"></a>00204             <span class="stringliteral">&quot; --annotation-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.na28.annot.db&quot;</span> +
<a name="l00205"></a>00205             <span class="stringliteral">&quot; --probe-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.probe_tab&quot;</span> +
<a name="l00206"></a>00206             <span class="stringliteral">&quot; --snp-qc-snp-list &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.snplist.txt&quot;</span> +
<a name="l00207"></a>00207             <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-As_NA18547_A08_01_NN_20081218.CEL&quot;</span> +
<a name="l00208"></a>00208             <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-As_NA18603_B03_01_NN_20081218.CEL&quot;</span> +
<a name="l00209"></a>00209             <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-As_NA18971_C02_01_NN_20081218.CEL&quot;</span> +
<a name="l00210"></a>00210             <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL&quot;</span> +
<a name="l00211"></a>00211             <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL&quot;</span> +
<a name="l00212"></a>00212             <span class="stringliteral">&quot; --out-dir &quot;</span> + outdir +
<a name="l00213"></a>00213             <span class="stringliteral">&quot; --male-gender-ratio-cutoff=1.3&quot;</span> +
<a name="l00214"></a>00214             <span class="stringliteral">&quot; --female-gender-ratio-cutoff=1.0&quot;</span> +
<a name="l00215"></a>00215             <span class="stringliteral">&quot; --xx-cutoff=0.8&quot;</span> +
<a name="l00216"></a>00216             <span class="stringliteral">&quot; --xx-cutoff-high=1.07&quot;</span> +
<a name="l00217"></a>00217             <span class="stringliteral">&quot; --y-cutoff=0.65&quot;</span> +
<a name="l00218"></a>00218             <span class="stringliteral">&quot; --analysis log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span> +
<a name="l00219"></a>00219             <span class="stringliteral">&quot; --analysis kernel-smooth.sigma_span=50&quot;</span> +
<a name="l00220"></a>00220             <span class="stringliteral">&quot; --analysis cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span> +
<a name="l00221"></a>00221             <span class="stringliteral">&quot; --analysis cn-cyto2-gender.cutoff=0.5&quot;</span> +
<a name="l00222"></a>00222             <span class="stringliteral">&quot; --analysis cn-segment&quot;</span> +
<a name="l00223"></a>00223             <span class="stringliteral">&quot; --analysis loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span> +
<a name="l00224"></a>00224             <span class="stringliteral">&quot; --analysis mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span> +
<a name="l00225"></a>00225             <span class="stringliteral">&quot; --analysis allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span> +
<a name="l00226"></a>00226             <span class="stringliteral">&quot; --local-gc-background-correction-reference-method pdnn-reference-method&quot;</span> +
<a name="l00227"></a>00227             <span class="stringliteral">&quot; --wave-correction-reference-method wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span> +
<a name="l00228"></a>00228             <span class="stringliteral">&quot; --local-gc-background-intensity-adjustment-method pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span> +
<a name="l00229"></a>00229             <span class="stringliteral">&quot; --image-correction-intensity-adjustment-method high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span> +
<a name="l00230"></a>00230             <span class="stringliteral">&quot; --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3&quot;</span>
<a name="l00231"></a>00231             <span class="stringliteral">&quot; --reference-chromosome=22&quot;</span>;
<a name="l00232"></a>00232           <a class="code" href="classRegressionTest.html" title="Utility class for helping to do regression testing.">RegressionTest</a> test(name.c_str(), command.c_str(), checks, <span class="keyword">false</span>);
<a name="l00233"></a>00233           test.setSuite(*<span class="keyword">this</span>,  outdir, outdir + <span class="stringliteral">&quot;/apt-copynumber-cyto.log&quot;</span>, outdir + <span class="stringliteral">&quot;/valgrind.log&quot;</span>);
<a name="l00234"></a>00234           <span class="keywordflow">if</span>(!test.pass()) {
<a name="l00235"></a>00235             <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Error in: &quot;</span> + name + <span class="stringliteral">&quot;(): &quot;</span> + test.getErrorMsg());
<a name="l00236"></a>00236           }
<a name="l00237"></a>00237 
<a name="l00238"></a>00238         }
<a name="l00239"></a>00239 
<a name="l00240"></a>00240   <span class="keywordtype">void</span> batch()
<a name="l00241"></a>00241   {
<a name="l00242"></a>00242     <span class="comment">//Fs::rm_rf(&quot;test-generated/qt-batch/&quot;, false);</span>
<a name="l00243"></a>00243     <span class="keywordtype">string</span> idata=<span class="stringliteral">&quot;../../../regression-data/data/idata&quot;</span>;
<a name="l00244"></a>00244     <span class="keywordtype">string</span> outdir = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(m_testDir, <span class="stringliteral">&quot;qt-batch&quot;</span>);
<a name="l00245"></a>00245           vector&lt;RegressionCheck *&gt; checks;
<a name="l00246"></a>00246           m_OutPrefix = <span class="stringliteral">&quot;qt-batch&quot;</span>;
<a name="l00247"></a>00247           m_OutModifier = <span class="stringliteral">&quot;ca&quot;</span>;
<a name="l00248"></a>00248           m_ToCompare =  vCytogenetics_Array_SixFiles;
<a name="l00249"></a>00249           <span class="keywordtype">string</span> name = <span class="stringliteral">&quot;qt-batch&quot;</span>;
<a name="l00250"></a>00250           <span class="keywordtype">string</span> command =
<a name="l00251"></a>00251             string(<span class="stringliteral">&quot;./apt-copynumber-cyto&quot;</span>)+
<a name="l00252"></a>00252             <span class="stringliteral">&quot; -v 1&quot;</span>+
<a name="l00253"></a>00253             <span class="stringliteral">&quot; --cyto2 true&quot;</span> +
<a name="l00254"></a>00254             <span class="stringliteral">&quot; --force true&quot;</span> +
<a name="l00255"></a>00255             <span class="stringliteral">&quot; --text-output true&quot;</span> +
<a name="l00256"></a>00256             <span class="stringliteral">&quot; --cnchp-output false&quot;</span> +
<a name="l00257"></a>00257             <span class="stringliteral">&quot; --cychp-output true&quot;</span> +
<a name="l00258"></a>00258             <span class="stringliteral">&quot; --set-analysis-name ca&quot;</span> +
<a name="l00259"></a>00259             <span class="stringliteral">&quot; --reference-output &quot;</span> + m_testDir + <span class="stringliteral">&quot;/qt-batch/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span> +
<a name="l00260"></a>00260             <span class="stringliteral">&quot; --snp-reference-input-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.na28.snpref.a5&quot;</span> +
<a name="l00261"></a>00261             <span class="stringliteral">&quot; --snp-reference-output-file &quot;</span> +
<a name="l00262"></a>00262             m_testDir +
<a name="l00263"></a>00263             <span class="stringliteral">&quot;/qt-batch/Cytogenetics_Array.10_cels.snpref.a5&quot;</span> +
<a name="l00264"></a>00264             <span class="stringliteral">&quot; --spf-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.spf&quot;</span> +
<a name="l00265"></a>00265             <span class="stringliteral">&quot; --chrX-probes &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.chrXprobes&quot;</span> +
<a name="l00266"></a>00266             <span class="stringliteral">&quot; --chrY-probes &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.chrYprobes&quot;</span> +
<a name="l00267"></a>00267             <span class="stringliteral">&quot; --annotation-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.na28.annot.db&quot;</span> +
<a name="l00268"></a>00268             <span class="stringliteral">&quot; --probe-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.probe_tab&quot;</span> +
<a name="l00269"></a>00269             <span class="stringliteral">&quot; --snp-qc-snp-list &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.snplist.txt&quot;</span> +
<a name="l00270"></a>00270             <span class="stringliteral">&quot; --gender-override-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/HapMap-LOHReference-Full-Genders.txt&quot;</span> +
<a name="l00271"></a>00271             <span class="stringliteral">&quot; --genotype-call-override-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/HapMapCalls_LOHReferenceSet-FullArray.txt&quot;</span> +
<a name="l00272"></a>00272             <span class="stringliteral">&quot; --cel-files &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/CELFileList_5F5M.txt&quot;</span> +
<a name="l00273"></a>00273             <span class="stringliteral">&quot; --out-dir &quot;</span> + outdir +
<a name="l00274"></a>00274             <span class="stringliteral">&quot; --male-gender-ratio-cutoff=1.3&quot;</span> +
<a name="l00275"></a>00275             <span class="stringliteral">&quot; --female-gender-ratio-cutoff=1.0&quot;</span> +
<a name="l00276"></a>00276             <span class="stringliteral">&quot; --xx-cutoff=0.8&quot;</span> +
<a name="l00277"></a>00277             <span class="stringliteral">&quot; --xx-cutoff-high=1.07&quot;</span> +
<a name="l00278"></a>00278             <span class="stringliteral">&quot; --y-cutoff=0.65&quot;</span> +
<a name="l00279"></a>00279             <span class="stringliteral">&quot; --analysis log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span> +
<a name="l00280"></a>00280             <span class="stringliteral">&quot; --analysis kernel-smooth.sigma_span=50&quot;</span> +
<a name="l00281"></a>00281             <span class="stringliteral">&quot; --analysis cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span> +
<a name="l00282"></a>00282             <span class="stringliteral">&quot; --analysis cn-cyto2-gender.cutoff=0.5&quot;</span> +
<a name="l00283"></a>00283             <span class="stringliteral">&quot; --analysis cn-segment&quot;</span> +
<a name="l00284"></a>00284             <span class="stringliteral">&quot; --analysis loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span> +
<a name="l00285"></a>00285             <span class="stringliteral">&quot; --analysis mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span> +
<a name="l00286"></a>00286             <span class="stringliteral">&quot; --analysis allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span> +
<a name="l00287"></a>00287             <span class="stringliteral">&quot; --local-gc-background-correction-reference-method pdnn-reference-method&quot;</span> +
<a name="l00288"></a>00288             <span class="stringliteral">&quot; --wave-correction-reference-method wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span> +
<a name="l00289"></a>00289             <span class="stringliteral">&quot; --local-gc-background-intensity-adjustment-method pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span> +
<a name="l00290"></a>00290             <span class="stringliteral">&quot; --image-correction-intensity-adjustment-method high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span> +
<a name="l00291"></a>00291             <span class="stringliteral">&quot; --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3 &quot;</span>+
<a name="l00292"></a>00292             <span class="stringliteral">&quot; --reference-chromosome=22&quot;</span>;
<a name="l00293"></a>00293 
<a name="l00294"></a>00294         <span class="keywordtype">bool</span> bPassed = <span class="keyword">true</span>;
<a name="l00295"></a>00295         std::set&lt;std::string&gt; setIgnore;
<a name="l00296"></a>00296         std::string strFileName1 = <span class="stringliteral">&quot;&quot;</span>+idata+<span class="stringliteral">&quot;/copynumber-cyto/qt-batch/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>;
<a name="l00297"></a>00297         std::string strFileName2 = m_testDir + <span class="stringliteral">&quot;/qt-batch/Cytogenetics_Array.10_cels.snpref.a5&quot;</span>;
<a name="l00298"></a>00298 <span class="comment">//        checks.push_back(new File5EquivalentCheck(strFileName1, strFileName2, &quot;SNPReference&quot;, &quot;Parameters&quot;, setIgnore, 0.0001, 1.0));</span>
<a name="l00299"></a>00299         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;SNPReference&quot;</span>, <span class="stringliteral">&quot;EMSNParameters&quot;</span>, setIgnore, 0.0001, 1.0));
<a name="l00300"></a>00300         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;SNPReference&quot;</span>, <span class="stringliteral">&quot;EMSNArray&quot;</span>, setIgnore, 0.0001, 1.0));
<a name="l00301"></a>00301         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;SNPReference&quot;</span>, <span class="stringliteral">&quot;SNPReference&quot;</span>, setIgnore, 0.0003, 1.0));
<a name="l00302"></a>00302 
<a name="l00303"></a>00303         strFileName1 = <span class="stringliteral">&quot;&quot;</span>+idata+<span class="stringliteral">&quot;/copynumber-cyto/qt-batch/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>;
<a name="l00304"></a>00304         strFileName2 = m_testDir + <span class="stringliteral">&quot;/qt-batch/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span>;
<a name="l00305"></a>00305 <span class="comment">//        checks.push_back(new File5EquivalentCheck(strFileName1, strFileName2, &quot;Cyto2&quot;, &quot;Parameters&quot;, setIgnore, 0.0001, 1.0));</span>
<a name="l00306"></a>00306         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;AntigenomicProbes&quot;</span>, setIgnore, 0.0001, 1.0));
<a name="l00307"></a>00307         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;MedianSignals&quot;</span>, setIgnore, 0.5, 0.999999));
<a name="l00308"></a>00308         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;ProbeEffects&quot;</span>, setIgnore, 0.0001, 0.999999));
<a name="l00309"></a>00309         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;SNPPosteriors&quot;</span>, setIgnore, 0.0001, 1.0));
<a name="l00310"></a>00310         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;SketchCN&quot;</span>, setIgnore, 0.0001, 1.0));
<a name="l00311"></a>00311         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;SketchSNP&quot;</span>, setIgnore, 0.0001, 1.0));
<a name="l00312"></a>00312         checks.push_back(<span class="keyword">new</span> <a class="code" href="classFile5EquivalentCheck.html">File5EquivalentCheck</a>(strFileName1, strFileName2, <span class="stringliteral">&quot;Cyto2&quot;</span>, <span class="stringliteral">&quot;WaveCorrection&quot;</span>, setIgnore, 0.0001, 0.999999));
<a name="l00313"></a>00313                 <a class="code" href="classRegressionTest.html" title="Utility class for helping to do regression testing.">RegressionTest</a> test(name.c_str(), command.c_str(), checks);
<a name="l00314"></a>00314                 test.setSuite(*<span class="keyword">this</span>,  outdir, outdir + <span class="stringliteral">&quot;/apt-copynumber-cyto.log&quot;</span>, outdir + <span class="stringliteral">&quot;/valgrind.log&quot;</span>);
<a name="l00315"></a>00315                 bPassed = test.pass();
<a name="l00316"></a>00316                 numPassed += test.getNumPassed();
<a name="l00317"></a>00317                 numFailed += test.getNumFailed();
<a name="l00318"></a>00318                 <span class="keywordflow">if</span>(!bPassed) {
<a name="l00319"></a>00319                   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Error in: &quot;</span> + name + <span class="stringliteral">&quot;(): &quot;</span> + test.getErrorMsg());
<a name="l00320"></a>00320                 }
<a name="l00321"></a>00321         <span class="keywordflow">if</span> (bPassed)
<a name="l00322"></a>00322                   {
<a name="l00323"></a>00323 
<a name="l00324"></a>00324                     std::set&lt;std::string&gt; setSetIgnore;
<a name="l00325"></a>00325 <span class="preprocessor">#ifndef __LP64__</span>
<a name="l00326"></a>00326 <span class="preprocessor"></span>                    {
<a name="l00327"></a>00327                       std::string ignoreString = <span class="stringliteral">&quot;Segments.Mosaicism&quot;</span>;
<a name="l00328"></a>00328                       std::string ignoreString2 = <span class="stringliteral">&quot;ProbeSets.AllelePeaks&quot;</span>;
<a name="l00329"></a>00329                       setSetIgnore.insert(ignoreString);
<a name="l00330"></a>00330                       setSetIgnore.insert(ignoreString2);
<a name="l00331"></a>00331                     }
<a name="l00332"></a>00332 <span class="preprocessor">#endif</span>
<a name="l00333"></a>00333 <span class="preprocessor"></span><span class="preprocessor">#ifdef __APPLE__</span>
<a name="l00334"></a>00334 <span class="preprocessor"></span>                    {
<a name="l00335"></a>00335                       std::string ignoreString = <span class="stringliteral">&quot;Segments.Mosaicism&quot;</span>;
<a name="l00336"></a>00336                       std::string ignoreString2 = <span class="stringliteral">&quot;ProbeSets.AllelePeaks&quot;</span>;
<a name="l00337"></a>00337                       setSetIgnore.insert(ignoreString);
<a name="l00338"></a>00338                       setSetIgnore.insert(ignoreString2);
<a name="l00339"></a>00339                     }
<a name="l00340"></a>00340 <span class="preprocessor">#endif</span>
<a name="l00341"></a>00341 <span class="preprocessor"></span>                    std::string ignoreString = <span class="stringliteral">&quot;ProbeSets.AllelePeaks&quot;</span>;
<a name="l00342"></a>00342                     setSetIgnore.insert(ignoreString);
<a name="l00343"></a>00343 
<a name="l00344"></a>00344                     <span class="keywordtype">string</span> outdir2 = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(m_testDir, <span class="stringliteral">&quot;qt-batch2&quot;</span>);
<a name="l00345"></a>00345                     <span class="comment">//Fs::rm_rf(outdir2, false);</span>
<a name="l00346"></a>00346                     vector&lt;RegressionCheck *&gt; checks2;
<a name="l00347"></a>00347                     m_OutPrefix = <span class="stringliteral">&quot;qt-batch2&quot;</span>;
<a name="l00348"></a>00348                     m_OutModifier = <span class="stringliteral">&quot;ca&quot;</span>;
<a name="l00349"></a>00349                     m_SetSetIgnore = setSetIgnore;
<a name="l00350"></a>00350                     m_ToCompare =  vCytogenetics_Array_SixFiles;
<a name="l00351"></a>00351                     checks2.push_back(<span class="keyword">this</span>);
<a name="l00352"></a>00352                     <span class="keywordtype">string</span> name2 = <span class="stringliteral">&quot;qt-batch2&quot;</span>;
<a name="l00353"></a>00353                     <span class="keywordtype">string</span> command2 =
<a name="l00354"></a>00354                       string(<span class="stringliteral">&quot;./apt-copynumber-cyto&quot;</span>) +
<a name="l00355"></a>00355                       <span class="stringliteral">&quot; -v 3&quot;</span> +
<a name="l00356"></a>00356                       <span class="stringliteral">&quot; --cyto2 true&quot;</span> +
<a name="l00357"></a>00357                       <span class="stringliteral">&quot; --force true&quot;</span> +
<a name="l00358"></a>00358                       <span class="stringliteral">&quot; --text-output true&quot;</span> +
<a name="l00359"></a>00359                       <span class="stringliteral">&quot; --cnchp-output false&quot;</span> +
<a name="l00360"></a>00360                       <span class="stringliteral">&quot; --cychp-output true&quot;</span> +
<a name="l00361"></a>00361                       <span class="stringliteral">&quot; --set-analysis-name ca&quot;</span> +
<a name="l00362"></a>00362                       <span class="stringliteral">&quot; --reference-input &quot;</span> +
<a name="l00363"></a>00363                       m_testDir +
<a name="l00364"></a>00364                       <span class="stringliteral">&quot;/qt-batch/Cytogenetics_Array.10_cels.REF_MODEL&quot;</span> +
<a name="l00365"></a>00365                       <span class="stringliteral">&quot; --snp-reference-input-file &quot;</span> + m_testDir + <span class="stringliteral">&quot;/qt-batch/Cytogenetics_Array.10_cels.snpref.a5&quot;</span> +
<a name="l00366"></a>00366                       <span class="stringliteral">&quot; --spf-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.spf&quot;</span> +
<a name="l00367"></a>00367                       <span class="stringliteral">&quot; --chrX-probes &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.chrXprobes&quot;</span> +
<a name="l00368"></a>00368                       <span class="stringliteral">&quot; --chrY-probes &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.chrYprobes&quot;</span> +
<a name="l00369"></a>00369                       <span class="stringliteral">&quot; --annotation-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.na28.annot.db&quot;</span> +
<a name="l00370"></a>00370                       <span class="stringliteral">&quot; --probe-file &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.probe_tab&quot;</span> +
<a name="l00371"></a>00371                       <span class="stringliteral">&quot; --snp-qc-snp-list &quot;</span>+idata+<span class="stringliteral">&quot;/lib/Cytogenetics_Array/Cytogenetics_Array.snplist.txt&quot;</span> +
<a name="l00372"></a>00372                       <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-As_NA18547_A08_01_NN_20081218.CEL&quot;</span> +
<a name="l00373"></a>00373                       <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-As_NA18603_B03_01_NN_20081218.CEL&quot;</span> +
<a name="l00374"></a>00374                       <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-As_NA18971_C02_01_NN_20081218.CEL&quot;</span> +
<a name="l00375"></a>00375                       <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-Cs_NA10857_A10_01_NN_20081218.CEL&quot;</span> +
<a name="l00376"></a>00376                       <span class="stringliteral">&quot; --cels &quot;</span>+idata+<span class="stringliteral">&quot;/cel/Cytogenetics_Array/HapMap-Cs_NA12003_D04_01_NN_20081218.CEL&quot;</span> +
<a name="l00377"></a>00377                       <span class="stringliteral">&quot; --out-dir &quot;</span> + outdir2 +
<a name="l00378"></a>00378                       <span class="stringliteral">&quot; --male-gender-ratio-cutoff=1.3&quot;</span> +
<a name="l00379"></a>00379                       <span class="stringliteral">&quot; --female-gender-ratio-cutoff=1.0&quot;</span> +
<a name="l00380"></a>00380                       <span class="stringliteral">&quot; --xx-cutoff=0.8&quot;</span> +
<a name="l00381"></a>00381                       <span class="stringliteral">&quot; --xx-cutoff-high=1.07&quot;</span> +
<a name="l00382"></a>00382                       <span class="stringliteral">&quot; --y-cutoff=0.65&quot;</span> +
<a name="l00383"></a>00383                       <span class="stringliteral">&quot; --analysis log2-ratio-cyto2.gc-correction=true.median-autosome-median-normalization=true.median-smooth-marker-count=5.trim-high=2.0.trim-low=-2.5&quot;</span> +
<a name="l00384"></a>00384                       <span class="stringliteral">&quot; --analysis kernel-smooth.sigma_span=50&quot;</span> +
<a name="l00385"></a>00385                       <span class="stringliteral">&quot; --analysis cn-cyto2.diagonal-weight=0.995.mapd-weight=0.22.min-segment-size=5.hmm-confidence-weight=0.6&quot;</span> +
<a name="l00386"></a>00386                       <span class="stringliteral">&quot; --analysis cn-cyto2-gender.cutoff=0.5&quot;</span> +
<a name="l00387"></a>00387                       <span class="stringliteral">&quot; --analysis cn-segment&quot;</span> +
<a name="l00388"></a>00388                       <span class="stringliteral">&quot; --analysis loh-cyto2.lohCNSegSeparation=1000000.minInformation=100.lambdaCritical=8.0&quot;</span> +
<a name="l00389"></a>00389                       <span class="stringliteral">&quot; --analysis mosaicism.marker-bandwidth=6000.confidence-window=251&quot;</span> +
<a name="l00390"></a>00390                       <span class="stringliteral">&quot; --analysis allele-peaks.step=100.window=250.point-count=128.bandwidth=0.45.cutoff=0.05&quot;</span> +
<a name="l00391"></a>00391                       <span class="stringliteral">&quot; --local-gc-background-correction-reference-method pdnn-reference-method&quot;</span> +
<a name="l00392"></a>00392                       <span class="stringliteral">&quot; --wave-correction-reference-method wave-correction-reference-method.trim=2.0.percentile=0.75.demean=false&quot;</span> +
<a name="l00393"></a>00393                       <span class="stringliteral">&quot; --local-gc-background-intensity-adjustment-method pdnn-intensity-adjustment-method.predicted-intensity-bin-count=20.gc-bin-count=20.residual-trim=2.0&quot;</span> +
<a name="l00394"></a>00394                       <span class="stringliteral">&quot; --image-correction-intensity-adjustment-method high-pass-filter-intensity-adjustment-method.data-block-rows=320.data-block-cols=2015.mini-block-rows=8.mini-block-cols=8.global-smooth-weight=256.local-smooth-weight=64.converged=0.0001&quot;</span> +
<a name="l00395"></a>00395                       <span class="stringliteral">&quot; --wave-correction-log2ratio-adjustment-method wave-correction-log2ratio-adjustment-method.bandwidth=101.bin-count=25.wave-count=3 &quot;</span> +
<a name="l00396"></a>00396                       <span class="stringliteral">&quot; --reference-chromosome=22&quot;</span>;
<a name="l00397"></a>00397 
<a name="l00398"></a>00398                     <a class="code" href="classRegressionTest.html" title="Utility class for helping to do regression testing.">RegressionTest</a> test2(name2.c_str(), command2.c_str(), checks2, <span class="keyword">false</span>);
<a name="l00399"></a>00399                     test2.setSuite(*<span class="keyword">this</span>,  outdir2, outdir2 + <span class="stringliteral">&quot;/apt-copynumber-cyto.log&quot;</span>, outdir2 + <span class="stringliteral">&quot;/valgrind.batch2.log&quot;</span>);
<a name="l00400"></a>00400                     <span class="keywordflow">if</span>(!test2.pass()) {
<a name="l00401"></a>00401                       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Error in: &quot;</span> + name + <span class="stringliteral">&quot;(): &quot;</span> + test.getErrorMsg());
<a name="l00402"></a>00402                     }
<a name="l00403"></a>00403                   }
<a name="l00404"></a>00404         <span class="keywordflow">else</span> {numFailed++;}
<a name="l00405"></a>00405   }
<a name="l00406"></a>00406 };
<a name="l00407"></a>00407 
<a name="l00408"></a>00408 <span class="keywordtype">int</span> <a class="code" href="canary-test_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="Everybody&amp;#39;s favorite function.">main</a>(<span class="keywordtype">int</span> argc, <span class="keywordtype">char</span>* argv[]) {
<a name="l00409"></a>00409   <span class="keywordflow">try</span> {
<a name="l00410"></a>00410 
<a name="l00411"></a>00411     <span class="comment">// Quick test uses ~ 100MB of disk space. </span>
<a name="l00412"></a>00412     <a class="code" href="classFsTestDir.html">FsTestDir</a> testDir;
<a name="l00413"></a>00413     testDir.<a class="code" href="classFsTestDir.html#ac14fb44eb179d1094ef5c9d127a39bbb">setTestDir</a>(<span class="stringliteral">&quot;copynumber/copynumber-cyto-qt&quot;</span>, <span class="keyword">true</span>);
<a name="l00414"></a>00414 
<a name="l00415"></a>00415     ofstream logOut;
<a name="l00416"></a>00416     <span class="keywordtype">string</span> logName;
<a name="l00417"></a>00417     logName = <a class="code" href="classFs.html#af44351abee4cbc47f94b83cb2f62899f" title="join strings with the path sep. /// while we would want people to keep the path with FsPath...">Fs::join</a>(testDir.asString(), <span class="stringliteral">&quot;qt-test-regression-copynumber-cyto-quick.log&quot;</span>);
<a name="l00418"></a>00418     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Log file: &quot;</span> + logName);
<a name="l00419"></a>00419     <a class="code" href="classFs.html#a1cc6e82074c84dd92bd7cdcf7d374b7c" title="Open an ofstream for writing to.">Fs::mustOpenToWrite</a>(logOut, logName.c_str());
<a name="l00420"></a>00420     <a class="code" href="classLogStream.html" title="Log messages with timestamps.">LogStream</a> log(3, &amp;logOut);
<a name="l00421"></a>00421 
<a name="l00422"></a>00422     Verbose::pushMsgHandler(&amp;log);
<a name="l00423"></a>00423     <a class="code" href="classVerbose.html#af20e052bc19f95d3a949d1daa1647fa8" title="Functions to add and remove handlers for communcation functions.">Verbose::pushProgressHandler</a>(&amp;log);
<a name="l00424"></a>00424     Verbose::pushWarnHandler(&amp;log);
<a name="l00425"></a>00425 
<a name="l00426"></a>00426     <a class="code" href="classVerbose.html#a416deece7d5098dc2e8759ef5fb134e3" title="Set the level of verbosity desired.">Verbose::setLevel</a>(3);
<a name="l00427"></a>00427     <a class="code" href="classtest__regression__copynumber__cyto__quick.html">test_regression_copynumber_cyto_quick</a> test;
<a name="l00428"></a>00428     test.m_testDir = testDir.asString();
<a name="l00429"></a>00429     test.parseArgv(argv);
<a name="l00430"></a>00430     test.doValgrind();
<a name="l00431"></a>00431     <span class="keywordtype">string</span> database = test.getDatabase();
<a name="l00432"></a>00432 
<a name="l00433"></a>00433     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;Execute regression test: test-regression-copynumber-cyto-quick&quot;</span>);
<a name="l00434"></a>00434 
<a name="l00435"></a>00435     test.run();
<a name="l00436"></a>00436 
<a name="l00437"></a>00437     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1, <span class="stringliteral">&quot;NumPassed: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(test.numPassed) + <span class="stringliteral">&quot; NumFailed: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(test.numFailed) + <span class="stringliteral">&quot; for test-regression-copynumber-cyto-quick&quot;</span>);
<a name="l00438"></a>00438     logOut.close();
<a name="l00439"></a>00439     <span class="keywordflow">return</span> test.numFailed != 0;
<a name="l00440"></a>00440   }
<a name="l00441"></a>00441   <span class="keywordflow">catch</span>(...) {
<a name="l00442"></a>00442       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(1,<span class="stringliteral">&quot;Unexpected Error: uncaught exception.&quot;</span>);
<a name="l00443"></a>00443       <span class="keywordflow">return</span> 1;
<a name="l00444"></a>00444   }
<a name="l00445"></a>00445   <span class="keywordflow">return</span> 1;
<a name="l00446"></a>00446 }
<a name="l00447"></a>00447 
<a name="l00448"></a>00448 ;
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:58:54 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
