<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: translation/TranslationTable.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_b6d6d4a2e58a0f89851f849ab50b1d34.html">translation</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>TranslationTable.cpp</h1>  </div>
</div>
<div class="contents">
<a href="TranslationTable_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2008 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment">/**</span>
<a name="l00022"></a>00022 <span class="comment"> * @file   TranslationTable.cpp</span>
<a name="l00023"></a>00023 <span class="comment"> * @author Mybrid Spalding</span>
<a name="l00024"></a>00024 <span class="comment"> * @date   Tue Mar 18 09:02:01 PDT 2008</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief  The primary business object class that manipulates the various input data models,</span>
<a name="l00026"></a>00026 <span class="comment"> *         creating other business objects appropriate for the translation algorithm and invokes translation. </span>
<a name="l00027"></a>00027 <span class="comment"> */</span>
<a name="l00028"></a>00028 
<a name="l00029"></a>00029 
<a name="l00030"></a>00030 <span class="preprocessor">#include &quot;<a class="code" href="TranslationTable_8h.html" title="The primary business object class that manipulates the various input data models, creating other busi...">translation/TranslationTable.h</a>&quot;</span>
<a name="l00031"></a>00031 <span class="comment">//</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="CopyNumberTableModel_8h.html" title="DEPRECATED DMET2 ONLY: Class for wrapping the DMET2 copy number report TSV file.">translation/CopyNumberTableModel.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="preprocessor">#include &quot;<a class="code" href="ExperimentGeneResults_8h.html" title="CallResults wrapper class for all gene translated experiment results to be consumed by various report...">translation/ExperimentGeneResults.h</a>&quot;</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="GeneCall_8h.html" title="Business object that is the heart of the matter, class with the algorithm that translates experiment ...">translation/GeneCall.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="GenotypeTableModel_8h.html" title="Genotype Data (CHP) class.">translation/GenotypeTableModel.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;<a class="code" href="TranslationTableModel_8h.html" title="Single instance data model class that wraps the translation file in the TranslationInputTsvModel.">translation/TranslationTableModel.h</a>&quot;</span>
<a name="l00037"></a>00037 <span class="comment">//</span>
<a name="l00038"></a>00038 <span class="preprocessor">#include &quot;<a class="code" href="Err_8h.html" title="Some utility error functions, currently prints message and aborts().">util/Err.h</a>&quot;</span> <span class="comment">// includes Verbose.h</span>
<a name="l00039"></a>00039 <span class="comment">//</span>
<a name="l00040"></a>00040 
<a name="l00041"></a>00041 <span class="keyword">using namespace </span>std;
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00044"></a>00044 <span class="comment">/**</span>
<a name="l00045"></a>00045 <span class="comment"> * _errorMessageDuplicateProbeSets</span>
<a name="l00046"></a>00046 <span class="comment"> * Synopsis:</span>
<a name="l00047"></a>00047 <span class="comment"> * Helper error message convienance function to make the code easier to read.</span>
<a name="l00048"></a>00048 <span class="comment"> *</span>
<a name="l00049"></a>00049 <span class="comment"> * @param rte - RunTimeEnviornment</span>
<a name="l00050"></a>00050 <span class="comment"> * @param ttm - TranslationTableModel.</span>
<a name="l00051"></a>00051 <span class="comment"> * @param row - the row in question, 0 based.</span>
<a name="l00052"></a>00052 <span class="comment"> *</span>
<a name="l00053"></a>00053 <span class="comment"> * @return false - indicates it is not ok to perform analysis.</span>
<a name="l00054"></a>00054 <span class="comment"> */</span>
<a name="l00055"></a>00055 <span class="comment">/*****************************************************************************/</span>
<a name="l00056"></a>00056 <span class="keyword">static</span> <span class="keywordtype">bool</span> _errorMessageDuplicateProbeSets(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00057"></a>00057     <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm,
<a name="l00058"></a>00058     <span class="keyword">const</span> <span class="keywordtype">int</span> row)
<a name="l00059"></a>00059 {
<a name="l00060"></a>00060 
<a name="l00061"></a>00061 
<a name="l00062"></a>00062   <span class="keywordtype">bool</span> okToContinue = <span class="keyword">false</span>;
<a name="l00063"></a>00063   std::stringstream msgSStr;
<a name="l00064"></a>00064   msgSStr &lt;&lt; rte.m_adtOpts.m_inputGenoFile &lt;&lt; <span class="stringliteral">&quot;: invalid ProbeSet with Allele [A1-A29] or Reference column for row: &quot;</span> &lt;&lt; endl;
<a name="l00065"></a>00065   std::string rowAsString = ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ac209f9165791ba0d30aa905965eba1f6" title="TranslationInputTsvTableModel::getRowAsString Synopsis:">getRowAsString</a>(row);
<a name="l00066"></a>00066   msgSStr &lt;&lt; <span class="stringliteral">&quot;gene row [&quot;</span> &lt;&lt; row &lt;&lt; <span class="stringliteral">&quot;]: &quot;</span> &lt;&lt; rowAsString &lt;&lt; endl;
<a name="l00067"></a>00067   <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(ADT_VERBOSE_EXCEPTION, msgSStr.str());
<a name="l00068"></a>00068 
<a name="l00069"></a>00069   <span class="keywordflow">return</span> okToContinue;
<a name="l00070"></a>00070 
<a name="l00071"></a>00071 }
<a name="l00072"></a>00072 <span class="comment">// end TranslationTable::_errorMessageDuplicateProbeSets</span>
<a name="l00073"></a>00073 <span class="comment">/*****************************************************************************/</span>
<a name="l00074"></a>00074 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00075"></a>00075 <span class="comment">/**</span>
<a name="l00076"></a>00076 <span class="comment"> * _errorMessageGeneRowOrder</span>
<a name="l00077"></a>00077 <span class="comment"> * Synopsis:</span>
<a name="l00078"></a>00078 <span class="comment"> * Helper error message convienance function to make the code easier to read.</span>
<a name="l00079"></a>00079 <span class="comment"> *</span>
<a name="l00080"></a>00080 <span class="comment"> * @param rte - RunTimeEnviornment</span>
<a name="l00081"></a>00081 <span class="comment"> * @param ttm - TranslationTableModel.</span>
<a name="l00082"></a>00082 <span class="comment"> * @param row - the row in question, 0 based.</span>
<a name="l00083"></a>00083 <span class="comment"> *</span>
<a name="l00084"></a>00084 <span class="comment"> * @return false - indicates it is not ok to perform analysis.</span>
<a name="l00085"></a>00085 <span class="comment"> */</span>
<a name="l00086"></a>00086 <span class="comment">/*****************************************************************************/</span>
<a name="l00087"></a>00087 <span class="keyword">static</span> <span class="keywordtype">bool</span> _errorMessageGeneRowOrder(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00088"></a>00088     <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm,
<a name="l00089"></a>00089     <span class="keyword">const</span> <span class="keywordtype">int</span> row,
<a name="l00090"></a>00090     <span class="keyword">const</span> std::string &amp; geneInProcess)
<a name="l00091"></a>00091 {
<a name="l00092"></a>00092 
<a name="l00093"></a>00093   <span class="keywordtype">bool</span> okToContinue = <span class="keyword">false</span>;
<a name="l00094"></a>00094   std::stringstream msgSStr;
<a name="l00095"></a>00095   msgSStr &lt;&lt; rte.m_adtOpts.m_inputGenoFile &lt;&lt; <span class="stringliteral">&quot;: invalid format, gene row is not under the correct gene header: &quot;</span> &lt;&lt; endl;
<a name="l00096"></a>00096   std::string rowAsString = ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ac209f9165791ba0d30aa905965eba1f6" title="TranslationInputTsvTableModel::getRowAsString Synopsis:">getRowAsString</a>(row);
<a name="l00097"></a>00097   msgSStr &lt;&lt; <span class="stringliteral">&quot;gene row [&quot;</span> &lt;&lt; row &lt;&lt; <span class="stringliteral">&quot;]: &quot;</span> &lt;&lt; rowAsString &lt;&lt; endl;
<a name="l00098"></a>00098   msgSStr &lt;&lt; <span class="stringliteral">&quot;gene header : &quot;</span>  &lt;&lt; geneInProcess ;
<a name="l00099"></a>00099   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES,  msgSStr.str());
<a name="l00100"></a>00100 
<a name="l00101"></a>00101   <span class="keywordflow">return</span> okToContinue;
<a name="l00102"></a>00102 
<a name="l00103"></a>00103 }
<a name="l00104"></a>00104 <span class="comment">// end _errorMessageGeneRowOrder</span>
<a name="l00105"></a>00105 <span class="comment">/*****************************************************************************/</span>
<a name="l00106"></a>00106 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00107"></a>00107 <span class="comment">/**</span>
<a name="l00108"></a>00108 <span class="comment"> * TranslationTable::TranslationTable</span>
<a name="l00109"></a>00109 <span class="comment"> * Synopsis:</span>
<a name="l00110"></a>00110 <span class="comment"> * Main constructor.</span>
<a name="l00111"></a>00111 <span class="comment"> * Truth be told, this object should really be refactored to</span>
<a name="l00112"></a>00112 <span class="comment"> * just &quot;Translation&quot; and not &quot;TranslationTable&quot;. There was a time where</span>
<a name="l00113"></a>00113 <span class="comment"> * object was closely coupled with the translation table but no more.</span>
<a name="l00114"></a>00114 <span class="comment"> *</span>
<a name="l00115"></a>00115 <span class="comment"> * This object is primarily comprised of a hash of gene call business objects</span>
<a name="l00116"></a>00116 <span class="comment"> * derived from the translation table. </span>
<a name="l00117"></a>00117 <span class="comment"> * The key is the gene name. The value is a std::vector of GenCalls corresponding</span>
<a name="l00118"></a>00118 <span class="comment"> * to the &quot;Reference&quot;, &quot;Variant&quot;, and &quot;A1&quot; ... &quot;AN&quot; columns of the</span>
<a name="l00119"></a>00119 <span class="comment"> * translation table. Non-haplotype markers will only have GeneCalls</span>
<a name="l00120"></a>00120 <span class="comment"> * that correspond to the &quot;Reference&quot; and &quot;Variant&quot; columns. </span>
<a name="l00121"></a>00121 <span class="comment"> * There will be a GeneCall per each non-haplotype marker plus one for the</span>
<a name="l00122"></a>00122 <span class="comment"> * single possible Haplotype set of markers.</span>
<a name="l00123"></a>00123 <span class="comment"> *</span>
<a name="l00124"></a>00124 <span class="comment"> * During translation, each GeneCall will be attempted to match</span>
<a name="l00125"></a>00125 <span class="comment"> * the CHP experiment data and if a match is find then the underlying</span>
<a name="l00126"></a>00126 <span class="comment"> * CallSet contained within the GeneCall will be attached to</span>
<a name="l00127"></a>00127 <span class="comment"> * the CallResults object. </span>
<a name="l00128"></a>00128 <span class="comment"> *</span>
<a name="l00129"></a>00129 <span class="comment"> * 1.) GeneCall with CallSets for making translation calls. </span>
<a name="l00130"></a>00130 <span class="comment"> * 2.) Zero Copy Number CallSets per gene. </span>
<a name="l00131"></a>00131 <span class="comment"> *</span>
<a name="l00132"></a>00132 <span class="comment"> * @param rte - the single instance run time environment</span>
<a name="l00133"></a>00133 <span class="comment"> * @param ttm - the single instance TranslationTableModel </span>
<a name="l00134"></a>00134 <span class="comment"> * @param cntm - CopyNumberTableModel which carries the override alleles.</span>
<a name="l00135"></a>00135 <span class="comment"> */</span>
<a name="l00136"></a>00136 <span class="comment">/*****************************************************************************/</span>
<a name="l00137"></a><a class="code" href="classTranslationTable.html#aa9cc2dd171c88126df9139955c1391d5">00137</a> <a class="code" href="classTranslationTable.html#aa9cc2dd171c88126df9139955c1391d5" title="TranslationTable::TranslationTable Synopsis: Main constructor.">TranslationTable::TranslationTable</a>(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00138"></a>00138                                    <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm,
<a name="l00139"></a>00139                                    <span class="keyword">const</span> <a class="code" href="classCopyNumberTableModel.html">CopyNumberTableModel</a> &amp; cntm)
<a name="l00140"></a>00140 {
<a name="l00141"></a>00141 
<a name="l00142"></a>00142 
<a name="l00143"></a>00143   std::string    geneInProcess;
<a name="l00144"></a>00144   std::string    priorGene;
<a name="l00145"></a>00145   <span class="keywordtype">int</span>       geneHeaderRow           = 0;
<a name="l00146"></a>00146   <a class="code" href="classGeneCall.html">GeneCall</a> *nextMarkerGeneCall      = NULL;
<a name="l00147"></a>00147   <a class="code" href="classGeneCall.html">GeneCall</a> *nextHaplotypeGeneCall   = NULL;
<a name="l00148"></a>00148   <span class="keywordtype">bool</span>      okToContinue            = <span class="keyword">true</span>;
<a name="l00149"></a>00149   <span class="keywordtype">bool</span>      rowIsHaplotype          = <span class="keyword">false</span>;
<a name="l00150"></a>00150   <span class="keywordtype">int</span>       haplotypeCount          = 0;
<a name="l00151"></a>00151 
<a name="l00152"></a>00152   <span class="comment">//Profile profile1;</span>
<a name="l00153"></a>00153   <span class="keywordtype">int</span> numRows = ttm.size();
<a name="l00154"></a>00154 
<a name="l00155"></a>00155   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> row = 0; (row &lt; numRows) &amp;&amp; okToContinue ;  row++) {
<a name="l00156"></a>00156 
<a name="l00157"></a>00157     std::string rowAsString = ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ac209f9165791ba0d30aa905965eba1f6" title="TranslationInputTsvTableModel::getRowAsString Synopsis:">getRowAsString</a>(row);
<a name="l00158"></a>00158 
<a name="l00159"></a>00159     <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#a590375d07f5f5bf813275858d0466114" title="TranslationTableModel::ignoreRow Synopsis: A setter and getter for ignoring rows in the translation t...">ignoreRow</a>(row)) {
<a name="l00160"></a>00160       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <span class="stringliteral">&quot;IGNORING: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(row + 2) + <span class="stringliteral">&quot;:&quot;</span> + rowAsString);
<a name="l00161"></a>00161       <span class="keywordflow">continue</span>;
<a name="l00162"></a>00162     } <span class="keywordflow">else</span> {
<a name="l00163"></a>00163       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <span class="stringliteral">&quot;PROCESSING: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(row + 2) + <span class="stringliteral">&quot;:&quot;</span> + rowAsString);
<a name="l00164"></a>00164     }
<a name="l00165"></a>00165 
<a name="l00166"></a>00166     std::string gene = ttm.m_rows[row][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_GENE)];
<a name="l00167"></a>00167     std::string probeSet = ttm.m_rows[row][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_PROBE_SET_ID)];
<a name="l00168"></a>00168 
<a name="l00169"></a>00169 
<a name="l00170"></a>00170     <span class="keywordflow">if</span> (gene != geneInProcess)  {
<a name="l00171"></a>00171       <span class="comment">// Starting a new GENE.</span>
<a name="l00172"></a>00172 
<a name="l00173"></a>00173       <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#acf1658f2cce028d22fa3f8666682acde" title="TranslationTableModel::validateRowIsHeader Synopsis:">validateRowIsHeader</a>(row)) {
<a name="l00174"></a>00174 
<a name="l00175"></a>00175         m_geneGroupCompleteCallSet[gene] = <a class="code" href="classCallSet.html">CallSet</a>(gene + <span class="stringliteral">&quot;:CompleteSet&quot;</span>);
<a name="l00176"></a>00176         priorGene = geneInProcess;
<a name="l00177"></a>00177         geneInProcess = gene;
<a name="l00178"></a>00178         rowIsHaplotype = <span class="keyword">false</span>;
<a name="l00179"></a>00179         geneHeaderRow = row;
<a name="l00180"></a>00180         haplotypeCount = 0;
<a name="l00181"></a>00181 
<a name="l00182"></a>00182         <span class="keywordflow">if</span> (nextHaplotypeGeneCall != NULL) {
<a name="l00183"></a>00183           nextHaplotypeGeneCall =
<a name="l00184"></a>00184             _addGeneCallToGroup(rte, priorGene, nextHaplotypeGeneCall);
<a name="l00185"></a>00185 
<a name="l00186"></a>00186         }
<a name="l00187"></a>00187 
<a name="l00188"></a>00188         <span class="comment">// Copy Number is hard code for DMET2</span>
<a name="l00189"></a>00189         <span class="comment">// Copy Number is data driven for DMET3</span>
<a name="l00190"></a>00190         <span class="keywordflow">if</span> (rte.m_adtOpts.m_dmet2Calling) {
<a name="l00191"></a>00191           <span class="keywordflow">if</span> (<a class="code" href="CopyNumberTableModel_8cpp.html#a0149ddb188582ad1c76c785e4ee6669b" title="isGeneCopyNumberSensitive Synopsis:">isGeneCopyNumberSensitive</a>(gene)) {
<a name="l00192"></a>00192             std::map&lt;std::string, std::string&gt;::const_iterator itSS;
<a name="l00193"></a>00193 
<a name="l00194"></a>00194             <span class="keywordflow">for</span> (itSS =  cntm.m_experimentCopyNumberCall.begin();
<a name="l00195"></a>00195                  itSS !=  cntm.m_experimentCopyNumberCall.end(); itSS++) {
<a name="l00196"></a>00196               std::string geneExperiment = gene + <span class="stringliteral">&quot;:&quot;</span> + itSS-&gt;first;
<a name="l00197"></a>00197               m_geneExperimentCopyNumberCall[geneExperiment] = itSS-&gt;second;
<a name="l00198"></a>00198             }
<a name="l00199"></a>00199           }
<a name="l00200"></a>00200         }
<a name="l00201"></a>00201       } <span class="keywordflow">else</span> {
<a name="l00202"></a>00202         okToContinue
<a name="l00203"></a>00203         = _errorMessageGeneRowOrder(rte, ttm, row, geneInProcess) ;
<a name="l00204"></a>00204       }
<a name="l00205"></a>00205 
<a name="l00206"></a>00206       <span class="keywordflow">continue</span>;
<a name="l00207"></a>00207 
<a name="l00208"></a>00208     } <span class="comment">// if starting new gene and encounted the gene header row</span>
<a name="l00209"></a>00209 
<a name="l00210"></a>00210     rowIsHaplotype = ttm.<a class="code" href="classTranslationTableModel.html#af313345b99cd8a53fa1efd617d671be2" title="TranslationTableModel::isRowHaplotype Synopsis:">isRowHaplotype</a>(row);
<a name="l00211"></a>00211 
<a name="l00212"></a>00212     <span class="comment">// We currently only support one Haplotype per gene.</span>
<a name="l00213"></a>00213     <span class="comment">// The haplotype rows may be out of order so just continue to append</span>
<a name="l00214"></a>00214     <span class="comment">// the current haplotype gene call.</span>
<a name="l00215"></a>00215     <span class="keywordflow">if</span> (rowIsHaplotype &amp;&amp; ! nextHaplotypeGeneCall) {
<a name="l00216"></a>00216       nextHaplotypeGeneCall = <span class="keyword">new</span> <a class="code" href="classGeneCall.html">GeneCall</a>(rte, ttm, geneHeaderRow, rowIsHaplotype);
<a name="l00217"></a>00217     }
<a name="l00218"></a>00218 
<a name="l00219"></a>00219     <span class="comment">// For multi-allelic probeSets then the nextMarkerGeneCall is the</span>
<a name="l00220"></a>00220     <span class="comment">// previous version.</span>
<a name="l00221"></a>00221     <span class="keywordtype">bool</span> okToAddMarkerGeneCall = <span class="keyword">true</span>;
<a name="l00222"></a>00222     <span class="keywordflow">if</span> ((nextMarkerGeneCall = _findMarkerGeneCall(geneInProcess,  probeSet))  != NULL) {
<a name="l00223"></a>00223 
<a name="l00224"></a>00224       okToAddMarkerGeneCall = <span class="keyword">false</span>;
<a name="l00225"></a>00225 
<a name="l00226"></a>00226     } <span class="keywordflow">else</span> {
<a name="l00227"></a>00227       nextMarkerGeneCall = <span class="keyword">new</span> <a class="code" href="classGeneCall.html">GeneCall</a>(rte, ttm, geneHeaderRow, <span class="keyword">false</span>);
<a name="l00228"></a>00228     }
<a name="l00229"></a>00229 
<a name="l00230"></a>00230     <span class="keywordtype">bool</span> okCallElement = <span class="keyword">true</span>;
<a name="l00231"></a>00231     <span class="comment">// Add a marker GeneCall for both Haplotype and Marker.</span>
<a name="l00232"></a>00232     <span class="keywordtype">int</span> allMarkerAlleleIndex[] = { ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_REFERENCE), ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_VARIANT) };
<a name="l00233"></a>00233 
<a name="l00234"></a>00234     <span class="comment">// Make sure the reference and variant are not the same.</span>
<a name="l00235"></a>00235     std::string ref = ttm.m_rows[row][allMarkerAlleleIndex[0]];
<a name="l00236"></a>00236     std::string var = ttm.m_rows[row][allMarkerAlleleIndex[1]];
<a name="l00237"></a>00237 
<a name="l00238"></a>00238     <span class="keywordflow">if</span> (ref == var) {
<a name="l00239"></a>00239       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(gene + <span class="stringliteral">&quot; &quot;</span> + probeSet + <span class="stringliteral">&quot;: translation file reference and variant base are identical.&quot;</span>);
<a name="l00240"></a>00240     }
<a name="l00241"></a>00241     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; (i &lt; 2) &amp;&amp; okCallElement; i++) {
<a name="l00242"></a>00242 
<a name="l00243"></a>00243       okCallElement = okCallElement &amp;&amp; nextMarkerGeneCall-&gt;
<a name="l00244"></a>00244                       addCallElementToAlleleCallSet(ttm, geneHeaderRow, row, allMarkerAlleleIndex[i], <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00245"></a>00245 
<a name="l00246"></a>00246       m_geneGroupCompleteCallSet[gene].
<a name="l00247"></a>00247       addCallElement(ttm, row, allMarkerAlleleIndex[i], <span class="keyword">true</span>);
<a name="l00248"></a>00248 
<a name="l00249"></a>00249     }
<a name="l00250"></a>00250 
<a name="l00251"></a>00251     <span class="keywordflow">if</span> (rowIsHaplotype) {
<a name="l00252"></a>00252       <span class="comment">// Add to Allele call sets</span>
<a name="l00253"></a>00253       std::vector&lt;int&gt; alleleColumns = ttm.<a class="code" href="classTranslationTableModel.html#af072cea2e362c21c8433939b9f6bde66" title="TranslationTableModel::getHaplotypeAlleleColumns Synopsis:">getHaplotypeAlleleColumns</a>(geneHeaderRow);
<a name="l00254"></a>00254       <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; okCallElement &amp;&amp; (j &lt; alleleColumns.size()); j++) {
<a name="l00255"></a>00255         <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(nextHaplotypeGeneCall, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00256"></a>00256 
<a name="l00257"></a>00257         <span class="keywordflow">if</span> (ttm.m_rows[row][alleleColumns[j]].empty()) {
<a name="l00258"></a>00258           <span class="keywordflow">continue</span>;
<a name="l00259"></a>00259         }
<a name="l00260"></a>00260 
<a name="l00261"></a>00261         okCallElement = okCallElement &amp;&amp;
<a name="l00262"></a>00262                         nextHaplotypeGeneCall-&gt;
<a name="l00263"></a>00263                         addCallElementToAlleleCallSet(ttm, geneHeaderRow, row, alleleColumns[j], <span class="keyword">true</span>, <span class="keyword">true</span>);
<a name="l00264"></a>00264         m_geneGroupCompleteCallSet[gene].
<a name="l00265"></a>00265         addCallElement(ttm, row, alleleColumns[j], <span class="keyword">true</span>);
<a name="l00266"></a>00266       }
<a name="l00267"></a>00267     } <span class="keywordflow">else</span> <span class="keywordflow">if</span> ( ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3 ){
<a name="l00268"></a>00268       <span class="comment">// For non-haplotype markers the copy number base needs to be added.</span>
<a name="l00269"></a>00269       <span class="keywordtype">int</span> copyNumberColumn =  ttm.<a class="code" href="classTranslationTableModel.html#a9748ec04f92938fdd4e04bf685d86186" title="TranslationTableModel::getCopyNumberColumn: Synopsis:">getCopyNumberColumn</a>(row);
<a name="l00270"></a>00270       <span class="keywordflow">if</span> (copyNumberColumn &gt; -1) {
<a name="l00271"></a>00271 
<a name="l00272"></a>00272         okCallElement = okCallElement &amp;&amp; nextMarkerGeneCall-&gt;
<a name="l00273"></a>00273                         addCallElementToAlleleCallSet(ttm, geneHeaderRow, row, copyNumberColumn, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00274"></a>00274 
<a name="l00275"></a>00275         m_geneGroupCompleteCallSet[gene].
<a name="l00276"></a>00276         addCallElement(ttm, row, copyNumberColumn, <span class="keyword">true</span>);
<a name="l00277"></a>00277 
<a name="l00278"></a>00278       }
<a name="l00279"></a>00279     }
<a name="l00280"></a>00280 
<a name="l00281"></a>00281     <span class="comment">// Add a single marker call</span>
<a name="l00282"></a>00282     <span class="keywordflow">if</span> (okToAddMarkerGeneCall &amp;&amp; okCallElement) {
<a name="l00283"></a>00283       nextMarkerGeneCall = _addGeneCallToGroup(rte, gene, nextMarkerGeneCall);
<a name="l00284"></a>00284     } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (!okCallElement) {
<a name="l00285"></a>00285       okToContinue = _errorMessageDuplicateProbeSets(rte, ttm, row);
<a name="l00286"></a>00286     }
<a name="l00287"></a>00287 
<a name="l00288"></a>00288   } <span class="comment">// for each row</span>
<a name="l00289"></a>00289 
<a name="l00290"></a>00290   <span class="keywordflow">if</span> (okToContinue &amp;&amp; (nextHaplotypeGeneCall != NULL)) {
<a name="l00291"></a>00291     nextHaplotypeGeneCall =
<a name="l00292"></a>00292       _addGeneCallToGroup(rte, geneInProcess, nextHaplotypeGeneCall);
<a name="l00293"></a>00293 
<a name="l00294"></a>00294   }
<a name="l00295"></a>00295   std::map&lt;std::string, CallSet&gt;::iterator itSCS;
<a name="l00296"></a>00296   itSCS = m_geneGroupCompleteCallSet.begin();
<a name="l00297"></a>00297 
<a name="l00298"></a>00298   <span class="keywordflow">for</span> ( ; itSCS != m_geneGroupCompleteCallSet.end(); itSCS++ ) {
<a name="l00299"></a>00299     std::map&lt;std::string, CallElement &gt;::iterator itCE;
<a name="l00300"></a>00300     <span class="keywordflow">for</span> (  itCE = itSCS-&gt;second.m_ceSet.begin();
<a name="l00301"></a>00301            itCE != itSCS-&gt;second.m_ceSet.end();   itCE++ ) {
<a name="l00302"></a>00302       itCE-&gt;second.setCompleteSetSize();
<a name="l00303"></a>00303     }
<a name="l00304"></a>00304   }
<a name="l00305"></a>00305 
<a name="l00306"></a>00306   <span class="keywordflow">if</span> (ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3) {
<a name="l00307"></a>00307     okToContinue = okToContinue &amp;&amp; _duplicateCallSetCheck(rte);
<a name="l00308"></a>00308     okToContinue = okToContinue &amp;&amp; _reconcileCopyNumberCallSet(rte, ttm);
<a name="l00309"></a>00309   }
<a name="l00310"></a>00310 
<a name="l00311"></a>00311   <span class="keywordflow">if</span> (!okToContinue) {
<a name="l00312"></a>00312     std::stringstream msgSStr;
<a name="l00313"></a>00313     msgSStr &lt;&lt; rte.m_adtOpts.m_inputGenoFile ;
<a name="l00314"></a>00314     msgSStr &lt;&lt; <span class="stringliteral">&quot;: business logic errors detected.&quot;</span>;
<a name="l00315"></a>00315     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(msgSStr.str());
<a name="l00316"></a>00316   }
<a name="l00317"></a>00317 
<a name="l00318"></a>00318 
<a name="l00319"></a>00319 }
<a name="l00320"></a>00320 <span class="comment">// end TranslationTable::TranslationTable</span>
<a name="l00321"></a>00321 <span class="comment">/*****************************************************************************/</span>
<a name="l00322"></a>00322 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00323"></a>00323 <span class="comment">/**</span>
<a name="l00324"></a>00324 <span class="comment"> * TranslationTable::getGeneCallGroup</span>
<a name="l00325"></a>00325 <span class="comment"> * Synopsis:</span>
<a name="l00326"></a>00326 <span class="comment"> *</span>
<a name="l00327"></a>00327 <span class="comment"> * A selector for the m_geneGroup.</span>
<a name="l00328"></a>00328 <span class="comment"> * </span>
<a name="l00329"></a>00329 <span class="comment"> * Typically a GeneCallGroup will be made up of SNPs (single</span>
<a name="l00330"></a>00330 <span class="comment"> * markers) and one possilbe Haplotype CallSet. In any case, the</span>
<a name="l00331"></a>00331 <span class="comment"> * entire set is returned with this method. </span>
<a name="l00332"></a>00332 <span class="comment"> *</span>
<a name="l00333"></a>00333 <span class="comment"> * @param geneName - TranslationTableModel gene.</span>
<a name="l00334"></a>00334 <span class="comment"> * @returns std::vector&lt;GeneCall&gt; - the group of GeneCalls found. a</span>
<a name="l00335"></a>00335 <span class="comment"> */</span>
<a name="l00336"></a>00336 <span class="comment">/*****************************************************************************/</span>
<a name="l00337"></a><a class="code" href="classTranslationTable.html#af445ef0873068bc4e39de2e0ef6f03ca">00337</a> std::vector&lt;GeneCall&gt; <a class="code" href="classTranslationTable.html#af445ef0873068bc4e39de2e0ef6f03ca" title="TranslationTable::getGeneCallGroup Synopsis:">TranslationTable::getGeneCallGroup</a>(<span class="keyword">const</span> std::string &amp; geneName)
<a name="l00338"></a>00338 {
<a name="l00339"></a>00339   <span class="keywordflow">return</span> m_geneGroup[geneName];
<a name="l00340"></a>00340 
<a name="l00341"></a>00341 }
<a name="l00342"></a>00342 <span class="comment">// end TranslationTable::getGeneCallGroup</span>
<a name="l00343"></a>00343 <span class="comment">/*****************************************************************************/</span>
<a name="l00344"></a>00344 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00345"></a>00345 <span class="comment">/**</span>
<a name="l00346"></a>00346 <span class="comment"> * TranslationTable::hasGeneCall</span>
<a name="l00347"></a>00347 <span class="comment"> * Synopsis:</span>
<a name="l00348"></a>00348 <span class="comment"> * </span>
<a name="l00349"></a>00349 <span class="comment"> * Interrogate the std::vector of GeneCalls if a particular gene name</span>
<a name="l00350"></a>00350 <span class="comment"> * is available for making a call. If not then there is no experiment</span>
<a name="l00351"></a>00351 <span class="comment"> * analysis to be done.</span>
<a name="l00352"></a>00352 <span class="comment"> *</span>
<a name="l00353"></a>00353 <span class="comment"> * @return true - if the passed in gene name has any corresponding GeneCall.</span>
<a name="l00354"></a>00354 <span class="comment"> */</span>
<a name="l00355"></a>00355 <span class="comment">/*****************************************************************************/</span>
<a name="l00356"></a><a class="code" href="classTranslationTable.html#aab8447bc4372e07e3c50facf27b41c53">00356</a> <span class="keywordtype">bool</span> <a class="code" href="classTranslationTable.html#aab8447bc4372e07e3c50facf27b41c53" title="TranslationTable::hasGeneCall Synopsis:">TranslationTable::hasGeneCall</a>(<span class="keyword">const</span> std::string &amp; geneName)
<a name="l00357"></a>00357 {
<a name="l00358"></a>00358 
<a name="l00359"></a>00359   <span class="keywordflow">return</span> (m_geneGroup[geneName].size() != 0);
<a name="l00360"></a>00360 
<a name="l00361"></a>00361 }
<a name="l00362"></a>00362 <span class="comment">// end TranslationTable::hasGeneCall</span>
<a name="l00363"></a>00363 <span class="comment">/*****************************************************************************/</span>
<a name="l00364"></a>00364 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00365"></a>00365 <span class="comment">/**</span>
<a name="l00366"></a>00366 <span class="comment"> * TranslationTable::translateExperimentGene</span>
<a name="l00367"></a>00367 <span class="comment"> * Snyopsis:</span>
<a name="l00368"></a>00368 <span class="comment"> * Perform translation and invoke the translation algorithm found in the</span>
<a name="l00369"></a>00369 <span class="comment"> * GeneCall business object.</span>
<a name="l00370"></a>00370 <span class="comment"> * Marshalls all the requisite data for making each translation</span>
<a name="l00371"></a>00371 <span class="comment"> * call and then again packages up the results in the CallResults</span>
<a name="l00372"></a>00372 <span class="comment"> * business object whenever a call is made.</span>
<a name="l00373"></a>00373 <span class="comment"> * </span>
<a name="l00374"></a>00374 <span class="comment"> * To re-state this routine is mostly</span>
<a name="l00375"></a>00375 <span class="comment"> * just a wrapper around the GeneCall method of similar name,</span>
<a name="l00376"></a>00376 <span class="comment"> * translateExperimentCall. The difference being is that this</span>
<a name="l00377"></a>00377 <span class="comment"> * method iterates over all the GeneCall&#39;s in a group and summarizes</span>
<a name="l00378"></a>00378 <span class="comment"> * the results in the return ExperimentGeneResults.</span>
<a name="l00379"></a>00379 <span class="comment"> *</span>
<a name="l00380"></a>00380 <span class="comment"> * So the way  this works is that each GeneCall acts as a filter on the</span>
<a name="l00381"></a>00381 <span class="comment"> * input genotype experiment data. Genotype data is grouped into genes within</span>
<a name="l00382"></a>00382 <span class="comment"> * the GenotypeTableModel being passed in. Each GeneCall corresponds to</span>
<a name="l00383"></a>00383 <span class="comment"> * precisely one gene marker or probe set, or the gene&#39;s</span>
<a name="l00384"></a>00384 <span class="comment"> * single Haplotype marker</span>
<a name="l00385"></a>00385 <span class="comment"> * set. The input genotype data (CHP) represented by the</span>
<a name="l00386"></a>00386 <span class="comment"> * GenotypeTableModel is searched for the probe sets required for</span>
<a name="l00387"></a>00387 <span class="comment"> * GeneCall. The two chromatid CallSets are built from the</span>
<a name="l00388"></a>00388 <span class="comment"> * GenotypTableModel for the GeneCall in question and then the</span>
<a name="l00389"></a>00389 <span class="comment"> * translation algorithm is invoked to find a match.</span>
<a name="l00390"></a>00390 <span class="comment"> *</span>
<a name="l00391"></a>00391 <span class="comment"> * If the input genotype data (CHP) does not contain the probe sets</span>
<a name="l00392"></a>00392 <span class="comment"> * required for the GeneCall in question then the GeneCall is skipped.</span>
<a name="l00393"></a>00393 <span class="comment"> * There is an exception with respect to the Haplotype set of markers.</span>
<a name="l00394"></a>00394 <span class="comment"> * If the program option is given to enforce that entire Haplotype</span>
<a name="l00395"></a>00395 <span class="comment"> * sets be complete for translation and markers are found missing</span>
<a name="l00396"></a>00396 <span class="comment"> * within the input genotype data (CHP), then an exception will be thrown.</span>
<a name="l00397"></a>00397 <span class="comment"> *</span>
<a name="l00398"></a>00398 <span class="comment"> * In the common case, an incomplete haplotype group results in the</span>
<a name="l00399"></a>00399 <span class="comment"> * haplotype GeneCall being skipped, but realize that each individual</span>
<a name="l00400"></a>00400 <span class="comment"> * marker has an individual marker GeneCall and so will still have its</span>
<a name="l00401"></a>00401 <span class="comment"> * marker &quot;Ref/Var&quot; call made. </span>
<a name="l00402"></a>00402 <span class="comment"> * </span>
<a name="l00403"></a>00403 <span class="comment"> * </span>
<a name="l00404"></a>00404 <span class="comment"> * </span>
<a name="l00405"></a>00405 <span class="comment"> * @param rte - the single instance run time environment</span>
<a name="l00406"></a>00406 <span class="comment"> * @param gtm - the experiment Genotype Tabel model filtered for a single gene. </span>
<a name="l00407"></a>00407 <span class="comment"> * @param ttm - the single instance translation table model. </span>
<a name="l00408"></a>00408 <span class="comment"> * @param cntm - DMET2 ONLY: the copy number table model</span>
<a name="l00409"></a>00409 <span class="comment"> * @param egr - the experiment gene results for return</span>
<a name="l00410"></a>00410 <span class="comment"> *</span>
<a name="l00411"></a>00411 <span class="comment"> * @return - egr, experiment gene results is filled in with the calls for the entire experiment gene. </span>
<a name="l00412"></a>00412 <span class="comment"></span>
<a name="l00413"></a>00413 <span class="comment"> */</span>
<a name="l00414"></a>00414 <span class="comment">/*****************************************************************************/</span>
<a name="l00415"></a><a class="code" href="classTranslationTable.html#aad1fde62c93ecd4ad2951eecc35706f7">00415</a> <span class="keywordtype">void</span> <a class="code" href="classTranslationTable.html#aad1fde62c93ecd4ad2951eecc35706f7" title="TranslationTable::translateExperimentGene Snyopsis: Perform translation and invoke the translation al...">TranslationTable::translateExperimentGene</a>(<a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,  <a class="code" href="classGenotypeTableModel.html">GenotypeTableModel</a>  &amp; gtm, <span class="keyword">const</span> <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp;ttm, <span class="keyword">const</span> <a class="code" href="classCopyNumberTableModel.html">CopyNumberTableModel</a> &amp; cntm, <a class="code" href="classExperimentGeneResults.html">ExperimentGeneResults</a> *egr)
<a name="l00416"></a>00416 {
<a name="l00417"></a>00417 
<a name="l00418"></a>00418   <span class="comment">// Copy the static stuff from the model.</span>
<a name="l00419"></a>00419 
<a name="l00420"></a>00420   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(egr, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00421"></a>00421 
<a name="l00422"></a>00422   <span class="keywordflow">if</span> (! <a class="code" href="classTranslationTable.html#aab8447bc4372e07e3c50facf27b41c53" title="TranslationTable::hasGeneCall Synopsis:">hasGeneCall</a>(gtm.m_geneName)) {
<a name="l00423"></a>00423 
<a name="l00424"></a>00424     <span class="comment">// This data should be passed through for reporting purposes.</span>
<a name="l00425"></a>00425     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_TMI, <span class="stringliteral">&quot;Experiment Gene with no call, ignoring: &quot;</span> + gtm.m_experimentName + <span class="stringliteral">&quot;, &quot;</span> + gtm.m_geneName);
<a name="l00426"></a>00426     <span class="keywordflow">return</span>;
<a name="l00427"></a>00427   }
<a name="l00428"></a>00428 
<a name="l00429"></a>00429   <span class="comment">// Each CallSet will have a pair of CallElements corresponding to a</span>
<a name="l00430"></a>00430   <span class="comment">// ALLELE1 and ALLELE2 for the row.</span>
<a name="l00431"></a>00431   <span class="keywordflow">if</span> (rte.m_adtOpts.m_profile) {
<a name="l00432"></a>00432     rte.m_profiles[<span class="stringliteral">&quot;call_loop1&quot;</span>]-&gt;begin();
<a name="l00433"></a>00433   }
<a name="l00434"></a>00434 
<a name="l00435"></a>00435   std::vector&lt;CallSet&gt; invalidGTMRows = _invalidateExperimentGeneProbeSetBaseValues(rte, gtm, ttm);
<a name="l00436"></a>00436 
<a name="l00437"></a>00437   <span class="keywordflow">if</span> (invalidGTMRows.size() &gt; 0) {
<a name="l00438"></a>00438     <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(ADT_VERBOSE_NORMAL, gtm.m_geneName + <span class="stringliteral">&quot;: Invalid gene markers not found in translation file&quot;</span>);
<a name="l00439"></a>00439     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; invalidGTMRows.size(); i++) {
<a name="l00440"></a>00440       invalidGTMRows[i].describeVerbose(ADT_VERBOSE_NORMAL);
<a name="l00441"></a>00441     }
<a name="l00442"></a>00442     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Invalid experiment gene marker values detected.&quot;</span>);
<a name="l00443"></a>00443   }
<a name="l00444"></a>00444   <span class="keywordflow">if</span> (rte.m_adtOpts.m_profile) {
<a name="l00445"></a>00445     rte.m_profiles[<span class="stringliteral">&quot;call_loop1&quot;</span>]-&gt;end();
<a name="l00446"></a>00446   }
<a name="l00447"></a>00447 
<a name="l00448"></a>00448   std::string geneExperiment = gtm.getGeneName() + <span class="stringliteral">&quot;:&quot;</span> + gtm.m_experimentName;
<a name="l00449"></a>00449   std::string dmet2CopyNumberCall;
<a name="l00450"></a>00450 
<a name="l00451"></a>00451   <span class="keywordflow">if</span> (rte.m_adtOpts.m_dmet2Calling) {
<a name="l00452"></a>00452     std::map&lt;std::string, std::string&gt;::const_iterator itSS
<a name="l00453"></a>00453     = m_geneExperimentCopyNumberCall.find(geneExperiment);
<a name="l00454"></a>00454 
<a name="l00455"></a>00455     <span class="keywordflow">if</span> (itSS != m_geneExperimentCopyNumberCall.end()) {
<a name="l00456"></a>00456       dmet2CopyNumberCall = itSS-&gt;second;
<a name="l00457"></a>00457     }
<a name="l00458"></a>00458   }
<a name="l00459"></a>00459 
<a name="l00460"></a>00460   <span class="comment">// This is a std::vector because a GeneCall represents one Haplotype grouping</span>
<a name="l00461"></a>00461   <span class="comment">// call and many marker calls.</span>
<a name="l00462"></a>00462 
<a name="l00463"></a>00463   std::vector&lt;GeneCall&gt; geneCallToMatch = <a class="code" href="classTranslationTable.html#af445ef0873068bc4e39de2e0ef6f03ca" title="TranslationTable::getGeneCallGroup Synopsis:">getGeneCallGroup</a>(gtm.getGeneName());
<a name="l00464"></a>00464 
<a name="l00465"></a>00465   std::vector&lt;GeneCall&gt;::iterator matchCall;
<a name="l00466"></a>00466 
<a name="l00467"></a>00467   <span class="keywordflow">for</span> (matchCall = geneCallToMatch.begin();
<a name="l00468"></a>00468        matchCall != geneCallToMatch.end(); matchCall++) {
<a name="l00469"></a>00469 
<a name="l00470"></a>00470     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_CALL, <span class="stringliteral">&quot;CALL: &quot;</span> + gtm.m_geneName);
<a name="l00471"></a>00471 
<a name="l00472"></a>00472     <span class="comment">//matchCall-&gt;describeVerbose( rte, ADT_VERBOSE_CALL );</span>
<a name="l00473"></a>00473 
<a name="l00474"></a>00474     <span class="comment">// Ok, the gtm will contain all the</span>
<a name="l00475"></a>00475     <span class="comment">//  markers needed for all GeneCalls</span>
<a name="l00476"></a>00476     <span class="comment">// This constructer filters the gtm with only elements found in the</span>
<a name="l00477"></a>00477     <span class="comment">// referenceSet</span>
<a name="l00478"></a>00478     <span class="keywordflow">if</span> (rte.m_adtOpts.m_profile) {
<a name="l00479"></a>00479       rte.m_profiles[<span class="stringliteral">&quot;call_loop2&quot;</span>]-&gt;begin();
<a name="l00480"></a>00480     }
<a name="l00481"></a>00481     <a class="code" href="classCallSet.html">CallSet</a> chromatid1(rte, gtm, matchCall-&gt;getReferenceCallSet(), 1);
<a name="l00482"></a>00482 
<a name="l00483"></a>00483     <a class="code" href="classCallSet.html">CallSet</a> chromatid2(rte, gtm, matchCall-&gt;getReferenceCallSet(), 2);
<a name="l00484"></a>00484 
<a name="l00485"></a>00485     <span class="keywordflow">if</span> (rte.m_adtOpts.m_profile) {
<a name="l00486"></a>00486       rte.m_profiles[<span class="stringliteral">&quot;call_loop2&quot;</span>]-&gt;end();
<a name="l00487"></a>00487     }
<a name="l00488"></a>00488 
<a name="l00489"></a>00489     <span class="keywordflow">if</span> (chromatid1.isEmpty()) {
<a name="l00490"></a>00490       <span class="comment">// This experiment didn&#39;t contain any data for this call,</span>
<a name="l00491"></a>00491       <span class="comment">// this is expected, just continue.</span>
<a name="l00492"></a>00492       <span class="keywordflow">continue</span>;
<a name="l00493"></a>00493     }
<a name="l00494"></a>00494 
<a name="l00495"></a>00495     <span class="keywordflow">if</span> (rte.m_adtOpts.m_profile) {
<a name="l00496"></a>00496       rte.m_profiles[<span class="stringliteral">&quot;call_loop3&quot;</span>]-&gt;begin();
<a name="l00497"></a>00497     }
<a name="l00498"></a>00498     <a class="code" href="classCallResults.html">CallResults</a> call =
<a name="l00499"></a>00499       matchCall-&gt;
<a name="l00500"></a>00500       translateExperimentCall(rte,
<a name="l00501"></a>00501                               m_geneGroupCompleteCallSet[gtm.m_geneName],
<a name="l00502"></a>00502                               gtm,
<a name="l00503"></a>00503                               chromatid1, chromatid2,
<a name="l00504"></a>00504                               dmet2CopyNumberCall,
<a name="l00505"></a>00505                               m_geneCopyNumberZeroCallSet[gtm.m_geneName]);
<a name="l00506"></a>00506 
<a name="l00507"></a>00507     <span class="keywordflow">if</span> (rte.m_adtOpts.m_profile) {
<a name="l00508"></a>00508       rte.m_profiles[<span class="stringliteral">&quot;call_loop3&quot;</span>]-&gt;end();
<a name="l00509"></a>00509     }
<a name="l00510"></a>00510 
<a name="l00511"></a>00511     <span class="comment">// Set the TranslatedExperimentResults data</span>
<a name="l00512"></a>00512     egr-&gt;<a class="code" href="classExperimentGeneResults.html#a13722310b26cf978fd3ab2c4325c798e" title="ExperimentGeneResults::appendCallResults Synopsis:">appendCallResults</a>(call, *matchCall);
<a name="l00513"></a>00513 
<a name="l00514"></a>00514   } <span class="comment">// for each experiment gene call</span>
<a name="l00515"></a>00515 
<a name="l00516"></a>00516 
<a name="l00517"></a>00517   <span class="keywordflow">return</span>;
<a name="l00518"></a>00518 }
<a name="l00519"></a>00519 <span class="comment">// end   TranslationTable::translateExperimentGene</span>
<a name="l00520"></a>00520 <span class="comment">/*****************************************************************************/</span>
<a name="l00521"></a>00521 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00522"></a>00522 <span class="comment">/**</span>
<a name="l00523"></a>00523 <span class="comment"> * TranslationTable::_addGeneCallToGroup</span>
<a name="l00524"></a>00524 <span class="comment"> * Synopsis:</span>
<a name="l00525"></a>00525 <span class="comment"> * Helper function for the constructor, TranslationTable::TranslationTable</span>
<a name="l00526"></a>00526 <span class="comment"> * </span>
<a name="l00527"></a>00527 <span class="comment"> * Add a completed GeneCall to a gene&#39;s call Group.</span>
<a name="l00528"></a>00528 <span class="comment"> *</span>
<a name="l00529"></a>00529 <span class="comment"> * @param rte - RunTimeEnvironment</span>
<a name="l00530"></a>00530 <span class="comment"> * @param gene - std::string name of the Gene from the translation table.</span>
<a name="l00531"></a>00531 <span class="comment"> * @param addGeneCall - pointer to the GeneCall to add</span>
<a name="l00532"></a>00532 <span class="comment"> * @param markerProbeSet - </span>
<a name="l00533"></a>00533 <span class="comment"> *</span>
<a name="l00534"></a>00534 <span class="comment"> * @returns NULL, to reset a the pointer if needed.</span>
<a name="l00535"></a>00535 <span class="comment"> */</span>
<a name="l00536"></a>00536 <span class="comment">/*****************************************************************************/</span>
<a name="l00537"></a>00537 <a class="code" href="classGeneCall.html">GeneCall</a> * TranslationTable::_addGeneCallToGroup(
<a name="l00538"></a>00538   <span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte,
<a name="l00539"></a>00539   <span class="keyword">const</span> std::string &amp; gene,
<a name="l00540"></a>00540   <a class="code" href="classGeneCall.html">GeneCall</a> * addGeneCall)
<a name="l00541"></a>00541 {
<a name="l00542"></a>00542 
<a name="l00543"></a>00543   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(addGeneCall, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00544"></a>00544 
<a name="l00545"></a>00545   <span class="keywordflow">if</span> (addGeneCall-&gt;getReferenceCallSet().<a class="code" href="classCallSet.html#adfd65845f5730a7e1ad09cf9823786d6" title="inline">getCallType</a>() == ADT_CALL_TYPE_HAPLOTYPE_GROUP) {
<a name="l00546"></a>00546     _completeAlleleSets(addGeneCall);
<a name="l00547"></a>00547   } <span class="keywordflow">else</span> {
<a name="l00548"></a>00548     <span class="comment">// Erase the allele name CallSets for MARKERS.</span>
<a name="l00549"></a>00549     <span class="comment">// The first two CallSets are always Reference and Variant.</span>
<a name="l00550"></a>00550     <span class="comment">// The copy number 0 set will be reconciled in another API.</span>
<a name="l00551"></a>00551     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = addGeneCall-&gt;m_alleleSet.size() - 1; i &gt;= 2; i--) {
<a name="l00552"></a>00552       <span class="keywordflow">if</span> (addGeneCall-&gt;m_alleleSet[i].m_copyNumber != 0) {
<a name="l00553"></a>00553         addGeneCall-&gt;m_alleleSet.erase(addGeneCall-&gt;m_alleleSet.begin() + i);
<a name="l00554"></a>00554       }
<a name="l00555"></a>00555     }
<a name="l00556"></a>00556   }
<a name="l00557"></a>00557 
<a name="l00558"></a>00558   m_geneGroup[gene].push_back(*addGeneCall);
<a name="l00559"></a>00559 
<a name="l00560"></a>00560   <span class="comment">//m_geneGroup[gene][m_geneGroup[gene].size() -1].describeVerbose(rte, ADT_VERBOSE_NORMAL);</span>
<a name="l00561"></a>00561 
<a name="l00562"></a>00562   <span class="keyword">delete</span> addGeneCall;
<a name="l00563"></a>00563 
<a name="l00564"></a>00564   <span class="keywordflow">return</span>  NULL;
<a name="l00565"></a>00565 
<a name="l00566"></a>00566 }
<a name="l00567"></a>00567 <span class="comment">// end TranslationTable::_addGeneCallToGroup</span>
<a name="l00568"></a>00568 <span class="comment">/*****************************************************************************/</span>
<a name="l00569"></a>00569 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00570"></a>00570 <span class="comment">/**</span>
<a name="l00571"></a>00571 <span class="comment"> * TranslationTable::_completeAlleleSets</span>
<a name="l00572"></a>00572 <span class="comment"> * Synopsis:</span>
<a name="l00573"></a>00573 <span class="comment"> * Helper function to TranslationTable::_addGeneCallToGroup</span>
<a name="l00574"></a>00574 <span class="comment"> * </span>
<a name="l00575"></a>00575 <span class="comment"> * The allele bases  in the input translation table represent</span>
<a name="l00576"></a>00576 <span class="comment"> * changes with respect to the reference set. Fill in the blanks where the</span>
<a name="l00577"></a>00577 <span class="comment"> * reference set ProbeSet bases are added when not found in the allele set.</span>
<a name="l00578"></a>00578 <span class="comment"> * For SNP markers this means filling in the reference value for the</span>
<a name="l00579"></a>00579 <span class="comment"> * first column which matches the reference.</span>
<a name="l00580"></a>00580 <span class="comment"> *</span>
<a name="l00581"></a>00581 <span class="comment"> * @param  gc - pointer to the GeneCall class in question.</span>
<a name="l00582"></a>00582 <span class="comment"> *</span>
<a name="l00583"></a>00583 <span class="comment"> */</span>
<a name="l00584"></a>00584 <span class="comment">/*****************************************************************************/</span>
<a name="l00585"></a>00585 <span class="keywordtype">void</span> TranslationTable::_completeAlleleSets(<a class="code" href="classGeneCall.html">GeneCall</a> * gc)
<a name="l00586"></a>00586 {
<a name="l00587"></a>00587 
<a name="l00588"></a>00588   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(gc, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00589"></a>00589 
<a name="l00590"></a>00590   <a class="code" href="classCallSet.html">CallSet</a> &amp; referenceSet = gc-&gt;getReferenceCallSet();
<a name="l00591"></a>00591 
<a name="l00592"></a>00592   std::map&lt;std::string, CallElement&gt;::iterator iCEit;
<a name="l00593"></a>00593 
<a name="l00594"></a>00594 
<a name="l00595"></a>00595   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; gc-&gt;m_alleleSet.size(); i++) {
<a name="l00596"></a>00596 
<a name="l00597"></a>00597     <a class="code" href="classCallSet.html">CallSet</a> &amp; alleleSet = gc-&gt;<a class="code" href="classGeneCall.html#a30757e06c17f4315d3686a5341f4b547" title="GeneCall::getAlleleCallSet Synopsis:">getAlleleCallSet</a>(i);
<a name="l00598"></a>00598 
<a name="l00599"></a>00599     alleleSet.m_type = ADT_CALL_TYPE_HAPLOTYPE_GROUP;
<a name="l00600"></a>00600     alleleSet.m_isDescriptive = <span class="keyword">true</span>;
<a name="l00601"></a>00601 
<a name="l00602"></a>00602     <span class="comment">// For each CallElement in the reference set,</span>
<a name="l00603"></a>00603     <span class="comment">// search for the ProbeSet in the allele set and if</span>
<a name="l00604"></a>00604     <span class="comment">// it is not there, then add the reference CallElement to the allele set.</span>
<a name="l00605"></a>00605     <span class="keywordflow">for</span> (iCEit = referenceSet.m_ceSet.begin();
<a name="l00606"></a>00606          iCEit != referenceSet.m_ceSet.end();
<a name="l00607"></a>00607          iCEit++) {
<a name="l00608"></a>00608 
<a name="l00609"></a>00609       std::map&lt;std::string, CallElement&gt;::iterator jCEit;
<a name="l00610"></a>00610       <span class="keywordtype">bool</span> okToAdd = <span class="keyword">true</span>;
<a name="l00611"></a>00611 
<a name="l00612"></a>00612       <span class="keywordflow">for</span> (jCEit = alleleSet.m_ceSet.begin();
<a name="l00613"></a>00613            (jCEit != alleleSet.m_ceSet.end()) &amp;&amp;
<a name="l00614"></a>00614            okToAdd;
<a name="l00615"></a>00615            jCEit++) {
<a name="l00616"></a>00616 
<a name="l00617"></a>00617         <span class="keywordflow">if</span> (jCEit-&gt;first == iCEit-&gt;first)
<a name="l00618"></a>00618           okToAdd = <span class="keyword">false</span>;
<a name="l00619"></a>00619 
<a name="l00620"></a>00620 
<a name="l00621"></a>00621       } <span class="comment">// for each CallElement in the alleleSet, find the corresponding</span>
<a name="l00622"></a>00622       <span class="comment">// CallElement probeSet in the referenceSet.</span>
<a name="l00623"></a>00623 
<a name="l00624"></a>00624       <span class="keywordflow">if</span> (okToAdd) {
<a name="l00625"></a>00625         alleleSet.m_ceSet[iCEit-&gt;first] = iCEit-&gt;second;
<a name="l00626"></a>00626       }
<a name="l00627"></a>00627 
<a name="l00628"></a>00628     } <span class="comment">// for each CallElement in the referenceSet</span>
<a name="l00629"></a>00629 
<a name="l00630"></a>00630 
<a name="l00631"></a>00631   } <span class="comment">// for each allele call, complete the allele set</span>
<a name="l00632"></a>00632 
<a name="l00633"></a>00633 
<a name="l00634"></a>00634   <span class="keywordflow">return</span>;
<a name="l00635"></a>00635 
<a name="l00636"></a>00636 }
<a name="l00637"></a>00637 <span class="comment">// end TranslationTable::_completeAlleleSets</span>
<a name="l00638"></a>00638 <span class="comment">/*****************************************************************************/</span>
<a name="l00639"></a>00639 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00640"></a>00640 <span class="comment">/**</span>
<a name="l00641"></a>00641 <span class="comment"> * TranslationTable::_duplicateCallSetCheck</span>
<a name="l00642"></a>00642 <span class="comment"> * Synopsis:</span>
<a name="l00643"></a>00643 <span class="comment"> * Helper function to the constructor, TranslationTable::TranslationTable</span>
<a name="l00644"></a>00644 <span class="comment"> *</span>
<a name="l00645"></a>00645 <span class="comment"> * The input translation table is created by humans and is</span>
<a name="l00646"></a>00646 <span class="comment"> * not guaranteed to have a unique set of</span>
<a name="l00647"></a>00647 <span class="comment"> * defining markers per haplotype alllele. The algorithm </span>
<a name="l00648"></a>00648 <span class="comment"> * behavior during translation is to match the first</span>
<a name="l00649"></a>00649 <span class="comment"> * allele set of defining markers and ignore the others.</span>
<a name="l00650"></a>00650 <span class="comment"> * This API is used to make the determinate if any duplicates</span>
<a name="l00651"></a>00651 <span class="comment"> * exist and the calllee can use this to abort the program or</span>
<a name="l00652"></a>00652 <span class="comment"> * not depending on the business requirements. </span>
<a name="l00653"></a>00653 <span class="comment"> * </span>
<a name="l00654"></a>00654 <span class="comment"> *</span>
<a name="l00655"></a>00655 <span class="comment"> * @param rte  - the single run time environment instance</span>
<a name="l00656"></a>00656 <span class="comment"> * </span>
<a name="l00657"></a>00657 <span class="comment"> * @returns true - the duplicates do not exist </span>
<a name="l00658"></a>00658 <span class="comment"> */</span>
<a name="l00659"></a>00659 <span class="comment">/*****************************************************************************/</span>
<a name="l00660"></a>00660 <span class="keywordtype">bool</span> TranslationTable::_duplicateCallSetCheck(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte)
<a name="l00661"></a>00661 {
<a name="l00662"></a>00662 
<a name="l00663"></a>00663   std::map&lt;std::string, std::vector&lt;GeneCall&gt; &gt;::iterator itMSVG;
<a name="l00664"></a>00664 
<a name="l00665"></a>00665   ADT_VERBOSE_ENUM level = ADT_VERBOSE_NORMAL;
<a name="l00666"></a>00666 
<a name="l00667"></a>00667   <span class="keywordtype">bool</span> okDuplicateCheck = <span class="keyword">true</span>;
<a name="l00668"></a>00668 
<a name="l00669"></a>00669   <span class="comment">// Foreach gene</span>
<a name="l00670"></a>00670   <span class="keywordflow">for</span> (itMSVG = m_geneGroup.begin(); itMSVG != m_geneGroup.end(); itMSVG++) {
<a name="l00671"></a>00671 
<a name="l00672"></a>00672     std::string gene = itMSVG-&gt;first;
<a name="l00673"></a>00673 
<a name="l00674"></a>00674     <span class="comment">// Foreach GeneCall allele call set</span>
<a name="l00675"></a>00675     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; itMSVG-&gt;second.size(); i++) {
<a name="l00676"></a>00676 
<a name="l00677"></a>00677       <span class="comment">// Foreach call set</span>
<a name="l00678"></a>00678       <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 1; j &lt; itMSVG-&gt;second[i].m_alleleSet.size(); j++) {
<a name="l00679"></a>00679 
<a name="l00680"></a>00680         <span class="comment">// Check subsequent call sets for duplicates.</span>
<a name="l00681"></a>00681 
<a name="l00682"></a>00682         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = j + 1; k &lt; itMSVG-&gt;second[i].m_alleleSet.size(); k++) {
<a name="l00683"></a>00683 
<a name="l00684"></a>00684           <span class="keywordflow">if</span> (!rte.m_adtOpts.m_audit.empty() &amp;&amp; itMSVG-&gt;second[i].m_alleleSet[j] == itMSVG-&gt;second[i].m_alleleSet[k]) {
<a name="l00685"></a>00685             <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(level, <span class="stringliteral">&quot;Invalid translation table, duplicate allele set of variants detected.&quot;</span>);
<a name="l00686"></a>00686             itMSVG-&gt;second[i].m_alleleSet[j].describeVerbose(level);
<a name="l00687"></a>00687             itMSVG-&gt;second[i].m_alleleSet[k].describeVerbose(level);
<a name="l00688"></a>00688             <span class="comment">//Verbose::warn(level, &quot;Translation file, duplicate allele set of variants detected.&quot;);</span>
<a name="l00689"></a>00689             okDuplicateCheck = <span class="keyword">false</span>;;
<a name="l00690"></a>00690           }
<a name="l00691"></a>00691         }
<a name="l00692"></a>00692       }
<a name="l00693"></a>00693     }
<a name="l00694"></a>00694   }
<a name="l00695"></a>00695 
<a name="l00696"></a>00696 
<a name="l00697"></a>00697   <span class="keywordflow">return</span> okDuplicateCheck;
<a name="l00698"></a>00698 }
<a name="l00699"></a>00699 <span class="comment">// end TranslationTable::_duplicateCallSetCheck</span>
<a name="l00700"></a>00700 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00701"></a>00701 <span class="comment">/**</span>
<a name="l00702"></a>00702 <span class="comment"> * TranslationTable::_findMarkerGeneCall</span>
<a name="l00703"></a>00703 <span class="comment"> * Synopsis:</span>
<a name="l00704"></a>00704 <span class="comment"> * Helper function to the constructor, TranslationTable::TranslationTable</span>
<a name="l00705"></a>00705 <span class="comment"> *</span>
<a name="l00706"></a>00706 <span class="comment"> * When constructing marker GeneCalls, each marker gets a new GeneCall</span>
<a name="l00707"></a>00707 <span class="comment"> * except for multiallelic markers. These markers will already have GeneCalls</span>
<a name="l00708"></a>00708 <span class="comment"> * in the m_geneGroup when the marker occurs for more than once in the</span>
<a name="l00709"></a>00709 <span class="comment"> * input translation table. </span>
<a name="l00710"></a>00710 <span class="comment"> * This method finds the appropriate GeneCall for the multiallelic  marker</span>
<a name="l00711"></a>00711 <span class="comment"> * that already has a GeneCall. </span>
<a name="l00712"></a>00712 <span class="comment"> * </span>
<a name="l00713"></a>00713 <span class="comment"> *</span>
<a name="l00714"></a>00714 <span class="comment"> * @param gene - the multiallelic probeSet</span>
<a name="l00715"></a>00715 <span class="comment"> * @param probeSet - the marker</span>
<a name="l00716"></a>00716 <span class="comment"> * </span>
<a name="l00717"></a>00717 <span class="comment"> * @return - The corresponding GeneCall for the inputs.</span>
<a name="l00718"></a>00718 <span class="comment"> */</span>
<a name="l00719"></a>00719 <span class="comment">/*****************************************************************************/</span>
<a name="l00720"></a>00720 <a class="code" href="classGeneCall.html">GeneCall</a> * TranslationTable::_findMarkerGeneCall(std::string gene, std::string probeSet)
<a name="l00721"></a>00721 {
<a name="l00722"></a>00722 
<a name="l00723"></a>00723   <span class="keywordflow">if</span> (! m_geneGroupCompleteCallSet[gene].hasProbeSet(probeSet)) {
<a name="l00724"></a>00724     <span class="keywordflow">return</span> NULL;
<a name="l00725"></a>00725   }
<a name="l00726"></a>00726 
<a name="l00727"></a>00727   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; m_geneGroup[gene].size(); i++) {
<a name="l00728"></a>00728 
<a name="l00729"></a>00729     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; m_geneGroup[gene][i].m_alleleSet.size(); j++) {
<a name="l00730"></a>00730       <span class="keywordflow">if</span> (m_geneGroup[gene][i].m_alleleSet[j].hasProbeSet(probeSet) &amp;&amp;
<a name="l00731"></a>00731           m_geneGroup[gene][i].m_alleleSet[j].getCallType() == ADT_CALL_TYPE_MARKER) {
<a name="l00732"></a>00732         <span class="keywordflow">return</span> &amp;(m_geneGroup[gene][i]);
<a name="l00733"></a>00733       }
<a name="l00734"></a>00734     }
<a name="l00735"></a>00735   }
<a name="l00736"></a>00736   <span class="comment">// Programming error. The complete set cannot have an probeSet that</span>
<a name="l00737"></a>00737   <span class="comment">// is not in a GeneCall.</span>
<a name="l00738"></a>00738 
<a name="l00739"></a>00739   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(<span class="keyword">false</span>, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00740"></a>00740 
<a name="l00741"></a>00741   <span class="keywordflow">return</span> NULL;
<a name="l00742"></a>00742 
<a name="l00743"></a>00743 }
<a name="l00744"></a>00744 <span class="comment">// end TranslationTable::_findMarkerGeneCall</span>
<a name="l00745"></a>00745 <span class="comment">/*****************************************************************************/</span>
<a name="l00746"></a>00746 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00747"></a>00747 <span class="comment">/**</span>
<a name="l00748"></a>00748 <span class="comment"> * TranslationTable::_invalidateExperimentGeneProbeSetBaseValues</span>
<a name="l00749"></a>00749 <span class="comment"> * Synopsis:</span>
<a name="l00750"></a>00750 <span class="comment"> * Helper function to TranslationTable::translateExperimentGene</span>
<a name="l00751"></a>00751 <span class="comment"> * </span>
<a name="l00752"></a>00752 <span class="comment"> * This is just and optimiaztion fucntion where the *MAJORITY* </span>
<a name="l00753"></a>00753 <span class="comment"> * of inbound GenotypeTableModel probe sets will have ProbeSet/Base pairs which</span>
<a name="l00754"></a>00754 <span class="comment"> * are not recognized for any GeneCall. This helper method modifies</span>
<a name="l00755"></a>00755 <span class="comment"> * the passed in GenotypeTableModel by marking rows in the</span>
<a name="l00756"></a>00756 <span class="comment"> * &quot;m_rows&quot; attribute that are unknown with a special experiment name.</span>
<a name="l00757"></a>00757 <span class="comment"> * The invalid rows are also returned</span>
<a name="l00758"></a>00758 <span class="comment"> * in the form of a CallSet where each ALLELE1 and ALLELE2</span>
<a name="l00759"></a>00759 <span class="comment"> * CallElement corresponds to one CallSet and one GTM row.</span>
<a name="l00760"></a>00760 <span class="comment"> *</span>
<a name="l00761"></a>00761 <span class="comment"> * @param rte - the single instance RunTimeEnvironment</span>
<a name="l00762"></a>00762 <span class="comment"> * @param gtm - a GenotypeTableModel for a particulare gene experiment.</span>
<a name="l00763"></a>00763 <span class="comment"> * @param ttm - the single instance translation table model</span>
<a name="l00764"></a>00764 <span class="comment"> *</span>
<a name="l00765"></a>00765 <span class="comment"> * @return invalidRows - a std::vector&lt;CallSet&gt; where each CallSet represents a</span>
<a name="l00766"></a>00766 <span class="comment"> *                       row where ALLELE1 or ALLELE2 had an invalid base.</span>
<a name="l00767"></a>00767 <span class="comment"> */</span>
<a name="l00768"></a>00768 <span class="comment">/*****************************************************************************/</span>
<a name="l00769"></a>00769 std::vector&lt;CallSet&gt; TranslationTable::_invalidateExperimentGeneProbeSetBaseValues(
<a name="l00770"></a>00770   <span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classGenotypeTableModel.html">GenotypeTableModel</a> &amp; gtm, <span class="keyword">const</span> <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm)
<a name="l00771"></a>00771 {
<a name="l00772"></a>00772 
<a name="l00773"></a>00773 
<a name="l00774"></a>00774   std::vector&lt;CallSet&gt; invalidGTMRows;
<a name="l00775"></a>00775 
<a name="l00776"></a>00776   <span class="comment">// Unknown gene names should be handled before calling this API.</span>
<a name="l00777"></a>00777 
<a name="l00778"></a>00778   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(m_geneGroupCompleteCallSet.find(gtm.getGeneName()) != m_geneGroupCompleteCallSet.end(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00779"></a>00779 
<a name="l00780"></a>00780 
<a name="l00781"></a>00781   <span class="keyword">const</span> <a class="code" href="classCallSet.html">CallSet</a> &amp; testCS = m_geneGroupCompleteCallSet[gtm.getGeneName()];
<a name="l00782"></a>00782 
<a name="l00783"></a>00783   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; gtm.<a class="code" href="classGenotypeTableModel.html#ad413a9b5e58f44ef33a42a641806cbed" title="getGeneRowSize: Synopsis:">getGeneRowSize</a>(gtm.getGeneName()); i++) {
<a name="l00784"></a>00784 
<a name="l00785"></a>00785     <span class="keywordtype">int</span> row = gtm.<a class="code" href="classGenotypeTableModel.html#a82ce9bc2679a496b3592fddb14224f48" title="GenotypeTableModel::getGeneRow: Synopsis:">getGeneRow</a>(gtm.getGeneName(), i);
<a name="l00786"></a>00786 
<a name="l00787"></a>00787     std::string probeSet = gtm.m_rows[row][GT_PROBE_SET_INDEX];
<a name="l00788"></a>00788     std::string geneName = gtm.m_rows[row][GT_GENES_INDEX];
<a name="l00789"></a>00789     std::string base1    = gtm.m_rows[row][GT_ALLELE1_INDEX];
<a name="l00790"></a>00790     std::string base2    = gtm.m_rows[row][GT_ALLELE2_INDEX];
<a name="l00791"></a>00791 
<a name="l00792"></a>00792     <span class="keywordflow">if</span> (base2.empty()) {
<a name="l00793"></a>00793       base2 = base1;
<a name="l00794"></a>00794     }
<a name="l00795"></a>00795 
<a name="l00796"></a>00796     <a class="code" href="classCallElement.html">CallElement</a> allele1(probeSet, base1);
<a name="l00797"></a>00797     <a class="code" href="classCallElement.html">CallElement</a> allele2(probeSet, base2);
<a name="l00798"></a>00798 
<a name="l00799"></a>00799     <span class="keywordflow">if</span> (testCS.hasProbeSet(probeSet)) {
<a name="l00800"></a>00800 
<a name="l00801"></a>00801       <span class="comment">// Non-haplotype markers are always valid.</span>
<a name="l00802"></a>00802       <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#a6cdb03c7403f576e8ec4d932072ac446" title="TranslationTableModel::isHaplotypeMarker Synopsis: Is a row a Haplotype row? Yes if the &amp;#39;Haplotyp...">isHaplotypeMarker</a>(probeSet)) {
<a name="l00803"></a>00803 
<a name="l00804"></a>00804 
<a name="l00805"></a>00805         <span class="comment">// Wildcard markers are always valid.</span>
<a name="l00806"></a>00806         <span class="keywordflow">if</span> ((! allele1.hasWildCards() &amp;&amp; !testCS.contains(allele1)) ||
<a name="l00807"></a>00807             (! allele2.hasWildCards() &amp;&amp; !testCS.contains(allele2))) {
<a name="l00808"></a>00808 
<a name="l00809"></a>00809           <span class="comment">// Just ignore the row by setting the Experiment Name to something diffferent.</span>
<a name="l00810"></a>00810           gtm.m_rows[row][GT_EXPERIMENT_INDEX] =
<a name="l00811"></a>00811             <span class="stringliteral">&quot;Invalid ALLELE Base For ProbeSet&quot;</span>;
<a name="l00812"></a>00812           <span class="comment">//Flag an exception by returning an invalid CallSEt.</span>
<a name="l00813"></a>00813           <a class="code" href="classCallSet.html">CallSet</a> invalidCS(geneName);
<a name="l00814"></a>00814           invalidCS.m_ceSet[probeSet] =  allele1 ;
<a name="l00815"></a>00815           invalidGTMRows.push_back(invalidCS);
<a name="l00816"></a>00816         }
<a name="l00817"></a>00817       }
<a name="l00818"></a>00818     } <span class="keywordflow">else</span> {
<a name="l00819"></a>00819 
<a name="l00820"></a>00820       <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(ttm.m_probeSetRowIndex.count(probeSet) == 0, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00821"></a>00821       <span class="comment">// Just ignore the row by setting the Experiment Name to something diffferent.</span>
<a name="l00822"></a>00822       gtm.m_rows[row][GT_EXPERIMENT_INDEX] =
<a name="l00823"></a>00823         <span class="stringliteral">&quot;Invalid ProbeSet for Gene&quot;</span>;
<a name="l00824"></a>00824     }
<a name="l00825"></a>00825 
<a name="l00826"></a>00826   }
<a name="l00827"></a>00827 
<a name="l00828"></a>00828   <span class="keywordflow">return</span> invalidGTMRows;
<a name="l00829"></a>00829 
<a name="l00830"></a>00830 }
<a name="l00831"></a>00831 <span class="comment">// end TranslationTable::_invalidateExperimentGeneProbeSetBaseValues</span>
<a name="l00832"></a>00832 <span class="comment">/*****************************************************************************/</span>
<a name="l00833"></a>00833 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00834"></a>00834 <span class="comment">/**</span>
<a name="l00835"></a>00835 <span class="comment"> * TranslationTable::_reconcileCopyNumberCallSet:</span>
<a name="l00836"></a>00836 <span class="comment"> * Synopsis:</span>
<a name="l00837"></a>00837 <span class="comment"> * Helper function to the constructor, TranslationTable::TranslationTable.</span>
<a name="l00838"></a>00838 <span class="comment"> *</span>
<a name="l00839"></a>00839 <span class="comment"> * The format of the translation table was hacked to use an allele column</span>
<a name="l00840"></a>00840 <span class="comment"> * (A1-AN) to carry the copy number zero designation. This routines</span>
<a name="l00841"></a>00841 <span class="comment"> * reconciles the hack and creates the appropriate business object</span>
<a name="l00842"></a>00842 <span class="comment"> * and then marks &quot;allele&quot; column as one to be ignored. </span>
<a name="l00843"></a>00843 <span class="comment"> *</span>
<a name="l00844"></a>00844 <span class="comment"> * How to know if an allele name is copy number zero?</span>
<a name="l00845"></a>00845 <span class="comment"> * There is no naming conention, instead the first row after the allele</span>
<a name="l00846"></a>00846 <span class="comment"> * name column is checked for a &quot;0&quot; value. If the row immediately below</span>
<a name="l00847"></a>00847 <span class="comment"> * the allele name contains a &quot;0&quot; then the allele is the single</span>
<a name="l00848"></a>00848 <span class="comment"> * copy number zero designation. </span>
<a name="l00849"></a>00849 <span class="comment"> *</span>
<a name="l00850"></a>00850 <span class="comment"> * 1.) It is an exception to have more than one copy number 0 call set.</span>
<a name="l00851"></a>00851 <span class="comment"> * 2.) Remove the copy number 0 allele call set from the GeneCall set.</span>
<a name="l00852"></a>00852 <span class="comment"> * 3.) Instantiate the copy number 0 member CallSet set for the GeneCall.</span>
<a name="l00853"></a>00853 <span class="comment"> *</span>
<a name="l00854"></a>00854 <span class="comment"> * @params rte - the RunTimeEnvironment</span>
<a name="l00855"></a>00855 <span class="comment"> @ @params ttm - the translation table model for updates.</span>
<a name="l00856"></a>00856 <span class="comment"> *</span>
<a name="l00857"></a>00857 <span class="comment"> * @return - false on error, the parent is responsible for throwing an exception.</span>
<a name="l00858"></a>00858 <span class="comment"> *           ttm is updated to comment out the copy number allele</span>
<a name="l00859"></a>00859 <span class="comment"> *           designations.</span>
<a name="l00860"></a>00860 <span class="comment"> */</span>
<a name="l00861"></a>00861 <span class="comment">/*****************************************************************************/</span>
<a name="l00862"></a>00862 <span class="keywordtype">bool</span> TranslationTable::_reconcileCopyNumberCallSet(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm)
<a name="l00863"></a>00863 {
<a name="l00864"></a>00864 
<a name="l00865"></a>00865   std::map&lt;std::string, std::vector&lt;GeneCall&gt; &gt;::iterator itMSVG;
<a name="l00866"></a>00866 
<a name="l00867"></a>00867   <span class="comment">// Foreach gene</span>
<a name="l00868"></a>00868   <span class="keywordflow">for</span> (itMSVG = m_geneGroup.begin(); itMSVG != m_geneGroup.end(); itMSVG++) {
<a name="l00869"></a>00869 
<a name="l00870"></a>00870     std::string gene = itMSVG-&gt;first;
<a name="l00871"></a>00871 
<a name="l00872"></a>00872     <span class="comment">// Foreach GeneCall allele call set</span>
<a name="l00873"></a>00873     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; itMSVG-&gt;second.size(); i++) {
<a name="l00874"></a>00874 
<a name="l00875"></a>00875 
<a name="l00876"></a>00876       <span class="comment">// Skip over reference and variant sets, start at 2.</span>
<a name="l00877"></a>00877       <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 2; j &lt; itMSVG-&gt;second[i].m_alleleSet.size(); j++) {
<a name="l00878"></a>00878 
<a name="l00879"></a>00879         <span class="keywordflow">if</span> (itMSVG-&gt;second[i].m_alleleSet[j].m_copyNumber == 0) {
<a name="l00880"></a>00880 
<a name="l00881"></a>00881           <span class="comment">// Multiple check</span>
<a name="l00882"></a>00882           <span class="keywordflow">if</span> (m_geneCopyNumberZeroCallSet[gene].getCallType() != ADT_CALL_TYPE_NULL) {
<a name="l00883"></a>00883             <span class="keywordflow">if</span> (m_geneCopyNumberZeroCallSet[gene].m_name == itMSVG-&gt;second[i].m_alleleSet[j].m_name) {
<a name="l00884"></a>00884               <span class="keywordflow">continue</span>;
<a name="l00885"></a>00885             }
<a name="l00886"></a>00886             <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(ADT_VERBOSE_EXCEPTION, <span class="stringliteral">&quot;Fatal error,  translation file has more than one copy number 0 designation for gene: &quot;</span> + itMSVG-&gt;first + <span class="stringliteral">&quot;: &quot;</span> +  m_geneCopyNumberZeroCallSet[gene].m_name + <span class="stringliteral">&quot;, &quot;</span>  + itMSVG-&gt;second[i].m_alleleSet[j].m_name);
<a name="l00887"></a>00887             <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00888"></a>00888           }
<a name="l00889"></a>00889           ttm.m_geneCopyNumberIndicator[gene] = <span class="keyword">true</span>;
<a name="l00890"></a>00890           m_geneCopyNumberZeroCallSet[gene] = itMSVG-&gt;second[i].m_alleleSet[j];
<a name="l00891"></a>00891           m_geneCopyNumberZeroCallSet[gene].m_type = ADT_CALL_TYPE_COPY_NUMBER;
<a name="l00892"></a>00892           m_geneCopyNumberZeroCallSet[gene].size() &gt; 1 ?
<a name="l00893"></a>00893           m_geneCopyNumberZeroCallSet[gene].m_type = ADT_CALL_TYPE_HAPLOTYPE_GROUP :
<a name="l00894"></a>00894               m_geneCopyNumberZeroCallSet[gene].m_type = ADT_CALL_TYPE_MARKER ;
<a name="l00895"></a>00895           itMSVG-&gt;second[i].m_alleleSet.erase(itMSVG-&gt;second[i].m_alleleSet.begin() + j);
<a name="l00896"></a>00896           <span class="comment">//m_geneCopyNumberZeroCallSet[gene].describeVerbose( ADT_VERBOSE_NORMAL );</span>
<a name="l00897"></a>00897           <span class="comment">// Delete the copy number 0 so that reporting doesn&#39;t</span>
<a name="l00898"></a>00898           <span class="comment">// use it in relevant alleles.</span>
<a name="l00899"></a>00899           std::map&lt;std::string, CallElement &gt;::iterator itMSCE = m_geneCopyNumberZeroCallSet[gene].m_ceSet.begin();
<a name="l00900"></a>00900           std::string probeSet = itMSCE-&gt;second.m_probeSet;
<a name="l00901"></a>00901           <span class="keywordtype">int</span> probeSetRow = ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(probeSet);
<a name="l00902"></a>00902           <span class="keywordtype">int</span> headerRow = ttm.<a class="code" href="classTranslationTableModel.html#abd5a952b613ca007ed4fc4ac4e12f524" title="TranslationTableModel::getHeaderRowgetHeaderRow: Synopsis:">getHeaderRow</a>(probeSetRow);
<a name="l00903"></a>00903 
<a name="l00904"></a>00904           <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_ALLELE_START); i &lt; ttm.<a class="code" href="classTranslationInputTsvTableModel.html#ae33ba37f59967a7bd0caefe58de785ed" title="TranslationInputTsvTableModel::columnCount Synopsis:">columnCount</a>(); k++) {
<a name="l00905"></a>00905 
<a name="l00906"></a>00906             <span class="keywordflow">if</span> (ttm.m_rows[headerRow][k] == m_geneCopyNumberZeroCallSet[gene].m_name) {
<a name="l00907"></a>00907               <span class="comment">//ttm.m_rows[probeSetRow][k] = &quot;#&quot; + m_geneCopyNumberZeroCallSet[gene].m_name;</span>
<a name="l00908"></a>00908               ttm.m_rows[probeSetRow][k].clear();
<a name="l00909"></a>00909               ttm.m_rows[headerRow][k].clear();
<a name="l00910"></a>00910               <span class="keywordflow">break</span>;
<a name="l00911"></a>00911             }
<a name="l00912"></a>00912           }
<a name="l00913"></a>00913         } <span class="keywordflow">else</span> <span class="keywordflow">if</span> (itMSVG-&gt;second[i].m_alleleSet[j].getCallType() == ADT_CALL_TYPE_MARKER) {
<a name="l00914"></a>00914           <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(itMSVG-&gt;second[i].m_alleleSet.size() == 2, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00915"></a>00915         }
<a name="l00916"></a>00916       }
<a name="l00917"></a>00917     }
<a name="l00918"></a>00918 
<a name="l00919"></a>00919   }
<a name="l00920"></a>00920 
<a name="l00921"></a>00921   <span class="keywordflow">return</span> <span class="keyword">true</span>;
<a name="l00922"></a>00922 }
<a name="l00923"></a>00923 <span class="comment">// TranslationTable::_reconcileCopyNumberCallSet</span>
<a name="l00924"></a>00924 <span class="comment">/*****************************************************************************/</span>
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:04 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
