<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Affymetrix Power Tools: translation/TranslationInputStreamTableModel.cpp Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.1 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directories</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>File&nbsp;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><a class="el" href="dir_b6d6d4a2e58a0f89851f849ab50b1d34.html">translation</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<h1>TranslationInputStreamTableModel.cpp</h1>  </div>
</div>
<div class="contents">
<a href="TranslationInputStreamTableModel_8cpp.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00002"></a>00002 <span class="comment"></span><span class="comment">//</span>
<a name="l00003"></a>00003 <span class="comment">// Copyright (C) 2008 Affymetrix, Inc.</span>
<a name="l00004"></a>00004 <span class="comment">//</span>
<a name="l00005"></a>00005 <span class="comment">// This program is free software; you can redistribute it and/or modify </span>
<a name="l00006"></a>00006 <span class="comment">// it under the terms of the GNU General Public License (version 2) as </span>
<a name="l00007"></a>00007 <span class="comment">// published by the Free Software Foundation.</span>
<a name="l00008"></a>00008 <span class="comment">// </span>
<a name="l00009"></a>00009 <span class="comment">// This program is distributed in the hope that it will be useful, </span>
<a name="l00010"></a>00010 <span class="comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of </span>
<a name="l00011"></a>00011 <span class="comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU </span>
<a name="l00012"></a>00012 <span class="comment">// General Public License for more details.</span>
<a name="l00013"></a>00013 <span class="comment">// </span>
<a name="l00014"></a>00014 <span class="comment">// You should have received a copy of the GNU General Public License </span>
<a name="l00015"></a>00015 <span class="comment">// along with this program;if not, write to the </span>
<a name="l00016"></a>00016 <span class="comment">// </span>
<a name="l00017"></a>00017 <span class="comment">// Free Software Foundation, Inc., </span>
<a name="l00018"></a>00018 <span class="comment">// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</span>
<a name="l00019"></a>00019 <span class="comment">//</span><span class="comment"></span>
<a name="l00020"></a>00020 <span class="comment">////////////////////////////////////////////////////////////////</span>
<a name="l00021"></a>00021 <span class="comment"></span><span class="comment">/**</span>
<a name="l00022"></a>00022 <span class="comment"> * @file   TranslationInputStreamTableModel.cpp</span>
<a name="l00023"></a>00023 <span class="comment"> * @author Mybrid Spalding</span>
<a name="l00024"></a>00024 <span class="comment"> * @date   Fri Jun 13 13:09:30 PDT 2008</span>
<a name="l00025"></a>00025 <span class="comment"> * @brief  Started as base class data model for either CHP or TSV files, but ended as an exclusive class to GenotypeTableModel and contains GenotypeTableModel data. </span>
<a name="l00026"></a>00026 <span class="comment"> */</span>
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="comment">//</span>
<a name="l00029"></a>00029 <span class="preprocessor">#include &quot;<a class="code" href="TranslationInputStreamTableModel_8h.html" title="Started as base class data model for either CHP or TSV files, but ended as an exclusive class to Geno...">translation/TranslationInputStreamTableModel.h</a>&quot;</span>
<a name="l00030"></a>00030 <span class="comment">//</span>
<a name="l00031"></a>00031 <span class="preprocessor">#include &quot;<a class="code" href="CallElement_8h.html" title="Atomic class element for translation.">translation/CallElement.h</a>&quot;</span>
<a name="l00032"></a>00032 <span class="preprocessor">#include &quot;<a class="code" href="TranslationTableModel_8h.html" title="Single instance data model class that wraps the translation file in the TranslationInputTsvModel.">translation/TranslationTableModel.h</a>&quot;</span>
<a name="l00033"></a>00033 <span class="comment">//</span>
<a name="l00034"></a>00034 <span class="preprocessor">#include &quot;<a class="code" href="GenericDataTypes_8h.html">calvin_files/data/src/GenericDataTypes.h</a>&quot;</span>
<a name="l00035"></a>00035 <span class="preprocessor">#include &quot;<a class="code" href="AffymetrixGuid_8h.html">calvin_files/utils/src/AffymetrixGuid.h</a>&quot;</span>
<a name="l00036"></a>00036 <span class="preprocessor">#include &quot;<a class="code" href="GenoCallCoder_8h.html" title="Genotype call en/decoder.">calvin_files/utils/src/GenoCallCoder.h</a>&quot;</span>
<a name="l00037"></a>00037 <span class="preprocessor">#include &quot;<a class="code" href="StringUtils_8h.html">calvin_files/utils/src/StringUtils.h</a>&quot;</span>
<a name="l00038"></a>00038 <span class="preprocessor">#include &quot;<a class="code" href="CHPFileData_8h.html">file/CHPFileData.h</a>&quot;</span>
<a name="l00039"></a>00039 <span class="preprocessor">#include &quot;<a class="code" href="Fs_8h.html" title="///">util/Fs.h</a>&quot;</span>
<a name="l00040"></a>00040 <span class="preprocessor">#include &quot;<a class="code" href="Util_8h.html" title="General Utilities.">util/Util.h</a>&quot;</span>
<a name="l00041"></a>00041 <span class="comment">//</span>
<a name="l00042"></a>00042 <span class="preprocessor">#include &lt;iostream&gt;</span>
<a name="l00043"></a>00043 <span class="preprocessor">#include &lt;sstream&gt;</span>
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 <span class="keyword">using namespace </span>std;
<a name="l00046"></a>00046 <span class="keyword">using namespace </span>affx;
<a name="l00047"></a>00047 <span class="keyword">using namespace </span>affymetrix_calvin_data;
<a name="l00048"></a>00048 <span class="keyword">using namespace </span>affymetrix_calvin_io;
<a name="l00049"></a>00049 <span class="keyword">using namespace </span>affymetrix_calvin_parameter;
<a name="l00050"></a>00050 <span class="keyword">using namespace </span>affymetrix_calvin_utilities;
<a name="l00051"></a>00051 <span class="comment"></span>
<a name="l00052"></a>00052 <span class="comment">///////////////////////////////////////////////////////////////////////////////</span>
<a name="l00053"></a>00053 <span class="comment"></span><span class="comment">// BEGIN TranslationInputStreamTableModel</span><span class="comment"></span>
<a name="l00054"></a>00054 <span class="comment">///////////////////////////////////////////////////////////////////////////////</span>
<a name="l00055"></a>00055 <span class="comment"></span>
<a name="l00056"></a>00056 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00057"></a>00057 <span class="comment">/**</span>
<a name="l00058"></a>00058 <span class="comment"> * TranslationInputStreamTableModel::TranslationInputStreamTableModel</span>
<a name="l00059"></a>00059 <span class="comment"> * Synopsis:</span>
<a name="l00060"></a>00060 <span class="comment"> *</span>
<a name="l00061"></a>00061 <span class="comment"> * CHP constructor.</span>
<a name="l00062"></a>00062 <span class="comment"> *</span>
<a name="l00063"></a>00063 <span class="comment"> * Note that most CHP data is completely ignored excepting the alleles</span>
<a name="l00064"></a>00064 <span class="comment"> * and the probe set id.</span>
<a name="l00065"></a>00065 <span class="comment"> * Be only using three pieces of  CHP data from the CHP file</span>
<a name="l00066"></a>00066 <span class="comment"> * (probe set id, allele1 and allele2) allows the design to be coallece CHP</span>
<a name="l00067"></a>00067 <span class="comment"> * files into the TSV data model for the DMET2, Genotype Short Report.</span>
<a name="l00068"></a>00068 <span class="comment"> * This is signficant because research</span>
<a name="l00069"></a>00069 <span class="comment"> * at Affy after DMET3 is release will able to use the Genotype Short Report</span>
<a name="l00070"></a>00070 <span class="comment"> * input for DMET3. Creaing text files of sample data can be done in Excel</span>
<a name="l00071"></a>00071 <span class="comment"> *  and this avoids creating CHP files soley for the purpose of</span>
<a name="l00072"></a>00072 <span class="comment"> *  running this translation program. </span>
<a name="l00073"></a>00073 <span class="comment"> * </span>
<a name="l00074"></a>00074 <span class="comment"> * @param rte - the single instance run time environment.</span>
<a name="l00075"></a>00075 <span class="comment"> * @param tsvColumnDefinitions  - CHP data is coerced into TSV fields. </span>
<a name="l00076"></a>00076 <span class="comment"> * @param tcdSize - size of tsvColumnDefinitions</span>
<a name="l00077"></a>00077 <span class="comment"> * @param gotm  - the genotype override table model.</span>
<a name="l00078"></a>00078 <span class="comment"> *</span>
<a name="l00079"></a>00079 <span class="comment"> */</span>
<a name="l00080"></a>00080 <span class="comment">/*****************************************************************************/</span>
<a name="l00081"></a><a class="code" href="classTranslationInputStreamTableModel.html#ac2e959523d51f94d7272bcd4ab647adf">00081</a> TranslationInputStreamTableModel::TranslationInputStreamTableModel(
<a name="l00082"></a>00082   <span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a>&amp; rte,
<a name="l00083"></a>00083   <span class="keyword">const</span> <a class="code" href="classTittmColumnDefinition.html">TittmColumnDefinition</a> tsvColumnDefinitions[],
<a name="l00084"></a>00084   <span class="keywordtype">size_t</span> tcdSize,
<a name="l00085"></a>00085   <a class="code" href="classGenotypeOverrideTableModel.html">GenotypeOverrideTableModel</a> *gotm) : m_gotm(gotm)
<a name="l00086"></a>00086 {
<a name="l00087"></a>00087 
<a name="l00088"></a>00088   m_type                 = ADT_EXPERIMENT_STREAM_TYPE_CHP;
<a name="l00089"></a>00089   m_chpData              = NULL;
<a name="l00090"></a>00090   m_genoCallCoder        = NULL;
<a name="l00091"></a>00091   m_callsFileReader      = NULL;
<a name="l00092"></a>00092 
<a name="l00093"></a>00093   _initializeColumnDefinitions(rte, tsvColumnDefinitions, tcdSize);
<a name="l00094"></a>00094 
<a name="l00095"></a>00095   <span class="keywordflow">if</span> (rte.m_adtOpts.m_streamType == ADT_EXPERIMENT_STREAM_TYPE_BATCH_FOLDER) {
<a name="l00096"></a>00096     <span class="comment">// read blob file in batch folder</span>
<a name="l00097"></a>00097     m_blockReader.initFolder(rte.m_adtOpts.m_inputBatchFolder);
<a name="l00098"></a>00098     <span class="keywordflow">if</span> (!m_blockReader.open()) {
<a name="l00099"></a>00099       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(rte.m_adtOpts.m_inputBatchFolder + <span class="stringliteral">&quot;: failed to open batch folder with blob for reading.&quot;</span>);
<a name="l00100"></a>00100     }
<a name="l00101"></a>00101     m_type = ADT_EXPERIMENT_STREAM_TYPE_BATCH_FOLDER;
<a name="l00102"></a>00102   }
<a name="l00103"></a>00103   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (rte.m_adtOpts.m_streamType == ADT_EXPERIMENT_STREAM_TYPE_CALLS_FILE) {
<a name="l00104"></a>00104     <span class="comment">// read calls file</span>
<a name="l00105"></a>00105     m_callsFileReader = <span class="keyword">new</span> <a class="code" href="classCallsFileReader.html">CallsFileReader</a>();
<a name="l00106"></a>00106     <span class="keywordflow">if</span> (!m_callsFileReader-&gt;readFile(rte.m_adtOpts.m_inputCallsFile)) {
<a name="l00107"></a>00107       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(rte.m_adtOpts.m_inputCallsFile + <span class="stringliteral">&quot;: failed to read input calls file.&quot;</span>);
<a name="l00108"></a>00108     }
<a name="l00109"></a>00109     m_type = ADT_EXPERIMENT_STREAM_TYPE_CALLS_FILE;
<a name="l00110"></a>00110   }
<a name="l00111"></a>00111   <span class="keywordflow">else</span> {
<a name="l00112"></a>00112     <span class="comment">// Ok, we shouldn&#39;t be trapping this exception here, but</span>
<a name="l00113"></a>00113     <span class="comment">// during the input validation. This is a programming error</span>
<a name="l00114"></a>00114     <span class="comment">// so just assert.</span>
<a name="l00115"></a>00115     <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(rte.m_adtOpts.m_inputExperimentFiles.size() != 0, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00116"></a>00116 
<a name="l00117"></a>00117     <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; rte.m_adtOpts.m_inputExperimentFiles.size(); i++) {
<a name="l00118"></a>00118       m_chpExperimentFileQueue.push_back(
<a name="l00119"></a>00119         rte.m_adtOpts.m_inputExperimentFiles[i]);
<a name="l00120"></a>00120     }
<a name="l00121"></a>00121   }
<a name="l00122"></a>00122 
<a name="l00123"></a>00123   <span class="keywordflow">return</span>;
<a name="l00124"></a>00124 
<a name="l00125"></a>00125 }
<a name="l00126"></a>00126 <span class="comment">// end TranslationInputStreamTableModel::TranslationInputStreamTableModel</span>
<a name="l00127"></a>00127 <span class="comment">/*****************************************************************************/</span>
<a name="l00128"></a>00128 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00129"></a>00129 <span class="comment">/**</span>
<a name="l00130"></a>00130 <span class="comment"> * TranslationInputStreamTableModel::TranslationInputStreamTableModel</span>
<a name="l00131"></a>00131 <span class="comment"> * Synopsis:</span>
<a name="l00132"></a>00132 <span class="comment"> *</span>
<a name="l00133"></a>00133 <span class="comment"> * Tsv constructor.</span>
<a name="l00134"></a>00134 <span class="comment"> *</span>
<a name="l00135"></a>00135 <span class="comment"> * A DMET2 constructor for the Genotype Short Report input of genotype</span>
<a name="l00136"></a>00136 <span class="comment"> * data. The schema for this object is the schema for the Genotype Short</span>
<a name="l00137"></a>00137 <span class="comment"> * Report even for CHP files. </span>
<a name="l00138"></a>00138 <span class="comment"> *</span>
<a name="l00139"></a>00139 <span class="comment"> * @param rte - the single instance run time environment. </span>
<a name="l00140"></a>00140 <span class="comment"> * @param inputTsvFileName - the file with the Genotype Short Report data. </span>
<a name="l00141"></a>00141 <span class="comment"> * @param tsvColumnDefinitions - the Genotype Short Report columns</span>
<a name="l00142"></a>00142 <span class="comment"> * @param tcdSize - the tsvColumnDefinitions size</span>
<a name="l00143"></a>00143 <span class="comment"> * @param strictTsvColumnOrder - when true the columns must be in order. </span>
<a name="l00144"></a>00144 <span class="comment"> * @param tsvExperimentColumn - if column order is not strict, which column contains the experiment. </span>
<a name="l00145"></a>00145 <span class="comment"> * @param tsvGeneColumn - if column order is not strict, which column contains the gene. </span>
<a name="l00146"></a>00146 <span class="comment"> * @param tsvProbeSetColumn - if column order is not strict, which column contains the probe set id.. </span>
<a name="l00147"></a>00147 <span class="comment"> * @param gotm - the single instance Genotype Override Table Model.</span>
<a name="l00148"></a>00148 <span class="comment"> *</span>
<a name="l00149"></a>00149 <span class="comment"> */</span>
<a name="l00150"></a>00150 <span class="comment">/*****************************************************************************/</span>
<a name="l00151"></a><a class="code" href="classTranslationInputStreamTableModel.html#a0eb088c9391c0254b7ea6c6f59f20f18">00151</a> TranslationInputStreamTableModel::TranslationInputStreamTableModel(
<a name="l00152"></a>00152   <span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a>&amp; rte,
<a name="l00153"></a>00153   <span class="keyword">const</span> std::string&amp; inputTsvFileName,
<a name="l00154"></a>00154   <span class="keyword">const</span> <a class="code" href="classTittmColumnDefinition.html">TittmColumnDefinition</a> tsvColumnDefinitions[],
<a name="l00155"></a>00155   <span class="keywordtype">size_t</span> tcdSize,
<a name="l00156"></a>00156   <span class="keywordtype">bool</span> strictTsvColumnOrder,
<a name="l00157"></a>00157   <span class="keywordtype">int</span> tsvExperimentColumn,
<a name="l00158"></a>00158   <span class="keywordtype">int</span> tsvGeneColumn,
<a name="l00159"></a>00159   <span class="keywordtype">int</span> tsvProbeSetColumn,
<a name="l00160"></a>00160   <a class="code" href="classGenotypeOverrideTableModel.html">GenotypeOverrideTableModel</a> *gotm) : m_gotm(gotm)
<a name="l00161"></a>00161 {
<a name="l00162"></a>00162 
<a name="l00163"></a>00163   m_type                 = ADT_EXPERIMENT_STREAM_TYPE_TSV;
<a name="l00164"></a>00164   m_strictTsvColumnOrder = strictTsvColumnOrder;
<a name="l00165"></a>00165   m_tsvExperimentColumn  = tsvExperimentColumn;
<a name="l00166"></a>00166   m_tsvGeneColumn        = tsvGeneColumn;
<a name="l00167"></a>00167   m_tsvFileName          = inputTsvFileName;
<a name="l00168"></a>00168   m_tsvProbeSetColumn    = tsvProbeSetColumn;
<a name="l00169"></a>00169   m_chpData              = NULL;
<a name="l00170"></a>00170   m_genoCallCoder        = NULL;
<a name="l00171"></a>00171   m_callsFileReader          = NULL;
<a name="l00172"></a>00172 
<a name="l00173"></a>00173   _initializeColumnDefinitions(rte, tsvColumnDefinitions, tcdSize);
<a name="l00174"></a>00174 
<a name="l00175"></a>00175 
<a name="l00176"></a>00176   _openTsvFile(rte);
<a name="l00177"></a>00177 
<a name="l00178"></a>00178 
<a name="l00179"></a>00179 }
<a name="l00180"></a>00180 <span class="comment">// end TranslationInputStreamTableModel::TranslationInputStreamTableModel</span>
<a name="l00181"></a>00181 <span class="comment">/*****************************************************************************/</span>
<a name="l00182"></a>00182 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00183"></a>00183 <span class="comment">/**</span>
<a name="l00184"></a>00184 <span class="comment"> * TranslationInputStreamTableModel::columnCount</span>
<a name="l00185"></a>00185 <span class="comment"> * Synopsis:</span>
<a name="l00186"></a>00186 <span class="comment"> * </span>
<a name="l00187"></a>00187 <span class="comment"> * Interogate an already instantiated table for the number of columns.</span>
<a name="l00188"></a>00188 <span class="comment"> * This method assumes ALL rows have the same column width and just</span>
<a name="l00189"></a>00189 <span class="comment"> * returns the width of the column defintion structure. .</span>
<a name="l00190"></a>00190 <span class="comment"> *</span>
<a name="l00191"></a>00191 <span class="comment"> */</span>
<a name="l00192"></a>00192 <span class="comment">/*****************************************************************************/</span>
<a name="l00193"></a><a class="code" href="classTranslationInputStreamTableModel.html#ad6a203e5f8a0b0009b3a43d313b7220b">00193</a> <span class="keywordtype">int</span> <a class="code" href="classTranslationInputStreamTableModel.html#ad6a203e5f8a0b0009b3a43d313b7220b" title="TranslationInputStreamTableModel::columnCount Synopsis:">TranslationInputStreamTableModel::columnCount</a>()<span class="keyword"> const</span>
<a name="l00194"></a>00194 <span class="keyword"></span>{
<a name="l00195"></a>00195 
<a name="l00196"></a>00196   <span class="keywordflow">if</span> (m_tsvColumnDefinition.size()  &gt; 0) {
<a name="l00197"></a>00197     <span class="keywordflow">return</span> m_tsvColumnDefinition.size();
<a name="l00198"></a>00198   }
<a name="l00199"></a>00199 
<a name="l00200"></a>00200   <span class="keywordflow">return</span> 0;
<a name="l00201"></a>00201 
<a name="l00202"></a>00202 }
<a name="l00203"></a>00203 <span class="comment">// end TranslationInputStreamTableModel::columnCount</span>
<a name="l00204"></a>00204 <span class="comment">/*****************************************************************************/</span>
<a name="l00205"></a>00205 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00206"></a>00206 <span class="comment">/**</span>
<a name="l00207"></a>00207 <span class="comment"> * TranslationInputStreamTableModel::clearData</span>
<a name="l00208"></a>00208 <span class="comment"> * Synopsis</span>
<a name="l00209"></a>00209 <span class="comment"> * </span>
<a name="l00210"></a>00210 <span class="comment"> * It is required that memory usage should constant as files and</span>
<a name="l00211"></a>00211 <span class="comment"> * experiments are streamed. This method resets the object to</span>
<a name="l00212"></a>00212 <span class="comment"> * no data other than the meta data required to open the next stream</span>
<a name="l00213"></a>00213 <span class="comment"> * (file). </span>
<a name="l00214"></a>00214 <span class="comment"> *</span>
<a name="l00215"></a>00215 <span class="comment"> * Calls to this API should clear all memory in between experiments.</span>
<a name="l00216"></a>00216 <span class="comment"> *</span>
<a name="l00217"></a>00217 <span class="comment"> */</span>
<a name="l00218"></a>00218 <span class="comment">/*****************************************************************************/</span>
<a name="l00219"></a><a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5">00219</a> <span class="keywordtype">void</span>  <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">TranslationInputStreamTableModel::clearData</a>()
<a name="l00220"></a>00220 {
<a name="l00221"></a>00221 
<a name="l00222"></a>00222   m_rows.clear() ;
<a name="l00223"></a>00223 
<a name="l00224"></a>00224   <span class="keywordflow">if</span> (m_chpData != NULL) {
<a name="l00225"></a>00225     <span class="keyword">delete</span> m_chpData;
<a name="l00226"></a>00226     m_chpData = NULL;
<a name="l00227"></a>00227   }
<a name="l00228"></a>00228 
<a name="l00229"></a>00229 
<a name="l00230"></a>00230   <span class="keywordflow">return</span>;
<a name="l00231"></a>00231 
<a name="l00232"></a>00232 }
<a name="l00233"></a>00233 <span class="comment">// end TranslationInputStreamTableModel::clearData</span>
<a name="l00234"></a>00234 <span class="comment">/*****************************************************************************/</span>
<a name="l00235"></a>00235 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00236"></a>00236 <span class="comment">/**</span>
<a name="l00237"></a>00237 <span class="comment"> * TranslationInputStreamTableModel::getColumnName</span>
<a name="l00238"></a>00238 <span class="comment"> * Synopsis:</span>
<a name="l00239"></a>00239 <span class="comment"> *</span>
<a name="l00240"></a>00240 <span class="comment"> * Get the column name for a column index, i.e. integer -&gt; std::string.</span>
<a name="l00241"></a>00241 <span class="comment"> * It&#39;s an assertion error to pass in an out of bound index. </span>
<a name="l00242"></a>00242 <span class="comment"> *</span>
<a name="l00243"></a>00243 <span class="comment"> * @param column - the 0 based column index for the column in question.</span>
<a name="l00244"></a>00244 <span class="comment"> * </span>
<a name="l00245"></a>00245 <span class="comment"> * @return m_columnName - the name corresponding to column. </span>
<a name="l00246"></a>00246 <span class="comment"> */</span>
<a name="l00247"></a>00247 <span class="comment">/*****************************************************************************/</span>
<a name="l00248"></a><a class="code" href="classTranslationInputStreamTableModel.html#a28ffd80e00964059f7e34c62c177df4a">00248</a> std::string <a class="code" href="classTranslationInputStreamTableModel.html#a28ffd80e00964059f7e34c62c177df4a" title="TranslationInputStreamTableModel::getColumnName Synopsis:">TranslationInputStreamTableModel::getColumnName</a>(<span class="keywordtype">int</span> column)
<a name="l00249"></a>00249 {
<a name="l00250"></a>00250 
<a name="l00251"></a>00251   <span class="comment">// This should always be true unless some constructor is created</span>
<a name="l00252"></a>00252   <span class="comment">// that doesn&#39;t initialize this.</span>
<a name="l00253"></a>00253   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(m_tsvColumnDefinition.size(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00254"></a>00254 
<a name="l00255"></a>00255   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>((column &gt;= 0) &amp;&amp; (column &lt; <a class="code" href="classTranslationInputStreamTableModel.html#ad6a203e5f8a0b0009b3a43d313b7220b" title="TranslationInputStreamTableModel::columnCount Synopsis:">columnCount</a>()), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00256"></a>00256 
<a name="l00257"></a>00257   <span class="keywordflow">return</span> m_tsvColumnDefinition[column].m_columnName;
<a name="l00258"></a>00258 
<a name="l00259"></a>00259 }
<a name="l00260"></a>00260 <span class="comment">// end TranslationInputStreamTableModel::getColumnName</span>
<a name="l00261"></a>00261 <span class="comment">/*****************************************************************************/</span>
<a name="l00262"></a>00262 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00263"></a>00263 <span class="comment">/**</span>
<a name="l00264"></a>00264 <span class="comment"> * TranslationInputStreamTableModel::getRowAsString</span>
<a name="l00265"></a>00265 <span class="comment"> * Synopsis:</span>
<a name="l00266"></a>00266 <span class="comment"> * </span>
<a name="l00267"></a>00267 <span class="comment"> * Debug convienance routine that std::stringifies and entire row. Also</span>
<a name="l00268"></a>00268 <span class="comment"> * used in error messages. All columns are &quot;separated&quot; by a single comma.</span>
<a name="l00269"></a>00269 <span class="comment"> * If the data within a column contains a column there is no attempt made</span>
<a name="l00270"></a>00270 <span class="comment"> * quote the columns. Empty columns will have comma sequences as &quot;,,,&quot;. </span>
<a name="l00271"></a>00271 <span class="comment"> * column 1 2   3 </span>
<a name="l00272"></a>00272 <span class="comment"> * data   A B,C D =&gt; &quot;A,B,C,D&quot;</span>
<a name="l00273"></a>00273 <span class="comment"> * </span>
<a name="l00274"></a>00274 <span class="comment"> * @param row - integer of the row, 0 based</span>
<a name="l00275"></a>00275 <span class="comment"> * </span>
<a name="l00276"></a>00276 <span class="comment"> * @return std::string - the comma separated std::string</span>
<a name="l00277"></a>00277 <span class="comment"> */</span>
<a name="l00278"></a>00278 <span class="comment">/*****************************************************************************/</span>
<a name="l00279"></a><a class="code" href="classTranslationInputStreamTableModel.html#a9dd29d23cbf5a0d0320a905f017fd4ea">00279</a> std::string <a class="code" href="classTranslationInputStreamTableModel.html#a9dd29d23cbf5a0d0320a905f017fd4ea" title="TranslationInputStreamTableModel::getRowAsString Synopsis:">TranslationInputStreamTableModel::getRowAsString</a>(<span class="keyword">const</span> <span class="keywordtype">int</span> row)<span class="keyword">  const</span>
<a name="l00280"></a>00280 <span class="keyword"></span>{
<a name="l00281"></a>00281 
<a name="l00282"></a>00282   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>((row &gt;= 0) &amp;&amp; (row &lt; size()), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00283"></a>00283 
<a name="l00284"></a>00284   std::stringstream rowSStr;
<a name="l00285"></a>00285 
<a name="l00286"></a>00286   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0 ; j &lt; <a class="code" href="classTranslationInputStreamTableModel.html#ad6a203e5f8a0b0009b3a43d313b7220b" title="TranslationInputStreamTableModel::columnCount Synopsis:">columnCount</a>(); j++) {
<a name="l00287"></a>00287     rowSStr &lt;&lt; m_rows[row][j];
<a name="l00288"></a>00288 
<a name="l00289"></a>00289     <span class="keywordflow">if</span> (j &lt; (<a class="code" href="classTranslationInputStreamTableModel.html#ad6a203e5f8a0b0009b3a43d313b7220b" title="TranslationInputStreamTableModel::columnCount Synopsis:">columnCount</a>() - 1)) {
<a name="l00290"></a>00290       rowSStr &lt;&lt; <span class="stringliteral">&quot;,&quot;</span>;
<a name="l00291"></a>00291     }
<a name="l00292"></a>00292 
<a name="l00293"></a>00293   } <span class="comment">// for each column</span>
<a name="l00294"></a>00294 
<a name="l00295"></a>00295   <span class="keywordflow">return</span> rowSStr.str();
<a name="l00296"></a>00296 
<a name="l00297"></a>00297 }
<a name="l00298"></a>00298 <span class="comment">// end TranslationInputStreamTableModel::getRowAsString</span>
<a name="l00299"></a>00299 <span class="comment">/*****************************************************************************/</span>
<a name="l00300"></a>00300 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00301"></a>00301 <span class="comment">/**</span>
<a name="l00302"></a>00302 <span class="comment"> * TranslationInputStreamTableModel::readNextExperiment:</span>
<a name="l00303"></a>00303 <span class="comment"> * Synopsis:</span>
<a name="l00304"></a>00304 <span class="comment"> *</span>
<a name="l00305"></a>00305 <span class="comment"> *  A wrapper that wraps the appropriate TSV for CHP read next experiment</span>
<a name="l00306"></a>00306 <span class="comment"> * method. The TranslationEngine calls this method.       </span>
<a name="l00307"></a>00307 <span class="comment"> *</span>
<a name="l00308"></a>00308 <span class="comment"> * </span>
<a name="l00309"></a>00309 <span class="comment"> * </span>
<a name="l00310"></a>00310 <span class="comment"> * @param rte - the single instance run time environment. </span>
<a name="l00311"></a>00311 <span class="comment"> * @param ttm - the single instance translation table model. </span>
<a name="l00312"></a>00312 <span class="comment"> * @param geneExperimentCopyNumberCall</span>
<a name="l00313"></a>00313 <span class="comment"> *            - The copy number 0 designation per gene std::map. </span>
<a name="l00314"></a>00314 <span class="comment"> * </span>
<a name="l00315"></a>00315 <span class="comment"> * @return - the number of records read or 0 if no more records are left.</span>
<a name="l00316"></a>00316 <span class="comment"> */</span>
<a name="l00317"></a>00317 <span class="comment">/*****************************************************************************/</span>
<a name="l00318"></a><a class="code" href="classTranslationInputStreamTableModel.html#ad0b554465fe47987c3da766bd203a9eb">00318</a> <span class="keywordtype">int</span> <a class="code" href="classTranslationInputStreamTableModel.html#ad0b554465fe47987c3da766bd203a9eb" title="TranslationInputStreamTableModel::readNextExperiment: Synopsis:">TranslationInputStreamTableModel::readNextExperiment</a>(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm, <span class="keyword">const</span> std::map&lt; std::string, std::string&gt; &amp; geneExperimentCopyNumberCall, std::map&lt; std::string, int&gt; &amp; geneCopyNumber)
<a name="l00319"></a>00319 {
<a name="l00320"></a>00320 
<a name="l00321"></a>00321   <span class="keywordflow">if</span> (m_type == ADT_EXPERIMENT_STREAM_TYPE_BATCH_FOLDER) {
<a name="l00322"></a>00322     <span class="keywordflow">return</span> _readNextExperimentBatchFolder(rte, ttm, geneExperimentCopyNumberCall, geneCopyNumber);
<a name="l00323"></a>00323   }
<a name="l00324"></a>00324   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (m_type == ADT_EXPERIMENT_STREAM_TYPE_CALLS_FILE) {
<a name="l00325"></a>00325     <span class="keywordflow">return</span> _readNextExperimentCallsFile(rte, ttm, geneExperimentCopyNumberCall, geneCopyNumber);
<a name="l00326"></a>00326   }
<a name="l00327"></a>00327   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (m_type == ADT_EXPERIMENT_STREAM_TYPE_TSV) {
<a name="l00328"></a>00328     <span class="keywordflow">return</span> _readNextExperimentTsv(rte, ttm, geneExperimentCopyNumberCall);
<a name="l00329"></a>00329   }
<a name="l00330"></a>00330   <span class="keywordflow">else</span> <span class="keywordflow">if</span> (m_type == ADT_EXPERIMENT_STREAM_TYPE_CHP) {
<a name="l00331"></a>00331     <span class="keywordflow">return</span> _readNextExperimentCHP(rte, ttm, geneExperimentCopyNumberCall, geneCopyNumber);
<a name="l00332"></a>00332   }
<a name="l00333"></a>00333 
<a name="l00334"></a>00334   <span class="keywordflow">return</span> 0;
<a name="l00335"></a>00335 
<a name="l00336"></a>00336 }
<a name="l00337"></a>00337 <span class="comment">// end TranslationInputStreamTableModel::readNextExperiment</span>
<a name="l00338"></a>00338 <span class="comment">/*****************************************************************************/</span>
<a name="l00339"></a>00339 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00340"></a>00340 <span class="comment">/**</span>
<a name="l00341"></a>00341 <span class="comment"> * TranslationInputStreamTableModel::sortTsvFileByExperiment:</span>
<a name="l00342"></a>00342 <span class="comment"> * Synopsis:</span>
<a name="l00343"></a>00343 <span class="comment"> *</span>
<a name="l00344"></a>00344 <span class="comment"> * Genotype Short Report files are NOT sorted by experiment, but need</span>
<a name="l00345"></a>00345 <span class="comment"> * to be in order to be streamed. If this program detects an unsorted</span>
<a name="l00346"></a>00346 <span class="comment"> * Genotype Short Report then it aborts. However, before it aborts</span>
<a name="l00347"></a>00347 <span class="comment"> * it sorts the TSV file and creates a new copy of the file</span>
<a name="l00348"></a>00348 <span class="comment"> * with &quot;.sorted&quot; as an extension. In this way the user can</span>
<a name="l00349"></a>00349 <span class="comment"> * simply call the program a second time using the &quot;.sorted&quot; file</span>
<a name="l00350"></a>00350 <span class="comment"> * and then continue on. </span>
<a name="l00351"></a>00351 <span class="comment"> * </span>
<a name="l00352"></a>00352 <span class="comment"> * This is utility routine used independent of translating calls. Use this</span>
<a name="l00353"></a>00353 <span class="comment"> * routine with its own option switch as a convienance to sort</span>
<a name="l00354"></a>00354 <span class="comment"> * an input genotype short report file. Sorting is a requirement</span>
<a name="l00355"></a>00355 <span class="comment"> * for streaming. </span>
<a name="l00356"></a>00356 <span class="comment"> *</span>
<a name="l00357"></a>00357 <span class="comment"> *</span>
<a name="l00358"></a>00358 <span class="comment"> * @rte - the single instance RunTimeEnvironment which contains the name of the file to be sorted. </span>
<a name="l00359"></a>00359 <span class="comment"> *</span>
<a name="l00360"></a>00360 <span class="comment"> * @return - void, creates a new file of sorted records. Throws exceptions</span>
<a name="l00361"></a>00361 <span class="comment"> *           on IO errors.</span>
<a name="l00362"></a>00362 <span class="comment"> */</span>
<a name="l00363"></a>00363 <span class="comment">/*****************************************************************************/</span>
<a name="l00364"></a><a class="code" href="classTranslationInputStreamTableModel.html#ab3371d6ef32696cdaccbfc30790a4bd6">00364</a> <span class="keywordtype">void</span> <a class="code" href="classTranslationInputStreamTableModel.html#ab3371d6ef32696cdaccbfc30790a4bd6" title="TranslationInputStreamTableModel::sortTsvFileByExperiment: Synopsis:">TranslationInputStreamTableModel::sortTsvFileByExperiment</a>(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte)
<a name="l00365"></a>00365 {
<a name="l00366"></a>00366 
<a name="l00367"></a>00367   <span class="comment">// This should never happen unless a new constructor is created</span>
<a name="l00368"></a>00368   <span class="comment">// that doesn&#39;t initialize columns.</span>
<a name="l00369"></a>00369   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(m_tsvColumnDefinition.size(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00370"></a>00370 
<a name="l00371"></a>00371   <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">TsvFile</a> sortTsv;
<a name="l00372"></a>00372 
<a name="l00373"></a>00373   std::string sortName = m_tsvFileName + <span class="stringliteral">&quot;.sort.txt&quot;</span>;
<a name="l00374"></a>00374 
<a name="l00375"></a>00375   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, <span class="stringliteral">&quot;sortTsvFileByExperiment: &quot;</span> + m_tsvFileName
<a name="l00376"></a>00376                +  <span class="stringliteral">&quot; -&gt; &quot;</span> + sortName);
<a name="l00377"></a>00377 
<a name="l00378"></a>00378   <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; m_tsvColumnDefinition.size(); i++) {
<a name="l00379"></a>00379 
<a name="l00380"></a>00380     sortTsv.defineColumn(0, i, m_tsvColumnDefinition[i].m_columnName);
<a name="l00381"></a>00381   }
<a name="l00382"></a>00382 
<a name="l00383"></a>00383   <span class="comment">//</span>
<a name="l00384"></a>00384   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a0b08c1eea21d4b6a893f57df16521a46" title="remove whitespace from value?">m_optAutoTrim</a>   = <span class="keyword">true</span>; <span class="comment">// remove &#39;&quot;&#39;s</span>
<a name="l00385"></a>00385   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7b96cab331d9eaec414cfeed75a55e3d" title="Quoting Character.">m_optQuoteChar1</a> = 0;    <span class="comment">// ignore &quot;&#39;&quot;s</span>
<a name="l00386"></a>00386 
<a name="l00387"></a>00387 
<a name="l00388"></a>00388   <span class="keywordflow">if</span> (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(m_tsvFileName) != TSV_OK) {
<a name="l00389"></a>00389     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_tsvFileName + <span class="stringliteral">&quot;: failed opening input Tsv file.&quot;</span>);
<a name="l00390"></a>00390   }
<a name="l00391"></a>00391 
<a name="l00392"></a>00392   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a6aad66cac289e3eeed0bc06ba0e2587b" title="Start at the beginning of the headers ///.">headersBegin</a>();
<a name="l00393"></a>00393 
<a name="l00394"></a>00394   std::string key;
<a name="l00395"></a>00395   std::string value;
<a name="l00396"></a>00396   sortTsv.addHeaderComment(<span class="stringliteral">&quot; File generated by sortTsvFileByExperiment.&quot;</span>);
<a name="l00397"></a>00397   <span class="keywordflow">while</span> (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a86c80445dd06ed593e194f85e8ad8946" title="Advance to the next header of the file and set the values ///.">headersNext</a>(key, value) == TSV_OK) {
<a name="l00398"></a>00398     sortTsv.addHeader(key, value);
<a name="l00399"></a>00399   }
<a name="l00400"></a>00400 
<a name="l00401"></a>00401   sortTsv.writeTsv_v1(sortName);
<a name="l00402"></a>00402 
<a name="l00403"></a>00403   <span class="keywordtype">int</span> row = 0;
<a name="l00404"></a>00404   <span class="keywordflow">while</span> (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == TSV_OK) {
<a name="l00405"></a>00405 
<a name="l00406"></a>00406     m_rows.resize(row + 1);
<a name="l00407"></a>00407     m_rows[row].resize(m_tsvColumnCount, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00408"></a>00408 
<a name="l00409"></a>00409     <span class="comment">// Read in a row, one column at a time.</span>
<a name="l00410"></a>00410     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> tsvColumn = 0; tsvColumn &lt; m_tsvColumnCount; tsvColumn++) {
<a name="l00411"></a>00411       std::string columnValue;
<a name="l00412"></a>00412       <span class="keywordflow">if</span> (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0, tsvColumn, columnValue) != TSV_OK) {
<a name="l00413"></a>00413         <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;tsv.get != TSV_OK&quot;</span>);
<a name="l00414"></a>00414       }
<a name="l00415"></a>00415       <span class="keywordflow">if</span> (!columnValue.empty()) {
<a name="l00416"></a>00416         m_rows[row][m_tsvIndexToColumnDefinitionIndex[tsvColumn]] = columnValue;
<a name="l00417"></a>00417       }
<a name="l00418"></a>00418     }
<a name="l00419"></a>00419 
<a name="l00420"></a>00420     m_experimentRowIndex[m_rows[row][m_tsvExperimentColumn]].push_back(row);
<a name="l00421"></a>00421     row++;
<a name="l00422"></a>00422   }
<a name="l00423"></a>00423 
<a name="l00424"></a>00424   std::map&lt;std::string, std::vector&lt; int &gt; &gt;::iterator itSV;
<a name="l00425"></a>00425 
<a name="l00426"></a>00426   row = 0;
<a name="l00427"></a>00427   <span class="keywordflow">for</span> (itSV = m_experimentRowIndex.begin(); itSV != m_experimentRowIndex.end(); itSV++) {
<a name="l00428"></a>00428 
<a name="l00429"></a>00429     <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; itSV-&gt;second.size() ; i++) {
<a name="l00430"></a>00430       <span class="keywordflow">for</span> (<span class="keywordtype">int</span> tsvColumn = 0; tsvColumn &lt; m_tsvColumnCount; tsvColumn++) {
<a name="l00431"></a>00431         sortTsv.set(0, tsvColumn,  m_rows[itSV-&gt;second[i]][tsvColumn]);
<a name="l00432"></a>00432       }
<a name="l00433"></a>00433       sortTsv.writeLevel(0);
<a name="l00434"></a>00434       row++;
<a name="l00435"></a>00435     }
<a name="l00436"></a>00436   }
<a name="l00437"></a>00437 
<a name="l00438"></a>00438   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(row) + <span class="stringliteral">&quot; sorted Genotype records written to &quot;</span> + sortName);
<a name="l00439"></a>00439 
<a name="l00440"></a>00440   sortTsv.close();
<a name="l00441"></a>00441 
<a name="l00442"></a>00442   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00443"></a>00443 
<a name="l00444"></a>00444   <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">clearData</a>();
<a name="l00445"></a>00445 
<a name="l00446"></a>00446   <span class="keywordflow">return</span>;
<a name="l00447"></a>00447 
<a name="l00448"></a>00448 }
<a name="l00449"></a>00449 <span class="comment">// end TranslationInputStreamTableModel::sortTsvFileByExperiment</span>
<a name="l00450"></a>00450 <span class="comment">/*****************************************************************************/</span>
<a name="l00451"></a>00451 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00452"></a>00452 <span class="comment">/**</span>
<a name="l00453"></a>00453 <span class="comment"> * TranslationInputStreamTableModel::_closeTsvFile:</span>
<a name="l00454"></a>00454 <span class="comment"> * Synopsis:</span>
<a name="l00455"></a>00455 <span class="comment"> *</span>
<a name="l00456"></a>00456 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_close().</span>
<a name="l00457"></a>00457 <span class="comment"> *</span>
<a name="l00458"></a>00458 <span class="comment"> * Call tsv.close() on the m_tsv. Made a wrapper in case messaging</span>
<a name="l00459"></a>00459 <span class="comment"> * or further clean up work is ever  needed. </span>
<a name="l00460"></a>00460 <span class="comment"> *</span>
<a name="l00461"></a>00461 <span class="comment"> *</span>
<a name="l00462"></a>00462 <span class="comment"> * @return - void</span>
<a name="l00463"></a>00463 <span class="comment"> */</span>
<a name="l00464"></a>00464 <span class="comment">/*****************************************************************************/</span>
<a name="l00465"></a>00465 <span class="keywordtype">void</span> TranslationInputStreamTableModel::_closeTsvFile()
<a name="l00466"></a>00466 {
<a name="l00467"></a>00467 
<a name="l00468"></a>00468   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#ae7932be2156c04a562f327b53833aed7" title="Close the file.">close</a>();
<a name="l00469"></a>00469 
<a name="l00470"></a>00470 }
<a name="l00471"></a>00471 <span class="comment">// end TranslationInputStreamTableModel::_closeTsvFile</span>
<a name="l00472"></a>00472 <span class="comment">/*****************************************************************************/</span>
<a name="l00473"></a>00473 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00474"></a>00474 <span class="comment">/**</span>
<a name="l00475"></a>00475 <span class="comment"> * TranslationInputStreamTableModel::_getAllelesFromProbeSet:</span>
<a name="l00476"></a>00476 <span class="comment"> * Synopsis:</span>
<a name="l00477"></a>00477 <span class="comment"> *</span>
<a name="l00478"></a>00478 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_readNextExperimentCHP</span>
<a name="l00479"></a>00479 <span class="comment"> *</span>
<a name="l00480"></a>00480 <span class="comment"> * Translates a probe set id&#39;s textual &#39;allele1&#39; and &#39;allele2&#39;</span>
<a name="l00481"></a>00481 <span class="comment"> * from the CHP file using the APIs found in the GenoCallCoder. </span>
<a name="l00482"></a>00482 <span class="comment"> *</span>
<a name="l00483"></a>00483 <span class="comment"> *  The CHP calls are encoded as &quot;A/A&quot; or &quot;A/B&quot; or &quot;B/B&quot;, etc. These</span>
<a name="l00484"></a>00484 <span class="comment"> *  need to be translated into the actual alleles in the annotation file.</span>
<a name="l00485"></a>00485 <span class="comment"> *  Do the look up in the GenoCallCoder and return the bases.</span>
<a name="l00486"></a>00486 <span class="comment"> * </span>
<a name="l00487"></a>00487 <span class="comment"> * Note: the report alleles are not retrieved here but instead are</span>
<a name="l00488"></a>00488 <span class="comment"> * imported at start up time using all the probe sets and alleles in the</span>
<a name="l00489"></a>00489 <span class="comment"> * translation table.</span>
<a name="l00490"></a>00490 <span class="comment"> *</span>
<a name="l00491"></a>00491 <span class="comment"> * Also note: the CHP will contain many more probe sets then are defined in</span>
<a name="l00492"></a>00492 <span class="comment"> * the translation table. When this API returns false then this is</span>
<a name="l00493"></a>00493 <span class="comment"> * the deciding factor when to ignore a probe set marker in the</span>
<a name="l00494"></a>00494 <span class="comment"> * CHP file and the data is dropped on the floor. </span>
<a name="l00495"></a>00495 <span class="comment"> *</span>
<a name="l00496"></a>00496 <span class="comment"> * @param rte - the single instance RunTimeEnvironment.</span>
<a name="l00497"></a>00497 <span class="comment"> * @param ttm - the single instance translation table.</span>
<a name="l00498"></a>00498 <span class="comment"> * @param dbad - the DmetBiAllelicData with probe set and call data.</span>
<a name="l00499"></a>00499 <span class="comment"> *               Note: BiAllelic is a misnomer, multi allelic is represented as</span>
<a name="l00500"></a>00500 <span class="comment"> *               well.</span>
<a name="l00501"></a>00501 <span class="comment"> * @param genoCallDecoder - the object that does translation from &quot;A/A&quot; -&gt; &quot;TGT/TTT&quot;.</span>
<a name="l00502"></a>00502 <span class="comment"> * @param allele1 - returned base for the ALLELE1 column.</span>
<a name="l00503"></a>00503 <span class="comment"> * @param allele2 - returned base for the ALLELE2 column.</span>
<a name="l00504"></a>00504 <span class="comment"> *</span>
<a name="l00505"></a>00505 <span class="comment"> * @return - true if the probeSet exists  in the translation table,</span>
<a name="l00506"></a>00506 <span class="comment"> *           the allele1 and allele2 std::strings</span>
<a name="l00507"></a>00507 <span class="comment"> */</span>
<a name="l00508"></a>00508 <span class="comment">/*****************************************************************************/</span>
<a name="l00509"></a>00509 <span class="keywordtype">bool</span> TranslationInputStreamTableModel::_getAllelesFromProbeSet(<span class="keyword">const</span> <span class="keyword">class</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm,  <a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html">affymetrix_calvin_data::DmetBiAllelicData</a> &amp; dbad,  std::string &amp; allele1, std::string &amp; allele2)
<a name="l00510"></a>00510 {
<a name="l00511"></a>00511 
<a name="l00512"></a>00512 
<a name="l00513"></a>00513 
<a name="l00514"></a>00514   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(!dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>.empty(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l00515"></a>00515 
<a name="l00516"></a>00516   std::string probeSet    = dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>;
<a name="l00517"></a>00517 
<a name="l00518"></a>00518   <span class="keywordflow">if</span> (ttm.m_probeSetRowIndex.count(probeSet) == 0) {
<a name="l00519"></a>00519     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00520"></a>00520   }
<a name="l00521"></a>00521 
<a name="l00522"></a>00522 
<a name="l00523"></a>00523   std::string decodedAbstractCall;
<a name="l00524"></a>00524   std::string decodedCall;
<a name="l00525"></a>00525   std::string reportCall;
<a name="l00526"></a>00526 
<a name="l00527"></a>00527   allele1.clear();
<a name="l00528"></a>00528   allele2.clear();
<a name="l00529"></a>00529 
<a name="l00530"></a>00530   <span class="keywordtype">bool</span> probeSetSwitchStrand = ((ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3) &amp;&amp; ttm.m_rows[ttm.m_probeSetRowIndex[probeSet][0]][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_SWITCH_STRAND)] == <span class="stringliteral">&quot;Y&quot;</span>);
<a name="l00531"></a>00531 
<a name="l00532"></a>00532 
<a name="l00533"></a>00533   decodedCall = m_genoCallCoder-&gt;<a class="code" href="classGenoCallCoder.html#aa50308f74c47225263f4abd43686e257" title="Decode from call code number to design allele name string for given probeset id.">genotypeCallNumToReferenceAllele</a>(probeSet, dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a1ee624790702d9a9ad16b97e29ce3e58">call</a>);
<a name="l00534"></a>00534 
<a name="l00535"></a>00535   <span class="keywordflow">if</span> (decodedCall.empty()) {
<a name="l00536"></a>00536     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Empty reference call returned by abstractAlleleToReferenceAllele.&quot;</span>);
<a name="l00537"></a>00537   }
<a name="l00538"></a>00538 
<a name="l00539"></a>00539   std::string first, second, copyNumber;
<a name="l00540"></a>00540 
<a name="l00541"></a>00541 
<a name="l00542"></a>00542   <span class="keywordflow">if</span> (!pcrecpp::RE(<span class="stringliteral">&quot;^([\\w\\d\\-]+)$&quot;</span>).FullMatch(decodedCall, &amp;first)) {
<a name="l00543"></a>00543 
<a name="l00544"></a>00544     <span class="keywordflow">if</span> (! pcrecpp::RE(<span class="stringliteral">&quot;^(\\w+|\\-)/(\\w+|\\-)$&quot;</span>).FullMatch(decodedCall, &amp;first, &amp;second)) {
<a name="l00545"></a>00545       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(probeSet + <span class="stringliteral">&quot;: unknown call? &quot;</span> + decodedCall);
<a name="l00546"></a>00546     }
<a name="l00547"></a>00547   }
<a name="l00548"></a>00548 
<a name="l00549"></a>00549   <span class="comment">// Copy Number 0 translation</span>
<a name="l00550"></a>00550 
<a name="l00551"></a>00551   <span class="keywordflow">if</span> ((first == <span class="stringliteral">&quot;17&quot;</span>) || (first == <span class="stringliteral">&quot;ZeroCopyNumber&quot;</span>)) {
<a name="l00552"></a>00552     first = <span class="stringliteral">&quot;0&quot;</span>;
<a name="l00553"></a>00553   }
<a name="l00554"></a>00554 
<a name="l00555"></a>00555   allele1 = first;
<a name="l00556"></a>00556 
<a name="l00557"></a>00557   <span class="keywordflow">if</span> (first == <span class="stringliteral">&quot;NotAvailable&quot;</span>) {
<a name="l00558"></a>00558     <span class="comment">// Returning false causes the marker to be ignored, exactly the behavior</span>
<a name="l00559"></a>00559     <span class="comment">// required.</span>
<a name="l00560"></a>00560     <span class="comment">//    return false;</span>
<a name="l00561"></a>00561   }
<a name="l00562"></a>00562 
<a name="l00563"></a>00563   <span class="keywordflow">if</span> (!second.empty()) {
<a name="l00564"></a>00564     allele2 = second;
<a name="l00565"></a>00565   }
<a name="l00566"></a>00566 
<a name="l00567"></a>00567 
<a name="l00568"></a>00568   <span class="keywordflow">if</span> (probeSetSwitchStrand) {
<a name="l00569"></a>00569 
<a name="l00570"></a>00570     allele1 = <a class="code" href="classCallElement.html#a11386b2d1d7a7a2497f51a3112567f9a" title="CallElement::reverseComplement.">CallElement::reverseComplement</a>(allele1);
<a name="l00571"></a>00571     allele2 = <a class="code" href="classCallElement.html#a11386b2d1d7a7a2497f51a3112567f9a" title="CallElement::reverseComplement.">CallElement::reverseComplement</a>(allele2);
<a name="l00572"></a>00572 
<a name="l00573"></a>00573   }
<a name="l00574"></a>00574 
<a name="l00575"></a>00575   <span class="keywordflow">return</span> <span class="keyword">true</span>;
<a name="l00576"></a>00576 
<a name="l00577"></a>00577 }
<a name="l00578"></a>00578 <span class="comment">// end TranslationInputStreamTableModel::_getAllelesFromProbeSet</span>
<a name="l00579"></a>00579 
<a name="l00580"></a>00580 <span class="keywordtype">bool</span> TranslationInputStreamTableModel::_getAllelesFromAbstractAlleles(<span class="keyword">const</span> <span class="keyword">class</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm, 
<a name="l00581"></a>00581                                                 <span class="keyword">const</span> std::string&amp; sProbeset, <span class="keyword">const</span> std::string&amp; sCall, std::string &amp; allele1, std::string &amp; allele2)
<a name="l00582"></a>00582 {
<a name="l00583"></a>00583 
<a name="l00584"></a>00584   <span class="keywordflow">if</span> (ttm.m_probeSetRowIndex.count(sProbeset) == 0) {
<a name="l00585"></a>00585     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00586"></a>00586   }
<a name="l00587"></a>00587 
<a name="l00588"></a>00588 
<a name="l00589"></a>00589   std::string decodedAbstractCall;
<a name="l00590"></a>00590   std::string decodedCall;
<a name="l00591"></a>00591   std::string reportCall;
<a name="l00592"></a>00592 
<a name="l00593"></a>00593   allele1.clear();
<a name="l00594"></a>00594   allele2.clear();
<a name="l00595"></a>00595 
<a name="l00596"></a>00596   <span class="keywordtype">bool</span> probeSetSwitchStrand = ((ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3) &amp;&amp; ttm.m_rows[ttm.m_probeSetRowIndex[sProbeset][0]][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_SWITCH_STRAND)] == <span class="stringliteral">&quot;Y&quot;</span>);
<a name="l00597"></a>00597 
<a name="l00598"></a>00598 
<a name="l00599"></a>00599   <span class="keywordflow">if</span> (sCall == <span class="stringliteral">&quot;NoCall&quot;</span>) {
<a name="l00600"></a>00600     decodedCall = <span class="stringliteral">&quot;NoCall/NoCall&quot;</span>;
<a name="l00601"></a>00601   }
<a name="l00602"></a>00602   <span class="keywordflow">else</span> {
<a name="l00603"></a>00603     decodedCall = m_genoCallCoder-&gt;<a class="code" href="classGenoCallCoder.html#ae14bfa54eac30fa7d2b303a960d36c31" title="Translate abstract allele to design allele name for given probeset id.">abstractAlleleToReferenceAllele</a>(sProbeset, sCall);
<a name="l00604"></a>00604   }
<a name="l00605"></a>00605 
<a name="l00606"></a>00606   <span class="keywordflow">if</span> (decodedCall.empty()) {
<a name="l00607"></a>00607     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Empty reference call returned by abstractAlleleToReferenceAllele.&quot;</span>);
<a name="l00608"></a>00608   }
<a name="l00609"></a>00609 
<a name="l00610"></a>00610   std::string first, second, copyNumber;
<a name="l00611"></a>00611 
<a name="l00612"></a>00612 
<a name="l00613"></a>00613   <span class="keywordflow">if</span> (!pcrecpp::RE(<span class="stringliteral">&quot;^([\\w\\d\\-]+)$&quot;</span>).FullMatch(decodedCall, &amp;first)) {
<a name="l00614"></a>00614 
<a name="l00615"></a>00615     <span class="keywordflow">if</span> (! pcrecpp::RE(<span class="stringliteral">&quot;^(\\w+|\\-)/(\\w+|\\-)$&quot;</span>).FullMatch(decodedCall, &amp;first, &amp;second)) {
<a name="l00616"></a>00616       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(sProbeset + <span class="stringliteral">&quot;: unknown call? &quot;</span> + decodedCall);
<a name="l00617"></a>00617     }
<a name="l00618"></a>00618   }
<a name="l00619"></a>00619 
<a name="l00620"></a>00620   <span class="comment">// Copy Number 0 translation</span>
<a name="l00621"></a>00621 
<a name="l00622"></a>00622   <span class="keywordflow">if</span> ((first == <span class="stringliteral">&quot;17&quot;</span>) || (first == <span class="stringliteral">&quot;ZeroCopyNumber&quot;</span>)) {
<a name="l00623"></a>00623     first = <span class="stringliteral">&quot;0&quot;</span>;
<a name="l00624"></a>00624   }
<a name="l00625"></a>00625 
<a name="l00626"></a>00626   allele1 = first;
<a name="l00627"></a>00627 
<a name="l00628"></a>00628   <span class="keywordflow">if</span> (first == <span class="stringliteral">&quot;NotAvailable&quot;</span>) {
<a name="l00629"></a>00629     <span class="comment">// Returning false causes the marker to be ignored, exactly the behavior</span>
<a name="l00630"></a>00630     <span class="comment">// required.</span>
<a name="l00631"></a>00631     <span class="comment">//    return false;</span>
<a name="l00632"></a>00632   }
<a name="l00633"></a>00633 
<a name="l00634"></a>00634   <span class="keywordflow">if</span> (!second.empty()) {
<a name="l00635"></a>00635     allele2 = second;
<a name="l00636"></a>00636   }
<a name="l00637"></a>00637 
<a name="l00638"></a>00638 
<a name="l00639"></a>00639   <span class="keywordflow">if</span> (probeSetSwitchStrand) {
<a name="l00640"></a>00640 
<a name="l00641"></a>00641     allele1 = <a class="code" href="classCallElement.html#a11386b2d1d7a7a2497f51a3112567f9a" title="CallElement::reverseComplement.">CallElement::reverseComplement</a>(allele1);
<a name="l00642"></a>00642     allele2 = <a class="code" href="classCallElement.html#a11386b2d1d7a7a2497f51a3112567f9a" title="CallElement::reverseComplement.">CallElement::reverseComplement</a>(allele2);
<a name="l00643"></a>00643 
<a name="l00644"></a>00644   }
<a name="l00645"></a>00645 
<a name="l00646"></a>00646   <span class="keywordflow">return</span> <span class="keyword">true</span>;
<a name="l00647"></a>00647 
<a name="l00648"></a>00648 }
<a name="l00649"></a>00649 
<a name="l00650"></a>00650 <span class="comment">/*****************************************************************************/</span>
<a name="l00651"></a>00651 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00652"></a>00652 <span class="comment">/**</span>
<a name="l00653"></a>00653 <span class="comment"> * TranslationInputStreamTableModel::_getGeneFromProbeSet:</span>
<a name="l00654"></a>00654 <span class="comment"> * Synopsis:</span>
<a name="l00655"></a>00655 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_readNextExperimentCHP</span>
<a name="l00656"></a>00656 <span class="comment"> *</span>
<a name="l00657"></a>00657 <span class="comment"> * The CHP file doesn&#39;t contain the gene. However, the assay id (probe set)</span>
<a name="l00658"></a>00658 <span class="comment"> * is distinct per translation table so it can be looked up in the</span>
<a name="l00659"></a>00659 <span class="comment"> * translation table.</span>
<a name="l00660"></a>00660 <span class="comment"> *</span>
<a name="l00661"></a>00661 <span class="comment"> * @param probeSet - the marker for which to get the gene</span>
<a name="l00662"></a>00662 <span class="comment"> * @param ttm - the single instance translation table containing the gene</span>
<a name="l00663"></a>00663 <span class="comment"> *</span>
<a name="l00664"></a>00664 <span class="comment"> * @return - gene as std::string, empty if the probeSet doesn&#39;t exist</span>
<a name="l00665"></a>00665 <span class="comment"> */</span>
<a name="l00666"></a>00666 <span class="comment">/*****************************************************************************/</span>
<a name="l00667"></a>00667 std::string TranslationInputStreamTableModel::_getGeneFromProbeSet(<span class="keyword">const</span> std::string probeSet,   <span class="keyword">const</span> <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm)
<a name="l00668"></a>00668 {
<a name="l00669"></a>00669 
<a name="l00670"></a>00670   std::string gene;
<a name="l00671"></a>00671 
<a name="l00672"></a>00672   <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(probeSet) &lt; 0) {
<a name="l00673"></a>00673     <span class="keywordflow">return</span> gene;
<a name="l00674"></a>00674   }
<a name="l00675"></a>00675 
<a name="l00676"></a>00676   gene = ttm.m_rows[ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(probeSet)][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_GENE)];
<a name="l00677"></a>00677 
<a name="l00678"></a>00678   <span class="keywordflow">return</span> gene;
<a name="l00679"></a>00679 }
<a name="l00680"></a>00680 <span class="comment">// end TranslationInputStreamTableModel::_getGeneFromProbeSet</span>
<a name="l00681"></a>00681 <span class="comment">/*****************************************************************************/</span>
<a name="l00682"></a>00682 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00683"></a>00683 <span class="comment">/**</span>
<a name="l00684"></a>00684 <span class="comment"> * searchParams</span>
<a name="l00685"></a>00685 <span class="comment"> * Synopsis:</span>
<a name="l00686"></a>00686 <span class="comment"> *</span>
<a name="l00687"></a>00687 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_getCHPGuid and</span>
<a name="l00688"></a>00688 <span class="comment"> * TranslationInputStreamTableModel::_initializeNextExperimentCHPFile.</span>
<a name="l00689"></a>00689 <span class="comment"> *</span>
<a name="l00690"></a>00690 <span class="comment"> * The calvin_files directory contains APIs to search for headers  or</span>
<a name="l00691"></a>00691 <span class="comment"> * parameters in the CHP file&#39;s various inheritance tree. However, these</span>
<a name="l00692"></a>00692 <span class="comment"> * APIs are specific the the class implemented for. This search</span>
<a name="l00693"></a>00693 <span class="comment"> * searchs the entire inheritance graph in one convienant API.</span>
<a name="l00694"></a>00694 <span class="comment"> * This function also convienantly handles all the conversion from</span>
<a name="l00695"></a>00695 <span class="comment"> * WCS to std::string. </span>
<a name="l00696"></a>00696 <span class="comment"> *</span>
<a name="l00697"></a>00697 <span class="comment"> * Note: This is not the most efficient API but the two functions</span>
<a name="l00698"></a>00698 <span class="comment"> * being helped are</span>
<a name="l00699"></a>00699 <span class="comment"> * initialization functions only used at start up and hence performance</span>
<a name="l00700"></a>00700 <span class="comment"> * is not a factor. </span>
<a name="l00701"></a>00701 <span class="comment"> *</span>
<a name="l00702"></a>00702 <span class="comment"> * @param gdh - GenericDataHeader from m_chpData</span>
<a name="l00703"></a>00703 <span class="comment"> * @params searchName - what to search for.</span>
<a name="l00704"></a>00704 <span class="comment"> * @params searchValue - the return value. </span>
<a name="l00705"></a>00705 <span class="comment"> *</span>
<a name="l00706"></a>00706 <span class="comment"> * @return - searchValue, if found</span>
<a name="l00707"></a>00707 <span class="comment"> */</span>
<a name="l00708"></a>00708 <span class="comment">/*****************************************************************************/</span>
<a name="l00709"></a>00709 <span class="keyword">static</span> <span class="keywordtype">void</span> searchParams(<a class="code" href="classaffymetrix__calvin__io_1_1GenericDataHeader.html">affymetrix_calvin_io::GenericDataHeader</a> &amp; gdh, <span class="keyword">const</span> std::string searchName ,  std::string &amp;searchValue )
<a name="l00710"></a>00710 {
<a name="l00711"></a>00711 
<a name="l00712"></a>00712   <span class="keywordflow">if</span> (!searchValue.empty()) {
<a name="l00713"></a>00713     <span class="keywordflow">return</span>;
<a name="l00714"></a>00714   }
<a name="l00715"></a>00715 
<a name="l00716"></a>00716   ParameterNameValueTypeIt itPNVT;
<a name="l00717"></a>00717   ParameterNameValueTypeIt jtPNVT;
<a name="l00718"></a>00718 
<a name="l00719"></a>00719   gdh.<a class="code" href="classaffymetrix__calvin__io_1_1GenericDataHeader.html#ace9b5edeef5b6f87241159fc91d44e0f">GetNameValIterators</a>(itPNVT, jtPNVT);
<a name="l00720"></a>00720 
<a name="l00721"></a>00721   <span class="keywordflow">for</span> (; itPNVT != jtPNVT; itPNVT++) {
<a name="l00722"></a>00722 
<a name="l00723"></a>00723     <span class="keywordflow">if</span> (StringUtils::ConvertWCSToMBS(itPNVT-&gt;GetName()) == searchName ) {
<a name="l00724"></a>00724       searchValue = StringUtils::ConvertWCSToMBS(itPNVT-&gt;ToString());
<a name="l00725"></a>00725       <span class="keywordflow">return</span>;
<a name="l00726"></a>00726     }
<a name="l00727"></a>00727   }
<a name="l00728"></a>00728 
<a name="l00729"></a>00729   std::vector&lt;GenericDataHeader&gt;::iterator itGDH;
<a name="l00730"></a>00730   std::vector&lt;GenericDataHeader&gt;::iterator jtGDH;
<a name="l00731"></a>00731 
<a name="l00732"></a>00732   gdh.<a class="code" href="classaffymetrix__calvin__io_1_1GenericDataHeader.html#a85d01fa491de6b637e5c700cfe4d6a6c">GetParentIterators</a>(itGDH, jtGDH);
<a name="l00733"></a>00733 
<a name="l00734"></a>00734   <span class="keywordflow">for</span> (; itGDH != jtGDH ; itGDH++) {
<a name="l00735"></a>00735     searchParams(*itGDH, searchName, searchValue);
<a name="l00736"></a>00736   }
<a name="l00737"></a>00737 
<a name="l00738"></a>00738   <span class="keywordflow">return</span>;
<a name="l00739"></a>00739 
<a name="l00740"></a>00740 }
<a name="l00741"></a>00741 <span class="comment">// end searchParams</span>
<a name="l00742"></a>00742 <span class="comment">/*****************************************************************************/</span>
<a name="l00743"></a>00743 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00744"></a>00744 <span class="comment">/**</span>
<a name="l00745"></a>00745 <span class="comment"> * TranslationInputStreamTableModel::_getCHPGuid:</span>
<a name="l00746"></a>00746 <span class="comment"> * Synopsis:</span>
<a name="l00747"></a>00747 <span class="comment"> *</span>
<a name="l00748"></a>00748 <span class="comment"> * Helper function to _initializeNextExperimentCHPFile</span>
<a name="l00749"></a>00749 <span class="comment"> * </span>
<a name="l00750"></a>00750 <span class="comment"> *   Returns the CHP file assay id. Broken out due to confusion about</span>
<a name="l00751"></a>00751 <span class="comment"> * which GUID is the right guid. Also, some prototype CHP files stuff</span>
<a name="l00752"></a>00752 <span class="comment"> * the GUID in a different place than that required by the </span>
<a name="l00753"></a>00753 <span class="comment"> * canonical gd.ArrayIdentifier(). If gd.ArrayIdentifier fails then</span>
<a name="l00754"></a>00754 <span class="comment"> * searchParams is called. Only prototype CHPS files have exhibited this</span>
<a name="l00755"></a>00755 <span class="comment"> * behavior. </span>
<a name="l00756"></a>00756 <span class="comment"> *</span>
<a name="l00757"></a>00757 <span class="comment"> * </span>
<a name="l00758"></a>00758 <span class="comment"> * </span>
<a name="l00759"></a>00759 <span class="comment"> * @param rte - the single instance run time environment. </span>
<a name="l00760"></a>00760 <span class="comment"> * @param gd - GenericData from m_chpData</span>
<a name="l00761"></a>00761 <span class="comment"> * </span>
<a name="l00762"></a>00762 <span class="comment"> * @return - guid, The GUID std::string.</span>
<a name="l00763"></a>00763 <span class="comment"> */</span>
<a name="l00764"></a>00764 <span class="comment">/*****************************************************************************/</span>
<a name="l00765"></a>00765 <span class="comment">/*****************************************************************************/</span>
<a name="l00766"></a>00766 std::string TranslationInputStreamTableModel::_getCHPGuid(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classaffymetrix__calvin__io_1_1GenericData.html">affymetrix_calvin_io::GenericData</a> &amp; gd)
<a name="l00767"></a>00767 {
<a name="l00768"></a>00768 
<a name="l00769"></a>00769 
<a name="l00770"></a>00770   <span class="comment">//m_chpGuid = m_chpData-&gt;GetFileHeader()-&gt;GetGenericDataHdr()-&gt;GetFileId();</span>
<a name="l00771"></a>00771   <span class="comment">//m_chpGuid = gd.ArrayIdentifier();</span>
<a name="l00772"></a>00772 
<a name="l00773"></a>00773   AffymetrixGuidType guid;
<a name="l00774"></a>00774 
<a name="l00775"></a>00775   guid = gd.<a class="code" href="classaffymetrix__calvin__io_1_1GenericData.html#a80d269da981e6f3f4537d0138ade91ef">ArrayIdentifier</a>();
<a name="l00776"></a>00776 
<a name="l00777"></a>00777 
<a name="l00778"></a>00778   <span class="keywordflow">if</span> (guid.empty()) {
<a name="l00779"></a>00779     <a class="code" href="classaffymetrix__calvin__io_1_1GenericDataHeader.html">GenericDataHeader</a>* gdh = gd.<a class="code" href="classaffymetrix__calvin__io_1_1GenericData.html#aa4dde02ef7050992a36bcd7c2c231c29">Header</a>().GetGenericDataHdr();
<a name="l00780"></a>00780 
<a name="l00781"></a>00781     <span class="keywordflow">if</span> (gdh  == NULL)  {
<a name="l00782"></a>00782       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;Can&#39;t find header: &quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="ArrayId_8h.html#ab3c490e6f65d1699fa846b945c6821b9">ARRAY_TYPE_IDENTIFIER</a>));
<a name="l00783"></a>00783     }
<a name="l00784"></a>00784     searchParams(*gdh, StringUtils::ConvertWCSToMBS(<a class="code" href="AffymetrixParameterConsts_8h.html#adc4e814d463aa4345d174497af6e13c6">ARRAY_ID_PARAM_NAME</a>) ,guid);
<a name="l00785"></a>00785     <span class="keywordflow">if</span> (guid.empty()) {
<a name="l00786"></a>00786       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_chpExperimentFile + <span class="stringliteral">&quot;: invalid CHP file, missing Array ID GUID (&quot;</span> + <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<a class="code" href="ArrayId_8h.html#ab3c490e6f65d1699fa846b945c6821b9">ARRAY_TYPE_IDENTIFIER</a>) + <span class="stringliteral">&quot;).&quot;</span>);
<a name="l00787"></a>00787     }
<a name="l00788"></a>00788   }
<a name="l00789"></a>00789 
<a name="l00790"></a>00790   <span class="keywordflow">return</span> guid;
<a name="l00791"></a>00791 
<a name="l00792"></a>00792 }
<a name="l00793"></a>00793 <span class="comment">// end TranslationInputStreamTableModel::_getCHPGuid</span>
<a name="l00794"></a>00794 <span class="comment">/*****************************************************************************/</span>
<a name="l00795"></a>00795 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00796"></a>00796 <span class="comment">/**</span>
<a name="l00797"></a>00797 <span class="comment"> * TranslationInputStreamTableModel::_getExternalIdFromProbeSet:</span>
<a name="l00798"></a>00798 <span class="comment"> * Synopsis:</span>
<a name="l00799"></a>00799 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_readNextExperimentCHP</span>
<a name="l00800"></a>00800 <span class="comment"> *</span>
<a name="l00801"></a>00801 <span class="comment"> * The CHP file doesn&#39;t contain the external id. However, the probe set id</span>
<a name="l00802"></a>00802 <span class="comment"> * is distinct per translation table so it can be looked up in the</span>
<a name="l00803"></a>00803 <span class="comment"> * translation table.</span>
<a name="l00804"></a>00804 <span class="comment"> *</span>
<a name="l00805"></a>00805 <span class="comment"> * @param probeSet - the marker</span>
<a name="l00806"></a>00806 <span class="comment"> * @param ttm      - the single instance translation table</span>
<a name="l00807"></a>00807 <span class="comment"> *</span>
<a name="l00808"></a>00808 <span class="comment"> * @return - external id (common name)  as std::string, empty if the probeSet doesn&#39;t exist</span>
<a name="l00809"></a>00809 <span class="comment"> */</span>
<a name="l00810"></a>00810 <span class="comment">/*****************************************************************************/</span>
<a name="l00811"></a>00811 std::string TranslationInputStreamTableModel::_getExternalIdFromProbeSet(<span class="keyword">const</span> std::string probeSet,   <span class="keyword">const</span> <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm)
<a name="l00812"></a>00812 {
<a name="l00813"></a>00813 
<a name="l00814"></a>00814   std::string externalId;
<a name="l00815"></a>00815 
<a name="l00816"></a>00816   <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(probeSet) &lt; 0) {
<a name="l00817"></a>00817     <span class="keywordflow">return</span> externalId;
<a name="l00818"></a>00818   }
<a name="l00819"></a>00819 
<a name="l00820"></a>00820   externalId = ttm.m_rows[ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(probeSet)][ttm.<a class="code" href="classTranslationTableModel.html#a8e519cfda65a7d4399dc0de52f7a4041" title="TranslationTableModel::getColumnIndex: Synopsis: DMET2 &amp;amp; DMET3 columns can be retrieved via this ...">getColumnIndex</a>(ADT_DMET3_TT_COMMON_NAME)];
<a name="l00821"></a>00821 
<a name="l00822"></a>00822   <span class="keywordflow">return</span> externalId;
<a name="l00823"></a>00823 }
<a name="l00824"></a>00824 <span class="comment">// end TranslationInputStreamTableModel::_getExternalIdFromProbeSet</span>
<a name="l00825"></a>00825 <span class="comment">/*****************************************************************************/</span>
<a name="l00826"></a>00826 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00827"></a>00827 <span class="comment">/**</span>
<a name="l00828"></a>00828 <span class="comment"> * TranslationInputStreamTableModel::_initializeColumnDefinitions</span>
<a name="l00829"></a>00829 <span class="comment"> * Synopsis: </span>
<a name="l00830"></a>00830 <span class="comment"> *</span>
<a name="l00831"></a>00831 <span class="comment"> * Helper function to the constructers</span>
<a name="l00832"></a>00832 <span class="comment"> * TranslationInputStreamTableModel::TranslationInputStreamTableModel</span>
<a name="l00833"></a>00833 <span class="comment"> * </span>
<a name="l00834"></a>00834 <span class="comment"> * Initialize the definition of columns data structure. This is compared</span>
<a name="l00835"></a>00835 <span class="comment"> * against the input TSV file when it is opened for integrity. </span>
<a name="l00836"></a>00836 <span class="comment"> * </span>
<a name="l00837"></a>00837 <span class="comment"> * @param rte - the single instance run time environment. </span>
<a name="l00838"></a>00838 <span class="comment"> * @param columnDefinitions - the typically hard coded columns. </span>
<a name="l00839"></a>00839 <span class="comment"> * @param tcdSize - the size of columnDefinitions. </span>
<a name="l00840"></a>00840 <span class="comment"> *</span>
<a name="l00841"></a>00841 <span class="comment"> */</span>
<a name="l00842"></a>00842 <span class="comment">/*****************************************************************************/</span>
<a name="l00843"></a>00843 <span class="keywordtype">void</span> TranslationInputStreamTableModel::_initializeColumnDefinitions(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <span class="keyword">const</span> <a class="code" href="classTittmColumnDefinition.html">TittmColumnDefinition</a> columnDefinitions[], <span class="keywordtype">int</span> tcdSize)
<a name="l00844"></a>00844 {
<a name="l00845"></a>00845 
<a name="l00846"></a>00846   m_tsvColumnDefinition.clear();
<a name="l00847"></a>00847 
<a name="l00848"></a>00848   <span class="comment">// Iterate so we can doublecheck the m_index value.</span>
<a name="l00849"></a>00849   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; tcdSize; i++) {
<a name="l00850"></a>00850 
<a name="l00851"></a>00851     <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(columnDefinitions[i].m_index == i, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00852"></a>00852 
<a name="l00853"></a>00853     m_tsvColumnDefinition.push_back(columnDefinitions[i]);
<a name="l00854"></a>00854   }
<a name="l00855"></a>00855 
<a name="l00856"></a>00856   <span class="keywordflow">return</span>;
<a name="l00857"></a>00857 }
<a name="l00858"></a>00858 <span class="comment">// end TranslationInputStreamTableModel::_initializeColumnDefinitions</span>
<a name="l00859"></a>00859 <span class="comment">/*****************************************************************************/</span>
<a name="l00860"></a>00860 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l00861"></a>00861 <span class="comment">/**</span>
<a name="l00862"></a>00862 <span class="comment"> * TranslationInputStreamTableModel::_initializeNextExperimentCHPFile:</span>
<a name="l00863"></a>00863 <span class="comment"> * Synopsis:</span>
<a name="l00864"></a>00864 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_readNextExperimentCHP</span>
<a name="l00865"></a>00865 <span class="comment"> * </span>
<a name="l00866"></a>00866 <span class="comment"> * Open the next CHP file, read the headers and postion things to read the</span>
<a name="l00867"></a>00867 <span class="comment"> * first data record.</span>
<a name="l00868"></a>00868 <span class="comment"> *</span>
<a name="l00869"></a>00869 <span class="comment"> * Initialize the global report alleles.</span>
<a name="l00870"></a>00870 <span class="comment"> * The report alleles global to all experiments are </span>
<a name="l00871"></a>00871 <span class="comment"> * initialized here. There are two reasons for this:</span>
<a name="l00872"></a>00872 <span class="comment"> * 1.) The GenoCallCoder requires parameters taken from a CHP file</span>
<a name="l00873"></a>00873 <span class="comment"> * specific to this set of CHP files.</span>
<a name="l00874"></a>00874 <span class="comment"> * 2.) The first CHP file is read for the first time here. The parameters</span>
<a name="l00875"></a>00875 <span class="comment"> * are take from that file here. </span>
<a name="l00876"></a>00876 <span class="comment"> * </span>
<a name="l00877"></a>00877 <span class="comment"> *</span>
<a name="l00878"></a>00878 <span class="comment"> * @param rte - the RunTimeEnvironment with the command line list</span>
<a name="l00879"></a>00879 <span class="comment"> *              of all the experiment CHP files.</span>
<a name="l00880"></a>00880 <span class="comment"> * @param ttm - the single instance translation table model updated with the</span>
<a name="l00881"></a>00881 <span class="comment"> * report alleles on the first call. </span>
<a name="l00882"></a>00882 <span class="comment"> *</span>
<a name="l00883"></a>00883 <span class="comment"> * @return - true if ok, false otherwise. Throws Exception on IO error. Updates the ttm with the report alleles on first call. </span>
<a name="l00884"></a>00884 <span class="comment"> */</span>
<a name="l00885"></a>00885 <span class="comment">/*****************************************************************************/</span>
<a name="l00886"></a>00886 <span class="keywordtype">bool</span> TranslationInputStreamTableModel::_initializeNextExperimentCHPFile(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm)
<a name="l00887"></a>00887 {
<a name="l00888"></a>00888 
<a name="l00889"></a>00889   <span class="keywordtype">bool</span> okToContinue = <span class="keyword">true</span>;
<a name="l00890"></a>00890 
<a name="l00891"></a>00891   <span class="keywordflow">if</span> (m_chpExperimentFileQueue.size() == 0) {
<a name="l00892"></a>00892     <span class="keywordflow">return</span> <span class="keyword">false</span>;
<a name="l00893"></a>00893   }
<a name="l00894"></a>00894 
<a name="l00895"></a>00895   m_chpExperimentFile = m_chpExperimentFileQueue.front();
<a name="l00896"></a>00896   m_chpExperimentFileQueue.pop_front();
<a name="l00897"></a>00897 
<a name="l00898"></a>00898 
<a name="l00899"></a>00899   <span class="comment">// Open the file and read the headers.</span>
<a name="l00900"></a>00900   <span class="comment">// I traced this &quot;Read&quot; call back to GenericFileReader.cpp</span>
<a name="l00901"></a>00901   <span class="comment">// and &quot;Read&quot; only reads the headers.</span>
<a name="l00902"></a>00902 
<a name="l00903"></a>00903 
<a name="l00904"></a>00904   m_chpData = <span class="keyword">new</span> <a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html">affymetrix_calvin_io::CHPMultiDataData</a>;
<a name="l00905"></a>00905 
<a name="l00906"></a>00906   <a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataFileReader.html">affymetrix_calvin_io::CHPMultiDataFileReader</a> reader;
<a name="l00907"></a>00907   reader.<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataFileReader.html#ab9777d64f28f0838e6e136b674f3e0f4">SetFilename</a>(m_chpExperimentFile);
<a name="l00908"></a>00908 
<a name="l00909"></a>00909   reader.<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataFileReader.html#ad7e9ca41a35f725631161c78f88d6c95">Read</a>(*m_chpData);
<a name="l00910"></a>00910 
<a name="l00911"></a>00911   <a class="code" href="classaffymetrix__calvin__io_1_1FileHeader.html">FileHeader</a> *chpFileHeader = m_chpData-&gt;<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html#a9cd1af909c16df6a06f6009a9888f426">GetFileHeader</a>();
<a name="l00912"></a>00912 
<a name="l00913"></a>00913   <a class="code" href="namespaceaffymetrix__calvin__io.html#a18875db4705ece08c05e229ecfa36e8c">DataGroupHdrIt</a> itDG, jtDG;
<a name="l00914"></a>00914 
<a name="l00915"></a>00915   chpFileHeader-&gt;GetDataGroupIts(itDG, jtDG);
<a name="l00916"></a>00916 
<a name="l00917"></a>00917   <span class="keywordflow">for</span> (; itDG != jtDG; itDG++) {
<a name="l00918"></a>00918 
<a name="l00919"></a>00919     <a class="code" href="namespaceaffymetrix__calvin__io.html#a16b5cb6d5b13f2acf2b0c8cfcae812e2">DataSetHdrIt</a> itDSH, jtDSH;
<a name="l00920"></a>00920 
<a name="l00921"></a>00921     itDG-&gt;GetDataSetIterators(itDSH, jtDSH);
<a name="l00922"></a>00922 
<a name="l00923"></a>00923   }
<a name="l00924"></a>00924 
<a name="l00925"></a>00925   ParameterNameValueTypeList summaryList = m_chpData-&gt;<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html#a94cf924a077870696383f24824c3ef69">GetSummaryParams</a>();
<a name="l00926"></a>00926 
<a name="l00927"></a>00927 
<a name="l00928"></a>00928   ParameterNameValueTypeList::iterator  itPNV;
<a name="l00929"></a>00929 
<a name="l00930"></a>00930   <span class="keywordflow">for</span> (itPNV = summaryList.begin(); itPNV != summaryList.end(); itPNV++) {
<a name="l00931"></a>00931     std::string name = StringUtils::ConvertWCSToMBS(itPNV-&gt;GetName());
<a name="l00932"></a>00932   }
<a name="l00933"></a>00933 
<a name="l00934"></a>00934   ParameterNameValueTypeList algorithmList = m_chpData-&gt;<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html#a07cb3bccf9d8618064ad17293abbd3d7">GetAlgParams</a>();
<a name="l00935"></a>00935 
<a name="l00936"></a>00936   <a class="code" href="classaffymetrix__calvin__io_1_1GenericData.html">GenericData</a> &amp; gd = m_chpData-&gt;<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html#a23c5338c17c342b7731a08728f7d6315">GetGenericData</a>();
<a name="l00937"></a>00937 
<a name="l00938"></a>00938   WStringVector names;
<a name="l00939"></a>00939   gd.<a class="code" href="classaffymetrix__calvin__io_1_1GenericData.html#a06cbe073cc803a554ede2a3d553ce8cb">DataGroupNames</a>(names);
<a name="l00940"></a>00940 
<a name="l00941"></a>00941   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; 3; i++) {
<a name="l00942"></a>00942 
<a name="l00943"></a>00943     std::string dataTypeName = StringUtils::ConvertWCSToMBS(MultiDataDataSetNames[ADT_CHP_DATA_TYPES[i]]);
<a name="l00944"></a>00944 
<a name="l00945"></a>00945     m_chpDataTypeSize[ADT_CHP_DATA_TYPES[i]] = m_chpData-&gt;<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html#a99bf358894f9dd3cf2aae2bacc2e2c98">GetEntryCount</a>(ADT_CHP_DATA_TYPES[i]);
<a name="l00946"></a>00946 
<a name="l00947"></a>00947     <span class="keywordflow">if</span> (m_chpDataTypeSize[ADT_CHP_DATA_TYPES[i]] == 0) {
<a name="l00948"></a>00948       <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(ADT_VERBOSE_EXCEPTION, dataTypeName  + <span class="stringliteral">&quot;: group has no records (0 size)\n&quot;</span>);
<a name="l00949"></a>00949     }
<a name="l00950"></a>00950 
<a name="l00951"></a>00951   }
<a name="l00952"></a>00952 
<a name="l00953"></a>00953   <span class="comment">// EXPERIMENT NAME, set to the file name, minus directories</span>
<a name="l00954"></a>00954   m_experimentName = <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(m_chpExperimentFile);
<a name="l00955"></a>00955   <span class="keywordtype">bool</span> okMatch = pcrecpp::RE(<span class="stringliteral">&quot;\\.[cC][hH][Pp]$&quot;</span>).PartialMatch(m_experimentName);
<a name="l00956"></a>00956 
<a name="l00957"></a>00957   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(okMatch, <span class="stringliteral">&quot;&quot;</span>);
<a name="l00958"></a>00958 
<a name="l00959"></a>00959   <span class="keywordflow">if</span> (m_experimentName.empty()) {
<a name="l00960"></a>00960     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_chpExperimentFile + <span class="stringliteral">&quot;: invalid experiment file name.&quot;</span>);
<a name="l00961"></a>00961   }
<a name="l00962"></a>00962 
<a name="l00963"></a>00963   m_chpGuid = _getCHPGuid(rte, gd);
<a name="l00964"></a>00964 
<a name="l00965"></a>00965   <span class="comment">//call-encoding-version=1.0 (ASCII)</span>
<a name="l00966"></a>00966   <span class="comment">//call-encoding-max-alleles=6 (Int32)</span>
<a name="l00967"></a>00967   <span class="comment">//call-encoding-data-size=UCHAR (ASCII)</span>
<a name="l00968"></a>00968   std::string versionHeader = <span class="stringliteral">&quot;call-encoding-version&quot;</span>;
<a name="l00969"></a>00969   std::string maxAllelesHeader = <span class="stringliteral">&quot;call-encoding-max-alleles&quot;</span>;
<a name="l00970"></a>00970   std::string dataSizeHeader = <span class="stringliteral">&quot;call-encoding-data-size&quot;</span>;
<a name="l00971"></a>00971   std::string version, maxAlleles, dataSize;
<a name="l00972"></a>00972 
<a name="l00973"></a>00973   
<a name="l00974"></a>00974   <a class="code" href="classaffymetrix__calvin__io_1_1GenericDataHeader.html">GenericDataHeader</a>* gdh = gd.<a class="code" href="classaffymetrix__calvin__io_1_1GenericData.html#aa4dde02ef7050992a36bcd7c2c231c29">Header</a>().GetGenericDataHdr();
<a name="l00975"></a>00975   
<a name="l00976"></a>00976   
<a name="l00977"></a>00977   searchParams(  *gdh, versionHeader, version );
<a name="l00978"></a>00978   searchParams(  *gdh, maxAllelesHeader, maxAlleles );
<a name="l00979"></a>00979   searchParams(  *gdh, dataSizeHeader, dataSize );
<a name="l00980"></a>00980 
<a name="l00981"></a>00981   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(m_chpExperimentFile) + <span class="stringliteral">&quot; &quot;</span> + versionHeader +  <span class="stringliteral">&quot;: {&quot;</span> + version + <span class="stringliteral">&quot;}&quot;</span> );
<a name="l00982"></a>00982   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(m_chpExperimentFile) + <span class="stringliteral">&quot; &quot;</span> +  maxAllelesHeader + <span class="stringliteral">&quot;: {&quot;</span> + maxAlleles +  <span class="stringliteral">&quot;}&quot;</span> );
<a name="l00983"></a>00983   <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(m_chpExperimentFile) + <span class="stringliteral">&quot; &quot;</span> + dataSizeHeader + <span class="stringliteral">&quot;: {&quot;</span> + dataSize+  <span class="stringliteral">&quot;}&quot;</span> );
<a name="l00984"></a>00984 
<a name="l00985"></a>00985   <span class="keywordflow">if</span> ( m_genoCallCoder == NULL ) {
<a name="l00986"></a>00986     <span class="keywordflow">if</span> ( rte.m_adtOpts.m_prototypeCHPFiles ) {
<a name="l00987"></a>00987       m_genoCallCoder = <span class="keyword">new</span> <a class="code" href="classGenoCallCoder.html" title="GenoCallCoder - Class for encoding and decoding numeric genotype call codes to and from allele string...">GenoCallCoder</a>(6, <span class="stringliteral">&quot;UCHAR&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="charliteral">&#39;/&#39;</span>, rte.m_adtOpts.m_inputAnnotationFile);
<a name="l00988"></a>00988     }
<a name="l00989"></a>00989     <span class="keywordflow">else</span> {
<a name="l00990"></a>00990       m_genoCallCoder = <span class="keyword">new</span> <a class="code" href="classGenoCallCoder.html" title="GenoCallCoder - Class for encoding and decoding numeric genotype call codes to and from allele string...">GenoCallCoder</a>(atoi(maxAlleles.c_str()), dataSize, version, <span class="charliteral">&#39;/&#39;</span>, rte.m_adtOpts.m_inputAnnotationFile);
<a name="l00991"></a>00991     }
<a name="l00992"></a>00992 
<a name="l00993"></a>00993     <span class="keywordflow">if</span> ((ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3) &amp;&amp;  !rte.m_adtOpts.m_ignoreReportAllele) {
<a name="l00994"></a>00994 
<a name="l00995"></a>00995       ttm.<a class="code" href="classTranslationTableModel.html#a101b8a8c3530648d99137c542b500a46" title="TranslationTableModel::setReportAlleles Synopsis:">setReportAlleles</a>( *m_genoCallCoder);
<a name="l00996"></a>00996     }
<a name="l00997"></a>00997 
<a name="l00998"></a>00998     m_chpVersion = version;
<a name="l00999"></a>00999     m_chpDataEncodingSize = dataSize;
<a name="l01000"></a>01000     m_chpMaxAlleles = maxAlleles;
<a name="l01001"></a>01001   }
<a name="l01002"></a>01002   <span class="keywordflow">else</span> <span class="keywordflow">if</span> ( ! rte.m_adtOpts.m_prototypeCHPFiles &amp;&amp; ((version != m_chpVersion) || (dataSize != m_chpDataEncodingSize) ||  (maxAlleles != m_chpMaxAlleles ) ) ){
<a name="l01003"></a>01003       <a class="code" href="classVerbose.html#a34e1d936178741eeb78de93c5497e1b8" title="Print a warning message.">Verbose::warn</a>(ADT_VERBOSE_NORMAL, std::string(<span class="stringliteral">&quot;Conflicting CHP type: &quot;</span>) + <a class="code" href="classFs.html#a16925724086de1c12ea58bad4f8df75c" title="like the unix function of the same name. ///">Fs::basename</a>(m_chpExperimentFile) + <span class="stringliteral">&quot; &quot;</span> + versionHeader +  <span class="stringliteral">&quot;: {&quot;</span> + version + <span class="stringliteral">&quot;},  &quot;</span> +  maxAllelesHeader + <span class="stringliteral">&quot;: {&quot;</span> + maxAlleles +  <span class="stringliteral">&quot;}, &quot;</span> + dataSizeHeader + <span class="stringliteral">&quot;: {&quot;</span> + dataSize +  <span class="stringliteral">&quot;}&quot;</span> );
<a name="l01004"></a>01004       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(std::string(<span class="stringliteral">&quot;CHP files with different versions detected, was expecting: &quot;</span>) + versionHeader + <span class="stringliteral">&quot;: {&quot;</span> + m_chpVersion + <span class="stringliteral">&quot;}, &quot;</span> + maxAllelesHeader + <span class="stringliteral">&quot;: {&quot;</span> + m_chpMaxAlleles + <span class="stringliteral">&quot;}, &quot;</span> + dataSizeHeader + <span class="stringliteral">&quot;: {&quot;</span> + m_chpDataEncodingSize + <span class="stringliteral">&quot;}&quot;</span>);
<a name="l01005"></a>01005 
<a name="l01006"></a>01006   }
<a name="l01007"></a>01007 
<a name="l01008"></a>01008 
<a name="l01009"></a>01009 
<a name="l01010"></a>01010   <span class="keywordflow">return</span> okToContinue;
<a name="l01011"></a>01011 
<a name="l01012"></a>01012 
<a name="l01013"></a>01013 }
<a name="l01014"></a>01014 <span class="comment">// end TranslationInputStreamTableModel::_initializeNextExperimentCHPFile</span>
<a name="l01015"></a>01015 <span class="comment">/*****************************************************************************/</span>
<a name="l01016"></a>01016 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l01017"></a>01017 <span class="comment">/**</span>
<a name="l01018"></a>01018 <span class="comment"> * TranslationInputStreamTableModel::_initializeTsvColumns</span>
<a name="l01019"></a>01019 <span class="comment"> * Synopsis:</span>
<a name="l01020"></a>01020 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_openTsvFile.</span>
<a name="l01021"></a>01021 <span class="comment"> * </span>
<a name="l01022"></a>01022 <span class="comment"> * Initialize the TsvFile object columns from the column definitions.</span>
<a name="l01023"></a>01023 <span class="comment"> * The TSV file does not have to be in column order, but this</span>
<a name="l01024"></a>01024 <span class="comment"> * method always puts them in column order internally so that</span>
<a name="l01025"></a>01025 <span class="comment"> * indexing by column index integer can take place. </span>
<a name="l01026"></a>01026 <span class="comment"> *</span>
<a name="l01027"></a>01027 <span class="comment"> *</span>
<a name="l01028"></a>01028 <span class="comment"> * @param m_tsvFileName - The name of the TsvFile in question used for messaging.</span>
<a name="l01029"></a>01029 <span class="comment"> * @param tsv           - The TsvFile object in question corresponding m_tsvFileName. </span>
<a name="l01030"></a>01030 <span class="comment"> *</span>
<a name="l01031"></a>01031 <span class="comment"> * @returns false - on error.</span>
<a name="l01032"></a>01032 <span class="comment"> *</span>
<a name="l01033"></a>01033 <span class="comment"> */</span>
<a name="l01034"></a>01034 <span class="comment">/*****************************************************************************/</span>
<a name="l01035"></a>01035 <span class="keywordtype">bool</span> TranslationInputStreamTableModel::_initializeTsvColumns(<span class="keyword">const</span> std::string &amp; inputTsvFileName, <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> &amp; tsv)
<a name="l01036"></a>01036 {
<a name="l01037"></a>01037 
<a name="l01038"></a>01038   <span class="keywordtype">bool</span>         okTsvColumns = <span class="keyword">true</span>;
<a name="l01039"></a>01039   std::stringstream msgSStr;
<a name="l01040"></a>01040 
<a name="l01041"></a>01041   m_tsvIndexToColumnDefinitionIndex.resize(tsv.<a class="code" href="classaffx_1_1TsvFile.html#a94ad901d12ab05cce0f050f99665970f" title="The number of columns for this level ///.">getColumnCount</a>(0), -1);
<a name="l01042"></a>01042 
<a name="l01043"></a>01043 
<a name="l01044"></a>01044   <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> column = 0; column &lt; m_tsvColumnDefinition.size(); column++) {
<a name="l01045"></a>01045 
<a name="l01046"></a>01046     <span class="comment">// find the tsv column index for this</span>
<a name="l01047"></a>01047     <span class="keywordtype">int</span> tsvColumn
<a name="l01048"></a>01048     = tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2dbd33415a815e221ae61cf3feac5500" title="Covert a column index to a column index ///.">cname2cidx</a>(0, m_tsvColumnDefinition[column].m_columnName);
<a name="l01049"></a>01049 
<a name="l01050"></a>01050     <span class="keywordflow">if</span> (tsvColumn &lt; 0) {
<a name="l01051"></a>01051       okTsvColumns = <span class="keyword">false</span>;
<a name="l01052"></a>01052       msgSStr &lt;&lt; inputTsvFileName &lt;&lt; <span class="stringliteral">&quot;: invalid TsvFile header first-line, &quot;</span>;
<a name="l01053"></a>01053       msgSStr &lt;&lt; <span class="stringliteral">&quot;column [&quot;</span> &lt;&lt; (column + 1) &lt;&lt; <span class="stringliteral">&quot;], \&quot;&quot;</span>;
<a name="l01054"></a>01054       msgSStr &lt;&lt; m_tsvColumnDefinition[column].m_columnName &lt;&lt; <span class="stringliteral">&quot;\&quot;  not found.&quot;</span> &lt;&lt; endl;
<a name="l01055"></a>01055       <span class="keywordflow">continue</span>;
<a name="l01056"></a>01056     }
<a name="l01057"></a>01057 
<a name="l01058"></a>01058     m_tsvIndexToColumnDefinitionIndex[tsvColumn]
<a name="l01059"></a>01059     = m_tsvColumnDefinition[column].m_index;
<a name="l01060"></a>01060 
<a name="l01061"></a>01061   }
<a name="l01062"></a>01062 
<a name="l01063"></a>01063   <span class="keywordflow">if</span> (!okTsvColumns) {
<a name="l01064"></a>01064     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, msgSStr.str(), <span class="keyword">false</span>);
<a name="l01065"></a>01065   }
<a name="l01066"></a>01066 
<a name="l01067"></a>01067   <span class="keywordflow">return</span> okTsvColumns;
<a name="l01068"></a>01068 
<a name="l01069"></a>01069 }
<a name="l01070"></a>01070 <span class="comment">// end TranslationInputStreamTableModel::_initializeTsvColumns</span>
<a name="l01071"></a>01071 <span class="comment">/*****************************************************************************/</span>
<a name="l01072"></a>01072 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l01073"></a>01073 <span class="comment">/**</span>
<a name="l01074"></a>01074 <span class="comment"> * TranslationInputStreamTableModel::_matchTsvColumnsWithColumnDefinitions</span>
<a name="l01075"></a>01075 <span class="comment"> * Synopsis:</span>
<a name="l01076"></a>01076 <span class="comment"> * Helper function to TranslationInputStreamTableModel::_openTsvFile</span>
<a name="l01077"></a>01077 <span class="comment"> *</span>
<a name="l01078"></a>01078 <span class="comment"> * It is not normal Affymetrix practice to require to have only the</span>
<a name="l01079"></a>01079 <span class="comment"> * set of columns required nor is it normal practice to have those</span>
<a name="l01080"></a>01080 <span class="comment"> * columns in any order. The TsvFile object can access columns by</span>
<a name="l01081"></a>01081 <span class="comment"> * name independent of order. However, there exists requirements</span>
<a name="l01082"></a>01082 <span class="comment"> * for this application to enforce column order as well as column order</span>
<a name="l01083"></a>01083 <span class="comment"> * enforce only columns diseried.</span>
<a name="l01084"></a>01084 <span class="comment"> * </span>
<a name="l01085"></a>01085 <span class="comment"> * This helper is used to enforce strict column ordering when required.</span>
<a name="l01086"></a>01086 <span class="comment"> * Inspect the first line of the TsvFile and insure the appropriate header</span>
<a name="l01087"></a>01087 <span class="comment"> * column names exist one-to-one and in the same order</span>
<a name="l01088"></a>01088 <span class="comment"> * with all the column definitions. Note this is only required for certain</span>
<a name="l01089"></a>01089 <span class="comment"> * files such as the human generated TranslationTable.xls spreadsheet.</span>
<a name="l01090"></a>01090 <span class="comment"> *</span>
<a name="l01091"></a>01091 <span class="comment"> * @param inputTsvFileName - the Tsv file in question</span>
<a name="l01092"></a>01092 <span class="comment"> * @param tsv - the TsvFile object that is opened and corresponds to</span>
<a name="l01093"></a>01093 <span class="comment"> *              inputTsvFileName</span>
<a name="l01094"></a>01094 <span class="comment"> *</span>
<a name="l01095"></a>01095 <span class="comment"> * @return false - the TSV columns are not complete or in column order. </span>
<a name="l01096"></a>01096 <span class="comment"> */</span>
<a name="l01097"></a>01097 <span class="comment">/*****************************************************************************/</span>
<a name="l01098"></a>01098 <span class="keywordtype">bool</span> TranslationInputStreamTableModel::_matchTsvColumnsWithColumnDefinitions(<span class="keyword">const</span> std::string &amp; inputTsvFileName, <a class="code" href="classaffx_1_1TsvFile.html" title="A class for reading and writing Tab Seperated Value (TSV) files. /// See the TsvFile format document ...">affx::TsvFile</a> &amp; tsv)
<a name="l01099"></a>01099 {
<a name="l01100"></a>01100 
<a name="l01101"></a>01101   <span class="keywordtype">bool</span> okHeaderLine = <span class="keyword">true</span>;
<a name="l01102"></a>01102 
<a name="l01103"></a>01103   <span class="keywordtype">int</span> invalidHeaderCount = 0;
<a name="l01104"></a>01104 
<a name="l01105"></a>01105   std::stringstream msgSStr;
<a name="l01106"></a>01106 
<a name="l01107"></a>01107   <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; (i &lt; tsv.<a class="code" href="classaffx_1_1TsvFile.html#a94ad901d12ab05cce0f050f99665970f" title="The number of columns for this level ///.">getColumnCount</a>(0)) &amp;&amp;
<a name="l01108"></a>01108        (i &lt; m_tsvColumnDefinition.size()) ; i++) {
<a name="l01109"></a>01109 
<a name="l01110"></a>01110     std::string columnName;
<a name="l01111"></a>01111 
<a name="l01112"></a>01112     <span class="keywordflow">if</span> (tsv.<a class="code" href="classaffx_1_1TsvFile.html#a0958f996be8dea8c5e5d2a0bac36819b" title="Get the name of a column by its level and index.">cidx2cname</a>(0, i, columnName) == TSV_OK) {
<a name="l01113"></a>01113 
<a name="l01114"></a>01114       <span class="keywordflow">if</span> (columnName != m_tsvColumnDefinition[i].m_columnName) {
<a name="l01115"></a>01115         okHeaderLine = <span class="keyword">false</span>;
<a name="l01116"></a>01116         msgSStr &lt;&lt; inputTsvFileName &lt;&lt; <span class="stringliteral">&quot;: invalid TsvFile header first-line, &quot;</span>;
<a name="l01117"></a>01117         msgSStr &lt;&lt; <span class="stringliteral">&quot;column [&quot;</span> &lt;&lt; (i + 1) &lt;&lt; <span class="stringliteral">&quot;]: found \&quot;&quot;</span> &lt;&lt; columnName &lt;&lt; <span class="stringliteral">&quot;\&quot; but was expecting \&quot;&quot;</span>;
<a name="l01118"></a>01118         msgSStr &lt;&lt; m_tsvColumnDefinition[i].m_columnName &lt;&lt; <span class="stringliteral">&quot;\&quot;.&quot;</span> &lt;&lt; endl;
<a name="l01119"></a>01119         invalidHeaderCount++;
<a name="l01120"></a>01120       }
<a name="l01121"></a>01121     } <span class="keywordflow">else</span> {
<a name="l01122"></a>01122 
<a name="l01123"></a>01123       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;tsv.cidx2cname returned unexpected error&quot;</span>);
<a name="l01124"></a>01124 
<a name="l01125"></a>01125     } <span class="comment">// if there was a TSV_ERROR</span>
<a name="l01126"></a>01126 
<a name="l01127"></a>01127   } <span class="comment">// for each column, inspect it</span>
<a name="l01128"></a>01128 
<a name="l01129"></a>01129   <span class="keywordflow">if</span> (invalidHeaderCount &gt; 0) {
<a name="l01130"></a>01130     <span class="keywordflow">if</span> (invalidHeaderCount &lt; 4) {
<a name="l01131"></a>01131       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, msgSStr.str(), <span class="keyword">false</span>);
<a name="l01132"></a>01132     } <span class="keywordflow">else</span> {
<a name="l01133"></a>01133       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, inputTsvFileName + <span class="stringliteral">&quot;: invalid TsvFile is missing the header line.&quot;</span>, <span class="keyword">false</span>);
<a name="l01134"></a>01134     }
<a name="l01135"></a>01135   }
<a name="l01136"></a>01136 
<a name="l01137"></a>01137   <span class="keywordflow">return</span> okHeaderLine;
<a name="l01138"></a>01138 
<a name="l01139"></a>01139 }
<a name="l01140"></a>01140 <span class="comment">// end TranslationInputStreamTableModel::_matchTsvColumnsWithColumnDefinitions</span>
<a name="l01141"></a>01141 <span class="comment">/*****************************************************************************/</span>
<a name="l01142"></a>01142 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l01143"></a>01143 <span class="comment">/**</span>
<a name="l01144"></a>01144 <span class="comment"> * TranslationInputStreamTableModel::_openTsvFile:</span>
<a name="l01145"></a>01145 <span class="comment"> * Synopsis:</span>
<a name="l01146"></a>01146 <span class="comment"> * Helper function to the TSV file constructor,</span>
<a name="l01147"></a>01147 <span class="comment"> * TranslationInputStreamTableModel::TranslationInputStreamTableModel</span>
<a name="l01148"></a>01148 <span class="comment"> *</span>
<a name="l01149"></a>01149 <span class="comment"> * Opens the Genotype Short Report TSV file, ready to read the first row</span>
<a name="l01150"></a>01150 <span class="comment"> * using the streaming readNextExperimentTSV API. Clears any data</span>
<a name="l01151"></a>01151 <span class="comment"> * from a previous open. </span>
<a name="l01152"></a>01152 <span class="comment"> *</span>
<a name="l01153"></a>01153 <span class="comment"> *</span>
<a name="l01154"></a>01154 <span class="comment"> * @param rte - the single instance RunTimeEnviroment.</span>
<a name="l01155"></a>01155 <span class="comment"> *</span>
<a name="l01156"></a>01156 <span class="comment"> * @return - void, throws an exception on open error.</span>
<a name="l01157"></a>01157 <span class="comment"> */</span>
<a name="l01158"></a>01158 <span class="comment">/*****************************************************************************/</span>
<a name="l01159"></a>01159 <span class="keywordtype">void</span> TranslationInputStreamTableModel::_openTsvFile(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte)
<a name="l01160"></a>01160 {
<a name="l01161"></a>01161 
<a name="l01162"></a>01162   <span class="comment">// This should never happen unless a new constructor is created</span>
<a name="l01163"></a>01163   <span class="comment">// that doesn&#39;t initialize columns.</span>
<a name="l01164"></a>01164 
<a name="l01165"></a>01165   <a class="code" href="Err_8h.html#ad024208f4ba6a4e75df4499ea1639a05" title="Calls Err::apt_err_assert with the filename and linenumber set. /// We want to avoid evaluating the m...">APT_ERR_ASSERT</a>(m_tsvColumnDefinition.size(), <span class="stringliteral">&quot;&quot;</span>);
<a name="l01166"></a>01166 
<a name="l01167"></a>01167   <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">clearData</a>();
<a name="l01168"></a>01168 
<a name="l01169"></a>01169   <span class="keywordflow">if</span> (<a class="code" href="classVerbose.html#a54ad403b97185f459a64fe3a732aab7c" title="This function gets around the problem of static variable initialization as local static variables wor...">Verbose::getParam</a>().m_Verbosity &gt; ADT_VERBOSE_INPUT_FILES) {
<a name="l01170"></a>01170     <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_INPUT_FILES, <a class="code" href="Convert_8h.html#a9e68e488e4da0371aea30596313537df" title="Little template function to make string conversion easy.">ToStr</a>(<span class="stringliteral">&quot;### TranslationInputStreamTableModel::readFile(&#39;&quot;</span>) + m_tsvFileName + <span class="stringliteral">&quot;&#39;)\n&quot;</span>);
<a name="l01171"></a>01171   }
<a name="l01172"></a>01172 
<a name="l01173"></a>01173   <span class="comment">//</span>
<a name="l01174"></a>01174   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a0b08c1eea21d4b6a893f57df16521a46" title="remove whitespace from value?">m_optAutoTrim</a>   = <span class="keyword">true</span>; <span class="comment">// remove &#39;&quot;&#39;s</span>
<a name="l01175"></a>01175   m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7b96cab331d9eaec414cfeed75a55e3d" title="Quoting Character.">m_optQuoteChar1</a> = 0;    <span class="comment">// ignore &quot;&#39;&quot;s</span>
<a name="l01176"></a>01176 
<a name="l01177"></a>01177   <span class="keywordflow">if</span> (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a763ce0bbf19b4134d0445aa58e0f5011" title="Opens a file -- attempts to guess some defaults.">open</a>(m_tsvFileName) != TSV_OK) {
<a name="l01178"></a>01178     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_tsvFileName + <span class="stringliteral">&quot;: failed opening input Tsv file.&quot;</span>);
<a name="l01179"></a>01179   }
<a name="l01180"></a>01180 
<a name="l01181"></a>01181   <span class="keywordtype">bool</span> okColumnDefinitions = <span class="keyword">true</span>;
<a name="l01182"></a>01182 
<a name="l01183"></a>01183   <span class="keywordflow">if</span> (m_strictTsvColumnOrder) {
<a name="l01184"></a>01184     okColumnDefinitions =
<a name="l01185"></a>01185       _matchTsvColumnsWithColumnDefinitions(m_tsvFileName, m_tsv);
<a name="l01186"></a>01186   }
<a name="l01187"></a>01187 
<a name="l01188"></a>01188   <span class="keywordflow">if</span> (okColumnDefinitions) {
<a name="l01189"></a>01189     okColumnDefinitions = _initializeTsvColumns(m_tsvFileName, m_tsv);
<a name="l01190"></a>01190   }
<a name="l01191"></a>01191 
<a name="l01192"></a>01192   m_tsvColumnCount = m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a94ad901d12ab05cce0f050f99665970f" title="The number of columns for this level ///.">getColumnCount</a>(0);
<a name="l01193"></a>01193   m_tsvRow = 0;
<a name="l01194"></a>01194 
<a name="l01195"></a>01195   <span class="keywordflow">if</span> (! okColumnDefinitions) {
<a name="l01196"></a>01196     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_tsvFileName + <span class="stringliteral">&quot;: invalid input file detected.&quot;</span>);
<a name="l01197"></a>01197   }
<a name="l01198"></a>01198 
<a name="l01199"></a>01199 
<a name="l01200"></a>01200 
<a name="l01201"></a>01201   <span class="keywordflow">return</span>;
<a name="l01202"></a>01202 
<a name="l01203"></a>01203 }
<a name="l01204"></a>01204 <span class="comment">// TranslationInputStreamTableModel::_openTsvFile</span>
<a name="l01205"></a>01205 <span class="comment">/*****************************************************************************/</span>
<a name="l01206"></a>01206 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l01207"></a>01207 <span class="comment">/**</span>
<a name="l01208"></a>01208 <span class="comment"> * TranslationInputStreamTableModel::_readNextExperimentCHP:</span>
<a name="l01209"></a>01209 <span class="comment"> * Synopsis:</span>
<a name="l01210"></a>01210 <span class="comment"> * Helper function to TranslationInputStreamTableModel::readNextExperiment.</span>
<a name="l01211"></a>01211 <span class="comment"> *</span>
<a name="l01212"></a>01212 <span class="comment"> * Streaming CHP files, where each CHP file is one experiment.</span>
<a name="l01213"></a>01213 <span class="comment"> * Read in the entire files worth of experiment gene data.</span>
<a name="l01214"></a>01214 <span class="comment"> * Fills in the &#39;m_rows&#39; table with experiment data. </span>
<a name="l01215"></a>01215 <span class="comment"> *</span>
<a name="l01216"></a>01216 <span class="comment"> * This is the place where override data from the override file supplants each</span>
<a name="l01217"></a>01217 <span class="comment"> * record of the CHP file as it is read in. This is how translation</span>
<a name="l01218"></a>01218 <span class="comment"> * is accomplished via the override file. The override data is</span>
<a name="l01219"></a>01219 <span class="comment"> * swapped for the CHP data here. </span>
<a name="l01220"></a>01220 <span class="comment"> *</span>
<a name="l01221"></a>01221 <span class="comment"> * </span>
<a name="l01222"></a>01222 <span class="comment"> * @param rte - the single instance run time environment </span>
<a name="l01223"></a>01223 <span class="comment"> * @param ttm - the single instance translation table model</span>
<a name="l01224"></a>01224 <span class="comment"> * @param geneExperimentCopyNumberCall</span>
<a name="l01225"></a>01225 <span class="comment"> *            - the DMET2 std::map per gene copy number alleles</span>
<a name="l01226"></a>01226 <span class="comment"> * @param geneCopyNumber</span>
<a name="l01227"></a>01227 <span class="comment"> *            - the DMET3 std::map of gene copy number</span>
<a name="l01228"></a>01228 <span class="comment"> * </span>
<a name="l01229"></a>01229 <span class="comment"> * @return - The number for records (markers) read for the experiment.</span>
<a name="l01230"></a>01230 <span class="comment"> */</span>
<a name="l01231"></a>01231 <span class="comment">/*****************************************************************************/</span>
<a name="l01232"></a>01232 <span class="keywordtype">int</span> TranslationInputStreamTableModel::_readNextExperimentCHP(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm, <span class="keyword">const</span> std::map&lt;std::string, std::string&gt; &amp;geneExperimentCopyNumberCall, std::map&lt;std::string, int&gt; &amp; geneCopyNumber)
<a name="l01233"></a>01233 {
<a name="l01234"></a>01234 
<a name="l01235"></a>01235   <span class="keywordtype">int</span> chpRow       = 0;
<a name="l01236"></a>01236   <span class="keywordtype">int</span> recordsRead  = 0;
<a name="l01237"></a>01237 
<a name="l01238"></a>01238   <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">clearData</a>();
<a name="l01239"></a>01239 
<a name="l01240"></a>01240   <span class="keywordflow">if</span> (!_initializeNextExperimentCHPFile(rte, ttm)) {
<a name="l01241"></a>01241     <span class="keywordflow">return</span> 0;
<a name="l01242"></a>01242   }
<a name="l01243"></a>01243 
<a name="l01244"></a>01244   std::vector&lt; affymetrix_calvin_io::MultiDataType &gt; dataTypes;
<a name="l01245"></a>01245   dataTypes.push_back(ADT_CHP_DATA_TYPES[BI_TYPE]);
<a name="l01246"></a>01246   dataTypes.push_back(ADT_CHP_DATA_TYPES[MULTI_TYPE]);
<a name="l01247"></a>01247 
<a name="l01248"></a>01248   <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> j = 0; j &lt;  dataTypes.size(); j++) {
<a name="l01249"></a>01249 
<a name="l01250"></a>01250     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; m_chpDataTypeSize[dataTypes[j]] ; i++) {
<a name="l01251"></a>01251 
<a name="l01252"></a>01252       <a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html">affymetrix_calvin_data::DmetBiAllelicData</a> dbad;
<a name="l01253"></a>01253 
<a name="l01254"></a>01254       m_chpData-&gt;<a class="code" href="classaffymetrix__calvin__io_1_1CHPMultiDataData.html#ad3f5d15b49302799f49a0cd9e70aceca">GetEntry</a>(dataTypes[j], i , dbad);
<a name="l01255"></a>01255 
<a name="l01256"></a>01256       <span class="comment">// Not all probe sets are going to exist in the translation table,</span>
<a name="l01257"></a>01257       <span class="comment">// if not then just drop the probeSet data on the floor.</span>
<a name="l01258"></a>01258       <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>) &lt; 0) {
<a name="l01259"></a>01259         <span class="keywordflow">continue</span>;
<a name="l01260"></a>01260       }
<a name="l01261"></a>01261 
<a name="l01262"></a>01262 
<a name="l01263"></a>01263       std::string allele1, allele2, originalAllele1, originalAllele2;
<a name="l01264"></a>01264 
<a name="l01265"></a>01265       originalAllele1.clear();
<a name="l01266"></a>01266       originalAllele2.clear();
<a name="l01267"></a>01267 
<a name="l01268"></a>01268       <span class="keywordflow">if</span> (! _getAllelesFromProbeSet(rte, ttm, dbad, allele1, allele2)) {
<a name="l01269"></a>01269         <span class="keywordflow">continue</span>;
<a name="l01270"></a>01270       }
<a name="l01271"></a>01271 
<a name="l01272"></a>01272       <span class="comment">// Genotype Override</span>
<a name="l01273"></a>01273       <span class="keywordflow">if</span> (m_gotm != NULL) {
<a name="l01274"></a>01274         std::vector&lt; std::string &gt; alleleOverride = m_gotm-&gt;<a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">getOverrideAlleles</a>(m_experimentName, dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>);
<a name="l01275"></a>01275         <span class="keywordflow">if</span> (alleleOverride.size() &gt; 0) {
<a name="l01276"></a>01276           <span class="keywordflow">if</span> ( <a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(allele1)  ) {
<a name="l01277"></a>01277             m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[m_experimentName + <span class="stringliteral">&quot;::&quot;</span> + dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>] = allele1;
<a name="l01278"></a>01278           } <span class="keywordflow">else</span> {
<a name="l01279"></a>01279             m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[m_experimentName + <span class="stringliteral">&quot;::&quot;</span> + dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>] = allele1 + <span class="stringliteral">&quot;/&quot;</span> + allele2;
<a name="l01280"></a>01280           }
<a name="l01281"></a>01281           
<a name="l01282"></a>01282           originalAllele1 = allele1;
<a name="l01283"></a>01283           originalAllele2 = allele2;
<a name="l01284"></a>01284           allele1 = alleleOverride[0];
<a name="l01285"></a>01285           <span class="keywordflow">if</span> ( alleleOverride.size() == 2 ) {
<a name="l01286"></a>01286             allele2 = alleleOverride[1];
<a name="l01287"></a>01287           }
<a name="l01288"></a>01288           <span class="keywordflow">else</span> {
<a name="l01289"></a>01289             allele2 = <span class="stringliteral">&quot;&quot;</span>;
<a name="l01290"></a>01290           }
<a name="l01291"></a>01291         }
<a name="l01292"></a>01292       }
<a name="l01293"></a>01293 
<a name="l01294"></a>01294       std::vector&lt;std::string&gt; newRow;
<a name="l01295"></a>01295       chpRow++;
<a name="l01296"></a>01296 
<a name="l01297"></a>01297       newRow.resize(GT_END_FIELD_INDEX, <span class="stringliteral">&quot;&quot;</span>);
<a name="l01298"></a>01298 
<a name="l01299"></a>01299       newRow[GT_SAMPLE_INDEX]      = <span class="stringliteral">&quot;N/A&quot;</span>;
<a name="l01300"></a>01300       newRow[GT_EXPERIMENT_INDEX]  = m_experimentName;
<a name="l01301"></a>01301       newRow[GT_GENES_INDEX]       = _getGeneFromProbeSet(dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>, ttm);
<a name="l01302"></a>01302       newRow[GT_EXTERNAL_ID_INDEX] = _getExternalIdFromProbeSet(dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>, ttm);
<a name="l01303"></a>01303       newRow[GT_PROBE_SET_INDEX]   = dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>;
<a name="l01304"></a>01304       newRow[GT_ALLELE1_INDEX]     = allele1;
<a name="l01305"></a>01305       newRow[GT_ALLELE2_INDEX]     = allele2;
<a name="l01306"></a>01306       newRow[GT_ORIGINAL_ALLELE1_INDEX] = originalAllele1;
<a name="l01307"></a>01307 
<a name="l01308"></a>01308       newRow[GT_ORIGINAL_ALLELE2_INDEX] = originalAllele2;
<a name="l01309"></a>01309 
<a name="l01310"></a>01310       m_fileRowIndex[m_rows.size()] = i;
<a name="l01311"></a>01311 
<a name="l01312"></a>01312       recordsRead++;
<a name="l01313"></a>01313       m_experimentRowIndex[m_experimentName].push_back(chpRow - 1);
<a name="l01314"></a>01314 
<a name="l01315"></a>01315       m_probeSetRowIndex[dbad.<a class="code" href="structaffymetrix__calvin__data_1_1__DmetBiAllelicData.html#a27db260346b24b34b5e73e8e731b5819">name</a>] = m_rows.size();
<a name="l01316"></a>01316       m_rows.push_back(newRow);
<a name="l01317"></a>01317 
<a name="l01318"></a>01318       <span class="keywordflow">if</span> ( geneCopyNumber.count(newRow[GT_GENES_INDEX]) == 0 ) {
<a name="l01319"></a>01319         geneCopyNumber[newRow[GT_GENES_INDEX]] = 2;
<a name="l01320"></a>01320       }
<a name="l01321"></a>01321       
<a name="l01322"></a>01322       <span class="keywordflow">if</span> ( allele1  == <span class="stringliteral">&quot;0&quot;</span> ) {
<a name="l01323"></a>01323         geneCopyNumber[newRow[GT_GENES_INDEX]]  = 0;
<a name="l01324"></a>01324       }
<a name="l01325"></a>01325       <span class="keywordflow">else</span> <span class="keywordflow">if</span> ( !<a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(allele1) &amp;&amp;
<a name="l01326"></a>01326                 (geneCopyNumber[newRow[GT_GENES_INDEX]] == 2)
<a name="l01327"></a>01327                 &amp;&amp; ( allele2.empty() ) ){
<a name="l01328"></a>01328         geneCopyNumber[newRow[GT_GENES_INDEX]] = 1;
<a name="l01329"></a>01329       }
<a name="l01330"></a>01330     }
<a name="l01331"></a>01331 
<a name="l01332"></a>01332   }
<a name="l01333"></a>01333 
<a name="l01334"></a>01334 
<a name="l01335"></a>01335   <span class="keywordflow">return</span> recordsRead;
<a name="l01336"></a>01336 }
<a name="l01337"></a>01337 <span class="comment">// end  TranslationInputStreamTableModel::_readNextExperimentCHP</span>
<a name="l01338"></a>01338 
<a name="l01339"></a>01339 
<a name="l01340"></a>01340 <span class="comment">/*****************************************************************************/</span>
<a name="l01341"></a>01341 <span class="comment">/*****************************************************************************/</span><span class="comment"></span>
<a name="l01342"></a>01342 <span class="comment">/**</span>
<a name="l01343"></a>01343 <span class="comment"> * TranslationInputStreamTableModel::_readNextExperimentTsv</span>
<a name="l01344"></a>01344 <span class="comment"> * Synopsis:</span>
<a name="l01345"></a>01345 <span class="comment"> * Helper function to TranslationInputStreamTableModel::readNextExperiment.</span>
<a name="l01346"></a>01346 <span class="comment"> *</span>
<a name="l01347"></a>01347 <span class="comment"> * Stream experiments from the Genotype Short Report file one experiment</span>
<a name="l01348"></a>01348 <span class="comment"> * at a time. Fills in the &#39;m_rows&#39; table with experiment data. </span>
<a name="l01349"></a>01349 <span class="comment"> *</span>
<a name="l01350"></a>01350 <span class="comment"> * @param rte - the single instance run time environment. </span>
<a name="l01351"></a>01351 <span class="comment"> * @param ttm  - the single instance translation table model</span>
<a name="l01352"></a>01352 <span class="comment"> * @param geneExperimentCopyNumberCall </span>
<a name="l01353"></a>01353 <span class="comment"> *            - the DMET2 std::map of per gene copy number alleles</span>
<a name="l01354"></a>01354 <span class="comment"> * @param geneCopyNumber</span>
<a name="l01355"></a>01355 <span class="comment"> *            - the DMET3 std::map of gene copy number</span>
<a name="l01356"></a>01356 <span class="comment"> * </span>
<a name="l01357"></a>01357 <span class="comment"> * @return - The number for records (markers) read for the experiment.</span>
<a name="l01358"></a>01358 <span class="comment"> */</span>
<a name="l01359"></a>01359 <span class="comment">/*****************************************************************************/</span>
<a name="l01360"></a>01360 <span class="keywordtype">int</span> TranslationInputStreamTableModel::_readNextExperimentTsv(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <span class="keyword">const</span> <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm, <span class="keyword">const</span> std::map&lt;std::string, std::string&gt; &amp;geneExperimentCopyNumberCall)
<a name="l01361"></a>01361 {
<a name="l01362"></a>01362 
<a name="l01363"></a>01363   <span class="keywordtype">bool</span>   okExperimentGene   = <span class="keyword">true</span>;
<a name="l01364"></a>01364   <span class="keywordtype">int</span>    recordsRead        = 0;
<a name="l01365"></a>01365   std::string currentExperiment;
<a name="l01366"></a>01366 
<a name="l01367"></a>01367   <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">clearData</a>();
<a name="l01368"></a>01368 
<a name="l01369"></a>01369   <span class="keywordflow">if</span> (m_tsvNextRow.size() &gt; 0) {
<a name="l01370"></a>01370     currentExperiment = m_tsvNextRow[m_tsvExperimentColumn];
<a name="l01371"></a>01371     <span class="keywordflow">if</span> (m_experimentRowIndex.count(currentExperiment) &gt; 0) {
<a name="l01372"></a>01372       <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(ADT_VERBOSE_NORMAL, <span class="stringliteral">&quot;\n\nDMET2 genotype short report input files need to be sorted by experiment before they can be input.&quot;</span>);
<a name="l01373"></a>01373       <a class="code" href="classTranslationInputStreamTableModel.html#ab3371d6ef32696cdaccbfc30790a4bd6" title="TranslationInputStreamTableModel::sortTsvFileByExperiment: Synopsis:">sortTsvFileByExperiment</a>(rte);
<a name="l01374"></a>01374       <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(currentExperiment + <span class="stringliteral">&quot;: duplicate experiment detected.&quot;</span>);
<a name="l01375"></a>01375     }
<a name="l01376"></a>01376 
<a name="l01377"></a>01377     m_fileRowIndex[m_rows.size()] = m_tsvRow;
<a name="l01378"></a>01378     m_rows.push_back(m_tsvNextRow);
<a name="l01379"></a>01379     m_tsvNextRow.clear();
<a name="l01380"></a>01380     recordsRead++;
<a name="l01381"></a>01381   }
<a name="l01382"></a>01382 
<a name="l01383"></a>01383   std::string whiteSpace = <span class="stringliteral">&quot; &quot;</span>;
<a name="l01384"></a>01384   <span class="keywordtype">bool</span> okColumnValues = <span class="keyword">true</span>;
<a name="l01385"></a>01385 
<a name="l01386"></a>01386 
<a name="l01387"></a>01387   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> row = 0; okColumnValues &amp;&amp; okExperimentGene &amp;&amp;
<a name="l01388"></a>01388        (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a2022895d082d1d635e7f59fccab12afb" title="Skip to the next level of data which matches seek_clvl ///.">nextLevel</a>(0) == TSV_OK); row++) {
<a name="l01389"></a>01389     <span class="comment">//</span>
<a name="l01390"></a>01390     std::vector&lt;std::string&gt; newRow;
<a name="l01391"></a>01391     m_tsvRow++;
<a name="l01392"></a>01392 
<a name="l01393"></a>01393     newRow.resize(m_tsvColumnCount + 4, <span class="stringliteral">&quot;&quot;</span>);
<a name="l01394"></a>01394 
<a name="l01395"></a>01395     <span class="keywordtype">bool</span> okRow = <span class="keyword">true</span>;
<a name="l01396"></a>01396 
<a name="l01397"></a>01397     <span class="comment">// Read in a row, one column at a time.</span>
<a name="l01398"></a>01398     <span class="keywordflow">for</span> (<span class="keywordtype">int</span> tsvColumn = 0; tsvColumn &lt; m_tsvColumnCount; tsvColumn++) {
<a name="l01399"></a>01399 
<a name="l01400"></a>01400       std::stringstream lineSStr;
<a name="l01401"></a>01401       std::string       columnName = <a class="code" href="classTranslationInputStreamTableModel.html#a28ffd80e00964059f7e34c62c177df4a" title="TranslationInputStreamTableModel::getColumnName Synopsis:">getColumnName</a>(m_tsvIndexToColumnDefinitionIndex[tsvColumn]);
<a name="l01402"></a>01402       <span class="comment">// get the data</span>
<a name="l01403"></a>01403       std::string columnValue;
<a name="l01404"></a>01404 
<a name="l01405"></a>01405       <span class="keywordflow">if</span> (m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0, tsvColumn, columnValue) != TSV_OK) {
<a name="l01406"></a>01406         <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(<span class="stringliteral">&quot;tsv.get != TSV_OK&quot;</span>);
<a name="l01407"></a>01407       }
<a name="l01408"></a>01408 
<a name="l01409"></a>01409 
<a name="l01410"></a>01410       <a class="code" href="classUtil.html#aa912567f9cdde6941ba305f4278b5828" title="Cut off any preceding and trailing white space.">Util::trimString</a>(columnValue, whiteSpace.c_str());
<a name="l01411"></a>01411 
<a name="l01412"></a>01412       <span class="comment">// empty and allowed?</span>
<a name="l01413"></a>01413       <span class="keywordflow">if</span> (columnValue.empty()) {
<a name="l01414"></a>01414         <span class="keywordflow">if</span> (m_tsvColumnDefinition[m_tsvIndexToColumnDefinitionIndex[tsvColumn]].m_emptyOk == <span class="keyword">false</span>) {
<a name="l01415"></a>01415           lineSStr &lt;&lt; m_tsvFileName &lt;&lt; <span class="stringliteral">&quot;: ROW &quot;</span> &lt;&lt; m_tsvRow  &lt;&lt; <span class="stringliteral">&quot; : &quot;</span>;
<a name="l01416"></a>01416           lineSStr &lt;&lt; columnName &lt;&lt; <span class="stringliteral">&quot; : ERROR missing required column.&quot;</span>;
<a name="l01417"></a>01417           <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(rte.m_currentVerbosity, lineSStr.str());
<a name="l01418"></a>01418           okColumnValues = <span class="keyword">false</span>;
<a name="l01419"></a>01419         }
<a name="l01420"></a>01420       }
<a name="l01421"></a>01421       <span class="comment">// validate it against regexp</span>
<a name="l01422"></a>01422       <span class="keywordflow">else</span> <span class="keywordflow">if</span> (!m_tsvColumnDefinition[m_tsvIndexToColumnDefinitionIndex[tsvColumn]].m_validRE.FullMatch(columnValue)) {
<a name="l01423"></a>01423         lineSStr &lt;&lt; m_tsvFileName &lt;&lt; <span class="stringliteral">&quot;: ROW &quot;</span> &lt;&lt; m_tsvRow  &lt;&lt; <span class="stringliteral">&quot; : COLUMN [&quot;</span>;
<a name="l01424"></a>01424         lineSStr &lt;&lt; tsvColumn &lt;&lt; <span class="stringliteral">&quot;] &quot;</span> &lt;&lt; columnName &lt;&lt; <span class="stringliteral">&quot; : ERROR invalid value: \&quot;&quot;</span>;
<a name="l01425"></a>01425         lineSStr &lt;&lt; columnValue &lt;&lt; <span class="stringliteral">&quot;\&quot;&quot;</span>;
<a name="l01426"></a>01426         lineSStr &lt;&lt; <span class="stringliteral">&quot; (expecting &quot;</span> &lt;&lt; m_tsvColumnDefinition[m_tsvIndexToColumnDefinitionIndex[tsvColumn]].m_validRE.pattern();
<a name="l01427"></a>01427         lineSStr &lt;&lt; <span class="stringliteral">&quot;)&quot;</span>;
<a name="l01428"></a>01428         lineSStr &lt;&lt; endl;
<a name="l01429"></a>01429 
<a name="l01430"></a>01430         <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0 ; j &lt; m_tsvColumnCount;  j++) {
<a name="l01431"></a>01431           m_tsv.<a class="code" href="classaffx_1_1TsvFile.html#a7a81092ce134e8c9116a1a5bcfe287d1" title="Get the string value of a field ///.">get</a>(0, j, columnValue);
<a name="l01432"></a>01432           lineSStr &lt;&lt; columnValue &lt;&lt; <span class="stringliteral">&quot;,&quot;</span>;
<a name="l01433"></a>01433         }
<a name="l01434"></a>01434         <a class="code" href="classVerbose.html#ac4034f68f4c8d2b49cd6340984b940ce" title="Print a message to the stream.">Verbose::out</a>(rte.m_currentVerbosity, lineSStr.str());
<a name="l01435"></a>01435         okColumnValues = <span class="keyword">false</span>;
<a name="l01436"></a>01436       }
<a name="l01437"></a>01437       <span class="comment">// Ingore row</span>
<a name="l01438"></a>01438       <span class="keywordflow">else</span> <span class="keywordflow">if</span> ((m_tsvColumnDefinition[m_tsvIndexToColumnDefinitionIndex[tsvColumn]].m_ignoreRE != NULL) &amp;&amp;
<a name="l01439"></a>01439                (m_tsvColumnDefinition[m_tsvIndexToColumnDefinitionIndex[tsvColumn]].m_ignoreRE-&gt;PartialMatch(columnValue))) {
<a name="l01440"></a>01440         okRow = <span class="keyword">false</span>;
<a name="l01441"></a>01441 
<a name="l01442"></a>01442       }
<a name="l01443"></a>01443       <span class="comment">// put the column data in our row.</span>
<a name="l01444"></a>01444       newRow[m_tsvIndexToColumnDefinitionIndex[tsvColumn]] = columnValue;
<a name="l01445"></a>01445     }
<a name="l01446"></a>01446     <span class="comment">// save our row to the data.</span>
<a name="l01447"></a>01447 
<a name="l01448"></a>01448     <span class="comment">// Genotype Override</span>
<a name="l01449"></a>01449     <span class="keywordflow">if</span> (m_gotm != NULL) {
<a name="l01450"></a>01450       std::vector&lt; std::string &gt; <span class="keyword">override</span> = m_gotm-&gt;<a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">getOverrideAlleles</a>(newRow[m_tsvExperimentColumn],  newRow[m_tsvProbeSetColumn]);
<a name="l01451"></a>01451       <span class="keywordflow">if</span> (<span class="keyword">override</span>.size() ) {
<a name="l01452"></a>01452         cerr &lt;&lt; newRow[m_tsvExperimentColumn] &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; newRow[m_tsvProbeSetColumn] &lt;&lt; <span class="stringliteral">&quot; has override &quot;</span> &lt;&lt; <span class="keyword">override</span>[0] &lt;&lt; endl;
<a name="l01453"></a>01453         m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[newRow[m_tsvExperimentColumn] + <span class="stringliteral">&quot;::&quot;</span> + newRow[m_tsvProbeSetColumn]] = newRow[GT_ALLELE1_INDEX];
<a name="l01454"></a>01454         newRow[GT_ALLELE1_INDEX] = <span class="keyword">override</span>[0];
<a name="l01455"></a>01455         newRow[GT_ALLELE2_INDEX] = <span class="stringliteral">&quot;&quot;</span>;
<a name="l01456"></a>01456         <span class="keywordflow">if</span> ( <span class="keyword">override</span>.size() == 2 ) {
<a name="l01457"></a>01457           newRow[GT_ALLELE2_INDEX] = <span class="keyword">override</span>[1];
<a name="l01458"></a>01458         }
<a name="l01459"></a>01459       }
<a name="l01460"></a>01460     }
<a name="l01461"></a>01461     <span class="comment">// DMET2 Copy Number Zero</span>
<a name="l01462"></a>01462     <span class="keywordflow">if</span> ( ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3 ) {
<a name="l01463"></a>01463       std::string key = newRow[GT_GENES_INDEX] + <span class="stringliteral">&quot;:&quot;</span> + newRow[GT_EXPERIMENT_INDEX];
<a name="l01464"></a>01464       <span class="keywordflow">if</span> ( geneExperimentCopyNumberCall.count(key) &gt; 0 ) {
<a name="l01465"></a>01465         newRow[GT_ALLELE1_INDEX] = <span class="stringliteral">&quot;0&quot;</span>;
<a name="l01466"></a>01466         newRow[GT_ALLELE2_INDEX] = <span class="stringliteral">&quot;&quot;</span>;
<a name="l01467"></a>01467       }
<a name="l01468"></a>01468         
<a name="l01469"></a>01469     }
<a name="l01470"></a>01470     
<a name="l01471"></a>01471     <span class="keywordflow">if</span> (okRow) {
<a name="l01472"></a>01472 
<a name="l01473"></a>01473       <span class="keywordflow">if</span> (currentExperiment.empty()) {
<a name="l01474"></a>01474         currentExperiment = newRow[m_tsvExperimentColumn];
<a name="l01475"></a>01475       }
<a name="l01476"></a>01476       <span class="keywordflow">if</span> (currentExperiment == newRow[m_tsvExperimentColumn]) {
<a name="l01477"></a>01477         m_fileRowIndex[m_rows.size()] = m_tsvRow;
<a name="l01478"></a>01478         recordsRead++;
<a name="l01479"></a>01479         m_experimentRowIndex[currentExperiment].push_back(m_tsvRow - 1);
<a name="l01480"></a>01480 
<a name="l01481"></a>01481         m_rows.push_back(newRow);
<a name="l01482"></a>01482       } <span class="keywordflow">else</span> {
<a name="l01483"></a>01483         m_tsvNextRow = newRow;
<a name="l01484"></a>01484         okExperimentGene = <span class="keyword">false</span>;
<a name="l01485"></a>01485       }
<a name="l01486"></a>01486     }
<a name="l01487"></a>01487   }
<a name="l01488"></a>01488 
<a name="l01489"></a>01489 
<a name="l01490"></a>01490   <span class="keywordflow">if</span> (! okColumnValues) {
<a name="l01491"></a>01491     <a class="code" href="Err_8h.html#a7e26a673d8901b5bc880dfab285f5309" title="Calls Err::apt_err_abort with the filename and linenumber set.">APT_ERR_ABORT</a>(m_tsvFileName + <span class="stringliteral">&quot;: invalid input file data detected.&quot;</span>);
<a name="l01492"></a>01492   }
<a name="l01493"></a>01493 
<a name="l01494"></a>01494   <span class="keywordflow">if</span> (m_rows.size() != recordsRead) {
<a name="l01495"></a>01495     m_rows.resize(recordsRead);
<a name="l01496"></a>01496   }
<a name="l01497"></a>01497 
<a name="l01498"></a>01498   <span class="keywordflow">return</span> recordsRead;
<a name="l01499"></a>01499 
<a name="l01500"></a>01500 }
<a name="l01501"></a>01501 <span class="comment">// end TranslationInputStreamTableModel::_readExperimentNextGeneTsv</span>
<a name="l01502"></a>01502 <span class="comment">/*****************************************************************************/</span>
<a name="l01503"></a>01503 
<a name="l01504"></a>01504 
<a name="l01505"></a>01505 <span class="keywordtype">int</span> TranslationInputStreamTableModel::_readNextExperimentBatchFolder(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm, <span class="keyword">const</span> std::map&lt;std::string, std::string&gt; &amp;geneExperimentCopyNumberCall, std::map&lt;std::string, int&gt; &amp; geneCopyNumber)
<a name="l01506"></a>01506 {
<a name="l01507"></a>01507   <span class="keywordtype">int</span> chpRow       = 0;
<a name="l01508"></a>01508   <span class="keywordtype">int</span> recordsRead  = 0;
<a name="l01509"></a>01509 
<a name="l01510"></a>01510   <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">clearData</a>();
<a name="l01511"></a>01511 
<a name="l01512"></a>01512   <span class="keywordflow">if</span> (!m_blockReader.nextSample()) {
<a name="l01513"></a>01513     <span class="keywordflow">return</span> 0;
<a name="l01514"></a>01514   }
<a name="l01515"></a>01515 
<a name="l01516"></a>01516   m_experimentName = m_blockReader.getSampleName();
<a name="l01517"></a>01517 
<a name="l01518"></a>01518   <span class="keywordflow">if</span> (m_genoCallCoder == NULL) {
<a name="l01519"></a>01519     <span class="comment">//</span>
<a name="l01520"></a>01520     <span class="comment">// todo: should use call code table from IndexHeader section</span>
<a name="l01521"></a>01521     <span class="comment">//</span>
<a name="l01522"></a>01522     <span class="keywordtype">int</span> maxAlleles = m_blockReader.getMaxAlleles();
<a name="l01523"></a>01523     m_genoCallCoder = <span class="keyword">new</span> <a class="code" href="classGenoCallCoder.html" title="GenoCallCoder - Class for encoding and decoding numeric genotype call codes to and from allele string...">GenoCallCoder</a>(maxAlleles, <span class="stringliteral">&quot;UCHAR&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="charliteral">&#39;/&#39;</span>, rte.m_adtOpts.m_inputAnnotationFile);
<a name="l01524"></a>01524 
<a name="l01525"></a>01525     <span class="keywordflow">if</span> (ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3 &amp;&amp; !rte.m_adtOpts.m_ignoreReportAllele) {
<a name="l01526"></a>01526       ttm.<a class="code" href="classTranslationTableModel.html#a101b8a8c3530648d99137c542b500a46" title="TranslationTableModel::setReportAlleles Synopsis:">setReportAlleles</a>(*m_genoCallCoder);
<a name="l01527"></a>01527     }
<a name="l01528"></a>01528 
<a name="l01529"></a>01529     m_chpVersion = <span class="stringliteral">&quot;1.0&quot;</span>;
<a name="l01530"></a>01530     m_chpDataEncodingSize = <span class="stringliteral">&quot;UCHAR&quot;</span>;
<a name="l01531"></a>01531     m_chpMaxAlleles = 2;
<a name="l01532"></a>01532   }
<a name="l01533"></a>01533 
<a name="l01534"></a>01534   vector&lt;pair&lt;IndexRow, SampleList&gt; &gt;&amp; vProbeSets = m_blockReader.getProbeSets();
<a name="l01535"></a>01535   <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; (int)vProbeSets.size(); i++) {
<a name="l01536"></a>01536       <a class="code" href="classIndexRow.html">IndexRow</a>&amp; row = vProbeSets[i].first;
<a name="l01537"></a>01537       <a class="code" href="classSampleList.html">SampleList</a>&amp; samples = vProbeSets[i].second;
<a name="l01538"></a>01538       
<a name="l01539"></a>01539       <span class="keywordtype">string</span> sProbeset = row.getProbeSetId();
<a name="l01540"></a>01540       <span class="keywordtype">int</span> numProbeSetRows = ttm.getProbeSetNumRows(sProbeset);
<a name="l01541"></a>01541       <span class="keywordflow">if</span> (numProbeSetRows == 0) {
<a name="l01542"></a>01542         <span class="keywordflow">continue</span>;
<a name="l01543"></a>01543       }
<a name="l01544"></a>01544 
<a name="l01545"></a>01545       <span class="keywordtype">string</span> strCallName = m_blockReader.getSampleCall(samples);
<a name="l01546"></a>01546       <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; numProbeSetRows; j++) {
<a name="l01547"></a>01547           <span class="keywordtype">string</span> allele1, allele2, originalAllele1, originalAllele2;
<a name="l01548"></a>01548 
<a name="l01549"></a>01549           originalAllele1.clear();
<a name="l01550"></a>01550           originalAllele2.clear();
<a name="l01551"></a>01551 
<a name="l01552"></a>01552           <span class="keywordflow">if</span> (!_getAllelesFromAbstractAlleles(rte, ttm, sProbeset, strCallName, allele1, allele2)) {
<a name="l01553"></a>01553             <span class="keywordflow">break</span>;
<a name="l01554"></a>01554           }
<a name="l01555"></a>01555 
<a name="l01556"></a>01556           <span class="comment">// Genotype Override</span>
<a name="l01557"></a>01557           <span class="keywordflow">if</span> (m_gotm != NULL) {
<a name="l01558"></a>01558             vector&lt;string&gt; alleleOverride = m_gotm-&gt;<a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">getOverrideAlleles</a>(m_experimentName, sProbeset);
<a name="l01559"></a>01559             <span class="keywordflow">if</span> (alleleOverride.size() &gt; 0) {
<a name="l01560"></a>01560               <span class="keywordflow">if</span> ( <a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(allele1)  ) {
<a name="l01561"></a>01561                 m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[m_experimentName + <span class="stringliteral">&quot;::&quot;</span> + sProbeset] = allele1;
<a name="l01562"></a>01562               } 
<a name="l01563"></a>01563               <span class="keywordflow">else</span> {
<a name="l01564"></a>01564                 m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[m_experimentName + <span class="stringliteral">&quot;::&quot;</span> + sProbeset] = allele1 + <span class="stringliteral">&quot;/&quot;</span> + allele2;
<a name="l01565"></a>01565               }
<a name="l01566"></a>01566           
<a name="l01567"></a>01567               originalAllele1 = allele1;
<a name="l01568"></a>01568               originalAllele2 = allele2;
<a name="l01569"></a>01569               allele1 = alleleOverride[0];
<a name="l01570"></a>01570               allele2 = (alleleOverride.size() == 2 ? alleleOverride[1] : <span class="stringliteral">&quot;&quot;</span>);
<a name="l01571"></a>01571             }
<a name="l01572"></a>01572           }
<a name="l01573"></a>01573 
<a name="l01574"></a>01574           vector&lt;string&gt; newRow;
<a name="l01575"></a>01575           chpRow++;
<a name="l01576"></a>01576 
<a name="l01577"></a>01577           newRow.resize(GT_END_FIELD_INDEX, <span class="stringliteral">&quot;&quot;</span>);
<a name="l01578"></a>01578 
<a name="l01579"></a>01579           newRow[GT_SAMPLE_INDEX]      = <span class="stringliteral">&quot;N/A&quot;</span>;
<a name="l01580"></a>01580           newRow[GT_EXPERIMENT_INDEX]  = m_experimentName;
<a name="l01581"></a>01581           newRow[GT_GENES_INDEX]       = _getGeneFromProbeSet(sProbeset, ttm);
<a name="l01582"></a>01582           newRow[GT_EXTERNAL_ID_INDEX] = _getExternalIdFromProbeSet(sProbeset, ttm);
<a name="l01583"></a>01583           newRow[GT_PROBE_SET_INDEX]   = sProbeset;
<a name="l01584"></a>01584           newRow[GT_ALLELE1_INDEX]     = allele1;
<a name="l01585"></a>01585           newRow[GT_ALLELE2_INDEX]     = allele2;
<a name="l01586"></a>01586           newRow[GT_ORIGINAL_ALLELE1_INDEX] = originalAllele1;
<a name="l01587"></a>01587 
<a name="l01588"></a>01588           newRow[GT_ORIGINAL_ALLELE2_INDEX] = originalAllele2;
<a name="l01589"></a>01589 
<a name="l01590"></a>01590           m_fileRowIndex[m_rows.size()] = recordsRead;
<a name="l01591"></a>01591 
<a name="l01592"></a>01592           recordsRead++;
<a name="l01593"></a>01593           m_experimentRowIndex[m_experimentName].push_back(chpRow - 1);
<a name="l01594"></a>01594 
<a name="l01595"></a>01595           m_probeSetRowIndex[sProbeset] = m_rows.size();
<a name="l01596"></a>01596           m_rows.push_back(newRow);
<a name="l01597"></a>01597 
<a name="l01598"></a>01598           <span class="keywordflow">if</span> (geneCopyNumber.count(newRow[GT_GENES_INDEX]) == 0) {
<a name="l01599"></a>01599             geneCopyNumber[newRow[GT_GENES_INDEX]] = 2;
<a name="l01600"></a>01600           }
<a name="l01601"></a>01601       
<a name="l01602"></a>01602           <span class="keywordflow">if</span> (allele1  == <span class="stringliteral">&quot;0&quot;</span>) {
<a name="l01603"></a>01603             geneCopyNumber[newRow[GT_GENES_INDEX]]  = 0;
<a name="l01604"></a>01604           }
<a name="l01605"></a>01605           <span class="keywordflow">else</span> <span class="keywordflow">if</span> (!<a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(allele1) &amp;&amp; geneCopyNumber[newRow[GT_GENES_INDEX]] == 2 &amp;&amp; allele2 == <span class="stringliteral">&quot;&quot;</span>){
<a name="l01606"></a>01606             geneCopyNumber[newRow[GT_GENES_INDEX]] = 1;
<a name="l01607"></a>01607           }
<a name="l01608"></a>01608 
<a name="l01609"></a>01609       } <span class="comment">// for (numProbeSetRows)</span>
<a name="l01610"></a>01610 
<a name="l01611"></a>01611   } <span class="comment">// for (indexTable.rows)</span>
<a name="l01612"></a>01612 
<a name="l01613"></a>01613   <span class="keywordflow">return</span> recordsRead;
<a name="l01614"></a>01614 }
<a name="l01615"></a>01615 
<a name="l01616"></a>01616 
<a name="l01617"></a>01617 <span class="keywordtype">int</span> TranslationInputStreamTableModel::_readNextExperimentCallsFile(<span class="keyword">const</span> <a class="code" href="classRunTimeEnvironment.html">RunTimeEnvironment</a> &amp; rte, <a class="code" href="classTranslationTableModel.html">TranslationTableModel</a> &amp; ttm, <span class="keyword">const</span> std::map&lt;std::string, std::string&gt; &amp;geneExperimentCopyNumberCall, std::map&lt;std::string, int&gt; &amp; geneCopyNumber)
<a name="l01618"></a>01618 {
<a name="l01619"></a>01619   <span class="keywordtype">int</span> chpRow       = 0;
<a name="l01620"></a>01620   <span class="keywordtype">int</span> recordsRead  = 0;
<a name="l01621"></a>01621 
<a name="l01622"></a>01622   <a class="code" href="classTranslationInputStreamTableModel.html#a3d17ce385639930b744346d6b0cdb1a5" title="TranslationInputStreamTableModel::clearData Synopsis.">clearData</a>();
<a name="l01623"></a>01623 
<a name="l01624"></a>01624   <span class="keywordflow">if</span> (m_callsFileReader==NULL || !m_callsFileReader-&gt;nextExperiment()) {
<a name="l01625"></a>01625     <span class="keywordflow">return</span> 0;
<a name="l01626"></a>01626   }
<a name="l01627"></a>01627 
<a name="l01628"></a>01628   m_experimentName = m_callsFileReader-&gt;getExperimentName();
<a name="l01629"></a>01629 
<a name="l01630"></a>01630   <span class="keywordflow">if</span> ( m_genoCallCoder == NULL ) {
<a name="l01631"></a>01631     <span class="comment">//</span>
<a name="l01632"></a>01632     <span class="comment">// TODO: We don&#39;t have access to all the meta-data from the chip files yet. Should put these in the header of the call file.</span>
<a name="l01633"></a>01633     <span class="comment">// In the mean time I&#39;m just guessing.</span>
<a name="l01634"></a>01634     <span class="comment">// </span>
<a name="l01635"></a>01635     m_genoCallCoder = <span class="keyword">new</span> <a class="code" href="classGenoCallCoder.html" title="GenoCallCoder - Class for encoding and decoding numeric genotype call codes to and from allele string...">GenoCallCoder</a>(m_callsFileReader-&gt;getMaxAlleles(), <span class="stringliteral">&quot;UCHAR&quot;</span>, <span class="stringliteral">&quot;1.0&quot;</span>, <span class="charliteral">&#39;/&#39;</span>, rte.m_adtOpts.m_inputAnnotationFile);
<a name="l01636"></a>01636 
<a name="l01637"></a>01637     <span class="keywordflow">if</span> ((ttm.getType() == ADT_TRANSLATION_TABLE_TYPE_DMET3) &amp;&amp;  !rte.m_adtOpts.m_ignoreReportAllele) {
<a name="l01638"></a>01638       ttm.<a class="code" href="classTranslationTableModel.html#a101b8a8c3530648d99137c542b500a46" title="TranslationTableModel::setReportAlleles Synopsis:">setReportAlleles</a>( *m_genoCallCoder);
<a name="l01639"></a>01639     }
<a name="l01640"></a>01640 
<a name="l01641"></a>01641     m_chpVersion = <span class="stringliteral">&quot;1.0&quot;</span>;
<a name="l01642"></a>01642     m_chpDataEncodingSize = <span class="stringliteral">&quot;UCHAR&quot;</span>;
<a name="l01643"></a>01643     m_chpMaxAlleles = 2;
<a name="l01644"></a>01644   }
<a name="l01645"></a>01645 
<a name="l01646"></a>01646   <span class="keywordflow">while</span> (m_callsFileReader-&gt;nextRow()) {
<a name="l01647"></a>01647   
<a name="l01648"></a>01648       std::string sProbeset = m_callsFileReader-&gt;getProbeset();
<a name="l01649"></a>01649       <span class="comment">// Not all probe sets are going to exist in the translation table,</span>
<a name="l01650"></a>01650       <span class="comment">// if not then just drop the probeSet data on the floor.</span>
<a name="l01651"></a>01651       <span class="keywordflow">if</span> (ttm.<a class="code" href="classTranslationTableModel.html#a078d4379d11cfc09e6d074c2ac0155c1" title="TranslationTableModel::getProbeSetRowIndex Synopsis:">getProbeSetRowIndex</a>(sProbeset) &lt; 0) {
<a name="l01652"></a>01652         <span class="keywordflow">continue</span>;
<a name="l01653"></a>01653       }
<a name="l01654"></a>01654 
<a name="l01655"></a>01655       std::string allele1, allele2, originalAllele1, originalAllele2;
<a name="l01656"></a>01656 
<a name="l01657"></a>01657       originalAllele1.clear();
<a name="l01658"></a>01658       originalAllele2.clear();
<a name="l01659"></a>01659 
<a name="l01660"></a>01660       <span class="keywordflow">if</span> (! _getAllelesFromAbstractAlleles(rte, ttm, sProbeset, m_callsFileReader-&gt;getCall(), allele1, allele2)) {
<a name="l01661"></a>01661         <span class="keywordflow">continue</span>;
<a name="l01662"></a>01662       }
<a name="l01663"></a>01663 
<a name="l01664"></a>01664       <span class="comment">// Genotype Override</span>
<a name="l01665"></a>01665       <span class="keywordflow">if</span> (m_gotm != NULL) {
<a name="l01666"></a>01666         std::vector&lt; std::string &gt; alleleOverride = m_gotm-&gt;<a class="code" href="classGenotypeOverrideTableModel.html#aeddcc3a332fd7ed3462c69bed74b17b5" title="GenotypeOverrideTableModel::getOverrideAlleles: Synopsis:">getOverrideAlleles</a>(m_experimentName, sProbeset);
<a name="l01667"></a>01667         <span class="keywordflow">if</span> (alleleOverride.size() &gt; 0) {
<a name="l01668"></a>01668           <span class="keywordflow">if</span> ( <a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(allele1)  ) {
<a name="l01669"></a>01669             m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[m_experimentName + <span class="stringliteral">&quot;::&quot;</span> + sProbeset] = allele1;
<a name="l01670"></a>01670           } <span class="keywordflow">else</span> {
<a name="l01671"></a>01671             m_gotm-&gt;m_chpFileProbeSetOriginalBasecall[m_experimentName + <span class="stringliteral">&quot;::&quot;</span> + sProbeset] = allele1 + <span class="stringliteral">&quot;/&quot;</span> + allele2;
<a name="l01672"></a>01672           }
<a name="l01673"></a>01673           
<a name="l01674"></a>01674           originalAllele1 = allele1;
<a name="l01675"></a>01675           originalAllele2 = allele2;
<a name="l01676"></a>01676           allele1 = alleleOverride[0];
<a name="l01677"></a>01677           <span class="keywordflow">if</span> ( alleleOverride.size() == 2 ) {
<a name="l01678"></a>01678             allele2 = alleleOverride[1];
<a name="l01679"></a>01679           }
<a name="l01680"></a>01680           <span class="keywordflow">else</span> {
<a name="l01681"></a>01681             allele2 = <span class="stringliteral">&quot;&quot;</span>;
<a name="l01682"></a>01682           }
<a name="l01683"></a>01683         }
<a name="l01684"></a>01684       }
<a name="l01685"></a>01685 
<a name="l01686"></a>01686       std::vector&lt;std::string&gt; newRow;
<a name="l01687"></a>01687       chpRow++;
<a name="l01688"></a>01688 
<a name="l01689"></a>01689       newRow.resize(GT_END_FIELD_INDEX, <span class="stringliteral">&quot;&quot;</span>);
<a name="l01690"></a>01690 
<a name="l01691"></a>01691       newRow[GT_SAMPLE_INDEX]      = <span class="stringliteral">&quot;N/A&quot;</span>;
<a name="l01692"></a>01692       newRow[GT_EXPERIMENT_INDEX]  = m_experimentName;
<a name="l01693"></a>01693       newRow[GT_GENES_INDEX]       = _getGeneFromProbeSet(sProbeset, ttm);
<a name="l01694"></a>01694       newRow[GT_EXTERNAL_ID_INDEX] = _getExternalIdFromProbeSet(sProbeset, ttm);
<a name="l01695"></a>01695       newRow[GT_PROBE_SET_INDEX]   = sProbeset;
<a name="l01696"></a>01696       newRow[GT_ALLELE1_INDEX]     = allele1;
<a name="l01697"></a>01697       newRow[GT_ALLELE2_INDEX]     = allele2;
<a name="l01698"></a>01698       newRow[GT_ORIGINAL_ALLELE1_INDEX] = originalAllele1;
<a name="l01699"></a>01699 
<a name="l01700"></a>01700       newRow[GT_ORIGINAL_ALLELE2_INDEX] = originalAllele2;
<a name="l01701"></a>01701 
<a name="l01702"></a>01702       m_fileRowIndex[m_rows.size()] = recordsRead;
<a name="l01703"></a>01703 
<a name="l01704"></a>01704       recordsRead++;
<a name="l01705"></a>01705       m_experimentRowIndex[m_experimentName].push_back(chpRow - 1);
<a name="l01706"></a>01706 
<a name="l01707"></a>01707       m_probeSetRowIndex[sProbeset] = m_rows.size();
<a name="l01708"></a>01708       m_rows.push_back(newRow);
<a name="l01709"></a>01709 
<a name="l01710"></a>01710       <span class="keywordflow">if</span> ( geneCopyNumber.count(newRow[GT_GENES_INDEX]) == 0 ) {
<a name="l01711"></a>01711         geneCopyNumber[newRow[GT_GENES_INDEX]] = 2;
<a name="l01712"></a>01712       }
<a name="l01713"></a>01713       
<a name="l01714"></a>01714       <span class="keywordflow">if</span> ( allele1  == <span class="stringliteral">&quot;0&quot;</span> ) {
<a name="l01715"></a>01715         geneCopyNumber[newRow[GT_GENES_INDEX]]  = 0;
<a name="l01716"></a>01716       }
<a name="l01717"></a>01717       <span class="keywordflow">else</span> <span class="keywordflow">if</span> ( !<a class="code" href="classCallElement.html#aae6a8be2dfee570fe4ddd451cb8a74c3" title="CallElement::isWildCardBase Synopsis:">CallElement::isWildCardBase</a>(allele1) &amp;&amp;
<a name="l01718"></a>01718                 (geneCopyNumber[newRow[GT_GENES_INDEX]] == 2)
<a name="l01719"></a>01719                 &amp;&amp; ( allele2.empty() ) ){
<a name="l01720"></a>01720         geneCopyNumber[newRow[GT_GENES_INDEX]] = 1;
<a name="l01721"></a>01721       }
<a name="l01722"></a>01722 
<a name="l01723"></a>01723 
<a name="l01724"></a>01724   } <span class="comment">// while (m_callsFileReader-&gt;nextRow())</span>
<a name="l01725"></a>01725 
<a name="l01726"></a>01726 
<a name="l01727"></a>01727   <span class="keywordflow">return</span> recordsRead;
<a name="l01728"></a>01728 }
</pre></div></div>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Wed Mar 23 2016 12:59:04 for Affymetrix Power Tools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.1 </small></address>
</body>
</html>
